--- a/net/minecraft/core/particles/BlockParticleOption.java
+++ b/net/minecraft/core/particles/BlockParticleOption.java
@@ -51,4 +_,15 @@
    public BlockState m_123642_() {
       return this.f_123626_;
    }
+
+   //FORGE: Add a source pos property, so we can provide models with additional model data
+   private net.minecraft.core.BlockPos pos;
+   public BlockParticleOption setPos(net.minecraft.core.BlockPos pos) {
+      this.pos = pos;
+      return this;
+   }
+
+   public net.minecraft.core.BlockPos getPos() {
+      return pos;
+   }
 }
