--- a/net/minecraft/entity/projectile/AbstractArrowEntity.java
+++ b/net/minecraft/entity/projectile/AbstractArrowEntity.java
@@ -130,7 +_,7 @@
 
       BlockPos blockpos = this.blockPosition();
       BlockState blockstate = this.level.getBlockState(blockpos);
-      if (!blockstate.isAir() && !flag) {
+      if (!blockstate.isAir(this.level, blockpos) && !flag) {
          VoxelShape voxelshape = blockstate.getCollisionShape(this.level, blockpos);
          if (!voxelshape.isEmpty()) {
             Vector3d vector3d1 = this.position();
@@ -184,7 +_,7 @@
                }
             }
 
-            if (raytraceresult != null && !flag) {
+            if (raytraceresult != null && raytraceresult.getType() != RayTraceResult.Type.MISS && !flag && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
                this.onHit(raytraceresult);
                this.hasImpulse = true;
             }
