--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityFishHook.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityFishHook.java
@@ -269,8 +269,8 @@
     {
         ItemStack itemstack = this.field_146042_b.func_184614_ca();
         ItemStack itemstack1 = this.field_146042_b.func_184592_cb();
-        boolean flag = itemstack.func_77973_b() == Items.field_151112_aM;
-        boolean flag1 = itemstack1.func_77973_b() == Items.field_151112_aM;
+        boolean flag = itemstack.func_77973_b() instanceof net.minecraft.item.ItemFishingRod;
+        boolean flag1 = itemstack1.func_77973_b() instanceof net.minecraft.item.ItemFishingRod;
 
         if (!this.field_146042_b.field_70128_L && this.field_146042_b.func_70089_S() && (flag || flag1) && this.func_70068_e(this.field_146042_b) <= 1024.0D)
         {
@@ -506,6 +506,7 @@
         {
             int i = 0;
 
+            net.minecraftforge.event.entity.player.ItemFishedEvent event = null;
             if (this.field_146043_c != null)
             {
                 this.func_184527_k();
@@ -515,9 +516,17 @@
             else if (this.field_146045_ax > 0)
             {
                 LootContext.Builder lootcontext$builder = new LootContext.Builder((WorldServer)this.field_70170_p);
-                lootcontext$builder.func_186469_a((float)this.field_191518_aw + this.field_146042_b.func_184817_da());
+                lootcontext$builder.func_186469_a((float)this.field_191518_aw + this.field_146042_b.func_184817_da()).func_186470_a(this.field_146042_b).func_186472_a(this); // Forge: add player & looted entity to LootContext
+                List<ItemStack> result = this.field_70170_p.func_184146_ak().func_186521_a(LootTableList.field_186387_al).func_186462_a(this.field_70146_Z, lootcontext$builder.func_186471_a());
+                event = new net.minecraftforge.event.entity.player.ItemFishedEvent(result, this.field_146051_au ? 2 : 1, this);
+                net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+                if (event.isCanceled())
+                {
+                    this.func_70106_y();
+                    return event.getRodDamage();
+                }
 
-                for (ItemStack itemstack : this.field_70170_p.func_184146_ak().func_186521_a(LootTableList.field_186387_al).func_186462_a(this.field_70146_Z, lootcontext$builder.func_186471_a()))
+                for (ItemStack itemstack : result)
                 {
                     EntityItem entityitem = new EntityItem(this.field_70170_p, this.field_70165_t, this.field_70163_u, this.field_70161_v, itemstack);
                     double d0 = this.field_146042_b.field_70165_t - this.field_70165_t;
@@ -547,7 +556,7 @@
             }
 
             this.func_70106_y();
-            return i;
+            return event == null ? i : event.getRodDamage();
         }
         else
         {
