--- a/net/minecraft/entity/projectile/EntityLargeFireball.java
+++ b/net/minecraft/entity/projectile/EntityLargeFireball.java
@@ -33,7 +33,7 @@
             this.func_174815_a(this.field_70235_a, p_70227_1_.field_72308_g);
          }
 
-         boolean flag = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
+         boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this.field_70235_a);
          this.field_70170_p.func_72885_a((Entity)null, this.field_70165_t, this.field_70163_u, this.field_70161_v, (float)this.field_92057_e, flag, flag);
          this.func_70106_y();
       }
