--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityLlamaSpit.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityLlamaSpit.java
@@ -80,7 +80,7 @@
             raytraceresult = new RayTraceResult(entity);
         }
 
-        if (raytraceresult != null)
+        if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
         {
             this.func_190536_a(raytraceresult);
         }
