--- a/net/minecraft/entity/projectile/EntityLlamaSpit.java
+++ b/net/minecraft/entity/projectile/EntityLlamaSpit.java
@@ -69,7 +69,7 @@
          raytraceresult = new RayTraceResult(entity);
       }
 
-      if (raytraceresult != null) {
+      if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
          this.func_190536_a(raytraceresult);
       }
 
