--- a/net/minecraft/entity/projectile/EntityArrow.java
+++ b/net/minecraft/entity/projectile/EntityArrow.java
@@ -165,7 +165,7 @@
 
       BlockPos blockpos = new BlockPos(this.field_145791_d, this.field_145792_e, this.field_145789_f);
       IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos);
-      if (!iblockstate.func_196958_f() && !flag) {
+      if (!iblockstate.isAir(this.field_70170_p, blockpos) && !flag) {
          VoxelShape voxelshape = iblockstate.func_196952_d(this.field_70170_p, blockpos);
          if (!voxelshape.func_197766_b()) {
             for(AxisAlignedBB axisalignedbb : voxelshape.func_197756_d()) {
@@ -223,7 +223,7 @@
             }
          }
 
-         if (raytraceresult != null && !flag) {
+         if (raytraceresult != null && !flag && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
             this.func_184549_a(raytraceresult);
             this.field_70160_al = true;
          }
@@ -314,7 +314,7 @@
          this.field_70254_i = true;
          this.field_70249_b = 7;
          this.func_70243_d(false);
-         if (!iblockstate.func_196958_f()) {
+         if (!iblockstate.isAir(field_70170_p, blockpos)) {
             this.field_195056_av.func_196950_a(this.field_70170_p, blockpos, this);
          }
       }
