--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityFishHook.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityFishHook.java
@@ -551,6 +551,12 @@
         float f = this.field_70170_p.field_73012_v.nextFloat();
         int i = EnchantmentHelper.func_151386_g(this.field_146042_b);
         int j = EnchantmentHelper.func_151387_h(this.field_146042_b);
+        if (true)
+        {
+            this.field_146042_b.func_71064_a(net.minecraftforge.common.FishingHooks.getFishableCategory(f, i, j).stat, 1);
+            return net.minecraftforge.common.FishingHooks.getRandomFishable(this.field_70146_Z, f, i, j);
+        }
+
         float f1 = 0.1F - (float)i * 0.025F - (float)j * 0.01F;
         float f2 = 0.05F + (float)i * 0.01F - (float)j * 0.01F;
         f1 = MathHelper.func_76131_a(f1, 0.0F, 1.0F);
