--- a/net/minecraft/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/entity/projectile/FireworkRocketEntity.java
@@ -155,6 +155,13 @@
 
    }
 
+   @Override
+   protected void func_70227_a(RayTraceResult result) {
+      if (result.func_216346_c() == RayTraceResult.Type.MISS || !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, result)) {
+         super.func_70227_a(result);
+      }
+   }
+
    private void func_213893_k() {
       this.field_70170_p.func_72960_a(this, (byte)17);
       this.func_191510_k();
