--- a/net/minecraft/entity/projectile/ThrowableEntity.java
+++ b/net/minecraft/entity/projectile/ThrowableEntity.java
@@ -63,7 +_,7 @@
          }
       }
 
-      if (raytraceresult.getType() != RayTraceResult.Type.MISS && !flag) {
+      if (raytraceresult.getType() != RayTraceResult.Type.MISS && !flag && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
          this.onHit(raytraceresult);
       }
 
