--- a/net/minecraft/world/server/Ticket.java
+++ b/net/minecraft/world/server/Ticket.java
@@ -9,9 +9,14 @@
    private long field_219483_d;
 
    protected Ticket(TicketType<T> p_i226095_1_, int p_i226095_2_, T p_i226095_3_) {
+      this(p_i226095_1_, p_i226095_2_, p_i226095_3_, false);
+   }
+
+   public Ticket(TicketType<T> p_i226095_1_, int p_i226095_2_, T p_i226095_3_, boolean forceTicks) {
       this.field_219480_a = p_i226095_1_;
       this.field_219481_b = p_i226095_2_;
       this.field_219482_c = p_i226095_3_;
+      this.forceTicks = forceTicks;
    }
 
    public int compareTo(Ticket<?> p_compareTo_1_) {
@@ -31,16 +36,16 @@
          return false;
       } else {
          Ticket<?> ticket = (Ticket)p_equals_1_;
-         return this.field_219481_b == ticket.field_219481_b && Objects.equals(this.field_219480_a, ticket.field_219480_a) && Objects.equals(this.field_219482_c, ticket.field_219482_c);
+         return this.field_219481_b == ticket.field_219481_b && Objects.equals(this.field_219480_a, ticket.field_219480_a) && Objects.equals(this.field_219482_c, ticket.field_219482_c) && this.forceTicks == ticket.forceTicks;
       }
    }
 
    public int hashCode() {
-      return Objects.hash(this.field_219480_a, this.field_219481_b, this.field_219482_c);
+      return Objects.hash(this.field_219480_a, this.field_219481_b, this.field_219482_c, forceTicks);
    }
 
    public String toString() {
-      return "Ticket[" + this.field_219480_a + " " + this.field_219481_b + " (" + this.field_219482_c + ")] at " + this.field_219483_d;
+      return "Ticket[" + this.field_219480_a + " " + this.field_219481_b + " (" + this.field_219482_c + ")] at " + this.field_219483_d + " force ticks " + forceTicks;
    }
 
    public TicketType<T> func_219479_a() {
@@ -59,4 +64,12 @@
       long i = this.field_219480_a.func_223184_b();
       return i != 0L && p_223182_1_ - this.field_219483_d > i;
    }
+
+   // Forge Start
+   private final boolean forceTicks;
+
+   public boolean isForceTicks() {
+      return forceTicks;
+   }
+   // Forge End
 }
