--- a/net/minecraft/entity/player/EntityPlayerMP.java
+++ b/net/minecraft/entity/player/EntityPlayerMP.java
@@ -420,6 +420,7 @@
    }
 
    public void func_70645_a(DamageSource p_70645_1_) {
+      if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this,  p_70645_1_)) return;
       boolean flag = this.field_70170_p.func_82736_K().func_82766_b("showDeathMessages");
       if (flag) {
          ITextComponent itextcomponent = this.func_110142_aN().func_151521_b();
@@ -451,8 +452,10 @@
 
       this.func_192030_dh();
       if (!this.field_70170_p.func_82736_K().func_82766_b("keepInventory") && !this.func_175149_v()) {
+         this.captureDrops(new java.util.ArrayList<>());
          this.func_190776_cN();
          this.field_71071_by.func_70436_m();
+         if (!field_70170_p.field_72995_K) net.minecraftforge.event.ForgeEventFactory.onPlayerDrops(this, p_70645_1_, captureDrops(null), field_70718_bc > 0);
       }
 
       this.func_96123_co().func_197893_a(ScoreCriteria.field_96642_c, this.func_195047_I_(), Score::func_96648_a);
@@ -537,8 +540,10 @@
       return this.field_71133_b.func_71219_W();
    }
 
+   @Override
    @Nullable
-   public Entity func_212321_a(DimensionType p_212321_1_) {
+   public Entity changeDimension(DimensionType p_212321_1_, net.minecraftforge.common.util.ITeleporter teleporter) {
+      if (!net.minecraftforge.common.ForgeHooks.onTravelToDimension(this, p_212321_1_)) return null;
       this.field_184851_cj = true;
       if (this.field_71093_bK == DimensionType.OVERWORLD && p_212321_1_ == DimensionType.NETHER) {
          this.field_193110_cw = new Vec3d(this.field_70165_t, this.field_70163_u, this.field_70161_v);
@@ -546,7 +551,7 @@
          this.field_193110_cw = null;
       }
 
-      if (this.field_71093_bK == DimensionType.THE_END && p_212321_1_ == DimensionType.THE_END) {
+      if (this.field_71093_bK == DimensionType.THE_END && p_212321_1_ == DimensionType.THE_END && teleporter.isVanilla()) {
          this.field_70170_p.func_72900_e(this);
          if (!this.field_71136_j) {
             this.field_71136_j = true;
@@ -560,7 +565,7 @@
             p_212321_1_ = DimensionType.THE_END;
          }
 
-         this.field_71133_b.func_184103_al().func_187242_a(this, p_212321_1_);
+         this.field_71133_b.func_184103_al().changePlayerDimension(this, p_212321_1_, teleporter);
          this.field_71135_a.func_147359_a(new SPacketEffect(1032, BlockPos.field_177992_a, 0, false));
          this.field_71144_ck = -1;
          this.field_71149_ch = -1.0F;
@@ -662,7 +667,7 @@
       int k = MathHelper.func_76128_c(this.field_70161_v);
       BlockPos blockpos = new BlockPos(i, j, k);
       IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos);
-      if (iblockstate.func_196958_f()) {
+      if (iblockstate.isAir(field_70170_p, blockpos)) {
          BlockPos blockpos1 = blockpos.func_177977_b();
          IBlockState iblockstate1 = this.field_70170_p.func_180495_p(blockpos1);
          Block block = iblockstate1.func_177230_c();
@@ -693,6 +698,7 @@
          this.field_71070_bA = p_180468_1_.func_174876_a(this.field_71071_by, this);
          this.field_71070_bA.field_75152_c = this.field_71139_cq;
          this.field_71070_bA.func_75132_a(this);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(this, this.field_71070_bA));
       }
    }
 
@@ -724,6 +730,7 @@
 
          this.field_71070_bA.field_75152_c = this.field_71139_cq;
          this.field_71070_bA.func_75132_a(this);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(this, this.field_71070_bA));
       }
    }
 
@@ -732,6 +739,7 @@
       this.field_71070_bA = new ContainerMerchant(this.field_71071_by, p_180472_1_, this.field_70170_p);
       this.field_71070_bA.field_75152_c = this.field_71139_cq;
       this.field_71070_bA.func_75132_a(this);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(this, this.field_71070_bA));
       IInventory iinventory = ((ContainerMerchant)this.field_71070_bA).func_75174_d();
       ITextComponent itextcomponent = p_180472_1_.func_145748_c_();
       this.field_71135_a.func_147359_a(new SPacketOpenWindow(this.field_71139_cq, "minecraft:villager", itextcomponent, iinventory.func_70302_i_()));
@@ -755,6 +763,7 @@
       this.field_71070_bA = new ContainerHorseInventory(this.field_71071_by, p_184826_2_, p_184826_1_, this);
       this.field_71070_bA.field_75152_c = this.field_71139_cq;
       this.field_71070_bA.func_75132_a(this);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(this, this.field_71070_bA));
    }
 
    public void func_184814_a(ItemStack p_184814_1_, EnumHand p_184814_2_) {
@@ -817,6 +826,7 @@
 
    public void func_71128_l() {
       this.field_71070_bA.func_75134_a(this);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Close(this, this.field_71070_bA));
       this.field_71070_bA = this.field_71069_bz;
    }
 
@@ -946,6 +956,20 @@
       this.field_193110_cw = p_193104_1_.field_193110_cw;
       this.func_192029_h(p_193104_1_.func_192023_dk());
       this.func_192031_i(p_193104_1_.func_192025_dl());
+
+      this.spawnPosMap = p_193104_1_.spawnPosMap;
+      this.spawnForcedMap = p_193104_1_.spawnForcedMap;
+      if(p_193104_1_.field_71093_bK != net.minecraft.world.dimension.DimensionType.OVERWORLD) {
+          this.field_71077_c = p_193104_1_.field_71077_c;
+          this.field_82248_d = p_193104_1_.field_82248_d;
+      }
+
+      //Copy over a section of the Entity Data from the old player.
+      //Allows mods to specify data that persists after players respawn.
+      NBTTagCompound old = p_193104_1_.getEntityData();
+      if (old.func_74764_b(PERSISTED_NBT_TAG))
+          getEntityData().func_74782_a(PERSISTED_NBT_TAG, old.func_74781_a(PERSISTED_NBT_TAG));
+      net.minecraftforge.event.ForgeEventFactory.onPlayerClone(this, p_193104_1_, !p_193104_2_);
    }
 
    protected void func_70670_a(PotionEffect p_70670_1_) {
@@ -1174,13 +1198,13 @@
       this.func_184210_p();
       if (p_200619_1_ == this.field_70170_p) {
          this.field_71135_a.func_147364_a(p_200619_2_, p_200619_4_, p_200619_6_, p_200619_8_, p_200619_9_);
-      } else {
+      } else if (net.minecraftforge.common.ForgeHooks.onTravelToDimension(this, p_200619_1_.field_73011_w.func_186058_p())) {
          WorldServer worldserver = this.func_71121_q();
          this.field_71093_bK = p_200619_1_.field_73011_w.func_186058_p();
-         this.field_71135_a.func_147359_a(new SPacketRespawn(this.field_71093_bK, worldserver.func_175659_aa(), worldserver.func_72912_H().func_76067_t(), this.field_71134_c.func_73081_b()));
+         this.field_71135_a.func_147359_a(new SPacketRespawn(this.field_71093_bK, p_200619_1_.func_175659_aa(), p_200619_1_.func_72912_H().func_76067_t(), this.field_71134_c.func_73081_b())); // Forge: Use new dimensions information
          this.field_71133_b.func_184103_al().func_187243_f(this);
-         worldserver.func_72973_f(this);
-         this.field_70128_L = false;
+         worldserver.removeEntityDangerously(this, true); //Forge: The player entity itself is moved, and not cloned. So we need to keep the data alive with no matching invalidate call later.
+         this.revive();
          this.func_70012_b(p_200619_2_, p_200619_4_, p_200619_6_, p_200619_8_, p_200619_9_);
          if (this.func_70089_S()) {
             worldserver.func_72866_a(this, false);
@@ -1194,6 +1218,7 @@
          this.field_71134_c.func_73080_a(p_200619_1_);
          this.field_71133_b.func_184103_al().func_72354_b(this, p_200619_1_);
          this.field_71133_b.func_184103_al().func_72385_f(this);
+         net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerChangedDimensionEvent(this, p_200619_1_.field_73011_w.func_186058_p(), this.field_71093_bK);
       }
 
    }
