--- ../src-base/minecraft/net/minecraft/entity/player/InventoryPlayer.java
+++ ../src-work/minecraft/net/minecraft/entity/player/InventoryPlayer.java
@@ -272,12 +272,8 @@
         {
             if (this.field_70462_a[j] == null)
             {
-                this.field_70462_a[j] = new ItemStack(item, 0, p_70452_1_.func_77960_j());
-
-                if (p_70452_1_.func_77942_o())
-                {
-                    this.field_70462_a[j].func_77982_d((NBTTagCompound)p_70452_1_.func_77978_p().func_74737_b());
-                }
+                this.field_70462_a[j] = p_70452_1_.func_77946_l(); // Forge: Replace Item clone above to preserve item capabilities when picking the item up.
+                this.field_70462_a[j].field_77994_a = 0;
             }
 
             int k = i;
@@ -315,6 +311,14 @@
                 this.field_70462_a[i].func_77945_a(this.field_70458_d.field_70170_p, this.field_70458_d, i, this.field_70461_c == i);
             }
         }
+
+        for (int i = 0; i < field_70460_b.length; i++)
+        {
+            if (field_70460_b[i] != null)
+            {
+                field_70460_b[i].func_77973_b().onArmorTick(field_70458_d.field_70170_p, field_70458_d, field_70460_b[i]);
+            }
+        }
     }
 
     public boolean func_146026_a(Item p_146026_1_)
