--- ../src-base/minecraft/net/minecraft/entity/player/InventoryPlayer.java
+++ ../src-work/minecraft/net/minecraft/entity/player/InventoryPlayer.java
@@ -292,7 +292,8 @@
 
         if (itemstack.func_190926_b())
         {
-            itemstack = new ItemStack(item, 0, p_191973_2_.func_77960_j());
+            itemstack = p_191973_2_.func_77946_l(); // Forge: Replace Item clone above to preserve item capabilities when picking the item up.
+            itemstack.func_190920_e(0);
 
             if (p_191973_2_.func_77942_o())
             {
@@ -363,6 +364,13 @@
                 }
             }
         }
+        for (ItemStack is : field_70460_b) // FORGE: Tick armor on animation ticks
+        {
+            if (!is.func_190926_b())
+            {
+                is.func_77973_b().onArmorTick(field_70458_d.field_70170_p, field_70458_d, is);
+            }
+        }
     }
 
     public boolean func_70441_a(ItemStack p_70441_1_)
@@ -444,6 +452,8 @@
                 CrashReportCategory crashreportcategory = crashreport.func_85058_a("Item being added");
                 crashreportcategory.func_71507_a("Item ID", Integer.valueOf(Item.func_150891_b(p_191971_2_.func_77973_b())));
                 crashreportcategory.func_71507_a("Item data", Integer.valueOf(p_191971_2_.func_77960_j()));
+                crashreportcategory.func_189529_a("Registry Name", () -> String.valueOf(p_191971_2_.func_77973_b().getRegistryName()));
+                crashreportcategory.func_189529_a("Item Class", () -> p_191971_2_.func_77973_b().getClass().getName());
                 crashreportcategory.func_189529_a("Item name", new ICrashReportDetail<String>()
                 {
                     public String call() throws Exception
