--- a/net/minecraft/potion/Effect.java
+++ b/net/minecraft/potion/Effect.java
@@ -20,7 +20,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public class Effect {
+public class Effect extends net.minecraftforge.registries.ForgeRegistryEntry<Effect> implements net.minecraftforge.common.extensions.IForgeEffect {
    private final Map<IAttribute, AttributeModifier> field_111188_I = Maps.newHashMap();
    private final EffectType field_220305_b;
    private final int field_76414_N;
@@ -134,7 +134,6 @@
       return new TranslationTextComponent(this.func_76393_a());
    }
 
-   @OnlyIn(Dist.CLIENT)
    public EffectType func_220303_e() {
       return this.field_220305_b;
    }
@@ -149,7 +148,6 @@
       return this;
    }
 
-   @OnlyIn(Dist.CLIENT)
    public Map<IAttribute, AttributeModifier> func_111186_k() {
       return this.field_111188_I;
    }
@@ -180,7 +178,6 @@
       return p_111183_2_.func_111164_d() * (double)(p_111183_1_ + 1);
    }
 
-   @OnlyIn(Dist.CLIENT)
    public boolean func_188408_i() {
       return this.field_220305_b == EffectType.BENEFICIAL;
    }
