--- ../src-base/minecraft/net/minecraft/potion/Potion.java
+++ ../src-work/minecraft/net/minecraft/potion/Potion.java
@@ -21,7 +21,7 @@
 
 public class Potion
 {
-    public static final Potion[] field_76425_a = new Potion[32];
+    public static final Potion[] field_76425_a = new Potion[256];
     private static final Map<ResourceLocation, Potion> field_180150_I = Maps.<ResourceLocation, Potion>newHashMap();
     public static final Potion field_180151_b = null;
     public static final Potion field_76424_c = (new Potion(1, new ResourceLocation("speed"), false, 8171462)).func_76390_b("potion.moveSpeed").func_76399_b(0, 0).func_111184_a(SharedMonsterAttributes.field_111263_d, "91AEAA56-376B-4498-935B-2F7F68070635", 0.20000000298023224D, 2);
@@ -64,8 +64,10 @@
     private double field_76412_L;
     private boolean field_76413_M;
 
+    @Deprecated // use the Constructor without potion ID or everything will explode.
     protected Potion(int p_i45897_1_, ResourceLocation p_i45897_2_, boolean p_i45897_3_, int p_i45897_4_)
     {
+        p_i45897_1_ = net.minecraftforge.fml.common.registry.GameRegistry.registerPotion(this, p_i45897_2_, p_i45897_1_);
         this.field_76415_H = p_i45897_1_;
         field_76425_a[p_i45897_1_] = this;
         field_180150_I.put(p_i45897_2_, this);
@@ -83,14 +85,18 @@
         this.field_76414_N = p_i45897_4_;
     }
 
+    protected Potion(ResourceLocation location, boolean badEffect, int potionColor) {
+        this(-1, location, badEffect, potionColor);
+    }
+
     public static Potion func_180142_b(String p_180142_0_)
     {
-        return (Potion)field_180150_I.get(new ResourceLocation(p_180142_0_));
+        return net.minecraftforge.fml.common.registry.GameData.getPotionRegistry().func_82594_a(new ResourceLocation(p_180142_0_));
     }
 
     public static Set<ResourceLocation> func_181168_c()
     {
-        return field_180150_I.keySet();
+        return net.minecraftforge.fml.common.registry.GameData.getPotionRegistry().func_148742_b();
     }
 
     protected Potion func_76399_b(int p_76399_1_, int p_76399_2_)
@@ -310,4 +316,27 @@
     {
         return p_111183_2_.func_111164_d() * (double)(p_111183_1_ + 1);
     }
+
+    /* ======================================== FORGE START =====================================*/
+
+    /**
+     * If the standard PotionEffect text (name and duration) should be drawn when this potion is active.
+     * @param effect the active PotionEffect
+     * @return true to draw the standard text
+     */
+    public boolean shouldRenderInvText(PotionEffect effect)
+    {
+        return true;
+    }
+
+    /**
+     * Called to draw the this Potion onto the player's inventory when it's active.
+     * This can be used to e.g. render Potion icons from your own texture.
+     * @param x the x coordinate
+     * @param y the y coordinate
+     * @param effect the active PotionEffect
+     * @param mc the Minecraft instance, for convenience
+     */
+    @SideOnly(Side.CLIENT)
+    public void renderInventoryEffect(int x, int y, PotionEffect effect, net.minecraft.client.Minecraft mc) { }
 }
