--- ../src-base/minecraft/net/minecraft/potion/PotionEffect.java
+++ ../src-work/minecraft/net/minecraft/potion/PotionEffect.java
@@ -19,6 +19,8 @@
     @SideOnly(Side.CLIENT)
     private boolean field_100013_f;
     private boolean field_188421_h;
+    /** List of ItemStack that can cure the potion effect **/
+    private java.util.List<net.minecraft.item.ItemStack> curativeItems;
 
     public PotionEffect(Potion p_i46811_1_)
     {
@@ -51,6 +53,7 @@
         this.field_76461_c = p_i1577_1_.field_76461_c;
         this.field_82724_e = p_i1577_1_.field_82724_e;
         this.field_188421_h = p_i1577_1_.field_188421_h;
+        this.curativeItems = p_i1577_1_.curativeItems;
     }
 
     public void func_76452_a(PotionEffect p_76452_1_)
@@ -200,7 +203,7 @@
 
     public static PotionEffect func_82722_b(NBTTagCompound p_82722_0_)
     {
-        int i = p_82722_0_.func_74771_c("Id");
+        int i = p_82722_0_.func_74771_c("Id") & 0xFF;
         Potion potion = Potion.func_188412_a(i);
 
         if (potion == null)
@@ -240,4 +243,58 @@
     {
         return this.field_100013_f;
     }
+
+    /* ======================================== FORGE START =====================================*/
+    /***
+     * Returns a list of curative items for the potion effect
+     * @return The list (ItemStack) of curative items for the potion effect
+     */
+    public java.util.List<net.minecraft.item.ItemStack> getCurativeItems()
+    {
+        if (this.curativeItems == null) //Lazy load this so that we don't create a circular dep on Items.
+        {
+            this.curativeItems = new java.util.ArrayList<net.minecraft.item.ItemStack>();
+            this.curativeItems.add(new net.minecraft.item.ItemStack(net.minecraft.init.Items.field_151117_aB));
+        }
+        return this.curativeItems;
+    }
+
+    /***
+     * Checks the given ItemStack to see if it is in the list of curative items for the potion effect
+     * @param stack The ItemStack being checked against the list of curative items for the potion effect
+     * @return {@code true} if the given ItemStack is in the list of curative items for the potion effect, {@code false} otherwise
+     */
+    public boolean isCurativeItem(net.minecraft.item.ItemStack stack)
+    {
+        for (net.minecraft.item.ItemStack curativeItem : this.getCurativeItems())
+        {
+            if (curativeItem.func_77969_a(stack))
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /***
+     * Sets the array of curative items for the potion effect
+     * @param curativeItems The list of ItemStacks being set to the potion effect
+     */
+    public void setCurativeItems(java.util.List<net.minecraft.item.ItemStack> curativeItems)
+    {
+        this.curativeItems = curativeItems;
+    }
+
+    /***
+     * Adds the given stack to list of curative items for the potion effect
+     * @param stack The ItemStack being added to the curative item list
+     */
+    public void addCurativeItem(net.minecraft.item.ItemStack stack)
+    {
+        if (!this.isCurativeItem(stack))
+        {
+            this.getCurativeItems().add(stack);
+        }
+    }
 }
