--- a/net/minecraft/potion/PotionBrewing.java
+++ b/net/minecraft/potion/PotionBrewing.java
@@ -68,7 +68,7 @@
 
       for(int j = field_185214_b.size(); i < j; ++i) {
          PotionBrewing.MixPredicate<Item> mixpredicate = field_185214_b.get(i);
-         if (mixpredicate.field_185198_a == item && mixpredicate.field_185199_b.test(p_185206_1_)) {
+         if (mixpredicate.field_185198_a.get() == item && mixpredicate.field_185199_b.test(p_185206_1_)) {
             return true;
          }
       }
@@ -82,7 +82,7 @@
 
       for(int j = field_185213_a.size(); i < j; ++i) {
          PotionBrewing.MixPredicate<PotionType> mixpredicate = field_185213_a.get(i);
-         if (mixpredicate.field_185198_a == potiontype && mixpredicate.field_185199_b.test(p_185209_1_)) {
+         if (mixpredicate.field_185198_a.get() == potiontype && mixpredicate.field_185199_b.test(p_185209_1_)) {
             return true;
          }
       }
@@ -98,8 +98,8 @@
 
          for(int j = field_185214_b.size(); i < j; ++i) {
             PotionBrewing.MixPredicate<Item> mixpredicate = field_185214_b.get(i);
-            if (mixpredicate.field_185198_a == item && mixpredicate.field_185199_b.test(p_185212_0_)) {
-               return PotionUtils.func_185188_a(new ItemStack((IItemProvider)mixpredicate.field_185200_c), potiontype);
+            if (mixpredicate.field_185198_a.get() == item && mixpredicate.field_185199_b.test(p_185212_0_)) {
+               return PotionUtils.func_185188_a(new ItemStack((IItemProvider)mixpredicate.field_185200_c.get()), potiontype);
             }
          }
 
@@ -107,8 +107,8 @@
 
          for(int k = field_185213_a.size(); i < k; ++i) {
             PotionBrewing.MixPredicate<PotionType> mixpredicate1 = field_185213_a.get(i);
-            if (mixpredicate1.field_185198_a == potiontype && mixpredicate1.field_185199_b.test(p_185212_0_)) {
-               return PotionUtils.func_185188_a(new ItemStack(item), (PotionType)mixpredicate1.field_185200_c);
+            if (mixpredicate1.field_185198_a.get() == potiontype && mixpredicate1.field_185199_b.test(p_185212_0_)) {
+               return PotionUtils.func_185188_a(new ItemStack(item), (PotionType)mixpredicate1.field_185200_c.get());
             }
          }
       }
@@ -201,15 +201,15 @@
       field_185213_a.add(new PotionBrewing.MixPredicate<>(p_193357_0_, Ingredient.func_199804_a(p_193357_1_), p_193357_2_));
    }
 
-   static class MixPredicate<T> {
-      private final T field_185198_a;
+   static class MixPredicate<T extends net.minecraftforge.registries.ForgeRegistryEntry<T>> {
+      private final net.minecraftforge.registries.IRegistryDelegate<T> field_185198_a;
       private final Ingredient field_185199_b;
-      private final T field_185200_c;
+      private final net.minecraftforge.registries.IRegistryDelegate<T> field_185200_c;
 
-      public MixPredicate(T p_i47570_1_, Ingredient p_i47570_2_, T p_i47570_3_) {
-         this.field_185198_a = p_i47570_1_;
-         this.field_185199_b = p_i47570_2_;
-         this.field_185200_c = p_i47570_3_;
+      public MixPredicate(T inputIn, Ingredient reagentIn, T outputIn) {
+         this.field_185198_a = inputIn.delegate;
+         this.field_185199_b = reagentIn;
+         this.field_185200_c = outputIn.delegate;
       }
    }
 }
