--- ../src_base/minecraft/net/minecraft/potion/Potion.java
+++ ../src_work/minecraft/net/minecraft/potion/Potion.java
@@ -104,7 +104,7 @@
     /**
      * Sets the index for the icon displayed in the player's inventory when the status is active.
      */
-    protected Potion setIconIndex(int par1, int par2)
+    public Potion setIconIndex(int par1, int par2)
     {
         this.statusIconIndex = par1 + par2 * 8;
         return this;
