--- a/net/minecraft/advancements/critereon/ItemPredicate.java
+++ b/net/minecraft/advancements/critereon/ItemPredicate.java
@@ -50,7 +_,7 @@
          return false;
       } else {
          if (!this.enchantments.isEmpty()) {
-            Map<Enchantment, Integer> map = EnchantmentHelper.deserializeEnchantments(p_45050_.getEnchantmentTags());
+            Map<Enchantment, Integer> map = p_45050_.getAllEnchantments();
 
             for(EnchantmentPredicate enchantmentpredicate : this.enchantments) {
                if (!enchantmentpredicate.containedIn(map)) {
