--- a/net/minecraft/data/tags/BannerPatternTagsProvider.java
+++ b/net/minecraft/data/tags/BannerPatternTagsProvider.java
@@ -9,8 +_,14 @@
 import net.minecraft.world.level.block.entity.BannerPatterns;
 
 public class BannerPatternTagsProvider extends TagsProvider<BannerPattern> {
+    /** @deprecated Forge: Use the {@linkplain #BannerPatternTagsProvider(PackOutput, CompletableFuture, String, net.minecraftforge.common.data.ExistingFileHelper) mod id variant} */
+    @Deprecated
     public BannerPatternTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.BANNER_PATTERN, lookupProvider);
+    }
+
+    public BannerPatternTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.BANNER_PATTERN, lookupProvider, modId, existingFileHelper);
     }
 
     @Override
