--- a/net/minecraft/data/tags/BiomeTagsProvider.java
+++ b/net/minecraft/data/tags/BiomeTagsProvider.java
@@ -15,6 +_,10 @@
         super(p_255800_, Registries.BIOME, p_256205_);
     }
 
+    public BiomeTagsProvider(PackOutput p_255800_, CompletableFuture<HolderLookup.Provider> p_256205_, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(p_255800_, Registries.BIOME, p_256205_, modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(HolderLookup.Provider p_256485_) {
         this.tag(BiomeTags.IS_DEEP_OCEAN).add(Biomes.DEEP_FROZEN_OCEAN).add(Biomes.DEEP_COLD_OCEAN).add(Biomes.DEEP_OCEAN).add(Biomes.DEEP_LUKEWARM_OCEAN);
