--- a/net/minecraft/data/tags/BlockItemTagAppender.java
+++ b/net/minecraft/data/tags/BlockItemTagAppender.java
@@ -56,4 +_,28 @@
         this.original.addOptionalTag(tag);
         return this;
     }
+
+    @Override
+    public TagAppender<Element> addOptional(net.minecraft.resources.Identifier location) {
+        this.original.addOptional(location);
+        return this;
+    }
+
+    @Override
+    public TagAppender<Element> replace(boolean value) {
+        this.original.replace(value);
+        return this;
+    }
+
+    @Override
+    public TagAppender<Element> remove(final net.minecraft.resources.Identifier location) {
+        this.original.remove(location);
+        return this;
+    }
+
+    @Override
+    public TagAppender<Element> remove(TagKey<Element> tag) {
+        this.original.remove(tag);
+        return this;
+    }
 }
