--- a/net/minecraft/data/tags/CatVariantTagsProvider.java
+++ b/net/minecraft/data/tags/CatVariantTagsProvider.java
@@ -12,6 +_,10 @@
         super(p_256547_, Registries.CAT_VARIANT, p_256090_);
     }
 
+    public CatVariantTagsProvider(PackOutput p_256547_, CompletableFuture<HolderLookup.Provider> p_256090_, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(p_256547_, Registries.CAT_VARIANT, p_256090_, modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(HolderLookup.Provider p_256078_) {
         this.tag(CatVariantTags.DEFAULT_SPAWNS)
