--- a/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
+++ b/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
@@ -13,6 +_,10 @@
         super(p_256604_, Registries.FLAT_LEVEL_GENERATOR_PRESET, p_255962_);
     }
 
+    public FlatLevelGeneratorPresetTagsProvider(PackOutput p_256604_, CompletableFuture<HolderLookup.Provider> p_255962_, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(p_256604_, Registries.FLAT_LEVEL_GENERATOR_PRESET, p_255962_, modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(HolderLookup.Provider p_255741_) {
         this.tag(FlatLevelGeneratorPresetTags.VISIBLE)
