--- a/net/minecraft/data/tags/FluidTagsProvider.java
+++ b/net/minecraft/data/tags/FluidTagsProvider.java
@@ -13,6 +_,10 @@
         super(p_255941_, Registries.FLUID, p_256600_, p_256474_ -> p_256474_.builtInRegistryHolder().key());
     }
 
+    public FluidTagsProvider(PackOutput p_255941_, CompletableFuture<HolderLookup.Provider> p_256600_, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(p_255941_, Registries.FLUID, p_256600_, p_256474_ -> p_256474_.builtInRegistryHolder().key(), modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(HolderLookup.Provider p_256366_) {
         this.tag(FluidTags.WATER).add(Fluids.WATER, Fluids.FLOWING_WATER);
