--- a/net/minecraft/data/tags/GameEventTagsProvider.java
+++ b/net/minecraft/data/tags/GameEventTagsProvider.java
@@ -6,8 +_,13 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 
 public class GameEventTagsProvider extends TagsProvider<GameEvent> {
+   /** @deprecated Forge: Use the {@link #GameEventTagsProvider(DataGenerator, String, net.minecraftforge.common.data.ExistingFileHelper) mod id variant} */
+   @Deprecated
    public GameEventTagsProvider(DataGenerator p_176826_) {
       super(p_176826_, Registry.f_175412_);
+   }
+   public GameEventTagsProvider(DataGenerator p_176826_, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+      super(p_176826_, Registry.f_175412_, modId, existingFileHelper);
    }
 
    protected void m_6577_() {
