--- a/net/minecraft/data/tags/InstrumentTagsProvider.java
+++ b/net/minecraft/data/tags/InstrumentTagsProvider.java
@@ -13,6 +_,10 @@
         super(p_256418_, Registries.INSTRUMENT, p_256038_);
     }
 
+    public InstrumentTagsProvider(PackOutput p_256418_, CompletableFuture<HolderLookup.Provider> p_256038_, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(p_256418_, Registries.INSTRUMENT, p_256038_, modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(HolderLookup.Provider p_256291_) {
         this.tag(InstrumentTags.REGULAR_GOAT_HORNS)
