--- a/net/minecraft/data/tags/PaintingVariantTagsProvider.java
+++ b/net/minecraft/data/tags/PaintingVariantTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.entity.decoration.painting.PaintingVariants;
 
 public class PaintingVariantTagsProvider extends TagsProvider<PaintingVariant> {
+    /** @deprecated Forge: Use the {@linkplain #PaintingVariantTagsProvider(PackOutput, CompletableFuture, String, net.minecraftforge.common.data.ExistingFileHelper) mod id variant} */
     public PaintingVariantTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.PAINTING_VARIANT, lookupProvider);
+    }
+
+    public PaintingVariantTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.PAINTING_VARIANT, lookupProvider, modId, existingFileHelper);
     }
 
     @Override
