--- a/net/minecraft/data/tags/PoiTypeTagsProvider.java
+++ b/net/minecraft/data/tags/PoiTypeTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.entity.ai.village.poi.PoiTypes;
 
 public class PoiTypeTagsProvider extends TagsProvider<PoiType> {
+    /** @deprecated Forge: Use the {@linkplain #PoiTypeTagsProvider(PackOutput, CompletableFuture, String, net.minecraftforge.common.data.ExistingFileHelper) mod id variant} */
     public PoiTypeTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.POINT_OF_INTEREST_TYPE, lookupProvider);
+    }
+
+    public PoiTypeTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.POINT_OF_INTEREST_TYPE, lookupProvider, modId, existingFileHelper);
     }
 
     @Override
