--- a/net/minecraft/data/tags/PoiTypeTagsProvider.java
+++ b/net/minecraft/data/tags/PoiTypeTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.entity.ai.village.poi.PoiTypes;
 
 public class PoiTypeTagsProvider extends TagsProvider<PoiType> {
+   /** @deprecated Forge: Use the {@linkplain #PoiTypeTagsProvider(PackOutput, CompletableFuture, String, net.minecraftforge.common.data.ExistingFileHelper) mod id variant} */
+   @Deprecated
    public PoiTypeTagsProvider(PackOutput p_256012_, CompletableFuture<HolderLookup.Provider> p_256617_) {
       super(p_256012_, Registries.f_256805_, p_256617_);
+   }
+   public PoiTypeTagsProvider(PackOutput p_256012_, CompletableFuture<HolderLookup.Provider> p_256617_, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+      super(p_256012_, Registries.f_256805_, p_256617_, modId, existingFileHelper);
    }
 
    protected void m_6577_(HolderLookup.Provider p_256206_) {
