--- a/net/minecraft/data/tags/TagsProvider.java
+++ b/net/minecraft/data/tags/TagsProvider.java
@@ -32,10 +_,24 @@
    protected final DataGenerator f_126539_;
    protected final Registry<T> f_126540_;
    protected final Map<ResourceLocation, Tag.Builder> f_126543_ = Maps.newLinkedHashMap();
+   protected final String modId;
+   protected final net.minecraftforge.common.data.ExistingFileHelper existingFileHelper;
+   private final net.minecraftforge.common.data.ExistingFileHelper.IResourceType resourceType;
 
+   /**
+    * @see #TagsProvider(DataGenerator, Registry, String, net.minecraftforge.common.data.ExistingFileHelper)
+    * @deprecated Forge: Use the mod id variant
+    */
+   @Deprecated
    protected TagsProvider(DataGenerator p_126546_, Registry<T> p_126547_) {
+      this(p_126546_, p_126547_, "vanilla", null);
+   }
+   protected TagsProvider(DataGenerator p_126546_, Registry<T> p_126547_, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
       this.f_126539_ = p_126546_;
       this.f_126540_ = p_126547_;
+      this.modId = modId;
+      this.existingFileHelper = existingFileHelper;
+      this.resourceType = new net.minecraftforge.common.data.ExistingFileHelper.ResourceType(net.minecraft.server.packs.PackType.SERVER_DATA, ".json", TagManager.m_203918_(p_126547_.m_123023_()));
    }
 
    protected abstract void m_6577_();
@@ -46,12 +_,13 @@
       this.f_126543_.forEach((p_176835_, p_176836_) -> {
          List<Tag.BuilderEntry> list = p_176836_.m_13330_().filter((p_176832_) -> {
             return !p_176832_.f_13338_().m_142746_(this.f_126540_::m_7804_, this.f_126543_::containsKey);
-         }).toList();
+         }).filter(this::missing).collect(Collectors.toList()); // Forge: Add validation via existing resources
          if (!list.isEmpty()) {
             throw new IllegalArgumentException(String.format("Couldn't define tag %s as it is missing following references: %s", p_176835_, list.stream().map(Objects::toString).collect(Collectors.joining(","))));
          } else {
             JsonObject jsonobject = p_176836_.m_13334_();
             Path path = this.m_6648_(p_176835_);
+            if (path == null) return; // Forge: Allow running this data provider without writing it. Recipe provider needs valid tags.
 
             try {
                String s = f_126542_.toJson((JsonElement)jsonobject);
@@ -88,6 +_,17 @@
       });
    }
 
+   private boolean missing(Tag.BuilderEntry reference) {
+      Tag.Entry entry = reference.f_13338_();
+      // We only care about non-optional tag entries, this is the only type that can reference a resource and needs validation
+      // Optional tags should not be validated
+
+      if (entry instanceof Tag.TagEntry nonOptionalEntry) {
+         return existingFileHelper == null || !existingFileHelper.exists(nonOptionalEntry.getId(), resourceType);
+      }
+      return false;
+   }
+
    protected Path m_6648_(ResourceLocation p_126561_) {
       ResourceKey<? extends Registry<T>> resourcekey = this.f_126540_.m_123023_();
       return this.f_126539_.m_123916_().resolve("data/" + p_126561_.m_135827_() + "/" + TagManager.m_203918_(resourcekey) + "/" + p_126561_.m_135815_() + ".json");
@@ -95,16 +_,17 @@
 
    protected TagsProvider.TagAppender<T> m_206424_(TagKey<T> p_206425_) {
       Tag.Builder tag$builder = this.m_206426_(p_206425_);
-      return new TagsProvider.TagAppender<>(tag$builder, this.f_126540_, "vanilla");
+      return new TagsProvider.TagAppender<>(tag$builder, this.f_126540_, modId);
    }
 
    protected Tag.Builder m_206426_(TagKey<T> p_206427_) {
       return this.f_126543_.computeIfAbsent(p_206427_.f_203868_(), (p_176838_) -> {
+         existingFileHelper.trackGenerated(p_176838_, resourceType);
          return new Tag.Builder();
       });
    }
 
-   protected static class TagAppender<T> {
+   public static class TagAppender<T> implements net.minecraftforge.common.extensions.IForgeTagAppender<T> {
       private final Tag.Builder f_126568_;
       public final Registry<T> f_126569_;
       private final String f_126570_;
@@ -150,6 +_,19 @@
             this.f_126568_.m_13327_(p_126587_, this.f_126570_);
          });
          return this;
+      }
+
+      public TagsProvider.TagAppender<T> add(Tag.Entry tag) {
+          f_126568_.m_13307_(tag, f_126570_);
+          return this;
+      }
+
+      public Tag.Builder getInternalBuilder() {
+          return f_126568_;
+      }
+
+      public String getModID() {
+          return f_126570_;
       }
    }
 }
