--- a/net/minecraft/data/tags/TagsProvider.java
+++ b/net/minecraft/data/tags/TagsProvider.java
@@ -31,28 +_,53 @@
     private final CompletableFuture<TagsProvider.TagLookup<T>> parentProvider;
     protected final ResourceKey<? extends Registry<T>> registryKey;
     protected final Map<Identifier, TagBuilder> builders = Maps.newLinkedHashMap();
+    protected final String modId;
+    @org.jetbrains.annotations.Nullable
+    protected final net.minecraftforge.common.data.ExistingFileHelper existingFileHelper;
+    private final net.minecraftforge.common.data.ExistingFileHelper.IResourceType resourceType;
+    private final net.minecraftforge.common.data.ExistingFileHelper.IResourceType elementResourceType; // FORGE: Resource type for validating required references to datapack registry elements.
 
     protected TagsProvider(
         final PackOutput output, final ResourceKey<? extends Registry<T>> registryKey, final CompletableFuture<HolderLookup.Provider> lookupProvider
     ) {
-        this(output, registryKey, lookupProvider, CompletableFuture.completedFuture(TagsProvider.TagLookup.empty()));
+        this(output, registryKey, lookupProvider, "vanilla", null);
+    }
+
+    protected TagsProvider(final PackOutput output, final ResourceKey<? extends Registry<T>> registryKey, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+       this(output, registryKey, lookupProvider, CompletableFuture.completedFuture(TagsProvider.TagLookup.empty()), modId, existingFileHelper);
+    }
+
+    protected TagsProvider(PackOutput output, ResourceKey<? extends Registry<T>> registryKey, CompletableFuture<HolderLookup.Provider> lookupProvider, CompletableFuture<TagsProvider.TagLookup<T>> parentProvider) {
+        this(output, registryKey, lookupProvider, parentProvider, "vanilla", null);
     }
 
     protected TagsProvider(
         final PackOutput output,
         final ResourceKey<? extends Registry<T>> registryKey,
         final CompletableFuture<HolderLookup.Provider> lookupProvider,
-        final CompletableFuture<TagsProvider.TagLookup<T>> parentProvider
+        final CompletableFuture<TagsProvider.TagLookup<T>> parentProvider,
+        final String modId,
+        final @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper
     ) {
         this.pathProvider = output.createRegistryTagsPathProvider(registryKey);
         this.registryKey = registryKey;
         this.parentProvider = parentProvider;
         this.lookupProvider = lookupProvider;
+        this.modId = modId;
+        this.existingFileHelper = existingFileHelper;
+        this.resourceType = new net.minecraftforge.common.data.ExistingFileHelper.ResourceType(net.minecraft.server.packs.PackType.SERVER_DATA, ".json", net.minecraft.core.registries.Registries.tagsDirPath(registryKey));
+        this.elementResourceType = new net.minecraftforge.common.data.ExistingFileHelper.ResourceType(net.minecraft.server.packs.PackType.SERVER_DATA, ".json", net.minecraft.core.registries.Registries.elementsDirPath(registryKey));
+    }
+
+    // Forge: Allow customizing the path for a given tag or returning null
+    @org.jetbrains.annotations.Nullable
+    protected Path getPath(Identifier id) {
+        return this.pathProvider.json(id);
     }
 
     @Override
     public String getName() {
-        return "Tags for " + this.registryKey.identifier();
+        return "Tags for " + this.registryKey.identifier() + " mod id " + this.modId;
     }
 
     protected abstract void addTags(HolderLookup.Provider registries);
@@ -70,7 +_,13 @@
             .thenCombineAsync(this.parentProvider, (x$0, x$1) -> new CombinedData<>(x$0, (TagsProvider.TagLookup<T>)x$1), Util.backgroundExecutor())
             .thenCompose(
                 c -> {
-                    HolderLookup.RegistryLookup<T> lookup = c.contents.lookupOrThrow(this.registryKey);
+                    HolderLookup.RegistryLookup<T> lookup = c.contents.lookup(this.registryKey).orElseThrow(() -> {
+                       // FORGE: Throw a more descriptive error message if this is a Forge registry without tags enabled
+                       if (net.minecraftforge.registries.RegistryManager.ACTIVE.getRegistry(this.registryKey) != null) {
+                          return new IllegalStateException("Forge registry " + this.registryKey.identifier() + " does not have support for tags");
+                       }
+                       return new IllegalStateException("Registry " + this.registryKey.identifier() + " not found");
+                    });
                     Predicate<Identifier> elementCheck = id -> lookup.get(ResourceKey.create(this.registryKey, id)).isPresent();
                     Predicate<Identifier> tagCheck = id -> this.builders.containsKey(id) || c.parent.contains(TagKey.create(this.registryKey, id));
                     return CompletableFuture.allOf(
@@ -82,7 +_,7 @@
                                     Identifier id = entry.getKey();
                                     TagBuilder builder = entry.getValue();
                                     List<TagEntry> entries = builder.build();
-                                    List<TagEntry> unresolvedEntries = entries.stream().filter(e -> !e.verifyIfPresent(elementCheck, tagCheck)).toList();
+                                    List<TagEntry> unresolvedEntries = java.util.stream.Stream.concat(entries.stream(), builder.getRemoveEntries()).filter(e -> !e.verifyIfPresent(elementCheck, tagCheck)).filter(this::missing).toList();
                                     if (!unresolvedEntries.isEmpty()) {
                                         throw new IllegalArgumentException(
                                             String.format(
@@ -94,8 +_,11 @@
                                         );
                                     }
 
-                                    Path path = this.pathProvider.json(id);
-                                    return DataProvider.saveStable(cache, c.contents, TagFile.CODEC, new TagFile(entries, builder.shouldReplace()), path);
+                                    Path path = this.getPath(id);
+                                    if (path == null) {
+                                        return CompletableFuture.completedFuture(null); // Forge: Allow running this data provider without writing it. Recipe provider needs valid tags.
+                                    }
+                                    return DataProvider.saveStable(cache, c.contents, TagFile.CODEC, new TagFile(entries, builder.shouldReplace(), builder.getRemoveEntries().toList()), path);
                                 }
                             )
                             .toArray(CompletableFuture[]::new)
@@ -105,7 +_,12 @@
     }
 
     protected TagBuilder getOrCreateRawBuilder(final TagKey<T> tag) {
-        return this.builders.computeIfAbsent(tag.location(), k -> TagBuilder.create());
+        return this.builders.computeIfAbsent(tag.location(), k -> {
+            if (existingFileHelper != null) {
+                existingFileHelper.trackGenerated(k, resourceType);
+            }
+            return TagBuilder.create();
+        });
     }
 
     public CompletableFuture<TagsProvider.TagLookup<T>> contentsGetter() {
@@ -120,15 +_,24 @@
         });
     }
 
+
+    private boolean missing(TagEntry reference) {
+        // Optional tags should not be validated
+        if (reference.isRequired()) {
+           return existingFileHelper == null || !existingFileHelper.exists(reference.getId(), reference.isTag() ? resourceType : elementResourceType);
+        }
+        return false;
+    }
+
     protected TagAppender<T> tag(final TagKey<T> tag) {
         TagBuilder builder = this.getOrCreateRawBuilder(tag);
-        return TagAppender.forBuilder(builder);
+        return TagAppender.forBuilder(builder, this.modId);
     }
 
     protected TagAppender<T> tag(final TagKey<T> tag, final boolean replace) {
         TagBuilder builder = this.getOrCreateRawBuilder(tag);
         builder.setReplace(replace);
-        return TagAppender.forBuilder(builder);
+        return TagAppender.forBuilder(builder, this.modId);
     }
 
     @FunctionalInterface
