--- ../src-base/minecraft/net/minecraft/item/crafting/FurnaceRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/FurnaceRecipes.java
@@ -4,6 +4,7 @@
 import java.util.Iterator;
 import java.util.Map;
 import java.util.Map.Entry;
+
 import net.minecraft.block.Block;
 import net.minecraft.init.Blocks;
 import net.minecraft.init.Items;
@@ -72,6 +73,7 @@
 
     public void func_151394_a(ItemStack p_151394_1_, ItemStack p_151394_2_, float p_151394_3_)
     {
+        if(p_151394_1_ == null || p_151394_2_ == null) throw new IllegalArgumentException("Found ItemStack with no associated item");
         this.field_77604_b.put(p_151394_1_, p_151394_2_);
         this.field_77605_c.put(p_151394_2_, Float.valueOf(p_151394_3_));
     }
@@ -107,6 +109,9 @@
 
     public float func_151398_b(ItemStack p_151398_1_)
     {
+        float ret = p_151398_1_.func_77973_b().getSmeltingExperience(p_151398_1_);
+        if (ret != -1) return ret;
+
         Iterator iterator = this.field_77605_c.entrySet().iterator();
         Entry entry;
 
