--- ../src-base/minecraft/net/minecraft/item/crafting/FurnaceRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/FurnaceRecipes.java
@@ -72,6 +72,7 @@
 
     public void func_151394_a(ItemStack p_151394_1_, ItemStack p_151394_2_, float p_151394_3_)
     {
+        if (func_151395_a(p_151394_1_) != null) { net.minecraftforge.fml.common.FMLLog.info("Ignored smelting recipe with conflicting input: " + p_151394_1_ + " = " + p_151394_2_); return; }
         this.field_77604_b.put(p_151394_1_, p_151394_2_);
         this.field_77605_c.put(p_151394_2_, Float.valueOf(p_151394_3_));
     }
@@ -101,6 +102,9 @@
 
     public float func_151398_b(ItemStack p_151398_1_)
     {
+        float ret = p_151398_1_.func_77973_b().getSmeltingExperience(p_151398_1_);
+        if (ret != -1) return ret;
+
         for (Entry<ItemStack, Float> entry : this.field_77605_c.entrySet())
         {
             if (this.func_151397_a(p_151398_1_, (ItemStack)entry.getKey()))
