--- a/net/minecraft/item/crafting/IRecipeSerializer.java
+++ b/net/minecraft/item/crafting/IRecipeSerializer.java
@@ -5,7 +5,7 @@
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.registry.Registry;
 
-public interface IRecipeSerializer<T extends IRecipe<?>> {
+public interface IRecipeSerializer<T extends IRecipe<?>> extends net.minecraftforge.registries.IForgeRegistryEntry<IRecipeSerializer<?>> {
    IRecipeSerializer<ShapedRecipe> field_222157_a = func_222156_a("crafting_shaped", new ShapedRecipe.Serializer());
    IRecipeSerializer<ShapelessRecipe> field_222158_b = func_222156_a("crafting_shapeless", new ShapelessRecipe.Serializer());
    SpecialRecipeSerializer<ArmorDyeRecipe> field_222159_c = func_222156_a("crafting_special_armordye", new SpecialRecipeSerializer<>(ArmorDyeRecipe::new));
@@ -30,6 +30,7 @@
 
    T func_199425_a_(ResourceLocation p_199425_1_, JsonObject p_199425_2_);
 
+   @javax.annotation.Nullable
    T func_199426_a_(ResourceLocation p_199426_1_, PacketBuffer p_199426_2_);
 
    void func_199427_a_(PacketBuffer p_199427_1_, T p_199427_2_);
