--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeRepairItem.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeRepairItem.java
@@ -7,10 +7,8 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeRepairItem implements IRecipe
+public class RecipeRepairItem extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
@@ -28,7 +26,7 @@
                 {
                     ItemStack itemstack1 = list.get(0);
 
-                    if (itemstack.func_77973_b() != itemstack1.func_77973_b() || itemstack1.func_190916_E() != 1 || itemstack.func_190916_E() != 1 || !itemstack1.func_77973_b().func_77645_m())
+                    if (itemstack.func_77973_b() != itemstack1.func_77973_b() || itemstack1.func_190916_E() != 1 || itemstack.func_190916_E() != 1 || !itemstack1.func_77973_b().isRepairable())
                     {
                         return false;
                     }
@@ -55,7 +53,7 @@
                 {
                     ItemStack itemstack1 = list.get(0);
 
-                    if (itemstack.func_77973_b() != itemstack1.func_77973_b() || itemstack1.func_190916_E() != 1 || itemstack.func_190916_E() != 1 || !itemstack1.func_77973_b().func_77645_m())
+                    if (itemstack.func_77973_b() != itemstack1.func_77973_b() || itemstack1.func_190916_E() != 1 || itemstack.func_190916_E() != 1 || !itemstack1.func_77973_b().isRepairable())
                     {
                         return ItemStack.field_190927_a;
                     }
@@ -68,13 +66,13 @@
             ItemStack itemstack2 = list.get(0);
             ItemStack itemstack3 = list.get(1);
 
-            if (itemstack2.func_77973_b() == itemstack3.func_77973_b() && itemstack2.func_190916_E() == 1 && itemstack3.func_190916_E() == 1 && itemstack2.func_77973_b().func_77645_m())
+            if (itemstack2.func_77973_b() == itemstack3.func_77973_b() && itemstack2.func_190916_E() == 1 && itemstack3.func_190916_E() == 1 && itemstack2.func_77973_b().isRepairable())
             {
-                Item item = itemstack2.func_77973_b();
-                int j = item.func_77612_l() - itemstack2.func_77952_i();
-                int k = item.func_77612_l() - itemstack3.func_77952_i();
-                int l = j + k + item.func_77612_l() * 5 / 100;
-                int i1 = item.func_77612_l() - l;
+                // FORGE: Make itemstack sensitive // Item item = itemstack2.getItem();
+                int j = itemstack2.func_77958_k() - itemstack2.func_77952_i();
+                int k = itemstack2.func_77958_k() - itemstack3.func_77952_i();
+                int l = j + k + itemstack2.func_77958_k() * 5 / 100;
+                int i1 = itemstack2.func_77958_k() - l;
 
                 if (i1 < 0)
                 {
@@ -100,11 +98,7 @@
         for (int i = 0; i < nonnulllist.size(); ++i)
         {
             ItemStack itemstack = p_179532_1_.func_70301_a(i);
-
-            if (itemstack.func_77973_b().func_77634_r())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
@@ -115,7 +109,6 @@
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
     public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
     {
         return p_194133_1_ * p_194133_2_ >= 2;
