--- a/net/minecraft/item/crafting/RecipeItemHelper.java
+++ b/net/minecraft/item/crafting/RecipeItemHelper.java
@@ -26,9 +26,13 @@
    }
 
    public void func_194112_a(ItemStack p_194112_1_) {
-      if (!p_194112_1_.func_190926_b()) {
-         int i = func_194113_b(p_194112_1_);
-         int j = p_194112_1_.func_190916_E();
+      this.accountStack(p_194112_1_, -1);
+   }
+
+   public void accountStack(ItemStack stack, int forceCount) {
+      if (!stack.func_190926_b()) {
+         int i = func_194113_b(stack);
+         int j = forceCount == -1 ? stack.func_190916_E() : forceCount;
          this.func_194117_b(i, j);
       }
 
