--- ../src_base/minecraft/net/minecraft/item/crafting/RecipeFireworks.java
+++ ../src_work/minecraft/net/minecraft/item/crafting/RecipeFireworks.java
@@ -91,9 +91,9 @@
             {
                 this.field_92102_a = new ItemStack(Item.firework);
 
+                nbttagcompound = new NBTTagCompound();
                 if (l > 0)
                 {
-                    nbttagcompound = new NBTTagCompound();
                     nbttagcompound1 = new NBTTagCompound("Fireworks");
                     NBTTagList nbttaglist = new NBTTagList("Explosions");
 
@@ -110,9 +110,9 @@
                     nbttagcompound1.setTag("Explosions", nbttaglist);
                     nbttagcompound1.setByte("Flight", (byte)j);
                     nbttagcompound.setTag("Fireworks", nbttagcompound1);
-                    this.field_92102_a.setTagCompound(nbttagcompound);
-                }
-
+                }
+
+                this.field_92102_a.setTagCompound(nbttagcompound);
                 return true;
             }
             else if (j == 1 && i == 0 && l == 0 && k > 0 && j1 <= 1)
