--- a/net/minecraft/item/crafting/RecipesArmorDyes.java
+++ b/net/minecraft/item/crafting/RecipesArmorDyes.java
@@ -10,8 +10,6 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.world.World;
-import net.minecraftforge.api.distmarker.Dist;
-import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class RecipesArmorDyes extends IRecipeHidden {
    public RecipesArmorDyes(ResourceLocation p_i48173_1_) {
@@ -35,7 +33,7 @@
 
                   itemstack = itemstack1;
                } else {
-                  if (!(itemstack1.func_77973_b() instanceof ItemDye)) {
+                  if (!itemstack1.func_77973_b().func_206844_a(net.minecraftforge.common.Tags.Items.DYES)) {
                      return false;
                   }
 
@@ -79,11 +77,12 @@
                   ++j;
                }
             } else {
-               if (!(item instanceof ItemDye)) {
+               net.minecraft.item.EnumDyeColor color = net.minecraft.item.EnumDyeColor.getColor(itemstack1);
+               if (color == null) {
                   return ItemStack.field_190927_a;
                }
 
-               float[] afloat = ((ItemDye)item).func_195962_g().func_193349_f();
+               float[] afloat = color.func_193349_f();
                int l1 = (int)(afloat[0] * 255.0F);
                int i2 = (int)(afloat[1] * 255.0F);
                int k2 = (int)(afloat[2] * 255.0F);
@@ -114,7 +113,6 @@
       }
    }
 
-   @OnlyIn(Dist.CLIENT)
    public boolean func_194133_a(int p_194133_1_, int p_194133_2_) {
       return p_194133_1_ * p_194133_2_ >= 2;
    }
