--- a/net/minecraft/item/crafting/RecipeBookCloning.java
+++ b/net/minecraft/item/crafting/RecipeBookCloning.java
@@ -9,8 +9,6 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.world.World;
-import net.minecraftforge.api.distmarker.Dist;
-import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class RecipeBookCloning extends IRecipeHidden {
    public RecipeBookCloning(ResourceLocation p_i48170_1_) {
@@ -86,8 +84,8 @@
 
       for(int i = 0; i < nonnulllist.size(); ++i) {
          ItemStack itemstack = p_179532_1_.func_70301_a(i);
-         if (itemstack.func_77973_b().func_77634_r()) {
-            nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
+         if (itemstack.hasContainerItem()) {
+            nonnulllist.set(i, itemstack.getContainerItem());
          } else if (itemstack.func_77973_b() instanceof ItemWrittenBook) {
             ItemStack itemstack1 = itemstack.func_77946_l();
             itemstack1.func_190920_e(1);
@@ -103,7 +101,6 @@
       return RecipeSerializers.field_199578_d;
    }
 
-   @OnlyIn(Dist.CLIENT)
    public boolean func_194133_a(int p_194133_1_, int p_194133_2_) {
       return p_194133_1_ >= 3 && p_194133_2_ >= 3;
    }
