--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeTippedArrow.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeTippedArrow.java
@@ -7,10 +7,8 @@
 import net.minecraft.potion.PotionUtils;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeTippedArrow implements IRecipe
+public class RecipeTippedArrow extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
@@ -83,7 +81,6 @@
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
     public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
     {
         return p_194133_1_ >= 2 && p_194133_2_ >= 2;
