--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeRepairItem.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeRepairItem.java
@@ -27,7 +27,7 @@
                 {
                     ItemStack itemstack1 = (ItemStack)list.get(0);
 
-                    if (itemstack.getItem() != itemstack1.getItem() || itemstack1.stackSize != 1 || itemstack.stackSize != 1 || !itemstack1.getItem().isDamageable())
+                    if (itemstack.getItem() != itemstack1.getItem() || itemstack1.stackSize != 1 || itemstack.stackSize != 1 || !itemstack1.getItem().isRepairable())
                     {
                         return false;
                     }
@@ -54,7 +54,7 @@
                 {
                     ItemStack itemstack1 = (ItemStack)list.get(0);
 
-                    if (itemstack.getItem() != itemstack1.getItem() || itemstack1.stackSize != 1 || itemstack.stackSize != 1 || !itemstack1.getItem().isDamageable())
+                    if (itemstack.getItem() != itemstack1.getItem() || itemstack1.stackSize != 1 || itemstack.stackSize != 1 || !itemstack1.getItem().isRepairable())
                     {
                         return null;
                     }
@@ -104,11 +104,7 @@
         for (int i = 0; i < aitemstack.length; ++i)
         {
             ItemStack itemstack = p_179532_1_.getStackInSlot(i);
-
-            if (itemstack != null && itemstack.getItem().hasContainerItem())
-            {
-                aitemstack[i] = new ItemStack(itemstack.getItem().getContainerItem());
-            }
+            aitemstack[i] = net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack);
         }
 
         return aitemstack;
