--- ../src-base/minecraft/net/minecraft/item/crafting/RecipesBanners.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipesBanners.java
@@ -12,12 +12,10 @@
 import net.minecraft.tileentity.TileEntityBanner;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class RecipesBanners
 {
-    public static class RecipeAddPattern implements IRecipe
+    public static class RecipeAddPattern extends net.minecraftforge.fml.common.registry.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
         {
             public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
             {
@@ -79,9 +77,10 @@
                     {
                         ItemStack itemstack2 = p_77572_1_.func_70301_a(j);
 
-                        if (itemstack2.func_77973_b() == Items.field_151100_aR)
+                        int color = getColor(itemstack2);
+                        if (color != -1)
                         {
-                            k = itemstack2.func_77960_j();
+                            k = color;
                             break;
                         }
                     }
@@ -120,11 +119,7 @@
                 for (int i = 0; i < nonnulllist.size(); ++i)
                 {
                     ItemStack itemstack = p_179532_1_.func_70301_a(i);
-
-                    if (itemstack.func_77973_b().func_77634_r())
-                    {
-                        nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
-                    }
+                    nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
                 }
 
                 return nonnulllist;
@@ -150,7 +145,7 @@
 
                                 if (!itemstack.func_190926_b() && itemstack.func_77973_b() != Items.field_179564_cE)
                                 {
-                                    if (itemstack.func_77973_b() == Items.field_151100_aR)
+                                    if (isDye(itemstack))
                                     {
                                         if (flag2)
                                         {
@@ -190,7 +185,7 @@
 
                                 if (!itemstack1.func_190926_b() && itemstack1.func_77973_b() != Items.field_179564_cE)
                                 {
-                                    if (itemstack1.func_77973_b() != Items.field_151100_aR)
+                                    if (!isDye(itemstack1))
                                     {
                                         flag = false;
                                         break;
@@ -237,14 +232,45 @@
                 return true;
             }
 
-            @SideOnly(Side.CLIENT)
             public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
             {
                 return p_194133_1_ >= 3 && p_194133_2_ >= 3;
             }
+
+            private static String[] colors = { "Black", "Red", "Green", "Brown", "Blue", "Purple", "Cyan", "LightGray", "Gray", "Pink", "Lime", "Yellow", "LightBlue", "Magenta", "Orange", "White" };
+            @SuppressWarnings("unchecked") //Why java...
+            private static java.util.List<ItemStack>[] colored = new java.util.List[colors.length];
+            private static java.util.List<ItemStack> dyes;
+            private static boolean hasInit = false;
+            private static void init()
+            {
+                if (hasInit) return;
+                for (int x = 0; x < colors.length; x++)
+                    colored[x] = net.minecraftforge.oredict.OreDictionary.getOres("dye" + colors[x]);
+                dyes = net.minecraftforge.oredict.OreDictionary.getOres("dye");
+                hasInit = true;
+            }
+            private boolean isDye(ItemStack stack)
+            {
+                init();
+                for (ItemStack ore : dyes)
+                    if (net.minecraftforge.oredict.OreDictionary.itemMatches(ore, stack, false))
+                        return true;
+                return false;
+            }
+            private int getColor(ItemStack stack)
+            {
+                init();
+                if (stack == null) return -1;
+                for (int x = 0; x < colored.length; x++)
+                    for (ItemStack ore : colored[x])
+                        if (net.minecraftforge.oredict.OreDictionary.itemMatches(ore, stack, true))
+                            return x;
+                return -1;
+            }
         }
 
-    public static class RecipeDuplicatePattern implements IRecipe
+    public static class RecipeDuplicatePattern extends net.minecraftforge.fml.common.registry.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
         {
             public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
             {
@@ -344,9 +370,9 @@
 
                     if (!itemstack.func_190926_b())
                     {
-                        if (itemstack.func_77973_b().func_77634_r())
+                        if (itemstack.func_77973_b().hasContainerItem(itemstack))
                         {
-                            nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
+                            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
                         }
                         else if (itemstack.func_77942_o() && TileEntityBanner.func_175113_c(itemstack) > 0)
                         {
@@ -365,7 +391,6 @@
                 return true;
             }
 
-            @SideOnly(Side.CLIENT)
             public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
             {
                 return p_194133_1_ * p_194133_2_ >= 2;
