--- a/net/minecraft/item/crafting/ShulkerBoxColoringRecipe.java
+++ b/net/minecraft/item/crafting/ShulkerBoxColoringRecipe.java
@@ -25,7 +_,7 @@
             if (Block.byItem(itemstack.getItem()) instanceof ShulkerBoxBlock) {
                ++i;
             } else {
-               if (!(itemstack.getItem() instanceof DyeItem)) {
+               if (!itemstack.getItem().is(net.minecraftforge.common.Tags.Items.DYES)) {
                   return false;
                }
 
@@ -43,7 +_,7 @@
 
    public ItemStack assemble(CraftingInventory p_77572_1_) {
       ItemStack itemstack = ItemStack.EMPTY;
-      DyeItem dyeitem = (DyeItem)Items.WHITE_DYE;
+      net.minecraft.item.DyeColor dyecolor = net.minecraft.item.DyeColor.WHITE;
 
       for(int i = 0; i < p_77572_1_.getContainerSize(); ++i) {
          ItemStack itemstack1 = p_77572_1_.getItem(i);
@@ -51,13 +_,14 @@
             Item item = itemstack1.getItem();
             if (Block.byItem(item) instanceof ShulkerBoxBlock) {
                itemstack = itemstack1;
-            } else if (item instanceof DyeItem) {
-               dyeitem = (DyeItem)item;
+            } else {
+               net.minecraft.item.DyeColor tmp = net.minecraft.item.DyeColor.getColor(itemstack1);
+               if (tmp != null) dyecolor = tmp;
             }
          }
       }
 
-      ItemStack itemstack2 = ShulkerBoxBlock.getColoredItemStack(dyeitem.getDyeColor());
+      ItemStack itemstack2 = ShulkerBoxBlock.getColoredItemStack(dyecolor);
       if (itemstack.hasTag()) {
          itemstack2.setTag(itemstack.getTag().copy());
       }
