--- a/net/minecraft/item/crafting/SingleItemRecipe.java
+++ b/net/minecraft/item/crafting/SingleItemRecipe.java
@@ -60,7 +_,7 @@
       return this.field_222132_b.func_77946_l();
    }
 
-   public static class Serializer<T extends SingleItemRecipe> implements IRecipeSerializer<T> {
+   public static class Serializer<T extends SingleItemRecipe> extends net.minecraftforge.registries.ForgeRegistryEntry<IRecipeSerializer<?>> implements IRecipeSerializer<T> {
       final SingleItemRecipe.Serializer.IRecipeFactory<T> field_222180_t;
 
       protected Serializer(SingleItemRecipe.Serializer.IRecipeFactory<T> p_i50146_1_) {
@@ -78,7 +_,7 @@
 
          String s1 = JSONUtils.func_151200_h(p_199425_2_, "result");
          int i = JSONUtils.func_151203_m(p_199425_2_, "count");
-         ItemStack itemstack = new ItemStack(Registry.field_212630_s.func_82594_a(new ResourceLocation(s1)), i);
+         ItemStack itemstack = net.minecraftforge.common.crafting.CraftingHelper.itemFromJson(p_199425_2_);
          return this.field_222180_t.create(p_199425_1_, s, ingredient, itemstack);
       }
 
