--- a/net/minecraft/item/crafting/SingleItemRecipe.java
+++ b/net/minecraft/item/crafting/SingleItemRecipe.java
@@ -60,7 +60,7 @@
       return this.field_222132_b.func_77946_l();
    }
 
-   public static class Serializer<T extends SingleItemRecipe> implements IRecipeSerializer<T> {
+   public static class Serializer<T extends SingleItemRecipe> extends net.minecraftforge.registries.ForgeRegistryEntry<IRecipeSerializer<?>> implements IRecipeSerializer<T> {
       final SingleItemRecipe.Serializer.IRecipeFactory<T> field_222180_t;
 
       protected Serializer(SingleItemRecipe.Serializer.IRecipeFactory<T> p_i50146_1_) {
