--- a/net/minecraft/item/crafting/SingleItemRecipe.java
+++ b/net/minecraft/item/crafting/SingleItemRecipe.java
@@ -60,7 +_,7 @@
       return this.result.copy();
    }
 
-   public static class Serializer<T extends SingleItemRecipe> implements IRecipeSerializer<T> {
+   public static class Serializer<T extends SingleItemRecipe> extends net.minecraftforge.registries.ForgeRegistryEntry<IRecipeSerializer<?>> implements IRecipeSerializer<T> {
       final SingleItemRecipe.Serializer.IRecipeFactory<T> factory;
 
       protected Serializer(SingleItemRecipe.Serializer.IRecipeFactory<T> p_i50146_1_) {
