--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenIceSpike.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenIceSpike.java
@@ -51,7 +51,7 @@
                             IBlockState iblockstate = worldIn.getBlockState(position.add(i1, k, j1));
                             Block block = iblockstate.getBlock();
 
-                            if (iblockstate.func_185904_a() == Material.air || block == Blocks.dirt || block == Blocks.snow || block == Blocks.ice)
+                            if (iblockstate.getBlock().isAir(iblockstate, worldIn, position.add(i1, k, j1)) || block == Blocks.dirt || block == Blocks.snow || block == Blocks.ice)
                             {
                                 this.setBlockAndNotifyAdequately(worldIn, position.add(i1, k, j1), Blocks.packed_ice.getDefaultState());
                             }
@@ -61,7 +61,7 @@
                                 iblockstate = worldIn.getBlockState(position.add(i1, -k, j1));
                                 block = iblockstate.getBlock();
 
-                                if (iblockstate.func_185904_a() == Material.air || block == Blocks.dirt || block == Blocks.snow || block == Blocks.ice)
+                                if (iblockstate.getBlock().isAir(iblockstate, worldIn, position.add(i1, -k, j1)) || block == Blocks.dirt || block == Blocks.snow || block == Blocks.ice)
                                 {
                                     this.setBlockAndNotifyAdequately(worldIn, position.add(i1, -k, j1), Blocks.packed_ice.getDefaultState());
                                 }
@@ -99,7 +99,7 @@
                         IBlockState iblockstate1 = worldIn.getBlockState(blockpos);
                         Block block1 = iblockstate1.getBlock();
 
-                        if (iblockstate1.func_185904_a() != Material.air && block1 != Blocks.dirt && block1 != Blocks.snow && block1 != Blocks.ice && block1 != Blocks.packed_ice)
+                        if (!iblockstate1.getBlock().isAir(iblockstate1, worldIn, blockpos) && block1 != Blocks.dirt && block1 != Blocks.snow && block1 != Blocks.ice && block1 != Blocks.packed_ice)
                         {
                             break;
                         }
