--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenCanopyTree.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenCanopyTree.java
@@ -33,9 +33,10 @@
         if (k >= 1 && k + i + 1 < 256)
         {
             BlockPos blockpos = position.down();
-            Block block = worldIn.getBlockState(blockpos).getBlock();
+            IBlockState state = worldIn.getBlockState(blockpos);
+            boolean isSoil = state.getBlock().canSustainPlant(state, worldIn, blockpos, net.minecraft.util.EnumFacing.UP, ((net.minecraft.block.BlockSapling)Blocks.sapling));
 
-            if (block != Blocks.grass && block != Blocks.dirt)
+            if (!(isSoil && position.getY() < worldIn.getHeight() - i - 1))
             {
                 return false;
             }
@@ -45,10 +46,10 @@
             }
             else
             {
-                this.func_175921_a(worldIn, blockpos);
-                this.func_175921_a(worldIn, blockpos.east());
-                this.func_175921_a(worldIn, blockpos.south());
-                this.func_175921_a(worldIn, blockpos.south().east());
+                this.onPlantGrow(worldIn, blockpos, position);
+                this.onPlantGrow(worldIn, blockpos.east(), position);
+                this.onPlantGrow(worldIn, blockpos.south(), position);
+                this.onPlantGrow(worldIn, blockpos.south().east(), position);
                 EnumFacing enumfacing = EnumFacing.Plane.HORIZONTAL.random(rand);
                 int i1 = i - rand.nextInt(4);
                 int j1 = 2 - rand.nextInt(3);
@@ -67,9 +68,9 @@
 
                     int k2 = k + j2;
                     BlockPos blockpos1 = new BlockPos(k1, k2, l1);
-                    Material material = worldIn.getBlockState(blockpos1).func_185904_a();
+                    state = worldIn.getBlockState(blockpos1);
 
-                    if (material == Material.air || material == Material.leaves)
+                    if (state.getBlock().isAir(state, worldIn, blockpos1) || state.getBlock().isLeaves(state, worldIn, blockpos1))
                     {
                         this.func_181639_b(worldIn, blockpos1);
                         this.func_181639_b(worldIn, blockpos1.east());
@@ -187,7 +188,7 @@
             {
                 for (int k1 = -i1; k1 <= i1; ++k1)
                 {
-                    if (!this.func_150523_a(p_181638_1_.getBlockState(blockpos$mutableblockpos.set(i + j1, j + l, k + k1)).getBlock()))
+                    if (!this.isReplaceable(p_181638_1_, blockpos$mutableblockpos.set(i + j1, j + l, k + k1)))
                     {
                         return false;
                     }
@@ -209,11 +210,18 @@
     private void func_150526_a(World worldIn, int p_150526_2_, int p_150526_3_, int p_150526_4_)
     {
         BlockPos blockpos = new BlockPos(p_150526_2_, p_150526_3_, p_150526_4_);
-        Material material = worldIn.getBlockState(blockpos).func_185904_a();
+        IBlockState state = worldIn.getBlockState(blockpos);
 
-        if (material == Material.air)
+        if (state.getBlock().isAir(state, worldIn, blockpos))
         {
             this.setBlockAndNotifyAdequately(worldIn, blockpos, field_181641_b);
         }
     }
+
+    //Just a helper macro
+    private void onPlantGrow(World world, BlockPos pos, BlockPos source)
+    {
+        IBlockState state = world.getBlockState(pos);
+        state.getBlock().onPlantGrow(state, world, pos, source);
+    }
 }
