--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenAbstractTree.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenAbstractTree.java
@@ -31,4 +31,10 @@
             this.setBlockAndNotifyAdequately(worldIn, p_175921_2_, Blocks.dirt.getDefaultState());
         }
     }
+
+    public boolean isReplaceable(World world, BlockPos pos)
+    {
+        net.minecraft.block.state.IBlockState state = world.getBlockState(pos);
+        return state.getBlock().isAir(world, pos) || state.getBlock().isLeaves(world, pos) || state.getBlock().isWood(world, pos) || func_150523_a(state.getBlock());
+    }
 }
