--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenBigMushroom.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenBigMushroom.java
@@ -55,9 +55,10 @@
                     {
                         if (k >= 0 && k < 256)
                         {
-                            Block block = worldIn.getBlockState(new BlockPos(l, k, i1)).getBlock();
+                            BlockPos pos = new BlockPos(l, k, i1);
+                            net.minecraft.block.state.IBlockState state = worldIn.getBlockState(pos);
 
-                            if (block.getMaterial() != Material.air && block.getMaterial() != Material.leaves)
+                            if (!state.getBlock().isAir(worldIn, pos) && !state.getBlock().isLeaves(worldIn, pos))
                             {
                                 flag = false;
                             }
@@ -188,7 +189,7 @@
                                 {
                                     BlockPos blockpos1 = new BlockPos(i2, l, j1);
 
-                                    if (!worldIn.getBlockState(blockpos1).getBlock().isFullBlock())
+                                    if (worldIn.getBlockState(blockpos1).getBlock().canBeReplacedByLeaves(worldIn, blockpos1))
                                     {
                                         this.func_175905_a(worldIn, blockpos1, Block.getBlockById(Block.getIdFromBlock(Blocks.brown_mushroom_block) + i), k1);
                                     }
@@ -199,9 +200,10 @@
 
                     for (l = 0; l < j; ++l)
                     {
-                        Block block2 = worldIn.getBlockState(p_180709_3_.up(l)).getBlock();
+                        BlockPos upN = p_180709_3_.up(l);
+                        net.minecraft.block.state.IBlockState state = worldIn.getBlockState(upN);
 
-                        if (!block2.isFullBlock())
+                        if (state.getBlock().canBeReplacedByLeaves(worldIn, upN))
                         {
                             this.func_175905_a(worldIn, p_180709_3_.up(l), Block.getBlockById(Block.getIdFromBlock(Blocks.brown_mushroom_block) + i), 10);
                         }
