--- a/net/minecraft/world/level/material/Fluid.java
+++ b/net/minecraft/world/level/material/Fluid.java
@@ -20,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public abstract class Fluid {
+public abstract class Fluid implements net.minecraftforge.common.extensions.IForgeFluid {
    public static final IdMapper<FluidState> f_76104_ = new IdMapper<>();
    protected final StateDefinition<Fluid, FluidState> f_76105_;
    private FluidState f_76103_;
@@ -101,6 +_,13 @@
    }
 
    public abstract VoxelShape m_7999_(FluidState p_76137_, BlockGetter p_76138_, BlockPos p_76139_);
+
+   private net.minecraftforge.fluids.FluidType forgeFluidType;
+   @Override
+   public net.minecraftforge.fluids.FluidType getFluidType() {
+      if (forgeFluidType == null) forgeFluidType = net.minecraftforge.common.ForgeHooks.getVanillaFluidType(this);
+      return forgeFluidType;
+   }
 
    public Optional<SoundEvent> m_142520_() {
       return Optional.empty();
