--- ../src-base/minecraft/net/minecraft/network/play/client/CPacketClickWindow.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/CPacketClickWindow.java
@@ -55,7 +55,7 @@
         p_148840_1_.writeByte(this.field_149553_c);
         p_148840_1_.writeShort(this.field_149550_d);
         p_148840_1_.func_179249_a(this.field_149549_f);
-        p_148840_1_.func_150788_a(this.field_149551_e);
+        net.minecraftforge.common.util.PacketUtil.writeItemStackFromClientToServer(p_148840_1_, this.field_149551_e);
     }
 
     public int func_149548_c()
