--- ../src-base/minecraft/net/minecraft/inventory/SlotCrafting.java
+++ ../src-work/minecraft/net/minecraft/inventory/SlotCrafting.java
@@ -1,5 +1,6 @@
 package net.minecraft.inventory;
 
+import cpw.mods.fml.common.registry.GameRegistry;
 import net.minecraft.block.Block;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.Item;
@@ -89,6 +90,7 @@
 
     public void func_82870_a(EntityPlayer p_82870_1_, ItemStack p_82870_2_)
     {
+        GameRegistry.onItemCrafted(p_82870_1_, p_82870_2_, field_75239_a);
         this.func_75208_c(p_82870_2_);
 
         for (int var3 = 0; var3 < this.field_75239_a.func_70302_i_(); ++var3)
