--- ../src-base/minecraft/net/minecraft/server/network/NetHandlerLoginServer.java
+++ ../src-work/minecraft/net/minecraft/server/network/NetHandlerLoginServer.java
@@ -62,7 +62,7 @@
             this.func_147326_c();
         }
 
-        if (this.field_147336_h++ == 600)
+        if (this.field_147336_h++ == net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.LOGIN_TIMEOUT)
         {
             this.func_147322_a("Took too long to log in");
         }
@@ -113,7 +113,7 @@
             }
 
             this.field_147333_a.func_179290_a(new S02PacketLoginSuccess(this.field_147337_i));
-            this.field_147327_f.func_71203_ab().func_72355_a(this.field_147333_a, this.field_147327_f.func_71203_ab().func_148545_a(this.field_147337_i));
+            net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.fmlServerHandshake(this.field_147327_f.func_71203_ab(), this.field_147333_a, this.field_147327_f.func_71203_ab().func_148545_a(this.field_147337_i));
         }
     }
 
