--- a/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -21,6 +_,7 @@
    }
 
    public void m_7322_(ClientIntentionPacket p_9975_) {
+      if (!net.minecraftforge.server.ServerLifecycleHooks.handleServerLogin(p_9975_, this.f_9966_)) return;
       switch (p_9975_.f_134723_()) {
          case LOGIN:
             this.f_9966_.m_294993_(ClientIntent.LOGIN);
@@ -42,7 +_,7 @@
             ServerStatus serverstatus = this.f_9965_.m_129928_();
             if (this.f_9965_.m_6373_() && serverstatus != null) {
                this.f_9966_.m_294993_(ClientIntent.STATUS);
-               this.f_9966_.m_129505_(new ServerStatusPacketListenerImpl(serverstatus, this.f_9966_));
+               this.f_9966_.m_129505_(new ServerStatusPacketListenerImpl(serverstatus, this.f_9966_, this.f_9965_.getStatusJson()));
             } else {
                this.f_9966_.m_129507_(f_9964_);
             }
