--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -180,7 +_,7 @@
          throw new IllegalStateException("Protocol error", cryptexception);
       }
 
-      Thread thread = new Thread("User Authenticator #" + UNIQUE_THREAD_ID.incrementAndGet()) {
+      Thread thread = new Thread(net.minecraftforge.fml.util.thread.SidedThreadGroups.SERVER, "User Authenticator #" + UNIQUE_THREAD_ID.incrementAndGet()) {
          public void run() {
             String s1 = Objects.requireNonNull(ServerLoginPacketListenerImpl.this.requestedUsername, "Player name not initialized");
 
@@ -220,6 +_,7 @@
    }
 
    public void handleCustomQueryPacket(ServerboundCustomQueryAnswerPacket p_297965_) {
+      if (!net.minecraftforge.common.ForgeHooks.onCustomPayload(p_297965_, this.connection))
       this.disconnect(DISCONNECT_UNEXPECTED_QUERY);
    }
 
@@ -230,6 +_,11 @@
       this.connection.setListener(serverconfigurationpacketlistenerimpl);
       serverconfigurationpacketlistenerimpl.startConfiguration();
       this.state = ServerLoginPacketListenerImpl.State.ACCEPTED;
+   }
+
+   @Nullable
+   public GameProfile getAuthenticatedProfile() {
+       return this.authenticatedProfile;
    }
 
    protected static GameProfile createOfflineProfile(String p_297551_) {
