--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -46,17 +_,20 @@
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final Component DISCONNECT_REASON_INVALID_DATA = Component.translatable("multiplayer.disconnect.invalid_player_data");
     private static final Component DISCONNECT_REASON_CONFIGURATION_ERROR = Component.translatable("multiplayer.disconnect.configuration_error");
+    private static final ConfigurationTask.Type VANILLA_START = new ConfigurationTask.Type("vanilla_start");
     private final GameProfile gameProfile;
     private final Queue<ConfigurationTask> configurationTasks = new ConcurrentLinkedQueue<>();
     private @Nullable ConfigurationTask currentTask;
     private ClientInformation clientInformation;
     private @Nullable SynchronizeRegistriesTask synchronizeRegistriesTask;
     private @Nullable PrepareSpawnTask prepareSpawnTask;
+    private final net.minecraftforge.network.config.ConfigurationTaskContext taskContext;
 
     public ServerConfigurationPacketListenerImpl(final MinecraftServer server, final Connection connection, final CommonListenerCookie cookie) {
         super(server, connection, cookie);
         this.gameProfile = cookie.gameProfile();
         this.clientInformation = cookie.clientInformation();
+        this.taskContext = new net.minecraftforge.network.config.ConfigurationTaskContext(connection, this::send, this::finishCurrentTask);
     }
 
     @Override
@@ -80,22 +_,27 @@
         return this.connection.isConnected();
     }
 
-    public void startConfiguration() {
+    public void vanillaStart() {
         this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
         ServerLinks serverLinks = this.server.serverLinks();
         if (!serverLinks.isEmpty()) {
             this.send(new ClientboundServerLinksPacket(serverLinks.untrust()));
         }
 
+        this.send(new ClientboundUpdateEnabledFeaturesPacket(FeatureFlags.REGISTRY.toNames(this.server.getWorldData().enabledFeatures())));
+    }
+
+    public void startConfiguration() {
+        net.minecraftforge.event.ForgeEventFactory.gatherLoginConfigTasks(this.connection, this.configurationTasks::add);
         LayeredRegistryAccess<RegistryLayer> registries = this.server.registries();
         List<KnownPack> knownPacks = this.server
             .getResourceManager()
             .listPacks()
             .flatMap(packResources -> packResources.location().knownPackInfo().stream())
             .toList();
-        this.send(new ClientboundUpdateEnabledFeaturesPacket(FeatureFlags.REGISTRY.toNames(this.server.getWorldData().enabledFeatures())));
         this.synchronizeRegistriesTask = new SynchronizeRegistriesTask(knownPacks, registries);
         this.configurationTasks.add(this.synchronizeRegistriesTask);
+        this.configurationTasks.add(new net.minecraftforge.network.config.SimpleConfigurationTask(VANILLA_START, this::vanillaStart));
         this.addOptionalTasks();
         this.returnToWorld();
     }
@@ -213,7 +_,7 @@
                 this.currentTask = task;
 
                 try {
-                    task.start(this::send);
+                    task.start(this.taskContext);
                 } catch (Exception e) {
                     LOGGER.error("Failed to start configuration task {}", task.type(), e);
                     this.disconnect(DISCONNECT_REASON_CONFIGURATION_ERROR);
