--- a/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
@@ -12,12 +_,17 @@
 public class ServerStatusPacketListenerImpl implements ServerStatusPacketListener {
    private static final Component DISCONNECT_REASON = Component.translatable("multiplayer.status.request_handled");
    private final ServerStatus status;
+   private final @org.jetbrains.annotations.Nullable String statusCache; // FORGE: cache status JSON
    private final Connection connection;
    private boolean hasRequestedStatus;
 
    public ServerStatusPacketListenerImpl(ServerStatus p_272864_, Connection p_273586_) {
+      this(p_272864_, p_273586_, null);
+   }
+   public ServerStatusPacketListenerImpl(ServerStatus p_272864_, Connection p_273586_, @org.jetbrains.annotations.Nullable String statusCache) {
       this.status = p_272864_;
       this.connection = p_273586_;
+      this.statusCache = statusCache;
    }
 
    public void onDisconnect(Component p_10091_) {
@@ -32,7 +_,7 @@
          this.connection.disconnect(DISCONNECT_REASON);
       } else {
          this.hasRequestedStatus = true;
-         this.connection.send(new ClientboundStatusResponsePacket(this.status));
+         this.connection.send(new ClientboundStatusResponsePacket(this.status, this.statusCache));
       }
    }
 
