--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -188,7 +_,7 @@
             throw new IllegalStateException("Protocol error", e);
         }
 
-        Thread thread = new Thread("User Authenticator #" + UNIQUE_THREAD_ID.incrementAndGet()) {
+        Thread thread = new Thread(net.minecraftforge.fml.util.thread.SidedThreadGroups.SERVER, "User Authenticator #" + UNIQUE_THREAD_ID.incrementAndGet()) {
             @Override
             public void run() {
                 String name = Objects.requireNonNull(ServerLoginPacketListenerImpl.this.requestedUsername, "Player name not initialized");
@@ -234,6 +_,7 @@
 
     @Override
     public void handleCustomQueryPacket(final ServerboundCustomQueryAnswerPacket packet) {
+        if (!net.minecraftforge.common.ForgeHooks.onCustomPayload(packet, this.connection))
         this.disconnect(ServerCommonPacketListenerImpl.DISCONNECT_UNEXPECTED_QUERY);
     }
 
@@ -256,6 +_,11 @@
     @Override
     public void handleCookieResponse(final ServerboundCookieResponsePacket packet) {
         this.disconnect(ServerCommonPacketListenerImpl.DISCONNECT_UNEXPECTED_QUERY);
+    }
+
+    @Nullable
+    public GameProfile getAuthenticatedProfile() {
+        return this.authenticatedProfile;
     }
 
     private enum State {
