--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1297,9 +_,10 @@
                     }
                 case SWAP_ITEM_WITH_OFFHAND:
                     if (!this.player.isSpectator()) {
-                        ItemStack swap = this.player.getItemInHand(InteractionHand.OFF_HAND);
-                        this.player.setItemInHand(InteractionHand.OFF_HAND, this.player.getItemInHand(InteractionHand.MAIN_HAND));
-                        this.player.setItemInHand(InteractionHand.MAIN_HAND, swap);
+                        var event = net.minecraftforge.event.ForgeEventFactory.onLivingSwapHandItems(this.player);
+                        if (event == null) return;
+                        this.player.setItemInHand(InteractionHand.OFF_HAND, event.getItemSwappedToOffHand());
+                        this.player.setItemInHand(InteractionHand.MAIN_HAND, event.getItemSwappedToMainHand());
                         this.player.stopUsingItem();
                     }
 
@@ -1515,8 +_,9 @@
                 }
 
                 CompletableFuture<FilteredText> filteredFuture = this.filterTextPacket(signedMessage.signedContent());
-                Component decorated = this.server.getChatDecorator().decorate(this.player, signedMessage.decoratedContent());
+                Component decorated = net.minecraftforge.common.ForgeHooks.onServerChatSubmittedEvent(this.player, signedMessage.decoratedContent());
                 this.chatMessageChain.append(filteredFuture, filtered -> {
+                    if (decorated == null) return;
                     PlayerChatMessage filteredMessage = signedMessage.withUnsignedContent(decorated).filter(filtered.mask());
                     this.broadcastChatMessage(filteredMessage);
                 });
@@ -2195,6 +_,7 @@
 
     @Override
     public void handleCustomPayload(final ServerboundCustomPayloadPacket packet) {
+        net.minecraftforge.common.ForgeHooks.onCustomPayload(packet.payload(), this.connection);
     }
 
     @Override
