--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -82,6 +_,7 @@
 
     @Override
     public void handleCustomPayload(ServerboundCustomPayloadPacket p_300164_) {
+        net.minecraftforge.common.ForgeHooks.onCustomPayload(p_300164_.payload(), this.connection);
     }
 
     @Override
@@ -180,5 +_,9 @@
 
     protected CommonListenerCookie createCookie(ClientInformation p_297318_) {
         return new CommonListenerCookie(this.playerProfile(), this.latency, p_297318_, this.transferred);
+    }
+
+    public Connection getConnection() {
+        return this.connection;
     }
 }
