--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -40,6 +_,7 @@
 import org.slf4j.Logger;
 
 public class ServerConnectionListener {
+    private static final int READ_TIMEOUT = Integer.parseInt(System.getProperty("forge.readTimeout", "30"));
     private static final Logger LOGGER = LogUtils.getLogger();
     private final MinecraftServer server;
     public volatile boolean running;
@@ -53,6 +_,8 @@
     }
 
     public void startTcpServerListener(final @Nullable InetAddress address, final int port) throws IOException {
+        var fixedAddress = address != null ? address : new java.net.InetSocketAddress(port).getAddress();
+        net.minecraftforge.network.DualStackUtils.checkIPv6(fixedAddress);
         synchronized (this.channels) {
             EventLoopGroupHolder eventLoopGroupHolder = EventLoopGroupHolder.remote(this.server.useNativeTransport());
             this.channels
@@ -68,7 +_,7 @@
                                     } catch (ChannelException var5) {
                                     }
 
-                                    ChannelPipeline pipeline = channel.pipeline().addLast("timeout", new ReadTimeoutHandler(30));
+                                    ChannelPipeline pipeline = channel.pipeline().addLast("timeout", new ReadTimeoutHandler(READ_TIMEOUT));
                                     if (ServerConnectionListener.this.server.repliesToStatus()) {
                                         pipeline.addLast("legacy_query", new LegacyQueryHandler(ServerConnectionListener.this.getServer()));
                                     }
@@ -87,7 +_,7 @@
                             }
                         )
                         .group(eventLoopGroupHolder.eventLoopGroup())
-                        .localAddress(address, port)
+                        .localAddress(fixedAddress, port)
                         .bind()
                         .syncUninterruptibly()
                 );
