--- a/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
@@ -12,12 +_,17 @@
 public class ServerStatusPacketListenerImpl implements ServerStatusPacketListener {
    private static final Component f_10081_ = Component.m_237115_("multiplayer.status.request_handled");
    private final ServerStatus f_271204_;
+   private final @org.jetbrains.annotations.Nullable String statusCache; // FORGE: cache status JSON
    private final Connection f_10083_;
    private boolean f_10084_;
 
    public ServerStatusPacketListenerImpl(ServerStatus p_272864_, Connection p_273586_) {
+      this(p_272864_, p_273586_, null);
+   }
+   public ServerStatusPacketListenerImpl(ServerStatus p_272864_, Connection p_273586_, @org.jetbrains.annotations.Nullable String statusCache) {
       this.f_271204_ = p_272864_;
       this.f_10083_ = p_273586_;
+      this.statusCache = statusCache;
    }
 
    public void m_7026_(Component p_10091_) {
@@ -32,7 +_,7 @@
          this.f_10083_.m_129507_(f_10081_);
       } else {
          this.f_10084_ = true;
-         this.f_10083_.m_129512_(new ClientboundStatusResponsePacket(this.f_271204_));
+         this.f_10083_.m_129512_(new ClientboundStatusResponsePacket(this.f_271204_, this.statusCache));
       }
    }
 
