--- ../src-base/minecraft/net/minecraft/util/BlockPos.java
+++ ../src-work/minecraft/net/minecraft/util/BlockPos.java
@@ -257,6 +257,21 @@
         };
     }
 
+    /**
+     * Returns a version of this BlockPos that is guaranteed to be Immutable.
+     * In most cases this will return 'this', but if 'this' is an instance of
+     * MutableBlockPos it will return a new instance of BlockPos with the same values.
+     *
+     * When storing a parameter given to you for an extended period of time, make sure you
+     * use this in case the value is changed internally.
+     *
+     * @return An immutable BlockPos.
+     */
+    public BlockPos getImmutable()
+    {
+        return this;
+    }
+
     public static final class MutableBlockPos extends BlockPos
         {
             private int field_177997_b;
@@ -298,5 +313,7 @@
                 this.field_177996_d = p_181079_3_;
                 return this;
             }
+
+            @Override public BlockPos getImmutable() { return new BlockPos(this); }
         }
 }
