--- a/net/minecraft/util/ClassInheritanceMultiMap.java
+++ b/net/minecraft/util/ClassInheritanceMultiMap.java
@@ -12,7 +12,8 @@
 import java.util.Set;
 
 public class ClassInheritanceMultiMap<T> extends AbstractSet<T> {
-   private static final Set<Class<?>> field_181158_a = Sets.newHashSet();
+   // Forge: Use concurrent collection to allow creating chunks from multiple threads safely
+   private static final Set<Class<?>> field_181158_a = java.util.Collections.newSetFromMap(new java.util.concurrent.ConcurrentHashMap<Class<?>, Boolean>());
    private final Map<Class<?>, List<T>> field_180218_a = Maps.newHashMap();
    private final Set<Class<?>> field_180216_b = Sets.newIdentityHashSet();
    private final Class<T> field_180217_c;
