--- ../src-base/minecraft/net/minecraft/util/CryptManager.java
+++ ../src-work/minecraft/net/minecraft/util/CryptManager.java
@@ -1,5 +1,6 @@
 package net.minecraft.util;
 
+import cpw.mods.fml.common.asm.ReobfuscationMarker;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.io.InputStream;
@@ -35,6 +36,7 @@
 import org.bouncycastle.crypto.params.ParametersWithIV;
 import org.bouncycastle.jce.provider.BouncyCastleProvider;
 
+@ReobfuscationMarker
 public class CryptManager
 {
     @SideOnly(Side.CLIENT)
