--- a/net/minecraft/util/Direction.java
+++ b/net/minecraft/util/Direction.java
@@ -119,7 +119,6 @@
       return func_82600_a(this.field_176759_h);
    }
 
-   @OnlyIn(Dist.CLIENT)
    public Direction func_176732_a(Direction.Axis p_176732_1_) {
       switch(p_176732_1_) {
       case X:
@@ -160,7 +159,6 @@
       }
    }
 
-   @OnlyIn(Dist.CLIENT)
    private Direction func_176744_n() {
       switch(this) {
       case NORTH:
@@ -178,7 +176,6 @@
       }
    }
 
-   @OnlyIn(Dist.CLIENT)
    private Direction func_176738_p() {
       switch(this) {
       case EAST:
@@ -231,7 +228,6 @@
    }
 
    @Nullable
-   @OnlyIn(Dist.CLIENT)
    public static Direction func_176739_a(@Nullable String p_176739_0_) {
       return p_176739_0_ == null ? null : field_176761_p.get(p_176739_0_.toLowerCase(Locale.ROOT));
    }
@@ -353,7 +349,6 @@
       }
 
       @Nullable
-      @OnlyIn(Dist.CLIENT)
       public static Direction.Axis func_176717_a(String p_176717_0_) {
          return field_176725_d.get(p_176717_0_.toLowerCase(Locale.ROOT));
       }
