--- ../src-base/minecraft/net/minecraft/util/IChatComponent.java
+++ ../src-work/minecraft/net/minecraft/util/IChatComponent.java
@@ -31,7 +31,6 @@
 
     String func_150260_c();
 
-    @SideOnly(Side.CLIENT)
     String func_150254_d();
 
     List<IChatComponent> func_150253_a();
@@ -91,6 +90,7 @@
 
                         if (jsonobject.has("with"))
                         {
+                            if (!jsonobject.get("with").isJsonArray()) throw new JsonParseException("'with' tag has to be an array");
                             JsonArray jsonarray = jsonobject.getAsJsonArray("with");
                             Object[] aobject = new Object[jsonarray.size()];
 
@@ -143,6 +143,7 @@
                     }
 
                     if (jsonobject.has("extra"))
+                    if (!jsonobject.get("extra").isJsonArray()) ichatcomponent.func_150257_a(p_deserialize_3_.<IChatComponent>deserialize(jsonobject.get("extra"), IChatComponent.class)); else
                     {
                         JsonArray jsonarray2 = jsonobject.getAsJsonArray("extra");
 
