--- ../src_base/minecraft/net/minecraft/util/MovingObjectPosition.java
+++ ../src_work/minecraft/net/minecraft/util/MovingObjectPosition.java
@@ -27,7 +27,13 @@
 
     /** The hit entity */
     public Entity entityHit;
-
+    
+    /** Used to determine what sub-segment is hit */
+    public int subHit = -1;
+    
+    /** Used to add extra hit info */
+    public Object hitInfo = null;
+    
     public MovingObjectPosition(int par1, int par2, int par3, int par4, Vec3 par5Vec3)
     {
         this.typeOfHit = EnumMovingObjectType.TILE;
