--- ../src-base/minecraft/net/minecraft/util/MovingObjectPosition.java
+++ ../src-work/minecraft/net/minecraft/util/MovingObjectPosition.java
@@ -13,6 +13,12 @@
     public Entity field_72308_g;
     private static final String __OBFID = "CL_00000610";
 
+    /** Used to determine what sub-segment is hit */
+    public int subHit = -1;
+
+    /** Used to add extra hit info */
+    public Object hitInfo = null;
+
     public MovingObjectPosition(int p_i2303_1_, int p_i2303_2_, int p_i2303_3_, int p_i2303_4_, Vec3 p_i2303_5_)
     {
         this(p_i2303_1_, p_i2303_2_, p_i2303_3_, p_i2303_4_, p_i2303_5_, true);
