--- ../src-base/minecraft/net/minecraft/util/ServerRecipeBookHelper.java
+++ ../src-work/minecraft/net/minecraft/util/ServerRecipeBookHelper.java
@@ -55,6 +55,11 @@
                 this.field_194335_f = ((ContainerPlayer)container).field_75179_f;
                 this.field_194336_g = ((ContainerPlayer)container).field_75181_e;
             }
+            else if (container instanceof net.minecraftforge.common.crafting.IRecipeContainer)
+            {
+                this.field_194335_f = ((net.minecraftforge.common.crafting.IRecipeContainer)container).getCraftResult();
+                this.field_194336_g = ((net.minecraftforge.common.crafting.IRecipeContainer)container).getCraftMatrix();
+            }
 
             if (this.field_194335_f != null && this.field_194336_g != null)
             {
@@ -199,11 +204,11 @@
         int i = this.field_194336_g.func_174922_i();
         int j = this.field_194336_g.func_174923_h();
 
-        if (this.field_194333_d instanceof ShapedRecipes)
+        if (this.field_194333_d instanceof net.minecraftforge.common.crafting.IShapedRecipe)
         {
-            ShapedRecipes shapedrecipes = (ShapedRecipes)this.field_194333_d;
-            i = shapedrecipes.func_192403_f();
-            j = shapedrecipes.func_192404_g();
+            net.minecraftforge.common.crafting.IShapedRecipe shapedrecipes = (net.minecraftforge.common.crafting.IShapedRecipe)this.field_194333_d;
+            i = shapedrecipes.getRecipeWidth();
+            j = shapedrecipes.getRecipeHeight();
         }
 
         int j1 = 1;
