--- ../src-base/minecraft/net/minecraft/util/Session.java
+++ ../src-work/minecraft/net/minecraft/util/Session.java
@@ -16,9 +16,24 @@
     private final String token;
     private final Session.Type sessionType;
     private static final String __OBFID = "CL_00000659";
+    /** Forge: Cache of the local session's GameProfile properties. */
+    private com.mojang.authlib.properties.PropertyMap properties;
 
     public Session(String usernameIn, String playerIDIn, String tokenIn, String sessionTypeIn)
     {
+        if (usernameIn == null || usernameIn.isEmpty())
+        {
+            usernameIn = "MissingName";
+            playerIDIn = tokenIn = "NotValid";
+            System.out.println("=========================================================");
+            System.out.println("Warning the username was not set for this session, typically");
+            System.out.println("this means you installed Forge incorrectly. We have set your");
+            System.out.println("name to \"MissingName\" and your session to nothing. Please");
+            System.out.println("check your instllation and post a console log from the launcher");
+            System.out.println("when asking for help!");
+            System.out.println("=========================================================");
+        }
+
         this.username = usernameIn;
         this.playerID = playerIDIn;
         this.token = tokenIn;
@@ -63,6 +78,19 @@
         return this.sessionType;
     }
 
+    /* ======================================== FORGE START ===================================== */
+    //For internal use only. Modders should never need to use this.
+    public void setProperties(com.mojang.authlib.properties.PropertyMap properties)
+    {
+        if(this.properties == null) this.properties = properties;
+    }
+
+    public boolean hasCachedProperties()
+    {
+        return properties != null;
+    }
+    /* ========================================= FORGE END ====================================== */
+
     @SideOnly(Side.CLIENT)
     public static enum Type
     {
