--- ../src-base/minecraft/net/minecraft/util/StringTranslate.java
+++ ../src-work/minecraft/net/minecraft/util/StringTranslate.java
@@ -23,9 +23,29 @@
 
     public StringTranslate()
     {
+        InputStream inputstream = StringTranslate.class.getResourceAsStream("/assets/minecraft/lang/en_US.lang");
+        inject(this, inputstream);
+    }
+
+    public static void inject(InputStream inputstream)
+    {
+        inject(field_74817_a, inputstream);
+    }
+
+    private static void inject(StringTranslate inst, InputStream inputstream)
+    {
+        java.util.HashMap<String, String> map = parseLangFile(inputstream);
+        inst.field_74816_c.putAll(map);
+        inst.field_150511_e = System.currentTimeMillis();
+    }
+
+    public static java.util.HashMap<String,String> parseLangFile(InputStream inputstream)
+    {
+        java.util.HashMap<String,String> table = Maps.newHashMap();
         try
         {
-            InputStream inputstream = StringTranslate.class.getResourceAsStream("/assets/minecraft/lang/en_US.lang");
+            inputstream = net.minecraftforge.fml.common.FMLCommonHandler.instance().loadLanguage(table, inputstream);
+            if (inputstream == null) return table;
 
             for (String s : IOUtils.readLines(inputstream, Charsets.UTF_8))
             {
@@ -37,17 +57,17 @@
                     {
                         String s1 = astring[0];
                         String s2 = field_111053_a.matcher(astring[1]).replaceAll("%$1s");
-                        this.field_74816_c.put(s1, s2);
+                        table.put(s1, s2);
                     }
                 }
             }
 
-            this.field_150511_e = System.currentTimeMillis();
         }
-        catch (IOException var7)
+        catch (Exception var7)
         {
             ;
         }
+        return table;
     }
 
     static StringTranslate func_74808_a()
