--- a/net/minecraft/util/Util.java
+++ b/net/minecraft/util/Util.java
@@ -313,7 +_,7 @@
                 .getSchema(DataFixUtils.makeKey(SharedConstants.getCurrentVersion().dataVersion().version()))
                 .getChoiceType(reference, name);
         } catch (IllegalArgumentException e) {
-            LOGGER.error("No data fixer registered for {}", name);
+            LOGGER.debug("No data fixer registered for {}", name);
             if (SharedConstants.IS_RUNNING_IN_IDE) {
                 throw e;
             }
