--- ../src-base/minecraft/net/minecraft/util/Vec3.java
+++ ../src-work/minecraft/net/minecraft/util/Vec3.java
@@ -1,8 +1,5 @@
 package net.minecraft.util;
 
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
-
 public class Vec3
 {
     public final double field_72450_a;
@@ -36,7 +33,6 @@
         this((double)p_i46377_1_.func_177958_n(), (double)p_i46377_1_.func_177956_o(), (double)p_i46377_1_.func_177952_p());
     }
 
-    @SideOnly(Side.CLIENT)
     public Vec3 func_72444_a(Vec3 p_72444_1_)
     {
         return new Vec3(p_72444_1_.field_72450_a - this.field_72450_a, p_72444_1_.field_72448_b - this.field_72448_b, p_72444_1_.field_72449_c - this.field_72449_c);
@@ -53,7 +49,6 @@
         return this.field_72450_a * p_72430_1_.field_72450_a + this.field_72448_b * p_72430_1_.field_72448_b + this.field_72449_c * p_72430_1_.field_72449_c;
     }
 
-    @SideOnly(Side.CLIENT)
     public Vec3 func_72431_c(Vec3 p_72431_1_)
     {
         return new Vec3(this.field_72448_b * p_72431_1_.field_72449_c - this.field_72449_c * p_72431_1_.field_72448_b, this.field_72449_c * p_72431_1_.field_72450_a - this.field_72450_a * p_72431_1_.field_72449_c, this.field_72450_a * p_72431_1_.field_72448_b - this.field_72448_b * p_72431_1_.field_72450_a);
@@ -79,7 +74,6 @@
         return new Vec3(this.field_72450_a + p_72441_1_, this.field_72448_b + p_72441_3_, this.field_72449_c + p_72441_5_);
     }
 
-    @SideOnly(Side.CLIENT)
     public double func_72438_d(Vec3 p_72438_1_)
     {
         double d0 = p_72438_1_.field_72450_a - this.field_72450_a;
