--- ../src_base/minecraft/net/minecraft/util/WeightedRandomChestContent.java
+++ ../src_work/minecraft/net/minecraft/util/WeightedRandomChestContent.java
@@ -5,6 +5,7 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.tileentity.TileEntityChest;
 import net.minecraft.tileentity.TileEntityDispenser;
+import net.minecraftforge.common.ChestGenHooks;
 
 public class WeightedRandomChestContent extends WeightedRandomItem
 {
@@ -19,6 +20,8 @@
 
     /** The maximum chance of item generating. */
     private int theMaximumChanceToGenerateItem;
+    
+    public final ItemStack itemStack;
 
     public WeightedRandomChestContent(int par1, int par2, int par3, int par4, int par5)
     {
@@ -27,8 +30,16 @@
         this.theItemDamage = par2;
         this.theMinimumChanceToGenerateItem = par3;
         this.theMaximumChanceToGenerateItem = par4;
+        itemStack = new ItemStack(par1, 1, par2);
     }
-
+    
+    public WeightedRandomChestContent(ItemStack stack, int min, int max, int weight)
+    {
+        super(weight);
+        itemStack = stack;
+        theMinimumChanceToGenerateItem = min;
+        theMaximumChanceToGenerateItem = max;
+    }
     /**
      * Generates the Chest contents.
      */
@@ -37,18 +48,11 @@
         for (int var4 = 0; var4 < par3; ++var4)
         {
             WeightedRandomChestContent var5 = (WeightedRandomChestContent)WeightedRandom.getRandomItem(par0Random, par1ArrayOfWeightedRandomChestContent);
-            int var6 = var5.theMinimumChanceToGenerateItem + par0Random.nextInt(var5.theMaximumChanceToGenerateItem - var5.theMinimumChanceToGenerateItem + 1);
+            ItemStack[] stacks = ChestGenHooks.generateStacks(par0Random, var5.itemStack, var5.theMinimumChanceToGenerateItem, var5.theMaximumChanceToGenerateItem);
 
-            if (Item.itemsList[var5.theItemId].getItemStackLimit() >= var6)
+            for (ItemStack item : stacks)
             {
-                par2TileEntityChest.setInventorySlotContents(par0Random.nextInt(par2TileEntityChest.getSizeInventory()), new ItemStack(var5.theItemId, var6, var5.theItemDamage));
-            }
-            else
-            {
-                for (int var7 = 0; var7 < var6; ++var7)
-                {
-                    par2TileEntityChest.setInventorySlotContents(par0Random.nextInt(par2TileEntityChest.getSizeInventory()), new ItemStack(var5.theItemId, 1, var5.theItemDamage));
-                }
+                par2TileEntityChest.setInventorySlotContents(par0Random.nextInt(par2TileEntityChest.getSizeInventory()), item);
             }
         }
     }
@@ -61,18 +65,11 @@
         for (int var4 = 0; var4 < par3; ++var4)
         {
             WeightedRandomChestContent var5 = (WeightedRandomChestContent)WeightedRandom.getRandomItem(par0Random, par1ArrayOfWeightedRandomChestContent);
-            int var6 = var5.theMinimumChanceToGenerateItem + par0Random.nextInt(var5.theMaximumChanceToGenerateItem - var5.theMinimumChanceToGenerateItem + 1);
+            ItemStack[] stacks = ChestGenHooks.generateStacks(par0Random, var5.itemStack, var5.theMinimumChanceToGenerateItem, var5.theMaximumChanceToGenerateItem);
 
-            if (Item.itemsList[var5.theItemId].getItemStackLimit() >= var6)
+            for (ItemStack item : stacks)
             {
-                par2TileEntityDispenser.setInventorySlotContents(par0Random.nextInt(par2TileEntityDispenser.getSizeInventory()), new ItemStack(var5.theItemId, var6, var5.theItemDamage));
-            }
-            else
-            {
-                for (int var7 = 0; var7 < var6; ++var7)
-                {
-                    par2TileEntityDispenser.setInventorySlotContents(par0Random.nextInt(par2TileEntityDispenser.getSizeInventory()), new ItemStack(var5.theItemId, 1, var5.theItemDamage));
-                }
+                par2TileEntityDispenser.setInventorySlotContents(par0Random.nextInt(par2TileEntityDispenser.getSizeInventory()), item);
             }
         }
     }
