--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -81,9 +_,17 @@
     private volatile @Nullable DisconnectionDetails delayedDisconnect;
     private @Nullable BandwidthDebugMonitor bandwidthDebugMonitor;
     private @Nullable UUID intendedProfileId;
+    private java.util.function.Consumer<Connection> activationHandler;
+    private final net.minecraftforge.common.util.PacketLogger packetLogger = new net.minecraftforge.common.util.PacketLogger(this);
+    private ProtocolInfo<?> outboundProtocol = null;
+    private ProtocolInfo<?> inboundProtocol = null;
 
     public Connection(final PacketFlow receiving) {
         this.receiving = receiving;
+        if (this.receiving == PacketFlow.SERVERBOUND)
+            this.inboundProtocol = INITIAL_PROTOCOL;
+        else
+            this.outboundProtocol = INITIAL_PROTOCOL;
     }
 
     @Override
@@ -91,6 +_,7 @@
         super.channelActive(ctx);
         this.channel = ctx.channel();
         this.address = this.channel.remoteAddress();
+        if (activationHandler != null) activationHandler.accept(this);
         if (this.delayedDisconnect != null) {
             this.disconnect(this.delayedDisconnect);
         }
@@ -152,6 +_,7 @@
 
             if (packetListener.shouldHandleMessage(packet)) {
                 try {
+                    packetLogger.recv(packet);
                     genericsFtw(packet, packetListener);
                 } catch (RunningOnDifferentThreadException var5) {
                 } catch (RejectedExecutionException ignored) {
@@ -201,6 +_,7 @@
             throw new IllegalStateException("Invalid inbound protocol: " + protocol.id());
         }
 
+        this.inboundProtocol = protocol;
         this.packetListener = packetListener;
         this.disconnectListener = null;
         UnconfiguredPipelineHandler.InboundConfigurationTask configMessage = UnconfiguredPipelineHandler.setupInboundProtocol(protocol);
@@ -218,7 +_,7 @@
             throw new IllegalStateException("Invalid outbound protocol: " + protocol.id());
         }
 
-        UnconfiguredPipelineHandler.OutboundConfigurationTask configMessage = UnconfiguredPipelineHandler.setupOutboundProtocol(protocol);
+        UnconfiguredPipelineHandler.OutboundConfigurationTask configMessage = UnconfiguredPipelineHandler.setupOutboundProtocol(protocol).andThen(_ -> this.outboundProtocol = protocol);
         BundlerInfo bundlerInfo = protocol.bundlerInfo();
         if (bundlerInfo != null) {
             PacketBundleUnpacker newUnbundler = new PacketBundleUnpacker(bundlerInfo);
@@ -265,7 +_,8 @@
         this.disconnectListener = listener;
         this.runOnceConnected(connection -> {
             this.setupInboundProtocol(inbound, listener);
-            connection.sendPacket(new ClientIntentionPacket(SharedConstants.getCurrentVersion().protocolVersion(), hostName, port, intent), null, true);
+            // TODO: Change this to be a immediately sent login custom payload packet?
+            connection.sendPacket(new ClientIntentionPacket(SharedConstants.getCurrentVersion().protocolVersion(), net.minecraftforge.network.NetworkContext.enhanceHostName(hostName), port, intent), null, true);
             this.setupOutboundProtocol(outbound);
         });
     }
@@ -309,10 +_,13 @@
         if (listener != null) {
             ChannelFuture future = flush ? this.channel.writeAndFlush(packet) : this.channel.write(packet);
             future.addListener(listener);
+            future.addListener(f -> packetLogger.send(packet));
         } else if (flush) {
             this.channel.writeAndFlush(packet, this.channel.voidPromise());
+            packetLogger.send(packet);
         } else {
             this.channel.write(packet, this.channel.voidPromise());
+            packetLogger.send(packet);
         }
     }
 
@@ -381,7 +_,7 @@
         if (this.address == null) {
             return "local";
         } else {
-            return logIPs ? this.address.toString() : "IP hidden";
+            return logIPs ? net.minecraftforge.network.DualStackUtils.getAddressString(this.address) : "IP hidden";
         }
     }
 
@@ -426,7 +_,9 @@
     }
 
     public static ChannelFuture connect(final InetSocketAddress address, final EventLoopGroupHolder eventLoopGroupHolder, final Connection connection) {
-        return new Bootstrap().group(eventLoopGroupHolder.eventLoopGroup()).handler(new ChannelInitializer<Channel>() {
+        net.minecraftforge.network.DualStackUtils.checkIPv6(address.getAddress());
+        connection.activationHandler = net.minecraftforge.network.NetworkRegistry::onConnectionStart;
+        return new Bootstrap().group(eventLoopGroupHolder.eventLoopGroup(true)).handler(new ChannelInitializer<Channel>() {
             @Override
             protected void initChannel(final Channel channel) {
                 try {
@@ -491,7 +_,8 @@
 
     public static Connection connectToLocalServer(final SocketAddress address) {
         final Connection connection = new Connection(PacketFlow.CLIENTBOUND);
-        new Bootstrap().group(EventLoopGroupHolder.local().eventLoopGroup()).handler(new ChannelInitializer<Channel>() {
+        connection.activationHandler = net.minecraftforge.network.NetworkRegistry::onConnectionStart;
+        new Bootstrap().group(EventLoopGroupHolder.local().eventLoopGroup(true)).handler(new ChannelInitializer<Channel>() {
             @Override
             protected void initChannel(final Channel channel) {
                 ChannelPipeline pipeline = channel.pipeline();
@@ -590,6 +_,22 @@
 
     public float getAverageSentPackets() {
         return this.averageSentPackets;
+    }
+
+    public Channel channel() {
+        return this.channel;
+    }
+
+    public ConnectionProtocol getProtocol() {
+        return outboundProtocol != null ? outboundProtocol.id() : inboundProtocol.id();
+    }
+
+    public ProtocolInfo<?> getInboundProtocolInfo() {
+        return this.inboundProtocol;
+    }
+
+    public ProtocolInfo<?> getOutputboundProtocolInfo() {
+        return this.outboundProtocol;
     }
 
     public void setBandwidthLogger(final LocalSampleLogger bandwidthLogger) {
