--- a/net/minecraft/network/FriendlyByteBuf.java
+++ b/net/minecraft/network/FriendlyByteBuf.java
@@ -78,7 +_,7 @@
 import org.joml.Quaternionf;
 import org.joml.Vector3f;
 
-public class FriendlyByteBuf extends ByteBuf {
+public class FriendlyByteBuf extends ByteBuf implements net.minecraftforge.common.extensions.IForgeFriendlyByteBuf {
    private static final int f_178335_ = 5;
    private static final int f_178336_ = 10;
    public static final int f_178337_ = 2097152;
@@ -622,6 +_,15 @@
    }
 
    public FriendlyByteBuf m_130055_(ItemStack p_130056_) {
+      return writeItemStack(p_130056_, true);
+   }
+
+   /**
+    * Most ItemStack serialization is Server to Client,and doesn't need to know the FULL tag details.
+    * One exception is items from the creative menu, which must be sent from Client to Server with their full NBT.
+    * If you want to send the FULL tag set limitedTag to false
+    */
+   public FriendlyByteBuf writeItemStack(ItemStack p_130056_, boolean limitedTag) {
       if (p_130056_.m_41619_()) {
          this.writeBoolean(false);
       } else {
@@ -630,8 +_,8 @@
          this.m_236818_(BuiltInRegistries.f_257033_, item);
          this.writeByte(p_130056_.m_41613_());
          CompoundTag compoundtag = null;
-         if (item.m_41465_() || item.m_41468_()) {
-            compoundtag = p_130056_.m_41783_();
+         if (item.isDamageable(p_130056_) || item.m_41468_()) {
+            compoundtag = limitedTag ? p_130056_.getShareTag() : p_130056_.m_41783_();
          }
 
          this.m_130079_(compoundtag);
@@ -647,7 +_,7 @@
          Item item = this.m_236816_(BuiltInRegistries.f_257033_);
          int i = this.readByte();
          ItemStack itemstack = new ItemStack(item, i);
-         itemstack.m_41751_(this.m_130260_());
+         itemstack.readShareTag(this.m_130260_());
          return itemstack;
       }
    }
