--- a/net/minecraft/network/NetHandlerPlayServer.java
+++ b/net/minecraft/network/NetHandlerPlayServer.java
@@ -1265,5 +1265,7 @@
    }
 
    public void processCustomPayload(CPacketCustomPayload packetIn) {
+      PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.player.getServerWorld());
+      net.minecraftforge.fml.network.NetworkHooks.onCustomPayload(packetIn, this.netManager);
    }
 }
