--- ../src-base/minecraft/net/minecraft/network/NetLoginHandler.java
+++ ../src-work/minecraft/net/minecraft/network/NetLoginHandler.java
@@ -11,11 +11,15 @@
 import java.util.List;
 import java.util.Random;
 import javax.crypto.SecretKey;
+
+import cpw.mods.fml.common.network.FMLNetworkHandler;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.network.packet.NetHandler;
 import net.minecraft.network.packet.Packet;
 import net.minecraft.network.packet.Packet1Login;
 import net.minecraft.network.packet.Packet205ClientCommand;
+import net.minecraft.network.packet.Packet250CustomPayload;
 import net.minecraft.network.packet.Packet252SharedKey;
 import net.minecraft.network.packet.Packet253ServerAuthData;
 import net.minecraft.network.packet.Packet254ServerPing;
@@ -54,7 +58,7 @@
             this.func_72529_d();
         }
 
-        if (this.field_72535_g++ == 600)
+        if (this.field_72535_g++ == 6000)
         {
             this.func_72527_a("Took too long to log in");
         }
@@ -155,11 +159,18 @@
         }
     }
 
-    public void func_72455_a(Packet1Login p_72455_1_) {}
+    public void func_72455_a(Packet1Login p_72455_1_)
+    {
+        FMLNetworkHandler.handleLoginPacketOnServer(this, p_72455_1_);
+    }
 
     public void func_72529_d()
     {
-        String s = this.field_72534_f.func_71203_ab().func_72399_a(this.field_72538_b.func_74430_c(), this.field_72543_h);
+        FMLNetworkHandler.onConnectionReceivedFromClient(this, this.field_72534_f, this.field_72538_b.func_74430_c(), this.field_72543_h);
+    }
+
+    public void completeConnection(String s)
+    {
 
         if (s != null)
         {
@@ -282,4 +293,22 @@
     {
         return p_72531_0_.field_72544_i = p_72531_1_;
     }
+    
+
+    public void func_72501_a(Packet250CustomPayload p_72501_1_)
+    {
+        FMLNetworkHandler.handlePacket250Packet(p_72501_1_, field_72538_b, this);
+    }
+
+    @Override
+    public void handleVanilla250Packet(Packet250CustomPayload payload)
+    {
+        // NOOP for login
+    }
+
+    @Override
+    public EntityPlayer getPlayer()
+    {
+        return null;
+    }
 }
