--- ../src-base/minecraft/net/minecraft/network/PacketBuffer.java
+++ ../src-work/minecraft/net/minecraft/network/PacketBuffer.java
@@ -336,7 +336,7 @@
 
             if (p_150788_1_.func_77973_b().func_77645_m() || p_150788_1_.func_77973_b().func_77651_p())
             {
-                nbttagcompound = p_150788_1_.func_77978_p();
+                nbttagcompound = p_150788_1_.func_77973_b().getNBTShareTag(p_150788_1_);
             }
 
             this.func_150786_a(nbttagcompound);
@@ -358,7 +358,7 @@
             int j = this.readByte();
             int k = this.readShort();
             ItemStack itemstack = new ItemStack(Item.func_150899_d(i), j, k);
-            itemstack.func_77982_d(this.func_150793_b());
+            itemstack.func_77973_b().readNBTShareTag(itemstack, this.func_150793_b());
             return itemstack;
         }
     }
