--- a/net/minecraft/network/PacketEncoder.java
+++ b/net/minecraft/network/PacketEncoder.java
@@ -48,7 +_,7 @@
                   JvmProfiler.f_185340_.m_183508_(k, integer, p_130545_.channel().remoteAddress(), j);
                }
             } catch (Throwable throwable) {
-               f_130538_.error(throwable);
+               f_130538_.error("Error encoding packet", throwable); // Forge: Get Minecraft to spit out more information about the Throwable we got.
                if (p_130546_.m_6588_()) {
                   throw new SkipPacketException(throwable);
                } else {
