--- ../src_base/minecraft/net/minecraft/network/TcpConnection.java
+++ ../src_work/minecraft/net/minecraft/network/TcpConnection.java
@@ -16,8 +16,6 @@
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
-import java.util.Queue;
-import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.atomic.AtomicInteger;
 import javax.crypto.SecretKey;
 import net.minecraft.logging.ILogAgent;
@@ -27,6 +25,8 @@
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.CryptManager;
 
+import java.io.IOException; // CraftBukkit
+
 public class TcpConnection implements INetworkManager
 {
     public static AtomicInteger field_74471_a = new AtomicInteger();
@@ -39,10 +39,10 @@
     private final ILogAgent tcpConLogAgent;
 
     /** The socket used by this network manager. */
-    private Socket networkSocket;
+    public Socket networkSocket; // CraftBukkit - private -> public
 
     /** The InetSocketAddress of the remote endpoint */
-    private final SocketAddress remoteSocketAddress;
+    private SocketAddress remoteSocketAddress; // Spigot - remove final
 
     /** The input stream connected to the socket. */
     private volatile DataInputStream socketInputStream;
@@ -61,13 +61,13 @@
     /**
      * Linked list of packets that have been read and are awaiting processing.
      */
-    private Queue readPackets;
+    private java.util.Queue readPackets; // CraftBukkit - Concurrent linked queue
 
     /** Linked list of packets awaiting sending. */
-    private List dataPackets;
+    private List dataPackets = Collections.synchronizedList(new ArrayList());
 
     /** Linked list of packets with chunk data that are awaiting sending. */
-    private List chunkDataPackets;
+    private List chunkDataPackets = Collections.synchronizedList(new ArrayList());
 
     /** A reference to the NetHandler object. */
     private NetHandler theNetHandler;
@@ -117,7 +117,7 @@
     {
         this.sendQueueLock = new Object();
         this.isRunning = true;
-        this.readPackets = new ConcurrentLinkedQueue();
+        this.readPackets = new java.util.concurrent.ConcurrentLinkedQueue();
         this.dataPackets = Collections.synchronizedList(new ArrayList());
         this.chunkDataPackets = Collections.synchronizedList(new ArrayList());
         this.terminationReason = "";
@@ -218,7 +218,8 @@
                 }
             }
 
-            if (this.chunkDataPacketsDelay-- <= 0 && (this.field_74468_e == 0 || !this.chunkDataPackets.isEmpty() && MinecraftServer.getSystemTimeMillis() - ((Packet)this.chunkDataPackets.get(0)).creationTimeMillis >= (long)this.field_74468_e))
+            // CraftBukkit - don't allow low priority packet to be sent unless it was placed in the queue before the first packet on the high priority queue TODO: is this still right?
+            if ((flag || this.chunkDataPacketsDelay-- <= 0) && !this.chunkDataPackets.isEmpty() && (this.dataPackets.isEmpty() || ((Packet) this.dataPackets.get(0)).creationTimeMillis > ((Packet) this.chunkDataPackets.get(0)).creationTimeMillis))
             {
                 packet = this.func_74460_a(true);
 
@@ -378,7 +379,7 @@
      */
     private void onNetworkError(Exception par1Exception)
     {
-        par1Exception.printStackTrace();
+        par1Exception.printStackTrace(); // CraftBukkit - Remove console spam
         this.networkShutdown("disconnect.genericReason", new Object[] {"Internal exception: " + par1Exception.toString()});
     }
 
@@ -453,14 +454,18 @@
 
         int i = 1000;
 
-        while (i-- >= 0)
-        {
-            Packet packet = (Packet)this.readPackets.poll();
-
-            if (packet != null && !this.theNetHandler.isConnectionClosed())
-            {
-                packet.processPacket(this.theNetHandler);
-            }
+        while (!this.readPackets.isEmpty() && i-- >= 0)
+        {
+            Packet packet = (Packet) this.readPackets.poll(); // CraftBukkit - remove -> poll
+
+            // CraftBukkit start
+            if (this.theNetHandler instanceof NetLoginHandler ? ((NetLoginHandler) this.theNetHandler).connectionComplete : ((NetServerHandler) this.theNetHandler).connectionClosed)
+            {
+                continue;
+            }
+
+            // CraftBukkit end
+            packet.processPacket(this.theNetHandler);
         }
 
         this.wakeThreads();
@@ -593,4 +598,9 @@
     {
         return par0TcpConnection.writeThread;
     }
+
+    public void setSocketAddress(SocketAddress address)
+    {
+        remoteSocketAddress = address;    // Spigot
+    }
 }
