--- a/net/minecraft/world/level/entity/TransientEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/TransientEntitySectionManager.java
@@ -82,11 +_,13 @@
 
     private class Callback implements EntityInLevelCallback {
         private final T entity;
+        private final Entity realEntity;
         private long currentSectionKey;
         private EntitySection<T> currentSection;
 
         private Callback(final T entity, final long currentSectionKey, final EntitySection<T> currentSection) {
             this.entity = entity;
+            this.realEntity = entity instanceof Entity ? (Entity)entity : null;
             this.currentSectionKey = currentSectionKey;
             this.currentSection = currentSection;
         }
@@ -105,6 +_,7 @@
                 TransientEntitySectionManager.this.removeSectionIfEmpty(this.currentSectionKey, this.currentSection);
                 EntitySection<T> newSection = TransientEntitySectionManager.this.sectionStorage.getOrCreateSection(newSectionPos);
                 newSection.add(this.entity);
+                long oldSectionKey = currentSectionKey;
                 this.currentSection = newSection;
                 this.currentSectionKey = newSectionPos;
                 TransientEntitySectionManager.this.callbacks.onSectionChange(this.entity);
@@ -116,6 +_,9 @@
                     } else if (!wasTicking && isTicking) {
                         TransientEntitySectionManager.this.callbacks.onTickingStart(this.entity);
                     }
+                }
+                if (this.realEntity != null) {
+                    net.minecraftforge.event.ForgeEventFactory.onEntityEnterSection(this.realEntity, oldSectionKey, newSectionPos);
                 }
             }
         }
