--- a/net/minecraft/world/level/entity/TransientEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/TransientEntitySectionManager.java
@@ -92,11 +_,13 @@
 
    class Callback implements EntityInLevelCallback {
       private final T entity;
+      private final Entity realEntity;
       private long currentSectionKey;
       private EntitySection<T> currentSection;
 
       Callback(T p_157673_, long p_157674_, EntitySection<T> p_157675_) {
          this.entity = p_157673_;
+         this.realEntity = p_157673_ instanceof Entity ? (Entity) p_157673_ : null;
          this.currentSectionKey = p_157674_;
          this.currentSection = p_157675_;
       }
@@ -113,6 +_,7 @@
             TransientEntitySectionManager.this.removeSectionIfEmpty(this.currentSectionKey, this.currentSection);
             EntitySection<T> entitysection = TransientEntitySectionManager.this.sectionStorage.getOrCreateSection(i);
             entitysection.add(this.entity);
+            long oldSectionKey = currentSectionKey;
             this.currentSection = entitysection;
             this.currentSectionKey = i;
             TransientEntitySectionManager.this.callbacks.onSectionChange(this.entity);
@@ -125,6 +_,7 @@
                   TransientEntitySectionManager.this.callbacks.onTickingStart(this.entity);
                }
             }
+            if (this.realEntity != null) net.minecraftforge.common.ForgeHooks.onEntityEnterSection(this.realEntity, oldSectionKey, i);
          }
 
       }
