--- a/net/minecraft/world/level/entity/TransientEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/TransientEntitySectionManager.java
@@ -105,6 +_,7 @@
          BlockPos blockpos = this.f_157668_.m_142538_();
          long i = SectionPos.m_175568_(blockpos);
          if (i != this.f_157669_) {
+            net.minecraftforge.common.ForgeHooks.onEntityEnterSection(f_157668_, this.f_157669_, i);
             Visibility visibility = this.f_157670_.m_156848_();
             if (!this.f_157670_.m_156846_(this.f_157668_)) {
                TransientEntitySectionManager.f_157635_.warn("Entity {} wasn't found in section {} (moving to {})", this.f_157668_, SectionPos.m_123184_(this.f_157669_), i);
