--- a/net/minecraft/client/renderer/block/model/multipart/Multipart.java
+++ b/net/minecraft/client/renderer/block/model/multipart/Multipart.java
@@ -81,11 +81,12 @@
    }
 
    @Nullable
-   public IBakedModel func_209558_a(Function<ResourceLocation, IUnbakedModel> p_209558_1_, Function<ResourceLocation, TextureAtlasSprite> p_209558_2_, ModelRotation p_209558_3_, boolean p_209558_4_) {
+   @Override
+   public IBakedModel bake(Function<ResourceLocation, IUnbakedModel> p_209558_1_, Function<ResourceLocation, TextureAtlasSprite> p_209558_2_, net.minecraftforge.common.model.IModelState p_209558_3_, boolean p_209558_4_, net.minecraft.client.renderer.vertex.VertexFormat format) {
       MultipartBakedModel.Builder multipartbakedmodel$builder = new MultipartBakedModel.Builder();
 
       for(Selector selector : this.getSelectors()) {
-         IBakedModel ibakedmodel = selector.getVariantList().func_209558_a(p_209558_1_, p_209558_2_, p_209558_3_, p_209558_4_);
+         IBakedModel ibakedmodel = selector.getVariantList().bake(p_209558_1_, p_209558_2_, p_209558_3_, p_209558_4_, format);
          if (ibakedmodel != null) {
             multipartbakedmodel$builder.putModel(selector.getPredicate(this.stateContainer), ibakedmodel);
          }
