--- ../src_base/minecraft/net/minecraft/world/chunk/storage/AnvilChunkLoader.java
+++ ../src_work/minecraft/net/minecraft/world/chunk/storage/AnvilChunkLoader.java
@@ -27,6 +27,8 @@
 import net.minecraft.world.chunk.NibbleArray;
 import net.minecraft.world.storage.IThreadedFileIO;
 import net.minecraft.world.storage.ThreadedFileIOBase;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.world.ChunkDataEvent;
 
 public class AnvilChunkLoader implements IChunkLoader, IThreadedFileIO
 {
@@ -108,6 +110,7 @@
                 var5 = this.readChunkFromNBT(par1World, par4NBTTagCompound.getCompoundTag("Level"));
             }
 
+            MinecraftForge.EVENT_BUS.post(new ChunkDataEvent.Load(var5, par4NBTTagCompound));
             return var5;
         }
     }
@@ -123,6 +126,7 @@
             var3.setTag("Level", var4);
             this.writeChunkToNBT(par2Chunk, par1World, var4);
             this.func_75824_a(par2Chunk.getChunkCoordIntPair(), var3);
+            MinecraftForge.EVENT_BUS.post(new ChunkDataEvent.Save(par2Chunk, var3));
         }
         catch (Exception var5)
         {
