--- a/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
+++ b/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
@@ -19,12 +19,8 @@
 
    public IChunkLoader func_75763_a(Dimension p_75763_1_) {
       File file1 = this.func_75765_b();
-      if (p_75763_1_ instanceof NetherDimension) {
-         File file3 = new File(file1, "DIM-1");
-         file3.mkdirs();
-         return new AnvilChunkLoader(file3, this.field_186341_a);
-      } else if (p_75763_1_ instanceof EndDimension) {
-         File file2 = new File(file1, "DIM1");
+      if (p_75763_1_.getSaveFolder() != null) {
+         File file2 = new File(file1, p_75763_1_.getSaveFolder());
          file2.mkdirs();
          return new AnvilChunkLoader(file2, this.field_186341_a);
       } else {
