--- ../src_base/minecraft/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
+++ ../src_work/minecraft/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
@@ -24,15 +24,9 @@
         File file1 = this.getWorldDirectory();
         File file2;
 
-        if (par1WorldProvider instanceof WorldProviderHell)
+        if (par1WorldProvider.getSaveFolder() != null)
         {
-            file2 = new File(file1, "DIM-1");
-            file2.mkdirs();
-            return new AnvilChunkLoader(file2);
-        }
-        else if (par1WorldProvider instanceof WorldProviderEnd)
-        {
-            file2 = new File(file1, "DIM1");
+            file2 = new File(file1, par1WorldProvider.getSaveFolder());
             file2.mkdirs();
             return new AnvilChunkLoader(file2);
         }
