--- ../src-base/minecraft/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/AnvilSaveHandler.java
@@ -23,18 +23,12 @@
         File file1 = this.func_75765_b();
         File file2;
 
-        if (p_75763_1_ instanceof WorldProviderHell)
+        if (p_75763_1_.getSaveFolder() != null)
         {
-            file2 = new File(file1, "DIM-1");
+            file2 = new File(file1, p_75763_1_.getSaveFolder());
             file2.mkdirs();
             return new AnvilChunkLoader(file2);
         }
-        else if (p_75763_1_ instanceof WorldProviderEnd)
-        {
-            file2 = new File(file1, "DIM1");
-            file2.mkdirs();
-            return new AnvilChunkLoader(file2);
-        }
         else
         {
             return new AnvilChunkLoader(file1);
