--- ../src-base/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
@@ -58,7 +58,7 @@
             }
         }
 
-        this.field_177488_d.func_186013_a(p_177484_1_, p_177484_2_, p_177484_3_, p_177484_4_);
+        this.field_177488_d.func_186013_a(p_177484_1_, p_177484_2_, p_177484_3_, p_177484_4_.getClean());
     }
 
     public boolean func_76663_a()
