--- ../src-base/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
@@ -101,6 +101,39 @@
         }
     }
 
+    // This is a copy (sort of) of the method below it, make sure they stay in sync
+    public synchronized boolean chunkExists(int x, int z)
+    {
+        if (this.outOfBounds(x, z)) return false;
+
+        try
+        {
+            int offset = this.getOffset(x, z);
+
+            if (offset == 0) return false;
+
+            int sectorNumber = offset >> 8;
+            int numSectors = offset & 255;
+
+            if (sectorNumber + numSectors > this.sectorFree.size()) return false;
+
+            this.dataFile.seek((long)(sectorNumber * 4096));
+            int length = this.dataFile.readInt();
+
+            if (length > 4096 * numSectors || length <= 0) return false;
+
+            byte version = this.dataFile.readByte();
+
+            if (version == 1 || version == 2) return true;
+        }
+        catch (IOException ioexception)
+        {
+            return false;
+        }
+
+        return false;
+    }
+
     public synchronized DataInputStream getChunkDataInputStream(int x, int z)
     {
         if (this.outOfBounds(x, z))
