--- ../src_base/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
+++ ../src_work/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
@@ -103,6 +103,61 @@
             ioexception.printStackTrace();
         }
     }
+
+    // CraftBukkit start - This is a copy (sort of) of the method below it, make sure they stay in sync
+
+    public synchronized boolean chunkExists(int i, int j)
+    {
+        if (this.outOfBounds(i, j))
+        {
+            return false;
+        }
+        else
+        {
+            try
+            {
+                int k = this.getOffset(i, j);
+
+                if (k == 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    int l = k >> 8;
+                    int i1 = k & 255;
+
+                    if (l + i1 > this.sectorFree.size())
+                    {
+                        return false;
+                    }
+
+                    this.dataFile.seek((long)(l * 4096));
+                    int j1 = this.dataFile.readInt();
+
+                    if (j1 > 4096 * i1 || j1 <= 0)
+                    {
+                        return false;
+                    }
+
+                    byte b0 = this.dataFile.readByte();
+
+                    if (b0 == 1 || b0 == 2)
+                    {
+                        return true;
+                    }
+                }
+            }
+            catch (IOException ioexception)
+            {
+                return false;
+            }
+        }
+
+        return false;
+    }
+
+    // CraftBukkit end
 
     /**
      * args: x, y - get uncompressed chunk stream from the region file
@@ -276,7 +331,7 @@
                 }
             }
 
-            this.setChunkTimestamp(par1, par2, (int)(MinecraftServer.func_130071_aq() / 1000L));
+            this.setChunkTimestamp(par1, par2, (int)(MinecraftServer.getSystemTimeMillis() / 1000L));
         }
         catch (IOException ioexception)
         {
