--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketSpawnMob.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketSpawnMob.java
@@ -38,7 +38,7 @@
     {
         this.field_149042_a = p_i46973_1_.func_145782_y();
         this.field_186894_b = p_i46973_1_.func_110124_au();
-        this.field_149040_b = EntityList.field_191308_b.func_148757_b(p_i46973_1_.getClass());
+        this.field_149040_b = EntityList.getID(p_i46973_1_.getClass());
         this.field_149041_c = p_i46973_1_.field_70165_t;
         this.field_149038_d = p_i46973_1_.field_70163_u;
         this.field_149039_e = p_i46973_1_.field_70161_v;
