--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -872,6 +_,17 @@
         @Override
         protected void breed() {
             Fox offspring = (Fox)this.animal.getBreedOffspring(this.level, this.partner);
+            var event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(animal, partner, offspring);
+            var eventWasCancelled = net.minecraftforge.event.entity.living.BabyEntitySpawnEvent.BUS.post(event);
+            offspring = (Fox)event.getChild();
+            if (eventWasCancelled) {
+                //Reset the "inLove" state for the animals
+                this.animal.setAge(6000);
+                this.partner.setAge(6000);
+                this.animal.resetLove();
+                this.partner.resetLove();
+                return;
+            }
             if (offspring != null) {
                 ServerPlayer animalLoveCause = this.animal.getLoveCause();
                 ServerPlayer partnerLoveCause = this.partner.getLoveCause();
@@ -949,7 +_,7 @@
         }
 
         protected void onReachedTarget() {
-            if (getServerLevel(Fox.this.level()).getGameRules().get(GameRules.MOB_GRIEFING)) {
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(getServerLevel(Fox.this.level()), Fox.this)) {
                 BlockState state = Fox.this.level().getBlockState(this.blockPos);
                 if (state.is(Blocks.SWEET_BERRY_BUSH)) {
                     this.pickSweetBerries(state);
@@ -1008,7 +_,7 @@
 
         @Override
         public boolean canUse() {
-            return Fox.this.isInWater() && Fox.this.getFluidHeight(FluidTags.WATER) > 0.25 || Fox.this.isInLava();
+            return Fox.this.isInWater() && Fox.this.getFluidHeight(FluidTags.WATER) > 0.25 || Fox.this.isInLava() || Fox.this.isInFluidType((fluidType, height) -> Fox.this.canSwimInFluidType(fluidType) && height > 0.25D);
         }
     }
 
