--- a/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
@@ -29,18 +_,22 @@
         double spawnY = center.y() + direction.getStepY() * 1.125F;
         double spawnZ = center.z() + direction.getStepZ() * justOutsideDispenser;
         BlockPos frontPos = source.pos().relative(direction);
+        AbstractBoat boat = this.type.create(level, EntitySpawnReason.DISPENSER);
+        if (boat == null) {
+            return this.defaultDispenseItemBehavior.dispense(source, dispensed);
+        }
+        boat.setYRot(direction.toYRot());
         double yOffset;
-        if (level.getFluidState(frontPos).is(FluidTags.WATER)) {
+        if (boat.canBoatInFluid(level.getFluidState(frontPos))) {
             yOffset = 1.0;
         } else {
-            if (!level.getBlockState(frontPos).isAir() || !level.getFluidState(frontPos.below()).is(FluidTags.WATER)) {
+            if (!level.getBlockState(frontPos).isAir() || !boat.canBoatInFluid(level.getFluidState(frontPos.below()))) {
                 return this.defaultDispenseItemBehavior.dispense(source, dispensed);
             }
 
             yOffset = 0.0;
         }
 
-        AbstractBoat boat = this.type.create(level, EntitySpawnReason.DISPENSER);
         if (boat != null) {
             boat.setInitialPos(spawnX, spawnY + yOffset, spawnZ);
             EntityType.<AbstractBoat>createDefaultStackConfig(level, dispensed, null).apply(boat);
