--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -141,7 +_,7 @@
                 DispensibleContainerItem bucket = (DispensibleContainerItem)dispensed.getItem();
                 BlockPos target = source.pos().relative(source.state().getValue(DispenserBlock.FACING));
                 Level level = source.level();
-                if (bucket.emptyContents(null, level, target, null)) {
+                if (bucket.emptyContents(null, level, target, null, dispensed)) {
                     bucket.checkExtraContent(null, level, dispensed, target);
                     return this.consumeWithRemainder(source, dispensed, new ItemStack(Items.BUCKET));
                 } else {
