--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartHopper.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartHopper.java
@@ -56,6 +56,7 @@
 
     public boolean func_184230_a(EntityPlayer p_184230_1_, ItemStack p_184230_2_, EnumHand p_184230_3_)
     {
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.minecart.MinecartInteractEvent(this, playerIn))) return true;
         if (!this.worldObj.isRemote)
         {
             p_184230_1_.displayGUIChest(this);
