--- a/net/minecraft/enchantment/Enchantments.java
+++ b/net/minecraft/enchantment/Enchantments.java
@@ -3,6 +3,7 @@
 import net.minecraft.inventory.EquipmentSlotType;
 import net.minecraft.util.registry.Registry;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Enchantments {
    private static final EquipmentSlotType[] field_222195_L = new EquipmentSlotType[]{EquipmentSlotType.HEAD, EquipmentSlotType.CHEST, EquipmentSlotType.LEGS, EquipmentSlotType.FEET};
    public static final Enchantment field_180310_c = func_222191_a("protection", new ProtectionEnchantment(Enchantment.Rarity.COMMON, ProtectionEnchantment.Type.ALL, field_222195_L));
