--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartEmpty.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartEmpty.java
@@ -19,6 +19,7 @@
 
     public boolean func_130002_c(EntityPlayer p_130002_1_)
     {
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.minecart.MinecartInteractEvent(this, p_130002_1_))) return true;
         if (this.field_70153_n != null && this.field_70153_n instanceof EntityPlayer && this.field_70153_n != p_130002_1_)
         {
             return true;
