--- ../src-base/minecraft/net/minecraft/entity/item/EntityXPOrb.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityXPOrb.java
@@ -208,6 +208,7 @@
         {
             if (this.field_70532_c == 0 && p_70100_1_.field_71090_bL == 0)
             {
+                if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerPickupXpEvent(p_70100_1_, this))) return;
                 p_70100_1_.field_71090_bL = 2;
                 p_70100_1_.func_71001_a(this, 1);
                 ItemStack itemstack = EnchantmentHelper.func_92099_a(Enchantments.field_185296_A, p_70100_1_);
