--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartHopper.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartHopper.java
@@ -53,6 +53,7 @@
     // JAVADOC METHOD $$ func_130002_c
     public boolean interactFirst(EntityPlayer par1EntityPlayer)
     {
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.minecart.MinecartInteractEvent(this, par1EntityPlayer))) return true;
         if (!this.worldObj.isRemote)
         {
             par1EntityPlayer.displayGUIHopperMinecart(this);
