--- ../src-base/minecraft/net/minecraft/entity/item/EntityEnderPearl.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityEnderPearl.java
@@ -61,6 +61,9 @@
 
                 if (entityplayermp.field_71135_a.func_147362_b().func_150724_d() && entityplayermp.field_70170_p == this.field_70170_p && !entityplayermp.func_70608_bn())
                 {
+                    net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(entityplayermp, this.field_70165_t, this.field_70163_u, this.field_70161_v, 5.0F);
+                    if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event))
+                    { // Don't indent to lower patch size
                     if (this.field_70146_Z.nextFloat() < 0.05F && this.field_70170_p.func_82736_K().func_82766_b("doMobSpawning"))
                     {
                         EntityEndermite entityendermite = new EntityEndermite(this.field_70170_p);
@@ -74,9 +77,10 @@
                         entitylivingbase.func_70078_a((Entity)null);
                     }
 
-                    entitylivingbase.func_70634_a(this.field_70165_t, this.field_70163_u, this.field_70161_v);
+                    entitylivingbase.func_70634_a(event.targetX, event.targetY, event.targetZ);
                     entitylivingbase.field_70143_R = 0.0F;
-                    entitylivingbase.func_70097_a(DamageSource.field_76379_h, 5.0F);
+                    entitylivingbase.func_70097_a(DamageSource.field_76379_h, event.attackDamage);
+                    }
                 }
             }
             else if (entitylivingbase != null)
