--- a/net/minecraft/entity/item/EntityXPOrb.java
+++ b/net/minecraft/entity/item/EntityXPOrb.java
@@ -117,7 +117,8 @@
       this.func_70091_d(MoverType.SELF, this.field_70159_w, this.field_70181_x, this.field_70179_y);
       float f = 0.98F;
       if (this.field_70122_E) {
-         f = this.field_70170_p.func_180495_p(new BlockPos(MathHelper.func_76128_c(this.field_70165_t), MathHelper.func_76128_c(this.func_174813_aQ().field_72338_b) - 1, MathHelper.func_76128_c(this.field_70161_v))).func_177230_c().func_208618_m() * 0.98F;
+         BlockPos underPos = new BlockPos(MathHelper.func_76128_c(this.field_70165_t), MathHelper.func_76128_c(this.func_174813_aQ().field_72338_b) - 1, MathHelper.func_76128_c(this.field_70161_v));
+         f = this.field_70170_p.func_180495_p(underPos).getSlipperiness(field_70170_p, underPos, this) * 0.98F;
       }
 
       this.field_70159_w *= (double)f;
@@ -150,6 +151,7 @@
    }
 
    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_) {
+      if (this.field_70170_p.field_72995_K || this.field_70128_L) return false; //Forge: Fixes MC-53850
       if (this.func_180431_b(p_70097_1_)) {
          return false;
       } else {
@@ -178,12 +180,14 @@
    public void func_70100_b_(EntityPlayer p_70100_1_) {
       if (!this.field_70170_p.field_72995_K) {
          if (this.field_70532_c == 0 && p_70100_1_.field_71090_bL == 0) {
+            if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerPickupXpEvent(p_70100_1_, this))) return;
             p_70100_1_.field_71090_bL = 2;
             p_70100_1_.func_71001_a(this, 1);
             ItemStack itemstack = EnchantmentHelper.func_92099_a(Enchantments.field_185296_A, p_70100_1_);
             if (!itemstack.func_190926_b() && itemstack.func_77951_h()) {
-               int i = Math.min(this.func_184514_c(this.field_70530_e), itemstack.func_77952_i());
-               this.field_70530_e -= this.func_184515_b(i);
+               float ratio = itemstack.func_77973_b().getXpRepairRatio(itemstack);
+               int i = Math.min(roundAverage(this.field_70530_e * ratio), itemstack.func_77952_i());
+               this.field_70530_e -= roundAverage(i / ratio);
                itemstack.func_196085_b(itemstack.func_77952_i() - i);
             }
 
@@ -261,4 +265,9 @@
    public boolean func_70075_an() {
       return false;
    }
+   
+   private static int roundAverage(float value) {
+      double floor = Math.floor(value);
+      return (int) floor + (Math.random() < value - floor ? 1 : 0);
+   }
 }
