--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartFurnace.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartFurnace.java
@@ -124,6 +124,7 @@
     // JAVADOC METHOD $$ func_130002_c
     public boolean interactFirst(EntityPlayer par1EntityPlayer)
     {
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.minecart.MinecartInteractEvent(this, par1EntityPlayer))) return true;
         ItemStack itemstack = par1EntityPlayer.inventory.getCurrentItem();
 
         if (itemstack != null && itemstack.getItem() == Items.coal)
