--- ../src-base/minecraft/net/minecraft/entity/item/EntityItemFrame.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityItemFrame.java
@@ -122,7 +122,7 @@
     {
         if (p_110131_1_ != null)
         {
-            if (p_110131_1_.func_77973_b() == Items.field_151098_aY)
+            if (p_110131_1_.func_77973_b() instanceof net.minecraft.item.ItemMap)
             {
                 MapData mapdata = ((ItemMap)p_110131_1_.func_77973_b()).func_77873_a(p_110131_1_, this.field_70170_p);
                 mapdata.field_76203_h.remove("frame-" + this.func_145782_y());
