--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartMobSpawner.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartMobSpawner.java
@@ -33,6 +33,9 @@
         {
             return new BlockPos(EntityMinecartMobSpawner.this);
         }
+        public net.minecraft.entity.Entity getSpawnerEntity() {
+            return EntityMinecartMobSpawner.this;
+        }
     };
 
     public EntityMinecartMobSpawner(World p_i46752_1_)
