--- a/net/minecraft/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/entity/item/FallingBlockEntity.java
@@ -153,7 +_,7 @@
                               ((FallingBlock)block).onLand(this.level, blockpos1, this.blockState, blockstate, this);
                            }
 
-                           if (this.blockData != null && block instanceof ITileEntityProvider) {
+                           if (this.blockData != null && this.blockState.hasTileEntity()) {
                               TileEntity tileentity = this.level.getBlockEntity(blockpos1);
                               if (tileentity != null) {
                                  CompoundNBT compoundnbt = tileentity.save(new CompoundNBT());
