--- a/net/minecraft/entity/item/LeashKnotEntity.java
+++ b/net/minecraft/entity/item/LeashKnotEntity.java
@@ -45,6 +45,7 @@
 
    protected void func_174856_o() {
       this.func_226288_n_((double)this.field_174861_a.func_177958_n() + 0.5D, (double)this.field_174861_a.func_177956_o() + 0.5D, (double)this.field_174861_a.func_177952_p() + 0.5D);
+      if (this.isAddedToWorld() && this.field_70170_p instanceof net.minecraft.world.server.ServerWorld) ((net.minecraft.world.server.ServerWorld)this.field_70170_p).func_217464_b(this); // Forge: Process chunk registration after moving.
    }
 
    public void func_174859_a(Direction p_174859_1_) {
