--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -138,8 +_,8 @@
 
         for (int i = 0; i < 10 && checkPos.getY() <= lastSetBlock; i++) {
             BlockState state = level.getBlockState(checkPos);
-            if (state.getBlock() instanceof BeaconBeamBlock beaconBeamBlock) {
-                int color = beaconBeamBlock.getColor().getTextureDiffuseColor();
+            int color = state.getBeaconColorMultiplier(level, checkPos, pos);
+            if (color != -1) {
                 if (entity.checkingBeamSections.size() <= 1) {
                     lastBeamSection = new BeaconBeamOwner.Section(color);
                     entity.checkingBeamSections.add(lastBeamSection);
