--- a/net/minecraft/world/level/block/entity/BlockEntityType.java
+++ b/net/minecraft/world/level/block/entity/BlockEntityType.java
@@ -85,7 +_,7 @@
    }
 
    public boolean m_155262_(BlockState p_155263_) {
-      return this.f_58915_.contains(p_155263_.m_60734_());
+      return this.f_58915_.contains(p_155263_.m_60734_()) || additionalValidBlocks.contains(p_155263_.m_60734_());
    }
 
    @Nullable
@@ -115,5 +_,13 @@
       public BlockEntityType<T> m_58966_(Type<?> p_58967_) {
          return new BlockEntityType<>(this.f_58959_, this.f_58960_, p_58967_);
       }
+   }
+
+   // FORGE START. INTERNAL USE ONLY!
+   private final java.util.Set<Block> additionalValidBlocks = new java.util.HashSet<>();
+
+   public static void fireAddBlockEvent() {
+       net.minecraftforge.registries.ForgeRegistries.BLOCK_ENTITY_TYPES.forEach(t -> t.additionalValidBlocks.clear());
+       net.minecraftforge.common.ForgeHooks.onBlockEntityBlocks().forEach((type, blocks) -> type.additionalValidBlocks.addAll(blocks));
    }
 }
