--- a/net/minecraft/world/level/block/entity/HangingSignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HangingSignBlockEntity.java
@@ -13,6 +_,10 @@
         super(BlockEntityType.HANGING_SIGN, p_250603_, p_251674_);
     }
 
+    public HangingSignBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState state) {
+        super(type, pos, state);
+    }
+
     @Override
     public int getTextLineHeight() {
         return 9;
