--- a/net/minecraft/client/multiplayer/chat/ChatListener.java
+++ b/net/minecraft/client/multiplayer/chat/ChatListener.java
@@ -104,7 +_,9 @@
       Instant instant = Instant.now();
       this.m_246163_((MessageSignature)null, () -> {
          Component component = p_251256_.m_240977_(p_250375_);
-         this.f_240348_.f_91065_.m_93076_().m_93785_(component);
+         Component forgeComponent = net.minecraftforge.client.ForgeHooksClient.onClientChat(p_251256_, component, Util.f_137441_);
+         if (forgeComponent == null) return false;
+         this.f_240348_.f_91065_.m_93076_().m_93785_(forgeComponent);
          this.m_241119_(p_251256_, p_250375_);
          this.m_240498_(component, instant);
          this.f_240659_ = Util.m_137550_();
@@ -121,12 +_,16 @@
          MessageSignature messagesignature = p_249430_.f_244279_();
          FilterMask filtermask = p_249430_.f_242992_();
          if (filtermask.m_243095_()) {
-            this.f_240348_.f_91065_.m_93076_().m_240964_(p_249231_, messagesignature, guimessagetag);
+            Component forgeComponent = net.minecraftforge.client.ForgeHooksClient.onClientPlayerChat(p_251766_, p_249231_, p_249430_, p_249430_.m_245167_());
+            if (forgeComponent == null) return false;
+            this.f_240348_.f_91065_.m_93076_().m_240964_(forgeComponent, messagesignature, guimessagetag);
             this.m_241119_(p_251766_, p_249430_.m_245692_());
          } else {
             Component component = filtermask.m_246134_(p_249430_.m_245728_());
             if (component != null) {
-               this.f_240348_.f_91065_.m_93076_().m_240964_(p_251766_.m_240977_(component), messagesignature, guimessagetag);
+               Component forgeComponent = net.minecraftforge.client.ForgeHooksClient.onClientPlayerChat(p_251766_, p_251766_.m_240977_(component), p_249430_, p_249430_.m_245167_());
+               if (forgeComponent == null) return false;
+               this.f_240348_.f_91065_.m_93076_().m_240964_(forgeComponent, messagesignature, guimessagetag);
                this.m_241119_(p_251766_, component);
             }
          }
@@ -159,6 +_,8 @@
 
    public void m_240494_(Component p_240522_, boolean p_240642_) {
       if (!this.f_240348_.f_91066_.m_231833_().m_231551_() || !this.f_240348_.m_91246_(this.m_240473_(p_240522_))) {
+         p_240522_ = net.minecraftforge.client.ForgeHooksClient.onClientSystemChat(p_240522_, p_240642_);
+         if (p_240522_ == null) return;
          if (p_240642_) {
             this.f_240348_.f_91065_.m_93063_(p_240522_, false);
          } else {
