--- ../src-base/minecraft/net/minecraft/client/renderer/texture/TextureMap.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/texture/TextureMap.java
@@ -96,6 +96,7 @@
         this.field_94252_e.clear();
         this.field_94258_i.clear();
         int j = Integer.MAX_VALUE;
+        cpw.mods.fml.common.ProgressManager.ProgressBar bar = cpw.mods.fml.common.ProgressManager.push("Texture stitching", this.field_110574_e.size());
         Iterator iterator = this.field_110574_e.entrySet().iterator();
         TextureAtlasSprite textureatlassprite;
 
@@ -108,6 +109,7 @@
 
             try
             {
+                bar.step(resourcelocation1.func_110623_a());
                 IResource iresource = p_110571_1_.func_110536_a(resourcelocation1);
                 BufferedImage[] abufferedimage = new BufferedImage[1 + this.field_147636_j];
                 abufferedimage[0] = ImageIO.read(iresource.func_110527_b());
@@ -156,12 +158,14 @@
             }
             catch (RuntimeException runtimeexception)
             {
-                field_147635_d.error("Unable to parse metadata from " + resourcelocation1, runtimeexception);
+                //logger.error("Unable to parse metadata from " + resourcelocation1, runtimeexception);
+                cpw.mods.fml.client.FMLClientHandler.instance().trackBrokenTexture(resourcelocation1, runtimeexception.getMessage());
                 continue;
             }
             catch (IOException ioexception1)
             {
-                field_147635_d.error("Using missing texture, unable to load " + resourcelocation1, ioexception1);
+                //logger.error("Using missing texture, unable to load " + resourcelocation1, ioexception1);
+                cpw.mods.fml.client.FMLClientHandler.instance().trackMissingTexture(resourcelocation1);
                 continue;
             }
 
@@ -169,6 +173,7 @@
             stitcher.func_110934_a(textureatlassprite);
         }
 
+        cpw.mods.fml.common.ProgressManager.pop(bar);
         int i1 = MathHelper.func_151239_c(j);
 
         if (i1 < this.field_147636_j)
