--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -116,6 +_,7 @@
          CriteriaTriggers.f_10585_.m_70959_((ServerPlayer)this.f_35263_, this, p_35274_.m_45368_());
       }
 
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.TradeWithVillagerEvent(this.f_35263_, p_35274_, this));
    }
 
    protected abstract void m_8058_(MerchantOffer p_35299_);
@@ -164,9 +_,9 @@
    }
 
    @Nullable
-   public Entity m_5489_(ServerLevel p_35295_) {
+   public Entity changeDimension(ServerLevel p_35295_, net.minecraftforge.common.util.ITeleporter teleporter) {
       this.m_7996_();
-      return super.m_5489_(p_35295_);
+      return super.changeDimension(p_35295_, teleporter);
    }
 
    protected void m_7996_() {
