--- a/net/minecraft/world/entity/npc/CatSpawner.java
+++ b/net/minecraft/world/entity/npc/CatSpawner.java
@@ -83,8 +_,8 @@
         if (cat == null) {
             return 0;
         } else {
+            cat.moveTo(p_35334_, 0.0F, 0.0F); // Fix MC-147659: Some witch huts spawn the incorrect cat
             cat.finalizeSpawn(p_35335_, p_35335_.getCurrentDifficultyAt(p_35334_), MobSpawnType.NATURAL, null);
-            cat.moveTo(p_35334_, 0.0F, 0.0F);
             p_35335_.addFreshEntityWithPassengers(cat);
             return 1;
         }
