--- a/net/minecraft/world/entity/npc/VillagerTrades.java
+++ b/net/minecraft/world/entity/npc/VillagerTrades.java
@@ -191,6 +_,7 @@
       private final int villagerXp;
 
       public EmeraldsForVillagerTypeItem(int p_35669_, int p_35670_, int p_35671_, Map<VillagerType, Item> p_35672_) {
+         if (Boolean.valueOf(false)) // FORGE: Modders can add custom villager types, so remove this validation, weird condition is to make the compiler not strip this dead code, breaking some mixins (see #10402)
          BuiltInRegistries.VILLAGER_TYPE.stream().filter((p_35680_) -> {
             return !p_35672_.containsKey(p_35680_);
          }).findAny().ifPresent((p_258962_) -> {
@@ -205,7 +_,9 @@
       @Nullable
       public MerchantOffer getOffer(Entity p_219685_, RandomSource p_219686_) {
          if (p_219685_ instanceof VillagerDataHolder villagerdataholder) {
-            ItemStack itemstack = new ItemStack(this.trades.get(villagerdataholder.getVillagerData().getType()), this.cost);
+            Item item = this.trades.get(villagerdataholder.getVillagerData().getType());
+            if (item == null) return null; // FORGE: Account for modded villager types by returning null if there is no trade
+            ItemStack itemstack = new ItemStack(item, this.cost);
             return new MerchantOffer(itemstack, new ItemStack(Items.EMERALD), this.maxUses, this.villagerXp, 0.05F);
          } else {
             return null;
