--- a/net/minecraft/world/level/gameevent/GameEvent.java
+++ b/net/minecraft/world/level/gameevent/GameEvent.java
@@ -2,7 +_,8 @@
 
 import net.minecraft.core.Registry;
 
-public class GameEvent {
+@net.minecraftforge.registries.ObjectHolder("minecraft")
+public class GameEvent extends net.minecraftforge.registries.ForgeRegistryEntry<GameEvent> {
    public static final GameEvent f_157791_ = m_157822_("block_attach");
    public static final GameEvent f_157792_ = m_157822_("block_change");
    public static final GameEvent f_157793_ = m_157822_("block_close");
@@ -51,10 +_,15 @@
    public static final int f_157788_ = 16;
    private final String f_157789_;
    private final int f_157790_;
+   private final net.minecraftforge.common.util.ReverseTagWrapper<GameEvent> reverseTags = new net.minecraftforge.common.util.ReverseTagWrapper<>(this, net.minecraft.tags.GameEventTags::m_144306_);
 
    public GameEvent(String p_157819_, int p_157820_) {
       this.f_157789_ = p_157819_;
       this.f_157790_ = p_157820_;
+   }
+
+   public java.util.Set<net.minecraft.resources.ResourceLocation> getTags() {
+      return reverseTags.getTagNames();
    }
 
    public String m_157821_() {
