--- a/net/minecraft/command/arguments/BlockStateParser.java
+++ b/net/minecraft/command/arguments/BlockStateParser.java
@@ -169,7 +_,7 @@
 
    private boolean hasBlockEntity(ITagCollection<Block> p_212598_1_) {
       if (this.state != null) {
-         return this.state.getBlock().isEntityBlock();
+         return this.state.hasTileEntity();
       } else {
          if (this.tag != null) {
             ITag<Block> itag = p_212598_1_.getTag(this.tag);
@@ -283,7 +_,7 @@
             p_197255_1_.suggest(String.valueOf('['));
          }
 
-         if (this.state.getBlock().isEntityBlock()) {
+         if (this.state.hasTileEntity()) {
             p_197255_1_.suggest(String.valueOf('{'));
          }
       }
