--- a/net/minecraft/network/chat/ChatDecorator.java
+++ b/net/minecraft/network/chat/ChatDecorator.java
@@ -29,7 +_,7 @@
 
    default CompletableFuture<FilteredText<PlayerChatMessage>> m_236964_(@Nullable ServerPlayer p_236965_, FilteredText<Component> p_236966_, MessageSignature p_236967_, boolean p_236968_) {
       return this.m_236969_(p_236965_, p_236966_).thenApply((p_236956_) -> {
-         return PlayerChatMessage.m_237223_(p_236966_, p_236956_, p_236967_, p_236968_);
+         return PlayerChatMessage.filteredSigned(p_236966_, p_236956_, p_236967_, p_236968_, p_236965_);
       });
    }
 }
