--- a/net/minecraft/network/chat/TranslatableComponent.java
+++ b/net/minecraft/network/chat/TranslatableComponent.java
@@ -21,7 +_,7 @@
    @Nullable
    private Language f_131300_;
    private final List<FormattedText> f_131301_ = Lists.newArrayList();
-   private static final Pattern f_131302_ = Pattern.compile("%(?:(\\d+)\\$)?([A-Za-z%]|$)");
+   private static final Pattern f_131302_ = Pattern.compile("%(?:(\\d+)\\$)?([+-]?,?\\d*\\.?\\d*)?([A-Za-z%]|$)");
 
    public TranslatableComponent(String p_131305_) {
       this.f_131298_ = p_131305_;
@@ -70,11 +_,12 @@
                this.f_131301_.add(FormattedText.m_130775_(s));
             }
 
-            String s4 = matcher.group(2);
+            String s4 = matcher.group(3);
             String s1 = p_131322_.substring(k, l);
             if ("%".equals(s4) && "%%".equals(s1)) {
                this.f_131301_.add(f_131296_);
             } else {
+               if (!"f".equals(s4) && !"d".equals(s4))
                if (!"s".equals(s4)) {
                   throw new TranslatableFormatException(this, "Unsupported format: '" + s1 + "'");
                }
@@ -82,11 +_,18 @@
                String s2 = matcher.group(1);
                int i1 = s2 != null ? Integer.parseInt(s2) - 1 : i++;
                if (i1 < this.f_131299_.length) {
+                  if (!(this.f_131299_[i1] instanceof Component) && this.f_131299_[i1] != null)
+                     this.f_131301_.add(FormattedText.m_130775_(("%"+matcher.group(2)+s4).formatted(this.f_131299_[i1])));
+                  else
                   this.f_131301_.add(this.m_131313_(i1));
                }
             }
          }
 
+         if (j == 0) {
+            // if we failed to match above, lets try the messageformat handler instead.
+            j = net.minecraftforge.fmllegacy.TextComponentMessageFormatHandler.handle(this, this.f_131301_, this.f_131299_, p_131322_);
+         }
          if (j < p_131322_.length()) {
             String s3 = p_131322_.substring(j);
             if (s3.indexOf(37) != -1) {
