--- a/net/minecraft/world/raid/Raid.java
+++ b/net/minecraft/world/raid/Raid.java
@@ -201,12 +201,14 @@
 
    public void func_221295_o() {
       this.field_221353_q = false;
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.RaidEvent.Stop(this))) return;
       this.field_221355_s.func_201360_b();
       this.field_221360_x = Raid.Status.STOPPED;
    }
 
    public void func_221307_p() {
       if (!this.func_221310_d()) {
+    	 if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.RaidEvent.Tick(this))) return;
          if (this.field_221360_x == Raid.Status.ONGOING) {
             boolean flag = this.field_221353_q;
             this.field_221353_q = this.field_221348_l.func_175667_e(this.field_221347_k);
@@ -296,6 +298,7 @@
             while(this.func_221318_F()) {
                BlockPos blockpos = this.field_221362_z.isPresent() ? this.field_221362_z.get() : this.func_221298_a(k, 20);
                if (blockpos != null) {
+            	  if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.RaidEvent.Start(this))) return;
                   this.field_221349_m = true;
                   this.func_221294_b(blockpos);
                   if (!flag3) {
@@ -469,6 +472,7 @@
             }
 
             this.func_221317_a(i, abstractraiderentity, p_221294_1_, false);
+            if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.RaidEvent.RaidersSpawn(this, abstractraiderentity))) return;
             if (raid$wavemember.field_221285_g == EntityType.field_220352_aU) {
                AbstractRaiderEntity abstractraiderentity1 = null;
                if (i == this.func_221306_a(Difficulty.NORMAL)) {
@@ -655,6 +659,7 @@
 
    private int func_221335_a(Raid.WaveMember p_221335_1_, Random p_221335_2_, int p_221335_3_, DifficultyInstance p_221335_4_, boolean p_221335_5_) {
       Difficulty difficulty = p_221335_4_.func_203095_a();
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.RaidEvent.RaidersSpawn(this, p_221335_1_.field_221285_g.func_200721_a(field_221348_l)))) return 0;
       boolean flag = difficulty == Difficulty.EASY;
       boolean flag1 = difficulty == Difficulty.NORMAL;
       int i;
@@ -682,7 +687,6 @@
       default:
          return 0;
       }
-
       return i > 0 ? p_221335_2_.nextInt(i + 1) : 0;
    }
 
@@ -770,7 +774,7 @@
       }
    }
 
-   static enum WaveMember {
+   public static enum WaveMember implements net.minecraftforge.common.IExtensibleEnum {
       VINDICATOR(EntityType.field_200758_ax, new int[]{0, 0, 2, 0, 1, 4, 2, 5}),
       EVOKER(EntityType.field_200806_t, new int[]{0, 0, 0, 0, 0, 1, 1, 2}),
       PILLAGER(EntityType.field_220350_aJ, new int[]{0, 4, 3, 3, 4, 4, 4, 2}),
@@ -785,5 +789,13 @@
          this.field_221285_g = p_i50602_3_;
          this.field_221286_h = p_i50602_4_;
       }
+      
+      public static WaveMember create(String name, EntityType<? extends AbstractRaiderEntity> typeIn, int[] waveCountsIn) {
+         throw new IllegalStateException("Enum not extended");
+      }
+      
+      public void init() {
+    	  field_221284_f = values();
+      }
    }
 }
