--- a/com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
+++ b/com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
@@ -45,9 +_,9 @@
       }
    };
    private final RealmsNotificationsScreen.DataFetcherConfiguration f_273849_ = new RealmsNotificationsScreen.DataFetcherConfiguration() {
-      public DataFetcher.Subscription m_274316_(RealmsDataFetcher p_275731_) {
-         DataFetcher.Subscription datafetcher$subscription = p_275731_.f_238549_.m_239139_();
-         RealmsNotificationsScreen.this.m_274585_(p_275731_, datafetcher$subscription);
+      public DataFetcher.Subscription m_274316_(RealmsDataFetcher p_275318_) {
+         DataFetcher.Subscription datafetcher$subscription = p_275318_.f_238549_.m_239139_();
+         RealmsNotificationsScreen.this.m_274585_(p_275318_, datafetcher$subscription);
          return datafetcher$subscription;
       }
 
