--- a/com/mojang/realmsclient/gui/screens/RealmsGenericErrorScreen.java
+++ b/com/mojang/realmsclient/gui/screens/RealmsGenericErrorScreen.java
@@ -68,6 +_,15 @@
       return Component.m_237119_().m_7220_(this.f_200947_.f_287789_).m_130946_(": ").m_7220_(this.f_200947_.f_287787_);
    }
 
+    @Override
+    public boolean m_7933_(int key, int scanCode, int modifiers) {
+       if (key == org.lwjgl.glfw.GLFW.GLFW_KEY_ESCAPE) {
+          f_96541_.m_91152_(this.f_88665_);
+          return true;
+       }
+       return super.m_7933_(key, scanCode, modifiers);
+    }
+
    public void m_88315_(GuiGraphics p_283497_, int p_88680_, int p_88681_, float p_88682_) {
       this.m_280273_(p_283497_);
       p_283497_.m_280653_(this.f_96547_, this.f_200947_.f_287789_, this.f_96543_ / 2, 80, 16777215);
