--- ../src-base/minecraft/net/minecraft/client/gui/achievement/GuiStats.java
+++ ../src-work/minecraft/net/minecraft/client/gui/achievement/GuiStats.java
@@ -5,6 +5,8 @@
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
+
+import net.minecraft.block.Block;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.audio.PositionedSoundRecord;
 import net.minecraft.client.gui.GuiButton;
@@ -443,17 +445,18 @@
             for (StatCrafting statcrafting : StatList.field_75939_e)
             {
                 boolean flag = false;
-                int i = Item.func_150891_b(statcrafting.func_150959_a());
+                int i = Block.func_149682_b(Block.func_149634_a(statcrafting.func_150959_a()));
+                int itemId = Item.func_150891_b(statcrafting.func_150959_a());
 
                 if (GuiStats.this.field_146546_t.func_77444_a(statcrafting) > 0)
                 {
                     flag = true;
                 }
-                else if (StatList.field_75929_E[i] != null && GuiStats.this.field_146546_t.func_77444_a(StatList.field_75929_E[i]) > 0)
+                else if (StatList.field_75929_E[itemId] != null && GuiStats.this.field_146546_t.func_77444_a(StatList.field_75929_E[itemId]) > 0)
                 {
                     flag = true;
                 }
-                else if (StatList.field_75928_D[i] != null && GuiStats.this.field_146546_t.func_77444_a(StatList.field_75928_D[i]) > 0)
+                else if (StatList.field_75928_D[itemId] != null && GuiStats.this.field_146546_t.func_77444_a(StatList.field_75928_D[itemId]) > 0)
                 {
                     flag = true;
                 }
@@ -468,8 +471,15 @@
             {
                 public int compare(StatCrafting p_compare_1_, StatCrafting p_compare_2_)
                 {
-                    int j = Item.func_150891_b(p_compare_1_.func_150959_a());
-                    int k = Item.func_150891_b(p_compare_2_.func_150959_a());
+                    int j;
+                    int k;
+                    if (StatsBlock.this.field_148217_o == 2) {
+                        j = Block.func_149682_b(Block.func_149634_a(p_compare_1_.func_150959_a()));
+                        k = Block.func_149682_b(Block.func_149634_a(p_compare_2_.func_150959_a()));
+                    } else {
+                        j = Item.func_150891_b(p_compare_1_.func_150959_a());
+                        k = Item.func_150891_b(p_compare_2_.func_150959_a());
+                    }
                     StatBase statbase = null;
                     StatBase statbase1 = null;
 
@@ -748,7 +758,7 @@
             super(p_i45510_2_, GuiStats.this.field_146294_l, GuiStats.this.field_146295_m, 32, GuiStats.this.field_146295_m - 64, GuiStats.this.field_146289_q.field_78288_b * 4);
             this.func_148130_a(false);
 
-            for (EntityList.EntityEggInfo entitylist$entityegginfo : EntityList.field_75627_a.values())
+            for (EntityList.EntityEggInfo entitylist$entityegginfo : com.google.common.collect.Iterables.concat(EntityList.field_75627_a.values(), net.minecraftforge.fml.common.registry.EntityRegistry.getEggs().values()))
             {
                 if (GuiStats.this.field_146546_t.func_77444_a(entitylist$entityegginfo.field_151512_d) > 0 || GuiStats.this.field_146546_t.func_77444_a(entitylist$entityegginfo.field_151513_e) > 0)
                 {
@@ -784,7 +794,7 @@
         protected void func_180791_a(int p_180791_1_, int p_180791_2_, int p_180791_3_, int p_180791_4_, int p_180791_5_, int p_180791_6_)
         {
             EntityList.EntityEggInfo entitylist$entityegginfo = (EntityList.EntityEggInfo)this.field_148222_l.get(p_180791_1_);
-            String s = I18n.func_135052_a("entity." + EntityList.func_75617_a(entitylist$entityegginfo.field_75613_a) + ".name", new Object[0]);
+            String s = I18n.func_135052_a("entity." + entitylist$entityegginfo.name + ".name", new Object[0]);
             int i = GuiStats.this.field_146546_t.func_77444_a(entitylist$entityegginfo.field_151512_d);
             int j = GuiStats.this.field_146546_t.func_77444_a(entitylist$entityegginfo.field_151513_e);
             String s1 = I18n.func_135052_a("stat.entityKills", new Object[] {Integer.valueOf(i), s});
