--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -101,6 +_,7 @@
 
          this.f_64068_ = p_64080_.m_128471_("DragonKilled");
          this.f_64069_ = p_64080_.m_128471_("PreviouslyKilled");
+         this.f_64071_ = !p_64080_.m_128471_("LegacyScanPerformed"); // Forge: fix MC-105080
          if (p_64080_.m_128471_("IsRespawning")) {
             this.f_64073_ = DragonRespawnAnimation.START;
          }
@@ -136,6 +_,7 @@
 
       compoundtag.m_128379_("DragonKilled", this.f_64068_);
       compoundtag.m_128379_("PreviouslyKilled", this.f_64069_);
+      compoundtag.m_128379_("LegacyScanPerformed", !this.f_64071_); // Forge: fix MC-105080
       if (this.f_64072_ != null) {
          compoundtag.m_128365_("ExitPortalLocation", NbtUtils.m_129224_(this.f_64072_));
       }
@@ -508,6 +_,13 @@
             endcrystal.m_31052_((BlockPos)null);
          }
       }
-
+   }
+
+   public void addPlayer(ServerPlayer player) {
+      this.f_64060_.m_6543_(player);
+   }
+
+   public void removePlayer(ServerPlayer player) {
+      this.f_64060_.m_6539_(player);
    }
 }
