--- ../src_base/minecraft/net/minecraft/entity/boss/BossStatus.java
+++ ../src_work/minecraft/net/minecraft/entity/boss/BossStatus.java
@@ -13,7 +13,7 @@
 
     public static void func_82824_a(IBossDisplayData par0IBossDisplayData, boolean par1)
     {
-        healthScale = par0IBossDisplayData.func_110183_aJ() / par0IBossDisplayData.getMaxHealth();
+        healthScale = par0IBossDisplayData.func_110143_aJ() / par0IBossDisplayData.func_110138_aP();
         statusBarLength = 100;
         bossName = par0IBossDisplayData.getEntityName();
         field_82825_d = par1;
