--- ../src-base/minecraft/net/minecraft/entity/boss/EntityDragon.java
+++ ../src-work/minecraft/net/minecraft/entity/boss/EntityDragon.java
@@ -495,9 +495,9 @@
                 {
                     Block block = this.worldObj.getBlock(k1, l1, i2);
 
-                    if (block.getMaterial() != Material.air)
+                    if (!block.isAir(worldObj, k1, l1, i2))
                     {
-                        if (block != Blocks.obsidian && block != Blocks.end_stone && block != Blocks.bedrock && this.worldObj.getGameRules().getGameRuleBooleanValue("mobGriefing"))
+                        if (block.canEntityDestroy(worldObj, k1, l1, i2, this) && this.worldObj.getGameRules().getGameRuleBooleanValue("mobGriefing"))
                         {
                             flag1 = this.worldObj.setBlockToAir(k1, l1, i2) || flag1;
                         }
