--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -169,6 +_,7 @@
       }
 
       if (p_60195_ == 0) {
+         if (net.minecraftforge.event.ForgeEventFactory.onPistonMovePre(p_60193_, p_60194_, direction, true)) return false;
          if (!this.moveBlocks(p_60193_, p_60194_, direction, true)) {
             return false;
          }
@@ -177,6 +_,7 @@
          p_60193_.playSound((Player)null, p_60194_, SoundEvents.PISTON_EXTEND, SoundSource.BLOCKS, 0.5F, p_60193_.random.nextFloat() * 0.25F + 0.6F);
          p_60193_.gameEvent(GameEvent.BLOCK_ACTIVATE, p_60194_, GameEvent.Context.of(blockstate));
       } else if (p_60195_ == 1 || p_60195_ == 2) {
+         if (net.minecraftforge.event.ForgeEventFactory.onPistonMovePre(p_60193_, p_60194_, direction, false)) return false;
          BlockEntity blockentity1 = p_60193_.getBlockEntity(p_60194_.relative(direction));
          if (blockentity1 instanceof PistonMovingBlockEntity) {
             ((PistonMovingBlockEntity)blockentity1).finalTick();
@@ -217,6 +_,7 @@
          p_60193_.gameEvent(GameEvent.BLOCK_DEACTIVATE, p_60194_, GameEvent.Context.of(blockstate1));
       }
 
+      net.minecraftforge.event.ForgeEventFactory.onPistonMovePost(p_60193_, p_60194_, direction, (p_60195_ == 0));
       return true;
    }
 
@@ -353,6 +_,10 @@
 
    public BlockState rotate(BlockState p_60215_, Rotation p_60216_) {
       return p_60215_.setValue(FACING, p_60216_.rotate(p_60215_.getValue(FACING)));
+   }
+
+   public BlockState rotate(BlockState state, net.minecraft.world.level.LevelAccessor world, BlockPos pos, Rotation direction) {
+       return state.getValue(EXTENDED) ? state : super.rotate(state, world, pos, direction);
    }
 
    public BlockState mirror(BlockState p_60212_, Mirror p_60213_) {
