--- a/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceCondition.java
@@ -18,7 +_,8 @@
    }
 
    public boolean test(LootContext p_81930_) {
-      return p_81930_.m_230907_().m_188501_() < this.f_81921_;
+      int i = p_81930_.getLootingModifier();
+      return p_81930_.m_230907_().m_188501_() < this.f_81921_ + (float) i;
    }
 
    public static LootItemCondition.Builder m_81927_(float p_81928_) {
