--- a/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
@@ -33,12 +_,7 @@
     }
 
     public boolean test(LootContext p_81967_) {
-        Entity entity = p_81967_.getParamOrNull(LootContextParams.KILLER_ENTITY);
-        int i = 0;
-        if (entity instanceof LivingEntity) {
-            i = EnchantmentHelper.getMobLooting((LivingEntity)entity);
-        }
-
+        int i = p_81967_.getLootingModifier();
         return p_81967_.getRandom().nextFloat() < this.percent + (float)i * this.lootingMultiplier;
     }
 
