--- a/net/minecraft/world/level/storage/loot/predicates/LootItemConditionType.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemConditionType.java
@@ -3,7 +_,7 @@
 import net.minecraft.world.level.storage.loot.Serializer;
 import net.minecraft.world.level.storage.loot.SerializerType;
 
-public class LootItemConditionType extends SerializerType<LootItemCondition> {
+public class LootItemConditionType extends net.minecraftforge.registries.ForgeSerializerTypeRegistryEntry<LootItemConditionType, LootItemCondition> {
    public LootItemConditionType(Serializer<? extends LootItemCondition> p_81810_) {
       super(p_81810_);
    }
