--- a/net/minecraft/world/level/storage/loot/predicates/LootItemConditions.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemConditions.java
@@ -6,6 +_,7 @@
 import net.minecraft.world.level.storage.loot.GsonAdapterFactory;
 import net.minecraft.world.level.storage.loot.Serializer;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class LootItemConditions {
    public static final LootItemConditionType f_81811_ = m_81831_("inverted", new InvertedLootItemCondition.Serializer());
    public static final LootItemConditionType f_81812_ = m_81831_("alternative", new AlternativeLootItemCondition.Serializer());
