--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -104,6 +_,14 @@
       this.noPhysics = true;
       this.noCulling = true;
       this.phaseManager = new EnderDragonPhaseManager(this);
+      this.setId(ENTITY_COUNTER.getAndAdd(this.subEntities.length + 1) + 1); // Forge: Fix MC-158205: Make sure part ids are successors of parent mob id
+   }
+
+   @Override
+   public void setId(int p_20235_) {
+      super.setId(p_20235_);
+      for (int i = 0; i < this.subEntities.length; i++) // Forge: Fix MC-158205: Set part ids to successors of parent mob id
+         this.subEntities[i].setId(p_20235_ + i + 1);
    }
 
    public void setDragonFight(EndDragonFight p_287736_) {
@@ -159,7 +_,11 @@
       return adouble;
    }
 
+   @org.jetbrains.annotations.Nullable private Player unlimitedLastHurtByPlayer = null;
    public void aiStep() {
+      // lastHurtByPlayer is cleared after 100 ticks, capture it indefinitely in unlimitedLastHurtByPlayer for LivingExperienceDropEvent
+      if (this.lastHurtByPlayer != null) this.unlimitedLastHurtByPlayer = lastHurtByPlayer;
+      if (this.unlimitedLastHurtByPlayer != null && this.unlimitedLastHurtByPlayer.isRemoved()) this.unlimitedLastHurtByPlayer = null;
       this.processFlappingMovement();
       if (this.level().isClientSide) {
          this.setHealth(this.getHealth());
@@ -433,7 +_,7 @@
                BlockPos blockpos = new BlockPos(k1, l1, i2);
                BlockState blockstate = this.level().getBlockState(blockpos);
                if (!blockstate.isAir() && !blockstate.is(BlockTags.DRAGON_TRANSPARENT)) {
-                  if (this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) && !blockstate.is(BlockTags.DRAGON_IMMUNE)) {
+                  if (net.minecraftforge.common.ForgeHooks.canEntityDestroy(this.level(), blockpos, this) && !blockstate.is(BlockTags.DRAGON_IMMUNE)) {
                      flag1 = this.level().removeBlock(blockpos, false) || flag1;
                   } else {
                      flag = true;
@@ -524,7 +_,8 @@
 
       if (this.level() instanceof ServerLevel) {
          if (this.dragonDeathTime > 150 && this.dragonDeathTime % 5 == 0 && flag) {
-            ExperienceOrb.award((ServerLevel)this.level(), this.position(), Mth.floor((float)i * 0.08F));
+            int award = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.unlimitedLastHurtByPlayer, Mth.floor((float)i * 0.08F));
+            ExperienceOrb.award((ServerLevel) this.level(), this.position(), award);
          }
 
          if (this.dragonDeathTime == 1 && !this.isSilent()) {
@@ -535,7 +_,8 @@
       this.move(MoverType.SELF, new Vec3(0.0D, (double)0.1F, 0.0D));
       if (this.dragonDeathTime == 200 && this.level() instanceof ServerLevel) {
          if (flag) {
-            ExperienceOrb.award((ServerLevel)this.level(), this.position(), Mth.floor((float)i * 0.2F));
+            int award = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.unlimitedLastHurtByPlayer, Mth.floor((float)i * 0.2F));
+            ExperienceOrb.award((ServerLevel) this.level(), this.position(), award);
          }
 
          if (this.dragonFight != null) {
@@ -858,8 +_,19 @@
       return false;
    }
 
+   @Override
+   public boolean isMultipartEntity() {
+      return true;
+   }
+
+   @Override
+   public net.minecraftforge.entity.PartEntity<?>[] getParts() {
+      return this.subEntities;
+   }
+
    public void recreateFromPacket(ClientboundAddEntityPacket p_218825_) {
       super.recreateFromPacket(p_218825_);
+      if (true) return; // Forge: Fix MC-158205: Moved into setId()
       EnderDragonPart[] aenderdragonpart = this.getSubEntities();
 
       for(int i = 0; i < aenderdragonpart.length; ++i) {
