--- a/net/minecraft/client/model/HumanoidModel.java
+++ b/net/minecraft/client/model/HumanoidModel.java
@@ -391,6 +_,8 @@
                 break;
             case SPEAR:
                 SpearAnimations.thirdPersonHandUse(this.rightArm, this.head, true, state.getUseItemStackForArm(HumanoidArm.RIGHT), state);
+            default:
+                state.rightArmPose.applyTransform(this, state, net.minecraft.world.entity.HumanoidArm.RIGHT);
         }
     }
 
@@ -436,6 +_,8 @@
                 break;
             case SPEAR:
                 SpearAnimations.thirdPersonHandUse(this.leftArm, this.head, false, state.getUseItemStackForArm(HumanoidArm.LEFT), state);
+            default:
+                state.leftArmPose.applyTransform(this, state, net.minecraft.world.entity.HumanoidArm.LEFT);
         }
     }
 
@@ -497,7 +_,7 @@
     }
 
     @OnlyIn(Dist.CLIENT)
-    public enum ArmPose {
+    public enum ArmPose implements net.minecraftforge.common.IExtensibleEnum {
         EMPTY(false, false),
         ITEM(false, false),
         BLOCK(false, false),
@@ -523,10 +_,29 @@
         ArmPose(final boolean twoHanded, final boolean affectsOffhandPose) {
             this.twoHanded = twoHanded;
             this.affectsOffhandPose = affectsOffhandPose;
+            this.forgeArmPose = null;
         }
 
         public boolean isTwoHanded() {
             return this.twoHanded;
+        }
+
+        @javax.annotation.Nullable
+        private final net.minecraftforge.client.IArmPoseTransformer forgeArmPose;
+
+        private ArmPose(boolean twoHanded, boolean affectsOffhandPose, @javax.annotation.Nonnull net.minecraftforge.client.IArmPoseTransformer forgeArmPose) {
+            this.twoHanded = twoHanded;
+            this.affectsOffhandPose = affectsOffhandPose;
+            com.google.common.base.Preconditions.checkNotNull(forgeArmPose, "Cannot create new ArmPose with null transformer!");
+            this.forgeArmPose = forgeArmPose;
+        }
+
+        public static ArmPose create(String name, boolean twoHanded, boolean affectsOffhandPose, @javax.annotation.Nonnull net.minecraftforge.client.IArmPoseTransformer forgeArmPose) {
+            throw new IllegalStateException("Enum not extended");
+        }
+
+        public <T extends HumanoidRenderState> void applyTransform(HumanoidModel<T> model, T state, net.minecraft.world.entity.HumanoidArm arm) {
+            if (this.forgeArmPose != null) this.forgeArmPose.applyTransform(model, state, arm);
         }
 
         public boolean affectsOffhandPose() {
