--- a/net/minecraft/client/model/ListModel.java
+++ b/net/minecraft/client/model/ListModel.java
@@ -27,4 +_,9 @@
    }
 
    public abstract Iterable<ModelPart> m_6195_();
+
+   @Override
+   public void gatherAnimatedParts(java.util.function.BiConsumer<String, ModelPart> consumer) {
+      net.minecraftforge.client.animation.IAnimatedModel.fillAnimatedPartsFromIterable(consumer, this.m_6195_(), "part");
+   }
 }
