--- ../src-base/minecraft/net/minecraft/client/model/ModelBox.java
+++ ../src-work/minecraft/net/minecraft/client/model/ModelBox.java
@@ -4,7 +4,6 @@
 import cpw.mods.fml.relauncher.SideOnly;
 import net.minecraft.client.renderer.Tessellator;
 
-@SideOnly(Side.CLIENT)
 public class ModelBox
 {
     // JAVADOC FIELD $$ field_78253_h
@@ -86,6 +85,7 @@
     }
 
     // JAVADOC METHOD $$ func_78245_a
+    @SideOnly(Side.CLIENT)
     public void render(Tessellator par1Tessellator, float par2)
     {
         for (int i = 0; i < this.quadList.length; ++i)
