--- ../src-base/minecraft/net/minecraft/server/network/NetHandlerLoginServer.java
+++ ../src-work/minecraft/net/minecraft/server/network/NetHandlerLoginServer.java
@@ -66,7 +66,7 @@
             if (entityplayermp == null)
             {
                 this.currentLoginState = NetHandlerLoginServer.LoginState.READY_TO_ACCEPT;
-                this.server.func_184103_al().initializeConnectionToPlayer(this.networkManager, this.field_181025_l);
+                net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.fmlServerHandshake(this.server.func_184103_al(), this.networkManager, this.field_181025_l);
                 this.field_181025_l = null;
             }
         }
@@ -130,7 +130,7 @@
             }
             else
             {
-                this.server.func_184103_al().initializeConnectionToPlayer(this.networkManager, this.server.func_184103_al().createPlayerForUser(this.loginGameProfile));
+                net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.fmlServerHandshake(this.server.func_184103_al(), this.networkManager, this.server.func_184103_al().createPlayerForUser(this.loginGameProfile));
             }
         }
     }
