--- a/net/minecraft/client/gui/screens/packs/TransferableSelectionList.java
+++ b/net/minecraft/client/gui/screens/packs/TransferableSelectionList.java
@@ -65,7 +_,7 @@
         Component header = Component.empty().append(this.title).withStyle(ChatFormatting.UNDERLINE, ChatFormatting.BOLD);
         this.addEntry(new TransferableSelectionList.HeaderEntry(this.minecraft.font, header), (int)(9.0F * 1.5F));
         this.setSelected(null);
-        entries.forEach(e -> {
+        entries.filter(PackSelectionModel.Entry::notHidden).forEach(e -> {
             TransferableSelectionList.PackEntry entry = new TransferableSelectionList.PackEntry(this.minecraft, this, e);
             this.addEntry(entry);
             if (transferredEntry != null && transferredEntry.getId().equals(e.getId())) {
