--- a/net/minecraft/client/ClientRecipeBook.java
+++ b/net/minecraft/client/ClientRecipeBook.java
@@ -54,7 +_,7 @@
       for(Recipe<?> recipe : p_90643_) {
          if (!recipe.m_5598_() && !recipe.m_142505_()) {
             RecipeBookCategories recipebookcategories = m_90646_(recipe);
-            String s = recipe.m_6076_();
+            String s = recipe.m_6076_().isEmpty() ? recipe.m_6423_().toString() : recipe.m_6076_(); // FORGE: Group value defaults to the recipe's ID if the recipe's explicit group is empty.
             if (s.isEmpty()) {
                map.computeIfAbsent(recipebookcategories, (p_90645_) -> {
                   return Lists.newArrayList();
@@ -141,6 +_,8 @@
          } else if (recipetype == RecipeType.f_44113_) {
             return RecipeBookCategories.SMITHING;
          } else {
+            RecipeBookCategories categories = net.minecraftforge.client.RecipeBookManager.findCategories((RecipeType) recipetype, p_90647_);
+            if (categories != null) return categories;
             f_90618_.warn("Unknown recipe category: {}/{}", LogUtils.defer(() -> {
                return BuiltInRegistries.f_256990_.m_7981_(p_90647_.m_6671_());
             }), LogUtils.defer(p_90647_::m_6423_));
