--- a/net/minecraft/client/GameSettings.java
+++ b/net/minecraft/client/GameSettings.java
@@ -170,6 +_,7 @@
    public boolean syncWrites;
 
    public GameSettings(Minecraft p_i46326_1_, File p_i46326_2_) {
+      setForgeKeybindProperties();
       this.minecraft = p_i46326_1_;
       this.optionsFile = new File(p_i46326_2_, "options.txt");
       if (p_i46326_1_.is64Bit() && Runtime.getRuntime().maxMemory() >= 1000000000L) {
@@ -529,7 +_,11 @@
 
                for(KeyBinding keybinding : this.keyMappings) {
                   if (s.equals("key_" + keybinding.getName())) {
-                     keybinding.setKey(InputMappings.getKey(s1));
+                     if (s1.indexOf(':') != -1) {
+                        String[] pts = s1.split(":");
+                        keybinding.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.valueFromString(pts[1]), InputMappings.getKey(pts[0]));
+                     } else
+                        keybinding.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.NONE, InputMappings.getKey(s1));
                   }
                }
 
@@ -576,6 +_,7 @@
    }
 
    public void save() {
+      if (net.minecraftforge.fml.client.ClientModLoader.isLoading()) return; //Don't save settings before mods add keybindigns and the like to prevent them from being deleted.
       try (PrintWriter printwriter = new PrintWriter(new OutputStreamWriter(new FileOutputStream(this.optionsFile), StandardCharsets.UTF_8))) {
          printwriter.println("version:" + SharedConstants.getCurrentVersion().getWorldVersion());
          printwriter.println("autoJump:" + AbstractOption.AUTO_JUMP.get(this));
@@ -661,7 +_,7 @@
          printwriter.println("syncChunkWrites:" + this.syncWrites);
 
          for(KeyBinding keybinding : this.keyMappings) {
-            printwriter.println("key_" + keybinding.getName() + ":" + keybinding.saveString());
+            printwriter.println("key_" + keybinding.getName() + ":" + keybinding.saveString() + (keybinding.getKeyModifier() != net.minecraftforge.client.settings.KeyModifier.NONE ? ":" + keybinding.getKeyModifier() : ""));
          }
 
          for(SoundCategory soundcategory : SoundCategory.values()) {
@@ -758,6 +_,23 @@
       }
 
       p_198017_1_.setSelected(set);
+   }
+
+   private void setForgeKeybindProperties() {
+      net.minecraftforge.client.settings.KeyConflictContext inGame = net.minecraftforge.client.settings.KeyConflictContext.IN_GAME;
+      keyUp.setKeyConflictContext(inGame);
+      keyLeft.setKeyConflictContext(inGame);
+      keyDown.setKeyConflictContext(inGame);
+      keyRight.setKeyConflictContext(inGame);
+      keyJump.setKeyConflictContext(inGame);
+      keyShift.setKeyConflictContext(inGame);
+      keySprint.setKeyConflictContext(inGame);
+      keyAttack.setKeyConflictContext(inGame);
+      keyChat.setKeyConflictContext(inGame);
+      keyPlayerList.setKeyConflictContext(inGame);
+      keyCommand.setKeyConflictContext(inGame);
+      keyTogglePerspective.setKeyConflictContext(inGame);
+      keySmoothCamera.setKeyConflictContext(inGame);
    }
 
    public PointOfView getCameraType() {
