--- ../src-base/minecraft/net/minecraft/client/Minecraft.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src-work/minecraft/net/minecraft/client/Minecraft.java	0000-00-00 00:00:00.000000000 -0000
@@ -126,6 +126,10 @@
 import org.lwjgl.opengl.PixelFormat;
 import org.lwjgl.util.glu.GLU;
 
+import cpw.mods.fml.client.FMLClientHandler;
+import cpw.mods.fml.relauncher.ArgsWrapper;
+import cpw.mods.fml.relauncher.FMLEmbeddingRelauncher;
+
 public abstract class Minecraft implements Runnable, IPlayerUsage
 {
     public static byte[] field_71444_a = new byte[10485760];
@@ -305,7 +309,7 @@
         this.func_71357_I();
         this.field_71466_p = new FontRenderer(this.field_71474_y, "/font/default.png", this.field_71446_o, false);
         this.field_71464_q = new FontRenderer(this.field_71474_y, "/font/alternate.png", this.field_71446_o, false);
-
+        FMLClientHandler.instance().beginMinecraftLoading(this);
         if (this.field_71474_y.field_74363_ab != null)
         {
             StringTranslate.func_74808_a().func_74810_a(this.field_71474_y.field_74363_ab);
@@ -1883,6 +1887,12 @@
 
     public static void main(String[] p_main_0_)
     {
+        FMLEmbeddingRelauncher.relaunch(new ArgsWrapper(p_main_0_));
+    }
+
+    public static void fmlReentry(ArgsWrapper wrapper)
+    {
+        String[] p_main_0_ = wrapper.args;
         HashMap var1 = new HashMap();
         boolean var2 = false;
         boolean var3 = true;
