--- a/net/minecraft/client/Minecraft.java
+++ b/net/minecraft/client/Minecraft.java
@@ -405,6 +405,7 @@
       this.languageManager = new LanguageManager(this.gameSettings.language);
       this.resourceManager.func_199006_a(this.languageManager);
       this.gameSettings.func_198017_a(this.resourcePackRepository);
+      net.minecraftforge.fml.client.ClientModLoader.begin(this, this.resourcePackRepository, this.resourceManager, this.field_195554_ax);
       this.refreshResources();
       this.renderEngine = new TextureManager(this.resourceManager);
       this.resourceManager.func_199006_a(this.renderEngine);
@@ -461,6 +462,7 @@
       this.resourceManager.func_199006_a(this.searchTreeManager);
       GlStateManager.viewport(0, 0, this.field_195558_d.func_198109_k(), this.field_195558_d.func_198091_l());
       this.effectRenderer = new ParticleManager(this.world, this.renderEngine);
+      net.minecraftforge.fml.client.ClientModLoader.end();
       this.ingameGUI = new GuiIngame(this);
       if (this.serverName != null) {
          this.displayGuiScreen(new GuiConnecting(new GuiMainMenu(), this, this.serverName, this.serverPort));
@@ -470,6 +472,7 @@
 
       this.debugRenderer = new DebugRenderer(this);
       GLFW.glfwSetErrorCallback(this::func_195545_a).free();
+      net.minecraftforge.fml.client.ClientModLoader.complete();
       if (this.gameSettings.fullScreen && !this.field_195558_d.func_198113_j()) {
          this.field_195558_d.func_198077_g();
       }
@@ -1392,6 +1395,7 @@
       NetworkManager networkmanager = NetworkManager.provideLocalClient(socketaddress);
       networkmanager.setNetHandler(new NetHandlerLoginClient(networkmanager, this, (GuiScreen)null, (p_209507_0_) -> {
       }));
+      net.minecraftforge.fml.network.NetworkHooks.registerClientLoginChannel(networkmanager);
       networkmanager.sendPacket(new CPacketHandshake(socketaddress.toString(), 0, EnumConnectionState.LOGIN));
       networkmanager.sendPacket(new CPacketLoginStart(this.getSession().getProfile()));
       this.networkManager = networkmanager;
