--- ../src-base/minecraft/net/minecraft/client/Minecraft.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src-work/minecraft/net/minecraft/client/Minecraft.java	0000-00-00 00:00:00.000000000 -0000
@@ -117,6 +117,8 @@
 import org.lwjgl.opengl.PixelFormat;
 import org.lwjgl.util.glu.GLU;
 
+import cpw.mods.fml.client.FMLClientHandler;
+
 public abstract class Minecraft implements Runnable
 {
     public static byte[] field_28006_b = new byte[10485760];
@@ -288,6 +290,7 @@
         this.field_6315_n = new RenderEngine(this.field_6298_C, this.field_6304_y);
         this.func_6257_q();
         this.field_6314_o = new FontRenderer(this.field_6304_y, "/font/default.png", this.field_6315_n, false);
+        FMLClientHandler.instance().onPreLoad(this);
         this.field_40007_r = new FontRenderer(this.field_6304_y, "/font/alternate.png", this.field_6315_n, false);
 
         if (this.field_6304_y.field_44018_Q != null)
@@ -330,6 +333,9 @@
         GL11.glMatrixMode(GL11.GL_PROJECTION);
         GL11.glLoadIdentity();
         GL11.glMatrixMode(GL11.GL_MODELVIEW);
+        FMLClientHandler.instance().onLoadComplete();
+        this.field_9231_Y.setup();
+        this.field_9232_X.setup();
         this.func_6250_c("Startup");
         this.field_6286_O = new OpenGlCapsChecker();
         this.field_6301_A.func_340_a(this.field_6304_y);
@@ -744,9 +750,11 @@
                     this.field_6327_b.func_6467_a(this.field_9237_P.field_1378_c);
                 }
 
+                FMLClientHandler.instance().onRenderTickStart(this.field_9237_P.field_1378_c);
                 Profiler.func_40661_c("gameRenderer");
                 this.field_9243_r.func_4136_b(this.field_9237_P.field_1378_c);
                 Profiler.func_40662_b();
+                FMLClientHandler.instance().onRenderTickEnd(this.field_9237_P.field_1378_c);
             }
 
             GL11.glFlush();
@@ -1340,6 +1348,7 @@
             this.func_28001_B();
         }
 
+        FMLClientHandler.instance().onPreWorldTick();
         Profiler.func_40663_a("stats");
         this.field_25001_G.func_27178_d();
         Profiler.func_40661_c("gui");
@@ -1730,6 +1739,7 @@
         }
 
         Profiler.func_40662_b();
+        FMLClientHandler.instance().onPostWorldTick();
         this.field_6287_N = System.currentTimeMillis();
     }
 
@@ -2233,7 +2243,7 @@
         return this.field_6322_g instanceof EntityClientPlayerMP ? ((EntityClientPlayerMP)this.field_6322_g).field_797_bg : null;
     }
 
-    public static void main(String[] p_main_0_)
+    public static void main(String[] p_main_0_) throws Exception
     {
         String var1 = null;
         String var2 = null;
@@ -2250,7 +2260,10 @@
         {
             var2 = p_main_0_[1];
         }
-
+        FMLClientHandler.instance().preGameLoad();
+    }
+    public static void fmlReentry(String var1, String var2)
+    {
         func_6269_a(var1, var2);
     }
 
