--- a/net/minecraft/client/Screenshot.java
+++ b/net/minecraft/client/Screenshot.java
@@ -56,13 +_,23 @@
          file2 = new File(file1, p_92307_);
       }
 
+      net.minecraftforge.client.event.ScreenshotEvent event = net.minecraftforge.client.ForgeHooksClient.onScreenshot(nativeimage, file2);
+      if (event.isCanceled()) {
+         p_92311_.accept(event.getCancelMessage());
+         return;
+      }
+      final File target = event.getScreenshotFile();
+
       Util.ioPool().execute(() -> {
          try {
-            nativeimage.writeToFile(file2);
+            nativeimage.writeToFile(target);
             Component component = Component.literal(file2.getName()).withStyle(ChatFormatting.UNDERLINE).withStyle((p_168608_) -> {
-               return p_168608_.withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_FILE, file2.getAbsolutePath()));
+               return p_168608_.withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_FILE, target.getAbsolutePath()));
             });
-            p_92311_.accept(Component.translatable("screenshot.success", component));
+            if (event.getResultMessage() != null)
+               p_92311_.accept(event.getResultMessage());
+            else
+               p_92311_.accept(Component.translatable("screenshot.success", component));
          } catch (Exception exception) {
             LOGGER.warn("Couldn't save screenshot", (Throwable)exception);
             p_92311_.accept(Component.translatable("screenshot.failure", exception.getMessage()));
