--- a/net/minecraft/client/Screenshot.java
+++ b/net/minecraft/client/Screenshot.java
@@ -54,14 +_,24 @@
             file2 = new File(file1, p_92307_);
         }
 
+        var event = net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenshot(nativeimage, file2);
+        if (event.isCanceled()) {
+           p_92311_.accept(event.getCancelMessage());
+           return;
+        }
+        final File target = event.getScreenshotFile();
+
         Util.ioPool()
             .execute(
                 () -> {
                     try {
-                        nativeimage.writeToFile(file2);
-                        Component component = Component.literal(file2.getName())
+                        nativeimage.writeToFile(target);
+                        Component component = Component.literal(target.getName())
                             .withStyle(ChatFormatting.UNDERLINE)
-                            .withStyle(p_168608_ -> p_168608_.withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_FILE, file2.getAbsolutePath())));
+                            .withStyle(p_168608_ -> p_168608_.withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_FILE, target.getAbsolutePath())));
+                        if (event.getResultMessage() != null)
+                            p_92311_.accept(event.getResultMessage());
+                        else
                         p_92311_.accept(Component.translatable("screenshot.success", component));
                     } catch (Exception exception) {
                         LOGGER.warn("Couldn't save screenshot", (Throwable)exception);
