--- a/net/minecraft/client/renderer/vertex/VertexFormat.java
+++ b/net/minecraft/client/renderer/vertex/VertexFormat.java
@@ -16,6 +16,7 @@
    private int field_177354_e = -1;
    private final List<Integer> field_177351_f = Lists.newArrayList();
    private int field_177352_g = -1;
+   private int hashCode;
 
    public VertexFormat(VertexFormat p_i46097_1_) {
       this();
@@ -37,6 +38,7 @@
       this.field_177351_f.clear();
       this.field_177352_g = -1;
       this.field_177353_d = 0;
+      this.hashCode = 0;
    }
 
    public VertexFormat func_181721_a(VertexFormatElement p_181721_1_) {
@@ -58,6 +60,7 @@
          }
 
          this.field_177353_d += p_181721_1_.func_177368_f();
+         this.hashCode = 0;
          return this;
       }
    }
@@ -152,9 +155,11 @@
    }
 
    public int hashCode() {
+      if (this.hashCode != 0) return this.hashCode;
       int i = this.field_177355_b.hashCode();
       i = 31 * i + this.field_177356_c.hashCode();
       i = 31 * i + this.field_177353_d;
+      this.hashCode = i;
       return i;
    }
 }
