--- a/net/minecraft/client/renderer/vertex/VertexFormatElement.java
+++ b/net/minecraft/client/renderer/vertex/VertexFormatElement.java
@@ -88,6 +_,12 @@
       this.field_177380_c.func_227901_a_(this.field_177377_d);
    }
 
+    //Forge Start
+    public int getElementCount() {
+       return field_177378_e;
+    }
+    //Forge End
+
    @OnlyIn(Dist.CLIENT)
    public static enum Type {
       FLOAT(4, "Float", 5126),
