--- ../src-base/minecraft/net/minecraft/client/renderer/vertex/VertexFormatElement.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/vertex/VertexFormatElement.java
@@ -114,13 +114,20 @@
     @SideOnly(Side.CLIENT)
     public static enum EnumType
     {
-        FLOAT(4, "Float", 5126),
-        UBYTE(1, "Unsigned Byte", 5121),
-        BYTE(1, "Byte", 5120),
-        USHORT(2, "Unsigned Short", 5123),
-        SHORT(2, "Short", 5122),
-        UINT(4, "Unsigned Int", 5125),
-        INT(4, "Int", 5124);
+        FLOAT(4, "Float", org.lwjgl.opengl.GL11.GL_FLOAT),
+        UBYTE(1, "Unsigned Byte", org.lwjgl.opengl.GL11.GL_UNSIGNED_BYTE),
+        BYTE(1, "Byte", org.lwjgl.opengl.GL11.GL_BYTE),
+        USHORT(2, "Unsigned Short", org.lwjgl.opengl.GL11.GL_UNSIGNED_SHORT),
+        SHORT(2, "Short", org.lwjgl.opengl.GL11.GL_SHORT),
+        UINT(4, "Unsigned Int", org.lwjgl.opengl.GL11.GL_UNSIGNED_INT),
+        INT(4, "Int", org.lwjgl.opengl.GL11.GL_INT);
+        // Commented for now, might be added in the future if anyone needs them
+        //HALF_FLOAT(2, "Half Float", org.lwjgl.opengl.GL30.GL_HALF_FLOAT),
+        //DOUBLE(8, "Double", org.lwjgl.opengl.GL11.GL_DOUBLE),
+        //INT_2_10_10_10_REV(4, "Int 2-10-10-10 reversed", org.lwjgl.opengl.GL33.GL_INT_2_10_10_10_REV),
+        //UINT_2_10_10_10_REV(4, "Unsigned Int 2-10-10-10 reversed", org.lwjgl.opengl.GL12.GL_UNSIGNED_INT_2_10_10_10_REV),
+        //UINT_10F_11F_11F_REV(4, "Unsigned Int 10F 11F 11F reversed", GL_UNSIGNED_INT_10F_11F_11F_REV);
+
         private final int field_177407_h;
         private final String field_177408_i;
         private final int field_177405_j;
@@ -157,9 +164,17 @@
         NORMAL("Normal"),
         COLOR("Vertex Color"),
         UV("UV"),
+        // As of 1.8 - unused in vanilla; use GENERIC for now
+        @Deprecated
         MATRIX("Bone Matrix"),
+        @Deprecated
         BLEND_WEIGHT("Blend Weight"),
-        PADDING("Padding");
+        PADDING("Padding"),
+        GENERIC("Generic Attribute");
+
+        public void preDraw(VertexFormatElement element, int stride, java.nio.ByteBuffer buffer) { net.minecraftforge.client.ForgeHooksClient.preDraw(this, element, stride, buffer); }
+        public void postDraw(VertexFormatElement element, int stride, java.nio.ByteBuffer buffer) { net.minecraftforge.client.ForgeHooksClient.postDraw(this, element, stride, buffer); }
+
         private final String field_177392_h;
 
         private static final String __OBFID = "CL_00002397";
