--- ../src-base/minecraft/net/minecraft/client/gui/GuiChat.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiChat.java
@@ -215,13 +215,14 @@
             this.mc.ingameGUI.getChatGUI().printChatMessageWithOptionalDeletion(new ChatComponentText(stringbuilder.toString()), 1);
         }
 
-        this.inputField.writeText((String)this.foundPlayerNames.get(this.autocompleteIndex++));
+        this.inputField.writeText(net.minecraft.util.EnumChatFormatting.getTextWithoutFormattingCodes((String)this.foundPlayerNames.get(this.autocompleteIndex++)));
     }
 
     private void sendAutocompleteRequest(String p_146405_1_, String p_146405_2_)
     {
         if (p_146405_1_.length() >= 1)
         {
+            net.minecraftforge.client.ClientCommandHandler.instance.autoComplete(p_146405_1_, p_146405_2_);
             BlockPos blockpos = null;
 
             if (this.mc.objectMouseOver != null && this.mc.objectMouseOver.typeOfHit == MovingObjectPosition.MovingObjectType.BLOCK)
@@ -281,6 +282,12 @@
             this.playerNamesFound = false;
             this.foundPlayerNames.clear();
 
+            String[] complete = net.minecraftforge.client.ClientCommandHandler.instance.latestAutoComplete;
+            if (complete != null)
+            {
+                p_146406_1_ = com.google.common.collect.ObjectArrays.concat(complete, p_146406_1_, String.class);
+            }
+
             for (String s : p_146406_1_)
             {
                 if (s.length() > 0)
@@ -291,6 +298,7 @@
 
             String s1 = this.inputField.getText().substring(this.inputField.func_146197_a(-1, this.inputField.getCursorPosition(), false));
             String s2 = StringUtils.getCommonPrefix(p_146406_1_);
+            s2 = net.minecraft.util.EnumChatFormatting.getTextWithoutFormattingCodes(s2);
 
             if (s2.length() > 0 && !s1.equalsIgnoreCase(s2))
             {
