--- ../src-base/minecraft/net/minecraft/client/gui/GuiUtilRenderComponents.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiUtilRenderComponents.java
@@ -73,6 +73,7 @@
                         s3 = s4;
                     }
 
+                    s3 = FontRenderer.getFormatFromString(s2) + s3; //Forge: Fix chat formatting not surviving line wrapping.
                     ChatComponentText chatcomponenttext2 = new ChatComponentText(s3);
                     chatcomponenttext2.setChatStyle(ichatcomponent1.getChatStyle().createShallowCopy());
                     list1.add(j + 1, chatcomponenttext2);
