--- a/net/minecraft/commands/arguments/EntityArgument.java
+++ b/net/minecraft/commands/arguments/EntityArgument.java
@@ -128,7 +_,7 @@
         if (contextBuilder.getSource() instanceof SharedSuggestionProvider source) {
             StringReader reader = new StringReader(builder.getInput());
             reader.setCursor(builder.getStart());
-            EntitySelectorParser parser = new EntitySelectorParser(reader, source.permissions().hasPermission(Permissions.COMMANDS_ENTITY_SELECTORS));
+            EntitySelectorParser parser = new EntitySelectorParser(reader, net.minecraftforge.common.ForgeHooks.canUseEntitySelectors(source));
 
             try {
                 parser.parse();
