--- ../src_base/minecraft/net/minecraft/scoreboard/ServerScoreboard.java
+++ ../src_work/minecraft/net/minecraft/scoreboard/ServerScoreboard.java
@@ -31,7 +31,7 @@
 
         if (this.field_96553_b.contains(par1Score.func_96645_d()))
         {
-            this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet207SetScore(par1Score, 0));
+            this.sendAll(new Packet207SetScore(par1Score, 0)); // CraftBukkit - Internal packet method
         }
 
         this.func_96551_b();
@@ -40,7 +40,7 @@
     public void func_96516_a(String par1Str)
     {
         super.func_96516_a(par1Str);
-        this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet207SetScore(par1Str));
+        this.sendAll(new Packet207SetScore(par1Str)); // CraftBukkit - Internal packet method
         this.func_96551_b();
     }
 
@@ -53,7 +53,7 @@
         {
             if (this.func_96552_h(scoreobjective1) > 0)
             {
-                this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet208SetDisplayObjective(par1, par2ScoreObjective));
+                this.sendAll(new Packet208SetDisplayObjective(par1, par2ScoreObjective)); // CraftBukkit - Internal packet method
             }
             else
             {
@@ -65,7 +65,7 @@
         {
             if (this.field_96553_b.contains(par2ScoreObjective))
             {
-                this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet208SetDisplayObjective(par1, par2ScoreObjective));
+                this.sendAll(new Packet208SetDisplayObjective(par1, par2ScoreObjective)); // CraftBukkit - Internal packet method
             }
             else
             {
@@ -76,10 +76,10 @@
         this.func_96551_b();
     }
 
-    public void func_96521_a(String par1Str, ScorePlayerTeam par2ScorePlayerTeam)
-    {
-        super.func_96521_a(par1Str, par2ScorePlayerTeam);
-        this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet209SetPlayerTeam(par2ScorePlayerTeam, Arrays.asList(new String[] {par1Str}), 3));
+    public void addPlayerToTeam(String par1Str, ScorePlayerTeam par2ScorePlayerTeam)
+    {
+        super.addPlayerToTeam(par1Str, par2ScorePlayerTeam);
+        this.sendAll(new Packet209SetPlayerTeam(par2ScorePlayerTeam, Arrays.asList(new String[] { par1Str}), 3)); // CraftBukkit - Internal packet method
         this.func_96551_b();
     }
 
@@ -90,7 +90,7 @@
     public void removePlayerFromTeam(String par1Str, ScorePlayerTeam par2ScorePlayerTeam)
     {
         super.removePlayerFromTeam(par1Str, par2ScorePlayerTeam);
-        this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet209SetPlayerTeam(par2ScorePlayerTeam, Arrays.asList(new String[] {par1Str}), 4));
+        this.sendAll(new Packet209SetPlayerTeam(par2ScorePlayerTeam, Arrays.asList(new String[] { par1Str}), 4)); // CraftBukkit - Internal packet method
         this.func_96551_b();
     }
 
@@ -106,7 +106,7 @@
 
         if (this.field_96553_b.contains(par1ScoreObjective))
         {
-            this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet206SetObjective(par1ScoreObjective, 2));
+            this.sendAll(new Packet206SetObjective(par1ScoreObjective, 2)); // CraftBukkit - Internal packet method
         }
 
         this.func_96551_b();
@@ -127,21 +127,21 @@
     public void func_96523_a(ScorePlayerTeam par1ScorePlayerTeam)
     {
         super.func_96523_a(par1ScorePlayerTeam);
-        this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet209SetPlayerTeam(par1ScorePlayerTeam, 0));
+        this.sendAll(new Packet209SetPlayerTeam(par1ScorePlayerTeam, 0)); // CraftBukkit - Internal packet method
         this.func_96551_b();
     }
 
     public void func_96538_b(ScorePlayerTeam par1ScorePlayerTeam)
     {
         super.func_96538_b(par1ScorePlayerTeam);
-        this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet209SetPlayerTeam(par1ScorePlayerTeam, 2));
+        this.sendAll(new Packet209SetPlayerTeam(par1ScorePlayerTeam, 2)); // CraftBukkit - Internal packet method
         this.func_96551_b();
     }
 
     public void func_96513_c(ScorePlayerTeam par1ScorePlayerTeam)
     {
         super.func_96513_c(par1ScorePlayerTeam);
-        this.field_96555_a.getConfigurationManager().sendPacketToAllPlayers(new Packet209SetPlayerTeam(par1ScorePlayerTeam, 1));
+        this.sendAll(new Packet209SetPlayerTeam(par1ScorePlayerTeam, 1)); // CraftBukkit - Internal packet method
         this.func_96551_b();
     }
 
@@ -190,6 +190,12 @@
         while (iterator.hasNext())
         {
             EntityPlayerMP entityplayermp = (EntityPlayerMP)iterator.next();
+
+            if (entityplayermp.getBukkitEntity().getScoreboard().getHandle() != this)
+            {
+                continue;    // CraftBukkit - Only players on this board
+            }
+
             Iterator iterator1 = list.iterator();
 
             while (iterator1.hasNext())
@@ -226,6 +232,12 @@
         while (iterator.hasNext())
         {
             EntityPlayerMP entityplayermp = (EntityPlayerMP)iterator.next();
+
+            if (entityplayermp.getBukkitEntity().getScoreboard().getHandle() != this)
+            {
+                continue;    // CraftBukkit - Only players on this board
+            }
+
             Iterator iterator1 = list.iterator();
 
             while (iterator1.hasNext())
@@ -252,4 +264,17 @@
 
         return i;
     }
+
+    // CraftBukkit start - Send to players
+    private void sendAll(Packet packet)
+    {
+        for (EntityPlayerMP entityplayermp : (List<EntityPlayerMP>) this.field_96555_a.getConfigurationManager().playerEntityList)
+        {
+            if (entityplayermp.getBukkitEntity().getScoreboard().getHandle() == this)
+            {
+                entityplayermp.playerNetServerHandler.sendPacketToPlayer(packet);
+            }
+        }
+    }
+    // CraftBukkit end
 }
