--- a/net/minecraft/client/resources/model/BakedModel.java
+++ b/net/minecraft/client/resources/model/BakedModel.java
@@ -13,7 +_,9 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface BakedModel {
+public interface BakedModel extends net.minecraftforge.client.extensions.IForgeBakedModel {
+   /**@deprecated Forge: Use {@link net.minecraftforge.client.extensions.IForgeBakedModel#getQuads(BlockState, Direction, RandomSource, net.minecraftforge.client.model.data.ModelData, net.minecraft.client.renderer.RenderType)}*/
+   @Deprecated
    List<BakedQuad> m_213637_(@Nullable BlockState p_235039_, @Nullable Direction p_235040_, RandomSource p_235041_);
 
    boolean m_7541_();
@@ -24,9 +_,13 @@
 
    boolean m_7521_();
 
+   /**@deprecated Forge: Use {@link net.minecraftforge.client.extensions.IForgeBakedModel#getParticleIcon(net.minecraftforge.client.model.data.ModelData)}*/
+   @Deprecated
    TextureAtlasSprite m_6160_();
 
-   ItemTransforms m_7442_();
+   /**@deprecated Forge: Use {@link net.minecraftforge.client.extensions.IForgeBakedModel#applyTransform(ItemTransforms.TransformType, com.mojang.blaze3d.vertex.PoseStack, boolean)} instead */
+   @Deprecated
+   default ItemTransforms m_7442_() { return ItemTransforms.f_111786_; }
 
    ItemOverrides m_7343_();
 }
