--- ../src-base/minecraft/net/minecraft/client/resources/model/ModelManager.java
+++ ../src-work/minecraft/net/minecraft/client/resources/model/ModelManager.java
@@ -24,9 +24,10 @@
 
     public void func_110549_a(IResourceManager p_110549_1_)
     {
-        ModelBakery modelbakery = new ModelBakery(p_110549_1_, this.field_174956_b, this.field_174957_c);
+        ModelBakery modelbakery = new net.minecraftforge.client.model.ModelLoader(p_110549_1_, this.field_174956_b, this.field_174957_c);
         this.field_174958_a = modelbakery.func_177570_a();
         this.field_174955_d = (IBakedModel)this.field_174958_a.func_82594_a(ModelBakery.field_177604_a);
+        net.minecraftforge.client.ForgeHooksClient.onModelBake(this, this.field_174958_a, modelbakery);
         this.field_174957_c.func_178124_c();
     }
 
