--- a/net/minecraft/item/BlockItem.java
+++ b/net/minecraft/item/BlockItem.java
@@ -36,14 +36,17 @@
    }
 
    public ActionResultType func_195939_a(ItemUseContext p_195939_1_) {
-      ActionResultType actionresulttype = this.func_195942_a(new BlockItemUseContext(p_195939_1_));
+      BlockItemUseContext ctx = new BlockItemUseContext(p_195939_1_);
+      ActionResultType actionresulttype = ActionResultType.FAIL;
+      if (!ctx.func_196012_c())
+         actionresulttype = this.func_195942_a(new net.minecraftforge.common.util.SameBlockPlaceContext(ctx));
+      if (actionresulttype != ActionResultType.SUCCESS)
+         actionresulttype = this.func_195942_a(ctx);
       return actionresulttype != ActionResultType.SUCCESS && this.func_219971_r() ? this.func_77659_a(p_195939_1_.field_196006_g, p_195939_1_.field_196001_b, p_195939_1_.field_221534_c).func_188397_a() : actionresulttype;
    }
 
    public ActionResultType func_195942_a(BlockItemUseContext p_195942_1_) {
-      if (!p_195942_1_.func_196011_b()) {
-         return ActionResultType.FAIL;
-      } else {
+      {
          BlockItemUseContext blockitemusecontext = this.func_219984_b(p_195942_1_);
          if (blockitemusecontext == null) {
             return ActionResultType.FAIL;
@@ -51,7 +54,7 @@
             BlockState blockstate = this.func_195945_b(blockitemusecontext);
             if (blockstate == null) {
                return ActionResultType.FAIL;
-            } else if (!this.func_195941_b(blockitemusecontext, blockstate)) {
+            } else if (!this.tryPlaceBlock(blockitemusecontext, blockstate)) {
                return ActionResultType.FAIL;
             } else {
                BlockPos blockpos = blockitemusecontext.func_195995_a();
@@ -69,7 +72,7 @@
                   }
                }
 
-               SoundType soundtype = blockstate1.func_215695_r();
+               SoundType soundtype = blockstate1.getSoundType(world, blockpos, p_195942_1_.func_195999_j());
                world.func_184133_a(playerentity, blockpos, this.func_219983_a(blockstate1), SoundCategory.BLOCKS, (soundtype.func_185843_a() + 1.0F) / 2.0F, soundtype.func_185847_b() * 0.8F);
                itemstack.func_190918_g(1);
                return ActionResultType.SUCCESS;
@@ -189,10 +192,25 @@
    }
 
    public Block func_179223_d() {
+      return this.getBlockRaw() == null ? null : this.getBlockRaw().delegate.get();
+   }
+
+   private Block getBlockRaw() {
       return this.field_150939_a;
    }
 
    public void func_195946_a(Map<Block, Item> p_195946_1_, Item p_195946_2_) {
       p_195946_1_.put(this.func_179223_d(), p_195946_2_);
    }
+
+   public void removeFromBlockToItemMap(Map<Block, Item> blockToItemMap, Item itemIn) {
+      blockToItemMap.remove(this.func_179223_d());
+   }
+
+   protected boolean tryPlaceBlock(BlockItemUseContext context, BlockState state) {
+      if (context.func_195991_k().canAddBlockState(context.func_195995_a(), state) && context.func_195991_k().addBlockState(context.func_195995_a(), state, 11)) return true;
+      if (context instanceof net.minecraftforge.common.util.SameBlockPlaceContext) return false;
+      if (!context.func_196012_c() && !context.func_195991_k().func_180495_p(context.func_195995_a()).func_196953_a(context)) return false;
+      return func_195941_b(context, state);
+   }
 }
