--- a/net/minecraft/item/EnumRarity.java
+++ b/net/minecraft/item/EnumRarity.java
@@ -2,7 +2,7 @@
 
 import net.minecraft.util.text.TextFormatting;
 
-public enum EnumRarity {
+public enum EnumRarity implements net.minecraftforge.common.IExtensibleEnum {
    COMMON(TextFormatting.WHITE),
    UNCOMMON(TextFormatting.YELLOW),
    RARE(TextFormatting.AQUA),
@@ -13,4 +13,8 @@
    private EnumRarity(TextFormatting p_i48837_3_) {
       this.field_77937_e = p_i48837_3_;
    }
+   
+   public static EnumRarity create(String enumName, TextFormatting p_i48837_3_) {
+       throw new IllegalStateException("Enum not extended");
+   }
 }
