--- a/net/minecraft/item/HoeItem.java
+++ b/net/minecraft/item/HoeItem.java
@@ -31,7 +31,9 @@
    public ActionResultType func_195939_a(ItemUseContext p_195939_1_) {
       World world = p_195939_1_.func_195991_k();
       BlockPos blockpos = p_195939_1_.func_195995_a();
-      if (p_195939_1_.func_196000_l() != Direction.DOWN && world.func_180495_p(blockpos.func_177984_a()).func_196958_f()) {
+      int hook = net.minecraftforge.event.ForgeEventFactory.onHoeUse(p_195939_1_);
+      if (hook != 0) return hook > 0 ? ActionResultType.SUCCESS : ActionResultType.FAIL;
+      if (p_195939_1_.func_196000_l() != Direction.DOWN && world.func_175623_d(blockpos.func_177984_a())) {
          BlockState blockstate = field_195973_b.get(world.func_180495_p(blockpos).func_177230_c());
          if (blockstate != null) {
             PlayerEntity playerentity = p_195939_1_.func_195999_j();
