--- ../src-base/minecraft/net/minecraft/item/ItemSeedFood.java
+++ ../src-work/minecraft/net/minecraft/item/ItemSeedFood.java
@@ -6,7 +6,7 @@
 import net.minecraft.util.EnumFacing;
 import net.minecraft.world.World;
 
-public class ItemSeedFood extends ItemFood
+public class ItemSeedFood extends ItemFood implements net.minecraftforge.common.IPlantable
 {
     private Block field_150908_b;
     private Block field_82809_c;
@@ -28,7 +28,7 @@
         {
             return false;
         }
-        else if (p_180614_3_.func_180495_p(p_180614_4_).func_177230_c() == this.field_82809_c && p_180614_3_.func_175623_d(p_180614_4_.func_177984_a()))
+        else if (p_180614_3_.func_180495_p(p_180614_4_).func_177230_c().canSustainPlant(p_180614_3_, p_180614_4_, EnumFacing.UP, this) && p_180614_3_.func_175623_d(p_180614_4_.func_177984_a()))
         {
             p_180614_3_.func_175656_a(p_180614_4_.func_177984_a(), this.field_150908_b.func_176223_P());
             --p_180614_1_.field_77994_a;
@@ -39,4 +39,16 @@
             return false;
         }
     }
+
+    @Override
+    public net.minecraftforge.common.EnumPlantType getPlantType(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return net.minecraftforge.common.EnumPlantType.Crop;
+    }
+
+    @Override
+    public net.minecraft.block.state.IBlockState getPlant(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return this.field_150908_b.func_176223_P();
+    }
 }
