--- ../src-base/minecraft/net/minecraft/item/ItemBed.java
+++ ../src-work/minecraft/net/minecraft/item/ItemBed.java
@@ -62,7 +62,7 @@
                     p_180614_2_.func_180501_a(blockpos, iblockstate2.func_177226_a(BlockBed.field_176472_a, BlockBed.EnumPartType.HEAD), 10);
                     p_180614_2_.func_175722_b(p_180614_3_, block, false);
                     p_180614_2_.func_175722_b(blockpos, iblockstate1.func_177230_c(), false);
-                    SoundType soundtype = iblockstate2.func_177230_c().func_185467_w();
+                    SoundType soundtype = iblockstate2.func_177230_c().getSoundType(iblockstate2, p_180614_2_, p_180614_3_, p_180614_1_);
                     p_180614_2_.func_184133_a((EntityPlayer)null, p_180614_3_, soundtype.func_185841_e(), SoundCategory.BLOCKS, (soundtype.func_185843_a() + 1.0F) / 2.0F, soundtype.func_185847_b() * 0.8F);
                     itemstack.func_190918_g(1);
                     return EnumActionResult.SUCCESS;
