--- ../src-base/minecraft/net/minecraft/item/ItemLilyPad.java
+++ ../src-work/minecraft/net/minecraft/item/ItemLilyPad.java
@@ -48,8 +48,17 @@
                 BlockPos blockpos1 = blockpos.func_177984_a();
                 IBlockState iblockstate = p_77659_1_.func_180495_p(blockpos);
 
-                if (iblockstate.func_185904_a() == Material.field_151586_h && ((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue() == 0 && p_77659_1_.func_175623_d(blockpos1))
+                if (iblockstate.func_185904_a().getCanWaterPlants() && ((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue() == 0 && p_77659_1_.func_175623_d(blockpos1))
                 {
+                    // special case for handling block placement with water lilies
+                    net.minecraftforge.common.util.BlockSnapshot blocksnapshot = net.minecraftforge.common.util.BlockSnapshot.getBlockSnapshot(p_77659_1_, blockpos1);
+                    p_77659_1_.func_175656_a(blockpos1, Blocks.field_150392_bi.func_176223_P());
+                    if (net.minecraftforge.event.ForgeEventFactory.onPlayerBlockPlace(p_77659_2_, blocksnapshot, net.minecraft.util.EnumFacing.UP, p_77659_3_).isCanceled())
+                    {
+                        blocksnapshot.restore(true, false);
+                        return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
+                    }
+
                     p_77659_1_.func_180501_a(blockpos1, Blocks.field_150392_bi.func_176223_P(), 11);
 
                     if (p_77659_2_ instanceof EntityPlayerMP)
