--- ../src-base/minecraft/net/minecraft/item/ItemMap.java
+++ ../src-work/minecraft/net/minecraft/item/ItemMap.java
@@ -57,7 +57,7 @@
             mapdata = new MapData(s);
             mapdata.field_76197_d = 3;
             mapdata.func_176054_a((double)p_77873_2_.func_72912_H().func_76079_c(), (double)p_77873_2_.func_72912_H().func_76074_e(), mapdata.field_76197_d);
-            mapdata.field_76200_c = (byte)p_77873_2_.field_73011_w.func_177502_q();
+            mapdata.field_76200_c = p_77873_2_.field_73011_w.func_177502_q();
             mapdata.func_76185_a();
             p_77873_2_.func_72823_a(s, mapdata);
         }
