--- ../src-base/minecraft/net/minecraft/item/ItemMonsterPlacer.java
+++ ../src-work/minecraft/net/minecraft/item/ItemMonsterPlacer.java
@@ -89,7 +89,7 @@
             p_180614_4_ = p_180614_4_.func_177972_a(p_180614_6_);
             double d0 = 0.0D;
 
-            if (p_180614_6_ == EnumFacing.UP && iblockstate instanceof BlockFence)
+            if (p_180614_6_ == EnumFacing.UP && iblockstate.func_177230_c() instanceof BlockFence) //Forge: Fix Vanilla bug comparing state instead of block
             {
                 d0 = 0.5D;
             }
