--- ../src-base/minecraft/net/minecraft/item/ItemFlintAndSteel.java
+++ ../src-work/minecraft/net/minecraft/item/ItemFlintAndSteel.java
@@ -27,7 +27,7 @@
         }
         else
         {
-            if (worldIn.getBlockState(pos).getBlock().getMaterial() == Material.air)
+            if (worldIn.isAirBlock(pos))
             {
                 worldIn.playSoundEffect((double)pos.getX() + 0.5D, (double)pos.getY() + 0.5D, (double)pos.getZ() + 0.5D, "fire.ignite", 1.0F, itemRand.nextFloat() * 0.4F + 0.8F);
                 worldIn.setBlockState(pos, Blocks.fire.getDefaultState());
