--- ../src-base/minecraft/net/minecraft/item/ItemSkull.java
+++ ../src-work/minecraft/net/minecraft/item/ItemSkull.java
@@ -42,13 +42,18 @@
         }
         else
         {
+            if (p_180614_3_.func_180495_p(p_180614_4_).func_177230_c().func_176200_f(p_180614_3_, p_180614_4_))
+            {
+                p_180614_6_ = EnumFacing.UP;
+                p_180614_4_ = p_180614_4_.func_177977_b();
+            }
             IBlockState iblockstate = p_180614_3_.func_180495_p(p_180614_4_);
             Block block = iblockstate.func_177230_c();
             boolean flag = block.func_176200_f(p_180614_3_, p_180614_4_);
 
             if (!flag)
             {
-                if (!p_180614_3_.func_180495_p(p_180614_4_).func_185904_a().func_76220_a())
+                if (!p_180614_3_.func_180495_p(p_180614_4_).func_185904_a().func_76220_a() && !p_180614_3_.isSideSolid(p_180614_4_, p_180614_6_, true))
                 {
                     return EnumActionResult.FAIL;
                 }
