--- ../src_base/minecraft/net/minecraft/item/ItemHoe.java
+++ ../src_work/minecraft/net/minecraft/item/ItemHoe.java
@@ -6,6 +6,9 @@
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.world.World;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event.Result;
+import net.minecraftforge.event.entity.player.UseHoeEvent;
 
 public class ItemHoe extends Item
 {
@@ -32,10 +35,22 @@
         }
         else
         {
+            UseHoeEvent event = new UseHoeEvent(par2EntityPlayer, par1ItemStack, par3World, par4, par5, par6);
+            if (MinecraftForge.EVENT_BUS.post(event))
+            {
+                return false;
+            }
+
+            if (event.getResult() == Result.ALLOW)
+            {
+                par1ItemStack.damageItem(1, par2EntityPlayer);
+                return true;
+            }
+
             int i1 = par3World.getBlockId(par4, par5, par6);
-            int j1 = par3World.getBlockId(par4, par5 + 1, par6);
+            boolean air = par3World.isAirBlock(par4, par5 + 1, par6);
 
-            if (par7 != 0 && j1 == 0 && (i1 == Block.grass.blockID || i1 == Block.dirt.blockID))
+            if (par7 != 0 && air && (i1 == Block.grass.blockID || i1 == Block.dirt.blockID))
             {
                 Block block = Block.tilledField;
                 par3World.playSoundEffect((double)((float)par4 + 0.5F), (double)((float)par5 + 0.5F), (double)((float)par6 + 0.5F), block.stepSound.getStepSound(), (block.stepSound.getVolume() + 1.0F) / 2.0F, block.stepSound.getPitch() * 0.8F);
