--- a/net/minecraft/item/ItemBow.java
+++ b/net/minecraft/item/ItemBow.java
@@ -22,7 +22,7 @@
          if (p_210310_2_ == null) {
             return 0.0F;
          } else {
-            return p_210310_2_.func_184607_cu().func_77973_b() != Items.field_151031_f ? 0.0F : (float)(p_210310_0_.func_77988_m() - p_210310_2_.func_184605_cv()) / 20.0F;
+            return !(p_210310_2_.func_184607_cu().func_77973_b() instanceof ItemBow) ? 0.0F : (float)(p_210310_0_.func_77988_m() - p_210310_2_.func_184605_cv()) / 20.0F;
          }
       });
       this.func_185043_a(new ResourceLocation("pulling"), (p_210309_0_, p_210309_1_, p_210309_2_) -> {
@@ -56,18 +56,23 @@
          EntityPlayer entityplayer = (EntityPlayer)p_77615_3_;
          boolean flag = entityplayer.field_71075_bZ.field_75098_d || EnchantmentHelper.func_77506_a(Enchantments.field_185312_x, p_77615_1_) > 0;
          ItemStack itemstack = this.func_185060_a(entityplayer);
+
+         int i = this.func_77626_a(p_77615_1_) - p_77615_4_;
+         i = net.minecraftforge.event.ForgeEventFactory.onArrowLoose(p_77615_1_, p_77615_2_, entityplayer, i, !itemstack.func_190926_b() || flag);
+         if (i < 0) return;
+
          if (!itemstack.func_190926_b() || flag) {
             if (itemstack.func_190926_b()) {
                itemstack = new ItemStack(Items.field_151032_g);
             }
 
-            int i = this.func_77626_a(p_77615_1_) - p_77615_4_;
             float f = func_185059_b(i);
             if (!((double)f < 0.1D)) {
-               boolean flag1 = flag && itemstack.func_77973_b() == Items.field_151032_g;
+               boolean flag1 = entityplayer.field_71075_bZ.field_75098_d || (itemstack.func_77973_b() instanceof ItemArrow && ((ItemArrow) itemstack.func_77973_b()).isInfinite(itemstack, p_77615_1_, entityplayer));
                if (!p_77615_2_.field_72995_K) {
                   ItemArrow itemarrow = (ItemArrow)(itemstack.func_77973_b() instanceof ItemArrow ? itemstack.func_77973_b() : Items.field_151032_g);
                   EntityArrow entityarrow = itemarrow.func_200887_a(p_77615_2_, itemstack, entityplayer);
+                  entityarrow = this.customizeArrow(entityarrow);
                   entityarrow.func_184547_a(entityplayer, entityplayer.field_70125_A, entityplayer.field_70177_z, 0.0F, f * 3.0F, 1.0F);
                   if (f == 1.0F) {
                      entityarrow.func_70243_d(true);
@@ -130,6 +135,10 @@
    public ActionResult<ItemStack> func_77659_a(World p_77659_1_, EntityPlayer p_77659_2_, EnumHand p_77659_3_) {
       ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
       boolean flag = !this.func_185060_a(p_77659_2_).func_190926_b();
+
+      ActionResult<ItemStack> ret = net.minecraftforge.event.ForgeEventFactory.onArrowNock(itemstack, p_77659_1_, p_77659_2_, p_77659_3_, flag);
+      if (ret != null) return ret;
+
       if (!p_77659_2_.field_71075_bZ.field_75098_d && !flag) {
          return flag ? new ActionResult<>(EnumActionResult.PASS, itemstack) : new ActionResult<>(EnumActionResult.FAIL, itemstack);
       } else {
@@ -141,4 +150,8 @@
    public int func_77619_b() {
       return 1;
    }
+
+   public EntityArrow customizeArrow(EntityArrow arrow) {
+      return arrow;
+   }
 }
