--- ../src-base/minecraft/net/minecraft/item/ItemBlockSpecial.java
+++ ../src-work/minecraft/net/minecraft/item/ItemBlockSpecial.java
@@ -55,8 +55,11 @@
                     iblockstate1.func_177230_c().func_180633_a(p_180614_3_, p_180614_4_, iblockstate1, p_180614_2_, p_180614_1_);
                 }
 
-                SoundType soundtype = this.field_150935_a.func_185467_w();
-                p_180614_3_.func_184133_a(p_180614_2_, p_180614_4_, soundtype.func_185841_e(), SoundCategory.BLOCKS, (soundtype.func_185843_a() + 1.0F) / 2.0F, soundtype.func_185847_b() * 0.8F);
+                Block block1 = iblockstate1.func_177230_c();
+                net.minecraft.util.SoundEvent placeSound = block1.getSoundEvent(p_180614_3_, p_180614_4_, p_180614_2_, net.minecraftforge.common.EnumSoundEventType.BLOCK_PLACE);
+                float volume = block1.getSoundVolume(p_180614_3_, p_180614_4_, p_180614_2_);
+                float pitch = block1.getSoundPitch(p_180614_3_, p_180614_4_, p_180614_2_);
+                p_180614_3_.func_184133_a(p_180614_2_, p_180614_4_, placeSound, SoundCategory.BLOCKS, (volume + 1.0F) / 2.0F, pitch * 0.8F);
                 --p_180614_1_.field_77994_a;
                 return EnumActionResult.SUCCESS;
             }
