--- ../src-base/minecraft/net/minecraft/item/ItemSkull.java
+++ ../src-work/minecraft/net/minecraft/item/ItemSkull.java
@@ -45,13 +45,18 @@
         }
         else
         {
+            if (p_180614_2_.func_180495_p(p_180614_3_).func_177230_c().func_176200_f(p_180614_2_, p_180614_3_))
+            {
+                p_180614_5_ = EnumFacing.UP;
+                p_180614_3_ = p_180614_3_.func_177977_b();
+            }
             IBlockState iblockstate = p_180614_2_.func_180495_p(p_180614_3_);
             Block block = iblockstate.func_177230_c();
             boolean flag = block.func_176200_f(p_180614_2_, p_180614_3_);
 
             if (!flag)
             {
-                if (!p_180614_2_.func_180495_p(p_180614_3_).func_185904_a().func_76220_a())
+                if (!p_180614_2_.func_180495_p(p_180614_3_).func_185904_a().func_76220_a() && !p_180614_2_.isSideSolid(p_180614_3_, p_180614_5_, true))
                 {
                     return EnumActionResult.FAIL;
                 }
@@ -128,7 +133,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_150895_a(CreativeTabs p_150895_1_, NonNullList<ItemStack> p_150895_2_)
     {
         if (this.func_194125_a(p_150895_1_))
