--- ../src_base/minecraft/net/minecraft/item/ItemFlintAndSteel.java
+++ ../src_work/minecraft/net/minecraft/item/ItemFlintAndSteel.java
@@ -4,6 +4,10 @@
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.world.World;
+import net.minecraftforge.common.BlockSnapshot;
+import net.minecraftforge.event.ForgeEventFactory;
+import net.minecraftforge.event.block.BlockIgniteEvent.IgniteCause;
+import net.minecraftforge.event.block.BlockPlaceEvent;
 
 public class ItemFlintAndSteel extends Item
 {
@@ -21,6 +25,8 @@
      */
     public boolean onItemUse(ItemStack par1ItemStack, EntityPlayer par2EntityPlayer, World par3World, int par4, int par5, int par6, int par7, float par8, float par9, float par10)
     {
+        int clickedX = par4, clickedY = par5, clickedZ = par6; // CraftBukkit
+
         if (par7 == 0)
         {
             --par5;
@@ -57,16 +63,18 @@
         }
         else
         {
-            int i1 = par3World.getBlockId(par4, par5, par6);
-
-            if (i1 == 0)
+            if (par3World.isAirBlock(par4, par5, par6))
             {
                 par3World.playSoundEffect((double)par4 + 0.5D, (double)par5 + 0.5D, (double)par6 + 0.5D, "fire.ignite", 1.0F, itemRand.nextFloat() * 0.4F + 0.8F);
-                par3World.setBlock(par4, par5, par6, Block.fire.blockID);
+                ItemBlock itemblock = (ItemBlock)Item.itemsList[Block.fire.blockID];
+                if (!itemblock.placeBlockAt(par3World, par2EntityPlayer, null, par4, par5, par6, par7, Block.fire.blockID, par3World.getBlockMetadata(par4, par5, par6), clickedX, clickedY, clickedZ))
+                {
+                    return false;
+                }
             }
 
             par1ItemStack.damageItem(1, par2EntityPlayer);
             return true;
         }
     }
-}
+}