--- ../src-base/minecraft/net/minecraft/item/ItemMap.java
+++ ../src-work/minecraft/net/minecraft/item/ItemMap.java
@@ -59,7 +59,7 @@
             mapdata = new MapData(s);
             mapdata.scale = 3;
             mapdata.func_176054_a((double)worldIn.getWorldInfo().getSpawnX(), (double)worldIn.getWorldInfo().getSpawnZ(), mapdata.scale);
-            mapdata.dimension = (byte)worldIn.provider.getDimensionId();
+            mapdata.dimension = worldIn.provider.getDimensionId();
             mapdata.markDirty();
             worldIn.setItemData(s, mapdata);
         }
