--- ../src_base/minecraft/net/minecraft/item/ItemDye.java
+++ ../src_work/minecraft/net/minecraft/item/ItemDye.java
@@ -21,6 +21,12 @@
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
 
+import net.minecraftforge.common.FakePlayerFactory;
+import net.minecraftforge.common.ForgeHooks;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event.Result;
+import net.minecraftforge.event.entity.player.BonemealEvent;
+
 public class ItemDye extends Item
 {
     /** List of dye color names */
@@ -141,7 +147,27 @@
 
     public static boolean func_96604_a(ItemStack par0ItemStack, World par1World, int par2, int par3, int par4)
     {
+        return applyBonemeal(par0ItemStack, par1World, par2, par3, par4, FakePlayerFactory.getMinecraft(par1World));
+    }
+
+    public static boolean applyBonemeal(ItemStack par0ItemStack, World par1World, int par2, int par3, int par4, EntityPlayer player)
+    {
         int l = par1World.getBlockId(par2, par3, par4);
+
+        BonemealEvent event = new BonemealEvent(player, par1World, l, par2, par3, par4);
+        if (MinecraftForge.EVENT_BUS.post(event))
+        {
+            return false;
+        }
+
+        if (event.getResult() == Result.ALLOW)
+        {
+            if (!par1World.isRemote)
+            {
+                par0ItemStack.stackSize--;
+            }
+            return true;
+        }
 
         if (l == Block.sapling.blockID)
         {
@@ -244,16 +270,9 @@
                                             par1World.setBlockAndMetadataWithNotify(j1, k1, l1, Block.tallGrass.blockID, 1, 3);
                                         }
                                     }
-                                    else if (itemRand.nextInt(3) != 0)
+                                    else
                                     {
-                                        if (Block.plantYellow.canBlockStay(par1World, j1, k1, l1))
-                                        {
-                                            par1World.func_94575_c(j1, k1, l1, Block.plantYellow.blockID);
-                                        }
-                                    }
-                                    else if (Block.plantRed.canBlockStay(par1World, j1, k1, l1))
-                                    {
-                                        par1World.func_94575_c(j1, k1, l1, Block.plantRed.blockID);
+                                        ForgeHooks.plantGrass(par1World, j1, k1, l1);
                                     }
                                 }
                             }
