--- ../src_base/minecraft/net/minecraft/item/ItemDye.java
+++ ../src_work/minecraft/net/minecraft/item/ItemDye.java
@@ -21,6 +21,12 @@
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
 
+import net.minecraftforge.common.FakePlayerFactory;
+import net.minecraftforge.common.ForgeHooks;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event.Result;
+import net.minecraftforge.event.entity.player.BonemealEvent;
+
 public class ItemDye extends Item
 {
     /** List of dye color names */
@@ -73,7 +79,7 @@
         {
             if (par1ItemStack.getItemDamage() == 15)
             {
-                if (func_96604_a(par1ItemStack, par3World, par4, par5, par6))
+                if (applyBonemeal(par1ItemStack, par3World, par4, par5, par6, par2EntityPlayer))
                 {
                     if (!par3World.isRemote)
                     {
@@ -141,7 +147,27 @@
 
     public static boolean func_96604_a(ItemStack par0ItemStack, World par1World, int par2, int par3, int par4)
     {
+        return applyBonemeal(par0ItemStack, par1World, par2, par3, par4, FakePlayerFactory.getMinecraft(par1World));
+    }
+
+    public static boolean applyBonemeal(ItemStack par0ItemStack, World par1World, int par2, int par3, int par4, EntityPlayer player)
+    {
         int l = par1World.getBlockId(par2, par3, par4);
+
+        BonemealEvent event = new BonemealEvent(player, par1World, l, par2, par3, par4);
+        if (MinecraftForge.EVENT_BUS.post(event))
+        {
+            return false;
+        }
+
+        if (event.getResult() == Result.ALLOW)
+        {
+            if (!par1World.isRemote)
+            {
+                par0ItemStack.stackSize--;
+            }
+            return true;
+        }
 
         if (l == Block.sapling.blockID)
         {
@@ -244,16 +270,9 @@
                                             par1World.setBlock(j1, k1, l1, Block.tallGrass.blockID, 1, 3);
                                         }
                                     }
-                                    else if (itemRand.nextInt(3) != 0)
+                                    else
                                     {
-                                        if (Block.plantYellow.canBlockStay(par1World, j1, k1, l1))
-                                        {
-                                            par1World.setBlock(j1, k1, l1, Block.plantYellow.blockID);
-                                        }
-                                    }
-                                    else if (Block.plantRed.canBlockStay(par1World, j1, k1, l1))
-                                    {
-                                        par1World.setBlock(j1, k1, l1, Block.plantRed.blockID);
+                                        ForgeHooks.plantGrass(par1World, j1, k1, l1);
                                     }
                                 }
                             }
@@ -318,6 +337,16 @@
                 par0World.spawnParticle("happyVillager", (double)((float)par1 + itemRand.nextFloat()), (double)par2 + (double)itemRand.nextFloat() * block.getBlockBoundsMaxY(), (double)((float)par3 + itemRand.nextFloat()), d0, d1, d2);
             }
         }
+        else
+        {
+            for (int j1 = 0; j1 < par4; ++j1)
+            {
+                double d0 = itemRand.nextGaussian() * 0.02D;
+                double d1 = itemRand.nextGaussian() * 0.02D;
+                double d2 = itemRand.nextGaussian() * 0.02D;
+                par0World.spawnParticle("happyVillager", (double)((float)par1 + itemRand.nextFloat()), (double)par2 + (double)itemRand.nextFloat() * 1.0f, (double)((float)par3 + itemRand.nextFloat()), d0, d1, d2);
+            }
+        }
     }
 
     /**
