--- ../src-base/minecraft/net/minecraft/item/ItemRecord.java
+++ ../src-work/minecraft/net/minecraft/item/ItemRecord.java
@@ -76,4 +76,15 @@
     {
         return (ItemRecord)field_150928_b.get(p_150926_0_);
     }
+
+    /**
+     * Retrieves the resource location of the sound to play for this record.
+     *
+     * @param name The name of the record to play
+     * @return The resource location for the audio, null to use default.
+     */
+    public net.minecraft.util.ResourceLocation getRecordResource(String name)
+    {
+        return new net.minecraft.util.ResourceLocation(name);
+    }
 }
