--- ../src-base/minecraft/net/minecraft/item/ItemArmor.java
+++ ../src-work/minecraft/net/minecraft/item/ItemArmor.java
@@ -43,7 +43,7 @@
         }
     };
     public final EntityEquipmentSlot field_77881_a;
-    public final int field_77879_b;
+    public final int armorAmount;
     public final int field_77880_c;
     private final ItemArmor.ArmorMaterial field_77878_bZ;
 
@@ -83,7 +83,7 @@
         this.field_77878_bZ = p_i46750_1_;
         this.field_77881_a = p_i46750_3_;
         this.field_77880_c = p_i46750_2_;
-        this.field_77879_b = p_i46750_1_.func_78044_b(p_i46750_3_);
+        this.armorAmount = p_i46750_1_.getArmorAmount(p_i46750_3_);
         this.func_77656_e(p_i46750_1_.func_78046_a(p_i46750_3_));
         this.field_77777_bU = 1;
         this.func_77637_a(CreativeTabs.field_78037_j);
@@ -216,7 +216,7 @@
 
         if (p_111205_1_ == this.field_77881_a)
         {
-            multimap.put(SharedMonsterAttributes.field_188791_g.func_111108_a(), new AttributeModifier(field_185084_n[p_111205_1_.func_188454_b()], "Armor modifier", (double)this.field_77879_b, 0));
+            multimap.put(SharedMonsterAttributes.field_188791_g.func_111108_a(), new AttributeModifier(field_185084_n[p_111205_1_.func_188454_b()], "Armor modifier", (double)this.armorAmount, 0));
         }
 
         return multimap;
@@ -232,15 +232,17 @@
 
         private final String field_179243_f;
         private final int field_78048_f;
-        private final int[] field_78049_g;
+        private final int[] armorAmountArray;
         private final int field_78055_h;
         private final SoundEvent field_185020_j;
+        //Added by forge for custom Armor materials.
+        public Item customCraftingMaterial = null;
 
         private ArmorMaterial(String p_i46796_3_, int p_i46796_4_, int[] p_i46796_5_, int p_i46796_6_, SoundEvent p_i46796_7_)
         {
             this.field_179243_f = p_i46796_3_;
             this.field_78048_f = p_i46796_4_;
-            this.field_78049_g = p_i46796_5_;
+            this.armorAmountArray = p_i46796_5_;
             this.field_78055_h = p_i46796_6_;
             this.field_185020_j = p_i46796_7_;
         }
@@ -250,9 +252,9 @@
             return ItemArmor.field_77882_bY[p_78046_1_.func_188454_b()] * this.field_78048_f;
         }
 
-        public int func_78044_b(EntityEquipmentSlot p_78044_1_)
+        public int getArmorAmount(EntityEquipmentSlot armorType)
         {
-            return this.field_78049_g[p_78044_1_.func_188454_b()];
+            return this.armorAmountArray[armorType.func_188454_b()];
         }
 
         public int func_78045_a()
@@ -267,7 +269,15 @@
 
         public Item func_151685_b()
         {
-            return this == LEATHER ? Items.field_151116_aA : (this == CHAIN ? Items.field_151042_j : (this == GOLD ? Items.field_151043_k : (this == IRON ? Items.field_151042_j : (this == DIAMOND ? Items.field_151045_i : null))));
+            switch (this)
+            {
+                case LEATHER: return Items.field_151116_aA;
+                case CHAIN:   return Items.field_151042_j;
+                case GOLD:    return Items.field_151043_k;
+                case IRON:    return Items.field_151042_j;
+                case DIAMOND: return Items.field_151045_i;
+                default:      return customCraftingMaterial;
+            }
         }
 
         @SideOnly(Side.CLIENT)
