--- ../src_base/minecraft/net/minecraft/item/ItemBlock.java
+++ ../src_work/minecraft/net/minecraft/item/ItemBlock.java
@@ -19,6 +19,7 @@
         super(par1);
         this.blockID = par1 + 256;
         this.setIconIndex(Block.blocksList[par1 + 256].getBlockTextureFromSide(2));
+        isDefaultTexture = Block.blocksList[par1 + 256].isDefaultTexture;
     }
 
     /**
@@ -41,7 +42,8 @@
         {
             par7 = 1;
         }
-        else if (var11 != Block.vine.blockID && var11 != Block.tallGrass.blockID && var11 != Block.deadBush.blockID)
+        else if (var11 != Block.vine.blockID && var11 != Block.tallGrass.blockID && var11 != Block.deadBush.blockID
+                && (Block.blocksList[var11] == null || !Block.blocksList[var11].isBlockReplaceable(par3World, par4, par5, par6)))
         {
             if (par7 == 0)
             {
@@ -92,14 +94,8 @@
             int var13 = this.getMetadata(par1ItemStack.getItemDamage());
             int var14 = Block.blocksList[this.blockID].onBlockPlaced(par3World, par4, par5, par6, par7, par8, par9, par10, var13);
 
-            if (par3World.setBlockAndMetadataWithNotify(par4, par5, par6, this.blockID, var14))
-            {
-                if (par3World.getBlockId(par4, par5, par6) == this.blockID)
-                {
-                    Block.blocksList[this.blockID].onBlockPlacedBy(par3World, par4, par5, par6, par2EntityPlayer);
-                    Block.blocksList[this.blockID].onPostBlockPlaced(par3World, par4, par5, par6, var14);
-                }
-
+            if (placeBlockAt(par1ItemStack, par2EntityPlayer, par3World, par4, par5, par6, par7, par8, par9, par10, var14))
+            {
                 par3World.playSoundEffect((double)((float)par4 + 0.5F), (double)((float)par5 + 0.5F), (double)((float)par6 + 0.5F), var12.stepSound.getPlaceSound(), (var12.stepSound.getVolume() + 1.0F) / 2.0F, var12.stepSound.getPitch() * 0.8F);
                 --par1ItemStack.stackSize;
             }
@@ -125,7 +121,8 @@
         {
             par5 = 1;
         }
-        else if (var8 != Block.vine.blockID && var8 != Block.tallGrass.blockID && var8 != Block.deadBush.blockID)
+        else if (var8 != Block.vine.blockID && var8 != Block.tallGrass.blockID && var8 != Block.deadBush.blockID
+                && (Block.blocksList[var8] == null || !Block.blocksList[var8].isBlockReplaceable(par1World, par2, par3, par4)))
         {
             if (par5 == 0)
             {
@@ -190,4 +187,28 @@
     {
         Block.blocksList[this.blockID].getSubBlocks(par1, par2CreativeTabs, par3List);
     }
+
+    /**
+     * Called to actually place the block, after the location is determined
+     * and all permission checks have been made.
+     *
+     * @param stack The item stack that was used to place the block. This can be changed inside the method.
+     * @param player The player who is placing the block. Can be null if the block is not being placed by a player.
+     * @param side The side the player (or machine) right-clicked on.
+     */
+    public boolean placeBlockAt(ItemStack stack, EntityPlayer player, World world, int x, int y, int z, int side, float hitX, float hitY, float hitZ, int metadata)
+    {
+       if (!world.setBlockAndMetadataWithNotify(x, y, z, this.blockID, metadata))
+       {
+               return false;
+       }
+
+       if (world.getBlockId(x, y, z) == this.blockID)
+       {
+           Block.blocksList[this.blockID].onBlockPlacedBy(world, x, y, z, player);
+           Block.blocksList[this.blockID].onPostBlockPlaced(world, x, y, z, metadata);
+       }
+
+       return true;
+    }
 }
