--- ../src_base/minecraft/net/minecraft/item/ItemFishingRod.java
+++ ../src_work/minecraft/net/minecraft/item/ItemFishingRod.java
@@ -8,6 +8,7 @@
 import net.minecraft.entity.projectile.EntityFishHook;
 import net.minecraft.util.Icon;
 import net.minecraft.world.World;
+import org.bukkit.event.player.PlayerFishEvent; // CraftBukkit
 
 public class ItemFishingRod extends Item
 {
@@ -56,11 +57,22 @@
         }
         else
         {
+            // CraftBukkit start
+            EntityFishHook hook = new EntityFishHook(par2World, par3EntityPlayer);
+            PlayerFishEvent playerFishEvent = new PlayerFishEvent((org.bukkit.entity.Player) par3EntityPlayer.getBukkitEntity(), null, (org.bukkit.entity.Fish) hook.getBukkitEntity(), PlayerFishEvent.State.FISHING);
+            par2World.getServer().getPluginManager().callEvent(playerFishEvent);
+
+            if (playerFishEvent.isCancelled())
+            {
+                return par1ItemStack;
+            }
+
+            // CraftBukkit end
             par2World.playSoundAtEntity(par3EntityPlayer, "random.bow", 0.5F, 0.4F / (itemRand.nextFloat() * 0.4F + 0.8F));
 
             if (!par2World.isRemote)
             {
-                par2World.spawnEntityInWorld(new EntityFishHook(par2World, par3EntityPlayer));
+                par2World.spawnEntityInWorld(hook); // CraftBukkit - moved creation up
             }
 
             par3EntityPlayer.swingItem();
