--- ../src-base/minecraft/net/minecraft/item/ItemDoor.java
+++ ../src-work/minecraft/net/minecraft/item/ItemDoor.java
@@ -46,7 +46,7 @@
                 int j = enumfacing.func_82599_e();
                 boolean flag = i < 0 && p_180614_9_ < 0.5F || i > 0 && p_180614_9_ > 0.5F || j < 0 && p_180614_7_ > 0.5F || j > 0 && p_180614_7_ < 0.5F;
                 func_179235_a(p_180614_3_, p_180614_4_, enumfacing, this.field_179236_a, flag);
-                SoundType soundtype = this.field_179236_a.func_185467_w();
+                SoundType soundtype = p_180614_3_.func_180495_p(p_180614_4_).func_177230_c().getSoundType(p_180614_3_.func_180495_p(p_180614_4_), p_180614_3_, p_180614_4_, p_180614_2_);
                 p_180614_3_.func_184133_a(p_180614_2_, p_180614_4_, soundtype.func_185841_e(), SoundCategory.BLOCKS, (soundtype.func_185843_a() + 1.0F) / 2.0F, soundtype.func_185847_b() * 0.8F);
                 --p_180614_1_.field_77994_a;
                 return EnumActionResult.SUCCESS;
@@ -62,8 +62,8 @@
     {
         BlockPos blockpos = p_179235_1_.func_177972_a(p_179235_2_.func_176746_e());
         BlockPos blockpos1 = p_179235_1_.func_177972_a(p_179235_2_.func_176735_f());
-        int i = (p_179235_0_.func_180495_p(blockpos1).func_185915_l() ? 1 : 0) + (p_179235_0_.func_180495_p(blockpos1.func_177984_a()).func_185915_l() ? 1 : 0);
-        int j = (p_179235_0_.func_180495_p(blockpos).func_185915_l() ? 1 : 0) + (p_179235_0_.func_180495_p(blockpos.func_177984_a()).func_185915_l() ? 1 : 0);
+        int i = (p_179235_0_.func_180495_p(blockpos1).isNormalCube(p_179235_0_, blockpos1) ? 1 : 0) + (p_179235_0_.func_180495_p(blockpos1.func_177984_a()).isNormalCube(p_179235_0_, blockpos1.func_177984_a()) ? 1 : 0);
+        int j = (p_179235_0_.func_180495_p(blockpos).isNormalCube(p_179235_0_, blockpos) ? 1 : 0) + (p_179235_0_.func_180495_p(blockpos.func_177984_a()).isNormalCube(p_179235_0_, blockpos.func_177984_a()) ? 1 : 0);
         boolean flag = p_179235_0_.func_180495_p(blockpos1).func_177230_c() == p_179235_3_ || p_179235_0_.func_180495_p(blockpos1.func_177984_a()).func_177230_c() == p_179235_3_;
         boolean flag1 = p_179235_0_.func_180495_p(blockpos).func_177230_c() == p_179235_3_ || p_179235_0_.func_180495_p(blockpos.func_177984_a()).func_177230_c() == p_179235_3_;
 
