--- ../src_base/minecraft/net/minecraft/item/ItemBucket.java
+++ ../src_work/minecraft/net/minecraft/item/ItemBucket.java
@@ -9,6 +9,10 @@
 import net.minecraft.util.MovingObjectPosition;
 import net.minecraft.world.World;
 
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event;
+import net.minecraftforge.event.entity.player.FillBucketEvent;
+
 public class ItemBucket extends Item
 {
     /** field for checking if the bucket has been filled. */
@@ -40,6 +44,32 @@
         }
         else
         {
+            FillBucketEvent event = new FillBucketEvent(par3EntityPlayer, par1ItemStack, par2World, var12);
+            if (MinecraftForge.EVENT_BUS.post(event))
+            {
+                return par1ItemStack;
+            }
+
+            if (event.getResult() == Event.Result.ALLOW)
+            {
+                if (par3EntityPlayer.capabilities.isCreativeMode)
+                {
+                    return par1ItemStack;
+                }
+
+                if (--par1ItemStack.stackSize <= 0)
+                {
+                    return event.result;
+                }
+
+                if (!par3EntityPlayer.inventory.addItemStackToInventory(event.result))
+                {
+                    par3EntityPlayer.dropPlayerItem(event.result);
+                }
+
+                return par1ItemStack;
+            }
+
             if (var12.typeOfHit == EnumMovingObjectType.TILE)
             {
                 int var13 = var12.blockX;
@@ -144,7 +174,7 @@
                         return par1ItemStack;
                     }
 
-                    if (this.tryPlaceContainedLiquid(par2World, var5, var7, var9, var13, var14, var15) && !par3EntityPlayer.capabilities.isCreativeMode)
+                    if (this.tryPlaceContainedLiquid(par2World, var5, var7, var9, var13, var14, var15, par3EntityPlayer) && !par3EntityPlayer.capabilities.isCreativeMode)
                     {
                         return new ItemStack(Item.bucketEmpty);
                     }
@@ -159,10 +189,15 @@
         }
     }
 
+    public boolean tryPlaceContainedLiquid(World par1World, double par2, double par4, double par6, int par8, int par9, int par10)
+    {
+        return tryPlaceContainedLiquid(par1World, par2, par4, par6, par8, par9, par10, null);
+    }
+    
     /**
      * Attempts to place the liquid contained inside the bucket.
      */
-    public boolean tryPlaceContainedLiquid(World par1World, double par2, double par4, double par6, int par8, int par9, int par10)
+    public boolean tryPlaceContainedLiquid(World par1World, double par2, double par4, double par6, int par8, int par9, int par10, EntityPlayer player)
     {
         if (this.isFull <= 0)
         {
@@ -185,7 +220,14 @@
             }
             else
             {
-                par1World.setBlockAndMetadataWithNotify(par8, par9, par10, this.isFull, 0);
+                if (player != null)
+                {
+                    par1World.setBlockAndMetadataWithNotify(par8, par9, par10, this.isFull, 0, player);
+                }
+                else
+                {
+                    par1World.setBlockAndMetadataWithNotify(par8, par9, par10, this.isFull, 0);
+                }
             }
 
             return true;
