--- a/net/minecraft/item/MinecartItem.java
+++ b/net/minecraft/item/MinecartItem.java
@@ -26,7 +_,7 @@
          double d2 = p_82487_1_.z() + (double)direction.getStepZ() * 1.125D;
          BlockPos blockpos = p_82487_1_.getPos().relative(direction);
          BlockState blockstate = world.getBlockState(blockpos);
-         RailShape railshape = blockstate.getBlock() instanceof AbstractRailBlock ? blockstate.getValue(((AbstractRailBlock)blockstate.getBlock()).getShapeProperty()) : RailShape.NORTH_SOUTH;
+         RailShape railshape = blockstate.getBlock() instanceof AbstractRailBlock ? ((AbstractRailBlock)blockstate.getBlock()).getRailDirection(blockstate, world, blockpos, null) : RailShape.NORTH_SOUTH;
          double d3;
          if (blockstate.is(BlockTags.RAILS)) {
             if (railshape.isAscending()) {
@@ -79,7 +_,7 @@
       } else {
          ItemStack itemstack = p_195939_1_.getItemInHand();
          if (!world.isClientSide) {
-            RailShape railshape = blockstate.getBlock() instanceof AbstractRailBlock ? blockstate.getValue(((AbstractRailBlock)blockstate.getBlock()).getShapeProperty()) : RailShape.NORTH_SOUTH;
+            RailShape railshape = blockstate.getBlock() instanceof AbstractRailBlock ? ((AbstractRailBlock)blockstate.getBlock()).getRailDirection(blockstate, world, blockpos, null) : RailShape.NORTH_SOUTH;
             double d0 = 0.0D;
             if (railshape.isAscending()) {
                d0 = 0.5D;
