--- a/net/minecraft/world/gen/layer/BiomeLayer.java
+++ b/net/minecraft/world/gen/layer/BiomeLayer.java
@@ -5,6 +5,8 @@
 import net.minecraft.world.gen.INoiseRandom;
 import net.minecraft.world.gen.layer.traits.IC0Transformer;
 
+import java.util.Collections;
+
 public class BiomeLayer implements IC0Transformer {
    private static final int field_202727_a = Registry.field_212624_m.func_148757_b(Biomes.field_150583_P);
    private static final int field_202728_b = Registry.field_212624_m.func_148757_b(Biomes.field_76769_d);
@@ -46,21 +48,21 @@
                return p_202726_1_.func_202696_a(3) == 0 ? field_202733_g : field_202734_h;
             }
 
-            return this.field_151623_c[p_202726_1_.func_202696_a(this.field_151623_c.length)];
+               return Registry.field_212624_m.func_148757_b(getWeightedBiomeEntry(net.minecraftforge.common.BiomeManager.BiomeType.DESERT, p_202726_1_).biome);
          case 2:
             if (i > 0) {
                return field_202732_f;
             }
 
-            return field_202745_s[p_202726_1_.func_202696_a(field_202745_s.length)];
+               return Registry.field_212624_m.func_148757_b(getWeightedBiomeEntry(net.minecraftforge.common.BiomeManager.BiomeType.WARM, p_202726_1_).biome);
          case 3:
             if (i > 0) {
                return field_202737_k;
             }
 
-            return field_202746_t[p_202726_1_.func_202696_a(field_202746_t.length)];
+               return Registry.field_212624_m.func_148757_b(getWeightedBiomeEntry(net.minecraftforge.common.BiomeManager.BiomeType.COOL, p_202726_1_).biome);
          case 4:
-            return field_202747_u[p_202726_1_.func_202696_a(field_202747_u.length)];
+               return Registry.field_212624_m.func_148757_b(getWeightedBiomeEntry(net.minecraftforge.common.BiomeManager.BiomeType.ICY, p_202726_1_).biome);
          default:
             return field_202735_i;
          }
@@ -68,4 +70,11 @@
          return p_202726_2_;
       }
    }
+
+   protected net.minecraftforge.common.BiomeManager.BiomeEntry getWeightedBiomeEntry(net.minecraftforge.common.BiomeManager.BiomeType type, INoiseRandom context) {
+      java.util.List<net.minecraftforge.common.BiomeManager.BiomeEntry> biomeList = Collections.emptyList(); //biomes[type.ordinal()]; //TODO WorldType stuff,
+      int totalWeight = net.minecraft.util.WeightedRandom.func_76272_a(biomeList);
+      int weight = net.minecraftforge.common.BiomeManager.isTypeListModded(type)?context.func_202696_a(totalWeight):context.func_202696_a(totalWeight / 10) * 10;
+      return (net.minecraftforge.common.BiomeManager.BiomeEntry)net.minecraft.util.WeightedRandom.func_180166_a(biomeList, weight);
+   }
 }
