--- ../src-base/minecraft/net/minecraft/world/World.java
+++ ../src-work/minecraft/net/minecraft/world/World.java
@@ -1811,13 +1811,14 @@
             }
         }
 
-        this.field_72989_L = false;
 
         if (!this.field_73000_b.isEmpty())
         {
             this.field_73009_h.removeAll(this.field_73000_b);
             this.field_73000_b.clear();
         }
+
+        this.field_72989_L = false;
 
         this.field_72984_F.func_76318_c("pendingTileEntities");
 
@@ -3646,7 +3647,7 @@
     @SideOnly(Side.CLIENT)
     public double func_72919_O()
     {
-        return this.field_72986_A.func_76067_t() == WorldType.field_77138_c ? 0.0D : 63.0D;
+        return this.field_72986_A.func_76067_t().getHorizon(this);
     }
 
     public CrashReportCategory func_72914_a(CrashReport p_72914_1_)
