--- ../src-base/minecraft/net/minecraft/entity/passive/EntityMooshroom.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityMooshroom.java
@@ -13,7 +13,7 @@
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
 
-public class EntityMooshroom extends EntityCow
+public class EntityMooshroom extends EntityCow implements net.minecraftforge.common.IShearable
 {
     public EntityMooshroom(World worldIn)
     {
@@ -78,6 +78,35 @@
         return new EntityMooshroom(this.worldObj);
     }
 
+    @Override public boolean isShearable(ItemStack item, net.minecraft.world.IBlockAccess world, net.minecraft.util.BlockPos pos){ return getGrowingAge() >= 0; }
+    @Override
+    public java.util.List<ItemStack> onSheared(ItemStack item, net.minecraft.world.IBlockAccess world, net.minecraft.util.BlockPos pos, int fortune)
+    {
+        this.setDead();
+        this.worldObj.spawnParticle(EnumParticleTypes.EXPLOSION_LARGE, this.posX, this.posY + (double)(this.height / 2.0F), this.posZ, 0.0D, 0.0D, 0.0D, new int[0]);
+
+        EntityCow entitycow = new EntityCow(this.worldObj);
+        entitycow.setLocationAndAngles(this.posX, this.posY, this.posZ, this.rotationYaw, this.rotationPitch);
+        entitycow.setHealth(this.getHealth());
+        entitycow.renderYawOffset = this.renderYawOffset;
+
+        if (this.hasCustomName())
+        {
+            entitycow.setCustomNameTag(this.getCustomNameTag());
+        }
+
+        this.worldObj.spawnEntityInWorld(entitycow);
+
+        java.util.List<ItemStack> ret = new java.util.ArrayList<ItemStack>();
+        for (int i = 0; i < 5; ++i)
+        {
+            ret.add(new ItemStack(Blocks.red_mushroom));
+        }
+
+        this.playSound("mob.sheep.shear", 1.0F, 1.0F);
+        return ret;
+    }
+
     protected ResourceLocation func_184647_J()
     {
         return LootTableList.field_186400_H;
