--- a/net/minecraft/world/level/biome/MobSpawnSettings.java
+++ b/net/minecraft/world/level/biome/MobSpawnSettings.java
@@ -49,6 +_,8 @@
     private final float creatureGenerationProbability;
     private final Map<MobCategory, WeightedRandomList<MobSpawnSettings.SpawnerData>> spawners;
     private final Map<EntityType<?>, MobSpawnSettings.MobSpawnCost> mobSpawnCosts;
+    private final java.util.Set<MobCategory> typesView;
+    private final java.util.Set<EntityType<?>> costView;
 
     MobSpawnSettings(
         float p_196689_,
@@ -58,6 +_,8 @@
         this.creatureGenerationProbability = p_196689_;
         this.spawners = ImmutableMap.copyOf(p_196690_);
         this.mobSpawnCosts = ImmutableMap.copyOf(p_196691_);
+        this.typesView = java.util.Collections.unmodifiableSet(this.spawners.keySet());
+        this.costView = java.util.Collections.unmodifiableSet(this.mobSpawnCosts.keySet());
     }
 
     public WeightedRandomList<MobSpawnSettings.SpawnerData> getMobs(MobCategory p_151799_) {
@@ -71,6 +_,14 @@
 
     public float getCreatureProbability() {
         return this.creatureGenerationProbability;
+    }
+
+    public java.util.Set<MobCategory> getSpawnerTypes() {
+        return this.typesView;
+    }
+
+    public java.util.Set<EntityType<?>> getEntityTypes() {
+        return this.costView;
     }
 
     public static class Builder {
