--- ../src-base/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
@@ -95,6 +95,12 @@
         }
     }
 
+    @Deprecated // TODO: remove (1.13)
+    public synchronized boolean chunkExists(int x, int z)
+    {
+        return func_76709_c(x, z);
+    }
+
     @Nullable
 
     public synchronized DataInputStream func_76704_a(int p_76704_1_, int p_76704_2_)
