--- a/net/minecraft/world/entity/monster/Evoker.java
+++ b/net/minecraft/world/entity/monster/Evoker.java
@@ -275,7 +_,7 @@
             return false;
          } else if (Evoker.this.f_19797_ < this.f_33775_) {
             return false;
-         } else if (!Evoker.this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_)) {
+         } else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(Evoker.this.m_9236_(), Evoker.this)) {
             return false;
          } else {
             List<Sheep> list = Evoker.this.m_9236_().m_45971_(Sheep.class, this.f_32705_, Evoker.this, Evoker.this.m_20191_().m_82377_(16.0D, 4.0D, 16.0D));
