--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -85,10 +_,12 @@
     }
 
     protected void doFreezeConversion() {
-        this.convertTo(EntityType.STRAY, true);
+        if (!net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.STRAY, (timer) -> this.conversionTime = timer)) return;
+        var result = this.convertTo(EntityType.STRAY, true);
         if (!this.isSilent()) {
             this.level().levelEvent(null, 1048, this.blockPosition(), 0);
         }
+        net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, result);
     }
 
     @Override
