--- a/net/minecraft/world/entity/monster/Spider.java
+++ b/net/minecraft/world/entity/monster/Spider.java
@@ -119,7 +_,12 @@
    }
 
    public boolean m_7301_(MobEffectInstance p_33809_) {
-      return p_33809_.m_19544_() == MobEffects.f_19614_ ? false : super.m_7301_(p_33809_);
+      if (p_33809_.m_19544_() == MobEffects.f_19614_) {
+         net.minecraftforge.event.entity.living.MobEffectEvent.Applicable event = new net.minecraftforge.event.entity.living.MobEffectEvent.Applicable(this, p_33809_);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         return event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW;
+      }
+      return super.m_7301_(p_33809_);
    }
 
    public boolean m_33816_() {
