--- a/net/minecraft/world/level/storage/loot/providers/nbt/LootNbtProviderType.java
+++ b/net/minecraft/world/level/storage/loot/providers/nbt/LootNbtProviderType.java
@@ -3,7 +_,7 @@
 import net.minecraft.world.level.storage.loot.Serializer;
 import net.minecraft.world.level.storage.loot.SerializerType;
 
-public class LootNbtProviderType extends SerializerType<NbtProvider> {
+public class LootNbtProviderType extends net.minecraftforge.registries.ForgeSerializerTypeRegistryEntry<LootNbtProviderType, NbtProvider> {
    public LootNbtProviderType(Serializer<? extends NbtProvider> p_165621_) {
       super(p_165621_);
    }
