--- ../src-base/minecraft/net/minecraft/item/ItemSeedFood.java
+++ ../src-work/minecraft/net/minecraft/item/ItemSeedFood.java
@@ -8,7 +8,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
-public class ItemSeedFood extends ItemFood
+public class ItemSeedFood extends ItemFood implements net.minecraftforge.common.IPlantable
 {
     private Block crops;
     private Block soilId;
@@ -33,4 +33,16 @@
             return EnumActionResult.FAIL;
         }
     }
+
+    @Override
+    public net.minecraftforge.common.EnumPlantType getPlantType(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return net.minecraftforge.common.EnumPlantType.Crop;
+    }
+
+    @Override
+    public net.minecraft.block.state.IBlockState getPlant(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return this.crops.getDefaultState();
+    }
 }
