--- ../src-base/minecraft/net/minecraft/item/ItemEmptyMap.java
+++ ../src-work/minecraft/net/minecraft/item/ItemEmptyMap.java
@@ -24,7 +24,7 @@
         worldIn.setItemData(s, mapdata);
         mapdata.scale = 0;
         mapdata.func_176054_a(playerIn.posX, playerIn.posZ, mapdata.scale);
-        mapdata.dimension = (byte)worldIn.provider.getDimensionId();
+        mapdata.dimension = worldIn.provider.getDimensionId();
         mapdata.markDirty();
         --itemStackIn.stackSize;
 
