--- ../src-base/minecraft/net/minecraft/item/ItemEmptyMap.java
+++ ../src-work/minecraft/net/minecraft/item/ItemEmptyMap.java
@@ -25,7 +25,7 @@
         worldIn.setItemData(s, mapdata);
         mapdata.scale = 0;
         mapdata.calculateMapCenter(playerIn.posX, playerIn.posZ, mapdata.scale);
-        mapdata.dimension = (byte)worldIn.provider.func_186058_p().func_186068_a();
+        mapdata.dimension = worldIn.provider.func_186058_p().func_186068_a();
         mapdata.field_186210_e = true;
         mapdata.markDirty();
         --itemStackIn.stackSize;
