--- ../src-base/minecraft/net/minecraft/item/ItemSeeds.java
+++ ../src-work/minecraft/net/minecraft/item/ItemSeeds.java
@@ -7,7 +7,7 @@
 import net.minecraft.util.EnumFacing;
 import net.minecraft.world.World;
 
-public class ItemSeeds extends Item
+public class ItemSeeds extends Item implements net.minecraftforge.common.IPlantable
 {
     private Block crops;
     private Block soilBlockID;
@@ -29,7 +29,7 @@
         {
             return false;
         }
-        else if (worldIn.getBlockState(pos).getBlock() == this.soilBlockID && worldIn.isAirBlock(pos.up()))
+        else if (worldIn.getBlockState(pos).getBlock().canSustainPlant(worldIn, pos, EnumFacing.UP, this) && worldIn.isAirBlock(pos.up()))
         {
             worldIn.setBlockState(pos.up(), this.crops.getDefaultState());
             --stack.stackSize;
@@ -40,4 +40,16 @@
             return false;
         }
     }
+
+    @Override
+    public net.minecraftforge.common.EnumPlantType getPlantType(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return this.crops == net.minecraft.init.Blocks.nether_wart ? net.minecraftforge.common.EnumPlantType.Nether : net.minecraftforge.common.EnumPlantType.Crop;
+    }
+
+    @Override
+    public net.minecraft.block.state.IBlockState getPlant(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return this.crops.getDefaultState();
+    }
 }
