--- a/net/minecraft/network/protocol/status/ServerStatus.java
+++ b/net/minecraft/network/protocol/status/ServerStatus.java
@@ -15,9 +_,9 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.util.ExtraCodecs;
 
-public record ServerStatus(Component f_134900_, Optional<ServerStatus.Players> f_134901_, Optional<ServerStatus.Version> f_134902_, Optional<ServerStatus.Favicon> f_134903_, boolean f_242955_) {
+public record ServerStatus(Component f_134900_, Optional<ServerStatus.Players> f_134901_, Optional<ServerStatus.Version> f_134902_, Optional<ServerStatus.Favicon> f_134903_, boolean f_242955_, Optional<net.minecraftforge.network.ServerStatusPing> forgeData) {
    public static final Codec<ServerStatus> f_271163_ = RecordCodecBuilder.create((p_273270_) -> {
-      return p_273270_.group(ExtraCodecs.f_252442_.optionalFieldOf("description", CommonComponents.f_237098_).forGetter(ServerStatus::f_134900_), ServerStatus.Players.f_271480_.optionalFieldOf("players").forGetter(ServerStatus::f_134901_), ServerStatus.Version.f_271272_.optionalFieldOf("version").forGetter(ServerStatus::f_134902_), ServerStatus.Favicon.f_271281_.optionalFieldOf("favicon").forGetter(ServerStatus::f_134903_), Codec.BOOL.optionalFieldOf("enforcesSecureChat", Boolean.valueOf(false)).forGetter(ServerStatus::f_242955_)).apply(p_273270_, ServerStatus::new);
+      return p_273270_.group(ExtraCodecs.f_252442_.optionalFieldOf("description", CommonComponents.f_237098_).forGetter(ServerStatus::f_134900_), ServerStatus.Players.f_271480_.optionalFieldOf("players").forGetter(ServerStatus::f_134901_), ServerStatus.Version.f_271272_.optionalFieldOf("version").forGetter(ServerStatus::f_134902_), ServerStatus.Favicon.f_271281_.optionalFieldOf("favicon").forGetter(ServerStatus::f_134903_), Codec.BOOL.optionalFieldOf("enforcesSecureChat", Boolean.valueOf(false)).forGetter(ServerStatus::f_242955_), net.minecraftforge.network.ServerStatusPing.CODEC.optionalFieldOf("forgeData").forGetter(ServerStatus::forgeData)).apply(p_273270_, ServerStatus::new);
    });
 
    public static record Favicon(byte[] f_271462_) {
