--- ../src_base/minecraft/net/minecraft/logging/LogAgent.java
+++ ../src_work/minecraft/net/minecraft/logging/LogAgent.java
@@ -72,7 +72,7 @@
 
     public void logWarningFormatted(String par1Str, Object ... par2ArrayOfObj)
     {
-        this.serverLogger.log(Level.WARNING, par1Str, par2ArrayOfObj);
+        this.serverLogger.log(Level.WARNING, String.format(par1Str, par2ArrayOfObj));
     }
 
     public void logWarningException(String par1Str, Throwable par2Throwable)
