--- a/net/minecraft/data/recipes/CraftingRecipeBuilder.java
+++ b/net/minecraft/data/recipes/CraftingRecipeBuilder.java
@@ -31,8 +_,15 @@
          this.category = p_250313_;
       }
 
+      private net.minecraftforge.common.crafting.conditions.ICondition condition;
+      public <R extends CraftingResult> R withCondition(net.minecraftforge.common.crafting.conditions.ICondition condition) {
+         this.condition = condition;
+         return (R)this;
+      }
+
       public void serializeRecipeData(JsonObject p_250456_) {
          p_250456_.addProperty("category", this.category.getSerializedName());
+         net.minecraftforge.common.ForgeHooks.writeCondition(this.condition, p_250456_);
       }
    }
 }
