--- a/net/minecraft/data/recipes/FinishedRecipe.java
+++ b/net/minecraft/data/recipes/FinishedRecipe.java
@@ -23,4 +_,12 @@
 
    @Nullable
    AdvancementHolder advancement();
+
+   @Nullable
+   default AdvancementData advancementData() {
+       var adv = this.advancement();
+       return adv == null ? null : new AdvancementData(adv.id(), adv.value().serializeToJson());
+   }
+
+   record AdvancementData(ResourceLocation id, JsonObject data) {}
 }
