--- a/net/minecraft/data/recipes/RecipeProvider.java
+++ b/net/minecraft/data/recipes/RecipeProvider.java
@@ -93,7 +_,9 @@
                AdvancementHolder advancementholder = p_300066_.advancement();
                if (advancementholder != null) {
                   JsonObject jsonobject = advancementholder.value().serializeToJson();
-                  list.add(DataProvider.saveStable(p_254020_, jsonobject, RecipeProvider.this.advancementPathProvider.json(advancementholder.id())));
+                  var saveAdvancementFuture = saveAdvancement(p_254020_, jsonobject, advancementholder, p_300066_);
+                  if (saveAdvancementFuture != null)
+                     list.add(saveAdvancementFuture);
                }
 
             }
@@ -106,6 +_,16 @@
       return CompletableFuture.allOf(list.toArray((p_253414_) -> {
          return new CompletableFuture[p_253414_];
       }));
+   }
+
+   /**
+    * Called every time a recipe is saved to also save the advancement JSON if it exists.
+    *
+    * @return A completable future that saves the advancement to disk, or null to cancel saving the advancement.
+    */
+   @Nullable
+   protected CompletableFuture<?> saveAdvancement(CachedOutput output, JsonObject advancementJson, AdvancementHolder holder, FinishedRecipe finishedRecipe) {
+      return DataProvider.saveStable(output, advancementJson, this.advancementPathProvider.json(holder.id()));
    }
 
    protected CompletableFuture<?> buildAdvancement(CachedOutput p_253674_, AdvancementHolder p_297687_) {
