--- a/net/minecraft/data/recipes/ShapedRecipeBuilder.java
+++ b/net/minecraft/data/recipes/ShapedRecipeBuilder.java
@@ -101,7 +_,14 @@
       this.ensureValid(p_126142_);
       Advancement.Builder advancement$builder = p_298334_.advancement().addCriterion("has_the_recipe", RecipeUnlockedTrigger.unlocked(p_126142_)).rewards(AdvancementRewards.Builder.recipe(p_126142_)).requirements(AdvancementRequirements.Strategy.OR);
       this.criteria.forEach(advancement$builder::addCriterion);
-      p_298334_.accept(new ShapedRecipeBuilder.Result(p_126142_, this.result, this.count, this.group == null ? "" : this.group, determineBookCategory(this.category), this.rows, this.key, advancement$builder.build(p_126142_.withPrefix("recipes/" + this.category.getFolderName() + "/")), this.showNotification));
+      p_298334_.accept(new ShapedRecipeBuilder.Result(p_126142_, this.result, this.count, this.group == null ? "" : this.group, determineBookCategory(this.category), this.rows, this.key, advancement$builder.build(p_126142_.withPrefix("recipes/" + this.category.getFolderName() + "/")), this.showNotification).withCondition(this.condition));
+   }
+
+   @Nullable
+   private net.minecraftforge.common.crafting.conditions.ICondition condition;
+   public ShapedRecipeBuilder condition(net.minecraftforge.common.crafting.conditions.ICondition condition) {
+       this.condition = condition;
+       return this;
    }
 
    private void ensureValid(ResourceLocation p_126144_) {
