--- ../src-base/minecraft/net/minecraft/network/status/server/S00PacketServerInfo.java
+++ ../src-work/minecraft/net/minecraft/network/status/server/S00PacketServerInfo.java
@@ -34,7 +34,7 @@
 
     public void writePacketData(PacketBuffer buf) throws IOException
     {
-        buf.writeString(GSON.toJson((Object)this.response));
+        buf.writeString(this.response.getJson());
     }
 
     public void processPacket(INetHandlerStatusClient handler)
