--- a/net/minecraft/command/impl/LocateCommand.java
+++ b/net/minecraft/command/impl/LocateCommand.java
@@ -25,11 +25,18 @@
          return p_198533_0_.func_197034_c(2);
       });
 
+      if (false)
       for(Entry<String, Structure<?>> entry : Structure.field_236365_a_.entrySet()) {
          literalargumentbuilder = literalargumentbuilder.then(Commands.func_197057_a(entry.getKey()).executes((p_241056_1_) -> {
             return func_241053_a_(p_241056_1_.getSource(), entry.getValue());
          }));
       }
+      else {
+      for (Structure<?> structureFeature : net.minecraftforge.registries.ForgeRegistries.STRUCTURE_FEATURES) {
+         String name = structureFeature.getRegistryName().toString().replace("minecraft:", "");
+         literalargumentbuilder = literalargumentbuilder.then(Commands.func_197057_a(name)
+               .executes(ctx -> func_241053_a_(ctx.getSource(), structureFeature)));
+      }}
 
       p_198528_0_.register(literalargumentbuilder);
    }
