--- a/net/minecraft/command/impl/LocateCommand.java
+++ b/net/minecraft/command/impl/LocateCommand.java
@@ -20,6 +20,14 @@
    public static void func_198528_a(CommandDispatcher<CommandSource> p_198528_0_) {
       p_198528_0_.register(Commands.func_197057_a("locate").requires((p_198533_0_) -> {
          return p_198533_0_.func_197034_c(2);
+      }).then(Commands.func_197056_a("structure_type", net.minecraftforge.common.command.StructureArgument.structure())
+              .suggests(net.minecraftforge.common.command.StructureArgument.suggestions())
+              .executes(ctx -> {
+                  return func_198534_a(ctx.getSource(), ctx.getArgument("structure_type", net.minecraft.util.ResourceLocation.class).toString().replace("minecraft:", ""));
+      })));
+      if (true) return; // Forge: replace this code with above extensible version
+      p_198528_0_.register(Commands.func_197057_a("locate").requires((p_198533_0_) -> {
+         return p_198533_0_.func_197034_c(2);
       }).then(Commands.func_197057_a("Pillager_Outpost").executes((p_198530_0_) -> {
          return func_198534_a(p_198530_0_.getSource(), "Pillager_Outpost");
       })).then(Commands.func_197057_a("Mineshaft").executes((p_198535_0_) -> {
