--- a/net/minecraft/command/impl/SpreadPlayersCommand.java
+++ b/net/minecraft/command/impl/SpreadPlayersCommand.java
@@ -162,7 +_,8 @@
             spreadplayerscommand$position = p_241072_2_[i++];
          }
 
-         entity.func_223102_j((double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198713_a) + 0.5D, (double)spreadplayerscommand$position.func_198710_a(p_241072_1_, p_241072_3_), (double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198714_b) + 0.5D);
+         net.minecraft.util.math.vector.Vector3d target = net.minecraftforge.event.ForgeEventFactory.onEntityTeleportSpreadPlayersCommand(entity, (double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198713_a) + 0.5D, (double)spreadplayerscommand$position.func_198710_a(p_241072_1_, p_241072_3_), (double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198714_b) + 0.5D);
+         if (target != null) entity.func_223102_j(target.field_72450_a, target.field_72448_b, target.field_72449_c);
          double d2 = Double.MAX_VALUE;
 
          for(SpreadPlayersCommand.Position spreadplayerscommand$position1 : p_241072_2_) {
