--- a/net/minecraft/client/gui/screens/controls/KeyBindsList.java
+++ b/net/minecraft/client/gui/screens/controls/KeyBindsList.java
@@ -53,7 +_,7 @@
    }
 
    protected int m_5756_() {
-      return super.m_5756_() + 15;
+      return super.m_5756_() + 15 + 20;
    }
 
    public int m_5759_() {
@@ -109,7 +_,7 @@
       KeyEntry(final KeyMapping p_193916_, final Component p_193917_) {
          this.f_193910_ = p_193916_;
          this.f_193911_ = p_193917_;
-         this.f_193912_ = new Button(0, 0, 75, 20, p_193917_, (p_193939_) -> {
+         this.f_193912_ = new Button(0, 0, 75 + 20 /* Forge: Add space */, 20, p_193917_, (p_193939_) -> {
             KeyBindsList.this.f_193858_.f_193975_ = p_193916_;
          }) {
             protected MutableComponent m_5646_() {
@@ -117,6 +_,7 @@
             }
          };
          this.f_193913_ = new Button(0, 0, 50, 20, new TranslatableComponent("controls.reset"), (p_193935_) -> {
+            this.f_193910_.setToDefault();
             KeyBindsList.this.f_93386_.f_91066_.m_92159_(p_193916_, p_193916_.m_90861_());
             KeyMapping.m_90854_();
          }) {
@@ -130,7 +_,7 @@
          boolean flag = KeyBindsList.this.f_193858_.f_193975_ == this.f_193910_;
          float f = (float)(p_193926_ + 90 - KeyBindsList.this.f_193859_);
          KeyBindsList.this.f_93386_.f_91062_.m_92889_(p_193923_, this.f_193911_, f, (float)(p_193925_ + p_193928_ / 2 - 9 / 2), 16777215);
-         this.f_193913_.f_93620_ = p_193926_ + 190;
+         this.f_193913_.f_93620_ = p_193926_ + 190 + 20;
          this.f_193913_.f_93621_ = p_193925_;
          this.f_193913_.f_93623_ = !this.f_193910_.m_90864_();
          this.f_193913_.m_6305_(p_193923_, p_193929_, p_193930_, p_193932_);
@@ -138,11 +_,12 @@
          this.f_193912_.f_93621_ = p_193925_;
          this.f_193912_.m_93666_(this.f_193910_.m_90863_());
          boolean flag1 = false;
+         boolean keyCodeModifierConflict = true; // gracefully handle conflicts like SHIFT vs SHIFT+G
          if (!this.f_193910_.m_90862_()) {
             for(KeyMapping keymapping : KeyBindsList.this.f_93386_.f_91066_.f_92059_) {
                if (keymapping != this.f_193910_ && this.f_193910_.m_90850_(keymapping)) {
                   flag1 = true;
-                  break;
+                  keyCodeModifierConflict &= keymapping.hasKeyModifierConflict(this.f_193910_);
                }
             }
          }
@@ -150,7 +_,7 @@
          if (flag) {
             this.f_193912_.m_93666_((new TextComponent("> ")).m_7220_(this.f_193912_.m_6035_().m_6881_().m_130940_(ChatFormatting.YELLOW)).m_130946_(" <").m_130940_(ChatFormatting.YELLOW));
          } else if (flag1) {
-            this.f_193912_.m_93666_(this.f_193912_.m_6035_().m_6881_().m_130940_(ChatFormatting.RED));
+            this.f_193912_.m_93666_(this.f_193912_.m_6035_().m_6881_().m_130940_(keyCodeModifierConflict ? ChatFormatting.GOLD : ChatFormatting.RED));
          }
 
          this.f_193912_.m_6305_(p_193923_, p_193929_, p_193930_, p_193932_);
