--- a/net/minecraft/client/renderer/entity/player/AvatarRenderer.java
+++ b/net/minecraft/client/renderer/entity/player/AvatarRenderer.java
@@ -98,7 +_,7 @@
             return HumanoidModel.ArmPose.EMPTY;
         }
 
-        if (!avatar.swinging && itemInHand.is(Items.CROSSBOW) && CrossbowItem.isCharged(itemInHand)) {
+        if (!avatar.swinging && itemInHand.getItem() instanceof CrossbowItem && CrossbowItem.isCharged(itemInHand)) {
             return HumanoidModel.ArmPose.CROSSBOW_HOLD;
         }
 
@@ -141,7 +_,9 @@
         if (attack != null && attack.type() == SwingAnimationType.STAB && avatar.swinging) {
             return HumanoidModel.ArmPose.SPEAR;
         } else {
-            return itemInHand.is(ItemTags.SPEARS) ? HumanoidModel.ArmPose.SPEAR : HumanoidModel.ArmPose.ITEM;
+            if (itemInHand.is(ItemTags.SPEARS)) return HumanoidModel.ArmPose.SPEAR;
+            var pose = net.minecraftforge.client.extensions.common.IClientItemExtensions.of(itemInHand).getArmPose(avatar, hand, itemInHand);
+            return pose == null ?  HumanoidModel.ArmPose.ITEM : pose;
         }
     }
 
@@ -241,12 +_,14 @@
     public void renderRightHand(
         final PoseStack poseStack, final SubmitNodeCollector submitNodeCollector, final int lightCoords, final Identifier skinTexture, final boolean hasSleeve
     ) {
+        if (!net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderArm(poseStack, submitNodeCollector, lightCoords, HumanoidArm.RIGHT))
         this.renderHand(poseStack, submitNodeCollector, lightCoords, skinTexture, this.model.rightArm, hasSleeve);
     }
 
     public void renderLeftHand(
         final PoseStack poseStack, final SubmitNodeCollector submitNodeCollector, final int lightCoords, final Identifier skinTexture, final boolean hasSleeve
     ) {
+        if (!net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderArm(poseStack, submitNodeCollector, lightCoords, HumanoidArm.LEFT))
         this.renderHand(poseStack, submitNodeCollector, lightCoords, skinTexture, this.model.leftArm, hasSleeve);
     }
 
@@ -304,5 +_,12 @@
 
     public static boolean isPlayerUpsideDown(final Player player) {
         return isUpsideDownName(player.getGameProfile().name());
+    }
+
+    @Override
+    public void submit(AvatarRenderState state, PoseStack pose, SubmitNodeCollector source, CameraRenderState camera) {
+        if (net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderAvatarPre(state, this, pose, source, camera)) return;
+        super.submit(state, pose, source, camera);
+        net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderAvatarPost(state, this, pose, source, camera);
     }
 }
