--- a/net/minecraft/client/gui/screens/LoadingOverlay.java
+++ b/net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -64,6 +_,8 @@
         return color & 16777215 | alpha << 24;
     }
 
+    protected boolean extractContentRenderState(final GuiGraphicsExtractor graphics, float alpha) { return true; }
+
     @Override
     public void extractRenderState(final GuiGraphicsExtractor graphics, final int mouseX, final int mouseY, final float a) {
         int width = graphics.guiWidth();
@@ -103,6 +_,7 @@
             logoAlpha = 1.0F;
         }
 
+        if (extractContentRenderState(graphics, logoAlpha)) {
         int contentX = (int)(graphics.guiWidth() * 0.5);
         int logoY = (int)(graphics.guiHeight() * 0.5);
         double logoHeight = Math.min(graphics.guiWidth() * 0.75, graphics.guiHeight()) * 0.25;
@@ -148,6 +_,7 @@
                 graphics, width / 2 - logoWidthHalf, barY - 5, width / 2 + logoWidthHalf, barY + 5, 1.0F - Mth.clamp(fadeOutAnim, 0.0F, 1.0F)
             );
         }
+        }
 
         if (fadeOutAnim >= 2.0F) {
             this.minecraft.gui.setOverlay(null);
@@ -157,6 +_,7 @@
     @Override
     public void tick() {
         if (this.fadeOutStart == -1L && this.reload.isDone() && this.isReadyToFadeOut()) {
+            this.fadeOutStart = Util.getMillis(); // Forge: Moved up to guard against inf loops caused by callback
             try {
                 this.reload.checkExceptions();
                 this.onFinish.accept(Optional.empty());
@@ -164,7 +_,6 @@
                 this.onFinish.accept(Optional.of(t));
             }
 
-            this.fadeOutStart = Util.getMillis();
             if (this.minecraft.gui.screen() != null) {
                 Window window = this.minecraft.getWindow();
                 this.minecraft.gui.screen().init(window.getGuiScaledWidth(), window.getGuiScaledHeight());
