--- a/net/minecraft/client/gui/screens/LoadingOverlay.java
+++ b/net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -119,6 +_,7 @@
       int k1 = (int)((double)this.f_96163_.m_91268_().m_85446_() * 0.8325D);
       float f6 = this.f_96164_.m_7750_();
       this.f_96167_ = Mth.m_14036_(this.f_96167_ * 0.95F + f6 * 0.050000012F, 0.0F, 1.0F);
+      net.minecraftforge.client.loading.ClientModLoader.renderProgressText();
       if (f < 1.0F) {
          this.m_96182_(p_96178_, i / 2 - j1, k1 - 5, i / 2 + j1, k1 + 5, 1.0F - Mth.m_14036_(f, 0.0F, 1.0F));
       }
@@ -128,6 +_,7 @@
       }
 
       if (this.f_96168_ == -1L && this.f_96164_.m_7746_() && (!this.f_96166_ || f1 >= 2.0F)) {
+         this.f_96168_ = Util.m_137550_(); // Moved up to guard against inf loops caused by callback
          try {
             this.f_96164_.m_7748_();
             this.f_96165_.accept(Optional.empty());
@@ -135,7 +_,6 @@
             this.f_96165_.accept(Optional.of(throwable));
          }
 
-         this.f_96168_ = Util.m_137550_();
          if (this.f_96163_.f_91080_ != null) {
             this.f_96163_.f_91080_.m_6575_(this.f_96163_, this.f_96163_.m_91268_().m_85445_(), this.f_96163_.m_91268_().m_85446_());
          }
