--- a/net/minecraft/world/level/levelgen/DebugLevelSource.java
+++ b/net/minecraft/world/level/levelgen/DebugLevelSource.java
@@ -144,4 +_,10 @@
     public int getSeaLevel() {
         return 63;
     }
+
+    public static void initValidStates() {
+        ALL_BLOCKS = StreamSupport.stream(BuiltInRegistries.BLOCK.spliterator(), false).flatMap(block -> block.getStateDefinition().getPossibleStates().stream()).collect(Collectors.toList());
+        GRID_WIDTH = Mth.ceil(Mth.sqrt(ALL_BLOCKS.size()));
+        GRID_HEIGHT = Mth.ceil((float) (ALL_BLOCKS.size() / GRID_WIDTH));
+    }
 }
