--- a/net/minecraft/world/level/levelgen/PhantomSpawner.java
+++ b/net/minecraft/world/level/levelgen/PhantomSpawner.java
@@ -42,13 +_,18 @@
                     for (ServerPlayer serverplayer : p_64576_.players()) {
                         if (!serverplayer.isSpectator()) {
                             BlockPos blockpos = serverplayer.blockPosition();
-                            if (!p_64576_.dimensionType().hasSkyLight() || blockpos.getY() >= p_64576_.getSeaLevel() && p_64576_.canSeeSky(blockpos)) {
                                 DifficultyInstance difficultyinstance = p_64576_.getCurrentDifficultyAt(blockpos);
+                            var vanillaPosition = (!p_64576_.dimensionType().hasSkyLight() || blockpos.getY() >= p_64576_.getSeaLevel() && p_64576_.canSeeSky(blockpos));
+                            var count = 1 + randomsource.nextInt(difficultyinstance.getDifficulty().getId() + 1);
+                            var event = net.minecraftforge.event.ForgeEventFactory.onPlayerSpawnPhantom(serverplayer, count);
+                            var eventResult = event.getResult();
+                            if (eventResult.isDenied()) continue;
+                            if (vanillaPosition || eventResult.isAllowed()) {
                                 if (difficultyinstance.isHarderThan(randomsource.nextFloat() * 3.0F)) {
                                     ServerStatsCounter serverstatscounter = serverplayer.getStats();
                                     int j = Mth.clamp(serverstatscounter.getValue(Stats.CUSTOM.get(Stats.TIME_SINCE_REST)), 1, Integer.MAX_VALUE);
                                     int k = 24000;
-                                    if (randomsource.nextInt(j) >= 72000) {
+                                    if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW || randomsource.nextInt(j) >= 72000) {
                                         BlockPos blockpos1 = blockpos.above(20 + randomsource.nextInt(15))
                                             .east(-10 + randomsource.nextInt(21))
                                             .south(-10 + randomsource.nextInt(21));
@@ -56,7 +_,7 @@
                                         FluidState fluidstate = p_64576_.getFluidState(blockpos1);
                                         if (NaturalSpawner.isValidEmptySpawnBlock(p_64576_, blockpos1, blockstate, fluidstate, EntityType.PHANTOM)) {
                                             SpawnGroupData spawngroupdata = null;
-                                            int l = 1 + randomsource.nextInt(difficultyinstance.getDifficulty().getId() + 1);
+                                            int l = event.getPhantomsToSpawn();
 
                                             for (int i1 = 0; i1 < l; i1++) {
                                                 Phantom phantom = EntityType.PHANTOM.create(p_64576_);
