--- a/net/minecraft/world/level/levelgen/PhantomSpawner.java
+++ b/net/minecraft/world/level/levelgen/PhantomSpawner.java
@@ -60,6 +_,9 @@
                                  for(int i1 = 0; i1 < l; ++i1) {
                                     Phantom phantom = EntityType.f_20509_.m_20615_(p_64576_);
                                     phantom.m_20035_(blockpos1, 0.0F, 0.0F);
+                                    // Note: This call to canEntitySpawn is erroneous, it is only provided for backwards compatibility.
+                                    if (net.minecraftforge.common.ForgeHooks.canEntitySpawn(phantom, p_64576_, blockpos1.m_123341_(), blockpos1.m_123342_(), blockpos1.m_123343_(), null, MobSpawnType.NATURAL) == -1) { i--; continue; }
+                                    if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(phantom, p_64576_, blockpos1.m_123341_(), blockpos1.m_123342_(), blockpos1.m_123343_(), null, MobSpawnType.NATURAL))
                                     spawngroupdata = phantom.m_6518_(p_64576_, difficultyinstance, MobSpawnType.NATURAL, spawngroupdata, (CompoundTag)null);
                                     p_64576_.m_47205_(phantom);
                                  }
