--- ../src_base/minecraft/net/minecraft/world/gen/structure/ComponentVillage.java
+++ ../src_work/minecraft/net/minecraft/world/gen/structure/ComponentVillage.java
@@ -7,6 +7,10 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.world.World;
 
+import net.minecraftforge.common.*;
+import net.minecraftforge.event.Event.*;
+import net.minecraftforge.event.terraingen.*;
+
 public abstract class ComponentVillage extends StructureComponent
 {
     protected int field_143015_k = -1;
@@ -14,6 +18,7 @@
     /** The number of villagers that have been spawned in this component. */
     private int villagersSpawned;
     private boolean field_143014_b;
+    private ComponentVillageStartPiece startPiece;
 
     public ComponentVillage() {}
 
@@ -24,6 +29,7 @@
         if (par1ComponentVillageStartPiece != null)
         {
             this.field_143014_b = par1ComponentVillageStartPiece.inDesert;
+            startPiece = par1ComponentVillageStartPiece;
         }
     }
 
@@ -157,6 +163,10 @@
      */
     protected int getBiomeSpecificBlock(int par1, int par2)
     {
+        BiomeEvent.GetVillageBlockID event = new BiomeEvent.GetVillageBlockID(startPiece == null ? null : startPiece.biome, par1, par2);
+        MinecraftForge.TERRAIN_GEN_BUS.post(event);
+        if (event.getResult() == Result.DENY) return event.replacement;
+
         if (this.field_143014_b)
         {
             if (par1 == Block.wood.blockID)
@@ -198,6 +208,10 @@
      */
     protected int getBiomeSpecificBlockMetadata(int par1, int par2)
     {
+        BiomeEvent.GetVillageBlockMeta event = new BiomeEvent.GetVillageBlockMeta(startPiece == null ? null : startPiece.biome, par1, par2);
+        MinecraftForge.TERRAIN_GEN_BUS.post(event);
+        if (event.getResult() == Result.DENY) return event.replacement;
+
         if (this.field_143014_b)
         {
             if (par1 == Block.wood.blockID)
