--- ../src-base/minecraft/net/minecraft/world/gen/structure/MapGenStronghold.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/MapGenStronghold.java
@@ -29,11 +29,19 @@
 
         for (Biome biome : Biome.field_185377_q)
         {
-            if (biome != null && biome.func_185355_j() > 0.0F)
+            if (biome != null && biome.func_185355_j() > 0.0F && !net.minecraftforge.common.BiomeManager.strongHoldBiomesBlackList.contains(biome))
             {
                 this.field_151546_e.add(biome);
             }
         }
+
+        for (Biome biome : net.minecraftforge.common.BiomeManager.strongHoldBiomes)
+        {
+            if (!this.field_151546_e.contains(biome))
+            {
+                this.field_151546_e.add(biome);
+            }
+        }
     }
 
     public MapGenStronghold(Map<String, String> p_i2068_1_)
