--- ../src-base/minecraft/net/minecraft/world/gen/structure/StructureComponent.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/StructureComponent.java
@@ -36,6 +36,10 @@
 
     public NBTTagCompound func_143010_b()
     {
+        if (MapGenStructureIO.func_143036_a(this) == null) // Friendlier error then the Null Stirng error below.
+        {
+            throw new RuntimeException("StructureComponent \"" + this.getClass().getName() + "\" missing ID Mapping, Modder see MapGenStructureIO");
+        }
         NBTTagCompound nbttagcompound = new NBTTagCompound();
         nbttagcompound.setString("id", MapGenStructureIO.func_143036_a(this));
         nbttagcompound.setTag("BB", this.boundingBox.func_151535_h());
