--- ../src-base/minecraft/net/minecraft/world/gen/structure/StructureNetherBridgePieces.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/StructureNetherBridgePieces.java
@@ -158,7 +158,7 @@
                     if (p_74875_3_.func_78890_b(j, i, k))
                     {
                         this.field_111021_b = false;
-                        this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 3, 2, 3, field_111019_a, 2 + p_74875_2_.nextInt(4));
+                        this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 3, 2, 3, net.minecraftforge.common.ChestGenHooks.getItems(net.minecraftforge.common.ChestGenHooks.NETHER_FORTRESS, p_74875_2_), net.minecraftforge.common.ChestGenHooks.getCount(net.minecraftforge.common.ChestGenHooks.NETHER_FORTRESS, p_74875_2_));
                     }
                 }
 
@@ -237,7 +237,7 @@
                     if (p_74875_3_.func_78890_b(j, i, k))
                     {
                         this.field_111020_b = false;
-                        this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 1, 2, 3, field_111019_a, 2 + p_74875_2_.nextInt(4));
+                        this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 1, 2, 3, net.minecraftforge.common.ChestGenHooks.getItems(net.minecraftforge.common.ChestGenHooks.NETHER_FORTRESS, p_74875_2_), net.minecraftforge.common.ChestGenHooks.getCount(net.minecraftforge.common.ChestGenHooks.NETHER_FORTRESS, p_74875_2_));
                     }
                 }
 
