--- ../src-base/minecraft/net/minecraft/world/gen/structure/StructureComponent.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/StructureComponent.java
@@ -34,6 +34,10 @@
 
     public NBTTagCompound func_143010_b()
     {
+        if (MapGenStructureIO.func_143036_a(this) == null) // Friendlier error then the Null Stirng error below.
+        {
+            throw new RuntimeException("StructureComponent \"" + this.getClass().getName() + "\" missing ID Mapping, Modder see MapGenStructureIO");
+        }
         NBTTagCompound nbttagcompound = new NBTTagCompound();
         nbttagcompound.func_74778_a("id", MapGenStructureIO.func_143036_a(this));
         nbttagcompound.func_74782_a("BB", this.field_74887_e.func_151535_h());
