--- a/net/minecraft/world/food/FoodData.java
+++ b/net/minecraft/world/food/FoodData.java
@@ -23,9 +_,15 @@
       this.saturationLevel = Math.min(this.saturationLevel + (float)p_38708_ * p_38709_ * 2.0F, (float)this.foodLevel);
    }
 
+   // Use the LivingEntity sensitive version in favour of this.
+   @Deprecated
    public void eat(Item p_38713_, ItemStack p_38714_) {
+      this.eat(p_38713_, p_38714_, null);
+   }
+
+   public void eat(Item p_38713_, ItemStack p_38714_, @org.jetbrains.annotations.Nullable net.minecraft.world.entity.LivingEntity entity) {
       if (p_38713_.isEdible()) {
-         FoodProperties foodproperties = p_38713_.getFoodProperties();
+         FoodProperties foodproperties = p_38714_.getFoodProperties(entity);
          this.eat(foodproperties.getNutrition(), foodproperties.getSaturationModifier());
       }
 
