--- a/net/minecraft/world/food/FoodData.java
+++ b/net/minecraft/world/food/FoodData.java
@@ -19,17 +_,34 @@
       this.f_38697_ = 5.0F;
    }
 
+   @Deprecated
    public void m_38707_(int p_38708_, float p_38709_) {
-      this.f_38696_ = Math.min(p_38708_ + this.f_38696_, 20);
-      this.f_38697_ = Math.min(this.f_38697_ + (float)p_38708_ * p_38709_ * 2.0F, (float)this.f_38696_);
-   }
-
+      this.eat(p_38708_, p_38709_, null);
+   }
+   public void eatInternal(int p_38708_, float p_38709_) {
+       this.f_38696_ = Math.min(p_38708_ + this.f_38696_, 20);
+       this.f_38697_ = Math.min(this.f_38697_ + (float)p_38708_ * p_38709_ * 2.0F, (float)this.f_38696_);
+   }
+   public void eat(int p_38708_, float p_38709_, @org.jetbrains.annotations.Nullable net.minecraft.world.entity.LivingEntity entity) {
+      float[] ret = net.minecraftforge.common.ForgeHooks.onLivingEat(entity, p_38708_, p_38709_);
+      if(ret != null)
+         this.eatInternal((int)ret[0], ret[1]);
+   }
+   // Use the LivingEntity sensitive version in favor of this.
+   @Deprecated
    public void m_38712_(Item p_38713_, ItemStack p_38714_) {
+      this.eat(p_38713_, p_38714_, null);
+   }
+   public java.util.List<com.mojang.datafixers.util.Pair<net.minecraft.world.effect.MobEffectInstance, Float>> eat(Item p_38713_, ItemStack p_38714_, @org.jetbrains.annotations.Nullable net.minecraft.world.entity.LivingEntity entity) {
       if (p_38713_.m_41472_()) {
-         FoodProperties foodproperties = p_38713_.m_41473_();
-         this.m_38707_(foodproperties.m_38744_(), foodproperties.m_38745_());
+         net.minecraftforge.event.entity.living.LivingEatEvent ret = net.minecraftforge.common.ForgeHooks.onLivingEat(entity, p_38714_);
+         if(ret != null)
+         {
+            this.eatInternal(ret.getFoodAmount(), ret.getSaturationAmount());
+            return ret.getEffects();
+         }
       }
-
+      return null;
    }
 
    public void m_38710_(Player p_38711_) {
