--- a/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -327,6 +_,7 @@
 
       this.f_30520_.m_19164_(this);
       this.m_7493_();
+      this.itemHandler = net.minecraftforge.common.util.LazyOptional.of(() -> new net.minecraftforge.items.wrapper.InvWrapper(this.f_30520_));
    }
 
    protected void m_7493_() {
@@ -374,9 +_,9 @@
    protected void m_7355_(BlockPos p_30584_, BlockState p_30585_) {
       if (!p_30585_.m_278721_()) {
          BlockState blockstate = this.m_9236_().m_8055_(p_30584_.m_7494_());
-         SoundType soundtype = p_30585_.m_60827_();
+         SoundType soundtype = p_30585_.getSoundType(m_9236_(), p_30584_, this);
          if (blockstate.m_60713_(Blocks.f_50125_)) {
-            soundtype = blockstate.m_60827_();
+            soundtype = blockstate.getSoundType(m_9236_(), p_30584_, this);
          }
 
          if (this.m_20160_() && this.f_30523_) {
@@ -786,6 +_,7 @@
       this.m_20334_(vec3.f_82479_, d1, vec3.f_82481_);
       this.m_30655_(true);
       this.f_19812_ = true;
+               net.minecraftforge.common.ForgeHooks.onLivingJump(this);
       if (p_275435_.f_82481_ > 0.0D) {
          float f = Mth.m_14031_(this.m_146908_() * ((float)Math.PI / 180F));
          float f1 = Mth.m_14089_(this.m_146908_() * ((float)Math.PI / 180F));
@@ -1120,6 +_,25 @@
 
       this.m_214179_(p_30555_.m_213780_());
       return super.m_6518_(p_30555_, p_30556_, p_30557_, p_30558_, p_30559_);
+   }
+
+   private net.minecraftforge.common.util.LazyOptional<?> itemHandler = null;
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable net.minecraft.core.Direction facing) {
+      if (this.m_6084_() && capability == net.minecraftforge.common.capabilities.ForgeCapabilities.ITEM_HANDLER && itemHandler != null)
+         return itemHandler.cast();
+      return super.getCapability(capability, facing);
+   }
+
+   @Override
+   public void invalidateCaps() {
+      super.invalidateCaps();
+      if (itemHandler != null) {
+         net.minecraftforge.common.util.LazyOptional<?> oldHandler = itemHandler;
+         itemHandler = null;
+         oldHandler.invalidate();
+      }
    }
 
    public boolean m_149511_(Container p_149512_) {
