--- a/net/minecraft/world/entity/animal/horse/Horse.java
+++ b/net/minecraft/world/entity/animal/horse/Horse.java
@@ -144,6 +_,8 @@
          this.playSound(SoundEvents.HORSE_BREATHE, p_30709_.getVolume() * 0.6F, p_30709_.getPitch());
       }
 
+      ItemStack stack = this.inventory.getItem(1);
+      if (isArmor(stack)) stack.onHorseArmorTick(level(), this);
    }
 
    protected SoundEvent getAmbientSound() {
