--- a/net/minecraft/data/loot/BlockLoot.java
+++ b/net/minecraft/data/loot/BlockLoot.java
@@ -250,7 +_,7 @@
       return LootTable.m_79147_();
    }
 
-   public void accept(BiConsumer<ResourceLocation, LootTable.Builder> p_124179_) {
+   protected void addTables() {
       this.m_124288_(Blocks.f_50122_);
       this.m_124288_(Blocks.f_50175_);
       this.m_124288_(Blocks.f_50228_);
@@ -1252,9 +_,13 @@
       this.m_124165_(Blocks.f_50142_, m_124125_());
       this.m_124165_(Blocks.f_152491_, m_124125_());
       this.m_124165_(Blocks.f_152499_, m_124125_());
+   }
+
+   public void accept(BiConsumer<ResourceLocation, LootTable.Builder> p_124179_) {
+      this.addTables();
       Set<ResourceLocation> set = Sets.newHashSet();
 
-      for(Block block : Registry.f_122824_) {
+      for(Block block : getKnownBlocks()) {
          ResourceLocation resourcelocation = block.m_60589_();
          if (resourcelocation != BuiltInLootTables.f_78712_ && set.add(resourcelocation)) {
             LootTable.Builder loottable$builder = this.f_124070_.remove(resourcelocation);
@@ -1279,6 +_,10 @@
 
    public static LootTable.Builder m_124137_(Block p_124138_) {
       return m_124161_(p_124138_, DoorBlock.f_52730_, DoubleBlockHalf.LOWER);
+   }
+
+   protected Iterable<Block> getKnownBlocks() {
+       return Registry.f_122824_;
    }
 
    public void m_124252_(Block p_124253_) {
