--- ../src_base/minecraft_server/net/minecraft/src/BlockNetherStalk.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/BlockNetherStalk.java	0000-00-00 00:00:00.000000000 -0000
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import java.util.ArrayList;
 import java.util.Random;
 
 public class BlockNetherStalk extends BlockFlower
@@ -66,30 +67,20 @@
         return 6;
     }
 
-    /**
-     * Drops the block items with a specified chance of dropping the specified items
-     */
-    public void dropBlockAsItemWithChance(World par1World, int par2, int par3, int par4, int par5, float par6, int par7)
+    @Override
+    public ArrayList<ItemStack> getBlockDropped(World world, int x, int y, int z, int metadata, int fortune)
     {
-        if (!par1World.isRemote)
+        ArrayList<ItemStack> ret = new ArrayList<ItemStack>();
+        int n = 1;
+        if (metadata >= 3)
         {
-            int var8 = 1;
-
-            if (par5 >= 3)
-            {
-                var8 = 2 + par1World.rand.nextInt(3);
-
-                if (par7 > 0)
-                {
-                    var8 += par1World.rand.nextInt(par7 + 1);
-                }
-            }
-
-            for (int var9 = 0; var9 < var8; ++var9)
-            {
-                this.dropBlockAsItem_do(par1World, par2, par3, par4, new ItemStack(Item.netherStalkSeeds));
-            }
+            n = 2 + world.rand.nextInt(3) + (fortune > 0 ? world.rand.nextInt(fortune + 1) : 0);
+        }
+        for (int m = 0; m < n; m++)
+        {
+            ret.add(new ItemStack(Item.netherStalkSeeds));
         }
+        return ret;
     }
 
     /**
