--- ../src_base/minecraft_server/net/minecraft/src/BlockLadder.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/BlockLadder.java	0000-00-00 00:00:00.000000000 -0000
@@ -71,7 +71,10 @@
      */
     public boolean canPlaceBlockAt(World par1World, int par2, int par3, int par4)
     {
-        return par1World.isBlockNormalCube(par2 - 1, par3, par4) ? true : (par1World.isBlockNormalCube(par2 + 1, par3, par4) ? true : (par1World.isBlockNormalCube(par2, par3, par4 - 1) ? true : par1World.isBlockNormalCube(par2, par3, par4 + 1)));
+        return par1World.isBlockSolidOnSide(par2 - 1, par3, par4, 5) ||
+               par1World.isBlockSolidOnSide(par2 + 1, par3, par4, 4) ||
+               par1World.isBlockSolidOnSide(par2, par3, par4 - 1, 3) ||
+               par1World.isBlockSolidOnSide(par2, par3, par4 + 1, 2);
     }
 
     /**
@@ -82,22 +85,22 @@
     {
         int var6 = par1World.getBlockMetadata(par2, par3, par4);
 
-        if ((var6 == 0 || par5 == 2) && par1World.isBlockNormalCube(par2, par3, par4 + 1))
+        if ((var6 == 0 || par5 == 2) && par1World.isBlockSolidOnSide(par2, par3, par4 + 1, 2))
         {
             var6 = 2;
         }
 
-        if ((var6 == 0 || par5 == 3) && par1World.isBlockNormalCube(par2, par3, par4 - 1))
+        if ((var6 == 0 || par5 == 3) && par1World.isBlockSolidOnSide(par2, par3, par4 - 1, 3))
         {
             var6 = 3;
         }
 
-        if ((var6 == 0 || par5 == 4) && par1World.isBlockNormalCube(par2 + 1, par3, par4))
+        if ((var6 == 0 || par5 == 4) && par1World.isBlockSolidOnSide(par2 + 1, par3, par4, 4))
         {
             var6 = 4;
         }
 
-        if ((var6 == 0 || par5 == 5) && par1World.isBlockNormalCube(par2 - 1, par3, par4))
+        if ((var6 == 0 || par5 == 5) && par1World.isBlockSolidOnSide(par2 - 1, par3, par4, 5))
         {
             var6 = 5;
         }
@@ -114,22 +117,22 @@
         int var6 = par1World.getBlockMetadata(par2, par3, par4);
         boolean var7 = false;
 
-        if (var6 == 2 && par1World.isBlockNormalCube(par2, par3, par4 + 1))
+        if (var6 == 2 && par1World.isBlockSolidOnSide(par2, par3, par4 + 1, 2))
         {
             var7 = true;
         }
 
-        if (var6 == 3 && par1World.isBlockNormalCube(par2, par3, par4 - 1))
+        if (var6 == 3 && par1World.isBlockSolidOnSide(par2, par3, par4 - 1, 3))
         {
             var7 = true;
         }
 
-        if (var6 == 4 && par1World.isBlockNormalCube(par2 + 1, par3, par4))
+        if (var6 == 4 && par1World.isBlockSolidOnSide(par2 + 1, par3, par4, 4))
         {
             var7 = true;
         }
 
-        if (var6 == 5 && par1World.isBlockNormalCube(par2 - 1, par3, par4))
+        if (var6 == 5 && par1World.isBlockSolidOnSide(par2 - 1, par3, par4, 5))
         {
             var7 = true;
         }
@@ -150,4 +153,10 @@
     {
         return 1;
     }
+    
+    @Override
+    public boolean isLadder(World world, int x, int y, int z) 
+    {
+    	return true;
+    }
 }
