--- ../src_base/minecraft_server/net/minecraft/src/ModLoader.java
+++ ../src_work/minecraft_server/net/minecraft/src/ModLoader.java
@@ -538,7 +538,7 @@
      */
     public static void registerEntityID(Class<? extends Entity> entityClass, String entityName, int id)
     {
-        EntityRegistry.registerGlobalEntityID(entityClass, entityName, id);
+        EntityRegistry.registerEntityID(entityClass, entityName, id);
     }
 
     /**
@@ -552,7 +552,7 @@
      */
     public static void registerEntityID(Class<? extends Entity> entityClass, String entityName, int id, int background, int foreground)
     {
-        EntityRegistry.registerGlobalEntityID(entityClass, entityName, id, background, foreground);
+        EntityRegistry.registerEntityID(entityClass, entityName, id, background, foreground);
     }
 
     public static void registerKey(BaseMod mod, Object keyHandler, boolean allowRepeat)
