--- ../src_base/minecraft_server/net/minecraft/src/WorldGenTallGrass.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/WorldGenTallGrass.java	0000-00-00 00:00:00.000000000 -0000
@@ -18,10 +18,16 @@
     {
         int var11;
 
-        for (boolean var6 = false; ((var11 = par1World.getBlockId(par3, par4, par5)) == 0 || var11 == Block.leaves.blockID) && par4 > 0; --par4)
+        Block block = null;
+        do 
         {
-            ;
-        }
+            block = Block.blocksList[par1World.getBlockId(par3,  par4, par5)];
+            if (block != null && !block.isLeaves(par1World, par3, par4, par5))
+            {
+                break;
+            }
+            par4--;
+        } while (par4 > 0);
 
         for (int var7 = 0; var7 < 128; ++var7)
         {
