--- a/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java
+++ b/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java
@@ -3,7 +3,7 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
 import com.mojang.datafixers.util.Pair;
-import it.unimi.dsi.fastutil.objects.ObjectArrays;
+
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
@@ -16,25 +16,33 @@
 import net.minecraft.world.gen.feature.template.GravityStructureProcessor;
 import net.minecraft.world.gen.feature.template.StructureProcessor;
 import net.minecraft.world.gen.feature.template.TemplateManager;
+import net.minecraftforge.common.JigsawCategory;
+import org.apache.logging.log4j.LogManager;
 
 public class JigsawPattern {
    public static final JigsawPattern field_214949_a = new JigsawPattern(new ResourceLocation("empty"), new ResourceLocation("empty"), ImmutableList.of(), JigsawPattern.PlacementBehaviour.RIGID);
    public static final JigsawPattern field_214950_b = new JigsawPattern(new ResourceLocation("invalid"), new ResourceLocation("invalid"), ImmutableList.of(), JigsawPattern.PlacementBehaviour.RIGID);
    private final ResourceLocation field_214951_c;
-   private final ImmutableList<Pair<JigsawPiece, Integer>> field_214952_d;
-   private final List<JigsawPiece> field_214953_e;
+   private final ImmutableList<net.minecraftforge.common.JigsawCategory> field_214952_d;
+   private final List<net.minecraftforge.common.JigsawCategory.PieceItem> field_214953_e;
    private final ResourceLocation field_214954_f;
    private final JigsawPattern.PlacementBehaviour field_214955_g;
    private int field_214956_h = Integer.MIN_VALUE;
 
    public JigsawPattern(ResourceLocation p_i51397_1_, ResourceLocation p_i51397_2_, List<Pair<JigsawPiece, Integer>> p_i51397_3_, JigsawPattern.PlacementBehaviour p_i51397_4_) {
+      this(p_i51397_1_,p_i51397_2_,p_i51397_4_, Lists.newArrayList(new net.minecraftforge.common.JigsawCategory(net.minecraftforge.common.JigsawCategory.DEFAULTCATEGORY, Lists.newArrayList(p_i51397_3_))));
+   }
+
+   public JigsawPattern(ResourceLocation p_i51397_1_, ResourceLocation p_i51397_2_, JigsawPattern.PlacementBehaviour p_i51397_4_, List<net.minecraftforge.common.JigsawCategory> p_i51397_3_) {
+      p_i51397_3_ = net.minecraftforge.event.ForgeEventFactory.onJigsawPatternInit(p_i51397_1_, Lists.newArrayList(p_i51397_3_));
       this.field_214951_c = p_i51397_1_;
       this.field_214952_d = ImmutableList.copyOf(p_i51397_3_);
       this.field_214953_e = Lists.newArrayList();
 
-      for(Pair<JigsawPiece, Integer> pair : p_i51397_3_) {
-         for(Integer integer = 0; integer < pair.getSecond(); integer = integer + 1) {
-            this.field_214953_e.add(pair.getFirst().func_214845_a(p_i51397_4_));
+      int totalCategoryWeight = p_i51397_3_.stream().mapToInt(JigsawCategory::getCategoryWeight).sum();
+      for(net.minecraftforge.common.JigsawCategory category : p_i51397_3_) {
+         for(Pair<JigsawPiece, Integer> pair : category.getPieces()){
+            this.field_214953_e.add(new net.minecraftforge.common.JigsawCategory.PieceItem(((float)category.getCategoryWeight() / (float)totalCategoryWeight)*((float)pair.getSecond() / (float)category.getTotalPieceWeight()), pair.getFirst().func_214845_a(p_i51397_4_)));
          }
       }
 
@@ -45,7 +53,7 @@
    public int func_214945_a(TemplateManager p_214945_1_) {
       if (this.field_214956_h == Integer.MIN_VALUE) {
          this.field_214956_h = this.field_214953_e.stream().mapToInt((p_214942_1_) -> {
-            return p_214942_1_.func_214852_a(p_214945_1_, BlockPos.field_177992_a, Rotation.NONE).func_78882_c();
+            return p_214942_1_.getPiece().func_214852_a(p_214945_1_, BlockPos.field_177992_a, Rotation.NONE).func_78882_c();
          }).max().orElse(0);
       }
 
@@ -57,11 +65,20 @@
    }
 
    public JigsawPiece func_214944_a(Random p_214944_1_) {
-      return this.field_214953_e.get(p_214944_1_.nextInt(this.field_214953_e.size()));
+      return net.minecraftforge.common.JigsawCategory.WeightedFloatRandom.getRandomItem(p_214944_1_, this.field_214953_e, 1).getPiece();
    }
 
    public List<JigsawPiece> func_214943_b(Random p_214943_1_) {
-      return ImmutableList.copyOf(ObjectArrays.shuffle(this.field_214953_e.toArray(new JigsawPiece[0]), p_214943_1_));
+      List<net.minecraftforge.common.JigsawCategory.PieceItem> full = Lists.newArrayList(this.field_214953_e);
+      List<JigsawPiece> pieces = Lists.newArrayList();
+      float weight = 1;
+      while (!full.isEmpty()){
+         net.minecraftforge.common.JigsawCategory.PieceItem picked = net.minecraftforge.common.JigsawCategory.WeightedFloatRandom.getRandomItem(p_214943_1_, full, weight);
+         full.remove(picked);
+         weight -= picked.getWeight();
+         pieces.add(picked.getPiece());
+      }
+      return pieces;
    }
 
    public ResourceLocation func_214947_b() {
@@ -72,7 +89,7 @@
       return this.field_214953_e.size();
    }
 
-   public static enum PlacementBehaviour {
+   public static enum PlacementBehaviour implements net.minecraftforge.common.IExtensibleEnum {
       TERRAIN_MATCHING("terrain_matching", ImmutableList.of(new GravityStructureProcessor(Heightmap.Type.WORLD_SURFACE_WG, -1))),
       RIGID("rigid", ImmutableList.of());
 
@@ -98,5 +115,15 @@
       public ImmutableList<StructureProcessor> func_214937_b() {
          return this.field_214941_e;
       }
+
+      public static PlacementBehaviour create(String enumName, String p_i50487_3_, ImmutableList<StructureProcessor> p_i50487_4_) {
+         throw new IllegalStateException("Enum not extended");
+      }
+
+      @Override
+      @Deprecated
+      public void init() {
+         field_214939_c.put(func_214936_a(), this);
+      }
    }
 }
