--- a/net/minecraft/world/gen/feature/jigsaw/JigsawPiece.java
+++ b/net/minecraft/world/gen/feature/jigsaw/JigsawPiece.java
@@ -17,6 +17,7 @@
 public abstract class JigsawPiece {
    @Nullable
    private volatile JigsawPattern.PlacementBehaviour field_214855_a;
+   private static final net.minecraft.util.ResourceLocation REGISTRYNAME = new net.minecraft.util.ResourceLocation("empty");
 
    protected JigsawPiece(JigsawPattern.PlacementBehaviour p_i51398_1_) {
       this.field_214855_a = p_i51398_1_;
@@ -62,4 +63,12 @@
    public int func_214850_d() {
       return 1;
    }
+
+   /**
+    * @return RegistryName of the JigsawPiece
+    */
+   public net.minecraft.util.ResourceLocation getRegistryName(){
+      return REGISTRYNAME;
+   }
+
 }
