--- a/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java
+++ b/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java
@@ -28,6 +28,7 @@
    private int field_214956_h = Integer.MIN_VALUE;
 
    public JigsawPattern(ResourceLocation p_i51397_1_, ResourceLocation p_i51397_2_, List<Pair<JigsawPiece, Integer>> p_i51397_3_, JigsawPattern.PlacementBehaviour p_i51397_4_) {
+      p_i51397_3_ = net.minecraftforge.event.ForgeEventFactory.onJigsawPatternInit(p_i51397_1_, Lists.newArrayList(p_i51397_3_));
       this.field_214951_c = p_i51397_1_;
       this.field_214952_d = ImmutableList.copyOf(p_i51397_3_);
       this.field_214953_e = Lists.newArrayList();
@@ -72,7 +73,7 @@
       return this.field_214953_e.size();
    }
 
-   public static enum PlacementBehaviour {
+   public static enum PlacementBehaviour implements net.minecraftforge.common.IExtensibleEnum {
       TERRAIN_MATCHING("terrain_matching", ImmutableList.of(new GravityStructureProcessor(Heightmap.Type.WORLD_SURFACE_WG, -1))),
       RIGID("rigid", ImmutableList.of());
 
@@ -98,5 +99,15 @@
       public ImmutableList<StructureProcessor> func_214937_b() {
          return this.field_214941_e;
       }
+
+      public static PlacementBehaviour create(String enumName, String p_i50487_3_, ImmutableList<StructureProcessor> p_i50487_4_) {
+         throw new IllegalStateException("Enum not extended");
+      }
+
+      @Override
+      @Deprecated
+      public void init() {
+         field_214939_c.put(func_214936_a(), this);
+      }
    }
 }
