--- a/net/minecraft/world/gen/feature/jigsaw/SingleJigsawPiece.java
+++ b/net/minecraft/world/gen/feature/jigsaw/SingleJigsawPiece.java
@@ -41,6 +41,16 @@
       this.field_214862_b = ImmutableList.copyOf(p_i51401_2_);
    }
 
+   public SingleJigsawPiece(ResourceLocation location, List<StructureProcessor> processors, JigsawPattern.PlacementBehaviour p_i51401_3_) {
+      super(p_i51401_3_);
+      this.field_214861_a = location;
+      this.field_214862_b = ImmutableList.copyOf(processors);
+   }
+
+   public SingleJigsawPiece(ResourceLocation location){
+      this(location, ImmutableList.of(), JigsawPattern.PlacementBehaviour.RIGID);
+   }
+
    @Deprecated
    public SingleJigsawPiece(String p_i51402_1_) {
       this(p_i51402_1_, ImmutableList.of());
@@ -89,7 +99,7 @@
       if (!template.func_189962_a(p_225575_2_, p_225575_4_, placementsettings, 18)) {
          return false;
       } else {
-         for(Template.BlockInfo template$blockinfo : Template.func_215387_a(p_225575_2_, p_225575_4_, placementsettings, this.func_214857_a(p_225575_1_, p_225575_4_, p_225575_5_, false))) {
+         for(Template.BlockInfo template$blockinfo : Template.processBlockInfos(template, p_225575_2_, p_225575_4_, placementsettings, this.func_214857_a(p_225575_1_, p_225575_4_, p_225575_5_, false))) {
             this.func_214846_a(p_225575_2_, template$blockinfo, p_225575_4_, p_225575_5_, p_225575_7_, p_225575_6_);
          }
 
@@ -123,4 +133,9 @@
    public String toString() {
       return "Single[" + this.field_214861_a + "]";
    }
+
+   @Override
+   public ResourceLocation getRegistryName() {
+      return field_214861_a;
+   }
 }
