--- a/net/minecraft/world/gen/feature/structure/FortressStructure.java
+++ b/net/minecraft/world/gen/feature/structure/FortressStructure.java
@@ -16,7 +16,7 @@
 import net.minecraft.world.gen.feature.template.TemplateManager;
 
 public class FortressStructure extends Structure<NoFeatureConfig> {
-   private static final List<MobSpawnInfo.Spawners> field_202381_d = ImmutableList.of(new MobSpawnInfo.Spawners(EntityType.field_200792_f, 10, 2, 3), new MobSpawnInfo.Spawners(EntityType.field_233592_ba_, 5, 4, 4), new MobSpawnInfo.Spawners(EntityType.field_200722_aA, 8, 5, 5), new MobSpawnInfo.Spawners(EntityType.field_200741_ag, 2, 5, 5), new MobSpawnInfo.Spawners(EntityType.field_200771_K, 3, 4, 4));
+   private static final List<MobSpawnInfo.Spawners> field_202381_d = com.google.common.collect.Lists.newArrayList(new MobSpawnInfo.Spawners(EntityType.field_200792_f, 10, 2, 3), new MobSpawnInfo.Spawners(EntityType.field_233592_ba_, 5, 4, 4), new MobSpawnInfo.Spawners(EntityType.field_200722_aA, 8, 5, 5), new MobSpawnInfo.Spawners(EntityType.field_200741_ag, 2, 5, 5), new MobSpawnInfo.Spawners(EntityType.field_200771_K, 3, 4, 4));
 
    public FortressStructure(Codec<NoFeatureConfig> p_i231972_1_) {
       super(p_i231972_1_);
