--- a/net/minecraft/world/gen/feature/structure/SwampHutStructure.java
+++ b/net/minecraft/world/gen/feature/structure/SwampHutStructure.java
@@ -13,8 +13,8 @@
 import net.minecraft.world.gen.feature.template.TemplateManager;
 
 public class SwampHutStructure extends Structure<NoFeatureConfig> {
-   private static final List<MobSpawnInfo.Spawners> field_202384_d = ImmutableList.of(new MobSpawnInfo.Spawners(EntityType.field_200759_ay, 1, 1, 1));
-   private static final List<MobSpawnInfo.Spawners> field_214559_aS = ImmutableList.of(new MobSpawnInfo.Spawners(EntityType.field_220360_g, 1, 1, 1));
+   private static final List<MobSpawnInfo.Spawners> field_202384_d = com.google.common.collect.Lists.newArrayList(new MobSpawnInfo.Spawners(EntityType.field_200759_ay, 1, 1, 1));
+   private static final List<MobSpawnInfo.Spawners> field_214559_aS = com.google.common.collect.Lists.newArrayList(new MobSpawnInfo.Spawners(EntityType.field_220360_g, 1, 1, 1));
 
    public SwampHutStructure(Codec<NoFeatureConfig> p_i231998_1_) {
       super(p_i231998_1_);
