--- a/net/minecraft/init/Enchantments.java
+++ b/net/minecraft/init/Enchantments.java
@@ -5,6 +5,7 @@
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.registry.IRegistry;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Enchantments {
    public static final Enchantment field_180310_c = func_185295_a("protection");
    public static final Enchantment field_77329_d = func_185295_a("fire_protection");
