--- a/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
+++ b/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
@@ -12,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class BlockEntityRenderers {
-    private static final Map<BlockEntityType<?>, BlockEntityRendererProvider<?>> PROVIDERS = Maps.newHashMap();
+    private static final Map<BlockEntityType<?>, BlockEntityRendererProvider<?>> PROVIDERS = new java.util.concurrent.ConcurrentHashMap<>();
 
     public static <T extends BlockEntity> void register(BlockEntityType<? extends T> p_173591_, BlockEntityRendererProvider<T> p_173592_) {
         PROVIDERS.put(p_173591_, p_173592_);
