--- a/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
+++ b/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
@@ -14,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class BlockEntityRenderers {
-    private static final Map<BlockEntityType<?>, BlockEntityRendererProvider<?, ?>> PROVIDERS = Maps.newHashMap();
+    private static final Map<BlockEntityType<?>, BlockEntityRendererProvider<?, ?>> PROVIDERS = new java.util.concurrent.ConcurrentHashMap<>();
 
     public static <T extends BlockEntity, S extends BlockEntityRenderState> void register(
         final BlockEntityType<? extends T> type, final BlockEntityRendererProvider<T, S> renderer
