--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -115,7 +_,8 @@
 import net.minecraft.world.scores.Team;
 import org.slf4j.Logger;
 
-public abstract class Player extends LivingEntity {
+public abstract class Player extends LivingEntity implements net.minecraftforge.common.extensions.IForgePlayer {
+   public static final String PERSISTED_NBT_TAG = "PlayerPersisted";
    private static final Logger f_219722_ = LogUtils.getLogger();
    public static final int f_150082_ = 16;
    public static final int f_150083_ = 20;
@@ -168,6 +_,9 @@
    @Nullable
    public FishingHook f_36083_;
    protected float f_263750_;
+   private final java.util.Collection<MutableComponent> prefixes = new java.util.LinkedList<>();
+   private final java.util.Collection<MutableComponent> suffixes = new java.util.LinkedList<>();
+   @Nullable private Pose forcedPose;
 
    public Player(Level p_250508_, BlockPos p_250289_, float p_251702_, GameProfile p_252153_) {
       super(EntityType.f_20532_, p_250508_);
@@ -193,7 +_,7 @@
    }
 
    public static AttributeSupplier.Builder m_36340_() {
-      return LivingEntity.m_21183_().m_22268_(Attributes.f_22281_, 1.0D).m_22268_(Attributes.f_22279_, (double)0.1F).m_22266_(Attributes.f_22283_).m_22266_(Attributes.f_22286_);
+       return LivingEntity.m_21183_().m_22268_(Attributes.f_22281_, 1.0D).m_22268_(Attributes.f_22279_, (double) 0.1F).m_22266_(Attributes.f_22283_).m_22266_(Attributes.f_22286_).m_22266_(net.minecraftforge.common.ForgeMod.BLOCK_REACH.get()).m_22266_(Attributes.f_22282_).m_22266_(net.minecraftforge.common.ForgeMod.ENTITY_REACH.get());
    }
 
    protected void m_8097_() {
@@ -207,6 +_,7 @@
    }
 
    public void m_8119_() {
+      net.minecraftforge.event.ForgeEventFactory.onPlayerPreTick(this);
       this.f_19794_ = this.m_5833_();
       if (this.m_5833_()) {
          this.m_6853_(false);
@@ -222,7 +_,7 @@
             this.f_36110_ = 100;
          }
 
-         if (!this.m_9236_().f_46443_ && this.m_9236_().m_46461_()) {
+         if (!this.m_9236_().f_46443_ && !net.minecraftforge.event.ForgeEventFactory.fireSleepingTimeCheck(this, m_21257_())) {
             this.m_6145_(false, true);
          }
       } else if (this.f_36110_ > 0) {
@@ -277,6 +_,7 @@
       this.m_36372_();
       this.f_36087_.m_41518_();
       this.m_7594_();
+      net.minecraftforge.event.ForgeEventFactory.onPlayerPostTick(this);
    }
 
    public boolean m_36341_() {
@@ -352,6 +_,10 @@
    }
 
    protected void m_7594_() {
+      if(forcedPose != null) {
+         this.m_20124_(forcedPose);
+         return;
+      }
       if (this.m_20175_(Pose.SWIMMING)) {
          Pose pose;
          if (this.m_21255_()) {
@@ -573,6 +_,7 @@
    }
 
    public void m_6667_(DamageSource p_36152_) {
+      if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this, p_36152_)) return;
       super.m_6667_(p_36152_);
       this.m_20090_();
       if (!this.m_5833_()) {
@@ -622,7 +_,7 @@
 
    @Nullable
    public ItemEntity m_36176_(ItemStack p_36177_, boolean p_36178_) {
-      return this.m_7197_(p_36177_, false, p_36178_);
+      return net.minecraftforge.common.ForgeHooks.onPlayerTossEvent(this, p_36177_, p_36178_);
    }
 
    @Nullable
@@ -660,7 +_,12 @@
       }
    }
 
+   @Deprecated //Use location sensitive version below
    public float m_36281_(BlockState p_36282_) {
+      return getDigSpeed(p_36282_, null);
+   }
+
+   public float getDigSpeed(BlockState p_36282_, @Nullable BlockPos pos) {
       float f = this.f_36093_.m_36020_(p_36282_);
       if (f > 1.0F) {
          int i = EnchantmentHelper.m_44926_(this);
@@ -702,11 +_,12 @@
          f /= 5.0F;
       }
 
+      f = net.minecraftforge.event.ForgeEventFactory.getBreakSpeed(this, p_36282_, f, pos);
       return f;
    }
 
    public boolean m_36298_(BlockState p_36299_) {
-      return !p_36299_.m_60834_() || this.f_36093_.m_36056_().m_41735_(p_36299_);
+      return net.minecraftforge.event.ForgeEventFactory.doPlayerHarvestCheck(this, p_36299_, !p_36299_.m_60834_() || this.f_36093_.m_36056_().m_41735_(p_36299_));
    }
 
    public void m_7378_(CompoundTag p_36215_) {
@@ -792,6 +_,7 @@
    }
 
    public boolean m_6469_(DamageSource p_36154_, float p_36155_) {
+      if (!net.minecraftforge.common.ForgeHooks.onPlayerAttack(this, p_36154_, p_36155_)) return false;
       if (this.m_6673_(p_36154_)) {
          return false;
       } else if (this.f_36077_.f_35934_ && !p_36154_.m_269533_(DamageTypeTags.f_268738_)) {
@@ -826,7 +_,7 @@
 
    protected void m_6728_(LivingEntity p_36295_) {
       super.m_6728_(p_36295_);
-      if (p_36295_.m_213824_()) {
+      if (p_36295_.m_21205_().canDisableShield(this.f_20935_, this, p_36295_)) {
          this.m_36384_(true);
       }
 
@@ -855,7 +_,7 @@
    }
 
    protected void m_7909_(float p_36383_) {
-      if (this.f_20935_.m_150930_(Items.f_42740_)) {
+      if (this.f_20935_.canPerformAction(net.minecraftforge.common.ToolActions.SHIELD_BLOCK)) {
          if (!this.m_9236_().f_46443_) {
             this.m_36246_(Stats.f_12982_.m_12902_(this.f_20935_.m_41720_()));
          }
@@ -865,6 +_,8 @@
             InteractionHand interactionhand = this.m_7655_();
             this.f_20935_.m_41622_(i, this, (p_219739_) -> {
                p_219739_.m_21190_(interactionhand);
+               net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, this.f_20935_, interactionhand);
+               m_5810_(); // Forge: fix MC-168573
             });
             if (this.f_20935_.m_41619_()) {
                if (interactionhand == InteractionHand.MAIN_HAND) {
@@ -883,10 +_,13 @@
 
    protected void m_6475_(DamageSource p_36312_, float p_36313_) {
       if (!this.m_6673_(p_36312_)) {
+         p_36313_ = net.minecraftforge.common.ForgeHooks.onLivingHurt(this, p_36312_, p_36313_);
+         if (p_36313_ <= 0) return;
          p_36313_ = this.m_21161_(p_36312_, p_36313_);
          p_36313_ = this.m_6515_(p_36312_, p_36313_);
          float f1 = Math.max(p_36313_ - this.m_6103_(), 0.0F);
          this.m_7911_(this.m_6103_() - (p_36313_ - f1));
+         f1 = net.minecraftforge.common.ForgeHooks.onLivingDamage(this, p_36312_, f1);
          float f = p_36313_ - f1;
          if (f > 0.0F && f < 3.4028235E37F) {
             this.m_36222_(Stats.f_12933_, Math.round(f * 10.0F));
@@ -949,6 +_,8 @@
 
          return InteractionResult.PASS;
       } else {
+         InteractionResult cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntity(this, p_36158_, p_36159_);
+         if (cancelResult != null) return cancelResult;
          ItemStack itemstack = this.m_21120_(p_36159_);
          ItemStack itemstack1 = itemstack.m_41777_();
          InteractionResult interactionresult = p_36158_.m_6096_(this, p_36159_);
@@ -957,6 +_,9 @@
                itemstack.m_41764_(itemstack1.m_41613_());
             }
 
+            if (!this.f_36077_.f_35937_ && itemstack.m_41619_()) {
+               net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, itemstack1, p_36159_);
+            }
             return interactionresult;
          } else {
             if (!itemstack.m_41619_() && p_36158_ instanceof LivingEntity) {
@@ -968,6 +_,7 @@
                if (interactionresult1.m_19077_()) {
                   this.m_9236_().m_214171_(GameEvent.f_223708_, p_36158_.m_20182_(), GameEvent.Context.m_223717_(this));
                   if (itemstack.m_41619_() && !this.f_36077_.f_35937_) {
+                     net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, itemstack1, p_36159_);
                      this.m_21008_(p_36159_, ItemStack.f_41583_);
                   }
 
@@ -997,6 +_,7 @@
       return !this.f_36077_.f_35935_;
    }
 
+   // Forge: Don't update this method to use IForgeEntity#getStepHeight() - https://github.com/MinecraftForge/MinecraftForge/issues/8922
    protected Vec3 m_5763_(Vec3 p_36201_, MoverType p_36202_) {
       if (!this.f_36077_.f_35935_ && p_36201_.f_82480_ <= 0.0D && (p_36202_ == MoverType.SELF || p_36202_ == MoverType.PLAYER) && this.m_36343_() && this.m_36386_()) {
          double d0 = p_36201_.f_82479_;
@@ -1047,11 +_,13 @@
       return p_36201_;
    }
 
+   // Forge: Don't update this method to use IForgeEntity#getStepHeight() - https://github.com/MinecraftForge/MinecraftForge/issues/9376
    private boolean m_36386_() {
       return this.m_20096_() || this.f_19789_ < this.m_274421_() && !this.m_9236_().m_45756_(this, this.m_20191_().m_82386_(0.0D, (double)(this.f_19789_ - this.m_274421_()), 0.0D));
    }
 
    public void m_5706_(Entity p_36347_) {
+      if (!net.minecraftforge.common.ForgeHooks.onPlayerAttackTarget(this, p_36347_)) return;
       if (p_36347_.m_6097_()) {
          if (!p_36347_.m_7313_(this)) {
             float f = (float)this.m_21133_(Attributes.f_22281_);
@@ -1065,11 +_,10 @@
             float f2 = this.m_36403_(0.5F);
             f *= 0.2F + f2 * f2 * 0.8F;
             f1 *= f2;
-            this.m_36334_();
             if (f > 0.0F || f1 > 0.0F) {
                boolean flag = f2 > 0.9F;
                boolean flag1 = false;
-               int i = 0;
+               float i = (float)this.m_21133_(Attributes.f_22282_); // Forge: Initialize this value to the attack knockback attribute of the player, which is by default 0
                i += EnchantmentHelper.m_44894_(this);
                if (this.m_20142_() && flag) {
                   this.m_9236_().m_6263_((Player)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), SoundEvents.f_12314_, this.m_5720_(), 1.0F, 1.0F);
@@ -1079,8 +_,10 @@
 
                boolean flag2 = flag && this.f_19789_ > 0.0F && !this.m_20096_() && !this.m_6147_() && !this.m_20069_() && !this.m_21023_(MobEffects.f_19610_) && !this.m_20159_() && p_36347_ instanceof LivingEntity;
                flag2 = flag2 && !this.m_20142_();
+               net.minecraftforge.event.entity.player.CriticalHitEvent hitResult = net.minecraftforge.common.ForgeHooks.getCriticalHit(this, p_36347_, flag2, flag2 ? 1.5F : 1.0F);
+               flag2 = hitResult != null;
                if (flag2) {
-                  f *= 1.5F;
+                  f *= hitResult.getDamageModifier();
                }
 
                f += f1;
@@ -1088,9 +_,7 @@
                double d0 = (double)(this.f_19787_ - this.f_19867_);
                if (flag && !flag2 && !flag1 && this.m_20096_() && d0 < (double)this.m_6113_()) {
                   ItemStack itemstack = this.m_21120_(InteractionHand.MAIN_HAND);
-                  if (itemstack.m_41720_() instanceof SwordItem) {
-                     flag3 = true;
-                  }
+                  flag3 = itemstack.canPerformAction(net.minecraftforge.common.ToolActions.SWORD_SWEEP);
                }
 
                float f4 = 0.0F;
@@ -1121,8 +_,9 @@
                   if (flag3) {
                      float f3 = 1.0F + EnchantmentHelper.m_44821_(this) * f;
 
-                     for(LivingEntity livingentity : this.m_9236_().m_45976_(LivingEntity.class, p_36347_.m_20191_().m_82377_(1.0D, 0.25D, 1.0D))) {
-                        if (livingentity != this && livingentity != p_36347_ && !this.m_7307_(livingentity) && (!(livingentity instanceof ArmorStand) || !((ArmorStand)livingentity).m_31677_()) && this.m_20280_(livingentity) < 9.0D) {
+                     for(LivingEntity livingentity : this.m_9236_().m_45976_(LivingEntity.class, this.m_21120_(InteractionHand.MAIN_HAND).getSweepHitBox(this, p_36347_))) {
+                        double entityReachSq = Mth.m_144952_(this.getEntityReach()); // Use entity reach instead of constant 9.0. Vanilla uses bottom center-to-center checks here, so don't update this to use canReach, since it uses closest-corner checks.
+                        if (livingentity != this && livingentity != p_36347_ && !this.m_7307_(livingentity) && (!(livingentity instanceof ArmorStand) || !((ArmorStand)livingentity).m_31677_()) && this.m_20280_(livingentity) < entityReachSq) {
                            livingentity.m_147240_((double)0.4F, (double)Mth.m_14031_(this.m_146908_() * ((float)Math.PI / 180F)), (double)(-Mth.m_14089_(this.m_146908_() * ((float)Math.PI / 180F))));
                            livingentity.m_6469_(this.m_269291_().m_269075_(this), f3);
                         }
@@ -1163,13 +_,15 @@
                   EnchantmentHelper.m_44896_(this, p_36347_);
                   ItemStack itemstack1 = this.m_21205_();
                   Entity entity = p_36347_;
-                  if (p_36347_ instanceof EnderDragonPart) {
-                     entity = ((EnderDragonPart)p_36347_).f_31010_;
+                  if (p_36347_ instanceof net.minecraftforge.entity.PartEntity) {
+                     entity = ((net.minecraftforge.entity.PartEntity<?>) p_36347_).getParent();
                   }
 
                   if (!this.m_9236_().f_46443_ && !itemstack1.m_41619_() && entity instanceof LivingEntity) {
+                     ItemStack copy = itemstack1.m_41777_();
                      itemstack1.m_41640_((LivingEntity)entity, this);
                      if (itemstack1.m_41619_()) {
+                        net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, copy, InteractionHand.MAIN_HAND);
                         this.m_21008_(InteractionHand.MAIN_HAND, ItemStack.f_41583_);
                      }
                   }
@@ -1195,6 +_,7 @@
                   }
                }
             }
+            this.m_36334_(); // FORGE: Moved from beginning of attack() so that getAttackStrengthScale() returns an accurate value during all attack events
 
          }
       }
@@ -1211,7 +_,7 @@
       }
 
       if (this.f_19796_.m_188501_() < f) {
-         this.m_36335_().m_41524_(Items.f_42740_, 100);
+         this.m_36335_().m_41524_(this.m_21211_().m_41720_(), 100);
          this.m_5810_();
          this.m_9236_().m_7605_(this, (byte)30);
       }
@@ -1275,6 +_,7 @@
    }
 
    public void m_6145_(boolean p_36226_, boolean p_36227_) {
+      net.minecraftforge.event.ForgeEventFactory.onPlayerWakeup(this, p_36226_, p_36227_);
       super.m_5796_();
       if (this.m_9236_() instanceof ServerLevel && p_36227_) {
          ((ServerLevel)this.m_9236_()).m_8878_();
@@ -1300,7 +_,7 @@
       } else if (block instanceof BedBlock && BedBlock.m_49488_(p_36131_)) {
          return BedBlock.m_260958_(EntityType.f_20532_, p_36131_, p_36132_, blockstate.m_61143_(BedBlock.f_54117_), p_36133_);
       } else if (!p_36134_) {
-         return Optional.empty();
+         return blockstate.getRespawnPosition(EntityType.f_20532_, p_36131_, p_36132_, p_36133_, null);
       } else {
          boolean flag = block.m_48673_(blockstate);
          BlockState blockstate1 = p_36131_.m_8055_(p_36132_.m_7494_());
@@ -1481,6 +_,7 @@
 
    public boolean m_142535_(float p_150093_, float p_150094_, DamageSource p_150095_) {
       if (this.f_36077_.f_35936_) {
+         net.minecraftforge.event.ForgeEventFactory.onPlayerFall(this, p_150093_, p_150093_);
          return false;
       } else {
          if (p_150093_ >= 2.0F) {
@@ -1494,7 +_,7 @@
    public boolean m_36319_() {
       if (!this.m_20096_() && !this.m_21255_() && !this.m_20069_() && !this.m_21023_(MobEffects.f_19620_)) {
          ItemStack itemstack = this.m_6844_(EquipmentSlot.CHEST);
-         if (itemstack.m_150930_(Items.f_42741_) && ElytraItem.m_41140_(itemstack)) {
+         if (itemstack.canElytraFly(this)) {
             this.m_36320_();
             return true;
          }
@@ -1522,13 +_,13 @@
    protected void m_7355_(BlockPos p_282121_, BlockState p_282194_) {
       if (this.m_20069_()) {
          this.m_280447_();
-         this.m_280568_(p_282194_);
+         this.playMuffledStepSound(p_282194_, p_282121_);
       } else {
          BlockPos blockpos = this.m_276951_(p_282121_);
          if (!p_282121_.equals(blockpos)) {
             BlockState blockstate = this.m_9236_().m_8055_(blockpos);
             if (blockstate.m_204336_(BlockTags.f_276549_)) {
-               this.m_276961_(blockstate, p_282194_);
+               this.playCombinationStepSounds(blockstate, p_282194_, blockpos, p_282121_);
             } else {
                super.m_7355_(blockpos, blockstate);
             }
@@ -1556,6 +_,10 @@
    }
 
    public void m_6756_(int p_36291_) {
+      net.minecraftforge.event.entity.player.PlayerXpEvent.XpChange event = new net.minecraftforge.event.entity.player.PlayerXpEvent.XpChange(this, p_36291_);
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return;
+      p_36291_ = event.getAmount();
+
       this.m_36401_(p_36291_);
       this.f_36080_ += (float)p_36291_ / (float)this.m_36323_();
       this.f_36079_ = Mth.m_14045_(this.f_36079_ + p_36291_, 0, Integer.MAX_VALUE);
@@ -1584,7 +_,7 @@
    }
 
    public void m_7408_(ItemStack p_36172_, int p_36173_) {
-      this.f_36078_ -= p_36173_;
+      m_6749_(-p_36173_);
       if (this.f_36078_ < 0) {
          this.f_36078_ = 0;
          this.f_36080_ = 0.0F;
@@ -1595,6 +_,10 @@
    }
 
    public void m_6749_(int p_36276_) {
+      net.minecraftforge.event.entity.player.PlayerXpEvent.LevelChange event = new net.minecraftforge.event.entity.player.PlayerXpEvent.LevelChange(this, p_36276_);
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return;
+      p_36276_ = event.getLevels();
+
       this.f_36078_ += p_36276_;
       if (this.f_36078_ < 0) {
          this.f_36078_ = 0;
@@ -1790,7 +_,11 @@
    }
 
    public Component m_5446_() {
-      MutableComponent mutablecomponent = PlayerTeam.m_83348_(this.m_5647_(), this.m_7755_());
+      if (this.displayname == null) this.displayname = net.minecraftforge.event.ForgeEventFactory.getPlayerDisplayName(this, this.m_7755_());
+      MutableComponent mutablecomponent = Component.m_237113_("");
+      mutablecomponent = prefixes.stream().reduce(mutablecomponent, MutableComponent::m_7220_);
+      mutablecomponent = mutablecomponent.m_7220_(PlayerTeam.m_83348_(this.m_5647_(), this.displayname));
+      mutablecomponent = suffixes.stream().reduce(mutablecomponent, MutableComponent::m_7220_);
       return this.m_36218_(mutablecomponent);
    }
 
@@ -1927,24 +_,24 @@
          Predicate<ItemStack> predicate = ((ProjectileWeaponItem)p_36349_.m_41720_()).m_6442_();
          ItemStack itemstack = ProjectileWeaponItem.m_43010_(this, predicate);
          if (!itemstack.m_41619_()) {
-            return itemstack;
+            return net.minecraftforge.common.ForgeHooks.getProjectile(this, p_36349_, itemstack);
          } else {
             predicate = ((ProjectileWeaponItem)p_36349_.m_41720_()).m_6437_();
 
             for(int i = 0; i < this.f_36093_.m_6643_(); ++i) {
                ItemStack itemstack1 = this.f_36093_.m_8020_(i);
                if (predicate.test(itemstack1)) {
-                  return itemstack1;
+                  return net.minecraftforge.common.ForgeHooks.getProjectile(this, p_36349_, itemstack1);
                }
             }
 
-            return this.f_36077_.f_35937_ ? new ItemStack(Items.f_42412_) : ItemStack.f_41583_;
+            return net.minecraftforge.common.ForgeHooks.getProjectile(this, p_36349_, this.f_36077_.f_35937_ ? new ItemStack(Items.f_42412_) : ItemStack.f_41583_);
          }
       }
    }
 
    public ItemStack m_5584_(Level p_36185_, ItemStack p_36186_) {
-      this.m_36324_().m_38712_(p_36186_.m_41720_(), p_36186_);
+      this.m_36324_().eat(p_36186_.m_41720_(), p_36186_, this);
       this.m_36246_(Stats.f_12982_.m_12902_(p_36186_.m_41720_()));
       p_36185_.m_6263_((Player)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), SoundEvents.f_12321_, SoundSource.PLAYERS, 0.5F, p_36185_.f_46441_.m_188501_() * 0.1F + 0.9F);
       if (this instanceof ServerPlayer) {
@@ -2052,5 +_,63 @@
       public Component m_36423_() {
          return this.f_36413_;
       }
+   }
+
+   // =========== FORGE START ==============//
+   public Collection<MutableComponent> getPrefixes() {
+       return this.prefixes;
+   }
+
+   public Collection<MutableComponent> getSuffixes() {
+       return this.suffixes;
+   }
+
+   private Component displayname = null;
+   /**
+    * Force the displayed name to refresh, by firing {@link net.minecraftforge.event.entity.player.PlayerEvent.NameFormat}, using the real player name as event parameter.
+    */
+   public void refreshDisplayName() {
+      this.displayname = net.minecraftforge.event.ForgeEventFactory.getPlayerDisplayName(this, this.m_7755_());
+   }
+
+   private final net.minecraftforge.common.util.LazyOptional<net.minecraftforge.items.IItemHandler>
+         playerMainHandler = net.minecraftforge.common.util.LazyOptional.of(
+               () -> new net.minecraftforge.items.wrapper.PlayerMainInvWrapper(f_36093_));
+
+   private final net.minecraftforge.common.util.LazyOptional<net.minecraftforge.items.IItemHandler>
+         playerEquipmentHandler = net.minecraftforge.common.util.LazyOptional.of(
+               () -> new net.minecraftforge.items.wrapper.CombinedInvWrapper(
+                     new net.minecraftforge.items.wrapper.PlayerArmorInvWrapper(f_36093_),
+                     new net.minecraftforge.items.wrapper.PlayerOffhandInvWrapper(f_36093_)));
+
+   private final net.minecraftforge.common.util.LazyOptional<net.minecraftforge.items.IItemHandler>
+         playerJoinedHandler = net.minecraftforge.common.util.LazyOptional.of(
+               () -> new net.minecraftforge.items.wrapper.PlayerInvWrapper(f_36093_));
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+      if (this.m_6084_() && capability == net.minecraftforge.common.capabilities.ForgeCapabilities.ITEM_HANDLER) {
+         if (facing == null) return playerJoinedHandler.cast();
+         else if (facing.m_122434_().m_122478_()) return playerMainHandler.cast();
+         else if (facing.m_122434_().m_122479_()) return playerEquipmentHandler.cast();
+      }
+      return super.getCapability(capability, facing);
+   }
+
+   /**
+    * Force a pose for the player. If set, the vanilla pose determination and clearance check is skipped. Make sure the pose is clear yourself (e.g. in PlayerTick).
+    * This has to be set just once, do not set it every tick.
+    * Make sure to clear (null) the pose if not required anymore and only use if necessary.
+    */
+   public void setForcedPose(@Nullable Pose pose) {
+      this.forcedPose = pose;
+   }
+
+   /**
+    * @return The forced pose if set, null otherwise
+    */
+   @Nullable
+   public Pose getForcedPose() {
+      return this.forcedPose;
    }
 }
