--- a/net/minecraft/client/renderer/block/dispatch/BlockStateModel.java
+++ b/net/minecraft/client/renderer/block/dispatch/BlockStateModel.java
@@ -20,9 +_,11 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface BlockStateModel {
+public interface BlockStateModel extends net.minecraftforge.client.extensions.IForgeBlockStateModel {
+    /**@deprecated Forge: Use {@link net.minecraftforge.client.extensions.IForgeBlockStateModel#collectParts(RandomSource, List<BlockStateModelPart>, net.minecraftforge.client.model.data.ModelData)}*/
     void collectParts(RandomSource random, List<BlockStateModelPart> output);
 
+    /**@deprecated Forge: Use {@link net.minecraftforge.client.extensions.IForgeBlockStateModel#particleMaterial(net.minecraftforge.client.model.data.ModelData)}*/
     Material.Baked particleMaterial();
 
     @BakedQuad.MaterialFlags int materialFlags();
