--- a/net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
+++ b/net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
@@ -21,6 +_,7 @@
    });
    public static final LootContextParamSet CHEST = register("chest", (p_81452_) -> {
       p_81452_.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY);
+      p_81452_.optional(LootContextParams.KILLER_ENTITY); //Forge: Chest Minecarts can have killers.
    });
    public static final LootContextParamSet COMMAND = register("command", (p_81450_) -> {
       p_81450_.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY);
@@ -30,6 +_,7 @@
    });
    public static final LootContextParamSet FISHING = register("fishing", (p_81446_) -> {
       p_81446_.required(LootContextParams.ORIGIN).required(LootContextParams.TOOL).optional(LootContextParams.THIS_ENTITY);
+      p_81446_.optional(LootContextParams.KILLER_ENTITY).optional(LootContextParams.THIS_ENTITY); //Forge: Allow fisher, and bobber
    });
    public static final LootContextParamSet ENTITY = register("entity", (p_81444_) -> {
       p_81444_.required(LootContextParams.THIS_ENTITY).required(LootContextParams.ORIGIN).required(LootContextParams.DAMAGE_SOURCE).optional(LootContextParams.KILLER_ENTITY).optional(LootContextParams.DIRECT_KILLER_ENTITY).optional(LootContextParams.LAST_DAMAGE_PLAYER);
