--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -46,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
-public abstract class AbstractMinecart extends VehicleEntity {
+public abstract class AbstractMinecart extends VehicleEntity implements net.minecraftforge.common.extensions.IForgeAbstractMinecart {
     private static final Vec3 LOWERED_PASSENGER_ATTACHMENT = new Vec3(0.0, 0.0, 0.0);
     private static final EntityDataAccessor<Optional<BlockState>> DATA_ID_CUSTOM_DISPLAY_BLOCK = SynchedEntityData.defineId(
         AbstractMinecart.class, EntityDataSerializers.OPTIONAL_BLOCK_STATE
@@ -96,15 +_,12 @@
             }
         )
     );
+    private boolean canBePushed = true;
 
     protected AbstractMinecart(final EntityType<?> type, final Level level) {
         super(type, level);
         this.blocksBuilding = true;
-        if (useExperimentalMovement(level)) {
-            this.behavior = new NewMinecartBehavior(this);
-        } else {
-            this.behavior = new OldMinecartBehavior(this);
-        }
+        this.behavior = behaviorFactory.apply(this, level);
     }
 
     protected AbstractMinecart(final EntityType<?> type, final Level level, final double x, final double y, final double z) {
@@ -166,7 +_,7 @@
 
     @Override
     public boolean isPushable() {
-        return true;
+        return canBePushed;
     }
 
     @Override
@@ -367,22 +_,31 @@
     }
 
     protected void comeOffTrack(final ServerLevel level) {
-        double maxSpeed = this.getMaxSpeed(level);
+        double maxSpeed = this.onGround() ? this.getMaxSpeed(level) : getMaxSpeedAirLateral();
         Vec3 movement = this.getDeltaMovement();
         this.setDeltaMovement(Mth.clamp(movement.x, -maxSpeed, maxSpeed), movement.y, Mth.clamp(movement.z, -maxSpeed, maxSpeed));
         if (this.onGround()) {
             this.setDeltaMovement(this.getDeltaMovement().scale(0.5));
         }
 
+        if (getMaxSpeedAirVertical() > 0 && getDeltaMovement().y > getMaxSpeedAirVertical()) {
+            if (Math.abs(getDeltaMovement().x) < 0.3f && Math.abs(getDeltaMovement().z) < 0.3f) {
+                setDeltaMovement(new Vec3(getDeltaMovement().x, 0.15f, getDeltaMovement().z));
+            } else {
+                setDeltaMovement(new Vec3(getDeltaMovement().x, getMaxSpeedAirVertical(), getDeltaMovement().z));
+            }
+        }
+
         this.move(MoverType.SELF, this.getDeltaMovement());
         if (!this.onGround()) {
             this.setDeltaMovement(this.getDeltaMovement().scale(this.getAirDrag()));
         }
     }
 
+    private float airDrag = DEFAULT_AIR_DRAG;
     @Override
     protected float getAirDrag() {
-        return 0.95F;
+        return airDrag;
     }
 
     protected double makeStepAlongTrack(final BlockPos pos, final RailShape shape, final double movementLeft) {
@@ -438,7 +_,7 @@
     public Vec3 getRedstoneDirection(final BlockPos pos) {
         BlockState state = this.level().getBlockState(pos);
         if (state.is(Blocks.POWERED_RAIL) && state.getValue(PoweredRailBlock.POWERED)) {
-            RailShape shape = state.getValue(((BaseRailBlock)state.getBlock()).getShapeProperty());
+            RailShape shape = ((BaseRailBlock)state.getBlock()).getRailDirection(state, this.level(), pos, this);
             if (shape == RailShape.EAST_WEST) {
                 if (this.isRedstoneConductor(pos.west())) {
                     return new Vec3(1.0, 0.0, 0.0);
@@ -609,5 +_,40 @@
 
     public boolean isFurnace() {
         return false;
+    }
+
+    private static java.util.function.BiFunction<AbstractMinecart, Level, MinecartBehavior> behaviorFactory = (cart, level) -> {
+        return useExperimentalMovement(level) ? new NewMinecartBehavior(cart) : new OldMinecartBehavior(cart);
+    };
+    public static void setBehaviorFactory(java.util.function.BiFunction<AbstractMinecart, Level, MinecartBehavior> factory) { behaviorFactory = factory; }
+    public static java.util.function.BiFunction<AbstractMinecart, Level, MinecartBehavior> getBehaviorFactory() { return behaviorFactory; }
+
+              private boolean canUseRail = true;
+    @Override public  boolean canUseRail() { return canUseRail; }
+    @Override public  void    setCanUseRail(boolean value) { this.canUseRail = value; }
+              private float    currentSpeedCapOnRail = getMaxCartSpeedOnRail();
+    @Override public  float getCurrentCartSpeedCapOnRail() { return currentSpeedCapOnRail; }
+    @Override public  void  setCurrentCartSpeedCapOnRail(float value) { currentSpeedCapOnRail = Math.min(value, getMaxCartSpeedOnRail()); }
+              private Float    maxSpeedAirLateral = null;
+    @Override public  float getMaxSpeedAirLateral() { return maxSpeedAirLateral == null ? (float)this.getMaxSpeed((ServerLevel)this.level()) : maxSpeedAirLateral; }
+    @Override public  void  setMaxSpeedAirLateral(float value) { maxSpeedAirLateral = value; }
+              private float    maxSpeedAirVertical = DEFAULT_MAX_SPEED_AIR_VERTICAL;
+    @Override public  float getMaxSpeedAirVertical() { return maxSpeedAirVertical; }
+    @Override public  void  setMaxSpeedAirVertical(float value) { maxSpeedAirVertical = value; }
+    @Override public  void  setAirDrag(float value) { airDrag = value; }
+    @Override
+    public double getMaxSpeedWithRail() { //Non-default because getMaximumSpeed is protected
+        if (!canUseRail()) {
+            return getMaxSpeed((ServerLevel)this.level());
+        }
+        BlockPos pos = this.getCurrentRailPosition();
+        BlockState state = this.level().getBlockState(pos);
+
+        if (!state.is(BlockTags.RAILS)) {
+            return getMaxSpeed((ServerLevel)this.level());
+        }
+
+        float railMaxSpeed = ((BaseRailBlock)state.getBlock()).getRailMaxSpeed(state, this.level(), pos, this);
+        return Math.min(railMaxSpeed, getCurrentCartSpeedCapOnRail());
     }
 }
