--- a/net/minecraft/world/entity/vehicle/minecart/MinecartSpawner.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartSpawner.java
@@ -19,6 +_,11 @@
         public void broadcastEvent(final Level level, final BlockPos pos, final int id) {
             level.broadcastEntityEvent(MinecartSpawner.this, (byte)id);
         }
+
+        @Override
+        public net.minecraft.world.entity.@org.jspecify.annotations.Nullable Entity getSpawnerEntity() {
+            return MinecartSpawner.this;
+        }
     };
     private final Runnable ticker;
 
