--- a/net/minecraft/world/entity/vehicle/minecart/MinecartCommandBlock.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartCommandBlock.java
@@ -87,6 +_,8 @@
 
     @Override
     public InteractionResult interact(final Player player, final InteractionHand hand, final Vec3 location) {
+        InteractionResult ret = super.interact(player, hand, location);
+        if (ret.consumesAction()) return ret;
         if (!player.canUseGameMasterBlocks()) {
             return InteractionResult.PASS;
         }
