--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -153,7 +_,7 @@
 
     public void adjustToRails(final BlockPos targetBlockPos, final BlockState currentState, final boolean instant) {
         if (BaseRailBlock.isRail(currentState)) {
-            RailShape shape = currentState.getValue(((BaseRailBlock)currentState.getBlock()).getShapeProperty());
+            RailShape shape = ((BaseRailBlock)currentState.getBlock()).getRailDirection(currentState, this.level(), targetBlockPos, this.minecart);
             Pair<Vec3i, Vec3i> exits = AbstractMinecart.exits(shape);
             Vec3 exit0 = new Vec3(exits.getFirst()).scale(0.5);
             Vec3 exit1 = new Vec3(exits.getSecond()).scale(0.5);
@@ -246,12 +_,12 @@
             if (onRails) {
                 this.minecart.resetFallDistance();
                 this.minecart.setOldPosAndRot();
-                if (currentState.is(Blocks.ACTIVATOR_RAIL)) {
+                if (currentState.getBlock() instanceof PoweredRailBlock power && power.isActivatorRail()) {
                     this.minecart
                         .activateMinecart(level, currentPos.getX(), currentPos.getY(), currentPos.getZ(), currentState.getValue(PoweredRailBlock.POWERED));
                 }
 
-                RailShape shape = currentState.getValue(((BaseRailBlock)currentState.getBlock()).getShapeProperty());
+                RailShape shape = ((BaseRailBlock)currentState.getBlock()).getRailDirection(currentState, this.level(), currentPos, this.minecart);
                 Vec3 newDeltaMovement = this.calculateTrackSpeed(level, initialStepDeltaMovement.horizontal(), trackIteration, currentPos, currentState, shape);
                 if (trackIteration.firstIteration) {
                     trackIteration.movementLeft = newDeltaMovement.horizontalDistance();
@@ -386,7 +_,7 @@
     }
 
     private Vec3 calculateHaltTrackSpeed(final Vec3 deltaMovement, final BlockState state) {
-        if (state.is(Blocks.POWERED_RAIL) && !state.getValue(PoweredRailBlock.POWERED)) {
+        if (state.getBlock() instanceof PoweredRailBlock powered && !powered.isActivatorRail() && !state.getValue(PoweredRailBlock.POWERED)) {
             return deltaMovement.length() < 0.03 ? Vec3.ZERO : deltaMovement.scale(0.5);
         } else {
             return deltaMovement;
@@ -394,7 +_,7 @@
     }
 
     private Vec3 calculateBoostTrackSpeed(final Vec3 deltaMovement, final BlockPos pos, final BlockState state) {
-        if (state.is(Blocks.POWERED_RAIL) && state.getValue(PoweredRailBlock.POWERED)) {
+        if (state.getBlock() instanceof PoweredRailBlock powered && !powered.isActivatorRail() && state.getValue(PoweredRailBlock.POWERED)) {
             if (deltaMovement.length() > 0.01) {
                 return deltaMovement.normalize().scale(deltaMovement.length() + 0.06);
             }
@@ -448,7 +_,7 @@
         BlockState newBlockState = this.level().getBlockState(BlockPos.containing(newPosition));
         if (inHill) {
             if (BaseRailBlock.isRail(newBlockState)) {
-                RailShape newRailShape = newBlockState.getValue(((BaseRailBlock)newBlockState.getBlock()).getShapeProperty());
+                RailShape newRailShape = ((BaseRailBlock)newBlockState.getBlock()).getRailDirection(newBlockState, this.level(), BlockPos.containing(newPosition), this.minecart);
                 if (this.restAtVShape(shape, newRailShape)) {
                     return 0.0;
                 }
