--- a/net/minecraft/world/level/storage/loot/LootDataManager.java
+++ b/net/minecraft/world/level/storage/loot/LootDataManager.java
@@ -53,7 +_,7 @@
          Map<ResourceLocation, JsonElement> map1 = new HashMap<>();
          SimpleJsonResourceReloadListener.m_278771_(p_279441_, p_279205_.m_278624_(), p_279205_.m_278857_(), map1);
          map1.forEach((p_279416_, p_279151_) -> {
-            p_279205_.m_278763_(p_279416_, p_279151_).ifPresent((p_279295_) -> {
+            p_279205_.deserialize(p_279416_, p_279151_, p_279441_).ifPresent((p_279295_) -> {
                map.put(p_279416_, p_279295_);
             });
          });
