--- a/net/minecraft/world/level/storage/loot/LootParams.java
+++ b/net/minecraft/world/level/storage/loot/LootParams.java
@@ -125,7 +_,7 @@
 
         public LootParams create(LootContextParamSet p_287701_) {
             Set<LootContextParam<?>> set = Sets.difference(this.params.keySet(), p_287701_.getAllowed());
-            if (!set.isEmpty()) {
+            if (false && !set.isEmpty()) { // Forge: Allow mods to pass custom loot parameters (not part of the vanilla loot table) to the loot context.
                 throw new IllegalArgumentException("Parameters not allowed in this parameter set: " + set);
             } else {
                 Set<LootContextParam<?>> set1 = Sets.difference(p_287701_.getRequired(), this.params.keySet());
