--- a/net/minecraft/block/KelpTopBlock.java
+++ b/net/minecraft/block/KelpTopBlock.java
@@ -52,8 +52,9 @@
       } else {
          BlockPos blockpos = p_225534_3_.func_177984_a();
          BlockState blockstate = p_225534_2_.func_180495_p(blockpos);
-         if (blockstate.func_177230_c() == Blocks.field_150355_j && p_225534_1_.func_177229_b(field_203163_a) < 25 && p_225534_4_.nextDouble() < 0.14D) {
+         if (blockstate.func_177230_c() == Blocks.field_150355_j && p_225534_1_.func_177229_b(field_203163_a) < 25 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_225534_2_, blockpos, p_225534_1_, p_225534_4_.nextDouble() < 0.14D)) {
             p_225534_2_.func_175656_a(blockpos, p_225534_1_.func_177231_a(field_203163_a));
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_225534_2_, blockpos, p_225534_1_);
          }
 
       }
