--- a/net/minecraft/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/entity/ai/goal/GoalSelector.java
@@ -27,16 +27,29 @@
    private final Supplier<IProfiler> field_75781_c;
    private final EnumSet<Goal.Flag> field_220893_f = EnumSet.noneOf(Goal.Flag.class);
    private int field_75779_e = 3;
+   private boolean isTicking;
+   private final Set<PrioritizedGoal> asyncAddedGoals = Sets.newLinkedHashSet();
+   private final Set<Goal> asyncRemovedGoals = Sets.newLinkedHashSet();
 
    public GoalSelector(Supplier<IProfiler> p_i231546_1_) {
       this.field_75781_c = p_i231546_1_;
    }
 
    public void func_75776_a(int p_75776_1_, Goal p_75776_2_) {
+      if (isTicking) {
+         asyncAddedGoals.add(new PrioritizedGoal(p_75776_1_, p_75776_2_));
+         asyncRemovedGoals.remove(p_75776_2_);
+         return;
+      }
       this.field_220892_d.add(new PrioritizedGoal(p_75776_1_, p_75776_2_));
    }
 
    public void func_85156_a(Goal p_85156_1_) {
+      if (isTicking) {
+         asyncRemovedGoals.add(p_85156_1_);
+         asyncAddedGoals.removeIf(prioritizedGoal -> prioritizedGoal.func_220772_j() == p_85156_1_);
+         return;
+      }
       this.field_220892_d.stream().filter((p_220882_1_) -> {
          return p_220882_1_.func_220772_j() == p_85156_1_;
       }).filter(PrioritizedGoal::func_220773_g).forEach(PrioritizedGoal::func_75251_c);
@@ -77,7 +90,14 @@
       });
       iprofiler.func_76319_b();
       iprofiler.func_76320_a("goalTick");
+      isTicking = true;
       this.func_220888_c().forEach(PrioritizedGoal::func_75246_d);
+      // Forge: Correctly handle goals added/removed during goal ticking, see #7509
+      isTicking = false;
+      for (PrioritizedGoal added : asyncAddedGoals) func_75776_a(added.func_220770_h(), added.func_220772_j());
+      asyncAddedGoals.clear();
+      for (Goal removed : asyncRemovedGoals) func_85156_a(removed);
+      asyncRemovedGoals.clear();
       iprofiler.func_76319_b();
    }
 
