--- a/net/minecraft/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/entity/ai/goal/RangedBowAttackGoal.java
@@ -36,7 +_,7 @@
    }
 
    protected boolean isHoldingBow() {
-      return this.mob.isHolding(Items.BOW);
+      return this.mob.isHolding(item -> item instanceof BowItem);
    }
 
    public boolean canContinueToUse() {
@@ -117,7 +_,7 @@
                }
             }
          } else if (--this.attackTime <= 0 && this.seeTime >= -60) {
-            this.mob.startUsingItem(ProjectileHelper.getWeaponHoldingHand(this.mob, Items.BOW));
+            this.mob.startUsingItem(ProjectileHelper.getWeaponHoldingHand(this.mob, item -> item instanceof BowItem));
          }
 
       }
