--- ../src-base/minecraft/net/minecraft/command/CommandHandler.java
+++ ../src-work/minecraft/net/minecraft/command/CommandHandler.java
@@ -48,6 +48,16 @@
         }
         else if (icommand.canCommandSenderUse(sender))
         {
+            net.minecraftforge.event.CommandEvent event = new net.minecraftforge.event.CommandEvent(icommand, sender, astring);
+            if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event))
+            {
+                if (event.exception != null)
+                {
+                    com.google.common.base.Throwables.propagateIfPossible(event.exception);
+                }
+                return 1;
+            }
+
             if (i > -1)
             {
                 List list = PlayerSelector.matchEntities(sender, astring[i], Entity.class);
