--- a/net/minecraft/util/registry/Bootstrap.java
+++ b/net/minecraft/util/registry/Bootstrap.java
@@ -48,6 +48,8 @@
                IDispenseItemBehavior.func_218401_c();
                ArgumentTypes.func_197483_a();
                TagRegistryManager.func_242197_b();
+               net.minecraftforge.registries.GameData.vanillaSnapshot();
+               if (false) // Forge: skip redirectOutputToLog, Forge already redirects stdout and stderr output to log so that they print with more context
                func_179868_d();
             }
          }
@@ -103,7 +105,6 @@
             Commands.func_242986_b();
          }
 
-         GlobalEntityTypeAttributes.func_233834_a_();
       }
    }
 
