--- a/net/minecraft/util/registry/DynamicRegistries.java
+++ b/net/minecraft/util/registry/DynamicRegistries.java
@@ -90,6 +_,8 @@
       RegistryKey<? extends Registry<E>> registrykey = p_243607_2_.key();
       boolean flag = !registrykey.equals(Registry.NOISE_GENERATOR_SETTINGS_REGISTRY) && !registrykey.equals(Registry.DIMENSION_TYPE_REGISTRY);
       Registry<E> registry = BUILTIN.registryOrThrow(registrykey);
+      if (!registrykey.equals(Registry.DIMENSION_TYPE_REGISTRY))
+         registry = ((Registry<Registry<E>>)WorldGenRegistries.REGISTRY).get((RegistryKey<Registry<E>>)registrykey);
       MutableRegistry<E> mutableregistry = p_243607_0_.registryOrThrow(registrykey);
 
       for(Entry<RegistryKey<E>, E> entry : registry.entrySet()) {
