--- ../src-base/minecraft/net/minecraft/util/registry/RegistryNamespacedDefaultedByKey.java
+++ ../src-work/minecraft/net/minecraft/util/registry/RegistryNamespacedDefaultedByKey.java
@@ -63,4 +63,11 @@
         V v = super.func_186801_a(p_186801_1_);
         return (V)(v == null ? this.field_148761_e : v);
     }
+
+    //Bypass functions to allow querying this registry WITHOUT getting the defaulted value.
+    // MODDERS DO NOT USE THIS IS FOR FORGE INTERNAL CHECKS
+    public int getIDForObjectBypass(V bypass) { return super.func_148757_b(bypass); }
+    public K getNameForObjectBypass(V value) { return super.func_177774_c(value); }
+    public V getObjectBypass(K name) { return super.func_82594_a(name); }
+    public V getObjectByIdBypass(int id){ return super.func_148754_a(id); }
 }
