--- a/net/minecraft/util/registry/SimpleRegistry.java
+++ b/net/minecraft/util/registry/SimpleRegistry.java
@@ -195,7 +195,7 @@
             simpleregistry.func_218382_a(entry.field_243547_b, entry.field_243546_a, entry.field_243548_c, p_243539_1_);
          }
 
-         return simpleregistry;
+         return net.minecraftforge.registries.GameData.setDynamicRegistry(simpleregistry);
       }, (p_243544_0_) -> {
          Builder<SimpleRegistry.Entry<T>> builder = ImmutableList.builder();
 
