--- a/net/minecraft/util/registry/WorldGenRegistries.java
+++ b/net/minecraft/util/registry/WorldGenRegistries.java
@@ -45,7 +45,7 @@
       return ProcessorLists.field_244102_b;
    });
    public static final Registry<JigsawPattern> field_243656_h = func_243667_a(Registry.field_243555_ax, JigsawPatternRegistry::func_244093_a);
-   public static final Registry<Biome> field_243657_i = func_243667_a(Registry.field_239720_u_, () -> {
+   /**@deprecated Use {@link net.minecraftforge.registries.ForgeRegistries#BIOMES}*/@Deprecated public static final Registry<Biome> field_243657_i = forge(Registry.field_239720_u_, () -> {
       return BiomeRegistry.field_244200_a;
    });
    public static final Registry<DimensionSettings> field_243658_j = func_243667_a(Registry.field_243549_ar, DimensionSettings::func_242746_i);
@@ -54,6 +54,10 @@
       return func_243665_a(p_243667_0_, Lifecycle.stable(), p_243667_1_);
    }
 
+   private static <T extends net.minecraftforge.registries.IForgeRegistryEntry<T>> Registry<T> forge(RegistryKey<? extends Registry<T>> key, Supplier<T> def) {
+      return func_243666_a(key, net.minecraftforge.registries.GameData.getWrapper(key, Lifecycle.stable()), def, Lifecycle.stable());
+   }
+
    private static <T> Registry<T> func_243665_a(RegistryKey<? extends Registry<T>> p_243665_0_, Lifecycle p_243665_1_, Supplier<T> p_243665_2_) {
       return func_243666_a(p_243665_0_, new SimpleRegistry<>(p_243665_0_, p_243665_1_), p_243665_2_, p_243665_1_);
    }
