--- a/net/minecraft/world/entity/animal/Cat.java
+++ b/net/minecraft/world/entity/animal/Cat.java
@@ -379,9 +_,9 @@
                     }
                 } else if (this.isFood(itemstack) && this.getHealth() < this.getMaxHealth()) {
                     if (!this.level().isClientSide()) {
-                        this.usePlayerItem(p_28153_, p_28154_, itemstack);
                         FoodProperties foodproperties = itemstack.get(DataComponents.FOOD);
                         this.heal(foodproperties != null ? (float)foodproperties.nutrition() : 1.0F);
+                        this.usePlayerItem(p_28153_, p_28154_, itemstack);
                     }
 
                     return InteractionResult.sidedSuccess(this.level().isClientSide());
@@ -441,7 +_,7 @@
     }
 
     private void tryToTame(Player p_333297_) {
-        if (this.random.nextInt(3) == 0) {
+        if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_333297_)) {
             this.tame(p_333297_);
             this.setOrderedToSit(true);
             this.level().broadcastEntityEvent(this, (byte)7);
