--- a/net/minecraft/world/entity/animal/Parrot.java
+++ b/net/minecraft/world/entity/animal/Parrot.java
@@ -245,7 +_,7 @@
          }
 
          if (!this.m_9236_().f_46443_) {
-            if (this.f_19796_.m_188503_(10) == 0) {
+            if (this.f_19796_.m_188503_(10) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_29414_)) {
                this.m_21828_(p_29414_);
                this.m_9236_().m_7605_(this, (byte)7);
             } else {
