--- a/net/minecraft/world/entity/animal/Pig.java
+++ b/net/minecraft/world/entity/animal/Pig.java
@@ -203,7 +_,7 @@
    }
 
    public void thunderHit(ServerLevel p_29473_, LightningBolt p_29474_) {
-      if (p_29473_.getDifficulty() != Difficulty.PEACEFUL) {
+      if (p_29473_.getDifficulty() != Difficulty.PEACEFUL && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.ZOMBIFIED_PIGLIN, (timer) -> {})) {
          ZombifiedPiglin zombifiedpiglin = EntityType.ZOMBIFIED_PIGLIN.create(p_29473_);
          if (zombifiedpiglin != null) {
             zombifiedpiglin.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
@@ -216,6 +_,7 @@
             }
 
             zombifiedpiglin.setPersistenceRequired();
+            net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zombifiedpiglin);
             p_29473_.addFreshEntity(zombifiedpiglin);
             this.discard();
          } else {
