--- a/net/minecraft/world/entity/animal/Wolf.java
+++ b/net/minecraft/world/entity/animal/Wolf.java
@@ -321,7 +_,7 @@
                   itemstack.m_41774_(1);
                }
 
-               this.m_5634_((float)item.m_41473_().m_38744_());
+               this.m_5634_((float)item.getFoodProperties(itemstack, this).m_38744_());
                this.m_146859_(GameEvent.f_157771_, this.m_146901_());
                return InteractionResult.SUCCESS;
             }
@@ -353,7 +_,7 @@
                itemstack.m_41774_(1);
             }
 
-            if (this.f_19796_.nextInt(3) == 0) {
+            if (this.f_19796_.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_30412_)) {
                this.m_21828_(p_30412_);
                this.f_21344_.m_26573_();
                this.m_6710_((LivingEntity)null);
@@ -393,7 +_,7 @@
 
    public boolean m_6898_(ItemStack p_30440_) {
       Item item = p_30440_.m_41720_();
-      return item.m_41472_() && item.m_41473_().m_38746_();
+      return item.m_41472_() && item.getFoodProperties(p_30440_, this).m_38746_();
    }
 
    public int m_5792_() {
