--- ../src-base/minecraft/net/minecraft/pathfinding/SwimNodeProcessor.java
+++ ../src-work/minecraft/net/minecraft/pathfinding/SwimNodeProcessor.java
@@ -67,7 +67,7 @@
                 {
                     IBlockState iblockstate = this.field_176169_a.func_180495_p(blockpos$mutableblockpos.func_181079_c(i, j, k));
 
-                    if (iblockstate.func_185904_a() != Material.field_151586_h)
+                    if (!iblockstate.func_185904_a().isSwimmable())
                     {
                         return PathNodeType.BLOCKED;
                     }
