--- ../src-base/minecraft/net/minecraft/network/login/server/SPacketLoginSuccess.java
+++ ../src-work/minecraft/net/minecraft/network/login/server/SPacketLoginSuccess.java
@@ -26,7 +26,7 @@
     {
         String s = p_148837_1_.func_150789_c(36);
         String s1 = p_148837_1_.func_150789_c(16);
-        UUID uuid = UUID.fromString(s);
+        UUID uuid = s.length() > 0 ? UUID.fromString(s) : null;
         this.field_149602_a = new GameProfile(uuid, s1);
     }
 
