--- ../src-base/minecraft/net/minecraft/client/renderer/ActiveRenderInfo.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/ActiveRenderInfo.java
@@ -30,6 +30,11 @@
 
     public static void func_74583_a(EntityPlayer p_74583_0_, boolean p_74583_1_)
     {
+        updateRenderInfo((Entity) p_74583_0_, p_74583_1_);
+    }
+
+    public static void updateRenderInfo(Entity p_74583_0_, boolean p_74583_1_)
+    {
         GlStateManager.func_179111_a(2982, field_178812_b);
         GlStateManager.func_179111_a(2983, field_178813_c);
         GlStateManager.func_187445_a(2978, field_178814_a);
@@ -81,7 +86,7 @@
             }
         }
 
-        return iblockstate;
+        return iblockstate.func_177230_c().getStateAtViewpoint(iblockstate, p_186703_0_, blockpos, vec3d);
     }
 
     public static float func_178808_b()
@@ -108,4 +113,14 @@
     {
         return field_74596_h;
     }
+
+    /* ======================================== FORGE START =====================================*/
+
+    /**
+     * Vector from render view entity position (corrected for partialTickTime) to the middle of screen
+     */
+    public static Vec3d getCameraPosition()
+    {
+        return field_178811_e;
+    }
 }
