--- ../src-base/minecraft/net/minecraft/client/renderer/RenderItem.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/RenderItem.java
@@ -71,7 +71,7 @@
     public RenderItem(TextureManager p_i46552_1_, ModelManager p_i46552_2_, ItemColors p_i46552_3_)
     {
         this.field_175057_n = p_i46552_1_;
-        this.field_175059_m = new ItemModelMesher(p_i46552_2_);
+        this.field_175059_m = new net.minecraftforge.client.ItemModelMesherForge(p_i46552_2_);
         this.func_175041_b();
         this.field_184395_f = p_i46552_3_;
     }
@@ -137,7 +137,7 @@
             {
                 GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
                 GlStateManager.func_179091_B();
-                TileEntityItemStackRenderer.field_147719_a.func_179022_a(p_180454_1_);
+                p_180454_1_.func_77973_b().getTileEntityItemStackRenderer().func_179022_a(p_180454_1_);
             }
             else
             {
@@ -218,7 +218,7 @@
                 k = k | -16777216;
             }
 
-            this.func_191969_a(p_191970_1_, bakedquad, k);
+            net.minecraftforge.client.model.pipeline.LightUtil.renderQuadColor(p_191970_1_, bakedquad, k);
         }
     }
 
@@ -240,17 +240,7 @@
     public IBakedModel func_184393_a(ItemStack p_184393_1_, @Nullable World p_184393_2_, @Nullable EntityLivingBase p_184393_3_)
     {
         IBakedModel ibakedmodel = this.field_175059_m.func_178089_a(p_184393_1_);
-        Item item = p_184393_1_.func_77973_b();
-
-        if (item != null && item.func_185040_i())
-        {
-            ResourceLocation resourcelocation = ibakedmodel.func_188617_f().func_188021_a(p_184393_1_, p_184393_2_, p_184393_3_);
-            return resourcelocation == null ? ibakedmodel : this.field_175059_m.func_178083_a().func_174953_a(new ModelResourceLocation(resourcelocation, "inventory"));
-        }
-        else
-        {
-            return ibakedmodel;
-        }
+        return ibakedmodel.func_188617_f().handleItemState(ibakedmodel, p_184393_1_, p_184393_2_, p_184393_3_);
     }
 
     public void func_184392_a(ItemStack p_184392_1_, EntityLivingBase p_184392_2_, ItemCameraTransforms.TransformType p_184392_3_, boolean p_184392_4_)
@@ -274,14 +264,9 @@
             GlStateManager.func_179147_l();
             GlStateManager.func_187428_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
             GlStateManager.func_179094_E();
-            ItemCameraTransforms itemcameratransforms = p_184394_2_.func_177552_f();
-            ItemCameraTransforms.func_188034_a(itemcameratransforms.func_181688_b(p_184394_3_), p_184394_4_);
+            // TODO: check if negative scale is a thing
+            p_184394_2_ = net.minecraftforge.client.ForgeHooksClient.handleCameraTransforms(p_184394_2_, p_184394_3_, p_184394_4_);
 
-            if (this.func_183005_a(itemcameratransforms.func_181688_b(p_184394_3_)))
-            {
-                GlStateManager.func_187407_a(GlStateManager.CullFace.FRONT);
-            }
-
             this.func_180454_a(p_184394_1_, p_184394_2_);
             GlStateManager.func_187407_a(GlStateManager.CullFace.BACK);
             GlStateManager.func_179121_F();
@@ -314,7 +299,7 @@
         GlStateManager.func_187401_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
         GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
         this.func_180452_a(p_191962_2_, p_191962_3_, p_191962_4_.func_177556_c());
-        p_191962_4_.func_177552_f().func_181689_a(ItemCameraTransforms.TransformType.GUI);
+        p_191962_4_ = net.minecraftforge.client.ForgeHooksClient.handleCameraTransforms(p_191962_4_, ItemCameraTransforms.TransformType.GUI, false);
         this.func_180454_a(p_191962_1_, p_191962_4_);
         GlStateManager.func_179118_c();
         GlStateManager.func_179101_C();
@@ -367,6 +352,7 @@
                         return String.valueOf((Object)p_184391_2_.func_77973_b());
                     }
                 });
+                crashreportcategory.func_189529_a("Registry Name", () -> String.valueOf(p_184391_2_.func_77973_b().getRegistryName()));
                 crashreportcategory.func_189529_a("Item Aux", new ICrashReportDetail<String>()
                 {
                     public String call() throws Exception
@@ -413,9 +399,12 @@
                 p_180453_1_.func_175063_a(s, (float)(p_180453_3_ + 19 - 2 - p_180453_1_.func_78256_a(s)), (float)(p_180453_4_ + 6 + 3), 16777215);
                 GlStateManager.func_179145_e();
                 GlStateManager.func_179126_j();
+                // Fixes opaque cooldown overlay a bit lower
+                // TODO: check if enabled blending still screws things up down the line.
+                GlStateManager.func_179147_l();
             }
 
-            if (p_180453_2_.func_77951_h())
+            if (p_180453_2_.func_77973_b().showDurabilityBar(p_180453_2_))
             {
                 GlStateManager.func_179140_f();
                 GlStateManager.func_179097_i();
@@ -424,11 +413,10 @@
                 GlStateManager.func_179084_k();
                 Tessellator tessellator = Tessellator.func_178181_a();
                 BufferBuilder bufferbuilder = tessellator.func_178180_c();
-                float f = (float)p_180453_2_.func_77952_i();
-                float f1 = (float)p_180453_2_.func_77958_k();
-                float f2 = Math.max(0.0F, (f1 - f) / f1);
-                int i = Math.round(13.0F - f * 13.0F / f1);
-                int j = MathHelper.func_181758_c(f2 / 3.0F, 1.0F, 1.0F);
+                double health = p_180453_2_.func_77973_b().getDurabilityForDisplay(p_180453_2_);
+                int rgbfordisplay = p_180453_2_.func_77973_b().getRGBDurabilityForDisplay(p_180453_2_);
+                int i = Math.round(13.0F - (float)health * 13.0F);
+                int j = rgbfordisplay;
                 this.func_181565_a(bufferbuilder, p_180453_3_ + 2, p_180453_4_ + 13, 13, 2, 0, 0, 0, 255);
                 this.func_181565_a(bufferbuilder, p_180453_3_ + 2, p_180453_4_ + 13, i, 1, j >> 16 & 255, j >> 8 & 255, j & 255, 255);
                 GlStateManager.func_179147_l();
@@ -1099,6 +1087,7 @@
         this.func_175029_a(Blocks.field_185779_df, TileEntityStructure.Mode.LOAD.func_185110_a(), "structure_block");
         this.func_175029_a(Blocks.field_185779_df, TileEntityStructure.Mode.CORNER.func_185110_a(), "structure_block");
         this.func_175029_a(Blocks.field_185779_df, TileEntityStructure.Mode.DATA.func_185110_a(), "structure_block");
+        net.minecraftforge.client.model.ModelLoader.onRegisterItems(this.field_175059_m);
     }
 
     public void func_110549_a(IResourceManager p_110549_1_)
