--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -468,7 +_,7 @@
             for (Player player : players) {
                 if (this.canAttack(player) && !this.isAlliedTo(player)) {
                     hasPotentialTarget = true;
-                    if ((!active || LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM.test(player))
+                    if ((!active || net.minecraftforge.common.ForgeHooks.isNotDisguised(this).test(player))
                         && this.isLookingAtMe(
                             player, 0.5, false, true, this.getEyeY(), this.getY() + 0.5 * this.getScale(), (this.getEyeY() + this.getY()) / 2.0
                         )) {
