--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -489,7 +_,7 @@
             for (Player player : list) {
                 if (this.canAttack(player) && !this.isAlliedTo(player)) {
                     flag1 = true;
-                    if ((!flag || LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM.test(player))
+                    if ((!flag || net.minecraftforge.common.ForgeHooks.isNotDisguised(this).test(player))
                         && this.isLookingAtMe(
                             player,
                             0.5,
