--- ../src-base/minecraft/net/minecraft/stats/StatList.java
+++ ../src-work/minecraft/net/minecraft/stats/StatList.java
@@ -118,7 +118,7 @@
 
             if (item != null)
             {
-                int i = Item.getIdFromItem(item);
+                int i = Item.getIdFromItem(item); //TODO: Hook FML's stat change event and re-assign these
                 objectCraftStats[i] = (new StatCrafting("stat.craftItem." + i, new ChatComponentTranslation("stat.craftItem", new Object[] {(new ItemStack(item)).func_151000_E()}), item)).registerStat();
             }
         }
@@ -136,7 +136,7 @@
 
             if (Item.getItemFromBlock(block) != null)
             {
-                int i = Block.getIdFromBlock(block);
+                int i = Block.getIdFromBlock(block); //TODO: Hook FML's stat change event and re-assign these
 
                 if (block.getEnableStats())
                 {
@@ -159,7 +159,7 @@
 
             if (item != null)
             {
-                int i = Item.getIdFromItem(item);
+                int i = Item.getIdFromItem(item); //TODO: Hook FML's stat change event and re-assign these
                 objectUseStats[i] = (new StatCrafting("stat.useItem." + i, new ChatComponentTranslation("stat.useItem", new Object[] {(new ItemStack(item)).func_151000_E()}), item)).registerStat();
 
                 if (!(item instanceof ItemBlock))
@@ -182,7 +182,7 @@
 
             if (item != null)
             {
-                int i = Item.getIdFromItem(item);
+                int i = Item.getIdFromItem(item); //TODO: Hook FML's stat change event and re-assign these
 
                 if (item.isDamageable())
                 {
