--- ../src-base/minecraft/net/minecraft/network/play/server/S3FPacketCustomPayload.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/S3FPacketCustomPayload.java
@@ -46,7 +46,11 @@
     public void writePacketData(PacketBuffer buf) throws IOException
     {
         buf.writeString(this.channel);
+        synchronized(this.data) { //This may be access multiple times, from multiple threads, lets be safe.
+        this.data.markReaderIndex();
         buf.writeBytes((ByteBuf)this.data);
+        this.data.resetReaderIndex();
+        }
     }
 
     public void processPacket(INetHandlerPlayClient handler)
