--- a/net/minecraft/server/integrated/IntegratedServer.java
+++ b/net/minecraft/server/integrated/IntegratedServer.java
@@ -56,9 +_,10 @@
       this.func_71188_g(true);
       this.func_71245_h(true);
       this.func_244801_P();
+      if (!net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
       this.func_240800_l__();
       this.func_71205_p(this.func_71214_G() + " - " + this.func_240793_aU_().func_76065_j());
-      return true;
+      return net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStarting(this);
    }
 
    public void func_71217_p(BooleanSupplier p_71217_1_) {
@@ -172,6 +_,7 @@
    }
 
    public void func_71263_m(boolean p_71263_1_) {
+      if (func_71278_l())
       this.func_213167_f(() -> {
          for(ServerPlayerEntity serverplayerentity : Lists.newArrayList(this.func_184103_al().func_181057_v())) {
             if (!serverplayerentity.func_110124_au().equals(this.field_211528_n)) {
