--- a/net/minecraft/server/integrated/IntegratedServer.java
+++ b/net/minecraft/server/integrated/IntegratedServer.java
@@ -171,6 +171,7 @@
    }
 
    public void func_71263_m(boolean p_71263_1_) {
+      if (func_71278_l())
       this.func_213167_f(() -> {
          for(ServerPlayerEntity serverplayerentity : Lists.newArrayList(this.func_184103_al().func_181057_v())) {
             if (!serverplayerentity.func_110124_au().equals(this.field_211528_n)) {
