--- ../src-base/minecraft/net/minecraft/server/integrated/IntegratedServer.java
+++ ../src-work/minecraft/net/minecraft/server/integrated/IntegratedServer.java
@@ -23,6 +23,9 @@
 import net.minecraft.world.WorldType;
 import net.minecraft.world.demo.DemoWorldServer;
 import net.minecraft.world.storage.ISaveHandler;
+import net.minecraftforge.common.DimensionManager;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.world.WorldEvent;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -54,44 +57,23 @@
     protected void func_71247_a(String p_71247_1_, String p_71247_2_, long p_71247_3_, WorldType p_71247_5_, String p_71247_6_)
     {
         this.func_71237_c(p_71247_1_);
-        this.field_71305_c = new WorldServer[3];
-        this.field_71312_k = new long[this.field_71305_c.length][100];
         ISaveHandler isavehandler = this.func_71254_M().func_75804_a(p_71247_1_, true);
 
-        for (int j = 0; j < this.field_71305_c.length; ++j)
+        WorldServer overWorld = (func_71242_L() ? new DemoWorldServer(this, isavehandler, p_71247_2_, 0, field_71304_b) : new WorldServer(this, isavehandler, p_71247_2_, 0, field_71350_m, field_71304_b));
+        for (int dim : DimensionManager.getStaticDimensionIDs())
         {
-            byte b0 = 0;
+            WorldServer world = (dim == 0 ? overWorld : new WorldServerMulti(this, isavehandler, p_71247_2_, dim, field_71350_m, overWorld, field_71304_b));
+            world.func_72954_a(new WorldManager(this, world));
 
-            if (j == 1)
+            if (!this.func_71264_H())
             {
-                b0 = -1;
+                world.func_72912_H().func_76060_a(func_71265_f());
             }
 
-            if (j == 2)
-            {
-                b0 = 1;
-            }
-
-            if (j == 0)
-            {
-                if (this.func_71242_L())
-                {
-                    this.field_71305_c[j] = new DemoWorldServer(this, isavehandler, p_71247_2_, b0, this.field_71304_b);
-                }
-                else
-                {
-                    this.field_71305_c[j] = new WorldServer(this, isavehandler, p_71247_2_, b0, this.field_71350_m, this.field_71304_b);
-                }
-            }
-            else
-            {
-                this.field_71305_c[j] = new WorldServerMulti(this, isavehandler, p_71247_2_, b0, this.field_71350_m, this.field_71305_c[0], this.field_71304_b);
-            }
-
-            this.field_71305_c[j].func_72954_a(new WorldManager(this, this.field_71305_c[j]));
-            this.func_71203_ab().func_72364_a(this.field_71305_c);
+            MinecraftForge.EVENT_BUS.post(new WorldEvent.Load(world));
         }
 
+        this.func_71203_ab().func_72364_a(new WorldServer[]{ overWorld });
         this.func_147139_a(this.func_147135_j());
         this.func_71222_d();
     }
