--- a/net/minecraft/entity/boss/dragon/phase/FlamingSittingPhase.java
+++ b/net/minecraft/entity/boss/dragon/phase/FlamingSittingPhase.java
@@ -58,7 +58,7 @@
          double d2 = this.field_188661_a.field_70986_h.field_70163_u + (double)(this.field_188661_a.field_70986_h.func_213302_cg() / 2.0F);
          BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos(d0, d2, d1);
 
-         while(this.field_188661_a.field_70170_p.func_175623_d(blockpos$mutableblockpos)) {
+         while(this.field_188661_a.field_70170_p.func_175623_d(blockpos$mutableblockpos) && d2 >= 0) { //Forge: Fix infinite loop if ground is missing.
             --d2;
             blockpos$mutableblockpos.func_189532_c(d0, d2, d1);
          }
