--- ../src-base/minecraft/net/minecraft/server/management/PlayerList.java
+++ ../src-work/minecraft/net/minecraft/server/management/PlayerList.java
@@ -101,7 +101,7 @@
         this.maxPlayers = 8;
     }
 
-    public void initializeConnectionToPlayer(NetworkManager netManager, EntityPlayerMP playerIn)
+    public void initializeConnectionToPlayer(NetworkManager netManager, EntityPlayerMP playerIn, NetHandlerPlayServer nethandlerplayserver)
     {
         GameProfile gameprofile = playerIn.getGameProfile();
         PlayerProfileCache playerprofilecache = this.mcServer.getPlayerProfileCache();
@@ -110,6 +110,17 @@
         playerprofilecache.addEntry(gameprofile);
         NBTTagCompound nbttagcompound = this.readPlayerDataFromFile(playerIn);
         playerIn.setWorld(this.mcServer.worldServerForDimension(playerIn.dimension));
+
+        World playerWorld = this.mcServer.worldServerForDimension(playerIn.dimension);
+        if (playerWorld == null)
+        {
+            playerIn.dimension = 0;
+            playerWorld = this.mcServer.worldServerForDimension(0);
+            BlockPos spawnPoint = playerWorld.provider.getRandomizedSpawnPoint();
+            playerIn.setPosition(spawnPoint.getX(), spawnPoint.getY(), spawnPoint.getZ());
+        }
+
+        playerIn.setWorld(playerWorld);
         playerIn.theItemInWorldManager.setWorld((WorldServer)playerIn.worldObj);
         String s1 = "local";
 
@@ -123,7 +134,7 @@
         WorldInfo worldinfo = worldserver.getWorldInfo();
         BlockPos blockpos = worldserver.getSpawnPoint();
         this.setPlayerGameTypeBasedOnOther(playerIn, (EntityPlayerMP)null, worldserver);
-        NetHandlerPlayServer nethandlerplayserver = new NetHandlerPlayServer(this.mcServer, netManager, playerIn);
+        playerIn.playerNetServerHandler = nethandlerplayserver;
         nethandlerplayserver.sendPacket(new SPacketJoinGame(playerIn.getEntityId(), playerIn.theItemInWorldManager.getGameType(), worldinfo.isHardcoreModeEnabled(), worldserver.provider.func_186058_p().func_186068_a(), worldserver.getDifficulty(), this.getMaxPlayers(), worldinfo.getTerrainType(), worldserver.getGameRules().getBoolean("reducedDebugInfo")));
         nethandlerplayserver.sendPacket(new SPacketCustomPayload("MC|Brand", (new PacketBuffer(Unpooled.buffer())).writeString(this.getServerInstance().getServerModName())));
         nethandlerplayserver.sendPacket(new SPacketServerDifficulty(worldinfo.getDifficulty(), worldinfo.isDifficultyLocked()));
@@ -213,6 +224,7 @@
         }
 
         playerIn.addSelfToInternalCraftingInventory();
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerLoggedIn(playerIn);
     }
 
     protected void sendScoreboard(ServerScoreboard scoreboardIn, EntityPlayerMP playerIn)
@@ -302,6 +314,7 @@
             nbttagcompound1 = this.mcServer.func_184110_aI().func_188257_a(FixTypes.PLAYER, nbttagcompound);
             playerIn.readFromNBT(nbttagcompound1);
             logger.debug("loading single player");
+            net.minecraftforge.event.ForgeEventFactory.firePlayerLoadingEvent(playerIn, this.playerNBTManagerObj, playerIn.getUniqueID().toString());
         }
         else
         {
@@ -311,8 +324,24 @@
         return nbttagcompound1;
     }
 
+    public NBTTagCompound getPlayerNBT(EntityPlayerMP player)
+    {
+        // Hacky method to allow loading the NBT for a player prior to login
+        NBTTagCompound nbttagcompound = this.mcServer.worldServers[0].getWorldInfo().getPlayerNBTTagCompound();
+        if (player.getName().equals(this.mcServer.getServerOwner()) && nbttagcompound != null)
+        {
+            return nbttagcompound;
+        }
+        else
+        {
+            return ((net.minecraft.world.storage.SaveHandler)this.playerNBTManagerObj).getPlayerNBT(player);
+        }
+    }
+
     protected void writePlayerData(EntityPlayerMP playerIn)
     {
+        if (playerIn.playerNetServerHandler == null) return;
+
         this.playerNBTManagerObj.writePlayerData(playerIn);
         StatisticsFile statisticsfile = (StatisticsFile)this.playerStatFiles.get(playerIn.getUniqueID());
 
@@ -334,6 +363,7 @@
             playerIn.playerNetServerHandler.sendPacket(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] {(EntityPlayerMP)this.playerEntityList.get(i)}));
         }
 
+        net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.getCurrentPlayerCount());
         worldserver.spawnEntityInWorld(playerIn);
         this.preparePlayer(playerIn, (WorldServer)null);
     }
@@ -345,6 +375,7 @@
 
     public void playerLoggedOut(EntityPlayerMP playerIn)
     {
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerLoggedOut(playerIn);
         WorldServer worldserver = playerIn.getServerForPlayer();
         playerIn.triggerAchievement(StatList.leaveGameStat);
         this.writePlayerData(playerIn);
@@ -379,6 +410,7 @@
             this.uuidToPlayerMap.remove(uuid);
             this.playerStatFiles.remove(uuid);
         }
+        net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.getCurrentPlayerCount());
 
         this.sendPacketToAllPlayers(new SPacketPlayerListItem(SPacketPlayerListItem.Action.REMOVE_PLAYER, new EntityPlayerMP[] {playerIn}));
     }
@@ -462,13 +494,23 @@
 
     public EntityPlayerMP recreatePlayerEntity(EntityPlayerMP playerIn, int dimension, boolean conqueredEnd)
     {
+        World world = mcServer.worldServerForDimension(dimension);
+        if (world == null)
+        {
+            dimension = 0;
+        }
+        else if (!world.provider.canRespawnHere())
+        {
+            dimension = world.provider.getRespawnDimension(playerIn);
+        }
+
         playerIn.getServerForPlayer().getEntityTracker().removePlayerFromTrackers(playerIn);
         playerIn.getServerForPlayer().getEntityTracker().untrackEntity(playerIn);
         playerIn.getServerForPlayer().func_184164_w().removePlayer(playerIn);
         this.playerEntityList.remove(playerIn);
         this.mcServer.worldServerForDimension(playerIn.dimension).removePlayerEntityDangerously(playerIn);
-        BlockPos blockpos = playerIn.getBedLocation();
-        boolean flag = playerIn.isSpawnForced();
+        BlockPos blockpos = playerIn.getBedLocation(dimension);
+        boolean flag = playerIn.isSpawnForced(dimension);
         playerIn.dimension = dimension;
         PlayerInteractionManager playerinteractionmanager;
 
@@ -484,7 +526,8 @@
         EntityPlayerMP entityplayermp = new EntityPlayerMP(this.mcServer, this.mcServer.worldServerForDimension(playerIn.dimension), playerIn.getGameProfile(), playerinteractionmanager);
         entityplayermp.playerNetServerHandler = playerIn.playerNetServerHandler;
         entityplayermp.clonePlayer(playerIn, conqueredEnd);
-        entityplayermp.setEntityId(playerIn.getEntityId());
+        entityplayermp.dimension = dimension;
+       entityplayermp.setEntityId(playerIn.getEntityId());
         entityplayermp.func_174817_o(playerIn);
         entityplayermp.func_184819_a(playerIn.func_184591_cq());
 
@@ -531,6 +574,7 @@
         this.uuidToPlayerMap.put(entityplayermp.getUniqueID(), entityplayermp);
         entityplayermp.addSelfToInternalCraftingInventory();
         entityplayermp.setHealth(entityplayermp.getHealth());
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerRespawnEvent(entityplayermp);
         return entityplayermp;
     }
 
@@ -545,15 +589,20 @@
 
     public void func_187242_a(EntityPlayerMP p_187242_1_, int p_187242_2_)
     {
+        transferPlayerToDimension(p_187242_1_, p_187242_2_, mcServer.worldServerForDimension(p_187242_2_).getDefaultTeleporter());
+    }
+
+    public void transferPlayerToDimension(EntityPlayerMP p_187242_1_, int p_187242_2_, net.minecraft.world.Teleporter teleporter)
+    {
         int i = p_187242_1_.dimension;
         WorldServer worldserver = this.mcServer.worldServerForDimension(p_187242_1_.dimension);
         p_187242_1_.dimension = p_187242_2_;
         WorldServer worldserver1 = this.mcServer.worldServerForDimension(p_187242_1_.dimension);
-        p_187242_1_.playerNetServerHandler.sendPacket(new SPacketRespawn(p_187242_1_.dimension, p_187242_1_.worldObj.getDifficulty(), p_187242_1_.worldObj.getWorldInfo().getTerrainType(), p_187242_1_.theItemInWorldManager.getGameType()));
+        p_187242_1_.playerNetServerHandler.sendPacket(new SPacketRespawn(p_187242_1_.dimension, worldserver1.getDifficulty(), worldserver1.getWorldInfo().getTerrainType(), p_187242_1_.theItemInWorldManager.getGameType()));
         this.func_187243_f(p_187242_1_);
         worldserver.removePlayerEntityDangerously(p_187242_1_);
         p_187242_1_.isDead = false;
-        this.transferEntityToWorld(p_187242_1_, i, worldserver, worldserver1);
+        this.transferEntityToWorld(p_187242_1_, i, worldserver, worldserver1, teleporter);
         this.preparePlayer(p_187242_1_, worldserver);
         p_187242_1_.playerNetServerHandler.setPlayerLocation(p_187242_1_.posX, p_187242_1_.posY, p_187242_1_.posZ, p_187242_1_.rotationYaw, p_187242_1_.rotationPitch);
         p_187242_1_.theItemInWorldManager.setWorld(worldserver1);
@@ -565,17 +614,27 @@
         {
             p_187242_1_.playerNetServerHandler.sendPacket(new SPacketEntityEffect(p_187242_1_.getEntityId(), potioneffect));
         }
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerChangedDimensionEvent(p_187242_1_, i, p_187242_2_);
     }
 
     public void transferEntityToWorld(Entity entityIn, int p_82448_2_, WorldServer p_82448_3_, WorldServer p_82448_4_)
     {
-        double d0 = entityIn.posX;
-        double d1 = entityIn.posZ;
+        transferEntityToWorld(entityIn, p_82448_2_, p_82448_3_, p_82448_4_, p_82448_4_.getDefaultTeleporter());
+    }
+
+    @SuppressWarnings("unused")
+    public void transferEntityToWorld(Entity entityIn, int p_82448_2_, WorldServer p_82448_3_, WorldServer p_82448_4_, net.minecraft.world.Teleporter teleporter)
+    {
+        net.minecraft.world.WorldProvider pOld = p_82448_3_.provider;
+        net.minecraft.world.WorldProvider pNew = p_82448_4_.provider;
+        double moveFactor = pOld.getMovementFactor() / pNew.getMovementFactor();
+        double d0 = entityIn.posX * moveFactor;
+        double d1 = entityIn.posZ * moveFactor;
         double d2 = 8.0D;
         float f = entityIn.rotationYaw;
         p_82448_3_.theProfiler.startSection("moving");
 
-        if (entityIn.dimension == -1)
+        if (false && entityIn.dimension == -1)
         {
             d0 = MathHelper.clamp_double(d0 / d2, p_82448_4_.getWorldBorder().minX() + 16.0D, p_82448_4_.getWorldBorder().maxX() - 16.0D);
             d1 = MathHelper.clamp_double(d1 / d2, p_82448_4_.getWorldBorder().minZ() + 16.0D, p_82448_4_.getWorldBorder().maxZ() - 16.0D);
@@ -586,7 +645,7 @@
                 p_82448_3_.updateEntityWithOptionalForce(entityIn, false);
             }
         }
-        else if (entityIn.dimension == 0)
+        else if (false && entityIn.dimension == 0)
         {
             d0 = MathHelper.clamp_double(d0 * d2, p_82448_4_.getWorldBorder().minX() + 16.0D, p_82448_4_.getWorldBorder().maxX() - 16.0D);
             d1 = MathHelper.clamp_double(d1 * d2, p_82448_4_.getWorldBorder().minZ() + 16.0D, p_82448_4_.getWorldBorder().maxZ() - 16.0D);
@@ -597,7 +656,8 @@
                 p_82448_3_.updateEntityWithOptionalForce(entityIn, false);
             }
         }
-        else
+
+        if (entityIn.dimension == 1)
         {
             BlockPos blockpos;
 
@@ -632,7 +692,7 @@
             if (entityIn.isEntityAlive())
             {
                 entityIn.setLocationAndAngles(d0, entityIn.posY, d1, entityIn.rotationYaw, entityIn.rotationPitch);
-                p_82448_4_.getDefaultTeleporter().placeInPortal(entityIn, f);
+                teleporter.placeInPortal(entityIn, f);
                 p_82448_4_.spawnEntityInWorld(entityIn);
                 p_82448_4_.updateEntityWithOptionalForce(entityIn, false);
             }
