--- ../src-base/minecraft/net/minecraft/client/gui/GuiErrorScreen.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiErrorScreen.java
@@ -22,6 +22,8 @@
         this.field_73887_h.add(new GuiButton(0, this.field_73880_f / 2 - 100, 140, StatCollector.func_74838_a("gui.cancel")));
     }
 
+    public GuiErrorScreen(){}
+
     public void func_73863_a(int p_73863_1_, int p_73863_2_, float p_73863_3_)
     {
         this.func_73733_a(0, 0, this.field_73880_f, this.field_73881_g, -12574688, -11530224);
