--- a/net/minecraft/core/RegistryAccess.java
+++ b/net/minecraft/core/RegistryAccess.java
@@ -186,6 +_,7 @@
       ResourceKey<? extends Registry<E>> resourcekey = p_211083_.f_123101_();
       Registry<E> registry = BuiltinRegistries.f_206379_.m_175515_(resourcekey);
 
+      net.minecraftforge.registries.GameData.postVanillaRegisterEvent(registry);
       for(Entry<ResourceKey<E>, E> entry : registry.m_6579_()) {
          ResourceKey<E> resourcekey1 = entry.getKey();
          E e = entry.getValue();
