--- a/net/minecraft/core/RegistryAccess.java
+++ b/net/minecraft/core/RegistryAccess.java
@@ -185,6 +_,8 @@
    private static <E> void m_211081_(RegistryResourceAccess.InMemoryStorage p_211082_, RegistryAccess.RegistryData<E> p_211083_) {
       ResourceKey<? extends Registry<E>> resourcekey = p_211083_.f_123101_();
       Registry<E> registry = BuiltinRegistries.f_206379_.m_175515_(resourcekey);
+      if (!resourcekey.equals(Registry.f_122818_))
+         registry = ((Registry<Registry<E>>)BuiltinRegistries.f_123858_).m_6246_((ResourceKey<Registry<E>>)resourcekey);
 
       for(Entry<ResourceKey<E>, E> entry : registry.m_6579_()) {
          ResourceKey<E> resourcekey1 = entry.getKey();
