--- a/net/minecraft/core/RegistryAccess.java
+++ b/net/minecraft/core/RegistryAccess.java
@@ -37,7 +_,7 @@
 
 public abstract class RegistryAccess {
    static final Logger f_123047_ = LogManager.getLogger();
-   static final Map<ResourceKey<? extends Registry<?>>, RegistryAccess.RegistryData<?>> f_123048_ = Util.m_137537_(() -> {
+   static final Map<ResourceKey<? extends Registry<?>>, RegistryAccess.RegistryData<?>> f_123048_ = new java.util.HashMap<>(Util.m_137537_(() -> {
       Builder<ResourceKey<? extends Registry<?>>, RegistryAccess.RegistryData<?>> builder = ImmutableMap.builder();
       m_123057_(builder, Registry.f_122818_, DimensionType.f_63843_, DimensionType.f_63843_);
       m_123057_(builder, Registry.f_122885_, Biome.f_47429_, Biome.f_47430_);
@@ -50,7 +_,7 @@
       m_123053_(builder, Registry.f_122878_, NoiseGeneratorSettings.f_64430_);
       m_123053_(builder, Registry.f_194568_, NormalNoise.NoiseParameters.f_192851_);
       return builder.build();
-   });
+   }));
    private static final RegistryAccess.RegistryHolder f_123049_ = Util.m_137537_(() -> {
       RegistryAccess.RegistryHolder registryaccess$registryholder = new RegistryAccess.RegistryHolder();
       DimensionType.m_188315_(registryaccess$registryholder);
@@ -109,6 +_,8 @@
       ResourceKey<? extends Registry<E>> resourcekey = p_194617_.f_123101_();
       boolean flag = !resourcekey.equals(Registry.f_122878_) && !resourcekey.equals(Registry.f_122818_);
       Registry<E> registry = f_123049_.m_175515_(resourcekey);
+      if (BuiltinRegistries.f_123858_.m_7804_(resourcekey.m_135782_()))
+         registry = ((Registry<Registry<E>>)BuiltinRegistries.f_123858_).m_6246_((ResourceKey<Registry<E>>)resourcekey);
       WritableRegistry<E> writableregistry = p_194615_.m_175512_(resourcekey);
 
       for(Entry<ResourceKey<E>, E> entry : registry.m_6579_()) {
@@ -224,4 +_,6 @@
          });
       }
    }
+
+   public static void addExtensions() { net.minecraftforge.common.ForgeHooks.injectRegistryAccessExtensions(f_123048_, (Map<ResourceKey<? extends Registry<?>>, MappedRegistry<?>>) f_123049_.f_123113_); }
 }
