--- a/net/minecraft/enchantment/EnchantmentFrostWalker.java
+++ b/net/minecraft/enchantment/EnchantmentFrostWalker.java
@@ -44,7 +44,8 @@
                IBlockState iblockstate1 = p_185266_1_.func_180495_p(blockpos$mutableblockpos);
                if (iblockstate1.func_196958_f()) {
                   IBlockState iblockstate2 = p_185266_1_.func_180495_p(blockpos$mutableblockpos1);
-                  if (iblockstate2.func_185904_a() == Material.field_151586_h && iblockstate2.func_177229_b(BlockFlowingFluid.field_176367_b) == 0 && iblockstate.func_196955_c(p_185266_1_, blockpos$mutableblockpos1) && p_185266_1_.func_195584_a(iblockstate, blockpos$mutableblockpos1)) {
+                  boolean isFull = iblockstate2.func_177230_c() == Blocks.field_150355_j && iblockstate2.func_177229_b(BlockFlowingFluid.field_176367_b) == 0; //TODO: Forge, modded waters?
+                  if (iblockstate2.func_185904_a() == Material.field_151586_h && isFull && iblockstate.func_196955_c(p_185266_1_, blockpos$mutableblockpos1) && p_185266_1_.func_195584_a(iblockstate, blockpos$mutableblockpos1) && !net.minecraftforge.event.ForgeEventFactory.onBlockPlace(p_185266_0_, new net.minecraftforge.common.util.BlockSnapshot(p_185266_1_, blockpos$mutableblockpos1.func_185334_h(), iblockstate2), net.minecraft.util.EnumFacing.UP)) {
                      p_185266_1_.func_175656_a(blockpos$mutableblockpos1, iblockstate);
                      p_185266_1_.func_205220_G_().func_205360_a(blockpos$mutableblockpos1.func_185334_h(), Blocks.field_185778_de, MathHelper.func_76136_a(p_185266_0_.func_70681_au(), 60, 120));
                   }
