--- a/net/minecraft/enchantment/Enchantment.java
+++ b/net/minecraft/enchantment/Enchantment.java
@@ -18,7 +_,8 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public abstract class Enchantment {
+public abstract class Enchantment extends net.minecraftforge.registries.ForgeRegistryEntry<Enchantment> {
+   private final net.minecraftforge.common.util.ReverseTagWrapper<Enchantment> reverseTags = new net.minecraftforge.common.util.ReverseTagWrapper<>(this, () -> net.minecraft.tags.TagCollectionManager.getInstance().getCustomTypeCollection(net.minecraftforge.registries.ForgeRegistries.ENCHANTMENTS));
    private final EquipmentSlotType[] slots;
    private final Enchantment.Rarity rarity;
    public final EnchantmentType category;
@@ -37,6 +_,14 @@
       this.slots = p_i46731_3_;
    }
 
+   public java.util.Set<net.minecraft.util.ResourceLocation> getTags() {
+      return reverseTags.getTagNames();
+   }
+
+   public boolean isIn(net.minecraft.tags.ITag<Enchantment> tag) {
+      return tag.contains(this);
+   }
+
    public Map<EquipmentSlotType, ItemStack> getSlotItems(LivingEntity p_222181_1_) {
       Map<EquipmentSlotType, ItemStack> map = Maps.newEnumMap(EquipmentSlotType.class);
 
@@ -114,7 +_,7 @@
    }
 
    public boolean canEnchant(ItemStack p_92089_1_) {
-      return this.category.canEnchant(p_92089_1_.getItem());
+      return canApplyAtEnchantingTable(p_92089_1_);
    }
 
    public void doPostAttack(LivingEntity p_151368_1_, Entity p_151368_2_, int p_151368_3_) {
@@ -136,6 +_,24 @@
    }
 
    public boolean isDiscoverable() {
+      return true;
+   }
+
+   /**
+    * This applies specifically to applying at the enchanting table. The other method {@link #canApply(ItemStack)}
+    * applies for <i>all possible</i> enchantments.
+    * @param stack
+    * @return
+    */
+   public boolean canApplyAtEnchantingTable(ItemStack stack) {
+      return stack.canApplyAtEnchantingTable(this);
+   }
+
+   /**
+    * Is this enchantment allowed to be enchanted on books via Enchantment Table
+    * @return false to disable the vanilla feature
+    */
+   public boolean isAllowedOnBooks() {
       return true;
    }
 
