--- ../src_base/minecraft/net/minecraft/enchantment/EnchantmentHelper.java
+++ ../src_work/minecraft/net/minecraft/enchantment/EnchantmentHelper.java
@@ -310,7 +310,7 @@
     public static int calcItemStackEnchantability(Random par0Random, int par1, int par2, ItemStack par3ItemStack)
     {
         Item var4 = par3ItemStack.getItem();
-        int var5 = var4.getItemEnchantability();
+        int var5 = var4.getItemEnchantability(par3ItemStack.getItemDamage());
 
         if (var5 <= 0)
         {
@@ -370,7 +370,7 @@
     public static List buildEnchantmentList(Random par0Random, ItemStack par1ItemStack, int par2)
     {
         Item var3 = par1ItemStack.getItem();
-        int var4 = var3.getItemEnchantability();
+        int var4 = var3.getItemEnchantability(par1ItemStack.getItemDamage());
 
         if (var4 <= 0)
         {
@@ -463,7 +463,7 @@
         {
             Enchantment var8 = var5[var7];
 
-            if (var8 != null && (var8.type.canEnchantItem(var2) || var4))
+            if (var8 != null && (var8.func_92037_a(par1ItemStack) || var4))
             {
                 for (int var9 = var8.getMinLevel(); var9 <= var8.getMaxLevel(); ++var9)
                 {
