--- ../src_base/minecraft/net/minecraft/enchantment/EnchantmentHelper.java
+++ ../src_work/minecraft/net/minecraft/enchantment/EnchantmentHelper.java
@@ -104,7 +104,7 @@
 
             if (par1ItemStack.itemID == Item.enchantedBook.itemID)
             {
-                Item.enchantedBook.getEnchantedItemStack_do(par1ItemStack, new EnchantmentData(i, ((Integer)par0Map.get(Integer.valueOf(i))).intValue()));
+                Item.enchantedBook.addEnchantment(par1ItemStack, new EnchantmentData(i, ((Integer)par0Map.get(Integer.valueOf(i))).intValue()));
             }
         }
 
@@ -351,7 +351,7 @@
 
                 if (flag)
                 {
-                    Item.enchantedBook.getEnchantedItemStack_do(par1ItemStack, enchantmentdata);
+                    Item.enchantedBook.addEnchantment(par1ItemStack, enchantmentdata);
                 }
                 else
                 {
@@ -463,7 +463,10 @@
         {
             Enchantment enchantment = aenchantment[k];
 
-            if (enchantment != null && (enchantment.type.canEnchantItem(item) || flag))
+            if (enchantment == null) continue;
+
+            flag = (par1ItemStack.itemID == Item.book.itemID) && enchantment.isAllowedOnBooks();
+            if (enchantment.canApplyAtEnchantingTable(par1ItemStack) || flag)
             {
                 for (int l = enchantment.getMinLevel(); l <= enchantment.getMaxLevel(); ++l)
                 {
