--- a/net/minecraft/enchantment/EnumEnchantmentType.java
+++ b/net/minecraft/enchantment/EnumEnchantmentType.java
@@ -13,7 +13,7 @@
 import net.minecraft.item.ItemTool;
 import net.minecraft.item.ItemTrident;
 
-public enum EnumEnchantmentType {
+public enum EnumEnchantmentType implements net.minecraftforge.common.IExtensibleEnum {
    ALL {
       public boolean func_77557_a(Item p_77557_1_) {
          for(EnumEnchantmentType enumenchantmenttype : EnumEnchantmentType.values()) {
@@ -90,5 +90,16 @@
    private EnumEnchantmentType() {
    }
 
-   public abstract boolean func_77557_a(Item p_77557_1_);
+   private java.util.function.Predicate<Item> delegate;
+   private EnumEnchantmentType(java.util.function.Predicate<Item> delegate) {
+      this.delegate = delegate;
+   }
+
+   public static EnumEnchantmentType create(String name, java.util.function.Predicate<Item> delegate) {
+      throw new IllegalStateException("Enum not extended");
+   }
+
+   public boolean func_77557_a(Item p_77557_1_) {
+      return this.delegate == null ? false : this.delegate.test(p_77557_1_);
+   }
 }
