--- a/net/minecraft/world/entity/monster/piglin/StopHoldingItemIfNoLongerAdmiring.java
+++ b/net/minecraft/world/entity/monster/piglin/StopHoldingItemIfNoLongerAdmiring.java
@@ -10,7 +_,7 @@
       return BehaviorBuilder.create((p_259197_) -> {
          return p_259197_.group(p_259197_.absent(MemoryModuleType.ADMIRING_ITEM)).apply(p_259197_, (p_259512_) -> {
             return (p_296847_, p_296848_, p_296849_) -> {
-               if (!p_296848_.getOffhandItem().isEmpty() && !p_296848_.getOffhandItem().is(Items.SHIELD)) {
+               if (!p_296848_.getOffhandItem().isEmpty() && !p_296848_.getOffhandItem().canPerformAction(net.minecraftforge.common.ToolActions.SHIELD_BLOCK)) {
                   PiglinAi.stopHoldingOffHandItem(p_296848_, true);
                   return true;
                } else {
