--- a/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
@@ -37,7 +_,7 @@
     }
 
     private boolean isHoldingCrossbow() {
-        return this.mob.isHolding(Items.CROSSBOW);
+        return this.mob.isHolding(is -> is.getItem() instanceof CrossbowItem);
     }
 
     @Override
@@ -99,7 +_,7 @@
             this.mob.getLookControl().setLookAt(livingentity, 30.0F, 30.0F);
             if (this.crossbowState == RangedCrossbowAttackGoal.CrossbowState.UNCHARGED) {
                 if (!flag2) {
-                    this.mob.startUsingItem(ProjectileUtil.getWeaponHoldingHand(this.mob, Items.CROSSBOW));
+                    this.mob.startUsingItem(ProjectileUtil.getWeaponHoldingHand(this.mob, item -> item instanceof CrossbowItem));
                     this.crossbowState = RangedCrossbowAttackGoal.CrossbowState.CHARGING;
                     this.mob.setChargingCrossbow(true);
                 }
