--- ../src-base/minecraft/net/minecraft/world/BossInfo.java
+++ ../src-work/minecraft/net/minecraft/world/BossInfo.java
@@ -2,8 +2,6 @@
 
 import java.util.UUID;
 import net.minecraft.util.text.ITextComponent;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public abstract class BossInfo
 {
@@ -35,7 +33,6 @@
         return this.field_186749_a;
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_186739_a(ITextComponent p_186739_1_)
     {
         this.field_186749_a = p_186739_1_;
@@ -56,7 +53,6 @@
         return this.field_186751_c;
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_186745_a(BossInfo.Color p_186745_1_)
     {
         this.field_186751_c = p_186745_1_;
@@ -67,7 +63,6 @@
         return this.field_186752_d;
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_186746_a(BossInfo.Overlay p_186746_1_)
     {
         this.field_186752_d = p_186746_1_;
