--- ../src-base/minecraft/net/minecraft/world/ChunkCache.java
+++ ../src-work/minecraft/net/minecraft/world/ChunkCache.java
@@ -8,6 +8,7 @@
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.world.biome.BiomeGenBase;
 import net.minecraft.world.chunk.Chunk;
+import net.minecraftforge.common.util.ForgeDirection;
 
 public class ChunkCache implements IBlockAccess
 {
@@ -91,6 +92,8 @@
     {
         int l = (p_147438_1_ >> 4) - this.field_72818_a;
         int i1 = (p_147438_3_ >> 4) - this.field_72816_b;
+        if (l < 0 || l >= field_72817_c.length || i1 < 0 || i1 >= field_72817_c[l].length) return null;
+        if (field_72817_c[l][i1] == null) return null;
         return this.field_72817_c[l][i1].func_150806_e(p_147438_1_ & 15, p_147438_2_, p_147438_3_ & 15);
     }
 
@@ -122,6 +125,8 @@
         {
             int l = (p_72805_1_ >> 4) - this.field_72818_a;
             int i1 = (p_72805_3_ >> 4) - this.field_72816_b;
+            if (l < 0 || l >= field_72817_c.length || i1 < 0 || i1 >= field_72817_c[l].length) return 0;
+            if (field_72817_c[l][i1] == null) return 0;
             return this.field_72817_c[l][i1].func_76628_c(p_72805_1_ & 15, p_72805_2_, p_72805_3_ & 15);
         }
     }
@@ -137,10 +142,9 @@
         return this.field_72815_e.func_72807_a(p_72807_1_, p_72807_2_);
     }
 
-    @SideOnly(Side.CLIENT)
     public boolean func_147437_c(int p_147437_1_, int p_147437_2_, int p_147437_3_)
     {
-        return this.func_147439_a(p_147437_1_, p_147437_2_, p_147437_3_).func_149688_o() == Material.field_151579_a;
+        return this.func_147439_a(p_147437_1_, p_147437_2_, p_147437_3_).isAir(this, p_147437_1_, p_147437_2_, p_147437_3_);
     }
 
     @SideOnly(Side.CLIENT)
@@ -241,4 +245,15 @@
     {
         return 256;
     }
+
+    @Override
+    public boolean isSideSolid(int x, int y, int z, ForgeDirection side, boolean _default)
+    {
+        if (x < -30000000 || z < -30000000 || x >= 30000000 || z >= 30000000)
+        {
+            return _default;
+        }
+
+        return func_147439_a(x, y, z).isSideSolid(this, x, y, z, side);
+    }
 }
