--- ../src-base/minecraft/net/minecraft/world/DimensionType.java
+++ ../src-work/minecraft/net/minecraft/world/DimensionType.java
@@ -13,6 +13,7 @@
     private final String field_186075_e;
     private final String field_186076_f;
     private final Class <? extends WorldProvider > field_186077_g;
+    private boolean shouldLoadSpawn = false;
 
     private DimensionType(int p_i46672_3_, String p_i46672_4_, String p_i46672_5_, Class <? extends WorldProvider > p_i46672_6_)
     {
@@ -20,6 +21,7 @@
         this.field_186075_e = p_i46672_4_;
         this.field_186076_f = p_i46672_5_;
         this.field_186077_g = p_i46672_6_;
+        this.shouldLoadSpawn = p_i46672_3_ == 0;
     }
 
     public int func_186068_a()
@@ -74,4 +76,18 @@
 
         throw new IllegalArgumentException("Invalid dimension id " + p_186069_0_);
     }
+
+    public boolean shouldLoadSpawn(){ return this.shouldLoadSpawn; }
+    public DimensionType setLoadSpawn(boolean value) { this.shouldLoadSpawn = value; return this; }
+
+    private static Class<?>[] ENUM_ARGS = {int.class, String.class, String.class, Class.class};
+    static { net.minecraftforge.common.util.EnumHelper.testEnum(DimensionType.class, ENUM_ARGS); }
+    public static DimensionType register(String name, String suffix, int id, Class<? extends WorldProvider> provider, boolean keepLoaded)
+    {
+        String enum_name = name.replace(" ", "_").toLowerCase();
+        DimensionType ret = net.minecraftforge.common.util.EnumHelper.addEnum(DimensionType.class, enum_name, ENUM_ARGS,
+                id, name, suffix, provider);
+        return ret.setLoadSpawn(keepLoaded);
+    }
+    //TODO: Unregister? There is no way to really delete a enum value...
 }
