--- a/net/minecraft/world/Dimension.java
+++ b/net/minecraft/world/Dimension.java
@@ -36,6 +36,15 @@
       this.field_236058_g_ = p_i231900_2_;
    }
 
+   public static Codec<Dimension> basedOn(Dimension base, RegistryKey<Dimension> key, net.minecraft.util.registry.WorldSettingsImport<com.google.gson.JsonElement> parser) {
+      return RecordCodecBuilder.create(inst ->
+              inst.group(
+                      com.mojang.serialization.MapCodec.unit(()->base.field_236057_f_).forGetter(d -> d.field_236057_f_),
+                      base.field_236058_g_.dispatch(key, parser).forGetter(d -> d.field_236058_g_)
+              ).apply(inst, Dimension::new)
+      );
+   }
+
    public Supplier<DimensionType> func_236059_a_() {
       return this.field_236057_f_;
    }
