--- ../src_base/minecraft/net/minecraft/world/Explosion.java
+++ ../src_work/minecraft/net/minecraft/world/Explosion.java
@@ -18,6 +18,15 @@
 import net.minecraft.util.MathHelper;
 import net.minecraft.util.Vec3;
 
+// CraftBukkit start
+import org.bukkit.Bukkit;
+import org.bukkit.event.entity.EntityDamageByBlockEvent;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.entity.EntityExplodeEvent;
+import org.bukkit.Location;
+// CraftBukkit end
+
 public class Explosion
 {
     /** whether or not the explosion sets fire to blocks around it */
@@ -37,12 +46,13 @@
     /** A list of ChunkPositions of blocks affected by this explosion */
     public List affectedBlockPositions = new ArrayList();
     private Map field_77288_k = new HashMap();
+    public boolean wasCanceled = false; // CraftBukkit
 
     public Explosion(World par1World, Entity par2Entity, double par3, double par5, double par7, float par9)
     {
         this.worldObj = par1World;
         this.exploder = par2Entity;
-        this.explosionSize = par9;
+        this.explosionSize = (float) Math.max(par9, 0.0); // CraftBukkit - clamp bad values
         this.explosionX = par3;
         this.explosionY = par5;
         this.explosionZ = par7;
@@ -53,6 +63,13 @@
      */
     public void doExplosionA()
     {
+        // CraftBukkit start
+        if (this.explosionSize < 0.1F)
+        {
+            return;
+        }
+
+        // CraftBukkit end
         float f = this.explosionSize;
         HashSet hashset = new HashSet();
         int i;
@@ -92,11 +109,11 @@
                             if (k1 > 0)
                             {
                                 Block block = Block.blocksList[k1];
-                                float f3 = this.exploder != null ? this.exploder.func_82146_a(this, this.worldObj, l, i1, j1, block) : block.getExplosionResistance(this.exploder);
+                                float f3 = this.exploder != null ? this.exploder.func_82146_a(this, this.worldObj, l, i1, j1, block) : block.getExplosionResistance(this.exploder, worldObj, l, i1, j1, explosionX, explosionY, explosionZ);
                                 f1 -= (f3 + 0.3F) * f2;
                             }
 
-                            if (f1 > 0.0F && (this.exploder == null || this.exploder.func_96091_a(this, this.worldObj, l, i1, j1, k1, f1)))
+                            if (f1 > 0.0F && (this.exploder == null || this.exploder.func_96091_a(this, this.worldObj, l, i1, j1, k1, f1)) && i1 < 256 && i1 >= 0)   // CraftBukkit - don't wrap explosions
                             {
                                 hashset.add(new ChunkPosition(l, i1, j1));
                             }
@@ -140,16 +157,67 @@
                     d2 /= d8;
                     double d9 = (double)this.worldObj.getBlockDensity(vec3, entity.boundingBox);
                     double d10 = (1.0D - d7) * d9;
-                    entity.attackEntityFrom(DamageSource.setExplosionSource(this), (float)((int)((d10 * d10 + d10) / 2.0D * 8.0D * (double)this.explosionSize + 1.0D)));
-                    double d11 = EnchantmentProtection.func_92092_a(entity, d10);
-                    entity.motionX += d0 * d11;
-                    entity.motionY += d1 * d11;
-                    entity.motionZ += d2 * d11;
-
-                    if (entity instanceof EntityPlayer)
-                    {
-                        this.field_77288_k.put((EntityPlayer)entity, this.worldObj.getWorldVec3Pool().getVecFromPool(d0 * d10, d1 * d10, d2 * d10));
-                    }
+                    // CraftBukkit start - Explosion damage hook
+                    org.bukkit.entity.Entity damagee = (entity == null) ? null : entity.getBukkitEntity();
+                    float damageDone = (float)((int)((d10 * d10 + d10) / 2.0D * 8.0D * (double) this.explosionSize + 1.0D));
+
+                    if (damagee == null)
+                    {
+                        // nothing was hurt
+                    }
+                    else if (this.exploder == null)     // Block explosion (without an entity source; bed etc.)
+                    {
+                        EntityDamageByBlockEvent event = new EntityDamageByBlockEvent(null, damagee, EntityDamageEvent.DamageCause.BLOCK_EXPLOSION, damageDone);
+                        Bukkit.getPluginManager().callEvent(event);
+
+                        if (!event.isCancelled())
+                        {
+                            damagee.setLastDamageCause(event);
+                            entity.attackEntityFrom(DamageSource.setExplosionSource(this), (float) event.getDamage());
+                            double d11 = EnchantmentProtection.func_92092_a(entity, d10);
+                            entity.motionX += d0 * d11;
+                            entity.motionY += d1 * d11;
+                            entity.motionZ += d2 * d11;
+
+                            if (entity instanceof EntityPlayer)
+                            {
+                                this.field_77288_k.put((EntityPlayer) entity, this.worldObj.getWorldVec3Pool().getVecFromPool(d0 * d10, d1 * d10, d2 * d10));
+                            }
+                        }
+                    }
+                    else
+                    {
+                        final org.bukkit.entity.Entity damager = this.exploder.getBukkitEntity();
+                        final EntityDamageEvent.DamageCause damageCause;
+
+                        if (damager instanceof org.bukkit.entity.TNTPrimed)
+                        {
+                            damageCause = EntityDamageEvent.DamageCause.BLOCK_EXPLOSION;
+                        }
+                        else
+                        {
+                            damageCause = EntityDamageEvent.DamageCause.ENTITY_EXPLOSION;
+                        }
+
+                        EntityDamageByEntityEvent event = new EntityDamageByEntityEvent(damager, damagee, damageCause, damageDone);
+                        Bukkit.getPluginManager().callEvent(event);
+
+                        if (!event.isCancelled())
+                        {
+                            entity.getBukkitEntity().setLastDamageCause(event);
+                            entity.attackEntityFrom(DamageSource.setExplosionSource(this), (float) event.getDamage());
+                            entity.motionX += d0 * d10;
+                            entity.motionY += d1 * d10;
+                            entity.motionZ += d2 * d10;
+
+                            if (entity instanceof EntityPlayer)
+                            {
+                                this.field_77288_k.put((EntityPlayer) entity, this.worldObj.getWorldVec3Pool().getVecFromPool(d0 * d10, d1 * d10, d2 * d10));
+                            }
+                        }
+                    }
+
+                    // CraftBukkit end
                 }
             }
         }
@@ -182,6 +250,40 @@
 
         if (this.isSmoking)
         {
+            // CraftBukkit start
+            org.bukkit.World bworld = this.worldObj.getWorld();
+            org.bukkit.entity.Entity explode = this.exploder == null ? null : this.exploder.getBukkitEntity();
+            Location location = new Location(bworld, this.explosionX, this.explosionY, this.explosionZ);
+            List<org.bukkit.block.Block> blockList = new ArrayList<org.bukkit.block.Block>();
+
+            for (int i1 = this.affectedBlockPositions.size() - 1; i1 >= 0; i1--)
+            {
+                ChunkPosition cpos = (ChunkPosition) this.affectedBlockPositions.get(i1);
+                org.bukkit.block.Block block = bworld.getBlockAt(cpos.x, cpos.y, cpos.z);
+
+                if (block.getType() != org.bukkit.Material.AIR)
+                {
+                    blockList.add(block);
+                }
+            }
+
+            EntityExplodeEvent event = new EntityExplodeEvent(explode, location, blockList, 0.3F);
+            this.worldObj.getServer().getPluginManager().callEvent(event);
+            this.affectedBlockPositions.clear();
+
+            for (org.bukkit.block.Block block : event.blockList())
+            {
+                ChunkPosition coords = new ChunkPosition(block.getX(), block.getY(), block.getZ());
+                affectedBlockPositions.add(coords);
+            }
+
+            if (event.isCancelled())
+            {
+                this.wasCanceled = true;
+                return;
+            }
+
+            // CraftBukkit end
             iterator = this.affectedBlockPositions.iterator();
 
             while (iterator.hasNext())
@@ -191,6 +293,7 @@
                 j = chunkposition.y;
                 k = chunkposition.z;
                 l = this.worldObj.getBlockId(i, j, k);
+                org.spigotmc.OrebfuscatorManager.updateNearbyBlocks(worldObj, i, j, k); // Spigot (Orebfuscator)
 
                 if (par1)
                 {
@@ -219,11 +322,11 @@
 
                     if (block.canDropFromExplosion(this))
                     {
-                        block.dropBlockAsItemWithChance(this.worldObj, i, j, k, this.worldObj.getBlockMetadata(i, j, k), 1.0F / this.explosionSize, 0);
-                    }
-
-                    this.worldObj.setBlock(i, j, k, 0, 0, 3);
-                    block.onBlockDestroyedByExplosion(this.worldObj, i, j, k, this);
+                        // CraftBukkit - add yield
+                        block.dropBlockAsItemWithChance(this.worldObj, i, j, k, this.worldObj.getBlockMetadata(i, j, k), event.getYield(), 0);
+                    }
+
+                    block.onBlockExploded(this.worldObj, i, j, k, this);
                 }
             }
         }
@@ -243,7 +346,13 @@
 
                 if (l == 0 && Block.opaqueCubeLookup[i1] && this.explosionRNG.nextInt(3) == 0)
                 {
-                    this.worldObj.setBlock(i, j, k, Block.fire.blockID);
+                    // CraftBukkit start - Ignition by explosion
+                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(this.worldObj, i, j, k, this).isCancelled())
+                    {
+                        this.worldObj.setBlock(i, j, k, Block.fire.blockID);
+                    }
+
+                    // CraftBukkit end
                 }
             }
         }
