--- a/net/minecraft/world/Explosion.java
+++ b/net/minecraft/world/Explosion.java
@@ -92,7 +92,7 @@
                      BlockPos blockpos = new BlockPos(d4, d6, d8);
                      IBlockState iblockstate = this.field_77287_j.func_180495_p(blockpos);
                      IFluidState ifluidstate = this.field_77287_j.func_204610_c(blockpos);
-                     float f2 = Math.max(iblockstate.func_177230_c().func_149638_a(), ifluidstate.func_210200_l());
+                     float f2 = Math.max(iblockstate.getExplosionResistance(field_77287_j, blockpos, field_77283_e, this), ifluidstate.getExplosionResistance(field_77287_j, blockpos, field_77283_e, this));
                      if (f2 > 0.0F) {
                         if (this.field_77283_e != null) {
                            f2 = this.field_77283_e.func_180428_a(this, this.field_77287_j, blockpos, iblockstate, ifluidstate, f2);
