--- ../src_base/minecraft/net/minecraft/world/Explosion.java
+++ ../src_work/minecraft/net/minecraft/world/Explosion.java
@@ -92,7 +92,7 @@
                             if (k1 > 0)
                             {
                                 Block block = Block.blocksList[k1];
-                                float f3 = this.exploder != null ? this.exploder.getBlockExplosionResistance(this, this.worldObj, l, i1, j1, block) : block.getExplosionResistance(this.exploder);
+                                float f3 = this.exploder != null ? this.exploder.getBlockExplosionResistance(this, this.worldObj, l, i1, j1, block) : block.getExplosionResistance(this.exploder, worldObj, l, i1, j1, explosionX, explosionY, explosionZ);
                                 f1 -= (f3 + 0.3F) * f2;
                             }
 
@@ -222,8 +222,7 @@
                         block.dropBlockAsItemWithChance(this.worldObj, i, j, k, this.worldObj.getBlockMetadata(i, j, k), 1.0F / this.explosionSize, 0);
                     }
 
-                    this.worldObj.setBlock(i, j, k, 0, 0, 3);
-                    block.onBlockDestroyedByExplosion(this.worldObj, i, j, k, this);
+                    block.onBlockExploded(this.worldObj, i, j, k, this);
                 }
             }
         }
