--- ../src-base/minecraft/net/minecraft/world/IBlockAccess.java
+++ ../src-work/minecraft/net/minecraft/world/IBlockAccess.java
@@ -29,4 +29,14 @@
 
     @SideOnly(Side.CLIENT)
     WorldType func_175624_G();
+
+    /**
+     * FORGE: isSideSolid, pulled up from {@link World}
+     *
+     * @param pos Position
+     * @param side Side
+     * @param _default default return value
+     * @return if the block is solid on the side
+     */
+    boolean isSideSolid(BlockPos pos, EnumFacing side, boolean _default);
 }
