--- ../src_base/minecraft/net/minecraft/world/IBlockAccess.java
+++ ../src_work/minecraft/net/minecraft/world/IBlockAccess.java
@@ -6,6 +6,7 @@
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.Vec3Pool;
 import net.minecraft.world.biome.BiomeGenBase;
+import net.minecraftforge.common.ForgeDirection;
 
 public interface IBlockAccess
 {
@@ -59,8 +60,6 @@
      */
     boolean isBlockNormalCube(int i, int j, int k);
 
-    @SideOnly(Side.CLIENT)
-
     /**
      * Returns true if the block at the specified coordinates is empty
      */
@@ -103,4 +102,16 @@
      * Is this block powering in the specified direction Args: x, y, z, direction
      */
     int isBlockProvidingPowerTo(int i, int j, int k, int l);
+
+    /**
+     * FORGE: isBlockSolidOnSide, pulled up from {@link World}
+     *
+     * @param x X coord
+     * @param y Y coord
+     * @param z Z coord
+     * @param side Side
+     * @param _default default return value
+     * @return if the block is solid on the side
+     */
+    boolean isBlockSolidOnSide(int x, int y, int z, ForgeDirection side, boolean _default);
 }
