--- a/net/minecraft/world/IWorldReader.java
+++ b/net/minecraft/world/IWorldReader.java
@@ -68,7 +_,7 @@
    }
 
    default boolean isEmptyBlock(BlockPos p_175623_1_) {
-      return this.getBlockState(p_175623_1_).isAir();
+      return this.getBlockState(p_175623_1_).isAir(this, p_175623_1_);
    }
 
    default boolean canSeeSkyFromBelowWater(BlockPos p_175710_1_) {
@@ -155,6 +_,10 @@
    @Deprecated
    default boolean hasChunkAt(BlockPos p_175667_1_) {
       return this.hasChunk(p_175667_1_.getX() >> 4, p_175667_1_.getZ() >> 4);
+   }
+
+   default boolean isAreaLoaded(BlockPos center, int range) {
+      return this.hasChunksAt(center.offset(-range, -range, -range), center.offset(range, range, range));
    }
 
    @Deprecated
