--- a/net/minecraft/world/Teleporter.java
+++ b/net/minecraft/world/Teleporter.java
@@ -16,7 +16,7 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.dimension.DimensionType;
 
-public class Teleporter {
+public class Teleporter implements net.minecraftforge.common.util.ITeleporter {
    private static final BlockPortal field_196236_a = (BlockPortal)Blocks.field_150427_aO;
    protected final WorldServer field_85192_a;
    protected final Random field_77187_a;
@@ -335,4 +335,13 @@
          this.field_85087_d = p_i45747_3_;
       }
    }
+   
+   @Override
+   public void placeEntity(World world, Entity entity, float yaw)
+   {
+      if (entity instanceof EntityPlayerMP)
+         func_180266_a(entity, yaw);
+      else
+         func_180620_b(entity, yaw);
+   }
 }
