--- a/net/minecraft/world/WorldServer.java
+++ b/net/minecraft/world/WorldServer.java
@@ -85,7 +85,7 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class WorldServer extends World implements IThreadListener {
+public class WorldServer extends World implements IThreadListener, net.minecraftforge.common.extensions.IForgeWorldServer {
    private static final Logger LOGGER = LogManager.getLogger();
    private final MinecraftServer server;
    private final EntityTracker entityTracker;
@@ -150,6 +150,7 @@
          this.getWorldBorder().setTransition(this.worldInfo.getBorderSize());
       }
 
+      this.initCapabilities();
       return this;
    }
 
@@ -957,4 +958,12 @@
    public NetworkTagManager func_205772_D() {
       return this.server.func_199731_aO();
    }
+
+   private long[] tickTime = new long[100];
+   public void setTickTime(int tick, long time) {
+      this.tickTime[tick % 100] = time;
+   }
+   public long[] getTickTimes() {
+      return this.tickTime;
+   }
 }
