--- ../src-base/minecraft/net/minecraft/client/gui/achievement/GuiStats.java
+++ ../src-work/minecraft/net/minecraft/client/gui/achievement/GuiStats.java
@@ -5,6 +5,8 @@
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
+
+import net.minecraft.block.Block;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.audio.PositionedSoundRecord;
 import net.minecraft.client.gui.GuiButton;
@@ -443,17 +445,18 @@
             for (StatCrafting statcrafting : StatList.objectMineStats)
             {
                 boolean flag = false;
-                int i = Item.getIdFromItem(statcrafting.func_150959_a());
+                int i = Block.getIdFromBlock(Block.getBlockFromItem(statcrafting.func_150959_a()));
+                int itemId = Item.getIdFromItem(statcrafting.func_150959_a());
 
                 if (GuiStats.this.field_146546_t.readStat(statcrafting) > 0)
                 {
                     flag = true;
                 }
-                else if (StatList.objectUseStats[i] != null && GuiStats.this.field_146546_t.readStat(StatList.objectUseStats[i]) > 0)
+                else if (StatList.objectUseStats[itemId] != null && GuiStats.this.field_146546_t.readStat(StatList.objectUseStats[itemId]) > 0)
                 {
                     flag = true;
                 }
-                else if (StatList.objectCraftStats[i] != null && GuiStats.this.field_146546_t.readStat(StatList.objectCraftStats[i]) > 0)
+                else if (StatList.objectCraftStats[itemId] != null && GuiStats.this.field_146546_t.readStat(StatList.objectCraftStats[itemId]) > 0)
                 {
                     flag = true;
                 }
@@ -468,8 +471,15 @@
             {
                 public int compare(StatCrafting p_compare_1_, StatCrafting p_compare_2_)
                 {
-                    int j = Item.getIdFromItem(p_compare_1_.func_150959_a());
-                    int k = Item.getIdFromItem(p_compare_2_.func_150959_a());
+                    int j;
+                    int k;
+                    if (StatsBlock.this.field_148217_o == 2) {
+                        j = Block.getIdFromBlock(Block.getBlockFromItem(p_compare_1_.func_150959_a()));
+                        k = Block.getIdFromBlock(Block.getBlockFromItem(p_compare_2_.func_150959_a()));
+                    } else {
+                        j = Item.getIdFromItem(p_compare_1_.func_150959_a());
+                        k = Item.getIdFromItem(p_compare_2_.func_150959_a());
+                    }
                     StatBase statbase = null;
                     StatBase statbase1 = null;
 
@@ -748,7 +758,7 @@
             super(mcIn, GuiStats.this.width, GuiStats.this.height, 32, GuiStats.this.height - 64, GuiStats.this.fontRendererObj.FONT_HEIGHT * 4);
             this.setShowSelectionBox(false);
 
-            for (EntityList.EntityEggInfo entitylist$entityegginfo : EntityList.entityEggs.values())
+            for (EntityList.EntityEggInfo entitylist$entityegginfo : com.google.common.collect.Iterables.concat(EntityList.entityEggs.values(), net.minecraftforge.fml.common.registry.EntityRegistry.getEggs().values()))
             {
                 if (GuiStats.this.field_146546_t.readStat(entitylist$entityegginfo.field_151512_d) > 0 || GuiStats.this.field_146546_t.readStat(entitylist$entityegginfo.field_151513_e) > 0)
                 {
@@ -784,7 +794,7 @@
         protected void drawSlot(int entryID, int p_180791_2_, int p_180791_3_, int p_180791_4_, int mouseXIn, int mouseYIn)
         {
             EntityList.EntityEggInfo entitylist$entityegginfo = (EntityList.EntityEggInfo)this.field_148222_l.get(entryID);
-            String s = I18n.format("entity." + EntityList.getStringFromID(entitylist$entityegginfo.spawnedID) + ".name", new Object[0]);
+            String s = I18n.format("entity." + entitylist$entityegginfo.name + ".name", new Object[0]);
             int i = GuiStats.this.field_146546_t.readStat(entitylist$entityegginfo.field_151512_d);
             int j = GuiStats.this.field_146546_t.readStat(entitylist$entityegginfo.field_151513_e);
             String s1 = I18n.format("stat.entityKills", new Object[] {Integer.valueOf(i), s});
