--- a/net/minecraft/block/AbstractFireBlock.java
+++ b/net/minecraft/block/AbstractFireBlock.java
@@ -124,6 +_,7 @@
       if (!p_220082_4_.is(p_220082_1_.getBlock())) {
          if (inPortalDimension(p_220082_2_)) {
             Optional<PortalSize> optional = PortalSize.findEmptyPortalShape(p_220082_2_, p_220082_3_, Direction.Axis.X);
+            optional =  net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_220082_2_, p_220082_3_, optional);
             if (optional.isPresent()) {
                optional.get().createPortalBlocks();
                return;
