--- a/net/minecraft/block/AbstractFireBlock.java
+++ b/net/minecraft/block/AbstractFireBlock.java
@@ -124,6 +_,7 @@
       if (!p_220082_4_.func_203425_a(p_220082_1_.func_177230_c())) {
          if (func_242649_a(p_220082_2_)) {
             Optional<PortalSize> optional = PortalSize.func_242964_a(p_220082_2_, p_220082_3_, Direction.Axis.X);
+            optional =  net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_220082_2_, p_220082_3_, optional);
             if (optional.isPresent()) {
                optional.get().func_150859_c();
                return;
