--- a/net/minecraft/block/BedBlock.java
+++ b/net/minecraft/block/BedBlock.java
@@ -88,7 +88,9 @@
             }
          }
 
-         if (p_225533_2_.field_73011_w.func_76567_e() && p_225533_2_.func_226691_t_(p_225533_3_) != Biomes.field_76778_j) {
+         net.minecraftforge.common.extensions.IForgeDimension.SleepResult sleepResult = p_225533_2_.field_73011_w.canSleepAt(p_225533_4_, p_225533_3_);
+         if (sleepResult != net.minecraftforge.common.extensions.IForgeDimension.SleepResult.BED_EXPLODES) {
+            if (sleepResult == net.minecraftforge.common.extensions.IForgeDimension.SleepResult.DENY) return ActionResultType.SUCCESS;
             if (p_225533_1_.func_177229_b(field_176471_b)) {
                if (!this.func_226861_a_(p_225533_2_, p_225533_3_)) {
                   p_225533_4_.func_146105_b(new TranslationTextComponent("block.minecraft.bed.occupied"), true);
