--- ../src-base/minecraft/net/minecraft/block/BlockGrass.java
+++ ../src-work/minecraft/net/minecraft/block/BlockGrass.java
@@ -59,7 +59,7 @@
     {
         if (!p_180650_1_.field_72995_K)
         {
-            if (p_180650_1_.func_175671_l(p_180650_2_.func_177984_a()) < 4 && p_180650_1_.func_180495_p(p_180650_2_.func_177984_a()).func_177230_c().func_149717_k() > 2)
+            if (p_180650_1_.func_175671_l(p_180650_2_.func_177984_a()) < 4 && p_180650_1_.func_180495_p(p_180650_2_.func_177984_a()).func_177230_c().getLightOpacity(p_180650_1_, p_180650_2_.func_177984_a()) > 2)
             {
                 p_180650_1_.func_175656_a(p_180650_2_, Blocks.field_150346_d.func_176223_P());
             }
@@ -73,7 +73,7 @@
                         Block block = p_180650_1_.func_180495_p(blockpos1.func_177984_a()).func_177230_c();
                         IBlockState iblockstate1 = p_180650_1_.func_180495_p(blockpos1);
 
-                        if (iblockstate1.func_177230_c() == Blocks.field_150346_d && iblockstate1.func_177229_b(BlockDirt.field_176386_a) == BlockDirt.DirtType.DIRT && p_180650_1_.func_175671_l(blockpos1.func_177984_a()) >= 4 && block.func_149717_k() <= 2)
+                        if (iblockstate1.func_177230_c() == Blocks.field_150346_d && iblockstate1.func_177229_b(BlockDirt.field_176386_a) == BlockDirt.DirtType.DIRT && p_180650_1_.func_175671_l(blockpos1.func_177984_a()) >= 4 && block.getLightOpacity(p_180650_1_, blockpos1.func_177984_a()) <= 2)
                         {
                             p_180650_1_.func_175656_a(blockpos1, Blocks.field_150349_c.func_176223_P());
                         }
@@ -120,18 +120,11 @@
                         continue;
                     }
                 }
-                else if (p_176474_1_.func_180495_p(blockpos2).func_177230_c().field_149764_J == Material.field_151579_a)
+                else if (p_176474_1_.func_175623_d(blockpos2))
                 {
                     if (p_176474_2_.nextInt(8) == 0)
                     {
-                        BlockFlower.EnumFlowerType enumflowertype = p_176474_1_.func_180494_b(blockpos2).func_180623_a(p_176474_2_, blockpos2);
-                        BlockFlower blockflower = enumflowertype.func_176964_a().func_180346_a();
-                        IBlockState iblockstate1 = blockflower.func_176223_P().func_177226_a(blockflower.func_176494_l(), enumflowertype);
-
-                        if (blockflower.func_180671_f(p_176474_1_, blockpos2, iblockstate1))
-                        {
-                            p_176474_1_.func_180501_a(blockpos2, iblockstate1, 3);
-                        }
+                        p_176474_1_.func_180494_b(blockpos2).plantFlower(p_176474_1_, p_176474_2_, blockpos2);
                     }
                     else
                     {
