--- a/net/minecraft/block/Blocks.java
+++ b/net/minecraft/block/Blocks.java
@@ -25,6 +25,7 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.gen.feature.Features;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Blocks {
    public static final Block field_150350_a = func_222382_a("air", new AirBlock(AbstractBlock.Properties.func_200945_a(Material.field_151579_a).func_200942_a().func_222380_e().func_235859_g_()));
    public static final Block field_150348_b = func_222382_a("stone", new Block(AbstractBlock.Properties.func_200949_a(Material.field_151576_e, MaterialColor.field_151665_m).func_235861_h_().func_200948_a(1.5F, 6.0F)));
@@ -119,7 +120,7 @@
    public static final Block field_196601_az = func_222382_a("green_bed", func_235422_a_(DyeColor.GREEN));
    public static final Block field_196550_aA = func_222382_a("red_bed", func_235422_a_(DyeColor.RED));
    public static final Block field_196551_aB = func_222382_a("black_bed", func_235422_a_(DyeColor.BLACK));
-   public static final Block field_196552_aC = func_222382_a("powered_rail", new PoweredRailBlock(AbstractBlock.Properties.func_200945_a(Material.field_151594_q).func_200942_a().func_200943_b(0.7F).func_200947_a(SoundType.field_185852_e)));
+   public static final Block field_196552_aC = func_222382_a("powered_rail", new PoweredRailBlock(AbstractBlock.Properties.func_200945_a(Material.field_151594_q).func_200942_a().func_200943_b(0.7F).func_200947_a(SoundType.field_185852_e), true));
    public static final Block field_150319_E = func_222382_a("detector_rail", new DetectorRailBlock(AbstractBlock.Properties.func_200945_a(Material.field_151594_q).func_200942_a().func_200943_b(0.7F).func_200947_a(SoundType.field_185852_e)));
    public static final Block field_150320_F = func_222382_a("sticky_piston", func_235432_a_(true));
    public static final Block field_196553_aF = func_222382_a("cobweb", new WebBlock(AbstractBlock.Properties.func_200945_a(Material.field_151569_G).func_200942_a().func_235861_h_().func_200943_b(4.0F)));
@@ -937,15 +938,4 @@
    public static void func_235419_a_() {
       Block.field_176229_d.forEach(AbstractBlock.AbstractBlockState::func_215692_c);
    }
-
-   static {
-      for(Block block : Registry.field_212618_g) {
-         for(BlockState blockstate : block.func_176194_O().func_177619_a()) {
-            Block.field_176229_d.func_195867_b(blockstate);
-         }
-
-         block.func_220068_i();
-      }
-
-   }
 }
