--- ../src-base/minecraft/net/minecraft/block/BlockBush.java
+++ ../src-work/minecraft/net/minecraft/block/BlockBush.java
@@ -6,8 +6,13 @@
 import net.minecraft.init.Blocks;
 import net.minecraft.util.AxisAlignedBB;
 import net.minecraft.world.World;
+import net.minecraft.world.IBlockAccess;
+import net.minecraftforge.common.EnumPlantType;
+import net.minecraftforge.common.util.ForgeDirection;
+import net.minecraftforge.common.IPlantable;
+import static net.minecraftforge.common.EnumPlantType.*;
 
-public class BlockBush extends Block
+public class BlockBush extends Block implements IPlantable
 {
     private static final String __OBFID = "CL_00000208";
 
@@ -27,7 +32,7 @@
 
     public boolean func_149742_c(World p_149742_1_, int p_149742_2_, int p_149742_3_, int p_149742_4_)
     {
-        return super.func_149742_c(p_149742_1_, p_149742_2_, p_149742_3_, p_149742_4_) && this.func_149854_a(p_149742_1_.func_147439_a(p_149742_2_, p_149742_3_ - 1, p_149742_4_));
+        return super.func_149742_c(p_149742_1_, p_149742_2_, p_149742_3_, p_149742_4_) && this.func_149718_j(p_149742_1_, p_149742_2_, p_149742_3_, p_149742_4_);
     }
 
     protected boolean func_149854_a(Block p_149854_1_)
@@ -57,7 +62,7 @@
 
     public boolean func_149718_j(World p_149718_1_, int p_149718_2_, int p_149718_3_, int p_149718_4_)
     {
-        return this.func_149854_a(p_149718_1_.func_147439_a(p_149718_2_, p_149718_3_ - 1, p_149718_4_));
+        return  p_149718_1_.func_147439_a(p_149718_2_, p_149718_3_ - 1, p_149718_4_).canSustainPlant(p_149718_1_, p_149718_2_, p_149718_3_ - 1, p_149718_4_, ForgeDirection.UP, this);
     }
 
     public AxisAlignedBB func_149668_a(World p_149668_1_, int p_149668_2_, int p_149668_3_, int p_149668_4_)
@@ -79,4 +84,37 @@
     {
         return 1;
     }
+
+    @Override
+    public EnumPlantType getPlantType(IBlockAccess world, int x, int y, int z)
+    {
+        if (this == Blocks.field_150464_aj)          return Crop;
+        if (this == Blocks.field_150459_bM)        return Crop;
+        if (this == Blocks.field_150469_bN)       return Crop;
+        if (this == Blocks.field_150394_bc)     return Crop;
+        if (this == Blocks.field_150393_bb)   return Crop;
+        if (this == Blocks.field_150330_I)       return Desert;
+        if (this == Blocks.field_150392_bi)      return Water;
+        if (this == Blocks.field_150337_Q)   return Cave;
+        if (this == Blocks.field_150338_P) return Cave;
+        if (this == Blocks.field_150388_bm)    return Nether;
+        if (this == Blocks.field_150345_g)        return Plains;
+        if (this == Blocks.field_150329_H)      return Plains;
+        if (this == Blocks.field_150398_cm)   return Plains;
+        if (this == Blocks.field_150328_O)     return Plains;
+        if (this == Blocks.field_150327_N)  return Plains;        
+        return Plains;
+    }
+
+    @Override
+    public Block getPlant(IBlockAccess world, int x, int y, int z)
+    {
+        return this;
+    }
+
+    @Override
+    public int getPlantMetadata(IBlockAccess world, int x, int y, int z)
+    {
+        return world.func_72805_g(x, y, z);
+    }
 }
