--- a/net/minecraft/block/BlockTorch.java
+++ b/net/minecraft/block/BlockTorch.java
@@ -37,9 +37,7 @@
 
    public boolean func_196260_a(IBlockState p_196260_1_, IWorldReaderBase p_196260_2_, BlockPos p_196260_3_) {
       IBlockState iblockstate = p_196260_2_.func_180495_p(p_196260_3_.func_177977_b());
-      Block block = iblockstate.func_177230_c();
-      boolean flag = block instanceof BlockFence || block instanceof BlockStainedGlass || block == Blocks.field_150359_w || block == Blocks.field_150463_bK || block == Blocks.field_196723_eg || iblockstate.func_185896_q();
-      return flag && block != Blocks.field_185775_db;
+      return iblockstate.canPlaceTorchOnTop(p_196260_2_, p_196260_3_);
    }
 
    @OnlyIn(Dist.CLIENT)
