--- ../src_base/minecraft/net/minecraft/block/BlockNetherStalk.java
+++ ../src_work/minecraft/net/minecraft/block/BlockNetherStalk.java
@@ -2,11 +2,14 @@
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
+
+import java.util.ArrayList;
 import java.util.Random;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+import net.minecraftforge.common.ForgeDirection;
 
 public class BlockNetherStalk extends BlockFlower
 {
@@ -33,7 +36,8 @@
      */
     public boolean canBlockStay(World par1World, int par2, int par3, int par4)
     {
-        return this.canThisPlantGrowOnThisBlockID(par1World.getBlockId(par2, par3 - 1, par4));
+        Block block = Block.blocksList[par1World.getBlockId(par2, par3 - 1, par4)];
+        return (block != null && block.canSustainPlant(par1World, par2, par3 - 1, par4, ForgeDirection.UP, this));
     }
 
     /**
@@ -73,25 +77,7 @@
      */
     public void dropBlockAsItemWithChance(World par1World, int par2, int par3, int par4, int par5, float par6, int par7)
     {
-        if (!par1World.isRemote)
-        {
-            int var8 = 1;
-
-            if (par5 >= 3)
-            {
-                var8 = 2 + par1World.rand.nextInt(3);
-
-                if (par7 > 0)
-                {
-                    var8 += par1World.rand.nextInt(par7 + 1);
-                }
-            }
-
-            for (int var9 = 0; var9 < var8; ++var9)
-            {
-                this.dropBlockAsItem_do(par1World, par2, par3, par4, new ItemStack(Item.netherStalkSeeds));
-            }
-        }
+        super.dropBlockAsItemWithChance(par1World, par2, par3, par4, par5, par6, par7);
     }
 
     /**
@@ -119,4 +105,23 @@
     {
         return Item.netherStalkSeeds.itemID;
     }
+
+    @Override
+    public ArrayList<ItemStack> getBlockDropped(World world, int x, int y, int z, int metadata, int fortune)
+    {
+        ArrayList<ItemStack> ret = new ArrayList<ItemStack>();
+        int count = 1;
+
+        if (metadata >= 3)
+        {
+            count = 2 + world.rand.nextInt(3) + (fortune > 0 ? world.rand.nextInt(fortune + 1) : 0);
+        }
+
+        for (int i = 0; i < count; i++)
+        {
+            ret.add(new ItemStack(Item.netherStalkSeeds));
+        }
+
+        return ret;
+    }
 }
