--- ../src_base/minecraft/net/minecraft/block/BlockCactus.java
+++ ../src_work/minecraft/net/minecraft/block/BlockCactus.java
@@ -7,12 +7,18 @@
 import net.minecraft.client.renderer.texture.IconRegister;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.util.AxisAlignedBB;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.Icon;
 import net.minecraft.world.World;
 
-public class BlockCactus extends Block
+import net.minecraftforge.common.EnumPlantType;
+import net.minecraftforge.common.ForgeDirection;
+import net.minecraftforge.common.IPlantable;
+import org.bukkit.event.entity.EntityDamageByBlockEvent;  // CraftBukkit
+
+public class BlockCactus extends Block implements IPlantable
 {
     @SideOnly(Side.CLIENT)
     private Icon cactusTopIcon;
@@ -44,9 +50,9 @@
             {
                 int i1 = par1World.getBlockMetadata(par2, par3, par4);
 
-                if (i1 == 15)
+                if (i1 >= (byte) range(3, (par1World.growthOdds / par1World.spigotConfig.cactusModifier * 15) + 0.5F, 15)) // Spigot
                 {
-                    par1World.setBlock(par2, par3 + 1, par4, this.blockID);
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(par1World, par2, par3 + 1, par4, this.blockID, 0); // CraftBukkit
                     par1World.setBlockMetadataWithNotify(par2, par3, par4, 0, 4);
                     this.onNeighborBlockChange(par1World, par2, par3 + 1, par4, this.blockID);
                 }
@@ -158,7 +164,7 @@
         else
         {
             int l = par1World.getBlockId(par2, par3 - 1, par4);
-            return l == Block.cactus.blockID || l == Block.sand.blockID;
+            return blocksList[l] != null && blocksList[l].canSustainPlant(par1World, par2, par3 - 1, par4, ForgeDirection.UP, this);
         }
     }
 
@@ -167,6 +173,24 @@
      */
     public void onEntityCollidedWithBlock(World par1World, int par2, int par3, int par4, Entity par5Entity)
     {
+        // CraftBukkit start - EntityDamageByBlock event
+        if (par5Entity instanceof EntityLivingBase)
+        {
+            org.bukkit.block.Block damager = par1World.getWorld().getBlockAt(par2, par3, par4);
+            org.bukkit.entity.Entity damagee = (par5Entity == null) ? null : par5Entity.getBukkitEntity();
+            EntityDamageByBlockEvent event = new EntityDamageByBlockEvent(damager, damagee, org.bukkit.event.entity.EntityDamageEvent.DamageCause.CONTACT, 1D);
+            par1World.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled())
+            {
+                damagee.setLastDamageCause(event);
+                par5Entity.attackEntityFrom(DamageSource.cactus, (float) event.getDamage());
+            }
+
+            return;
+        }
+
+        // CraftBukkit end
         par5Entity.attackEntityFrom(DamageSource.cactus, 1.0F);
     }
 
@@ -182,4 +206,22 @@
         this.cactusTopIcon = par1IconRegister.registerIcon(this.getTextureName() + "_top");
         this.cactusBottomIcon = par1IconRegister.registerIcon(this.getTextureName() + "_bottom");
     }
+
+    @Override
+    public EnumPlantType getPlantType(World world, int x, int y, int z)
+    {
+        return EnumPlantType.Desert;
+    }
+
+    @Override
+    public int getPlantID(World world, int x, int y, int z)
+    {
+        return blockID;
+    }
+
+    @Override
+    public int getPlantMetadata(World world, int x, int y, int z)
+    {
+        return -1;
+    }
 }
