--- ../src-base/minecraft/net/minecraft/block/BlockSand.java
+++ ../src-work/minecraft/net/minecraft/block/BlockSand.java
@@ -28,7 +28,6 @@
         return ((BlockSand.EnumType)p_180651_1_.func_177229_b(field_176504_a)).func_176688_a();
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
     {
         for (BlockSand.EnumType blocksand$enumtype : BlockSand.EnumType.values())
