--- ../src-base/minecraft/net/minecraft/block/BlockRailPowered.java
+++ ../src-work/minecraft/net/minecraft/block/BlockRailPowered.java
@@ -23,9 +23,17 @@
     });
     public static final PropertyBool field_176569_M = PropertyBool.func_177716_a("powered");
 
+    private final boolean isActivator;
+
     protected BlockRailPowered()
     {
+        this(false);
+    }
+
+    protected BlockRailPowered(boolean isActivator)
+    {
         super(true);
+        this.isActivator = isActivator;
         this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176568_b, BlockRailBase.EnumRailDirection.NORTH_SOUTH).func_177226_a(field_176569_M, Boolean.valueOf(false)));
     }
 
@@ -146,7 +154,7 @@
     {
         IBlockState iblockstate = p_176567_1_.func_180495_p(p_176567_2_);
 
-        if (iblockstate.func_177230_c() != this)
+        if (!(iblockstate.func_177230_c() instanceof BlockRailPowered) || isActivator != ((BlockRailPowered)iblockstate.func_177230_c()).isActivator)
         {
             return false;
         }
