--- ../src-base/minecraft/net/minecraft/block/BlockLilyPad.java
+++ ../src-work/minecraft/net/minecraft/block/BlockLilyPad.java
@@ -1,7 +1,9 @@
 package net.minecraft.block;
 
 import java.util.List;
+
 import javax.annotation.Nullable;
+
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.creativetab.CreativeTabs;
@@ -22,6 +24,7 @@
         this.func_149647_a(CreativeTabs.field_78031_c);
     }
 
+    @Override
     public void func_185477_a(IBlockState p_185477_1_, World p_185477_2_, BlockPos p_185477_3_, AxisAlignedBB p_185477_4_, List<AxisAlignedBB> p_185477_5_, @Nullable Entity p_185477_6_, boolean p_185477_7_)
     {
         if (!(p_185477_6_ instanceof EntityBoat))
@@ -30,6 +33,7 @@
         }
     }
 
+    @Override
     public void func_180634_a(World p_180634_1_, BlockPos p_180634_2_, IBlockState p_180634_3_, Entity p_180634_4_)
     {
         super.func_180634_a(p_180634_1_, p_180634_2_, p_180634_3_, p_180634_4_);
@@ -40,23 +44,26 @@
         }
     }
 
+    @Override
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
         return field_185523_a;
     }
 
+    @Override
     protected boolean func_185514_i(IBlockState p_185514_1_)
     {
         return p_185514_1_.func_177230_c() == Blocks.field_150355_j || p_185514_1_.func_185904_a() == Material.field_151588_w;
     }
 
+    @Override
     public boolean func_180671_f(World p_180671_1_, BlockPos p_180671_2_, IBlockState p_180671_3_)
     {
         if (p_180671_2_.func_177956_o() >= 0 && p_180671_2_.func_177956_o() < 256)
         {
             IBlockState iblockstate = p_180671_1_.func_180495_p(p_180671_2_.func_177977_b());
             Material material = iblockstate.func_185904_a();
-            return material == Material.field_151586_h && ((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue() == 0 || material == Material.field_151588_w;
+            return material.func_76224_d() && iblockstate.func_177229_b(BlockLiquid.field_176367_b).intValue() == 0 || material == Material.field_151588_w;
         }
         else
         {
@@ -64,6 +71,7 @@
         }
     }
 
+    @Override
     public int func_176201_c(IBlockState p_176201_1_)
     {
         return 0;
