--- ../src_base/minecraft/net/minecraft/block/BlockCactus.java
+++ ../src_work/minecraft/net/minecraft/block/BlockCactus.java
@@ -10,7 +10,11 @@
 import net.minecraft.util.DamageSource;
 import net.minecraft.world.World;
 
-public class BlockCactus extends Block
+import net.minecraftforge.common.EnumPlantType;
+import net.minecraftforge.common.ForgeDirection;
+import net.minecraftforge.common.IPlantable;
+
+public class BlockCactus extends Block implements IPlantable
 {
     protected BlockCactus(int par1, int par2)
     {
@@ -149,7 +153,7 @@
         else
         {
             int var5 = par1World.getBlockId(par2, par3 - 1, par4);
-            return var5 == Block.cactus.blockID || var5 == Block.sand.blockID;
+            return blocksList[var5] != null && blocksList[var5].canSustainPlant(par1World, par2, par3 - 1, par4, ForgeDirection.UP, this);
         }
     }
 
@@ -160,4 +164,22 @@
     {
         par5Entity.attackEntityFrom(DamageSource.cactus, 1);
     }
+
+    @Override
+    public EnumPlantType getPlantType(World world, int x, int y, int z)
+    {
+        return EnumPlantType.Desert;
+    }
+
+    @Override
+    public int getPlantID(World world, int x, int y, int z)
+    {
+        return blockID;
+    }
+
+    @Override
+    public int getPlantMetadata(World world, int x, int y, int z)
+    {
+        return -1;
+    }
 }
