--- a/net/minecraft/block/BlockCactus.java
+++ b/net/minecraft/block/BlockCactus.java
@@ -21,7 +21,7 @@
 import net.minecraft.world.IWorldReaderBase;
 import net.minecraft.world.World;
 
-public class BlockCactus extends Block {
+public class BlockCactus extends Block implements net.minecraftforge.common.IPlantable {
    public static final IntegerProperty field_176587_a = BlockStateProperties.field_208171_X;
    protected static final VoxelShape field_196400_b = Block.func_208617_a(1.0D, 0.0D, 1.0D, 15.0D, 15.0D, 15.0D);
    protected static final VoxelShape field_196401_c = Block.func_208617_a(1.0D, 0.0D, 1.0D, 15.0D, 16.0D, 15.0D);
@@ -32,6 +32,7 @@
    }
 
    public void func_196267_b(IBlockState p_196267_1_, World p_196267_2_, BlockPos p_196267_3_, Random p_196267_4_) {
+      if (!p_196267_2_.func_205050_e(p_196267_3_, 1)) return; // Forge: prevent growing cactus from loading unloaded chunks with block update
       if (!p_196267_1_.func_196955_c(p_196267_2_, p_196267_3_)) {
          p_196267_2_.func_175655_b(p_196267_3_, true);
       } else {
@@ -44,6 +45,7 @@
 
             if (i < 3) {
                int j = p_196267_1_.func_177229_b(field_176587_a);
+               if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_196267_2_, blockpos, p_196267_1_, true)) {
                if (j == 15) {
                   p_196267_2_.func_175656_a(blockpos, this.func_176223_P());
                   IBlockState iblockstate = p_196267_1_.func_206870_a(field_176587_a, Integer.valueOf(0));
@@ -52,7 +54,8 @@
                } else {
                   p_196267_2_.func_180501_a(p_196267_3_, p_196267_1_.func_206870_a(field_176587_a, Integer.valueOf(j + 1)), 4);
                }
-
+               net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_196267_2_, p_196267_3_, p_196267_1_);
+               }
             }
          }
       }
@@ -91,8 +94,8 @@
          }
       }
 
-      Block block = p_196260_2_.func_180495_p(p_196260_3_.func_177977_b()).func_177230_c();
-      return (block == Blocks.field_150434_aF || block == Blocks.field_150354_m || block == Blocks.field_196611_F) && !p_196260_2_.func_180495_p(p_196260_3_.func_177984_a()).func_185904_a().func_76224_d();
+      IBlockState soil = p_196260_2_.func_180495_p(p_196260_3_.func_177977_b());
+      return soil.canSustainPlant(p_196260_2_, p_196260_3_.func_177977_b(), EnumFacing.UP, this) && !p_196260_2_.func_180495_p(p_196260_3_.func_177984_a()).func_185904_a().func_76224_d();
    }
 
    public void func_196262_a(IBlockState p_196262_1_, World p_196262_2_, BlockPos p_196262_3_, Entity p_196262_4_) {
@@ -114,4 +117,14 @@
    public boolean func_196266_a(IBlockState p_196266_1_, IBlockReader p_196266_2_, BlockPos p_196266_3_, PathType p_196266_4_) {
       return false;
    }
+
+   @Override
+   public net.minecraftforge.common.EnumPlantType getPlantType(IBlockReader world, BlockPos pos) {
+       return net.minecraftforge.common.EnumPlantType.Desert;
+   }
+
+   @Override
+   public IBlockState getPlant(IBlockReader world, BlockPos pos) {
+       return func_176223_P();
+   }
 }
