--- ../src_base/minecraft/net/minecraft/block/BlockTrapDoor.java
+++ ../src_work/minecraft/net/minecraft/block/BlockTrapDoor.java
@@ -11,8 +11,13 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import net.minecraftforge.common.ForgeDirection;
+
 public class BlockTrapDoor extends Block
 {
+    /** Set this to allow trapdoors to remain free-floating */
+    public static boolean disableValidation = false;
+
     protected BlockTrapDoor(int par1, Material par2Material)
     {
         super(par1, par2Material);
@@ -144,7 +149,7 @@
      */
     public boolean onBlockActivated(World par1World, int par2, int par3, int par4, EntityPlayer par5EntityPlayer, int par6, float par7, float par8, float par9)
     {
-        if (this.blockMaterial == Material.iron)
+        if (this.getBlockMaterial(par1World.getBlockMetadata(par2, par3, par4)) == Material.iron)
         {
             return true;
         }
@@ -201,7 +206,7 @@
                 --var7;
             }
 
-            if (!isValidSupportBlock(par1World.getBlockId(var7, par3, var8)))
+            if (!(isValidSupportBlock(par1World.getBlockId(var7, par3, var8), par1World.getBlockId(var7, par3, var8)) || par1World.isBlockSolidOnSide(var7, par3, var8, ForgeDirection.getOrientation((var6 & 3) + 2))))
             {
                 par1World.setBlockWithNotify(par2, par3, par4, 0);
                 this.dropBlockAsItem(par1World, par2, par3, par4, var6, 0);
@@ -266,6 +271,10 @@
      */
     public boolean canPlaceBlockOnSide(World par1World, int par2, int par3, int par4, int par5)
     {
+        if (disableValidation)
+        {
+            return true;
+        }
         if (par5 == 0)
         {
             return false;
@@ -296,7 +305,7 @@
                 --par2;
             }
 
-            return isValidSupportBlock(par1World.getBlockId(par2, par3, par4));
+            return isValidSupportBlock(par1World.getBlockId(par2, par3, par4), par1World.getBlockMetadata(par2, par3, par4)) || par1World.isBlockSolidOnSide(par2, par3, par4, ForgeDirection.UP);
         }
     }
 
@@ -309,8 +318,12 @@
      * Checks if the block ID is a valid support block for the trap door to connect with. If it is not the trapdoor is
      * dropped into the world.
      */
-    private static boolean isValidSupportBlock(int par0)
-    {
+    private static boolean isValidSupportBlock(int par0, int par1)
+    {
+        if (disableValidation)
+        {
+            return true;
+        }
         if (par0 <= 0)
         {
             return false;
@@ -318,7 +331,7 @@
         else
         {
             Block var1 = Block.blocksList[par0];
-            return var1 != null && var1.blockMaterial.isOpaque() && var1.renderAsNormalBlock() || var1 == Block.glowStone || var1 instanceof BlockHalfSlab || var1 instanceof BlockStairs;
+            return var1 != null && var1.getBlockMaterial(par1).isOpaque() && var1.renderAsNormalBlock() || var1 == Block.glowStone || var1 instanceof BlockHalfSlab || var1 instanceof BlockStairs;
         }
     }
 }
