--- ../src_base/minecraft/net/minecraft/block/BlockDaylightDetector.java
+++ ../src_work/minecraft/net/minecraft/block/BlockDaylightDetector.java
@@ -90,6 +90,7 @@
 
             if (l != i1)
             {
+                i1 = org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(par1World, par2, par3, par4, l, i1).getNewCurrent(); // CraftBukkit - Call BlockRedstoneEvent
                 par1World.setBlockMetadataWithNotify(par2, par3, par4, i1, 3);
             }
         }
@@ -146,7 +147,7 @@
      */
     public void registerIcons(IconRegister par1IconRegister)
     {
-        this.iconArray[0] = par1IconRegister.registerIcon(this.func_111023_E() + "_top");
-        this.iconArray[1] = par1IconRegister.registerIcon(this.func_111023_E() + "_side");
+        this.iconArray[0] = par1IconRegister.registerIcon(this.getTextureName() + "_top");
+        this.iconArray[1] = par1IconRegister.registerIcon(this.getTextureName() + "_side");
     }
 }
