--- a/net/minecraft/block/BlockOre.java
+++ b/net/minecraft/block/BlockOre.java
@@ -49,23 +49,27 @@
 
    public void func_196255_a(IBlockState p_196255_1_, World p_196255_2_, BlockPos p_196255_3_, float p_196255_4_, int p_196255_5_) {
       super.func_196255_a(p_196255_1_, p_196255_2_, p_196255_3_, p_196255_4_, p_196255_5_);
-      if (this.func_199769_a(p_196255_1_, p_196255_2_, p_196255_3_, p_196255_5_) != this) {
+   }
+
+   @Override
+   public int getExpDrop(IBlockState state, net.minecraft.world.IWorldReader reader, BlockPos pos, int fortune) {
+      World world = reader instanceof World ? (World)reader : null;
+      if (world == null || this.func_199769_a(state, world, pos, fortune) != this) {
          int i = 0;
          if (this == Blocks.field_150365_q) {
-            i = MathHelper.func_76136_a(p_196255_2_.field_73012_v, 0, 2);
+            i = MathHelper.func_76136_a(RANDOM, 0, 2);
          } else if (this == Blocks.field_150482_ag) {
-            i = MathHelper.func_76136_a(p_196255_2_.field_73012_v, 3, 7);
+            i = MathHelper.func_76136_a(RANDOM, 3, 7);
          } else if (this == Blocks.field_150412_bA) {
-            i = MathHelper.func_76136_a(p_196255_2_.field_73012_v, 3, 7);
+            i = MathHelper.func_76136_a(RANDOM, 3, 7);
          } else if (this == Blocks.field_150369_x) {
-            i = MathHelper.func_76136_a(p_196255_2_.field_73012_v, 2, 5);
+            i = MathHelper.func_76136_a(RANDOM, 2, 5);
          } else if (this == Blocks.field_196766_fg) {
-            i = MathHelper.func_76136_a(p_196255_2_.field_73012_v, 2, 5);
+            i = MathHelper.func_76136_a(RANDOM, 2, 5);
          }
-
-         this.func_180637_b(p_196255_2_, p_196255_3_, i);
+         return i;
       }
-
+      return 0;
    }
 
    public ItemStack func_185473_a(IBlockReader p_185473_1_, BlockPos p_185473_2_, IBlockState p_185473_3_) {
