--- ../src-base/minecraft/net/minecraft/block/BlockBush.java
+++ ../src-work/minecraft/net/minecraft/block/BlockBush.java
@@ -6,8 +6,13 @@
 import net.minecraft.init.Blocks;
 import net.minecraft.util.AxisAlignedBB;
 import net.minecraft.world.World;
+import net.minecraft.world.IBlockAccess;
+import net.minecraftforge.common.EnumPlantType;
+import net.minecraftforge.common.util.ForgeDirection;
+import net.minecraftforge.common.IPlantable;
+import static net.minecraftforge.common.EnumPlantType.*;
 
-public class BlockBush extends Block
+public class BlockBush extends Block implements IPlantable
 {
     private static final String __OBFID = "CL_00000208";
 
@@ -28,7 +33,7 @@
     // JAVADOC METHOD $$ func_149742_c
     public boolean canPlaceBlockAt(World p_149742_1_, int p_149742_2_, int p_149742_3_, int p_149742_4_)
     {
-        return super.canPlaceBlockAt(p_149742_1_, p_149742_2_, p_149742_3_, p_149742_4_) && this.canPlaceBlockOn(p_149742_1_.getBlock(p_149742_2_, p_149742_3_ - 1, p_149742_4_));
+        return super.canPlaceBlockAt(p_149742_1_, p_149742_2_, p_149742_3_, p_149742_4_) && this.canBlockStay(p_149742_1_, p_149742_2_, p_149742_3_, p_149742_4_);
     }
 
     // JAVADOC METHOD $$ func_149854_a
@@ -63,7 +68,7 @@
     // JAVADOC METHOD $$ func_149718_j
     public boolean canBlockStay(World p_149718_1_, int p_149718_2_, int p_149718_3_, int p_149718_4_)
     {
-        return this.canPlaceBlockOn(p_149718_1_.getBlock(p_149718_2_, p_149718_3_ - 1, p_149718_4_));
+        return  p_149718_1_.getBlock(p_149718_2_, p_149718_3_ - 1, p_149718_4_).canSustainPlant(p_149718_1_, p_149718_2_, p_149718_3_ - 1, p_149718_4_, ForgeDirection.UP, this);
     }
 
     // JAVADOC METHOD $$ func_149668_a
@@ -89,4 +94,37 @@
     {
         return 1;
     }
+
+    @Override
+    public EnumPlantType getPlantType(IBlockAccess world, int x, int y, int z)
+    {
+        if (this == Blocks.wheat)          return Crop;
+        if (this == Blocks.carrots)        return Crop;
+        if (this == Blocks.potatoes)       return Crop;
+        if (this == Blocks.melon_stem)     return Crop;
+        if (this == Blocks.pumpkin_stem)   return Crop;
+        if (this == Blocks.deadbush)       return Desert;
+        if (this == Blocks.waterlily)      return Water;
+        if (this == Blocks.red_mushroom)   return Cave;
+        if (this == Blocks.brown_mushroom) return Cave;
+        if (this == Blocks.nether_wart)    return Nether;
+        if (this == Blocks.sapling)        return Plains;
+        if (this == Blocks.tallgrass)      return Plains;
+        if (this == Blocks.double_plant)   return Plains;
+        if (this == Blocks.red_flower)     return Plains;
+        if (this == Blocks.yellow_flower)  return Plains;        
+        return Plains;
+    }
+
+    @Override
+    public Block getPlant(IBlockAccess world, int x, int y, int z)
+    {
+        return this;
+    }
+
+    @Override
+    public int getPlantMetadata(IBlockAccess world, int x, int y, int z)
+    {
+        return world.getBlockMetadata(x, y, z);
+    }
 }
