--- ../src_base/minecraft/net/minecraft/block/BlockSapling.java
+++ ../src_work/minecraft/net/minecraft/block/BlockSapling.java
@@ -13,6 +13,8 @@
 import net.minecraft.world.gen.feature.WorldGenTaiga2;
 import net.minecraft.world.gen.feature.WorldGenTrees;
 import net.minecraft.world.gen.feature.WorldGenerator;
+
+import net.minecraftforge.event.terraingen.TerrainGen;
 
 public class BlockSapling extends BlockFlower
 {
@@ -65,6 +67,8 @@
      */
     public void growTree(World par1World, int par2, int par3, int par4, Random par5Random)
     {
+        if (!TerrainGen.saplingGrowTree(par1World, par5Random, par2, par3, par4)) return;
+
         int var6 = par1World.getBlockMetadata(par2, par3, par4) & 3;
         Object var7 = null;
         int var8 = 0;
