--- ../src-base/minecraft/net/minecraft/block/BlockCactus.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCactus.java
@@ -18,7 +18,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class BlockCactus extends Block
+public class BlockCactus extends Block implements net.minecraftforge.common.IPlantable
 {
     public static final PropertyInteger field_176587_a = PropertyInteger.func_177719_a("age", 0, 15);
     protected static final AxisAlignedBB field_185593_b = new AxisAlignedBB(0.0625D, 0.0D, 0.0625D, 0.9375D, 0.9375D, 0.9375D);
@@ -48,18 +48,22 @@
             if (i < 3)
             {
                 int j = ((Integer)p_180650_3_.func_177229_b(field_176587_a)).intValue();
-
-                if (j == 15)
-                {
-                    p_180650_1_.func_175656_a(blockpos, this.func_176223_P());
-                    IBlockState iblockstate = p_180650_3_.func_177226_a(field_176587_a, Integer.valueOf(0));
-                    p_180650_1_.func_180501_a(p_180650_2_, iblockstate, 4);
-                    iblockstate.func_189546_a(p_180650_1_, blockpos, this);
+                net.minecraftforge.event.world.BlockEvent ev = new net.minecraftforge.event.world.BlockEvent.CropGrowEvent.Pre(p_180650_1_,p_180650_2_,p_180650_3_);
+                net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(ev);
+                if(ev.getResult() != net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) {
+                    if (j == 15)
+                    {
+                        p_180650_1_.func_175656_a(blockpos, this.func_176223_P());
+                        IBlockState iblockstate = p_180650_3_.func_177226_a(field_176587_a, Integer.valueOf(0));
+                        p_180650_1_.func_180501_a(p_180650_2_, iblockstate, 4);
+                        iblockstate.func_189546_a(p_180650_1_, blockpos, this);
+                    }
+                    else
+                    {
+                        p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(field_176587_a, Integer.valueOf(j + 1)), 4);
+                    }
+                    net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.BlockEvent.CropGrowEvent.Post(p_180650_1_, p_180650_2_, p_180650_3_, p_180650_1_.func_180495_p(p_180650_2_)));
                 }
-                else
-                {
-                    p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(field_176587_a, Integer.valueOf(j + 1)), 4);
-                }
             }
         }
     }
@@ -110,8 +114,8 @@
             }
         }
 
-        Block block = p_176586_1_.func_180495_p(p_176586_2_.func_177977_b()).func_177230_c();
-        return block == Blocks.field_150434_aF || block == Blocks.field_150354_m && !p_176586_1_.func_180495_p(p_176586_2_.func_177984_a()).func_185904_a().func_76224_d();
+        IBlockState state = p_176586_1_.func_180495_p(p_176586_2_.func_177977_b());
+        return state.func_177230_c().canSustainPlant(state, p_176586_1_, p_176586_2_.func_177977_b(), EnumFacing.UP, this) && !p_176586_1_.func_180495_p(p_176586_2_.func_177984_a()).func_185904_a().func_76224_d();
     }
 
     public void func_180634_a(World p_180634_1_, BlockPos p_180634_2_, IBlockState p_180634_3_, Entity p_180634_4_)
@@ -135,6 +139,18 @@
         return ((Integer)p_176201_1_.func_177229_b(field_176587_a)).intValue();
     }
 
+    @Override
+    public net.minecraftforge.common.EnumPlantType getPlantType(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return net.minecraftforge.common.EnumPlantType.Desert;
+    }
+
+    @Override
+    public IBlockState getPlant(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return func_176223_P();
+    }
+
     protected BlockStateContainer func_180661_e()
     {
         return new BlockStateContainer(this, new IProperty[] {field_176587_a});
