--- ../src-base/minecraft/net/minecraft/block/BlockTrapDoor.java
+++ ../src-work/minecraft/net/minecraft/block/BlockTrapDoor.java
@@ -240,6 +240,18 @@
         return new BlockStateContainer(this, new IProperty[] {field_176284_a, field_176283_b, field_176285_M});
     }
 
+    @Override
+    public boolean isLadder(IBlockState state, IBlockAccess world, BlockPos pos, EntityLivingBase entity)
+    {
+        if (state.func_177229_b(field_176283_b))
+        {
+            IBlockState down = world.func_180495_p(pos.func_177977_b());
+            if (down.func_177230_c() == net.minecraft.init.Blocks.field_150468_ap)
+                return down.func_177229_b(BlockLadder.field_176382_a) == state.func_177229_b(field_176284_a);
+        }
+        return false;
+    }
+
     public static enum DoorHalf implements IStringSerializable
     {
         TOP("top"),
