--- ../src_base/minecraft/net/minecraft/block/BlockOre.java
+++ ../src_work/minecraft/net/minecraft/block/BlockOre.java
@@ -59,34 +59,36 @@
     public void dropBlockAsItemWithChance(World par1World, int par2, int par3, int par4, int par5, float par6, int par7)
     {
         super.dropBlockAsItemWithChance(par1World, par2, par3, par4, par5, par6, par7);
-
-        if (this.idDropped(par5, par1World.rand, par7) != this.blockID)
+    }
+    
+    @Override
+    public int getExpDrop(World world, int data, int enchantmentLevel)
+    {
+        int j1 = 0;
+        if (this.idDropped(data, world.rand, enchantmentLevel) != this.blockID)
         {
-            int j1 = 0;
-
             if (this.blockID == Block.oreCoal.blockID)
             {
-                j1 = MathHelper.getRandomIntegerInRange(par1World.rand, 0, 2);
+                j1 = MathHelper.getRandomIntegerInRange(world.rand, 0, 2);
             }
             else if (this.blockID == Block.oreDiamond.blockID)
             {
-                j1 = MathHelper.getRandomIntegerInRange(par1World.rand, 3, 7);
+                j1 = MathHelper.getRandomIntegerInRange(world.rand, 3, 7);
             }
             else if (this.blockID == Block.oreEmerald.blockID)
             {
-                j1 = MathHelper.getRandomIntegerInRange(par1World.rand, 3, 7);
+                j1 = MathHelper.getRandomIntegerInRange(world.rand, 3, 7);
             }
             else if (this.blockID == Block.oreLapis.blockID)
             {
-                j1 = MathHelper.getRandomIntegerInRange(par1World.rand, 2, 5);
+                j1 = MathHelper.getRandomIntegerInRange(world.rand, 2, 5);
             }
             else if (this.blockID == Block.oreNetherQuartz.blockID)
             {
-                j1 = MathHelper.getRandomIntegerInRange(par1World.rand, 2, 5);
+                j1 = MathHelper.getRandomIntegerInRange(world.rand, 2, 5);
             }
-
-            this.dropXpOnBlockBreak(par1World, par2, par3, par4, j1);
         }
+        return j1;
     }
 
     /**
