--- ../src-base/minecraft/net/minecraft/block/BlockFurnace.java
+++ ../src-work/minecraft/net/minecraft/block/BlockFurnace.java
@@ -2,7 +2,9 @@
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
+
 import java.util.Random;
+
 import net.minecraft.block.material.Material;
 import net.minecraft.client.renderer.texture.IIconRegister;
 import net.minecraft.entity.EntityLivingBase;
@@ -18,6 +20,7 @@
 import net.minecraft.tileentity.TileEntityFurnace;
 import net.minecraft.util.IIcon;
 import net.minecraft.util.MathHelper;
+import net.minecraft.util.Util;
 import net.minecraft.world.World;
 
 public class BlockFurnace extends BlockContainer
@@ -185,7 +188,7 @@
     {
         if (!field_149934_M)
         {
-            TileEntityFurnace tileentityfurnace = (TileEntityFurnace)p_149749_1_.getTileEntity(p_149749_2_, p_149749_3_, p_149749_4_);
+            TileEntityFurnace tileentityfurnace = Util.getTileBehavior(p_149749_1_, p_149749_2_, p_149749_3_, p_149749_4_, TileEntityFurnace.class);
 
             if (tileentityfurnace != null)
             {
