--- ../src-base/minecraft/net/minecraft/block/BlockSilverfish.java
+++ ../src-work/minecraft/net/minecraft/block/BlockSilverfish.java
@@ -75,7 +75,6 @@
         return new ItemStack(this, 1, p_185473_3_.func_177230_c().func_176201_c(p_185473_3_));
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
     {
         for (BlockSilverfish.EnumType blocksilverfish$enumtype : BlockSilverfish.EnumType.values())
