--- a/net/minecraft/block/BlockRedstoneTorch.java
+++ b/net/minecraft/block/BlockRedstoneTorch.java
@@ -24,7 +24,7 @@
 
 public class BlockRedstoneTorch extends BlockTorch {
    public static final BooleanProperty field_196528_a = BlockStateProperties.field_208190_q;
-   private static final Map<IBlockReader, List<BlockRedstoneTorch.Toggle>> field_196529_b = Maps.newHashMap();
+   private static final Map<IBlockReader, List<BlockRedstoneTorch.Toggle>> field_196529_b = new java.util.WeakHashMap<IBlockReader, List<BlockRedstoneTorch.Toggle>>(); // FORGE - fix vanilla MC-101233
 
    protected BlockRedstoneTorch(Block.Properties p_i48342_1_) {
       super(p_i48342_1_);
