--- ../src-base/minecraft/net/minecraft/block/BlockPlanks.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPlanks.java
@@ -31,7 +31,6 @@
         return ((BlockPlanks.EnumType)p_180651_1_.func_177229_b(field_176383_a)).func_176839_a();
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
     {
         for (BlockPlanks.EnumType blockplanks$enumtype : BlockPlanks.EnumType.values())
