--- ../src-base/minecraft/net/minecraft/block/BlockCrops.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCrops.java
@@ -2,6 +2,7 @@
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
+import java.util.ArrayList;
 import java.util.Random;
 import net.minecraft.client.renderer.texture.IIconRegister;
 import net.minecraft.creativetab.CreativeTabs;
@@ -12,6 +13,7 @@
 import net.minecraft.util.IIcon;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
+import net.minecraftforge.common.util.ForgeDirection;
 
 public class BlockCrops extends BlockBush implements IGrowable
 {
@@ -89,11 +91,11 @@
             {
                 float f1 = 0.0F;
 
-                if (p_149864_1_.func_147439_a(l, p_149864_3_ - 1, i1) == Blocks.field_150458_ak)
+                if (p_149864_1_.func_147439_a(l, p_149864_3_ - 1, i1).canSustainPlant(p_149864_1_, l, p_149864_3_ - 1, i1, ForgeDirection.UP, this))
                 {
                     f1 = 1.0F;
 
-                    if (p_149864_1_.func_72805_g(l, p_149864_3_ - 1, i1) > 0)
+                    if (p_149864_1_.func_147439_a(l, p_149864_3_ - 1, i1).isFertile(p_149864_1_, l, p_149864_3_ - 1, i1))
                     {
                         f1 = 3.0F;
                     }
@@ -145,22 +147,6 @@
     public void func_149690_a(World p_149690_1_, int p_149690_2_, int p_149690_3_, int p_149690_4_, int p_149690_5_, float p_149690_6_, int p_149690_7_)
     {
         super.func_149690_a(p_149690_1_, p_149690_2_, p_149690_3_, p_149690_4_, p_149690_5_, p_149690_6_, 0);
-
-        if (!p_149690_1_.field_72995_K)
-        {
-            if (p_149690_5_ >= 7)
-            {
-                int j1 = 3 + p_149690_7_;
-
-                for (int k1 = 0; k1 < j1; ++k1)
-                {
-                    if (p_149690_1_.field_73012_v.nextInt(15) <= p_149690_5_)
-                    {
-                        this.func_149642_a(p_149690_1_, p_149690_2_, p_149690_3_, p_149690_4_, new ItemStack(this.func_149866_i(), 1, 0));
-                    }
-                }
-            }
-        }
     }
 
     public Item func_149650_a(int p_149650_1_, Random p_149650_2_, int p_149650_3_)
@@ -204,4 +190,23 @@
     {
         this.func_149863_m(p_149853_1_, p_149853_3_, p_149853_4_, p_149853_5_);
     }
+
+    @Override
+    public ArrayList<ItemStack> getDrops(World world, int x, int y, int z, int metadata, int fortune)
+    {
+        ArrayList<ItemStack> ret = super.getDrops(world, x, y, z, metadata, fortune);
+
+        if (metadata >= 7)
+        {
+            for (int i = 0; i < 3 + fortune; ++i)
+            {
+                if (world.field_73012_v.nextInt(15) <= metadata)
+                {
+                    ret.add(new ItemStack(this.func_149866_i(), 1, 0));
+                }
+            }
+        }
+
+        return ret;
+    }
 }
