--- ../src-base/minecraft/net/minecraft/block/BlockPistonBase.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPistonBase.java
@@ -240,7 +240,7 @@
                     }
                 }
 
-                if (!flag1 && iblockstate.func_185904_a() != Material.field_151579_a && func_185646_a(iblockstate, p_189539_2_, blockpos, enumfacing.func_176734_d(), false) && (iblockstate.func_185905_o() == EnumPushReaction.NORMAL || block == Blocks.field_150331_J || block == Blocks.field_150320_F))
+                if (!flag1 && !iblockstate.func_177230_c().isAir(iblockstate, p_189539_2_, p_189539_3_) && func_185646_a(iblockstate, p_189539_2_, blockpos, enumfacing.func_176734_d(), false) && (iblockstate.func_185905_o() == EnumPushReaction.NORMAL || block == Blocks.field_150331_J || block == Blocks.field_150320_F))
                 {
                     this.func_176319_a(p_189539_2_, p_189539_3_, enumfacing, false);
                 }
@@ -306,7 +306,7 @@
                     return false;
                 }
 
-                return !block.func_149716_u();
+                return !block.hasTileEntity(p_185646_0_);
             }
             else
             {
@@ -352,7 +352,9 @@
             {
                 BlockPos blockpos1 = (BlockPos)list2.get(j);
                 IBlockState iblockstate = p_176319_1_.func_180495_p(blockpos1);
-                iblockstate.func_177230_c().func_176226_b(p_176319_1_, blockpos1, iblockstate, 0);
+                // Forge: With our change to how snowballs are dropped this needs to disallow to mimic vanilla behavior.
+                float chance = iblockstate.func_177230_c() instanceof BlockSnow ? -1.0f : 1.0f;
+                iblockstate.func_177230_c().func_180653_a(p_176319_1_, blockpos1, iblockstate, chance, 0);
                 p_176319_1_.func_175698_g(blockpos1);
                 --k;
                 aiblockstate[k] = iblockstate;
