--- ../src-base/minecraft/net/minecraft/block/BlockBed.java
+++ ../src-work/minecraft/net/minecraft/block/BlockBed.java
@@ -84,7 +84,7 @@
                 }
             }
 
-            if (p_180639_1_.field_73011_w.func_76567_e() && p_180639_1_.func_180494_b(p_180639_2_) != Biomes.field_76778_j)
+            if (p_180639_1_.field_73011_w.canSleepHere(p_180639_4_, p_180639_2_))
             {
                 if (((Boolean)p_180639_3_.func_177229_b(field_176471_b)).booleanValue())
                 {
