--- ../src_base/minecraft/net/minecraft/block/BlockPistonBase.java
+++ ../src_work/minecraft/net/minecraft/block/BlockPistonBase.java
@@ -380,7 +380,7 @@
                 return false;
             }
 
-            return !(Block.blocksList[par0] instanceof BlockContainer);
+            return !par1World.blockHasTileEntity(par2, par3, par4);
         }
     }
 
@@ -398,7 +398,7 @@
         {
             if (l1 < 13)
             {
-                if (j1 <= 0 || j1 >= 255)
+                if (j1 <= 0 || j1 >= par0World.getHeight() - 1)
                 {
                     return false;
                 }
@@ -448,7 +448,7 @@
 
             if (l1 < 13)
             {
-                if (j1 <= 0 || j1 >= 255)
+                if (j1 <= 0 || j1 >= par1World.getHeight() - 1)
                 {
                     return false;
                 }
