--- ../src-base/minecraft/net/minecraft/block/BlockCocoa.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCocoa.java
@@ -43,13 +43,19 @@
         {
             this.func_176500_f(p_180650_1_, p_180650_2_, p_180650_3_);
         }
-        else if (p_180650_1_.field_73012_v.nextInt(5) == 0)
+        else
         {
-            int i = ((Integer)p_180650_3_.func_177229_b(field_176501_a)).intValue();
-
-            if (i < 2)
+            net.minecraftforge.event.world.BlockEvent ev = new net.minecraftforge.event.world.BlockEvent.CropGrowEvent.Pre(p_180650_1_,p_180650_2_,p_180650_3_);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(ev);
+            if(ev.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW || (ev.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.DEFAULT && p_180650_1_.field_73012_v.nextInt(5) == 0))
             {
-                p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(field_176501_a, Integer.valueOf(i + 1)), 2);
+                int i = ((Integer)p_180650_3_.func_177229_b(field_176501_a)).intValue();
+
+                if (i < 2)
+                {
+                    p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(field_176501_a, Integer.valueOf(i + 1)), 2);
+                    net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.BlockEvent.CropGrowEvent.Post(p_180650_1_, p_180650_2_, p_180650_3_, p_180650_1_.func_180495_p(p_180650_2_)));
+                }
             }
         }
     }
@@ -131,7 +137,14 @@
 
     public void func_180653_a(World p_180653_1_, BlockPos p_180653_2_, IBlockState p_180653_3_, float p_180653_4_, int p_180653_5_)
     {
-        int i = ((Integer)p_180653_3_.func_177229_b(field_176501_a)).intValue();
+        super.func_180653_a(p_180653_1_, p_180653_2_, p_180653_3_, p_180653_4_, p_180653_5_);
+    }
+
+    @Override
+    public java.util.List<ItemStack> getDrops(IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
+        java.util.List<ItemStack> dropped = super.getDrops(world, pos, state, fortune);
+        int i = ((Integer)state.func_177229_b(field_176501_a)).intValue();
         int j = 1;
 
         if (i >= 2)
@@ -141,8 +154,9 @@
 
         for (int k = 0; k < j; ++k)
         {
-            func_180635_a(p_180653_1_, p_180653_2_, new ItemStack(Items.field_151100_aR, 1, EnumDyeColor.BROWN.func_176767_b()));
+            dropped.add(new ItemStack(Items.field_151100_aR, 1, EnumDyeColor.BROWN.func_176767_b()));
         }
+        return dropped;
     }
 
     public ItemStack func_185473_a(World p_185473_1_, BlockPos p_185473_2_, IBlockState p_185473_3_)
