--- ../src-base/minecraft/net/minecraft/block/BlockBanner.java
+++ ../src-work/minecraft/net/minecraft/block/BlockBanner.java
@@ -104,14 +104,7 @@
 
     public void func_180653_a(World p_180653_1_, BlockPos p_180653_2_, IBlockState p_180653_3_, float p_180653_4_, int p_180653_5_)
     {
-        ItemStack itemstack = this.func_185549_e(p_180653_1_, p_180653_2_, p_180653_3_);
-
-        if (itemstack != null)
         {
-            func_180635_a(p_180653_1_, p_180653_2_, itemstack);
-        }
-        else
-        {
             super.func_180653_a(p_180653_1_, p_180653_2_, p_180653_3_, p_180653_4_, p_180653_5_);
         }
     }
@@ -138,6 +131,28 @@
         }
     }
 
+    @Override
+    public java.util.List<ItemStack> getDrops(IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
+        TileEntity te = world.func_175625_s(pos);
+
+        java.util.List<ItemStack> ret = new java.util.ArrayList<ItemStack>();
+        if (te instanceof TileEntityBanner)
+        {
+            TileEntityBanner banner = (TileEntityBanner)te;
+            ItemStack item = new ItemStack(Items.field_179564_cE, 1, banner.func_175115_b());
+            NBTTagCompound nbt = new NBTTagCompound();
+            TileEntityBanner.func_181020_a(nbt, banner.func_175115_b(), banner.func_181021_d());
+            item.func_77983_a("BlockEntityTag", nbt);
+            ret.add(item);
+        }
+        else
+        {
+            ret.add(new ItemStack(Items.field_179564_cE, 1, 0));
+        }
+        return ret;
+    }
+
     public static class BlockBannerHanging extends BlockBanner
         {
             protected static final AxisAlignedBB field_185551_d = new AxisAlignedBB(0.0D, 0.0D, 0.875D, 1.0D, 0.78125D, 1.0D);
