--- ../src-base/minecraft/net/minecraft/block/BlockEnderChest.java
+++ ../src-work/minecraft/net/minecraft/block/BlockEnderChest.java
@@ -95,7 +95,7 @@
 
         if (inventoryenderchest != null && tileentity instanceof TileEntityEnderChest)
         {
-            if (p_180639_1_.func_180495_p(p_180639_2_.func_177984_a()).func_185915_l())
+            if (p_180639_1_.func_180495_p(p_180639_2_.func_177984_a()).isNormalCube(p_180639_1_, p_180639_2_.func_177984_a()))
             {
                 return true;
             }
