--- ../src-base/minecraft/net/minecraft/block/BlockMobSpawner.java
+++ ../src-work/minecraft/net/minecraft/block/BlockMobSpawner.java
@@ -37,10 +37,14 @@
     public void func_180653_a(World p_180653_1_, BlockPos p_180653_2_, IBlockState p_180653_3_, float p_180653_4_, int p_180653_5_)
     {
         super.func_180653_a(p_180653_1_, p_180653_2_, p_180653_3_, p_180653_4_, p_180653_5_);
-        int i = 15 + p_180653_1_.field_73012_v.nextInt(15) + p_180653_1_.field_73012_v.nextInt(15);
-        this.func_180637_b(p_180653_1_, p_180653_2_, i);
     }
 
+    @Override
+    public int getExpDrop(net.minecraft.world.IBlockAccess world, BlockPos pos, int fortune)
+    {
+        return 15 + RANDOM.nextInt(15) + RANDOM.nextInt(15);
+    }
+
     public boolean func_149662_c()
     {
         return false;
