--- ../src-base/minecraft/net/minecraft/block/BlockPrismarine.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPrismarine.java
@@ -60,7 +60,6 @@
         return this.func_176223_P().func_177226_a(field_176332_a, BlockPrismarine.EnumType.func_176810_a(p_176203_1_));
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
     {
         p_149666_2_.add(new ItemStack(this, 1, field_176331_b));
