--- ../src_base/minecraft/net/minecraft/block/BlockCrops.java
+++ ../src_work/minecraft/net/minecraft/block/BlockCrops.java
@@ -2,11 +2,14 @@
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
+
+import java.util.ArrayList;
 import java.util.Random;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+import net.minecraftforge.common.ForgeDirection;
 
 public class BlockCrops extends BlockFlower
 {
@@ -92,11 +95,11 @@
                 int var19 = par1World.getBlockId(var17, par3 - 1, var18);
                 float var20 = 0.0F;
 
-                if (var19 == Block.tilledField.blockID)
+                if (blocksList[var19] != null && blocksList[var19].canSustainPlant(par1World, var17, par3 - 1, var18, ForgeDirection.UP, this))
                 {
                     var20 = 1.0F;
 
-                    if (par1World.getBlockMetadata(var17, par3 - 1, var18) > 0)
+                    if (blocksList[var19].isFertile(par1World, var17, par3 - 1, var18))
                     {
                         var20 = 3.0F;
                     }
@@ -162,22 +165,25 @@
     public void dropBlockAsItemWithChance(World par1World, int par2, int par3, int par4, int par5, float par6, int par7)
     {
         super.dropBlockAsItemWithChance(par1World, par2, par3, par4, par5, par6, 0);
-
-        if (!par1World.isRemote)
-        {
-            if (par5 >= 7)
+    }
+
+    @Override 
+    public ArrayList<ItemStack> getBlockDropped(World world, int x, int y, int z, int metadata, int fortune)
+    {
+        ArrayList<ItemStack> ret = super.getBlockDropped(world, x, y, z, metadata, fortune);
+
+        if (metadata >= 7)
+        {
+            for (int n = 0; n < 3 + fortune; n++)
             {
-                int var8 = 3 + par7;
-
-                for (int var9 = 0; var9 < var8; ++var9)
-                {
-                    if (par1World.rand.nextInt(15) <= par5)
-                    {
-                        this.dropBlockAsItem_do(par1World, par2, par3, par4, new ItemStack(this.getSeedItem(), 1, 0));
-                    }
+                if (world.rand.nextInt(15) <= metadata)
+                {
+                    ret.add(new ItemStack(this.getSeedItem(), 1, 0));
                 }
             }
         }
+
+        return ret;
     }
 
     /**
