--- ../src_base/minecraft/net/minecraft/block/BlockBasePressurePlate.java
+++ ../src_work/minecraft/net/minecraft/block/BlockBasePressurePlate.java
@@ -10,6 +10,7 @@
 import net.minecraft.util.AxisAlignedBB;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
 
 public abstract class BlockBasePressurePlate extends Block
 {
@@ -154,6 +155,18 @@
         int i1 = this.getPlateState(par1World, par2, par3, par4);
         boolean flag = par5 > 0;
         boolean flag1 = i1 > 0;
+        // CraftBukkit start - Interact Pressure Plate
+        org.bukkit.World bworld = par1World.getWorld();
+        org.bukkit.plugin.PluginManager manager = par1World.getServer().getPluginManager();
+
+        if (flag != flag1)
+        {
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bworld.getBlockAt(par2, par3, par4), par5, i1);
+            manager.callEvent(eventRedstone);
+            flag1 = eventRedstone.getNewCurrent() > 0;
+            i1 = eventRedstone.getNewCurrent();
+        }
+        // CraftBukkit end
 
         if (par5 != i1)
         {
