--- ../src-base/minecraft/net/minecraft/block/BlockFalling.java
+++ ../src-work/minecraft/net/minecraft/block/BlockFalling.java
@@ -86,6 +86,7 @@
 
     public static boolean func_180685_d(World p_180685_0_, BlockPos p_180685_1_)
     {
+        if (p_180685_0_.func_175623_d(p_180685_1_)) return true;
         Block block = p_180685_0_.func_180495_p(p_180685_1_).func_177230_c();
         Material material = block.field_149764_J;
         return block == Blocks.field_150480_ab || material == Material.field_151579_a || material == Material.field_151586_h || material == Material.field_151587_i;
