--- ../src_base/minecraft/net/minecraft/block/BlockDaylightDetector.java
+++ ../src_work/minecraft/net/minecraft/block/BlockDaylightDetector.java
@@ -90,6 +90,7 @@
 
             if (l != i1)
             {
+                i1 = org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(par1World, par2, par3, par4, l, i1).getNewCurrent(); // CraftBukkit - Call BlockRedstoneEvent
                 par1World.setBlockMetadataWithNotify(par2, par3, par4, i1, 3);
             }
         }
