--- ../src_base/minecraft/net/minecraft/block/BlockDetectorRail.java
+++ ../src_work/minecraft/net/minecraft/block/BlockDetectorRail.java
@@ -14,6 +14,7 @@
 import net.minecraft.util.Icon;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
 
 public class BlockDetectorRail extends BlockRailBase
 {
@@ -108,6 +109,15 @@
             flag1 = true;
         }
 
+        // CraftBukkit start
+        if (flag != flag1)
+        {
+            org.bukkit.block.Block block = par1World.getWorld().getBlockAt(par2, par3, par4);
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, flag ? 15 : 0, flag1 ? 15 : 0);
+            par1World.getServer().getPluginManager().callEvent(eventRedstone);
+            flag1 = eventRedstone.getNewCurrent() > 0;
+        }
+        // CraftBukkit end
         if (flag1 && !flag)
         {
             par1World.setBlockMetadataWithNotify(par2, par3, par4, par5 | 8, 3);
