--- ../src-base/minecraft/net/minecraft/block/BlockFalling.java
+++ ../src-work/minecraft/net/minecraft/block/BlockFalling.java
@@ -88,7 +88,7 @@
     {
         Block block = p_149831_0_.getBlock(p_149831_1_, p_149831_2_, p_149831_3_);
 
-        if (block.blockMaterial == Material.air)
+        if (block.isAir(p_149831_0_, p_149831_1_, p_149831_2_, p_149831_3_))
         {
             return true;
         }
@@ -98,6 +98,7 @@
         }
         else
         {
+            //TODO: King, take a look here when doing liquids!
             Material material = block.blockMaterial;
             return material == Material.water ? true : material == Material.lava;
         }
