--- ../src_base/minecraft/net/minecraft/block/BlockCocoa.java
+++ ../src_work/minecraft/net/minecraft/block/BlockCocoa.java
@@ -2,6 +2,8 @@
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
+
+import java.util.ArrayList;
 import java.util.Random;
 import net.minecraft.block.material.Material;
 import net.minecraft.entity.EntityLiving;
@@ -175,7 +177,14 @@
      */
     public void dropBlockAsItemWithChance(World par1World, int par2, int par3, int par4, int par5, float par6, int par7)
     {
-        int var8 = func_72219_c(par5);
+        super.dropBlockAsItemWithChance(par1World, par2, par3, par4, par5, par6, 0);
+    }
+
+    @Override
+    public ArrayList<ItemStack> getBlockDropped(World world, int x, int y, int z, int metadata, int fortune)
+    {
+        ArrayList<ItemStack> dropped = super.getBlockDropped(world, x, y, z, metadata, fortune);
+        int var8 = func_72219_c(metadata);
         byte var9 = 1;
 
         if (var8 >= 2)
@@ -185,8 +194,9 @@
 
         for (int var10 = 0; var10 < var9; ++var10)
         {
-            this.dropBlockAsItem_do(par1World, par2, par3, par4, new ItemStack(Item.dyePowder, 1, 3));
-        }
+            dropped.add(new ItemStack(Item.dyePowder, 1, 3));
+        }
+        return dropped;
     }
 
     @SideOnly(Side.CLIENT)
@@ -206,4 +216,10 @@
     {
         return 3;
     }
+
+    @Override
+    public int idDropped(int par1, Random par2Random, int par3)
+    {
+        return 0;
+    }
 }
