--- a/net/minecraft/block/BlockMobSpawner.java
+++ b/net/minecraft/block/BlockMobSpawner.java
@@ -13,6 +13,8 @@
 import net.minecraft.world.World;
 
 public class BlockMobSpawner extends BlockContainer {
+   private static final java.util.Random RAND = new java.util.Random();
+
    protected BlockMobSpawner(Block.Properties p_i48364_1_) {
       super(p_i48364_1_);
    }
@@ -27,10 +29,13 @@
 
    public void func_196255_a(IBlockState p_196255_1_, World p_196255_2_, BlockPos p_196255_3_, float p_196255_4_, int p_196255_5_) {
       super.func_196255_a(p_196255_1_, p_196255_2_, p_196255_3_, p_196255_4_, p_196255_5_);
-      int i = 15 + p_196255_2_.field_73012_v.nextInt(15) + p_196255_2_.field_73012_v.nextInt(15);
-      this.func_180637_b(p_196255_2_, p_196255_3_, i);
    }
 
+   @Override
+   public int getExpDrop(IBlockState state, net.minecraft.world.IWorldReader world, BlockPos pos, int fortune) {
+      return 15 + RAND.nextInt(15) + RAND.nextInt(15);
+   }
+
    public EnumBlockRenderType func_149645_b(IBlockState p_149645_1_) {
       return EnumBlockRenderType.MODEL;
    }
