--- ../src-base/minecraft/net/minecraft/block/BlockLiquid.java
+++ ../src-work/minecraft/net/minecraft/block/BlockLiquid.java
@@ -271,7 +271,7 @@
 
             for (EnumFacing enumfacing : EnumFacing.values())
             {
-                if (enumfacing != EnumFacing.DOWN && p_176365_1_.func_180495_p(p_176365_2_.func_177972_a(enumfacing)).func_185904_a() == Material.field_151586_h)
+                if (enumfacing != EnumFacing.DOWN && p_176365_1_.func_180495_p(p_176365_2_.func_177972_a(enumfacing)).func_185904_a().canMixWithLava())
                 {
                     flag = true;
                     break;
@@ -358,7 +358,7 @@
                 double d5 = d1 - 1.05D;
                 double d7 = d2 + (double)p_180655_4_.nextFloat();
 
-                if (this.field_149764_J == Material.field_151586_h)
+                if (this.field_149764_J.hasLiquidParticles())
                 {
                     p_180655_2_.func_175688_a(EnumParticleTypes.DRIP_WATER, d3, d5, d7, 0.0D, 0.0D, 0.0D);
                 }
@@ -452,4 +452,29 @@
     {
         return BlockFaceShape.UNDEFINED;
     }
+
+    @Override
+    @SideOnly (Side.CLIENT)
+    public Vec3d getFogColor(World world, BlockPos pos, IBlockState state, Entity entity, Vec3d originalColor, float partialTicks)
+    {
+        Vec3d viewport = net.minecraft.client.renderer.ActiveRenderInfo.func_178806_a(entity, partialTicks);
+
+        if (state.func_185904_a().func_76224_d())
+        {
+            float height = 0.0F;
+            if (state.func_177230_c() instanceof BlockLiquid)
+            {
+                height = func_149801_b(state.func_177229_b(field_176367_b)) - 0.11111111F;
+            }
+            float f1 = (float) (pos.func_177956_o() + 1) - height;
+            if (viewport.field_72448_b > (double)f1)
+            {
+                BlockPos upPos = pos.func_177984_a();
+                IBlockState upState = world.func_180495_p(upPos);
+                return upState.func_177230_c().getFogColor(world, upPos, upState, entity, originalColor, partialTicks);
+            }
+        }
+
+        return super.getFogColor(world, pos, state, entity, originalColor, partialTicks);
+    }
 }
