--- ../src-base/minecraft/net/minecraft/block/BlockSandStone.java
+++ ../src-work/minecraft/net/minecraft/block/BlockSandStone.java
@@ -31,7 +31,6 @@
         return ((BlockSandStone.EnumType)p_180651_1_.func_177229_b(field_176297_a)).func_176675_a();
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
     {
         for (BlockSandStone.EnumType blocksandstone$enumtype : BlockSandStone.EnumType.values())
