--- a/net/minecraft/block/Blocks.java
+++ b/net/minecraft/block/Blocks.java
@@ -25,6 +_,7 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.gen.feature.Features;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Blocks {
    public static final Block AIR = register("air", new AirBlock(AbstractBlock.Properties.of(Material.AIR).noCollission().noDrops().air()));
    public static final Block STONE = register("stone", new Block(AbstractBlock.Properties.of(Material.STONE, MaterialColor.STONE).requiresCorrectToolForDrops().strength(1.5F, 6.0F)));
@@ -119,7 +_,7 @@
    public static final Block GREEN_BED = register("green_bed", bed(DyeColor.GREEN));
    public static final Block RED_BED = register("red_bed", bed(DyeColor.RED));
    public static final Block BLACK_BED = register("black_bed", bed(DyeColor.BLACK));
-   public static final Block POWERED_RAIL = register("powered_rail", new PoweredRailBlock(AbstractBlock.Properties.of(Material.DECORATION).noCollission().strength(0.7F).sound(SoundType.METAL)));
+   public static final Block POWERED_RAIL = register("powered_rail", new PoweredRailBlock(AbstractBlock.Properties.of(Material.DECORATION).noCollission().strength(0.7F).sound(SoundType.METAL), true));
    public static final Block DETECTOR_RAIL = register("detector_rail", new DetectorRailBlock(AbstractBlock.Properties.of(Material.DECORATION).noCollission().strength(0.7F).sound(SoundType.METAL)));
    public static final Block STICKY_PISTON = register("sticky_piston", pistonBase(true));
    public static final Block COBWEB = register("cobweb", new WebBlock(AbstractBlock.Properties.of(Material.WEB).noCollission().requiresCorrectToolForDrops().strength(4.0F)));
@@ -936,16 +_,5 @@
 
    public static void rebuildCache() {
       Block.BLOCK_STATE_REGISTRY.forEach(AbstractBlock.AbstractBlockState::initCache);
-   }
-
-   static {
-      for(Block block : Registry.BLOCK) {
-         for(BlockState blockstate : block.getStateDefinition().getPossibleStates()) {
-            Block.BLOCK_STATE_REGISTRY.add(blockstate);
-         }
-
-         block.getLootTable();
-      }
-
    }
 }
