--- ../src_base/minecraft/net/minecraft/block/BlockFarmland.java
+++ ../src_work/minecraft/net/minecraft/block/BlockFarmland.java
@@ -11,6 +11,13 @@
 import net.minecraft.util.Icon;
 import net.minecraft.world.World;
 
+import net.minecraftforge.common.ForgeDirection;
+import net.minecraftforge.common.IPlantable;
+// CraftBukkit start
+import org.bukkit.event.entity.EntityInteractEvent;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
+
 public class BlockFarmland extends Block
 {
     @SideOnly(Side.CLIENT)
@@ -77,6 +84,15 @@
             }
             else if (!this.isCropsNearby(par1World, par2, par3, par4))
             {
+                // CraftBukkit start
+                org.bukkit.block.Block block = par1World.getWorld().getBlockAt(par2, par3, par4);
+
+                if (CraftEventFactory.callBlockFadeEvent(block, Block.dirt.blockID).isCancelled())
+                {
+                    return;
+                }
+
+                // CraftBukkit end
                 par1World.setBlock(par2, par3, par4, Block.dirt.blockID);
             }
         }
@@ -98,6 +114,24 @@
                 return;
             }
 
+            // CraftBukkit start - Interact soil
+            org.bukkit.event.Cancellable cancellable;
+
+            if (par5Entity instanceof EntityPlayer)
+            {
+                cancellable = CraftEventFactory.callPlayerInteractEvent((EntityPlayer) par5Entity, org.bukkit.event.block.Action.PHYSICAL, par2, par3, par4, -1, null);
+            }
+            else
+            {
+                cancellable = new EntityInteractEvent(par5Entity.getBukkitEntity(), par1World.getWorld().getBlockAt(par2, par3, par4));
+                par1World.getServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+            }
+
+            if (cancellable.isCancelled())
+            {
+                return;
+            }
+            // CraftBukkit end
             par1World.setBlock(par2, par3, par4, Block.dirt.blockID);
         }
     }
@@ -115,7 +149,8 @@
             {
                 int j1 = par1World.getBlockId(l, par3 + 1, i1);
 
-                if (j1 == Block.crops.blockID || j1 == Block.melonStem.blockID || j1 == Block.pumpkinStem.blockID || j1 == Block.potato.blockID || j1 == Block.carrot.blockID)
+                Block plant = blocksList[j1];
+                if (plant instanceof IPlantable && canSustainPlant(par1World, par2, par3, par4, ForgeDirection.UP, (IPlantable)plant))
                 {
                     return true;
                 }
@@ -188,7 +223,7 @@
      */
     public void registerIcons(IconRegister par1IconRegister)
     {
-        this.field_94441_a = par1IconRegister.registerIcon(this.func_111023_E() + "_wet");
-        this.field_94440_b = par1IconRegister.registerIcon(this.func_111023_E() + "_dry");
+        this.field_94441_a = par1IconRegister.registerIcon(this.getTextureName() + "_wet");
+        this.field_94440_b = par1IconRegister.registerIcon(this.getTextureName() + "_dry");
     }
 }
