--- ../src-base/minecraft/net/minecraft/block/BlockStoneSlabNew.java
+++ ../src-work/minecraft/net/minecraft/block/BlockStoneSlabNew.java
@@ -74,7 +74,6 @@
         return BlockStoneSlabNew.EnumType.func_176916_a(p_185674_1_.func_77960_j() & 7);
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
     {
         for (BlockStoneSlabNew.EnumType blockstoneslabnew$enumtype : BlockStoneSlabNew.EnumType.values())
