--- a/net/minecraft/block/BubbleColumnBlock.java
+++ b/net/minecraft/block/BubbleColumnBlock.java
@@ -80,9 +80,11 @@
       Block block = blockstate.func_177230_c();
       if (block == Blocks.field_203203_C) {
          return blockstate.func_177229_b(field_203160_a);
-      } else {
-         return block != Blocks.field_150425_aM;
       }
+      if (block instanceof IBubbleSource) {
+         return ((IBubbleSource) block).isDragging();
+      }
+      return true;
    }
 
    public int func_149738_a(IWorldReader p_149738_1_) {
@@ -126,7 +128,7 @@
 
    public boolean func_196260_a(BlockState p_196260_1_, IWorldReader p_196260_2_, BlockPos p_196260_3_) {
       Block block = p_196260_2_.func_180495_p(p_196260_3_.func_177977_b()).func_177230_c();
-      return block == Blocks.field_203203_C || block == Blocks.field_196814_hQ || block == Blocks.field_150425_aM;
+      return block == Blocks.field_203203_C || block instanceof IBubbleSource;
    }
 
    public VoxelShape func_220053_a(BlockState p_220053_1_, IBlockReader p_220053_2_, BlockPos p_220053_3_, ISelectionContext p_220053_4_) {
