--- a/net/minecraft/block/BushBlock.java
+++ b/net/minecraft/block/BushBlock.java
@@ -7,13 +_,15 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReader;
 
-public class BushBlock extends Block {
+public class BushBlock extends Block implements net.minecraftforge.common.IPlantable {
    public BushBlock(AbstractBlock.Properties p_i48437_1_) {
       super(p_i48437_1_);
    }
 
+   @Deprecated
+   // FORGE: Use IPlantable.getPlantType instead
    protected boolean func_200014_a_(BlockState p_200014_1_, IBlockReader p_200014_2_, BlockPos p_200014_3_) {
-      return p_200014_1_.func_203425_a(Blocks.field_196658_i) || p_200014_1_.func_203425_a(Blocks.field_150346_d) || p_200014_1_.func_203425_a(Blocks.field_196660_k) || p_200014_1_.func_203425_a(Blocks.field_196661_l) || p_200014_1_.func_203425_a(Blocks.field_150458_ak);
+      return getPlantType(p_200014_2_, p_200014_3_).test(p_200014_2_, p_200014_3_, p_200014_1_, Direction.UP);
    }
 
    public BlockState func_196271_a(BlockState p_196271_1_, Direction p_196271_2_, BlockState p_196271_3_, IWorld p_196271_4_, BlockPos p_196271_5_, BlockPos p_196271_6_) {
@@ -22,6 +_,8 @@
 
    public boolean func_196260_a(BlockState p_196260_1_, IWorldReader p_196260_2_, BlockPos p_196260_3_) {
       BlockPos blockpos = p_196260_3_.func_177977_b();
+      if (p_196260_1_.func_177230_c() == this) //Forge: This function is called during world gen and placement, before this block is set, so if we are not 'here' then assume it's the pre-check.
+         return p_196260_2_.func_180495_p(blockpos).canSustainPlant(p_196260_2_, blockpos, Direction.UP, this);
       return this.func_200014_a_(p_196260_2_.func_180495_p(blockpos), p_196260_2_, blockpos);
    }
 
@@ -31,5 +_,18 @@
 
    public boolean func_196266_a(BlockState p_196266_1_, IBlockReader p_196266_2_, BlockPos p_196266_3_, PathType p_196266_4_) {
       return p_196266_4_ == PathType.AIR && !this.field_235688_at_ ? true : super.func_196266_a(p_196266_1_, p_196266_2_, p_196266_3_, p_196266_4_);
+   }
+
+   @Override
+   public BlockState getPlant(IBlockReader world, BlockPos pos) {
+      BlockState state = world.func_180495_p(pos);
+      if (state.func_177230_c() != this) return func_176223_P();
+      return state;
+   }
+
+   @Override
+   public net.minecraftforge.common.PlantType getPlantType(IBlockReader world, BlockPos pos)
+   {
+      return net.minecraftforge.common.PlantType.PLAINS;
    }
 }
