--- a/net/minecraft/block/ChestBlock.java
+++ b/net/minecraft/block/ChestBlock.java
@@ -325,7 +_,8 @@
    }
 
    public BlockState func_185471_a(BlockState p_185471_1_, Mirror p_185471_2_) {
-      return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176459_a)));
+      BlockState rotated = p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176459_a)));
+      return p_185471_2_ == Mirror.NONE ? rotated : rotated.func_206870_a(field_196314_b, rotated.func_177229_b(field_196314_b).func_208081_a());  // Forge: Fixed MC-134110 Structure mirroring breaking apart double chests
    }
 
    protected void func_206840_a(StateContainer.Builder<Block, BlockState> p_206840_1_) {
