--- a/net/minecraft/block/FlowerPotBlock.java
+++ b/net/minecraft/block/FlowerPotBlock.java
@@ -19,7 +19,7 @@
 import net.minecraft.world.World;
 
 public class FlowerPotBlock extends Block {
-   private static final Map<Block, Block> field_196451_b = Maps.newHashMap();
+   private static final Map<Block, Block> field_196451_b = Maps.newHashMap();  //TODO: Delegates
    protected static final VoxelShape field_196450_a = Block.func_208617_a(5.0D, 0.0D, 5.0D, 11.0D, 6.0D, 11.0D);
    private final Block field_196452_c;
 
