--- a/net/minecraft/block/MagmaBlock.java
+++ b/net/minecraft/block/MagmaBlock.java
@@ -72,7 +72,7 @@
    }
 
    public boolean func_220067_a(BlockState p_220067_1_, IBlockReader p_220067_2_, BlockPos p_220067_3_, EntityType<?> p_220067_4_) {
-      return p_220067_4_.func_220338_c();
+      return p_220067_4_ != null && p_220067_4_.func_220338_c();
    }
 
    public boolean func_201783_b(BlockState p_201783_1_, IBlockReader p_201783_2_, BlockPos p_201783_3_) {
