--- a/net/minecraft/block/MushroomBlock.java
+++ b/net/minecraft/block/MushroomBlock.java
@@ -62,7 +62,7 @@
       BlockState blockstate = p_196260_2_.func_180495_p(blockpos);
       Block block = blockstate.func_177230_c();
       if (block != Blocks.field_150391_bh && block != Blocks.field_196661_l) {
-         return p_196260_2_.func_201669_a(p_196260_3_, 0) < 13 && this.func_200014_a_(blockstate, p_196260_2_, blockpos);
+         return p_196260_2_.func_201669_a(p_196260_3_, 0) < 13 && blockstate.canSustainPlant(p_196260_2_, blockpos, net.minecraft.util.Direction.UP, this);
       } else {
          return true;
       }
