--- a/net/minecraft/block/NetherWartBlock.java
+++ b/net/minecraft/block/NetherWartBlock.java
@@ -26,7 +_,7 @@
    }
 
    protected boolean func_200014_a_(BlockState p_200014_1_, IBlockReader p_200014_2_, BlockPos p_200014_3_) {
-      return p_200014_1_.func_203425_a(Blocks.field_150425_aM);
+      return p_200014_1_.func_235714_a_(net.minecraftforge.common.Tags.Blocks.GROWABLE_NETHER_WART);
    }
 
    public boolean func_149653_t(BlockState p_149653_1_) {
@@ -35,9 +_,10 @@
 
    public void func_225542_b_(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
       int i = p_225542_1_.func_177229_b(field_176486_a);
-      if (i < 3 && p_225542_4_.nextInt(10) == 0) {
+      if (i < 3 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_225542_2_, p_225542_3_, p_225542_1_, p_225542_4_.nextInt(10) == 0)) {
          p_225542_1_ = p_225542_1_.func_206870_a(field_176486_a, Integer.valueOf(i + 1));
          p_225542_2_.func_180501_a(p_225542_3_, p_225542_1_, 2);
+         net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_225542_2_, p_225542_3_, p_225542_1_);
       }
 
    }
