--- a/net/minecraft/block/NetherPortalBlock.java
+++ b/net/minecraft/block/NetherPortalBlock.java
@@ -70,7 +70,7 @@
 
    public static boolean func_176548_d(IWorld p_176548_0_, BlockPos p_176548_1_) {
       NetherPortalBlock.Size netherportalblock$size = func_201816_b(p_176548_0_, p_176548_1_);
-      if (netherportalblock$size != null) {
+      if (netherportalblock$size != null && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_176548_0_, p_176548_1_, netherportalblock$size)) {
          netherportalblock$size.func_150859_c();
          return true;
       } else {
