--- a/net/minecraft/block/RedstoneWireBlock.java
+++ b/net/minecraft/block/RedstoneWireBlock.java
@@ -184,7 +_,7 @@
       BlockState blockstate = p_235545_1_.getBlockState(blockpos);
       if (p_235545_4_) {
          boolean flag = this.canSurviveOn(p_235545_1_, blockpos, blockstate);
-         if (flag && shouldConnectTo(p_235545_1_.getBlockState(blockpos.above()))) {
+         if (flag && canConnectTo(p_235545_1_.getBlockState(blockpos.above()), p_235545_1_, blockpos.above(), null) ) {
             if (blockstate.isFaceSturdy(p_235545_1_, blockpos, p_235545_3_.getOpposite())) {
                return RedstoneSide.UP;
             }
@@ -193,7 +_,7 @@
          }
       }
 
-      return !shouldConnectTo(blockstate, p_235545_3_) && (blockstate.isRedstoneConductor(p_235545_1_, blockpos) || !shouldConnectTo(p_235545_1_.getBlockState(blockpos.below()))) ? RedstoneSide.NONE : RedstoneSide.SIDE;
+      return !canConnectTo(blockstate, p_235545_1_, blockpos, p_235545_3_) && (blockstate.isRedstoneConductor(p_235545_1_, blockpos) || !canConnectTo(p_235545_1_.getBlockState(blockpos.below()), p_235545_1_, blockpos.below(), null)) ? RedstoneSide.NONE : RedstoneSide.SIDE;
    }
 
    public boolean canSurvive(BlockState p_196260_1_, IWorldReader p_196260_2_, BlockPos p_196260_3_) {
@@ -335,11 +_,7 @@
       }
    }
 
-   protected static boolean shouldConnectTo(BlockState p_176346_0_) {
-      return shouldConnectTo(p_176346_0_, (Direction)null);
-   }
-
-   protected static boolean shouldConnectTo(BlockState p_176343_0_, @Nullable Direction p_176343_1_) {
+   protected static boolean canConnectTo(BlockState p_176343_0_, IBlockReader world, BlockPos pos, @Nullable Direction p_176343_1_) {
       if (p_176343_0_.is(Blocks.REDSTONE_WIRE)) {
          return true;
       } else if (p_176343_0_.is(Blocks.REPEATER)) {
@@ -348,7 +_,7 @@
       } else if (p_176343_0_.is(Blocks.OBSERVER)) {
          return p_176343_1_ == p_176343_0_.getValue(ObserverBlock.FACING);
       } else {
-         return p_176343_0_.isSignalSource() && p_176343_1_ != null;
+         return p_176343_0_.canConnectRedstone(world, pos, p_176343_1_) && p_176343_1_ != null;
       }
    }
 
