--- a/net/minecraft/block/SeaGrassBlock.java
+++ b/net/minecraft/block/SeaGrassBlock.java
@@ -17,7 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
 
-public class SeaGrassBlock extends BushBlock implements IGrowable, ILiquidContainer {
+public class SeaGrassBlock extends BushBlock implements IGrowable, ILiquidContainer, net.minecraftforge.common.IForgeShearable {
    protected static final VoxelShape field_207798_a = Block.func_208617_a(2.0D, 0.0D, 2.0D, 14.0D, 12.0D, 14.0D);
 
    public SeaGrassBlock(AbstractBlock.Properties p_i48780_1_) {
@@ -28,8 +_,8 @@
       return field_207798_a;
    }
 
-   protected boolean func_200014_a_(BlockState p_200014_1_, IBlockReader p_200014_2_, BlockPos p_200014_3_) {
-      return p_200014_1_.func_224755_d(p_200014_2_, p_200014_3_, Direction.UP) && !p_200014_1_.func_203425_a(Blocks.field_196814_hQ);
+   public net.minecraftforge.common.PlantType getPlantType(IBlockReader world, BlockPos pos) {
+      return net.minecraftforge.common.PlantType.SEA_GRASS;
    }
 
    @Nullable
