--- a/net/minecraft/block/SoulSandBlock.java
+++ b/net/minecraft/block/SoulSandBlock.java
@@ -10,8 +10,9 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
+import net.minecraftforge.block.IBubbleSource;
 
-public class SoulSandBlock extends Block {
+public class SoulSandBlock extends Block implements IBubbleSource {
    protected static final VoxelShape field_196509_a = Block.func_208617_a(0.0D, 0.0D, 0.0D, 16.0D, 14.0D, 16.0D);
 
    public SoulSandBlock(Block.Properties p_i48326_1_) {
