--- a/net/minecraft/block/SpawnerBlock.java
+++ b/net/minecraft/block/SpawnerBlock.java
@@ -19,10 +19,13 @@
 
    public void func_220062_a(BlockState p_220062_1_, World p_220062_2_, BlockPos p_220062_3_, ItemStack p_220062_4_) {
       super.func_220062_a(p_220062_1_, p_220062_2_, p_220062_3_, p_220062_4_);
-      int i = 15 + p_220062_2_.field_73012_v.nextInt(15) + p_220062_2_.field_73012_v.nextInt(15);
-      this.func_180637_b(p_220062_2_, p_220062_3_, i);
    }
 
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.IWorldReader world, BlockPos pos, int fortune, int silktouch) {
+      return 15 + RANDOM.nextInt(15) + RANDOM.nextInt(15);
+   }
+
    public BlockRenderType func_149645_b(BlockState p_149645_1_) {
       return BlockRenderType.MODEL;
    }
