--- a/net/minecraft/block/SpreadableSnowyDirtBlock.java
+++ b/net/minecraft/block/SpreadableSnowyDirtBlock.java
@@ -33,6 +_,7 @@
 
    public void func_225542_b_(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
       if (!func_220257_b(p_225542_1_, p_225542_2_, p_225542_3_)) {
+         if (!p_225542_2_.isAreaLoaded(p_225542_3_, 3)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light and spreading
          p_225542_2_.func_175656_a(p_225542_3_, Blocks.field_150346_d.func_176223_P());
       } else {
          if (p_225542_2_.func_201696_r(p_225542_3_.func_177984_a()) >= 9) {
