--- a/net/minecraft/block/SpreadableSnowyDirtBlock.java
+++ b/net/minecraft/block/SpreadableSnowyDirtBlock.java
@@ -8,7 +8,7 @@
 import net.minecraft.world.World;
 import net.minecraft.world.lighting.LightEngine;
 
-public abstract class SpreadableSnowyDirtBlock extends SnowyDirtBlock {
+public abstract class SpreadableSnowyDirtBlock extends SnowyDirtBlock implements net.minecraftforge.common.extensions.IForgeSpreadable {
    protected SpreadableSnowyDirtBlock(Block.Properties p_i48324_1_) {
       super(p_i48324_1_);
    }
@@ -31,16 +31,18 @@
 
    public void func_196267_b(BlockState p_196267_1_, World p_196267_2_, BlockPos p_196267_3_, Random p_196267_4_) {
       if (!p_196267_2_.field_72995_K) {
+         if (!p_196267_2_.isAreaLoaded(p_196267_3_, 3)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light and spreading
          if (!func_220257_b(p_196267_1_, p_196267_2_, p_196267_3_)) {
-            p_196267_2_.func_175656_a(p_196267_3_, Blocks.field_150346_d.func_176223_P());
+            p_196267_2_.func_175656_a(p_196267_3_, degeneratesInto().func_176223_P());
          } else {
             if (p_196267_2_.func_201696_r(p_196267_3_.func_177984_a()) >= 9) {
                BlockState blockstate = this.func_176223_P();
 
                for(int i = 0; i < 4; ++i) {
                   BlockPos blockpos = p_196267_3_.func_177982_a(p_196267_4_.nextInt(3) - 1, p_196267_4_.nextInt(5) - 3, p_196267_4_.nextInt(3) - 1);
-                  if (p_196267_2_.func_180495_p(blockpos).func_177230_c() == Blocks.field_150346_d && func_220256_c(blockstate, p_196267_2_, blockpos)) {
+                  if (p_196267_2_.func_180495_p(blockpos).func_177230_c().canSustainSpreadableOfType(this.getBlock()) && func_220256_c(blockstate, p_196267_2_, blockpos) && net.minecraftforge.common.ForgeHooks.onSpreadableSpreadPre(p_196267_2_, blockpos, p_196267_2_.func_180495_p(blockpos), blockstate)) {
                      p_196267_2_.func_175656_a(blockpos, blockstate.func_206870_a(field_196382_a, Boolean.valueOf(p_196267_2_.func_180495_p(blockpos.func_177984_a()).func_177230_c() == Blocks.field_150433_aE)));
+                     net.minecraftforge.common.ForgeHooks.onSpreadableSpreadPost(p_196267_2_, blockpos, p_196267_2_.func_180495_p(blockpos));
                   }
                }
             }
