--- a/net/minecraft/block/SweetBerryBushBlock.java
+++ b/net/minecraft/block/SweetBerryBushBlock.java
@@ -76,7 +76,7 @@
          return ActionResultType.PASS;
       } else if (i > 1) {
          int j = 1 + p_225533_2_.field_73012_v.nextInt(2);
-         func_180635_a(p_225533_2_, p_225533_3_, new ItemStack(Items.field_222112_pR, j + (flag ? 1 : 0)));
+         net.minecraftforge.event.ForgeEventFactory.fireBlockDropLootEvent(p_225533_2_, p_225533_3_, p_225533_1_, p_225533_4_, new ItemStack(Items.field_222112_pR, j + (flag ? 1 : 0)));
          p_225533_2_.func_184133_a((PlayerEntity)null, p_225533_3_, SoundEvents.field_219693_lB, SoundCategory.BLOCKS, 1.0F, 0.8F + p_225533_2_.field_73012_v.nextFloat() * 0.4F);
          p_225533_2_.func_180501_a(p_225533_3_, p_225533_1_.func_206870_a(field_220125_a, Integer.valueOf(1)), 2);
          return ActionResultType.SUCCESS;
