--- a/net/minecraft/block/VineBlock.java
+++ b/net/minecraft/block/VineBlock.java
@@ -22,7 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.server.ServerWorld;
 
-public class VineBlock extends Block {
+public class VineBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public static final BooleanProperty field_176277_a = SixWayBlock.field_196496_z;
    public static final BooleanProperty field_176273_b = SixWayBlock.field_196488_a;
    public static final BooleanProperty field_176278_M = SixWayBlock.field_196490_b;
@@ -152,7 +_,7 @@
    }
 
    public void func_225542_b_(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
-      if (p_225542_2_.field_73012_v.nextInt(4) == 0) {
+      if (p_225542_2_.field_73012_v.nextInt(4) == 0 && p_225542_2_.isAreaLoaded(p_225542_3_, 4)) { // Forge: check area to prevent loading unloaded chunks
          Direction direction = Direction.func_239631_a_(p_225542_4_);
          BlockPos blockpos = p_225542_3_.func_177984_a();
          if (direction.func_176740_k().func_176722_c() && !p_225542_1_.func_177229_b(func_176267_a(direction))) {
