--- a/net/minecraft/block/WebBlock.java
+++ b/net/minecraft/block/WebBlock.java
@@ -5,7 +_,7 @@
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraft.world.World;
 
-public class WebBlock extends Block {
+public class WebBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public WebBlock(AbstractBlock.Properties p_i48296_1_) {
       super(p_i48296_1_);
    }
