--- a/net/minecraft/block/WoodType.java
+++ b/net/minecraft/block/WoodType.java
@@ -32,4 +_,11 @@
    public String func_227048_b_() {
       return this.field_227045_h_;
    }
+
+   /**
+    * Use this to create a new {@link WoodType}. Make sure to register its rendering by enqueuing Atlases.addWoodType(...) during client setup..
+    */
+   public static WoodType create(String name) {
+      return new WoodType(name);
+   }
 }
