--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/ModelBlockDefinition.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/ModelBlockDefinition.java
@@ -31,7 +31,7 @@
 
     public static ModelBlockDefinition parseFromReader(Reader p_178331_0_)
     {
-        return (ModelBlockDefinition)GSON.fromJson(p_178331_0_, ModelBlockDefinition.class);
+        return net.minecraftforge.client.model.BlockStateLoader.load(p_178331_0_, GSON);
     }
 
     public ModelBlockDefinition(Map<String, VariantList> p_i46572_1_, Multipart p_i46572_2_)
