--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -245,17 +_,21 @@
     }
 
     protected void convertToZombieType(final ServerLevel level, final EntityType<? extends Zombie> zombieType) {
+        if (!net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, zombieType, timer -> this.conversionTime = timer)) return;
         this.convertTo(
             zombieType,
             ConversionParams.single(this, true, true),
-            newZombie -> newZombie.handleAttributes(
-                level.getCurrentDifficultyAt(newZombie.blockPosition()).getSpecialMultiplier(), EntitySpawnReason.CONVERSION
-            )
+            newZombie -> {
+                newZombie.handleAttributes(level.getCurrentDifficultyAt(newZombie.blockPosition()).getSpecialMultiplier(), EntitySpawnReason.CONVERSION);
+                net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, newZombie);
+            }
         );
     }
 
     @VisibleForTesting
     public boolean convertVillagerToZombieVillager(final ServerLevel level, final Villager villager) {
+        if (!net.minecraftforge.event.ForgeEventFactory.canLivingConvert(villager, EntityTypes.ZOMBIE_VILLAGER, (timer) -> {}))
+            return false;
         ZombieVillager zombieVillager = villager.convertTo(
             EntityTypes.ZOMBIE_VILLAGER,
             ConversionParams.single(villager, true, true),
@@ -268,6 +_,7 @@
                 zombie.setGossips(villager.getGossips().copy());
                 zombie.setTradeOffers(villager.getOffers().copy());
                 zombie.setVillagerXp(villager.getVillagerXp());
+                net.minecraftforge.event.ForgeEventFactory.onLivingConvert(villager, zombie);
                 if (!this.isSilent()) {
                     level.levelEvent(null, 1026, this.blockPosition(), 0);
                 }
@@ -291,15 +_,26 @@
             target = (LivingEntity)source.getEntity();
         }
 
-        if (target != null
+        var vanilla = (target != null
             && level.getDifficulty() == Difficulty.HARD
             && this.random.nextFloat() < this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE)
-            && level.isSpawningMonsters()) {
+            && level.isSpawningMonsters());
+        {
             int x = Mth.floor(this.getX());
             int y = Mth.floor(this.getY());
             int z = Mth.floor(this.getZ());
             EntityType<? extends Zombie> type = this.getType();
-            Zombie reinforcement = type.create(level, EntitySpawnReason.REINFORCEMENT);
+
+            var event = net.minecraftforge.event.ForgeEventFactory.fireZombieSummonAid(this, level(), x, y, z, target, this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE));
+
+            Zombie reinforcement = null;
+            if (event.getResult().isAllowed() || (vanilla && event.getResult().isDefault())) {
+                if (event.getCustomSummonedAid() != null)
+                    reinforcement = event.getCustomSummonedAid();
+                else
+                    reinforcement = type.create(this.level(), EntitySpawnReason.REINFORCEMENT);
+            }
+
             if (reinforcement == null) {
                 return true;
             }
