--- ../src-base/minecraft/net/minecraft/client/renderer/block/statemap/StateMap.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/statemap/StateMap.java
@@ -38,7 +38,7 @@
         }
         else
         {
-            s = this.func_187490_a(this.name, map);
+            s = String.format("%s:%s", Block.blockRegistry.getNameForObject(state.getBlock()).getResourceDomain(), this.func_187490_a(this.name, map));
         }
 
         if (this.suffix != null)
