--- a/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
+++ b/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
@@ -3,8 +_,14 @@
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.resources.ResourceLocation;
 
-public record DiscardedPayload(ResourceLocation f_291401_) implements CustomPacketPayload {
+public record DiscardedPayload(ResourceLocation f_291401_, @org.jetbrains.annotations.Nullable FriendlyByteBuf data) implements CustomPacketPayload {
+   public DiscardedPayload(ResourceLocation f_291401_) {
+      this(f_291401_, null);
+   }
+
    public void m_293110_(FriendlyByteBuf p_301050_) {
+       if (data != null)
+          p_301050_.writeBytes(data.slice());
    }
 
    public ResourceLocation m_292644_() {
