--- a/net/minecraft/network/syncher/EntityDataSerializers.java
+++ b/net/minecraft/network/syncher/EntityDataSerializers.java
@@ -142,16 +_,16 @@
    public static final EntityDataSerializer<Quaternionf> QUATERNION = EntityDataSerializer.simple(FriendlyByteBuf::writeQuaternion, FriendlyByteBuf::readQuaternion);
 
    public static void registerSerializer(EntityDataSerializer<?> p_135051_) {
-      SERIALIZERS.add(p_135051_);
+      if (SERIALIZERS.add(p_135051_) >= 256) throw new RuntimeException("Vanilla DataSerializer ID limit exceeded");
    }
 
    @Nullable
    public static EntityDataSerializer<?> getSerializer(int p_135049_) {
-      return SERIALIZERS.byId(p_135049_);
+      return net.minecraftforge.common.ForgeHooks.getSerializer(p_135049_, SERIALIZERS);
    }
 
    public static int getSerializedId(EntityDataSerializer<?> p_135053_) {
-      return SERIALIZERS.getId(p_135053_);
+      return net.minecraftforge.common.ForgeHooks.getSerializerId(p_135053_, SERIALIZERS);
    }
 
    private EntityDataSerializers() {
