--- a/net/minecraft/world/item/alchemy/PotionBrewing.java
+++ b/net/minecraft/world/item/alchemy/PotionBrewing.java
@@ -19,17 +_,37 @@
     private final List<Ingredient> containers;
     private final List<PotionBrewing.Mix<Potion>> potionMixes;
     private final List<PotionBrewing.Mix<Item>> containerMixes;
+    private final List<net.minecraftforge.common.brewing.IBrewingRecipe> recipes;
 
     private PotionBrewing(
         final List<Ingredient> containers, final List<PotionBrewing.Mix<Potion>> potionMixes, final List<PotionBrewing.Mix<Item>> containerMixes
     ) {
+        this(containers, potionMixes, containerMixes, null);
+    }
+
+    PotionBrewing(final List<Ingredient> containers, final List<PotionBrewing.Mix<Potion>> potionMixes, final List<PotionBrewing.Mix<Item>> containerMixes, Builder builder) {
         this.containers = containers;
         this.potionMixes = potionMixes;
         this.containerMixes = containerMixes;
+        var tmp = new ArrayList<net.minecraftforge.common.brewing.IBrewingRecipe>();
+        tmp.add(new net.minecraftforge.common.brewing.VanillaBrewingRecipe(this, this::mixVanilla));
+        if (builder != null) {
+            tmp.addAll(builder.recipes);
+        }
+        this.recipes = java.util.Collections.unmodifiableList(tmp);
     }
 
     public boolean isIngredient(final ItemStack ingredient) {
-        return this.isContainerIngredient(ingredient) || this.isPotionIngredient(ingredient);
+        if (ingredient.isEmpty()) {
+            return false;
+        }
+
+        for (var recipe : recipes) {
+            if (recipe.isIngredient(ingredient)) {
+                return true;
+            }
+        }
+        return false;
     }
 
     private boolean isContainer(final ItemStack input) {
@@ -73,9 +_,10 @@
     }
 
     public boolean hasMix(final ItemStack source, final ItemStack ingredient) {
-        return !this.isContainer(source) ? false : this.hasContainerMix(source, ingredient) || this.hasPotionMix(source, ingredient);
+        return !mix(ingredient, source).isEmpty();
     }
 
+    /** @deprecated Forge: use hasMix(ItemStack, ItemStack)*/
     public boolean hasContainerMix(final ItemStack source, final ItemStack ingredient) {
         for (PotionBrewing.Mix<Item> mix : this.containerMixes) {
             if (source.is(mix.from) && mix.ingredient.test(ingredient)) {
@@ -86,6 +_,7 @@
         return false;
     }
 
+    /** @deprecated Forge: use hasMix(ItemStack, ItemStack)*/
     public boolean hasPotionMix(final ItemStack source, final ItemStack ingredient) {
         Optional<Holder<Potion>> potion = source.getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY).potion();
         if (potion.isEmpty()) {
@@ -102,6 +_,19 @@
     }
 
     public ItemStack mix(final ItemStack ingredient, final ItemStack source) {
+        if (source.isEmpty() || source.getCount() != 1) return ItemStack.EMPTY;
+        if (ingredient.isEmpty()) return ItemStack.EMPTY;
+
+        for (var recipe : recipes) {
+            ItemStack output = recipe.getOutput(source, ingredient);
+            if (!output.isEmpty()) {
+                return output;
+            }
+        }
+        return ItemStack.EMPTY;
+    }
+
+    private ItemStack mixVanilla(final ItemStack ingredient, final ItemStack source) {
         if (source.isEmpty()) {
             return source;
         }
@@ -126,9 +_,29 @@
         return source;
     }
 
+    /**
+     * Returns true if the passed ItemStack is a valid input for the start of a recipe
+     */
+    public boolean isValidInput(ItemStack stack) {
+        for (var recipe : recipes) {
+            if (recipe.isInput(stack)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns an unmodifiable list containing all the recipes in the registry
+     */
+    public List<net.minecraftforge.common.brewing.IBrewingRecipe> getRecipes() {
+        return this.recipes;
+    }
+
     public static PotionBrewing bootstrap(final FeatureFlagSet enabledFeatures) {
         PotionBrewing.Builder builder = new PotionBrewing.Builder(enabledFeatures);
         addVanillaMixes(builder);
+        net.minecraftforge.event.ForgeEventFactory.onBrewingRecipeRegister(builder, enabledFeatures);
         return builder.build();
     }
 
@@ -197,6 +_,7 @@
         private final List<PotionBrewing.Mix<Potion>> potionMixes = new ArrayList<>();
         private final List<PotionBrewing.Mix<Item>> containerMixes = new ArrayList<>();
         private final FeatureFlagSet enabledFeatures;
+        private final List<net.minecraftforge.common.brewing.IBrewingRecipe> recipes = new ArrayList<>();
 
         public Builder(final FeatureFlagSet enabledFeatures) {
             this.enabledFeatures = enabledFeatures;
@@ -236,8 +_,13 @@
             }
         }
 
+        public Builder add(net.minecraftforge.common.brewing.IBrewingRecipe recipe) {
+            this.recipes.add(recipe);
+            return this;
+        }
+
         public PotionBrewing build() {
-            return new PotionBrewing(List.copyOf(this.containers), List.copyOf(this.potionMixes), List.copyOf(this.containerMixes));
+            return new PotionBrewing(List.copyOf(this.containers), List.copyOf(this.potionMixes), List.copyOf(this.containerMixes), this);
         }
     }
 
