--- ../src-base/minecraft/net/minecraft/nbt/CompressedStreamTools.java
+++ ../src-work/minecraft/net/minecraft/nbt/CompressedStreamTools.java
@@ -1,7 +1,5 @@
 package net.minecraft.nbt;
 
-import cpw.mods.fml.relauncher.Side;
-import cpw.mods.fml.relauncher.SideOnly;
 import java.io.BufferedInputStream;
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
@@ -90,7 +88,6 @@
         return bytearrayoutputstream.toByteArray();
     }
 
-    @SideOnly(Side.CLIENT)
     public static void func_74793_a(NBTTagCompound p_74793_0_, File p_74793_1_) throws IOException
     {
         File file2 = new File(p_74793_1_.getAbsolutePath() + "_tmp");
@@ -176,7 +173,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     public static void func_74795_b(NBTTagCompound p_74795_0_, File p_74795_1_) throws IOException
     {
         DataOutputStream dataoutputstream = new DataOutputStream(new FileOutputStream(p_74795_1_));
@@ -191,7 +187,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     public static NBTTagCompound func_74797_a(File p_74797_0_) throws IOException
     {
         if (!p_74797_0_.exists())
