--- a/net/minecraft/nbt/CompressedStreamTools.java
+++ b/net/minecraft/nbt/CompressedStreamTools.java
@@ -106,10 +106,12 @@
 
    private static INBT func_152455_a(DataInput p_152455_0_, int p_152455_1_, NBTSizeTracker p_152455_2_) throws IOException {
       byte b0 = p_152455_0_.readByte();
+      p_152455_2_.func_152450_a(8); // Forge: Count everything!
       if (b0 == 0) {
          return EndNBT.field_229686_b_;
       } else {
-         p_152455_0_.readUTF();
+         p_152455_2_.readUTF(p_152455_0_.readUTF()); // Forge: Count this string.
+         p_152455_2_.func_152450_a(32); // Forge: 4 extra bytes for the object allocation.
 
          try {
             return NBTTypes.func_229710_a_(b0).func_225649_b_(p_152455_0_, p_152455_1_, p_152455_2_);
