--- a/net/minecraft/nbt/CompoundTag.java
+++ b/net/minecraft/nbt/CompoundTag.java
@@ -41,6 +_,7 @@
             while((b0 = CompoundTag.m_128420_(p_128485_, p_128487_)) != 0) {
                String s = CompoundTag.m_128432_(p_128485_, p_128487_);
                p_128487_.m_128926_((long)(28 + 2 * s.length()));
+               p_128487_.m_128926_(4); //Forge: 4 extra bytes for the object allocation.
                Tag tag = CompoundTag.m_128413_(TagTypes.m_129397_(b0), s, p_128485_, p_128486_ + 1, p_128487_);
                if (map.put(s, tag) == null) {
                   p_128487_.m_128926_(36L);
@@ -167,6 +_,7 @@
 
    @Nullable
    public Tag m_128365_(String p_128366_, Tag p_128367_) {
+      if (p_128367_ == null) throw new IllegalArgumentException("Invalid null NBT value with key " + p_128366_);
       return this.f_128329_.put(p_128366_, p_128367_);
    }
 
@@ -459,11 +_,12 @@
    }
 
    static byte m_128420_(DataInput p_128421_, NbtAccounter p_128422_) throws IOException {
+      p_128422_.m_128926_(2);
       return p_128421_.readByte();
    }
 
    static String m_128432_(DataInput p_128433_, NbtAccounter p_128434_) throws IOException {
-      return p_128433_.readUTF();
+      return p_128434_.readUTF(p_128433_.readUTF());
    }
 
    static Tag m_128413_(TagType<?> p_128414_, String p_128415_, DataInput p_128416_, int p_128417_, NbtAccounter p_128418_) {
