--- a/net/minecraft/nbt/CompoundTag.java
+++ b/net/minecraft/nbt/CompoundTag.java
@@ -20,6 +_,10 @@
 import net.minecraft.ReportedException;
 
 public class CompoundTag implements Tag {
+    public static net.minecraftforge.common.util.INBTBuilder.Builder builder() {
+        return (new net.minecraftforge.common.util.INBTBuilder(){}).nbt();
+    }
+
     public static final Codec<CompoundTag> CODEC = Codec.PASSTHROUGH
         .comapFlatMap(
             p_308555_ -> {
