--- a/net/minecraft/nbt/NbtIo.java
+++ b/net/minecraft/nbt/NbtIo.java
@@ -154,10 +_,12 @@
 
    private static Tag readUnnamedTag(DataInput p_128931_, NbtAccounter p_128933_) throws IOException {
       byte b0 = p_128931_.readByte();
+      p_128933_.accountBytes(1); // Forge: Count everything!
       if (b0 == 0) {
          return EndTag.INSTANCE;
       } else {
-         StringTag.skipString(p_128931_);
+         p_128933_.readUTF(p_128931_.readUTF()); //Forge: Count this string.
+         p_128933_.accountBytes(4); //Forge: 4 extra bytes for the object allocation.
          return readTagSafe(p_128931_, p_128933_, b0);
       }
    }
