--- ../src_base/minecraft/net/minecraft/nbt/NBTTagList.java
+++ ../src_work/minecraft/net/minecraft/nbt/NBTTagList.java
@@ -1,7 +1,5 @@
 package net.minecraft.nbt;
 
-import cpw.mods.fml.relauncher.Side;
-import cpw.mods.fml.relauncher.SideOnly;
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
@@ -99,8 +97,6 @@
         this.tagList.add(par1NBTBase);
     }
 
-    @SideOnly(Side.CLIENT)
-
     /**
      * Removes a tag at the given index.
      */
@@ -108,6 +104,16 @@
     {
         return (NBTBase)this.tagList.remove(par1);
     }
+
+    // MCPC+ start- workaround for MoCreatures crash
+    /**
+     * Removes a tag at the given index.
+     */
+    public NBTBase func_74744_a(int par1)
+    {
+        return (NBTBase)this.tagList.remove(par1);
+    }
+    // MCPC+ end
 
     /**
      * Retrieves the tag at the specified index from the list.
