--- a/net/minecraft/world/chunk/Chunk.java
+++ b/net/minecraft/world/chunk/Chunk.java
@@ -55,7 +55,7 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class Chunk implements IChunk {
+public class Chunk extends net.minecraftforge.common.capabilities.CapabilityProvider<Chunk> implements IChunk, net.minecraftforge.common.extensions.IForgeChunk {
    private static final Logger field_150817_t = LogManager.getLogger();
    public static final ChunkSection field_186036_a = null;
    private final ChunkSection[] field_76652_q = new ChunkSection[16];
@@ -95,6 +95,7 @@
    }
 
    public Chunk(World p_i49379_1_, int p_i49379_2_, int p_i49379_3_, Biome[] p_i49379_4_, UpgradeData p_i49379_5_, ITickList<Block> p_i49379_6_, ITickList<Fluid> p_i49379_7_, long p_i49379_8_) {
+      super(Chunk.class);
       this.field_76645_j = new ClassInheritanceMultiMap[16];
       this.field_76637_e = p_i49379_1_;
       this.field_76635_g = p_i49379_2_;
@@ -116,6 +117,7 @@
       this.field_201621_s = p_i49379_6_;
       this.field_205325_u = p_i49379_7_;
       this.field_111204_q = p_i49379_8_;
+      this.gatherCapabilities();
    }
 
    public Chunk(World p_i48703_1_, ChunkPrimer p_i48703_2_, int p_i48703_3_, int p_i48703_4_) {
@@ -414,6 +416,7 @@
          Block block = p_177436_2_.func_177230_c();
          Block block1 = iblockstate.func_177230_c();
          ChunkSection chunksection = this.field_76652_q[j >> 4];
+         int j1 = iblockstate.func_200016_a(this.field_76637_e, p_177436_1_); // Relocate old light value lookup here, so that it is called before TE is removed.
          boolean flag = false;
          if (chunksection == field_186036_a) {
             if (p_177436_2_.func_196958_f()) {
@@ -432,7 +435,7 @@
          this.field_76634_f.get(Heightmap.Type.WORLD_SURFACE).func_202270_a(i, j, k, p_177436_2_);
          if (!this.field_76637_e.field_72995_K) {
             iblockstate.func_196947_b(this.field_76637_e, p_177436_1_, p_177436_2_, p_177436_3_);
-         } else if (block1 != block && block1 instanceof ITileEntityProvider) {
+         } else if (block1 != block && iblockstate.hasTileEntity()) {
             this.field_76637_e.func_175713_t(p_177436_1_);
          }
 
@@ -443,14 +446,13 @@
                this.func_76603_b();
             } else {
                int i1 = p_177436_2_.func_200016_a(this.field_76637_e, p_177436_1_);
-               int j1 = iblockstate.func_200016_a(this.field_76637_e, p_177436_1_);
                this.func_76615_h(i, j, k, p_177436_2_);
                if (i1 != j1 && (i1 < j1 || this.func_177413_a(EnumLightType.SKY, p_177436_1_) > 0 || this.func_177413_a(EnumLightType.BLOCK, p_177436_1_) > 0)) {
                   this.func_76595_e(i, k);
                }
             }
 
-            if (block1 instanceof ITileEntityProvider) {
+            if (iblockstate.hasTileEntity()) {
                TileEntity tileentity = this.func_177424_a(p_177436_1_, Chunk.EnumCreateEntityType.CHECK);
                if (tileentity != null) {
                   tileentity.func_145836_u();
@@ -461,10 +463,10 @@
                p_177436_2_.func_196945_a(this.field_76637_e, p_177436_1_, iblockstate);
             }
 
-            if (block instanceof ITileEntityProvider) {
+            if (p_177436_2_.hasTileEntity()) {
                TileEntity tileentity1 = this.func_177424_a(p_177436_1_, Chunk.EnumCreateEntityType.CHECK);
                if (tileentity1 == null) {
-                  tileentity1 = ((ITileEntityProvider)block).func_196283_a_(this.field_76637_e);
+                  tileentity1 = p_177436_2_.createTileEntity(this.field_76637_e);
                   this.field_76637_e.func_175690_a(p_177436_1_, tileentity1);
                } else {
                   tileentity1.func_145836_u();
@@ -579,11 +581,13 @@
          k = this.field_76645_j.length - 1;
       }
 
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityEvent.EnteringChunk(p_76612_1_, this.field_76635_g, this.field_76647_h, p_76612_1_.field_70176_ah, p_76612_1_.field_70164_aj));
       p_76612_1_.field_70175_ag = true;
       p_76612_1_.field_70176_ah = this.field_76635_g;
       p_76612_1_.field_70162_ai = k;
       p_76612_1_.field_70164_aj = this.field_76647_h;
       this.field_76645_j[k].add(p_76612_1_);
+      this.func_76630_e(); // Forge - ensure chunks are marked to save after an entity add
    }
 
    public void func_201607_a(Heightmap.Type p_201607_1_, long[] p_201607_2_) {
@@ -604,6 +608,7 @@
       }
 
       this.field_76645_j[p_76608_2_].remove(p_76608_1_);
+      this.func_76630_e(); // Forge - ensure chunks are marked to save after entity removals
    }
 
    public boolean func_177444_d(BlockPos p_177444_1_) {
@@ -621,7 +626,7 @@
    private TileEntity func_177422_i(BlockPos p_177422_1_) {
       IBlockState iblockstate = this.func_180495_p(p_177422_1_);
       Block block = iblockstate.func_177230_c();
-      return !block.func_149716_u() ? null : ((ITileEntityProvider)block).func_196283_a_(this.field_76637_e);
+      return !iblockstate.hasTileEntity() ? null : iblockstate.createTileEntity(this.field_76637_e);
    }
 
    @Nullable
@@ -632,6 +637,11 @@
    @Nullable
    public TileEntity func_177424_a(BlockPos p_177424_1_, Chunk.EnumCreateEntityType p_177424_2_) {
       TileEntity tileentity = this.field_150816_i.get(p_177424_1_);
+      if (tileentity != null && tileentity.func_145837_r()) {
+         field_150816_i.remove(p_177424_1_);
+         tileentity = null;
+      }
+
       if (tileentity == null) {
          NBTTagCompound nbttagcompound = this.field_201618_i.remove(p_177424_1_);
          if (nbttagcompound != null) {
@@ -647,11 +657,8 @@
             tileentity = this.func_177422_i(p_177424_1_);
             this.field_76637_e.func_175690_a(p_177424_1_, tileentity);
          } else if (p_177424_2_ == Chunk.EnumCreateEntityType.QUEUED) {
-            this.field_177447_w.add(p_177424_1_);
+            this.field_177447_w.add(p_177424_1_.func_185334_h());
          }
-      } else if (tileentity.func_145837_r()) {
-         this.field_150816_i.remove(p_177424_1_);
-         return null;
       }
 
       return tileentity;
@@ -666,9 +673,10 @@
    }
 
    public void func_177426_a(BlockPos p_177426_1_, TileEntity p_177426_2_) {
+      if (p_177426_2_.func_145831_w() != this.field_76637_e) //Forge don't call unless it's changed, could screw up bad mods.
       p_177426_2_.func_145834_a(this.field_76637_e);
       p_177426_2_.func_174878_a(p_177426_1_);
-      if (this.func_180495_p(p_177426_1_).func_177230_c() instanceof ITileEntityProvider) {
+      if (this.func_180495_p(p_177426_1_).hasTileEntity()) {
          if (this.field_150816_i.containsKey(p_177426_1_)) {
             this.field_150816_i.get(p_177426_1_).func_145843_s();
          }
@@ -702,9 +710,11 @@
          }));
       }
 
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(this));
    }
 
    public void func_76623_d() {
+      java.util.Arrays.stream(field_76645_j).forEach(multimap -> com.google.common.collect.Lists.newArrayList(multimap.func_180215_b(net.minecraft.entity.player.EntityPlayer.class)).forEach(player -> field_76637_e.func_72866_a(player, false))); // FORGE - Fix for MC-92916
       this.field_76636_d = false;
 
       for(TileEntity tileentity : this.field_150816_i.values()) {
@@ -714,6 +724,7 @@
       for(ClassInheritanceMultiMap<Entity> classinheritancemultimap : this.field_76645_j) {
          this.field_76637_e.func_175681_c(classinheritancemultimap);
       }
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Unload(this));
 
    }
 
@@ -722,8 +733,8 @@
    }
 
    public void func_177414_a(@Nullable Entity p_177414_1_, AxisAlignedBB p_177414_2_, List<Entity> p_177414_3_, Predicate<? super Entity> p_177414_4_) {
-      int i = MathHelper.func_76128_c((p_177414_2_.field_72338_b - 2.0D) / 16.0D);
-      int j = MathHelper.func_76128_c((p_177414_2_.field_72337_e + 2.0D) / 16.0D);
+      int i = MathHelper.func_76128_c((p_177414_2_.field_72338_b - this.field_76637_e.getMaxEntityRadius()) / 16.0D);
+      int j = MathHelper.func_76128_c((p_177414_2_.field_72337_e + this.field_76637_e.getMaxEntityRadius()) / 16.0D);
       i = MathHelper.func_76125_a(i, 0, this.field_76645_j.length - 1);
       j = MathHelper.func_76125_a(j, 0, this.field_76645_j.length - 1);
 
@@ -751,8 +762,8 @@
    }
 
    public <T extends Entity> void func_177430_a(Class<? extends T> p_177430_1_, AxisAlignedBB p_177430_2_, List<T> p_177430_3_, @Nullable Predicate<? super T> p_177430_4_) {
-      int i = MathHelper.func_76128_c((p_177430_2_.field_72338_b - 2.0D) / 16.0D);
-      int j = MathHelper.func_76128_c((p_177430_2_.field_72337_e + 2.0D) / 16.0D);
+      int i = MathHelper.func_76128_c((p_177430_2_.field_72338_b - this.field_76637_e.getMaxEntityRadius()) / 16.0D);
+      int j = MathHelper.func_76128_c((p_177430_2_.field_72337_e + this.field_76637_e.getMaxEntityRadius()) / 16.0D);
       i = MathHelper.func_76125_a(i, 0, this.field_76645_j.length - 1);
       j = MathHelper.func_76125_a(j, 0, this.field_76645_j.length - 1);
 
@@ -791,7 +802,7 @@
 
       while(!this.field_177447_w.isEmpty()) {
          BlockPos blockpos = this.field_177447_w.poll();
-         if (this.func_177424_a(blockpos, Chunk.EnumCreateEntityType.CHECK) == null && this.func_180495_p(blockpos).func_177230_c().func_149716_u()) {
+         if (this.func_177424_a(blockpos, Chunk.EnumCreateEntityType.CHECK) == null && this.func_180495_p(blockpos).hasTileEntity()) {
             TileEntity tileentity = this.func_177422_i(blockpos);
             this.field_76637_e.func_175690_a(blockpos, tileentity);
             this.field_76637_e.func_175704_b(blockpos, blockpos);
@@ -841,6 +852,10 @@
 
    @OnlyIn(Dist.CLIENT)
    public void func_186033_a(PacketBuffer p_186033_1_, int p_186033_2_, boolean p_186033_3_) {
+      for (TileEntity tileEntity : field_150816_i.values()) {
+         tileEntity.func_145836_u();
+         tileEntity.func_195044_w();
+      }
       if (p_186033_3_) {
          this.field_150816_i.clear();
       } else {
@@ -929,10 +944,10 @@
             for(int i1 = 0; i1 < 16; ++i1) {
                BlockPos blockpos1 = blockpos.func_177982_a(k, (j << 4) + i1, l);
                boolean flag = i1 == 0 || i1 == 15 || k == 0 || k == 15 || l == 0 || l == 15;
-               if (this.field_76652_q[j] == field_186036_a && flag || this.field_76652_q[j] != field_186036_a && this.field_76652_q[j].func_177485_a(k, i1, l).func_196958_f()) {
+               if (this.field_76652_q[j] == field_186036_a && flag || this.field_76652_q[j] != field_186036_a && this.field_76652_q[j].func_177485_a(k, i1, l).isAir(field_76637_e, blockpos1)) {
                   for(EnumFacing enumfacing : EnumFacing.values()) {
                      BlockPos blockpos2 = blockpos1.func_177972_a(enumfacing);
-                     if (this.field_76637_e.func_180495_p(blockpos2).func_185906_d() > 0) {
+                     if (this.field_76637_e.func_180495_p(blockpos2).getLightValue(field_76637_e, blockpos2) > 0) {
                         this.field_76637_e.func_175664_x(blockpos2);
                      }
                   }
@@ -1097,9 +1112,9 @@
    private TileEntity func_212815_a(BlockPos p_212815_1_, NBTTagCompound p_212815_2_) {
       TileEntity tileentity;
       if ("DUMMY".equals(p_212815_2_.func_74779_i("id"))) {
-         Block block = this.func_180495_p(p_212815_1_).func_177230_c();
-         if (block instanceof ITileEntityProvider) {
-            tileentity = ((ITileEntityProvider)block).func_196283_a_(this.field_76637_e);
+         IBlockState state = this.func_180495_p(p_212815_1_);
+         if (state.hasTileEntity()) {
+            tileentity = state.createTileEntity(this.field_76637_e);
          } else {
             tileentity = null;
             field_150817_t.warn("Tried to load a DUMMY block entity @ {} but found not block entity block {} at location", p_212815_1_, this.func_180495_p(p_212815_1_));
@@ -1170,4 +1185,30 @@
       QUEUED,
       CHECK;
    }
+
+   /**
+    * <strong>FOR INTERNAL USE ONLY</strong>
+    * <p>
+    * Only public for use in {@link AnvilChunkLoader}.
+    */
+   @java.lang.Deprecated
+   @javax.annotation.Nullable
+   public final NBTTagCompound writeCapsToNBT() {
+      return this.serializeCaps();
+   }
+
+   /**
+    * <strong>FOR INTERNAL USE ONLY</strong>
+    * <p>
+    * Only public for use in {@link AnvilChunkLoader}.
+    */
+   @java.lang.Deprecated
+   public final void readCapsFromNBT(NBTTagCompound tag) {
+      this.deserializeCaps(tag);
+   }
+
+   @Override
+   public World getWorldForge() {
+      return func_177412_p();
+   }
 }
