--- a/net/minecraft/world/chunk/ChunkPrimer.java
+++ b/net/minecraft/world/chunk/ChunkPrimer.java
@@ -138,13 +_,13 @@
          if (this.sections[j >> 4] == Chunk.EMPTY_SECTION && p_177436_2_.is(Blocks.AIR)) {
             return p_177436_2_;
          } else {
-            if (p_177436_2_.getLightEmission() > 0) {
+            if (p_177436_2_.getLightValue(this, p_177436_1_) > 0) {
                this.lights.add(new BlockPos((i & 15) + this.getPos().getMinBlockX(), j, (k & 15) + this.getPos().getMinBlockZ()));
             }
 
             ChunkSection chunksection = this.getOrCreateSection(j >> 4);
             BlockState blockstate = chunksection.setBlockState(i & 15, j & 15, k & 15, p_177436_2_);
-            if (this.status.isOrAfter(ChunkStatus.FEATURES) && p_177436_2_ != blockstate && (p_177436_2_.getLightBlock(this, p_177436_1_) != blockstate.getLightBlock(this, p_177436_1_) || p_177436_2_.getLightEmission() != blockstate.getLightEmission() || p_177436_2_.useShapeForLightOcclusion() || blockstate.useShapeForLightOcclusion())) {
+            if (this.status.isOrAfter(ChunkStatus.FEATURES) && p_177436_2_ != blockstate && (p_177436_2_.getLightBlock(this, p_177436_1_) != blockstate.getLightBlock(this, p_177436_1_) || p_177436_2_.getLightValue(this, p_177436_1_) != blockstate.getLightValue(this, p_177436_1_) || p_177436_2_.useShapeForLightOcclusion() || blockstate.useShapeForLightOcclusion())) {
                WorldLightManager worldlightmanager = this.getLightEngine();
                worldlightmanager.checkBlock(p_177436_1_);
             }
