--- ../src-base/minecraft/net/minecraft/world/chunk/Chunk.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/Chunk.java
@@ -68,6 +68,7 @@
     private int field_76649_t;
     private final ConcurrentLinkedQueue<BlockPos> field_177447_w;
     public boolean field_189550_d;
+    public short[] neighborLightChecks = null;
 
     public Chunk(World p_i1995_1_, int p_i1995_2_, int p_i1995_3_)
     {
@@ -179,7 +180,7 @@
                 {
                     IBlockState iblockstate = this.func_186032_a(j, l - 1, k);
 
-                    if (iblockstate.func_185891_c() != 0)
+                    if (this.func_150808_b(j, l - 1, k) != 0)
                     {
                         this.field_76634_f[k << 4 | j] = l;
 
@@ -359,13 +360,15 @@
 
         if (j != i)
         {
-            this.field_76637_e.func_72975_g(p_76615_1_ + this.field_76635_g * 16, p_76615_3_ + this.field_76647_h * 16, j, i);
+            if (false) this.field_76637_e.func_72975_g(p_76615_1_ + this.field_76635_g * 16, p_76615_3_ + this.field_76647_h * 16, j, i); //Forge: Useless, since heightMap is not updated yet (See #3871)
             this.field_76634_f[p_76615_3_ << 4 | p_76615_1_] = j;
             int k = this.field_76635_g * 16 + p_76615_1_;
             int l = this.field_76647_h * 16 + p_76615_3_;
 
             if (this.field_76637_e.field_73011_w.func_191066_m())
             {
+                net.minecraftforge.common.lighting.LightingHooks.relightSkylightColumn(this.field_76637_e, this, p_76615_1_, p_76615_3_, i, j); //Forge: Optimized version of World.markBlocksDirtyVertical; heightMap is now updated (See #3871)
+            } else if (false) { //Forge: Don't mess up the light cache; World.checkLight already does all necessary steps (See #3871)
                 if (j < i)
                 {
                     for (int j1 = j; j1 < i; ++j1)
@@ -435,6 +438,7 @@
             {
                 this.field_82912_p = l1;
             }
+            if (true) return; //Forge: Following checks are not needed if the light cache is not messed up (See #3871)
 
             if (this.field_76637_e.field_73011_w.func_191066_m())
             {
@@ -452,12 +456,13 @@
 
     public int func_177437_b(BlockPos p_177437_1_)
     {
-        return this.func_177435_g(p_177437_1_).func_185891_c();
+        return this.func_177435_g(p_177437_1_).getLightOpacity(this.field_76637_e, p_177437_1_);
     }
 
     private int func_150808_b(int p_150808_1_, int p_150808_2_, int p_150808_3_)
     {
-        return this.func_186032_a(p_150808_1_, p_150808_2_, p_150808_3_).func_185891_c();
+        IBlockState state = this.func_186032_a(p_150808_1_, p_150808_2_, p_150808_3_); //Forge: Can sometimes be called before we are added to the global world list. So use the less accurate one during that. It'll be recalculated later
+        return !field_76636_d ? state.func_185891_c() : state.getLightOpacity(field_76637_e, new BlockPos(this.field_76635_g << 4 | p_150808_1_ & 15, p_150808_2_, this.field_76647_h << 4 | p_150808_3_ & 15));
     }
 
     public IBlockState func_177435_g(BlockPos p_177435_1_)
@@ -539,6 +544,7 @@
         {
             Block block = p_177436_2_.func_177230_c();
             Block block1 = iblockstate.func_177230_c();
+            int k1 = iblockstate.getLightOpacity(this.field_76637_e, p_177436_1_); // Relocate old light value lookup here, so that it is called before TE is removed.
             ExtendedBlockStorage extendedblockstorage = this.field_76652_q[j >> 4];
             boolean flag = false;
 
@@ -556,14 +562,19 @@
 
             extendedblockstorage.func_177484_a(i, j & 15, k, p_177436_2_);
 
-            if (block1 != block)
+            //if (block1 != block)
             {
                 if (!this.field_76637_e.field_72995_K)
                 {
+                    if (block1 != block) //Only fire block breaks when the block changes.
                     block1.func_180663_b(this.field_76637_e, p_177436_1_, iblockstate);
+                    TileEntity te = this.func_177424_a(p_177436_1_, Chunk.EnumCreateEntityType.CHECK);
+                    if (te != null && te.shouldRefresh(this.field_76637_e, p_177436_1_, iblockstate, p_177436_2_)) this.field_76637_e.func_175713_t(p_177436_1_);
                 }
-                else if (block1 instanceof ITileEntityProvider)
+                else if (block1.hasTileEntity(iblockstate))
                 {
+                    TileEntity te = this.func_177424_a(p_177436_1_, Chunk.EnumCreateEntityType.CHECK);
+                    if (te != null && te.shouldRefresh(this.field_76637_e, p_177436_1_, iblockstate, p_177436_2_))
                     this.field_76637_e.func_175713_t(p_177436_1_);
                 }
             }
@@ -580,8 +591,7 @@
                 }
                 else
                 {
-                    int j1 = p_177436_2_.func_185891_c();
-                    int k1 = iblockstate.func_185891_c();
+                    int j1 = p_177436_2_.getLightOpacity(this.field_76637_e, p_177436_1_);
 
                     if (j1 > 0)
                     {
@@ -595,34 +605,25 @@
                         this.func_76615_h(i, j, k);
                     }
 
-                    if (j1 != k1 && (j1 < k1 || this.func_177413_a(EnumSkyBlock.SKY, p_177436_1_) > 0 || this.func_177413_a(EnumSkyBlock.BLOCK, p_177436_1_) > 0))
+                    if (false) //Forge: Error correction is unnecessary as these are fixed (See #3871)
                     {
                         this.func_76595_e(i, k);
                     }
                 }
 
-                if (block1 instanceof ITileEntityProvider)
+                // If capturing blocks, only run block physics for TE's. Non-TE's are handled in ForgeHooks.onPlaceItemIntoWorld
+                if (!this.field_76637_e.field_72995_K && block1 != block && (!this.field_76637_e.captureBlockSnapshots || block.hasTileEntity(p_177436_2_)))
                 {
-                    TileEntity tileentity = this.func_177424_a(p_177436_1_, Chunk.EnumCreateEntityType.CHECK);
-
-                    if (tileentity != null)
-                    {
-                        tileentity.func_145836_u();
-                    }
-                }
-
-                if (!this.field_76637_e.field_72995_K && block1 != block)
-                {
                     block.func_176213_c(this.field_76637_e, p_177436_1_, p_177436_2_);
                 }
 
-                if (block instanceof ITileEntityProvider)
+                if (block.hasTileEntity(p_177436_2_))
                 {
                     TileEntity tileentity1 = this.func_177424_a(p_177436_1_, Chunk.EnumCreateEntityType.CHECK);
 
                     if (tileentity1 == null)
                     {
-                        tileentity1 = ((ITileEntityProvider)block).func_149915_a(this.field_76637_e, block.func_176201_c(p_177436_2_));
+                        tileentity1 = block.createTileEntity(this.field_76637_e, p_177436_2_);
                         this.field_76637_e.func_175690_a(p_177436_1_, tileentity1);
                     }
 
@@ -640,6 +641,12 @@
 
     public int func_177413_a(EnumSkyBlock p_177413_1_, BlockPos p_177413_2_)
     {
+        this.field_76637_e.lightingEngine.procLightUpdates(p_177413_1_);
+        return this.getCachedLightFor(p_177413_1_, p_177413_2_);
+    }
+
+    public int getCachedLightFor(EnumSkyBlock p_177413_1_, BlockPos p_177413_2_)
+    {
         int i = p_177413_2_.func_177958_n() & 15;
         int j = p_177413_2_.func_177956_o();
         int k = p_177413_2_.func_177952_p() & 15;
@@ -690,6 +697,7 @@
 
     public int func_177443_a(BlockPos p_177443_1_, int p_177443_2_)
     {
+        this.field_76637_e.lightingEngine.procLightUpdates();
         int i = p_177443_1_.func_177958_n() & 15;
         int j = p_177443_1_.func_177956_o();
         int k = p_177443_1_.func_177952_p() & 15;
@@ -738,6 +746,7 @@
             k = this.field_76645_j.length - 1;
         }
 
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityEvent.EnteringChunk(p_76612_1_, this.field_76635_g, this.field_76647_h, p_76612_1_.field_70176_ah, p_76612_1_.field_70164_aj));
         p_76612_1_.field_70175_ag = true;
         p_76612_1_.field_70176_ah = this.field_76635_g;
         p_76612_1_.field_70162_ai = k;
@@ -778,7 +787,7 @@
     {
         IBlockState iblockstate = this.func_177435_g(p_177422_1_);
         Block block = iblockstate.func_177230_c();
-        return !block.func_149716_u() ? null : ((ITileEntityProvider)block).func_149915_a(this.field_76637_e, iblockstate.func_177230_c().func_176201_c(iblockstate));
+        return !block.hasTileEntity(iblockstate) ? null : block.createTileEntity(this.field_76637_e, iblockstate);
     }
 
     @Nullable
@@ -786,6 +795,12 @@
     {
         TileEntity tileentity = this.field_150816_i.get(p_177424_1_);
 
+        if (tileentity != null && tileentity.func_145837_r())
+        {
+            field_150816_i.remove(p_177424_1_);
+            tileentity = null;
+        }
+
         if (tileentity == null)
         {
             if (p_177424_2_ == Chunk.EnumCreateEntityType.IMMEDIATE)
@@ -795,14 +810,9 @@
             }
             else if (p_177424_2_ == Chunk.EnumCreateEntityType.QUEUED)
             {
-                this.field_177447_w.add(p_177424_1_);
+                this.field_177447_w.add(p_177424_1_.func_185334_h());
             }
         }
-        else if (tileentity.func_145837_r())
-        {
-            this.field_150816_i.remove(p_177424_1_);
-            return null;
-        }
 
         return tileentity;
     }
@@ -819,10 +829,11 @@
 
     public void func_177426_a(BlockPos p_177426_1_, TileEntity p_177426_2_)
     {
+        if (p_177426_2_.func_145831_w() != this.field_76637_e) //Forge don't call unless it's changed, could screw up bad mods.
         p_177426_2_.func_145834_a(this.field_76637_e);
         p_177426_2_.func_174878_a(p_177426_1_);
 
-        if (this.func_177435_g(p_177426_1_).func_177230_c() instanceof ITileEntityProvider)
+        if (this.func_177435_g(p_177426_1_).func_177230_c().hasTileEntity(this.func_177435_g(p_177426_1_)))
         {
             if (this.field_150816_i.containsKey(p_177426_1_))
             {
@@ -854,8 +865,9 @@
 
         for (ClassInheritanceMultiMap<Entity> classinheritancemultimap : this.field_76645_j)
         {
-            this.field_76637_e.func_175650_b(classinheritancemultimap);
+            this.field_76637_e.func_175650_b(com.google.common.collect.ImmutableList.copyOf(classinheritancemultimap));
         }
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(this));
     }
 
     public void func_76623_d()
@@ -871,6 +883,8 @@
         {
             this.field_76637_e.func_175681_c(classinheritancemultimap);
         }
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Unload(this));
+        net.minecraftforge.common.lighting.LightingHooks.scheduleRelightChecksForChunkBoundaries(this.field_76637_e, this);
     }
 
     public void func_76630_e()
@@ -880,8 +894,8 @@
 
     public void func_177414_a(@Nullable Entity p_177414_1_, AxisAlignedBB p_177414_2_, List<Entity> p_177414_3_, Predicate <? super Entity > p_177414_4_)
     {
-        int i = MathHelper.func_76128_c((p_177414_2_.field_72338_b - 2.0D) / 16.0D);
-        int j = MathHelper.func_76128_c((p_177414_2_.field_72337_e + 2.0D) / 16.0D);
+        int i = MathHelper.func_76128_c((p_177414_2_.field_72338_b - World.MAX_ENTITY_RADIUS) / 16.0D);
+        int j = MathHelper.func_76128_c((p_177414_2_.field_72337_e + World.MAX_ENTITY_RADIUS) / 16.0D);
         i = MathHelper.func_76125_a(i, 0, this.field_76645_j.length - 1);
         j = MathHelper.func_76125_a(j, 0, this.field_76645_j.length - 1);
 
@@ -918,8 +932,8 @@
 
     public <T extends Entity> void func_177430_a(Class <? extends T > p_177430_1_, AxisAlignedBB p_177430_2_, List<T> p_177430_3_, Predicate <? super T > p_177430_4_)
     {
-        int i = MathHelper.func_76128_c((p_177430_2_.field_72338_b - 2.0D) / 16.0D);
-        int j = MathHelper.func_76128_c((p_177430_2_.field_72337_e + 2.0D) / 16.0D);
+        int i = MathHelper.func_76128_c((p_177430_2_.field_72338_b - World.MAX_ENTITY_RADIUS) / 16.0D);
+        int j = MathHelper.func_76128_c((p_177430_2_.field_72337_e + World.MAX_ENTITY_RADIUS) / 16.0D);
         i = MathHelper.func_76125_a(i, 0, this.field_76645_j.length - 1);
         j = MathHelper.func_76125_a(j, 0, this.field_76645_j.length - 1);
 
@@ -997,6 +1011,8 @@
 
     protected void func_186034_a(IChunkGenerator p_186034_1_)
     {
+        if (populating && net.minecraftforge.common.ForgeModContainer.logCascadingWorldGeneration) logCascadingWorldGeneration();
+        populating = true;
         if (this.func_177419_t())
         {
             if (p_186034_1_.func_185933_a(this, this.field_76635_g, this.field_76647_h))
@@ -1008,8 +1024,10 @@
         {
             this.func_150809_p();
             p_186034_1_.func_185931_b(this.field_76635_g, this.field_76647_h);
+            net.minecraftforge.fml.common.registry.GameRegistry.generateWorld(this.field_76635_g, this.field_76647_h, this.field_76637_e, p_186034_1_, this.field_76637_e.func_72863_F());
             this.func_76630_e();
         }
+        populating = false;
     }
 
     public BlockPos func_177440_h(BlockPos p_177440_1_)
@@ -1064,7 +1082,7 @@
         {
             BlockPos blockpos = this.field_177447_w.poll();
 
-            if (this.func_177424_a(blockpos, Chunk.EnumCreateEntityType.CHECK) == null && this.func_177435_g(blockpos).func_177230_c().func_149716_u())
+            if (this.func_177424_a(blockpos, Chunk.EnumCreateEntityType.CHECK) == null && this.func_177435_g(blockpos).func_177230_c().hasTileEntity(this.func_177435_g(blockpos)))
             {
                 TileEntity tileentity = this.func_177422_i(blockpos);
                 this.field_76637_e.func_175690_a(blockpos, tileentity);
@@ -1128,6 +1146,13 @@
     @SideOnly(Side.CLIENT)
     public void func_186033_a(PacketBuffer p_186033_1_, int p_186033_2_, boolean p_186033_3_)
     {
+        for(TileEntity tileEntity : field_150816_i.values())
+        {
+            tileEntity.func_145836_u();
+            tileEntity.func_145832_p();
+            tileEntity.func_145838_q();
+        }
+
         boolean flag = this.field_76637_e.field_73011_w.func_191066_m();
 
         for (int i = 0; i < this.field_76652_q.length; ++i)
@@ -1176,10 +1201,16 @@
         this.field_76646_k = true;
         this.func_76590_a();
 
+        List<TileEntity> invalidList = new java.util.ArrayList<TileEntity>();
+
         for (TileEntity tileentity : this.field_150816_i.values())
         {
+            if (tileentity.shouldRefresh(this.field_76637_e, tileentity.func_174877_v(), tileentity.func_145838_q().func_176203_a(tileentity.func_145832_p()), func_177435_g(tileentity.func_174877_v())))
+                invalidList.add(tileentity);
             tileentity.func_145836_u();
         }
+
+        for (TileEntity te : invalidList) te.func_145843_s();
     }
 
     public Biome func_177411_a(BlockPos p_177411_1_, BiomeProvider p_177411_2_)
@@ -1244,13 +1275,13 @@
                     BlockPos blockpos1 = blockpos.func_177982_a(k, (j << 4) + i1, l);
                     boolean flag = i1 == 0 || i1 == 15 || k == 0 || k == 15 || l == 0 || l == 15;
 
-                    if (this.field_76652_q[j] == field_186036_a && flag || this.field_76652_q[j] != field_186036_a && this.field_76652_q[j].func_177485_a(k, i1, l).func_185904_a() == Material.field_151579_a)
+                    if (this.field_76652_q[j] == field_186036_a && flag || this.field_76652_q[j] != field_186036_a && this.field_76652_q[j].func_177485_a(k, i1, l).func_177230_c().isAir(this.field_76652_q[j].func_177485_a(k, i1, l), this.field_76637_e, blockpos1))
                     {
                         for (EnumFacing enumfacing : EnumFacing.values())
                         {
                             BlockPos blockpos2 = blockpos1.func_177972_a(enumfacing);
 
-                            if (this.field_76637_e.func_180495_p(blockpos2).func_185906_d() > 0)
+                            if (this.field_76637_e.func_180495_p(blockpos2).getLightValue(this.field_76637_e, blockpos2) > 0)
                             {
                                 this.field_76637_e.func_175664_x(blockpos2);
                             }
@@ -1381,7 +1412,7 @@
         {
             blockpos$mutableblockpos.func_181079_c(blockpos$mutableblockpos.func_177958_n(), l, blockpos$mutableblockpos.func_177952_p());
 
-            if (this.func_177435_g(blockpos$mutableblockpos).func_185906_d() > 0)
+            if (this.func_177435_g(blockpos$mutableblockpos).getLightValue(this.field_76637_e, blockpos$mutableblockpos) > 0)
             {
                 this.field_76637_e.func_175664_x(blockpos$mutableblockpos);
             }
@@ -1489,4 +1520,34 @@
         QUEUED,
         CHECK;
     }
+
+    /* ======================================== FORGE START =====================================*/
+    /**
+     * Removes the tile entity at the specified position, only if it's
+     * marked as invalid.
+     */
+    public void removeInvalidTileEntity(BlockPos pos)
+    {
+        if (field_76636_d)
+        {
+            TileEntity entity = (TileEntity)field_150816_i.get(pos);
+            if (entity != null && entity.func_145837_r())
+            {
+                field_150816_i.remove(pos);
+            }
+        }
+    }
+
+    private static boolean populating = false; // keep track of cascading chunk generation during chunk population
+
+    private void logCascadingWorldGeneration()
+    {
+        net.minecraftforge.fml.common.ModContainer activeModContainer = net.minecraftforge.fml.common.Loader.instance().activeModContainer();
+        String format = "{} loaded a new chunk ({}, {}  Dimension: {}) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.";
+
+        if (activeModContainer == null) // vanilla minecraft has problems too (MC-114332), log it at a quieter level.
+            net.minecraftforge.fml.common.FMLLog.log.debug(format, "Minecraft", this.field_76635_g, this.field_76647_h, this.field_76637_e.field_73011_w.getDimension());
+        else
+            net.minecraftforge.fml.common.FMLLog.log.warn(format, activeModContainer.getName(), this.field_76635_g, this.field_76647_h, this.field_76637_e.field_73011_w.getDimension());
+    }
 }
