--- ../src-base/minecraft/net/minecraft/client/renderer/InventoryEffectRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/InventoryEffectRenderer.java
@@ -28,7 +28,12 @@
 
     protected void updateActivePotionEffects()
     {
-        if (!this.mc.thePlayer.getActivePotionEffects().isEmpty())
+        boolean hasVisibleEffect = false;
+        for(PotionEffect potioneffect : this.mc.thePlayer.getActivePotionEffects()) {
+            Potion potion = Potion.potionTypes[potioneffect.getPotionID()];
+            if(potion.shouldRender(potioneffect)) { hasVisibleEffect = true; break; }
+        }
+        if (!this.mc.thePlayer.getActivePotionEffects().isEmpty() && hasVisibleEffect)
         {
             this.guiLeft = 160 + (this.width - this.xSize - 200) / 2;
             this.hasActivePotionEffects = true;
@@ -81,6 +86,8 @@
                     this.drawTexturedModalRect(i + 6, j + 7, 0 + i1 % 8 * 18, 198 + i1 / 8 * 18, 18, 18);
                 }
 
+                potion.renderInventoryEffect(i, j, potioneffect, mc);
+                if (!potion.shouldRenderInvText(potioneffect)) { j += l; continue; }
                 String s1 = I18n.format(potion.getName(), new Object[0]);
 
                 if (potioneffect.getAmplifier() == 1)
