--- a/net/minecraft/loot/functions/Smelt.java
+++ b/net/minecraft/loot/functions/Smelt.java
@@ -34,7 +34,7 @@
             ItemStack itemstack = optional.get().func_77571_b();
             if (!itemstack.func_190926_b()) {
                ItemStack itemstack1 = itemstack.func_77946_l();
-               itemstack1.func_190920_e(p_215859_1_.func_190916_E());
+               itemstack1.func_190920_e(p_215859_1_.func_190916_E() * itemstack.func_190916_E()); // Forge: Support smelting returning multiple
                return itemstack1;
             }
          }
