--- ../src-base/minecraft/net/minecraft/world/storage/MapData.java
+++ ../src-work/minecraft/net/minecraft/world/storage/MapData.java
@@ -21,7 +21,7 @@
 {
     public int field_76201_a;
     public int field_76199_b;
-    public byte field_76200_c;
+    public int field_76200_c; //FML byte -> int
     public byte field_76197_d;
     public byte[] field_76198_e = new byte[16384];
     public List<MapData.MapInfo> field_76196_g = Lists.<MapData.MapInfo>newArrayList();
@@ -44,7 +44,17 @@
 
     public void func_76184_a(NBTTagCompound p_76184_1_)
     {
-        this.field_76200_c = p_76184_1_.func_74771_c("dimension");
+        net.minecraft.nbt.NBTBase dimension = p_76184_1_.func_74781_a("dimension");
+
+        if (dimension instanceof net.minecraft.nbt.NBTTagByte)
+        {
+            this.field_76200_c = ((net.minecraft.nbt.NBTTagByte)dimension).func_150290_f();
+        }
+        else
+        {
+            this.field_76200_c = ((net.minecraft.nbt.NBTTagInt)dimension).func_150287_d();
+        }
+
         this.field_76201_a = p_76184_1_.func_74762_e("xCenter");
         this.field_76199_b = p_76184_1_.func_74762_e("zCenter");
         this.field_76197_d = p_76184_1_.func_74771_c("scale");
@@ -85,7 +95,7 @@
 
     public void func_76187_b(NBTTagCompound p_76187_1_)
     {
-        p_76187_1_.func_74774_a("dimension", this.field_76200_c);
+        p_76187_1_.func_74768_a("dimension", this.field_76200_c);
         p_76187_1_.func_74768_a("xCenter", this.field_76201_a);
         p_76187_1_.func_74768_a("zCenter", this.field_76199_b);
         p_76187_1_.func_74774_a("scale", this.field_76197_d);
@@ -164,7 +174,7 @@
             p_82567_8_ = p_82567_8_ + (p_82567_8_ < 0.0D ? -8.0D : 8.0D);
             b2 = (byte)((int)(p_82567_8_ * 16.0D / 360.0D));
 
-            if (this.field_76200_c < 0)
+            if (p_82567_2_.field_73011_w.shouldMapSpin(p_82567_3_, p_82567_4_, p_82567_6_, p_82567_8_))
             {
                 int k = (int)(p_82567_2_.func_72912_H().func_76073_f() / 10L);
                 b2 = (byte)(k * k * 34187121 + k * 121 >> 15 & 15);
