--- a/net/minecraft/world/storage/MapData.java
+++ b/net/minecraft/world/storage/MapData.java
@@ -198,8 +198,8 @@
       if (f >= -63.0F && f1 >= -63.0F && f <= 63.0F && f1 <= 63.0F) {
          p_191095_8_ = p_191095_8_ + (p_191095_8_ < 0.0D ? -8.0D : 8.0D);
          b2 = (byte)((int)(p_191095_8_ * 16.0D / 360.0D));
-         if (this.field_76200_c == DimensionType.NETHER && p_191095_2_ != null) {
-            int l = (int)(p_191095_2_.func_72912_H().func_76073_f() / 10L);
+         if (p_191095_2_ != null && p_191095_2_.func_201672_e().field_73011_w.shouldMapSpin(p_191095_3_, p_191095_4_, p_191095_6_, p_191095_8_)) {
+            int l = (int)(p_191095_2_.func_201672_e().func_72820_D() / 10L);
             b2 = (byte)(l * l * 34187121 + l * 121 >> 15 & 15);
          }
       } else {
