--- a/net/minecraft/world/storage/MapDecoration.java
+++ b/net/minecraft/world/storage/MapDecoration.java
@@ -82,6 +_,16 @@
       return 31 * i + Objects.hashCode(this.field_204310_e);
    }
 
+   /**
+    * Renders this decoration, useful for custom sprite sheets.
+    * @param index The index of this icon in the MapData's list. Used by vanilla to offset the Z-coordinate to prevent Z-fighting
+    * @return false to run vanilla logic for this decoration, true to skip it
+    */
+   @OnlyIn(Dist.CLIENT)
+   public boolean render(int index) {
+      return false;
+   }
+
    public static enum Type {
       PLAYER(false),
       FRAME(true),
