--- ../src-base/minecraft/net/minecraft/world/storage/MapData.java
+++ ../src-work/minecraft/net/minecraft/world/storage/MapData.java
@@ -20,7 +20,7 @@
 {
     public int field_76201_a;
     public int field_76199_b;
-    public byte field_76200_c;
+    public int field_76200_c; //FML byte -> int
     public boolean field_186210_e;
     public boolean field_191096_f;
     public byte field_76197_d;
@@ -45,7 +45,7 @@
 
     public void func_76184_a(NBTTagCompound p_76184_1_)
     {
-        this.field_76200_c = p_76184_1_.func_74771_c("dimension");
+        this.field_76200_c = p_76184_1_.func_74762_e("dimension");
         this.field_76201_a = p_76184_1_.func_74762_e("xCenter");
         this.field_76199_b = p_76184_1_.func_74762_e("zCenter");
         this.field_76197_d = p_76184_1_.func_74771_c("scale");
@@ -97,7 +97,7 @@
 
     public NBTTagCompound func_189551_b(NBTTagCompound p_189551_1_)
     {
-        p_189551_1_.func_74774_a("dimension", this.field_76200_c);
+        p_189551_1_.func_74768_a("dimension", this.field_76200_c);
         p_189551_1_.func_74768_a("xCenter", this.field_76201_a);
         p_189551_1_.func_74768_a("zCenter", this.field_76199_b);
         p_189551_1_.func_74774_a("scale", this.field_76197_d);
@@ -158,7 +158,11 @@
 
                 if (!this.field_76203_h.containsKey(nbttagcompound.func_74779_i("id")))
                 {
-                    this.func_191095_a(MapDecoration.Type.func_191159_a(nbttagcompound.func_74771_c("type")), p_76191_1_.field_70170_p, nbttagcompound.func_74779_i("id"), nbttagcompound.func_74769_h("x"), nbttagcompound.func_74769_h("z"), nbttagcompound.func_74769_h("rot"));
+                    // Forge support for old maps
+                    MapDecoration.Type type;
+                    if (nbttagcompound.func_150297_b("type", 99)) { type = MapDecoration.Type.func_191159_a(nbttagcompound.func_74771_c("type")); } // Byte
+                    else { type = MapDecoration.Type.byName(nbttagcompound.func_74779_i("type")); } // String
+                    this.func_191095_a(type, p_76191_1_.field_70170_p, nbttagcompound.func_74779_i("id"), nbttagcompound.func_74769_h("x"), nbttagcompound.func_74769_h("z"), nbttagcompound.func_74769_h("rot"));
                 }
             }
         }
@@ -179,7 +183,7 @@
         }
 
         NBTTagCompound nbttagcompound = new NBTTagCompound();
-        nbttagcompound.func_74774_a("type", p_191094_3_.func_191163_a());
+        nbttagcompound.func_74778_a("type", p_191094_3_.name());
         nbttagcompound.func_74778_a("id", p_191094_2_);
         nbttagcompound.func_74780_a("x", (double)p_191094_1_.func_177958_n());
         nbttagcompound.func_74780_a("z", (double)p_191094_1_.func_177952_p());
@@ -208,9 +212,9 @@
             p_191095_8_ = p_191095_8_ + (p_191095_8_ < 0.0D ? -8.0D : 8.0D);
             b2 = (byte)((int)(p_191095_8_ * 16.0D / 360.0D));
 
-            if (this.field_76200_c < 0)
+            if (p_191095_2_.field_73011_w.shouldMapSpin(p_191095_3_, p_191095_4_, p_191095_6_, p_191095_8_))
             {
-                int l = (int)(p_191095_2_.func_72912_H().func_76073_f() / 10L);
+                int l = (int)(p_191095_2_.func_72820_D() / 10L);
                 b2 = (byte)(l * l * 34187121 + l * 121 >> 15 & 15);
             }
         }
