--- ../src-base/minecraft/net/minecraft/world/storage/WorldInfo.java
+++ ../src-work/minecraft/net/minecraft/world/storage/WorldInfo.java
@@ -51,6 +51,7 @@
     private int field_176154_I = 5;
     private int field_176155_J = 15;
     private GameRules field_82577_x = new GameRules();
+    private java.util.Map<String, net.minecraft.nbt.NBTBase> additionalProperties;
 
     protected WorldInfo()
     {
@@ -438,7 +439,9 @@
 
     public void func_176142_i(int p_176142_1_)
     {
-        this.field_176157_p = p_176142_1_;
+    	if (!net.minecraftforge.common.ForgeHooks.onWeatherClear(MinecraftServer.func_71276_C().func_71218_a(field_76105_j).field_73059_b.field_73251_h)) {
+    		this.field_176157_p = p_176142_1_;
+    	}
     }
 
     public boolean func_76061_m()
@@ -448,7 +451,9 @@
 
     public void func_76069_a(boolean p_76069_1_)
     {
-        this.field_76102_o = p_76069_1_;
+    	if (!net.minecraftforge.common.ForgeHooks.onWeatherStorm(MinecraftServer.func_71276_C().func_71218_a(field_76105_j).field_73059_b.field_73251_h)) {
+    		this.field_76102_o = p_76069_1_;
+    	}
     }
 
     public int func_76071_n()
@@ -468,7 +473,9 @@
 
     public void func_76084_b(boolean p_76084_1_)
     {
-        this.field_76104_m = p_76084_1_;
+    	if (!net.minecraftforge.common.ForgeHooks.onWeatherRain(MinecraftServer.func_71276_C().func_71218_a(field_76105_j).field_73059_b.field_73251_h)) {
+    		this.field_76104_m = p_76084_1_;
+    	}
     }
 
     public int func_76083_p()
@@ -745,4 +752,24 @@
             }
         });
     }
+
+    /**
+     * Allow access to additional mod specific world based properties
+     * Used by FML to store mod list associated with a world, and maybe an id map
+     * Used by Forge to store the dimensions available to a world
+     * @param additionalProperties
+     */
+    public void setAdditionalProperties(java.util.Map<String,net.minecraft.nbt.NBTBase> additionalProperties)
+    {
+        // one time set for this
+        if (this.additionalProperties == null)
+        {
+            this.additionalProperties = additionalProperties;
+        }
+    }
+
+    public net.minecraft.nbt.NBTBase getAdditionalProperty(String additionalProperty)
+    {
+        return this.additionalProperties!=null? this.additionalProperties.get(additionalProperty) : null;
+    }
 }
