--- ../src_base/minecraft/net/minecraft/world/storage/WorldInfo.java
+++ ../src_work/minecraft/net/minecraft/world/storage/WorldInfo.java
@@ -151,6 +151,7 @@
         this.thunderTime = par1NBTTagCompound.getInteger("thunderTime");
         this.thundering = par1NBTTagCompound.getBoolean("thundering");
         this.hardcore = par1NBTTagCompound.getBoolean("hardcore");
+        this.dimension = par1NBTTagCompound.getInteger("dimension"); // MCPC+
 
         if (par1NBTTagCompound.hasKey("initialized"))
         {
@@ -227,6 +228,7 @@
         this.allowCommands = par1WorldInfo.allowCommands;
         this.initialized = par1WorldInfo.initialized;
         this.theGameRules = par1WorldInfo.theGameRules;
+        this.dimension = par1WorldInfo.dimension; // MCPC+
     }
 
     /**
@@ -274,6 +276,7 @@
         par1NBTTagCompound.setBoolean("allowCommands", this.allowCommands);
         par1NBTTagCompound.setBoolean("initialized", this.initialized);
         par1NBTTagCompound.setCompoundTag("GameRules", this.theGameRules.writeGameRulesToNBT());
+        par1NBTTagCompound.setInteger("dimension", this.dimension); // MCPC+
 
         if (par2NBTTagCompound != null)
         {
@@ -340,11 +343,17 @@
         return this.playerTag;
     }
 
-    /**
-     * Returns vanilla MC dimension (-1,0,1). For custom dimension compatibility, always prefer
-     * WorldProvider.dimensionID accessed from World.provider.dimensionID
-     */
-    public int getVanillaDimension()
+    // MCPC+ start
+    /**
+     * Sets the Dimension.
+     */
+    public void setDimension(int dim)
+    {
+        this.dimension = dim;
+    }
+    // MCPC+ end
+
+    public int getDimension()
     {
         return this.dimension;
     }
