--- a/net/minecraft/world/storage/WorldInfo.java
+++ b/net/minecraft/world/storage/WorldInfo.java
@@ -325,6 +325,7 @@
       NBTTagCompound nbttagcompound1 = new NBTTagCompound();
 
       for(Entry<DimensionType, NBTTagCompound> entry : this.field_186348_N.entrySet()) {
+         if (entry.getValue() == null || entry.getValue().isEmpty()) continue;
          nbttagcompound1.func_74782_a(String.valueOf(entry.getKey().func_186068_a()), entry.getValue());
       }
 
