--- a/net/minecraft/inventory/container/GrindstoneContainer.java
+++ b/net/minecraft/inventory/container/GrindstoneContainer.java
@@ -37,12 +_,12 @@
       this.field_217015_e = p_i50081_3_;
       this.func_75146_a(new Slot(this.field_217014_d, 0, 49, 19) {
          public boolean func_75214_a(ItemStack p_75214_1_) {
-            return p_75214_1_.func_77984_f() || p_75214_1_.func_77973_b() == Items.field_151134_bR || p_75214_1_.func_77948_v();
+            return true;
          }
       });
       this.func_75146_a(new Slot(this.field_217014_d, 1, 49, 40) {
          public boolean func_75214_a(ItemStack p_75214_1_) {
-            return p_75214_1_.func_77984_f() || p_75214_1_.func_77973_b() == Items.field_151134_bR || p_75214_1_.func_77948_v();
+            return true;
          }
       });
       this.func_75146_a(new Slot(this.field_217013_c, 2, 129, 34) {
@@ -111,61 +_,69 @@
       super.func_75130_a(p_75130_1_);
       if (p_75130_1_ == this.field_217014_d) {
          this.func_217010_e();
-      }
-
-   }
-
-   private void func_217010_e() {
-      ItemStack itemstack = this.field_217014_d.func_70301_a(0);
-      ItemStack itemstack1 = this.field_217014_d.func_70301_a(1);
-      boolean flag = !itemstack.func_190926_b() || !itemstack1.func_190926_b();
-      boolean flag1 = !itemstack.func_190926_b() && !itemstack1.func_190926_b();
-      if (!flag) {
-         this.field_217013_c.func_70299_a(0, ItemStack.field_190927_a);
-      } else {
-         boolean flag2 = !itemstack.func_190926_b() && itemstack.func_77973_b() != Items.field_151134_bR && !itemstack.func_77948_v() || !itemstack1.func_190926_b() && itemstack1.func_77973_b() != Items.field_151134_bR && !itemstack1.func_77948_v();
-         if (itemstack.func_190916_E() > 1 || itemstack1.func_190916_E() > 1 || !flag1 && flag2) {
-            this.field_217013_c.func_70299_a(0, ItemStack.field_190927_a);
-            this.func_75142_b();
-            return;
-         }
-
-         int j = 1;
-         int i;
-         ItemStack itemstack2;
-         if (flag1) {
-            if (itemstack.func_77973_b() != itemstack1.func_77973_b()) {
-               this.field_217013_c.func_70299_a(0, ItemStack.field_190927_a);
+         this.field_217015_e.func_221486_a((level, pos) -> {
+            java.util.Optional<net.minecraftforge.common.crafting.GrindstoneRecipe> optional = level.func_199532_z().func_215371_a(net.minecraftforge.common.ForgeMod.GRINDING, this.field_217014_d, level);
+            if (optional.isPresent()) {
+               net.minecraftforge.common.crafting.GrindstoneRecipe grindstoneRecipe = optional.get();
+               ItemStack result = grindstoneRecipe.func_77572_b(this.field_217013_c);
+               this.field_217013_c.func_70299_a(0, result);
                this.func_75142_b();
-               return;
-            }
-
-            Item item = itemstack.func_77973_b();
-            int k = item.func_77612_l() - itemstack.func_77952_i();
-            int l = item.func_77612_l() - itemstack1.func_77952_i();
-            int i1 = k + l + item.func_77612_l() * 5 / 100;
-            i = Math.max(item.func_77612_l() - i1, 0);
-            itemstack2 = this.func_217011_b(itemstack, itemstack1);
-            if (!itemstack2.func_77984_f()) {
-               if (!ItemStack.func_77989_b(itemstack, itemstack1)) {
-                  this.field_217013_c.func_70299_a(0, ItemStack.field_190927_a);
-                  this.func_75142_b();
-                  return;
-               }
-
-               j = 2;
-            }
-         } else {
-            boolean flag3 = !itemstack.func_190926_b();
-            i = flag3 ? itemstack.func_77952_i() : itemstack1.func_77952_i();
-            itemstack2 = flag3 ? itemstack : itemstack1;
-         }
-
-         this.field_217013_c.func_70299_a(0, this.func_217007_a(itemstack2, i, j));
+            }
+         });
       }
-
-      this.func_75142_b();
    }
+
+   private void func_217010_e() {
+       ItemStack itemstack = this.field_217014_d.func_70301_a(0);
+       ItemStack itemstack1 = this.field_217014_d.func_70301_a(1);
+       boolean flag = !itemstack.func_190926_b() || !itemstack1.func_190926_b();
+       boolean flag1 = !itemstack.func_190926_b() && !itemstack1.func_190926_b();
+       if (!flag) {
+          this.field_217013_c.func_70299_a(0, ItemStack.field_190927_a);
+       } else {
+          boolean flag2 = !itemstack.func_190926_b() && itemstack.func_77973_b() != Items.field_151134_bR && !itemstack.func_77948_v() || !itemstack1.func_190926_b() && itemstack1.func_77973_b() != Items.field_151134_bR && !itemstack1.func_77948_v();
+          if (itemstack.func_190916_E() > 1 || itemstack1.func_190916_E() > 1 || !flag1 && flag2) {
+             this.field_217013_c.func_70299_a(0, ItemStack.field_190927_a);
+             this.func_75142_b();
+             return;
+          }
+
+          int j = 1;
+          int i;
+          ItemStack itemstack2;
+          if (flag1) {
+             if (itemstack.func_77973_b() != itemstack1.func_77973_b()) {
+                this.field_217013_c.func_70299_a(0, ItemStack.field_190927_a);
+                this.func_75142_b();
+                return;
+             }
+
+             Item item = itemstack.func_77973_b();
+             int k = item.func_77612_l() - itemstack.func_77952_i();
+             int l = item.func_77612_l() - itemstack1.func_77952_i();
+             int i1 = k + l + item.func_77612_l() * 5 / 100;
+             i = Math.max(item.func_77612_l() - i1, 0);
+             itemstack2 = this.func_217011_b(itemstack, itemstack1);
+             if (!itemstack2.func_77984_f()) {
+                if (!ItemStack.func_77989_b(itemstack, itemstack1)) {
+                   this.field_217013_c.func_70299_a(0, ItemStack.field_190927_a);
+                   this.func_75142_b();
+                   return;
+                }
+
+                j = 2;
+             }
+          } else {
+             boolean flag3 = !itemstack.func_190926_b();
+             i = flag3 ? itemstack.func_77952_i() : itemstack1.func_77952_i();
+             itemstack2 = flag3 ? itemstack : itemstack1;
+          }
+
+          this.field_217013_c.func_70299_a(0, this.func_217007_a(itemstack2, i, j));
+       }
+
+       this.func_75142_b();
+    }
 
    private ItemStack func_217011_b(ItemStack p_217011_1_, ItemStack p_217011_2_) {
       ItemStack itemstack = p_217011_1_.func_77946_l();
