--- a/net/minecraft/inventory/container/RecipeBookContainer.java
+++ b/net/minecraft/inventory/container/RecipeBookContainer.java
@@ -31,4 +31,10 @@
 
    @OnlyIn(Dist.CLIENT)
    public abstract int func_203721_h();
+
+   public abstract net.minecraft.item.crafting.IRecipeType<?> getRecipeType();
+
+   public java.util.List<net.minecraft.client.util.RecipeBookCategories> getRecipeBookCategories() {
+      return net.minecraftforge.registries.ForgeRegistries.RECIPE_BOOK_CATEGORIES.getValues().stream().filter(category -> category.getRecipeType().equals(this.getRecipeType())).sorted(net.minecraft.client.util.RecipeBookCategories::compareTo).collect(java.util.stream.Collectors.toList());
+   }
 }
