--- a/net/minecraft/inventory/container/Slot.java
+++ b/net/minecraft/inventory/container/Slot.java
@@ -13,6 +13,9 @@
    public int field_75222_d;
    public int field_75223_e;
    public int field_75221_f;
+   protected String backgroundName = null;
+   protected net.minecraft.util.ResourceLocation backgroundLocation = null;
+   protected Object backgroundMap;
 
    public Slot(IInventory p_i1824_1_, int p_i1824_2_, int p_i1824_3_, int p_i1824_4_) {
       this.field_75224_c = p_i1824_1_;
@@ -75,7 +78,7 @@
    @Nullable
    @OnlyIn(Dist.CLIENT)
    public String func_178171_c() {
-      return null;
+      return backgroundName;
    }
 
    public ItemStack func_75209_a(int p_75209_1_) {
@@ -90,4 +93,62 @@
    public boolean func_111238_b() {
       return true;
    }
+
+   /**
+    * Gets the path of the texture file to use for the background image of this slot when drawing the GUI.
+    * @return The resource location for the background image
+    */
+   @OnlyIn(Dist.CLIENT)
+   public net.minecraft.util.ResourceLocation getBackgroundLocation() {
+       return (backgroundLocation == null ? net.minecraft.client.renderer.texture.AtlasTexture.field_110575_b : backgroundLocation);
+   }
+
+   /**
+    * Sets the texture file to use for the background image of the slot when it's empty.
+    * @param texture the resourcelocation for the texture
+    */
+   @OnlyIn(Dist.CLIENT)
+   public void setBackgroundLocation(net.minecraft.util.ResourceLocation texture) {
+       this.backgroundLocation = texture;
+   }
+
+   /**
+    * Sets which icon index to use as the background image of the slot when it's empty.
+    * @param name The icon to use, null for none
+    */
+   public void setBackgroundName(@Nullable String name) {
+       this.backgroundName = name;
+   }
+
+   @Nullable
+   @OnlyIn(Dist.CLIENT)
+   public net.minecraft.client.renderer.texture.TextureAtlasSprite getBackgroundSprite() {
+       String name = func_178171_c();
+       return name == null ? null : getBackgroundMap().func_110572_b(name);
+   }
+
+   @OnlyIn(Dist.CLIENT)
+   protected net.minecraft.client.renderer.texture.AtlasTexture getBackgroundMap() {
+      if (backgroundMap == null) backgroundMap = net.minecraft.client.Minecraft.func_71410_x().func_147117_R();
+      return (net.minecraft.client.renderer.texture.AtlasTexture)backgroundMap;
+   }
+
+   /**
+    * Retrieves the index in the inventory for this slot, this value should typically not
+    * be used, but can be useful for some occasions.
+    *
+    * @return Index in associated inventory for this slot.
+    */
+   public int getSlotIndex() {
+      return field_75225_a;
+   }
+
+   /**
+    * Checks if the other slot is in the same inventory, by comparing the inventory reference.
+    * @param other
+    * @return true if the other slot is in the same inventory
+    */
+   public boolean isSameInventory(Slot other) {
+      return this.field_75224_c == other.field_75224_c;
+   }
 }
