--- a/net/minecraft/world/dimension/EndDimension.java
+++ b/net/minecraft/world/dimension/EndDimension.java
@@ -20,9 +20,14 @@
 public class EndDimension extends Dimension {
    public static final BlockPos field_209958_g = new BlockPos(100, 50, 0);
    private DragonFightManager field_186064_g;
+   private final DimensionType type;
+   public EndDimension() { this(DimensionType.THE_END); }
+   public EndDimension(DimensionType type) {
+      this.type = type;
+   }
 
    public void func_76572_b() {
-      NBTTagCompound nbttagcompound = this.field_76579_a.func_72912_H().func_186347_a(DimensionType.THE_END);
+      NBTTagCompound nbttagcompound = this.field_76579_a.func_72912_H().func_186347_a(this.field_76579_a.field_73011_w.func_186058_p());
       this.field_186064_g = this.field_76579_a instanceof WorldServer ? new DragonFightManager((WorldServer)this.field_76579_a, nbttagcompound.func_74775_l("DragonFight")) : null;
       this.field_191067_f = false;
    }
@@ -99,7 +104,7 @@
    }
 
    public DimensionType func_186058_p() {
-      return DimensionType.THE_END;
+      return type;
    }
 
    public void func_186057_q() {
@@ -108,7 +113,7 @@
          nbttagcompound.func_74782_a("DragonFight", this.field_186064_g.func_186088_a());
       }
 
-      this.field_76579_a.func_72912_H().func_186345_a(DimensionType.THE_END, nbttagcompound);
+      this.field_76579_a.func_72912_H().func_186345_a(this.field_76579_a.field_73011_w.func_186058_p(), nbttagcompound);
    }
 
    public void func_186059_r() {
@@ -122,4 +127,18 @@
    public DragonFightManager func_186063_s() {
       return this.field_186064_g;
    }
+
+   @Override
+   public void func_186061_a(net.minecraft.entity.player.EntityPlayerMP player) {
+      if (field_186064_g != null) {
+         field_186064_g.addPlayer(player);
+      }
+   }
+
+   @Override
+   public void func_186062_b(net.minecraft.entity.player.EntityPlayerMP player) {
+      if (field_186064_g != null) {
+         field_186064_g.removePlayer(player);
+      }
+   }
 }
