--- a/net/minecraft/world/dimension/EndDimension.java
+++ b/net/minecraft/world/dimension/EndDimension.java
@@ -24,8 +24,8 @@
 
    public EndDimension(World p_i49932_1_, DimensionType p_i49932_2_) {
       super(p_i49932_1_, p_i49932_2_);
-      CompoundNBT compoundnbt = p_i49932_1_.func_72912_H().func_186347_a(DimensionType.field_223229_c_);
-      this.field_186064_g = p_i49932_1_ instanceof ServerWorld ? new DragonFightManager((ServerWorld)p_i49932_1_, compoundnbt.func_74775_l("DragonFight")) : null;
+      CompoundNBT compoundnbt = p_i49932_1_.func_72912_H().func_186347_a(p_i49932_2_);
+      this.field_186064_g = p_i49932_1_ instanceof ServerWorld ? new DragonFightManager((ServerWorld)p_i49932_1_, compoundnbt.func_74775_l("DragonFight"), this) : null;
    }
 
    public ChunkGenerator<?> func_186060_c() {
@@ -99,17 +99,13 @@
       return false;
    }
 
-   public DimensionType func_186058_p() {
-      return DimensionType.field_223229_c_;
-   }
-
    public void func_186057_q() {
       CompoundNBT compoundnbt = new CompoundNBT();
       if (this.field_186064_g != null) {
          compoundnbt.func_218657_a("DragonFight", this.field_186064_g.func_186088_a());
       }
 
-      this.field_76579_a.func_72912_H().func_186345_a(DimensionType.field_223229_c_, compoundnbt);
+      this.field_76579_a.func_72912_H().func_186345_a(this.field_76579_a.func_201675_m().func_186058_p(), compoundnbt);
    }
 
    public void func_186059_r() {
