--- ../src-base/minecraft/net/minecraft/world/storage/loot/functions/LootingEnchantBonus.java
+++ ../src-work/minecraft/net/minecraft/world/storage/loot/functions/LootingEnchantBonus.java
@@ -32,7 +32,7 @@
 
         if (entity instanceof EntityLivingBase)
         {
-            int i = EnchantmentHelper.func_185283_h((EntityLivingBase)entity);
+            int i = p_186553_3_.getLootingModifier();
 
             if (i == 0)
             {
