--- a/net/minecraft/world/storage/loot/functions/Smelt.java
+++ b/net/minecraft/world/storage/loot/functions/Smelt.java
@@ -42,8 +42,8 @@
 
    @Nullable
    public static IRecipe func_202880_a(LootContext p_202880_0_, ItemStack p_202880_1_) {
-      for(IRecipe irecipe : p_202880_0_.func_202879_g().func_199532_z().func_199510_b()) {
-         if (irecipe instanceof FurnaceRecipe && irecipe.func_192400_c().get(0).test(p_202880_1_)) {
+      for(IRecipe irecipe : p_202880_0_.func_202879_g().func_199532_z().getRecipes(net.minecraftforge.common.crafting.VanillaRecipeTypes.SMELTING)) {
+         if (irecipe.func_192400_c().get(0).test(p_202880_1_)) {
             return irecipe;
          }
       }
