--- a/net/minecraft/fluid/FlowingFluid.java
+++ b/net/minecraft/fluid/FlowingFluid.java
@@ -159,7 +_,7 @@
          BlockState blockstate = p_205576_1_.getBlockState(blockpos);
          FluidState fluidstate = blockstate.getFluidState();
          if (fluidstate.getType().isSame(this) && this.canPassThroughWall(direction, p_205576_1_, p_205576_2_, p_205576_3_, blockpos, blockstate)) {
-            if (fluidstate.isSource()) {
+            if (fluidstate.isSource() && net.minecraftforge.event.ForgeEventFactory.canCreateFluidSource(p_205576_1_, blockpos, blockstate, this.canConvertToSource())) {
                ++j;
             }
 
@@ -167,7 +_,7 @@
          }
       }
 
-      if (this.canConvertToSource() && j >= 2) {
+      if (j >= 2) {
          BlockState blockstate1 = p_205576_1_.getBlockState(p_205576_2_.below());
          FluidState fluidstate1 = blockstate1.getFluidState();
          if (blockstate1.getMaterial().isSolid() || this.isSourceBlockOfThisType(fluidstate1)) {
