--- a/net/minecraft/fluid/Fluid.java
+++ b/net/minecraft/fluid/Fluid.java
@@ -19,7 +19,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public abstract class Fluid {
+public abstract class Fluid extends net.minecraftforge.registries.ForgeRegistryEntry<Fluid> implements net.minecraftforge.common.extensions.IForgeFluid {
    public static final ObjectIntIdentityMap<IFluidState> field_207201_d = new ObjectIntIdentityMap<>();
    protected final StateContainer<Fluid, IFluidState> field_207202_e;
    private IFluidState field_207200_b;
@@ -102,4 +102,10 @@
    }
 
    public abstract VoxelShape func_215664_b(IFluidState p_215664_1_, IBlockReader p_215664_2_, BlockPos p_215664_3_);
+
+   private final net.minecraftforge.common.util.ReverseTagWrapper<Fluid> reverseTags = new net.minecraftforge.common.util.ReverseTagWrapper<>(this, net.minecraft.tags.FluidTags::getGeneration, net.minecraft.tags.FluidTags::getCollection);
+   @Override
+   public java.util.Set<net.minecraft.util.ResourceLocation> getTags() {
+      return reverseTags.getTagNames();
+   }
 }
