--- a/net/minecraft/fluid/Fluid.java
+++ b/net/minecraft/fluid/Fluid.java
@@ -19,7 +19,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public abstract class Fluid {
+public abstract class Fluid extends net.minecraftforge.registries.ForgeRegistryEntry<Fluid> implements net.minecraftforge.common.extensions.IForgeFluid {
    public static final ObjectIntIdentityMap<IFluidState> field_207201_d = new ObjectIntIdentityMap<>();
    protected final StateContainer<Fluid, IFluidState> field_207202_e;
    private IFluidState field_207200_b;
@@ -102,4 +102,45 @@
    }
 
    public abstract VoxelShape func_215664_b(IFluidState p_215664_1_, IBlockReader p_215664_2_, BlockPos p_215664_3_);
+
+   private final net.minecraftforge.common.util.ReverseTagWrapper<Fluid> reverseTags = new net.minecraftforge.common.util.ReverseTagWrapper<>(this, net.minecraft.tags.FluidTags::getGeneration, net.minecraft.tags.FluidTags::getCollection);
+   @Override
+   public java.util.Set<net.minecraft.util.ResourceLocation> getTags() {
+      return reverseTags.getTagNames();
+   }
+
+   /**
+    * Creates the fluid attributes object, which will contain all the extended values for the fluid that aren't part of the vanilla system.
+    * Do not call this from outside. To retrieve the values use {@link Fluid#getAttributes()}
+    */
+   protected net.minecraftforge.fluids.FluidAttributes createAttributes()
+   {
+      if (this instanceof EmptyFluid)
+         return net.minecraftforge.fluids.FluidAttributes.builder(this,
+                 new net.minecraft.util.ResourceLocation("white"),
+                 new net.minecraft.util.ResourceLocation("white"))
+                 .translationKey("block.minecraft.air")
+                 .color(0).density(0).temperature(0).luminosity(0).viscosity(0).density(0).build();
+      if (this instanceof WaterFluid)
+         return net.minecraftforge.fluids.FluidAttributes.builder(this,
+                 new net.minecraft.util.ResourceLocation("block/water_still"),
+                 new net.minecraft.util.ResourceLocation("block/water_flow"))
+                 .overlay(new net.minecraft.util.ResourceLocation("block/water_overlay"))
+                 .translationKey("block.minecraft.water")
+                 .color(0x3F76E4).useBiomeWaterColor().build();
+      if (this instanceof LavaFluid)
+         return net.minecraftforge.fluids.FluidAttributes.builder(this,
+                 new net.minecraft.util.ResourceLocation("block/lava_still"),
+                 new net.minecraft.util.ResourceLocation("block/lava_flow"))
+                 .translationKey("block.minecraft.lava")
+                 .luminosity(15).density(3000).viscosity(6000).temperature(1300).build();
+      throw new RuntimeException("Mod fluids must override createAttributes.");
+   }
+
+   private net.minecraftforge.fluids.FluidAttributes forgeFluidAttributes;
+   public final net.minecraftforge.fluids.FluidAttributes getAttributes() {
+      if (forgeFluidAttributes == null)
+         forgeFluidAttributes = createAttributes();
+      return forgeFluidAttributes;
+   }
 }
