--- a/net/minecraft/fluid/IFluidState.java
+++ b/net/minecraft/fluid/IFluidState.java
@@ -29,7 +29,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public interface IFluidState extends IStateHolder<IFluidState> {
+public interface IFluidState extends IStateHolder<IFluidState>, net.minecraftforge.common.extensions.IForgeFluidState {
    Fluid func_206886_c();
 
    default boolean func_206889_d() {
@@ -107,6 +107,7 @@
       return this.func_206886_c().func_207185_a(p_206884_1_);
    }
 
+   @Deprecated //Forge: Use more sensitive version.
    default float func_210200_l() {
       return this.func_206886_c().func_210195_d();
    }
