--- a/net/minecraft/fluid/WaterFluid.java
+++ b/net/minecraft/fluid/WaterFluid.java
@@ -60,7 +60,7 @@
    }
 
    protected void func_205580_a(IWorld p_205580_1_, BlockPos p_205580_2_, BlockState p_205580_3_) {
-      TileEntity tileentity = p_205580_3_.func_177230_c().func_235695_q_() ? p_205580_1_.func_175625_s(p_205580_2_) : null;
+      TileEntity tileentity = p_205580_3_.hasTileEntity() ? p_205580_1_.func_175625_s(p_205580_2_) : null;
       Block.func_220059_a(p_205580_3_, p_205580_1_, p_205580_2_, tileentity);
    }
 
