--- a/net/minecraft/world/gen/feature/AbstractTreeFeature.java
+++ b/net/minecraft/world/gen/feature/AbstractTreeFeature.java
@@ -34,6 +34,8 @@
    }
 
    protected static boolean func_214587_a(IWorldGenerationBaseReader p_214587_0_, BlockPos p_214587_1_) {
+      if (p_214587_0_ instanceof net.minecraft.world.IWorldReader) // FORGE: Redirect to state method when possible
+         return p_214587_0_.func_217375_a(p_214587_1_, state -> state.canBeReplacedByLogs((net.minecraft.world.IWorldReader)p_214587_0_, p_214587_1_));
       return p_214587_0_.func_217375_a(p_214587_1_, (p_214573_0_) -> {
          Block block = p_214573_0_.func_177230_c();
          return p_214573_0_.func_196958_f() || p_214573_0_.func_203425_a(BlockTags.field_206952_E) || func_227250_b_(block) || block.func_203417_a(BlockTags.field_200031_h) || block.func_203417_a(BlockTags.field_200030_g) || block == Blocks.field_150395_bd;
@@ -41,6 +43,8 @@
    }
 
    public static boolean func_214574_b(IWorldGenerationBaseReader p_214574_0_, BlockPos p_214574_1_) {
+      if (p_214574_0_ instanceof net.minecraft.world.IBlockReader) // FORGE: Redirect to state method when possible
+        return p_214574_0_.func_217375_a(p_214574_1_, state -> state.isAir((net.minecraft.world.IBlockReader)p_214574_0_, p_214574_1_));
       return p_214574_0_.func_217375_a(p_214574_1_, BlockState::func_196958_f);
    }
 
@@ -64,17 +68,27 @@
    }
 
    public static boolean func_214572_g(IWorldGenerationBaseReader p_214572_0_, BlockPos p_214572_1_) {
+      if (p_214572_0_ instanceof net.minecraft.world.IWorldReader) // FORGE: Redirect to state method when possible
+         return p_214572_0_.func_217375_a(p_214572_1_, state -> state.canBeReplacedByLeaves((net.minecraft.world.IWorldReader)p_214572_0_, p_214572_1_));
       return p_214572_0_.func_217375_a(p_214572_1_, (p_227223_0_) -> {
          return p_227223_0_.func_196958_f() || p_227223_0_.func_203425_a(BlockTags.field_206952_E);
       });
    }
 
+   @Deprecated //Forge: moved to isSoil
    public static boolean func_214589_h(IWorldGenerationBaseReader p_214589_0_, BlockPos p_214589_1_) {
       return p_214589_0_.func_217375_a(p_214589_1_, (p_227221_0_) -> {
          return func_227250_b_(p_227221_0_.func_177230_c());
       });
    }
 
+   protected static boolean isSoil(IWorldGenerationBaseReader reader, BlockPos pos, net.minecraftforge.common.IPlantable sapling) {
+      if (!(reader instanceof net.minecraft.world.IBlockReader) || sapling == null)
+         return func_214589_h(reader, pos);
+      return reader.func_217375_a(pos, state -> state.canSustainPlant((net.minecraft.world.IBlockReader)reader, pos, Direction.UP, sapling));
+   }
+
+   @Deprecated //Forge: moved to isSoilOrFarm
    protected static boolean func_214585_i(IWorldGenerationBaseReader p_214585_0_, BlockPos p_214585_1_) {
       return p_214585_0_.func_217375_a(p_214585_1_, (p_227220_0_) -> {
          Block block = p_227220_0_.func_177230_c();
@@ -82,6 +96,12 @@
       });
    }
 
+   protected static boolean isSoilOrFarm(IWorldGenerationBaseReader reader, BlockPos pos, net.minecraftforge.common.IPlantable sapling) {
+      if (!(reader instanceof net.minecraft.world.IBlockReader) || sapling == null)
+         return func_214585_i(reader, pos);
+      return reader.func_217375_a(pos, state -> state.canSustainPlant((net.minecraft.world.IBlockReader)reader, pos, Direction.UP, sapling));
+   }
+
    public static boolean func_214576_j(IWorldGenerationBaseReader p_214576_0_, BlockPos p_214576_1_) {
       return p_214576_0_.func_217375_a(p_214576_1_, (p_227218_0_) -> {
          Material material = p_227218_0_.func_185904_a();
@@ -89,6 +109,7 @@
       });
    }
 
+   @Deprecated //Forge: moved to setDirtAt
    protected void func_214584_a(IWorldGenerationReader p_214584_1_, BlockPos p_214584_2_) {
       if (!func_214578_c(p_214584_1_, p_214584_2_)) {
          this.func_202278_a(p_214584_1_, p_214584_2_, Blocks.field_150346_d.func_176223_P());
@@ -116,6 +137,13 @@
       }
    }
 
+   protected void setDirtAt(IWorldGenerationReader reader, BlockPos pos, BlockPos origin) {
+      if (!(reader instanceof IWorld)) {
+         func_214584_a(reader, pos);
+         return;
+      }
+      ((IWorld)reader).func_180495_p(pos).onPlantGrow((IWorld)reader, pos, origin);
+   }
    protected void func_202278_a(IWorldWriter p_202278_1_, BlockPos p_202278_2_, BlockState p_202278_3_) {
       this.func_208521_b(p_202278_1_, p_202278_2_, p_202278_3_);
    }
