--- a/net/minecraft/world/gen/feature/AbstractBigMushroomFeature.java
+++ b/net/minecraft/world/gen/feature/AbstractBigMushroomFeature.java
@@ -19,7 +19,7 @@
    protected void func_227210_a_(IWorld p_227210_1_, Random p_227210_2_, BlockPos p_227210_3_, BigMushroomFeatureConfig p_227210_4_, int p_227210_5_, BlockPos.Mutable p_227210_6_) {
       for(int i = 0; i < p_227210_5_; ++i) {
          p_227210_6_.func_189533_g(p_227210_3_).func_189534_c(Direction.UP, i);
-         if (!p_227210_1_.func_180495_p(p_227210_6_).func_200015_d(p_227210_1_, p_227210_6_)) {
+         if (p_227210_1_.func_180495_p(p_227210_6_).canBeReplacedByLogs(p_227210_1_, p_227210_6_)) {
             this.func_230367_a_(p_227210_1_, p_227210_6_, p_227210_4_.field_227273_b_.func_225574_a_(p_227210_2_, p_227210_3_));
          }
       }
@@ -48,7 +48,7 @@
                for(int l = -k; l <= k; ++l) {
                   for(int i1 = -k; i1 <= k; ++i1) {
                      BlockState blockstate = p_227209_1_.func_180495_p(p_227209_4_.func_239621_a_(p_227209_2_, l, j, i1));
-                     if (!blockstate.func_196958_f() && !blockstate.func_235714_a_(BlockTags.field_206952_E)) {
+                     if (!blockstate.isAir(p_227209_1_, p_227209_4_.func_239621_a_(p_227209_2_, l, j, i1)) && !blockstate.func_235714_a_(BlockTags.field_206952_E)) {
                         return false;
                      }
                   }
