--- a/net/minecraft/world/gen/feature/BaseTreeFeatureConfig.java
+++ b/net/minecraft/world/gen/feature/BaseTreeFeatureConfig.java
@@ -42,6 +42,7 @@
    public final int field_236680_i_;
    public final boolean field_236681_j_;
    public final Heightmap.Type field_236682_l_;
+   protected net.minecraftforge.common.IPlantable sapling = (net.minecraftforge.common.IPlantable)net.minecraft.block.Blocks.field_196674_t;
 
    protected BaseTreeFeatureConfig(BlockStateProvider p_i232020_1_, BlockStateProvider p_i232020_2_, FoliagePlacer p_i232020_3_, AbstractTrunkPlacer p_i232020_4_, AbstractFeatureSizeType p_i232020_5_, List<TreeDecorator> p_i232020_6_, int p_i232020_7_, boolean p_i232020_8_, Heightmap.Type p_i232020_9_) {
       this.field_227368_m_ = p_i232020_1_;
@@ -55,6 +56,15 @@
       this.field_236682_l_ = p_i232020_9_;
    }
 
+   protected BaseTreeFeatureConfig setSapling(net.minecraftforge.common.IPlantable value) {
+      this.sapling = value;
+      return this;
+   }
+
+   public net.minecraftforge.common.IPlantable getSapling() {
+      return this.sapling;
+   }
+
    public void func_227373_a_() {
       this.field_227372_q_ = true;
    }
