--- a/net/minecraft/world/gen/feature/BigRedMushroomFeature.java
+++ b/net/minecraft/world/gen/feature/BigRedMushroomFeature.java
@@ -26,7 +26,7 @@
                boolean flag5 = flag2 || flag3;
                if (i >= p_225564_4_ || flag4 != flag5) {
                   p_225564_5_.func_239621_a_(p_225564_3_, l, i, i1);
-                  if (!p_225564_1_.func_180495_p(p_225564_5_).func_200015_d(p_225564_1_, p_225564_5_)) {
+                  if (p_225564_1_.func_180495_p(p_225564_5_).canBeReplacedByLeaves(p_225564_1_, p_225564_5_)) {
                      this.func_230367_a_(p_225564_1_, p_225564_5_, p_225564_6_.field_227272_a_.func_225574_a_(p_225564_2_, p_225564_3_).func_206870_a(HugeMushroomBlock.field_196465_z, Boolean.valueOf(i >= p_225564_4_ - 1)).func_206870_a(HugeMushroomBlock.field_196464_y, Boolean.valueOf(l < -k)).func_206870_a(HugeMushroomBlock.field_196461_b, Boolean.valueOf(l > k)).func_206870_a(HugeMushroomBlock.field_196459_a, Boolean.valueOf(i1 < -k)).func_206870_a(HugeMushroomBlock.field_196463_c, Boolean.valueOf(i1 > k)));
                   }
                }
