--- a/net/minecraft/world/gen/feature/BushFeature.java
+++ b/net/minecraft/world/gen/feature/BushFeature.java
@@ -20,7 +20,7 @@
 
       for(int j = 0; j < 64; ++j) {
          BlockPos blockpos = p_212245_4_.func_177982_a(p_212245_3_.nextInt(8) - p_212245_3_.nextInt(8), p_212245_3_.nextInt(4) - p_212245_3_.nextInt(4), p_212245_3_.nextInt(8) - p_212245_3_.nextInt(8));
-         if (p_212245_1_.func_175623_d(blockpos) && (!p_212245_1_.func_201675_m().func_177495_o() || blockpos.func_177956_o() < 255) && blockstate.func_196955_c(p_212245_1_, blockpos)) {
+         if (p_212245_1_.func_175623_d(blockpos) && (!p_212245_1_.func_201675_m().func_177495_o() || blockpos.func_177956_o() < p_212245_1_.func_201672_e().func_201675_m().getHeight()) && blockstate.func_196955_c(p_212245_1_, blockpos)) {
             p_212245_1_.func_180501_a(blockpos, blockstate, 2);
             ++i;
          }
