--- a/net/minecraft/world/gen/feature/DeadBushFeature.java
+++ b/net/minecraft/world/gen/feature/DeadBushFeature.java
@@ -20,7 +20,7 @@
    }
 
    public boolean func_212245_a(IWorld p_212245_1_, ChunkGenerator<? extends GenerationSettings> p_212245_2_, Random p_212245_3_, BlockPos p_212245_4_, NoFeatureConfig p_212245_5_) {
-      for(BlockState blockstate = p_212245_1_.func_180495_p(p_212245_4_); (blockstate.func_196958_f() || blockstate.func_203425_a(BlockTags.field_206952_E)) && p_212245_4_.func_177956_o() > 0; blockstate = p_212245_1_.func_180495_p(p_212245_4_)) {
+      for(BlockState blockstate = p_212245_1_.func_180495_p(p_212245_4_); (blockstate.isAir(p_212245_1_, p_212245_4_) || blockstate.func_203425_a(BlockTags.field_206952_E)) && p_212245_4_.func_177956_o() > 0; blockstate = p_212245_1_.func_180495_p(p_212245_4_)) {
          p_212245_4_ = p_212245_4_.func_177977_b();
       }
 
