--- a/net/minecraft/world/gen/feature/HugeTreesFeature.java
+++ b/net/minecraft/world/gen/feature/HugeTreesFeature.java
@@ -22,6 +22,7 @@
       this.field_150538_d = p_i51481_4_;
       this.field_76520_b = p_i51481_5_;
       this.field_76521_c = p_i51481_6_;
+      setSapling((net.minecraftforge.common.IPlantable)net.minecraft.block.Blocks.field_196674_t);
    }
 
    protected int func_150533_a(Random p_150533_1_) {
@@ -35,7 +36,7 @@
 
    private boolean func_175926_c(IWorldGenerationBaseReader p_175926_1_, BlockPos p_175926_2_, int p_175926_3_) {
       boolean flag = true;
-      if (p_175926_2_.func_177956_o() >= 1 && p_175926_2_.func_177956_o() + p_175926_3_ + 1 <= 256) {
+      if (p_175926_2_.func_177956_o() >= 1 && p_175926_2_.func_177956_o() + p_175926_3_ + 1 <= p_175926_1_.getMaxHeight()) {
          for(int i = 0; i <= 1 + p_175926_3_; ++i) {
             int j = 2;
             if (i == 0) {
@@ -46,7 +47,7 @@
 
             for(int k = -j; k <= j && flag; ++k) {
                for(int l = -j; l <= j && flag; ++l) {
-                  if (p_175926_2_.func_177956_o() + i < 0 || p_175926_2_.func_177956_o() + i >= 256 || !func_214587_a(p_175926_1_, p_175926_2_.func_177982_a(k, i, l))) {
+                  if (p_175926_2_.func_177956_o() + i < 0 || p_175926_2_.func_177956_o() + i >= p_175926_1_.getMaxHeight() || !func_214587_a(p_175926_1_, p_175926_2_.func_177982_a(k, i, l))) {
                      flag = false;
                   }
                }
@@ -61,11 +62,12 @@
 
    private boolean func_202405_b(IWorldGenerationReader p_202405_1_, BlockPos p_202405_2_) {
       BlockPos blockpos = p_202405_2_.func_177977_b();
-      if (func_214589_h(p_202405_1_, blockpos) && p_202405_2_.func_177956_o() >= 2) {
-         this.func_214584_a(p_202405_1_, blockpos);
-         this.func_214584_a(p_202405_1_, blockpos.func_177974_f());
-         this.func_214584_a(p_202405_1_, blockpos.func_177968_d());
-         this.func_214584_a(p_202405_1_, blockpos.func_177968_d().func_177974_f());
+
+      if (isSoil(p_202405_1_, blockpos, getSapling()) && p_202405_2_.func_177956_o() >= 2) {
+         setDirtAt(p_202405_1_, blockpos, p_202405_2_);
+         setDirtAt(p_202405_1_, blockpos.func_177974_f(), p_202405_2_);
+         setDirtAt(p_202405_1_, blockpos.func_177968_d(), p_202405_2_);
+         setDirtAt(p_202405_1_, blockpos.func_177968_d().func_177974_f(), p_202405_2_);
          return true;
       } else {
          return false;
