--- a/net/minecraft/world/gen/feature/IceSpikeFeature.java
+++ b/net/minecraft/world/gen/feature/IceSpikeFeature.java
@@ -42,14 +_,14 @@
                   if ((i1 == 0 && j1 == 0 || !(f1 * f1 + f2 * f2 > f * f)) && (i1 != -l && i1 != l && j1 != -l && j1 != l || !(p_241855_3_.nextFloat() > 0.75F))) {
                      BlockState blockstate = p_241855_1_.func_180495_p(p_241855_4_.func_177982_a(i1, k, j1));
                      Block block = blockstate.func_177230_c();
-                     if (blockstate.func_196958_f() || func_227250_b_(block) || block == Blocks.field_196604_cC || block == Blocks.field_150432_aD) {
+                     if (blockstate.isAir(p_241855_1_, p_241855_4_.func_177982_a(i1, k, j1)) || func_227250_b_(block) || block == Blocks.field_196604_cC || block == Blocks.field_150432_aD) {
                         this.func_230367_a_(p_241855_1_, p_241855_4_.func_177982_a(i1, k, j1), Blocks.field_150403_cj.func_176223_P());
                      }
 
                      if (k != 0 && l > 1) {
                         blockstate = p_241855_1_.func_180495_p(p_241855_4_.func_177982_a(i1, -k, j1));
                         block = blockstate.func_177230_c();
-                        if (blockstate.func_196958_f() || func_227250_b_(block) || block == Blocks.field_196604_cC || block == Blocks.field_150432_aD) {
+                        if (blockstate.isAir(p_241855_1_, p_241855_4_.func_177982_a(i1, -k, j1)) || func_227250_b_(block) || block == Blocks.field_196604_cC || block == Blocks.field_150432_aD) {
                            this.func_230367_a_(p_241855_1_, p_241855_4_.func_177982_a(i1, -k, j1), Blocks.field_150403_cj.func_176223_P());
                         }
                      }
@@ -76,7 +_,7 @@
                while(blockpos.func_177956_o() > 50) {
                   BlockState blockstate1 = p_241855_1_.func_180495_p(blockpos);
                   Block block1 = blockstate1.func_177230_c();
-                  if (!blockstate1.func_196958_f() && !func_227250_b_(block1) && block1 != Blocks.field_196604_cC && block1 != Blocks.field_150432_aD && block1 != Blocks.field_150403_cj) {
+                  if (!blockstate1.isAir(p_241855_1_, blockpos) && !func_227250_b_(block1) && block1 != Blocks.field_196604_cC && block1 != Blocks.field_150432_aD && block1 != Blocks.field_150403_cj) {
                      break;
                   }
 
