--- a/net/minecraft/world/gen/feature/MegaPineTree.java
+++ b/net/minecraft/world/gen/feature/MegaPineTree.java
@@ -20,6 +20,7 @@
    public MegaPineTree(Function<Dynamic<?>, ? extends NoFeatureConfig> p_i51482_1_, boolean p_i51482_2_, boolean p_i51482_3_) {
       super(p_i51482_1_, p_i51482_2_, 13, 15, field_181633_e, field_181634_f);
       this.field_150542_e = p_i51482_3_;
+      setSapling((net.minecraftforge.common.IPlantable)Blocks.field_196675_u);
    }
 
    public boolean func_208519_a(Set<BlockPos> p_208519_1_, IWorldGenerationReader p_208519_2_, Random p_208519_3_, BlockPos p_208519_4_, MutableBoundingBox p_208519_5_) {
@@ -98,7 +99,7 @@
    private void func_175934_c(IWorldGenerationReader p_175934_1_, BlockPos p_175934_2_) {
       for(int i = 2; i >= -3; --i) {
          BlockPos blockpos = p_175934_2_.func_177981_b(i);
-         if (func_214589_h(p_175934_1_, blockpos)) {
+         if (isSoil(p_175934_1_, blockpos, getSapling())) {
             this.func_202278_a(p_175934_1_, blockpos, field_181635_g);
             break;
          }
