--- a/net/minecraft/world/gen/feature/MegaJungleFeature.java
+++ b/net/minecraft/world/gen/feature/MegaJungleFeature.java
@@ -16,6 +16,7 @@
 public class MegaJungleFeature extends HugeTreesFeature<NoFeatureConfig> {
    public MegaJungleFeature(Function<Dynamic<?>, ? extends NoFeatureConfig> p_i51483_1_, boolean p_i51483_2_, int p_i51483_3_, int p_i51483_4_, BlockState p_i51483_5_, BlockState p_i51483_6_) {
       super(p_i51483_1_, p_i51483_2_, p_i51483_3_, p_i51483_4_, p_i51483_5_, p_i51483_6_);
+      setSapling((net.minecraftforge.common.IPlantable)Blocks.field_196678_w);
    }
 
    public boolean func_208519_a(Set<BlockPos> p_208519_1_, IWorldGenerationReader p_208519_2_, Random p_208519_3_, BlockPos p_208519_4_, MutableBoundingBox p_208519_5_) {
