--- a/net/minecraft/world/gen/feature/MegaJungleFeature.java
+++ b/net/minecraft/world/gen/feature/MegaJungleFeature.java
@@ -16,6 +16,7 @@
 public class MegaJungleFeature extends HugeTreesFeature<NoFeatureConfig> {
    public MegaJungleFeature(Function<Dynamic<?>, ? extends NoFeatureConfig> p_i51483_1_, boolean p_i51483_2_, int p_i51483_3_, int p_i51483_4_, BlockState p_i51483_5_, BlockState p_i51483_6_) {
       super(p_i51483_1_, p_i51483_2_, p_i51483_3_, p_i51483_4_, p_i51483_5_, p_i51483_6_);
+      setSapling((net.minecraftforge.common.IPlantable)Blocks.field_196678_w);
    }
 
    public boolean func_208519_a(Set<BlockPos> p_208519_1_, IWorldGenerationReader p_208519_2_, Random p_208519_3_, BlockPos p_208519_4_, MutableBoundingBox p_208519_5_) {
@@ -47,7 +48,7 @@
 
          for(int i2 = 0; i2 < i; ++i2) {
             BlockPos blockpos = p_208519_4_.func_177981_b(i2);
-            if (func_214587_a(p_208519_2_, blockpos)) {
+            if (canBeReplacedByLog(p_208519_2_, blockpos)) {
                this.func_208520_a(p_208519_1_, p_208519_2_, blockpos, this.field_76520_b, p_208519_5_);
                if (i2 > 0) {
                   this.func_202407_a(p_208519_2_, p_208519_3_, blockpos.func_177976_e(), VineBlock.field_176278_M);
@@ -57,7 +58,7 @@
 
             if (i2 < i - 1) {
                BlockPos blockpos1 = blockpos.func_177974_f();
-               if (func_214587_a(p_208519_2_, blockpos1)) {
+               if (canBeReplacedByLog(p_208519_2_, blockpos1)) {
                   this.func_208520_a(p_208519_1_, p_208519_2_, blockpos1, this.field_76520_b, p_208519_5_);
                   if (i2 > 0) {
                      this.func_202407_a(p_208519_2_, p_208519_3_, blockpos1.func_177974_f(), VineBlock.field_176280_O);
@@ -66,7 +67,7 @@
                }
 
                BlockPos blockpos2 = blockpos.func_177968_d().func_177974_f();
-               if (func_214587_a(p_208519_2_, blockpos2)) {
+               if (canBeReplacedByLog(p_208519_2_, blockpos2)) {
                   this.func_208520_a(p_208519_1_, p_208519_2_, blockpos2, this.field_76520_b, p_208519_5_);
                   if (i2 > 0) {
                      this.func_202407_a(p_208519_2_, p_208519_3_, blockpos2.func_177974_f(), VineBlock.field_176280_O);
@@ -75,7 +76,7 @@
                }
 
                BlockPos blockpos3 = blockpos.func_177968_d();
-               if (func_214587_a(p_208519_2_, blockpos3)) {
+               if (canBeReplacedByLog(p_208519_2_, blockpos3)) {
                   this.func_208520_a(p_208519_1_, p_208519_2_, blockpos3, this.field_76520_b, p_208519_5_);
                   if (i2 > 0) {
                      this.func_202407_a(p_208519_2_, p_208519_3_, blockpos3.func_177976_e(), VineBlock.field_176278_M);
@@ -90,7 +91,7 @@
    }
 
    private void func_202407_a(IWorldGenerationReader p_202407_1_, Random p_202407_2_, BlockPos p_202407_3_, BooleanProperty p_202407_4_) {
-      if (p_202407_2_.nextInt(3) > 0 && func_214574_b(p_202407_1_, p_202407_3_)) {
+      if (p_202407_2_.nextInt(3) > 0 && isAir(p_202407_1_, p_202407_3_)) {
          this.func_202278_a(p_202407_1_, p_202407_3_, Blocks.field_150395_bd.func_176223_P().func_206870_a(p_202407_4_, Boolean.valueOf(true)));
       }
 
