--- a/net/minecraft/world/gen/feature/OreFeatureConfig.java
+++ b/net/minecraft/world/gen/feature/OreFeatureConfig.java
@@ -31,7 +31,7 @@
       this.field_202442_b = p_i51473_1_;
    }
 
-   public static enum FillerBlockType implements IStringSerializable {
+   public static enum FillerBlockType implements IStringSerializable, net.minecraftforge.common.IExtensibleEnum {
       NATURAL_STONE("natural_stone", (p_214739_0_) -> {
          if (p_214739_0_ == null) {
             return false;
@@ -68,6 +68,16 @@
          return field_214741_c.get(p_214736_0_);
       }
 
+      public static FillerBlockType create(String enumName, String p_i50618_3_, Predicate<BlockState> p_i50618_4_) {
+         throw new IllegalStateException("Enum not extended");
+      }
+
+      @Override
+      @Deprecated
+      public void init() {
+          field_214741_c.put(func_214737_a(), this);
+      }
+
       public Predicate<BlockState> func_214738_b() {
          return this.field_214743_e;
       }
