--- a/net/minecraft/world/gen/feature/SavannaTreeFeature.java
+++ b/net/minecraft/world/gen/feature/SavannaTreeFeature.java
@@ -23,7 +23,7 @@
    public boolean func_208519_a(Set<BlockPos> p_208519_1_, IWorldGenerationReader p_208519_2_, Random p_208519_3_, BlockPos p_208519_4_, MutableBoundingBox p_208519_5_) {
       int i = p_208519_3_.nextInt(3) + p_208519_3_.nextInt(3) + 5;
       boolean flag = true;
-      if (p_208519_4_.func_177956_o() >= 1 && p_208519_4_.func_177956_o() + i + 1 <= 256) {
+      if (p_208519_4_.func_177956_o() >= 1 && p_208519_4_.func_177956_o() + i + 1 <= p_208519_2_.getMaxHeight()) {
          for(int j = p_208519_4_.func_177956_o(); j <= p_208519_4_.func_177956_o() + 1 + i; ++j) {
             int k = 1;
             if (j == p_208519_4_.func_177956_o()) {
@@ -38,7 +38,7 @@
 
             for(int l = p_208519_4_.func_177958_n() - k; l <= p_208519_4_.func_177958_n() + k && flag; ++l) {
                for(int i1 = p_208519_4_.func_177952_p() - k; i1 <= p_208519_4_.func_177952_p() + k && flag; ++i1) {
-                  if (j >= 0 && j < 256) {
+                  if (j >= 0 && j < p_208519_2_.getMaxHeight()) {
                      if (!func_214587_a(p_208519_2_, blockpos$mutableblockpos.func_181079_c(l, j, i1))) {
                         flag = false;
                      }
@@ -51,8 +51,8 @@
 
          if (!flag) {
             return false;
-         } else if (func_214589_h(p_208519_2_, p_208519_4_.func_177977_b()) && p_208519_4_.func_177956_o() < 256 - i - 1) {
-            this.func_214584_a(p_208519_2_, p_208519_4_.func_177977_b());
+         } else if (isSoil(p_208519_2_, p_208519_4_.func_177977_b(), getSapling()) && p_208519_4_.func_177956_o() < p_208519_2_.getMaxHeight() - i - 1) {
+            this.setDirtAt(p_208519_2_, p_208519_4_.func_177977_b(), p_208519_4_);
             Direction direction = Direction.Plane.HORIZONTAL.func_179518_a(p_208519_3_);
             int k2 = i - p_208519_3_.nextInt(4) - 1;
             int l2 = 3 - p_208519_3_.nextInt(3);
