--- a/net/minecraft/world/gen/feature/TreeFeature.java
+++ b/net/minecraft/world/gen/feature/TreeFeature.java
@@ -23,7 +23,7 @@
          return false;
       } else {
          BlockPos blockpos = optional.get();
-         this.func_214584_a(p_225557_1_, blockpos.func_177977_b());
+         this.setDirtAt(p_225557_1_, blockpos.func_177977_b(), blockpos);
          p_225557_7_.field_227327_a_.func_225571_a_(p_225557_1_, p_225557_2_, p_225557_7_, i, j, k, blockpos, p_225557_5_);
          this.func_227213_a_(p_225557_1_, p_225557_2_, i, blockpos, p_225557_7_.field_227332_g_ + p_225557_2_.nextInt(p_225557_7_.field_227333_h_ + 1), p_225557_4_, p_225557_6_, p_225557_7_);
          return true;
