--- ../src_base/minecraft/net/minecraft/world/gen/feature/WorldGenTaiga1.java
+++ ../src_work/minecraft/net/minecraft/world/gen/feature/WorldGenTaiga1.java
@@ -43,7 +43,9 @@
                         {
                             k2 = par1World.getBlockId(i2, l1, j2);
 
-                            if (k2 != 0 && k2 != Block.leaves.blockID)
+                            Block block = Block.blocksList[k2];
+
+                            if (k2 != 0 && (block == null || !block.isLeaves(par1World, i2, l1, j2)))
                             {
                                 flag = false;
                             }
@@ -79,7 +81,10 @@
                             {
                                 int j3 = i3 - par5;
 
-                                if ((Math.abs(k2) != l2 || Math.abs(j3) != l2 || l2 <= 0) && !Block.opaqueCubeLookup[par1World.getBlockId(j2, i2, i3)])
+                                Block block = Block.blocksList[par1World.getBlockId(j2, i2, i3)];
+
+                                if ((Math.abs(k2) != l2 || Math.abs(j3) != l2 || l2 <= 0) && 
+                                    (block == null || block.canBeReplacedByLeaves(par1World, j2, i2, i3)))
                                 {
                                     this.setBlockAndMetadata(par1World, j2, i2, i3, Block.leaves.blockID, 1);
                                 }
@@ -100,7 +105,9 @@
                     {
                         j2 = par1World.getBlockId(par3, par4 + i2, par5);
 
-                        if (j2 == 0 || j2 == Block.leaves.blockID)
+                        Block block = Block.blocksList[j2];
+
+                        if (j2 == 0 || block == null || block.isLeaves(par1World, par3, par4 + i2, par5))
                         {
                             this.setBlockAndMetadata(par1World, par3, par4 + i2, par5, Block.wood.blockID, 1);
                         }
