--- ../src_base/minecraft/net/minecraft/world/gen/feature/WorldGenTaiga1.java
+++ ../src_work/minecraft/net/minecraft/world/gen/feature/WorldGenTaiga1.java
@@ -43,7 +43,9 @@
                         {
                             var15 = par1World.getBlockId(var13, var11, var14);
 
-                            if (var15 != 0 && var15 != Block.leaves.blockID)
+                            Block block = Block.blocksList[var15];
+
+                            if (var15 != 0 && (block == null || !block.isLeaves(par1World, var13, var11, var14)))
                             {
                                 var10 = false;
                             }
@@ -79,7 +81,10 @@
                             {
                                 int var17 = var16 - par5;
 
-                                if ((Math.abs(var15) != var18 || Math.abs(var17) != var18 || var18 <= 0) && !Block.opaqueCubeLookup[par1World.getBlockId(var14, var13, var16)])
+                                Block block = Block.blocksList[par1World.getBlockId(var14, var13, var16)];
+
+                                if ((Math.abs(var15) != var18 || Math.abs(var17) != var18 || var18 <= 0) && 
+                                    (block == null || block.canBeReplacedByLeaves(par1World, var14, var13, var16)))
                                 {
                                     this.setBlockAndMetadata(par1World, var14, var13, var16, Block.leaves.blockID, 1);
                                 }
@@ -100,7 +105,9 @@
                     {
                         var14 = par1World.getBlockId(par3, par4 + var13, par5);
 
-                        if (var14 == 0 || var14 == Block.leaves.blockID)
+                        Block block = Block.blocksList[var14];
+
+                        if (var14 == 0 || block == null || block.isLeaves(par1World, par3, par4 + var13, par5))
                         {
                             this.setBlockAndMetadata(par1World, par3, par4 + var13, par5, Block.wood.blockID, 1);
                         }
