--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenSwamp.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenSwamp.java
@@ -60,7 +60,7 @@
                         {
                             Block block = p_180709_1_.func_180495_p(blockpos$mutableblockpos.func_181079_c(l, j, i1)).func_177230_c();
 
-                            if (block.func_149688_o() != Material.field_151579_a && block.func_149688_o() != Material.field_151584_j)
+                            if (!block.isAir(p_180709_1_, blockpos$mutableblockpos.func_181079_c(l, j, i1)) && !block.isLeaves(p_180709_1_, blockpos$mutableblockpos.func_181079_c(l, j, i1)))
                             {
                                 if (block != Blocks.field_150355_j && block != Blocks.field_150358_i)
                                 {
@@ -86,11 +86,13 @@
             }
             else
             {
-                Block block1 = p_180709_1_.func_180495_p(p_180709_3_.func_177977_b()).func_177230_c();
+                BlockPos down = p_180709_3_.func_177977_b();
+                Block block1 = p_180709_1_.func_180495_p(down).func_177230_c();
+                boolean isSoil = block1.canSustainPlant(p_180709_1_, down, net.minecraft.util.EnumFacing.UP, ((net.minecraft.block.BlockSapling)Blocks.field_150345_g));
 
-                if ((block1 == Blocks.field_150349_c || block1 == Blocks.field_150346_d) && p_180709_3_.func_177956_o() < 256 - i - 1)
+                if (isSoil && p_180709_3_.func_177956_o() < 256 - i - 1)
                 {
-                    this.func_175921_a(p_180709_1_, p_180709_3_.func_177977_b());
+                    block1.onPlantGrow(p_180709_1_, p_180709_3_.func_177977_b(),p_180709_3_);
 
                     for (int l1 = p_180709_3_.func_177956_o() - 3 + i; l1 <= p_180709_3_.func_177956_o() + i; ++l1)
                     {
@@ -109,7 +111,7 @@
                                 {
                                     BlockPos blockpos = new BlockPos(k3, l1, j1);
 
-                                    if (!p_180709_1_.func_180495_p(blockpos).func_177230_c().func_149730_j())
+                                    if (p_180709_1_.func_180495_p(blockpos).func_177230_c().canBeReplacedByLeaves(p_180709_1_, blockpos))
                                     {
                                         this.func_175903_a(p_180709_1_, blockpos, field_181649_b);
                                     }
@@ -120,9 +122,10 @@
 
                     for (int i2 = 0; i2 < i; ++i2)
                     {
-                        Block block2 = p_180709_1_.func_180495_p(p_180709_3_.func_177981_b(i2)).func_177230_c();
+                        BlockPos upN = p_180709_3_.func_177981_b(i2);
+                        Block block2 = p_180709_1_.func_180495_p(upN).func_177230_c();
 
-                        if (block2.func_149688_o() == Material.field_151579_a || block2.func_149688_o() == Material.field_151584_j || block2 == Blocks.field_150358_i || block2 == Blocks.field_150355_j)
+                        if (block2.isAir(p_180709_1_, upN) || block2.isLeaves(p_180709_1_, upN) || block2 == Blocks.field_150358_i || block2 == Blocks.field_150355_j)
                         {
                             this.func_175903_a(p_180709_1_, p_180709_3_.func_177981_b(i2), field_181648_a);
                         }
@@ -140,29 +143,29 @@
                             {
                                 blockpos$mutableblockpos1.func_181079_c(i4, j2, j4);
 
-                                if (p_180709_1_.func_180495_p(blockpos$mutableblockpos1).func_177230_c().func_149688_o() == Material.field_151584_j)
+                                if (p_180709_1_.func_180495_p(blockpos$mutableblockpos1).func_177230_c().isLeaves(p_180709_1_, blockpos$mutableblockpos1))
                                 {
                                     BlockPos blockpos3 = blockpos$mutableblockpos1.func_177976_e();
                                     BlockPos blockpos4 = blockpos$mutableblockpos1.func_177974_f();
                                     BlockPos blockpos1 = blockpos$mutableblockpos1.func_177978_c();
                                     BlockPos blockpos2 = blockpos$mutableblockpos1.func_177968_d();
 
-                                    if (p_180709_2_.nextInt(4) == 0 && p_180709_1_.func_180495_p(blockpos3).func_177230_c().func_149688_o() == Material.field_151579_a)
+                                    if (p_180709_2_.nextInt(4) == 0 && p_180709_1_.func_180495_p(blockpos3).func_177230_c().isAir(p_180709_1_,blockpos3))
                                     {
                                         this.func_181647_a(p_180709_1_, blockpos3, BlockVine.field_176278_M);
                                     }
 
-                                    if (p_180709_2_.nextInt(4) == 0 && p_180709_1_.func_180495_p(blockpos4).func_177230_c().func_149688_o() == Material.field_151579_a)
+                                    if (p_180709_2_.nextInt(4) == 0 && p_180709_1_.func_180495_p(blockpos4).func_177230_c().isAir(p_180709_1_,blockpos4))
                                     {
                                         this.func_181647_a(p_180709_1_, blockpos4, BlockVine.field_176280_O);
                                     }
 
-                                    if (p_180709_2_.nextInt(4) == 0 && p_180709_1_.func_180495_p(blockpos1).func_177230_c().func_149688_o() == Material.field_151579_a)
+                                    if (p_180709_2_.nextInt(4) == 0 && p_180709_1_.func_180495_p(blockpos1).func_177230_c().isAir(p_180709_1_,blockpos1))
                                     {
                                         this.func_181647_a(p_180709_1_, blockpos1, BlockVine.field_176279_N);
                                     }
 
-                                    if (p_180709_2_.nextInt(4) == 0 && p_180709_1_.func_180495_p(blockpos2).func_177230_c().func_149688_o() == Material.field_151579_a)
+                                    if (p_180709_2_.nextInt(4) == 0 && p_180709_1_.func_180495_p(blockpos2).func_177230_c().isAir(p_180709_1_,blockpos2))
                                     {
                                         this.func_181647_a(p_180709_1_, blockpos2, BlockVine.field_176273_b);
                                     }
@@ -191,7 +194,7 @@
         this.func_175903_a(p_181647_1_, p_181647_2_, iblockstate);
         int i = 4;
 
-        for (p_181647_2_ = p_181647_2_.func_177977_b(); p_181647_1_.func_180495_p(p_181647_2_).func_177230_c().func_149688_o() == Material.field_151579_a && i > 0; --i)
+        for (p_181647_2_ = p_181647_2_.func_177977_b(); p_181647_1_.func_180495_p(p_181647_2_).func_177230_c().isAir(p_181647_1_,p_181647_2_) && i > 0; --i)
         {
             this.func_175903_a(p_181647_1_, p_181647_2_, iblockstate);
             p_181647_2_ = p_181647_2_.func_177977_b();
