--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenMegaJungle.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenMegaJungle.java
@@ -55,7 +55,7 @@
             {
                 BlockPos blockpos = p_180709_3_.func_177981_b(i2);
 
-                if (this.func_150523_a(p_180709_1_.func_180495_p(blockpos).func_177230_c()))
+                if (this.isAirLeaves(p_180709_1_,blockpos))
                 {
                     this.func_175903_a(p_180709_1_, blockpos, this.field_76520_b);
 
@@ -70,7 +70,7 @@
                 {
                     BlockPos blockpos1 = blockpos.func_177974_f();
 
-                    if (this.func_150523_a(p_180709_1_.func_180495_p(blockpos1).func_177230_c()))
+                    if (this.isAirLeaves(p_180709_1_,blockpos1))
                     {
                         this.func_175903_a(p_180709_1_, blockpos1, this.field_76520_b);
 
@@ -83,7 +83,7 @@
 
                     BlockPos blockpos2 = blockpos.func_177968_d().func_177974_f();
 
-                    if (this.func_150523_a(p_180709_1_.func_180495_p(blockpos2).func_177230_c()))
+                    if (this.isAirLeaves(p_180709_1_,blockpos2))
                     {
                         this.func_175903_a(p_180709_1_, blockpos2, this.field_76520_b);
 
@@ -96,7 +96,7 @@
 
                     BlockPos blockpos3 = blockpos.func_177968_d();
 
-                    if (this.func_150523_a(p_180709_1_.func_180495_p(blockpos3).func_177230_c()))
+                    if (this.isAirLeaves(p_180709_1_,blockpos3))
                     {
                         this.func_175903_a(p_180709_1_, blockpos3, this.field_76520_b);
 
@@ -130,4 +130,11 @@
             this.func_175925_a(p_175930_1_, p_175930_2_.func_177981_b(j), p_175930_3_ + 1 - j);
         }
     }
+
+    //Helper macro
+    private boolean isAirLeaves(World world, BlockPos pos)
+    {
+        net.minecraft.block.Block block = world.func_180495_p(pos).func_177230_c();
+        return block.isAir(world, pos) || block.isLeaves(world, pos);
+    }
 }
