--- ../src_base/minecraft/net/minecraft/world/gen/feature/WorldGenTaiga2.java
+++ ../src_work/minecraft/net/minecraft/world/gen/feature/WorldGenTaiga2.java
@@ -2,9 +2,13 @@
 
 import java.util.Random;
 import net.minecraft.block.Block;
+import net.minecraft.block.BlockSapling;
 import net.minecraft.world.World;
+import net.minecraftforge.common.ForgeDirection;
 
-public class WorldGenTaiga2 extends WorldGenerator
+import org.bukkit.BlockChangeDelegate; // CraftBukkit
+
+public class WorldGenTaiga2 extends WorldGenerator implements net.minecraft.block.BlockSapling.TreeGenerator   // CraftBukkit add interface
 {
     public WorldGenTaiga2(boolean par1)
     {
@@ -13,6 +17,13 @@
 
     public boolean generate(World par1World, Random par2Random, int par3, int par4, int par5)
     {
+        // CraftBukkit start - Moved to generate
+        return this.generate((BlockChangeDelegate) par1World, par2Random, par3, par4, par5);
+    }
+
+    public boolean generate(BlockChangeDelegate par1World, Random par2Random, int par3, int par4, int par5)
+    {
+        // CraftBukkit end
         int l = par2Random.nextInt(4) + 6;
         int i1 = 1 + par2Random.nextInt(2);
         int j1 = l - i1;
@@ -45,9 +56,11 @@
                     {
                         if (l1 >= 0 && l1 < 256)
                         {
-                            j2 = par1World.getBlockId(i2, l1, l2);
+                            j2 = par1World.getTypeId(i2, l1, l2);
 
-                            if (j2 != 0 && j2 != Block.leaves.blockID)
+                            Block block = Block.blocksList[j2];
+
+                            if (j2 != 0 && block != null && !block.isLeaves(par1World, i2, l1, l2))
                             {
                                 flag = false;
                             }
@@ -66,11 +79,13 @@
             }
             else
             {
-                l1 = par1World.getBlockId(par3, par4 - 1, par5);
+                l1 = par1World.getTypeId(par3, par4 - 1, par5);
+                Block soil = Block.blocksList[l1];
+                boolean isValidSoil = soil != null && soil.canSustainPlant(par1World, par3, par4 - 1, par5, ForgeDirection.UP, (BlockSapling)Block.sapling);
 
-                if ((l1 == Block.grass.blockID || l1 == Block.dirt.blockID) && par4 < 256 - l - 1)
+                if (isValidSoil && par4 < 256 - l - 1)
                 {
-                    this.setBlock(par1World, par3, par4 - 1, par5, Block.dirt.blockID);
+                    soil.onPlantGrow(par1World, par3, par4 - 1, par5, par3, par4, par5);
                     k2 = par2Random.nextInt(2);
                     i2 = 1;
                     byte b0 = 0;
@@ -89,9 +104,12 @@
                             {
                                 int i4 = l3 - par5;
 
-                                if ((Math.abs(k3) != k2 || Math.abs(i4) != k2 || k2 <= 0) && !Block.opaqueCubeLookup[par1World.getBlockId(i3, j3, l3)])
+                                Block block = Block.blocksList[par1World.getTypeId(i3, j3, l3)];
+
+                                if ((Math.abs(k3) != k2 || Math.abs(i4) != k2 || k2 <= 0) &&
+                                    (block == null || block.canBeReplacedByLeaves(par1World, i3, j3, l3)))
                                 {
-                                    this.setBlockAndMetadata(par1World, i3, j3, l3, Block.leaves.blockID, 1);
+                                    this.setTypeAndData(par1World, i3, j3, l3, Block.leaves.blockID, 1);
                                 }
                             }
                         }
@@ -117,11 +135,13 @@
 
                     for (j3 = 0; j3 < l - j2; ++j3)
                     {
-                        i3 = par1World.getBlockId(par3, par4 + j3, par5);
+                        i3 = par1World.getTypeId(par3, par4 + j3, par5);
 
-                        if (i3 == 0 || i3 == Block.leaves.blockID)
+                        Block block = Block.blocksList[i3];
+
+                        if (i3 == 0 || block == null || block.isLeaves(par1World, par3, par4 + j3, par5))
                         {
-                            this.setBlockAndMetadata(par1World, par3, par4 + j3, par5, Block.wood.blockID, 1);
+                            this.setTypeAndData(par1World, par3, par4 + j3, par5, Block.wood.blockID, 1);
                         }
                     }
 
