--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenAbstractTree.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenAbstractTree.java
@@ -31,4 +31,10 @@
             this.func_175903_a(p_175921_1_, p_175921_2_, Blocks.field_150346_d.func_176223_P());
         }
     }
+
+    public boolean isReplaceable(World world, BlockPos pos)
+    {
+        net.minecraft.block.state.IBlockState state = world.func_180495_p(pos);
+        return state.func_177230_c().isAir(world, pos) || state.func_177230_c().isLeaves(world, pos) || state.func_177230_c().isWood(world, pos) || func_150523_a(state.func_177230_c());
+    }
 }
