--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenHugeTrees.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenHugeTrees.java
@@ -2,9 +2,11 @@
 
 import java.util.Random;
 import net.minecraft.block.Block;
+import net.minecraft.block.BlockSapling;
 import net.minecraft.block.material.Material;
 import net.minecraft.init.Blocks;
 import net.minecraft.world.World;
+import net.minecraftforge.common.util.ForgeDirection;
 
 public abstract class WorldGenHugeTrees extends WorldGenAbstractTree
 {
@@ -63,7 +65,7 @@
                         {
                             Block block = p_150536_1_.func_147439_a(j1, i1, k1);
 
-                            if (!this.func_150523_a(block))
+                            if (!this.isReplaceable(p_150536_1_, j1, i1, k1))
                             {
                                 flag = false;
                             }
@@ -88,12 +90,13 @@
     {
         Block block = p_150532_1_.func_147439_a(p_150532_3_, p_150532_4_ - 1, p_150532_5_);
 
-        if ((block == Blocks.field_150349_c || block == Blocks.field_150346_d) && p_150532_4_ >= 2)
+        boolean isSoil = block.canSustainPlant(p_150532_1_, p_150532_3_, p_150532_4_ - 1, p_150532_5_, ForgeDirection.UP, (BlockSapling)Blocks.field_150345_g);
+        if (isSoil && p_150532_4_ >= 2)
         {
-            p_150532_1_.func_147465_d(p_150532_3_, p_150532_4_ - 1, p_150532_5_, Blocks.field_150346_d, 0, 2);
-            p_150532_1_.func_147465_d(p_150532_3_ + 1, p_150532_4_ - 1, p_150532_5_, Blocks.field_150346_d, 0, 2);
-            p_150532_1_.func_147465_d(p_150532_3_, p_150532_4_ - 1, p_150532_5_ + 1, Blocks.field_150346_d, 0, 2);
-            p_150532_1_.func_147465_d(p_150532_3_ + 1, p_150532_4_ - 1, p_150532_5_ + 1, Blocks.field_150346_d, 0, 2);
+            onPlantGrow(p_150532_1_, p_150532_3_,     p_150532_4_ - 1, p_150532_5_,     p_150532_3_, p_150532_4_, p_150532_5_);
+            onPlantGrow(p_150532_1_, p_150532_3_ + 1, p_150532_4_ - 1, p_150532_5_,     p_150532_3_, p_150532_4_, p_150532_5_);
+            onPlantGrow(p_150532_1_, p_150532_3_,     p_150532_4_ - 1, p_150532_5_ + 1, p_150532_3_, p_150532_4_, p_150532_5_);
+            onPlantGrow(p_150532_1_, p_150532_3_ + 1, p_150532_4_ - 1, p_150532_5_ + 1, p_150532_3_, p_150532_4_, p_150532_5_);
             return true;
         }
         else
@@ -125,7 +128,7 @@
                 {
                     Block block = p_150535_1_.func_147439_a(j1, p_150535_3_, l1);
 
-                    if (block.func_149688_o() == Material.field_151579_a || block.func_149688_o() == Material.field_151584_j)
+                    if (block.isAir(p_150535_1_, j1, p_150535_3_, l1) || block.isLeaves(p_150535_1_, j1, p_150535_3_, l1))
                     {
                         this.func_150516_a(p_150535_1_, j1, p_150535_3_, l1, Blocks.field_150362_t, this.field_76521_c);
                     }
@@ -150,7 +153,7 @@
                 {
                     Block block = p_150534_1_.func_147439_a(j1, p_150534_3_, l1);
 
-                    if (block.func_149688_o() == Material.field_151579_a || block.func_149688_o() == Material.field_151584_j)
+                    if (block.isAir(p_150534_1_, j1, p_150534_3_, l1) || block.isLeaves(p_150534_1_, j1, p_150534_3_, l1))
                     {
                         this.func_150516_a(p_150534_1_, j1, p_150534_3_, l1, Blocks.field_150362_t, this.field_76521_c);
                     }
@@ -158,4 +161,10 @@
             }
         }
     }
+
+    //Just a helper macro
+    private void onPlantGrow(World world, int x, int y, int z, int sourceX, int sourceY, int sourceZ)
+    {
+        world.func_147439_a(x, y, z).onPlantGrow(world, x, y, z, sourceX, sourceY, sourceZ);
+    }
 }
