--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenMinable.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenMinable.java
@@ -73,7 +73,7 @@
                                 {
                                     BlockPos blockpos = new BlockPos(l1, i2, j2);
 
-                                    if (this.field_175919_c.apply(p_180709_1_.func_180495_p(blockpos)))
+                                    if (p_180709_1_.func_180495_p(blockpos).func_177230_c().isReplaceableOreGen(p_180709_1_, blockpos, this.field_175919_c))
                                     {
                                         p_180709_1_.func_180501_a(blockpos, this.field_175920_a, 2);
                                     }
