--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenTallGrass.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenTallGrass.java
@@ -21,10 +21,15 @@
     {
         Block block;
 
-        while (((block = par1World.getBlock(par3, par4, par5)).getMaterial() == Material.air || block.getMaterial() == Material.leaves) && par4 > 0)
+        do
         {
+            block = par1World.getBlock(par3, par4, par5);
+            if (!(block.isLeaves(par1World, par3, par4, par5) || block.isAir(par1World, par3, par4, par5)))
+            {
+                break;
+            }
             --par4;
-        }
+        } while (par4 > 0);
 
         for (int l = 0; l < 128; ++l)
         {
