--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenDeadBush.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenDeadBush.java
@@ -19,10 +19,15 @@
     {
         Block block;
 
-        while (((block = p_76484_1_.func_147439_a(p_76484_3_, p_76484_4_, p_76484_5_)).func_149688_o() == Material.field_151579_a || block.func_149688_o() == Material.field_151584_j) && p_76484_4_ > 0)
+        do
         {
+            block = p_76484_1_.func_147439_a(p_76484_3_, p_76484_4_, p_76484_5_);
+            if (!(block.isLeaves(p_76484_1_, p_76484_3_, p_76484_4_, p_76484_5_) || block.isAir(p_76484_1_, p_76484_3_, p_76484_4_, p_76484_5_)))
+            {
+                break;
+            }
             --p_76484_4_;
-        }
+        } while (p_76484_4_ > 0);
 
         for (int l = 0; l < 4; ++l)
         {
