--- ../src_base/minecraft/net/minecraft/world/gen/feature/WorldGenHugeTrees.java
+++ ../src_work/minecraft/net/minecraft/world/gen/feature/WorldGenHugeTrees.java
@@ -58,7 +58,12 @@
                         {
                             l1 = par1World.getBlockId(j1, i1, k1);
 
-                            if (l1 != 0 && l1 != Block.leaves.blockID && l1 != Block.grass.blockID && l1 != Block.dirt.blockID && l1 != Block.wood.blockID && l1 != Block.sapling.blockID)
+                            if (l1 != 0 && 
+                               (Block.blocksList[l1] != null && !Block.blocksList[l1].isLeaves(par1World, j1, i1, k1)) && 
+                               l1 != Block.grass.blockID && 
+                               l1 != Block.dirt.blockID && 
+                               (Block.blocksList[l1] != null && !Block.blocksList[l1].isWood(par1World, j1, i1, k1)) &&
+                               l1 != Block.sapling.blockID)
                             {
                                 flag = false;
                             }
@@ -106,7 +111,7 @@
                     {
                         k1 = par1World.getBlockId(par3, par4 + j1, par5);
 
-                        if (k1 == 0 || k1 == Block.leaves.blockID)
+                        if (k1 == 0 || Block.blocksList[k1] == null || Block.blocksList[k1].isLeaves(par1World, par3, par4 + j1, par5))
                         {
                             this.setBlockAndMetadata(par1World, par3, par4 + j1, par5, Block.wood.blockID, this.woodMetadata);
 
@@ -128,7 +133,7 @@
                         {
                             k1 = par1World.getBlockId(par3 + 1, par4 + j1, par5);
 
-                            if (k1 == 0 || k1 == Block.leaves.blockID)
+                            if (k1 == 0 || Block.blocksList[k1] == null || Block.blocksList[k1].isLeaves(par1World, par3 + 1, par4 + j1, par5))
                             {
                                 this.setBlockAndMetadata(par1World, par3 + 1, par4 + j1, par5, Block.wood.blockID, this.woodMetadata);
 
@@ -148,7 +153,7 @@
 
                             k1 = par1World.getBlockId(par3 + 1, par4 + j1, par5 + 1);
 
-                            if (k1 == 0 || k1 == Block.leaves.blockID)
+                            if (k1 == 0 || Block.blocksList[k1] == null || Block.blocksList[k1].isLeaves(par1World, par3 + 1, par4 + j1, par5 + 1))
                             {
                                 this.setBlockAndMetadata(par1World, par3 + 1, par4 + j1, par5 + 1, Block.wood.blockID, this.woodMetadata);
 
@@ -168,7 +173,7 @@
 
                             k1 = par1World.getBlockId(par3, par4 + j1, par5 + 1);
 
-                            if (k1 == 0 || k1 == Block.leaves.blockID)
+                            if (k1 == 0 || Block.blocksList[k1] == null || Block.blocksList[k1].isLeaves(par1World, par3, par4 + j1, par5 + 1))
                             {
                                 this.setBlockAndMetadata(par1World, par3, par4 + j1, par5 + 1, Block.wood.blockID, this.woodMetadata);
 
@@ -219,7 +224,12 @@
                 {
                     int k2 = j2 - par3;
 
-                    if ((i2 >= 0 || k2 >= 0 || i2 * i2 + k2 * k2 <= k1 * k1) && (i2 <= 0 && k2 <= 0 || i2 * i2 + k2 * k2 <= (k1 + 1) * (k1 + 1)) && (par6Random.nextInt(4) != 0 || i2 * i2 + k2 * k2 <= (k1 - 1) * (k1 - 1)) && !Block.opaqueCubeLookup[par1World.getBlockId(l1, i1, j2)])
+                    Block block = Block.blocksList[par1World.getBlockId(l1, i1, j2)];
+
+                    if ((i2 >= 0 || k2 >= 0 || i2 * i2 + k2 * k2 <= k1 * k1) && 
+                        (i2 <= 0 && k2 <= 0 || i2 * i2 + k2 * k2 <= (k1 + 1) * (k1 + 1)) && 
+                        (par6Random.nextInt(4) != 0 || i2 * i2 + k2 * k2 <= (k1 - 1) * (k1 - 1)) && 
+                        (block == null || block.canBeReplacedByLeaves(par1World, l1, i1, j2)))
                     {
                         this.setBlockAndMetadata(par1World, l1, i1, j2, Block.leaves.blockID, this.leavesMetadata);
                     }
