--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenCanopyTree.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenCanopyTree.java
@@ -2,10 +2,12 @@
 
 import java.util.Random;
 import net.minecraft.block.Block;
+import net.minecraft.block.BlockSapling;
 import net.minecraft.block.material.Material;
 import net.minecraft.init.Blocks;
 import net.minecraft.util.Direction;
 import net.minecraft.world.World;
+import net.minecraftforge.common.util.ForgeDirection;
 
 public class WorldGenCanopyTree extends WorldGenAbstractTree
 {
@@ -48,7 +50,7 @@
                         {
                             Block block = p_76484_1_.func_147439_a(j1, i1, k1);
 
-                            if (!this.func_150523_a(block))
+                            if (!this.isReplaceable(p_76484_1_, j1, i1, k1))
                             {
                                 flag = false;
                             }
@@ -69,12 +71,13 @@
             {
                 Block block2 = p_76484_1_.func_147439_a(p_76484_3_, p_76484_4_ - 1, p_76484_5_);
 
-                if ((block2 == Blocks.field_150349_c || block2 == Blocks.field_150346_d) && p_76484_4_ < 256 - l - 1)
+                boolean isSoil = block2.canSustainPlant(p_76484_1_, p_76484_3_, p_76484_4_ - 1, p_76484_5_, ForgeDirection.UP, (BlockSapling)Blocks.field_150345_g);
+                if (isSoil && p_76484_4_ < 256 - l - 1)
                 {
-                    this.func_150515_a(p_76484_1_, p_76484_3_, p_76484_4_ - 1, p_76484_5_, Blocks.field_150346_d);
-                    this.func_150515_a(p_76484_1_, p_76484_3_ + 1, p_76484_4_ - 1, p_76484_5_, Blocks.field_150346_d);
-                    this.func_150515_a(p_76484_1_, p_76484_3_ + 1, p_76484_4_ - 1, p_76484_5_ + 1, Blocks.field_150346_d);
-                    this.func_150515_a(p_76484_1_, p_76484_3_, p_76484_4_ - 1, p_76484_5_ + 1, Blocks.field_150346_d);
+                    onPlantGrow(p_76484_1_, p_76484_3_,     p_76484_4_ - 1, p_76484_5_,     p_76484_3_, p_76484_4_, p_76484_5_);
+                    onPlantGrow(p_76484_1_, p_76484_3_ + 1, p_76484_4_ - 1, p_76484_5_,     p_76484_3_, p_76484_4_, p_76484_5_);
+                    onPlantGrow(p_76484_1_, p_76484_3_ + 1, p_76484_4_ - 1, p_76484_5_ + 1, p_76484_3_, p_76484_4_, p_76484_5_);
+                    onPlantGrow(p_76484_1_, p_76484_3_,     p_76484_4_ - 1, p_76484_5_ + 1, p_76484_3_, p_76484_4_, p_76484_5_);
                     int j3 = p_76484_2_.nextInt(4);
                     j1 = l - p_76484_2_.nextInt(4);
                     k1 = 2 - p_76484_2_.nextInt(3);
@@ -97,7 +100,7 @@
 
                         Block block1 = p_76484_1_.func_147439_a(k3, k2, l1);
 
-                        if (block1.func_149688_o() == Material.field_151579_a || block1.func_149688_o() == Material.field_151584_j)
+                        if (block1.isAir(p_76484_1_, k3, k2, l1) || block1.isLeaves(p_76484_1_, k3, k2, l1))
                         {
                             this.func_150516_a(p_76484_1_, k3, k2, l1, Blocks.field_150363_s, 1);
                             this.func_150516_a(p_76484_1_, k3 + 1, k2, l1, Blocks.field_150363_s, 1);
@@ -203,9 +206,15 @@
     {
         Block block = p_150526_1_.func_147439_a(p_150526_2_, p_150526_3_, p_150526_4_);
 
-        if (block.func_149688_o() == Material.field_151579_a)
+        if (block.isAir(p_150526_1_, p_150526_2_, p_150526_3_, p_150526_4_))
         {
             this.func_150516_a(p_150526_1_, p_150526_2_, p_150526_3_, p_150526_4_, Blocks.field_150361_u, 1);
         }
     }
+
+    //Just a helper macro
+    private void onPlantGrow(World world, int x, int y, int z, int sourceX, int sourceY, int sourceZ)
+    {
+        world.func_147439_a(x, y, z).onPlantGrow(world, x, y, z, sourceX, sourceY, sourceZ);
+    }
 }
