--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java
@@ -33,7 +33,8 @@
         }
         else
         {
-            p_175903_1_.func_180501_a(p_175903_2_, p_175903_3_, 2);
+            int flag = net.minecraftforge.common.ForgeModContainer.fixVanillaCascading ? 2| 16 : 2; //Forge: With bit 5 unset, it will notify neighbors and load adjacent chunks.
+            p_175903_1_.func_180501_a(p_175903_2_, p_175903_3_, flag);
         }
     }
 }
