--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenCanopyTree.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenCanopyTree.java
@@ -48,7 +48,7 @@
                     {
                         if (j >= 0 && j < 256)
                         {
-                            if (!this.func_150523_a(p_180709_1_.func_180495_p(new BlockPos(k, j, l)).func_177230_c()))
+                            if (!this.isReplaceable(p_180709_1_, new BlockPos(k, j, l)))
                             {
                                 flag = false;
                             }
@@ -67,14 +67,16 @@
             }
             else
             {
-                Block block = p_180709_1_.func_180495_p(p_180709_3_.func_177977_b()).func_177230_c();
+                BlockPos down = p_180709_3_.func_177977_b();
+                net.minecraft.block.state.IBlockState state = p_180709_1_.func_180495_p(down);
+                boolean isSoil = state.func_177230_c().canSustainPlant(p_180709_1_, down, net.minecraft.util.EnumFacing.UP, ((net.minecraft.block.BlockSapling)Blocks.field_150345_g));
 
-                if ((block == Blocks.field_150349_c || block == Blocks.field_150346_d) && p_180709_3_.func_177956_o() < 256 - i - 1)
+                if (isSoil && p_180709_3_.func_177956_o() < 256 - i - 1)
                 {
-                    this.func_175921_a(p_180709_1_, p_180709_3_.func_177977_b());
-                    this.func_175921_a(p_180709_1_, p_180709_3_.func_177982_a(1, -1, 0));
-                    this.func_175921_a(p_180709_1_, p_180709_3_.func_177982_a(1, -1, 1));
-                    this.func_175921_a(p_180709_1_, p_180709_3_.func_177982_a(0, -1, 1));
+                    this.onPlantGrow(p_180709_1_, p_180709_3_.func_177977_b(), p_180709_3_);
+                    this.onPlantGrow(p_180709_1_, p_180709_3_.func_177982_a(1, -1, 0), p_180709_3_);
+                    this.onPlantGrow(p_180709_1_, p_180709_3_.func_177982_a(1, -1, 1), p_180709_3_);
+                    this.onPlantGrow(p_180709_1_, p_180709_3_.func_177982_a(0, -1, 1), p_180709_3_);
                     EnumFacing enumfacing = EnumFacing.Plane.HORIZONTAL.func_179518_a(p_180709_2_);
                     k = i - p_180709_2_.nextInt(4);
                     l = 2 - p_180709_2_.nextInt(3);
@@ -96,9 +98,9 @@
                         }
 
                         BlockPos blockpos1 = new BlockPos(i1, i2, j1);
-                        Material material = p_180709_1_.func_180495_p(blockpos1).func_177230_c().func_149688_o();
+                        state = p_180709_1_.func_180495_p(blockpos1);
 
-                        if (material == Material.field_151579_a || material == Material.field_151584_j)
+                        if (state.func_177230_c().isAir(p_180709_1_, blockpos1) || state.func_177230_c().isLeaves(p_180709_1_, blockpos1))
                         {
                             this.func_175905_a(p_180709_1_, blockpos1, Blocks.field_150363_s, BlockPlanks.EnumType.DARK_OAK.func_176839_a() - 4);
                             this.func_175905_a(p_180709_1_, blockpos1.func_177974_f(), Blocks.field_150363_s, BlockPlanks.EnumType.DARK_OAK.func_176839_a() - 4);
@@ -202,11 +204,18 @@
 
     private void func_150526_a(World p_150526_1_, int p_150526_2_, int p_150526_3_, int p_150526_4_)
     {
-        Block block = p_150526_1_.func_180495_p(new BlockPos(p_150526_2_, p_150526_3_, p_150526_4_)).func_177230_c();
+        BlockPos pos = new BlockPos(p_150526_2_, p_150526_3_, p_150526_4_);
+        net.minecraft.block.state.IBlockState state = p_150526_1_.func_180495_p(pos);
 
-        if (block.func_149688_o() == Material.field_151579_a)
+        if (state.func_177230_c().isAir(p_150526_1_, pos))
         {
             this.func_175905_a(p_150526_1_, new BlockPos(p_150526_2_, p_150526_3_, p_150526_4_), Blocks.field_150361_u, 1);
         }
     }
+
+    //Just a helper macro
+    private void onPlantGrow(World world, BlockPos pos, BlockPos source)
+    {
+        world.func_180495_p(pos).func_177230_c().onPlantGrow(world, pos, source);
+    }
 }
