--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenForest.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenForest.java
@@ -2,9 +2,11 @@
 
 import java.util.Random;
 import net.minecraft.block.Block;
+import net.minecraft.block.BlockSapling;
 import net.minecraft.block.material.Material;
 import net.minecraft.init.Blocks;
 import net.minecraft.world.World;
+import net.minecraftforge.common.util.ForgeDirection;
 
 public class WorldGenForest extends WorldGenAbstractTree
 {
@@ -55,7 +57,7 @@
                         {
                             Block block = p_76484_1_.func_147439_a(j1, i1, k1);
 
-                            if (!this.func_150523_a(block))
+                            if (!this.isReplaceable(p_76484_1_, j1, i1, k1))
                             {
                                 flag = false;
                             }
@@ -76,9 +78,10 @@
             {
                 Block block2 = p_76484_1_.func_147439_a(p_76484_3_, p_76484_4_ - 1, p_76484_5_);
 
-                if ((block2 == Blocks.field_150349_c || block2 == Blocks.field_150346_d || block2 == Blocks.field_150458_ak) && p_76484_4_ < 256 - l - 1)
+                boolean isSoil = block2.canSustainPlant(p_76484_1_, p_76484_3_, p_76484_4_ - 1, p_76484_5_, ForgeDirection.UP, (BlockSapling)Blocks.field_150345_g);
+                if (isSoil && p_76484_4_ < 256 - l - 1)
                 {
-                    this.func_150515_a(p_76484_1_, p_76484_3_, p_76484_4_ - 1, p_76484_5_, Blocks.field_150346_d);
+                    block2.onPlantGrow(p_76484_1_, p_76484_3_, p_76484_4_ - 1, p_76484_5_, p_76484_3_, p_76484_4_, p_76484_5_);
                     int k2;
 
                     for (k2 = p_76484_4_ - 3 + l; k2 <= p_76484_4_ + l; ++k2)
@@ -98,7 +101,7 @@
                                 {
                                     Block block1 = p_76484_1_.func_147439_a(l2, k2, i2);
 
-                                    if (block1.func_149688_o() == Material.field_151579_a || block1.func_149688_o() == Material.field_151584_j)
+                                    if (block1.isAir(p_76484_1_, l2, k2, i2) || block1.isLeaves(p_76484_1_, l2, k2, i2))
                                     {
                                         this.func_150516_a(p_76484_1_, l2, k2, i2, Blocks.field_150362_t, 2);
                                     }
@@ -111,7 +114,7 @@
                     {
                         Block block3 = p_76484_1_.func_147439_a(p_76484_3_, p_76484_4_ + k2, p_76484_5_);
 
-                        if (block3.func_149688_o() == Material.field_151579_a || block3.func_149688_o() == Material.field_151584_j)
+                        if (block3.isAir(p_76484_1_, p_76484_3_, p_76484_4_ + k2, p_76484_5_) || block3.isLeaves(p_76484_1_, p_76484_3_, p_76484_4_ + k2, p_76484_5_))
                         {
                             this.func_150516_a(p_76484_1_, p_76484_3_, p_76484_4_ + k2, p_76484_5_, Blocks.field_150364_r, 2);
                         }
