--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenMinable.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenMinable.java
@@ -74,7 +74,7 @@
                                 {
                                     BlockPos blockpos1 = new BlockPos(l1, i2, j2);
 
-                                    if (this.field_175919_c.apply(p_180709_1_.func_180495_p(blockpos1)))
+                                    if (p_180709_1_.func_180495_p(blockpos1).func_177230_c().isReplaceableOreGen(p_180709_1_, blockpos1, this.field_175919_c))
                                     {
                                         p_180709_1_.func_180501_a(blockpos1, this.field_175920_a, 2);
                                     }
