--- a/net/minecraft/client/gui/widget/list/AbstractList.java
+++ b/net/minecraft/client/gui/widget/list/AbstractList.java
@@ -471,6 +471,13 @@
       p_238480_1_.field_230666_a_ = this;
    }
 
+   public int getWidth() { return this.field_230670_d_; }
+   public int getHeight() { return this.field_230671_e_; }
+   public int getTop() { return this.field_230672_i_; }
+   public int getBottom() { return this.field_230673_j_; }
+   public int getLeft() { return this.field_230675_l_; }
+   public int getRight() { return this.field_230674_k_; }
+
    @OnlyIn(Dist.CLIENT)
    public abstract static class AbstractListEntry<E extends AbstractList.AbstractListEntry<E>> implements IGuiEventListener {
       @Deprecated
