--- a/net/minecraft/world/item/BlockItem.java
+++ b/net/minecraft/world/item/BlockItem.java
@@ -83,17 +_,23 @@
             }
         }
 
-        SoundType soundType = placedState.getSoundType();
-        level.playSound(player, pos, this.getPlaceSound(placedState), SoundSource.BLOCKS, (soundType.getVolume() + 1.0F) / 2.0F, soundType.getPitch() * 0.8F);
+        SoundType soundType = placedState.getSoundType(level, pos, placeContext.getPlayer());
+        level.playSound(player, pos, this.getPlaceSound(placedState, level, pos, placeContext.getPlayer()), SoundSource.BLOCKS, (soundType.getVolume() + 1.0F) / 2.0F, soundType.getPitch() * 0.8F);
         level.gameEvent(GameEvent.BLOCK_PLACE, pos, GameEvent.Context.of(player, placedState));
         itemStack.consume(1, player);
         return InteractionResult.SUCCESS;
     }
 
+    @Deprecated //Forge: Use more sensitive version {@link BlockItem#getPlaceSound(BlockState, IBlockReader, BlockPos, Entity) }
     protected SoundEvent getPlaceSound(final BlockState blockState) {
         return blockState.getSoundType().getPlaceSound();
     }
 
+    //Forge: Sensitive version of BlockItem#getPlaceSound
+    protected SoundEvent getPlaceSound(BlockState state, Level world, BlockPos pos, Player entity) {
+        return state.getSoundType(world, pos, entity).getPlaceSound();
+    }
+
     public @Nullable BlockPlaceContext updatePlacementContext(final BlockPlaceContext context) {
         return context;
     }
@@ -188,6 +_,10 @@
 
     public void registerBlocks(final Map<Block, Item> map, final Item item) {
         map.put(this.getBlock(), item);
+    }
+
+    public void removeFromBlockToItemMap(Map<Block, Item> blockToItemMap, Item itemIn) {
+        blockToItemMap.remove(this.getBlock());
     }
 
     @Override
