--- a/net/minecraft/world/item/BrushItem.java
+++ b/net/minecraft/world/item/BrushItem.java
@@ -76,6 +_,7 @@
       } else {
          p_273619_.m_21253_();
       }
+      super.m_5929_(p_273467_, p_273619_, p_273316_, p_273101_); //TODO: Remove in 1.20, this is only here to trigger Forge's old onUseingTick method
    }
 
    public void m_271895_(Level p_272701_, BlockHitResult p_273758_, BlockState p_272753_, Vec3 p_272912_) {
