--- a/net/minecraft/world/item/CreativeModeTabs.java
+++ b/net/minecraft/world/item/CreativeModeTabs.java
@@ -1612,7 +_,7 @@
 
    private static void m_269182_(CreativeModeTab.Output p_270868_, HolderLookup<Enchantment> p_270903_, Set<EnchantmentCategory> p_270380_, CreativeModeTab.TabVisibility p_270407_) {
       p_270903_.m_214062_().map(Holder::m_203334_).filter((p_270008_) -> {
-         return p_270380_.contains(p_270008_.f_44672_);
+         return p_270008_.allowedInCreativeTab(Items.f_42690_, p_270380_);
       }).map((p_270038_) -> {
          return EnchantedBookItem.m_41161_(new EnchantmentInstance(p_270038_, p_270038_.m_6586_()));
       }).forEach((p_269989_) -> {
@@ -1622,7 +_,7 @@
 
    private static void m_269335_(CreativeModeTab.Output p_270961_, HolderLookup<Enchantment> p_270628_, Set<EnchantmentCategory> p_271024_, CreativeModeTab.TabVisibility p_270805_) {
       p_270628_.m_214062_().map(Holder::m_203334_).filter((p_269991_) -> {
-         return p_271024_.contains(p_269991_.f_44672_);
+         return p_269991_.allowedInCreativeTab(Items.f_42690_, p_271024_);
       }).flatMap((p_270024_) -> {
          return IntStream.rangeClosed(p_270024_.m_44702_(), p_270024_.m_6586_()).mapToObj((p_270006_) -> {
             return EnchantedBookItem.m_41161_(new EnchantmentInstance(p_270024_, p_270006_));
