--- a/net/minecraft/world/item/CreativeModeTab.java
+++ b/net/minecraft/world/item/CreativeModeTab.java
@@ -31,16 +_,40 @@
    private Consumer<List<ItemStack>> f_256965_;
    private final Supplier<ItemStack> f_256912_;
    private final CreativeModeTab.DisplayItemsGenerator f_256824_;
+   private net.minecraft.resources.ResourceLocation backgroundLocation;
+   private final boolean hasSearchBar;
+   private final int searchBarWidth;
+   private final net.minecraft.resources.ResourceLocation tabsImage;
+   private final int labelColor;
+   private final int slotColor;
+   public final List<net.minecraft.resources.ResourceLocation> tabsBefore;
+   public final List<net.minecraft.resources.ResourceLocation> tabsAfter;
 
-   CreativeModeTab(CreativeModeTab.Row p_260217_, int p_259557_, CreativeModeTab.Type p_260176_, Component p_260100_, Supplier<ItemStack> p_259543_, CreativeModeTab.DisplayItemsGenerator p_259085_) {
+   CreativeModeTab(CreativeModeTab.Row p_260217_, int p_259557_, CreativeModeTab.Type p_260176_, Component p_260100_, Supplier<ItemStack> p_259543_, CreativeModeTab.DisplayItemsGenerator p_259085_, net.minecraft.resources.ResourceLocation backgroundLocation, boolean hasSearchBar, int searchBarWidth, net.minecraft.resources.ResourceLocation tabsImage, int labelColor, int slotColor, List<net.minecraft.resources.ResourceLocation> tabsBefore, List<net.minecraft.resources.ResourceLocation> tabsAfter) {
       this.f_256931_ = p_260217_;
       this.f_256967_ = p_259557_;
       this.f_40764_ = p_260100_;
       this.f_256912_ = p_259543_;
       this.f_256824_ = p_259085_;
       this.f_256819_ = p_260176_;
-   }
-
+      this.backgroundLocation = backgroundLocation;
+      this.hasSearchBar = hasSearchBar;
+      this.searchBarWidth = searchBarWidth;
+      this.tabsImage = tabsImage;
+      this.labelColor = labelColor;
+      this.slotColor = slotColor;
+      this.tabsBefore = List.copyOf(tabsBefore);
+      this.tabsAfter = List.copyOf(tabsAfter);
+   }
+
+   protected CreativeModeTab(CreativeModeTab.Builder builder) {
+      this(builder.f_256796_, builder.f_256977_, builder.f_256847_, builder.f_256856_, builder.f_256981_, builder.f_256953_, builder.backgroundLocation, builder.hasSearchBar, builder.searchBarWidth, builder.tabsImage, builder.labelColor, builder.slotColor, builder.tabsBefore, builder.tabsAfter);
+   }
+
+   public static CreativeModeTab.Builder builder() {
+      return new CreativeModeTab.Builder(Row.TOP, 0);
+   }
+   /** @deprecated Forge: use {@link #builder()} **/ @Deprecated
    public static CreativeModeTab.Builder m_257815_(CreativeModeTab.Row p_259342_, int p_260312_) {
       return new CreativeModeTab.Builder(p_259342_, p_260312_);
    }
@@ -57,6 +_,10 @@
       return this.f_40770_;
    }
 
+   /**
+    * Forge: Use {@link #getBackgroundLocation()} instead.
+    */
+   @Deprecated
    public String m_40788_() {
       return this.f_40766_;
    }
@@ -98,7 +_,7 @@
       ResourceKey<CreativeModeTab> resourcekey = BuiltInRegistries.f_279662_.m_7854_(this).orElseThrow(() -> {
          return new IllegalStateException("Unregistered creative tab: " + this);
       });
-      this.f_256824_.m_257865_(p_270156_, creativemodetab$itemdisplaybuilder);
+      net.minecraftforge.common.ForgeHooks.onCreativeModeTabBuildContents(this, resourcekey, this.f_256824_, p_270156_, creativemodetab$itemdisplaybuilder);
       this.f_243839_ = creativemodetab$itemdisplaybuilder.f_244363_;
       this.f_243841_ = creativemodetab$itemdisplaybuilder.f_244585_;
       this.m_257466_();
@@ -127,9 +_,34 @@
 
    }
 
+   public net.minecraft.resources.ResourceLocation getBackgroundLocation() {
+      return backgroundLocation;
+   }
+
+   public boolean hasSearchBar() {
+      return this.hasSearchBar;
+   }
+
+   public int getSearchBarWidth() {
+      return searchBarWidth;
+   }
+
+   public net.minecraft.resources.ResourceLocation getTabsImage() {
+      return tabsImage;
+   }
+
+   public int getLabelColor() {
+      return labelColor;
+   }
+
+   public int getSlotColor() {
+      return slotColor;
+   }
+
    public static class Builder {
       private static final CreativeModeTab.DisplayItemsGenerator f_256756_ = (p_270422_, p_259433_) -> {
       };
+      private static final net.minecraft.resources.ResourceLocation CREATIVE_INVENTORY_TABS_IMAGE = new net.minecraft.resources.ResourceLocation("textures/gui/container/creative_inventory/tabs.png");
       private final CreativeModeTab.Row f_256796_;
       private final int f_256977_;
       private Component f_256856_ = Component.m_237119_();
@@ -142,6 +_,16 @@
       private boolean f_256854_ = false;
       private CreativeModeTab.Type f_256847_ = CreativeModeTab.Type.CATEGORY;
       private String f_257036_ = "items.png";
+      @org.jetbrains.annotations.Nullable
+      private net.minecraft.resources.ResourceLocation backgroundLocation;
+      private boolean hasSearchBar = false;
+      private int searchBarWidth = 89;
+      private net.minecraft.resources.ResourceLocation tabsImage = CREATIVE_INVENTORY_TABS_IMAGE;
+      private int labelColor = 4210752;
+      private int slotColor = -2130706433;
+      private java.util.function.Function<CreativeModeTab.Builder, CreativeModeTab> tabFactory = CreativeModeTab::new;
+      private final List<net.minecraft.resources.ResourceLocation> tabsBefore = new java.util.ArrayList<>();
+      private final List<net.minecraft.resources.ResourceLocation> tabsAfter = new java.util.ArrayList<>();
 
       public Builder(CreativeModeTab.Row p_259171_, int p_259661_) {
          this.f_256796_ = p_259171_;
@@ -180,11 +_,95 @@
 
       protected CreativeModeTab.Builder m_257623_(CreativeModeTab.Type p_259283_) {
          this.f_256847_ = p_259283_;
+         if (p_259283_ == Type.SEARCH)
+            return this.withSearchBar();
          return this;
       }
 
       public CreativeModeTab.Builder m_257609_(String p_259981_) {
-         this.f_257036_ = p_259981_;
+         return withBackgroundLocation(new net.minecraft.resources.ResourceLocation("textures/gui/container/creative_inventory/tab_" + p_259981_));
+      }
+
+      /**
+       * Sets the location of the tab background.
+       */
+      public CreativeModeTab.Builder withBackgroundLocation(net.minecraft.resources.ResourceLocation background) {
+         this.backgroundLocation = background;
+         return this;
+      }
+
+      /**
+       * Gives this tab a search bar.
+       * <p>Note that, if using a custom {@link #withBackgroundLocation(net.minecraft.resources.ResourceLocation) background image}, you will need to make sure that your image contains the input box and the scroll bar.</p>
+       */
+      public CreativeModeTab.Builder withSearchBar() {
+         this.hasSearchBar = true;
+         if (this.backgroundLocation == null)
+            return this.m_257609_("item_search.png");
+         return this;
+      }
+
+      /**
+       * Gives this tab a search bar, with a specific width.
+       * @param searchBarWidth the width of the search bar
+       */
+      public CreativeModeTab.Builder withSearchBar(int searchBarWidth) {
+         this.searchBarWidth = searchBarWidth;
+         return withSearchBar();
+      }
+
+      /**
+       * Sets the image of the tab to a custom resource location, instead of an item's texture.
+       */
+      public CreativeModeTab.Builder withTabsImage(net.minecraft.resources.ResourceLocation tabsImage) {
+         this.tabsImage = tabsImage;
+         return this;
+      }
+
+      /**
+       * Sets the color of the tab label.
+       */
+      public CreativeModeTab.Builder withLabelColor(int labelColor) {
+          this.labelColor = labelColor;
+          return this;
+      }
+
+      /**
+       * Sets the color of tab's slots.
+       */
+      public CreativeModeTab.Builder withSlotColor(int slotColor) {
+          this.slotColor = slotColor;
+          return this;
+      }
+
+      public CreativeModeTab.Builder withTabFactory(java.util.function.Function<CreativeModeTab.Builder, CreativeModeTab> tabFactory) {
+         this.tabFactory = tabFactory;
+         return this;
+      }
+
+      /** Define tabs that should come <i>before</i> this tab. This tab will be placed <strong>after</strong> the {@code tabs}. **/
+      public CreativeModeTab.Builder withTabsBefore(net.minecraft.resources.ResourceLocation... tabs) {
+         this.tabsBefore.addAll(List.of(tabs));
+         return this;
+      }
+
+      /** Define tabs that should come <i>after</i> this tab. This tab will be placed <strong>before</strong> the {@code tabs}.**/
+      public CreativeModeTab.Builder withTabsAfter(net.minecraft.resources.ResourceLocation... tabs) {
+         this.tabsAfter.addAll(List.of(tabs));
+         return this;
+      }
+
+      /** Define tabs that should come <i>before</i> this tab. This tab will be placed <strong>after</strong> the {@code tabs}. **/
+      @SafeVarargs
+      public final CreativeModeTab.Builder withTabsBefore(net.minecraft.resources.ResourceKey<CreativeModeTab>... tabs) {
+         java.util.stream.Stream.of(tabs).map(net.minecraft.resources.ResourceKey::m_135782_).forEach(this.tabsBefore::add);
+         return this;
+      }
+
+      /** Define tabs that should come <i>after</i> this tab. This tab will be placed <strong>before</strong> the {@code tabs}.**/
+      @SafeVarargs
+      public final CreativeModeTab.Builder withTabsAfter(net.minecraft.resources.ResourceKey<CreativeModeTab>... tabs) {
+         java.util.stream.Stream.of(tabs).map(net.minecraft.resources.ResourceKey::m_135782_).forEach(this.tabsAfter::add);
          return this;
       }
 
@@ -192,11 +_,12 @@
          if ((this.f_256847_ == CreativeModeTab.Type.HOTBAR || this.f_256847_ == CreativeModeTab.Type.INVENTORY) && this.f_256953_ != f_256756_) {
             throw new IllegalStateException("Special tabs can't have display items");
          } else {
-            CreativeModeTab creativemodetab = new CreativeModeTab(this.f_256796_, this.f_256977_, this.f_256847_, this.f_256856_, this.f_256981_, this.f_256953_);
+            CreativeModeTab creativemodetab = tabFactory.apply(this);
             creativemodetab.f_257018_ = this.f_256854_;
             creativemodetab.f_40768_ = this.f_256851_;
             creativemodetab.f_40767_ = this.f_256992_;
             creativemodetab.f_40766_ = this.f_257036_;
+            creativemodetab.backgroundLocation = this.backgroundLocation != null ? this.backgroundLocation : new net.minecraft.resources.ResourceLocation("textures/gui/container/creative_inventory/tab_" + this.f_257036_);
             return creativemodetab;
          }
       }
