--- a/net/minecraft/world/item/CreativeModeTabs.java
+++ b/net/minecraft/world/item/CreativeModeTabs.java
@@ -1612,7 +_,7 @@
 
    private static void generateEnchantmentBookTypesOnlyMaxLevel(CreativeModeTab.Output p_270868_, HolderLookup<Enchantment> p_270903_, Set<EnchantmentCategory> p_270380_, CreativeModeTab.TabVisibility p_270407_) {
       p_270903_.listElements().map(Holder::value).filter((p_270008_) -> {
-         return p_270380_.contains(p_270008_.category);
+         return p_270008_.allowedInCreativeTab(Items.ENCHANTED_BOOK, p_270380_);
       }).map((p_270038_) -> {
          return EnchantedBookItem.createForEnchantment(new EnchantmentInstance(p_270038_, p_270038_.getMaxLevel()));
       }).forEach((p_269989_) -> {
@@ -1622,7 +_,7 @@
 
    private static void generateEnchantmentBookTypesAllLevels(CreativeModeTab.Output p_270961_, HolderLookup<Enchantment> p_270628_, Set<EnchantmentCategory> p_271024_, CreativeModeTab.TabVisibility p_270805_) {
       p_270628_.listElements().map(Holder::value).filter((p_269991_) -> {
-         return p_271024_.contains(p_269991_.category);
+         return p_269991_.allowedInCreativeTab(Items.ENCHANTED_BOOK, p_271024_);
       }).flatMap((p_270024_) -> {
          return IntStream.rangeClosed(p_270024_.getMinLevel(), p_270024_.getMaxLevel()).mapToObj((p_270006_) -> {
             return EnchantedBookItem.createForEnchantment(new EnchantmentInstance(p_270024_, p_270006_));
