--- a/net/minecraft/world/item/DyeableHorseArmorItem.java
+++ b/net/minecraft/world/item/DyeableHorseArmorItem.java
@@ -4,4 +_,7 @@
    public DyeableHorseArmorItem(int p_41110_, String p_41111_, Item.Properties p_41112_) {
       super(p_41110_, p_41111_, p_41112_);
    }
+   public DyeableHorseArmorItem(int p_41110_, net.minecraft.resources.ResourceLocation p_41111_, Item.Properties p_41112_) {
+      super(p_41110_, p_41111_, p_41112_);
+   }
 }
