--- a/net/minecraft/world/item/DyeColor.java
+++ b/net/minecraft/world/item/DyeColor.java
@@ -44,6 +_,7 @@
     private final int textureDiffuseColor;
     private final int fireworkColor;
     private final int textColor;
+    private final net.minecraft.tags.TagKey<Item> tag;
 
     private DyeColor(final int p_41046_, final String p_41047_, final int p_41048_, final MapColor p_285297_, final int p_41050_, final int p_41051_) {
         this.id = p_41046_;
@@ -52,6 +_,7 @@
         this.textColor = p_41051_;
         this.textureDiffuseColor = FastColor.ARGB32.opaque(p_41048_);
         this.fireworkColor = p_41050_;
+        this.tag = net.minecraft.tags.ItemTags.create("forge", "dyes/" + p_41047_);
     }
 
     public int getId() {
@@ -102,5 +_,14 @@
     @Override
     public String getSerializedName() {
         return this.name;
+    }
+
+    public net.minecraft.tags.TagKey<Item> getTag() {
+        return tag;
+    }
+
+    @Nullable
+    public static DyeColor getColor(ItemStack stack) {
+        return net.minecraftforge.common.ForgeHooks.getDyeColorFromItemStack(stack);
     }
 }
