--- a/net/minecraft/world/item/ElytraItem.java
+++ b/net/minecraft/world/item/ElytraItem.java
@@ -27,6 +_,25 @@
       return this.m_269277_(this, p_41137_, p_41138_, p_41139_);
    }
 
+   @Override
+   public boolean canElytraFly(ItemStack stack, net.minecraft.world.entity.LivingEntity entity) {
+      return ElytraItem.m_41140_(stack);
+   }
+
+   @Override
+   public boolean elytraFlightTick(ItemStack stack, net.minecraft.world.entity.LivingEntity entity, int flightTicks) {
+      if (!entity.m_9236_().f_46443_) {
+         int nextFlightTick = flightTicks + 1;
+         if (nextFlightTick % 10 == 0) {
+            if (nextFlightTick % 20 == 0) {
+               stack.m_41622_(1, entity, e -> e.m_21166_(net.minecraft.world.entity.EquipmentSlot.CHEST));
+            }
+            entity.m_146850_(net.minecraft.world.level.gameevent.GameEvent.f_223705_);
+         }
+      }
+      return true;
+   }
+
    public SoundEvent m_150681_() {
       return SoundEvents.f_11674_;
    }
