--- a/net/minecraft/world/item/FishingRodItem.java
+++ b/net/minecraft/world/item/FishingRodItem.java
@@ -24,7 +_,11 @@
         if (p_41291_.fishing != null) {
             if (!p_41290_.isClientSide) {
                 int i = p_41291_.fishing.retrieve(itemstack);
+                ItemStack original = itemstack.copy();
                 itemstack.hurtAndBreak(i, p_41291_, LivingEntity.getSlotForHand(p_41292_));
+                if (itemstack.isEmpty()) {
+                    net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(p_41291_, original, LivingEntity.getSlotForHand(p_41292_));
+                }
             }
 
             p_41290_.playSound(
@@ -65,5 +_,10 @@
     @Override
     public int getEnchantmentValue() {
         return 1;
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.minecraftforge.common.ToolAction toolAction) {
+        return net.minecraftforge.common.ToolActions.DEFAULT_FISHING_ROD_ACTIONS.contains(toolAction);
     }
 }
