--- a/net/minecraft/world/item/HoeItem.java
+++ b/net/minecraft/world/item/HoeItem.java
@@ -26,13 +_,16 @@
    }, m_150849_(Blocks.f_50493_.m_49966_(), Items.f_151017_))));
 
    public HoeItem(Tier p_41336_, int p_41337_, float p_41338_, Item.Properties p_41339_) {
-      super((float)p_41337_, p_41338_, p_41336_, BlockTags.f_144281_, p_41339_);
+      super((float)p_41337_, p_41338_, p_41336_, BlockTags.f_144281_, p_41339_.addToolType(net.minecraftforge.common.ToolType.HOE, p_41336_.m_6604_()));
    }
 
    public InteractionResult m_6225_(UseOnContext p_41341_) {
       Level level = p_41341_.m_43725_();
       BlockPos blockpos = p_41341_.m_8083_();
       Pair<Predicate<UseOnContext>, Consumer<UseOnContext>> pair = f_41332_.get(level.m_8055_(blockpos).m_60734_());
+      int hook = net.minecraftforge.event.ForgeEventFactory.onHoeUse(p_41341_);
+      if (hook != 0) return hook > 0 ? InteractionResult.SUCCESS : InteractionResult.FAIL;
+      if (p_41341_.m_43719_() != Direction.DOWN && level.m_46859_(blockpos.m_7494_())) {
       if (pair == null) {
          return InteractionResult.PASS;
       } else {
@@ -55,6 +_,9 @@
             return InteractionResult.PASS;
          }
       }
+      }
+
+      return InteractionResult.PASS;
    }
 
    public static Consumer<UseOnContext> m_150858_(BlockState p_150859_) {
