--- a/net/minecraft/world/item/HoeItem.java
+++ b/net/minecraft/world/item/HoeItem.java
@@ -30,6 +_,8 @@
    }
 
    public InteractionResult m_6225_(UseOnContext p_41341_) {
+      int hook = net.minecraftforge.event.ForgeEventFactory.onHoeUse(p_41341_);
+      if (hook != 0) return hook > 0 ? InteractionResult.SUCCESS : InteractionResult.FAIL;
       Level level = p_41341_.m_43725_();
       BlockPos blockpos = p_41341_.m_8083_();
       Pair<Predicate<UseOnContext>, Consumer<UseOnContext>> pair = f_41332_.get(level.m_8055_(blockpos).m_60734_());
@@ -72,5 +_,10 @@
 
    public static boolean m_150856_(UseOnContext p_150857_) {
       return p_150857_.m_43719_() != Direction.DOWN && p_150857_.m_43725_().m_8055_(p_150857_.m_8083_().m_7494_()).m_60795_();
+   }
+
+   @Override
+   public boolean canPerformAction(ItemStack stack, net.minecraftforge.common.ToolAction toolAction) {
+      return net.minecraftforge.common.ToolActions.DEFAULT_HOE_ACTIONS.contains(toolAction);
    }
 }
