--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -99,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public final class ItemStack implements DataComponentHolder, ItemInstance {
+public final class ItemStack extends net.minecraftforge.common.capabilities.CapabilityProvider.ItemStacks implements DataComponentHolder, ItemInstance, net.minecraftforge.common.extensions.IForgeItemStack {
     private static final List<Component> OP_NBT_WARNING = List.of(
         Component.translatable("item.op_warning.line1").withStyle(ChatFormatting.RED, ChatFormatting.BOLD),
         Component.translatable("item.op_warning.line2").withStyle(ChatFormatting.RED),
@@ -257,12 +_,15 @@
     }
 
     private ItemStack(final Holder<Item> item, final int count, final PatchedDataComponentMap components) {
+        super(true);
         this.item = item;
         this.count = count;
         this.components = components;
+        gatherCapabilities(() -> this.item.get().getCapabilityProvider(this));
     }
 
     private ItemStack(final @Nullable Void nullMarker) {
+        super(false);
         this.item = null;
         this.components = new PatchedDataComponentMap(DataComponentMap.EMPTY);
     }
@@ -355,6 +_,15 @@
     }
 
     public InteractionResult useOn(final UseOnContext context) {
+        if (!context.getLevel().isClientSide()) return net.minecraftforge.common.ForgeHooks.onPlaceItemIntoWorld(context);
+        return onItemUse(context, (c) -> getItem().useOn(context));
+    }
+
+    public InteractionResult onItemUseFirst(UseOnContext context) {
+        return onItemUse(context, (c) -> getItem().onItemUseFirst(this, context));
+    }
+
+    private InteractionResult onItemUse(UseOnContext context, java.util.function.Function<UseOnContext, InteractionResult> callback) {
         Player player = context.getPlayer();
         BlockPos pos = context.getClickedPos();
         if (player != null && !player.getAbilities().mayBuild && !this.canPlaceOnBlockInAdventureMode(new BlockInWorld(context.getLevel(), pos, false))) {
@@ -362,7 +_,7 @@
         }
 
         Item usedItem = this.getItem();
-        InteractionResult result = usedItem.useOn(context);
+        InteractionResult result = callback.apply(context);
         if (player != null && result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
             player.awardStat(Stats.ITEM_USED.get(usedItem));
         }
@@ -442,18 +_,26 @@
     }
 
     public void hurtAndBreak(final int amount, final ServerLevel level, final @Nullable ServerPlayer player, final Consumer<Item> onBreak) {
-        int newAmount = this.processDurabilityChange(amount, level, player);
+        // FORGE: use context-sensitive sister of processDurabilityChange that calls IForgeItem.damageItem
+        int newAmount = this.processDurabilityChange(amount, level, player, true, onBreak);
         if (newAmount != 0) {
             this.applyDamage(this.getDamageValue() + newAmount, player, onBreak);
         }
     }
 
     private int processDurabilityChange(final int amount, final ServerLevel level, final @Nullable ServerPlayer player) {
+        return this.processDurabilityChange(amount, level, player, false, item -> { });
+    }
+
+    /** FORGE: context-sensitive sister of processDurabilityChange that calls IForgeItem.damageItem */
+    private int processDurabilityChange(int amount, final ServerLevel level, final @Nullable ServerPlayer player, final boolean canBreak, final Consumer<Item> onBreak) {
         if (!this.isDamageableItem()) {
             return 0;
         } else if (player != null && player.hasInfiniteMaterials()) {
             return 0;
         } else {
+            // FORGE: modify the base damage based on the item's impl of IForgeItem.damageItem
+            amount = this.damageItem(amount, level, player, canBreak, onBreak);
             return amount > 0 ? EnchantmentHelper.processDurabilityChange(level, this, amount) : amount;
         }
     }
@@ -490,7 +_,13 @@
     public void hurtAndBreak(final int amount, final LivingEntity owner, final EquipmentSlot slot) {
         if (owner.level() instanceof ServerLevel serverLevel) {
             this.hurtAndBreak(
-                amount, serverLevel, owner instanceof ServerPlayer player ? player : null, brokenItem -> owner.onEquippedItemBroken(brokenItem, slot)
+                amount, serverLevel, owner instanceof ServerPlayer player ? player : null, brokenItem -> {
+                    if (owner instanceof Player ownerPlayer) {
+                        net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(ownerPlayer, this, slot);
+                        if (ownerPlayer.getUseItem() == this) ownerPlayer.stopUsingItem(); // Forge: fix MC-168573
+                    }
+                    owner.onEquippedItemBroken(brokenItem, slot);
+                }
             );
         }
     }
@@ -857,6 +_,7 @@
             List<Component> lines = Lists.newArrayList();
             lines.add(this.getStyledHoverName());
             this.addDetailsToTooltip(context, display, player, tooltipFlag, lines::add);
+            net.minecraftforge.event.ForgeEventFactory.onItemTooltip(this, player, lines, tooltipFlag);
             return lines;
         }
     }
@@ -1105,6 +_,7 @@
         }
     }
 
+    /** @deprecated Forge: Use {@linkplain net.minecraftforge.common.extensions.IForgeItemStack#onDestroyed(ItemEntity, net.minecraft.world.damagesource.DamageSource) damage source sensitive version} */
     public void onDestroyed(final ItemEntity itemEntity) {
         this.getItem().onDestroyed(itemEntity);
     }
