--- a/net/minecraft/world/item/Items.java
+++ b/net/minecraft/world/item/Items.java
@@ -2155,7 +_,25 @@
     }
 
     public static Item registerBlock(Block p_252092_, Block... p_248886_) {
-        Item item = registerBlock(p_252092_);
+        Item item = registerBlock(p_252092_, (block_, prop_) -> {
+            return new BlockItem(block_, prop_) {
+                @Override
+                public void registerBlocks(java.util.Map<Block, Item> map, Item self) {
+                    super.registerBlocks(map, self);
+                    for (var b : p_248886_) {
+                        map.put(b, self);
+                    }
+                }
+
+                @Override
+                public void removeFromBlockToItemMap(java.util.Map<Block, Item> map, Item self) {
+                    super.removeFromBlockToItemMap(map, self);
+                    for (var b : p_248886_) {
+                        map.remove(b);
+                    }
+                }
+            };
+        });
 
         for (Block block : p_248886_) {
             Item.BY_BLOCK.put(block, item);
@@ -2194,10 +_,6 @@
 
     public static Item registerItem(ResourceKey<Item> p_364577_, Function<Item.Properties, Item> p_366144_, Item.Properties p_369845_) {
         Item item = p_366144_.apply(p_369845_.setId(p_364577_));
-        if (item instanceof BlockItem blockitem) {
-            blockitem.registerBlocks(Item.BY_BLOCK, item);
-        }
-
         return Registry.register(BuiltInRegistries.ITEM, p_364577_, item);
     }
 }
