--- a/net/minecraft/world/item/ItemDisplayContext.java
+++ b/net/minecraft/world/item/ItemDisplayContext.java
@@ -5,7 +_,7 @@
 import net.minecraft.util.ByIdMap;
 import net.minecraft.util.StringRepresentable;
 
-public enum ItemDisplayContext implements StringRepresentable {
+public enum ItemDisplayContext implements StringRepresentable, net.minecraftforge.common.IExtensibleEnum {
     NONE(0, "none"),
     THIRD_PERSON_LEFT_HAND(1, "thirdperson_lefthand"),
     THIRD_PERSON_RIGHT_HAND(2, "thirdperson_righthand"),
@@ -17,14 +_,16 @@
     FIXED(8, "fixed"),
     ON_SHELF(9, "on_shelf");
 
-    public static final Codec<ItemDisplayContext> CODEC = StringRepresentable.fromEnum(ItemDisplayContext::values);
-    public static final IntFunction<ItemDisplayContext> BY_ID = ByIdMap.continuous(ItemDisplayContext::getId, values(), ByIdMap.OutOfBoundsStrategy.ZERO);
+    public static final Codec<ItemDisplayContext> CODEC =  com.mojang.serialization.Codec.lazyInitialized(() -> StringRepresentable.fromEnum(ItemDisplayContext::values));
+    public static final IntFunction<ItemDisplayContext> BY_ID = id -> java.util.Objects.requireNonNullElse(((net.minecraftforge.registries.IForgeRegistryInternal<ItemDisplayContext>) net.minecraftforge.registries.ForgeRegistries.DISPLAY_CONTEXTS.get()).getValue(id < 0 ? Byte.MAX_VALUE + -id : id), NONE);
     private byte id;
     private final String name;
 
     ItemDisplayContext(final int id, final String name) {
         this.name = name;
         this.id = (byte)id;
+        this.isModded = false;
+        this.fallback = null;
     }
 
     @Override
@@ -43,4 +_,29 @@
     public boolean leftHand() {
         return this == FIRST_PERSON_LEFT_HAND || this == THIRD_PERSON_LEFT_HAND;
     }
+
+    private final boolean isModded;
+    @javax.annotation.Nullable
+    private final ItemDisplayContext fallback;
+
+    private ItemDisplayContext(net.minecraft.resources.Identifier serializeName, ItemDisplayContext fallback) {
+        this.id = 0; // If the context is never registered then it should be treated as the NONE one
+        this.name = java.util.Objects.requireNonNull(serializeName, "Modded ItemDisplayContexts must have a non-null serializeName").toString();
+        this.isModded = true;
+        this.fallback = fallback;
+    }
+
+    public boolean isModded() {
+        return isModded;
+    }
+
+    public @javax.annotation.Nullable ItemDisplayContext fallback() {
+        return fallback;
+    }
+
+    public static ItemDisplayContext create(String keyName, net.minecraft.resources.Identifier serializedName, @javax.annotation.Nullable ItemDisplayContext fallback) {
+        throw new IllegalStateException("Enum not extended!");
+    }
+
+    public static final net.minecraftforge.registries.IForgeRegistry.AddCallback<ItemDisplayContext> ADD_CALLBACK = (owner, stage, id, key, obj, oldObj) -> obj.id = id > Byte.MAX_VALUE ? (byte) -(id - Byte.MAX_VALUE) : (byte) id; // if the ID is > 127 start doing negative IDs
 }
