--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -58,8 +_,18 @@
 
    @Nullable
    public static MapItemSavedData getSavedData(ItemStack p_42854_, Level p_42855_) {
-      Integer integer = getMapId(p_42854_);
-      return getSavedData(integer, p_42855_);
+      // Forge: Add instance method so that mods can override
+      Item map = p_42854_.getItem();
+      if(map instanceof MapItem) {
+         return ((MapItem)map).getCustomMapData(p_42854_, p_42855_);
+      }
+      return null;
+   }
+
+   @Nullable
+   protected MapItemSavedData getCustomMapData(ItemStack p_42910_, Level p_42911_) {
+      Integer integer = getMapId(p_42910_);
+      return getSavedData(integer, p_42911_);
    }
 
    @Nullable
