--- a/net/minecraft/world/item/PickaxeItem.java
+++ b/net/minecraft/world/item/PickaxeItem.java
@@ -4,6 +_,6 @@
 
 public class PickaxeItem extends DiggerItem {
    public PickaxeItem(Tier p_42961_, int p_42962_, float p_42963_, Item.Properties p_42964_) {
-      super((float)p_42962_, p_42963_, p_42961_, BlockTags.f_144282_, p_42964_);
+      super((float)p_42962_, p_42963_, p_42961_, BlockTags.f_144282_, p_42964_.addToolType(net.minecraftforge.common.ToolType.PICKAXE, p_42961_.m_6604_()));
    }
 }
