--- a/net/minecraft/world/item/RecordItem.java
+++ b/net/minecraft/world/item/RecordItem.java
@@ -22,17 +_,42 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 
 public class RecordItem extends Item {
+   @Deprecated // Forge: refer to WorldRender#playRecord. Modders: there's no need to reflectively modify this map!
    private static final Map<SoundEvent, RecordItem> f_43032_ = Maps.newHashMap();
    private final int f_43033_;
+   @Deprecated // Forge: refer to soundSupplier
    private final SoundEvent f_43034_;
    private final int f_238749_;
+   private final java.util.function.Supplier<SoundEvent> soundSupplier;
 
+   @Deprecated // Forge: Use the constructor that takes a supplier instead
    public RecordItem(int p_239614_, SoundEvent p_239615_, Item.Properties p_239616_, int p_239617_) {
       super(p_239616_);
       this.f_43033_ = p_239614_;
       this.f_43034_ = p_239615_;
       this.f_238749_ = p_239617_ * 20;
       f_43032_.put(this.f_43034_, this);
+      this.soundSupplier = net.minecraftforge.registries.ForgeRegistries.SOUND_EVENTS.getDelegateOrThrow(this.f_43034_);
+   }
+
+   /**
+    * For mod use, allows to create a music disc without having to create a new
+    * SoundEvent before their registry event is fired.
+    *
+    * @param comparatorValue The value this music disc should output on the comparator. Must be between 0 and 15.
+    * @param soundSupplier A supplier that provides the sound that should be played. Use a
+    *                      {@link net.minecraftforge.registries.RegistryObject<SoundEvent>} or a
+    *                      {@link net.minecraft.core.Holder<SoundEvent>} for this parameter.
+    * @param builder A set of {@link Item.Properties} that describe this item.
+    * @param lengthInTicks The length of the music disc track in ticks.
+    */
+   public RecordItem(int comparatorValue, java.util.function.Supplier<SoundEvent> soundSupplier, Item.Properties builder, int lengthInTicks)
+   {
+      super(builder);
+      this.f_43033_ = comparatorValue;
+      this.f_43034_ = null;
+      this.soundSupplier = soundSupplier;
+      this.f_238749_ = lengthInTicks;
    }
 
    public InteractionResult m_6225_(UseOnContext p_43048_) {
@@ -80,7 +_,7 @@
    }
 
    public SoundEvent m_43051_() {
-      return this.f_43034_;
+      return this.soundSupplier.get();
    }
 
    public int m_43036_() {
