--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -49,6 +_,31 @@
       }
    }
 
+   @Override
+   public net.minecraft.world.InteractionResult m_6880_(ItemStack stack, net.minecraft.world.entity.player.Player playerIn, LivingEntity entity, net.minecraft.world.InteractionHand hand) {
+      if (entity instanceof net.minecraftforge.common.IForgeShearable target) {
+         if (entity.m_9236_().f_46443_) return net.minecraft.world.InteractionResult.SUCCESS;
+         BlockPos pos = BlockPos.m_274446_(entity.m_20182_());
+         if (target.isShearable(stack, entity.m_9236_(), pos)) {
+            java.util.List<ItemStack> drops = target.onSheared(playerIn, stack, entity.m_9236_(), pos,
+                    net.minecraft.world.item.enchantment.EnchantmentHelper.m_44843_(net.minecraft.world.item.enchantment.Enchantments.f_44987_, stack));
+            java.util.Random rand = new java.util.Random();
+            drops.forEach(d -> {
+               net.minecraft.world.entity.item.ItemEntity ent = entity.m_5552_(d, 1.0F);
+               ent.m_20256_(ent.m_20184_().m_82520_((double)((rand.nextFloat() - rand.nextFloat()) * 0.1F), (double)(rand.nextFloat() * 0.05F), (double)((rand.nextFloat() - rand.nextFloat()) * 0.1F)));
+            });
+            stack.m_41622_(1, playerIn, e -> e.m_21190_(hand));
+         }
+         return net.minecraft.world.InteractionResult.SUCCESS;
+      }
+      return net.minecraft.world.InteractionResult.PASS;
+   }
+
+   @Override
+   public boolean canPerformAction(ItemStack stack, net.minecraftforge.common.ToolAction toolAction) {
+      return net.minecraftforge.common.ToolActions.DEFAULT_SHEARS_ACTIONS.contains(toolAction);
+   }
+
    public InteractionResult m_6225_(UseOnContext p_186371_) {
       Level level = p_186371_.m_43725_();
       BlockPos blockpos = p_186371_.m_8083_();
