--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -43,9 +_,9 @@
         }
 
         Player player = context.getPlayer();
-        BlockState newState = FLATTENABLES.get(blockState.getBlock());
+        BlockState newState = blockState.getToolModifiedState(context, net.minecraftforge.common.ToolActions.SHOVEL_FLATTEN, false);
         BlockState updatedState = null;
-        if (newState != null && level.getBlockState(pos.above()).isAir()) {
+        if (newState != null && level.isEmptyBlock(pos.above())) {
             level.playSound(player, pos, SoundEvents.SHOVEL_FLATTEN, SoundSource.BLOCKS, 1.0F, 1.0F);
             updatedState = newState;
         } else if (blockState.getBlock() instanceof CampfireBlock && blockState.getValue(CampfireBlock.LIT)) {
@@ -70,5 +_,15 @@
         } else {
             return InteractionResult.PASS;
         }
+    }
+
+    @org.jetbrains.annotations.Nullable
+    public static BlockState getShovelPathingState(BlockState originalState) {
+        return FLATTENABLES.get(originalState.getBlock());
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.minecraftforge.common.ToolAction toolAction) {
+        return net.minecraftforge.common.ToolActions.DEFAULT_SHOVEL_ACTIONS.contains(toolAction);
     }
 }
