--- a/net/minecraft/world/item/StandingAndWallBlockItem.java
+++ b/net/minecraft/world/item/StandingAndWallBlockItem.java
@@ -48,4 +_,9 @@
       super.m_6192_(p_43252_, p_43253_);
       p_43252_.put(this.f_43246_, p_43253_);
    }
+
+   public void removeFromBlockToItemMap(Map<Block, Item> blockToItemMap, Item itemIn) {
+      super.removeFromBlockToItemMap(blockToItemMap, itemIn);
+      blockToItemMap.remove(this.f_43246_);
+   }
 }
