--- ../src-base/minecraft/net/minecraft/client/MinecraftApplet.java
+++ ../src-work/minecraft/net/minecraft/client/MinecraftApplet.java
@@ -1,5 +1,6 @@
 package net.minecraft.client;
 
+import cpw.mods.fml.relauncher.FMLRelauncher;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.applet.Applet;
@@ -15,6 +16,11 @@
     private Thread field_71482_c = null;
 
     public void init()
+    {
+        FMLRelauncher.appletEntry(this);
+    }
+
+    public void fmlInitReentry()
     {
         this.field_71483_a = new CanvasMinecraftApplet(this);
         boolean var1 = "true".equalsIgnoreCase(this.getParameter("fullscreen"));
@@ -62,6 +68,11 @@
 
     public void start()
     {
+        FMLRelauncher.appletStart(this);
+    }
+
+    public void fmlStartReentry()
+    {
         if (this.field_71481_b != null)
         {
             this.field_71481_b.field_71445_n = false;
