--- ../src-base/minecraft/net/minecraft/client/Minecraft.java
+++ ../src-work/minecraft/net/minecraft/client/Minecraft.java
@@ -9,6 +9,9 @@
 import com.google.common.util.concurrent.ListenableFutureTask;
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
+import cpw.mods.fml.client.FMLClientHandler;
+import cpw.mods.fml.common.FMLCommonHandler;
+import cpw.mods.fml.common.StartupQuery;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import io.netty.util.concurrent.GenericFutureListener;
@@ -169,6 +172,8 @@
 import org.lwjgl.opengl.PixelFormat;
 import org.lwjgl.util.glu.GLU;
 
+import com.google.common.collect.MapDifference;
+
 @SideOnly(Side.CLIENT)
 public class Minecraft implements IPlayerUsage
 {
@@ -344,21 +349,23 @@
         File file2 = new File(file1, "crash-" + (new SimpleDateFormat("yyyy-MM-dd_HH.mm.ss")).format(new Date()) + "-client.txt");
         System.out.println(p_71377_1_.func_71502_e());
 
+        int retVal;
         if (p_71377_1_.func_71497_f() != null)
         {
             System.out.println("#@!@# Game crashed! Crash report saved to: #@!@# " + p_71377_1_.func_71497_f());
-            System.exit(-1);
+            retVal = -1;
         }
         else if (p_71377_1_.func_147149_a(file2))
         {
             System.out.println("#@!@# Game crashed! Crash report saved to: #@!@# " + file2.getAbsolutePath());
-            System.exit(-1);
+            retVal = -1;
         }
         else
         {
             System.out.println("#@?@# Game crashed! Crash report could not be saved. #@?@#");
-            System.exit(-2);
+            retVal = -2;
         }
+        FMLCommonHandler.instance().handleExit(retVal);
     }
 
     public void func_71367_a(String p_71367_1_, int p_71367_2_)
@@ -471,11 +478,11 @@
         this.field_110451_am = new SimpleReloadableResourceManager(this.field_110452_an);
         this.field_135017_as = new LanguageManager(this.field_110452_an, this.field_71474_y.field_74363_ab);
         this.field_110451_am.func_110542_a(this.field_135017_as);
-        this.func_110436_a();
+        FMLClientHandler.instance().beginMinecraftLoading(this, this.field_110449_ao, this.field_110451_am);
         this.field_71446_o = new TextureManager(this.field_110451_am);
         this.field_110451_am.func_110542_a(this.field_71446_o);
         this.field_152350_aA = new SkinManager(this.field_71446_o, new File(this.field_110446_Y, "skins"), this.field_152355_az);
-        this.func_71357_I();
+        cpw.mods.fml.client.SplashProgress.drawVanillaScreen();
         this.field_147127_av = new SoundHandler(this.field_110451_am, this.field_71474_y);
         this.field_110451_am.func_110542_a(this.field_147127_av);
         this.field_147126_aw = new MusicTicker(this);
@@ -492,7 +499,10 @@
         this.field_110451_am.func_110542_a(this.field_71464_q);
         this.field_110451_am.func_110542_a(new GrassColorReloadListener());
         this.field_110451_am.func_110542_a(new FoliageColorReloadListener());
+        cpw.mods.fml.common.ProgressManager.ProgressBar bar= cpw.mods.fml.common.ProgressManager.push("Rendering Setup", 9, true);
+        bar.step("Loading Render Manager");
         RenderManager.field_78727_a.field_78721_f = new ItemRenderer(this);
+        bar.step("Loading Entity Renderer");
         this.field_71460_t = new EntityRenderer(this, this.field_110451_am);
         this.field_110451_am.func_110542_a(this.field_71460_t);
         AchievementList.field_76004_f.func_75988_a(new IStatStringFormat()
@@ -510,6 +520,7 @@
                 }
             }
         });
+        bar.step("Loading GL properties");
         this.field_71417_B = new MouseHelper();
         this.func_71361_d("Pre startup");
         GL11.glEnable(GL11.GL_TEXTURE_2D);
@@ -524,30 +535,39 @@
         GL11.glLoadIdentity();
         GL11.glMatrixMode(GL11.GL_MODELVIEW);
         this.func_71361_d("Startup");
+        bar.step("Render Global instance");
         this.field_71438_f = new RenderGlobal(this);
-        this.field_147128_au = new TextureMap(0, "textures/blocks");
+        bar.step("Building Blocks Texture");
+        this.field_147128_au = new TextureMap(0, "textures/blocks", true);
+        bar.step("Anisotropy and Mipmaps");
         this.field_147128_au.func_147632_b(this.field_71474_y.field_151443_J);
         this.field_147128_au.func_147633_a(this.field_71474_y.field_151442_I);
+        bar.step("Loading Blocks Texture");
         this.field_71446_o.func_130088_a(TextureMap.field_110575_b, this.field_147128_au);
-        this.field_71446_o.func_130088_a(TextureMap.field_110576_c, new TextureMap(1, "textures/items"));
+        bar.step("Loading Items Texture");
+        this.field_71446_o.func_130088_a(TextureMap.field_110576_c, new TextureMap(1, "textures/items", true));
+        bar.step("Viewport");
         GL11.glViewport(0, 0, this.field_71443_c, this.field_71440_d);
         this.field_71452_i = new EffectRenderer(this.field_71441_e, this.field_71446_o);
+        cpw.mods.fml.common.ProgressManager.pop(bar);
+        FMLClientHandler.instance().finishMinecraftLoading();
         this.func_71361_d("Post startup");
         this.field_71456_v = new GuiIngame(this);
 
         if (this.field_71475_ae != null)
         {
-            this.func_147108_a(new GuiConnecting(new GuiMainMenu(), this, this.field_71475_ae, this.field_71477_af));
+            FMLClientHandler.instance().connectToServerAtStartup(this.field_71475_ae, this.field_71477_af);
         }
         else
         {
             this.func_147108_a(new GuiMainMenu());
         }
 
-        this.field_71446_o.func_147645_c(this.field_152354_ay);
+        cpw.mods.fml.client.SplashProgress.clearVanillaResources(field_71446_o, field_152354_ay);
         this.field_152354_ay = null;
         this.field_71461_s = new LoadingScreenRenderer(this);
 
+        FMLClientHandler.instance().onInitializationComplete();
         if (this.field_71474_y.field_74353_u && !this.field_71431_Q)
         {
             this.func_71352_k();
@@ -970,9 +990,11 @@
 
         if (!this.field_71454_w)
         {
+            FMLCommonHandler.instance().onRenderTickStart(this.field_71428_T.field_74281_c);
             this.field_71424_I.func_76318_c("gameRenderer");
             this.field_71460_t.func_78480_b(this.field_71428_T.field_74281_c);
             this.field_71424_I.func_76319_b();
+            FMLCommonHandler.instance().onRenderTickEnd(this.field_71428_T.field_74281_c);
         }
 
         GL11.glFlush();
@@ -1554,6 +1576,8 @@
             --this.field_71467_ac;
         }
 
+        FMLCommonHandler.instance().onPreClientTick();
+
         this.field_71424_I.func_76320_a("gui");
 
         if (!this.field_71445_n)
@@ -1698,6 +1722,7 @@
                         this.field_71462_r.func_146274_d();
                     }
                 }
+                FMLCommonHandler.instance().fireMouseInput();
             }
 
             if (this.field_71429_W > 0)
@@ -1835,6 +1860,7 @@
                         }
                     }
                 }
+                FMLCommonHandler.instance().fireKeyInput();
             }
 
             for (j = 0; j < 9; ++j)
@@ -2025,12 +2051,15 @@
             this.field_71453_ak.func_74428_b();
         }
 
+        FMLCommonHandler.instance().onPostClientTick();
+
         this.field_71424_I.func_76319_b();
         this.field_71423_H = func_71386_F();
     }
 
     public void func_71371_a(String p_71371_1_, String p_71371_2_, WorldSettings p_71371_3_)
     {
+        FMLClientHandler.instance().startIntegratedServer(p_71371_1_, p_71371_2_, p_71371_3_);
         this.func_71403_a((WorldClient)null);
         System.gc();
         ISaveHandler isavehandler = this.field_71469_aa.func_75804_a(p_71371_1_, false);
@@ -2066,6 +2095,12 @@
 
         while (!this.field_71437_Z.func_71200_ad())
         {
+            if (!StartupQuery.check())
+            {
+                func_71403_a(null);
+                func_147108_a(null);
+                return;
+            }
             String s2 = this.field_71437_Z.func_71195_b_();
 
             if (s2 != null)
@@ -2141,6 +2176,7 @@
             this.field_110448_aq.func_148529_f();
             this.func_71351_a((ServerData)null);
             this.field_71455_al = false;
+            FMLClientHandler.instance().handleClientWorldClosing(this.field_71441_e);
         }
 
         this.field_147127_av.func_147690_c();
