--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -279,7 +_,7 @@
 
          if (this.f_31429_ > 0) {
             --this.f_31429_;
-            if (this.f_31429_ == 0 && this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_)) {
+            if (this.f_31429_ == 0 && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.m_9236_(), this)) {
                int j1 = Mth.m_14107_(this.m_20186_());
                int i2 = Mth.m_14107_(this.m_20185_());
                int j2 = Mth.m_14107_(this.m_20189_());
@@ -293,7 +_,7 @@
                         int i1 = j2 + k2;
                         BlockPos blockpos = new BlockPos(l2, l, i1);
                         BlockState blockstate = this.m_9236_().m_8055_(blockpos);
-                        if (m_31491_(blockstate)) {
+                        if (blockstate.canEntityDestroy(this.m_9236_(), blockpos, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, blockstate)) {
                            flag = this.m_9236_().m_46953_(blockpos, true, this) || flag;
                         }
                      }
@@ -314,6 +_,10 @@
       }
    }
 
+   /**
+    * @deprecated Forge: Use {@link BlockState#canEntityDestroy(net.minecraft.world.level.BlockGetter, BlockPos, Entity)} instead.
+    */
+   @Deprecated
    public static boolean m_31491_(BlockState p_31492_) {
       return !p_31492_.m_60795_() && !p_31492_.m_204336_(BlockTags.f_13070_);
    }
