--- ../src-base/minecraft/net/minecraft/stats/RecipeBook.java
+++ ../src-work/minecraft/net/minecraft/stats/RecipeBook.java
@@ -42,9 +42,17 @@
         this.field_194078_b.clear(i);
     }
 
+    @Deprecated //DO NOT USE
     protected static int func_194075_d(@Nullable IRecipe p_194075_0_)
     {
-        return CraftingManager.field_193380_a.func_148757_b(p_194075_0_);
+        int ret = CraftingManager.field_193380_a.func_148757_b(p_194075_0_);
+        if (ret == -1)
+        {
+            ret = ((net.minecraftforge.registries.ForgeRegistry<IRecipe>)net.minecraftforge.fml.common.registry.ForgeRegistries.RECIPES).getID(p_194075_0_.getRegistryName());
+            if (ret == -1)
+                throw new IllegalArgumentException(String.format("Attempted to get the ID for a unknown recipe: %s Name: %s", p_194075_0_, p_194075_0_.getRegistryName()));
+        }
+        return ret;
     }
 
     @SideOnly(Side.CLIENT)
