--- ../src-base/minecraft/net/minecraft/crash/CallableSuspiciousClasses.java
+++ ../src-work/minecraft/net/minecraft/crash/CallableSuspiciousClasses.java
@@ -130,13 +130,6 @@
 
     public Object call()
     {
-        try
-        {
-            return this.func_85085_a();
-        }
-        catch (Exception ex)
-        {
-            return null;
-        }
+        return "FML is installed";
     }
 }
