--- ../src-base/minecraft/net/minecraft/crash/CrashReportCategory.java
+++ ../src-work/minecraft/net/minecraft/crash/CrashReportCategory.java
@@ -132,8 +132,11 @@
         }
         else
         {
-            this.stackTrace = new StackTraceElement[astacktraceelement.length - 3 - par1];
-            System.arraycopy(astacktraceelement, 3 + par1, this.stackTrace, 0, this.stackTrace.length);
+            int len = astacktraceelement.length - 3 - par1;
+            // Really Mojang, Still, god damn...
+            if (len <= 0) len = astacktraceelement.length;
+            this.stackTrace = new StackTraceElement[len];
+            System.arraycopy(astacktraceelement, astacktraceelement.length - len, this.stackTrace, 0, this.stackTrace.length);
             return this.stackTrace.length;
         }
     }
