--- ../src-base/minecraft/net/minecraft/crash/CrashReportCategory.java
+++ ../src-work/minecraft/net/minecraft/crash/CrashReportCategory.java
@@ -127,8 +127,11 @@
         }
         else
         {
-            this.field_85075_d = new StackTraceElement[astacktraceelement.length - 3 - p_85073_1_];
-            System.arraycopy(astacktraceelement, 3 + p_85073_1_, this.field_85075_d, 0, this.field_85075_d.length);
+            int len = astacktraceelement.length - 3 - p_85073_1_;
+            // Really Mojang, Still, god damn...
+            if (len <= 0) len = astacktraceelement.length;
+            this.field_85075_d = new StackTraceElement[len];
+            System.arraycopy(astacktraceelement, astacktraceelement.length - len, this.field_85075_d, 0, this.field_85075_d.length);
             return this.field_85075_d.length;
         }
     }
