--- ../src-base/minecraft/net/minecraft/command/CommandEnchant.java
+++ ../src-work/minecraft/net/minecraft/command/CommandEnchant.java
@@ -85,7 +85,7 @@
                                 {
                                     Enchantment enchantment1 = Enchantment.func_185262_c(k);
 
-                                    if (!enchantment.func_77326_a(enchantment1))
+                                    if (!enchantment.func_77326_a(enchantment1) || !enchantment1.func_77326_a(enchantment)) //Forge BugFix: Let Both enchantments veto being together
                                     {
                                         throw new CommandException("commands.enchant.cantCombine", new Object[] {enchantment.func_77316_c(i), enchantment1.func_77316_c(nbttaglist.func_150305_b(j).func_74765_d("lvl"))});
                                     }
