--- ../src-base/minecraft/net/minecraft/command/EntitySelector.java
+++ ../src-work/minecraft/net/minecraft/command/EntitySelector.java
@@ -153,6 +153,7 @@
                         list2.addAll(func_184951_f(map));
                         list2.addAll(func_180698_a(map, vec3d));
                         list2.addAll(func_179662_g(map));
+                        list2.addAll(net.minecraftforge.fml.common.registry.GameRegistry.createEntitySelectors(map, s, p_179656_0_, vec3d));
 
                         if ("s".equalsIgnoreCase(s))
                         {
