--- ../src-base/common/net/minecraft/src/ServerConfigurationManager.java
+++ ../src-work/common/net/minecraft/src/ServerConfigurationManager.java
@@ -12,6 +12,10 @@
 import java.util.List;
 import java.util.Set;
 import java.util.logging.Logger;
+
+import cpw.mods.fml.common.network.FMLNetworkHandler;
+import cpw.mods.fml.common.network.NetworkRegistry;
+import cpw.mods.fml.common.registry.GameRegistry;
 import net.minecraft.server.MinecraftServer;
 
 public abstract class ServerConfigurationManager
@@ -81,6 +85,7 @@
         }
 
         p_72355_2_.func_71116_b();
+        FMLNetworkHandler.handlePlayerLogin(p_72355_2_, var6, p_72355_1_);
     }
 
     public void func_72364_a(WorldServer[] p_72364_1_)
@@ -154,6 +159,7 @@
 
     public void func_72367_e(EntityPlayerMP p_72367_1_)
     {
+        GameRegistry.onPlayerLogout(p_72367_1_);
         this.func_72391_b(p_72367_1_);
         WorldServer var2 = p_72367_1_.func_71121_q();
         var2.func_72900_e(p_72367_1_);
@@ -304,6 +310,7 @@
         var8.func_72838_d(var7);
         this.field_72404_b.add(var7);
         var7.func_71116_b();
+        GameRegistry.onPlayerRespawn(var7);
         return var7;
     }
 
@@ -329,6 +336,7 @@
             PotionEffect var7 = (PotionEffect)var6.next();
             p_72356_1_.field_71135_a.func_72567_b(new Packet41EntityEffect(p_72356_1_.field_70157_k, var7));
         }
+        GameRegistry.onPlayerChangedDimension(p_72356_1_);
     }
 
     public void func_82448_a(Entity p_82448_1_, int p_82448_2_, WorldServer p_82448_3_, WorldServer p_82448_4_)
