--- a/net/minecraft/network/play/ServerPlayNetHandler.java
+++ b/net/minecraft/network/play/ServerPlayNetHandler.java
@@ -180,6 +180,7 @@
 
    public void func_73660_a() {
       this.func_184342_d();
+      if (this.field_147369_b.func_70089_S()) //Don't tick players while they are dead/respawning.
       this.field_147369_b.func_71127_g();
       this.field_147369_b.func_70080_a(this.field_184349_l, this.field_184350_m, this.field_184351_n, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A);
       ++this.field_147368_e;
@@ -337,9 +338,11 @@
             }
 
             entity.func_70080_a(d3, d4, d5, f, f1);
+            this.field_147369_b.func_70080_a(d3, d4, d5, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A); // Forge - Resync player position on vehicle moving
             boolean flag2 = serverworld.func_195586_b(entity, entity.func_174813_aQ().func_186664_h(0.0625D));
             if (flag && (flag1 || !flag2)) {
                entity.func_70080_a(d0, d1, d2, f, f1);
+               this.field_147369_b.func_70080_a(d3, d4, d5, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A); // Forge - Resync player position on vehicle moving
                this.field_147371_a.func_179290_a(new SMoveVehiclePacket(entity));
                return;
             }
@@ -839,7 +842,9 @@
       Direction direction = blockraytraceresult.func_216354_b();
       this.field_147369_b.func_143004_u();
       if (blockpos.func_177956_o() < this.field_147367_d.func_71207_Z() - 1 || direction != Direction.UP && blockpos.func_177956_o() < this.field_147367_d.func_71207_Z()) {
-         if (this.field_184362_y == null && this.field_147369_b.func_70092_e((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 0.5D, (double)blockpos.func_177952_p() + 0.5D) < 64.0D && serverworld.func_175660_a(this.field_147369_b, blockpos)) {
+         double dist = field_147369_b.func_110148_a(net.minecraft.entity.player.PlayerEntity.REACH_DISTANCE).func_111126_e() + 3;
+         dist *= dist;
+         if (this.field_184362_y == null && this.field_147369_b.func_70092_e((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 0.5D, (double)blockpos.func_177952_p() + 0.5D) < dist && serverworld.func_175660_a(this.field_147369_b, blockpos)) {
             this.field_147369_b.field_71134_c.func_219441_a(this.field_147369_b, serverworld, itemstack, hand, blockraytraceresult);
          }
       } else {
@@ -959,7 +964,9 @@
          if (s.startsWith("/")) {
             this.func_147361_d(s);
          } else {
-            ITextComponent itextcomponent = new TranslationTextComponent("chat.type.text", this.field_147369_b.func_145748_c_(), s);
+            ITextComponent itextcomponent = new TranslationTextComponent("chat.type.text", this.field_147369_b.func_145748_c_(), net.minecraftforge.common.ForgeHooks.newChatWithLinks(s));
+            itextcomponent = net.minecraftforge.common.ForgeHooks.onServerChatEvent(this, s, itextcomponent);
+            if (itextcomponent == null) return;
             this.field_147367_d.func_184103_al().func_148544_a(itextcomponent, false);
          }
 
@@ -1057,6 +1064,7 @@
                this.field_147369_b.func_190775_a(entity, hand);
             } else if (p_147340_1_.func_149565_c() == CUseEntityPacket.Action.INTERACT_AT) {
                Hand hand1 = p_147340_1_.func_186994_b();
+               if (net.minecraftforge.common.ForgeHooks.onInteractEntityAt(field_147369_b, entity, p_147340_1_.func_179712_b(), hand1) != null) return;
                entity.func_184199_a(this.field_147369_b, p_147340_1_.func_179712_b(), hand1);
             } else if (p_147340_1_.func_149565_c() == CUseEntityPacket.Action.ATTACK) {
                if (entity instanceof ItemEntity || entity instanceof ExperienceOrbEntity || entity instanceof AbstractArrowEntity || entity == this.field_147369_b) {
@@ -1087,7 +1095,7 @@
                return;
             }
 
-            this.field_147369_b = this.field_147367_d.func_184103_al().func_72368_a(this.field_147369_b, DimensionType.field_223227_a_, false);
+            this.field_147369_b = this.field_147367_d.func_184103_al().func_72368_a(this.field_147369_b, this.field_147369_b.field_71093_bK, false);
             if (this.field_147367_d.func_71199_h()) {
                this.field_147369_b.func_71033_a(GameType.SPECTATOR);
                this.field_147369_b.func_71121_q().func_82736_K().func_223585_a(GameRules.field_223613_p).func_223570_a(false, this.field_147367_d);
@@ -1264,6 +1272,8 @@
    }
 
    public void func_147349_a(CCustomPayloadPacket p_147349_1_) {
+      PacketThreadUtil.func_218796_a(p_147349_1_, this, this.field_147369_b.func_71121_q());
+      net.minecraftforge.fml.network.NetworkHooks.onCustomPayload(p_147349_1_, this.field_147371_a);
    }
 
    public void func_217263_a(CSetDifficultyPacket p_217263_1_) {
