--- ../src-base/minecraft/net/minecraft/client/gui/recipebook/GuiRecipeOverlay.java
+++ ../src-work/minecraft/net/minecraft/client/gui/recipebook/GuiRecipeOverlay.java
@@ -230,11 +230,11 @@
             int k = 3;
             int l = 3;
 
-            if (this.field_193924_p instanceof ShapedRecipes)
+            if (this.field_193924_p instanceof net.minecraftforge.common.crafting.IShapedRecipe)
             {
-                ShapedRecipes shapedrecipes = (ShapedRecipes)this.field_193924_p;
-                k = shapedrecipes.func_192403_f();
-                l = shapedrecipes.func_192404_g();
+                net.minecraftforge.common.crafting.IShapedRecipe shapedrecipes = (net.minecraftforge.common.crafting.IShapedRecipe)this.field_193924_p;
+                k = shapedrecipes.getRecipeWidth();
+                l = shapedrecipes.getRecipeHeight();
             }
 
             Iterator<Ingredient> iterator = this.field_193924_p.func_192400_c().iterator();
