--- a/net/minecraft/client/gui/recipebook/RecipeBookGui.java
+++ b/net/minecraft/client/gui/recipebook/RecipeBookGui.java
@@ -95,7 +_,7 @@
       this.initFilterButtonTextures();
       this.tabButtons.clear();
 
-      for(RecipeBookCategories recipebookcategories : RecipeBookCategories.getCategories(this.menu.getRecipeBookType())) {
+      for(RecipeBookCategories recipebookcategories : this.menu.getRecipeBookCategories()) {
          this.tabButtons.add(new RecipeTabToggleWidget(recipebookcategories));
       }
 
@@ -434,7 +_,7 @@
 
          languagemanager.setSelected(language);
          this.minecraft.options.languageCode = language.getCode();
-         this.minecraft.reloadResourcePacks();
+         net.minecraftforge.client.ForgeHooksClient.refreshResources(this.minecraft, net.minecraftforge.resource.VanillaResourceType.LANGUAGES);
          this.minecraft.options.save();
       }
 
