--- a/net/minecraft/client/gui/recipebook/RecipeBookGui.java
+++ b/net/minecraft/client/gui/recipebook/RecipeBookGui.java
@@ -95,7 +95,7 @@
       this.func_205702_a();
       this.field_193018_j.clear();
 
-      for(RecipeBookCategories recipebookcategories : RecipeBookCategories.func_243236_a(this.field_201522_g.func_241850_m())) {
+      for(RecipeBookCategories recipebookcategories : this.field_201522_g.getRecipeBookCategories()) {
          this.field_193018_j.add(new RecipeTabToggleWidget(recipebookcategories));
       }
 
@@ -433,7 +433,7 @@
 
          languagemanager.func_135045_a(language);
          this.field_191888_F.field_71474_y.field_74363_ab = language.getCode();
-         this.field_191888_F.func_213237_g();
+         net.minecraftforge.client.ForgeHooksClient.refreshResources(this.field_191888_F, net.minecraftforge.resource.VanillaResourceType.LANGUAGES);
          this.field_191888_F.field_71474_y.func_74303_b();
       }
 
