--- a/net/minecraft/world/level/block/BambooBlock.java
+++ b/net/minecraft/world/level/block/BambooBlock.java
@@ -26,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class BambooBlock extends Block implements BonemealableBlock {
+public class BambooBlock extends Block implements BonemealableBlock, net.minecraftforge.common.IPlantable  {
    protected static final float f_152092_ = 3.0F;
    protected static final float f_152093_ = 5.0F;
    protected static final float f_152094_ = 1.5F;
@@ -110,10 +_,11 @@
 
    public void m_213898_(BlockState p_220738_, ServerLevel p_220739_, BlockPos p_220740_, RandomSource p_220741_) {
       if (p_220738_.m_61143_(f_48871_) == 0) {
-         if (p_220741_.m_188503_(3) == 0 && p_220739_.m_46859_(p_220740_.m_7494_()) && p_220739_.m_45524_(p_220740_.m_7494_(), 0) >= 9) {
+         if (p_220739_.m_46859_(p_220740_.m_7494_()) && p_220739_.m_45524_(p_220740_.m_7494_(), 0) >= 9) {
             int i = this.m_48932_(p_220739_, p_220740_) + 1;
-            if (i < 16) {
+            if (i < 16 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_220739_, p_220740_, p_220738_, p_220741_.m_188503_(3) == 0)) {
                this.m_220731_(p_220738_, p_220739_, p_220740_, p_220741_, i);
+               net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_220739_, p_220740_, p_220738_);
             }
          }
 
@@ -167,7 +_,7 @@
    }
 
    public float m_5880_(BlockState p_48901_, Player p_48902_, BlockGetter p_48903_, BlockPos p_48904_) {
-      return p_48902_.m_21205_().m_41720_() instanceof SwordItem ? 1.0F : super.m_5880_(p_48901_, p_48902_, p_48903_, p_48904_);
+      return p_48902_.m_21205_().canPerformAction(net.minecraftforge.common.ToolActions.SWORD_DIG) ? 1.0F : super.m_5880_(p_48901_, p_48902_, p_48903_, p_48904_);
    }
 
    protected void m_220731_(BlockState p_220732_, Level p_220733_, BlockPos p_220734_, RandomSource p_220735_, int p_220736_) {
@@ -208,5 +_,12 @@
       }
 
       return i;
+   }
+
+   @Override
+   public BlockState getPlant(BlockGetter world, BlockPos pos) {
+      BlockState state = world.m_8055_(pos);
+      if (state.m_60734_() != this) return m_49966_();
+      return state;
    }
 }
