--- a/net/minecraft/world/level/block/BambooSaplingBlock.java
+++ b/net/minecraft/world/level/block/BambooSaplingBlock.java
@@ -20,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class BambooSaplingBlock extends Block implements BonemealableBlock {
+public class BambooSaplingBlock extends Block implements BonemealableBlock, net.minecraftforge.common.IPlantable {
    protected static final float f_152100_ = 4.0F;
    protected static final VoxelShape f_48954_ = Block.m_49796_(4.0D, 0.0D, 4.0D, 12.0D, 12.0D, 12.0D);
 
@@ -41,6 +_,7 @@
    }
 
    public boolean m_7898_(BlockState p_48986_, LevelReader p_48987_, BlockPos p_48988_) {
+      if (p_48987_.m_8055_(p_48988_.m_7495_()).canSustainPlant(p_48987_, p_48988_.m_7495_(), Direction.UP, this)) {return true;}
       return p_48987_.m_8055_(p_48988_.m_7495_()).m_204336_(BlockTags.f_13065_);
    }
 
@@ -73,10 +_,15 @@
    }
 
    public float m_5880_(BlockState p_48981_, Player p_48982_, BlockGetter p_48983_, BlockPos p_48984_) {
-      return p_48982_.m_21205_().m_41720_() instanceof SwordItem ? 1.0F : super.m_5880_(p_48981_, p_48982_, p_48983_, p_48984_);
+      return p_48982_.m_21205_().canPerformAction(net.minecraftforge.common.ToolActions.SWORD_DIG) ? 1.0F : super.m_5880_(p_48981_, p_48982_, p_48983_, p_48984_);
    }
 
    protected void m_48972_(Level p_48973_, BlockPos p_48974_) {
       p_48973_.m_7731_(p_48974_.m_7494_(), Blocks.f_50571_.m_49966_().m_61124_(BambooStalkBlock.f_260716_, BambooLeaves.SMALL), 3);
+   }
+
+   @Override
+   public BlockState getPlant(BlockGetter level, BlockPos pos) {
+      return Blocks.f_50571_.m_49966_();
    }
 }
