--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -161,6 +_,7 @@
         if (!oldState.is(state.getBlock())) {
             if (inPortalDimension(level)) {
                 Optional<PortalShape> optionalShape = PortalShape.findEmptyPortalShape(level, pos, Direction.Axis.X);
+                optionalShape = net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(level, pos, optionalShape);
                 if (optionalShape.isPresent()) {
                     optionalShape.get().createPortalBlocks(level);
                     return;
@@ -204,7 +_,7 @@
         boolean hasObsidian = false;
 
         for (Direction face : Direction.values()) {
-            if (level.getBlockState(testPos.set(pos).move(face)).is(Blocks.OBSIDIAN)) {
+            if (level.getBlockState(testPos.set(pos).move(face)).isPortalFrame(level, testPos)) {
                 hasObsidian = true;
                 break;
             }
