--- a/net/minecraft/world/level/block/BaseRailBlock.java
+++ b/net/minecraft/world/level/block/BaseRailBlock.java
@@ -24,7 +_,7 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
 
-public abstract class BaseRailBlock extends Block implements SimpleWaterloggedBlock {
+public abstract class BaseRailBlock extends Block implements SimpleWaterloggedBlock, net.minecraftforge.common.extensions.IForgeBaseRailBlock {
     public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;
     private static final VoxelShape SHAPE_FLAT = Block.column(16.0, 0.0, 2.0);
     private static final VoxelShape SHAPE_SLOPE = Block.column(16.0, 0.0, 8.0);
@@ -81,7 +_,7 @@
         final BlockState state, final Level level, final BlockPos pos, final Block block, final @Nullable Orientation orientation, final boolean movedByPiston
     ) {
         if (!level.isClientSide() && level.getBlockState(pos).is(this)) {
-            RailShape shape = state.getValue(this.getShapeProperty());
+            RailShape shape = getRailDirection(state, level, pos, null);
             if (shouldBeRemoved(pos, level, shape)) {
                 dropResources(state, level, pos);
                 level.removeBlock(pos, movedByPiston);
@@ -120,7 +_,7 @@
     @Override
     protected void affectNeighborsAfterRemoval(final BlockState state, final ServerLevel level, final BlockPos pos, final boolean movedByPiston) {
         if (!movedByPiston) {
-            if (state.getValue(this.getShapeProperty()).isSlope()) {
+            if (getRailDirection(state, level, pos, null).isSlope()) {
                 level.updateNeighborsAt(pos.above(), this);
             }
 
@@ -141,6 +_,11 @@
         return state.setValue(this.getShapeProperty(), isEastWest ? RailShape.EAST_WEST : RailShape.NORTH_SOUTH).setValue(WATERLOGGED, isWaterSource);
     }
 
+    /**
+     * @deprecated Forge: Use {@link BaseRailBlock#getRailDirection(BlockState, BlockGetter, BlockPos, net.minecraft.world.entity.vehicle.AbstractMinecart)} for enhanced ability.
+     * If you do change this property, be aware that other functions in this/subclasses may break as they can make assumptions about this property
+     */
+    @Deprecated
     public abstract Property<RailShape> getShapeProperty();
 
     protected RailShape rotate(final RailShape shape, final Rotation rotation) {
@@ -296,5 +_,15 @@
     @Override
     protected FluidState getFluidState(final BlockState state) {
         return state.getValue(WATERLOGGED) ? Fluids.WATER.getSource(false) : super.getFluidState(state);
+    }
+
+    @Override
+    public boolean isFlexibleRail(final BlockState state, final BlockGetter level, final BlockPos pos) {
+        return !this.isStraight;
+    }
+
+    @Override
+    public RailShape getRailDirection(final BlockState state, final BlockGetter level, final BlockPos pos, final net.minecraft.world.entity.vehicle.minecart.@Nullable AbstractMinecart cart) {
+        return state.getValue(getShapeProperty());
     }
 }
