--- a/net/minecraft/world/level/block/DoublePlantBlock.java
+++ b/net/minecraft/world/level/block/DoublePlantBlock.java
@@ -54,6 +_,7 @@
          return super.canSurvive(p_52887_, p_52888_, p_52889_);
       } else {
          BlockState blockstate = p_52888_.getBlockState(p_52889_.below());
+         if (p_52887_.getBlock() != this) return super.canSurvive(p_52887_, p_52888_, p_52889_); //Forge: This function is called during world gen and placement, before this block is set, so if we are not 'here' then assume it's the pre-check.
          return blockstate.is(this) && blockstate.getValue(HALF) == DoubleBlockHalf.LOWER;
       }
    }
