--- a/net/minecraft/world/level/block/DropperBlock.java
+++ b/net/minecraft/world/level/block/DropperBlock.java
@@ -38,7 +_,7 @@
          p_52944_.m_46796_(1001, p_52945_, 0);
       } else {
          ItemStack itemstack = dispenserblockentity.m_8020_(i);
-         if (!itemstack.m_41619_()) {
+         if (!itemstack.m_41619_() && net.minecraftforge.items.VanillaInventoryCodeHooks.dropperInsertHook(p_52944_, p_52945_, dispenserblockentity, i, itemstack)) {
             Direction direction = p_52944_.m_8055_(p_52945_).m_61143_(f_52659_);
             Container container = HopperBlockEntity.m_59390_(p_52944_, p_52945_.m_121945_(direction));
             ItemStack itemstack1;
