--- a/net/minecraft/world/level/block/DropperBlock.java
+++ b/net/minecraft/world/level/block/DropperBlock.java
@@ -45,7 +_,7 @@
             p_52944_.levelEvent(1001, p_52945_, 0);
          } else {
             ItemStack itemstack = dispenserblockentity.getItem(i);
-            if (!itemstack.isEmpty()) {
+         if (!itemstack.isEmpty() && net.minecraftforge.items.VanillaInventoryCodeHooks.dropperInsertHook(p_52944_, p_52945_, dispenserblockentity, i, itemstack)) {
                Direction direction = p_52944_.getBlockState(p_52945_).getValue(FACING);
                Container container = HopperBlockEntity.getContainerAt(p_52944_, p_52945_.relative(direction));
                ItemStack itemstack1;
