--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -23,7 +_,7 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class LeavesBlock extends Block implements SimpleWaterloggedBlock {
+public class LeavesBlock extends Block implements SimpleWaterloggedBlock, net.minecraftforge.common.IForgeShearable {
    public static final int f_153563_ = 7;
    public static final IntegerProperty f_54418_ = BlockStateProperties.f_61414_;
    public static final BooleanProperty f_54419_ = BlockStateProperties.f_61447_;
