--- a/net/minecraft/world/level/block/RailState.java
+++ b/net/minecraft/world/level/block/RailState.java
@@ -16,14 +_,16 @@
     private BlockState state;
     private final boolean isStraight;
     private final List<BlockPos> connections = Lists.newArrayList();
+    private final boolean canMakeSlopes;
 
     public RailState(final Level level, final BlockPos pos, final BlockState state) {
         this.level = level;
         this.pos = pos;
         this.state = state;
         this.block = (BaseRailBlock)state.getBlock();
-        RailShape direction = state.getValue(this.block.getShapeProperty());
-        this.isStraight = this.block.isStraight();
+        RailShape direction = this.block.getRailDirection(state, level, pos, null);
+        this.isStraight = !this.block.isFlexibleRail(state, level, pos);
+        this.canMakeSlopes = this.block.canMakeSlopes(state, level, pos);
         this.updateConnections(direction);
     }
 
@@ -177,7 +_,7 @@
             }
         }
 
-        if (shape == RailShape.NORTH_SOUTH) {
+        if (shape == RailShape.NORTH_SOUTH && canMakeSlopes) {
             if (BaseRailBlock.isRail(this.level, north.above())) {
                 shape = RailShape.ASCENDING_NORTH;
             }
@@ -187,7 +_,7 @@
             }
         }
 
-        if (shape == RailShape.EAST_WEST) {
+        if (shape == RailShape.EAST_WEST && canMakeSlopes) {
             if (BaseRailBlock.isRail(this.level, east.above())) {
                 shape = RailShape.ASCENDING_EAST;
             }
@@ -199,6 +_,11 @@
 
         if (shape == null) {
             shape = RailShape.NORTH_SOUTH;
+        }
+
+        if (!this.block.isValidRailShape(shape)) { // Forge: allow rail block to decide if the new shape is valid
+            this.connections.remove(rail.pos);
+            return;
         }
 
         this.state = this.state.setValue(this.block.getShapeProperty(), shape);
@@ -303,7 +_,7 @@
             }
         }
 
-        if (shape == RailShape.NORTH_SOUTH) {
+        if (shape == RailShape.NORTH_SOUTH && canMakeSlopes) {
             if (BaseRailBlock.isRail(this.level, north.above())) {
                 shape = RailShape.ASCENDING_NORTH;
             }
@@ -313,7 +_,7 @@
             }
         }
 
-        if (shape == RailShape.EAST_WEST) {
+        if (shape == RailShape.EAST_WEST && canMakeSlopes) {
             if (BaseRailBlock.isRail(this.level, east.above())) {
                 shape = RailShape.ASCENDING_EAST;
             }
@@ -323,7 +_,7 @@
             }
         }
 
-        if (shape == null) {
+        if (shape == null || !this.block.isValidRailShape(shape)) { // Forge: allow rail block to decide if the new shape is valid
             shape = defaultShape;
         }
 
