--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -363,7 +_,7 @@
       } else if (p_55595_.m_60713_(Blocks.f_50455_)) {
          return p_55596_ == p_55595_.m_61143_(ObserverBlock.f_52588_);
       } else {
-         return p_55595_.m_60803_() && p_55596_ != null;
+         return p_55595_.canConnectRedstone(p_55596_);
       }
    }
 
