--- a/net/minecraft/world/level/block/SculkShriekerBlock.java
+++ b/net/minecraft/world/level/block/SculkShriekerBlock.java
@@ -127,7 +_,7 @@
     @Override
     protected void spawnAfterBreak(final BlockState state, final ServerLevel level, final BlockPos pos, final ItemStack tool, final boolean dropExperience) {
         super.spawnAfterBreak(state, level, pos, tool, dropExperience);
-        if (dropExperience) {
+        if (false && dropExperience) { // Forge: Moved to getExpDrop
             this.tryDropExperience(level, pos, tool, ConstantInt.of(5));
         }
     }
@@ -141,5 +_,10 @@
                 (innerLevel, pos, state, entity) -> VibrationSystem.Ticker.tick(innerLevel, entity.getVibrationData(), entity.getVibrationUser())
             )
             : null;
+    }
+
+    @Override
+    public int getExpDrop(BlockState state, net.minecraft.world.level.LevelReader level, RandomSource randomSource, BlockPos pos, int fortuneLevel, int silkTouchLevel) {
+       return silkTouchLevel == 0 ? 5 : 0;
     }
 }
