--- a/net/minecraft/world/level/block/SeagrassBlock.java
+++ b/net/minecraft/world/level/block/SeagrassBlock.java
@@ -20,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class SeagrassBlock extends BushBlock implements BonemealableBlock, LiquidBlockContainer {
+public class SeagrassBlock extends BushBlock implements BonemealableBlock, LiquidBlockContainer, net.minecraftforge.common.IForgeShearable {
    protected static final float f_154492_ = 6.0F;
    protected static final VoxelShape f_154493_ = Block.m_49796_(2.0D, 0.0D, 2.0D, 14.0D, 12.0D, 14.0D);
 
