--- a/net/minecraft/world/level/block/SpawnerBlock.java
+++ b/net/minecraft/world/level/block/SpawnerBlock.java
@@ -39,11 +_,12 @@
 
    public void spawnAfterBreak(BlockState p_222477_, ServerLevel p_222478_, BlockPos p_222479_, ItemStack p_222480_, boolean p_222481_) {
       super.spawnAfterBreak(p_222477_, p_222478_, p_222479_, p_222480_, p_222481_);
-      if (p_222481_) {
-         int i = 15 + p_222478_.random.nextInt(15) + p_222478_.random.nextInt(15);
-         this.popExperience(p_222478_, p_222479_, i);
-      }
-
+
+   }
+
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.level.LevelReader world, net.minecraft.util.RandomSource randomSource, BlockPos pos, int fortune, int silktouch) {
+      return 15 + randomSource.nextInt(15) + randomSource.nextInt(15);
    }
 
    public RenderShape getRenderShape(BlockState p_56794_) {
