--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -38,6 +_,7 @@
    }
 
    private boolean m_56807_(Level p_56808_, BlockPos p_56809_) {
+      BlockState spongeState = p_56808_.m_8055_(p_56809_);
       return BlockPos.m_276833_(p_56809_, 6, 65, (p_277519_, p_277492_) -> {
          for(Direction direction : f_276425_) {
             p_277492_.accept(p_277519_.m_121945_(direction));
@@ -49,7 +_,7 @@
          } else {
             BlockState blockstate = p_56808_.m_8055_(p_279054_);
             FluidState fluidstate = p_56808_.m_6425_(p_279054_);
-            if (!fluidstate.m_205070_(FluidTags.f_13131_)) {
+            if (!spongeState.canBeHydrated(p_56808_, p_56809_, fluidstate, p_279054_)) {
                return false;
             } else {
                Block block = blockstate.m_60734_();
