--- a/net/minecraft/world/level/block/StemBlock.java
+++ b/net/minecraft/world/level/block/StemBlock.java
@@ -44,22 +_,23 @@
    }
 
    public void randomTick(BlockState p_222538_, ServerLevel p_222539_, BlockPos p_222540_, RandomSource p_222541_) {
+      if (!p_222539_.isAreaLoaded(p_222540_, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
       if (p_222539_.getRawBrightness(p_222540_, 0) >= 9) {
          float f = CropBlock.getGrowthSpeed(this, p_222539_, p_222540_);
-         if (p_222541_.nextInt((int)(25.0F / f) + 1) == 0) {
+         if (net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_222539_, p_222540_, p_222538_, p_222541_.nextInt((int)(25.0F / f) + 1) == 0)) {
             int i = p_222538_.getValue(AGE);
             if (i < 7) {
-               p_222538_ = p_222538_.setValue(AGE, Integer.valueOf(i + 1));
-               p_222539_.setBlock(p_222540_, p_222538_, 2);
+               p_222539_.setBlock(p_222540_, p_222538_.setValue(AGE, Integer.valueOf(i + 1)), 2);
             } else {
                Direction direction = Direction.Plane.HORIZONTAL.getRandomDirection(p_222541_);
                BlockPos blockpos = p_222540_.relative(direction);
                BlockState blockstate = p_222539_.getBlockState(blockpos.below());
-               if (p_222539_.getBlockState(blockpos).isAir() && (blockstate.is(Blocks.FARMLAND) || blockstate.is(BlockTags.DIRT))) {
+               if (p_222539_.isEmptyBlock(blockpos) && (blockstate.canSustainPlant(p_222539_, blockpos.below(), Direction.UP, this.fruit) || blockstate.is(Blocks.FARMLAND) || blockstate.is(BlockTags.DIRT))) {
                   p_222539_.setBlockAndUpdate(blockpos, this.fruit.defaultBlockState());
                   p_222539_.setBlockAndUpdate(p_222540_, this.fruit.getAttachedStem().defaultBlockState().setValue(HorizontalDirectionalBlock.FACING, direction));
                }
             }
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_222539_, p_222540_, p_222538_);
          }
 
       }
@@ -93,5 +_,11 @@
 
    public StemGrownBlock getFruit() {
       return this.fruit;
+   }
+
+   //FORGE START
+   @Override
+   public net.minecraftforge.common.PlantType getPlantType(BlockGetter world, BlockPos pos) {
+      return net.minecraftforge.common.PlantType.CROP;
    }
 }
