--- a/net/minecraft/world/level/block/SweetBerryBushBlock.java
+++ b/net/minecraft/world/level/block/SweetBerryBushBlock.java
@@ -57,10 +_,11 @@
 
    public void m_213898_(BlockState p_222563_, ServerLevel p_222564_, BlockPos p_222565_, RandomSource p_222566_) {
       int i = p_222563_.m_61143_(f_57244_);
-      if (i < 3 && p_222566_.m_188503_(5) == 0 && p_222564_.m_45524_(p_222565_.m_7494_(), 0) >= 9) {
+      if (i < 3 && p_222564_.m_45524_(p_222565_.m_7494_(), 0) >= 9 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_222564_, p_222565_, p_222563_, p_222566_.m_188503_(5) == 0)) {
          BlockState blockstate = p_222563_.m_61124_(f_57244_, Integer.valueOf(i + 1));
          p_222564_.m_7731_(p_222565_, blockstate, 2);
          p_222564_.m_220407_(GameEvent.f_157792_, p_222565_, GameEvent.Context.m_223722_(blockstate));
+         net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_222564_, p_222565_, p_222563_);
       }
 
    }
