--- a/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -205,4 +_,14 @@
     protected BlockSetType getType() {
         return this.type;
     }
+
+    @Override
+    public boolean isLadder(BlockState state, net.minecraft.world.level.LevelReader world, BlockPos pos, net.minecraft.world.entity.LivingEntity entity) {
+        if (state.getValue(OPEN)) {
+            BlockPos downPos = pos.below();
+            BlockState down = world.getBlockState(downPos);
+            return down.getBlock().makesOpenTrapdoorAboveClimbable(down, world, downPos, state);
+        }
+        return false;
+    }
 }
