--- a/net/minecraft/world/level/block/WaterlilyBlock.java
+++ b/net/minecraft/world/level/block/WaterlilyBlock.java
@@ -33,9 +_,8 @@
       return f_58159_;
    }
 
-   protected boolean m_6266_(BlockState p_58174_, BlockGetter p_58175_, BlockPos p_58176_) {
-      FluidState fluidstate = p_58175_.m_6425_(p_58176_);
-      FluidState fluidstate1 = p_58175_.m_6425_(p_58176_.m_7494_());
-      return (fluidstate.m_76152_() == Fluids.f_76193_ || p_58174_.m_60767_() == Material.f_76276_) && fluidstate1.m_76152_() == Fluids.f_76191_;
+   @Override
+   public net.minecraftforge.common.PlantType getPlantType(BlockGetter world, BlockPos pos) {
+      return net.minecraftforge.common.PlantType.WATERLILY;
    }
 }
