--- a/net/minecraft/world/level/block/WebBlock.java
+++ b/net/minecraft/world/level/block/WebBlock.java
@@ -7,7 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
 
-public class WebBlock extends Block {
+public class WebBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public WebBlock(BlockBehaviour.Properties p_58178_) {
       super(p_58178_);
    }
