--- ../src-base/minecraft_server/net/minecraft/server/MinecraftServer.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src-work/minecraft_server/net/minecraft/server/MinecraftServer.java	0000-00-00 00:00:00.000000000 -0000
@@ -7,12 +7,17 @@
 import java.net.UnknownHostException;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.EnumSet;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Random;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+
+import cpw.mods.fml.common.FMLCommonHandler;
+import cpw.mods.fml.common.TickType;
+import cpw.mods.fml.server.FMLServerHandler;
 import net.minecraft.src.AnvilSaveConverter;
 import net.minecraft.src.AnvilSaveHandler;
 import net.minecraft.src.AxisAlignedBB;
@@ -100,6 +105,7 @@
         var1.start();
         ConsoleLogManager.func_641_a();
         field_6038_a.info("Starting minecraft server version 1.2.5");
+        FMLServerHandler.instance().onPreLoad(this);
 
         if (Runtime.getRuntime().maxMemory() / 1024L / 1024L < 512L)
         {
@@ -146,6 +152,7 @@
             field_6038_a.warning("To change this, set \"online-mode\" to \"true\" in the server.settings file.");
         }
 
+        FMLServerHandler.instance().onLoadComplete();
         this.field_6033_f = new ServerConfigurationManager(this);
         this.field_6028_k[0] = new EntityTracker(this, 0);
         this.field_6028_k[1] = new EntityTracker(this, -1);
@@ -353,6 +360,8 @@
             {
                 long var1 = System.currentTimeMillis();
 
+                FMLServerHandler.instance().onWorldLoadTick();
+
                 for (long var3 = 0L; this.field_6025_n; Thread.sleep(1L))
                 {
                     long var5 = System.currentTimeMillis();
@@ -444,9 +453,11 @@
 
     private void func_6018_h()
     {
+        FMLCommonHandler.instance().rescheduleTicks();
         long var1 = System.nanoTime();
         ArrayList var3 = new ArrayList();
         Iterator var4 = field_6037_b.keySet().iterator();
+        FMLServerHandler.instance().onPreServerTick();
 
         while (var4.hasNext())
         {
@@ -487,7 +498,9 @@
                     this.field_6033_f.func_28169_a(new Packet4UpdateTime(var7.func_22080_k()), var7.field_4272_q.field_6165_g);
                 }
 
+                FMLServerHandler.instance().onPreWorldTick(var7);
                 var7.func_22077_g();
+                FMLServerHandler.instance().onPostWorldTick(var7);
 
                 while (true)
                 {
@@ -533,6 +546,7 @@
         this.field_48076_G = Packet.field_48101_l;
         this.field_48082_x[this.field_9014_h % 100] = Packet.field_48102_m - this.field_48077_H;
         this.field_48077_H = Packet.field_48102_m;
+        FMLServerHandler.instance().onPostServerTick();
     }
 
     public void func_6010_a(String p_6010_1_, ICommandListener p_6010_2_)
@@ -715,7 +729,7 @@
 
     public String func_52003_getServerModName()
     {
-        return "vanilla";
+        return "fml";
     }
 
     public static boolean func_6015_a(MinecraftServer p_6015_0_)
