--- a/net/minecraft/client/renderer/entity/EntityRendererManager.java
+++ b/net/minecraft/client/renderer/entity/EntityRendererManager.java
@@ -183,13 +183,15 @@
       this.field_178637_m = new PlayerRenderer(this);
       this.field_178636_l.put("default", this.field_178637_m);
       this.field_178636_l.put("slim", new PlayerRenderer(this, true));
+   }
 
+   //FORGE: Validate the existence of a renderer for each entity type after modded entity renderers have been added
+   public void validateRendererExistence() {
       for(EntityType<?> entitytype : Registry.field_212629_r) {
          if (entitytype != EntityType.field_200729_aH && !this.field_78729_o.containsKey(entitytype)) {
             throw new IllegalStateException("No renderer registered for " + Registry.field_212629_r.func_177774_c(entitytype));
          }
       }
-
    }
 
    public <T extends Entity> EntityRenderer<? super T> func_78713_a(T p_78713_1_) {
@@ -275,6 +277,7 @@
    private void func_229093_a_(MatrixStack p_229093_1_, IVertexBuilder p_229093_2_, Entity p_229093_3_, float p_229093_4_) {
       float f = p_229093_3_.func_213311_cf() / 2.0F;
       this.func_229094_a_(p_229093_1_, p_229093_2_, p_229093_3_, 1.0F, 1.0F, 1.0F);
+      if(p_229093_3_.getParts().length > 0){ net.minecraftforge.entity.MultiPartEntityUtil.renderDebugBoundingBoxes(p_229093_1_,p_229093_2_,p_229093_3_,p_229093_4_); } else
       if (p_229093_3_ instanceof EnderDragonEntity) {
          double d0 = p_229093_3_.func_226277_ct_() - MathHelper.func_219803_d((double)p_229093_4_, p_229093_3_.field_70142_S, p_229093_3_.func_226277_ct_());
          double d1 = p_229093_3_.func_226278_cu_() - MathHelper.func_219803_d((double)p_229093_4_, p_229093_3_.field_70137_T, p_229093_3_.func_226278_cu_());
@@ -445,4 +448,8 @@
    public FontRenderer func_78716_a() {
       return this.field_78736_p;
    }
+
+   public Map<String, PlayerRenderer> getSkinMap() {
+      return (Map<String, PlayerRenderer>) java.util.Collections.unmodifiableMap(field_178636_l);
+   }
 }
