--- a/net/minecraft/client/renderer/entity/EntityRendererManager.java
+++ b/net/minecraft/client/renderer/entity/EntityRendererManager.java
@@ -445,4 +445,8 @@
    public FontRenderer func_78716_a() {
       return this.field_78736_p;
    }
+
+   public Map<String, PlayerRenderer> getSkinMap() {
+      return (Map<String, PlayerRenderer>) java.util.Collections.unmodifiableMap(field_178636_l);
+   }
 }
