--- a/net/minecraft/client/renderer/entity/EntityRendererManager.java
+++ b/net/minecraft/client/renderer/entity/EntityRendererManager.java
@@ -183,9 +183,8 @@
       this.field_178637_m = new PlayerRenderer(this);
       this.field_178636_l.put("default", this.field_178637_m);
       this.field_178636_l.put("slim", new PlayerRenderer(this, true));
-
       for(EntityType<?> entitytype : Registry.field_212629_r) {
-         if (entitytype != EntityType.field_200729_aH && !this.field_78729_o.containsKey(entitytype)) {
+         if (entitytype.getRegistryName().func_110624_b().equals("minecraft") && entitytype != EntityType.field_200729_aH && !this.field_78729_o.containsKey(entitytype)) { //FORGE: Skip check for modded entities as they have not had the chance to add their renderers
             throw new IllegalStateException("No renderer registered for " + Registry.field_212629_r.func_177774_c(entitytype));
          }
       }
@@ -445,4 +444,8 @@
    public FontRenderer func_78716_a() {
       return this.field_78736_p;
    }
+
+   public Map<String, PlayerRenderer> getSkinMap() {
+      return (Map<String, PlayerRenderer>) java.util.Collections.unmodifiableMap(field_178636_l);
+   }
 }
