--- a/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
@@ -61,7 +_,7 @@
       if (!flag) {
          BlockRenderDispatcher blockrenderdispatcher = this.f_115046_.m_91289_();
          ModelManager modelmanager = blockrenderdispatcher.m_110907_().m_110881_();
-         ModelResourceLocation modelresourcelocation = this.m_174212_(p_115076_, itemstack);
+         ModelResourceLocation modelresourcelocation = p_115076_.m_31822_().m_41720_() instanceof MapItem ? f_115045_ : f_115044_;
          p_115079_.m_85836_();
          p_115079_.m_85837_(-0.5D, -0.5D, -0.5D);
          blockrenderdispatcher.m_110937_().m_111067_(p_115079_.m_85850_(), p_115080_.m_6299_(Sheets.m_110789_()), (BlockState)null, modelmanager.m_119422_(modelresourcelocation), 1.0F, 1.0F, 1.0F, p_115081_, OverlayTexture.f_118083_);
@@ -69,22 +_,22 @@
       }
 
       if (!itemstack.m_41619_()) {
-         boolean flag1 = itemstack.m_150930_(Items.f_42573_);
+         MapItemSavedData mapitemsaveddata = MapItem.m_42853_(itemstack, p_115076_.f_19853_);
          if (flag) {
             p_115079_.m_85837_(0.0D, 0.0D, 0.5D);
          } else {
             p_115079_.m_85837_(0.0D, 0.0D, 0.4375D);
          }
 
-         int j = flag1 ? p_115076_.m_31823_() % 4 * 2 : p_115076_.m_31823_();
+         int j = mapitemsaveddata != null ? p_115076_.m_31823_() % 4 * 2 : p_115076_.m_31823_();
          p_115079_.m_85845_(Vector3f.f_122227_.m_122240_((float)j * 360.0F / 8.0F));
-         if (flag1) {
+         if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderItemInFrameEvent(p_115076_, this, p_115079_, p_115080_, p_115081_))) {
+         if (mapitemsaveddata != null) {
             p_115079_.m_85845_(Vector3f.f_122227_.m_122240_(180.0F));
             float f = 0.0078125F;
             p_115079_.m_85841_(0.0078125F, 0.0078125F, 0.0078125F);
             p_115079_.m_85837_(-64.0D, -64.0D, 0.0D);
             Integer integer = MapItem.m_151131_(itemstack);
-            MapItemSavedData mapitemsaveddata = MapItem.m_151128_(integer, p_115076_.f_19853_);
             p_115079_.m_85837_(0.0D, 0.0D, -1.0D);
             if (mapitemsaveddata != null) {
                int i = this.m_174208_(p_115076_, 15728850, p_115081_);
@@ -94,6 +_,7 @@
             int k = this.m_174208_(p_115076_, 15728880, p_115081_);
             p_115079_.m_85841_(0.5F, 0.5F, 0.5F);
             this.f_115047_.m_174269_(itemstack, ItemTransforms.TransformType.FIXED, k, OverlayTexture.f_118083_, p_115079_, p_115080_, p_115076_.m_142049_());
+         }
          }
       }
 
