--- a/net/minecraft/client/renderer/entity/LivingRenderer.java
+++ b/net/minecraft/client/renderer/entity/LivingRenderer.java
@@ -59,17 +59,19 @@
    }
 
    public void func_76986_a(T p_76986_1_, double p_76986_2_, double p_76986_4_, double p_76986_6_, float p_76986_8_, float p_76986_9_) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Pre<T, M>(p_76986_1_, this, p_76986_9_, p_76986_2_, p_76986_4_, p_76986_6_))) return;
       GlStateManager.pushMatrix();
       GlStateManager.disableCull();
       this.field_77045_g.field_217112_c = this.func_77040_d(p_76986_1_, p_76986_9_);
-      this.field_77045_g.field_217113_d = p_76986_1_.func_184218_aH();
+      boolean shouldSit = p_76986_1_.func_184218_aH() && (p_76986_1_.func_184187_bx() != null && p_76986_1_.func_184187_bx().shouldRiderSit());
+      this.field_77045_g.field_217113_d = shouldSit;
       this.field_77045_g.field_217114_e = p_76986_1_.func_70631_g_();
 
       try {
          float f = MathHelper.func_219805_h(p_76986_9_, p_76986_1_.field_70760_ar, p_76986_1_.field_70761_aq);
          float f1 = MathHelper.func_219805_h(p_76986_9_, p_76986_1_.field_70758_at, p_76986_1_.field_70759_as);
          float f2 = f1 - f;
-         if (p_76986_1_.func_184218_aH() && p_76986_1_.func_184187_bx() instanceof LivingEntity) {
+         if (shouldSit && p_76986_1_.func_184187_bx() instanceof LivingEntity) {
             LivingEntity livingentity = (LivingEntity)p_76986_1_.func_184187_bx();
             f = MathHelper.func_219805_h(p_76986_9_, livingentity.field_70760_ar, livingentity.field_70761_aq);
             f2 = f1 - f;
@@ -153,6 +155,7 @@
       GlStateManager.enableCull();
       GlStateManager.popMatrix();
       super.func_76986_a(p_76986_1_, p_76986_2_, p_76986_4_, p_76986_6_, p_76986_8_, p_76986_9_);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Post<T, M>(p_76986_1_, this, p_76986_9_, p_76986_2_, p_76986_4_, p_76986_6_));
    }
 
    public float func_188322_c(T p_188322_1_, float p_188322_2_) {
@@ -408,6 +411,7 @@
    }
 
    public void func_177067_a(T p_177067_1_, double p_177067_2_, double p_177067_4_, double p_177067_6_) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Specials.Pre<T, M>(p_177067_1_, this, p_177067_2_, p_177067_4_, p_177067_6_))) return;
       if (this.func_177070_b(p_177067_1_)) {
          double d0 = p_177067_1_.func_195048_a(this.field_76990_c.field_217783_c.func_216785_c());
          float f = p_177067_1_.func_213287_bg() ? 32.0F : 64.0F;
@@ -417,6 +421,7 @@
             this.func_188296_a(p_177067_1_, p_177067_2_, p_177067_4_, p_177067_6_, s, d0);
          }
       }
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Specials.Post<T, M>(p_177067_1_, this, p_177067_2_, p_177067_4_, p_177067_6_));
    }
 
    protected boolean func_177070_b(T p_177070_1_) {
