--- a/net/minecraft/client/renderer/entity/LivingRenderer.java
+++ b/net/minecraft/client/renderer/entity/LivingRenderer.java
@@ -48,14 +_,17 @@
    }
 
    public void render(T p_225623_1_, float p_225623_2_, float p_225623_3_, MatrixStack p_225623_4_, IRenderTypeBuffer p_225623_5_, int p_225623_6_) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Pre<T, M>(p_225623_1_, this, p_225623_3_, p_225623_4_, p_225623_5_, p_225623_6_))) return;
       p_225623_4_.pushPose();
       this.model.attackTime = this.getAttackAnim(p_225623_1_, p_225623_3_);
-      this.model.riding = p_225623_1_.isPassenger();
+
+      boolean shouldSit = p_225623_1_.isPassenger() && (p_225623_1_.getVehicle() != null && p_225623_1_.getVehicle().shouldRiderSit());
+      this.model.riding = shouldSit;
       this.model.young = p_225623_1_.isBaby();
       float f = MathHelper.rotLerp(p_225623_3_, p_225623_1_.yBodyRotO, p_225623_1_.yBodyRot);
       float f1 = MathHelper.rotLerp(p_225623_3_, p_225623_1_.yHeadRotO, p_225623_1_.yHeadRot);
       float f2 = f1 - f;
-      if (p_225623_1_.isPassenger() && p_225623_1_.getVehicle() instanceof LivingEntity) {
+      if (shouldSit && p_225623_1_.getVehicle() instanceof LivingEntity) {
          LivingEntity livingentity = (LivingEntity)p_225623_1_.getVehicle();
          f = MathHelper.rotLerp(p_225623_3_, livingentity.yBodyRotO, livingentity.yBodyRot);
          f2 = f1 - f;
@@ -92,7 +_,7 @@
       p_225623_4_.translate(0.0D, (double)-1.501F, 0.0D);
       float f8 = 0.0F;
       float f5 = 0.0F;
-      if (!p_225623_1_.isPassenger() && p_225623_1_.isAlive()) {
+      if (!shouldSit && p_225623_1_.isAlive()) {
          f8 = MathHelper.lerp(p_225623_3_, p_225623_1_.animationSpeedOld, p_225623_1_.animationSpeed);
          f5 = p_225623_1_.animationPosition - p_225623_1_.animationSpeed * (1.0F - p_225623_3_);
          if (p_225623_1_.isBaby()) {
@@ -125,6 +_,7 @@
 
       p_225623_4_.popPose();
       super.render(p_225623_1_, p_225623_2_, p_225623_3_, p_225623_4_, p_225623_5_, p_225623_6_);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Post<T, M>(p_225623_1_, this, p_225623_3_, p_225623_4_, p_225623_5_, p_225623_6_));
    }
 
    @Nullable
