--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RenderFish.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RenderFish.java
@@ -75,7 +75,7 @@
             int k = entityplayer.func_184591_cq() == EnumHandSide.RIGHT ? 1 : -1;
             ItemStack itemstack = entityplayer.func_184614_ca();
 
-            if (itemstack.func_77973_b() != Items.field_151112_aM)
+            if (!(itemstack.func_77973_b() instanceof net.minecraft.item.ItemFishingRod))
             {
                 k = -k;
             }
