--- a/net/minecraft/client/renderer/entity/RenderManager.java
+++ b/net/minecraft/client/renderer/entity/RenderManager.java
@@ -241,8 +241,13 @@
       this.field_178637_m = new RenderPlayer(this);
       this.field_178636_l.put("default", this.field_178637_m);
       this.field_178636_l.put("slim", new RenderPlayer(this, true));
+      net.minecraftforge.fml.client.registry.RenderingRegistry.loadEntityRenderers(this, this.field_78729_o);
    }
 
+   public Map<String, RenderPlayer> getSkinMap() {
+       return (Map<String, RenderPlayer>) java.util.Collections.unmodifiableMap(field_178636_l);
+   }
+
    public void func_178628_a(double p_178628_1_, double p_178628_3_, double p_178628_5_) {
       this.field_78725_b = p_178628_1_;
       this.field_78726_c = p_178628_3_;
@@ -277,10 +282,11 @@
       this.field_147941_i = p_180597_4_;
       this.field_78736_p = p_180597_2_;
       if (p_180597_3_ instanceof EntityLivingBase && ((EntityLivingBase)p_180597_3_).func_70608_bn()) {
-         IBlockState iblockstate = p_180597_1_.func_180495_p(new BlockPos(p_180597_3_));
+         BlockPos pos = new BlockPos(p_180597_3_);
+         IBlockState iblockstate = p_180597_1_.func_180495_p(pos);
          Block block = iblockstate.func_177230_c();
-         if (block instanceof BlockBed) {
-            int i = iblockstate.func_177229_b(BlockBed.field_185512_D).func_176736_b();
+         if (block.isBed(iblockstate, p_180597_1_, pos, p_180597_3_)) {
+            int i = block.getBedDirection(iblockstate, p_180597_1_, pos).func_176736_b();
             this.field_78735_i = (float)(i * 90 + 180);
             this.field_78732_j = 0.0F;
          }
