--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RenderManager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RenderManager.java
@@ -219,11 +219,13 @@
         if (p_147938_4_.isPlayerSleeping())
         {
             Block block = p_147938_1_.getBlock(MathHelper.floor_double(p_147938_4_.posX), MathHelper.floor_double(p_147938_4_.posY), MathHelper.floor_double(p_147938_4_.posZ));
+            int x = MathHelper.floor_double(p_147938_4_.posX);
+            int y = MathHelper.floor_double(p_147938_4_.posY);
+            int z = MathHelper.floor_double(p_147938_4_.posZ);
 
-            if (block == Blocks.bed)
+            if (block.isBed(p_147938_1_, x, y, z, p_147938_4_))
             {
-                int i = p_147938_1_.getBlockMetadata(MathHelper.floor_double(p_147938_4_.posX), MathHelper.floor_double(p_147938_4_.posY), MathHelper.floor_double(p_147938_4_.posZ));
-                int j = i & 3;
+                int j = block.getBedDirection(p_147938_1_, x, y, z);
                 this.playerViewY = (float)(j * 90 + 180);
                 this.playerViewX = 0.0F;
             }
