--- a/net/minecraft/util/text/translation/LanguageMap.java
+++ b/net/minecraft/util/text/translation/LanguageMap.java
@@ -26,21 +26,34 @@
    private long lastUpdateTimeInMilliseconds;
 
    public LanguageMap() {
-      try {
          InputStream inputstream = LanguageMap.class.getResourceAsStream("/assets/minecraft/lang/en_us.json");
+         injectLanguage(this, inputstream);
+   }
+
+   public static void injectLanguage(InputStream inputStream){
+      injectLanguage(field_197636_c, inputStream);
+   }
+
+   private static void injectLanguage(LanguageMap inst, InputStream inputStream){
+      final Map<String, String> map = parseLanguageFile(inputStream);
+      inst.languageList.putAll(map);
+      inst.lastUpdateTimeInMilliseconds = System.currentTimeMillis();
+   }
+
+   private static Map<String, String> parseLanguageFile(InputStream inputstream) {
+      Map<String,String> languageList = Maps.newHashMap();
+      try {
          JsonElement jsonelement = (JsonElement)(new Gson()).fromJson(new InputStreamReader(inputstream, StandardCharsets.UTF_8), JsonElement.class);
          JsonObject jsonobject = JsonUtils.getJsonObject(jsonelement, "strings");
 
          for(Entry<String, JsonElement> entry : jsonobject.entrySet()) {
             String s = NUMERIC_VARIABLE_PATTERN.matcher(JsonUtils.getString(entry.getValue(), entry.getKey())).replaceAll("%$1s");
-            this.languageList.put(entry.getKey(), s);
+            languageList.put(entry.getKey(), s);
          }
-
-         this.lastUpdateTimeInMilliseconds = Util.func_211177_b();
       } catch (JsonParseException jsonparseexception) {
          field_201045_a.error("Couldn't read strings from /assets/minecraft/lang/en_us.json", (Throwable)jsonparseexception);
       }
-
+      return languageList;
    }
 
    public static LanguageMap getInstance() {
