--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -76,6 +_,7 @@
     private ItemStack pickupItemStack = this.getDefaultPickupItem();
     @Nullable
     private ItemStack firedFromWeapon = null;
+    private final IntOpenHashSet ignoredEntities = new IntOpenHashSet();
 
     protected AbstractArrow(EntityType<? extends AbstractArrow> p_332730_, Level p_335646_) {
         super(p_332730_, p_335646_);
@@ -189,7 +_,7 @@
             this.shakeTime--;
         }
 
-        if (this.isInWaterOrRain()) {
+        if (this.isInWaterOrRain() || this.isInFluidType((fluidType, height) -> this.canFluidExtinguish(fluidType))) {
             this.clearFire();
         }
 
@@ -566,7 +_,7 @@
     protected boolean canHitEntity(Entity p_36743_) {
         return p_36743_ instanceof Player && this.getOwner() instanceof Player player && !player.canHarmPlayer((Player)p_36743_)
             ? false
-            : super.canHitEntity(p_36743_) && (this.piercingIgnoreEntityIds == null || !this.piercingIgnoreEntityIds.contains(p_36743_.getId()));
+            : super.canHitEntity(p_36743_) && (this.piercingIgnoreEntityIds == null || !this.piercingIgnoreEntityIds.contains(p_36743_.getId())) && !this.ignoredEntities.contains(p_36743_.getId());
     }
 
     @Override
