--- a/net/minecraft/world/entity/projectile/LargeFireball.java
+++ b/net/minecraft/world/entity/projectile/LargeFireball.java
@@ -31,7 +_,7 @@
     protected void onHit(HitResult p_37218_) {
         super.onHit(p_37218_);
         if (this.level() instanceof ServerLevel serverlevel) {
-            boolean flag = serverlevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+            boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(serverlevel, this.getOwner());
             this.level().explode(this, this.getX(), this.getY(), this.getZ(), this.explosionPower, flag, Level.ExplosionInteraction.MOB);
             this.discard();
         }
