--- a/net/minecraft/world/entity/projectile/ProjectileUtil.java
+++ b/net/minecraft/world/entity/projectile/ProjectileUtil.java
@@ -68,7 +_,7 @@
             Vec3 vec31 = optional.get();
             double d1 = p_37289_.m_82557_(vec31);
             if (d1 < d0 || d0 == 0.0D) {
-               if (entity1.m_20201_() == p_37288_.m_20201_()) {
+               if (entity1.m_20201_() == p_37288_.m_20201_() && !entity1.canRiderInteract()) {
                   if (d0 == 0.0D) {
                      entity = entity1;
                      vec3 = vec31;
@@ -138,8 +_,13 @@
       }
    }
 
+   @Deprecated // Forge: Use the version below that takes in a Predicate<Item> instead of an Item
    public static InteractionHand m_37297_(LivingEntity p_37298_, Item p_37299_) {
       return p_37298_.m_21205_().m_150930_(p_37299_) ? InteractionHand.MAIN_HAND : InteractionHand.OFF_HAND;
+   }
+
+   public static InteractionHand getWeaponHoldingHand(LivingEntity livingEntity, Predicate<Item> itemPredicate) {
+      return itemPredicate.test(livingEntity.m_21205_().m_41720_()) ? InteractionHand.MAIN_HAND : InteractionHand.OFF_HAND;
    }
 
    public static AbstractArrow m_37300_(LivingEntity p_37301_, ItemStack p_37302_, float p_37303_) {
