diff -u -r --strip-trailing-cr ../src_base/minecraft_server/net/minecraft/server/MinecraftServer.java ../src_work/minecraft_server/net/minecraft/server/MinecraftServer.java
--- ../src_base/minecraft_server/net/minecraft/server/MinecraftServer.java	2011-07-17 00:28:07.755771900 +0200
+++ ../src_work/minecraft_server/net/minecraft/server/MinecraftServer.java	2011-07-17 00:28:09.485870900 +0200
@@ -397,38 +397,7 @@
             {
                 System.exit(0);
             }
-            break MISSING_BLOCK_LABEL_350;
         }
-        try
-        {
-            stopServer();
-            serverStopped = true;
-        }
-        catch (Throwable throwable)
-        {
-            throwable.printStackTrace();
-        }
-        finally
-        {
-            System.exit(0);
-        }
-        break MISSING_BLOCK_LABEL_350;
-        Exception exception2;
-        exception2;
-        try
-        {
-            stopServer();
-            serverStopped = true;
-        }
-        catch (Throwable throwable3)
-        {
-            throwable3.printStackTrace();
-        }
-        finally
-        {
-            System.exit(0);
-        }
-        throw exception2;
     }
 
     private void doTick()

diff U3 ../src_base/minecraft_server/net/minecraft/src/MLProp.java ../src_work/minecraft_server/net/minecraft/src/MLProp.java
--- ../src_base/minecraft_server/net/minecraft/src/MLProp.java	Wed Nov 30 22:27:22 2011
+++ ../src_work/minecraft_server/net/minecraft/src/MLProp.java	Sun Dec 04 14:31:04 2011
@@ -8,4 +8,6 @@
 	String info() default "";
 	double min() default Double.NEGATIVE_INFINITY;
 	double max() default Double.POSITIVE_INFINITY;
+    
+    public int reobf() default 0;
 }
\ No newline at end of file
