--- a/net/minecraft/particles/ParticleType.java
+++ b/net/minecraft/particles/ParticleType.java
@@ -4,7 +4,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public abstract class ParticleType<T extends IParticleData> {
+public abstract class ParticleType<T extends IParticleData>  extends net.minecraftforge.registries.ForgeRegistryEntry<ParticleType<?>>{
    private final boolean field_197581_e;
    private final IParticleData.IDeserializer<T> field_197582_f;
 
