--- ../src-base/minecraft/net/minecraft/world/gen/structure/StructureMineshaftPieces.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/StructureMineshaftPieces.java
@@ -31,6 +31,12 @@
         MapGenStructureIO.registerStructureComponent(StructureMineshaftPieces.Stairs.class, "MSStairs");
     }
 
+    static
+    {
+        net.minecraftforge.common.ChestGenHooks.init(net.minecraftforge.common.ChestGenHooks.MINESHAFT_CORRIDOR, CHEST_CONTENT_WEIGHT_LIST, 8, 8);
+        net.minecraftforge.common.ChestGenHooks.addItem(net.minecraftforge.common.ChestGenHooks.MINESHAFT_CORRIDOR, new WeightedRandomChestContent(new net.minecraft.item.ItemStack(net.minecraft.init.Items.enchanted_book, 1, 0), 1, 1, 1));
+    }
+
     private static StructureComponent func_175892_a(List<StructureComponent> listIn, Random rand, int x, int y, int z, EnumFacing facing, int type)
     {
         int i = rand.nextInt(100);
