--- ../src_base/common/net/minecraft/src/BlockCrops.java
+++ ../src_work/common/net/minecraft/src/BlockCrops.java
@@ -2,7 +2,11 @@
 
 import cpw.mods.fml.common.Side;
 import cpw.mods.fml.common.asm.SideOnly;
+
+import java.util.ArrayList;
 import java.util.Random;
+
+import net.minecraftforge.common.ForgeDirection;
 
 public class BlockCrops extends BlockFlower
 {
@@ -88,11 +92,11 @@
                 int var19 = par1World.getBlockId(var17, par3 - 1, var18);
                 float var20 = 0.0F;
 
-                if (var19 == Block.tilledField.blockID)
+                if (blocksList[var19] != null && blocksList[var19].canSustainPlant(par1World, var17, par3 - 1, var18, ForgeDirection.UP, this))
                 {
                     var20 = 1.0F;
 
-                    if (par1World.getBlockMetadata(var17, par3 - 1, var18) > 0)
+                    if (blocksList[var19].isFertile(par1World, var17, par3 - 1, var18))
                     {
                         var20 = 3.0F;
                     }
@@ -152,22 +156,37 @@
     public void dropBlockAsItemWithChance(World par1World, int par2, int par3, int par4, int par5, float par6, int par7)
     {
         super.dropBlockAsItemWithChance(par1World, par2, par3, par4, par5, par6, 0);
-
-        if (!par1World.isRemote)
-        {
-            if (par5 >= 7)
-            {
-                int var8 = 3 + par7;
-
-                for (int var9 = 0; var9 < var8; ++var9)
-                {
-                    if (par1World.rand.nextInt(15) <= par5)
-                    {
-                        this.dropBlockAsItem_do(par1World, par2, par3, par4, new ItemStack(this.func_82532_h(), 1, 0));
-                    }
-                }
-            }
-        }
+    }
+
+    @Override 
+    public ArrayList<ItemStack> getBlockDropped(World world, int x, int y, int z, int metadata, int fortune)
+    {
+        ArrayList<ItemStack> ret = new ArrayList<ItemStack>();
+        if (metadata == 7)
+        {      
+            int count = quantityDropped(metadata, fortune, world.rand);
+            for(int i = 0; i < count; i++)
+            {
+                int id = idDropped(metadata, world.rand, 0);
+                if (id > 0)
+                {
+                    ret.add(new ItemStack(id, 1, damageDropped(metadata)));
+                }
+            }
+        }
+
+        if (metadata >= 7)
+        {
+            for (int n = 0; n < 3 + fortune; n++)
+            {
+                if (world.rand.nextInt(15) <= metadata)
+                {
+                    ret.add(new ItemStack(this.func_82532_h(), 1, 0));
+                }
+            }
+        }
+
+        return ret;
     }
 
     /**
