--- ../src_base/common/net/minecraft/src/BlockCactus.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/common/net/minecraft/src/BlockCactus.java	0000-00-00 00:00:00.000000000 -0000
@@ -1,8 +1,11 @@
 package net.minecraft.src;
 
+import java.util.Random;
+
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.block.CanBlockStayCheckEvent;
 import cpw.mods.fml.common.Side;
 import cpw.mods.fml.common.asm.SideOnly;
-import java.util.Random;
 
 public class BlockCactus extends Block
 {
@@ -124,6 +127,7 @@
      */
     public boolean canBlockStay(World par1World, int par2, int par3, int par4)
     {
+    	CanBlockStayCheckEvent event = new CanBlockStayCheckEvent(this, par1World, par2, par3, par4);
         if (par1World.getBlockMaterial(par2 - 1, par3, par4).isSolid())
         {
             return false;
@@ -140,6 +144,14 @@
         {
             return false;
         }
+        else if (MinecraftForge.EVENT_BUS.post(event))
+        {
+            return false;
+        }
+        else if (event.canBePlaced())
+        {
+            return true;
+        }
         else
         {
             int var5 = par1World.getBlockId(par2, par3 - 1, par4);
