--- ../src_base/common/net/minecraft/src/BlockReed.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/common/net/minecraft/src/BlockReed.java	0000-00-00 00:00:00.000000000 -0000
@@ -4,6 +4,9 @@
 import cpw.mods.fml.common.asm.SideOnly;
 import java.util.Random;
 
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.block.CanBlockStayCheckEvent;
+
 public class BlockReed extends Block
 {
     protected BlockReed(int par1, int par2)
@@ -51,6 +54,17 @@
      */
     public boolean canPlaceBlockAt(World par1World, int par2, int par3, int par4)
     {
+    	CanBlockStayCheckEvent event = new CanBlockStayCheckEvent(this, new ItemStack(par1World.getBlockId(par2, par3 - 1, par4), 1, par1World.getBlockMetadata(par2, par3 - 1, par4)));
+    	if(MinecraftForge.EVENT_BUS.post(event))
+    	{
+    		return false;
+    	}
+    	
+    	if(event.canBePlaced())
+    	{
+    		return true;
+    	}
+    	
         int var5 = par1World.getBlockId(par2, par3 - 1, par4);
         return var5 == this.blockID ? true : (var5 != Block.grass.blockID && var5 != Block.dirt.blockID && var5 != Block.sand.blockID ? false : (par1World.getBlockMaterial(par2 - 1, par3 - 1, par4) == Material.water ? true : (par1World.getBlockMaterial(par2 + 1, par3 - 1, par4) == Material.water ? true : (par1World.getBlockMaterial(par2, par3 - 1, par4 - 1) == Material.water ? true : par1World.getBlockMaterial(par2, par3 - 1, par4 + 1) == Material.water))));
     }
