--- ../src_base/common/net/minecraft/src/BlockDoor.java
+++ ../src_work/common/net/minecraft/src/BlockDoor.java
@@ -3,6 +3,7 @@
 import cpw.mods.fml.common.Side;
 import cpw.mods.fml.common.asm.SideOnly;
 import java.util.Random;
+import static net.minecraftforge.common.ForgeDirection.*;
 
 public class BlockDoor extends Block
 {
@@ -260,7 +261,7 @@
     {
         if (this.blockMaterial == Material.iron)
         {
-            return true;
+            return false; //Allow items to interact with the door
         }
         else
         {
