--- ../src_base/common/net/minecraft/src/ChunkProviderEnd.java
+++ ../src_work/common/net/minecraft/src/ChunkProviderEnd.java
@@ -2,6 +2,9 @@
 
 import java.util.List;
 import java.util.Random;
+
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.world.chunkprovider.InitChunkTerrainNoiseEvent;
 
 public class ChunkProviderEnd implements IChunkProvider
 {
@@ -41,6 +44,10 @@
         byte var7 = 33;
         int var8 = var5 + 1;
         this.densities = this.initializeNoiseField(this.densities, par1 * var5, 0, par2 * var5, var6, var7, var8);
+        
+        InitChunkTerrainNoiseEvent event = new InitChunkTerrainNoiseEvent(this.densities, this, this.endWorld); 
+        MinecraftForge.EVENT_BUS.post(event);
+        this.densities = event.noiseArray;
 
         for (int var9 = 0; var9 < var5; ++var9)
         {
