--- ../src_base/common/net/minecraft/src/ChunkProviderServer.java
+++ ../src_work/common/net/minecraft/src/ChunkProviderServer.java
@@ -7,6 +7,7 @@
 import java.util.List;
 import java.util.Set;
 
+import net.minecraftforge.common.DimensionManager;
 import cpw.mods.fml.common.registry.GameRegistry;
 
 public class ChunkProviderServer implements IChunkProvider
@@ -50,7 +51,7 @@
      */
     public void unloadChunksIfNotNearSpawn(int par1, int par2)
     {
-        if (this.currentServer.provider.canRespawnHere())
+        if (this.currentServer.provider.canRespawnHere() && DimensionManager.shouldLoadSpawn(currentServer.provider.worldType))
         {
             ChunkCoordinates var3 = this.currentServer.getSpawnPoint();
             int var4 = par1 * 16 + 8 - var3.posX;
@@ -286,6 +287,11 @@
                     this.chunksToUnload.remove(var2);
                     this.loadedChunkHashMap.remove(var2.longValue());
                     this.loadedChunks.remove(var3);
+                    if(loadedChunks.size() == 0 && !DimensionManager.shouldLoadSpawn(currentServer.provider.worldType))
+                    {
+                        DimensionManager.unloadWorld(currentServer.provider.worldType);
+                        return currentChunkProvider.unload100OldestChunks();
+                    }
                 }
             }
 
