--- ../src_base/common/net/minecraft/src/ColorizerGrass.java
+++ ../src_work/common/net/minecraft/src/ColorizerGrass.java
@@ -1,11 +1,19 @@
 package net.minecraft.src;
 
+import net.minecraftforge.client.IGrassColorizer;
 import cpw.mods.fml.common.Side;
 import cpw.mods.fml.common.asm.SideOnly;
 
 @SideOnly(Side.CLIENT)
 public class ColorizerGrass
 {
+    /**
+     * Interface for overriding the functions within this class. Normally you 
+     * could just extend the class, but since all methods are static, rather 
+     * than being called through an instance, that won't really work.
+     */
+    public static IGrassColorizer GrassColorizer = null;
+    
     /** Color buffer for grass */
     private static int[] grassBuffer = new int[65536];
 
@@ -19,6 +27,8 @@
      */
     public static int getGrassColor(double par0, double par2)
     {
+        if (GrassColorizer != null) return GrassColorizer.getGrassColor(par0, par2, grassBuffer);
+        
         par2 *= par0;
         int var4 = (int)((1.0D - par0) * 255.0D);
         int var5 = (int)((1.0D - par2) * 255.0D);
