--- ../src_base/common/net/minecraft/src/ComponentMineshaftCorridor.java
+++ ../src_work/common/net/minecraft/src/ComponentMineshaftCorridor.java
@@ -2,6 +2,9 @@
 
 import java.util.List;
 import java.util.Random;
+
+import net.minecraftforge.common.ChestGenHooks;
+import static net.minecraftforge.common.ChestGenHooks.*;
 
 public class ComponentMineshaftCorridor extends StructureComponent
 {
@@ -231,14 +234,16 @@
                 this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 1, 2, var10 - 1, Block.torchWood.blockID, 0);
                 this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 1, 2, var10 + 1, Block.torchWood.blockID, 0);
 
+                ChestGenHooks info = ChestGenHooks.getInfo(MINESHAFT_CORRIDOR);
+
                 if (par2Random.nextInt(100) == 0)
                 {
-                    this.generateStructureChestContents(par1World, par3StructureBoundingBox, par2Random, 2, 0, var10 - 1, StructureMineshaftPieces.func_78816_a(), 3 + par2Random.nextInt(4));
+                    this.generateStructureChestContents(par1World, par3StructureBoundingBox, par2Random, 2, 0, var10 - 1, info.getItems(), info.getCount(par2Random));
                 }
 
                 if (par2Random.nextInt(100) == 0)
                 {
-                    this.generateStructureChestContents(par1World, par3StructureBoundingBox, par2Random, 0, 0, var10 + 1, StructureMineshaftPieces.func_78816_a(), 3 + par2Random.nextInt(4));
+                    this.generateStructureChestContents(par1World, par3StructureBoundingBox, par2Random, 0, 0, var10 + 1, info.getItems(), info.getCount(par2Random));
                 }
 
                 if (this.hasSpiders && !this.spawnerPlaced)
