--- ../src_base/common/net/minecraft/src/ContainerFurnace.java
+++ ../src_work/common/net/minecraft/src/ContainerFurnace.java
@@ -119,7 +119,7 @@
             }
             else if (par2 != 1 && par2 != 0)
             {
-                if (FurnaceRecipes.smelting().getSmeltingResult(var5.getItem().shiftedIndex) != null)
+                if (FurnaceRecipes.smelting().getSmeltingResult(var5) != null)
                 {
                     if (!this.mergeItemStack(var5, 0, 1, false))
                     {
