--- ../src_base/common/net/minecraft/src/EntityItem.java
+++ ../src_work/common/net/minecraft/src/EntityItem.java
@@ -2,6 +2,9 @@
 
 import java.util.Iterator;
 
+import net.minecraftforge.common.IAdvancedDroppable;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.entity.player.EntityItemPickupEvent;
 import cpw.mods.fml.common.registry.GameRegistry;
 
 public class EntityItem extends Entity
@@ -59,6 +62,14 @@
     {
         super.onUpdate();
 
+        if (!this.worldObj.isRemote && (age == 0 || age == 4800)){ //CLASS EDIT
+                if (this.item.getItem() instanceof IAdvancedDroppable) {
+                        worldObj.spawnEntityInWorld(((IAdvancedDroppable)this.item.getItem()).createEntity(worldObj, this, this.item));
+                        this.setDead();
+                        return;
+                }
+        } //END EDIT
+
         if (this.delayBeforeCanPickup > 0)
         {
             --this.delayBeforeCanPickup;
@@ -122,6 +133,11 @@
         {
             this.setDead();
         }
+
+        if (this.item == null || this.item.stackSize <= 0)
+        {
+            this.setDead();
+        }
     }
 
     public boolean func_70289_a(EntityItem par1EntityItem)
@@ -225,7 +241,7 @@
         NBTTagCompound var2 = par1NBTTagCompound.getCompoundTag("Item");
         this.item = ItemStack.loadItemStackFromNBT(var2);
 
-        if (this.item == null)
+        if (this.item == null || this.item.stackSize <= 0)
         {
             this.setDead();
         }
@@ -238,9 +254,21 @@
     {
         if (!this.worldObj.isRemote)
         {
+            if (this.delayBeforeCanPickup > 0)
+            {
+                return;
+            }
+
+            EntityItemPickupEvent event = new EntityItemPickupEvent(par1EntityPlayer, this);
+
+            if (MinecraftForge.EVENT_BUS.post(event))
+            {
+                return;
+            }
+
             int var2 = this.item.stackSize;
 
-            if (this.delayBeforeCanPickup == 0 && par1EntityPlayer.inventory.addItemStackToInventory(this.item))
+            if (this.delayBeforeCanPickup <= 0 && (event.isHandled() || var2 <= 0 || par1EntityPlayer.inventory.addItemStackToInventory(this.item)))
             {
                 if (this.item.itemID == Block.wood.blockID)
                 {
