--- ../src-base/common/net/minecraft/src/EntityPlayerMP.java
+++ ../src-work/common/net/minecraft/src/EntityPlayerMP.java
@@ -270,6 +270,7 @@
     public void func_70645_a(DamageSource p_70645_1_)
     {
         this.field_71133_b.func_71203_ab().func_72384_a(new Packet3Chat(p_70645_1_.func_76360_b(this)));
+        this.field_71133_b.func_71203_ab().field_72406_a.info(p_70645_1_.func_76360_b(this));
         this.field_71071_by.func_70436_m();
     }
 
