--- /Volumes/Data/Users/tsingh/Desktop/patches/old/EntityPlayer.java	2012-08-18 00:28:36.000000000 +1000
+++ /Volumes/Data/Users/tsingh/Desktop/patches/new/EntityPlayer.java	2012-08-17 23:57:09.000000000 +1000
@@ -14,6 +14,7 @@
 import net.minecraftforge.event.entity.living.LivingHurtEvent;
 import net.minecraftforge.event.entity.player.AttackEntityEvent;
 import net.minecraftforge.event.entity.player.EntityInteractEvent;
+import net.minecraftforge.event.entity.player.NBTEvent;
 import net.minecraftforge.event.entity.player.PlayerDestroyItemEvent;
 import net.minecraftforge.event.entity.player.PlayerSleepInBedEvent;
 
@@ -796,6 +797,7 @@
             NBTTagList var3 = par1NBTTagCompound.getTagList("EnderItems");
             this.theInventoryEnderChest.loadInventoryFromNBT(var3);
         }
+        MinecraftForge.EVENT_BUS.post(new NBTEvent.NBTLoad(this, par1NBTTagCompound));
     }
 
     /**
@@ -822,6 +824,7 @@
         this.foodStats.writeNBT(par1NBTTagCompound);
         this.capabilities.writeCapabilitiesToNBT(par1NBTTagCompound);
         par1NBTTagCompound.setTag("EnderItems", this.theInventoryEnderChest.saveInventoryToNBT());
+        MinecraftForge.EVENT_BUS.post(new NBTEvent.NBTSave(this, par1NBTTagCompound));
     }
 
     /**
