--- C:/Users/Keithy/Desktop/EntityFishHook_orig.java	Fri Oct 26 13:24:29 2012
+++ C:/Users/Keithy/Desktop/EntityFishHook.java	Thu Nov 01 21:22:15 2012
@@ -5,6 +5,10 @@
 import java.util.Iterator;
 import java.util.List;
 
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event;
+import net.minecraftforge.event.entity.player.CatchFishEvent;
+
 public class EntityFishHook extends Entity
 {
     /** The tile this entity is on, X position */
@@ -474,7 +478,24 @@
             }
             else if (this.ticksCatchable > 0)
             {
-                EntityItem var13 = new EntityItem(this.worldObj, this.posX, this.posY, this.posZ, new ItemStack(Item.fishRaw));
+                CatchFishEvent event = new CatchFishEvent(this.angler, this.worldObj, this.posX, this.posY, this.posZ);
+                if(MinecraftForge.EVENT_BUS.post(event))
+                {
+                    this.setDead();
+                    this.angler.fishEntity = null;
+                    return event.damage;
+                }
+                
+                EntityItem var13;
+                if (event.getResult() == Event.Result.ALLOW)
+                {
+                    var13 = new EntityItem(this.worldObj, this.posX, this.posY, this.posZ, event.result);
+                } 
+                else 
+                { 
+                    var13 = new EntityItem(this.worldObj, this.posX, this.posY, this.posZ, new ItemStack(Item.fishRaw));
+                }
+                
                 double var3 = this.angler.posX - this.posX;
                 double var5 = this.angler.posY - this.posY;
                 double var7 = this.angler.posZ - this.posZ;
@@ -486,7 +507,7 @@
                 this.worldObj.spawnEntityInWorld(var13);
                 this.angler.addStat(StatList.fishCaughtStat, 1);
                 this.angler.worldObj.spawnEntityInWorld(new EntityXPOrb(this.angler.worldObj, this.angler.posX, this.angler.posY + 0.5D, this.angler.posZ + 0.5D, this.rand.nextInt(3) + 1));
-                var1 = 1;
+                var1 = event.damage;
             }
 
             if (this.inGround)
