--- ../src_base/common/net/minecraft/src/ItemTool.java
+++ ../src_work/common/net/minecraft/src/ItemTool.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import net.minecraftforge.common.ForgeHooks;
 import cpw.mods.fml.common.Side;
 import cpw.mods.fml.common.asm.SideOnly;
 
@@ -105,4 +106,15 @@
     {
         return this.toolMaterial.getToolCraftingMaterial() == par2ItemStack.itemID ? true : super.getIsRepairable(par1ItemStack, par2ItemStack);
     }
+
+    /** FORGE: Overridden to allow custom tool effectiveness */
+    @Override
+    public float getStrVsBlock(ItemStack stack, Block block, int meta) 
+    {
+        if (ForgeHooks.isToolEffective(stack, block, meta))
+        {
+            return efficiencyOnProperMaterial;
+        }
+        return getStrVsBlock(stack, block);
+    }
 }
