--- ../src_base/common/net/minecraft/src/ItemBucket.java
+++ ../src_work/common/net/minecraft/src/ItemBucket.java
@@ -1,4 +1,8 @@
 package net.minecraft.src;
+
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event;
+import net.minecraftforge.event.entity.player.FillBucketEvent;
 
 public class ItemBucket extends Item
 {
@@ -31,6 +35,32 @@
         }
         else
         {
+            FillBucketEvent event = new FillBucketEvent(par3EntityPlayer, par1ItemStack, par2World, var12);
+            if (MinecraftForge.EVENT_BUS.post(event))
+            {
+                return par1ItemStack;
+            }
+
+            if (event.getResult() == Event.Result.ALLOW)
+            {
+                if (par3EntityPlayer.capabilities.isCreativeMode)
+                {
+                    return par1ItemStack;
+                }
+
+                if (--par1ItemStack.stackSize <= 0)
+                {
+                    return event.result;
+                }
+
+                if (!par3EntityPlayer.inventory.addItemStackToInventory(event.result))
+                {
+                    par3EntityPlayer.dropPlayerItem(event.result);
+                }
+
+                return par1ItemStack;
+            }
+
             if (var12.typeOfHit == EnumMovingObjectType.TILE)
             {
                 int var13 = var12.blockX;
