--- ../src_base/common/net/minecraft/src/MapData.java
+++ ../src_work/common/net/minecraft/src/MapData.java
@@ -11,7 +11,7 @@
 {
     public int xCenter;
     public int zCenter;
-    public byte dimension;
+    public int dimension;
     public byte scale;
 
     /** colours */
@@ -44,7 +44,7 @@
      */
     public void readFromNBT(NBTTagCompound par1NBTTagCompound)
     {
-        this.dimension = par1NBTTagCompound.getByte("dimension");
+        this.dimension = par1NBTTagCompound.getInteger("dimension");
         this.xCenter = par1NBTTagCompound.getInteger("xCenter");
         this.zCenter = par1NBTTagCompound.getInteger("zCenter");
         this.scale = par1NBTTagCompound.getByte("scale");
@@ -98,7 +98,7 @@
      */
     public void writeToNBT(NBTTagCompound par1NBTTagCompound)
     {
-        par1NBTTagCompound.setByte("dimension", this.dimension);
+        par1NBTTagCompound.setInteger("dimension", this.dimension);
         par1NBTTagCompound.setInteger("xCenter", this.xCenter);
         par1NBTTagCompound.setInteger("zCenter", this.zCenter);
         par1NBTTagCompound.setByte("scale", this.scale);
