--- ../src_base/common/net/minecraft/src/MapGenVillage.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/common/net/minecraft/src/MapGenVillage.java	0000-00-00 00:00:00.000000000 -0000
@@ -1,19 +1,29 @@
 package net.minecraft.src;
 
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 import java.util.Random;
+import java.util.Set;
+
+import net.minecraftforge.common.BiomeManager;
+
+import com.google.common.collect.Sets;
 
 public class MapGenVillage extends MapGenStructure
 {
     /** A list of all the biomes villages can spawn in. */
-    public static List villageSpawnBiomes = Arrays.asList(new BiomeGenBase[] {BiomeGenBase.plains, BiomeGenBase.desert});
+    public static ArrayList villageSpawnBiomes;
 
     /** World terrain type, 0 for normal, 1 for flat map */
     private final int terrainType;
-
+    
     public MapGenVillage(int par1)
     {
+    	villageSpawnBiomes = new ArrayList<BiomeGenBase>();
+    	Collections.addAll(villageSpawnBiomes, new BiomeGenBase[] {BiomeGenBase.plains, BiomeGenBase.desert});
+    	villageSpawnBiomes.addAll(BiomeManager.getVillageBiomes());
         this.terrainType = par1;
     }
 
@@ -59,4 +69,5 @@
     {
         return new StructureVillageStart(this.worldObj, this.rand, par1, par2, this.terrainType);
     }
+    
 }
