--- ../src-base/common/net/minecraft/src/NetServerHandler.java
+++ ../src-work/common/net/minecraft/src/NetServerHandler.java
@@ -7,6 +7,8 @@
 import java.util.Iterator;
 import java.util.Random;
 import java.util.logging.Logger;
+
+import cpw.mods.fml.common.network.FMLNetworkHandler;
 import net.minecraft.server.MinecraftServer;
 
 public class NetServerHandler extends NetHandler
@@ -587,6 +589,7 @@
 
     public void func_72481_a(Packet3Chat p_72481_1_)
     {
+        p_72481_1_ = FMLNetworkHandler.handleChatMessage(this, p_72481_1_);
         if (this.field_72574_e.func_71126_v() == 2)
         {
             this.func_72567_b(new Packet3Chat("Cannot send chat message."));
@@ -952,6 +955,11 @@
     }
 
     public void func_72501_a(Packet250CustomPayload p_72501_1_)
+    {
+        FMLNetworkHandler.handlePacket250Packet(p_72501_1_, field_72575_b, this);
+    }
+
+    public void handleVanilla250Packet(Packet250CustomPayload p_72501_1_)
     {
         DataInputStream var2;
         ItemStack var3;
@@ -1114,4 +1122,17 @@
             }
         }
     }
+
+    @Override
+    public void func_72494_a(Packet131MapData p_72494_1_)
+    {
+        FMLNetworkHandler.handlePacket131Packet(this, p_72494_1_);
+    }
+
+    // modloader compat -- yuk!
+    @Override
+    public EntityPlayerMP getPlayer()
+    {
+        return field_72574_e;
+    }
 }
