--- ../src_base/common/net/minecraft/src/Packet51MapChunk.java
+++ ../src_work/common/net/minecraft/src/Packet51MapChunk.java
@@ -100,6 +100,13 @@
 
         var3 = 12288 * var2;
 
+        int msb = 0; //BugFix: MC does not read the MSB array from the packet properly, causing issues for servers that use blocks > 256
+        for (int x = 0; x < 16; x++)
+        {
+            msb += (yChMax >> x) & 1;
+        }
+        var3 += (2048 * msb);
+
         if (this.includeInitialize)
         {
             var3 += 256;
