--- ../src_base/common/net/minecraft/src/Potion.java
+++ ../src_work/common/net/minecraft/src/Potion.java
@@ -118,7 +118,7 @@
         {
             if (par1EntityLiving.getHealth() < par1EntityLiving.getMaxHealth())
             {
-                par1EntityLiving.heal(1);
+                par1EntityLiving.heal(1, net.minecraftforge.event.entity.living.LivingHealEvent.HealReason.MAGIC_REGEN);
             }
         }
         else if (this.id == poison.id)
@@ -141,7 +141,7 @@
         }
         else
         {
-            par1EntityLiving.heal(6 << par2);
+            par1EntityLiving.heal(6 << par2, net.minecraftforge.event.entity.living.LivingHealEvent.HealReason.MAGIC);
         }
     }
 
@@ -171,7 +171,7 @@
         else
         {
             var6 = (int)(par4 * (double)(6 << par3) + 0.5D);
-            par2EntityLiving.heal(var6);
+            par2EntityLiving.heal(var6, net.minecraftforge.event.entity.living.LivingHealEvent.HealReason.MAGIC);
         }
     }
 
