--- ../src_base/common/net/minecraft/src/RailLogic.java
+++ ../src_work/common/net/minecraft/src/RailLogic.java
@@ -18,6 +18,7 @@
     private List connectedTracks;
 
     final BlockRail rail;
+    private final boolean canMakeSlopes;
 
     public RailLogic(BlockRail par1BlockRail, World par2World, int par3, int par4, int par5)
     {
@@ -28,18 +29,11 @@
         this.trackY = par4;
         this.trackZ = par5;
         int var6 = par2World.getBlockId(par3, par4, par5);
-        int var7 = par2World.getBlockMetadata(par3, par4, par5);
-
-        if (BlockRail.isPoweredBlockRail((BlockRail)Block.blocksList[var6]))
-        {
-            this.isPoweredRail = true;
-            var7 &= -9;
-        }
-        else
-        {
-            this.isPoweredRail = false;
-        }
-
+
+        BlockRail target = (BlockRail)Block.blocksList[var6];
+        int var7 = target.getBasicRailMetadata(par2World, null, par3, par4, par5);
+        isPoweredRail = !target.isFlexibleRail(par2World, par3, par4, par5);
+        canMakeSlopes = target.canMakeSlopes(par2World, par3, par4, par5);
         this.setConnections(var7);
     }
 
@@ -258,7 +252,7 @@
             }
         }
 
-        if (var6 == 0)
+        if (var6 == 0 && canMakeSlopes)
         {
             if (BlockRail.isRailBlockAt(this.worldObj, this.trackX, this.trackY + 1, this.trackZ - 1))
             {
@@ -271,7 +265,7 @@
             }
         }
 
-        if (var6 == 1)
+        if (var6 == 1 && canMakeSlopes)
         {
             if (BlockRail.isRailBlockAt(this.worldObj, this.trackX + 1, this.trackY + 1, this.trackZ))
             {
@@ -422,7 +416,7 @@
             }
         }
 
-        if (var7 == 0)
+        if (var7 == 0 && canMakeSlopes)
         {
             if (BlockRail.isRailBlockAt(this.worldObj, this.trackX, this.trackY + 1, this.trackZ - 1))
             {
@@ -435,7 +429,7 @@
             }
         }
 
-        if (var7 == 1)
+        if (var7 == 1 && canMakeSlopes)
         {
             if (BlockRail.isRailBlockAt(this.worldObj, this.trackX + 1, this.trackY + 1, this.trackZ))
             {
