--- ../src_base/common/net/minecraft/src/SaveHandler.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/common/net/minecraft/src/SaveHandler.java	0000-00-00 00:00:00.000000000 -0000
@@ -8,6 +8,8 @@
 import java.io.IOException;
 import java.util.logging.Logger;
 
+import net.minecraftforge.common.NBTHelper;
+
 public class SaveHandler implements ISaveHandler, IPlayerFileData
 {
     /** Reference to the logger. */
@@ -120,6 +122,7 @@
      */
     public WorldInfo loadWorldInfo()
     {
+    	NBTHelper.loadNBT(this.saveDirectory);
         File var1 = new File(this.saveDirectory, "level.dat");
         NBTTagCompound var2;
         NBTTagCompound var3;
@@ -203,6 +206,7 @@
      */
     public void saveWorldInfo(WorldInfo par1WorldInfo)
     {
+    	NBTHelper.saveNBT(this.saveDirectory);
         NBTTagCompound var2 = par1WorldInfo.getNBTTagCompound();
         NBTTagCompound var3 = new NBTTagCompound();
         var3.setTag("Data", var2);
@@ -244,6 +248,7 @@
      */
     public void writePlayerData(EntityPlayer par1EntityPlayer)
     {
+    	NBTHelper.savePlayerNBT(this.saveDirectory, par1EntityPlayer);
         try
         {
             NBTTagCompound var2 = new NBTTagCompound();
@@ -270,6 +275,7 @@
      */
     public void readPlayerData(EntityPlayer par1EntityPlayer)
     {
+    	NBTHelper.loadPlayerNBT(this.saveDirectory, par1EntityPlayer);
         NBTTagCompound var2 = this.getPlayerData(par1EntityPlayer.username);
 
         if (var2 != null)
