--- ../src_base/common/net/minecraft/src/SlotFurnace.java
+++ ../src_work/common/net/minecraft/src/SlotFurnace.java
@@ -62,7 +62,7 @@
         if (!this.thePlayer.worldObj.isRemote)
         {
             int var2 = this.field_75228_b;
-            float var3 = FurnaceRecipes.smelting().getExperience(par1ItemStack.itemID);
+            float var3 = FurnaceRecipes.smelting().getExperience(par1ItemStack);
             int var4;
 
             if (var3 == 0.0F)
