--- ../src-base/common/net/minecraft/src/TcpConnection.java
+++ ../src-work/common/net/minecraft/src/TcpConnection.java
@@ -16,6 +16,8 @@
 import java.util.List;
 import java.util.concurrent.atomic.AtomicInteger;
 import javax.crypto.SecretKey;
+
+import cpw.mods.fml.common.network.FMLNetworkHandler;
 
 public class TcpConnection implements INetworkManager
 {
@@ -377,6 +379,7 @@
         if (this.field_74472_n && this.field_74473_o.isEmpty())
         {
             this.field_74485_r.func_72515_a(this.field_74481_v, this.field_74480_w);
+            FMLNetworkHandler.onConnectionClosed(this, this.field_74485_r.getPlayer());
         }
     }
 
