--- ../src_base/common/net/minecraft/src/TileEntityBrewingStand.java
+++ ../src_work/common/net/minecraft/src/TileEntityBrewingStand.java
@@ -4,7 +4,10 @@
 import cpw.mods.fml.common.asm.SideOnly;
 import java.util.List;
 
-public class TileEntityBrewingStand extends TileEntity implements IInventory
+import net.minecraftforge.common.ISidedInventory;
+import net.minecraftforge.common.ForgeDirection;
+
+public class TileEntityBrewingStand extends TileEntity implements IInventory, ISidedInventory
 {
     /** The itemstacks currently placed in the slots of the brewing stand */
     private ItemStack[] brewingItemStacks = new ItemStack[4];
@@ -158,7 +161,7 @@
 
             if (Item.itemsList[var1.itemID].hasContainerItem())
             {
-                this.brewingItemStacks[3] = new ItemStack(Item.itemsList[var1.itemID].getContainerItem());
+                this.brewingItemStacks[3] = Item.itemsList[var1.itemID].getContainerItemStack(brewingItemStacks[3]);
             }
             else
             {
@@ -325,4 +328,16 @@
 
         return var1;
     }
+
+    @Override
+    public int getStartInventorySide(ForgeDirection side)
+    {
+        return (side == ForgeDirection.UP ? 3 : 0);
+    }
+
+    @Override
+    public int getSizeInventorySide(ForgeDirection side)
+    {
+        return (side == ForgeDirection.UP ? 1 : 3);
+    }
 }
