--- ../src_base/common/net/minecraft/src/WorldProviderSurface.java
+++ ../src_work/common/net/minecraft/src/WorldProviderSurface.java
@@ -11,4 +11,12 @@
     {
         return "Overworld";
     }
+
+    @Override
+    @SideOnly(Side.CLIENT)
+    public SkyProvider getSkyProvider()
+    {
+        SkyProvider skyProvider = super.getSkyProvider();
+        return skyProvider != null ? skyProvider : SkyProviderOverworld.instance;
+    }
 }
