--- ../src_base/common/net/minecraft/src/WorldProviderEnd.java
+++ ../src_work/common/net/minecraft/src/WorldProviderEnd.java
@@ -141,4 +141,12 @@
     {
         return "The End";
     }
+
+    @Override
+    @SideOnly(Side.CLIENT)
+    public SkyProvider getSkyProvider()
+    {
+        SkyProvider skyProvider = super.getSkyProvider();
+        return skyProvider != null ? skyProvider : SkyProviderEnd.instance;
+    }
 }
