--- ../src_base/common/net/minecraft/src/WorldServer.java
+++ ../src_work/common/net/minecraft/src/WorldServer.java
@@ -10,6 +10,9 @@
 import java.util.Set;
 import java.util.TreeSet;
 import net.minecraft.server.MinecraftServer;
+import net.minecraftforge.common.DimensionManager;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.world.WorldEvent;
 
 public class WorldServer extends World
 {
@@ -50,6 +53,8 @@
     /** An IntHashMap of entity IDs (integers) to their Entity objects. */
     private IntHashMap entityIdMap;
 
+    protected Set<ChunkCoordIntPair> doneChunks = new HashSet<ChunkCoordIntPair>();
+
     public WorldServer(MinecraftServer par1MinecraftServer, ISaveHandler par2ISaveHandler, String par3Str, int par4, WorldSettings par5WorldSettings, Profiler par6Profiler)
     {
         super(par2ISaveHandler, par3Str, par5WorldSettings, WorldProvider.getProviderForDimension(par4), par6Profiler);
@@ -71,6 +76,7 @@
         {
             this.pendingTickListEntries = new TreeSet();
         }
+        DimensionManager.setWorld(par4, this);
     }
 
     /**
@@ -179,6 +185,9 @@
 
     private void resetRainAndThunder()
     {
+        this.provider.resetRainAndThunder();
+    }
+    public void resetRainAndThunderBody() {
         this.worldInfo.setRainTime(0);
         this.worldInfo.setRaining(false);
         this.worldInfo.setThunderTime(0);
@@ -253,6 +262,14 @@
         int var1 = 0;
         int var2 = 0;
         Iterator var3 = this.activeChunkSet.iterator();
+
+        doneChunks.retainAll(activeChunkSet);
+        if (doneChunks.size() == activeChunkSet.size())
+        {
+            doneChunks.clear();
+        }
+
+        final long time = -System.currentTimeMillis();
 
         while (var3.hasNext())
         {
@@ -263,13 +280,19 @@
             Chunk var7 = this.getChunkFromChunkCoords(var4.chunkXPos, var4.chunkZPos);
             this.moodSoundAndLightCheck(var5, var6, var7);
             this.theProfiler.endStartSection("tickChunk");
-            var7.updateSkylight();
-            this.theProfiler.endStartSection("thunder");
+            if (System.currentTimeMillis() + time <= 4 && doneChunks.add(var4)) {
+                var7.updateSkylight();
+            }
+            this.theProfiler.endStartSection("dimensionChunk");
+            this.theProfiler.startSection("unknown");
             int var8;
             int var9;
             int var10;
             int var11;
-
+            int var13;
+
+            if (!this.provider.tickBlocksAndAmbiance(var7)) {
+            this.theProfiler.startSection("thunder");
             if (this.rand.nextInt(100000) == 0 && this.isRaining() && this.isThundering())
             {
                 this.updateLCG = this.updateLCG * 3 + 1013904223;
@@ -286,7 +309,6 @@
             }
 
             this.theProfiler.endStartSection("iceandsnow");
-            int var13;
 
             if (this.rand.nextInt(16) == 0)
             {
@@ -321,6 +343,8 @@
                     }
                 }
             }
+            }
+            this.theProfiler.endSection();
 
             this.theProfiler.endStartSection("tickTiles");
             ExtendedBlockStorage[] var19 = var7.getBlockStorageArray();
@@ -559,15 +583,27 @@
     public List getAllTileEntityInBox(int par1, int par2, int par3, int par4, int par5, int par6)
     {
         ArrayList var7 = new ArrayList();
-        Iterator var8 = this.loadedTileEntityList.iterator();
-
-        while (var8.hasNext())
-        {
-            TileEntity var9 = (TileEntity)var8.next();
-
-            if (var9.xCoord >= par1 && var9.yCoord >= par2 && var9.zCoord >= par3 && var9.xCoord < par4 && var9.yCoord < par5 && var9.zCoord < par6)
-            {
-                var7.add(var9);
+
+        for(int x = (par1 >> 4); x <= (par4 >> 4); x++)
+        {
+            for(int z = (par3 >> 4); z <= (par6 >> 4); z++)
+            {
+                Chunk chunk = getChunkFromChunkCoords(x, z);
+                if (chunk != null)
+                {
+                    for(Object obj : chunk.chunkTileEntityMap.values())
+                    {
+                        TileEntity entity = (TileEntity)obj;
+                        if (!entity.isInvalid())
+                        {
+                            if (entity.xCoord >= par1 && entity.yCoord >= par2 && entity.zCoord >= par3 && 
+                                entity.xCoord <= par4 && entity.yCoord <= par5 && entity.zCoord <= par6)
+                            {
+                                var7.add(entity);
+                            }
+                        }
+                    }
+                }
             }
         }
 
@@ -579,6 +615,9 @@
      */
     public boolean canMineBlock(EntityPlayer par1EntityPlayer, int par2, int par3, int par4)
     {
+        return super.canMineBlock(par1EntityPlayer, par2, par3, par4);
+    }
+    public boolean canMineBlockBody(EntityPlayer par1EntityPlayer, int par2, int par3, int par4) {
         int var5 = MathHelper.abs_int(par2 - this.worldInfo.getSpawnX());
         int var6 = MathHelper.abs_int(par4 - this.worldInfo.getSpawnZ());
 
@@ -587,7 +626,7 @@
             var6 = var5;
         }
 
-        return var6 > 16 || this.mcServer.getConfigurationManager().areCommandsAllowed(par1EntityPlayer.username) || this.mcServer.isSinglePlayer();
+        return var6 > mcServer.spawnProtectionSize || this.mcServer.getConfigurationManager().areCommandsAllowed(par1EntityPlayer.username) || this.mcServer.isSinglePlayer();
     }
 
     protected void initialize(WorldSettings par1WorldSettings)
@@ -710,6 +749,7 @@
             }
 
             this.chunkProvider.saveChunks(par1, par2IProgressUpdate);
+            MinecraftForge.EVENT_BUS.post(new WorldEvent.Save(this));
         }
     }
 
@@ -906,7 +946,7 @@
         boolean var1 = this.isRaining();
         super.updateWeather();
 
-        if (var1 != this.isRaining())
+        if (var1 != this.isRaining() && this.provider.worldType == 0) //Only send data about weather changing on Overworld.  Does the native Weather model require rewrite to handle this right? Empirically this works fine. - XCompWiz
         {
             if (var1)
             {
