--- a/net/minecraft/network/protocol/game/ClientboundCustomPayloadPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundCustomPayloadPacket.java
@@ -4,7 +_,7 @@
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.resources.ResourceLocation;
 
-public class ClientboundCustomPayloadPacket implements Packet<ClientGamePacketListener> {
+public class ClientboundCustomPayloadPacket implements Packet<ClientGamePacketListener>, net.minecraftforge.fmllegacy.network.ICustomPacket<ClientboundCustomPayloadPacket> {
    private static final int f_178834_ = 1048576;
    public static final ResourceLocation f_132012_ = new ResourceLocation("brand");
    public static final ResourceLocation f_132013_ = new ResourceLocation("debug/path");
