--- a/net/minecraft/network/protocol/game/ServerboundContainerClickPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundContainerClickPacket.java
@@ -49,7 +_,7 @@
       p_133961_.writeByte(this.f_133941_);
       p_133961_.m_130068_(this.f_133944_);
       p_133961_.m_236831_(this.f_179569_, FriendlyByteBuf::writeShort, FriendlyByteBuf::m_130055_);
-      p_133961_.m_130055_(this.f_179568_);
+      p_133961_.writeItemStack(this.f_179568_, false); //Forge: Include full tag for C->S
    }
 
    public void m_5797_(ServerGamePacketListener p_133958_) {
