--- a/net/minecraft/entity/boss/dragon/EnderDragonEntity.java
+++ b/net/minecraft/entity/boss/dragon/EnderDragonEntity.java
@@ -408,8 +408,8 @@
                BlockPos blockpos = new BlockPos(k1, l1, i2);
                BlockState blockstate = this.field_70170_p.func_180495_p(blockpos);
                Block block = blockstate.func_177230_c();
-               if (!blockstate.func_196958_f() && blockstate.func_185904_a() != Material.field_151581_o) {
-                  if (this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223599_b) && !BlockTags.field_219754_W.func_199685_a_(block)) {
+               if (!blockstate.isAir(this.field_70170_p, blockpos) && blockstate.func_185904_a() != Material.field_151581_o) {
+                  if (net.minecraftforge.common.ForgeHooks.canEntityDestroy(this.field_70170_p, blockpos, this) && !BlockTags.field_219754_W.func_199685_a_(block)) {
                      flag1 = this.field_70170_p.func_217377_a(blockpos, false) || flag1;
                   } else {
                      flag = true;
