--- a/net/minecraft/tileentity/BeehiveTileEntity.java
+++ b/net/minecraft/tileentity/BeehiveTileEntity.java
@@ -23,12 +23,13 @@ import net.minecraft.nbt.NBTUtil;
 import net.minecraft.network.DebugPacketSender;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.EntityTypeTags;
+import net.minecraft.tileentity.IBeehiveTileEntity;
 import net.minecraft.util.Direction;
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;

-public class BeehiveTileEntity extends TileEntity implements ITickableTileEntity {
+public class BeehiveTileEntity extends TileEntity implements ITickableTileEntity, IBeehiveTileEntity {
     public List<BeehiveTileEntity.Bee> field_226958_a_ = Lists.newArrayList();
     @Nullable
     public BlockPos field_226959_b_ = null;
