--- a/net/minecraft/tileentity/BeehiveTileEntity.java
+++ b/net/minecraft/tileentity/BeehiveTileEntity.java
@@ -23,7 +23,7 @@
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;
 
-public class BeehiveTileEntity extends TileEntity implements ITickableTileEntity {
+public class BeehiveTileEntity extends TileEntity implements ITickableTileEntity, net.minecraftforge.common.extensions.IForgeBeehiveTileEntity {
    private final List<BeehiveTileEntity.Bee> field_226958_a_ = Lists.newArrayList();
    @Nullable
    private BlockPos field_226959_b_ = null;
