--- a/net/minecraft/tileentity/PistonTileEntity.java
+++ b/net/minecraft/tileentity/PistonTileEntity.java
@@ -105,7 +105,7 @@
          AxisAlignedBB axisalignedbb = this.func_190607_a(this.func_191515_a(list));
          List<Entity> list1 = this.field_145850_b.func_72839_b((Entity)null, AabbHelper.func_227019_a_(axisalignedbb, direction, d0).func_111270_a(axisalignedbb));
          if (!list1.isEmpty()) {
-            boolean flag = this.field_200231_a.func_177230_c() == Blocks.field_180399_cE;
+            boolean flag = this.field_200231_a.isSlimeBlock(); //TODO: Merge with isStickyBlock? Look into further how vanilla splits slime vs honey
 
             for(Entity entity : list1) {
                if (entity.func_184192_z() != PushReaction.IGNORE) {
