--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityBrewingStand.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityBrewingStand.java
@@ -12,6 +12,8 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagList;
 import net.minecraft.potion.PotionHelper;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.brewing.PotionBrewedEvent;
 
 public class TileEntityBrewingStand extends TileEntity implements ISidedInventory
 {
@@ -109,7 +111,7 @@
 
                 for (int i = 0; i < 3; ++i)
                 {
-                    if (this.brewingItemStacks[i] != null && this.brewingItemStacks[i].getItem() == Items.potionitem)
+                    if (this.brewingItemStacks[i] != null && this.brewingItemStacks[i].getItem() instanceof ItemPotion)
                     {
                         int j = this.brewingItemStacks[i].getItemDamage();
                         int k = this.func_145936_c(j, itemstack);
@@ -148,7 +150,7 @@
 
             for (int i = 0; i < 3; ++i)
             {
-                if (this.brewingItemStacks[i] != null && this.brewingItemStacks[i].getItem() == Items.potionitem)
+                if (this.brewingItemStacks[i] != null && this.brewingItemStacks[i].getItem() instanceof ItemPotion)
                 {
                     int j = this.brewingItemStacks[i].getItemDamage();
                     int k = this.func_145936_c(j, itemstack);
@@ -169,9 +171,9 @@
                 }
             }
 
-            if (itemstack.getItem().hasContainerItem())
+            if (itemstack.getItem().hasContainerItem(itemstack))
             {
-                this.brewingItemStacks[3] = new ItemStack(itemstack.getItem().getContainerItem());
+                this.brewingItemStacks[3] = itemstack.getItem().getContainerItem(itemstack);
             }
             else
             {
@@ -182,6 +184,7 @@
                     this.brewingItemStacks[3] = null;
                 }
             }
+            MinecraftForge.EVENT_BUS.post(new PotionBrewedEvent(brewingItemStacks));
         }
     }
 
@@ -304,7 +307,7 @@
     // JAVADOC METHOD $$ func_94041_b
     public boolean isItemValidForSlot(int par1, ItemStack par2ItemStack)
     {
-        return par1 == 3 ? par2ItemStack.getItem().isPotionIngredient(par2ItemStack) : par2ItemStack.getItem() == Items.potionitem || par2ItemStack.getItem() == Items.glass_bottle;
+        return par1 == 3 ? par2ItemStack.getItem().isPotionIngredient(par2ItemStack) : par2ItemStack.getItem() instanceof ItemPotion || par2ItemStack.getItem() == Items.glass_bottle;
     }
 
     @SideOnly(Side.CLIENT)
