--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityFurnace.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityFurnace.java
@@ -141,9 +141,9 @@
             }
         }
 
-        this.field_145956_a = p_145839_1_.func_74765_d("BurnTime");
-        this.field_174906_k = p_145839_1_.func_74765_d("CookTime");
-        this.field_174905_l = p_145839_1_.func_74765_d("CookTimeTotal");
+        this.field_145956_a = p_145839_1_.func_74762_e("BurnTime");
+        this.field_174906_k = p_145839_1_.func_74762_e("CookTime");
+        this.field_174905_l = p_145839_1_.func_74762_e("CookTimeTotal");
         this.field_145963_i = func_145952_a(this.field_145957_n[1]);
 
         if (p_145839_1_.func_150297_b("CustomName", 8))
@@ -155,9 +155,9 @@
     public void func_145841_b(NBTTagCompound p_145841_1_)
     {
         super.func_145841_b(p_145841_1_);
-        p_145841_1_.func_74777_a("BurnTime", (short)this.field_145956_a);
-        p_145841_1_.func_74777_a("CookTime", (short)this.field_174906_k);
-        p_145841_1_.func_74777_a("CookTimeTotal", (short)this.field_174905_l);
+        p_145841_1_.func_74768_a("BurnTime", this.field_145956_a);
+        p_145841_1_.func_74768_a("CookTime", this.field_174906_k);
+        p_145841_1_.func_74768_a("CookTimeTotal", this.field_174905_l);
         NBTTagList nbttaglist = new NBTTagList();
 
         for (int i = 0; i < this.field_145957_n.length; ++i)
@@ -223,8 +223,7 @@
 
                             if (this.field_145957_n[1].field_77994_a == 0)
                             {
-                                Item item = this.field_145957_n[1].func_77973_b().func_77668_q();
-                                this.field_145957_n[1] = item != null ? new ItemStack(item) : null;
+                                this.field_145957_n[1] = field_145957_n[1].func_77973_b().getContainerItem(field_145957_n[1]);
                             }
                         }
                     }
@@ -279,7 +278,11 @@
         else
         {
             ItemStack itemstack = FurnaceRecipes.func_77602_a().func_151395_a(this.field_145957_n[0]);
-            return itemstack == null ? false : (this.field_145957_n[2] == null ? true : (!this.field_145957_n[2].func_77969_a(itemstack) ? false : (this.field_145957_n[2].field_77994_a < this.func_70297_j_() && this.field_145957_n[2].field_77994_a < this.field_145957_n[2].func_77976_d() ? true : this.field_145957_n[2].field_77994_a < itemstack.func_77976_d())));
+            if (itemstack == null) return false;
+            if (this.field_145957_n[2] == null) return true;
+            if (!this.field_145957_n[2].func_77969_a(itemstack)) return false;
+            int result = field_145957_n[2].field_77994_a + itemstack.field_77994_a;
+            return result <= func_70297_j_() && result <= this.field_145957_n[2].func_77976_d(); //Forge BugFix: Make it respect stack sizes properly.
         }
     }
 
@@ -295,7 +298,7 @@
             }
             else if (this.field_145957_n[2].func_77973_b() == itemstack.func_77973_b())
             {
-                ++this.field_145957_n[2].field_77994_a;
+                this.field_145957_n[2].field_77994_a += itemstack.field_77994_a; // Forge BugFix: Results may have multiple items
             }
 
             if (this.field_145957_n[0].func_77973_b() == Item.func_150898_a(Blocks.field_150360_v) && this.field_145957_n[0].func_77960_j() == 1 && this.field_145957_n[1] != null && this.field_145957_n[1].func_77973_b() == Items.field_151133_ar)
@@ -342,7 +345,15 @@
                 }
             }
 
-            return item instanceof ItemTool && ((ItemTool)item).func_77861_e().equals("WOOD") ? 200 : (item instanceof ItemSword && ((ItemSword)item).func_150932_j().equals("WOOD") ? 200 : (item instanceof ItemHoe && ((ItemHoe)item).func_77842_f().equals("WOOD") ? 200 : (item == Items.field_151055_y ? 100 : (item == Items.field_151044_h ? 1600 : (item == Items.field_151129_at ? 20000 : (item == Item.func_150898_a(Blocks.field_150345_g) ? 100 : (item == Items.field_151072_bj ? 2400 : 0)))))));
+            if (item instanceof ItemTool && ((ItemTool)item).func_77861_e().equals("WOOD")) return 200;
+            if (item instanceof ItemSword && ((ItemSword)item).func_150932_j().equals("WOOD")) return 200;
+            if (item instanceof ItemHoe && ((ItemHoe)item).func_77842_f().equals("WOOD")) return 200;
+            if (item == Items.field_151055_y) return 100;
+            if (item == Items.field_151044_h) return 1600;
+            if (item == Items.field_151129_at) return 20000;
+            if (item == Item.func_150898_a(Blocks.field_150345_g)) return 100;
+            if (item == Items.field_151072_bj) return 2400;
+            return net.minecraftforge.fml.common.registry.GameRegistry.getFuelValue(p_145952_0_);
         }
     }
 
@@ -451,4 +462,21 @@
             this.field_145957_n[i] = null;
         }
     }
+
+    net.minecraftforge.items.IItemHandler handlerTop = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.UP);
+    net.minecraftforge.items.IItemHandler handlerBottom = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.DOWN);
+    net.minecraftforge.items.IItemHandler handlerSide = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.WEST);
+
+    @Override
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, net.minecraft.util.EnumFacing facing)
+    {
+        if (facing != null && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+            if (facing == EnumFacing.DOWN)
+                return (T) handlerBottom;
+            else if (facing == EnumFacing.UP)
+                return (T) handlerTop;
+            else
+                return (T) handlerSide;
+        return super.getCapability(capability, facing);
+    }
 }
