--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityHopper.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityHopper.java
@@ -404,17 +404,30 @@
 
             if (itemstack1 == null)
             {
-                p_145899_0_.setInventorySlotContents(p_145899_2_, p_145899_1_);
-                p_145899_1_ = null;
+                //Forge: BUGFIX: Again, make things respect max stack sizes.
+                int max = Math.min(p_145899_1_.getMaxStackSize(), p_145899_0_.getInventoryStackLimit());
+                if (max >= p_145899_1_.stackSize)
+                {
+                    p_145899_0_.setInventorySlotContents(p_145899_2_, p_145899_1_);
+                    p_145899_1_ = null;
+                }
+                else
+                {
+                    p_145899_0_.setInventorySlotContents(p_145899_2_, p_145899_1_.splitStack(max));
+                }
                 flag = true;
             }
             else if (func_145894_a(itemstack1, p_145899_1_))
             {
-                int k = p_145899_1_.getMaxStackSize() - itemstack1.stackSize;
-                int l = Math.min(p_145899_1_.stackSize, k);
-                p_145899_1_.stackSize -= l;
-                itemstack1.stackSize += l;
-                flag = l > 0;
+                //Forge: BUGFIX: Again, make things respect max stack sizes.
+                int max = Math.min(p_145899_1_.getMaxStackSize(), p_145899_0_.getInventoryStackLimit());
+                if (max > itemstack1.stackSize)
+                {
+                    int l = Math.min(p_145899_1_.stackSize, max - itemstack1.stackSize);
+                    p_145899_1_.stackSize -= l;
+                    itemstack1.stackSize += l;
+                    flag = l > 0;
+                }
             }
 
             if (flag)
