--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityLockableLoot.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityLockableLoot.java
@@ -73,7 +73,7 @@
 
             if (p_184281_1_ != null)
             {
-                lootcontext$builder.func_186469_a(p_184281_1_.func_184817_da());
+                lootcontext$builder.func_186469_a(p_184281_1_.func_184817_da()).func_186470_a(p_184281_1_); // Forge: add player to LootContext
             }
 
             loottable.func_186460_a(this, random, lootcontext$builder.func_186471_a());
