--- a/net/minecraft/tileentity/TileEntityBeacon.java
+++ b/net/minecraft/tileentity/TileEntityBeacon.java
@@ -141,9 +141,13 @@
                   this.field_174909_f.clear();
                   break;
                }
-
+               float[] customColor = iblockstate.func_177230_c().getBeaconColorMultiplier(iblockstate, this.field_145850_b, blockpos$mutableblockpos, func_174877_v());
+               if(customColor!=null) {
+                  afloat = customColor;
+               } else {
                tileentitybeacon$beamsegment.func_177262_a();
                continue;
+               }
             }
 
             afloat = ((BlockStainedGlassPane)block).func_196419_d().func_193349_f();
