--- a/net/minecraft/tileentity/TileEntityLockableLoot.java
+++ b/net/minecraft/tileentity/TileEntityLockableLoot.java
@@ -69,6 +69,7 @@
 
          LootContext.Builder lootcontext$builder = new LootContext.Builder((WorldServer)this.field_145850_b);
          lootcontext$builder.func_204313_a(this.field_174879_c);
+         lootcontext$builder.func_186470_a(p_184281_1_);
          if (p_184281_1_ != null) {
             lootcontext$builder.func_186469_a(p_184281_1_.func_184817_da());
          }
