--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityHopper.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityHopper.java
@@ -523,17 +523,30 @@
 
             if (itemstack1 == null)
             {
-                p_174916_0_.func_70299_a(p_174916_2_, p_174916_1_);
-                p_174916_1_ = null;
+                //Forge: BUGFIX: Again, make things respect max stack sizes.
+                int max = Math.min(p_174916_1_.func_77976_d(), p_174916_0_.func_70297_j_());
+                if (max >= p_174916_1_.field_77994_a)
+                {
+                    p_174916_0_.func_70299_a(p_174916_2_, p_174916_1_);
+                    p_174916_1_ = null;
+                }
+                else
+                {
+                    p_174916_0_.func_70299_a(p_174916_2_, p_174916_1_.func_77979_a(max));
+                }
                 flag = true;
             }
             else if (func_145894_a(itemstack1, p_174916_1_))
             {
-                int j = p_174916_1_.func_77976_d() - itemstack1.field_77994_a;
-                int k = Math.min(p_174916_1_.field_77994_a, j);
-                p_174916_1_.field_77994_a -= k;
-                itemstack1.field_77994_a += k;
-                flag = k > 0;
+                //Forge: BUGFIX: Again, make things respect max stack sizes.
+                int max = Math.min(p_174916_1_.func_77976_d(), p_174916_0_.func_70297_j_());
+                if (max > itemstack1.field_77994_a)
+                {
+                    int size = Math.min(p_174916_1_.field_77994_a, max - itemstack1.field_77994_a);
+                    p_174916_1_.field_77994_a -= size;
+                    itemstack1.field_77994_a += size;
+                    flag = size > 0;
+                }
             }
 
             if (flag)
