--- a/net/minecraft/client/renderer/entity/model/BipedModel.java
+++ b/net/minecraft/client/renderer/entity/model/BipedModel.java
@@ -62,6 +62,7 @@
 
    public void func_78088_a(T p_78088_1_, float p_78088_2_, float p_78088_3_, float p_78088_4_, float p_78088_5_, float p_78088_6_, float p_78088_7_) {
       this.func_212844_a_(p_78088_1_, p_78088_2_, p_78088_3_, p_78088_4_, p_78088_5_, p_78088_6_, p_78088_7_);
+      if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderBipedModelEvent(this, p_78088_1_, p_78088_2_, p_78088_3_, p_78088_4_, p_78088_5_, p_78088_6_, p_78088_7_))) return;
       GlStateManager.pushMatrix();
       if (this.field_217114_e) {
          float f = 2.0F;
