--- ../src-base/minecraft/net/minecraft/world/SpawnerAnimals.java
+++ ../src-work/minecraft/net/minecraft/world/SpawnerAnimals.java
@@ -17,6 +17,10 @@
 import net.minecraft.util.WeightedRandom;
 import net.minecraft.world.biome.BiomeGenBase;
 import net.minecraft.world.chunk.Chunk;
+import java.util.ArrayList;
+import java.util.Collections;
+import net.minecraftforge.fml.common.eventhandler.Event.Result;
+import net.minecraftforge.event.ForgeEventFactory;
 
 public final class SpawnerAnimals
 {
@@ -80,12 +84,15 @@
 
                 if ((!enumcreaturetype.getPeacefulCreature() || p_77192_3_) && (enumcreaturetype.getPeacefulCreature() || p_77192_2_) && (!enumcreaturetype.getAnimal() || p_77192_4_))
                 {
-                    i1 = p_77192_1_.countEntities(enumcreaturetype.getCreatureClass());
+                    i1 = p_77192_1_.countEntities(enumcreaturetype, true);
                     int i4 = enumcreaturetype.getMaxNumberOfCreature() * i / MOB_COUNT_DIV;
 
                     if (i1 <= i4)
                     {
                         Iterator iterator1 = this.eligibleChunksForSpawning.iterator();
+                        ArrayList<ChunkCoordIntPair> tmp = new ArrayList(eligibleChunksForSpawning);
+                        Collections.shuffle(tmp);
+                        iterator1 = tmp.iterator();
                         label115:
 
                         while (iterator1.hasNext())
@@ -153,8 +160,10 @@
 
                                                         entityliving.setLocationAndAngles((double)f, (double)l2, (double)f1, p_77192_1_.rand.nextFloat() * 360.0F, 0.0F);
 
-                                                        if (entityliving.getCanSpawnHere() && entityliving.handleLavaMovement())
+                                                        Result canSpawn = ForgeEventFactory.canEntitySpawn(entityliving, p_77192_1_, f, l2, f1);
+                                                        if (canSpawn == Result.ALLOW || (canSpawn == Result.DEFAULT && (entityliving.getCanSpawnHere() && entityliving.handleLavaMovement())))
                                                         {
+                                                            if (!ForgeEventFactory.doSpecialSpawn(entityliving, p_77192_1_, f1, l2, f1))
                                                             ientitylivingdata = entityliving.func_180482_a(p_77192_1_.getDifficultyForLocation(new BlockPos(entityliving)), ientitylivingdata);
 
                                                             if (entityliving.handleLavaMovement())
@@ -163,7 +172,7 @@
                                                                 p_77192_1_.spawnEntityInWorld(entityliving);
                                                             }
 
-                                                            if (i2 >= entityliving.getMaxSpawnedInChunk())
+                                                            if (i2 >= ForgeEventFactory.getMaxSpawnPackSize(entityliving))
                                                             {
                                                                 continue label115;
                                                             }
@@ -220,7 +229,7 @@
             {
                 BlockPos blockpos1 = pos.down();
 
-                if (!World.doesBlockHaveSolidTopSurface(worldIn, blockpos1))
+                if (!worldIn.getBlockState(blockpos1).getBlock().canCreatureSpawn(worldIn, blockpos1, p_180267_0_))
                 {
                     return false;
                 }
