--- ../src-base/minecraft/net/minecraft/world/WorldProviderSurface.java
+++ ../src-work/minecraft/net/minecraft/world/WorldProviderSurface.java
@@ -9,6 +9,6 @@
 
     public boolean func_186056_c(int p_186056_1_, int p_186056_2_)
     {
-        return !this.worldObj.isSpawnChunk(p_186056_1_, p_186056_2_);
+        return !this.worldObj.isSpawnChunk(p_186056_1_, p_186056_2_) || !net.minecraftforge.common.DimensionManager.shouldLoadSpawn(this.worldObj.provider.getDimension());
     }
 }
