--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -122,7 +_,7 @@
       String s7 = m_129638_(optionset, optionspec);
       Integer integer = m_129638_(optionset, optionspec1);
       CrashReport.m_127529_();
-      Bootstrap.m_135870_();
+      net.minecraftforge.fml.loading.BackgroundWaiter.runAndTick(()->Bootstrap.m_135870_(), net.minecraftforge.fml.loading.FMLLoader.progressWindowTick);
       Bootstrap.m_135889_();
       Util.m_137584_();
       User user = new User(optionspec9.value(optionset), s5, optionspec11.value(optionset), optionspec20.value(optionset));
