--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -123,8 +_,8 @@
             CrashReport.preload();
             logger = LogUtils.getLogger();
             s1 = "Bootstrap";
-            Bootstrap.bootStrap();
-            ClientBootstrap.bootstrap();
+            net.minecraftforge.fml.loading.BackgroundWaiter.runAndTick(Bootstrap::bootStrap, net.minecraftforge.fml.loading.FMLLoader.progressWindowTick);
+            net.minecraftforge.fml.loading.BackgroundWaiter.runAndTick(ClientBootstrap::bootstrap, net.minecraftforge.fml.loading.FMLLoader.progressWindowTick);
             GameLoadTimesEvent.INSTANCE.setBootstrapTime(Bootstrap.bootstrapDuration.get());
             Bootstrap.validate();
             s1 = "Argument parsing";
