--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -154,8 +_,8 @@
         CrashReport.preload();
 
         try {
-            Bootstrap.bootStrap();
             GameLoadTimesEvent.INSTANCE.setBootstrapTime(Bootstrap.bootstrapDuration.get());
+            net.minecraftforge.fml.loading.BackgroundWaiter.runAndTick(Bootstrap::bootStrap, net.minecraftforge.fml.loading.FMLLoader.progressWindowTick);
             Bootstrap.validate();
         } catch (Throwable throwable2) {
             CrashReport crashreport = CrashReport.forThrowable(throwable2, "Bootstrap");
