--- ../src_base/minecraft_server/net/minecraft/src/BlockTrapDoor.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/BlockTrapDoor.java	0000-00-00 00:00:00.000000000 -0000
@@ -2,6 +2,9 @@
 
 public class BlockTrapDoor extends Block
 {
+    /** Set this to allow trapdoors to remain free-floating */
+    public static boolean disableValidation = false;
+    
     protected BlockTrapDoor(int par1, Material par2Material)
     {
         super(par1, par2Material);
@@ -174,7 +177,7 @@
                 --var7;
             }
 
-            if (!isValidSupportBlock(par1World.getBlockId(var7, par3, var8)))
+            if (!(isValidSupportBlock(par1World.getBlockId(var7, par3, var8)) || par1World.isBlockSolidOnSide(var7, par3, var8, (var6 & 3) + 2)))
             {
                 par1World.setBlockWithNotify(par2, par3, par4, 0);
                 this.dropBlockAsItem(par1World, par2, par3, par4, var6, 0);
@@ -235,6 +238,11 @@
      */
     public boolean canPlaceBlockOnSide(World par1World, int par2, int par3, int par4, int par5)
     {
+        if (disableValidation)
+        {
+            return true;
+        }
+        
         if (par5 == 0)
         {
             return false;
@@ -265,7 +273,7 @@
                 --par2;
             }
 
-            return isValidSupportBlock(par1World.getBlockId(par2, par3, par4));
+            return isValidSupportBlock(par1World.getBlockId(par2, par3, par4)) || par1World.isBlockSolidOnSide(par2, par3, par4, par5);
         }
     }
 
