--- ../src_base/minecraft_server/net/minecraft/src/ChunkProviderServer.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/ChunkProviderServer.java	0000-00-00 00:00:00.000000000 -0000
@@ -9,6 +9,8 @@
 
 import cpw.mods.fml.server.FMLServerHandler;
 
+import net.minecraft.src.forge.ForgeHooks;
+
 public class ChunkProviderServer implements IChunkProvider
 {
     private Set droppedChunksSet = new HashSet();
@@ -51,6 +53,11 @@
 
     public void dropChunk(int par1, int par2)
     {
+        if(!ForgeHooks.canUnloadChunk(world.getChunkFromChunkCoords(par1, par2)))
+        {
+            return;
+        }
+        
         if (this.world.worldProvider.canRespawnHere())
         {
             ChunkCoordinates var3 = this.world.getSpawnPoint();
