--- ../src_base/minecraft_server/net/minecraft/src/ConsoleCommandHandler.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/ConsoleCommandHandler.java	0000-00-00 00:00:00.000000000 -0000
@@ -4,6 +4,7 @@
 import java.util.Set;
 import java.util.logging.Logger;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.src.forge.DimensionManager;
 
 public class ConsoleCommandHandler
 {
@@ -52,9 +53,9 @@
                         var5.savePlayerStates();
                     }
 
-                    for (var6 = 0; var6 < this.minecraftServer.worldMngr.length; ++var6)
+                    for (World world : DimensionManager.getWorlds())
                     {
-                        var7 = this.minecraftServer.worldMngr[var6];
+                        var7 = (WorldServer)world;
                         boolean var8 = var7.levelSaving;
                         var7.levelSaving = false;
                         var7.saveWorld(true, (IProgressUpdate)null);
@@ -67,20 +68,18 @@
                 {
                     this.sendNoticeToOps(var4, "Disabling level saving..");
 
-                    for (var6 = 0; var6 < this.minecraftServer.worldMngr.length; ++var6)
+                    for (World world : DimensionManager.getWorlds())
                     {
-                        var7 = this.minecraftServer.worldMngr[var6];
-                        var7.levelSaving = true;
+                        ((WorldServer)world).levelSaving = true;
                     }
                 }
                 else if (var2.toLowerCase().startsWith("save-on"))
                 {
                     this.sendNoticeToOps(var4, "Enabling level saving..");
 
-                    for (var6 = 0; var6 < this.minecraftServer.worldMngr.length; ++var6)
+                    for (World world : DimensionManager.getWorlds())
                     {
-                        var7 = this.minecraftServer.worldMngr[var6];
-                        var7.levelSaving = false;
+                        ((WorldServer)world).levelSaving = false;
                     }
                 }
                 else
@@ -356,20 +355,18 @@
 
                                             if ("add".equalsIgnoreCase(var20))
                                             {
-                                                for (var19 = 0; var19 < this.minecraftServer.worldMngr.length; ++var19)
+                                                for (World world : DimensionManager.getWorlds())
                                                 {
-                                                    var23 = this.minecraftServer.worldMngr[var19];
-                                                    var23.advanceTime(var23.getWorldTime() + (long)var18);
+                                                    world.advanceTime(world.getWorldTime() + (long)var18);
                                                 }
 
                                                 this.sendNoticeToOps(var4, "Added " + var18 + " to time");
                                             }
                                             else if ("set".equalsIgnoreCase(var20))
                                             {
-                                                for (var19 = 0; var19 < this.minecraftServer.worldMngr.length; ++var19)
+                                                for (World world : DimensionManager.getWorlds())
                                                 {
-                                                    var23 = this.minecraftServer.worldMngr[var19];
-                                                    var23.advanceTime((long)var18);
+                                                    world.advanceTime((long)var18);
                                                 }
 
                                                 this.sendNoticeToOps(var4, "Set time to " + var18);
@@ -414,7 +411,7 @@
                                     }
                                     else if (var2.toLowerCase().startsWith("toggledownfall"))
                                     {
-                                        this.minecraftServer.worldMngr[0].commandToggleDownfall();
+                                        ((WorldServer)DimensionManager.getWorld(0)).commandToggleDownfall();
                                         var3.log("Toggling rain and snow, hold on...");
                                     }
                                     else if (var2.toLowerCase().startsWith("banlist"))
