--- ../src_base/minecraft_server/net/minecraft/src/EnchantmentHelper.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/EnchantmentHelper.java	0000-00-00 00:00:00.000000000 -0000
@@ -357,7 +357,7 @@
         {
             Enchantment var7 = var4[var6];
 
-            if (var7 != null && var7.type.canEnchantItem(var2))
+            if (var7 != null && var7.canEnchantItem(par1ItemStack))
             {
                 for (int var8 = var7.getMinLevel(); var8 <= var7.getMaxLevel(); ++var8)
                 {
