--- ../src_base/minecraft_server/net/minecraft/src/EntityPlayer.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/EntityPlayer.java	0000-00-00 00:00:00.000000000 -0000
@@ -3,6 +3,10 @@
 import java.util.Iterator;
 import java.util.List;
 
+import net.minecraft.src.forge.ArmorProperties;
+import net.minecraft.src.forge.ForgeHooks;
+import net.minecraft.src.forge.ISpecialArmor;
+
 public abstract class EntityPlayer extends EntityLiving
 {
     /** Inventory of the player */
@@ -158,6 +162,7 @@
             }
             else
             {
+                itemInUse.getItem().onUsingItemTick(itemInUse, this, itemInUseCount);
                 if (this.itemInUseCount <= 25 && this.itemInUseCount % 4 == 0)
                 {
                     this.func_35208_b(var1, 5);
@@ -341,6 +346,10 @@
      */
     protected void closeScreen()
     {
+        if (craftingInventory != null) //TODO: This is a bugfix for chests staying open in SMP
+        {
+            craftingInventory.onCraftGuiClosed(this);
+        }
         this.craftingInventory = this.inventorySlots;
     }
 
@@ -514,7 +523,16 @@
 
     public EntityItem func_48347_R()
     {
-        return this.func_48349_a(this.inventory.decrStackSize(this.inventory.currentItem, 1), false);
+        ItemStack stack = inventory.getCurrentItem();
+        if (stack == null)
+        {
+            return null;
+        }
+        if (stack.getItem().onDroppedByPlayer(stack, this))
+        {
+            return func_48349_a(inventory.decrStackSize(inventory.currentItem, 1), false);
+        }
+        return null;
     }
 
     public EntityItem func_48348_b(ItemStack par1ItemStack)
@@ -573,10 +591,18 @@
 
     /**
      * Returns how strong the player is against the specified block at this moment
+     * Deprecated in favor of the metadata-sensitive version
      */
+    @Deprecated
     public float getCurrentPlayerStrVsBlock(Block par1Block)
     {
-        float var2 = this.inventory.getStrVsBlock(par1Block);
+        return getCurrentPlayerStrVsBlock(par1Block, 0);
+    }
+    
+    public float getCurrentPlayerStrVsBlock(Block par1Block, int meta)
+    {
+        ItemStack stack = inventory.getCurrentItem();
+        float var2 = (stack == null ? 1.0F : stack.getItem().getStrVsBlock(stack, par1Block, meta));        
         float var3 = var2;
         int var4 = EnchantmentHelper.getEfficiencyModifier(this.inventory);
 
@@ -868,7 +894,11 @@
             par2 = 1 + par2 >> 1;
         }
 
-        par2 = this.applyArmorCalculations(par1DamageSource, par2);
+        par2 = ArmorProperties.ApplyArmor(this, inventory.armorInventory, par1DamageSource, par2);
+        if (par2 <= 0)
+        {
+            return;
+        }
         par2 = this.applyPotionDamageCalculations(par1DamageSource, par2);
         this.addExhaustion(par1DamageSource.getHungerDamage());
         this.health -= par2;
@@ -926,7 +956,9 @@
      */
     public void destroyCurrentEquippedItem()
     {
+        ItemStack orig = inventory.getCurrentItem();
         this.inventory.setInventorySlotContents(this.inventory.currentItem, (ItemStack)null);
+        ForgeHooks.onDestroyCurrentItem(this, orig);
     }
 
     /**
@@ -955,6 +987,12 @@
      */
     public void attackTargetEntityWithCurrentItem(Entity par1Entity)
     {
+        ItemStack stack = getCurrentEquippedItem();
+        if (stack != null && stack.getItem().onLeftClickEntity(stack, this, par1Entity))
+        {
+            return;
+        }
+        
         if (par1Entity.func_48313_k_())
         {
             int var2 = this.inventory.getDamageVsEntity(par1Entity);
@@ -1093,6 +1131,12 @@
      */
     public EnumStatus sleepInBedAt(int par1, int par2, int par3)
     {
+        EnumStatus customSleep = ForgeHooks.sleepInBedAt(this, par1, par2, par3);
+        if (customSleep != null)
+        {
+            return customSleep;
+        }
+        
         if (!this.worldObj.isRemote)
         {
             if (this.isPlayerSleeping() || !this.isEntityAlive())
@@ -1643,4 +1687,16 @@
     {
         return !this.capabilities.isFlying;
     }
+    
+    /**
+     * Opens a Gui for the player. 
+     * 
+     * @param mod The mod associated with the gui
+     * @param ID The ID number for the Gui
+     * @param world The World
+     * @param X X Position
+     * @param Y Y Position
+     * @param Z Z Position
+     */
+    public void openGui(BaseMod mod, int ID, World world, int x, int y, int z){}
 }
