--- ../src_base/minecraft_server/net/minecraft/src/EntityPlayerMP.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/EntityPlayerMP.java	0000-00-00 00:00:00.000000000 -0000
@@ -2,6 +2,10 @@
 
 import java.util.*;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.src.forge.IGuiHandler;
+import net.minecraft.src.forge.MinecraftForge;
+import net.minecraft.src.forge.NetworkMod;
+import net.minecraft.src.forge.packets.PacketOpenGUI;
 
 public class EntityPlayerMP extends EntityPlayer
     implements ICrafting
@@ -405,6 +409,7 @@
     public void displayWorkbenchGUI(int i, int j, int k)
     {
         getNextWidowId();
+        closeCraftingGui();
         playerNetServerHandler.sendPacket(new Packet100OpenWindow(currentWindowId, 1, "Crafting", 9));
         craftingInventory = new ContainerWorkbench(inventory, worldObj, i, j, k);
         craftingInventory.windowId = currentWindowId;
@@ -414,6 +419,7 @@
     public void displayGUIEnchantment(int i, int j, int k)
     {
         getNextWidowId();
+        closeCraftingGui();
         playerNetServerHandler.sendPacket(new Packet100OpenWindow(currentWindowId, 4, "Enchanting", 9));
         craftingInventory = new ContainerEnchantment(inventory, worldObj, i, j, k);
         craftingInventory.windowId = currentWindowId;
@@ -423,6 +429,7 @@
     public void displayGUIChest(IInventory iinventory)
     {
         getNextWidowId();
+        closeCraftingGui();
         playerNetServerHandler.sendPacket(new Packet100OpenWindow(currentWindowId, 0, iinventory.getInvName(), iinventory.getSizeInventory()));
         craftingInventory = new ContainerChest(inventory, iinventory);
         craftingInventory.windowId = currentWindowId;
@@ -432,6 +439,7 @@
     public void displayGUIFurnace(TileEntityFurnace tileentityfurnace)
     {
         getNextWidowId();
+        closeCraftingGui();
         playerNetServerHandler.sendPacket(new Packet100OpenWindow(currentWindowId, 2, tileentityfurnace.getInvName(), tileentityfurnace.getSizeInventory()));
         craftingInventory = new ContainerFurnace(inventory, tileentityfurnace);
         craftingInventory.windowId = currentWindowId;
@@ -441,6 +449,7 @@
     public void displayGUIDispenser(TileEntityDispenser tileentitydispenser)
     {
         getNextWidowId();
+        closeCraftingGui();
         playerNetServerHandler.sendPacket(new Packet100OpenWindow(currentWindowId, 3, tileentitydispenser.getInvName(), tileentitydispenser.getSizeInventory()));
         craftingInventory = new ContainerDispenser(inventory, tileentitydispenser);
         craftingInventory.windowId = currentWindowId;
@@ -450,6 +459,7 @@
     public void displayGUIBrewingStand(TileEntityBrewingStand tileentitybrewingstand)
     {
         getNextWidowId();
+        closeCraftingGui();
         playerNetServerHandler.sendPacket(new Packet100OpenWindow(currentWindowId, 5, tileentitybrewingstand.getInvName(), tileentitybrewingstand.getSizeInventory()));
         craftingInventory = new ContainerBrewingStand(inventory, tileentitybrewingstand);
         craftingInventory.windowId = currentWindowId;
@@ -613,4 +623,38 @@
         EntityTracker entitytracker = mcServer.getEntityTracker(dimension);
         entitytracker.sendPacketToTrackedPlayersAndTrackedEntity(this, new Packet18Animation(entity, 7));
     }
+    
+    /**
+     * Opens a Gui for the player. 
+     * 
+     * @param mod The mod associated with the gui
+     * @param ID The ID number for the Gui
+     * @param world The World
+     * @param X X Position
+     * @param Y Y Position
+     * @param Z Z Position
+     */
+    @Override
+    public void openGui(BaseMod mod, int ID, World world, int X, int Y, int Z)
+    {
+        if (!(mod instanceof NetworkMod))
+        {
+            return;
+        }
+        IGuiHandler handler = MinecraftForge.getGuiHandler(mod);
+        if (handler != null)
+        {
+            Container container = handler.getGuiContainer(ID, this, world, X, Y, Z);
+            if (container != null)
+            {
+                getNextWidowId();
+                closeCraftingGui();
+                PacketOpenGUI pkt = new PacketOpenGUI(currentWindowId, MinecraftForge.getModID((NetworkMod)mod), ID, X, Y, Z);
+                playerNetServerHandler.sendPacket(pkt.getPacket());
+                craftingInventory = container; 
+                craftingInventory.windowId = currentWindowId;
+                craftingInventory.onCraftGuiOpened(this);
+            }
+        }
+    }
 }
