--- ../src_base/minecraft_server/net/minecraft/src/MLProp.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/MLProp.java	0000-00-00 00:00:00.000000000 -0000
@@ -1,22 +1,13 @@
-// Decompiled by Jad v1.5.8g. Copyright 2001 Pavel Kouznetsov.
-// Jad home page: http://www.kpdus.com/jad.html
-// Decompiler options: packimports(3) braces deadcode fieldsfirst 
-
 package net.minecraft.src;
+import java.lang.annotation.*;
 
-import java.lang.annotation.Annotation;
-
-public interface MLProp
-    extends Annotation
-{
-
-    public abstract String name();
-
-    public abstract String info();
-
-    public abstract double min();
-
-    public abstract double max();
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.FIELD)
+public @interface MLProp {
+	String name() default "";
+	String info() default "";
+	double min() default Double.NEGATIVE_INFINITY;
+	double max() default Double.POSITIVE_INFINITY;
 
-    public abstract int reobf();
+	public int reobf() default 0;
 }
