--- ../src_base/minecraft_server/net/minecraft/src/NetServerHandler.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/NetServerHandler.java	0000-00-00 00:00:00.000000000 -0000
@@ -8,6 +8,7 @@
 import cpw.mods.fml.server.FMLServerHandler;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.src.forge.ForgeHooks;
+import net.minecraft.src.forge.ForgeHooksServer;
 import net.minecraft.src.forge.MessageManager;
 import java.io.UnsupportedEncodingException;
 
@@ -357,7 +358,7 @@
         {
             boolean var3 = var2.disableSpawnProtection = var2.worldProvider.worldType != 0 || this.mcServer.configManager.isOp(this.playerEntity.username);
             boolean var4 = false;
-
+            
             if (par1Packet14BlockDig.status == 0)
             {
                 var4 = true;
@@ -371,7 +372,7 @@
             int var5 = par1Packet14BlockDig.xPosition;
             int var6 = par1Packet14BlockDig.yPosition;
             int var7 = par1Packet14BlockDig.zPosition;
-
+            
             if (var4)
             {
                 double var8 = this.playerEntity.posX - ((double)var5 + 0.5D);
@@ -401,10 +402,14 @@
             {
                 var20 = var9;
             }
-
+            
             if (par1Packet14BlockDig.status == 0)
             {
-                if (var20 <= mcServer.spawnProtectionSize && !var3)
+            	if (!ForgeHooksServer.canMineBlock(this.playerEntity, var5, var6, var7))
+            	{
+            		this.playerEntity.playerNetServerHandler.sendPacket(new Packet53BlockChange(var5, var6, var7, var2));
+            	}
+            	else if (var20 <= mcServer.spawnProtectionSize && !var3)
                 {
                     this.playerEntity.playerNetServerHandler.sendPacket(new Packet53BlockChange(var5, var6, var7, var2));
                 }
@@ -449,7 +454,7 @@
         int var7 = par1Packet15Place.zPosition;
         int var8 = par1Packet15Place.direction;
         boolean var9 = var2.disableSpawnProtection = var2.worldProvider.worldType != 0 || this.mcServer.configManager.isOp(this.playerEntity.username);
-
+        if (!ForgeHooksServer.canPlaceBlock(this.playerEntity, var5, var6, var7)){var4 = true;} 
         if (par1Packet15Place.direction == 255)
         {
             if (var3 == null)
