--- ../src_base/minecraft_server/net/minecraft/src/Packet230ModLoader.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft_server/net/minecraft/src/Packet230ModLoader.java	0000-00-00 00:00:00.000000000 -0000
@@ -93,10 +93,7 @@
 
                             byte[] var8 = new byte[var7];
 
-                            for (int var9 = 0; var9 < var7; var9 += var1.read(var8, var9, var7 - var9))
-                            {
-                                ;
-                            }
+                            var1.readFully(var8);
 
                             this.dataString[var6] = new String(var8);
                         }
