--- ../src_base/minecraft_server/net/minecraft/src/TileEntity.java	0000-00-00 00:00:00.000000000 -00
+++ ../src_work/minecraft_server/net/minecraft/src/TileEntity.java	0000-00-00 00:00:00.000000000 -00
@@ -61,6 +61,13 @@
     {
     }
 
+    /* FORGE: Return false if this TileEntity does not require updateEntity()
+     * calls.
+     */
+    public boolean canUpdate() {
+	    return true;
+    }
+
     public static TileEntity createAndLoadEntity(NBTTagCompound nbttagcompound)
     {
         TileEntity tileentity = null;
