--- a/net/minecraft/client/renderer/block/model/BlockElementFace.java
+++ b/net/minecraft/client/renderer/block/model/BlockElementFace.java
@@ -13,9 +_,18 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public record BlockElementFace(@Nullable Direction cullForDirection, int tintIndex, String texture, BlockFaceUV uv) {
+public record BlockElementFace(@Nullable Direction cullForDirection, int tintIndex, String texture, BlockFaceUV uv, @Nullable net.minecraftforge.client.model.ForgeFaceData data) {
     public static final int NO_TINT = -1;
 
+    public BlockElementFace(@Nullable Direction cullForDirection, int tintIndex, String texture, BlockFaceUV uv) {
+        this(cullForDirection, tintIndex, texture, uv, null);
+    }
+
+    @Override
+    public net.minecraftforge.client.model.ForgeFaceData data() {
+        return this.data == null ? net.minecraftforge.client.model.ForgeFaceData.DEFAULT : this.data;
+    }
+
     @OnlyIn(Dist.CLIENT)
     public static class Deserializer implements JsonDeserializer<BlockElementFace> {
         private static final int DEFAULT_TINT_INDEX = -1;
@@ -26,7 +_,7 @@
             int i = this.getTintIndex(jsonobject);
             String s = this.getTexture(jsonobject);
             BlockFaceUV blockfaceuv = p_111367_.deserialize(jsonobject, BlockFaceUV.class);
-            return new BlockElementFace(direction, i, s, blockfaceuv);
+            return new BlockElementFace(direction, i, s, blockfaceuv, net.minecraftforge.client.model.ForgeFaceData.read(jsonobject.get("forge_data"), null));
         }
 
         protected int getTintIndex(JsonObject p_111369_) {
