--- a/net/minecraft/client/renderer/block/model/BlockElementFace.java
+++ b/net/minecraft/client/renderer/block/model/BlockElementFace.java
@@ -19,12 +_,30 @@
    public final int f_111355_;
    public final String f_111356_;
    public final BlockFaceUV f_111357_;
+   @Nullable
+   private final net.minecraftforge.client.model.ForgeFaceData faceData; // If null, we defer to the parent BlockElement's ForgeFaceData, which is not nullable.
+   @Nullable
+   BlockElement parent; // Parent canot be set by the constructor due to instantiation ordering. This shouldn't really ever be null, but it could theoretically be.
 
    public BlockElementFace(@Nullable Direction p_111359_, int p_111360_, String p_111361_, BlockFaceUV p_111362_) {
+      this(p_111359_, p_111360_, p_111361_, p_111362_, null);
+   }
+
+   public BlockElementFace(@Nullable Direction p_111359_, int p_111360_, String p_111361_, BlockFaceUV p_111362_, @Nullable net.minecraftforge.client.model.ForgeFaceData faceData) {
       this.f_111354_ = p_111359_;
       this.f_111355_ = p_111360_;
       this.f_111356_ = p_111361_;
       this.f_111357_ = p_111362_;
+      this.faceData = faceData;
+   }
+
+   public net.minecraftforge.client.model.ForgeFaceData getFaceData() {
+      if(this.faceData != null) {
+         return this.faceData;
+      } else if(this.parent != null) {
+         return this.parent.getFaceData();
+      }
+      return net.minecraftforge.client.model.ForgeFaceData.DEFAULT;
    }
 
    @OnlyIn(Dist.CLIENT)
@@ -37,7 +_,8 @@
          int i = this.m_111368_(jsonobject);
          String s = this.m_111370_(jsonobject);
          BlockFaceUV blockfaceuv = p_111367_.deserialize(jsonobject, BlockFaceUV.class);
-         return new BlockElementFace(direction, i, s, blockfaceuv);
+         var faceData = net.minecraftforge.client.model.ForgeFaceData.read(jsonobject.get("forge_data"), null);
+         return new BlockElementFace(direction, i, s, blockfaceuv, faceData);
       }
 
       protected int m_111368_(JsonObject p_111369_) {
