--- a/net/minecraft/client/renderer/block/model/IBakedModel.java
+++ b/net/minecraft/client/renderer/block/model/IBakedModel.java
@@ -21,7 +21,18 @@
 
    TextureAtlasSprite getParticleTexture();
 
-   ItemCameraTransforms getItemCameraTransforms();
+   @Deprecated
+   default ItemCameraTransforms getItemCameraTransforms() { return ItemCameraTransforms.DEFAULT; }
 
    ItemOverrideList getOverrides();
+   
+   default boolean isAmbientOcclusion(IBlockState state) { return isAmbientOcclusion(); }
+   
+   /*
+    * Returns the pair of the model for the given perspective, and the matrix
+    * that should be applied to the GL state before rendering it (matrix may be null).
+    */
+   default org.apache.commons.lang3.tuple.Pair<? extends IBakedModel, javax.vecmath.Matrix4f> handlePerspective(ItemCameraTransforms.TransformType cameraTransformType) {
+       return net.minecraftforge.client.ForgeHooksClient.handlePerspective(this, cameraTransformType);
 }
+}
