--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/IBakedModel.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/IBakedModel.java
@@ -21,7 +21,21 @@
 
     TextureAtlasSprite func_177554_e();
 
-    ItemCameraTransforms func_177552_f();
+    @Deprecated
+    default ItemCameraTransforms func_177552_f() { return ItemCameraTransforms.field_178357_a; }
 
     ItemOverrideList func_188617_f();
+
+    default List<BakedQuad> getQuads(@Nullable IBlockState state, @Nullable EnumFacing side, long rand, @Nullable net.minecraft.util.BlockRenderLayer layer)
+    {
+        return func_188616_a(state, side, rand);
+    }
+
+    /*
+     * Returns the pair of the model for the given perspective, and the matrix
+     * that should be applied to the GL state before rendering it (matrix may be null).
+     */
+    default org.apache.commons.lang3.tuple.Pair<? extends IBakedModel, javax.vecmath.Matrix4f> handlePerspective(ItemCameraTransforms.TransformType cameraTransformType) {
+        return net.minecraftforge.client.ForgeHooksClient.handlePerspective(this, cameraTransformType);
+    }
 }
