--- a/net/minecraft/client/renderer/block/model/ItemModelGenerator.java
+++ b/net/minecraft/client/renderer/block/model/ItemModelGenerator.java
@@ -39,6 +_,8 @@
         map.put("particle", p_111672_.hasTexture("particle") ? Either.left(p_111672_.getMaterial("particle")) : map.get("layer0"));
         BlockModel blockmodel = new BlockModel(null, list, map, false, p_111672_.getGuiLight(), p_111672_.getTransforms(), p_111672_.getOverrides());
         blockmodel.name = p_111672_.name;
+        blockmodel.customData.copyFrom(p_111672_.customData);
+        blockmodel.customData.setGui3d(false);
         return blockmodel;
     }
 
