--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/ModelBlockDefinition.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/ModelBlockDefinition.java
@@ -34,7 +34,7 @@
 
     public static ModelBlockDefinition func_178331_a(Reader p_178331_0_)
     {
-        return (ModelBlockDefinition)field_178333_a.fromJson(p_178331_0_, ModelBlockDefinition.class);
+        return net.minecraftforge.client.model.BlockStateLoader.load(p_178331_0_, field_178333_a);
     }
 
     public ModelBlockDefinition(Map<String, VariantList> p_i46572_1_, Multipart p_i46572_2_)
