--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/ModelManager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/ModelManager.java
@@ -9,7 +9,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class ModelManager implements IResourceManagerReloadListener
+public class ModelManager implements net.minecraftforge.client.resource.ISensitiveResourceReloadListener
 {
     private IRegistry<ModelResourceLocation, IBakedModel> field_174958_a;
     private final TextureMap field_174956_b;
@@ -22,11 +22,13 @@
         this.field_174957_c = new BlockModelShapes(this);
     }
 
-    public void func_110549_a(IResourceManager p_110549_1_)
+    public void onResourceManagerReload(IResourceManager resourceManager, java.util.function.Predicate<net.minecraftforge.client.resource.IResourceType> resourcePredicate)
     {
-        ModelBakery modelbakery = new ModelBakery(p_110549_1_, this.field_174956_b, this.field_174957_c);
+        if (!resourcePredicate.test(net.minecraftforge.client.resource.VanillaResourceType.MODELS)) return;
+        net.minecraftforge.client.model.ModelLoader modelbakery = new net.minecraftforge.client.model.ModelLoader(resourceManager, this.field_174956_b, this.field_174957_c);
         this.field_174958_a = modelbakery.func_177570_a();
         this.field_174955_d = this.field_174958_a.func_82594_a(ModelBakery.field_177604_a);
+        net.minecraftforge.client.ForgeHooksClient.onModelBake(this, this.field_174958_a, modelbakery);
         this.field_174957_c.func_178124_c();
     }
 
