--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/ModelBlockDefinition.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/ModelBlockDefinition.java
@@ -32,11 +32,16 @@
     private final Map<String, VariantList> field_178332_b = Maps.<String, VariantList>newHashMap();
     private Multipart field_188005_c;
 
+    @Deprecated
     public static ModelBlockDefinition func_178331_a(Reader p_178331_0_)
     {
-        return (ModelBlockDefinition)JsonUtils.func_193839_a(field_178333_a, p_178331_0_, ModelBlockDefinition.class);
+        return parseFromReader(p_178331_0_, null);
     }
 
+    public static ModelBlockDefinition parseFromReader(Reader reader, net.minecraft.util.ResourceLocation location) {
+        return net.minecraftforge.client.model.BlockStateLoader.load(reader, location, field_178333_a);
+    }
+
     public ModelBlockDefinition(Map<String, VariantList> p_i46572_1_, Multipart p_i46572_2_)
     {
         this.field_188005_c = p_i46572_2_;
