--- ../src-base/minecraft/net/minecraft/client/gui/inventory/GuiContainer.java
+++ ../src-work/minecraft/net/minecraft/client/gui/inventory/GuiContainer.java
@@ -175,8 +175,11 @@
         GlStateManager.func_179109_b(0.0F, 0.0F, 32.0F);
         this.field_73735_i = 200.0F;
         this.field_146296_j.field_77023_b = 200.0F;
+        net.minecraft.client.gui.FontRenderer font = null;
+        if (p_146982_1_ != null) font = p_146982_1_.func_77973_b().getFontRenderer(p_146982_1_);
+        if (font == null) font = field_146289_q;
         this.field_146296_j.func_180450_b(p_146982_1_, p_146982_2_, p_146982_3_);
-        this.field_146296_j.func_180453_a(this.field_146289_q, p_146982_1_, p_146982_2_, p_146982_3_ - (this.field_147012_x == null ? 0 : 8), p_146982_4_);
+        this.field_146296_j.func_180453_a(font, p_146982_1_, p_146982_2_, p_146982_3_ - (this.field_147012_x == null ? 0 : 8), p_146982_4_);
         this.field_73735_i = 0.0F;
         this.field_146296_j.field_77023_b = 0.0F;
     }
@@ -239,13 +242,12 @@
 
         if (itemstack == null)
         {
-            String s1 = p_146977_1_.func_178171_c();
+            TextureAtlasSprite textureatlassprite = p_146977_1_.getBackgroundSprite();
 
-            if (s1 != null)
+            if (textureatlassprite != null)
             {
-                TextureAtlasSprite textureatlassprite = this.field_146297_k.func_147117_R().func_110572_b(s1);
                 GlStateManager.func_179140_f();
-                this.field_146297_k.func_110434_K().func_110577_a(TextureMap.field_110575_b);
+                this.field_146297_k.func_110434_K().func_110577_a(p_146977_1_.getBackgroundLocation());
                 this.func_175175_a(i, j, textureatlassprite, 16, 16);
                 GlStateManager.func_179145_e();
                 flag1 = true;
@@ -326,6 +328,7 @@
             int j = this.field_147003_i;
             int k = this.field_147009_r;
             boolean flag1 = p_73864_1_ < j || p_73864_2_ < k || p_73864_1_ >= j + this.field_146999_f || p_73864_2_ >= k + this.field_147000_g;
+            if (slot != null) flag1 = false; // Forge, prevent dropping of items through slots outside of GUI boundaries
             int l = -1;
 
             if (slot != null)
@@ -463,10 +466,12 @@
 
     protected void func_146286_b(int p_146286_1_, int p_146286_2_, int p_146286_3_)
     {
+        super.func_146286_b(p_146286_1_, p_146286_2_, p_146286_3_); //Forge, Call parent to release buttons
         Slot slot = this.func_146975_c(p_146286_1_, p_146286_2_);
         int i = this.field_147003_i;
         int j = this.field_147009_r;
         boolean flag = p_146286_1_ < i || p_146286_2_ < j || p_146286_1_ >= i + this.field_146999_f || p_146286_2_ >= j + this.field_147000_g;
+        if (slot != null) flag = false; // Forge, prevent dropping of items through slots outside of GUI boundaries
         int k = -1;
 
         if (slot != null)
@@ -685,4 +690,16 @@
             this.field_146297_k.field_71439_g.func_71053_j();
         }
     }
+
+    /* ======================================== FORGE START =====================================*/
+
+    /**
+     * Returns the slot that is currently displayed under the mouse.
+     */
+    public Slot getSlotUnderMouse()
+    {
+        return this.field_147006_u;
+    }
+
+    /* ======================================== FORGE END   =====================================*/
 }
