--- ../src_base/minecraft/net/minecraft/client/gui/inventory/GuiContainer.java
+++ ../src_work/minecraft/net/minecraft/client/gui/inventory/GuiContainer.java
@@ -333,7 +333,7 @@
             if (k >= 0)
             {
                 GL11.glDisable(GL11.GL_LIGHTING);
-                this.mc.renderEngine.bindTexture(this.mc.renderEngine.getTexture("/gui/items.png"));
+                this.mc.renderEngine.bindTexture(this.mc.renderEngine.getTexture(par1Slot.getBackgroundIconTexture()));
                 this.drawTexturedModalRect(i, j, k % 16 * 16, k / 16 * 16, 16, 16);
                 GL11.glEnable(GL11.GL_LIGHTING);
                 flag = true;
