--- a/net/minecraft/world/gen/blockplacer/BlockPlacerType.java
+++ b/net/minecraft/world/gen/blockplacer/BlockPlacerType.java
@@ -3,7 +_,7 @@
 import com.mojang.serialization.Codec;
 import net.minecraft.util.registry.Registry;
 
-public class BlockPlacerType<P extends BlockPlacer> {
+public class BlockPlacerType<P extends BlockPlacer> extends net.minecraftforge.registries.ForgeRegistryEntry<BlockPlacerType<?>> {
    public static final BlockPlacerType<SimpleBlockPlacer> field_227259_a_ = func_236438_a_("simple_block_placer", SimpleBlockPlacer.field_236446_b_);
    public static final BlockPlacerType<DoublePlantBlockPlacer> field_227260_b_ = func_236438_a_("double_plant_placer", DoublePlantBlockPlacer.field_236443_b_);
    public static final BlockPlacerType<ColumnBlockPlacer> field_227261_c_ = func_236438_a_("column_placer", ColumnBlockPlacer.field_236439_b_);
