--- a/net/minecraft/server/level/ThreadedLevelLightEngine.java
+++ b/net/minecraft/server/level/ThreadedLevelLightEngine.java
@@ -171,6 +_,7 @@
         }, () -> "lightChunk " + chunkpos + " " + p_9355_));
         return CompletableFuture.supplyAsync(() -> {
             p_9354_.setLightCorrect(true);
+            net.minecraftforge.common.ForgeHooks.fireLightingCalculatedEvent(p_9354_);
             return p_9354_;
         }, p_280982_ -> this.addTask(chunkpos.x, chunkpos.z, ThreadedLevelLightEngine.TaskType.POST_UPDATE, p_280982_));
     }
