--- a/net/minecraft/server/level/Ticket.java
+++ b/net/minecraft/server/level/Ticket.java
@@ -6,12 +_,18 @@
     private final TicketType<T> type;
     private final int ticketLevel;
     private final T key;
+    private final boolean forceTicks;
     private long createdTick;
 
     protected Ticket(TicketType<T> p_9425_, int p_9426_, T p_9427_) {
+        this(p_9425_, p_9426_, p_9427_, false);
+    }
+
+    public Ticket(TicketType<T> p_9425_, int p_9426_, T p_9427_, boolean forceTicks) {
         this.type = p_9425_;
         this.ticketLevel = p_9426_;
         this.key = p_9427_;
+        this.forceTicks = forceTicks;
     }
 
     public int compareTo(Ticket<?> p_9432_) {
@@ -31,18 +_,18 @@
         } else {
             return !(p_9439_ instanceof Ticket<?> ticket)
                 ? false
-                : this.ticketLevel == ticket.ticketLevel && Objects.equals(this.type, ticket.type) && Objects.equals(this.key, ticket.key);
+                : this.ticketLevel == ticket.ticketLevel && Objects.equals(this.type, ticket.type) && Objects.equals(this.key, ticket.key) && this.forceTicks == ticket.forceTicks;
         }
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(this.type, this.ticketLevel, this.key);
+        return Objects.hash(this.type, this.ticketLevel, this.key, this.forceTicks);
     }
 
     @Override
     public String toString() {
-        return "Ticket[" + this.type + " " + this.ticketLevel + " (" + this.key + ")] at " + this.createdTick;
+        return "Ticket[" + this.type + " " + this.ticketLevel + " (" + this.key + ")] at " + this.createdTick + " force ticks " + forceTicks;
     }
 
     public TicketType<T> getType() {
@@ -51,6 +_,10 @@
 
     public int getTicketLevel() {
         return this.ticketLevel;
+    }
+
+    public boolean isForceTicks() {
+        return this.forceTicks;
     }
 
     protected void setCreatedTick(long p_9430_) {
