--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -341,6 +_,7 @@
 
     @Override
     public boolean addFreshEntity(final Entity entity) {
+        if (entity instanceof net.minecraft.world.entity.Mob mob && mob.isSpawnCancelled()) return false;
         int xc = SectionPos.blockToSectionCoord(entity.getBlockX());
         int zc = SectionPos.blockToSectionCoord(entity.getBlockZ());
         this.getChunk(xc, zc).addEntity(entity);
