--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -297,6 +_,7 @@
    }
 
    public boolean m_7967_(Entity p_9580_) {
+      if (p_9580_ instanceof net.minecraft.world.entity.Mob mob && mob.isSpawnCancelled()) return false;
       int i = SectionPos.m_123171_(p_9580_.m_146903_());
       int j = SectionPos.m_123171_(p_9580_.m_146907_());
       this.m_6325_(i, j).m_6286_(p_9580_);
