--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -16,7 +_,7 @@
 import net.minecraft.world.entity.MobType;
 import net.minecraft.world.item.ItemStack;
 
-public abstract class Enchantment {
+public abstract class Enchantment implements net.minecraftforge.common.extensions.IForgeEnchantment {
    private final EquipmentSlot[] f_44671_;
    private final Enchantment.Rarity f_44674_;
    public final EnchantmentCategory f_44672_;
@@ -71,6 +_,7 @@
       return 0;
    }
 
+   @Deprecated // Forge: Use ItemStack aware version in IForgeEnchantment
    public float m_7335_(int p_44682_, MobType p_44683_) {
       return 0.0F;
    }
@@ -111,7 +_,7 @@
    }
 
    public boolean m_6081_(ItemStack p_44689_) {
-      return this.f_44672_.m_7454_(p_44689_.m_41720_());
+      return canApplyAtEnchantingTable(p_44689_);
    }
 
    public void m_7677_(LivingEntity p_44686_, Entity p_44687_, int p_44688_) {
@@ -133,6 +_,24 @@
    }
 
    public boolean m_6592_() {
+      return true;
+   }
+
+   /**
+    * This applies specifically to applying at the enchanting table. The other method {@link #canEnchant(ItemStack)}
+    * applies for <i>all possible</i> enchantments.
+    * @param stack
+    * @return
+    */
+   public boolean canApplyAtEnchantingTable(ItemStack stack) {
+      return stack.canApplyAtEnchantingTable(this);
+   }
+
+   /**
+    * Is this enchantment allowed to be enchanted on books via Enchantment Table
+    * @return false to disable the vanilla feature
+    */
+   public boolean isAllowedOnBooks() {
       return true;
    }
 
