--- a/net/minecraft/world/level/storage/loot/providers/number/LootNumberProviderType.java
+++ b/net/minecraft/world/level/storage/loot/providers/number/LootNumberProviderType.java
@@ -3,7 +_,7 @@
 import net.minecraft.world.level.storage.loot.Serializer;
 import net.minecraft.world.level.storage.loot.SerializerType;
 
-public class LootNumberProviderType extends SerializerType<NumberProvider> {
+public class LootNumberProviderType extends net.minecraftforge.registries.ForgeSerializerTypeRegistryEntry<LootNumberProviderType, NumberProvider> {
    public LootNumberProviderType(Serializer<? extends NumberProvider> p_165728_) {
       super(p_165728_);
    }
