--- a/net/minecraft/world/level/storage/loot/providers/number/NumberProviders.java
+++ b/net/minecraft/world/level/storage/loot/providers/number/NumberProviders.java
@@ -5,6 +_,7 @@
 import net.minecraft.world.level.storage.loot.GsonAdapterFactory;
 import net.minecraft.world.level.storage.loot.Serializer;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class NumberProviders {
    public static final LootNumberProviderType f_165731_ = m_165738_("constant", new ConstantValue.Serializer());
    public static final LootNumberProviderType f_165732_ = m_165738_("uniform", new UniformGenerator.Serializer());
