--- a/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
+++ b/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
@@ -5,8 +_,8 @@
 import net.minecraft.world.level.block.Blocks;
 
 public class CampfireCookingRecipe extends AbstractCookingRecipe {
-   public CampfireCookingRecipe(ResourceLocation p_43822_, String p_43823_, Ingredient p_43824_, ItemStack p_43825_, float p_43826_, int p_43827_) {
-      super(RecipeType.f_44111_, p_43822_, p_43823_, p_43824_, p_43825_, p_43826_, p_43827_);
+   public CampfireCookingRecipe(ResourceLocation p_43822_, String p_43823_, Ingredient p_43824_, ItemStack p_43825_, float p_43826_, int p_43827_, ResourceLocation category) {
+      super(RecipeType.f_44111_, p_43822_, p_43823_, p_43824_, p_43825_, p_43826_, p_43827_, category);
    }
 
    public ItemStack m_8042_() {
