--- a/net/minecraft/world/item/crafting/LegacyUpgradeRecipe.java
+++ b/net/minecraft/world/item/crafting/LegacyUpgradeRecipe.java
@@ -72,7 +_,7 @@
 
    public boolean m_142505_() {
       return Stream.of(this.f_265911_, this.f_265910_).anyMatch((p_267290_) -> {
-         return p_267290_.m_43908_().length == 0;
+         return net.minecraftforge.common.ForgeHooks.hasNoElements(p_267290_);
       });
    }
 
