--- a/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -18,12 +_,20 @@
    final String f_44242_;
    final ItemStack f_44243_;
    final NonNullList<Ingredient> f_44244_;
+   private final boolean isSimple;
+   private final String category;
 
    public ShapelessRecipe(ResourceLocation p_44246_, String p_44247_, ItemStack p_44248_, NonNullList<Ingredient> p_44249_) {
+      this(p_44246_, p_44247_, p_44248_, p_44249_, null);
+   }
+
+   public ShapelessRecipe(ResourceLocation p_44246_, String p_44247_, ItemStack p_44248_, NonNullList<Ingredient> p_44249_, String category) {
       this.f_44241_ = p_44246_;
       this.f_44242_ = p_44247_;
       this.f_44243_ = p_44248_;
       this.f_44244_ = p_44249_;
+      this.isSimple = p_44249_.stream().allMatch(Ingredient::isSimple);
+      this.category = category;
    }
 
    public ResourceLocation m_6423_() {
@@ -38,6 +_,10 @@
       return this.f_44242_;
    }
 
+   public String getCategory() {
+      return this.category;
+   }
+
    public ItemStack m_8043_() {
       return this.f_44243_;
    }
@@ -48,17 +_,20 @@
 
    public boolean m_5818_(CraftingContainer p_44262_, Level p_44263_) {
       StackedContents stackedcontents = new StackedContents();
+      java.util.List<ItemStack> inputs = new java.util.ArrayList<>();
       int i = 0;
 
       for(int j = 0; j < p_44262_.m_6643_(); ++j) {
          ItemStack itemstack = p_44262_.m_8020_(j);
          if (!itemstack.m_41619_()) {
             ++i;
+            if (isSimple)
             stackedcontents.m_36468_(itemstack, 1);
+            else inputs.add(itemstack);
          }
       }
 
-      return i == this.f_44244_.size() && stackedcontents.m_36475_(this, (IntList)null);
+      return i == this.f_44244_.size() && (isSimple ? stackedcontents.m_36475_(this, (IntList)null) : net.minecraftforge.common.util.RecipeMatcher.findMatches(inputs,  this.f_44244_) != null);
    }
 
    public ItemStack m_5874_(CraftingContainer p_44260_) {
@@ -70,16 +_,18 @@
    }
 
    public static class Serializer implements RecipeSerializer<ShapelessRecipe> {
+      private static final ResourceLocation NAME = new ResourceLocation("minecraft", "crafting_shapeless");
       public ShapelessRecipe m_6729_(ResourceLocation p_44290_, JsonObject p_44291_) {
          String s = GsonHelper.m_13851_(p_44291_, "group", "");
+         String c = GsonHelper.m_13851_(p_44291_, "forge:category", "");
          NonNullList<Ingredient> nonnulllist = m_44275_(GsonHelper.m_13933_(p_44291_, "ingredients"));
          if (nonnulllist.isEmpty()) {
             throw new JsonParseException("No ingredients for shapeless recipe");
-         } else if (nonnulllist.size() > 9) {
-            throw new JsonParseException("Too many ingredients for shapeless recipe");
+         } else if (nonnulllist.size() > ShapedRecipe.MAX_WIDTH * ShapedRecipe.MAX_HEIGHT) {
+            throw new JsonParseException("Too many ingredients for shapeless recipe. The maximum is " + (ShapedRecipe.MAX_WIDTH * ShapedRecipe.MAX_HEIGHT));
          } else {
             ItemStack itemstack = ShapedRecipe.m_151274_(GsonHelper.m_13930_(p_44291_, "result"));
-            return new ShapelessRecipe(p_44290_, s, itemstack, nonnulllist);
+            return new ShapelessRecipe(p_44290_, s, itemstack, nonnulllist, c);
          }
       }
 
@@ -88,7 +_,7 @@
 
          for(int i = 0; i < p_44276_.size(); ++i) {
             Ingredient ingredient = Ingredient.m_43917_(p_44276_.get(i));
-            if (!ingredient.m_43947_()) {
+            if (true || !ingredient.m_43947_()) { // FORGE: Skip checking if an ingredient is empty during shapeless recipe deserialization to prevent complex ingredients from caching tags too early. Can not be done using a config value due to sync issues.
                nonnulllist.add(ingredient);
             }
          }
@@ -98,6 +_,7 @@
 
       public ShapelessRecipe m_8005_(ResourceLocation p_44293_, FriendlyByteBuf p_44294_) {
          String s = p_44294_.m_130277_();
+         String c = p_44294_.m_130277_();
          int i = p_44294_.m_130242_();
          NonNullList<Ingredient> nonnulllist = NonNullList.m_122780_(i, Ingredient.f_43901_);
 
@@ -106,11 +_,12 @@
          }
 
          ItemStack itemstack = p_44294_.m_130267_();
-         return new ShapelessRecipe(p_44293_, s, itemstack, nonnulllist);
+         return new ShapelessRecipe(p_44293_, s, itemstack, nonnulllist, c);
       }
 
       public void m_6178_(FriendlyByteBuf p_44281_, ShapelessRecipe p_44282_) {
          p_44281_.m_130070_(p_44282_.f_44242_);
+         p_44281_.m_130070_(p_44282_.category);
          p_44281_.m_130130_(p_44282_.f_44244_.size());
 
          for(Ingredient ingredient : p_44282_.f_44244_) {
