--- a/net/minecraft/world/item/crafting/SimpleCookingSerializer.java
+++ b/net/minecraft/world/item/crafting/SimpleCookingSerializer.java
@@ -22,11 +_,17 @@
       CookingBookCategory cookingbookcategory = CookingBookCategory.f_244271_.m_262792_(GsonHelper.m_13851_(p_44348_, "category", (String)null), CookingBookCategory.MISC);
       JsonElement jsonelement = (JsonElement)(GsonHelper.m_13885_(p_44348_, "ingredient") ? GsonHelper.m_13933_(p_44348_, "ingredient") : GsonHelper.m_13930_(p_44348_, "ingredient"));
       Ingredient ingredient = Ingredient.m_288218_(jsonelement, false);
+      //Forge: Check if primitive string to keep vanilla or a object which can contain a count field.
+      if (!p_44348_.has("result")) throw new com.google.gson.JsonSyntaxException("Missing result, expected to find a string or object");
+      ItemStack itemstack;
+      if (p_44348_.get("result").isJsonObject()) itemstack = ShapedRecipe.m_151274_(GsonHelper.m_13930_(p_44348_, "result"));
+      else {
       String s1 = GsonHelper.m_13906_(p_44348_, "result");
       ResourceLocation resourcelocation = new ResourceLocation(s1);
-      ItemStack itemstack = new ItemStack(BuiltInRegistries.f_257033_.m_6612_(resourcelocation).orElseThrow(() -> {
+      itemstack = new ItemStack(BuiltInRegistries.f_257033_.m_6612_(resourcelocation).orElseThrow(() -> {
          return new IllegalStateException("Item: " + s1 + " does not exist");
       }));
+      }
       float f = GsonHelper.m_13820_(p_44348_, "experience", 0.0F);
       int i = GsonHelper.m_13824_(p_44348_, "cookingtime", this.f_44327_);
       return this.f_44328_.m_44352_(p_44347_, s, cookingbookcategory, ingredient, itemstack, f, i);
