--- a/net/minecraft/world/item/crafting/SmokingRecipe.java
+++ b/net/minecraft/world/item/crafting/SmokingRecipe.java
@@ -5,8 +_,8 @@
 import net.minecraft.world.level.block.Blocks;
 
 public class SmokingRecipe extends AbstractCookingRecipe {
-   public SmokingRecipe(ResourceLocation p_44469_, String p_44470_, Ingredient p_44471_, ItemStack p_44472_, float p_44473_, int p_44474_) {
-      super(RecipeType.f_44110_, p_44469_, p_44470_, p_44471_, p_44472_, p_44473_, p_44474_);
+   public SmokingRecipe(ResourceLocation p_44469_, String p_44470_, Ingredient p_44471_, ItemStack p_44472_, float p_44473_, int p_44474_, net.minecraft.client.RecipeBookCategories category) {
+      super(RecipeType.f_44110_, p_44469_, p_44470_, p_44471_, p_44472_, p_44473_, p_44474_, category);
    }
 
    public ItemStack m_8042_() {
