--- a/net/minecraft/world/entity/monster/spider/Spider.java
+++ b/net/minecraft/world/entity/monster/spider/Spider.java
@@ -124,7 +_,10 @@
 
     @Override
     public boolean canBeAffected(final MobEffectInstance newEffect) {
-        return newEffect.is(MobEffects.POISON) ? false : super.canBeAffected(newEffect);
+        if (newEffect.getEffect() == MobEffects.POISON) {
+            return net.minecraftforge.event.ForgeEventFactory.onLivingEffectCanApply(this, newEffect).getResult().isAllowed();
+        }
+        return super.canBeAffected(newEffect);
     }
 
     public boolean isClimbing() {
