--- ../src-base/minecraft/net/minecraft/village/VillageCollection.java
+++ ../src-work/minecraft/net/minecraft/village/VillageCollection.java
@@ -147,7 +147,7 @@
                 {
                     Village village = (Village)iterator.next();
                     int j = (int)village.func_75577_a().func_71569_e(villagedoorinfo.field_75481_a, villagedoorinfo.field_75479_b, villagedoorinfo.field_75480_c);
-                    int k = 32 + village.func_75568_b();
+                    float k = 32f + village.func_75568_b(); //BugFix: Avoid int wrapping
 
                     if (j > k * k)
                     {
