--- a/net/minecraft/client/gui/screens/worldselection/WorldCreationUiState.java
+++ b/net/minecraft/client/gui/screens/worldselection/WorldCreationUiState.java
@@ -226,7 +_,7 @@
    @Nullable
    public PresetEditor getPresetEditor() {
       Holder<WorldPreset> holder = this.getWorldType().preset();
-      return holder != null ? PresetEditor.EDITORS.get(holder.unwrapKey()) : null;
+      return holder != null ? holder.unwrapKey().map(net.minecraftforge.client.PresetEditorManager::get).orElse(null) : null; // FORGE: redirect lookup to expanded map
    }
 
    public List<WorldCreationUiState.WorldTypeEntry> getNormalPresetList() {
