--- a/net/minecraft/client/gui/screens/worldselection/WorldPreset.java
+++ b/net/minecraft/client/gui/screens/worldselection/WorldPreset.java
@@ -93,6 +_,9 @@
    WorldPreset(String p_101519_) {
       this.f_101516_ = new TranslatableComponent("generator." + p_101519_);
    }
+   public WorldPreset(Component displayName) {
+      this.f_101516_ = displayName;
+   }
 
    private static WorldGenSettings m_205493_(RegistryAccess p_205494_, WorldGenSettings p_205495_, Holder<Biome> p_205496_) {
       BiomeSource biomesource = new FixedBiomeSource(p_205496_);
@@ -134,4 +_,8 @@
    public interface PresetEditor {
       Screen m_101642_(CreateWorldScreen p_101643_, WorldGenSettings p_101644_);
    }
+
+   // Forge start
+   // For internal use only, automatically called for all ForgeWorldTypes. Register your ForgeWorldType in the forge registry!
+   public static void registerGenerator(WorldPreset gen) { f_101508_.add(gen); }
 }
