--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -178,7 +_,7 @@
         if (!state.isAir() && explosion.getBlockInteraction() != Explosion.BlockInteraction.TRIGGER_BLOCK) {
             Block block = state.getBlock();
             boolean doDropExperienceHack = explosion.getIndirectSourceEntity() instanceof Player;
-            if (block.dropFromExplosion(explosion)) {
+            if (state.canDropFromExplosion(level, pos, explosion)) {
                 BlockEntity blockEntity = state.hasBlockEntity() ? level.getBlockEntity(pos) : null;
                 LootParams.Builder params = new LootParams.Builder(level)
                     .withParameter(LootContextParams.ORIGIN, Vec3.atCenterOf(pos))
@@ -193,8 +_,7 @@
                 state.getDrops(params).forEach(stack -> onHit.accept(stack, pos));
             }
 
-            level.setBlock(pos, Blocks.AIR.defaultBlockState(), 3);
-            block.wasExploded(level, pos, explosion);
+            state.onBlockExploded(level, pos, explosion);
         }
     }
 
@@ -352,11 +_,14 @@
             return 0.0F;
         }
 
-        int modifier = player.hasCorrectToolForDrops(state) ? 30 : 100;
-        return player.getDestroySpeed(state) / destroySpeed / modifier;
+        int modifier = net.minecraftforge.common.ForgeHooks.isCorrectToolForDrops(state, player) ? 30 : 100;
+        return player.getDestroySpeed(state, pos) / destroySpeed / (float)modifier;
     }
 
     protected void spawnAfterBreak(final BlockState state, final ServerLevel level, final BlockPos pos, final ItemStack tool, final boolean dropExperience) {
+        if (dropExperience) {
+            net.minecraftforge.common.ForgeHooks.dropXpForBlock(state, level, pos, tool);
+        }
     }
 
     protected void attack(final BlockState state, final Level level, final BlockPos pos, final Player player) {
@@ -407,6 +_,8 @@
         return this.isRandomlyTicking;
     }
 
+    /** @deprecated Forge: Use more sensitive version {@link net.minecraftforge.common.extensions.IForgeBlockState#getSoundType(LevelReader, BlockPos, Entity) } */
+    @Deprecated
     protected SoundType getSoundType(final BlockState state) {
         return this.soundType;
     }
@@ -427,6 +_,10 @@
         return this.properties.destroyTime;
     }
 
+    protected boolean isAir(BlockState state) {
+       return ((BlockStateBase)state).isAir;
+    }
+
     public abstract static class BlockStateBase extends StateHolder<Block, BlockState> implements TypedInstance<Block> {
         private static final Direction[] DIRECTIONS = Direction.values();
         private static final VoxelShape[] EMPTY_OCCLUSION_SHAPES = Util.make(new VoxelShape[DIRECTIONS.length], s -> Arrays.fill(s, Shapes.empty()));
@@ -582,12 +_,14 @@
             return this.useShapeForLightOcclusion;
         }
 
+        /** @deprecated Forge: Use {@link BlockState#getLightEmission(BlockGetter, BlockPos)} instead */
+        @Deprecated
         public int getLightEmission() {
             return this.lightEmission;
         }
 
         public boolean isAir() {
-            return this.isAir;
+            return this.getBlock().isAir((BlockState)this);
         }
 
         public boolean ignitedByLava() {
@@ -600,9 +_,11 @@
         }
 
         public MapColor getMapColor(final BlockGetter level, final BlockPos pos) {
-            return this.mapColor;
+            return getBlock().getMapColor(this.asState(), level, pos, this.mapColor);
         }
 
+        /** @deprecated use {@link BlockState#rotate(LevelAccessor, BlockPos, Rotation)} */
+        @Deprecated
         public BlockState rotate(final Rotation rotation) {
             return this.getBlock().rotate(this.asState(), rotation);
         }
@@ -660,6 +_,8 @@
         }
 
         public PushReaction getPistonPushReaction() {
+            PushReaction reaction = getBlock().getPistonPushReaction(asState());
+            if (reaction != null) return reaction;
             return this.pushReaction;
         }
 
@@ -1013,7 +_,7 @@
         private BlockBehaviour.StateArgumentPredicate<EntityType<?>> isValidSpawn = (state, level, pos, entityType) -> state.isFaceSturdy(
                 level, pos, Direction.UP
             )
-            && state.getLightEmission() < 14;
+            && state.getLightEmission(level, pos) < 14;
         private BlockBehaviour.StatePredicate isRedstoneConductor = BlockBehaviour.BlockStateBase::isCollisionShapeFullBlock;
         private BlockBehaviour.StatePredicate isSuffocating = (state, level, pos) -> state.blocksMotion() && state.isCollisionShapeFullBlock(level, pos);
         private BlockBehaviour.StatePredicate isViewBlocking = this.isSuffocating;
