--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -94,9 +_,13 @@
                 ListTag entities = new ListTag();
                 chunk.getEntities().forEach(e -> {
                     TagValueOutput output = TagValueOutput.createWithContext(reporter.forChild(e.problemPath()), e.registryAccess());
+                    try {
                     if (e.save(output)) {
                         CompoundTag result = output.buildResult();
                         entities.add(result);
+                    }
+                    } catch (Exception exception) {
+                        LOGGER.error("An Entity type {} has thrown an exception trying to write state. It will not persist. Report this to the mod author", e.getType(), exception);
                     }
                 });
                 CompoundTag chunkTag = NbtUtils.addCurrentDataVersion(new CompoundTag());
