--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -111,6 +111,7 @@
          LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
       }
 
+      net.minecraftforge.fml.server.ServerModLoader.begin(this);
       LOGGER.info("Loading properties");
       this.settings = new PropertyManager(new File("server.properties"));
       this.eula = new ServerEula(new File("eula.txt"));
@@ -182,6 +183,7 @@
          if (!PreYggdrasilConverter.tryConvert(this.settings)) {
             return false;
          } else {
+            net.minecraftforge.fml.server.ServerModLoader.end();
             this.setPlayerList(new DedicatedPlayerList(this));
             long j = Util.func_211178_c();
             if (this.getFolderName() == null) {
