--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -77,6 +77,7 @@
    public boolean func_71197_b() throws IOException {
       Thread thread = new Thread("Server console handler") {
          public void run() {
+            if (net.minecraftforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
             BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
 
             String s1;
@@ -98,7 +99,9 @@
          field_155771_h.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
       }
 
+      net.minecraftforge.fml.server.ServerModLoader.begin(this);
       field_155771_h.info("Loading properties");
+      this.field_71340_o.func_219033_a(properties -> ServerProperties.func_218985_a(java.nio.file.Paths.get("server.properties")));
       ServerProperties serverproperties = this.field_71340_o.func_219034_a();
       if (this.func_71264_H()) {
          this.func_71189_e("127.0.0.1");
@@ -153,17 +156,20 @@
       if (!PreYggdrasilConverter.func_219587_e(this)) {
          return false;
       } else {
+         net.minecraftforge.fml.server.ServerModLoader.end();
          this.func_184105_a(new DedicatedPlayerList(this, this.field_240767_f_, this.field_240766_e_));
          long i = Util.func_211178_c();
          this.func_71191_d(serverproperties.field_219026_t);
          SkullTileEntity.func_184293_a(this.func_152358_ax());
          SkullTileEntity.func_184294_a(this.func_147130_as());
          PlayerProfileCache.func_187320_a(this.func_71266_T());
+         if (!net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
          field_155771_h.info("Preparing level \"{}\"", (Object)this.func_230542_k__());
          this.func_240800_l__();
          long j = Util.func_211178_c() - i;
          String s = String.format(Locale.ROOT, "%.3fs", (double)j / 1.0E9D);
          field_155771_h.info("Done ({})! For help, type \"help\"", (Object)s);
+         this.field_211151_aa = Util.func_211177_b(); //Forge: Update server time to prevent watchdog/spaming during long load.
          if (serverproperties.field_219027_u != null) {
             this.func_200252_aR().func_223585_a(GameRules.field_223620_w).func_223570_a(serverproperties.field_219027_u, this);
          }
@@ -175,6 +181,7 @@
          }
 
          if (serverproperties.field_219030_x) {
+         net.minecraftforge.fml.server.ServerModLoader.end();
             field_155771_h.info("Starting remote control listener");
             this.field_71339_n = new MainThread(this);
             this.field_71339_n.func_72602_a();
@@ -193,7 +200,8 @@
             ServerInfoMBean.func_233490_a_(this);
          }
 
-         return true;
+         // <3 you Grum for this, saves us ~30 patch files! --^
+         return net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStarting(this);
       }
    }
 
@@ -506,6 +514,11 @@
       return false;
    }
 
+   @Override //Forge: Enable formated text for colors in console.
+   public void func_145747_a(net.minecraft.util.text.ITextComponent message, java.util.UUID p_145747_2_) {
+      field_155771_h.info(message.getString());
+   }
+
    public int func_230512_b_(int p_230512_1_) {
       return this.func_213221_d_().field_241081_R_ * p_230512_1_ / 100;
    }
