--- a/net/minecraft/potion/Effects.java
+++ b/net/minecraft/potion/Effects.java
@@ -9,6 +9,7 @@
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.server.ServerWorld;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Effects {
    public static final Effect field_76424_c = func_220308_a(1, "speed", (new Effect(EffectType.BENEFICIAL, 8171462)).func_220304_a(SharedMonsterAttributes.field_111263_d, "91AEAA56-376B-4498-935B-2F7F68070635", (double)0.2F, AttributeModifier.Operation.MULTIPLY_TOTAL));
    public static final Effect field_76421_d = func_220308_a(2, "slowness", (new Effect(EffectType.HARMFUL, 5926017)).func_220304_a(SharedMonsterAttributes.field_111263_d, "7107DE5E-7CE8-4030-940E-514C1F160890", (double)-0.15F, AttributeModifier.Operation.MULTIPLY_TOTAL));
