--- a/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -45,7 +_,7 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 
-public abstract class AbstractMinecart extends Entity {
+public abstract class AbstractMinecart extends Entity implements net.minecraftforge.common.extensions.IForgeAbstractMinecart {
    private static final EntityDataAccessor<Integer> f_38079_ = SynchedEntityData.m_135353_(AbstractMinecart.class, EntityDataSerializers.f_135028_);
    private static final EntityDataAccessor<Integer> f_38080_ = SynchedEntityData.m_135353_(AbstractMinecart.class, EntityDataSerializers.f_135028_);
    private static final EntityDataAccessor<Float> f_38081_ = SynchedEntityData.m_135353_(AbstractMinecart.class, EntityDataSerializers.f_135029_);
@@ -76,6 +_,7 @@
       p_38135_.put(RailShape.NORTH_WEST, Pair.of(vec3i2, vec3i));
       p_38135_.put(RailShape.NORTH_EAST, Pair.of(vec3i2, vec3i1));
    });
+   private static net.minecraftforge.common.IMinecartCollisionHandler COLLISIONS = null;
    private int f_38070_;
    private double f_38071_;
    private double f_38072_;
@@ -85,6 +_,7 @@
    private double f_38076_;
    private double f_38077_;
    private double f_38078_;
+   private boolean canBePushed = true;
 
    protected AbstractMinecart(EntityType<?> p_38087_, Level p_38088_) {
       super(p_38087_, p_38088_);
@@ -99,6 +_,14 @@
       this.f_19856_ = p_38094_;
    }
 
+   public net.minecraftforge.common.IMinecartCollisionHandler getCollisionHandler() {
+      return COLLISIONS;
+   }
+
+   public static void registerCollisionHandler(@Nullable net.minecraftforge.common.IMinecartCollisionHandler handler) {
+      COLLISIONS = handler;
+   }
+
    public static AbstractMinecart m_38119_(Level p_38120_, double p_38121_, double p_38122_, double p_38123_, AbstractMinecart.Type p_38124_) {
       if (p_38124_ == AbstractMinecart.Type.CHEST) {
          return new MinecartChest(p_38120_, p_38121_, p_38122_, p_38123_);
@@ -133,7 +_,7 @@
    }
 
    public boolean m_6094_() {
-      return true;
+      return canBePushed;
    }
 
    protected Vec3 m_7643_(Direction.Axis p_38132_, BlockUtil.FoundRectangle p_38133_) {
@@ -303,9 +_,9 @@
          BlockPos blockpos = new BlockPos(k, i, j);
          BlockState blockstate = this.m_9236_().m_8055_(blockpos);
          this.f_287786_ = BaseRailBlock.m_49416_(blockstate);
-         if (this.f_287786_) {
+         if (canUseRail() && this.f_287786_) {
             this.m_6401_(blockpos, blockstate);
-            if (blockstate.m_60713_(Blocks.f_50285_)) {
+            if (blockstate.m_60734_() instanceof PoweredRailBlock && ((PoweredRailBlock) blockstate.m_60734_()).isActivatorRail()) {
                this.m_6025_(k, i, j, blockstate.m_61143_(PoweredRailBlock.f_55215_));
             }
          } else {
@@ -330,8 +_,11 @@
          }
 
          this.m_19915_(this.m_146908_(), this.m_146909_());
-         if (this.m_6064_() == AbstractMinecart.Type.RIDEABLE && this.m_20184_().m_165925_() > 0.01D) {
-            List<Entity> list = this.m_9236_().m_6249_(this, this.m_20191_().m_82377_((double)0.2F, 0.0D, (double)0.2F), EntitySelector.m_20421_(this));
+         AABB box;
+         if (getCollisionHandler() != null) box = getCollisionHandler().getMinecartCollisionBox(this);
+         else                               box = this.m_20191_().m_82377_(0.2F, 0.0D, 0.2F);
+         if (canBeRidden() && this.m_20184_().m_165925_() > 0.01D) {
+            List<Entity> list = this.m_9236_().m_6249_(this, box, EntitySelector.m_20421_(this));
             if (!list.isEmpty()) {
                for(int l = 0; l < list.size(); ++l) {
                   Entity entity1 = list.get(l);
@@ -343,7 +_,7 @@
                }
             }
          } else {
-            for(Entity entity : this.m_9236_().m_45933_(this, this.m_20191_().m_82377_((double)0.2F, 0.0D, (double)0.2F))) {
+            for(Entity entity : this.m_9236_().m_45933_(this, box)) {
                if (!this.m_20363_(entity) && entity.m_6094_() && entity instanceof AbstractMinecart) {
                   entity.m_7334_(this);
                }
@@ -368,16 +_,23 @@
    }
 
    protected void m_38163_() {
-      double d0 = this.m_7097_();
+      double d0 = this.m_20096_() ? this.m_7097_() : getMaxSpeedAirLateral();
       Vec3 vec3 = this.m_20184_();
       this.m_20334_(Mth.m_14008_(vec3.f_82479_, -d0, d0), vec3.f_82480_, Mth.m_14008_(vec3.f_82481_, -d0, d0));
       if (this.m_20096_()) {
          this.m_20256_(this.m_20184_().m_82490_(0.5D));
       }
 
+      if (getMaxSpeedAirVertical() > 0 && m_20184_().f_82480_ > getMaxSpeedAirVertical()) {
+          if(Math.abs(m_20184_().f_82479_) < 0.3f && Math.abs(m_20184_().f_82481_) < 0.3f)
+              m_20256_(new Vec3(m_20184_().f_82479_, 0.15f, m_20184_().f_82481_));
+          else
+              m_20256_(new Vec3(m_20184_().f_82479_, getMaxSpeedAirVertical(), m_20184_().f_82481_));
+      }
+
       this.m_6478_(MoverType.SELF, this.m_20184_());
       if (!this.m_20096_()) {
-         this.m_20256_(this.m_20184_().m_82490_(0.95D));
+         this.m_20256_(this.m_20184_().m_82490_(getDragAir()));
       }
 
    }
@@ -391,18 +_,19 @@
       d1 = (double)p_38156_.m_123342_();
       boolean flag = false;
       boolean flag1 = false;
-      if (p_38157_.m_60713_(Blocks.f_50030_)) {
+      BaseRailBlock baserailblock = (BaseRailBlock) p_38157_.m_60734_();
+      if (baserailblock instanceof PoweredRailBlock && !((PoweredRailBlock) baserailblock).isActivatorRail()) {
          flag = p_38157_.m_61143_(PoweredRailBlock.f_55215_);
          flag1 = !flag;
       }
 
-      double d3 = 0.0078125D;
+      double d3 = getSlopeAdjustment();
       if (this.m_20069_()) {
          d3 *= 0.2D;
       }
 
       Vec3 vec31 = this.m_20184_();
-      RailShape railshape = p_38157_.m_61143_(((BaseRailBlock)p_38157_.m_60734_()).m_7978_());
+      RailShape railshape = ((BaseRailBlock)p_38157_.m_60734_()).getRailDirection(p_38157_, this.m_9236_(), p_38156_, this);
       switch (railshape) {
          case ASCENDING_EAST:
             this.m_20256_(vec31.m_82520_(-d3, 0.0D, 0.0D));
@@ -448,7 +_,7 @@
          }
       }
 
-      if (flag1) {
+      if (flag1 && shouldDoRailFunctions()) {
          double d22 = this.m_20184_().m_165924_();
          if (d22 < 0.03D) {
             this.m_20256_(Vec3.f_82478_);
@@ -477,10 +_,7 @@
       d0 = d23 + d4 * d14;
       d2 = d10 + d5 * d14;
       this.m_6034_(d0, d1, d2);
-      double d24 = this.m_20160_() ? 0.75D : 1.0D;
-      double d25 = this.m_7097_();
-      vec31 = this.m_20184_();
-      this.m_6478_(MoverType.SELF, new Vec3(Mth.m_14008_(d24 * vec31.f_82479_, -d25, d25), 0.0D, Mth.m_14008_(d24 * vec31.f_82481_, -d25, d25)));
+      this.moveMinecartOnRail(p_38156_);
       if (vec3i.m_123342_() != 0 && Mth.m_14107_(this.m_20185_()) - p_38156_.m_123341_() == vec3i.m_123341_() && Mth.m_14107_(this.m_20189_()) - p_38156_.m_123343_() == vec3i.m_123343_()) {
          this.m_6034_(this.m_20185_(), this.m_20186_() + (double)vec3i.m_123342_(), this.m_20189_());
       } else if (vec3i1.m_123342_() != 0 && Mth.m_14107_(this.m_20185_()) - p_38156_.m_123341_() == vec3i1.m_123341_() && Mth.m_14107_(this.m_20189_()) - p_38156_.m_123343_() == vec3i1.m_123343_()) {
@@ -508,7 +_,10 @@
          this.m_20334_(d26 * (double)(j - p_38156_.m_123341_()), vec35.f_82480_, d26 * (double)(i - p_38156_.m_123343_()));
       }
 
-      if (flag) {
+      if (shouldDoRailFunctions())
+          baserailblock.onMinecartPass(p_38157_, m_9236_(), p_38156_, this);
+
+      if (flag && shouldDoRailFunctions()) {
          Vec3 vec36 = this.m_20184_();
          double d27 = vec36.m_165924_();
          if (d27 > 0.01D) {
@@ -572,7 +_,7 @@
 
       BlockState blockstate = this.m_9236_().m_8055_(new BlockPos(i, j, k));
       if (BaseRailBlock.m_49416_(blockstate)) {
-         RailShape railshape = blockstate.m_61143_(((BaseRailBlock)blockstate.m_60734_()).m_7978_());
+         RailShape railshape = ((BaseRailBlock)blockstate.m_60734_()).getRailDirection(blockstate, this.m_9236_(), new BlockPos(i, j, k), this);
          p_38098_ = (double)j;
          if (railshape.m_61745_()) {
             p_38098_ = (double)(j + 1);
@@ -611,7 +_,7 @@
 
       BlockState blockstate = this.m_9236_().m_8055_(new BlockPos(i, j, k));
       if (BaseRailBlock.m_49416_(blockstate)) {
-         RailShape railshape = blockstate.m_61143_(((BaseRailBlock)blockstate.m_60734_()).m_7978_());
+         RailShape railshape = ((BaseRailBlock)blockstate.m_60734_()).getRailDirection(blockstate, this.m_9236_(), new BlockPos(i, j, k), this);
          Pair<Vec3i, Vec3i> pair = m_38125_(railshape);
          Vec3i vec3i = pair.getFirst();
          Vec3i vec3i1 = pair.getSecond();
@@ -673,6 +_,10 @@
    }
 
    public void m_7334_(Entity p_38165_) {
+      if (getCollisionHandler() != null) {
+         getCollisionHandler().onEntityCollision(this, p_38165_);
+         return;
+      }
       if (!this.m_9236_().f_46443_) {
          if (!p_38165_.f_19794_ && !this.f_19794_) {
             if (!this.m_20363_(p_38165_)) {
@@ -706,11 +_,11 @@
 
                      Vec3 vec32 = this.m_20184_();
                      Vec3 vec33 = p_38165_.m_20184_();
-                     if (((AbstractMinecart)p_38165_).m_6064_() == AbstractMinecart.Type.FURNACE && this.m_6064_() != AbstractMinecart.Type.FURNACE) {
+                     if (((AbstractMinecart)p_38165_).isPoweredCart() && !this.isPoweredCart()) {
                         this.m_20256_(vec32.m_82542_(0.2D, 1.0D, 0.2D));
                         this.m_5997_(vec33.f_82479_ - d0, 0.0D, vec33.f_82481_ - d1);
                         p_38165_.m_20256_(vec33.m_82542_(0.95D, 1.0D, 0.95D));
-                     } else if (((AbstractMinecart)p_38165_).m_6064_() != AbstractMinecart.Type.FURNACE && this.m_6064_() == AbstractMinecart.Type.FURNACE) {
+                     } else if (!((AbstractMinecart)p_38165_).isPoweredCart() && this.isPoweredCart()) {
                         p_38165_.m_20256_(vec33.m_82542_(0.2D, 1.0D, 0.2D));
                         p_38165_.m_5997_(vec32.f_82479_ + d0, 0.0D, vec32.f_82481_ + d1);
                         this.m_20256_(vec32.m_82542_(0.95D, 1.0D, 0.95D));
@@ -809,6 +_,42 @@
    public void m_38138_(boolean p_38139_) {
       this.m_20088_().m_135381_(f_38084_, p_38139_);
    }
+
+   // Forge Start
+   private boolean canUseRail = true;
+   @Override public boolean canUseRail() { return canUseRail; }
+   @Override public void setCanUseRail(boolean value) { this.canUseRail = value; }
+   private float currentSpeedCapOnRail = getMaxCartSpeedOnRail();
+   @Override public float getCurrentCartSpeedCapOnRail() { return currentSpeedCapOnRail; }
+   @Override public void setCurrentCartSpeedCapOnRail(float value) { currentSpeedCapOnRail = Math.min(value, getMaxCartSpeedOnRail()); }
+   @org.jetbrains.annotations.Nullable private Float maxSpeedAirLateral = null;
+   @Override public float getMaxSpeedAirLateral() { return maxSpeedAirLateral == null ? (float) this.m_7097_() : maxSpeedAirLateral; }
+   @Override public void setMaxSpeedAirLateral(float value) { maxSpeedAirLateral = value; }
+   private float maxSpeedAirVertical = DEFAULT_MAX_SPEED_AIR_VERTICAL;
+   @Override public float getMaxSpeedAirVertical() { return maxSpeedAirVertical; }
+   @Override public void setMaxSpeedAirVertical(float value) { maxSpeedAirVertical = value; }
+   private double dragAir = DEFAULT_AIR_DRAG;
+   @Override public double getDragAir() { return dragAir; }
+   @Override public void setDragAir(double value) { dragAir = value; }
+   @Override
+   public double getMaxSpeedWithRail() { //Non-default because getMaximumSpeed is protected
+      if (!canUseRail()) return m_7097_();
+      BlockPos pos = this.getCurrentRailPosition();
+      BlockState state = this.m_9236_().m_8055_(pos);
+      if (!state.m_204336_(BlockTags.f_13034_)) return m_7097_();
+
+      float railMaxSpeed = ((BaseRailBlock)state.m_60734_()).getRailMaxSpeed(state, this.m_9236_(), pos, this);
+      return Math.min(railMaxSpeed, getCurrentCartSpeedCapOnRail());
+   }
+   @Override
+   public void moveMinecartOnRail(BlockPos pos) { //Non-default because getMaximumSpeed is protected
+      AbstractMinecart mc = this;
+      double d24 = mc.m_20160_() ? 0.75D : 1.0D;
+      double d25 = mc.getMaxSpeedWithRail();
+      Vec3 vec3d1 = mc.m_20184_();
+      mc.m_6478_(MoverType.SELF, new Vec3(Mth.m_14008_(d24 * vec3d1.f_82479_, -d25, d25), 0.0D, Mth.m_14008_(d24 * vec3d1.f_82481_, -d25, d25)));
+   }
+   // Forge end
 
    public ItemStack m_142340_() {
       Item item;
