--- a/net/minecraft/world/entity/vehicle/Boat.java
+++ b/net/minecraft/world/entity/vehicle/Boat.java
@@ -50,7 +_,7 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class Boat extends Entity implements VariantHolder<Boat.Type> {
+public class Boat extends Entity implements net.minecraftforge.common.extensions.IForgeBoat {
    private static final EntityDataAccessor<Integer> f_38282_ = SynchedEntityData.m_135353_(Boat.class, EntityDataSerializers.f_135028_);
    private static final EntityDataAccessor<Integer> f_38283_ = SynchedEntityData.m_135353_(Boat.class, EntityDataSerializers.f_135028_);
    private static final EntityDataAccessor<Float> f_38284_ = SynchedEntityData.m_135353_(Boat.class, EntityDataSerializers.f_135029_);
@@ -455,7 +_,7 @@
             for(int i2 = i1; i2 < j1; ++i2) {
                blockpos$mutableblockpos.m_122178_(l1, k1, i2);
                FluidState fluidstate = this.m_9236_().m_6425_(blockpos$mutableblockpos);
-               if (fluidstate.m_205070_(FluidTags.f_13131_)) {
+               if (this.canBoatInFluid(fluidstate)) {
                   f = Math.max(f, fluidstate.m_76155_(this.m_9236_(), blockpos$mutableblockpos));
                }
 
@@ -496,7 +_,7 @@
                      blockpos$mutableblockpos.m_122178_(l1, k2, i2);
                      BlockState blockstate = this.m_9236_().m_8055_(blockpos$mutableblockpos);
                      if (!(blockstate.m_60734_() instanceof WaterlilyBlock) && Shapes.m_83157_(blockstate.m_60812_(this.m_9236_(), blockpos$mutableblockpos).m_83216_((double)l1, (double)k2, (double)i2), voxelshape, BooleanOp.f_82689_)) {
-                        f += blockstate.m_60734_().m_49958_();
+                        f += blockstate.getFriction(this.m_9236_(), blockpos$mutableblockpos, this);
                         ++k1;
                      }
                   }
@@ -525,7 +_,7 @@
             for(int i2 = i1; i2 < j1; ++i2) {
                blockpos$mutableblockpos.m_122178_(k1, l1, i2);
                FluidState fluidstate = this.m_9236_().m_6425_(blockpos$mutableblockpos);
-               if (fluidstate.m_205070_(FluidTags.f_13131_)) {
+               if (this.canBoatInFluid(fluidstate)) {
                   float f = (float)l1 + fluidstate.m_76155_(this.m_9236_(), blockpos$mutableblockpos);
                   this.f_38277_ = Math.max((double)f, this.f_38277_);
                   flag |= aabb.f_82289_ < (double)f;
@@ -555,7 +_,7 @@
             for(int i2 = i1; i2 < j1; ++i2) {
                blockpos$mutableblockpos.m_122178_(k1, l1, i2);
                FluidState fluidstate = this.m_9236_().m_6425_(blockpos$mutableblockpos);
-               if (fluidstate.m_205070_(FluidTags.f_13131_) && d0 < (double)((float)blockpos$mutableblockpos.m_123342_() + fluidstate.m_76155_(this.m_9236_(), blockpos$mutableblockpos))) {
+               if (this.canBoatInFluid(fluidstate) && d0 < (double)((float)blockpos$mutableblockpos.m_123342_() + fluidstate.m_76155_(this.m_9236_(), blockpos$mutableblockpos))) {
                   if (!fluidstate.m_76170_()) {
                      return Boat.Status.UNDER_FLOWING_WATER;
                   }
@@ -773,7 +_,7 @@
             }
 
             this.m_183634_();
-         } else if (!this.m_9236_().m_6425_(this.m_20183_().m_7495_()).m_205070_(FluidTags.f_13131_) && p_38307_ < 0.0D) {
+         } else if (!this.canBoatInFluid(this.m_9236_().m_6425_(this.m_20183_().m_7495_())) && p_38307_ < 0.0D) {
             this.f_19789_ -= (float)p_38307_;
          }
 
@@ -829,7 +_,7 @@
    }
 
    protected boolean m_7310_(Entity p_38390_) {
-      return this.m_20197_().size() < this.m_213801_() && !this.m_204029_(FluidTags.f_13131_);
+      return this.m_20197_().size() < this.m_213801_() && !this.canBoatInFluid(this.getEyeInFluidType());
    }
 
    protected int m_213801_() {
@@ -858,6 +_,16 @@
 
    public boolean m_5842_() {
       return this.f_38279_ == Boat.Status.UNDER_WATER || this.f_38279_ == Boat.Status.UNDER_FLOWING_WATER;
+   }
+
+   // Forge: Fix MC-119811 by instantly completing lerp on board
+   @Override
+   protected void m_20348_(Entity passenger) {
+      super.m_20348_(passenger);
+      if (this.m_6109_() && this.f_38267_ > 0) {
+         this.f_38267_ = 0;
+         this.m_19890_(this.f_38268_, this.f_38269_, this.f_38270_, (float)this.f_38271_, (float)this.f_38272_);
+      }
    }
 
    public ItemStack m_142340_() {
