--- a/net/minecraft/world/entity/vehicle/MinecartFurnace.java
+++ b/net/minecraft/world/entity/vehicle/MinecartFurnace.java
@@ -110,6 +_,8 @@
    }
 
    public InteractionResult interact(Player p_38562_, InteractionHand p_38563_) {
+      InteractionResult ret = super.interact(p_38562_, p_38563_);
+      if (ret.consumesAction()) return ret;
       ItemStack itemstack = p_38562_.getItemInHand(p_38563_);
       if (INGREDIENT.test(itemstack) && this.fuel + 3600 <= 32000) {
          if (!p_38562_.getAbilities().instabuild) {
@@ -125,6 +_,11 @@
       }
 
       return InteractionResult.sidedSuccess(this.level().isClientSide);
+   }
+
+   @Override
+   public float getMaxCartSpeedOnRail() {
+      return 0.2f;
    }
 
    protected void addAdditionalSaveData(CompoundTag p_38567_) {
