--- a/net/minecraft/world/entity/vehicle/Minecart.java
+++ b/net/minecraft/world/entity/vehicle/Minecart.java
@@ -19,6 +_,8 @@
 
     @Override
     public InteractionResult interact(Player p_38483_, InteractionHand p_38484_) {
+        InteractionResult ret = super.interact(p_38483_, p_38484_);
+        if (ret.consumesAction()) return ret;
         if (p_38483_.isSecondaryUseActive()) {
             return InteractionResult.PASS;
         } else if (this.isVehicle()) {
