--- a/net/minecraft/entity/monster/AbstractSkeletonEntity.java
+++ b/net/minecraft/entity/monster/AbstractSkeletonEntity.java
@@ -154,7 +_,7 @@
       if (this.level != null && !this.level.isClientSide) {
          this.goalSelector.removeGoal(this.meleeGoal);
          this.goalSelector.removeGoal(this.bowGoal);
-         ItemStack itemstack = this.getItemInHand(ProjectileHelper.getWeaponHoldingHand(this, Items.BOW));
+         ItemStack itemstack = this.getItemInHand(ProjectileHelper.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.item.BowItem));
          if (itemstack.getItem() == Items.BOW) {
             int i = 20;
             if (this.level.getDifficulty() != Difficulty.HARD) {
@@ -171,8 +_,10 @@
    }
 
    public void performRangedAttack(LivingEntity p_82196_1_, float p_82196_2_) {
-      ItemStack itemstack = this.getProjectile(this.getItemInHand(ProjectileHelper.getWeaponHoldingHand(this, Items.BOW)));
+      ItemStack itemstack = this.getProjectile(this.getItemInHand(ProjectileHelper.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.item.BowItem)));
       AbstractArrowEntity abstractarrowentity = this.getArrow(itemstack, p_82196_2_);
+      if (this.getMainHandItem().getItem() instanceof net.minecraft.item.BowItem)
+         abstractarrowentity = ((net.minecraft.item.BowItem)this.getMainHandItem().getItem()).customArrow(abstractarrowentity);
       double d0 = p_82196_1_.getX() - this.getX();
       double d1 = p_82196_1_.getY(0.3333333333333333D) - abstractarrowentity.getY();
       double d2 = p_82196_1_.getZ() - this.getZ();
