--- a/net/minecraft/entity/monster/EntityCreeper.java
+++ b/net/minecraft/entity/monster/EntityCreeper.java
@@ -211,7 +211,7 @@
 
    private void func_146077_cc() {
       if (!this.field_70170_p.field_72995_K) {
-         boolean flag = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
+         boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this);
          float f = this.func_70830_n() ? 2.0F : 1.0F;
          this.field_70729_aU = true;
          this.field_70170_p.func_72876_a(this, this.field_70165_t, this.field_70163_u, this.field_70161_v, (float)this.field_82226_g * f, flag);
