--- a/net/minecraft/entity/monster/EntityEnderman.java
+++ b/net/minecraft/entity/monster/EntityEnderman.java
@@ -209,7 +209,9 @@
    }
 
    private boolean func_70825_j(double p_70825_1_, double p_70825_3_, double p_70825_5_) {
-      boolean flag = this.func_184595_k(p_70825_1_, p_70825_3_, p_70825_5_);
+      net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(this, p_70825_1_, p_70825_3_, p_70825_5_, 0);
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return false;
+      boolean flag = this.func_184595_k(event.getTargetX(), event.getTargetY(), event.getTargetZ());
       if (flag) {
          this.field_70170_p.func_184148_a((EntityPlayer)null, this.field_70169_q, this.field_70167_r, this.field_70166_s, SoundEvents.field_187534_aX, this.func_184176_by(), 1.0F, 1.0F);
          this.func_184185_a(SoundEvents.field_187534_aX, 1.0F, 1.0F);
@@ -356,7 +358,7 @@
       public boolean func_75250_a() {
          if (this.field_179475_a.func_195405_dq() == null) {
             return false;
-         } else if (!this.field_179475_a.field_70170_p.func_82736_K().func_82766_b("mobGriefing")) {
+         } else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_179475_a.field_70170_p, this.field_179475_a)) {
             return false;
          } else {
             return this.field_179475_a.func_70681_au().nextInt(2000) == 0;
@@ -373,7 +375,7 @@
          IBlockState iblockstate = iworld.func_180495_p(blockpos);
          IBlockState iblockstate1 = iworld.func_180495_p(blockpos.func_177977_b());
          IBlockState iblockstate2 = this.field_179475_a.func_195405_dq();
-         if (iblockstate2 != null && this.func_195924_a(iworld, blockpos, iblockstate2, iblockstate, iblockstate1)) {
+         if (iblockstate2 != null && this.func_195924_a(iworld, blockpos, iblockstate2, iblockstate, iblockstate1) && !net.minecraftforge.event.ForgeEventFactory.onBlockPlace(field_179475_a, new net.minecraftforge.common.util.BlockSnapshot(field_179475_a.field_70170_p, blockpos, iblockstate), net.minecraft.util.EnumFacing.UP)) {
             iworld.func_180501_a(blockpos, iblockstate2, 3);
             this.field_179475_a.func_195406_b((IBlockState)null);
          }
@@ -395,7 +397,7 @@
       public boolean func_75250_a() {
          if (this.field_179473_a.func_195405_dq() != null) {
             return false;
-         } else if (!this.field_179473_a.field_70170_p.func_82736_K().func_82766_b("mobGriefing")) {
+         } else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_179473_a.field_70170_p, this.field_179473_a)) {
             return false;
          } else {
             return this.field_179473_a.func_70681_au().nextInt(20) == 0;
