--- ../src-base/minecraft/net/minecraft/entity/monster/EntityIronGolem.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityIronGolem.java
@@ -294,7 +294,7 @@
     {
         if (!this.func_70850_q() && this.field_70717_bb != null && this.field_70857_d != null)
         {
-            this.field_70857_d.func_82688_a(this.field_70717_bb.func_70005_c_(), -5);
+            this.field_70857_d.modifyPlayerReputation(this.field_70717_bb.func_110124_au(), -5);
         }
 
         super.func_70645_a(p_70645_1_);
