--- a/net/minecraft/entity/monster/EntitySnowman.java
+++ b/net/minecraft/entity/monster/EntitySnowman.java
@@ -31,7 +31,7 @@
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
 
-public class EntitySnowman extends EntityGolem implements IRangedAttackMob {
+public class EntitySnowman extends EntityGolem implements IRangedAttackMob, net.minecraftforge.common.IShearable {
    private static final DataParameter<Byte> field_184749_a = EntityDataManager.func_187226_a(EntitySnowman.class, DataSerializers.field_187191_a);
 
    public EntitySnowman(World p_i1692_1_) {
@@ -85,7 +85,7 @@
             this.func_70097_a(DamageSource.field_76370_b, 1.0F);
          }
 
-         if (!this.field_70170_p.func_82736_K().func_82766_b("mobGriefing")) {
+         if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this)) {
             return;
          }
 
@@ -127,7 +127,7 @@
 
    protected boolean func_184645_a(EntityPlayer p_184645_1_, EnumHand p_184645_2_) {
       ItemStack itemstack = p_184645_1_.func_184586_b(p_184645_2_);
-      if (itemstack.func_77973_b() == Items.field_151097_aZ && this.func_184748_o() && !this.field_70170_p.field_72995_K) {
+      if (false &&  itemstack.func_77973_b() == Items.field_151097_aZ && this.func_184748_o() && !this.field_70170_p.field_72995_K) { //Forge: Moved to onSheared
          this.func_184747_a(false);
          itemstack.func_77972_a(1, p_184645_1_);
       }
@@ -166,4 +166,15 @@
 
    public void func_184724_a(boolean p_184724_1_) {
    }
+
+   @Override
+   public boolean isShearable(ItemStack item, net.minecraft.world.IWorldReader world, BlockPos pos) {
+      return this.func_184748_o();
+   }
+
+   @Override
+   public java.util.List<ItemStack> onSheared(ItemStack item, net.minecraft.world.IWorld world, BlockPos pos, int fortune) {
+      this.func_184747_a(false);
+      return new java.util.ArrayList<>();
+   }
 }
