--- ../src-base/minecraft/net/minecraft/entity/monster/EntityEndermite.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityEndermite.java
@@ -44,6 +44,7 @@
         this.field_70714_bg.func_75776_a(8, new EntityAILookIdle(this));
         this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, true, new Class[0]));
         this.field_70715_bh.func_75776_a(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
+        super.func_184651_r();
     }
 
     public float func_70047_e()
