--- ../src_base/minecraft/net/minecraft/entity/monster/EntitySlime.java
+++ ../src_work/minecraft/net/minecraft/entity/monster/EntitySlime.java
@@ -10,6 +10,7 @@
 import net.minecraft.world.WorldType;
 import net.minecraft.world.biome.BiomeGenBase;
 import net.minecraft.world.chunk.Chunk;
+import net.minecraftforge.common.ForgeHooks;
 
 public class EntitySlime extends EntityLiving implements IMob
 {
@@ -48,7 +49,7 @@
     public int getMaxHealth()
     {
         int var1 = this.getSlimeSize();
-        return var1 * var1;
+        return var1 * var1 * ForgeHooks.getEntityHealth(this.getClass().getName().replaceAll("Entity", ""));
     }
 
     /**
