--- a/net/minecraft/entity/monster/EntitySlime.java
+++ b/net/minecraft/entity/monster/EntitySlime.java
@@ -113,7 +113,7 @@
 
    public void func_70071_h_() {
       if (!this.field_70170_p.field_72995_K && this.field_70170_p.func_175659_aa() == EnumDifficulty.PEACEFUL && this.func_70809_q() > 0) {
-         this.field_70128_L = true;
+         this.func_70106_y(); //Forge: Kill entity with notification to caps/subclasses.
       }
 
       this.field_70811_b += (this.field_70813_a - this.field_70811_b) * 0.5F;
@@ -122,6 +122,7 @@
       if (this.field_70122_E && !this.field_175452_bi) {
          int i = this.func_70809_q();
 
+         if (spawnCustomParticles()) i = 0; // don't spawn particles if it's handled by the implementation itself
          for(int j = 0; j < i * 8; ++j) {
             float f = this.field_70146_Z.nextFloat() * ((float)Math.PI * 2F);
             float f1 = this.field_70146_Z.nextFloat() * 0.5F + 0.5F;
@@ -255,7 +256,7 @@
 
    public boolean func_205020_a(IWorld p_205020_1_, boolean p_205020_2_) {
       BlockPos blockpos = new BlockPos(MathHelper.func_76128_c(this.field_70165_t), 0, MathHelper.func_76128_c(this.field_70161_v));
-      if (p_205020_1_.func_72912_H().func_76067_t() == WorldType.field_77138_c && this.field_70146_Z.nextInt(4) != 1) {
+      if (p_205020_1_.func_72912_H().func_76067_t().handleSlimeSpawnReduction(field_70146_Z, field_70170_p)) {
          return false;
       } else {
          if (p_205020_1_.func_175659_aa() != EnumDifficulty.PEACEFUL) {
@@ -308,6 +309,12 @@
       return this.func_189101_db() ? SoundEvents.field_189110_fE : SoundEvents.field_187882_fq;
    }
 
+   /**
+    * Called when the slime spawns particles on landing, see onUpdate.
+    * Return true to prevent the spawning of the default particles.
+    */
+   protected boolean spawnCustomParticles() { return false; }
+
    static class AISlimeAttack extends EntityAIBase {
       private final EntitySlime field_179466_a;
       private int field_179465_b;
