--- a/net/minecraft/entity/monster/EntityIronGolem.java
+++ b/net/minecraft/entity/monster/EntityIronGolem.java
@@ -233,7 +233,7 @@
 
    public void func_70645_a(DamageSource p_70645_1_) {
       if (!this.func_70850_q() && this.field_70717_bb != null && this.field_70857_d != null) {
-         this.field_70857_d.func_82688_a(this.field_70717_bb.func_146103_bH().getName(), -5);
+         this.field_70857_d.modifyPlayerReputation(this.field_70717_bb.func_146103_bH().getId(), -5);
       }
 
       super.func_70645_a(p_70645_1_);
