--- ../src-base/minecraft/net/minecraft/entity/monster/EntityEnderman.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityEnderman.java
@@ -267,7 +267,9 @@
 
     private boolean func_70825_j(double p_70825_1_, double p_70825_3_, double p_70825_5_)
     {
-        boolean flag = this.func_184595_k(p_70825_1_, p_70825_3_, p_70825_5_);
+        net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(this, p_70825_1_, p_70825_3_, p_70825_5_, 0);
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return false;
+        boolean flag = this.func_184595_k(event.getTargetX(), event.getTargetY(), event.getTargetZ());
 
         if (flag)
         {
@@ -354,6 +356,18 @@
         }
     }
 
+    /*===================================== Forge Start ==============================*/
+    public static void setCarriable(Block block, boolean canCarry)
+    {
+        if (canCarry) field_70827_d.add(block);
+        else          field_70827_d.remove(block);
+    }
+    public static boolean getCarriable(Block block)
+    {
+        return field_70827_d.contains(block);
+    }
+    /*===================================== Forge End ==============================*/
+
     public boolean func_70823_r()
     {
         return ((Boolean)this.field_70180_af.func_187225_a(field_184719_bw)).booleanValue();
@@ -486,7 +500,7 @@
                 {
                     return false;
                 }
-                else if (!this.field_179475_a.field_70170_p.func_82736_K().func_82766_b("mobGriefing"))
+                else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_179475_a.field_70170_p, this.field_179475_a))
                 {
                     return false;
                 }
@@ -508,7 +522,7 @@
                 IBlockState iblockstate1 = world.func_180495_p(blockpos.func_177977_b());
                 IBlockState iblockstate2 = this.field_179475_a.func_175489_ck();
 
-                if (iblockstate2 != null && this.func_188518_a(world, blockpos, iblockstate2.func_177230_c(), iblockstate, iblockstate1))
+                if (iblockstate2 != null && this.func_188518_a(world, blockpos, iblockstate2.func_177230_c(), iblockstate, iblockstate1) && net.minecraftforge.event.ForgeEventFactory.onBlockPlace(field_179475_a, new net.minecraftforge.common.util.BlockSnapshot(world, blockpos, iblockstate2), net.minecraft.util.EnumFacing.UP).isCanceled())
                 {
                     world.func_180501_a(blockpos, iblockstate2, 3);
                     this.field_179475_a.func_175490_a((IBlockState)null);
@@ -551,7 +565,7 @@
                 {
                     return false;
                 }
-                else if (!this.field_179473_a.field_70170_p.func_82736_K().func_82766_b("mobGriefing"))
+                else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_179473_a.field_70170_p, this.field_179473_a))
                 {
                     return false;
                 }
