--- ../src-base/minecraft/net/minecraft/entity/monster/EntityGuardian.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityGuardian.java
@@ -75,6 +75,7 @@
         this.field_175481_bq.func_75248_a(3);
         entityaimovetowardsrestriction.func_75248_a(3);
         this.field_70715_bh.func_75776_a(1, new EntityAINearestAttackableTarget(this, EntityLivingBase.class, 10, true, false, new EntityGuardian.GuardianTargetSelector(this)));
+        super.func_184651_r();
     }
 
     protected void func_110147_ax()
