--- a/net/minecraft/entity/monster/EntityIllusionIllager.java
+++ b/net/minecraft/entity/monster/EntityIllusionIllager.java
@@ -181,6 +181,8 @@
 
    public void func_82196_d(EntityLivingBase p_82196_1_, float p_82196_2_) {
       EntityArrow entityarrow = this.func_193097_t(p_82196_2_);
+      if (this.func_184614_ca().func_77973_b() instanceof net.minecraft.item.ItemBow)
+         entityarrow = ((net.minecraft.item.ItemBow) this.func_184614_ca().func_77973_b()).customizeArrow(entityarrow);
       double d0 = p_82196_1_.field_70165_t - this.field_70165_t;
       double d1 = p_82196_1_.func_174813_aQ().field_72338_b + (double)(p_82196_1_.field_70131_O / 3.0F) - entityarrow.field_70163_u;
       double d2 = p_82196_1_.field_70161_v - this.field_70161_v;
