--- ../src-base/minecraft/net/minecraft/entity/monster/EntityGuardian.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityGuardian.java
@@ -205,7 +205,7 @@
 
     public float func_180484_a(BlockPos p_180484_1_)
     {
-        return this.field_70170_p.func_180495_p(p_180484_1_).func_185904_a() == Material.field_151586_h ? 10.0F + this.field_70170_p.func_175724_o(p_180484_1_) - 0.5F : super.func_180484_a(p_180484_1_);
+        return this.field_70170_p.func_180495_p(p_180484_1_).func_185904_a().isSwimmable() ? 10.0F + this.field_70170_p.func_175724_o(p_180484_1_) - 0.5F : super.func_180484_a(p_180484_1_);
     }
 
     public void func_70636_d()
