--- ../src-base/minecraft/net/minecraft/entity/monster/EntityZombie.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityZombie.java
@@ -41,6 +41,10 @@
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.EnumDifficulty;
 import net.minecraft.world.World;
+import net.minecraftforge.common.ForgeModContainer;
+import cpw.mods.fml.common.eventhandler.Event.Result;
+import net.minecraftforge.event.ForgeEventFactory;
+import net.minecraftforge.event.entity.living.ZombieEvent.SummonAidEvent;
 
 public class EntityZombie extends EntityMob
 {
@@ -78,7 +82,7 @@
         this.func_110148_a(SharedMonsterAttributes.field_111265_b).func_111128_a(40.0D);
         this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.23000000417232513D);
         this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(3.0D);
-        this.func_110140_aT().func_111150_b(field_110186_bp).func_111128_a(this.field_70146_Z.nextDouble() * 0.10000000149011612D);
+        this.func_110140_aT().func_111150_b(field_110186_bp).func_111128_a(this.field_70146_Z.nextDouble() * ForgeModContainer.zombieSummonBaseChance);
     }
 
     protected void func_70088_a()
@@ -233,12 +237,28 @@
                 entitylivingbase = (EntityLivingBase)p_70097_1_.func_76346_g();
             }
 
-            if (entitylivingbase != null && this.field_70170_p.field_73013_u == EnumDifficulty.HARD && (double)this.field_70146_Z.nextFloat() < this.func_110148_a(field_110186_bp).func_111126_e())
+
+            int i = MathHelper.func_76128_c(this.field_70165_t);
+            int j = MathHelper.func_76128_c(this.field_70163_u);
+            int k = MathHelper.func_76128_c(this.field_70161_v);
+
+            SummonAidEvent summonAid = ForgeEventFactory.fireZombieSummonAid(this, field_70170_p, i, j, k, entitylivingbase, this.func_110148_a(field_110186_bp).func_111126_e());
+            
+            if (summonAid.getResult() == Result.DENY)
             {
-                int i = MathHelper.func_76128_c(this.field_70165_t);
-                int j = MathHelper.func_76128_c(this.field_70163_u);
-                int k = MathHelper.func_76128_c(this.field_70161_v);
-                EntityZombie entityzombie = new EntityZombie(this.field_70170_p);
+                return true;
+            }
+            else if (summonAid.getResult() == Result.ALLOW || entitylivingbase != null && this.field_70170_p.field_73013_u == EnumDifficulty.HARD && (double)this.field_70146_Z.nextFloat() < this.func_110148_a(field_110186_bp).func_111126_e())
+            {
+                EntityZombie entityzombie;
+                if (summonAid.customSummonedAid != null && summonAid.getResult() == Result.ALLOW)
+                {
+                    entityzombie = summonAid.customSummonedAid;
+                }
+                else
+                {
+                    entityzombie = new EntityZombie(this.field_70170_p);
+                }
 
                 for (int l = 0; l < 50; ++l)
                 {
@@ -253,7 +273,7 @@
                         if (this.field_70170_p.func_72855_b(entityzombie.field_70121_D) && this.field_70170_p.func_72945_a(entityzombie, entityzombie.field_70121_D).isEmpty() && !this.field_70170_p.func_72953_d(entityzombie.field_70121_D))
                         {
                             this.field_70170_p.func_72838_d(entityzombie);
-                            entityzombie.func_70624_b(entitylivingbase);
+                            if (entitylivingbase != null) entityzombie.func_70624_b(entitylivingbase);
                             entityzombie.func_110161_a((IEntityLivingData)null);
                             this.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Zombie reinforcement caller charge", -0.05000000074505806D, 0));
                             entityzombie.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Zombie reinforcement callee charge", -0.05000000074505806D, 0));
@@ -439,7 +459,7 @@
 
         if (p_110161_1_1 == null)
         {
-            p_110161_1_1 = new EntityZombie.GroupData(this.field_70170_p.field_73012_v.nextFloat() < 0.05F, this.field_70170_p.field_73012_v.nextFloat() < 0.05F, null);
+            p_110161_1_1 = new EntityZombie.GroupData(this.field_70170_p.field_73012_v.nextFloat() < ForgeModContainer.zombieBabyChance, this.field_70170_p.field_73012_v.nextFloat() < 0.05F, null);
         }
 
         if (p_110161_1_1 instanceof EntityZombie.GroupData)
