--- ../src_base/minecraft/net/minecraft/entity/monster/EntityWitch.java
+++ ../src_work/minecraft/net/minecraft/entity/monster/EntityWitch.java
@@ -232,6 +232,8 @@
      */
     protected void dropFewItems(boolean par1, int par2)
     {
+        // CraftBukkit start
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
         int j = this.rand.nextInt(3) + 1;
 
         for (int k = 0; k < j; ++k)
@@ -244,11 +246,11 @@
                 l += this.rand.nextInt(par2 + 1);
             }
 
-            for (int j1 = 0; j1 < l; ++j1)
-            {
-                this.dropItem(i1, 1);
-            }
-        }
+            loot.add(new org.bukkit.inventory.ItemStack(i1, l));
+        }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+        // CraftBukkit end
     }
 
     /**
