--- ../src_base/minecraft/net/minecraft/entity/monster/EntityBlaze.java
+++ ../src_work/minecraft/net/minecraft/entity/monster/EntityBlaze.java
@@ -27,10 +27,10 @@
         this.experienceValue = 10;
     }
 
-    protected void func_110147_ax()
-    {
-        super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(6.0D);
+    protected void applyEntityAttributes()
+    {
+        super.applyEntityAttributes();
+        this.getEntityAttribute(SharedMonsterAttributes.attackDamage).setAttribute(6.0D);
     }
 
     protected void entityInit()
@@ -206,12 +206,17 @@
     {
         if (par1)
         {
+            // CraftBukkit start
+            java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
             int j = this.rand.nextInt(2 + par2);
 
-            for (int k = 0; k < j; ++k)
-            {
-                this.dropItem(Item.blazeRod.itemID, 1);
-            }
+            if (j > 0)
+            {
+                loot.add(new org.bukkit.inventory.ItemStack(Item.blazeRod.itemID, j));
+            }
+
+            org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+            // CraftBukkit end
         }
     }
 
