--- a/net/minecraft/entity/monster/HoglinEntity.java
+++ b/net/minecraft/entity/monster/HoglinEntity.java
@@ -117,7 +_,7 @@
       HoglinTasks.func_234377_a_(this);
       if (this.func_234364_eK_()) {
          ++this.field_234358_by_;
-         if (this.field_234358_by_ > 300) {
+         if (this.field_234358_by_ > 300 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.field_233590_aW_, (timer) -> this.field_234358_by_ = timer)) {
             this.func_241412_a_(SoundEvents.field_232715_fE_);
             this.func_234360_a_((ServerWorld)this.field_70170_p);
          }
@@ -212,6 +_,7 @@
       ZoglinEntity zoglinentity = this.func_233656_b_(EntityType.field_233590_aW_, true);
       if (zoglinentity != null) {
          zoglinentity.func_195064_c(new EffectInstance(Effects.field_76431_k, 200, 0));
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zoglinentity);
       }
 
    }
