--- a/net/minecraft/entity/monster/SpiderEntity.java
+++ b/net/minecraft/entity/monster/SpiderEntity.java
@@ -118,7 +_,12 @@
    }
 
    public boolean canBeAffected(EffectInstance p_70687_1_) {
-      return p_70687_1_.getEffect() == Effects.POISON ? false : super.canBeAffected(p_70687_1_);
+      if (p_70687_1_.getEffect() == Effects.POISON) {
+         net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent event = new net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent(this, p_70687_1_);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         return event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW;
+      }
+      return super.canBeAffected(p_70687_1_);
    }
 
    public boolean isClimbing() {
