--- a/net/minecraft/entity/monster/ZombifiedPiglinEntity.java
+++ b/net/minecraft/entity/monster/ZombifiedPiglinEntity.java
@@ -173,6 +_,7 @@
 
    public void readAdditionalSaveData(CompoundNBT p_70037_1_) {
       super.readAdditionalSaveData(p_70037_1_);
+      if(!level.isClientSide) //FORGE: allow this entity to be read from nbt on client. (Fixes MC-189565)
       this.readPersistentAngerSaveData((ServerWorld)this.level, p_70037_1_);
    }
 
