--- a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
+++ b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
@@ -46,7 +_,7 @@
     }
 
     protected boolean checkExtraStartConditions(final ServerLevel level, final Villager body) {
-        if (!level.getGameRules().get(GameRules.MOB_GRIEFING)) {
+        if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(level, body)) {
             return false;
         }
 
@@ -120,6 +_,11 @@
                             level.setBlockAndUpdate(this.aboveFarmlandPos, place);
                             level.gameEvent(GameEvent.BLOCK_PLACE, this.aboveFarmlandPos, GameEvent.Context.of(body, place));
                             ok = true;
+                        } else if (itemStack.getItem() instanceof net.minecraftforge.common.IPlantable) {
+                            if (((net.minecraftforge.common.IPlantable) itemStack.getItem()).getPlantType(level, aboveFarmlandPos) == net.minecraftforge.common.PlantType.CROP) {
+                                level.setBlock(aboveFarmlandPos, ((net.minecraftforge.common.IPlantable) itemStack.getItem()).getPlant(level, aboveFarmlandPos), 3);
+                                ok = true;
+                            }
                         }
 
                         if (ok) {
