--- a/net/minecraft/world/entity/ai/behavior/Swim.java
+++ b/net/minecraft/world/entity/ai/behavior/Swim.java
@@ -14,7 +_,7 @@
     }
 
     public static boolean shouldSwim(Mob p_327994_) {
-        return p_327994_.isInWater() && p_327994_.getFluidHeight(FluidTags.WATER) > p_327994_.getFluidJumpThreshold() || p_327994_.isInLava();
+        return p_327994_.isInWater() &&p_327994_.getFluidHeight(FluidTags.WATER) > p_327994_.getFluidJumpThreshold() || p_327994_.isInLava() || p_327994_.isInFluidType((fluidType, height) -> p_327994_.canSwimInFluidType(fluidType) && height > p_327994_.getFluidJumpThreshold());
     }
 
     protected boolean checkExtraStartConditions(ServerLevel p_24388_, Mob p_24389_) {
