--- a/net/minecraft/client/renderer/chunk/ChunkRenderDispatcher.java
+++ b/net/minecraft/client/renderer/chunk/ChunkRenderDispatcher.java
@@ -45,10 +45,14 @@
    private Vec3d field_217672_l = Vec3d.field_186680_a;
 
    public ChunkRenderDispatcher(boolean p_i51518_1_) {
+      this(p_i51518_1_, -1);
+   }
+
+   public ChunkRenderDispatcher(boolean p_i51518_1_, int countRenderBuilders) {
       int i = Math.max(1, (int)((double)Runtime.getRuntime().maxMemory() * 0.3D) / 10485760 - 1);
       int j = Runtime.getRuntime().availableProcessors();
       int k = p_i51518_1_ ? j : Math.min(j, 4);
-      int l = Math.max(1, Math.min(k * 2, i));
+      int l = countRenderBuilders < 0 ? Math.max(1, Math.min(k * 2, i)) : countRenderBuilders;
       this.field_178525_i = new ChunkRenderWorker(this, new RegionRenderCacheBuilder());
       List<RegionRenderCacheBuilder> list = Lists.newArrayListWithExpectedSize(l);
 
