--- a/net/minecraft/client/renderer/chunk/SectionCompiler.java
+++ b/net/minecraft/client/renderer/chunk/SectionCompiler.java
@@ -56,6 +_,7 @@
     public SectionCompiler.Results compile(
         final SectionPos sectionPos, final RenderSectionRegion region, final VertexSorting vertexSorting, final SectionBufferBuilderPack builders
     ) {
+        var modelDataMap = net.minecraft.client.Minecraft.getInstance().level.getModelDataManager().getAt(sectionPos);
         SectionCompiler.Results results = new SectionCompiler.Results();
         BlockPos minPos = sectionPos.origin();
         BlockPos maxPos = minPos.offset(15, 15, 15);
@@ -95,6 +_,9 @@
                     }
 
                     if (blockState.getRenderShape() == RenderShape.MODEL) {
+                        var model = this.blockModelSet.get(blockState);
+                        var modelData = modelDataMap.getOrDefault(pos, net.minecraftforge.client.model.data.ModelData.EMPTY);
+                        modelData = model.getModelData(region, pos, blockState, modelData);
                         blockRenderer.tesselateBlock(
                             ModelBlockRenderer.forceOpaque(this.cutoutLeaves, blockState) ? opaqueQuadOutput : quadOutput,
                             SectionPos.sectionRelative(pos.getX()),
@@ -103,8 +_,9 @@
                             region,
                             pos,
                             blockState,
-                            this.blockModelSet.get(blockState),
-                            blockState.getSeed(pos)
+                            model,
+                            blockState.getSeed(pos),
+                            modelData
                         );
                     }
                 } catch (Throwable t) {
