--- a/net/minecraft/client/renderer/model/IModelTransform.java
+++ b/net/minecraft/client/renderer/model/IModelTransform.java
@@ -5,7 +_,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface IModelTransform {
+public interface IModelTransform extends net.minecraftforge.client.extensions.IForgeModelTransform {
    default TransformationMatrix getRotation() {
       return TransformationMatrix.identity();
    }
