--- a/net/minecraft/client/renderer/model/ItemTransformVec3f.java
+++ b/net/minecraft/client/renderer/model/ItemTransformVec3f.java
@@ -14,7 +14,9 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
+/** @deprecated Use {@link net.minecraft.util.math.vector.TransformationMatrix} through {@link net.minecraftforge.client.extensions.IForgeBakedModel#handlePerspective */
 @OnlyIn(Dist.CLIENT)
+@Deprecated
 public class ItemTransformVec3f {
    public static final ItemTransformVec3f field_178366_a = new ItemTransformVec3f(new Vector3f(), new Vector3f(), new Vector3f(1.0F, 1.0F, 1.0F));
    public final Vector3f field_178364_b;
