--- a/net/minecraft/client/renderer/model/ItemCameraTransforms.java
+++ b/net/minecraft/client/renderer/model/ItemCameraTransforms.java
@@ -25,6 +25,7 @@
       this(ItemTransformVec3f.field_178366_a, ItemTransformVec3f.field_178366_a, ItemTransformVec3f.field_178366_a, ItemTransformVec3f.field_178366_a, ItemTransformVec3f.field_178366_a, ItemTransformVec3f.field_178366_a, ItemTransformVec3f.field_178366_a, ItemTransformVec3f.field_178366_a);
    }
 
+   /**@deprecated Modders shouldn't use this. This is handled by Forge.*/@Deprecated
    public ItemCameraTransforms(ItemCameraTransforms p_i46443_1_) {
       this.field_188036_k = p_i46443_1_.field_188036_k;
       this.field_188037_l = p_i46443_1_.field_188037_l;
@@ -36,6 +37,7 @@
       this.field_181700_p = p_i46443_1_.field_181700_p;
    }
 
+   /**@deprecated Modders shouldn't use this. This is handled by Forge.*/@Deprecated
    public ItemCameraTransforms(ItemTransformVec3f p_i46569_1_, ItemTransformVec3f p_i46569_2_, ItemTransformVec3f p_i46569_3_, ItemTransformVec3f p_i46569_4_, ItemTransformVec3f p_i46569_5_, ItemTransformVec3f p_i46569_6_, ItemTransformVec3f p_i46569_7_, ItemTransformVec3f p_i46569_8_) {
       this.field_188036_k = p_i46569_1_;
       this.field_188037_l = p_i46569_2_;
@@ -47,6 +49,7 @@
       this.field_181700_p = p_i46569_8_;
    }
 
+   /**@deprecated Modders shouldn't use this. This is handled by Forge.*/@Deprecated
    public ItemTransformVec3f func_181688_b(ItemCameraTransforms.TransformType p_181688_1_) {
       switch(p_181688_1_) {
       case THIRD_PERSON_LEFT_HAND:
