--- a/net/minecraft/client/renderer/model/Variant.java
+++ b/net/minecraft/client/renderer/model/Variant.java
@@ -12,7 +12,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public class Variant {
+public class Variant implements net.minecraftforge.client.model.ISmartVariant {
    private final ResourceLocation field_188050_a;
    private final ModelRotation field_188051_b;
    private final boolean field_188052_c;
@@ -29,10 +29,15 @@
       return this.field_188050_a;
    }
 
+   @Deprecated
    public ModelRotation func_188048_b() {
       return this.field_188051_b;
    }
 
+   public net.minecraftforge.common.model.IModelState getState() {
+      return this.field_188051_b;
+   }
+
    public boolean func_188049_c() {
       return this.field_188052_c;
    }
