--- a/net/minecraft/world/biome/provider/BiomeProvider.java
+++ b/net/minecraft/world/biome/provider/BiomeProvider.java
@@ -36,6 +36,10 @@
       this.field_226837_c_ = p_i231634_1_;
    }
 
+   public com.mojang.serialization.MapCodec<BiomeProvider> dispatch(net.minecraft.util.RegistryKey<net.minecraft.world.Dimension> key, net.minecraft.util.registry.WorldSettingsImport<com.google.gson.JsonElement> parser) {
+      return com.mojang.serialization.MapCodec.unit(this);
+   }
+
    protected abstract Codec<? extends BiomeProvider> func_230319_a_();
 
    @OnlyIn(Dist.CLIENT)
