--- a/net/minecraft/world/biome/provider/NetherBiomeProvider.java
+++ b/net/minecraft/world/biome/provider/NetherBiomeProvider.java
@@ -92,6 +92,20 @@
       this.field_235269_l_ = false;
    }
 
+   @Override
+   public MapCodec<BiomeProvider> dispatch(net.minecraft.util.RegistryKey<net.minecraft.world.Dimension> key, net.minecraft.util.registry.WorldSettingsImport<com.google.gson.JsonElement> parser) {
+      return RecordCodecBuilder.<NetherBiomeProvider>mapCodec(inst ->
+              inst.group(
+                      MapCodec.unit(this.field_235270_m_).forGetter(nbp -> nbp.field_235270_m_),
+                      net.minecraftforge.common.world.WorldGenAdditions.complexBiomeAdditions(this.field_235268_k_, key, parser).forGetter(nbp -> nbp.field_235268_k_),
+                      MapCodec.unit(this.field_242597_h).forGetter(nbp -> nbp.field_242597_h),
+                      MapCodec.unit(this.field_242598_i).forGetter(nbp -> nbp.field_242598_i),
+                      MapCodec.unit(this.field_242599_j).forGetter(nbp -> nbp.field_242599_j),
+                      MapCodec.unit(this.field_242600_k).forGetter(nbp -> nbp.field_242600_k)
+              ).apply(inst, NetherBiomeProvider::new)
+      ).xmap(bp -> bp, bp -> (NetherBiomeProvider) bp);
+   }
+
    protected Codec<? extends BiomeProvider> func_230319_a_() {
       return field_235263_f_;
    }
