--- ../src-base/minecraft/net/minecraft/client/renderer/tileentity/RenderItemFrame.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/tileentity/RenderItemFrame.java
@@ -99,15 +99,18 @@
             GlStateManager.disableLighting();
             int i = itemFrame.getRotation();
 
-            if (item == Items.filled_map)
+            if (item instanceof net.minecraft.item.ItemMap)
             {
                 i = i % 4 * 2;
             }
 
             GlStateManager.rotate((float)i * 360.0F / 8.0F, 0.0F, 0.0F, 1.0F);
 
-            if (item == Items.filled_map)
+            net.minecraftforge.client.event.RenderItemInFrameEvent event = new net.minecraftforge.client.event.RenderItemInFrameEvent(itemFrame, this);
+            if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event))
             {
+            if (item instanceof net.minecraft.item.ItemMap)
+            {
                 this.renderManager.renderEngine.bindTexture(mapBackgroundTextures);
                 GlStateManager.rotate(180.0F, 0.0F, 0.0F, 1.0F);
                 float f = 0.0078125F;
@@ -165,6 +168,7 @@
                     textureatlassprite.updateAnimation();
                 }
             }
+            }
 
             GlStateManager.enableLighting();
             GlStateManager.popMatrix();
