--- a/net/minecraft/world/entity/monster/cubemob/AbstractCubeMob.java
+++ b/net/minecraft/world/entity/monster/cubemob/AbstractCubeMob.java
@@ -125,6 +_,8 @@
             float size = this.getDimensions(this.getPose()).width() * 2.0F;
             float radius = size / 2.0F;
 
+            // Forge: Don't spawn particles if it's handled by the implementation itself
+            if (!spawnCustomParticles())
             for (int i = 0; i < size * 16.0F; i++) {
                 float dir = this.random.nextFloat() * (float) (Math.PI * 2);
                 float d = this.random.nextFloat() * 0.5F + 0.5F;
@@ -145,6 +_,12 @@
         this.wasOnGround = this.onGround();
         this.decreaseSquish();
     }
+
+    /**
+     * Called when the slime spawns particles on landing, see onUpdate.
+     * Return true to prevent the spawning of the default particles.
+     */
+    protected boolean spawnCustomParticles() { return false; }
 
     protected void decreaseSquish() {
         this.targetSquish *= 0.6F;
