--- a/net/minecraft/client/resources/ClientLanguageMap.java
+++ b/net/minecraft/client/resources/ClientLanguageMap.java
@@ -78,4 +78,9 @@
    public IReorderingProcessor func_241870_a(ITextProperties p_241870_1_) {
       return BidiReorderer.func_243508_a(p_241870_1_, this.field_239496_d_);
    }
+
+   @Override
+   public Map<String, String> getLanguageData() {
+      return field_239495_c_;
+   }
 }
