--- ../src-base/minecraft/net/minecraft/client/resources/FoliageColorReloadListener.java
+++ ../src-work/minecraft/net/minecraft/client/resources/FoliageColorReloadListener.java
@@ -8,12 +8,13 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class FoliageColorReloadListener implements IResourceManagerReloadListener
+public class FoliageColorReloadListener implements net.minecraftforge.client.resource.ISensitiveResourceReloadListener
 {
     private static final ResourceLocation field_130079_a = new ResourceLocation("textures/colormap/foliage.png");
 
-    public void func_110549_a(IResourceManager p_110549_1_)
+    public void onResourceManagerReload(IResourceManager p_110549_1_, net.minecraftforge.client.resource.ReloadRequirements requirements)
     {
+        if (!requirements.shouldReload(net.minecraftforge.client.resource.VanillaResourceType.TEXTURES)) return;
         try
         {
             ColorizerFoliage.func_77467_a(TextureUtil.func_110986_a(p_110549_1_, field_130079_a));
