--- ../src-base/minecraft/net/minecraft/client/resources/IResourceManagerReloadListener.java
+++ ../src-work/minecraft/net/minecraft/client/resources/IResourceManagerReloadListener.java
@@ -3,6 +3,11 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+/**
+ * @deprecated in favor of {@link net.minecraftforge.client.resource.ISensitiveResourceReloadListener}, which is
+ * sensitive to individual resource types being requested for reload.
+ */
+@Deprecated
 @SideOnly(Side.CLIENT)
 public interface IResourceManagerReloadListener
 {
