--- ../src-base/minecraft/net/minecraft/client/resources/LanguageManager.java
+++ ../src-work/minecraft/net/minecraft/client/resources/LanguageManager.java
@@ -16,7 +16,7 @@
 import org.apache.logging.log4j.Logger;
 
 @SideOnly(Side.CLIENT)
-public class LanguageManager implements IResourceManagerReloadListener
+public class LanguageManager implements net.minecraftforge.client.resource.ISensitiveResourceReloadListener
 {
     private static final Logger field_147648_b = LogManager.getLogger();
     private final MetadataSerializer field_135047_b;
@@ -63,8 +63,9 @@
         }
     }
 
-    public void func_110549_a(IResourceManager p_110549_1_)
+    public void onResourceManagerReload(IResourceManager resourceManager, net.minecraftforge.client.resource.ReloadRequirements requirements)
     {
+        if (!requirements.shouldReload(net.minecraftforge.client.resource.VanillaResourceType.LANGUAGES)) return;
         List<String> list = Lists.newArrayList("en_us");
 
         if (!"en_us".equals(this.field_135048_c))
@@ -72,7 +73,7 @@
             list.add(this.field_135048_c);
         }
 
-        field_135049_a.func_135022_a(p_110549_1_, list);
+        field_135049_a.func_135022_a(resourceManager, list);
         LanguageMap.func_135063_a(field_135049_a.field_135032_a);
     }
 
