--- ../src-base/minecraft/net/minecraft/client/resources/Locale.java
+++ ../src-work/minecraft/net/minecraft/client/resources/Locale.java
@@ -95,6 +95,8 @@
 
     private void func_135021_a(InputStream p_135021_1_) throws IOException
     {
+        p_135021_1_ = net.minecraftforge.fml.common.FMLCommonHandler.instance().loadLanguage(field_135032_a, p_135021_1_);
+        if (p_135021_1_ == null) return;
         for (String s : IOUtils.readLines(p_135021_1_, Charsets.UTF_8))
         {
             if (!s.isEmpty() && s.charAt(0) != 35)
