--- ../src-base/minecraft/net/minecraft/client/resources/SkinManager.java
+++ ../src-work/minecraft/net/minecraft/client/resources/SkinManager.java
@@ -129,7 +129,9 @@
 
                 if (hashmap.isEmpty() && p_152790_1_.getId().equals(Minecraft.func_71410_x().func_110432_I().func_148256_e().getId()))
                 {
-                    hashmap.putAll(SkinManager.this.field_152797_e.getTextures(SkinManager.this.field_152797_e.fillProfileProperties(p_152790_1_, false), false));
+                    // FORGE: Use already filled profile from session rather
+                    // than getting rate limited by filling the input profile
+                    hashmap.putAll(SkinManager.this.field_152797_e.getTextures(Minecraft.func_71410_x().func_110432_I().func_148256_e(), false));
                 }
 
                 Minecraft.func_71410_x().func_152344_a(new Runnable()
