--- a/net/minecraft/client/gui/screens/inventory/EffectRenderingInventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/EffectRenderingInventoryScreen.java
@@ -42,12 +_,17 @@
       if (!collection.isEmpty() && j >= 32) {
          RenderSystem.m_157429_(1.0F, 1.0F, 1.0F, 1.0F);
          boolean flag = j >= 120;
+         var event = net.minecraftforge.client.ForgeHooksClient.onScreenPotionSize(this, j, !flag, i);
+         if (event.isCanceled()) return;
+         flag = !event.isCompact();
+         i = event.getHorizontalOffset();
          int k = 33;
          if (collection.size() > 5) {
             k = 132 / (collection.size() - 1);
          }
 
-         Iterable<MobEffectInstance> iterable = Ordering.natural().sortedCopy(collection);
+
+         Iterable<MobEffectInstance> iterable = collection.stream().filter(net.minecraftforge.client.ForgeHooksClient::shouldRenderEffect).sorted().collect(java.util.stream.Collectors.toList());
          this.m_194002_(p_194015_, i, k, iterable, flag);
          this.m_194008_(p_194015_, i, k, iterable, flag);
          if (flag) {
@@ -95,6 +_,11 @@
       int i = this.f_97736_;
 
       for(MobEffectInstance mobeffectinstance : p_194012_) {
+         var renderer = net.minecraftforge.client.extensions.common.IClientMobEffectExtensions.of(mobeffectinstance);
+         if (renderer.renderInventoryIcon(mobeffectinstance, this, p_194009_, p_194010_ + (p_194013_ ? 6 : 7), i, this.m_93252_())) {
+            i += p_194011_;
+            continue;
+         }
          MobEffect mobeffect = mobeffectinstance.m_19544_();
          TextureAtlasSprite textureatlassprite = mobeffecttexturemanager.m_118732_(mobeffect);
          RenderSystem.m_157456_(0, textureatlassprite.m_118414_().m_118330_());
@@ -108,6 +_,11 @@
       int i = this.f_97736_;
 
       for(MobEffectInstance mobeffectinstance : p_98726_) {
+         var renderer = net.minecraftforge.client.extensions.common.IClientMobEffectExtensions.of(mobeffectinstance);
+         if (renderer.renderInventoryText(mobeffectinstance, this, p_98723_, p_98724_, i, this.m_93252_())) {
+            i += p_98725_;
+            continue;
+         }
          Component component = this.m_194000_(mobeffectinstance);
          this.f_96547_.m_92763_(p_98723_, component, (float)(p_98724_ + 10 + 18), (float)(i + 6), 16777215);
          String s = MobEffectUtil.m_19581_(mobeffectinstance, 1.0F);
