--- ../src-base/minecraft/net/minecraft/client/renderer/tileentity/TileEntityItemStackRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/tileentity/TileEntityItemStackRenderer.java
@@ -39,6 +39,7 @@
     private final TileEntityBed field_193843_g = new TileEntityBed();
     private final TileEntitySkull field_179023_f = new TileEntitySkull();
     private final ModelShield field_187318_g = new ModelShield();
+    public static final java.util.Map<net.minecraftforge.registries.IRegistryDelegate<Item>, TileEntityItemStackRenderer> ITEM_TO_RENDERER = new java.util.IdentityHashMap<>();
 
     public void func_179022_a(ItemStack p_179022_1_)
     {
@@ -118,6 +119,7 @@
         {
             TileEntityRendererDispatcher.field_147556_a.func_192855_a(field_191274_b[BlockShulkerBox.func_190955_b(item).func_176765_a()], 0.0D, 0.0D, 0.0D, 0.0F, p_192838_2_);
         }
+        else if (Block.func_149634_a(item) != Blocks.field_150486_ae) net.minecraftforge.client.ForgeHooksClient.renderTileItem(p_192838_1_.func_77973_b(), p_192838_1_.func_77960_j());
         else
         {
             TileEntityRendererDispatcher.field_147556_a.func_192855_a(this.field_147717_b, 0.0D, 0.0D, 0.0D, 0.0F, p_192838_2_);
