--- ../src-base/minecraft/net/minecraft/client/renderer/tileentity/TileEntityRendererDispatcher.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/tileentity/TileEntityRendererDispatcher.java
@@ -92,7 +92,7 @@
     @Nullable
     public <T extends TileEntity> TileEntitySpecialRenderer<T> func_147547_b(@Nullable TileEntity p_147547_1_)
     {
-        return p_147547_1_ == null ? null : this.func_147546_a(p_147547_1_.getClass());
+        return p_147547_1_ == null || p_147547_1_.func_145837_r() ? null : this.func_147546_a(p_147547_1_.getClass()); // Forge: fix MC-123363
     }
 
     public void func_190056_a(World p_190056_1_, TextureManager p_190056_2_, FontRenderer p_190056_3_, Entity p_190056_4_, RayTraceResult p_190056_5_, float p_190056_6_)
@@ -117,12 +117,15 @@
     {
         if (p_180546_1_.func_145835_a(this.field_147560_j, this.field_147561_k, this.field_147558_l) < p_180546_1_.func_145833_n())
         {
+            if(!net.minecraftforge.client.render.BatchedTileEntityRenderer.isDrawingBatch() || !p_180546_1_.hasFastRenderer())
+            {
             RenderHelper.func_74519_b();
             int i = this.field_147550_f.func_175626_b(p_180546_1_.func_174877_v(), 0);
             int j = i % 65536;
             int k = i / 65536;
             OpenGlHelper.func_77475_a(OpenGlHelper.field_77476_b, (float)j, (float)k);
             GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
+            }
             BlockPos blockpos = p_180546_1_.func_174877_v();
             this.func_192854_a(p_180546_1_, (double)blockpos.func_177958_n() - field_147554_b, (double)blockpos.func_177956_o() - field_147555_c, (double)blockpos.func_177952_p() - field_147552_d, p_180546_2_, p_180546_3_, 1.0F);
         }
@@ -146,6 +149,9 @@
         {
             try
             {
+                if(net.minecraftforge.client.render.BatchedTileEntityRenderer.isDrawingBatch() && p_192854_1_.hasFastRenderer())
+                    tileentityspecialrenderer.renderTileEntityFast(p_192854_1_, p_192854_2_, p_192854_4_, p_192854_6_, p_192854_8_, p_192854_9_, p_192854_10_, net.minecraftforge.client.render.BatchedTileEntityRenderer.getAll());
+                else
                 tileentityspecialrenderer.func_192841_a(p_192854_1_, p_192854_2_, p_192854_4_, p_192854_6_, p_192854_8_, p_192854_9_, p_192854_10_);
             }
             catch (Throwable throwable)
