--- ../src_base/minecraft/net/minecraft/src/BlockRedstoneWire.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/BlockRedstoneWire.java	0000-00-00 00:00:00.000000000 -0000
@@ -2,6 +2,8 @@
 
 import java.util.*;
 
+import net.minecraft.src.forge.IConnectRedstone;
+
 public class BlockRedstoneWire extends Block
 {
     private boolean wiresProvidePower;
@@ -47,7 +49,7 @@
 
     public boolean canPlaceBlockAt(World world, int i, int j, int k)
     {
-        return world.isBlockNormalCube(i, j - 1, k);
+        return world.isBlockSolidOnSide(i, j - 1, k, 1);
     }
 
     private void updateAndPropagateCurrentStrength(World world, int i, int j, int k)
@@ -467,6 +469,12 @@
             int j1 = iblockaccess.getBlockMetadata(i, j, k);
             return l == (j1 & 3) || l == Direction.footInvisibleFaceRemap[j1 & 3];
         }
+        
+        if (Block.blocksList[i1] instanceof IConnectRedstone)
+        {
+        	return ((IConnectRedstone)Block.blocksList[i1]).canConnectRedstone(iblockaccess, i, j, k, l);
+        }
+        
         return Block.blocksList[i1].canProvidePower() && l != -1;
     }
 
