--- ../src_base/minecraft/net/minecraft/src/BlockContainer.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/BlockContainer.java	0000-00-00 00:00:00.000000000 -0000
@@ -20,7 +20,7 @@
     public void onBlockAdded(World par1World, int par2, int par3, int par4)
     {
         super.onBlockAdded(par1World, par2, par3, par4);
-        par1World.setBlockTileEntity(par2, par3, par4, this.getBlockEntity());
+        par1World.setBlockTileEntity(par2, par3, par4, this.getBlockEntity(par1World.getBlockMetadata(par2, par3, par4)));
     }
 
     /**
@@ -51,4 +51,14 @@
             var7.receiveClientEvent(par5, par6);
         }
     }
+    
+    /**
+     * Metadata-sensitive version, to fix 1.8.1 regression.
+     * @param meta The current Metadata 
+     * @return And instance of the TileEntity class for this block
+     */
+    public TileEntity getBlockEntity(int meta)
+    {
+        return getBlockEntity();
+    }
 }
