--- ../src_base/minecraft/net/minecraft/src/BlockTrapDoor.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/BlockTrapDoor.java	0000-00-00 00:00:00.000000000 -0000
@@ -11,7 +11,10 @@
 
 public class BlockTrapDoor extends Block
 {
-
+	/* FORGE: Set this to allow trapdoors to remain free-floating
+	 */
+	public static boolean disableValidation=false;
+	
     protected BlockTrapDoor(int i, Material material)
     {
         super(i, material);
@@ -147,7 +150,7 @@
         {
             j1--;
         }
-        if(!func_41052_f(world.getBlockId(j1, j, k1)))
+        if(!(func_41052_f(world.getBlockId(j1, j, k1)) || world.isBlockSolidOnSide(j1, j, k1, (i1&3)+2)))
         {
             world.setBlockWithNotify(i, j, k, 0);
             dropBlockAsItem(world, i, j, k, i1, 0);
@@ -192,6 +195,7 @@
 
     public boolean canPlaceBlockOnSide(World world, int i, int j, int k, int l)
     {
+        if(disableValidation) return true;
         if(l == 0)
         {
             return false;
@@ -216,7 +220,7 @@
         {
             i--;
         }
-        return func_41052_f(world.getBlockId(i, j, k));
+        return func_41052_f(world.getBlockId(i, j, k)) || world.isBlockSolidOnSide(i, j, k, l);
     }
 
     public static boolean isTrapdoorOpen(int i)
@@ -226,6 +230,10 @@
 
     private static boolean func_41052_f(int i)
     {
+    	if  (disableValidation)
+    	{
+    		return true;
+    	}
         if(i <= 0)
         {
             return false;
