--- ../src_base/minecraft/net/minecraft/src/BlockPistonBase.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/BlockPistonBase.java	0000-00-00 00:00:00.000000000 -0000
@@ -361,7 +361,7 @@
                 return false;
             }
         }
-        return !(Block.blocksList[i] instanceof BlockContainer);
+        return !(Block.blocksList[i] != null &&  Block.blocksList[i].hasTileEntity(world.getBlockMetadata(j, k, l)));
     }
 
     private static boolean canExtend(World world, int i, int j, int k, int l)
