--- ../src_base/minecraft/net/minecraft/src/ContainerCreative.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/ContainerCreative.java	0000-00-00 00:00:00.000000000 -0000
@@ -68,17 +68,20 @@
             this.itemList.add(new ItemStack(var2[var12], 1, var13));
         }
 
-        for (var12 = 256; var12 < Item.itemsList.length; ++var12)
+        for (Block block : Block.blocksList)
         {
-            if (Item.itemsList[var12] != null && Item.itemsList[var12].shiftedIndex != Item.potion.shiftedIndex && Item.itemsList[var12].shiftedIndex != Item.monsterPlacer.shiftedIndex)
+            if (block != null)
             {
-                this.itemList.add(new ItemStack(Item.itemsList[var12]));
+                block.addCreativeItems((ArrayList)itemList);
             }
         }
-
-        for (var12 = 1; var12 < 16; ++var12)
+        
+        for (Item item : Item.itemsList)
         {
-            this.itemList.add(new ItemStack(Item.dyePowder.shiftedIndex, 1, var12));
+            if (item != null && !(item instanceof ItemBlock))
+            {
+                item.addCreativeItems((ArrayList)itemList);
+            }
         }
 
         Iterator var15 = EntityList.entityEggs.keySet().iterator();
