--- ../src_base/minecraft/net/minecraft/src/Entity.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/Entity.java	0000-00-00 00:00:00.000000000 -0000
@@ -5,7 +5,7 @@
 
 public abstract class Entity
 {
-    private static int nextEntityID = 0;
+    public static int nextEntityID = 0;
     public int entityId;
     public double renderDistanceWeight;
 
@@ -173,6 +173,9 @@
      */
     public boolean ignoreFrustumCheck;
     public boolean isAirBorne;
+    
+    /** Forge: Used to store custom data for each entity. */
+    private NBTTagCompound customEntityData;
 
     public Entity(World par1World)
     {
@@ -214,6 +217,29 @@
     }
 
     protected abstract void entityInit();
+    
+    /**
+     * Returns a NBTTagCompound that can be used to store custom data for this entity.
+     * It will be written, and read from disc, so it persists over world saves.
+     * @return A NBTTagCompound
+     */
+    public NBTTagCompound getEntityData()
+    {
+    	if (customEntityData == null)
+    	{
+    		customEntityData = new NBTTagCompound();
+    	}
+    	return customEntityData;
+    }
+    
+    /**
+     * Used in model rendering to determine if the entity riding this entity should be in the 'sitting' position.
+     * @return false to prevent an entity that is mounted to this entity from displaying the 'sitting' animation.
+     */
+    public boolean shouldRiderSit()
+    {
+        return true;
+    }
 
     public DataWatcher getDataWatcher()
     {
@@ -1318,6 +1344,11 @@
         par1NBTTagCompound.setShort("Fire", (short)this.fire);
         par1NBTTagCompound.setShort("Air", (short)this.getAir());
         par1NBTTagCompound.setBoolean("OnGround", this.onGround);
+        if (customEntityData != null)
+        {
+            par1NBTTagCompound.setCompoundTag("ForgeData", customEntityData);
+        }
+        par1NBTTagCompound.setInteger("entityID", entityId);
         this.writeEntityToNBT(par1NBTTagCompound);
     }
 
@@ -1359,6 +1390,14 @@
         this.onGround = par1NBTTagCompound.getBoolean("OnGround");
         this.setPosition(this.posX, this.posY, this.posZ);
         this.setRotation(this.rotationYaw, this.rotationPitch);
+        if (par1NBTTagCompound.hasKey("ForgeData"))
+        {
+            customEntityData = par1NBTTagCompound.getCompoundTag("ForgeData");
+        }
+        if (par1NBTTagCompound.hasKey("entityID"))
+        {
+        	entityId = par1NBTTagCompound.getInteger("entityID");
+        }
         this.readEntityFromNBT(par1NBTTagCompound);
     }
 
@@ -1717,7 +1756,7 @@
      */
     public boolean isRiding()
     {
-        return this.ridingEntity != null || this.getFlag(2);
+        return (this.ridingEntity != null && ridingEntity.shouldRiderSit()) || this.getFlag(2);
     }
 
     /**
