--- ../src_base/minecraft/net/minecraft/src/EntityPlayerSP.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/EntityPlayerSP.java	0000-00-00 00:00:00.000000000 -0000
@@ -2,6 +2,8 @@
 
 import java.util.Random;
 import net.minecraft.client.Minecraft;
+import net.minecraft.src.forge.IGuiHandler;
+import net.minecraft.src.forge.MinecraftForge;
 
 public class EntityPlayerSP extends EntityPlayer
 {
@@ -474,4 +476,28 @@
         experienceTotal = i;
         experienceLevel = j;
     }
+    
+    /**
+     * Opens a Gui for the player. 
+     * 
+     * @param mod The mod associated with the gui
+     * @param ID The ID number for the Gui
+     * @param world The World
+     * @param X X Position
+     * @param Y Y Position
+     * @param Z Z Position
+     */
+    @Override
+    public void openGui(BaseMod mod, int ID, World world, int X, int Y, int Z)
+    {
+        IGuiHandler handler = MinecraftForge.getGuiHandler(mod);
+        if (handler != null)
+        {
+            GuiScreen screen = handler.getGuiScreen(ID, this, world, X, Y, Z);
+            if (screen != null)
+            {
+                mc.displayGuiScreen(screen);
+            }
+        }
+    }
 }
