--- ../src_base/minecraft/net/minecraft/src/Explosion.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/Explosion.java	0000-00-00 00:00:00.000000000 -0000
@@ -5,6 +5,7 @@
 package net.minecraft.src;
 
 import java.util.*;
+import net.minecraft.src.forge.ISpecialResistance;
 
 // Referenced classes of package net.minecraft.src:
 //            World, MathHelper, Block, ChunkPosition, 
@@ -76,7 +77,14 @@
                         int k3 = worldObj.getBlockId(l2, i3, j3);
                         if(k3 > 0)
                         {
-                            f1 -= (Block.blocksList[k3].getExplosionResistance(exploder) + 0.3F) * f2;
+                            if (Block.blocksList[k3] instanceof ISpecialResistance){
+                            	ISpecialResistance isr = (ISpecialResistance)Block.blocksList[k3];
+                            	f1 -= (isr.getSpecialExplosionResistance(worldObj, l2, i3, j3, 
+                            			explosionX, explosionY, explosionZ, exploder) + 0.3F) * f2;
+                            } else {
+                                f1 -= (Block.blocksList[k3].getExplosionResistance(exploder) + 0.3F) * f2;
+                            }
+                            
                         }
                         if(f1 > 0.0F)
                         {
