--- ../src_base/minecraft/net/minecraft/src/ItemHoe.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/ItemHoe.java	0000-00-00 00:00:00.000000000 -0000
@@ -1,5 +1,7 @@
 package net.minecraft.src;
 
+import net.minecraft.src.forge.ForgeHooks;
+
 public class ItemHoe extends Item
 {
     public ItemHoe(int par1, EnumToolMaterial par2EnumToolMaterial)
@@ -21,6 +23,11 @@
         }
         else
         {
+            if (ForgeHooks.onUseHoe(par1ItemStack, par2EntityPlayer, par3World, par4, par5, par6))
+            {
+                par1ItemStack.damageItem(1, par2EntityPlayer);
+                return true;
+            }
             int var8 = par3World.getBlockId(par4, par5, par6);
             int var9 = par3World.getBlockId(par4, par5 + 1, par6);
 
