--- ../src_base/minecraft/net/minecraft/src/NetworkManager.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/NetworkManager.java	0000-00-00 00:00:00.000000000 -0000
@@ -11,6 +11,8 @@
 import java.util.Collections;
 import java.util.List;
 
+import net.minecraft.src.forge.ForgeHooks;
+
 public class NetworkManager
 {
     /** Synchronization object used for read and write threads. */
@@ -307,6 +309,7 @@
             {
                 ;
             }
+            ForgeHooks.onDisconnect(this, par1Str, par2ArrayOfObj);
         }
     }
 
@@ -430,4 +433,14 @@
     {
         return par0NetworkManager.writeThread;
     }
+    
+    /**
+     * Retrieves the current associated network handler.
+     * Added so modders don't have to use reflection.
+     * @return The current registered Network Handler
+     */
+    public NetHandler getNetHandler()
+    {
+        return netHandler;
+    }
 }
