--- ../src_base/minecraft/net/minecraft/src/Packet100OpenWindow.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/Packet100OpenWindow.java	0000-00-00 00:00:00.000000000 -0000
@@ -25,9 +25,9 @@
         throws IOException
     {
         windowId = datainputstream.readByte();
-        inventoryType = datainputstream.readByte();
+        inventoryType = datainputstream.readByte()&0xFF;
         windowTitle = readString(datainputstream, 16);
-        slotsCount = datainputstream.readByte();
+        slotsCount = datainputstream.readByte()&0xFF;
     }
 
     public void writePacketData(DataOutputStream dataoutputstream)
