--- ../src_base/minecraft/net/minecraft/src/PlayerControllerMP.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/PlayerControllerMP.java	0000-00-00 00:00:00.000000000 -0000
@@ -1,6 +1,7 @@
 package net.minecraft.src;
 
 import net.minecraft.client.Minecraft;
+import net.minecraft.src.forge.ForgeHooks;
 
 public class PlayerControllerMP extends PlayerController
 {
@@ -56,13 +57,19 @@
 
     public boolean onPlayerDestroyBlock(int i, int j, int k, int l)
     {
+    	ItemStack itemstack = mc.thePlayer.getCurrentEquippedItem();
+    	if (itemstack != null && itemstack.getItem().onBlockStartBreak(itemstack, i, j, k, mc.thePlayer))
+    	{
+    		return false;
+    	}
+    	
         if (creativeMode)
         {
             return super.onPlayerDestroyBlock(i, j, k, l);
         }
         int i1 = mc.theWorld.getBlockId(i, j, k);
         boolean flag = super.onPlayerDestroyBlock(i, j, k, l);
-        ItemStack itemstack = mc.thePlayer.getCurrentEquippedItem();
+
         if (itemstack != null)
         {
             itemstack.onDestroyBlock(i1, i, j, k, mc.thePlayer);
@@ -91,7 +98,7 @@
             {
                 Block.blocksList[i1].onBlockClicked(mc.theWorld, i, j, k, mc.thePlayer);
             }
-            if (i1 > 0 && Block.blocksList[i1].blockStrength(mc.thePlayer) >= 1.0F)
+            if(i1 > 0 && Block.blocksList[i1].blockStrength(mc.theWorld, mc.thePlayer, i, j, k) >= 1.0F)
             {
                 onPlayerDestroyBlock(i, j, k, l);
             }
@@ -138,7 +145,7 @@
                 return;
             }
             Block block = Block.blocksList[i1];
-            curBlockDamageMP += block.blockStrength(mc.thePlayer);
+            curBlockDamageMP += block.blockStrength(mc.theWorld, mc.thePlayer, i, j, k);
             if (stepSoundTickCounter % 4F == 0.0F && block != null)
             {
                 mc.sndManager.playSound(block.stepSound.getStepSound(), (float)i + 0.5F, (float)j + 0.5F, (float)k + 0.5F, (block.stepSound.getVolume() + 1.0F) / 8F, block.stepSound.getPitch() * 0.5F);
@@ -207,6 +214,16 @@
     {
         syncCurrentPlayItem();
         netClientHandler.addToSendQueue(new Packet15Place(i, j, k, l, entityplayer.inventory.getCurrentItem()));
+
+        if (itemstack != null) 
+        {
+        	Item item = itemstack.getItem();
+        	if (item.onItemUseFirst(itemstack, entityplayer, world, i, j, k, l)) 
+        	{
+        		return true;
+    		}
+        }
+        
         int i1 = world.getBlockId(i, j, k);
         if (i1 > 0 && Block.blocksList[i1].blockActivated(world, i, j, k, entityplayer))
         {
@@ -227,7 +244,15 @@
         }
         else
         {
-            return itemstack.useItem(entityplayer, world, i, j, k, l);
+            if (!itemstack.useItem(entityplayer, world, i, j, k, l))
+            {
+            	return false;
+            }
+        	if (itemstack.stackSize == 0)
+        	{
+        		ForgeHooks.onDestroyCurrentItem(entityplayer, itemstack);
+        	}
+        	return true;
         }
     }
 
