--- ../src_base/minecraft/net/minecraft/src/PlayerController.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/PlayerController.java	0000-00-00 00:00:00.000000000 -0000
@@ -1,6 +1,7 @@
 package net.minecraft.src;
 
 import net.minecraft.client.Minecraft;
+import net.minecraft.src.forge.ForgeHooks;
 
 public abstract class PlayerController
 {
@@ -39,7 +40,7 @@
         {
             var5.playAuxSFX(2001, par1, par2, par3, var6.blockID + (var5.getBlockMetadata(par1, par2, par3) << 12));
             int var7 = var5.getBlockMetadata(par1, par2, par3);
-            boolean var8 = var5.setBlockWithNotify(par1, par2, par3, 0);
+            boolean var8 = var6.removeBlockByPlayer(var5, mc.thePlayer, par1, par2, par3);
 
             if (var8)
             {
@@ -86,6 +87,7 @@
             if (var5.stackSize == 0)
             {
                 par1EntityPlayer.inventory.mainInventory[par1EntityPlayer.inventory.currentItem] = null;
+                ForgeHooks.onDestroyCurrentItem(par1EntityPlayer, var5);
             }
 
             return true;
