--- ../src_base/minecraft/net/minecraft/src/RenderEngine.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/RenderEngine.java	0000-00-00 00:00:00.000000000 -0000
@@ -14,6 +14,10 @@
 import java.util.List;
 import java.util.Map;
 import javax.imageio.ImageIO;
+
+import net.minecraft.src.forge.ForgeHooks;
+import net.minecraft.src.forge.ForgeHooksClient;
+
 import org.lwjgl.opengl.GL11;
 
 public class RenderEngine
@@ -48,7 +52,7 @@
     public boolean blurTexture = false;
 
     /** Texture pack */
-    private TexturePackList texturePack;
+    public TexturePackList texturePack;
 
     /** Missing texture image */
     private BufferedImage missingTextureImage = new BufferedImage(64, 64, 2);
@@ -156,6 +160,10 @@
         {
             try
             {
+                if (Tessellator.renderingWorldRenderer)
+                {
+                    System.out.printf("Warning: Texture %s not preloaded, will cause render glitches!\n", par1Str);
+                }
                 this.singleIntBuffer.clear();
                 GLAllocation.generateTextureNames(this.singleIntBuffer);
                 int var6 = this.singleIntBuffer.get(0);
@@ -199,6 +207,7 @@
                 }
 
                 this.textureMap.put(par1Str, Integer.valueOf(var6));
+                ForgeHooksClient.onTextureLoad(par1Str);
                 return var6;
             }
             catch (Exception var5)
