--- ../src_base/minecraft/net/minecraft/src/RenderEngine.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/RenderEngine.java	0000-00-00 00:00:00.000000000 -0000
@@ -48,7 +48,7 @@
     public boolean blurTexture = false;
 
     /** Texture pack */
-    private TexturePackList texturePack;
+    public TexturePackList texturePack;
 
     /** Missing texture image */
     private BufferedImage missingTextureImage = new BufferedImage(64, 64, 2);
@@ -156,6 +156,10 @@
         {
             try
             {
+                if (Tessellator.renderingWorldRenderer)
+                {
+                    System.out.printf("Warning: Texture %s not preloaded, will cause render glitches!\n", par1Str);
+                }
                 this.singleIntBuffer.clear();
                 GLAllocation.generateTextureNames(this.singleIntBuffer);
                 int var6 = this.singleIntBuffer.get(0);
