--- ../src_base/minecraft/net/minecraft/src/TileEntityRendererPiston.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/TileEntityRendererPiston.java	0000-00-00 00:00:00.000000000 -0000
@@ -1,6 +1,8 @@
 package net.minecraft.src;
 
 import net.minecraft.client.Minecraft;
+import net.minecraft.src.forge.ForgeHooksClient;
+
 import org.lwjgl.opengl.GL11;
 
 public class TileEntityRendererPiston extends TileEntitySpecialRenderer
@@ -29,7 +31,7 @@
             {
                 GL11.glShadeModel(GL11.GL_FLAT);
             }
-
+            ForgeHooksClient.beforeBlockRender(var9, blockRenderer);
             var10.startDrawingQuads();
             var10.setTranslation((double)((float)par2 - (float)par1TileEntityPiston.xCoord + par1TileEntityPiston.getOffsetX(par8)), (double)((float)par4 - (float)par1TileEntityPiston.yCoord + par1TileEntityPiston.getOffsetY(par8)), (double)((float)par6 - (float)par1TileEntityPiston.zCoord + par1TileEntityPiston.getOffsetZ(par8)));
             var10.setColorOpaque(1, 1, 1);
@@ -53,6 +55,7 @@
 
             var10.setTranslation(0.0D, 0.0D, 0.0D);
             var10.draw();
+            ForgeHooksClient.afterBlockRender(var9, blockRenderer);
             RenderHelper.enableStandardItemLighting();
         }
     }
