--- ../src_base/minecraft/net/minecraft/src/WorldGenShrub.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/WorldGenShrub.java	0000-00-00 00:00:00.000000000 -0000
@@ -2,6 +2,8 @@
 
 import java.util.Random;
 
+import net.minecraft.src.forge.MinecraftForge;
+
 public class WorldGenShrub extends WorldGenerator
 {
     private int field_48197_a;
@@ -17,7 +19,7 @@
     {
         int var15;
 
-        for (boolean var6 = false; ((var15 = par1World.getBlockId(par3, par4, par5)) == 0 || var15 == Block.leaves.blockID) && par4 > 0; --par4)
+        for (boolean var6 = false; ((var15 = par1World.getBlockId(par3, par4, par5)) == 0 || MinecraftForge.isLeaves(var15)) && par4 > 0; --par4)
         {
             ;
         }
@@ -42,7 +44,7 @@
                     {
                         int var14 = var13 - par5;
 
-                        if ((Math.abs(var12) != var10 || Math.abs(var14) != var10 || par2Random.nextInt(2) != 0) && !Block.opaqueCubeLookup[par1World.getBlockId(var11, var8, var13)])
+                        if ((Math.abs(var12) != var10 || Math.abs(var14) != var10 || par2Random.nextInt(2) != 0) && !Block.opaqueCubeLookup[par1World.getBlockId(var11, var8, var13)] && MinecraftForge.leavesCanReplace(par1World.getBlockId(var11, var8, var13)))
                         {
                             this.setBlockAndMetadata(par1World, var11, var8, var13, Block.leaves.blockID, this.field_48197_a);
                         }
