--- ../src_base/minecraft/net/minecraft/src/WorldGenTrees.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/WorldGenTrees.java	0000-00-00 00:00:00.000000000 -0000
@@ -2,6 +2,8 @@
 
 import java.util.Random;
 
+import net.minecraft.src.forge.MinecraftForge;
+
 public class WorldGenTrees extends WorldGenerator
 {
     /** The minimum height of a generated tree. */
@@ -64,7 +66,7 @@
                         {
                             var12 = par1World.getBlockId(var10, var8, var11);
 
-                            if (var12 != 0 && var12 != Block.leaves.blockID && var12 != Block.grass.blockID && var12 != Block.dirt.blockID && var12 != Block.wood.blockID)
+                            if (var12 != 0 && !MinecraftForge.isLeaves(var12) && var12 != Block.grass.blockID && var12 != Block.dirt.blockID && !MinecraftForge.isWood(var12))
                             {
                                 var7 = false;
                             }
@@ -119,7 +121,7 @@
                     {
                         var12 = par1World.getBlockId(par3, par4 + var11, par5);
 
-                        if (var12 == 0 || var12 == Block.leaves.blockID)
+                        if (var12 == 0 || MinecraftForge.isLeaves(var12))
                         {
                             this.setBlockAndMetadata(par1World, par3, par4 + var11, par5, Block.wood.blockID, this.metaWood);
 
@@ -159,7 +161,7 @@
                             {
                                 for (var15 = par5 - var13; var15 <= par5 + var13; ++var15)
                                 {
-                                    if (par1World.getBlockId(var14, var11, var15) == Block.leaves.blockID)
+                                    if (MinecraftForge.isLeaves(par1World.getBlockId(var14, var11, var15)))
                                     {
                                         if (par2Random.nextInt(4) == 0 && par1World.getBlockId(var14 - 1, var11, var15) == 0)
                                         {
