--- ../src_base/minecraft/net/minecraft/src/WorldRenderer.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/WorldRenderer.java	0000-00-00 00:00:00.000000000 -0000
@@ -3,6 +3,9 @@
 import java.util.ArrayList;
 import java.util.HashSet;
 import java.util.List;
+
+import net.minecraft.src.forge.ForgeHooksClient;
+
 import org.lwjgl.opengl.GL11;
 
 public class WorldRenderer
@@ -10,7 +13,6 @@
     /** Reference to the World object. */
     public World worldObj;
     private int glRenderList = -1;
-    private static Tessellator tessellator = Tessellator.instance;
     public static int chunksUpdated = 0;
     public int posX;
     public int posY;
@@ -179,11 +181,12 @@
                                         GL11.glTranslatef(-8.0F, -8.0F, -8.0F);
                                         GL11.glScalef(var19, var19, var19);
                                         GL11.glTranslatef(8.0F, 8.0F, 8.0F);
-                                        tessellator.startDrawingQuads();
-                                        tessellator.setTranslation((double)(-this.posX), (double)(-this.posY), (double)(-this.posZ));
+                                        ForgeHooksClient.beforeRenderPass(var11);
+                                        Tessellator.instance.startDrawingQuads();
+                                        Tessellator.instance.setTranslation((double)(-this.posX), (double)(-this.posY), (double)(-this.posZ));
                                     }
 
-                                    if (var11 == 0 && Block.blocksList[var18].hasTileEntity())
+                                    if (var11 == 0 && Block.blocksList[var18] != null && Block.blocksList[var18].hasTileEntity(var9.getBlockMetadata(var17, var15, var16)))
                                     {
                                         TileEntity var23 = var9.getBlockTileEntity(var17, var15, var16);
 
@@ -196,14 +199,17 @@
                                     Block var24 = Block.blocksList[var18];
                                     int var20 = var24.getRenderBlockPass();
 
-                                    if (var20 != var11)
+                                    if (var20 > var11)
                                     {
                                         var12 = true;
                                     }
-                                    else if (var20 == var11)
+                                    if (!ForgeHooksClient.canRenderInPass(var24, var11))
                                     {
-                                        var13 |= var10.renderBlockByRenderType(var24, var17, var15, var16);
+                                        continue;
                                     }
+                                    ForgeHooksClient.beforeBlockRender(var24, var10);
+                                    var13 |= var10.renderBlockByRenderType(var24, var17, var15, var16);
+                                    ForgeHooksClient.afterBlockRender(var24, var10);
                                 }
                             }
                         }
@@ -211,10 +217,11 @@
 
                     if (var14)
                     {
-                        this.bytesDrawn += tessellator.draw();
+                        ForgeHooksClient.afterRenderPass(var11);
+                        this.bytesDrawn += Tessellator.instance.draw();
                         GL11.glPopMatrix();
                         GL11.glEndList();
-                        tessellator.setTranslation(0.0D, 0.0D, 0.0D);
+                        Tessellator.instance.setTranslation(0.0D, 0.0D, 0.0D);
                     }
                     else
                     {
