--- a/net/minecraft/world/gen/feature/DarkOakTreeFeature.java
+++ b/net/minecraft/world/gen/feature/DarkOakTreeFeature.java
@@ -20,17 +20,17 @@
       int j = p_225557_3_.func_177958_n();
       int k = p_225557_3_.func_177956_o();
       int l = p_225557_3_.func_177952_p();
-      if (k >= 1 && k + i + 1 < 256) {
+      if (k >= 1 && k + i + 1 < p_225557_1_.getMaxHeight()) {
          BlockPos blockpos = p_225557_3_.func_177977_b();
-         if (!func_214589_h(p_225557_1_, blockpos)) {
+         if (!isSoil(p_225557_1_, blockpos, p_225557_7_.getSapling())) {
             return false;
          } else if (!this.func_214615_a(p_225557_1_, p_225557_3_, i)) {
             return false;
          } else {
-            this.func_214584_a(p_225557_1_, blockpos);
-            this.func_214584_a(p_225557_1_, blockpos.func_177974_f());
-            this.func_214584_a(p_225557_1_, blockpos.func_177968_d());
-            this.func_214584_a(p_225557_1_, blockpos.func_177968_d().func_177974_f());
+            this.setDirtAt(p_225557_1_, blockpos, p_225557_3_);
+            this.setDirtAt(p_225557_1_, blockpos.func_177974_f(), p_225557_3_);
+            this.setDirtAt(p_225557_1_, blockpos.func_177968_d(), p_225557_3_);
+            this.setDirtAt(p_225557_1_, blockpos.func_177968_d().func_177974_f(), p_225557_3_);
             Direction direction = Direction.Plane.HORIZONTAL.func_179518_a(p_225557_2_);
             int i1 = i - p_225557_2_.nextInt(4);
             int j1 = 2 - p_225557_2_.nextInt(3);
