--- a/net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
+++ b/net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
@@ -107,7 +_,7 @@
       this.m_5674_();
       this.f_100279_.clear();
 
-      for(RecipeBookCategories recipebookcategories : RecipeBookCategories.m_92269_(this.f_100271_.m_5867_())) {
+      for(RecipeBookCategories recipebookcategories : this.f_100271_.getRecipeBookCategories()) {
          this.f_100279_.add(new RecipeBookTabButton(recipebookcategories));
       }
 
@@ -319,7 +_,7 @@
       }
 
       if (itemstack != null && this.f_100272_.f_91080_ != null) {
-         this.f_100272_.f_91080_.m_96597_(p_100375_, this.f_100272_.f_91080_.m_96555_(itemstack), p_100378_, p_100379_);
+         this.f_100272_.f_91080_.renderComponentTooltip(p_100375_, this.f_100272_.f_91080_.m_96555_(itemstack), p_100378_, p_100379_, itemstack);
       }
 
    }
