--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -385,6 +_,7 @@
 
       this.minecraft.debugRenderer.clear();
       this.minecraft.player.resetPos();
+      net.minecraftforge.client.ForgeHooksClient.firePlayerLogin(this.minecraft.gameMode, this.minecraft.player, this.minecraft.getConnection().connection);
       this.minecraft.player.setId(p_105030_.playerId());
       this.level.addEntity(this.minecraft.player);
       this.minecraft.player.input = new KeyboardInput(this.minecraft.options);
@@ -1016,7 +_,9 @@
          localplayer1.getAttributes().assignValues(localplayer.getAttributes());
       }
 
+      localplayer1.updateSyncFields(localplayer); // Forge: fix MC-10657
       localplayer1.resetPos();
+      net.minecraftforge.client.ForgeHooksClient.firePlayerRespawn(this.minecraft.gameMode, localplayer, localplayer1, localplayer1.connection.connection);
       this.level.addEntity(localplayer1);
       localplayer1.setYRot(-180.0F);
       localplayer1.input = new KeyboardInput(this.minecraft.options);
@@ -1125,10 +_,7 @@
       PacketUtils.ensureRunningOnSameThread(p_104976_, this, this.minecraft);
       BlockPos blockpos = p_104976_.getPos();
       this.minecraft.level.getBlockEntity(blockpos, p_104976_.getType()).ifPresent((p_205557_) -> {
-         CompoundTag compoundtag = p_104976_.getTag();
-         if (compoundtag != null) {
-            p_205557_.load(compoundtag);
-         }
+         p_205557_.onDataPacket(connection, p_104976_);
 
          if (p_205557_ instanceof CommandBlockEntity && this.minecraft.screen instanceof CommandBlockEditScreen) {
             ((CommandBlockEditScreen)this.minecraft.screen).updateGui();
@@ -1276,7 +_,9 @@
 
    public void handleCommands(ClientboundCommandsPacket p_104990_) {
       PacketUtils.ensureRunningOnSameThread(p_104990_, this, this.minecraft);
-      this.commands = new CommandDispatcher<>(p_104990_.getRoot(CommandBuildContext.simple(this.registryAccess, this.enabledFeatures)));
+      var context = CommandBuildContext.simple(this.registryAccess, this.enabledFeatures);
+      this.commands = new CommandDispatcher<>(p_104990_.getRoot(context));
+      this.commands = net.minecraftforge.client.ClientCommandHandler.mergeServerCommands(this.commands, context);
    }
 
    public void handleStopSoundEvent(ClientboundStopSoundPacket p_105116_) {
@@ -1295,6 +_,7 @@
       ClientRecipeBook clientrecipebook = this.minecraft.player.getRecipeBook();
       clientrecipebook.setupCollections(this.recipeManager.getRecipes(), this.minecraft.level.registryAccess());
       this.minecraft.populateSearchTree(SearchRegistry.RECIPE_COLLECTIONS, clientrecipebook.getCollections());
+      net.minecraftforge.client.ForgeHooksClient.onRecipesUpdated(this.recipeManager);
    }
 
    public void handleLookAt(ClientboundPlayerLookAtPacket p_105054_) {
@@ -1393,7 +_,8 @@
          Blocks.rebuildCache();
       }
 
-      CreativeModeTabs.searchTab().rebuildSearchTree();
+      CreativeModeTabs.allTabs().stream().filter(net.minecraft.world.item.CreativeModeTab::hasSearchBar).forEach(net.minecraft.world.item.CreativeModeTab::rebuildSearchTree);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.TagsUpdatedEvent(this.registryAccess, true, connection.isMemoryConnection()));
    }
 
    public void handlePlayerCombatEnd(ClientboundPlayerCombatEndPacket p_171771_) {
@@ -2086,6 +_,8 @@
    }
 
    public void sendChat(String p_249888_) {
+      p_249888_ = net.minecraftforge.client.ForgeHooksClient.onClientSendMessage(p_249888_);
+      if (p_249888_.isEmpty()) return;
       Instant instant = Instant.now();
       long i = Crypt.SaltSupplier.getLong();
       LastSeenMessagesTracker.Update lastseenmessagestracker$update = this.lastSeenMessages.generateAndApplyUpdate();
@@ -2094,6 +_,7 @@
    }
 
    public void sendCommand(String p_250092_) {
+      if (net.minecraftforge.client.ClientCommandHandler.runCommand(p_250092_)) return;
       Instant instant = Instant.now();
       long i = Crypt.SaltSupplier.getLong();
       LastSeenMessagesTracker.Update lastseenmessagestracker$update = this.lastSeenMessages.generateAndApplyUpdate();
