--- a/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java
+++ b/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java
@@ -89,6 +_,7 @@
    }
 
    public void handleCustomPayload(ClientboundCustomPayloadPacket p_298103_) {
+      if (net.minecraftforge.common.ForgeHooks.onCustomPayload(p_298103_, this.connection)) return;
       CustomPacketPayload custompacketpayload = p_298103_.payload();
       if (!(custompacketpayload instanceof DiscardedPayload)) {
          PacketUtils.ensureRunningOnSameThread(p_298103_, this, this.minecraft);
