--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -360,6 +_,7 @@
 
       this.f_104888_.f_91064_.m_113434_();
       this.f_104888_.f_91074_.m_172530_();
+      net.minecraftforge.client.ForgeHooksClient.firePlayerLogin(this.f_104888_.f_91072_, this.f_104888_.f_91074_, this.f_104888_.m_91403_().m_6198_());
       int i = p_105030_.f_132360_();
       this.f_104888_.f_91074_.m_20234_(i);
       this.f_104889_.m_104630_(i, this.f_104888_.f_91074_);
@@ -371,6 +_,7 @@
       this.f_104888_.f_91074_.m_108711_(p_105030_.f_132372_());
       this.f_104888_.f_91072_.m_171805_(p_105030_.f_132363_(), p_105030_.f_132364_());
       this.f_104888_.f_91066_.m_193770_(p_105030_.f_132370_());
+      net.minecraftforge.network.NetworkHooks.sendMCRegistryPackets(f_104885_, "PLAY_TO_SERVER");
       this.f_104888_.f_91066_.m_92172_();
       this.f_104885_.m_129512_(new ServerboundCustomPayloadPacket(ServerboundCustomPayloadPacket.f_133979_, (new FriendlyByteBuf(Unpooled.buffer())).m_130070_(ClientBrandRetriever.m_129629_())));
       this.f_104888_.m_91309_().m_90739_();
@@ -712,7 +_,9 @@
 
    public void m_5784_(ClientboundChatPacket p_104986_) {
       PacketUtils.m_131363_(p_104986_, this, this.f_104888_);
-      this.f_104888_.f_91065_.m_93051_(p_104986_.m_131840_(), p_104986_.m_131836_(), p_104986_.m_131841_());
+      net.minecraft.network.chat.Component message = net.minecraftforge.event.ForgeEventFactory.onClientChat(p_104986_.m_131840_(), p_104986_.m_131836_(), p_104986_.m_131841_());
+      if (message == null) return;
+      this.f_104888_.f_91065_.m_93051_(p_104986_.m_131840_(), message, p_104986_.m_131841_());
    }
 
    public void m_7791_(ClientboundAnimatePacket p_104968_) {
@@ -894,8 +_,10 @@
          localplayer1.m_21204_().m_22159_(localplayer.m_21204_());
       }
 
+      localplayer1.updateSyncFields(localplayer); // Forge: fix MC-10657
       localplayer1.m_172530_();
       localplayer1.m_108748_(s);
+      net.minecraftforge.client.ForgeHooksClient.firePlayerRespawn(this.f_104888_.f_91072_, localplayer, localplayer1, localplayer1.f_108617_.m_6198_());
       this.f_104889_.m_104630_(i, localplayer1);
       localplayer1.m_146922_(-180.0F);
       localplayer1.f_108618_ = new KeyboardInput(this.f_104888_.f_91066_);
@@ -998,10 +_,7 @@
       PacketUtils.m_131363_(p_104976_, this, this.f_104888_);
       BlockPos blockpos = p_104976_.m_131704_();
       this.f_104888_.f_91073_.m_141902_(blockpos, p_104976_.m_195645_()).ifPresent((p_205557_) -> {
-         CompoundTag compoundtag = p_104976_.m_131708_();
-         if (compoundtag != null) {
-            p_205557_.m_142466_(compoundtag);
-         }
+         p_205557_.onDataPacket(f_104885_, p_104976_);
 
          if (p_205557_ instanceof CommandBlockEntity && this.f_104888_.f_91080_ instanceof CommandBlockEditScreen) {
             ((CommandBlockEditScreen)this.f_104888_.f_91080_).m_98398_();
@@ -1147,6 +_,7 @@
    public void m_7443_(ClientboundCommandsPacket p_104990_) {
       PacketUtils.m_131363_(p_104990_, this, this.f_104888_);
       this.f_104899_ = new CommandDispatcher<>(p_104990_.m_131884_());
+      this.f_104899_ = net.minecraftforge.client.ClientCommandHandler.mergeServerCommands(this.f_104899_);
    }
 
    public void m_7183_(ClientboundStopSoundPacket p_105116_) {
@@ -1168,6 +_,7 @@
       clientrecipebook.m_90625_(this.f_104900_.m_44051_());
       clientrecipebook.m_90639_().forEach(mutablesearchtree::m_8080_);
       mutablesearchtree.m_7729_();
+      net.minecraftforge.client.ForgeHooksClient.onRecipesUpdated(this.f_104900_);
    }
 
    public void m_7244_(ClientboundPlayerLookAtPacket p_105054_) {
@@ -1262,6 +_,7 @@
       }
 
       this.f_104888_.m_91171_(SearchRegistry.f_119942_).m_7729_();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.TagsUpdatedEvent(this.f_104903_));
    }
 
    private <T> void m_205560_(ResourceKey<? extends Registry<? extends T>> p_205561_, TagNetworkSerialization.NetworkPayload p_205562_) {
@@ -1815,7 +_,8 @@
             int i5 = friendlybytebuf.m_130242_();
             this.f_104888_.f_91064_.f_173815_.m_173830_(positionsource, i5);
          } else {
-            f_104883_.warn("Unknown custom packed identifier: {}", (Object)resourcelocation);
+            if (!net.minecraftforge.network.NetworkHooks.onCustomPayload(p_105004_, this.f_104885_))
+            f_104883_.warn("Unknown custom packet identifier: {}", (Object)resourcelocation);
          }
       } finally {
          if (friendlybytebuf != null) {
