--- ../src_base/minecraft/net/minecraft/client/multiplayer/NetClientHandler.java
+++ ../src_work/minecraft/net/minecraft/client/multiplayer/NetClientHandler.java
@@ -160,6 +160,11 @@
 import net.minecraft.world.storage.MapStorage;
 import org.lwjgl.input.Keyboard;
 
+import net.minecraftforge.client.ForgeHooksClient;
+import net.minecraftforge.client.event.ClientChatReceivedEvent;
+import net.minecraftforge.common.ForgeHooks;
+import net.minecraftforge.common.MinecraftForge;
+
 @SideOnly(Side.CLIENT)
 public class NetClientHandler extends NetHandler
 {
@@ -744,7 +749,7 @@
 
     public void handleKickDisconnect(Packet255KickDisconnect par1Packet255KickDisconnect)
     {
-        this.netManager.networkShutdown("disconnect.kicked", new Object[0]);
+        this.netManager.networkShutdown("disconnect.kicked", par1Packet255KickDisconnect.reason);
         this.disconnected = true;
         this.mc.loadWorld((WorldClient)null);
         this.mc.displayGuiScreen(new GuiDisconnected("disconnect.disconnected", "disconnect.genericReason", new Object[] {par1Packet255KickDisconnect.reason}));
@@ -810,7 +815,11 @@
     public void handleChat(Packet3Chat par1Packet3Chat)
     {
         par1Packet3Chat = FMLNetworkHandler.handleChatMessage(this, par1Packet3Chat);
-        this.mc.ingameGUI.getChatGUI().printChatMessage(par1Packet3Chat.message);
+        ClientChatReceivedEvent event = new ClientChatReceivedEvent(par1Packet3Chat.message);
+        if (!MinecraftForge.EVENT_BUS.post(event) && event.message != null)
+        {
+            this.mc.ingameGUI.getChatGUI().printChatMessage(par1Packet3Chat.message);
+        }
     }
 
     public void handleAnimation(Packet18Animation par1Packet18Animation)
@@ -1192,6 +1201,10 @@
                 {
                     var2.readFromNBT(par1Packet132TileEntityData.customParam1);
                 }
+                else
+                {
+                    var2.onDataPacket(netManager,  par1Packet132TileEntityData);
+                }
             }
         }
     }
