--- ../src-base/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java
+++ ../src-work/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java
@@ -122,10 +122,16 @@
             }
         }
 
-        if (this.field_78779_k.func_77145_d() && !this.field_78776_a.field_71439_g.func_184614_ca().func_190926_b() && this.field_78776_a.field_71439_g.func_184614_ca().func_77973_b() instanceof ItemSword)
+        ItemStack stack = field_78776_a.field_71439_g.func_184614_ca();
+        if (!stack.func_190926_b() && stack.func_77973_b().onBlockStartBreak(stack, p_187103_1_, field_78776_a.field_71439_g))
         {
             return false;
         }
+
+        if (this.field_78779_k.func_77145_d() && !stack.func_190926_b() && !stack.func_77973_b().canDestroyBlockInCreative(field_78776_a.field_71441_e, p_187103_1_, stack, field_78776_a.field_71439_g))
+        {
+            return false;
+        }
         else
         {
             World world = this.field_78776_a.field_71441_e;
@@ -143,19 +149,13 @@
             else
             {
                 world.func_175718_b(2001, p_187103_1_, Block.func_176210_f(iblockstate));
-                block.func_176208_a(world, p_187103_1_, iblockstate, this.field_78776_a.field_71439_g);
-                boolean flag = world.func_180501_a(p_187103_1_, Blocks.field_150350_a.func_176223_P(), 11);
 
-                if (flag)
-                {
-                    block.func_176206_d(world, p_187103_1_, iblockstate);
-                }
-
                 this.field_178895_c = new BlockPos(this.field_178895_c.func_177958_n(), -1, this.field_178895_c.func_177952_p());
 
                 if (!this.field_78779_k.func_77145_d())
                 {
                     ItemStack itemstack1 = this.field_78776_a.field_71439_g.func_184614_ca();
+                    ItemStack copyBeforeUse = itemstack1.func_77946_l();
 
                     if (!itemstack1.func_190926_b())
                     {
@@ -163,11 +163,18 @@
 
                         if (itemstack1.func_190926_b())
                         {
+                            net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this.field_78776_a.field_71439_g, copyBeforeUse, EnumHand.MAIN_HAND);
                             this.field_78776_a.field_71439_g.func_184611_a(EnumHand.MAIN_HAND, ItemStack.field_190927_a);
                         }
                     }
                 }
 
+                boolean flag = block.removedByPlayer(iblockstate, world, p_187103_1_, field_78776_a.field_71439_g, false);
+
+                if (flag)
+                {
+                    block.func_176206_d(world, p_187103_1_, iblockstate);
+                }
                 return flag;
             }
         }
@@ -207,6 +214,7 @@
             if (this.field_78779_k.func_77145_d())
             {
                 this.field_78774_b.func_147297_a(new CPacketPlayerDigging(CPacketPlayerDigging.Action.START_DESTROY_BLOCK, p_180511_1_, p_180511_2_));
+                if (!net.minecraftforge.common.ForgeHooks.onLeftClickBlock(this.field_78776_a.field_71439_g, p_180511_1_, p_180511_2_, net.minecraftforge.common.ForgeHooks.rayTraceEyeHitVec(this.field_78776_a.field_71439_g, func_78757_d() + 1)).isCanceled())
                 func_178891_a(this.field_78776_a, this, p_180511_1_, p_180511_2_);
                 this.field_78781_i = 5;
             }
@@ -218,14 +226,17 @@
                 }
 
                 this.field_78774_b.func_147297_a(new CPacketPlayerDigging(CPacketPlayerDigging.Action.START_DESTROY_BLOCK, p_180511_1_, p_180511_2_));
+                net.minecraftforge.event.entity.player.PlayerInteractEvent.LeftClickBlock event = net.minecraftforge.common.ForgeHooks.onLeftClickBlock(this.field_78776_a.field_71439_g, p_180511_1_, p_180511_2_, net.minecraftforge.common.ForgeHooks.rayTraceEyeHitVec(this.field_78776_a.field_71439_g, func_78757_d() + 1));
+
                 IBlockState iblockstate = this.field_78776_a.field_71441_e.func_180495_p(p_180511_1_);
                 boolean flag = iblockstate.func_185904_a() != Material.field_151579_a;
 
                 if (flag && this.field_78770_f == 0.0F)
                 {
+                    if (event.getUseBlock() != net.minecraftforge.fml.common.eventhandler.Event.Result.DENY)
                     iblockstate.func_177230_c().func_180649_a(this.field_78776_a.field_71441_e, p_180511_1_, this.field_78776_a.field_71439_g);
                 }
-
+                if (event.getUseItem() == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) return true;
                 if (flag && iblockstate.func_185903_a(this.field_78776_a.field_71439_g, this.field_78776_a.field_71439_g.field_70170_p, p_180511_1_) >= 1.0F)
                 {
                     this.func_187103_a(p_180511_1_);
@@ -289,7 +300,7 @@
 
                 if (this.field_78780_h % 4.0F == 0.0F)
                 {
-                    SoundType soundtype = block.func_185467_w();
+                    SoundType soundtype = block.getSoundType(iblockstate, field_78776_a.field_71441_e, p_180512_1_, field_78776_a.field_71439_g);
                     this.field_78776_a.func_147118_V().func_147682_a(new PositionedSoundRecord(soundtype.func_185846_f(), SoundCategory.NEUTRAL, (soundtype.func_185843_a() + 1.0F) / 8.0F, soundtype.func_185847_b() * 0.5F, p_180512_1_));
                 }
 
@@ -341,7 +352,7 @@
 
         if (!this.field_85183_f.func_190926_b() && !itemstack.func_190926_b())
         {
-            flag = itemstack.func_77973_b() == this.field_85183_f.func_77973_b() && ItemStack.func_77970_a(itemstack, this.field_85183_f) && (itemstack.func_77984_f() || itemstack.func_77960_j() == this.field_85183_f.func_77960_j());
+            flag = !net.minecraftforge.client.ForgeHooksClient.shouldCauseBlockBreakReset(this.field_85183_f, itemstack);
         }
 
         return p_178893_1_.equals(this.field_178895_c) && flag;
@@ -373,13 +384,34 @@
         }
         else
         {
+            net.minecraftforge.event.entity.player.PlayerInteractEvent.RightClickBlock event = net.minecraftforge.common.ForgeHooks
+                    .onRightClickBlock(p_187099_1_, p_187099_6_, p_187099_3_, p_187099_4_, net.minecraftforge.common.ForgeHooks.rayTraceEyeHitVec(p_187099_1_, func_78757_d() + 1));
+            if (event.isCanceled())
+            {
+                // Give the server a chance to fire event as well. That way server event is not dependant on client event.
+                this.field_78774_b.func_147297_a(new CPacketPlayerTryUseItemOnBlock(p_187099_3_, p_187099_4_, p_187099_6_, f, f1, f2));
+                return event.getCancellationResult();
+            }
+            EnumActionResult result = EnumActionResult.PASS;
+
             if (this.field_78779_k != GameType.SPECTATOR)
             {
+                EnumActionResult ret = itemstack.onItemUseFirst(p_187099_1_, p_187099_2_, p_187099_3_, p_187099_6_, p_187099_4_, f, f1, f2);
+                if (ret != EnumActionResult.PASS)
+                {
+                    // The server needs to process the item use as well. Otherwise onItemUseFirst won't ever be called on the server without causing weird bugs
+                    this.field_78774_b.func_147297_a(new CPacketPlayerTryUseItemOnBlock(p_187099_3_, p_187099_4_, p_187099_6_, f, f1, f2));
+                    return ret;
+                }
+
                 IBlockState iblockstate = p_187099_2_.func_180495_p(p_187099_3_);
+                boolean bypass = itemstack.func_190926_b() || itemstack.func_77973_b().doesSneakBypassUse(itemstack, p_187099_2_, p_187099_3_, p_187099_1_);
 
-                if ((!p_187099_1_.func_70093_af() || p_187099_1_.func_184614_ca().func_190926_b() && p_187099_1_.func_184592_cb().func_190926_b()) && iblockstate.func_177230_c().func_180639_a(p_187099_2_, p_187099_3_, iblockstate, p_187099_1_, p_187099_6_, p_187099_4_, f, f1, f2))
+                if ((!p_187099_1_.func_70093_af() || bypass || event.getUseBlock() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW))
                 {
-                    flag = true;
+                    if (event.getUseBlock() != net.minecraftforge.fml.common.eventhandler.Event.Result.DENY)
+                    flag = iblockstate.func_177230_c().func_180639_a(p_187099_2_, p_187099_3_, iblockstate, p_187099_1_, p_187099_6_, p_187099_4_, f, f1, f2);
+                    if (flag) result = EnumActionResult.SUCCESS;
                 }
 
                 if (!flag && itemstack.func_77973_b() instanceof ItemBlock)
@@ -395,7 +427,7 @@
 
             this.field_78774_b.func_147297_a(new CPacketPlayerTryUseItemOnBlock(p_187099_3_, p_187099_4_, p_187099_6_, f, f1, f2));
 
-            if (!flag && this.field_78779_k != GameType.SPECTATOR)
+            if (!flag && this.field_78779_k != GameType.SPECTATOR || event.getUseItem() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW)
             {
                 if (itemstack.func_190926_b())
                 {
@@ -421,14 +453,20 @@
                     {
                         int i = itemstack.func_77960_j();
                         int j = itemstack.func_190916_E();
+                        if (event.getUseItem() != net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) {
                         EnumActionResult enumactionresult = itemstack.func_179546_a(p_187099_1_, p_187099_2_, p_187099_3_, p_187099_6_, p_187099_4_, f, f1, f2);
                         itemstack.func_77964_b(i);
                         itemstack.func_190920_e(j);
                         return enumactionresult;
+                        } else return result;
                     }
                     else
                     {
-                        return itemstack.func_179546_a(p_187099_1_, p_187099_2_, p_187099_3_, p_187099_6_, p_187099_4_, f, f1, f2);
+                        ItemStack copyForUse = itemstack.func_77946_l();
+                        if (event.getUseItem() != net.minecraftforge.fml.common.eventhandler.Event.Result.DENY)
+                        result = itemstack.func_179546_a(p_187099_1_, p_187099_2_, p_187099_3_, p_187099_6_, p_187099_4_, f, f1, f2);
+                        if (itemstack.func_190926_b()) net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(p_187099_1_, copyForUse, p_187099_6_);
+                        return result;
                     }
                 }
             }
@@ -457,6 +495,8 @@
             }
             else
             {
+                EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onItemRightClickAction(p_187101_1_, p_187101_3_);
+                if (cancelResult != null) return cancelResult;
                 int i = itemstack.func_190916_E();
                 ActionResult<ItemStack> actionresult = itemstack.func_77957_a(p_187101_2_, p_187101_1_, p_187101_3_);
                 ItemStack itemstack1 = (ItemStack)actionresult.func_188398_b();
@@ -464,6 +504,10 @@
                 if (itemstack1 != itemstack || itemstack1.func_190916_E() != i)
                 {
                     p_187101_1_.func_184611_a(p_187101_3_, itemstack1);
+                    if (itemstack1.func_190926_b())
+                    {
+                        net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(p_187101_1_, itemstack, p_187101_3_);
+                    }
                 }
 
                 return actionresult.func_188397_a();
@@ -500,6 +544,9 @@
         this.func_78750_j();
         Vec3d vec3d = new Vec3d(p_187102_3_.field_72307_f.field_72450_a - p_187102_2_.field_70165_t, p_187102_3_.field_72307_f.field_72448_b - p_187102_2_.field_70163_u, p_187102_3_.field_72307_f.field_72449_c - p_187102_2_.field_70161_v);
         this.field_78774_b.func_147297_a(new CPacketUseEntity(p_187102_2_, p_187102_4_, vec3d));
+        if (this.field_78779_k == GameType.SPECTATOR) return EnumActionResult.PASS; // don't fire for spectators to match non-specific EntityInteract
+        EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntityAtAction(p_187102_1_, p_187102_2_, p_187102_3_, p_187102_4_);
+        if(cancelResult != null) return cancelResult;
         return this.field_78779_k == GameType.SPECTATOR ? EnumActionResult.PASS : p_187102_2_.func_184199_a(p_187102_1_, vec3d, p_187102_4_);
     }
 
