--- a/net/minecraft/client/multiplayer/PlayerControllerMP.java
+++ b/net/minecraft/client/multiplayer/PlayerControllerMP.java
@@ -105,6 +105,8 @@
          }
       }
 
+      if (field_78776_a.field_71439_g.func_184614_ca().onBlockStartBreak(p_187103_1_, field_78776_a.field_71439_g)) return false;
+
       World world = this.field_78776_a.field_71441_e;
       IBlockState iblockstate = world.func_180495_p(p_187103_1_);
       if (!this.field_78776_a.field_71439_g.func_184614_ca().func_77973_b().func_195938_a(iblockstate, world, p_187103_1_, this.field_78776_a.field_71439_g)) {
@@ -116,24 +118,23 @@
          } else if (iblockstate.func_196958_f()) {
             return false;
          } else {
-            block.func_176208_a(world, p_187103_1_, iblockstate, this.field_78776_a.field_71439_g);
             IFluidState ifluidstate = world.func_204610_c(p_187103_1_);
-            boolean flag = world.func_180501_a(p_187103_1_, ifluidstate.func_206883_i(), 11);
-            if (flag) {
-               block.func_176206_d(world, p_187103_1_, iblockstate);
-            }
 
             this.field_178895_c = new BlockPos(this.field_178895_c.func_177958_n(), -1, this.field_178895_c.func_177952_p());
             if (!this.field_78779_k.func_77145_d()) {
                ItemStack itemstack1 = this.field_78776_a.field_71439_g.func_184614_ca();
+               ItemStack copyBeforeUse = itemstack1.func_77946_l();
                if (!itemstack1.func_190926_b()) {
                   itemstack1.func_179548_a(world, iblockstate, p_187103_1_, this.field_78776_a.field_71439_g);
                   if (itemstack1.func_190926_b()) {
+                     net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this.field_78776_a.field_71439_g, copyBeforeUse, EnumHand.MAIN_HAND);
                      this.field_78776_a.field_71439_g.func_184611_a(EnumHand.MAIN_HAND, ItemStack.field_190927_a);
                   }
                }
             }
 
+            boolean flag = iblockstate.removedByPlayer(world, p_187103_1_, field_78776_a.field_71439_g, false, ifluidstate);
+            if (flag) block.func_176206_d(world, p_187103_1_, iblockstate);
             return flag;
          }
       }
@@ -164,21 +165,25 @@
          if (this.field_78779_k.func_77145_d()) {
             this.field_78776_a.func_193032_ao().func_193294_a(this.field_78776_a.field_71441_e, p_180511_1_, this.field_78776_a.field_71441_e.func_180495_p(p_180511_1_), 1.0F);
             this.field_78774_b.func_147297_a(new CPacketPlayerDigging(CPacketPlayerDigging.Action.START_DESTROY_BLOCK, p_180511_1_, p_180511_2_));
+            if (!net.minecraftforge.common.ForgeHooks.onLeftClickBlock(this.field_78776_a.field_71439_g, p_180511_1_, p_180511_2_, net.minecraftforge.common.ForgeHooks.rayTraceEyeHitVec(this.field_78776_a.field_71439_g, func_78757_d() + 1)).isCanceled())
             func_178891_a(this.field_78776_a, this, p_180511_1_, p_180511_2_);
             this.field_78781_i = 5;
          } else if (!this.field_78778_j || !this.func_178893_a(p_180511_1_)) {
             if (this.field_78778_j) {
                this.field_78774_b.func_147297_a(new CPacketPlayerDigging(CPacketPlayerDigging.Action.ABORT_DESTROY_BLOCK, this.field_178895_c, p_180511_2_));
             }
+            net.minecraftforge.event.entity.player.PlayerInteractEvent.LeftClickBlock event = net.minecraftforge.common.ForgeHooks.onLeftClickBlock(this.field_78776_a.field_71439_g, p_180511_1_, p_180511_2_, net.minecraftforge.common.ForgeHooks.rayTraceEyeHitVec(this.field_78776_a.field_71439_g, func_78757_d() + 1));
 
             IBlockState iblockstate = this.field_78776_a.field_71441_e.func_180495_p(p_180511_1_);
             this.field_78776_a.func_193032_ao().func_193294_a(this.field_78776_a.field_71441_e, p_180511_1_, iblockstate, 0.0F);
             this.field_78774_b.func_147297_a(new CPacketPlayerDigging(CPacketPlayerDigging.Action.START_DESTROY_BLOCK, p_180511_1_, p_180511_2_));
             boolean flag = !iblockstate.func_196958_f();
             if (flag && this.field_78770_f == 0.0F) {
+               if (event.getUseBlock() != net.minecraftforge.eventbus.api.Event.Result.DENY)
                iblockstate.func_196942_a(this.field_78776_a.field_71441_e, p_180511_1_, this.field_78776_a.field_71439_g);
             }
 
+            if (event.getUseItem() == net.minecraftforge.eventbus.api.Event.Result.DENY) return true;
             if (flag && iblockstate.func_185903_a(this.field_78776_a.field_71439_g, this.field_78776_a.field_71439_g.field_70170_p, p_180511_1_) >= 1.0F) {
                this.func_187103_a(p_180511_1_);
             } else {
@@ -227,7 +232,7 @@
          } else {
             this.field_78770_f += iblockstate.func_185903_a(this.field_78776_a.field_71439_g, this.field_78776_a.field_71439_g.field_70170_p, p_180512_1_);
             if (this.field_78780_h % 4.0F == 0.0F) {
-               SoundType soundtype = block.func_185467_w();
+               SoundType soundtype = iblockstate.getSoundType(field_78776_a.field_71441_e, p_180512_1_, field_78776_a.field_71439_g);
                this.field_78776_a.func_147118_V().func_147682_a(new SimpleSound(soundtype.func_185846_f(), SoundCategory.NEUTRAL, (soundtype.func_185843_a() + 1.0F) / 8.0F, soundtype.func_185847_b() * 0.5F, p_180512_1_));
             }
 
@@ -251,7 +256,8 @@
    }
 
    public float func_78757_d() {
-      return this.field_78779_k.func_77145_d() ? 5.0F : 4.5F;
+      float attrib = (float)field_78776_a.field_71439_g.func_110148_a(EntityPlayer.REACH_DISTANCE).func_111126_e();
+      return this.field_78779_k.func_77145_d() ? attrib : attrib - 0.5F;
    }
 
    public void func_78765_e() {
@@ -268,7 +274,7 @@
       ItemStack itemstack = this.field_78776_a.field_71439_g.func_184614_ca();
       boolean flag = this.field_85183_f.func_190926_b() && itemstack.func_190926_b();
       if (!this.field_85183_f.func_190926_b() && !itemstack.func_190926_b()) {
-         flag = itemstack.func_77973_b() == this.field_85183_f.func_77973_b() && ItemStack.func_77970_a(itemstack, this.field_85183_f) && (itemstack.func_77984_f() || itemstack.func_77952_i() == this.field_85183_f.func_77952_i());
+         flag = !this.field_85183_f.shouldCauseBlockBreakReset(itemstack);
       }
 
       return p_178893_1_.equals(this.field_178895_c) && flag;
@@ -292,17 +298,35 @@
          float f = (float)(p_187099_5_.field_72450_a - (double)p_187099_3_.func_177958_n());
          float f1 = (float)(p_187099_5_.field_72448_b - (double)p_187099_3_.func_177956_o());
          float f2 = (float)(p_187099_5_.field_72449_c - (double)p_187099_3_.func_177952_p());
+         net.minecraftforge.event.entity.player.PlayerInteractEvent.RightClickBlock event = net.minecraftforge.common.ForgeHooks
+                 .onRightClickBlock(p_187099_1_, p_187099_6_, p_187099_3_, p_187099_4_, net.minecraftforge.common.ForgeHooks.rayTraceEyeHitVec(p_187099_1_, func_78757_d() + 1));
+         if (event.isCanceled()) {
+            // Give the server a chance to fire event as well. That way server event is not dependant on client event.
+            this.field_78774_b.func_147297_a(new CPacketPlayerTryUseItemOnBlock(p_187099_3_, p_187099_4_, p_187099_6_, f, f1, f2));
+            return event.getCancellationResult();
+         }
          if (this.field_78779_k == GameType.SPECTATOR) {
             this.field_78774_b.func_147297_a(new CPacketPlayerTryUseItemOnBlock(p_187099_3_, p_187099_4_, p_187099_6_, f, f1, f2));
             return EnumActionResult.SUCCESS;
          } else {
-            boolean flag = !p_187099_1_.func_184614_ca().func_190926_b() || !p_187099_1_.func_184592_cb().func_190926_b();
+            EnumActionResult ret = itemstack.onItemUseFirst(new ItemUseContext(p_187099_1_, p_187099_1_.func_184586_b(p_187099_6_), p_187099_3_, p_187099_4_, f, f1, f2));
+            if (ret != EnumActionResult.PASS) {
+               // The server needs to process the item use as well. Otherwise onItemUseFirst won't ever be called on the server without causing weird bugs
+               this.field_78774_b.func_147297_a(new CPacketPlayerTryUseItemOnBlock(p_187099_3_, p_187099_4_, p_187099_6_, f, f1, f2));
+               return ret;
+            }
+            boolean flag = !(p_187099_1_.func_184614_ca().doesSneakBypassUse(p_187099_2_, p_187099_3_, p_187099_1_) && p_187099_1_.func_184592_cb().doesSneakBypassUse(p_187099_2_, p_187099_3_, p_187099_1_));
             boolean flag1 = p_187099_1_.func_70093_af() && flag;
-            if (!flag1 && p_187099_2_.func_180495_p(p_187099_3_).func_196943_a(p_187099_2_, p_187099_3_, p_187099_1_, p_187099_6_, p_187099_4_, f, f1, f2)) {
+            if (!flag1 || event.getUseBlock() == net.minecraftforge.eventbus.api.Event.Result.ALLOW) {
+               if (event.getUseBlock() != net.minecraftforge.eventbus.api.Event.Result.DENY) {
+                 if (p_187099_2_.func_180495_p(p_187099_3_).func_196943_a(p_187099_2_, p_187099_3_, p_187099_1_, p_187099_6_, p_187099_4_, f, f1, f2)) {
+                    this.field_78774_b.func_147297_a(new CPacketPlayerTryUseItemOnBlock(p_187099_3_, p_187099_4_, p_187099_6_, f, f1, f2));
+                    return EnumActionResult.SUCCESS;
+                 }
+               }
+            }
+            if (event.getUseItem() != net.minecraftforge.eventbus.api.Event.Result.DENY) {
                this.field_78774_b.func_147297_a(new CPacketPlayerTryUseItemOnBlock(p_187099_3_, p_187099_4_, p_187099_6_, f, f1, f2));
-               return EnumActionResult.SUCCESS;
-            } else {
-               this.field_78774_b.func_147297_a(new CPacketPlayerTryUseItemOnBlock(p_187099_3_, p_187099_4_, p_187099_6_, f, f1, f2));
                if (!itemstack.func_190926_b() && !p_187099_1_.func_184811_cZ().func_185141_a(itemstack.func_77973_b())) {
                   ItemUseContext itemusecontext = new ItemUseContext(p_187099_1_, p_187099_1_.func_184586_b(p_187099_6_), p_187099_3_, p_187099_4_, f, f1, f2);
                   EnumActionResult enumactionresult;
@@ -311,7 +335,9 @@
                      enumactionresult = itemstack.func_196084_a(itemusecontext);
                      itemstack.func_190920_e(i);
                   } else {
+                     ItemStack copyForUse = itemstack.func_77946_l();
                      enumactionresult = itemstack.func_196084_a(itemusecontext);
+                     if (itemstack.func_190926_b()) net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(p_187099_1_, copyForUse, p_187099_6_);
                   }
 
                   return enumactionresult;
@@ -319,6 +345,7 @@
                   return EnumActionResult.PASS;
                }
             }
+            return EnumActionResult.PASS;
          }
       }
    }
@@ -333,11 +360,14 @@
          if (p_187101_1_.func_184811_cZ().func_185141_a(itemstack.func_77973_b())) {
             return EnumActionResult.PASS;
          } else {
+            EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onItemRightClick(p_187101_1_, p_187101_3_);
+            if (cancelResult != null) return cancelResult;
             int i = itemstack.func_190916_E();
             ActionResult<ItemStack> actionresult = itemstack.func_77957_a(p_187101_2_, p_187101_1_, p_187101_3_);
             ItemStack itemstack1 = actionresult.func_188398_b();
             if (itemstack1 != itemstack || itemstack1.func_190916_E() != i) {
                p_187101_1_.func_184611_a(p_187101_3_, itemstack1);
+               if (itemstack1.func_190926_b()) net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(p_187101_1_, itemstack, p_187101_3_);
             }
 
             return actionresult.func_188397_a();
@@ -362,6 +392,9 @@
    public EnumActionResult func_187097_a(EntityPlayer p_187097_1_, Entity p_187097_2_, EnumHand p_187097_3_) {
       this.func_78750_j();
       this.field_78774_b.func_147297_a(new CPacketUseEntity(p_187097_2_, p_187097_3_));
+      if (this.field_78779_k == GameType.SPECTATOR) return EnumActionResult.PASS; // don't fire for spectators to match non-specific EntityInteract
+      EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntity(p_187097_1_, p_187097_2_, p_187097_3_);
+      if(cancelResult != null) return cancelResult;
       return this.field_78779_k == GameType.SPECTATOR ? EnumActionResult.PASS : p_187097_1_.func_190775_a(p_187097_2_, p_187097_3_);
    }
 
@@ -369,6 +402,9 @@
       this.func_78750_j();
       Vec3d vec3d = new Vec3d(p_187102_3_.field_72307_f.field_72450_a - p_187102_2_.field_70165_t, p_187102_3_.field_72307_f.field_72448_b - p_187102_2_.field_70163_u, p_187102_3_.field_72307_f.field_72449_c - p_187102_2_.field_70161_v);
       this.field_78774_b.func_147297_a(new CPacketUseEntity(p_187102_2_, p_187102_4_, vec3d));
+      if (this.field_78779_k == GameType.SPECTATOR) return EnumActionResult.PASS; // don't fire for spectators to match non-specific EntityInteract
+      EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntityAt(p_187102_1_, p_187102_2_, p_187102_3_, p_187102_4_);
+      if(cancelResult != null) return cancelResult;
       return this.field_78779_k == GameType.SPECTATOR ? EnumActionResult.PASS : p_187102_2_.func_184199_a(p_187102_1_, vec3d, p_187102_4_);
    }
 
