--- ../src-base/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java
+++ ../src-work/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java
@@ -28,6 +28,13 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldSettings;
 
+import net.minecraftforge.common.MinecraftForge;
+import cpw.mods.fml.common.eventhandler.Event;
+import net.minecraftforge.event.ForgeEventFactory;
+import net.minecraftforge.event.entity.player.PlayerDestroyItemEvent;
+import net.minecraftforge.event.entity.player.PlayerInteractEvent;
+import net.minecraftforge.event.entity.player.PlayerInteractEvent.Action;
+
 @SideOnly(Side.CLIENT)
 public class PlayerControllerMP
 {
@@ -88,6 +95,12 @@
 
     public boolean func_78751_a(int p_78751_1_, int p_78751_2_, int p_78751_3_, int p_78751_4_)
     {
+        ItemStack stack = field_78776_a.field_71439_g.func_71045_bC();
+        if (stack != null && stack.func_77973_b() != null && stack.func_77973_b().onBlockStartBreak(stack, p_78751_1_, p_78751_2_, p_78751_3_, field_78776_a.field_71439_g))
+        {
+            return false;
+        }
+
         if (this.field_78779_k.func_82752_c() && !this.field_78776_a.field_71439_g.func_82246_f(p_78751_1_, p_78751_2_, p_78751_3_))
         {
             return false;
@@ -109,7 +122,7 @@
             {
                 worldclient.func_72926_e(2001, p_78751_1_, p_78751_2_, p_78751_3_, Block.func_149682_b(block) + (worldclient.func_72805_g(p_78751_1_, p_78751_2_, p_78751_3_) << 12));
                 int i1 = worldclient.func_72805_g(p_78751_1_, p_78751_2_, p_78751_3_);
-                boolean flag = worldclient.func_147468_f(p_78751_1_, p_78751_2_, p_78751_3_);
+                boolean flag = block.removedByPlayer(worldclient, field_78776_a.field_71439_g, p_78751_1_, p_78751_2_, p_78751_3_);
 
                 if (flag)
                 {
@@ -142,6 +155,12 @@
     {
         if (!this.field_78779_k.func_82752_c() || this.field_78776_a.field_71439_g.func_82246_f(p_78743_1_, p_78743_2_, p_78743_3_))
         {
+            PlayerInteractEvent event = ForgeEventFactory.onPlayerInteract(this.field_78776_a.field_71439_g, Action.LEFT_CLICK_BLOCK, p_78743_1_, p_78743_2_, p_78743_3_, p_78743_4_, this.field_78776_a.field_71441_e);
+            if (event.isCanceled())
+            {
+                return;
+            }
+
             if (this.field_78779_k.func_77145_d())
             {
                 this.field_78774_b.func_147297_a(new C07PacketPlayerDigging(0, p_78743_1_, p_78743_2_, p_78743_3_, p_78743_4_));
@@ -157,13 +176,19 @@
 
                 this.field_78774_b.func_147297_a(new C07PacketPlayerDigging(0, p_78743_1_, p_78743_2_, p_78743_3_, p_78743_4_));
                 Block block = this.field_78776_a.field_71441_e.func_147439_a(p_78743_1_, p_78743_2_, p_78743_3_);
-                boolean flag = block.func_149688_o() != Material.field_151579_a;
+                float blockHardness = 1.0F;
+                boolean flag = !block.isAir(this.field_78776_a.field_71441_e, p_78743_1_, p_78743_2_, p_78743_3_);
 
-                if (flag && this.field_78770_f == 0.0F)
+                if (flag && this.field_78770_f == 0.0F && event.useBlock != Event.Result.DENY)
                 {
                     block.func_149699_a(this.field_78776_a.field_71441_e, p_78743_1_, p_78743_2_, p_78743_3_, this.field_78776_a.field_71439_g);
                 }
 
+                if (event.useItem == Event.Result.DENY)
+                {
+                    return;
+                }
+
                 if (flag && block.func_149737_a(this.field_78776_a.field_71439_g, this.field_78776_a.field_71439_g.field_70170_p, p_78743_1_, p_78743_2_, p_78743_3_) >= 1.0F)
                 {
                     this.func_78751_a(p_78743_1_, p_78743_2_, p_78743_3_, p_78743_4_);
@@ -298,17 +323,30 @@
 
     public boolean func_78760_a(EntityPlayer p_78760_1_, World p_78760_2_, ItemStack p_78760_3_, int p_78760_4_, int p_78760_5_, int p_78760_6_, int p_78760_7_, Vec3 p_78760_8_)
     {
+        PlayerInteractEvent event = ForgeEventFactory.onPlayerInteract(p_78760_1_, Action.RIGHT_CLICK_BLOCK, p_78760_4_, p_78760_5_, p_78760_6_, p_78760_7_, p_78760_2_);
+        if (event.isCanceled())
+        {
+            return false;
+        }
+
         this.func_78750_j();
         float f = (float)p_78760_8_.field_72450_a - (float)p_78760_4_;
         float f1 = (float)p_78760_8_.field_72448_b - (float)p_78760_5_;
         float f2 = (float)p_78760_8_.field_72449_c - (float)p_78760_6_;
         boolean flag = false;
 
-        if ((!p_78760_1_.func_70093_af() || p_78760_1_.func_70694_bm() == null) && p_78760_2_.func_147439_a(p_78760_4_, p_78760_5_, p_78760_6_).func_149727_a(p_78760_2_, p_78760_4_, p_78760_5_, p_78760_6_, p_78760_1_, p_78760_7_, f, f1, f2))
+        if (p_78760_3_ != null &&
+            p_78760_3_.func_77973_b() != null &&
+            p_78760_3_.func_77973_b().onItemUseFirst(p_78760_3_, p_78760_1_, p_78760_2_, p_78760_4_, p_78760_5_, p_78760_6_, p_78760_7_, f, f1, f2))
         {
-            flag = true;
+                return true;
         }
 
+        if (event.useBlock != Event.Result.DENY && (!p_78760_1_.func_70093_af() || p_78760_1_.func_70694_bm() == null || p_78760_1_.func_70694_bm().func_77973_b().doesSneakBypassUse(p_78760_2_, p_78760_4_, p_78760_5_, p_78760_6_, p_78760_1_)))
+        {
+            flag = p_78760_2_.func_147439_a(p_78760_4_, p_78760_5_, p_78760_6_).func_149727_a(p_78760_2_, p_78760_4_, p_78760_5_, p_78760_6_, p_78760_1_, p_78760_7_, f, f1, f2);
+        }
+
         if (!flag && p_78760_3_ != null && p_78760_3_.func_77973_b() instanceof ItemBlock)
         {
             ItemBlock itemblock = (ItemBlock)p_78760_3_.func_77973_b();
@@ -325,7 +363,7 @@
         {
             return true;
         }
-        else if (p_78760_3_ == null)
+        else if (p_78760_3_ == null || event.useItem == Event.Result.DENY)
         {
             return false;
         }
@@ -340,12 +378,26 @@
         }
         else
         {
-            return p_78760_3_.func_77943_a(p_78760_1_, p_78760_2_, p_78760_4_, p_78760_5_, p_78760_6_, p_78760_7_, f, f1, f2);
+            if (!p_78760_3_.func_77943_a(p_78760_1_, p_78760_2_, p_78760_4_, p_78760_5_, p_78760_6_, p_78760_7_, f, f1, f2))
+            {
+                return false;
+            }
+            if (p_78760_3_.field_77994_a <= 0)
+            {
+                MinecraftForge.EVENT_BUS.post(new PlayerDestroyItemEvent(p_78760_1_, p_78760_3_));
+            }
+            return true;
         }
     }
 
     public boolean func_78769_a(EntityPlayer p_78769_1_, World p_78769_2_, ItemStack p_78769_3_)
     {
+        PlayerInteractEvent event = ForgeEventFactory.onPlayerInteract(p_78769_1_, PlayerInteractEvent.Action.RIGHT_CLICK_AIR, 0, 0, 0, -1, p_78769_2_);
+        if (event.isCanceled() || event.useItem == Event.Result.DENY)
+        {
+            return false;
+        }
+
         this.func_78750_j();
         this.field_78774_b.func_147297_a(new C08PacketPlayerBlockPlacement(-1, -1, -1, 255, p_78769_1_.field_71071_by.func_70448_g(), 0.0F, 0.0F, 0.0F));
         int i = p_78769_3_.field_77994_a;
@@ -359,9 +411,10 @@
         {
             p_78769_1_.field_71071_by.field_70462_a[p_78769_1_.field_71071_by.field_70461_c] = itemstack1;
 
-            if (itemstack1.field_77994_a == 0)
+            if (itemstack1.field_77994_a <= 0)
             {
                 p_78769_1_.field_71071_by.field_70462_a[p_78769_1_.field_71071_by.field_70461_c] = null;
+                MinecraftForge.EVENT_BUS.post(new PlayerDestroyItemEvent(p_78769_1_, itemstack1));
             }
 
             return true;
