--- a/net/minecraft/client/multiplayer/PlayerController.java
+++ b/net/minecraft/client/multiplayer/PlayerController.java
@@ -93,6 +_,7 @@
    }
 
    public boolean destroyBlock(BlockPos p_187103_1_) {
+      if (minecraft.player.getMainHandItem().onBlockStartBreak(p_187103_1_, minecraft.player)) return false;
       if (this.minecraft.player.blockActionRestricted(this.minecraft.level, p_187103_1_, this.localPlayerMode)) {
          return false;
       } else {
@@ -104,12 +_,11 @@
             Block block = blockstate.getBlock();
             if ((block instanceof CommandBlockBlock || block instanceof StructureBlock || block instanceof JigsawBlock) && !this.minecraft.player.canUseGameMasterBlocks()) {
                return false;
-            } else if (blockstate.isAir()) {
+            } else if (blockstate.isAir(world, p_187103_1_)) {
                return false;
             } else {
-               block.playerWillDestroy(world, p_187103_1_, blockstate, this.minecraft.player);
                FluidState fluidstate = world.getFluidState(p_187103_1_);
-               boolean flag = world.setBlock(p_187103_1_, fluidstate.createLegacyBlock(), 11);
+               boolean flag = blockstate.removedByPlayer(world, p_187103_1_, minecraft.player, false, fluidstate);
                if (flag) {
                   block.destroy(world, p_187103_1_, blockstate);
                }
@@ -130,21 +_,25 @@
             BlockState blockstate = this.minecraft.level.getBlockState(p_180511_1_);
             this.minecraft.getTutorial().onDestroyBlock(this.minecraft.level, p_180511_1_, blockstate, 1.0F);
             this.sendBlockAction(CPlayerDiggingPacket.Action.START_DESTROY_BLOCK, p_180511_1_, p_180511_2_);
+            if (!net.minecraftforge.common.ForgeHooks.onLeftClickBlock(this.minecraft.player, p_180511_1_, p_180511_2_).isCanceled())
             this.destroyBlock(p_180511_1_);
             this.destroyDelay = 5;
          } else if (!this.isDestroying || !this.sameDestroyTarget(p_180511_1_)) {
             if (this.isDestroying) {
                this.sendBlockAction(CPlayerDiggingPacket.Action.ABORT_DESTROY_BLOCK, this.destroyBlockPos, p_180511_2_);
             }
+            net.minecraftforge.event.entity.player.PlayerInteractEvent.LeftClickBlock event = net.minecraftforge.common.ForgeHooks.onLeftClickBlock(this.minecraft.player, p_180511_1_, p_180511_2_);
 
             BlockState blockstate1 = this.minecraft.level.getBlockState(p_180511_1_);
             this.minecraft.getTutorial().onDestroyBlock(this.minecraft.level, p_180511_1_, blockstate1, 0.0F);
             this.sendBlockAction(CPlayerDiggingPacket.Action.START_DESTROY_BLOCK, p_180511_1_, p_180511_2_);
-            boolean flag = !blockstate1.isAir();
+            boolean flag = !blockstate1.isAir(this.minecraft.level, p_180511_1_);
             if (flag && this.destroyProgress == 0.0F) {
+               if (event.getUseBlock() != net.minecraftforge.eventbus.api.Event.Result.DENY)
                blockstate1.attack(this.minecraft.level, p_180511_1_, this.minecraft.player);
             }
 
+            if (event.getUseItem() == net.minecraftforge.eventbus.api.Event.Result.DENY) return true;
             if (flag && blockstate1.getDestroyProgress(this.minecraft.player, this.minecraft.player.level, p_180511_1_) >= 1.0F) {
                this.destroyBlock(p_180511_1_);
             } else {
@@ -184,22 +_,24 @@
          BlockState blockstate1 = this.minecraft.level.getBlockState(p_180512_1_);
          this.minecraft.getTutorial().onDestroyBlock(this.minecraft.level, p_180512_1_, blockstate1, 1.0F);
          this.sendBlockAction(CPlayerDiggingPacket.Action.START_DESTROY_BLOCK, p_180512_1_, p_180512_2_);
+         if (!net.minecraftforge.common.ForgeHooks.onLeftClickBlock(this.minecraft.player, p_180512_1_, p_180512_2_).isCanceled())
          this.destroyBlock(p_180512_1_);
          return true;
       } else if (this.sameDestroyTarget(p_180512_1_)) {
          BlockState blockstate = this.minecraft.level.getBlockState(p_180512_1_);
-         if (blockstate.isAir()) {
+         if (blockstate.isAir(this.minecraft.level, p_180512_1_)) {
             this.isDestroying = false;
             return false;
          } else {
             this.destroyProgress += blockstate.getDestroyProgress(this.minecraft.player, this.minecraft.player.level, p_180512_1_);
             if (this.destroyTicks % 4.0F == 0.0F) {
-               SoundType soundtype = blockstate.getSoundType();
+               SoundType soundtype = blockstate.getSoundType(this.minecraft.level, p_180512_1_, this.minecraft.player);
                this.minecraft.getSoundManager().play(new SimpleSound(soundtype.getHitSound(), SoundCategory.BLOCKS, (soundtype.getVolume() + 1.0F) / 8.0F, soundtype.getPitch() * 0.5F, p_180512_1_));
             }
 
             ++this.destroyTicks;
             this.minecraft.getTutorial().onDestroyBlock(this.minecraft.level, p_180512_1_, blockstate, MathHelper.clamp(this.destroyProgress, 0.0F, 1.0F));
+            if (net.minecraftforge.common.ForgeHooks.onLeftClickBlock(this.minecraft.player, p_180512_1_, p_180512_2_).getUseItem() == net.minecraftforge.eventbus.api.Event.Result.DENY) return true;
             if (this.destroyProgress >= 1.0F) {
                this.isDestroying = false;
                this.sendBlockAction(CPlayerDiggingPacket.Action.STOP_DESTROY_BLOCK, p_180512_1_, p_180512_2_);
@@ -218,7 +_,8 @@
    }
 
    public float getPickRange() {
-      return this.localPlayerMode.isCreative() ? 5.0F : 4.5F;
+      float attrib = (float)minecraft.player.getAttribute(net.minecraftforge.common.ForgeMod.REACH_DISTANCE.get()).getValue();
+      return this.localPlayerMode.isCreative() ? attrib : attrib - 0.5F;
    }
 
    public void tick() {
@@ -235,7 +_,7 @@
       ItemStack itemstack = this.minecraft.player.getMainHandItem();
       boolean flag = this.destroyingItem.isEmpty() && itemstack.isEmpty();
       if (!this.destroyingItem.isEmpty() && !itemstack.isEmpty()) {
-         flag = itemstack.getItem() == this.destroyingItem.getItem() && ItemStack.tagMatches(itemstack, this.destroyingItem) && (itemstack.isDamageableItem() || itemstack.getDamageValue() == this.destroyingItem.getDamageValue());
+         flag = !this.destroyingItem.shouldCauseBlockBreakReset(itemstack);
       }
 
       return p_178893_1_.equals(this.destroyBlockPos) && flag;
@@ -257,13 +_,27 @@
          return ActionResultType.FAIL;
       } else {
          ItemStack itemstack = p_217292_1_.getItemInHand(p_217292_3_);
+         net.minecraftforge.event.entity.player.PlayerInteractEvent.RightClickBlock event = net.minecraftforge.common.ForgeHooks
+                 .onRightClickBlock(p_217292_1_, p_217292_3_, blockpos, p_217292_4_);
+         if (event.isCanceled()) {
+            this.connection.send(new CPlayerTryUseItemOnBlockPacket(p_217292_3_, p_217292_4_));
+            return event.getCancellationResult();
+         }
          if (this.localPlayerMode == GameType.SPECTATOR) {
             this.connection.send(new CPlayerTryUseItemOnBlockPacket(p_217292_3_, p_217292_4_));
             return ActionResultType.SUCCESS;
          } else {
-            boolean flag = !p_217292_1_.getMainHandItem().isEmpty() || !p_217292_1_.getOffhandItem().isEmpty();
+            ItemUseContext itemusecontext = new ItemUseContext(p_217292_1_, p_217292_3_, p_217292_4_);
+            if (event.getUseItem() != net.minecraftforge.eventbus.api.Event.Result.DENY) {
+               ActionResultType result = itemstack.onItemUseFirst(itemusecontext);
+               if (result != ActionResultType.PASS) {
+                  this.connection.send(new CPlayerTryUseItemOnBlockPacket(p_217292_3_, p_217292_4_));
+                  return result;
+               }
+            }
+            boolean flag = !p_217292_1_.getMainHandItem().doesSneakBypassUse(p_217292_2_,blockpos,p_217292_1_) || !p_217292_1_.getOffhandItem().doesSneakBypassUse(p_217292_2_,blockpos,p_217292_1_);
             boolean flag1 = p_217292_1_.isSecondaryUseActive() && flag;
-            if (!flag1) {
+            if (event.getUseBlock() == net.minecraftforge.eventbus.api.Event.Result.ALLOW || (event.getUseBlock() != net.minecraftforge.eventbus.api.Event.Result.DENY && !flag1)) {
                ActionResultType actionresulttype = p_217292_2_.getBlockState(blockpos).use(p_217292_2_, p_217292_1_, p_217292_3_, p_217292_4_);
                if (actionresulttype.consumesAction()) {
                   this.connection.send(new CPlayerTryUseItemOnBlockPacket(p_217292_3_, p_217292_4_));
@@ -272,8 +_,8 @@
             }
 
             this.connection.send(new CPlayerTryUseItemOnBlockPacket(p_217292_3_, p_217292_4_));
-            if (!itemstack.isEmpty() && !p_217292_1_.getCooldowns().isOnCooldown(itemstack.getItem())) {
-               ItemUseContext itemusecontext = new ItemUseContext(p_217292_1_, p_217292_3_, p_217292_4_);
+            if (event.getUseItem() == net.minecraftforge.eventbus.api.Event.Result.DENY) return ActionResultType.PASS;
+            if (event.getUseItem() == net.minecraftforge.eventbus.api.Event.Result.ALLOW || (!itemstack.isEmpty() && !p_217292_1_.getCooldowns().isOnCooldown(itemstack.getItem()))) {
                ActionResultType actionresulttype1;
                if (this.localPlayerMode.isCreative()) {
                   int i = itemstack.getCount();
@@ -301,11 +_,14 @@
          if (p_187101_1_.getCooldowns().isOnCooldown(itemstack.getItem())) {
             return ActionResultType.PASS;
          } else {
+            ActionResultType cancelResult = net.minecraftforge.common.ForgeHooks.onItemRightClick(p_187101_1_, p_187101_3_);
+            if (cancelResult != null) return cancelResult;
             int i = itemstack.getCount();
             ActionResult<ItemStack> actionresult = itemstack.use(p_187101_2_, p_187101_1_, p_187101_3_);
             ItemStack itemstack1 = actionresult.getObject();
             if (itemstack1 != itemstack) {
                p_187101_1_.setItemInHand(p_187101_3_, itemstack1);
+               if (itemstack1.isEmpty()) net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(p_187101_1_, itemstack, p_187101_3_);
             }
 
             return actionresult.getResult();
@@ -334,6 +_,9 @@
    public ActionResultType interact(PlayerEntity p_187097_1_, Entity p_187097_2_, Hand p_187097_3_) {
       this.ensureHasSentCarriedItem();
       this.connection.send(new CUseEntityPacket(p_187097_2_, p_187097_3_, p_187097_1_.isShiftKeyDown()));
+      if (this.localPlayerMode == GameType.SPECTATOR) return ActionResultType.PASS; // don't fire for spectators to match non-specific EntityInteract
+      ActionResultType cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntity(p_187097_1_, p_187097_2_, p_187097_3_);
+      if(cancelResult != null) return cancelResult;
       return this.localPlayerMode == GameType.SPECTATOR ? ActionResultType.PASS : p_187097_1_.interactOn(p_187097_2_, p_187097_3_);
    }
 
@@ -341,6 +_,9 @@
       this.ensureHasSentCarriedItem();
       Vector3d vector3d = p_187102_3_.getLocation().subtract(p_187102_2_.getX(), p_187102_2_.getY(), p_187102_2_.getZ());
       this.connection.send(new CUseEntityPacket(p_187102_2_, p_187102_4_, vector3d, p_187102_1_.isShiftKeyDown()));
+      if (this.localPlayerMode == GameType.SPECTATOR) return ActionResultType.PASS; // don't fire for spectators to match non-specific EntityInteract
+      ActionResultType cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntityAt(p_187102_1_, p_187102_2_, p_187102_3_, p_187102_4_);
+      if(cancelResult != null) return cancelResult;
       return this.localPlayerMode == GameType.SPECTATOR ? ActionResultType.PASS : p_187102_2_.interactAt(p_187102_1_, vector3d, p_187102_4_);
    }
 
