--- a/net/minecraft/client/multiplayer/PlayerController.java
+++ b/net/minecraft/client/multiplayer/PlayerController.java
@@ -103,7 +103,9 @@
             Block block = blockstate.func_177230_c();
             if ((block instanceof CommandBlockBlock || block instanceof StructureBlock || block instanceof JigsawBlock) && !this.field_78776_a.field_71439_g.func_195070_dx()) {
                return false;
-            } else if (blockstate.func_196958_f()) {
+            } else if (blockstate.isAir(world, p_187103_1_)) {
+      if (field_78776_a.field_71439_g.func_184614_ca().onBlockStartBreak(p_187103_1_, field_78776_a.field_71439_g)) return false;
+
                return false;
             } else {
                block.func_176208_a(world, p_187103_1_, blockstate, this.field_78776_a.field_71439_g);
@@ -187,13 +189,13 @@
          return true;
       } else if (this.func_178893_a(p_180512_1_)) {
          BlockState blockstate = this.field_78776_a.field_71441_e.func_180495_p(p_180512_1_);
-         if (blockstate.func_196958_f()) {
+         if (blockstate.isAir(this.field_78776_a.field_71441_e, p_180512_1_)) {
             this.field_78778_j = false;
             return false;
          } else {
             this.field_78770_f += blockstate.func_185903_a(this.field_78776_a.field_71439_g, this.field_78776_a.field_71439_g.field_70170_p, p_180512_1_);
             if (this.field_78780_h % 4.0F == 0.0F) {
-               SoundType soundtype = blockstate.func_215695_r();
+               SoundType soundtype = blockstate.getSoundType(this.field_78776_a.field_71441_e, p_180512_1_, this.field_78776_a.field_71439_g);
                this.field_78776_a.func_147118_V().func_147682_a(new SimpleSound(soundtype.func_185846_f(), SoundCategory.NEUTRAL, (soundtype.func_185843_a() + 1.0F) / 8.0F, soundtype.func_185847_b() * 0.5F, p_180512_1_));
             }
 
@@ -217,7 +219,8 @@
    }
 
    public float func_78757_d() {
-      return this.field_78779_k.func_77145_d() ? 5.0F : 4.5F;
+      float attrib = (float)field_78776_a.field_71439_g.func_110148_a(PlayerEntity.REACH_DISTANCE).func_111126_e();
+      return this.field_78779_k.func_77145_d() ? attrib : attrib - 0.5F;
    }
 
    public void func_78765_e() {
@@ -234,7 +237,7 @@
       ItemStack itemstack = this.field_78776_a.field_71439_g.func_184614_ca();
       boolean flag = this.field_85183_f.func_190926_b() && itemstack.func_190926_b();
       if (!this.field_85183_f.func_190926_b() && !itemstack.func_190926_b()) {
-         flag = itemstack.func_77973_b() == this.field_85183_f.func_77973_b() && ItemStack.func_77970_a(itemstack, this.field_85183_f) && (itemstack.func_77984_f() || itemstack.func_77952_i() == this.field_85183_f.func_77952_i());
+         flag = !this.field_85183_f.shouldCauseBlockBreakReset(itemstack);
       }
 
       return p_178893_1_.equals(this.field_178895_c) && flag;
@@ -298,11 +301,14 @@
          if (p_187101_1_.func_184811_cZ().func_185141_a(itemstack.func_77973_b())) {
             return ActionResultType.PASS;
          } else {
+            ActionResultType cancelResult = net.minecraftforge.common.ForgeHooks.onItemRightClick(p_187101_1_, p_187101_3_);
+            if (cancelResult != null) return cancelResult;
             int i = itemstack.func_190916_E();
             ActionResult<ItemStack> actionresult = itemstack.func_77957_a(p_187101_2_, p_187101_1_, p_187101_3_);
             ItemStack itemstack1 = actionresult.func_188398_b();
             if (itemstack1 != itemstack || itemstack1.func_190916_E() != i) {
                p_187101_1_.func_184611_a(p_187101_3_, itemstack1);
+               if (itemstack1.func_190926_b()) net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(p_187101_1_, itemstack, p_187101_3_);
             }
 
             return actionresult.func_188397_a();
@@ -327,6 +333,9 @@
    public ActionResultType func_187097_a(PlayerEntity p_187097_1_, Entity p_187097_2_, Hand p_187097_3_) {
       this.func_78750_j();
       this.field_78774_b.func_147297_a(new CUseEntityPacket(p_187097_2_, p_187097_3_));
+      if (this.field_78779_k == GameType.SPECTATOR) return ActionResultType.PASS; // don't fire for spectators to match non-specific EntityInteract
+      ActionResultType cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntity(p_187097_1_, p_187097_2_, p_187097_3_);
+      if(cancelResult != null) return cancelResult;
       return this.field_78779_k == GameType.SPECTATOR ? ActionResultType.PASS : p_187097_1_.func_190775_a(p_187097_2_, p_187097_3_);
    }
 
@@ -334,6 +343,9 @@
       this.func_78750_j();
       Vec3d vec3d = p_187102_3_.func_216347_e().func_178786_a(p_187102_2_.field_70165_t, p_187102_2_.field_70163_u, p_187102_2_.field_70161_v);
       this.field_78774_b.func_147297_a(new CUseEntityPacket(p_187102_2_, p_187102_4_, vec3d));
+      if (this.field_78779_k == GameType.SPECTATOR) return ActionResultType.PASS; // don't fire for spectators to match non-specific EntityInteract
+      ActionResultType cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntityAt(p_187102_1_, p_187102_2_, p_187102_3_, p_187102_4_);
+      if(cancelResult != null) return cancelResult;
       return this.field_78779_k == GameType.SPECTATOR ? ActionResultType.PASS : p_187102_2_.func_184199_a(p_187102_1_, vec3d, p_187102_4_);
    }
 
