--- a/net/minecraft/client/multiplayer/ServerData.java
+++ b/net/minecraft/client/multiplayer/ServerData.java
@@ -32,6 +_,7 @@
    private byte[] f_271511_;
    private boolean f_105373_;
    private boolean f_242950_;
+   public net.minecraftforge.client.ExtendedServerListData forgeData = null;
 
    public ServerData(String p_105375_, String p_105376_, boolean p_105377_) {
       this.f_105362_ = p_105375_;
