--- a/net/minecraft/client/multiplayer/ServerData.java
+++ b/net/minecraft/client/multiplayer/ServerData.java
@@ -32,6 +_,7 @@
    private byte[] iconBytes;
    private ServerData.Type type;
    private boolean enforcesSecureChat;
+   public net.minecraftforge.client.ExtendedServerListData forgeData = null;
 
    public ServerData(String p_105375_, String p_105376_, ServerData.Type p_297678_) {
       this.name = p_105375_;
