--- a/net/minecraft/client/multiplayer/TagCollector.java
+++ b/net/minecraft/client/multiplayer/TagCollector.java
@@ -23,7 +_,7 @@
     }
 
     private static void refreshCommonTagDependentData() {
-        CreativeModeTabs.searchTab().rebuildSearchTree();
+        CreativeModeTabs.allTabs().stream().filter(tab -> tab.hasSearchBar()).forEach(tab -> tab.rebuildSearchTree());
     }
 
     private static void refreshBuiltInTagDependentData() {
