--- ../src-base/minecraft/net/minecraft/client/multiplayer/WorldClient.java
+++ ../src-work/minecraft/net/minecraft/client/multiplayer/WorldClient.java
@@ -34,6 +34,9 @@
 import net.minecraft.world.chunk.IChunkProvider;
 import net.minecraft.world.storage.SaveHandlerMP;
 
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.world.WorldEvent;
+
 @SideOnly(Side.CLIENT)
 public class WorldClient extends World
 {
@@ -56,8 +59,11 @@
         super(new SaveHandlerMP(), "MpServer", WorldProvider.getProviderForDimension(p_i45063_3_), p_i45063_2_, p_i45063_5_);
         this.sendQueue = p_i45063_1_;
         this.difficultySetting = p_i45063_4_;
-        this.setSpawnLocation(8, 64, 8);
         this.mapStorage = p_i45063_1_.mapStorageOrigin;
+        this.isRemote = true;
+        this.finishSetup();
+        this.setSpawnLocation(8, 64, 8);
+        MinecraftForge.EVENT_BUS.post(new WorldEvent.Load(this));
     }
 
     // JAVADOC METHOD $$ func_72835_b
@@ -275,6 +281,12 @@
     // JAVADOC METHOD $$ func_72979_l
     protected void updateWeather()
     {
+        super.updateWeather();
+    }
+
+    @Override
+    public void updateWeatherBody()
+    {
         if (!this.provider.hasNoSky)
         {
             ;
