--- a/net/minecraft/client/multiplayer/WorldClient.java
+++ b/net/minecraft/client/multiplayer/WorldClient.java
@@ -76,11 +76,13 @@
       this.field_200261_M = new Scoreboard();
       this.field_73035_a = p_i49845_1_;
       this.func_72912_H().func_176144_a(p_i49845_4_);
-      this.func_175652_B(new BlockPos(8, 64, 8));
       this.field_73011_w.func_76558_a(this);
+      this.func_175652_B(new BlockPos(8, 64, 8)); //Forge: Moved below registerWorld to prevent NPE in our redirect.
       this.field_73020_y = this.func_72970_h();
       this.func_72966_v();
       this.func_72947_a();
+      this.initCapabilities();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Load(this));
    }
 
    public void func_72835_b(BooleanSupplier p_72835_1_) {
@@ -179,6 +181,11 @@
 
    public void func_72900_e(Entity p_72900_1_) {
       super.func_72900_e(p_72900_1_);
+   }
+
+   @Override
+   public void removeEntity(Entity p_72900_1_, boolean keepData) {
+      super.removeEntity(p_72900_1_, keepData);
       this.field_73032_d.remove(p_72900_1_);
    }
 
