--- a/net/minecraft/client/gui/screens/advancements/AdvancementsScreen.java
+++ b/net/minecraft/client/gui/screens/advancements/AdvancementsScreen.java
@@ -52,6 +_,7 @@
     private final Map<AdvancementHolder, AdvancementTab> tabs = Maps.newLinkedHashMap();
     private @Nullable AdvancementTab selectedTab;
     private boolean isScrolling;
+    private static int tabPage, maxPages;
 
     public AdvancementsScreen(final ClientAdvancements advancements) {
         this(advancements, null);
@@ -65,6 +_,19 @@
 
     @Override
     protected void init() {
+        if (this.tabs.size() > AdvancementTabType.MAX_TABS) {
+            maxPages = this.tabs.size() / AdvancementTabType.MAX_TABS;
+            var pageLayout = new net.minecraft.client.gui.layouts.GridLayout();
+            var text = new net.minecraft.client.gui.components.StringWidget(Component.literal(String.format("%d / %d", maxPages + 1, maxPages + 1)), this.font);
+            var update = (java.util.function.Consumer<Boolean>)back -> {
+                tabPage = back ? Math.max(tabPage - 1, 0) : Math.min(tabPage + 1, maxPages);
+                text.setMessage(Component.literal(String.format("%d / %d", 1, maxPages + 1)));
+            };
+            pageLayout.addChild(Button.builder(Component.literal("<"), b -> update.accept(true)).size(20, 20).build(), 0, 0, pageLayout.defaultCellSetting().alignHorizontallyLeft());
+            pageLayout.addChild(text, 0, 1, pageLayout.defaultCellSetting().alignHorizontallyCenter());
+            pageLayout.addChild(Button.builder(Component.literal(">"), b -> update.accept(false)).size(20, 20).build(), 0, 2, pageLayout.defaultCellSetting().alignHorizontallyRight());
+            this.layout.addToHeader(pageLayout);
+        }
         this.layout.addTitleHeader(TITLE, this.font);
         this.tabs.clear();
         this.selectedTab = null;
@@ -119,7 +_,7 @@
             int yo = (this.height - 140) / 2;
 
             for (AdvancementTab tab : this.tabs.values()) {
-                if (tab.isMouseOver(xo, yo, event.x(), event.y())) {
+                if (tab.getPage() == tabPage && tab.isMouseOver(xo, yo, event.x(), event.y())) {
                     this.advancements.setSelectedTab(tab.getRootNode().holder(), true);
                     break;
                 }
@@ -208,10 +_,12 @@
         graphics.blit(RenderPipelines.GUI_TEXTURED, WINDOW_LOCATION, this.leftPos, this.topPos, 0.0F, 0.0F, 252, 140, 256, 256);
         if (this.tabs.size() > 1) {
             for (AdvancementTab tab : this.tabs.values()) {
+                if (tab.getPage() == tabPage)
                 tab.extractTab(graphics, this.leftPos, this.topPos, mouseX, mouseY, tab == this.selectedTab);
             }
 
             for (AdvancementTab tab : this.tabs.values()) {
+                if (tab.getPage() == tabPage)
                 tab.extractIcon(graphics, this.leftPos, this.topPos);
             }
         }
@@ -230,6 +_,7 @@
 
         if (this.tabs.size() > 1) {
             for (AdvancementTab tab : this.tabs.values()) {
+                if (tab.getPage() == tabPage)
                 if (tab.isMouseOver(this.leftPos, this.topPos, mouseX, mouseY)) {
                     graphics.setTooltipForNextFrame(this.font, tab.getTitle(), mouseX, mouseY);
                 }
