--- a/net/minecraft/client/gui/screens/advancements/AdvancementTab.java
+++ b/net/minecraft/client/gui/screens/advancements/AdvancementTab.java
@@ -36,6 +_,7 @@
    private int f_97141_ = Integer.MIN_VALUE;
    private float f_97142_;
    private boolean f_97143_;
+   private int page;
 
    public AdvancementTab(Minecraft p_97145_, AdvancementsScreen p_97146_, AdvancementTabType p_97147_, int p_97148_, Advancement p_97149_, DisplayInfo p_97150_) {
       this.f_97126_ = p_97145_;
@@ -50,6 +_,15 @@
       this.m_97175_(this.f_97134_, p_97149_);
    }
 
+   public AdvancementTab(Minecraft mc, AdvancementsScreen screen, AdvancementTabType type, int index, int page, Advancement adv, DisplayInfo info) {
+      this(mc, screen, type, index, adv, info);
+      this.page = page;
+   }
+
+   public int getPage() {
+      return page;
+   }
+
    public AdvancementTabType m_169538_() {
       return this.f_97128_;
    }
@@ -143,8 +_,8 @@
          return null;
       } else {
          for(AdvancementTabType advancementtabtype : AdvancementTabType.values()) {
-            if (p_97173_ < advancementtabtype.m_97210_()) {
-               return new AdvancementTab(p_97171_, p_97172_, advancementtabtype, p_97173_, p_97174_, p_97174_.m_138320_());
+            if ((p_97173_ % AdvancementTabType.MAX_TABS) < advancementtabtype.m_97210_()) {
+               return new AdvancementTab(p_97171_, p_97172_, advancementtabtype, p_97173_ % AdvancementTabType.MAX_TABS, p_97173_ / AdvancementTabType.MAX_TABS, p_97174_, p_97174_.m_138320_());
             }
 
             p_97173_ -= advancementtabtype.m_97210_();
