--- a/net/minecraft/util/datafix/fixes/StructuresBecomeConfiguredFix.java
+++ b/net/minecraft/util/datafix/fixes/StructuresBecomeConfiguredFix.java
@@ -180,7 +_,17 @@
         String s = p_207725_.asString("UNKNOWN").toLowerCase(Locale.ROOT);
         StructuresBecomeConfiguredFix.Conversion structuresbecomeconfiguredfix$conversion = CONVERSION_MAP.get(s);
         if (structuresbecomeconfiguredfix$conversion == null) {
-            return null;
+            // Forge: hook for mods to register conversions through RegisterStructureConversionsEvent
+            structuresbecomeconfiguredfix$conversion = net.minecraftforge.common.ForgeHooks.getStructureConversion(s);
+        }
+        if (structuresbecomeconfiguredfix$conversion == null) {
+
+            if (net.minecraftforge.common.ForgeHooks.checkStructureNamespace(s)) {
+                // Forge: pass-through structure IDs which have a non-"minecraft" namespace
+                return p_329413_.createString(s);
+            }
+            // Forge: Pass-through with "unknown." prefix, so deserializer logs and ignores rather than fixer throwing an exception and dropping chunk data
+            return p_329413_.createString("unknown." + s);
         } else {
             String s1 = structuresbecomeconfiguredfix$conversion.fallback;
             if (!structuresbecomeconfiguredfix$conversion.biomeMapping().isEmpty()) {
