--- a/net/minecraft/util/datafix/fixes/StructuresBecomeConfiguredFix.java
+++ b/net/minecraft/util/datafix/fixes/StructuresBecomeConfiguredFix.java
@@ -178,7 +_,16 @@
         String key = dynamicKey.asString("UNKNOWN").toLowerCase(Locale.ROOT);
         StructuresBecomeConfiguredFix.Conversion conversion = CONVERSION_MAP.get(key);
         if (conversion == null) {
-            return null;
+            // Forge: hook for mods to register conversions through RegisterStructureConversionsEvent
+            conversion = net.minecraftforge.common.ForgeHooks.getStructureConversion(key);
+        }
+        if (conversion == null) {
+            if (net.minecraftforge.common.ForgeHooks.checkStructureNamespace(key)) {
+                // Forge: pass-through structure IDs which have a non-"minecraft" namespace
+                return chunk.createString(key);
+            }
+            // Forge: Pass-through with "unknown." prefix, so deserializer logs and ignores rather than fixer throwing an exception and dropping chunk data
+            return chunk.createString("unknown." + key);
         }
 
         String resultingId = conversion.fallback;
