--- a/net/minecraft/entity/ai/brain/task/ShootTargetTask.java
+++ b/net/minecraft/entity/ai/brain/task/ShootTargetTask.java
@@ -24,7 +_,7 @@
 
    protected boolean func_212832_a_(ServerWorld p_212832_1_, E p_212832_2_) {
       LivingEntity livingentity = func_233887_a_(p_212832_2_);
-      return p_212832_2_.func_233631_a_(Items.field_222114_py) && BrainUtil.func_233876_c_(p_212832_2_, livingentity) && BrainUtil.func_233869_a_(p_212832_2_, livingentity, 0);
+      return p_212832_2_.func_233634_a_(item -> item instanceof CrossbowItem) && BrainUtil.func_233876_c_(p_212832_2_, livingentity) && BrainUtil.func_233869_a_(p_212832_2_, livingentity, 0);
    }
 
    protected boolean func_212834_g_(ServerWorld p_212834_1_, E p_212834_2_, long p_212834_3_) {
@@ -42,7 +_,7 @@
          p_212835_2_.func_184602_cy();
       }
 
-      if (p_212835_2_.func_233631_a_(Items.field_222114_py)) {
+      if (p_212835_2_.func_233634_a_(item -> item instanceof CrossbowItem)) {
          p_212835_2_.func_213671_a(false);
          CrossbowItem.func_220011_a(p_212835_2_.func_184607_cu(), false);
       }
@@ -51,7 +_,7 @@
 
    private void func_233888_a_(E p_233888_1_, LivingEntity p_233888_2_) {
       if (this.field_233886_c_ == ShootTargetTask.Status.UNCHARGED) {
-         p_233888_1_.func_184598_c(ProjectileHelper.func_221274_a(p_233888_1_, Items.field_222114_py));
+         p_233888_1_.func_184598_c(ProjectileHelper.getWeaponHoldingHand(p_233888_1_, item -> item instanceof CrossbowItem));
          this.field_233886_c_ = ShootTargetTask.Status.CHARGING;
          p_233888_1_.func_213671_a(true);
       } else if (this.field_233886_c_ == ShootTargetTask.Status.CHARGING) {
@@ -74,7 +_,7 @@
          }
       } else if (this.field_233886_c_ == ShootTargetTask.Status.READY_TO_ATTACK) {
          p_233888_1_.func_82196_d(p_233888_2_, 1.0F);
-         ItemStack itemstack1 = p_233888_1_.func_184586_b(ProjectileHelper.func_221274_a(p_233888_1_, Items.field_222114_py));
+         ItemStack itemstack1 = p_233888_1_.func_184586_b(ProjectileHelper.getWeaponHoldingHand(p_233888_1_, item -> item instanceof CrossbowItem));
          CrossbowItem.func_220011_a(itemstack1, false);
          this.field_233886_c_ = ShootTargetTask.Status.UNCHARGED;
       }
