--- a/net/minecraft/world/effect/MobEffect.java
+++ b/net/minecraft/world/effect/MobEffect.java
@@ -17,7 +_,7 @@
 import net.minecraft.world.entity.ai.attributes.AttributeModifier;
 import net.minecraft.world.entity.player.Player;
 
-public class MobEffect {
+public class MobEffect implements net.minecraftforge.common.extensions.IForgeMobEffect {
    private final Map<Attribute, AttributeModifier> f_19446_ = Maps.newHashMap();
    private final MobEffectCategory f_19447_;
    private final int f_19448_;
@@ -43,6 +_,7 @@
    protected MobEffect(MobEffectCategory p_19451_, int p_19452_) {
       this.f_19447_ = p_19451_;
       this.f_19448_ = p_19452_;
+      initClient();
    }
 
    public Optional<MobEffectInstance.FactorData> m_216881_() {
@@ -194,4 +_,29 @@
    public boolean m_19486_() {
       return this.f_19447_ == MobEffectCategory.BENEFICIAL;
    }
+
+   // FORGE START
+   private Object effectRenderer;
+
+   /*
+      DO NOT CALL, IT WILL DISAPPEAR IN THE FUTURE
+      Call RenderProperties.getEffectRenderer instead
+    */
+   public Object getEffectRendererInternal() {
+      return effectRenderer;
+   }
+
+   private void initClient() {
+      // Minecraft instance isn't available in datagen, so don't call initializeClient if in datagen
+      if (net.minecraftforge.fml.loading.FMLEnvironment.dist == net.minecraftforge.api.distmarker.Dist.CLIENT && !net.minecraftforge.fml.loading.FMLLoader.getLaunchHandler().isData()) {
+         initializeClient(properties -> {
+            this.effectRenderer = properties;
+         });
+      }
+   }
+
+   public void initializeClient(java.util.function.Consumer<net.minecraftforge.client.extensions.common.IClientMobEffectExtensions> consumer) {
+   }
+   // END FORGE
+
 }
