--- ../src-base/minecraft/net/minecraft/util/Vec3.java
+++ ../src-work/minecraft/net/minecraft/util/Vec3.java
@@ -1,8 +1,5 @@
 package net.minecraft.util;
 
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
-
 public class Vec3
 {
     public final double xCoord;
@@ -32,7 +29,6 @@
         this.zCoord = z;
     }
 
-    @SideOnly(Side.CLIENT)
     public Vec3 subtractReverse(Vec3 vec)
     {
         return new Vec3(vec.xCoord - this.xCoord, vec.yCoord - this.yCoord, vec.zCoord - this.zCoord);
@@ -49,7 +45,6 @@
         return this.xCoord * vec.xCoord + this.yCoord * vec.yCoord + this.zCoord * vec.zCoord;
     }
 
-    @SideOnly(Side.CLIENT)
     public Vec3 crossProduct(Vec3 vec)
     {
         return new Vec3(this.yCoord * vec.zCoord - this.zCoord * vec.yCoord, this.zCoord * vec.xCoord - this.xCoord * vec.zCoord, this.xCoord * vec.yCoord - this.yCoord * vec.xCoord);
