--- a/net/minecraft/world/storage/loot/conditions/RandomChanceWithLooting.java
+++ b/net/minecraft/world/storage/loot/conditions/RandomChanceWithLooting.java
@@ -20,11 +20,7 @@
    }
 
    public boolean func_186618_a(Random p_186618_1_, LootContext p_186618_2_) {
-      int i = 0;
-      if (p_186618_2_.func_186492_c() instanceof EntityLivingBase) {
-         i = EnchantmentHelper.func_185283_h((EntityLivingBase)p_186618_2_.func_186492_c());
-      }
-
+      int i = p_186618_2_.getLootingModifier();
       return p_186618_1_.nextFloat() < this.field_186627_a + (float)i * this.field_186628_b;
    }
 
