--- ../src-base/minecraft/net/minecraft/client/multiplayer/NetClientHandler.java
+++ ../src-work/minecraft/net/minecraft/client/multiplayer/NetClientHandler.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.multiplayer;
 
+import cpw.mods.fml.common.network.FMLNetworkHandler;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.io.BufferedReader;
@@ -174,11 +175,14 @@
     public int field_72556_d = 20;
     Random field_72557_e = new Random();
 
+    private static byte connectionCompatibilityLevel;
+
     public NetClientHandler(Minecraft p_i3103_1_, String p_i3103_2_, int p_i3103_3_) throws IOException
     {
         this.field_72563_h = p_i3103_1_;
         Socket var4 = new Socket(InetAddress.getByName(p_i3103_2_), p_i3103_3_);
         this.field_72555_g = new TcpConnection(var4, "Client", this);
+        FMLNetworkHandler.onClientConnectionToRemoteServer(this, p_i3103_2_, p_i3103_3_, this.field_72555_g);
     }
 
     public NetClientHandler(Minecraft p_i3104_1_, IntegratedServer p_i3104_2_) throws IOException
@@ -186,6 +190,7 @@
         this.field_72563_h = p_i3104_1_;
         this.field_72555_g = new MemoryConnection(this);
         p_i3104_2_.func_71343_a().func_71754_a((MemoryConnection)this.field_72555_g, p_i3104_1_.field_71449_j.field_74286_b);
+        FMLNetworkHandler.onClientConnectionToIntegratedServer(this, p_i3104_2_, this.field_72555_g);
     }
 
     public void func_72547_c()
@@ -256,6 +261,7 @@
 
     public void func_72513_a(Packet252SharedKey p_72513_1_)
     {
+        this.func_72552_c(FMLNetworkHandler.getFMLFakeLoginPacket());
         this.func_72552_c(new Packet205ClientCommand(0));
     }
 
@@ -271,6 +277,7 @@
         this.field_72563_h.field_71439_g.field_70157_k = p_72455_1_.field_73561_a;
         this.field_72556_d = p_72455_1_.field_73562_h;
         this.field_72563_h.field_71442_b.func_78746_a(p_72455_1_.field_73557_d);
+        FMLNetworkHandler.onConnectionEstablishedToServer(this, field_72555_g, p_72455_1_);
         this.field_72563_h.field_71474_y.func_82879_c();
     }
 
@@ -711,6 +718,7 @@
         {
             this.field_72555_g.func_74429_a(p_72546_1_);
             this.field_72555_g.func_74423_d();
+            FMLNetworkHandler.onConnectionClosed(this.field_72555_g, this.getPlayer());
         }
     }
 
@@ -750,6 +758,7 @@
 
     public void func_72481_a(Packet3Chat p_72481_1_)
     {
+        p_72481_1_ = FMLNetworkHandler.handleChatMessage(this, p_72481_1_);
         this.field_72563_h.field_71456_v.func_73827_b().func_73765_a(p_72481_1_.field_73476_b);
     }
 
@@ -1243,6 +1252,11 @@
 
     public void func_72494_a(Packet131MapData p_72494_1_)
     {
+        FMLNetworkHandler.handlePacket131Packet(this, p_72494_1_);
+    }
+
+    public void fmlPacket131Callback(Packet131MapData p_72494_1_)
+    {
         if (p_72494_1_.field_73438_a == Item.field_77744_bd.field_77779_bT)
         {
             ItemMap.func_77874_a(p_72494_1_.field_73436_b, this.field_72563_h.field_71441_e).func_76192_a(p_72494_1_.field_73437_c);
@@ -1352,6 +1366,11 @@
 
     public void func_72501_a(Packet250CustomPayload p_72501_1_)
     {
+        FMLNetworkHandler.handlePacket250Packet(p_72501_1_, field_72555_g, this);
+    }
+
+    public void handleVanilla250Packet(Packet250CustomPayload p_72501_1_)
+    {
         if ("MC|TPack".equals(p_72501_1_.field_73630_a))
         {
             String[] var2 = (new String(p_72501_1_.field_73629_c)).split("\u0000");
@@ -1396,4 +1415,20 @@
     {
         return this.field_72555_g;
     }
+
+    @Override
+    public EntityPlayer getPlayer()
+    {
+        return field_72563_h.field_71439_g;
+    }
+
+    public static void setConnectionCompatibilityLevel(byte connectionCompatibilityLevel)
+    {
+        NetClientHandler.connectionCompatibilityLevel = connectionCompatibilityLevel;
+    }
+
+    public static byte getConnectionCompatibilityLevel()
+    {
+        return connectionCompatibilityLevel;
+    }
 }
