--- a/net/minecraft/world/entity/ai/village/VillageSiege.java
+++ b/net/minecraft/world/entity/ai/village/VillageSiege.java
@@ -78,7 +_,9 @@
                   this.spawnX = blockpos.getX() + Mth.floor(Mth.cos(f) * 32.0F);
                   this.spawnY = blockpos.getY();
                   this.spawnZ = blockpos.getZ() + Mth.floor(Mth.sin(f) * 32.0F);
-                  if (this.findRandomSpawnPos(p_27008_, new BlockPos(this.spawnX, this.spawnY, this.spawnZ)) != null) {
+                  Vec3 siegeLocation = this.findRandomSpawnPos(p_27008_, new BlockPos(this.spawnX, this.spawnY, this.spawnZ));
+                  if (siegeLocation != null) {
+                     if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.village.VillageSiegeEvent(this, p_27008_, player, siegeLocation))) return false;
                      this.nextSpawnTime = 0;
                      this.zombiesToSpawn = 20;
                      break;
@@ -98,7 +_,7 @@
       if (vec3 != null) {
          Zombie zombie;
          try {
-            zombie = new Zombie(p_27017_);
+            zombie = EntityType.ZOMBIE.create(p_27017_); //Forge: Direct Initialization is deprecated, use EntityType.
             zombie.finalizeSpawn(p_27017_, p_27017_.getCurrentDifficultyAt(zombie.blockPosition()), MobSpawnType.EVENT, (SpawnGroupData)null, (CompoundTag)null);
          } catch (Exception exception) {
             LOGGER.warn("Failed to create zombie for village siege at {}", vec3, exception);
