--- a/net/minecraft/world/entity/ai/village/VillageSiege.java
+++ b/net/minecraft/world/entity/ai/village/VillageSiege.java
@@ -78,7 +_,9 @@
                   this.f_27002_ = blockpos.m_123341_() + Mth.m_14143_(Mth.m_14089_(f) * 32.0F);
                   this.f_27003_ = blockpos.m_123342_();
                   this.f_27004_ = blockpos.m_123343_() + Mth.m_14143_(Mth.m_14031_(f) * 32.0F);
-                  if (this.m_27009_(p_27008_, new BlockPos(this.f_27002_, this.f_27003_, this.f_27004_)) != null) {
+                  Vec3 siegeLocation = this.m_27009_(p_27008_, new BlockPos(this.f_27002_, this.f_27003_, this.f_27004_));
+                  if (siegeLocation != null) {
+                     if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.village.VillageSiegeEvent(this, p_27008_, player, siegeLocation))) return false;
                      this.f_27001_ = 0;
                      this.f_27000_ = 20;
                      break;
@@ -98,7 +_,7 @@
       if (vec3 != null) {
          Zombie zombie;
          try {
-            zombie = new Zombie(p_27017_);
+            zombie = EntityType.f_20501_.m_20615_(p_27017_); //Forge: Direct Initialization is deprecated, use EntityType.
             zombie.m_6518_(p_27017_, p_27017_.m_6436_(zombie.m_20183_()), MobSpawnType.EVENT, (SpawnGroupData)null, (CompoundTag)null);
          } catch (Exception exception) {
             f_26997_.warn("Failed to create zombie for village siege at {}", vec3, exception);
