--- ../src-base/minecraft/net/minecraft/world/gen/structure/template/Template.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/template/Template.java
@@ -514,6 +514,7 @@
             nbttaglist2.func_74742_a(NBTUtil.func_190009_a(new NBTTagCompound(), iblockstate));
         }
 
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().getDataFixer().writeVersionData(p_189552_1_); //Moved up for MC updating reasons.
         p_189552_1_.func_74782_a("palette", nbttaglist2);
         p_189552_1_.func_74782_a("blocks", nbttaglist);
         p_189552_1_.func_74782_a("entities", nbttaglist1);
