diff -r -U 3 minecraft_server\net\minecraft\src\CallableSuspiciousClasses.java minecraft_server_patched\net\minecraft\src\CallableSuspiciousClasses.java
--- minecraft_server\net\minecraft\src\CallableSuspiciousClasses.java	Mon Dec 17 22:50:59 2012
+++ minecraft_server_patched\net\minecraft\src\CallableSuspiciousClasses.java	Mon Dec 17 22:52:18 2012
@@ -21,9 +21,14 @@
 
    public String func_85085_a() {
       StringBuilder var1 = new StringBuilder();
-      Field var2 = ClassLoader.class.getDeclaredField("classes");
-      var2.setAccessible(true);
-      ArrayList var3 = new ArrayList((Vector)var2.get(CrashReport.class.getClassLoader()));
+      ArrayList var3;
+      try {
+         Field var2 = ClassLoader.class.getDeclaredField("classes");
+         var2.setAccessible(true);
+         var3 = new ArrayList((Vector)var2.get(CrashReport.class.getClassLoader()));
+      } catch(Exception ex) {
+    	 return "";
+      }
       boolean var4 = true;
       boolean var5 = !CrashReport.class.getCanonicalName().equals("net.minecraft.CrashReport");
       HashMap var6 = new HashMap();
diff -r -U 3 minecraft_server\net\minecraft\src\ComponentStrongholdRightTurn.java minecraft_server_patched\net\minecraft\src\ComponentStrongholdRightTurn.java
--- minecraft_server\net\minecraft\src\ComponentStrongholdRightTurn.java	Mon Dec 17 22:50:59 2012
+++ minecraft_server_patched\net\minecraft\src\ComponentStrongholdRightTurn.java	Mon Dec 17 22:52:18 2012
@@ -11,6 +11,10 @@
 
 public class ComponentStrongholdRightTurn extends ComponentStrongholdLeftTurn {
 
+   public ComponentStrongholdRightTurn(int p_i3843_1_, Random p_i3843_2_, StructureBoundingBox p_i3843_3_, int p_i3843_4_) {
+      super(p_i3843_1_, p_i3843_2_, p_i3843_3_, p_i3843_4_);
+   }
+
    public void func_74861_a(StructureComponent p_74861_1_, List p_74861_2_, Random p_74861_3_) {
       if(this.field_74885_f != 2 && this.field_74885_f != 3) {
          this.func_74989_b((ComponentStrongholdStairs2)p_74861_1_, p_74861_2_, p_74861_3_, 1, 1);
diff -r -U 3 minecraft_server\net\minecraft\src\DedicatedServer.java minecraft_server_patched\net\minecraft\src\DedicatedServer.java
--- minecraft_server\net\minecraft\src\DedicatedServer.java	Mon Dec 17 22:50:59 2012
+++ minecraft_server_patched\net\minecraft\src\DedicatedServer.java	Mon Dec 17 22:52:18 2012
@@ -203,7 +203,7 @@
       System.exit(0);
    }
 
-   protected void func_71190_q() {
+   public void func_71190_q() {
       super.func_71190_q();
       this.func_71333_ah();
    }
diff -r -U 3 minecraft_server\net\minecraft\src\EntityMooshroom.java minecraft_server_patched\net\minecraft\src\EntityMooshroom.java
--- minecraft_server\net\minecraft\src\EntityMooshroom.java	Mon Dec 17 22:51:00 2012
+++ minecraft_server_patched\net\minecraft\src\EntityMooshroom.java	Mon Dec 17 22:52:18 2012
@@ -56,10 +56,10 @@
       return new EntityMooshroom(this.field_70170_p);
    }
 
-   // $FF: synthetic method
-   public EntityCow func_70879_a(EntityAgeable p_70879_1_) {
-      return this.func_70879_a(p_70879_1_);
-   }
+//   // $FF: synthetic method
+//   public EntityCow func_70879_a(EntityAgeable p_70879_1_) {
+//      return this.func_70879_a(p_70879_1_);
+//   }
 
    // $FF: synthetic method
    public EntityAgeable func_90011_a(EntityAgeable p_90011_1_) {
diff -r -U 3 minecraft_server\net\minecraft\src\TcpMasterThread.java minecraft_server_patched\net\minecraft\src\TcpMasterThread.java
--- minecraft_server\net\minecraft\src\TcpMasterThread.java	Mon Dec 17 22:51:01 2012
+++ minecraft_server_patched\net\minecraft\src\TcpMasterThread.java	Mon Dec 17 22:52:18 2012
@@ -11,6 +11,7 @@
       this.field_74504_a = p_i3285_1_;
    }
 
+   @SuppressWarnings("deprecation")
    public void run() {
       try {
          Thread.sleep(5000L);
diff -r -U 3 minecraft_server\net\minecraft\src\ThreadedFileIOBase.java minecraft_server_patched\net\minecraft\src\ThreadedFileIOBase.java
--- minecraft_server\net\minecraft\src\ThreadedFileIOBase.java	Mon Dec 17 22:51:01 2012
+++ minecraft_server_patched\net\minecraft\src\ThreadedFileIOBase.java	Mon Dec 17 22:52:18 2012
@@ -20,7 +20,8 @@
    }
 
    public void run() {
-      this.func_75736_b();
+      while(true)
+         this.func_75736_b();
    }
 
    private void func_75736_b() {
