diff -r -U 3 minecraft/net/minecraft/src/RenderGlobal.java minecraft_patched/net/minecraft/src/RenderGlobal.java
--- minecraft/net/minecraft/src/RenderGlobal.java	2012-02-25 12:51:18.000000000 +1300
+++ minecraft_patched/net/minecraft/src/RenderGlobal.java	2012-02-25 13:13:03.850786399 +1300
@@ -954,7 +954,7 @@
             f2 = f6;
             f3 = f7;
         }
-        float f5 = 0.0004882812F;
+        float f5 = 0.0004882813F;
         double d = (float)field_1435_x + p_4141_1_;
         double d1 = field_1439_t.field_22009_h.field_9285_at + (field_1439_t.field_22009_h.field_611_ak - field_1439_t.field_22009_h.field_9285_at) * (double)p_4141_1_ + d * 0.029999999329447746D;
         double d2 = field_1439_t.field_22009_h.field_9283_av + (field_1439_t.field_22009_h.field_609_am - field_1439_t.field_22009_h.field_9283_av) * (double)p_4141_1_;
