--- a/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
+++ b/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
@@ -14,6 +_,14 @@
 import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
 
 public abstract class AbstractTreeGrower {
+   /**
+    * Forge: context-sensitive version to be able to use data pack provided features
+    */
+   @Nullable
+   protected Holder<? extends ConfiguredFeature<?, ?>> getConfiguredFeature(ServerLevel level, ChunkGenerator chunkGenerator, BlockPos pos, BlockState state, RandomSource random, boolean hasFlowers, ResourceKey<ConfiguredFeature<?, ?>> featureKey) {
+      return level.m_8891_().m_175515_(Registries.f_256911_).m_203636_(featureKey).orElse(null);
+   }
+
    @Nullable
    protected abstract ResourceKey<ConfiguredFeature<?, ?>> m_213888_(RandomSource p_222910_, boolean p_222911_);
 
@@ -22,11 +_,12 @@
       if (resourcekey == null) {
          return false;
       } else {
-         Holder<ConfiguredFeature<?, ?>> holder = p_222905_.m_8891_().m_175515_(Registries.f_256911_).m_203636_(resourcekey).orElse((Holder.Reference<ConfiguredFeature<?, ?>>)null);
-         if (holder == null) {
+         Holder<? extends ConfiguredFeature<?, ?>> holder = this.getConfiguredFeature(p_222905_, p_222906_, p_222907_, p_222908_, p_222909_, this.m_60011_(p_222905_, p_222907_), resourcekey);
+         net.minecraftforge.event.level.SaplingGrowTreeEvent event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_222905_, p_222909_, p_222907_, holder);
+         if (event.getResult().equals(net.minecraftforge.eventbus.api.Event.Result.DENY) || event.getFeature() == null) {
             return false;
          } else {
-            ConfiguredFeature<?, ?> configuredfeature = holder.m_203334_();
+            ConfiguredFeature<?, ?> configuredfeature = event.getFeature().m_203334_();
             BlockState blockstate = p_222905_.m_6425_(p_222907_).m_76188_();
             p_222905_.m_7731_(p_222907_, blockstate, 4);
             if (configuredfeature.m_224953_(p_222905_, p_222906_, p_222909_, p_222907_)) {
