--- a/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
+++ b/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
@@ -459,6 +_,9 @@
          }
 
          this.reader.skip();
+         EntitySelector forgeSelector = net.minecraftforge.common.command.EntitySelectorManager.parseSelector(this);
+         if (forgeSelector != null)
+            return forgeSelector;
          this.parseSelector();
       } else {
          this.parseNameOrUUID();
@@ -474,6 +_,7 @@
       p_121248_.suggest("@r", Component.translatable("argument.entity.selector.randomPlayer"));
       p_121248_.suggest("@s", Component.translatable("argument.entity.selector.self"));
       p_121248_.suggest("@e", Component.translatable("argument.entity.selector.allEntities"));
+      net.minecraftforge.common.command.EntitySelectorManager.fillSelectorSuggestions(p_121248_);
    }
 
    private CompletableFuture<Suggestions> suggestNameOrSelector(SuggestionsBuilder p_121287_, Consumer<SuggestionsBuilder> p_121288_) {
