--- a/net/minecraft/client/shader/ShaderGroup.java
+++ b/net/minecraft/client/shader/ShaderGroup.java
@@ -164,7 +164,8 @@
                         throw new JSONException("Render target '" + s4 + "' can't be used as depth buffer");
                      }
 
-                     ResourceLocation resourcelocation = new ResourceLocation("textures/effect/" + s4 + ".png");
+                     ResourceLocation rl = ResourceLocation.func_208304_a(s4);
+                     ResourceLocation resourcelocation = new ResourceLocation(rl.func_110624_b(), "textures/effect/" + rl.func_110623_a() + ".png");
                      IResource iresource = null;
 
                      try {
