--- ../src-base/minecraft/net/minecraft/client/shader/ShaderManager.java
+++ ../src-work/minecraft/net/minecraft/client/shader/ShaderManager.java
@@ -54,7 +54,8 @@
     public ShaderManager(IResourceManager p_i45087_1_, String p_i45087_2_) throws JsonException, IOException
     {
         JsonParser jsonparser = new JsonParser();
-        ResourceLocation resourcelocation = new ResourceLocation("shaders/program/" + p_i45087_2_ + ".json");
+        String[] rl = ResourceLocation.func_177516_a(p_i45087_2_);
+        ResourceLocation resourcelocation = new ResourceLocation(rl[0], "shaders/program/" + rl[1] + ".json");
         this.field_148007_m = p_i45087_2_;
         IResource iresource = null;
 
