--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RenderManager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RenderManager.java
@@ -204,7 +204,12 @@
         this.playerRenderer = new RenderPlayer(this);
         this.skinMap.put("default", this.playerRenderer);
         this.skinMap.put("slim", new RenderPlayer(this, true));
+        net.minecraftforge.fml.client.registry.RenderingRegistry.loadEntityRenderers(this, this.entityRenderMap);
     }
+	
+    public Map<String, RenderPlayer> getSkinMap() {
+        return (Map<String, RenderPlayer>) java.util.Collections.unmodifiableMap(skinMap);
+    }
 
     public void setRenderPosition(double renderPosXIn, double renderPosYIn, double renderPosZIn)
     {
