--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RendererLivingEntity.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RendererLivingEntity.java
@@ -40,6 +40,9 @@
     protected boolean renderOutlines = false;
     private static final String __OBFID = "CL_00001012";
 
+    public static float NAME_TAG_RANGE = 64.0f;
+    public static float NAME_TAG_RANGE_SNEAK = 32.0f;
+
     public RendererLivingEntity(RenderManager p_i46156_1_, ModelBase p_i46156_2_, float p_i46156_3_)
     {
         super(p_i46156_1_);
@@ -83,6 +86,7 @@
 
     public void doRender(EntityLivingBase entity, double x, double y, double z, float p_76986_8_, float partialTicks)
     {
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Pre(entity, this, x, y, z))) return;
         GlStateManager.pushMatrix();
         GlStateManager.disableCull();
         this.mainModel.swingProgress = this.getSwingProgress(entity, partialTicks);
@@ -193,6 +197,7 @@
         {
             super.doRender(entity, x, y, z, p_76986_8_, partialTicks);
         }
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Post(entity, this, x, y, z));
     }
 
     protected boolean func_177088_c(EntityLivingBase p_177088_1_)
@@ -472,10 +477,11 @@
 
     public void passSpecialRender(EntityLivingBase p_77033_1_, double p_77033_2_, double p_77033_4_, double p_77033_6_)
     {
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Specials.Pre(p_77033_1_, this, p_77033_2_, p_77033_4_, p_77033_6_))) return;
         if (this.canRenderName(p_77033_1_))
         {
             double d3 = p_77033_1_.getDistanceSqToEntity(this.renderManager.livingPlayer);
-            float f = p_77033_1_.isSneaking() ? 32.0F : 64.0F;
+            float f = p_77033_1_.isSneaking() ? NAME_TAG_RANGE_SNEAK : NAME_TAG_RANGE;
 
             if (d3 < (double)(f * f))
             {
@@ -522,6 +528,7 @@
                 }
             }
         }
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Specials.Post(p_77033_1_, this, p_77033_2_, p_77033_4_, p_77033_6_));
     }
 
     protected boolean canRenderName(EntityLivingBase targetEntity)
