--- a/net/minecraft/world/level/levelgen/feature/Feature.java
+++ b/net/minecraft/world/level/levelgen/feature/Feature.java
@@ -52,7 +_,7 @@
 import net.minecraft.world.level.levelgen.feature.configurations.UnderwaterMagmaConfiguration;
 import net.minecraft.world.level.levelgen.feature.configurations.VegetationPatchConfiguration;
 
-public abstract class Feature<FC extends FeatureConfiguration> {
+public abstract class Feature<FC extends FeatureConfiguration> extends net.minecraftforge.registries.ForgeRegistryEntry<Feature<?>> {
    public static final Feature<NoneFeatureConfiguration> f_65759_ = m_65807_("no_op", new NoOpFeature(NoneFeatureConfiguration.f_67815_));
    public static final Feature<TreeConfiguration> f_65760_ = m_65807_("tree", new TreeFeature(TreeConfiguration.f_68184_));
    public static final Feature<RandomPatchConfiguration> f_65761_ = m_65807_("flower", new RandomPatchFeature(RandomPatchConfiguration.f_67902_));
@@ -154,7 +_,7 @@
    }
 
    protected static boolean m_159747_(BlockState p_159748_) {
-      return p_159748_.m_204336_(BlockTags.f_13061_);
+      return p_159748_.m_204336_(net.minecraftforge.common.Tags.Blocks.STONE);
    }
 
    public static boolean m_159759_(BlockState p_159760_) {
