--- a/net/minecraft/world/level/levelgen/feature/JigsawFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/JigsawFeature.java
@@ -42,7 +_,7 @@
 
       public void m_142743_(RegistryAccess p_159927_, ChunkGenerator p_159928_, StructureManager p_159929_, ChunkPos p_159930_, Biome p_159931_, JigsawConfiguration p_159932_, LevelHeightAccessor p_159933_) {
          BlockPos blockpos = new BlockPos(p_159930_.m_45604_(), this.f_66168_.f_66146_, p_159930_.m_45605_());
-         Pools.m_127189_();
+         Pools.bootstrap();
          JigsawPlacement.m_161612_(p_159927_, p_159932_, PoolElementStructurePiece::new, p_159928_, p_159929_, blockpos, this, this.f_73564_, this.f_66168_.f_66147_, this.f_66168_.f_66148_, p_159933_);
       }
    }
