--- a/net/minecraft/world/level/levelgen/feature/NetherFortressFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/NetherFortressFeature.java
@@ -42,4 +_,11 @@
 
       p_197129_.m_192792_(p_197130_.f_192708_(), 48, 70);
    }
+
+   @Override
+   public List<MobSpawnSettings.SpawnerData> getDefaultSpawnList(net.minecraft.world.entity.MobCategory category) {
+      if (category == net.minecraft.world.entity.MobCategory.MONSTER)
+         return f_66381_.m_146338_();
+      return java.util.Collections.emptyList();
+   }
 }
