--- a/net/minecraft/world/level/levelgen/feature/OceanMonumentFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/OceanMonumentFeature.java
@@ -28,8 +_,11 @@
       super(p_66472_, PieceGeneratorSupplier.m_197349_(OceanMonumentFeature::m_197131_, OceanMonumentFeature::m_191045_));
    }
 
-   protected boolean m_5910_() {
-      return false;
+   @Override
+   public java.util.List<MobSpawnSettings.SpawnerData> getDefaultSpawnList(net.minecraft.world.entity.MobCategory category) {
+      if (category == net.minecraft.world.entity.MobCategory.MONSTER)
+         return f_66469_.m_146338_();
+      return java.util.Collections.emptyList();
    }
 
    private static boolean m_197131_(PieceGeneratorSupplier.Context<NoneFeatureConfiguration> p_197132_) {
