--- a/net/minecraft/world/level/levelgen/feature/PillagerOutpostFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/PillagerOutpostFeature.java
@@ -52,4 +_,11 @@
          return false;
       }
    }
+
+   @Override
+   public java.util.List<MobSpawnSettings.SpawnerData> getDefaultSpawnList(net.minecraft.world.entity.MobCategory category) {
+      if (category == net.minecraft.world.entity.MobCategory.MONSTER)
+         return f_66559_.m_146338_();
+      return java.util.Collections.emptyList();
+   }
 }
