--- a/net/minecraft/world/level/levelgen/feature/StructureFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/StructureFeature.java
@@ -49,7 +_,8 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public abstract class StructureFeature<C extends FeatureConfiguration> {
+@net.minecraftforge.registries.ObjectHolder("minecraft")
+public abstract class StructureFeature<C extends FeatureConfiguration> extends net.minecraftforge.registries.ForgeRegistryEntry<StructureFeature<?>> implements net.minecraftforge.common.extensions.IForgeStructureFeature {
    public static final BiMap<String, StructureFeature<?>> f_67012_ = HashBiMap.create();
    private static final Map<StructureFeature<?>, GenerationStep.Decoration> f_67032_ = Maps.newHashMap();
    private static final Logger f_67033_ = LogManager.getLogger();
@@ -250,15 +_,20 @@
    }
 
    public WeightedRandomList<MobSpawnSettings.SpawnerData> m_142494_() {
-      return MobSpawnSettings.f_151796_;
+      return getSpawnList(net.minecraft.world.entity.MobCategory.MONSTER);
    }
 
    public WeightedRandomList<MobSpawnSettings.SpawnerData> m_142498_() {
-      return MobSpawnSettings.f_151796_;
+      return getSpawnList(net.minecraft.world.entity.MobCategory.CREATURE);
    }
 
    public WeightedRandomList<MobSpawnSettings.SpawnerData> m_160477_() {
-      return MobSpawnSettings.f_151796_;
+      return getSpawnList(net.minecraft.world.entity.MobCategory.UNDERGROUND_WATER_CREATURE);
+   }
+
+   @Override
+   public final WeightedRandomList<MobSpawnSettings.SpawnerData> getSpawnList(net.minecraft.world.entity.MobCategory classification) {
+      return net.minecraftforge.common.world.StructureSpawnManager.getSpawnList(this, classification);
    }
 
    public interface StructureStartFactory<C extends FeatureConfiguration> {
