--- a/net/minecraft/world/level/levelgen/feature/StructureFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/StructureFeature.java
@@ -46,7 +_,7 @@
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureManager;
 import org.slf4j.Logger;
 
-public class StructureFeature<C extends FeatureConfiguration> {
+public abstract class StructureFeature<C extends FeatureConfiguration> extends net.minecraftforge.registries.ForgeRegistryEntry<StructureFeature<?>> {
    private static final Map<StructureFeature<?>, GenerationStep.Decoration> f_67032_ = Maps.newHashMap();
    private static final Logger f_67033_ = LogUtils.getLogger();
    public static final StructureFeature<JigsawConfiguration> f_67013_ = m_67089_("pillager_outpost", new PillagerOutpostFeature(JigsawConfiguration.f_67756_), GenerationStep.Decoration.SURFACE_STRUCTURES);
