--- a/net/minecraft/world/level/levelgen/feature/SwamplandHutFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/SwamplandHutFeature.java
@@ -22,4 +_,13 @@
    private static void m_197181_(StructurePiecesBuilder p_197182_, PieceGenerator.Context<NoneFeatureConfiguration> p_197183_) {
       p_197182_.m_142679_(new SwamplandHutPiece(p_197183_.f_192708_(), p_197183_.f_192705_().m_45604_(), p_197183_.f_192705_().m_45605_()));
    }
+
+   @Override
+   public java.util.List<MobSpawnSettings.SpawnerData> getDefaultSpawnList(net.minecraft.world.entity.MobCategory category) {
+      if (category == net.minecraft.world.entity.MobCategory.MONSTER)
+         return f_67169_.m_146338_();
+      else if (category == net.minecraft.world.entity.MobCategory.CREATURE)
+         return f_67170_.m_146338_();
+      return java.util.Collections.emptyList();
+   }
 }
