--- ../src-base/minecraft/net/minecraft/world/storage/MapData.java
+++ ../src-work/minecraft/net/minecraft/world/storage/MapData.java
@@ -45,7 +45,17 @@
 
     public void readFromNBT(NBTTagCompound nbt)
     {
-        this.dimension = nbt.getByte("dimension");
+        net.minecraft.nbt.NBTBase dimension = nbt.getTag("dimension");
+
+        if (dimension instanceof net.minecraft.nbt.NBTTagByte)
+        {
+            this.dimension = ((net.minecraft.nbt.NBTTagByte)dimension).getByte();
+        }
+        else
+        {
+            this.dimension = ((net.minecraft.nbt.NBTTagInt)dimension).getInt();
+        }
+
         this.xCenter = nbt.getInteger("xCenter");
         this.zCenter = nbt.getInteger("zCenter");
         this.scale = nbt.getByte("scale");
@@ -96,7 +106,7 @@
 
     public void writeToNBT(NBTTagCompound nbt)
     {
-        nbt.setByte("dimension", this.dimension);
+        nbt.setInteger("dimension", this.dimension);
         nbt.setInteger("xCenter", this.xCenter);
         nbt.setInteger("zCenter", this.zCenter);
         nbt.setByte("scale", this.scale);
@@ -176,7 +186,7 @@
             rotation = rotation + (rotation < 0.0D ? -8.0D : 8.0D);
             b2 = (byte)((int)(rotation * 16.0D / 360.0D));
 
-            if (this.dimension < 0)
+            if (worldIn.provider.shouldMapSpin(entityIdentifier, worldX, worldZ, rotation))
             {
                 int k = (int)(worldIn.getWorldInfo().getWorldTime() / 10L);
                 b2 = (byte)(k * k * 34187121 + k * 121 >> 15 & 15);
