--- a/net/minecraft/client/gui/Font.java
+++ b/net/minecraft/client/gui/Font.java
@@ -32,7 +_,7 @@
 import org.joml.Matrix4f;
 
 @OnlyIn(Dist.CLIENT)
-public class Font {
+public class Font implements net.minecraftforge.client.extensions.IForgeFont {
     private static final float EFFECT_DEPTH = 0.01F;
     private static final float OVER_EFFECT_DEPTH = 0.01F;
     private static final float UNDER_EFFECT_DEPTH = -0.01F;
