--- a/net/minecraft/client/gui/Font.java
+++ b/net/minecraft/client/gui/Font.java
@@ -31,7 +_,7 @@
 import org.joml.Vector3f;
 
 @OnlyIn(Dist.CLIENT)
-public class Font {
+public class Font implements net.minecraftforge.client.extensions.IForgeFont {
    private static final float f_168643_ = 0.01F;
    private static final Vector3f f_92712_ = new Vector3f(0.0F, 0.0F, 0.03F);
    public static final int f_193827_ = 8;
@@ -202,6 +_,8 @@
    public StringSplitter m_92865_() {
       return this.f_92714_;
    }
+
+   @Override public Font self() { return this; }
 
    @OnlyIn(Dist.CLIENT)
    public static enum DisplayMode {
