--- a/net/minecraft/client/gui/Gui.java
+++ b/net/minecraft/client/gui/Gui.java
@@ -130,7 +_,7 @@
    public float vignetteBrightness = 1.0F;
    protected int toolHighlightTimer;
    protected ItemStack lastToolHighlight = ItemStack.EMPTY;
-   protected final DebugScreenOverlay debugOverlay;
+   protected DebugScreenOverlay debugOverlay;
    protected final SubtitleOverlay subtitleOverlay;
    protected final SpectatorGui spectatorGui;
    protected final PlayerTabOverlay tabList;
@@ -453,6 +_,8 @@
 
          for(MobEffectInstance mobeffectinstance : Ordering.natural().reverse().sortedCopy(collection)) {
             MobEffect mobeffect = mobeffectinstance.getEffect();
+            var renderer = net.minecraftforge.client.extensions.common.IClientMobEffectExtensions.of(mobeffectinstance);
+            if (!renderer.isVisibleInGui(mobeffectinstance)) continue;
             if (mobeffectinstance.showIcon()) {
                int i = this.screenWidth;
                int j = 1;
@@ -481,6 +_,7 @@
                   }
                }
 
+               if (renderer.renderGuiIcon(mobeffectinstance, this, p_282812_, i, j, 0, f)) continue;
                TextureAtlasSprite textureatlassprite = mobeffecttexturemanager.get(mobeffect);
                int i1 = j;
                float f1 = f;
@@ -601,16 +_,21 @@
    }
 
    public void renderSelectedItemName(GuiGraphics p_283501_) {
+      renderSelectedItemName(p_283501_, 0);
+   }
+
+   public void renderSelectedItemName(GuiGraphics p_283501_, int yShift) {
       this.minecraft.getProfiler().push("selectedItemName");
       if (this.toolHighlightTimer > 0 && !this.lastToolHighlight.isEmpty()) {
-         MutableComponent mutablecomponent = Component.empty().append(this.lastToolHighlight.getHoverName()).withStyle(this.lastToolHighlight.getRarity().color);
+         MutableComponent mutablecomponent = Component.empty().append(this.lastToolHighlight.getHoverName()).withStyle(this.lastToolHighlight.getRarity().getStyleModifier());
          if (this.lastToolHighlight.hasCustomHoverName()) {
             mutablecomponent.withStyle(ChatFormatting.ITALIC);
          }
 
-         int i = this.getFont().width(mutablecomponent);
+         Component highlightTip = this.lastToolHighlight.getHighlightTip(mutablecomponent);
+         int i = this.getFont().width(highlightTip);
          int j = (this.screenWidth - i) / 2;
-         int k = this.screenHeight - 59;
+         int k = this.screenHeight - Math.max(yShift, 59);
          if (!this.minecraft.gameMode.canHurtPlayer()) {
             k += 14;
          }
@@ -622,7 +_,13 @@
 
          if (l > 0) {
             p_283501_.fill(j - 2, k - 2, j + i + 2, k + 9 + 2, this.minecraft.options.getBackgroundColor(0));
-            p_283501_.drawString(this.getFont(), mutablecomponent, j, k, 16777215 + (l << 24));
+            Font font = net.minecraftforge.client.extensions.common.IClientItemExtensions.of(lastToolHighlight).getFont(lastToolHighlight, net.minecraftforge.client.extensions.common.IClientItemExtensions.FontContext.SELECTED_ITEM_NAME);
+            if (font == null) {
+               p_283501_.drawString(this.getFont(), highlightTip, j, k, 16777215 + (l << 24));
+            } else {
+               j = (this.screenWidth - font.width(highlightTip)) / 2;
+               p_283501_.drawString(font, highlightTip, j, k, 16777215 + (l << 24));
+            }
          }
       }
 
@@ -1069,7 +_,7 @@
          ItemStack itemstack = this.minecraft.player.getInventory().getSelected();
          if (itemstack.isEmpty()) {
             this.toolHighlightTimer = 0;
-         } else if (!this.lastToolHighlight.isEmpty() && itemstack.is(this.lastToolHighlight.getItem()) && itemstack.getHoverName().equals(this.lastToolHighlight.getHoverName())) {
+         } else if (!this.lastToolHighlight.isEmpty() && itemstack.getItem() == this.lastToolHighlight.getItem() && (itemstack.getHoverName().equals(this.lastToolHighlight.getHoverName()) && itemstack.getHighlightTip(itemstack.getHoverName()).equals(lastToolHighlight.getHighlightTip(lastToolHighlight.getHoverName())))) {
             if (this.toolHighlightTimer > 0) {
                --this.toolHighlightTimer;
             }
