--- a/net/minecraft/client/gui/Gui.java
+++ b/net/minecraft/client/gui/Gui.java
@@ -64,7 +_,7 @@
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
-public class Gui {
+public class Gui extends net.minecraftforge.client.gui.overlay.ForgeLayerInstance {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final Component SOCIAL_INTERACTIONS_NOT_AVAILABLE = Component.translatable("multiplayer.socialInteractions.not_available");
     public static final Component SAVING_LEVEL = Component.translatable("menu.savingLevel");
@@ -80,6 +_,7 @@
     private @Nullable TutorialToast socialInteractionsToast;
 
     public Gui(final Minecraft minecraft, final Hud hud, final GuiRenderState guiRenderState) {
+        super(minecraft);
         this.minecraft = minecraft;
         this.hud = hud;
         this.splashManager = new SplashManager(minecraft.getUser());
@@ -175,7 +_,7 @@
             profiler.push("screen");
 
             try {
-                this.screen.extractRenderStateWithTooltipAndSubtitles(graphics, xMouse, yMouse, deltaTracker.getGameTimeDeltaTicks());
+                drawScreen(graphics, xMouse, yMouse, deltaTracker.getRealtimeDeltaTicks());
             } catch (Throwable t) {
                 CrashReport report = CrashReport.forThrowable(t, "Rendering screen");
                 CrashReportCategory category = report.addCategory("Screen render details");
@@ -224,12 +_,6 @@
             LOGGER.error("setScreen called from non-game thread");
         }
 
-        if (this.screen != null) {
-            this.screen.removed();
-        } else {
-            this.minecraft.setLastInputType(InputType.NONE);
-        }
-
         if (screen == null) {
             if (this.clientLevelTeardownInProgress) {
                 throw new IllegalStateException("Trying to return to in-game GUI during disconnection");
@@ -248,6 +_,22 @@
             }
         }
 
+        clearLayers();
+        Screen old = this.screen;
+        if (screen != null) {
+           var event = net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenOpening(old, screen);
+           if (event == null) return;
+           screen = event;
+        }
+
+        if (screen != old) {
+            if (old != null) {
+                net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenClose(old);
+                old.removed();
+            } else {
+                this.minecraft.setLastInputType(InputType.NONE);
+            }
+        }
         this.screen = screen;
         if (this.screen != null) {
             this.screen.added();
@@ -465,5 +_,10 @@
                         }
                     )
             );
+    }
+
+    @Override
+    protected void setScreenInternal(Screen value) {
+        this.screen = value;
     }
 }
