--- ../src-base/minecraft/net/minecraft/client/gui/GuiButton.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiButton.java
@@ -22,6 +22,7 @@
     public boolean field_146124_l;
     public boolean field_146125_m;
     protected boolean field_146123_n;
+    public int packedFGColour; //FML
 
     public GuiButton(int p_i1020_1_, int p_i1020_2_, int p_i1020_3_, String p_i1020_4_)
     {
@@ -75,6 +76,11 @@
             this.func_146119_b(p_191745_1_, p_191745_2_, p_191745_3_);
             int j = 14737632;
 
+            if (packedFGColour != 0)
+            {
+                j = packedFGColour;
+            }
+            else
             if (!this.field_146124_l)
             {
                 j = 10526880;
