--- a/net/minecraft/client/gui/GuiBossOverlay.java
+++ b/net/minecraft/client/gui/GuiBossOverlay.java
@@ -28,12 +28,17 @@
 
          for(BossInfoClient bossinfoclient : this.mapBossInfos.values()) {
             int k = i / 2 - 91;
+            net.minecraftforge.client.event.RenderGameOverlayEvent.BossInfo event =
+                    net.minecraftforge.client.ForgeHooksClient.bossBarRenderPre(this.client.field_195558_d, bossinfoclient, k, j, 10 + this.client.fontRenderer.FONT_HEIGHT);
+            if (!event.isCanceled()) {
             GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
             this.client.getTextureManager().bindTexture(GUI_BARS_TEXTURES);
             this.render(k, j, bossinfoclient);
             String s = bossinfoclient.getName().getFormattedText();
             this.client.fontRenderer.drawStringWithShadow(s, (float)(i / 2 - this.client.fontRenderer.getStringWidth(s) / 2), (float)(j - 9), 16777215);
-            j += 10 + this.client.fontRenderer.FONT_HEIGHT;
+            }
+            j += event.getIncrement();
+            net.minecraftforge.client.ForgeHooksClient.bossBarRenderPost(this.client.field_195558_d);
             if (j >= this.client.field_195558_d.func_198087_p() / 3) {
                break;
             }
