--- a/net/minecraft/client/gui/GuiConnecting.java
+++ b/net/minecraft/client/gui/GuiConnecting.java
@@ -62,6 +62,7 @@
                GuiConnecting.this.field_146371_g.func_150719_a(new NetHandlerLoginClient(GuiConnecting.this.field_146371_g, GuiConnecting.this.field_146297_k, GuiConnecting.this.field_146374_i, (p_209549_1_) -> {
                   GuiConnecting.this.func_209514_a(p_209549_1_);
                }));
+               net.minecraftforge.fml.network.NetworkHooks.registerClientLoginChannel(GuiConnecting.this.field_146371_g);
                GuiConnecting.this.field_146371_g.func_179290_a(new CPacketHandshake(p_146367_1_, p_146367_2_, EnumConnectionState.LOGIN));
                GuiConnecting.this.field_146371_g.func_179290_a(new CPacketLoginStart(GuiConnecting.this.field_146297_k.func_110432_I().func_148256_e()));
             } catch (UnknownHostException unknownhostexception) {
