--- ../src-base/minecraft/net/minecraft/client/gui/GuiCreateWorld.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiCreateWorld.java
@@ -323,7 +323,7 @@
             }
             else if (p_146284_1_.id == 8)
             {
-                this.mc.displayGuiScreen(new GuiCreateFlatWorld(this, this.field_146334_a));
+                WorldType.worldTypes[field_146331_K].onCustomizeButton(mc, this);
             }
         }
     }
@@ -341,7 +341,7 @@
         this.field_146326_C.visible = this.field_146344_y;
         this.field_146320_D.visible = this.field_146344_y;
         this.field_146321_E.visible = this.field_146344_y;
-        this.field_146322_F.visible = this.field_146344_y && WorldType.worldTypes[this.field_146331_K] == WorldType.FLAT;
+        this.field_146322_F.visible = this.field_146344_y && WorldType.worldTypes[this.field_146331_K].isCustomizable();
 
         if (this.field_146344_y)
         {
