--- ../src-base/minecraft/net/minecraft/client/gui/GuiCreateWorld.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiCreateWorld.java
@@ -320,7 +320,7 @@
             }
             else if (p_146284_1_.field_146127_k == 8)
             {
-                this.field_146297_k.func_147108_a(new GuiCreateFlatWorld(this, this.field_146334_a));
+                WorldType.field_77139_a[field_146331_K].onCustomizeButton(field_146297_k, this);
             }
         }
     }
@@ -338,7 +338,7 @@
         this.field_146326_C.field_146125_m = this.field_146344_y;
         this.field_146320_D.field_146125_m = this.field_146344_y;
         this.field_146321_E.field_146125_m = this.field_146344_y;
-        this.field_146322_F.field_146125_m = this.field_146344_y && WorldType.field_77139_a[this.field_146331_K] == WorldType.field_77138_c;
+        this.field_146322_F.field_146125_m = this.field_146344_y && WorldType.field_77139_a[this.field_146331_K].isCustomizable();
 
         if (this.field_146344_y)
         {
