--- ../src-base/minecraft/net/minecraft/client/gui/GuiKeyBindingList.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiKeyBindingList.java
@@ -62,7 +62,7 @@
 
     protected int func_148137_d()
     {
-        return super.func_148137_d() + 15;
+        return super.func_148137_d() + 35;
     }
 
     public int func_148139_c()
@@ -113,7 +113,7 @@
         {
             this.field_148282_b = p_i45029_2_;
             this.field_148283_c = I18n.func_135052_a(p_i45029_2_.func_151464_g(), new Object[0]);
-            this.field_148280_d = new GuiButton(0, 0, 0, 75, 20, I18n.func_135052_a(p_i45029_2_.func_151464_g(), new Object[0]));
+            this.field_148280_d = new GuiButton(0, 0, 0, 95, 20, I18n.func_135052_a(p_i45029_2_.func_151464_g(), new Object[0]));
             this.field_148281_e = new GuiButton(0, 0, 0, 50, 20, I18n.func_135052_a("controls.reset", new Object[0]));
         }
 
@@ -121,20 +121,20 @@
         {
             boolean flag = GuiKeyBindingList.this.field_148191_k.field_146491_f == this.field_148282_b;
             GuiKeyBindingList.this.field_148189_l.field_71466_p.func_78276_b(this.field_148283_c, p_180790_2_ + 90 - GuiKeyBindingList.this.field_148188_n, p_180790_3_ + p_180790_5_ / 2 - GuiKeyBindingList.this.field_148189_l.field_71466_p.field_78288_b / 2, 16777215);
-            this.field_148281_e.field_146128_h = p_180790_2_ + 190;
+            this.field_148281_e.field_146128_h = p_180790_2_ + 210;
             this.field_148281_e.field_146129_i = p_180790_3_;
-            this.field_148281_e.field_146124_l = this.field_148282_b.func_151463_i() != this.field_148282_b.func_151469_h();
+            this.field_148281_e.field_146124_l = this.field_148282_b.func_151463_i() != this.field_148282_b.func_151469_h() || this.field_148282_b.getKeyModifier() != this.field_148282_b.getKeyModifierDefault();
             this.field_148281_e.func_146112_a(GuiKeyBindingList.this.field_148189_l, p_180790_6_, p_180790_7_);
             this.field_148280_d.field_146128_h = p_180790_2_ + 105;
             this.field_148280_d.field_146129_i = p_180790_3_;
-            this.field_148280_d.field_146126_j = GameSettings.func_74298_c(this.field_148282_b.func_151463_i());
+            this.field_148280_d.field_146126_j = this.field_148282_b.getKeyModifier().getLocalizedComboName(GameSettings.func_74298_c(this.field_148282_b.func_151463_i()));
             boolean flag1 = false;
 
             if (this.field_148282_b.func_151463_i() != 0)
             {
                 for (KeyBinding keybinding : GuiKeyBindingList.this.field_148189_l.field_71474_y.field_74324_K)
                 {
-                    if (keybinding != this.field_148282_b && keybinding.func_151463_i() == this.field_148282_b.func_151463_i())
+                    if (keybinding != this.field_148282_b && keybinding.conflicts(this.field_148282_b))
                     {
                         flag1 = true;
                         break;
@@ -163,6 +163,7 @@
             }
             else if (this.field_148281_e.func_146116_c(GuiKeyBindingList.this.field_148189_l, p_148278_2_, p_148278_3_))
             {
+                this.field_148282_b.setKeyModifier(this.field_148282_b.getKeyModifierDefault());
                 GuiKeyBindingList.this.field_148189_l.field_71474_y.func_151440_a(this.field_148282_b, this.field_148282_b.func_151469_h());
                 KeyBinding.func_74508_b();
                 return true;
