--- ../src-base/minecraft/net/minecraft/client/gui/GuiMultiplayer.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiMultiplayer.java
@@ -40,6 +40,7 @@
     public GuiMultiplayer(GuiScreen p_i1040_1_)
     {
         this.field_146798_g = p_i1040_1_;
+        net.minecraftforge.fml.client.FMLClientHandler.instance().setupServerList();
     }
 
     public void func_73866_w_()
@@ -372,7 +373,7 @@
 
     private void func_146791_a(ServerData p_146791_1_)
     {
-        this.field_146297_k.func_147108_a(new GuiConnecting(this, this.field_146297_k, p_146791_1_));
+        net.minecraftforge.fml.client.FMLClientHandler.instance().connectToServer(this, p_146791_1_);
     }
 
     public void func_146790_a(int p_146790_1_)
