--- ../src-base/minecraft/net/minecraft/client/gui/GuiNewChat.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiNewChat.java
@@ -134,6 +134,10 @@
 
     public void func_146234_a(ITextComponent p_146234_1_, int p_146234_2_)
     {
+        net.minecraftforge.client.event.PrintChatMessageEvent event = net.minecraftforge.client.ForgeHooksClient.onPrintChatMessage(p_146234_1_, p_146234_2_);
+        if(event.isCanceled() || event.getMessage() == null) return;
+        p_146234_1_ = event.getMessage();
+        p_146234_2_ = event.getChatLineId();
         this.func_146237_a(p_146234_1_, p_146234_2_, this.field_146247_f.field_71456_v.func_73834_c(), false);
         field_146249_a.info("[CHAT] " + field_184062_f.join(field_184061_a.split(p_146234_1_.func_150260_c())));
     }
