--- ../src-base/minecraft/net/minecraft/client/gui/GuiSlot.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiSlot.java
@@ -294,16 +294,7 @@
         GL11.glDisable(GL11.GL_LIGHTING);
         GL11.glDisable(GL11.GL_FOG);
         Tessellator tessellator = Tessellator.field_78398_a;
-        this.field_148161_k.func_110434_K().func_110577_a(Gui.field_110325_k);
-        GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
-        float f1 = 32.0F;
-        tessellator.func_78382_b();
-        tessellator.func_78378_d(2105376);
-        tessellator.func_78374_a((double)this.field_148152_e, (double)this.field_148154_c, 0.0D, (double)((float)this.field_148152_e / f1), (double)((float)(this.field_148154_c + (int)this.field_148169_q) / f1));
-        tessellator.func_78374_a((double)this.field_148151_d, (double)this.field_148154_c, 0.0D, (double)((float)this.field_148151_d / f1), (double)((float)(this.field_148154_c + (int)this.field_148169_q) / f1));
-        tessellator.func_78374_a((double)this.field_148151_d, (double)this.field_148153_b, 0.0D, (double)((float)this.field_148151_d / f1), (double)((float)(this.field_148153_b + (int)this.field_148169_q) / f1));
-        tessellator.func_78374_a((double)this.field_148152_e, (double)this.field_148153_b, 0.0D, (double)((float)this.field_148152_e / f1), (double)((float)(this.field_148153_b + (int)this.field_148169_q) / f1));
-        tessellator.func_78381_a();
+        drawContainerBackground(tessellator);
         l1 = this.field_148152_e + this.field_148155_a / 2 - this.func_148139_c() / 2 + 2;
         i2 = this.field_148153_b + 4 - (int)this.field_148169_q;
 
@@ -475,4 +466,18 @@
     {
         return this.field_148149_f;
     }
+
+    protected void drawContainerBackground(Tessellator tessellator)
+    {
+        this.field_148161_k.func_110434_K().func_110577_a(Gui.field_110325_k);
+        GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
+        float f1 = 32.0F;
+        tessellator.func_78382_b();
+        tessellator.func_78378_d(2105376);
+        tessellator.func_78374_a((double)this.field_148152_e, (double)this.field_148154_c, 0.0D, (double)((float)this.field_148152_e / f1), (double)((float)(this.field_148154_c + (int)this.field_148169_q) / f1));
+        tessellator.func_78374_a((double)this.field_148151_d, (double)this.field_148154_c, 0.0D, (double)((float)this.field_148151_d / f1), (double)((float)(this.field_148154_c + (int)this.field_148169_q) / f1));
+        tessellator.func_78374_a((double)this.field_148151_d, (double)this.field_148153_b, 0.0D, (double)((float)this.field_148151_d / f1), (double)((float)(this.field_148153_b + (int)this.field_148169_q) / f1));
+        tessellator.func_78374_a((double)this.field_148152_e, (double)this.field_148153_b, 0.0D, (double)((float)this.field_148152_e / f1), (double)((float)(this.field_148153_b + (int)this.field_148169_q) / f1));
+        tessellator.func_78381_a();
+    }
 }
