--- a/net/minecraft/client/gui/GuiSlot.java
+++ b/net/minecraft/client/gui/GuiSlot.java
@@ -144,15 +144,8 @@
          GlStateManager.func_179106_n();
          Tessellator tessellator = Tessellator.func_178181_a();
          BufferBuilder bufferbuilder = tessellator.func_178180_c();
-         this.field_148161_k.func_110434_K().func_110577_a(Gui.field_110325_k);
-         GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
-         float f = 32.0F;
-         bufferbuilder.func_181668_a(7, DefaultVertexFormats.field_181709_i);
-         bufferbuilder.func_181662_b((double)this.field_148152_e, (double)this.field_148154_c, 0.0D).func_187315_a((double)((float)this.field_148152_e / 32.0F), (double)((float)(this.field_148154_c + (int)this.field_148169_q) / 32.0F)).func_181669_b(32, 32, 32, 255).func_181675_d();
-         bufferbuilder.func_181662_b((double)this.field_148151_d, (double)this.field_148154_c, 0.0D).func_187315_a((double)((float)this.field_148151_d / 32.0F), (double)((float)(this.field_148154_c + (int)this.field_148169_q) / 32.0F)).func_181669_b(32, 32, 32, 255).func_181675_d();
-         bufferbuilder.func_181662_b((double)this.field_148151_d, (double)this.field_148153_b, 0.0D).func_187315_a((double)((float)this.field_148151_d / 32.0F), (double)((float)(this.field_148153_b + (int)this.field_148169_q) / 32.0F)).func_181669_b(32, 32, 32, 255).func_181675_d();
-         bufferbuilder.func_181662_b((double)this.field_148152_e, (double)this.field_148153_b, 0.0D).func_187315_a((double)((float)this.field_148152_e / 32.0F), (double)((float)(this.field_148153_b + (int)this.field_148169_q) / 32.0F)).func_181669_b(32, 32, 32, 255).func_181675_d();
-         tessellator.func_78381_a();
+         // Forge: background rendering moved into separate method.
+         this.drawContainerBackground(tessellator);
          int k = this.field_148152_e + this.field_148155_a / 2 - this.func_148139_c() / 2 + 2;
          int l = this.field_148153_b + 4 - (int)this.field_148169_q;
          if (this.field_148165_u) {
@@ -369,4 +362,17 @@
    public int func_148146_j() {
       return this.field_148149_f;
    }
+
+   protected void drawContainerBackground(Tessellator tessellator) {
+      BufferBuilder buffer = tessellator.func_178180_c();
+      this.field_148161_k.func_110434_K().func_110577_a(Gui.field_110325_k);
+      GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
+      float scale = 32.0F;
+      buffer.func_181668_a(7, DefaultVertexFormats.field_181709_i);
+      buffer.func_181662_b((double)this.field_148152_e,  (double)this.field_148154_c, 0.0D).func_187315_a(this.field_148152_e  / scale, (this.field_148154_c + (int)this.field_148169_q) / scale).func_181669_b(32, 32, 32, 255).func_181675_d();
+      buffer.func_181662_b((double)this.field_148151_d, (double)this.field_148154_c, 0.0D).func_187315_a(this.field_148151_d / scale, (this.field_148154_c + (int)this.field_148169_q) / scale).func_181669_b(32, 32, 32, 255).func_181675_d();
+      buffer.func_181662_b((double)this.field_148151_d, (double)this.field_148153_b,    0.0D).func_187315_a(this.field_148151_d / scale, (this.field_148153_b    + (int)this.field_148169_q) / scale).func_181669_b(32, 32, 32, 255).func_181675_d();
+      buffer.func_181662_b((double)this.field_148152_e,  (double)this.field_148153_b,    0.0D).func_187315_a(this.field_148152_e  / scale, (this.field_148153_b    + (int)this.field_148169_q) / scale).func_181669_b(32, 32, 32, 255).func_181675_d();
+      tessellator.func_78381_a();
+   }
 }
