--- ../src-base/minecraft/net/minecraft/client/gui/GuiSpectator.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiSpectator.java
@@ -109,7 +109,7 @@
             GlStateManager.func_179131_c(f, f, f, p_175266_4_);
             p_175266_5_.func_178663_a(f, i);
             GlStateManager.func_179121_F();
-            String s = String.valueOf((Object)GameSettings.func_74298_c(this.field_175268_g.field_71474_y.field_151456_ac[p_175266_1_].func_151463_i()));
+            String s = String.valueOf(this.field_175268_g.field_71474_y.field_151456_ac[p_175266_1_].getDisplayName());
 
             if (i > 3 && p_175266_5_.func_178662_A_())
             {
