--- ../src-base/minecraft/net/minecraft/client/gui/GuiUtilRenderComponents.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiUtilRenderComponents.java
@@ -73,6 +73,7 @@
                         s3 = s4;
                     }
 
+                    s3 = FontRenderer.func_78282_e(s2) + s3; //Forge: Fix chat formatting not surviving line wrapping.
                     ChatComponentText chatcomponenttext2 = new ChatComponentText(s3);
                     chatcomponenttext2.func_150255_a(ichatcomponent1.func_150256_b().func_150232_l());
                     list1.add(j + 1, chatcomponenttext2);
