--- a/net/minecraft/client/gui/MapRenderer.java
+++ b/net/minecraft/client/gui/MapRenderer.java
@@ -118,6 +_,7 @@
 
          for(MapDecoration mapdecoration : this.f_93280_.m_164811_()) {
             if (!p_93294_ || mapdecoration.m_77809_()) {
+               if (mapdecoration.render(k)) { k++; continue; }
                p_93292_.m_85836_();
                p_93292_.m_252880_(0.0F + (float)mapdecoration.m_77804_() / 2.0F + 64.0F, 0.0F + (float)mapdecoration.m_77805_() / 2.0F + 64.0F, -0.02F);
                p_93292_.m_252781_(Axis.f_252403_.m_252977_((float)(mapdecoration.m_77806_() * 360) / 16.0F));
