--- a/net/minecraft/client/gui/ResourceLoadProgressGui.java
+++ b/net/minecraft/client/gui/ResourceLoadProgressGui.java
@@ -83,6 +83,7 @@
          this.func_212972_a(i / 2 - 150, j / 4 * 3, i / 2 + 150, j / 4 * 3 + 10, this.field_212978_f, 1.0F - MathHelper.func_76131_a(f, 0.0F, 1.0F));
       }
 
+      net.minecraftforge.fml.client.ClientModLoader.renderProgressText();
       if (f >= 2.0F) {
          this.field_212974_b.func_213268_a((LoadingGui)null);
       }
