--- ../src-base/minecraft/net/minecraft/client/gui/ServerListEntryNormal.java
+++ ../src-work/minecraft/net/minecraft/client/gui/ServerListEntryNormal.java
@@ -83,7 +83,7 @@
         boolean flag1 = this.field_148301_e.field_82821_f < 338;
         boolean flag2 = flag || flag1;
         this.field_148300_d.field_71466_p.func_78276_b(this.field_148301_e.field_78847_a, p_192634_2_ + 32 + 3, p_192634_3_ + 1, 16777215);
-        List<String> list = this.field_148300_d.field_71466_p.func_78271_c(this.field_148301_e.field_78843_d, p_192634_4_ - 32 - 2);
+        List<String> list = this.field_148300_d.field_71466_p.func_78271_c(net.minecraftforge.fml.client.FMLClientHandler.instance().fixDescription(this.field_148301_e.field_78843_d), p_192634_4_ - 32 - 2);
 
         for (int i = 0; i < Math.min(list.size(), 2); ++i)
         {
@@ -177,6 +177,11 @@
         int i1 = p_192634_6_ - p_192634_2_;
         int j1 = p_192634_7_ - p_192634_3_;
 
+        String tooltip = net.minecraftforge.fml.client.FMLClientHandler.instance().enhanceServerListEntry(this, this.field_148301_e, p_192634_2_, p_192634_4_, p_192634_3_, i1, j1);
+        if (tooltip != null)
+        {
+            this.field_148303_c.func_146793_a(tooltip);
+        } else
         if (i1 >= p_192634_4_ - 15 && i1 <= p_192634_4_ - 5 && j1 >= 0 && j1 <= 8)
         {
             this.field_148303_c.func_146793_a(s1);
