--- a/net/minecraft/client/particle/EnchantmentTableParticle.java
+++ b/net/minecraft/client/particle/EnchantmentTableParticle.java
@@ -73,6 +_,7 @@
          this.x = this.xStart + this.xd * (double)f;
          this.y = this.yStart + this.yd * (double)f - (double)(f1 * 1.2F);
          this.z = this.zStart + this.zd * (double)f;
+         this.setPos(this.x, this.y, this.z); // FORGE: update the particle's bounding box
       }
    }
 
