--- a/net/minecraft/client/particle/EnchantmentTableParticle.java
+++ b/net/minecraft/client/particle/EnchantmentTableParticle.java
@@ -73,6 +_,7 @@
          this.f_107212_ = this.f_106461_ + this.f_107215_ * (double)f;
          this.f_107213_ = this.f_106462_ + this.f_107216_ * (double)f - (double)(f1 * 1.2F);
          this.f_107214_ = this.f_106460_ + this.f_107217_ * (double)f;
+         this.m_107264_(this.f_107212_, this.f_107213_, this.f_107214_); // FORGE: update the particle's bounding box
       }
    }
 
