--- ../src_base/minecraft/net/minecraft/client/particle/EntityDiggingFX.java
+++ ../src_work/minecraft/net/minecraft/client/particle/EntityDiggingFX.java
@@ -11,20 +11,22 @@
 public class EntityDiggingFX extends EntityFX
 {
     private Block blockInstance;
+    private int side;
 
     public EntityDiggingFX(World par1World, double par2, double par4, double par6, double par8, double par10, double par12, Block par14Block, int par15, int par16, RenderEngine par17RenderEngine)
     {
         super(par1World, par2, par4, par6, par8, par10, par12);
         this.blockInstance = par14Block;
-        this.setParticleIcon(par17RenderEngine, par14Block.getIcon(0, par16));
+        this.setParticleIcon(par17RenderEngine, par14Block.getIcon(par15, par16));
         this.particleGravity = par14Block.blockParticleGravity;
         this.particleRed = this.particleGreen = this.particleBlue = 0.6F;
         this.particleScale /= 2.0F;
+        this.side = par15;
     }
 
     public EntityDiggingFX func_70596_a(int par1, int par2, int par3)
     {
-        if (this.blockInstance == Block.grass)
+        if (this.blockInstance == Block.grass && this.side != 1)
         {
             return this;
         }
