--- ../src-base/minecraft/net/minecraft/client/particle/EntityDiggingFX.java
+++ ../src-work/minecraft/net/minecraft/client/particle/EntityDiggingFX.java
@@ -12,20 +12,27 @@
 {
     private Block field_145784_a;
     private static final String __OBFID = "CL_00000932";
+    private int side;
 
     public EntityDiggingFX(World p_i1234_1_, double p_i1234_2_, double p_i1234_4_, double p_i1234_6_, double p_i1234_8_, double p_i1234_10_, double p_i1234_12_, Block p_i1234_14_, int p_i1234_15_)
     {
+        this(p_i1234_1_, p_i1234_2_, p_i1234_4_, p_i1234_6_, p_i1234_8_, p_i1234_10_, p_i1234_12_, p_i1234_14_, p_i1234_15_, p_i1234_1_.field_73012_v.nextInt(6));
+    }
+
+    public EntityDiggingFX(World p_i1234_1_, double p_i1234_2_, double p_i1234_4_, double p_i1234_6_, double p_i1234_8_, double p_i1234_10_, double p_i1234_12_, Block p_i1234_14_, int p_i1234_15_, int side)
+    {
         super(p_i1234_1_, p_i1234_2_, p_i1234_4_, p_i1234_6_, p_i1234_8_, p_i1234_10_, p_i1234_12_);
         this.field_145784_a = p_i1234_14_;
-        this.func_110125_a(p_i1234_14_.func_149691_a(0, p_i1234_15_));
+        this.func_110125_a(p_i1234_14_.func_149691_a(side, p_i1234_15_));
         this.field_70545_g = p_i1234_14_.field_149763_I;
         this.field_70552_h = this.field_70553_i = this.field_70551_j = 0.6F;
         this.field_70544_f /= 2.0F;
+        this.side = side;
     }
 
     public EntityDiggingFX func_70596_a(int p_70596_1_, int p_70596_2_, int p_70596_3_)
     {
-        if (this.field_145784_a == Blocks.field_150349_c)
+        if (this.field_145784_a == Blocks.field_150349_c && this.side != 1)
         {
             return this;
         }
