--- a/net/minecraft/client/particle/FireworkParticles.java
+++ b/net/minecraft/client/particle/FireworkParticles.java
@@ -263,6 +_,10 @@
                     intlist = IntList.of(DyeColor.BLACK.getFireworkColor());
                 }
 
+                var factory = net.minecraftforge.client.FireworkShapeFactoryRegistry.get(fireworkexplosion1.shape());
+                if (factory != null)
+                    factory.build(this, flag3, flag4, intlist.toIntArray(), intlist1.toIntArray());
+                else
                 switch (fireworkexplosion1.shape()) {
                     case SMALL_BALL:
                         this.createParticleBall(0.25, 2, intlist, intlist1, flag3, flag4);
