--- a/net/minecraft/client/particle/Particle.java
+++ b/net/minecraft/client/particle/Particle.java
@@ -226,4 +_,16 @@
    public Optional<ParticleGroup> m_142654_() {
       return Optional.empty();
    }
+
+    /**
+     * Forge added method that controls if a particle should be culled to it's bounding box.
+     * Default behaviour is culling enabled
+     */
+    public boolean shouldCull() {
+        return true;
+    }
+
+    public net.minecraft.world.phys.Vec3 getPos() {
+       return new net.minecraft.world.phys.Vec3(this.f_107212_, this.f_107213_, this.f_107214_);
+    }
 }
