--- ../src-base/minecraft/net/minecraft/client/particle/ParticleDigging.java
+++ ../src-work/minecraft/net/minecraft/client/particle/ParticleDigging.java
@@ -5,6 +5,7 @@
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.VertexBuffer;
+import net.minecraft.client.renderer.color.IBlockColor;
 import net.minecraft.entity.Entity;
 import net.minecraft.init.Blocks;
 import net.minecraft.util.math.BlockPos;
@@ -37,7 +38,7 @@
     {
         this.sourcePos = pos;
 
-        if (this.sourceState.getBlock() == Blocks.GRASS)
+        if (this.sourceState.getBlock() == Blocks.GRASS || (this.sourceState.getBlock() instanceof IBlockColor && ((IBlockColor) this.sourceState.getBlock()).preventParticleDiggingColoring(sourceState, sourcePos)) || Minecraft.getMinecraft().getBlockColors().getPreventParticleDiggingColoring(sourceState, sourcePos))
         {
             return this;
         }
