--- a/net/minecraft/client/particle/VibrationSignalParticle.java
+++ b/net/minecraft/client/particle/VibrationSignalParticle.java
@@ -76,6 +_,7 @@
                 this.x = Mth.lerp(d0, this.x, vec3.x());
                 this.y = Mth.lerp(d0, this.y, vec3.y());
                 this.z = Mth.lerp(d0, this.z, vec3.z());
+                this.setPos(this.x, this.y, this.z); // FORGE: Update the particle's bounding box
                 double d1 = this.x - vec3.x();
                 double d2 = this.y - vec3.y();
                 double d3 = this.z - vec3.z();
