--- a/net/minecraft/client/particle/VibrationSignalParticle.java
+++ b/net/minecraft/client/particle/VibrationSignalParticle.java
@@ -105,6 +_,7 @@
             this.f_107212_ = Mth.m_14139_(d0, this.f_107212_, vec3.m_7096_());
             this.f_107213_ = Mth.m_14139_(d0, this.f_107213_, vec3.m_7098_());
             this.f_107214_ = Mth.m_14139_(d0, this.f_107214_, vec3.m_7094_());
+            this.m_107264_(this.f_107212_, this.f_107213_, this.f_107214_); // FORGE: Update the particle's bounding box
             double d1 = this.f_107212_ - vec3.m_7096_();
             double d2 = this.f_107213_ - vec3.m_7098_();
             double d3 = this.f_107214_ - vec3.m_7094_();
