--- a/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -86,6 +_,8 @@
 
         boolean isSupported = BlockPos.betweenClosedStream(this.calculateSupportBox()).allMatch(pos -> {
             BlockState state = this.level().getBlockState(pos);
+            if (net.minecraft.world.level.block.Block.canSupportCenter(this.level(), pos, this.getDirection()))
+                return true;
             return state.isSolid() || DiodeBlock.isDiode(state);
         });
         return isSupported && this.canCoexist(false);
