--- a/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -59,6 +_,8 @@
         } else {
             boolean flag = BlockPos.betweenClosedStream(this.calculateSupportBox()).allMatch(p_341437_ -> {
                 BlockState blockstate = this.level().getBlockState(p_341437_);
+                if (net.minecraft.world.level.block.Block.canSupportCenter(this.level(), p_341437_, this.direction))
+                    return true;
                 return blockstate.isSolid() || DiodeBlock.isDiode(blockstate);
             });
             return !flag ? false : this.level().getEntities(this, this.getBoundingBox(), HANGING_ENTITY).isEmpty();
