--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -77,7 +_,7 @@
             return InteractionResult.SUCCESS;
         }
 
-        if (player.getItemInHand(hand).is(Items.SHEARS)) {
+        if (player.getItemInHand(hand).canPerformAction(net.minecraftforge.common.ToolActions.SHEARS_HARVEST)) {
             InteractionResult result = super.interact(player, hand, location);
             if (result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
                 return result;
