--- a/net/minecraft/tags/BlockTags.java
+++ b/net/minecraft/tags/BlockTags.java
@@ -131,6 +_,14 @@
       return f_13053_.m_13244_(p_13117_);
    }
 
+   public static net.minecraftforge.common.Tags.IOptionalNamedTag<Block> bindOptional(net.minecraft.resources.ResourceLocation name) {
+       return bindOptional(name, null);
+   }
+
+   public static net.minecraftforge.common.Tags.IOptionalNamedTag<Block> bindOptional(net.minecraft.resources.ResourceLocation name, @javax.annotation.Nullable java.util.Set<java.util.function.Supplier<Block>> defaults) {
+      return f_13053_.createOptional(name, defaults);
+   }
+
    public static TagCollection<Block> m_13115_() {
       return f_13053_.m_13246_();
    }
