--- a/net/minecraft/tags/BlockTags.java
+++ b/net/minecraft/tags/BlockTags.java
@@ -97,6 +97,14 @@
       return field_199899_c.func_232937_a_(p_199894_0_);
    }
 
+   public static net.minecraftforge.common.Tags.IOptionalNamedTag<Block> createOptional(ResourceLocation name) {
+       return createOptional(name, null);
+   }
+
+   public static net.minecraftforge.common.Tags.IOptionalNamedTag<Block> createOptional(ResourceLocation name, @javax.annotation.Nullable java.util.Set<java.util.function.Supplier<Block>> defaults) {
+      return field_199899_c.createOptional(name, defaults);
+   }
+
    public static ITagCollection<Block> func_199896_a() {
       return field_199899_c.func_232939_b_();
    }
