--- a/net/minecraft/tags/EnchantmentTags.java
+++ b/net/minecraft/tags/EnchantmentTags.java
@@ -45,4 +_,12 @@
     private static TagKey<Enchantment> create(String p_334094_) {
         return TagKey.create(Registries.ENCHANTMENT, ResourceLocation.withDefaultNamespace(p_334094_));
     }
+
+    public static TagKey<Enchantment> create(String namepsace, String path) {
+        return create(ResourceLocation.fromNamespaceAndPath(namepsace, path));
+    }
+
+    public static TagKey<Enchantment> create(ResourceLocation name) {
+        return TagKey.create(Registries.ENCHANTMENT, name);
+    }
 }
