--- a/net/minecraft/tags/EntityTypeTags.java
+++ b/net/minecraft/tags/EntityTypeTags.java
@@ -16,6 +_,14 @@
       return HELPER.bind(p_232896_0_);
    }
 
+   public static net.minecraftforge.common.Tags.IOptionalNamedTag<EntityType<?>> createOptional(ResourceLocation name) {
+       return createOptional(name, null);
+   }
+
+   public static net.minecraftforge.common.Tags.IOptionalNamedTag<EntityType<?>> createOptional(ResourceLocation name, @javax.annotation.Nullable java.util.Set<java.util.function.Supplier<EntityType<?>>> defaults) {
+      return HELPER.createOptional(name, defaults);
+   }
+
    public static ITagCollection<EntityType<?>> getAllTags() {
       return HELPER.getAllTags();
    }
