--- a/net/minecraft/tags/FluidTags.java
+++ b/net/minecraft/tags/FluidTags.java
@@ -14,4 +_,12 @@
     private static TagKey<Fluid> create(String p_203851_) {
         return TagKey.create(Registries.FLUID, ResourceLocation.withDefaultNamespace(p_203851_));
     }
+
+    public static TagKey<Fluid> create(String namepsace, String path) {
+        return create(ResourceLocation.fromNamespaceAndPath(namepsace, path));
+    }
+
+    public static TagKey<Fluid> create(ResourceLocation name) {
+        return TagKey.create(Registries.FLUID, name);
+    }
 }
