--- a/net/minecraft/tags/GameEventTags.java
+++ b/net/minecraft/tags/GameEventTags.java
@@ -14,4 +_,12 @@
     private static TagKey<GameEvent> create(String p_203853_) {
         return TagKey.create(Registries.GAME_EVENT, ResourceLocation.withDefaultNamespace(p_203853_));
     }
+
+    public static TagKey<GameEvent> create(String namepsace, String path) {
+        return create(ResourceLocation.fromNamespaceAndPath(namepsace, path));
+    }
+
+    public static TagKey<GameEvent> create(ResourceLocation name) {
+        return TagKey.create(Registries.GAME_EVENT, name);
+    }
 }
