--- a/net/minecraft/tags/ItemTags.java
+++ b/net/minecraft/tags/ItemTags.java
@@ -41,6 +41,24 @@
    public static final Tag<Item> field_212187_B = func_199901_a("trapdoors");
    public static final Tag<Item> field_202902_o = func_199901_a("boats");
    public static final Tag<Item> field_206964_G = func_199901_a("fishes");
+   public static final Tag<Item> DYES = func_199901_a("dyes");
+   public static final Tag<Item> DYES_BLACK = func_199901_a("dyes_black");
+   public static final Tag<Item> DYES_RED = func_199901_a("dyes_red");
+   public static final Tag<Item> DYES_GREEN = func_199901_a("dyes_green");
+   public static final Tag<Item> DYES_BROWN = func_199901_a("dyes_brown");
+   public static final Tag<Item> DYES_BLUE = func_199901_a("dyes_blue");
+   public static final Tag<Item> DYES_PURPLE = func_199901_a("dyes_purple");
+   public static final Tag<Item> DYES_CYAN = func_199901_a("dyes_cyan");
+   public static final Tag<Item> DYES_LIGHT_GRAY = func_199901_a("dyes_light_gray");
+   public static final Tag<Item> DYES_GRAY = func_199901_a("dyes_gray");
+   public static final Tag<Item> DYES_PINK = func_199901_a("dyes_pink");
+   public static final Tag<Item> DYES_LIME = func_199901_a("dyes_lime");
+   public static final Tag<Item> DYES_YELLOW = func_199901_a("dyes_yellow");
+   public static final Tag<Item> DYES_LIGHT_BLUE = func_199901_a("dyes_light_blue");
+   public static final Tag<Item> DYES_MAGENTA = func_199901_a("dyes_magenta");
+   public static final Tag<Item> DYES_ORANGE = func_199901_a("dyes_orange");
+   public static final Tag<Item> DYES_WHITE = func_199901_a("dyes_white");
+   public static final Tag<Item> MUSIC_DISCS = func_199901_a("music_discs");
 
    public static void func_199902_a(TagCollection<Item> p_199902_0_) {
       field_199906_c = p_199902_0_;
