--- a/net/minecraft/tags/ItemTags.java
+++ b/net/minecraft/tags/ItemTags.java
@@ -220,4 +_,12 @@
     private static TagKey<Item> bind(final String name) {
         return TagKey.create(Registries.ITEM, Identifier.withDefaultNamespace(name));
     }
+
+    public static TagKey<Item> create(String namepsace, String path) {
+        return create(Identifier.fromNamespaceAndPath(namepsace, path));
+    }
+
+    public static TagKey<Item> create(Identifier name) {
+        return TagKey.create(Registries.ITEM, name);
+    }
 }
