--- a/net/minecraft/tags/Tag.java
+++ b/net/minecraft/tags/Tag.java
@@ -43,7 +43,7 @@
       JsonObject jsonobject = new JsonObject();
       JsonArray jsonarray = new JsonArray();
 
-      for(Tag.ITagEntry<T> itagentry : this.field_200150_c) {
+      for(Tag.ITagEntry<T> itagentry : this.func_200570_b()) {
          itagentry.func_200576_a(jsonarray, p_200571_1_);
       }
 
@@ -53,7 +53,7 @@
    }
 
    public boolean func_199685_a_(T p_199685_1_) {
-      return this.field_199889_b.contains(p_199685_1_);
+      return this.func_199885_a().contains(p_199685_1_);
    }
 
    public Collection<T> func_199885_a() {
