--- a/net/minecraft/tags/TagBuilder.java
+++ b/net/minecraft/tags/TagBuilder.java
@@ -4,7 +_,17 @@
 import java.util.List;
 import net.minecraft.resources.ResourceLocation;
 
-public class TagBuilder {
+public class TagBuilder implements net.minecraftforge.common.extensions.IForgeRawTagBuilder {
+   // FORGE: Remove entries are used for datagen.
+   private final List<TagEntry> removeEntries = new ArrayList<>();
+   public java.util.stream.Stream<TagEntry> getRemoveEntries() { return this.removeEntries.stream(); }
+   // FORGE: Add an entry to be removed from this tag in datagen.
+   public TagBuilder remove(final TagEntry entry) {
+      this.removeEntries.add(entry);
+      return this;
+   }
+   // FORGE: is this tag set to replace or not?
+   private boolean replace = false;
    private final List<TagEntry> f_215897_ = new ArrayList<>();
 
    public static TagBuilder m_215899_() {
@@ -34,5 +_,21 @@
 
    public TagBuilder m_215909_(ResourceLocation p_215910_) {
       return this.m_215902_(TagEntry.m_215953_(p_215910_));
+   }
+
+   // FORGE: Set the replace property of this tag.
+   public TagBuilder replace(boolean value) {
+      this.replace = value;
+      return this;
+   }
+
+   // FORGE: Shorthand version of replace(true)
+   public TagBuilder replace() {
+      return replace(true);
+   }
+
+   // FORGE: Is this tag set to replace or not?
+   public boolean isReplace() {
+      return this.replace;
    }
 }
