--- a/net/minecraft/tags/TagManager.java
+++ b/net/minecraft/tags/TagManager.java
@@ -33,7 +_,7 @@
    }
 
    public CompletableFuture<Void> m_5540_(PreparableReloadListener.PreparationBarrier p_13482_, ResourceManager p_13483_, ProfilerFiller p_13484_, ProfilerFiller p_13485_, Executor p_13486_, Executor p_13487_) {
-      List<? extends CompletableFuture<? extends TagManager.LoadResult<?>>> list = this.f_144569_.m_206193_().map((p_203927_) -> {
+      List<? extends CompletableFuture<? extends TagManager.LoadResult<?>>> list = java.util.stream.Stream.concat(net.minecraftforge.registries.GameData.getTagWrapperRegistries(), this.f_144569_.m_206193_()).map((p_203927_) -> {
          return this.m_203907_(p_13483_, p_13486_, p_203927_);
       }).toList();
       return CompletableFuture.allOf(list.toArray((p_203906_) -> {
