--- a/net/minecraft/tags/TagManager.java
+++ b/net/minecraft/tags/TagManager.java
@@ -31,7 +_,7 @@
 
    public static String m_203918_(ResourceKey<? extends Registry<?>> p_203919_) {
       String s = f_203902_.get(p_203919_);
-      return s != null ? s : "tags/" + p_203919_.m_135782_().m_135815_();
+      return s != null ? s : "tags/" + net.minecraftforge.common.ForgeHooks.prefixNamespace(p_203919_.m_135782_());
    }
 
    public CompletableFuture<Void> m_5540_(PreparableReloadListener.PreparationBarrier p_13482_, ResourceManager p_13483_, ProfilerFiller p_13484_, ProfilerFiller p_13485_, Executor p_13486_, Executor p_13487_) {
