--- a/net/minecraft/world/biome/BiomeGenerationSettings.java
+++ b/net/minecraft/world/biome/BiomeGenerationSettings.java
@@ -64,6 +64,9 @@
    public List<Supplier<ConfiguredCarver<?>>> func_242489_a(GenerationStage.Carving p_242489_1_) {
       return this.field_242483_e.getOrDefault(p_242489_1_, ImmutableList.of());
    }
+   public Map<GenerationStage.Carving, List<Supplier<ConfiguredCarver<?>>>> getAllCarvers(){
+      return this.field_242483_e;
+   }
 
    public boolean func_242493_a(Structure<?> p_242493_1_) {
       return this.field_242485_g.stream().anyMatch((p_242494_1_) -> {
