--- ../src-base/minecraft/net/minecraft/world/biome/BiomeMesa.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeMesa.java
@@ -336,6 +336,7 @@
         protected void func_76797_b(World p_76797_1_, Random p_76797_2_)
         {
             super.func_76797_b(p_76797_1_, p_76797_2_);
+            if (net.minecraftforge.event.terraingen.TerrainGen.generateOre(p_76797_1_, p_76797_2_, field_76819_m, field_180294_c, net.minecraftforge.event.terraingen.OreGenEvent.GenerateMinable.EventType.GOLD))
             this.func_76795_a(p_76797_1_, p_76797_2_, 20, this.field_76819_m, 32, 80);
         }
     }
