--- ../src-base/minecraft/net/minecraft/world/biome/BiomeGenHills.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeGenHills.java
@@ -58,7 +58,7 @@
             int i1 = p_180624_2_.nextInt(16);
             BlockPos blockpos1 = p_180624_3_.func_177982_a(k, l, i1);
 
-            if (p_180624_1_.func_180495_p(blockpos1).func_177230_c() == Blocks.field_150348_b)
+            if (p_180624_1_.func_180495_p(blockpos1).func_177230_c().isReplaceableOreGen(p_180624_1_, blockpos1, net.minecraft.block.state.pattern.BlockHelper.func_177642_a(Blocks.field_150348_b)))
             {
                 p_180624_1_.func_180501_a(blockpos1, Blocks.field_150412_bA.func_176223_P(), 2);
             }
