--- ../src-base/minecraft/net/minecraft/world/biome/BiomeMesa.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeMesa.java
@@ -318,13 +318,13 @@
     @SideOnly(Side.CLIENT)
     public int func_180625_c(BlockPos p_180625_1_)
     {
-        return 10387789;
+        return getModdedBiomeFoliageColor(10387789);
     }
 
     @SideOnly(Side.CLIENT)
     public int func_180627_b(BlockPos p_180627_1_)
     {
-        return 9470285;
+        return getModdedBiomeGrassColor(9470285);
     }
 
     class Decorator extends BiomeDecorator
@@ -336,6 +336,7 @@
         protected void func_76797_b(World p_76797_1_, Random p_76797_2_)
         {
             super.func_76797_b(p_76797_1_, p_76797_2_);
+            if (net.minecraftforge.event.terraingen.TerrainGen.generateOre(p_76797_1_, p_76797_2_, field_76819_m, field_180294_c, net.minecraftforge.event.terraingen.OreGenEvent.GenerateMinable.EventType.GOLD))
             this.func_76795_a(p_76797_1_, p_76797_2_, 20, this.field_76819_m, 32, 80);
         }
     }
