--- ../src-base/minecraft/net/minecraft/world/biome/BiomeGenPlains.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeGenPlains.java
@@ -3,6 +3,7 @@
 import java.util.Random;
 import net.minecraft.block.BlockFlower;
 import net.minecraft.entity.passive.EntityHorse;
+import net.minecraft.init.Blocks;
 import net.minecraft.world.World;
 
 public class BiomeGenPlains extends BiomeGenBase
@@ -19,6 +20,15 @@
         this.field_76760_I.field_76832_z = -999;
         this.field_76760_I.field_76802_A = 4;
         this.field_76760_I.field_76803_B = 10;
+        this.flowers.clear();
+        this.addFlower(Blocks.field_150328_O,    4,  3);
+        this.addFlower(Blocks.field_150328_O,    5,  3);
+        this.addFlower(Blocks.field_150328_O,    6,  3);
+        this.addFlower(Blocks.field_150328_O,    7,  3);
+        this.addFlower(Blocks.field_150328_O,    0, 20);
+        this.addFlower(Blocks.field_150328_O,    3, 20);
+        this.addFlower(Blocks.field_150328_O,    8, 20);
+        this.addFlower(Blocks.field_150327_N, 0, 30);
     }
 
     public String func_150572_a(Random p_150572_1_, int p_150572_2_, int p_150572_3_, int p_150572_4_)
