--- ../src-base/minecraft/net/minecraft/world/biome/BiomeSwamp.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeSwamp.java
@@ -79,6 +79,7 @@
     {
         super.func_180624_a(p_180624_1_, p_180624_2_, p_180624_3_);
 
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, new net.minecraft.util.math.ChunkPos(p_180624_3_), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.FOSSIL))
         if (p_180624_2_.nextInt(64) == 0)
         {
             (new WorldGenFossils()).func_180709_b(p_180624_1_, p_180624_2_, p_180624_3_);
@@ -89,12 +90,18 @@
     public int func_180627_b(BlockPos p_180627_1_)
     {
         double d0 = field_180281_af.func_151601_a((double)p_180627_1_.func_177958_n() * 0.0225D, (double)p_180627_1_.func_177952_p() * 0.0225D);
-        return d0 < -0.1D ? 5011004 : 6975545;
+        return getModdedBiomeGrassColor(d0 < -0.1D ? 5011004 : 6975545);
     }
 
     @SideOnly(Side.CLIENT)
     public int func_180625_c(BlockPos p_180625_1_)
     {
-        return 6975545;
+        return getModdedBiomeFoliageColor(6975545);
     }
+
+    @Override
+    public void addDefaultFlowers()
+    {
+        addFlower(Blocks.field_150328_O.func_176223_P().func_177226_a(Blocks.field_150328_O.func_176494_l(), BlockFlower.EnumFlowerType.BLUE_ORCHID), 10);
+    }
 }
