--- ../src-base/minecraft/net/minecraft/world/biome/BiomeDesert.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeDesert.java
@@ -47,7 +47,7 @@
     {
         super.func_180624_a(p_180624_1_, p_180624_2_, p_180624_3_);
 
-        if (p_180624_2_.nextInt(1000) == 0)
+        for (int f_i_o = net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, p_180624_3_, net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.DESERT_WELL, p_180624_2_.nextInt(1000) == 0 ? 1 : 0); f_i_o > 0; f_i_o--)
         {
             int i = p_180624_2_.nextInt(16) + 8;
             int j = p_180624_2_.nextInt(16) + 8;
@@ -55,7 +55,7 @@
             (new WorldGenDesertWells()).func_180709_b(p_180624_1_, p_180624_2_, blockpos);
         }
 
-        if (p_180624_2_.nextInt(64) == 0)
+        for (int f_i_o = net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, p_180624_3_, net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.FOSSIL, p_180624_2_.nextInt(64) == 0 ? 1 : 0); f_i_o > 0; f_i_o--)
         {
             (new WorldGenFossils()).func_180709_b(p_180624_1_, p_180624_2_, p_180624_3_);
         }
