--- a/net/minecraft/world/biome/BiomeGenerationSettings.java
+++ b/net/minecraft/world/biome/BiomeGenerationSettings.java
@@ -103,6 +103,16 @@
       private final List<List<Supplier<ConfiguredFeature<?, ?>>>> field_242506_c = Lists.newArrayList();
       private final List<Supplier<StructureFeature<?, ?>>> field_242507_d = Lists.newArrayList();
 
+      public Builder() {
+      }
+
+      public Builder(BiomeGenerationSettings orig) {
+         field_242504_a = Optional.ofNullable(orig.field_242482_d);
+         orig.field_242483_e.forEach((k,l)->field_242505_b.put(k, new java.util.ArrayList<>(l)));
+         orig.field_242484_f.forEach(l->field_242506_c.add(new java.util.ArrayList<>(l)));
+         field_242507_d.addAll(orig.field_242485_g);
+      }
+
       public BiomeGenerationSettings.Builder func_242517_a(ConfiguredSurfaceBuilder<?> p_242517_1_) {
          return this.func_242519_a(() -> {
             return p_242517_1_;
