--- ../src-base/minecraft/net/minecraft/world/biome/BiomeGenJungle.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeGenJungle.java
@@ -68,10 +68,12 @@
         super.func_180624_a(p_180624_1_, p_180624_2_, p_180624_3_);
         int i = p_180624_2_.nextInt(16) + 8;
         int j = p_180624_2_.nextInt(16) + 8;
-        int k = p_180624_2_.nextInt(p_180624_1_.func_175645_m(p_180624_3_.func_177982_a(i, 0, j)).func_177956_o() * 2);
-        (new WorldGenMelon()).func_180709_b(p_180624_1_, p_180624_2_, p_180624_3_.func_177982_a(i, k, j));
+        int height = p_180624_1_.func_175645_m(p_180624_3_.func_177982_a(i, 0, j)).func_177956_o() * 2; // could == 0, which crashes nextInt
+        if (height < 1) height = 1;
+        int k = p_180624_2_.nextInt(height);
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, p_180624_3_, net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.PUMPKIN))(new WorldGenMelon()).func_180709_b(p_180624_1_, p_180624_2_, p_180624_3_.func_177982_a(i, k, j));
         WorldGenVines worldgenvines = new WorldGenVines();
-
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, p_180624_3_, net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.GRASS)){
         for (j = 0; j < 50; ++j)
         {
             k = p_180624_2_.nextInt(16) + 8;
@@ -79,5 +81,6 @@
             int i1 = p_180624_2_.nextInt(16) + 8;
             worldgenvines.func_180709_b(p_180624_1_, p_180624_2_, p_180624_3_.func_177982_a(k, 128, i1));
         }
+        }
     }
 }
