--- ../src-base/minecraft/net/minecraft/world/biome/BiomeGenSwamp.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeGenSwamp.java
@@ -30,6 +30,8 @@
         this.theBiomeDecorator.grassPerChunk = 5;
         this.waterColorMultiplier = 14745518;
         this.spawnableMonsterList.add(new BiomeGenBase.SpawnListEntry(EntitySlime.class, 1, 1, 1));
+        this.flowers.clear();
+        this.addFlower(Blocks.red_mushroom_block, 0, 10);
     }
 
     public WorldGenAbstractTree func_150567_a(Random p_150567_1_)
