--- ../src-base/minecraft/net/minecraft/world/biome/BiomeGenPlains.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeGenPlains.java
@@ -68,7 +68,7 @@
             this.field_76760_I.field_76802_A = 4;
             this.field_76760_I.field_76803_B = 10;
             field_180280_ag.func_180710_a(BlockDoublePlant.EnumPlantType.GRASS);
-
+            if(net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, p_180624_3_, net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.GRASS)){
             for (int i = 0; i < 7; ++i)
             {
                 int j = p_180624_2_.nextInt(16) + 8;
@@ -76,9 +76,10 @@
                 int l = p_180624_2_.nextInt(p_180624_1_.func_175645_m(p_180624_3_.func_177982_a(j, 0, k)).func_177956_o() + 32);
                 field_180280_ag.func_180709_b(p_180624_1_, p_180624_2_, p_180624_3_.func_177982_a(j, l, k));
             }
+            }
         }
 
-        if (this.field_150628_aC)
+        if (this.field_150628_aC && net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, p_180624_3_, net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.FLOWERS))
         {
             field_180280_ag.func_180710_a(BlockDoublePlant.EnumPlantType.SUNFLOWER);
 
@@ -103,4 +104,19 @@
         biomegenplains.field_150609_ah = 14273354;
         return biomegenplains;
     }
+
+    @Override
+    public void addDefaultFlowers()
+    {
+        BlockFlower red = net.minecraft.init.Blocks.field_150328_O;
+        BlockFlower yel = net.minecraft.init.Blocks.field_150327_N;
+        addFlower(red.func_176223_P().func_177226_a(red.func_176494_l(), BlockFlower.EnumFlowerType.ORANGE_TULIP), 3);
+        addFlower(red.func_176223_P().func_177226_a(red.func_176494_l(), BlockFlower.EnumFlowerType.RED_TULIP), 3);
+        addFlower(red.func_176223_P().func_177226_a(red.func_176494_l(), BlockFlower.EnumFlowerType.PINK_TULIP), 3);
+        addFlower(red.func_176223_P().func_177226_a(red.func_176494_l(), BlockFlower.EnumFlowerType.WHITE_TULIP), 3);
+        addFlower(red.func_176223_P().func_177226_a(red.func_176494_l(), BlockFlower.EnumFlowerType.POPPY), 20);
+        addFlower(red.func_176223_P().func_177226_a(red.func_176494_l(), BlockFlower.EnumFlowerType.HOUSTONIA), 20);
+        addFlower(red.func_176223_P().func_177226_a(red.func_176494_l(), BlockFlower.EnumFlowerType.OXEYE_DAISY), 20);
+        addFlower(yel.func_176223_P().func_177226_a(yel.func_176494_l(), BlockFlower.EnumFlowerType.DANDELION), 30);
+    }
 }
