--- ../src-base/minecraft/net/minecraft/world/biome/WorldChunkManager.java
+++ ../src-work/minecraft/net/minecraft/world/biome/WorldChunkManager.java
@@ -3,6 +3,7 @@
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Random;
 import net.minecraft.crash.CrashReport;
@@ -13,9 +14,13 @@
 import net.minecraft.world.WorldType;
 import net.minecraft.world.gen.layer.GenLayer;
 import net.minecraft.world.gen.layer.IntCache;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.terraingen.WorldTypeEvent;
+import static net.minecraft.world.biome.BiomeGenBase.*;
 
 public class WorldChunkManager
 {
+    public static ArrayList<BiomeGenBase> allowedBiomes = new ArrayList<BiomeGenBase>(Arrays.asList(field_76767_f, field_76772_c, field_76768_g, field_76784_u, field_76785_t, field_76782_w, field_76792_x));
     private GenLayer field_76944_d;
     private GenLayer field_76945_e;
     private BiomeCache field_76942_f;
@@ -26,19 +31,14 @@
     {
         this.field_76942_f = new BiomeCache(this);
         this.field_76943_g = new ArrayList();
-        this.field_76943_g.add(BiomeGenBase.field_76767_f);
-        this.field_76943_g.add(BiomeGenBase.field_76772_c);
-        this.field_76943_g.add(BiomeGenBase.field_76768_g);
-        this.field_76943_g.add(BiomeGenBase.field_76784_u);
-        this.field_76943_g.add(BiomeGenBase.field_76785_t);
-        this.field_76943_g.add(BiomeGenBase.field_76782_w);
-        this.field_76943_g.add(BiomeGenBase.field_76792_x);
+        this.field_76943_g.addAll(allowedBiomes);
     }
 
     public WorldChunkManager(long p_i1975_1_, WorldType p_i1975_3_)
     {
         this();
         GenLayer[] agenlayer = GenLayer.func_75901_a(p_i1975_1_, p_i1975_3_);
+        agenlayer = getModdedBiomeGenerators(p_i1975_3_, p_i1975_1_, agenlayer);
         this.field_76944_d = agenlayer[0];
         this.field_76945_e = agenlayer[1];
     }
@@ -242,4 +242,11 @@
     {
         this.field_76942_f.func_76838_a();
     }
+
+    public GenLayer[] getModdedBiomeGenerators(WorldType worldType, long seed, GenLayer[] original)
+    {
+        WorldTypeEvent.InitBiomeGens event = new WorldTypeEvent.InitBiomeGens(worldType, seed, original);
+        MinecraftForge.TERRAIN_GEN_BUS.post(event);
+        return event.newBiomeGens;
+    }
 }
