--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIAttackMelee.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIAttackMelee.java
@@ -21,6 +21,8 @@
     private double targetY;
     private double targetZ;
     protected final int field_188493_g = 20;
+    private int failedPathFindingPenalty = 0;
+    private boolean canPenalize = false;
 
     public EntityAIAttackMelee(EntityCreature creature, double speedIn, boolean useLongMemory)
     {
@@ -45,6 +47,19 @@
         }
         else
         {
+            if (canPenalize)
+            {
+                if (--this.delayCounter <= 0)
+                {
+                    this.entityPathEntity = this.attacker.getNavigator().getPathToEntityLiving(entitylivingbase);
+                    this.targetX = 4 + this.attacker.getRNG().nextInt(7);
+                    return this.entityPathEntity != null;
+                }
+                else
+                {
+                    return true;
+                }
+            }
             this.entityPathEntity = this.attacker.getNavigator().getPathToEntityLiving(entitylivingbase);
             return this.entityPathEntity != null;
         }
@@ -89,6 +104,23 @@
             this.targetZ = entitylivingbase.posZ;
             this.delayCounter = 4 + this.attacker.getRNG().nextInt(7);
 
+            if (this.canPenalize)
+            {
+                this.targetX += failedPathFindingPenalty;
+                if (this.attacker.getNavigator().getPath() != null)
+                {
+                    net.minecraft.pathfinding.PathPoint finalPathPoint = this.attacker.getNavigator().getPath().getFinalPathPoint();
+                    if (finalPathPoint != null && entitylivingbase.getDistanceSq(finalPathPoint.xCoord, finalPathPoint.yCoord, finalPathPoint.zCoord) < 1)
+                        failedPathFindingPenalty = 0;
+                    else
+                        failedPathFindingPenalty += 10;
+                }
+                else
+                {
+                    failedPathFindingPenalty += 10;
+                }
+            }
+
             if (d0 > 1024.0D)
             {
                 this.delayCounter += 10;
