--- a/net/minecraft/network/datasync/DataSerializers.java
+++ b/net/minecraft/network/datasync/DataSerializers.java
@@ -301,16 +_,16 @@
    };
 
    public static void registerSerializer(IDataSerializer<?> p_187189_0_) {
-      SERIALIZERS.add(p_187189_0_);
+      if (SERIALIZERS.add(p_187189_0_) >= 256) throw new RuntimeException("Vanilla DataSerializer ID limit exceeded");
    }
 
    @Nullable
    public static IDataSerializer<?> getSerializer(int p_187190_0_) {
-      return SERIALIZERS.byId(p_187190_0_);
+      return net.minecraftforge.common.ForgeHooks.getSerializer(p_187190_0_, SERIALIZERS);
    }
 
    public static int getSerializedId(IDataSerializer<?> p_187188_0_) {
-      return SERIALIZERS.getId(p_187188_0_);
+      return net.minecraftforge.common.ForgeHooks.getSerializerId(p_187188_0_, SERIALIZERS);
    }
 
    static {
