--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -435,6 +_,7 @@
                 ItemStack carried = this.getCarried();
                 player.updateTutorialInventoryAction(carried, slot.getItem(), clickAction);
                 if (!this.tryItemClickBehaviourOverride(player, clickAction, slot, clicked, carried)) {
+                    if (!net.minecraftforge.event.ForgeEventFactory.onItemStackedOn(carried, clicked, slot, clickAction, player, createCarriedSlotAccess()))
                     if (clicked.isEmpty()) {
                         if (!carried.isEmpty()) {
                             int amount = clickAction == ClickAction.PRIMARY ? carried.getCount() : 1;
@@ -648,7 +_,7 @@
                 ItemStack target = slot.getItem();
                 if (!target.isEmpty() && ItemStack.isSameItemSameComponents(itemStack, target)) {
                     int totalStack = target.getCount() + itemStack.getCount();
-                    int maxStackSize = slot.getMaxStackSize(target);
+                    int maxStackSize = Math.min(slot.getMaxStackSize(target), target.getMaxStackSize());
                     if (totalStack <= maxStackSize) {
                         itemStack.setCount(0);
                         target.setCount(totalStack);
