--- a/net/minecraft/world/inventory/EnchantmentMenu.java
+++ b/net/minecraft/world/inventory/EnchantmentMenu.java
@@ -61,7 +_,7 @@
         this.addSlot(new Slot(this.enchantSlots, 1, 35, 47) {
             @Override
             public boolean mayPlace(final ItemStack itemStack) {
-                return itemStack.is(Items.LAPIS_LAZULI);
+                return itemStack.is(net.minecraftforge.common.Tags.Items.ENCHANTING_FUELS);
             }
 
             @Override
@@ -89,23 +_,24 @@
             if (!itemStack.isEmpty() && itemStack.isEnchantable()) {
                 this.access.execute((level, pos) -> {
                     IdMap<Holder<Enchantment>> holders = level.registryAccess().lookupOrThrow(Registries.ENCHANTMENT).asHolderIdMap();
-                    int bookcases = 0;
+                    float bookcases = 0;
 
                     for (BlockPos offset : EnchantingTableBlock.BOOKSHELF_OFFSETS) {
                         if (EnchantingTableBlock.isValidBookShelf(level, pos, offset)) {
-                            bookcases++;
+                            bookcases += level.getBlockState(pos.offset(offset)).getEnchantPowerBonus(level, pos.offset(offset));
                         }
                     }
 
                     this.random.setSeed(this.enchantmentSeed.get());
 
                     for (int ixx = 0; ixx < 3; ixx++) {
-                        this.costs[ixx] = EnchantmentHelper.getEnchantmentCost(this.random, ixx, bookcases, itemStack);
+                        this.costs[ixx] = EnchantmentHelper.getEnchantmentCost(this.random, ixx, (int)bookcases, itemStack);
                         this.enchantClue[ixx] = -1;
                         this.levelClue[ixx] = -1;
                         if (this.costs[ixx] < ixx + 1) {
                             this.costs[ixx] = 0;
                         }
+                        this.costs[ixx] = net.minecraftforge.event.ForgeEventFactory.onEnchantmentLevelSet(level, pos, ixx, (int)bookcases, itemStack, costs[ixx]);
                     }
 
                     for (int ix = 0; ix < 3; ix++) {
@@ -234,7 +_,7 @@
                 if (!this.moveItemStackTo(stack, 2, 38, true)) {
                     return ItemStack.EMPTY;
                 }
-            } else if (stack.is(Items.LAPIS_LAZULI)) {
+            } else if (stack.is(net.minecraftforge.common.Tags.Items.ENCHANTING_FUELS)) {
                 if (!this.moveItemStackTo(stack, 1, 2, true)) {
                     return ItemStack.EMPTY;
                 }
