--- a/net/minecraft/world/inventory/ResultSlot.java
+++ b/net/minecraft/world/inventory/ResultSlot.java
@@ -47,6 +_,7 @@
     protected void checkTakeAchievements(ItemStack p_40185_) {
         if (this.removeCount > 0) {
             p_40185_.onCraftedBy(this.player.level(), this.player, this.removeCount);
+            net.minecraftforge.event.ForgeEventFactory.firePlayerCraftingEvent(this.player, p_40185_, this.craftSlots);
         }
 
         if (this.container instanceof RecipeCraftingHolder recipecraftingholder) {
@@ -63,7 +_,9 @@
         CraftingInput craftinginput = craftinginput$positioned.input();
         int i = craftinginput$positioned.left();
         int j = craftinginput$positioned.top();
+        net.minecraftforge.common.ForgeHooks.setCraftingPlayer(p_150638_);
         NonNullList<ItemStack> nonnulllist = p_150638_.level().getRecipeManager().getRemainingItemsFor(RecipeType.CRAFTING, craftinginput, p_150638_.level());
+        net.minecraftforge.common.ForgeHooks.setCraftingPlayer(null);
 
         for (int k = 0; k < craftinginput.height(); k++) {
             for (int l = 0; l < craftinginput.width(); l++) {
