--- ../src-base/minecraft/net/minecraft/client/settings/GameSettings.java
+++ ../src-work/minecraft/net/minecraft/client/settings/GameSettings.java
@@ -2,6 +2,8 @@
 
 import com.google.common.collect.Maps;
 import com.google.gson.Gson;
+
+import cpw.mods.fml.client.FMLClientHandler;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.io.BufferedReader;
@@ -1049,6 +1051,7 @@
 
     public void func_74303_b()
     {
+        if (FMLClientHandler.instance().isLoading()) return;
         try
         {
             PrintWriter printwriter = new PrintWriter(new FileWriter(this.field_74354_ai));
