--- ../src-base/minecraft/net/minecraft/src/NetClientHandler.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src-work/minecraft/net/minecraft/src/NetClientHandler.java	0000-00-00 00:00:00.000000000 -0000
@@ -20,6 +20,8 @@
 import net.minecraft.client.Minecraft;
 import org.lwjgl.input.Keyboard;
 
+import cpw.mods.fml.client.FMLClientHandler;
+
 public class NetClientHandler extends NetHandler
 {
     private boolean field_55324_f = false;
@@ -1165,4 +1167,13 @@
     {
         return this.field_1213_d;
     }
+
+    /* (non-Javadoc)
+     * @see net.minecraft.src.NetHandler#func_44028_a(net.minecraft.src.Packet250CustomPayload)
+     */
+//    @Override
+//    public void func_44028_a(Packet250CustomPayload p_44028_1_)
+//    {
+//        FMLClientHandler.instance().handlePacket250(p_44028_1_);
+//    }
 }
