--- ../src_base/minecraft/net/minecraft/src/ChunkProvider.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/ChunkProvider.java	0000-00-00 00:00:00.000000000 -0000
@@ -8,6 +8,8 @@
 
 import cpw.mods.fml.client.FMLClientHandler;
 
+import net.minecraft.src.forge.ForgeHooks;
+
 public class ChunkProvider implements IChunkProvider
 {
     /** A set of dropped chunks. Currently not used in single player. */
@@ -53,6 +55,11 @@
      */
     public void dropChunk(int par1, int par2)
     {
+        if(!ForgeHooks.canUnloadChunk(worldObj.getChunkFromChunkCoords(par1, par2)))
+        {
+            return;
+        }
+        
         ChunkCoordinates var3 = this.worldObj.getSpawnPoint();
         int var4 = par1 * 16 + 8 - var3.posX;
         int var5 = par2 * 16 + 8 - var3.posZ;
