--- ../src_base/minecraft/net/minecraft/src/GuiContainer.java
+++ ../src_work/minecraft/net/minecraft/src/GuiContainer.java
@@ -1,7 +1,10 @@
 package net.minecraft.src;
 
 import java.util.Iterator;
 import java.util.List;
+import net.minecraftforge.client.IItemRenderer;
+import net.minecraftforge.client.MinecraftForgeClient;
+import static net.minecraftforge.client.IItemRenderer.ItemRenderType.TOOLTIP;
 import org.lwjgl.input.Keyboard;
 import org.lwjgl.opengl.GL11;
 import org.lwjgl.opengl.GL12;
@@ -116,2 +118,9 @@
    protected void func_74184_a(ItemStack par1ItemStack, int x, int y)
    {
+    	IItemRenderer customRenderer = MinecraftForgeClient.getItemRenderer(par1ItemStack, TOOLTIP);
+    	if (customRenderer != null) 
+    	{
+    		customRenderer.renderItem(TOOLTIP, par1ItemStack, this.fontRenderer, x, y);
+    	}
+    	else
+    	{
@@ -183,1 +190,2 @@	        
            itemRenderer.zLevel = 0.0F;
        }
+    }
 }  