--- ../src_base/minecraft/net/minecraft/src/IntegratedServer.java
+++ ../src_work/minecraft/net/minecraft/src/IntegratedServer.java
@@ -8,6 +8,7 @@
 import cpw.mods.fml.common.FMLCommonHandler;
 import net.minecraft.client.Minecraft;
 import net.minecraft.server.MinecraftServer;
+import net.minecraftforge.common.DimensionManager;
 
 @SideOnly(Side.CLIENT)
 public class IntegratedServer extends MinecraftServer
@@ -48,44 +49,20 @@
     protected void loadAllWorlds(String par1Str, String par2Str, long par3, WorldType par5WorldType)
     {
         this.convertMapIfNeeded(par1Str);
-        this.worldServers = new WorldServer[3];
-        this.timeOfLastDimensionTick = new long[this.worldServers.length][100];
         ISaveHandler var6 = this.getActiveAnvilConverter().getSaveLoader(par1Str, true);
 
-        for (int var7 = 0; var7 < this.worldServers.length; ++var7)
-        {
-            byte var8 = 0;
-
-            if (var7 == 1)
+        WorldServer overWorld = (isDemo() ? new DemoWorldServer(this, var6, par2Str, 0, theProfiler) : new WorldServer(this, var6, par2Str, 0, theWorldSettings, theProfiler));
+        for (int dim : DimensionManager.getIDs())
+        {
+            WorldServer world = (dim == 0 ? overWorld : new WorldServerMulti(this, var6, par2Str, dim, theWorldSettings, overWorld, theProfiler));
+            world.addWorldAccess(new WorldManager(this, world));
+            if (!this.isSinglePlayer())
             {
-                var8 = -1;
+                world.getWorldInfo().setGameType(this.getGameType());
             }
-
-            if (var7 == 2)
-            {
-                var8 = 1;
-            }
-
-            if (var7 == 0)
-            {
-                if (this.isDemo())
-                {
-                    this.worldServers[var7] = new DemoWorldServer(this, var6, par2Str, var8, this.theProfiler);
-                }
-                else
-                {
-                    this.worldServers[var7] = new WorldServer(this, var6, par2Str, var8, this.theWorldSettings, this.theProfiler);
-                }
-            }
-            else
-            {
-                this.worldServers[var7] = new WorldServerMulti(this, var6, par2Str, var8, this.theWorldSettings, this.worldServers[0], this.theProfiler);
-            }
-
-            this.worldServers[var7].addWorldAccess(new WorldManager(this, this.worldServers[var7]));
-            this.getConfigurationManager().setPlayerManager(this.worldServers);
-        }
-
+        }
+
+        this.getConfigurationManager().setPlayerManager(new WorldServer[]{ overWorld });
         this.setDifficultyForAllWorlds(this.getDifficulty());
         this.initialWorldChunkLoad();
     }
@@ -106,6 +83,7 @@
         this.loadAllWorlds(this.getFolderName(), this.getWorldName(), this.theWorldSettings.getSeed(), this.theWorldSettings.getTerrainType());
         this.setMOTD(this.getServerOwner() + " - " + this.worldServers[0].getWorldInfo().getWorldName());
         FMLCommonHandler.instance().handleServerStarting(this);
+        spawnProtectionSize = 0;
         return true;
     }
 
