--- ../src-base/minecraft/net/minecraft/src/IntegratedServer.java
+++ ../src-work/minecraft/net/minecraft/src/IntegratedServer.java
@@ -4,6 +4,8 @@
 import cpw.mods.fml.common.asm.SideOnly;
 import java.io.File;
 import java.io.IOException;
+
+import cpw.mods.fml.common.FMLCommonHandler;
 import net.minecraft.client.Minecraft;
 import net.minecraft.server.MinecraftServer;
 
@@ -97,6 +99,7 @@
         this.func_71253_a(CryptManager.func_75891_b());
         this.func_71247_a(this.func_71270_I(), this.func_71221_J(), this.field_71350_m.func_77160_d(), this.field_71350_m.func_77165_h(), this.field_71350_m.func_82749_j());
         this.func_71205_p(this.func_71214_G() + " - " + this.field_71305_c[0].func_72912_H().func_76065_j());
+        FMLCommonHandler.instance().handleServerStarting(this);
         return true;
     }
 
