--- ../src_base/minecraft/net/minecraft/src/ItemHoe.java
+++ ../src_work/minecraft/net/minecraft/src/ItemHoe.java
@@ -1,4 +1,7 @@
 package net.minecraft.src;
+
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.entity.UseHoeEvent;
 
 public class ItemHoe extends Item
 {
@@ -21,6 +24,18 @@
         }
         else
         {
+            UseHoeEvent event = new UseHoeEvent(par2EntityPlayer, par1ItemStack, par3World, par4, par5, par6);
+            MinecraftForge.EVENT_BUS.post(event);
+            if (event.isCanceled())
+            {
+                return false;
+            }
+            if (event.isHandeled())
+            {
+                par1ItemStack.damageItem(1, par2EntityPlayer);
+                return true;
+            }
+
             int var11 = par3World.getBlockId(par4, par5, par6);
             int var12 = par3World.getBlockId(par4, par5 + 1, par6);
 
