--- ../src-base/minecraft/net/minecraft/src/NetClientHandler.java
+++ ../src-work/minecraft/net/minecraft/src/NetClientHandler.java
@@ -2,6 +2,8 @@
 
 import cpw.mods.fml.common.Side;
 import cpw.mods.fml.common.asm.SideOnly;
+import cpw.mods.fml.common.network.FMLNetworkHandler;
+
 import java.io.BufferedReader;
 import java.io.ByteArrayInputStream;
 import java.io.DataInputStream;
@@ -42,6 +44,7 @@
         this.field_72563_h = p_i3103_1_;
         Socket var4 = new Socket(InetAddress.getByName(p_i3103_2_), p_i3103_3_);
         this.field_72555_g = new TcpConnection(var4, "Client", this);
+        FMLNetworkHandler.onClientConnectionToRemoteServer(this, p_i3103_2_, p_i3103_3_, this.field_72555_g);
     }
 
     public NetClientHandler(Minecraft p_i3104_1_, IntegratedServer p_i3104_2_) throws IOException
@@ -49,6 +52,7 @@
         this.field_72563_h = p_i3104_1_;
         this.field_72555_g = new MemoryConnection(this);
         p_i3104_2_.func_71343_a().func_71754_a((MemoryConnection)this.field_72555_g, p_i3104_1_.field_71449_j.field_74286_b);
+        FMLNetworkHandler.onClientConnectionToIntegratedServer(this, p_i3104_2_, this.field_72555_g);
     }
 
     public void func_72547_c()
@@ -119,6 +123,7 @@
 
     public void func_72513_a(Packet252SharedKey p_72513_1_)
     {
+        this.func_72552_c(FMLNetworkHandler.getFMLFakeLoginPacket());
         this.func_72552_c(new Packet205ClientCommand(0));
     }
 
@@ -134,6 +139,7 @@
         this.field_72563_h.field_71439_g.field_70157_k = p_72455_1_.field_73561_a;
         this.field_72556_d = p_72455_1_.field_73562_h;
         this.field_72563_h.field_71442_b.func_78746_a(p_72455_1_.field_73557_d);
+        FMLNetworkHandler.onConnectionEstablishedToServer(this, field_72555_g, p_72455_1_);
         this.func_72552_c(new Packet204ClientInfo(this.field_72563_h.field_71474_y.field_74363_ab, this.field_72563_h.field_71474_y.field_74339_e, this.field_72563_h.field_71474_y.field_74343_n, this.field_72563_h.field_71474_y.field_74344_o, this.field_72563_h.field_71474_y.field_74318_M));
     }
 
@@ -1167,6 +1173,11 @@
 
     public void func_72501_a(Packet250CustomPayload p_72501_1_)
     {
+        FMLNetworkHandler.handlePacket250Packet(p_72501_1_, field_72555_g, this);
+    }
+
+    public void handleVanilla250Packet(Packet250CustomPayload p_72501_1_)
+    {
         if ("MC|TPack".equals(p_72501_1_.field_73630_a))
         {
             String[] var2 = (new String(p_72501_1_.field_73629_c)).split("\u0000");
@@ -1211,4 +1222,10 @@
     {
         return this.field_72555_g;
     }
+
+    @Override
+    public EntityPlayer getPlayer()
+    {
+        return field_72563_h.field_71439_g;
+    }
 }
