--- ../src_base/minecraft/net/minecraft/src/RenderEngine.java
+++ ../src_work/minecraft/net/minecraft/src/RenderEngine.java
@@ -20,6 +20,9 @@
 import java.util.logging.Logger;
 
 import javax.imageio.ImageIO;
+
+import net.minecraftforge.client.ForgeHooksClient;
+
 import org.lwjgl.opengl.GL11;
 
 import cpw.mods.fml.client.TextureFXManager;
@@ -167,6 +170,7 @@
 
             try
             {
+                ForgeHooksClient.onTextureLoadPre(par1Str);
                 this.singleIntBuffer.clear();
                 GLAllocation.generateTextureNames(this.singleIntBuffer);
                 int var3 = this.singleIntBuffer.get(0);
@@ -210,6 +214,7 @@
                 }
 
                 this.textureMap.put(par1Str, Integer.valueOf(var3));
+                ForgeHooksClient.onTextureLoad(par1Str, var6);
                 return var3;
             }
             catch (Exception var5)
