--- ../src_base/minecraft/net/minecraft/src/RenderGlobal.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/RenderGlobal.java	0000-00-00 00:00:00.000000000 -0000
@@ -1635,8 +1635,8 @@
     }
 
     public List tileEntities;
-    private World worldObj;
-    private RenderEngine renderEngine;
+    public World worldObj;
+    public RenderEngine renderEngine;
     private List worldRenderersToUpdate;
     private WorldRenderer sortedWorldRenderers[];
     private WorldRenderer worldRenderers[];
@@ -1644,8 +1644,8 @@
     private int renderChunksTall;
     private int renderChunksDeep;
     private int glRenderListBase;
-    private Minecraft mc;
-    private RenderBlocks globalRenderBlocks;
+    public Minecraft mc;
+    public RenderBlocks globalRenderBlocks;
     private IntBuffer glOcclusionQueryBase;
     private boolean occlusionEnabled;
     private int cloudOffsetX;
