--- ../src_base/minecraft/net/minecraft/src/RenderEngine.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/RenderEngine.java	0000-00-00 00:00:00.000000000 -0000
@@ -27,6 +27,8 @@
 import cpw.mods.fml.client.FMLClientHandler;
 import cpw.mods.fml.common.FMLCommonHandler;
 
+import net.minecraft.src.forge.ForgeHooksClient;
+
 public class RenderEngine
 {
     /** Use mipmaps for all bound textures (unused at present) */
@@ -59,7 +61,7 @@
     public boolean blurTexture = false;
 
     /** Texture pack */
-    private TexturePackList texturePack;
+    public TexturePackList texturePack;
 
     /** Missing texture image */
     private BufferedImage missingTextureImage = new BufferedImage(64, 64, 2);
@@ -169,6 +171,7 @@
         {
             try
             {
+                ForgeHooksClient.onTextureLoadPre(par1Str);
                 this.singleIntBuffer.clear();
                 GLAllocation.generateTextureNames(this.singleIntBuffer);
                 int var6 = this.singleIntBuffer.get(0);
@@ -212,6 +215,7 @@
                 }
 
                 this.textureMap.put(par1Str, Integer.valueOf(var6));
+                ForgeHooksClient.onTextureLoad(par1Str, var6);
                 return var6;
             }
             catch (Exception var5)
