--- ../src_base/minecraft/net/minecraft/src/RenderItem.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/RenderItem.java	0000-00-00 00:00:00.000000000 -0000
@@ -4,6 +4,10 @@
 
 package net.minecraft.src;
 
+import net.minecraft.src.forge.ForgeHooksClient;
+import net.minecraft.src.forge.ICustomItemRenderer;
+import net.minecraft.src.forge.MinecraftForgeClient;
+
 import java.util.Random;
 import org.lwjgl.opengl.GL11;
 
@@ -54,10 +58,36 @@
         }
         GL11.glTranslatef((float)d, (float)d1 + f2, (float)d2);
         GL11.glEnable(32826 /*GL_RESCALE_NORMAL_EXT*/);
-        if(itemstack.itemID < 256 && RenderBlocks.renderItemIn3d(Block.blocksList[itemstack.itemID].getRenderType()))
+
+        ICustomItemRenderer customRenderer = MinecraftForgeClient.getCustomItemRenderer(itemstack.itemID);
+        
+        if (customRenderer != null) {
+        	GL11.glRotatef(f3, 0.0F, 1.0F, 0.0F);
+            loadTexture("/terrain.png");
+            ForgeHooksClient.overrideTexture(itemstack.getItem());
+            float f4 = 0.25F;
+            f4 = 0.5F;
+            GL11.glScalef(f4, f4, f4);
+            for(int j = 0; j < byte0; j++)
+            {
+                GL11.glPushMatrix();
+                if(j > 0)
+                {
+                    float f5 = ((random.nextFloat() * 2.0F - 1.0F) * 0.2F) / f4;
+                    float f7 = ((random.nextFloat() * 2.0F - 1.0F) * 0.2F) / f4;
+                    float f9 = ((random.nextFloat() * 2.0F - 1.0F) * 0.2F) / f4;
+                    GL11.glTranslatef(f5, f7, f9);
+                }
+				ForgeHooksClient.renderCustomItem(customRenderer, renderBlocks,
+						itemstack.itemID, itemstack.getItemDamage(),
+						entityitem.getEntityBrightness(f1));
+                GL11.glPopMatrix();
+            }
+        } else if(itemstack.itemID < 256 && RenderBlocks.renderItemIn3d(Block.blocksList[itemstack.itemID].getRenderType()))
         {
             GL11.glRotatef(f3, 0.0F, 1.0F, 0.0F);
             loadTexture("/terrain.png");
+            ForgeHooksClient.overrideTexture(Block.blocksList[itemstack.itemID]);
             float f4 = 0.25F;
             int j = Block.blocksList[itemstack.itemID].getRenderType();
             if(j == 1 || j == 19 || j == 12 || j == 2)
@@ -108,9 +138,11 @@
             if(itemstack.itemID < 256)
             {
                 loadTexture("/terrain.png");
+                ForgeHooksClient.overrideTexture(Block.blocksList[itemstack.itemID]);
             } else
             {
                 loadTexture("/gui/items.png");
+                ForgeHooksClient.overrideTexture(Item.itemsList[itemstack.itemID]);
             }
             if(field_27004_a)
             {
@@ -162,10 +194,37 @@
 
     public void drawItemIntoGui(FontRenderer fontrenderer, RenderEngine renderengine, int i, int j, int k, int l, int i1)
     {
-        if(i < 256 && RenderBlocks.renderItemIn3d(Block.blocksList[i].getRenderType()))
+    	ICustomItemRenderer customRenderer = MinecraftForgeClient.getCustomItemRenderer(i);
+    	if (customRenderer != null) {
+    		  int j1 = i;
+              renderengine.bindTexture(renderengine.getTexture("/terrain.png"));
+              Item item = Item.itemsList[i];
+              ForgeHooksClient.overrideTexture (Item.itemsList [i]);
+              GL11.glPushMatrix();
+              GL11.glTranslatef(l - 2, i1 + 3, -3F);
+              GL11.glScalef(10F, 10F, 10F);
+              GL11.glTranslatef(1.0F, 0.5F, 1.0F);
+              GL11.glScalef(1.0F, 1.0F, -1F);
+              GL11.glRotatef(210F, 1.0F, 0.0F, 0.0F);
+              GL11.glRotatef(45F, 0.0F, 1.0F, 0.0F);
+              int l1 = Item.itemsList[i].getColorFromDamage(j);
+              float f2 = (float)(l1 >> 16 & 0xff) / 255F;
+              float f4 = (float)(l1 >> 8 & 0xff) / 255F;
+              float f5 = (float)(l1 & 0xff) / 255F;
+              if(field_27004_a)
+              {
+                  GL11.glColor4f(f2, f4, f5, 1.0F);
+              }
+              GL11.glRotatef(-90F, 0.0F, 1.0F, 0.0F);
+              renderBlocks.useInventoryTint = field_27004_a;
+              ForgeHooksClient.renderCustomItem(customRenderer, renderBlocks, i, j, 1.0F);
+              renderBlocks.useInventoryTint = true;
+              GL11.glPopMatrix();
+    	} else if(i < 256 && RenderBlocks.renderItemIn3d(Block.blocksList[i].getRenderType()))
         {
             int j1 = i;
             renderengine.bindTexture(renderengine.getTexture("/terrain.png"));
+            ForgeHooksClient.overrideTexture (Block.blocksList[i]);
             Block block = Block.blocksList[j1];
             GL11.glPushMatrix();
             GL11.glTranslatef(l - 2, i1 + 3, -3F + field_40268_b);
@@ -215,9 +274,11 @@
             if(i < 256)
             {
                 renderengine.bindTexture(renderengine.getTexture("/terrain.png"));
+                ForgeHooksClient.overrideTexture (Block.blocksList[i]);
             } else
             {
                 renderengine.bindTexture(renderengine.getTexture("/gui/items.png"));
+                ForgeHooksClient.overrideTexture(Item.itemsList[i]);
             }
             int k1 = Item.itemsList[i].getColorFromDamage(j);
             float f = (float)(k1 >> 16 & 0xff) / 255F;
