--- ../src_base/minecraft/net/minecraft/src/RenderManager.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/RenderManager.java	0000-00-00 00:00:00.000000000 -0000
@@ -124,12 +124,14 @@
 
         if (par4EntityLiving.isPlayerSleeping())
         {
-            int var7 = par1World.getBlockId(MathHelper.floor_double(par4EntityLiving.posX), MathHelper.floor_double(par4EntityLiving.posY), MathHelper.floor_double(par4EntityLiving.posZ));
+            int x = MathHelper.floor_double(par4EntityLiving.posX);
+            int y = MathHelper.floor_double(par4EntityLiving.posY);
+            int z = MathHelper.floor_double(par4EntityLiving.posZ);
+            Block block = Block.blocksList[par1World.getBlockId(x,  y, z)];
 
-            if (var7 == Block.bed.blockID)
+            if (block != null && block.isBed(par1World, x, y, z, par4EntityLiving))
             {
-                int var8 = par1World.getBlockMetadata(MathHelper.floor_double(par4EntityLiving.posX), MathHelper.floor_double(par4EntityLiving.posY), MathHelper.floor_double(par4EntityLiving.posZ));
-                int var9 = var8 & 3;
+                int var9 = block.getBedDirection(par1World, x, y, z);
                 this.playerViewY = (float)(var9 * 90 + 180);
                 this.playerViewX = 0.0F;
             }
