--- ../src_base/minecraft/net/minecraft/src/RenderSnowMan.java
+++ ../src_work/minecraft/net/minecraft/src/RenderSnowMan.java
@@ -1,4 +1,9 @@
 package net.minecraft.src;
+
+import net.minecraftforge.client.IItemRenderer;
+import static net.minecraftforge.client.IItemRenderer.ItemRenderType.*;
+import static net.minecraftforge.client.IItemRenderer.ItemRendererHelper.*;
+import net.minecraftforge.client.MinecraftForgeClient;
 
 import cpw.mods.fml.common.Side;
 import cpw.mods.fml.common.asm.SideOnly;
@@ -25,12 +30,15 @@
         super.renderEquippedItems(par1EntitySnowman, par2);
         ItemStack var3 = new ItemStack(Block.pumpkin, 1);
 
-        if (var3 != null && var3.getItem().shiftedIndex < 256)
+        if (var3 != null && var3.getItem() instanceof ItemBlock)
         {
             GL11.glPushMatrix();
             this.snowmanModel.head.postRender(0.0625F);
 
-            if (RenderBlocks.renderItemIn3d(Block.blocksList[var3.itemID].getRenderType()))
+            IItemRenderer customRenderer = MinecraftForgeClient.getItemRenderer(var3, EQUIPPED);
+            boolean is3D = (customRenderer != null && customRenderer.shouldUseRenderHelper(EQUIPPED, var3, BLOCK_3D));
+
+            if (is3D || RenderBlocks.renderItemIn3d(Block.blocksList[var3.itemID].getRenderType()))
             {
                 float var4 = 0.625F;
                 GL11.glTranslatef(0.0F, -0.34375F, 0.0F);
