--- ../src-base/minecraft/net/minecraft/src/TextureWatchFX.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src-work/minecraft/net/minecraft/src/TextureWatchFX.java	0000-00-00 00:00:00.000000000 -0000
@@ -2,7 +2,11 @@
 
 import java.awt.image.BufferedImage;
 import java.io.IOException;
+import java.util.logging.Level;
+
 import javax.imageio.ImageIO;
+
+import cpw.mods.fml.common.FMLCommonHandler;
 import net.minecraft.client.Minecraft;
 
 public class TextureWatchFX extends TextureFX
@@ -18,22 +22,39 @@
         super(Item.field_4023_aQ.func_27009_a(0));
         this.field_4225_g = p_i364_1_;
         this.field_1128_f = 1;
-
+        setup();
+    }
+    
+    public void setup()
+    {
+        super.setup();
+        field_4224_h = new int[tileSizeSquare];
+        field_4223_i = new int[tileSizeSquare];
         try
         {
-            BufferedImage var2 = ImageIO.read(Minecraft.class.getResource("/gui/items.png"));
-            int var3 = this.field_1126_b % 16 * 16;
-            int var4 = this.field_1126_b / 16 * 16;
-            var2.getRGB(var3, var4, 16, 16, this.field_4224_h, 0, 16);
-            var2 = ImageIO.read(Minecraft.class.getResource("/misc/dial.png"));
-            var2.getRGB(0, 0, 16, 16, this.field_4223_i, 0, 16);
+            BufferedImage var2 = ImageIO.read(field_4225_g.field_6298_C.field_6534_a.func_6481_a("/gui/items.png"));
+            int var3 = this.field_1126_b % 16 * iconTileSize;
+            int var4 = this.field_1126_b / 16 * iconTileSize;
+            var2.getRGB(var3, var4, iconTileSize, iconTileSize, this.field_4224_h, 0, iconTileSize);
+            var2 = ImageIO.read(field_4225_g.field_6298_C.field_6534_a.func_6481_a("/misc/dial.png"));
+            var2.getRGB(0, 0, iconTileSize, iconTileSize, this.field_4223_i, 0, iconTileSize);
         }
-        catch (IOException var5)
+        catch (Exception var5)
         {
-            var5.printStackTrace();
+            FMLCommonHandler.instance().getFMLLogger().log(Level.WARNING, String.format("A problem occurred with the watch texture: animation will be disabled"), var5);
+            errored = true;
         }
     }
 
+    @Override
+    public void onTexturePackChange(int newTileSize, int newTileSizeSquare, int newTileSizeMask, int newTileSizeSquareMask)
+    {
+        super.onTexturePackChange(newTileSize, newTileSizeSquare, newTileSizeMask, newTileSizeSquareMask);
+        if (field_4224_h.length!=tileSizeSquare) {
+            setup();
+        }
+    }
+    
     public void func_783_a()
     {
         double var1 = 0.0D;
@@ -77,7 +98,7 @@
         double var5 = Math.sin(this.field_4222_j);
         double var7 = Math.cos(this.field_4222_j);
 
-        for (int var9 = 0; var9 < 256; ++var9)
+        for (int var9 = 0; var9 < tileSizeSquare; ++var9)
         {
             int var10 = this.field_4224_h[var9] >> 24 & 255;
             int var11 = this.field_4224_h[var9] >> 16 & 255;
@@ -86,12 +107,12 @@
 
             if (var11 == var13 && var12 == 0 && var13 > 0)
             {
-                double var14 = -((double)(var9 % 16) / 15.0D - 0.5D);
-                double var16 = (double)(var9 / 16) / 15.0D - 0.5D;
+                double var14 = -((double)(var9 % iconTileSize) / tileSizeMask - 0.5D);
+                double var16 = (double)(var9 / iconTileSize) / tileSizeMask - 0.5D;
                 int var18 = var11;
-                int var19 = (int)((var14 * var7 + var16 * var5 + 0.5D) * 16.0D);
-                int var20 = (int)((var16 * var7 - var14 * var5 + 0.5D) * 16.0D);
-                int var21 = (var19 & 15) + (var20 & 15) * 16;
+                int var19 = (int)((var14 * var7 + var16 * var5 + 0.5D) * iconTileSize);
+                int var20 = (int)((var16 * var7 - var14 * var5 + 0.5D) * iconTileSize);
+                int var21 = (var19 & tileSizeMask) + (var20 & tileSizeMask) * iconTileSize;
                 var10 = this.field_4223_i[var21] >> 24 & 255;
                 var11 = (this.field_4223_i[var21] >> 16 & 255) * var11 / 255;
                 var12 = (this.field_4223_i[var21] >> 8 & 255) * var18 / 255;
