--- ../src-base/minecraft/net/minecraft/src/TexturePackList.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src-work/minecraft/net/minecraft/src/TexturePackList.java	0000-00-00 00:00:00.000000000 -0000
@@ -8,6 +8,8 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+
+import cpw.mods.fml.client.FMLClientHandler;
 import net.minecraft.client.Minecraft;
 
 public class TexturePackList
@@ -18,7 +20,7 @@
     private final File field_57507_d;
     private List field_6533_b = new ArrayList();
     private Map field_55280_f = new HashMap();
-    private TexturePackBase field_6534_a;
+    public TexturePackBase field_6534_a;
     private boolean field_57509_h;
 
     public TexturePackList(File p_i1506_1_, Minecraft p_i1506_2_)
