--- ../src-base/minecraft/net/minecraft/src/TextureFX.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src-work/minecraft/net/minecraft/src/TextureFX.java	0000-00-00 00:00:00.000000000 -0000
@@ -4,12 +4,17 @@
 
 public class TextureFX
 {
-    public byte[] field_1127_a = new byte[1024];
+    public byte[] field_1127_a = new byte[tileSizeSquare << 2];
     public int field_1126_b;
     public boolean field_1131_c = false;
     public int field_1130_d = 0;
     public int field_1129_e = 1;
     public int field_1128_f = 0;
+    public static int iconTileSize = 16;
+    public static int tileSizeSquare = 256;
+    public static int tileSizeMask = 15;
+    public static int tileSizeSquareMask = 255;
+    public boolean errored = false;
 
     public TextureFX(int p_i616_1_)
     {
@@ -29,4 +34,22 @@
             GL11.glBindTexture(GL11.GL_TEXTURE_2D, p_782_1_.func_1070_a("/gui/items.png"));
         }
     }
+    
+    public void onTexturePackChange(int newTileSize, int newTileSizeSquare, int newTileSizeMask, int newTileSizeSquareMask) {
+        errored = false;
+        setup();
+    }
+    public void setup() {
+        field_1127_a = new byte[tileSizeSquare << 2];
+    }
+
+    /**
+     * @return 
+     * 
+     */
+    public boolean unregister(RenderEngine engine)
+    {
+        engine.field_1604_f.remove(this);
+        return true;
+    }
 }
