--- ../src_base/minecraft/net/minecraft/src/TileEntityRenderer.java
+++ ../src_work/minecraft/net/minecraft/src/TileEntityRenderer.java
@@ -126,7 +126,9 @@
      */
     public void renderTileEntity(TileEntity par1TileEntity, float par2)
     {
-        if (par1TileEntity.getDistanceFrom(this.playerX, this.playerY, this.playerZ) < 4096.0D)
+        double dist = par1TileEntity.getRenderDistance();
+        dist *= dist;
+        if (par1TileEntity.getDistanceFrom(this.playerX, this.playerY, this.playerZ) < dist)
         {
             int var3 = this.worldObj.getLightBrightnessForSkyBlocks(par1TileEntity.xCoord, par1TileEntity.yCoord, par1TileEntity.zCoord, 0);
             int var4 = var3 % 65536;
