--- ../src-base/minecraft/net/minecraft/src/World.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src-work/minecraft/net/minecraft/src/World.java	0000-00-00 00:00:00.000000000 -0000
@@ -3322,7 +3322,7 @@
 
     public double func_46068_G()
     {
-        return this.field_22145_q.func_46133_t() == WorldType.field_48636_c ? 0.0D : 63.0D;
+        return this.field_22145_q.func_46133_t().getHorizon(this);
     }
 
     public CrashReport func_55266_a(CrashReport p_55266_1_)
