--- ../src_base/minecraft/net/minecraft/src/WorldProviderSurface.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src_work/minecraft/net/minecraft/src/WorldProviderSurface.java	0000-00-00 00:00:00.000000000 -0000
@@ -2,4 +2,22 @@
 
 public class WorldProviderSurface extends WorldProvider
 {
+
+    @Override
+    public String getSaveFolder() 
+    {
+        return null;
+    }
+
+    @Override
+    public String getWelcomeMessage() 
+    {
+        return null;
+    }
+
+    @Override
+    public String getDepartMessage() 
+    {
+        return null;
+    }
 }
