--- ../src-base/minecraft/net/minecraft/client/entity/EntityPlayerSP.java
+++ ../src-work/minecraft/net/minecraft/client/entity/EntityPlayerSP.java
@@ -400,6 +400,15 @@
         this.mc.ingameGUI.getChatGUI().printChatMessage(chatComponent);
     }
 
+    private boolean isHeadspaceFree(BlockPos pos, int height)
+    {
+        for (int y = 0; y < height; y++)
+        {
+            if (isOpenBlockSpace(pos.add(0, y, 0))) return false;
+        }
+        return true;
+    }
+
     protected boolean pushOutOfBlocks(double x, double y, double z)
     {
         if (this.noClip)
@@ -412,30 +421,34 @@
             double d0 = x - (double)blockpos.getX();
             double d1 = z - (double)blockpos.getZ();
 
-            if (!this.isOpenBlockSpace(blockpos))
+            int entHeight = Math.max(Math.round(this.height), 1);
+
+            boolean inTranslucentBlock = this.isHeadspaceFree(blockpos, entHeight);
+
+            if (inTranslucentBlock)
             {
                 int i = -1;
                 double d2 = 9999.0D;
 
-                if (this.isOpenBlockSpace(blockpos.west()) && d0 < d2)
+                if (!this.isHeadspaceFree(blockpos.west(), entHeight) && d0 < d2)
                 {
                     d2 = d0;
                     i = 0;
                 }
 
-                if (this.isOpenBlockSpace(blockpos.east()) && 1.0D - d0 < d2)
+                if (!this.isHeadspaceFree(blockpos.east(), entHeight) && 1.0D - d0 < d2)
                 {
                     d2 = 1.0D - d0;
                     i = 1;
                 }
 
-                if (this.isOpenBlockSpace(blockpos.north()) && d1 < d2)
+                if (!this.isHeadspaceFree(blockpos.north(), entHeight) && d1 < d2)
                 {
                     d2 = d1;
                     i = 4;
                 }
 
-                if (this.isOpenBlockSpace(blockpos.south()) && 1.0D - d1 < d2)
+                if (!this.isHeadspaceFree(blockpos.south(), entHeight) && 1.0D - d1 < d2)
                 {
                     d2 = 1.0D - d1;
                     i = 5;
