--- ../src-base/minecraft/net/minecraft/client/settings/GameSettings.java
+++ ../src-work/minecraft/net/minecraft/client/settings/GameSettings.java
@@ -165,6 +165,7 @@
 
     public GameSettings(Minecraft p_i46326_1_, File p_i46326_2_)
     {
+        setForgeKeybindProperties();
         this.field_74324_K = (KeyBinding[])ArrayUtils.addAll(new KeyBinding[] {this.field_74312_F, this.field_74313_G, this.field_74351_w, this.field_74370_x, this.field_74368_y, this.field_74366_z, this.field_74314_A, this.field_74311_E, this.field_151444_V, this.field_74316_C, this.field_151445_Q, this.field_74310_D, this.field_74321_H, this.field_74322_I, this.field_74323_J, this.field_151447_Z, this.field_151457_aa, this.field_151458_ab, this.field_152395_am, this.field_178883_an, this.field_186718_X, this.field_193629_ap, this.field_193630_aq, this.field_194146_ao}, this.field_151456_ac);
         this.field_74318_M = EnumDifficulty.NORMAL;
         this.field_74332_R = "";
@@ -188,6 +189,7 @@
 
     public GameSettings()
     {
+        setForgeKeybindProperties();
         this.field_74324_K = (KeyBinding[])ArrayUtils.addAll(new KeyBinding[] {this.field_74312_F, this.field_74313_G, this.field_74351_w, this.field_74370_x, this.field_74368_y, this.field_74366_z, this.field_74314_A, this.field_74311_E, this.field_151444_V, this.field_74316_C, this.field_151445_Q, this.field_74310_D, this.field_74321_H, this.field_74322_I, this.field_74323_J, this.field_151447_Z, this.field_151457_aa, this.field_151458_ab, this.field_152395_am, this.field_178883_an, this.field_186718_X, this.field_193629_ap, this.field_193630_aq, this.field_194146_ao}, this.field_151456_ac);
         this.field_74318_M = EnumDifficulty.NORMAL;
         this.field_74332_R = "";
@@ -299,7 +301,7 @@
                 this.field_74317_L.func_147117_R().func_147633_a(this.field_151442_I);
                 this.field_74317_L.func_110434_K().func_110577_a(TextureMap.field_110575_b);
                 this.field_74317_L.func_147117_R().func_174937_a(false, this.field_151442_I > 0);
-                this.field_74317_L.func_175603_A();
+                this.needsBlockModelRefresh = true; // FORGE: fix for MC-64581 very laggy mipmap slider
             }
         }
 
@@ -361,7 +363,7 @@
         if (p_74306_1_ == GameSettings.Options.ANAGLYPH)
         {
             this.field_74337_g = !this.field_74337_g;
-            this.field_74317_L.func_110436_a();
+            net.minecraftforge.fml.client.FMLClientHandler.instance().refreshResources(net.minecraftforge.client.resource.VanillaResourceType.TEXTURES);
         }
 
         if (p_74306_1_ == GameSettings.Options.GRAPHICS)
@@ -1068,7 +1070,12 @@
                     {
                         if (s1.equals("key_" + keybinding.func_151464_g()))
                         {
-                            keybinding.func_151462_b(Integer.parseInt(s2));
+                            if (s2.indexOf(':') != -1)
+                            {
+                                String[] t = s2.split(":");
+                                keybinding.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.valueFromString(t[1]), Integer.parseInt(t[0]));
+                            } else
+                            keybinding.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.NONE, Integer.parseInt(s2));
                         }
                     }
 
@@ -1132,6 +1139,7 @@
 
     public void func_74303_b()
     {
+        if (net.minecraftforge.fml.client.FMLClientHandler.instance().isLoading()) return;
         PrintWriter printwriter = null;
 
         try
@@ -1206,7 +1214,8 @@
 
             for (KeyBinding keybinding : this.field_74324_K)
             {
-                printwriter.println("key_" + keybinding.func_151464_g() + ":" + keybinding.func_151463_i());
+                String keyString = "key_" + keybinding.func_151464_g() + ":" + keybinding.func_151463_i();
+                printwriter.println(keybinding.getKeyModifier() != net.minecraftforge.client.settings.KeyModifier.NONE ? keyString + ":" + keybinding.getKeyModifier() : keyString);
             }
 
             for (SoundCategory soundcategory : SoundCategory.values())
@@ -1440,4 +1449,35 @@
             return p_148264_1_;
         }
     }
+
+    /******* Forge Start ***********/
+    private void setForgeKeybindProperties() {
+        net.minecraftforge.client.settings.KeyConflictContext inGame = net.minecraftforge.client.settings.KeyConflictContext.IN_GAME;
+        field_74351_w.setKeyConflictContext(inGame);
+        field_74370_x.setKeyConflictContext(inGame);
+        field_74368_y.setKeyConflictContext(inGame);
+        field_74366_z.setKeyConflictContext(inGame);
+        field_74314_A.setKeyConflictContext(inGame);
+        field_74311_E.setKeyConflictContext(inGame);
+        field_151444_V.setKeyConflictContext(inGame);
+        field_74312_F.setKeyConflictContext(inGame);
+        field_74310_D.setKeyConflictContext(inGame);
+        field_74321_H.setKeyConflictContext(inGame);
+        field_74323_J.setKeyConflictContext(inGame);
+        field_151457_aa.setKeyConflictContext(inGame);
+        field_151458_ab.setKeyConflictContext(inGame);
+        field_186718_X.setKeyConflictContext(inGame);
+    }
+
+    // FORGE: fix for MC-64581 very laggy mipmap slider
+    private boolean needsBlockModelRefresh = false;
+    public void onGuiClosed()
+    {
+        if (needsBlockModelRefresh)
+        {
+            net.minecraftforge.fml.client.FMLClientHandler.instance().scheduleResourcesRefresh(net.minecraftforge.client.resource.VanillaResourceType.MODELS);
+            this.needsBlockModelRefresh = false;
+        }
+    }
+    /******* Forge End ***********/
 }
