--- a/net/minecraft/client/settings/PointOfView.java
+++ b/net/minecraft/client/settings/PointOfView.java
@@ -4,7 +4,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public enum PointOfView {
+public enum PointOfView implements net.minecraftforge.common.IExtensibleEnum {
    FIRST_PERSON(true, false),
    THIRD_PERSON_BACK(false, false),
    THIRD_PERSON_FRONT(false, true);
@@ -29,4 +29,14 @@
    public PointOfView func_243194_c() {
       return field_243189_d[(this.ordinal() + 1) % field_243189_d.length];
    }
+
+   public static PointOfView create(String name, boolean fristPerson, boolean reverseCamera) {
+      throw new IllegalStateException("Enum not extended");
+   }
+
+   @Override
+   @Deprecated
+   public void init() {
+      field_243189_d = values();
+   }
 }
