--- a/net/minecraft/client/settings/ToggleableKeyBinding.java
+++ b/net/minecraft/client/settings/ToggleableKeyBinding.java
@@ -16,7 +_,7 @@
 
    public void setDown(boolean p_225593_1_) {
       if (this.needsToggle.getAsBoolean()) {
-         if (p_225593_1_) {
+         if (p_225593_1_ && isConflictContextAndModifierActive()) {
             super.setDown(!this.isDown());
          }
       } else {
@@ -24,4 +_,5 @@
       }
 
    }
+   @Override public boolean isDown() { return this.isDown && (isConflictContextAndModifierActive() || needsToggle.getAsBoolean()); }
 }
