--- ../src-base/minecraft/net/minecraft/inventory/ContainerEnchantment.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerEnchantment.java
@@ -14,6 +14,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import net.minecraftforge.oredict.OreDictionary;
 
 public class ContainerEnchantment extends Container
 {
@@ -68,9 +69,12 @@
         this.addSlotToContainer(new Slot(this.tableInventory, 1, 35, 47)
         {
             private static final String __OBFID = "CL_00002185";
+            java.util.List<ItemStack> ores = net.minecraftforge.oredict.OreDictionary.getOres("gemLapis");
             public boolean isItemValid(ItemStack stack)
             {
-                return stack.getItem() == Items.dye && EnumDyeColor.byDyeDamage(stack.getMetadata()) == EnumDyeColor.BLUE;
+                for (ItemStack ore : ores)
+                    if (net.minecraftforge.oredict.OreDictionary.itemMatches(ore, stack, false)) return true;
+                return false;
             }
         });
         int i;
@@ -152,6 +156,7 @@
                 {
                     i = 0;
                     int j;
+                    float power = 0;
 
                     for (j = -1; j <= 1; ++j)
                     {
@@ -159,37 +164,15 @@
                         {
                             if ((j != 0 || k != 0) && this.worldPointer.isAirBlock(this.position.add(k, 0, j)) && this.worldPointer.isAirBlock(this.position.add(k, 1, j)))
                             {
-                                if (this.worldPointer.getBlockState(this.position.add(k * 2, 0, j * 2)).getBlock() == Blocks.bookshelf)
-                                {
-                                    ++i;
-                                }
+                                power += net.minecraftforge.common.ForgeHooks.getEnchantPower(worldPointer, position.add(k * 2, 0, j * 2));
+                                power += net.minecraftforge.common.ForgeHooks.getEnchantPower(worldPointer, position.add(k * 2, 1, j * 2));
 
-                                if (this.worldPointer.getBlockState(this.position.add(k * 2, 1, j * 2)).getBlock() == Blocks.bookshelf)
-                                {
-                                    ++i;
-                                }
-
                                 if (k != 0 && j != 0)
                                 {
-                                    if (this.worldPointer.getBlockState(this.position.add(k * 2, 0, j)).getBlock() == Blocks.bookshelf)
-                                    {
-                                        ++i;
-                                    }
-
-                                    if (this.worldPointer.getBlockState(this.position.add(k * 2, 1, j)).getBlock() == Blocks.bookshelf)
-                                    {
-                                        ++i;
-                                    }
-
-                                    if (this.worldPointer.getBlockState(this.position.add(k, 0, j * 2)).getBlock() == Blocks.bookshelf)
-                                    {
-                                        ++i;
-                                    }
-
-                                    if (this.worldPointer.getBlockState(this.position.add(k, 1, j * 2)).getBlock() == Blocks.bookshelf)
-                                    {
-                                        ++i;
-                                    }
+                                    power += net.minecraftforge.common.ForgeHooks.getEnchantPower(worldPointer, position.add(k * 2, 0, j));
+                                    power += net.minecraftforge.common.ForgeHooks.getEnchantPower(worldPointer, position.add(k * 2, 1, j));
+                                    power += net.minecraftforge.common.ForgeHooks.getEnchantPower(worldPointer, position.add(k, 0, j * 2));
+                                    power += net.minecraftforge.common.ForgeHooks.getEnchantPower(worldPointer, position.add(k, 1, j * 2));
                                 }
                             }
                         }
@@ -199,7 +182,7 @@
 
                     for (j = 0; j < 3; ++j)
                     {
-                        this.enchantLevels[j] = EnchantmentHelper.calcItemStackEnchantability(this.rand, j, i, itemstack);
+                        this.enchantLevels[j] = EnchantmentHelper.calcItemStackEnchantability(this.rand, j, (int)power, itemstack);
                         this.field_178151_h[j] = -1;
 
                         if (this.enchantLevels[j] < j + 1)
