--- a/net/minecraft/client/resources/sounds/SoundInstance.java
+++ b/net/minecraft/client/resources/sounds/SoundInstance.java
@@ -55,4 +_,10 @@
       NONE,
       LINEAR;
    }
+
+   /*================================ FORGE START ================================================*/
+
+   default java.util.concurrent.CompletableFuture<net.minecraft.client.sounds.AudioStream> getStream(net.minecraft.client.sounds.SoundBufferLibrary soundBuffers, Sound sound, boolean looping) {
+      return soundBuffers.getStream(sound.getPath(), looping);
+   }
 }
