--- ../src-base/minecraft/net/minecraft/item/Item.java
+++ ../src-work/minecraft/net/minecraft/item/Item.java
@@ -1,5 +1,6 @@
 package net.minecraft.item;
 
+import cpw.mods.fml.common.registry.GameData;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.util.List;
@@ -207,6 +208,8 @@
         }
 
         field_77698_e[256 + p_i3659_1_] = this;
+
+        GameData.newItemAdded(this);
     }
 
     public Item func_77665_c(int p_77665_1_)
