--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -105,13 +_,14 @@
    public boolean m_7038_() {
       this.m_129823_(new PlayerList(this, this.m_247573_(), this.f_129745_, 1) {
       });
+      if (!net.minecraftforge.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
       this.m_130006_();
       ServerLevel serverlevel = this.m_129783_();
       serverlevel.m_8733_(this.f_177588_, 0.0F);
       int i = 20000000;
       serverlevel.m_8606_(20000000, 20000000, false, false);
       f_177585_.info("Started game test server");
-      return true;
+      return net.minecraftforge.server.ServerLifecycleHooks.handleServerStarting(this);
    }
 
    public void m_5705_(BooleanSupplier p_177619_) {
