--- a/net/minecraft/gametest/framework/GameTestInfo.java
+++ b/net/minecraft/gametest/framework/GameTestInfo.java
@@ -328,4 +_,24 @@
    public boolean rerunUntilFailed() {
       return this.rerunUntilFailed;
    }
+
+
+   // FORGE START
+
+   private final boolean useTestOrigin = net.minecraftforge.common.ForgeConfig.COMMON.fixGameTestBlockPosAsserts.get();
+
+   /**
+    * Forge: Gets the origin of the test structure based on the
+    * {@linkplain #getStructureBlockEntity() structure block entity}.
+    * <p>
+    * This is a backported method from 1.21.3 to fix how block positions are asserted.
+    *
+    * @return The origin of the test structure
+    *
+    * @see net.minecraftforge.common.ForgeConfig.Common#fixGameTestBlockPosAsserts
+    */
+   @SuppressWarnings("DataFlowIssue") // Nullability moment
+   public BlockPos getTestOrigin() {
+      return this.useTestOrigin ? StructureUtils.getStructureOrigin(this.getStructureBlockEntity()) : this.getStructureBlockPos();
+   }
 }
