--- a/net/minecraft/advancements/Advancement.java
+++ b/net/minecraft/advancements/Advancement.java
@@ -238,6 +_,11 @@
          return advancement;
       }
 
+      public Advancement save(Consumer<Advancement> consumer, ResourceLocation id, net.minecraftforge.common.data.ExistingFileHelper fileHelper) {
+         return net.minecraftforge.common.data.AdvancementBuilderHelper.build(id, consumer, fileHelper, this);
+      }
+
+
       public JsonObject m_138400_() {
          if (this.f_138337_ == null) {
             this.f_138337_ = this.f_138338_.m_15985_(this.f_138336_.keySet());
@@ -311,6 +_,7 @@
       }
 
       public static Advancement.Builder m_138380_(JsonObject p_138381_, DeserializationContext p_138382_) {
+         if ((p_138381_ = net.minecraftforge.common.crafting.ConditionalAdvancement.processConditional(p_138381_)) == null) return null;
          ResourceLocation resourcelocation = p_138381_.has("parent") ? new ResourceLocation(GsonHelper.m_13906_(p_138381_, "parent")) : null;
          DisplayInfo displayinfo = p_138381_.has("display") ? DisplayInfo.m_14981_(GsonHelper.m_13930_(p_138381_, "display")) : null;
          AdvancementRewards advancementrewards = p_138381_.has("rewards") ? AdvancementRewards.m_9991_(GsonHelper.m_13930_(p_138381_, "rewards")) : AdvancementRewards.f_9978_;
