--- a/net/minecraft/advancements/Advancement.java
+++ b/net/minecraft/advancements/Advancement.java
@@ -127,7 +_,7 @@
       return this.f_138305_;
    }
 
-   public static class Builder {
+   public static class Builder implements net.minecraftforge.common.extensions.IForgeAdvancementBuilder {
       private ResourceLocation f_138332_;
       private Advancement f_138333_;
       private DisplayInfo f_138334_;
@@ -311,6 +_,7 @@
       }
 
       public static Advancement.Builder m_138380_(JsonObject p_138381_, DeserializationContext p_138382_) {
+         if ((p_138381_ = net.minecraftforge.common.crafting.ConditionalAdvancement.processConditional(p_138381_)) == null) return null;
          ResourceLocation resourcelocation = p_138381_.has("parent") ? new ResourceLocation(GsonHelper.m_13906_(p_138381_, "parent")) : null;
          DisplayInfo displayinfo = p_138381_.has("display") ? DisplayInfo.m_14981_(GsonHelper.m_13930_(p_138381_, "display")) : null;
          AdvancementRewards advancementrewards = p_138381_.has("rewards") ? AdvancementRewards.m_9991_(GsonHelper.m_13930_(p_138381_, "rewards")) : AdvancementRewards.f_9978_;
