--- ../src-base/minecraft/net/minecraft/advancements/AdvancementManager.java
+++ ../src-work/minecraft/net/minecraft/advancements/AdvancementManager.java
@@ -66,6 +66,7 @@
         field_192784_c.func_192087_a();
         Map<ResourceLocation, Advancement.Builder> map = this.func_192781_c();
         this.func_192777_a(map);
+        this.field_193768_e |= net.minecraftforge.common.ForgeHooks.loadAdvancements(map);
         field_192784_c.func_192083_a(map);
 
         for (Advancement advancement : field_192784_c.func_192088_b())
@@ -101,6 +102,7 @@
                 if (astring.length == 2)
                 {
                     ResourceLocation resourcelocation = new ResourceLocation(astring[0], astring[1]);
+                    if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.advancement.AdvancementAboutToLoadEvent(resourcelocation))) continue;
 
                     try
                     {
@@ -172,6 +174,7 @@
                         Path path2 = path.relativize(path1);
                         String s = FilenameUtils.removeExtension(path2.toString()).replaceAll("\\\\", "/");
                         ResourceLocation resourcelocation = new ResourceLocation("minecraft", s);
+                        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.advancement.AdvancementAboutToLoadEvent(resourcelocation))) continue;
 
                         if (!p_192777_1_.containsKey(resourcelocation))
                         {
