--- a/net/minecraft/client/renderer/item/BlockModelWrapper.java
+++ b/net/minecraft/client/renderer/item/BlockModelWrapper.java
@@ -55,7 +_,7 @@
             aint[j] = this.tints.get(j).calculate(p_378482_, p_376532_, p_376906_);
         }
 
-        RenderType rendertype = ItemBlockRenderTypes.getRenderType(p_378482_);
+        RenderType rendertype = net.minecraftforge.client.RenderTypeHelper.getFallbackItemRenderType(p_378482_, this.model);
         itemstackrenderstate$layerrenderstate.setupBlockModel(this.model, rendertype);
     }
 
