--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -18,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public interface BlockGetter extends LevelHeightAccessor {
+public interface BlockGetter extends LevelHeightAccessor, net.minecraftforge.common.extensions.IForgeBlockGetter {
    @Nullable
    BlockEntity getBlockEntity(BlockPos p_45570_);
 
@@ -32,7 +_,7 @@
    FluidState getFluidState(BlockPos p_45569_);
 
    default int getLightEmission(BlockPos p_45572_) {
-      return this.getBlockState(p_45572_).getLightEmission();
+      return this.getBlockState(p_45572_).getLightEmission(this, p_45572_);
    }
 
    default int getMaxLightLevel() {
