--- ../src-base/minecraft/net/minecraft/server/management/ItemInWorldManager.java
+++ ../src-work/minecraft/net/minecraft/server/management/ItemInWorldManager.java
@@ -7,12 +7,23 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.ItemSword;
 import net.minecraft.network.play.server.S23PacketBlockChange;
+import net.minecraft.tileentity.TileEntity;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
 import net.minecraft.world.WorldSettings;
+import net.minecraftforge.common.ForgeHooks;
+import net.minecraftforge.common.MinecraftForge;
+import cpw.mods.fml.common.eventhandler.Event;
+import net.minecraftforge.event.ForgeEventFactory;
+import net.minecraftforge.event.entity.player.PlayerDestroyItemEvent;
+import net.minecraftforge.event.entity.player.PlayerInteractEvent;
+import net.minecraftforge.event.entity.player.PlayerInteractEvent.Action;
+import net.minecraftforge.event.world.BlockEvent;
 
 public class ItemInWorldManager
 {
+    /** Forge reach distance */
+    private double blockReachDistance = 5.0d;
     public World field_73092_a;
     public EntityPlayerMP field_73090_b;
     private WorldSettings.GameType field_73091_c;
@@ -126,6 +137,13 @@
     {
         if (!this.field_73091_c.func_82752_c() || this.field_73090_b.func_82246_f(p_73074_1_, p_73074_2_, p_73074_3_))
         {
+            PlayerInteractEvent event = ForgeEventFactory.onPlayerInteract(field_73090_b, Action.LEFT_CLICK_BLOCK, p_73074_1_, p_73074_2_, p_73074_3_, p_73074_4_, field_73092_a);
+            if (event.isCanceled())
+            {
+                field_73090_b.field_71135_a.func_147359_a(new S23PacketBlockChange(p_73074_1_, p_73074_2_, p_73074_3_, field_73092_a));
+                return;
+            }
+
             if (this.func_73083_d())
             {
                 if (!this.field_73092_a.func_72886_a((EntityPlayer)null, p_73074_1_, p_73074_2_, p_73074_3_, p_73074_4_))
@@ -135,19 +153,36 @@
             }
             else
             {
-                this.field_73092_a.func_72886_a((EntityPlayer)null, p_73074_1_, p_73074_2_, p_73074_3_, p_73074_4_);
                 this.field_73089_e = this.field_73100_i;
                 float f = 1.0F;
                 Block block = this.field_73092_a.func_147439_a(p_73074_1_, p_73074_2_, p_73074_3_);
 
-                if (block.func_149688_o() != Material.field_151579_a)
+
+                if (!block.isAir(field_73092_a, p_73074_1_, p_73074_2_, p_73074_3_))
                 {
-                    block.func_149699_a(this.field_73092_a, p_73074_1_, p_73074_2_, p_73074_3_, this.field_73090_b);
-                    f = block.func_149737_a(this.field_73090_b, this.field_73090_b.field_70170_p, p_73074_1_, p_73074_2_, p_73074_3_);
+                    if (event.useBlock != Event.Result.DENY)
+                    {
+                        block.func_149699_a(field_73092_a, p_73074_1_, p_73074_2_, p_73074_3_, field_73090_b);
+                        field_73092_a.func_72886_a(null, p_73074_1_, p_73074_2_, p_73074_3_, p_73074_4_);
+                    }
+                    else
+                    {
+                        field_73090_b.field_71135_a.func_147359_a(new S23PacketBlockChange(p_73074_1_, p_73074_2_, p_73074_3_, field_73092_a));
+                    }
+                    f = block.func_149737_a(field_73090_b, field_73090_b.field_70170_p, p_73074_1_, p_73074_2_, p_73074_3_);
                 }
 
-                if (block.func_149688_o() != Material.field_151579_a && f >= 1.0F)
+                if (event.useItem == Event.Result.DENY)
                 {
+                    if (f >= 1.0f)
+                    {
+                        field_73090_b.field_71135_a.func_147359_a(new S23PacketBlockChange(p_73074_1_, p_73074_2_, p_73074_3_, field_73092_a));
+                    }
+                    return;
+                }
+
+                if (!block.isAir(field_73092_a, p_73074_1_, p_73074_2_, p_73074_3_) && f >= 1.0F)
+                {
                     this.func_73084_b(p_73074_1_, p_73074_2_, p_73074_3_);
                 }
                 else
@@ -171,7 +206,7 @@
             int l = this.field_73100_i - this.field_73089_e;
             Block block = this.field_73092_a.func_147439_a(p_73082_1_, p_73082_2_, p_73082_3_);
 
-            if (block.func_149688_o() != Material.field_151579_a)
+            if (!block.isAir(field_73092_a, p_73082_1_, p_73082_2_, p_73082_3_))
             {
                 float f = block.func_149737_a(this.field_73090_b, this.field_73090_b.field_70170_p, p_73082_1_, p_73082_2_, p_73082_3_) * (float)(l + 1);
 
@@ -202,10 +237,14 @@
 
     private boolean func_73079_d(int p_73079_1_, int p_73079_2_, int p_73079_3_)
     {
+        return removeBlock(p_73079_1_, p_73079_2_, p_73079_3_, false);
+    }
+    private boolean removeBlock(int p_73079_1_, int p_73079_2_, int p_73079_3_, boolean canHarvest)
+    {
         Block block = this.field_73092_a.func_147439_a(p_73079_1_, p_73079_2_, p_73079_3_);
         int l = this.field_73092_a.func_72805_g(p_73079_1_, p_73079_2_, p_73079_3_);
         block.func_149681_a(this.field_73092_a, p_73079_1_, p_73079_2_, p_73079_3_, l, this.field_73090_b);
-        boolean flag = this.field_73092_a.func_147468_f(p_73079_1_, p_73079_2_, p_73079_3_);
+        boolean flag = block.removedByPlayer(field_73092_a, field_73090_b, p_73079_1_, p_73079_2_, p_73079_3_, canHarvest);
 
         if (flag)
         {
@@ -217,29 +256,32 @@
 
     public boolean func_73084_b(int p_73084_1_, int p_73084_2_, int p_73084_3_)
     {
-        if (this.field_73091_c.func_82752_c() && !this.field_73090_b.func_82246_f(p_73084_1_, p_73084_2_, p_73084_3_))
+        BlockEvent.BreakEvent event = ForgeHooks.onBlockBreakEvent(field_73092_a, field_73091_c, field_73090_b, p_73084_1_, p_73084_2_, p_73084_3_);
+        if (event.isCanceled())
         {
             return false;
         }
-        else if (this.field_73091_c.func_77145_d() && this.field_73090_b.func_70694_bm() != null && this.field_73090_b.func_70694_bm().func_77973_b() instanceof ItemSword)
-        {
-            return false;
-        }
         else
         {
+            ItemStack stack = field_73090_b.func_71045_bC();
+            if (stack != null && stack.func_77973_b().onBlockStartBreak(stack, p_73084_1_, p_73084_2_, p_73084_3_, field_73090_b))
+            {
+                return false;
+            }
             Block block = this.field_73092_a.func_147439_a(p_73084_1_, p_73084_2_, p_73084_3_);
             int l = this.field_73092_a.func_72805_g(p_73084_1_, p_73084_2_, p_73084_3_);
             this.field_73092_a.func_72889_a(this.field_73090_b, 2001, p_73084_1_, p_73084_2_, p_73084_3_, Block.func_149682_b(block) + (this.field_73092_a.func_72805_g(p_73084_1_, p_73084_2_, p_73084_3_) << 12));
-            boolean flag = this.func_73079_d(p_73084_1_, p_73084_2_, p_73084_3_);
+            boolean flag = false;
 
             if (this.func_73083_d())
             {
+                flag = this.func_73079_d(p_73084_1_, p_73084_2_, p_73084_3_);
                 this.field_73090_b.field_71135_a.func_147359_a(new S23PacketBlockChange(p_73084_1_, p_73084_2_, p_73084_3_, this.field_73092_a));
             }
             else
             {
                 ItemStack itemstack = this.field_73090_b.func_71045_bC();
-                boolean flag1 = this.field_73090_b.func_146099_a(block);
+                boolean flag1 = block.canHarvestBlock(field_73090_b, l);
 
                 if (itemstack != null)
                 {
@@ -251,12 +293,18 @@
                     }
                 }
 
+                flag = this.removeBlock(p_73084_1_, p_73084_2_, p_73084_3_, flag1);
                 if (flag && flag1)
                 {
                     block.func_149636_a(this.field_73092_a, this.field_73090_b, p_73084_1_, p_73084_2_, p_73084_3_, l);
                 }
             }
 
+            // Drop experience
+            if (!this.func_73083_d() && flag && event != null)
+            {
+                block.func_149657_c(this.field_73092_a, p_73084_1_, p_73084_2_, p_73084_3_, event.getExpToDrop());
+            }
             return flag;
         }
     }
@@ -288,6 +336,7 @@
             if (itemstack1.field_77994_a == 0)
             {
                 p_73085_1_.field_71071_by.field_70462_a[p_73085_1_.field_71071_by.field_70461_c] = null;
+                MinecraftForge.EVENT_BUS.post(new PlayerDestroyItemEvent(field_73090_b, itemstack1));
             }
 
             if (!p_73085_1_.func_71039_bw())
@@ -301,31 +350,70 @@
 
     public boolean func_73078_a(EntityPlayer p_73078_1_, World p_73078_2_, ItemStack p_73078_3_, int p_73078_4_, int p_73078_5_, int p_73078_6_, int p_73078_7_, float p_73078_8_, float p_73078_9_, float p_73078_10_)
     {
-        if ((!p_73078_1_.func_70093_af() || p_73078_1_.func_70694_bm() == null) && p_73078_2_.func_147439_a(p_73078_4_, p_73078_5_, p_73078_6_).func_149727_a(p_73078_2_, p_73078_4_, p_73078_5_, p_73078_6_, p_73078_1_, p_73078_7_, p_73078_8_, p_73078_9_, p_73078_10_))
+        PlayerInteractEvent event = ForgeEventFactory.onPlayerInteract(p_73078_1_, Action.RIGHT_CLICK_BLOCK, p_73078_4_, p_73078_5_, p_73078_6_, p_73078_7_, p_73078_2_);
+        if (event.isCanceled())
         {
-            return true;
+            field_73090_b.field_71135_a.func_147359_a(new S23PacketBlockChange(p_73078_4_, p_73078_5_, p_73078_6_, field_73092_a));
+            return false;
         }
-        else if (p_73078_3_ == null)
+
+        if (p_73078_3_ != null && p_73078_3_.func_77973_b().onItemUseFirst(p_73078_3_, p_73078_1_, p_73078_2_, p_73078_4_, p_73078_5_, p_73078_6_, p_73078_7_, p_73078_8_, p_73078_9_, p_73078_10_))
         {
-            return false;
+            if (p_73078_3_.field_77994_a <= 0) ForgeEventFactory.onPlayerDestroyItem(field_73090_b, p_73078_3_);
+            return true;
         }
-        else if (this.func_73083_d())
+
+        Block block = p_73078_2_.func_147439_a(p_73078_4_, p_73078_5_, p_73078_6_);
+        boolean isAir = block.isAir(p_73078_2_, p_73078_4_, p_73078_5_, p_73078_6_);
+        boolean useBlock = !p_73078_1_.func_70093_af() || p_73078_1_.func_70694_bm() == null;
+        if (!useBlock) useBlock = p_73078_1_.func_70694_bm().func_77973_b().doesSneakBypassUse(p_73078_2_, p_73078_4_, p_73078_5_, p_73078_6_, p_73078_1_);
+        boolean result = false;
+
+        if (useBlock)
         {
-            int i1 = p_73078_3_.func_77960_j();
-            int j1 = p_73078_3_.field_77994_a;
-            boolean flag = p_73078_3_.func_77943_a(p_73078_1_, p_73078_2_, p_73078_4_, p_73078_5_, p_73078_6_, p_73078_7_, p_73078_8_, p_73078_9_, p_73078_10_);
-            p_73078_3_.func_77964_b(i1);
-            p_73078_3_.field_77994_a = j1;
-            return flag;
+            if (event.useBlock != Event.Result.DENY)
+            {
+                result = block.func_149727_a(p_73078_2_, p_73078_4_, p_73078_5_, p_73078_6_, p_73078_1_, p_73078_7_, p_73078_8_, p_73078_9_, p_73078_10_);
+            }
+            else
+            {
+                field_73090_b.field_71135_a.func_147359_a(new S23PacketBlockChange(p_73078_4_, p_73078_5_, p_73078_6_, field_73092_a));
+                result = event.useItem != Event.Result.ALLOW;
+            }
         }
-        else
+
+        if (p_73078_3_ != null && !result && event.useItem != Event.Result.DENY)
         {
-            return p_73078_3_.func_77943_a(p_73078_1_, p_73078_2_, p_73078_4_, p_73078_5_, p_73078_6_, p_73078_7_, p_73078_8_, p_73078_9_, p_73078_10_);
+            int meta = p_73078_3_.func_77960_j();
+            int size = p_73078_3_.field_77994_a;
+            result = p_73078_3_.func_77943_a(p_73078_1_, p_73078_2_, p_73078_4_, p_73078_5_, p_73078_6_, p_73078_7_, p_73078_8_, p_73078_9_, p_73078_10_);
+            if (func_73083_d())
+            {
+                p_73078_3_.func_77964_b(meta);
+                p_73078_3_.field_77994_a = size;
+            }
+            if (p_73078_3_.field_77994_a <= 0) ForgeEventFactory.onPlayerDestroyItem(field_73090_b, p_73078_3_);
         }
+
+        /* Re-enable if this causes bukkit incompatibility, or re-write client side to only send a single packet per right click.
+        if (par3ItemStack != null && ((!result && event.useItem != Event.Result.DENY) || event.useItem == Event.Result.ALLOW))
+        {
+            this.tryUseItem(thisPlayerMP, par2World, par3ItemStack);
+        }*/
+        return result;
     }
 
     public void func_73080_a(WorldServer p_73080_1_)
     {
         this.field_73092_a = p_73080_1_;
     }
+
+    public double getBlockReachDistance()
+    {
+        return blockReachDistance;
+    }
+    public void setBlockReachDistance(double distance)
+    {
+        blockReachDistance = distance;
+    }
 }
