--- a/net/minecraft/client/gui/advancements/AdvancementsScreen.java
+++ b/net/minecraft/client/gui/advancements/AdvancementsScreen.java
@@ -29,6 +_,7 @@
    private final Map<Advancement, AdvancementTabGui> tabs = Maps.newLinkedHashMap();
    private AdvancementTabGui selectedTab;
    private boolean isScrolling;
+   private static int tabPage, maxPages;
 
    public AdvancementsScreen(ClientAdvancementManager p_i47383_1_) {
       super(NarratorChatListener.NO_TITLE);
@@ -44,7 +_,13 @@
       } else {
          this.advancements.setSelectedTab(this.selectedTab == null ? null : this.selectedTab.getAdvancement(), true);
       }
-
+      if (this.tabs.size() > AdvancementTabType.MAX_TABS) {
+          int guiLeft = (this.width - 252) / 2;
+          int guiTop = (this.height - 140) / 2;
+          addButton(new net.minecraft.client.gui.widget.button.Button(guiLeft,            guiTop - 50, 20, 20, new net.minecraft.util.text.StringTextComponent("<"), b -> tabPage = Math.max(tabPage - 1, 0       )));
+          addButton(new net.minecraft.client.gui.widget.button.Button(guiLeft + 252 - 20, guiTop - 50, 20, 20, new net.minecraft.util.text.StringTextComponent(">"), b -> tabPage = Math.min(tabPage + 1, maxPages)));
+          maxPages = this.tabs.size() / AdvancementTabType.MAX_TABS;
+      }
    }
 
    public void removed() {
@@ -62,7 +_,7 @@
          int j = (this.height - 140) / 2;
 
          for(AdvancementTabGui advancementtabgui : this.tabs.values()) {
-            if (advancementtabgui.isMouseOver(i, j, p_231044_1_, p_231044_3_)) {
+            if (advancementtabgui.getPage() == tabPage && advancementtabgui.isMouseOver(i, j, p_231044_1_, p_231044_3_)) {
                this.advancements.setSelectedTab(advancementtabgui.getAdvancement(), true);
                break;
             }
@@ -86,6 +_,12 @@
       int i = (this.width - 252) / 2;
       int j = (this.height - 140) / 2;
       this.renderBackground(p_230430_1_);
+      if (maxPages != 0) {
+          net.minecraft.util.text.ITextComponent page = new net.minecraft.util.text.StringTextComponent(String.format("%d / %d", tabPage + 1, maxPages + 1));
+         int width = this.font.width(page);
+         RenderSystem.disableLighting();
+         this.font.drawShadow(p_230430_1_, page.getVisualOrderText(), i + (252 / 2) - (width / 2), j - 44, -1);
+      }
       this.renderInside(p_230430_1_, p_230430_2_, p_230430_3_, i, j);
       this.renderWindow(p_230430_1_, i, j);
       this.renderTooltips(p_230430_1_, p_230430_2_, p_230430_3_, i, j);
@@ -132,6 +_,7 @@
          this.minecraft.getTextureManager().bind(TABS_LOCATION);
 
          for(AdvancementTabGui advancementtabgui : this.tabs.values()) {
+            if (advancementtabgui.getPage() == tabPage)
             advancementtabgui.drawTab(p_238695_1_, p_238695_2_, p_238695_3_, advancementtabgui == this.selectedTab);
          }
 
@@ -139,6 +_,7 @@
          RenderSystem.defaultBlendFunc();
 
          for(AdvancementTabGui advancementtabgui1 : this.tabs.values()) {
+            if (advancementtabgui1.getPage() == tabPage)
             advancementtabgui1.drawIcon(p_238695_2_, p_238695_3_, this.itemRenderer);
          }
 
@@ -161,7 +_,7 @@
 
       if (this.tabs.size() > 1) {
          for(AdvancementTabGui advancementtabgui : this.tabs.values()) {
-            if (advancementtabgui.isMouseOver(p_238697_4_, p_238697_5_, (double)p_238697_2_, (double)p_238697_3_)) {
+            if (advancementtabgui.getPage() == tabPage && advancementtabgui.isMouseOver(p_238697_4_, p_238697_5_, (double)p_238697_2_, (double)p_238697_3_)) {
                this.renderTooltip(p_238697_1_, advancementtabgui.getTitle(), p_238697_2_, p_238697_3_);
             }
          }
