--- ../src-base/minecraft/net/minecraft/client/gui/advancements/GuiAdvancementTab.java
+++ ../src-work/minecraft/net/minecraft/client/gui/advancements/GuiAdvancementTab.java
@@ -37,6 +37,7 @@
     private int field_191814_q = Integer.MIN_VALUE;
     private float field_191815_r;
     private boolean field_192992_s;
+    private int page;
 
     public GuiAdvancementTab(Minecraft p_i47589_1_, GuiScreenAdvancements p_i47589_2_, AdvancementTabType p_i47589_3_, int p_i47589_4_, Advancement p_i47589_5_, DisplayInfo p_i47589_6_)
     {
@@ -160,9 +161,9 @@
         {
             for (AdvancementTabType advancementtabtype : AdvancementTabType.values())
             {
-                if (p_193936_2_ < advancementtabtype.func_192650_a())
+                if ((p_193936_2_ % AdvancementTabType.field_192659_e) < advancementtabtype.func_192650_a())
                 {
-                    return new GuiAdvancementTab(p_193936_0_, p_193936_1_, advancementtabtype, p_193936_2_, p_193936_3_, p_193936_3_.func_192068_c());
+                    return new GuiAdvancementTab(p_193936_0_, p_193936_1_, advancementtabtype, p_193936_2_ % AdvancementTabType.field_192659_e, p_193936_2_ / AdvancementTabType.field_192659_e, p_193936_3_, p_193936_3_.func_192068_c());
                 }
 
                 p_193936_2_ -= advancementtabtype.func_192650_a();
@@ -222,4 +223,16 @@
     {
         return this.field_193938_f;
     }
+
+    /***** FORGE *****/
+    public int getPage()
+    {
+        return this.page;
+    }
+
+    public GuiAdvancementTab(Minecraft p_i47589_1_, GuiScreenAdvancements p_i47589_2_, AdvancementTabType p_i47589_3_, int p_i47589_4_, int page, Advancement p_i47589_5_, DisplayInfo p_i47589_6_)
+    {
+        this(p_i47589_1_, p_i47589_2_, p_i47589_3_, p_i47589_4_, p_i47589_5_, p_i47589_6_);
+        this.page = page;
+    }
 }
