--- ../src-base/minecraft/net/minecraft/world/biome/BiomeGenJungle.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeGenJungle.java
@@ -68,7 +68,9 @@
         super.decorate(worldIn, rand, pos);
         int i = rand.nextInt(16) + 8;
         int j = rand.nextInt(16) + 8;
-        int k = rand.nextInt(worldIn.getHeight(pos.add(i, 0, j)).getY() * 2);
+        int height = worldIn.getHeight(pos.add(i, 0, j)).getY() * 2; // could == 0, which crashes nextInt
+        if (height < 1) height = 1;
+        int k = rand.nextInt(height);
         (new WorldGenMelon()).generate(worldIn, rand, pos.add(i, k, j));
         WorldGenVines worldgenvines = new WorldGenVines();
 
