--- a/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
+++ b/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
@@ -13,7 +_,7 @@
 
    public void m_5779_(FriendlyByteBuf p_134737_) {
       p_134737_.m_130130_(this.f_134720_);
-      p_134737_.m_130070_(this.f_134721_);
+      p_134737_.m_130070_(this.f_134721_ + "\0" + net.minecraftforge.network.NetworkConstants.NETVERSION + "\0");
       p_134737_.writeShort(this.f_134722_);
       p_134737_.m_130130_(this.f_134723_.m_293527_());
    }
