--- a/net/minecraft/world/level/storage/DimensionDataStorage.java
+++ b/net/minecraft/world/level/storage/DimensionDataStorage.java
@@ -48,9 +_,13 @@
    @Nullable
    public <T extends SavedData> T get(SavedData.Factory p_297465_, String p_164860_) {
       SavedData saveddata = this.cache.get(p_164860_);
+      if (saveddata == net.minecraftforge.common.util.DummySavedData.DUMMY) return null;
       if (saveddata == null && !this.cache.containsKey(p_164860_)) {
          saveddata = this.readSavedData(p_297465_.deserializer(), p_297465_.type(), p_164860_);
          this.cache.put(p_164860_, saveddata);
+      } else if (saveddata == null) {
+         this.cache.put(p_164860_, net.minecraftforge.common.util.DummySavedData.DUMMY);
+         return null;
       }
 
       return (T)saveddata;
