--- a/net/minecraft/world/level/storage/SavedDataStorage.java
+++ b/net/minecraft/world/level/storage/SavedDataStorage.java
@@ -121,6 +_,9 @@
             }
 
             int version = NbtUtils.getDataVersion(tag, 1343);
+            // Forge: Allow the data fixer to be null, leaving the modder responsible for keeping track of their own data formats
+            if (type == null)
+                return tag;
             return type.update(this.fixerUpper, tag, version, newVersion);
         }
     }
