--- ../src-base/minecraft/net/minecraft/world/storage/WorldInfo.java
+++ ../src-work/minecraft/net/minecraft/world/storage/WorldInfo.java
@@ -1,8 +1,11 @@
 package net.minecraft.world.storage;
+
+import java.util.Map;
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import net.minecraft.crash.CrashReportCategory;
+import net.minecraft.nbt.NBTBase;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.world.EnumGameType;
 import net.minecraft.world.GameRules;
@@ -35,6 +38,7 @@
     private boolean field_76110_t;
     private boolean field_76109_u;
     private GameRules field_82577_x;
+    private Map<String, NBTBase> additionalProperties;
 
     protected WorldInfo()
     {
@@ -534,4 +538,24 @@
     {
         return p_85131_0_.field_76110_t;
     }
+
+    /**
+     * Allow access to additional mod specific world based properties
+     * Used by FML to store mod list associated with a world, and maybe an id map
+     * Used by Forge to store the dimensions available to a world
+     * @param additionalProperties
+     */
+    public void setAdditionalProperties(Map<String,NBTBase> additionalProperties)
+    {
+        // one time set for this
+        if (this.additionalProperties == null)
+        {
+            this.additionalProperties = additionalProperties;
+        }
+    }
+
+    public NBTBase getAdditionalProperty(String additionalProperty)
+    {
+        return this.additionalProperties!=null? this.additionalProperties.get(additionalProperty) : null;
+    }
 }
