--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -324,7 +_,7 @@
 
    public boolean m_7243_(ItemStack p_218387_) {
       ItemStack itemstack = this.m_21120_(InteractionHand.MAIN_HAND);
-      return !itemstack.m_41619_() && this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_) && this.f_218303_.m_19183_(p_218387_) && this.m_245613_(itemstack, p_218387_);
+      return !itemstack.m_41619_() && this.f_218303_.m_19183_(p_218387_) && this.m_245613_(itemstack, p_218387_) && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.m_9236_(), this);
    }
 
    private boolean m_245613_(ItemStack p_252278_, ItemStack p_250405_) {
