--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -326,7 +_,7 @@
 
    public boolean m_7243_(ItemStack p_218387_) {
       ItemStack itemstack = this.m_21120_(InteractionHand.MAIN_HAND);
-      return !itemstack.m_41619_() && itemstack.m_41726_(p_218387_) && this.f_218303_.m_19183_(p_218387_) && this.f_19853_.m_46469_().m_46207_(GameRules.f_46132_);
+      return !itemstack.m_41619_() && itemstack.m_41726_(p_218387_) && this.f_218303_.m_19183_(p_218387_) && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_19853_, this);
    }
 
    protected void m_7581_(ItemEntity p_218359_) {
@@ -362,7 +_,7 @@
    }
 
    private boolean m_239812_() {
-      return this.f_238682_ == null || !this.f_238682_.m_203195_(this.m_20182_(), (double)GameEvent.f_238690_.m_157827_()) || !this.f_19853_.m_8055_(this.f_238682_).m_60713_(Blocks.f_50131_);
+      return this.f_238682_ == null || !this.f_238682_.m_203195_(this.m_20182_(), (double)GameEvent.f_238690_.m_157827_()) || !(this.f_19853_.m_8055_(this.f_238682_).m_60734_() instanceof net.minecraftforge.common.extensions.IForgeJukebox jukebox && jukebox.canDance(this));
    }
 
    public float m_218394_(float p_218395_) {
