--- ../src-base/minecraft/net/minecraft/entity/EntityAgeable.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityAgeable.java
@@ -31,7 +31,7 @@
         {
             if (!this.field_70170_p.field_72995_K)
             {
-                Class <? extends Entity > oclass = EntityList.func_90035_a(itemstack.func_77960_j());
+                Class <? extends Entity > oclass = EntityList.field_75625_b.get(net.minecraft.item.ItemMonsterPlacer.getEntityName(itemstack));
 
                 if (oclass != null && this.getClass() == oclass)
                 {
