--- ../src-base/minecraft/net/minecraft/entity/EntityTrackerEntry.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityTrackerEntry.java
@@ -436,12 +436,14 @@
                             p_73117_1_.field_71135_a.func_147359_a(new S1DPacketEntityEffect(this.field_73132_a.func_145782_y(), potioneffect));
                         }
                     }
+                    net.minecraftforge.event.ForgeEventFactory.onStartEntityTracking(field_73132_a, p_73117_1_);
                 }
             }
             else if (this.field_73134_o.contains(p_73117_1_))
             {
                 this.field_73134_o.remove(p_73117_1_);
                 p_73117_1_.func_152339_d(this.field_73132_a);
+                net.minecraftforge.event.ForgeEventFactory.onStopEntityTracking(field_73132_a, p_73117_1_);
             }
         }
     }
