--- ../src-base/minecraft/net/minecraft/entity/EntityList.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityList.java
@@ -90,9 +90,11 @@
     private static final Map < Class <? extends Entity > , Integer > field_75624_e = Maps. < Class <? extends Entity > , Integer > newHashMap();
     private static final Map<String, Integer> field_180126_g = Maps.<String, Integer>newHashMap();
     public static final Map<String, EntityList.EntityEggInfo> field_75627_a = Maps.<String, EntityList.EntityEggInfo>newLinkedHashMap();
+    private static final Map<String, java.lang.reflect.Constructor<? extends Entity>> CACHED_CONSTRUCTORS = Maps.newHashMap();
 
     public static void func_75618_a(Class <? extends Entity > p_75618_0_, String p_75618_1_, int p_75618_2_)
     {
+        if (p_75618_2_ < 0 || p_75618_2_ > 255) throw new IllegalArgumentException("Attempted to register a entity with invalid ID: " + p_75618_2_ + " Name: " + p_75618_1_ + " Class: " + p_75618_0_);
         if (field_75625_b.containsKey(p_75618_1_))
         {
             throw new IllegalArgumentException("ID is already registered: " + p_75618_1_);
@@ -132,12 +134,20 @@
 
         try
         {
+            java.lang.reflect.Constructor<? extends Entity> ctr = CACHED_CONSTRUCTORS.get(p_75620_0_);
+            if (ctr == null)
+            {
             Class <? extends Entity > oclass = (Class)field_75625_b.get(p_75620_0_);
 
             if (oclass != null)
             {
-                entity = (Entity)oclass.getConstructor(new Class[] {World.class}).newInstance(new Object[] {p_75620_1_});
+                ctr = oclass.getConstructor(World.class);
+                CACHED_CONSTRUCTORS.put(p_75620_0_, ctr);
             }
+            }
+
+            if (ctr != null)
+                entity = ctr.newInstance(p_75620_1_);
         }
         catch (Exception exception)
         {
@@ -154,12 +164,7 @@
 
         try
         {
-            Class <? extends Entity > oclass = (Class)field_75625_b.get(p_75615_0_.func_74779_i("id"));
-
-            if (oclass != null)
-            {
-                entity = (Entity)oclass.getConstructor(new Class[] {World.class}).newInstance(new Object[] {p_75615_1_});
-            }
+            entity = func_75620_a(p_75615_0_.func_74779_i("id"), p_75615_1_);
         }
         catch (Exception exception)
         {
@@ -168,7 +173,17 @@
 
         if (entity != null)
         {
+            try
+            {
             entity.func_70020_e(p_75615_0_);
+            }
+            catch (Exception e)
+            {
+                net.minecraftforge.fml.common.FMLLog.log(org.apache.logging.log4j.Level.ERROR, e,
+                        "An Entity %s(%s) has thrown an exception during loading, its state cannot be restored. Report this to the mod author",
+                        p_75615_0_.func_74779_i("id"), field_75625_b.get(p_75615_0_.func_74779_i("id")).getName());
+                entity = null;
+            }
         }
         else
         {
@@ -208,7 +223,8 @@
     @Nullable
     public static Entity func_188429_b(String p_188429_0_, World p_188429_1_)
     {
-        return func_75616_a(func_180122_a(p_188429_0_), p_188429_1_);
+        Entity e = func_75620_a(p_188429_0_, p_188429_1_); // Forge: Support entities without global ID
+        return e == null ? func_75620_a("Pig", p_188429_1_) : e;
     }
 
     public static int func_75619_a(Entity p_75619_0_)
