--- ../src-base/minecraft/net/minecraft/entity/EntityList.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityList.java
@@ -189,7 +189,17 @@
         }
         else
         {
+            try
+            {
             entity.func_70020_e(p_75615_0_);
+            }
+            catch (Exception e)
+            {
+                net.minecraftforge.fml.common.FMLLog.log(org.apache.logging.log4j.Level.ERROR, e,
+                        "An Entity %s(%s) has thrown an exception during loading, its state cannot be restored. Report this to the mod author",
+                        p_75615_0_.func_74779_i("id"), entity.func_70005_c_());
+                entity = null;
+            }
         }
 
         return entity;
