--- ../src-base/minecraft/net/minecraft/entity/EntitySpawnPlacementRegistry.java
+++ ../src-work/minecraft/net/minecraft/entity/EntitySpawnPlacementRegistry.java
@@ -50,9 +50,14 @@
 
     public static EntityLiving.SpawnPlacementType func_180109_a(Class<?> p_180109_0_)
     {
-        return field_180110_a.get(p_180109_0_);
+        return field_180110_a.getOrDefault(p_180109_0_, EntityLiving.SpawnPlacementType.ON_GROUND);
     }
 
+    public static void setPlacementType(Class<? extends Entity> entityClass, EntityLiving.SpawnPlacementType placementType)
+    {
+        field_180110_a.putIfAbsent(entityClass, placementType);
+    }
+
     static
     {
         field_180110_a.put(EntityBat.class, EntityLiving.SpawnPlacementType.ON_GROUND);
