--- ../src-base/minecraft/net/minecraft/entity/EntityLivingBase.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityLivingBase.java
@@ -19,6 +19,7 @@
 import net.minecraft.entity.item.EntityItem;
 import net.minecraft.entity.item.EntityXPOrb;
 import net.minecraft.entity.monster.EntityZombie;
+import net.minecraft.entity.passive.EntityPig;
 import net.minecraft.entity.passive.EntityWolf;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.EntityPlayerMP;
@@ -48,6 +49,7 @@
 import net.minecraft.util.Vec3;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
+import net.minecraftforge.common.ForgeHooks;
 
 public abstract class EntityLivingBase extends Entity
 {
@@ -225,7 +227,7 @@
                 }
             }
 
-            if (!this.field_70170_p.field_72995_K && this.func_70115_ae() && this.field_70154_o instanceof EntityLivingBase)
+            if (!this.field_70170_p.field_72995_K && this.func_70115_ae() && this.field_70154_o != null && field_70154_o.shouldDismountInWater(this))
             {
                 this.func_70078_a((Entity)null);
             }
@@ -374,6 +376,7 @@
     {
         this.field_70755_b = p_70604_1_;
         this.field_70756_c = this.field_70173_aa;
+        ForgeHooks.onLivingSetAttackTarget(this, p_70604_1_);
     }
 
     public EntityLivingBase func_110144_aD()
@@ -661,7 +664,6 @@
         return this.func_70668_bt() == EnumCreatureAttribute.UNDEAD;
     }
 
-    @SideOnly(Side.CLIENT)
     public void func_70618_n(int p_70618_1_)
     {
         this.field_70713_bf.remove(Integer.valueOf(p_70618_1_));
@@ -710,6 +712,8 @@
 
     public void func_70691_i(float p_70691_1_)
     {
+        p_70691_1_ = net.minecraftforge.event.ForgeEventFactory.onLivingHeal(this, p_70691_1_);
+        if (p_70691_1_ <= 0) return;
         float f1 = this.func_110143_aJ();
 
         if (f1 > 0.0F)
@@ -730,6 +734,7 @@
 
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
+        if (ForgeHooks.onLivingAttack(this, p_70097_1_, p_70097_2_)) return false;
         if (this.func_85032_ar())
         {
             return false;
@@ -796,9 +801,9 @@
                         this.field_70718_bc = 100;
                         this.field_70717_bb = (EntityPlayer)entity;
                     }
-                    else if (entity instanceof EntityWolf)
+                    else if (entity instanceof net.minecraft.entity.passive.EntityTameable)
                     {
-                        EntityWolf entitywolf = (EntityWolf)entity;
+                        net.minecraft.entity.passive.EntityTameable entitywolf = (net.minecraft.entity.passive.EntityTameable)entity;
 
                         if (entitywolf.func_70909_n())
                         {
@@ -883,6 +888,7 @@
 
     public void func_70645_a(DamageSource p_70645_1_)
     {
+        if (ForgeHooks.onLivingDeath(this, p_70645_1_)) return;
         Entity entity = p_70645_1_.func_76346_g();
         EntityLivingBase entitylivingbase = this.func_94060_bK();
 
@@ -908,6 +914,10 @@
                 i = EnchantmentHelper.func_77519_f((EntityLivingBase)entity);
             }
 
+            captureDrops = true;
+            capturedDrops.clear();
+            int j = 0;
+
             if (this.func_146066_aG() && this.field_70170_p.func_82736_K().func_82766_b("doMobLoot"))
             {
                 this.func_70628_a(this.field_70718_bc > 0, i);
@@ -915,7 +925,7 @@
 
                 if (this.field_70718_bc > 0)
                 {
-                    int j = this.field_70146_Z.nextInt(200) - i;
+                    j = this.field_70146_Z.nextInt(200) - i;
 
                     if (j < 5)
                     {
@@ -923,6 +933,16 @@
                     }
                 }
             }
+
+            captureDrops = false;
+
+            if (!ForgeHooks.onLivingDrops(this, p_70645_1_, capturedDrops, i, field_70718_bc > 0, j))
+            {
+                for (EntityItem item : capturedDrops)
+                {
+                    field_70170_p.func_72838_d(item);
+                }
+            }
         }
 
         this.field_70170_p.func_72960_a(this, (byte)3);
@@ -971,7 +991,7 @@
         int j = MathHelper.func_76128_c(this.field_70121_D.field_72338_b);
         int k = MathHelper.func_76128_c(this.field_70161_v);
         Block block = this.field_70170_p.func_147439_a(i, j, k);
-        return block == Blocks.field_150468_ap || block == Blocks.field_150395_bd;
+        return ForgeHooks.isLivingOnLadder(block, field_70170_p, i, j, k, this);
     }
 
     public boolean func_70089_S()
@@ -981,6 +1001,8 @@
 
     protected void func_70069_a(float p_70069_1_)
     {
+        p_70069_1_ = ForgeHooks.onLivingFall(this, p_70069_1_);
+        if (p_70069_1_ <= 0) return;
         super.func_70069_a(p_70069_1_);
         PotionEffect potioneffect = this.func_70660_b(Potion.field_76430_j);
         float f1 = potioneffect != null ? (float)(potioneffect.func_76458_c() + 1) : 0.0F;
@@ -1060,7 +1082,7 @@
         {
             if (this instanceof EntityZombie)
             {
-                p_70672_2_ = p_70672_2_;
+                //par2 = par2; // Forge: Noop Warning
             }
 
             int i;
@@ -1104,6 +1126,8 @@
     {
         if (!this.func_85032_ar())
         {
+            p_70665_2_ = ForgeHooks.onLivingHurt(this, p_70665_1_, p_70665_2_);
+            if (p_70665_2_ <= 0) return;
             p_70665_2_ = this.func_70655_b(p_70665_1_, p_70665_2_);
             p_70665_2_ = this.func_70672_c(p_70665_1_, p_70665_2_);
             float f1 = p_70665_2_;
@@ -1152,6 +1176,17 @@
 
     public void func_71038_i()
     {
+        ItemStack stack = this.func_70694_bm();
+
+        if (stack != null && stack.func_77973_b() != null)
+        {
+            Item item = stack.func_77973_b();
+            if (item.onEntitySwing(this, stack))
+            {
+                return;
+            }
+        }
+
         if (!this.field_82175_bq || this.field_110158_av >= this.func_82166_i() / 2 || this.field_110158_av < 0)
         {
             this.field_110158_av = -1;
@@ -1347,6 +1382,7 @@
         }
 
         this.field_70160_al = true;
+        ForgeHooks.onLivingJump(this);
     }
 
     public void func_70612_e(float p_70612_1_, float p_70612_2_)
@@ -1521,6 +1557,7 @@
 
     public void func_70071_h_()
     {
+        if (ForgeHooks.onLivingUpdate(this)) return;
         super.func_70071_h_();
 
         if (!this.field_70170_p.field_72995_K)
@@ -2007,6 +2044,44 @@
         return this.func_96124_cp() != null ? this.func_96124_cp().func_142054_a(p_142012_1_) : false;
     }
 
+    /***
+     * Removes all potion effects that have curativeItem as a curative item for its effect
+     * @param curativeItem The itemstack we are using to cure potion effects
+     */
+    public void curePotionEffects(ItemStack curativeItem)
+    {
+        Iterator<Integer> potionKey = field_70713_bf.keySet().iterator();
+
+        if (field_70170_p.field_72995_K)
+        {
+            return;
+        }
+
+        while (potionKey.hasNext())
+        {
+            Integer key = potionKey.next();
+            PotionEffect effect = (PotionEffect)field_70713_bf.get(key);
+
+            if (effect.isCurativeItem(curativeItem))
+            {
+                potionKey.remove();
+                func_70688_c(effect);
+            }
+        }
+    }
+
+    /**
+     * Returns true if the entity's rider (EntityPlayer) should face forward when mounted.
+     * currently only used in vanilla code by pigs.
+     *
+     * @param player The player who is riding the entity.
+     * @return If the player should orient the same direction as this entity.
+     */
+    public boolean shouldRiderFaceForward(EntityPlayer player)
+    {
+        return this instanceof EntityPig;
+    }
+
     public void func_152111_bt() {}
 
     public void func_152112_bu() {}
