--- a/net/minecraft/entity/IShearable.java
+++ b/net/minecraft/entity/IShearable.java
@@ -2,8 +_,11 @@
 
 import net.minecraft.util.SoundCategory;
 
+@Deprecated // Forge: Use IForgeShearable
 public interface IShearable {
+   @Deprecated // Forge: Use IForgeShearable
    void func_230263_a_(SoundCategory p_230263_1_);
 
+   @Deprecated // Forge: Use IForgeShearable
    boolean func_230262_K__();
 }
