--- a/net/minecraft/entity/IShearable.java
+++ b/net/minecraft/entity/IShearable.java
@@ -2,8 +2,11 @@
 
 import net.minecraft.util.SoundCategory;
 
+/**@deprecated Use {@link net.minecraftforge.common.IForgeShearable}*/@Deprecated
 public interface IShearable {
+   /**@deprecated Use {@link net.minecraftforge.common.IForgeShearable#onSheared(net.minecraft.entity.player.PlayerEntity, net.minecraft.item.ItemStack, net.minecraft.world.World, net.minecraft.util.math.BlockPos, int)}*/@Deprecated
    void func_230263_a_(SoundCategory p_230263_1_);
 
+   /**@deprecated Use {@link net.minecraftforge.common.IForgeShearable#isShearable(net.minecraft.item.ItemStack, net.minecraft.world.World, net.minecraft.util.math.BlockPos)}*/@Deprecated
    boolean func_230262_K__();
 }
