--- a/net/minecraft/world/entity/ai/attributes/Attributes.java
+++ b/net/minecraft/world/entity/ai/attributes/Attributes.java
@@ -2,6 +_,7 @@
 
 import net.minecraft.core.Registry;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Attributes {
    public static final Attribute f_22276_ = m_22290_("generic.max_health", (new RangedAttribute("attribute.name.generic.max_health", 20.0D, 1.0D, 1024.0D)).m_22084_(true));
    public static final Attribute f_22277_ = m_22290_("generic.follow_range", new RangedAttribute("attribute.name.generic.follow_range", 32.0D, 0.0D, 2048.0D));
