--- ../src-base/minecraft/net/minecraft/network/play/client/C12PacketUpdateSign.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/C12PacketUpdateSign.java
@@ -32,7 +32,7 @@
 
         for (int i = 0; i < 4; ++i)
         {
-            this.field_149590_d[i] = p_148837_1_.func_179258_d();
+            this.field_149590_d[i] = IChatComponent.Serializer.func_150699_a(p_148837_1_.func_150789_c(384));
         }
     }
 
