--- ../src-base/minecraft/net/minecraft/network/play/client/C17PacketCustomPayload.java
+++ ../src-work/minecraft/net/minecraft/network/play/client/C17PacketCustomPayload.java
@@ -47,7 +47,11 @@
     public void func_148840_b(PacketBuffer p_148840_1_) throws IOException
     {
         p_148840_1_.func_180714_a(this.field_149562_a);
+        synchronized(this.field_149561_c) { //This may be access multiple times, from multiple threads, lets be safe.
+        this.field_149561_c.markReaderIndex();
         p_148840_1_.writeBytes((ByteBuf)this.field_149561_c);
+        this.field_149561_c.resetReaderIndex();
+        }
     }
 
     public void func_148833_a(INetHandlerPlayServer p_148833_1_)
