--- a/net/minecraft/world/item/trading/MerchantOffer.java
+++ b/net/minecraft/world/item/trading/MerchantOffer.java
@@ -90,7 +_,7 @@
       } else {
          int i = this.baseCostA.getCount();
          int j = Math.max(0, Mth.floor((float)(i * this.demand) * this.priceMultiplier));
-         return this.baseCostA.copyWithCount(Mth.clamp(i + j + this.specialPriceDiff, 1, this.baseCostA.getItem().getMaxStackSize()));
+         return this.baseCostA.copyWithCount(Mth.clamp(i + j + this.specialPriceDiff, 1, this.baseCostA.getMaxStackSize()));
       }
    }
 
@@ -194,7 +_,7 @@
          return true;
       } else {
          ItemStack itemstack = p_45366_.copy();
-         if (itemstack.getItem().canBeDepleted()) {
+         if (itemstack.getItem().isDamageable(itemstack)) {
             itemstack.setDamageValue(itemstack.getDamageValue());
          }
 
