--- a/net/minecraft/client/renderer/entity/layers/ArmorLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/ArmorLayer.java
@@ -55,24 +55,25 @@
          ArmorItem armoritem = (ArmorItem)itemstack.func_77973_b();
          if (armoritem.func_185083_B_() == p_188361_9_) {
             A a = this.func_215337_a(p_188361_9_);
+            a = getArmorModelHook(p_188361_1_, itemstack, p_188361_9_, a);
             ((BipedModel)this.func_215332_c()).func_217148_a(a);
             a.func_212843_a_(p_188361_1_, p_188361_2_, p_188361_3_, p_188361_4_);
             this.func_188359_a(a, p_188361_9_);
             boolean flag = this.func_188363_b(p_188361_9_);
-            this.func_215333_a(this.func_177181_a(armoritem, flag));
-            if (armoritem instanceof DyeableArmorItem) {
-               int i = ((DyeableArmorItem)armoritem).func_200886_f(itemstack);
+            this.func_215333_a(this.getArmorResource(p_188361_1_, itemstack, p_188361_9_, null));
+            if (armoritem instanceof net.minecraft.item.IDyeableArmorItem) { // Allow this for anything, not only cloth
+               int i = ((net.minecraft.item.IDyeableArmorItem)armoritem).func_200886_f(itemstack);
                float f = (float)(i >> 16 & 255) / 255.0F;
                float f1 = (float)(i >> 8 & 255) / 255.0F;
                float f2 = (float)(i & 255) / 255.0F;
                GlStateManager.color4f(this.field_177184_f * f, this.field_177185_g * f1, this.field_177192_h * f2, this.field_177187_e);
                a.func_78088_a(p_188361_1_, p_188361_2_, p_188361_3_, p_188361_5_, p_188361_6_, p_188361_7_, p_188361_8_);
-               this.func_215333_a(this.func_177178_a(armoritem, flag, "overlay"));
+               this.func_215333_a(this.getArmorResource(p_188361_1_, itemstack, p_188361_9_, "overlay"));
             }
 
             GlStateManager.color4f(this.field_177184_f, this.field_177185_g, this.field_177192_h, this.field_177187_e);
             a.func_78088_a(p_188361_1_, p_188361_2_, p_188361_3_, p_188361_5_, p_188361_6_, p_188361_7_, p_188361_8_);
-            if (!this.field_177193_i && itemstack.func_77948_v()) {
+            if (!this.field_177193_i && itemstack.func_77962_s()) {
                func_215338_a(this::func_215333_a, p_188361_1_, a, p_188361_2_, p_188361_3_, p_188361_4_, p_188361_5_, p_188361_6_, p_188361_7_, p_188361_8_);
             }
 
@@ -125,15 +126,60 @@
       gamerenderer.func_191514_d(false);
    }
 
+   @Deprecated //Use the more sensitive version getArmorResource below
    private ResourceLocation func_177181_a(ArmorItem p_177181_1_, boolean p_177181_2_) {
       return this.func_177178_a(p_177181_1_, p_177181_2_, (String)null);
    }
 
+   @Deprecated //Use the more sensitive version getArmorResource below
    private ResourceLocation func_177178_a(ArmorItem p_177178_1_, boolean p_177178_2_, @Nullable String p_177178_3_) {
       String s = "textures/models/armor/" + p_177178_1_.func_200880_d().func_200897_d() + "_layer_" + (p_177178_2_ ? 2 : 1) + (p_177178_3_ == null ? "" : "_" + p_177178_3_) + ".png";
       return field_177191_j.computeIfAbsent(s, ResourceLocation::new);
    }
 
+
+   /*=================================== FORGE START =========================================*/
+
+   /**
+    * Hook to allow item-sensitive armor model. for LayerBipedArmor.
+    */
+   protected A getArmorModelHook(LivingEntity entity, ItemStack itemStack, EquipmentSlotType slot, A model) {
+      return model;
+   }
+
+   /**
+    * More generic ForgeHook version of the above function, it allows for Items to have more control over what texture they provide.
+    *
+    * @param entity Entity wearing the armor
+    * @param stack ItemStack for the armor
+    * @param slot Slot ID that the item is in
+    * @param type Subtype, can be null or "overlay"
+    * @return ResourceLocation pointing at the armor's texture
+    */
+   public ResourceLocation getArmorResource(net.minecraft.entity.Entity entity, ItemStack stack, EquipmentSlotType slot, @javax.annotation.Nullable String type) {
+      ArmorItem item = (ArmorItem)stack.func_77973_b();
+      String texture = item.func_200880_d().func_200897_d();
+      String domain = "minecraft";
+      int idx = texture.indexOf(':');
+      if (idx != -1)
+      {
+         domain = texture.substring(0, idx);
+         texture = texture.substring(idx + 1);
+      }
+      String s1 = String.format("%s:textures/models/armor/%s_layer_%d%s.png", domain, texture, (func_188363_b(slot) ? 2 : 1), type == null ? "" : String.format("_%s", type));
+
+      s1 = net.minecraftforge.client.ForgeHooksClient.getArmorTexture(entity, stack, s1, slot, type);
+      ResourceLocation resourcelocation = (ResourceLocation)field_177191_j.get(s1);
+
+      if (resourcelocation == null)
+      {
+         resourcelocation = new ResourceLocation(s1);
+         field_177191_j.put(s1, resourcelocation);
+      }
+
+      return resourcelocation;
+   }
+   /*=================================== FORGE END ===========================================*/
    protected abstract void func_188359_a(A p_188359_1_, EquipmentSlotType p_188359_2_);
 
    protected abstract void func_177194_a(A p_177194_1_);
