--- a/net/minecraft/client/renderer/entity/layers/ElytraLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/ElytraLayer.java
@@ -23,7 +_,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public class ElytraLayer<T extends LivingEntity, M extends EntityModel<T>> extends RenderLayer<T, M> {
+public class ElytraLayer<T extends LivingEntity, M extends EntityModel<T>> extends RenderLayer<T, M> implements net.minecraftforge.client.extensions.IForgeElytraLayer<T> {
     public static final ResourceLocation WINGS_LOCATION = new ResourceLocation("textures/entity/elytra.png");
     private final ElytraModel<T> elytraModel;
 
@@ -45,7 +_,7 @@
         float p_116960_
     ) {
         ItemStack itemstack = p_116954_.getItemBySlot(EquipmentSlot.CHEST);
-        if (itemstack.is(Items.ELYTRA)) {
+        if (shouldRender(itemstack, p_116954_)) {
             ResourceLocation resourcelocation;
             if (p_116954_ instanceof AbstractClientPlayer abstractclientplayer) {
                 PlayerSkin playerskin = abstractclientplayer.getSkin();
@@ -54,10 +_,10 @@
                 } else if (playerskin.capeTexture() != null && abstractclientplayer.isModelPartShown(PlayerModelPart.CAPE)) {
                     resourcelocation = playerskin.capeTexture();
                 } else {
-                    resourcelocation = WINGS_LOCATION;
+                    resourcelocation = getElytraTexture(itemstack, p_116954_);
                 }
             } else {
-                resourcelocation = WINGS_LOCATION;
+                resourcelocation = getElytraTexture(itemstack, p_116954_);
             }
 
             p_116951_.pushPose();
