--- a/net/minecraft/client/renderer/entity/layers/HumanoidArmorLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/HumanoidArmorLayer.java
@@ -66,12 +_,13 @@
         Equippable equippable = itemStack.get(DataComponents.EQUIPPABLE);
         if (equippable != null && shouldRender(equippable, slot)) {
             A model = this.getArmorModel(state, slot);
+            var newModel = this.getArmorModel(state, slot, itemStack, model);
             EquipmentClientInfo.LayerType layerType = state.isBaby && state.entityType != EntityTypes.ARMOR_STAND
                 ? EquipmentClientInfo.LayerType.HUMANOID_BABY
                 : (this.usesInnerModel(slot) ? EquipmentClientInfo.LayerType.HUMANOID_LEGGINGS : EquipmentClientInfo.LayerType.HUMANOID);
             this.equipmentRenderer
                 .renderLayers(
-                    layerType, equippable.assetId().orElseThrow(), model, state, itemStack, poseStack, submitNodeCollector, lightCoords, state.outlineColor
+                    layerType, equippable.assetId().orElseThrow(), newModel, state, itemStack, poseStack, submitNodeCollector, lightCoords, state.outlineColor
                 );
         }
     }
@@ -82,5 +_,12 @@
 
     private boolean usesInnerModel(final EquipmentSlot slot) {
         return slot == EquipmentSlot.LEGS;
+    }
+
+    /**
+     * Hook to allow item-sensitive armor model. for HumanoidArmorLayer.
+     */
+    protected net.minecraft.client.model.Model getArmorModel(S state, EquipmentSlot slot, ItemStack itemStack, A _default) {
+       return net.minecraftforge.client.ForgeHooksClient.getArmorModel(state, itemStack, slot, _default);
     }
 }
