--- ../src-base/minecraft/net/minecraft/client/renderer/entity/layers/LayerBipedArmor.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/layers/LayerBipedArmor.java
@@ -49,4 +49,10 @@
     {
         p_177194_1_.func_178719_a(false);
     }
+
+    @Override
+    protected ModelBiped getArmorModelHook(net.minecraft.entity.EntityLivingBase entity, net.minecraft.item.ItemStack itemStack, int slot, ModelBiped model)
+    {
+        return net.minecraftforge.client.ForgeHooksClient.getArmorModel(entity, itemStack, slot, model);
+    }
 }
