--- ../src-base/minecraft/net/minecraft/block/state/IBlockProperties.java
+++ ../src-work/minecraft/net/minecraft/block/state/IBlockProperties.java
@@ -28,9 +28,13 @@
 
     boolean func_189884_a(Entity p_189884_1_);
 
+    @Deprecated //Forge location aware version below
     int func_185891_c();
+    int getLightOpacity(IBlockAccess world, BlockPos pos);
 
+    @Deprecated //Forge location aware version below
     int func_185906_d();
+    int getLightValue(IBlockAccess world, BlockPos pos);
 
     @SideOnly(Side.CLIENT)
     boolean func_185895_e();
@@ -95,8 +99,14 @@
 
     RayTraceResult func_185910_a(World p_185910_1_, BlockPos p_185910_2_, Vec3d p_185910_3_, Vec3d p_185910_4_);
 
+    @Deprecated // Forge: Use isSideSolid(IBlockAccess, BlockPos, EnumFacing.UP) instead
     boolean func_185896_q();
 
+    //Forge added functions
+    boolean doesSideBlockRendering(IBlockAccess world, BlockPos pos, EnumFacing side);
+    boolean isSideSolid(IBlockAccess world, BlockPos pos, EnumFacing side);
+    boolean isStickyBlock(IBlockAccess world, BlockPos pos);
+
     Vec3d func_191059_e(IBlockAccess p_191059_1_, BlockPos p_191059_2_);
 
     boolean func_191058_s();
