--- ../src_base/minecraft/net/minecraft/world/gen/ChunkProviderServer.java
+++ ../src_work/minecraft/net/minecraft/world/gen/ChunkProviderServer.java
@@ -6,6 +6,9 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
+
+import net.minecraftforge.common.DimensionManager;
+import net.minecraftforge.common.ForgeChunkManager;
 
 import cpw.mods.fml.common.registry.GameRegistry;
 import net.minecraft.crash.CrashReport;
@@ -66,7 +69,7 @@
      */
     public void unloadChunksIfNotNearSpawn(int par1, int par2)
     {
-        if (this.worldObj.provider.canRespawnHere())
+        if (this.worldObj.provider.canRespawnHere() && DimensionManager.shouldLoadSpawn(this.worldObj.provider.dimensionId))
         {
             ChunkCoordinates var3 = this.worldObj.getSpawnPoint();
             int var4 = par1 * 16 + 8 - var3.posX;
@@ -109,7 +112,11 @@
 
         if (var5 == null)
         {
-            var5 = this.safeLoadChunk(par1, par2);
+            var5 = ForgeChunkManager.fetchDormantChunk(var3, this.worldObj);
+            if (var5 == null)
+            {
+                var5 = this.safeLoadChunk(par1, par2);
+            }
 
             if (var5 == null)
             {
@@ -306,6 +313,11 @@
     {
         if (!this.worldObj.canNotSave)
         {
+            for (ChunkCoordIntPair forced : this.worldObj.getPersistentChunks().keySet())
+            {
+                this.chunksToUnload.remove(ChunkCoordIntPair.chunkXZ2Int(forced.chunkXPos, forced.chunkZPos));
+            }
+
             for (int var1 = 0; var1 < 100; ++var1)
             {
                 if (!this.chunksToUnload.isEmpty())
@@ -318,6 +330,11 @@
                     this.chunksToUnload.remove(var2);
                     this.loadedChunkHashMap.remove(var2.longValue());
                     this.loadedChunks.remove(var3);
+                    ForgeChunkManager.putDormantChunk(ChunkCoordIntPair.chunkXZ2Int(var3.xPosition, var3.zPosition), var3);
+                    if(loadedChunks.size() == 0 && ForgeChunkManager.getPersistentChunksFor(this.worldObj).size() == 0 && !DimensionManager.shouldLoadSpawn(this.worldObj.provider.dimensionId)) {
+                        DimensionManager.unloadWorld(this.worldObj.provider.dimensionId);
+                        return currentChunkProvider.unload100OldestChunks();
+                    }
                 }
             }
 
