--- ../src-base/minecraft/net/minecraft/world/gen/ChunkGeneratorOverworld.java
+++ ../src-work/minecraft/net/minecraft/world/gen/ChunkGeneratorOverworld.java
@@ -61,6 +61,16 @@
 
     public ChunkGeneratorOverworld(World p_i46668_1_, long p_i46668_2_, boolean p_i46668_4_, String p_i46668_5_)
     {
+        {
+            field_186003_v = net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(field_186003_v, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.CAVE);
+            field_186004_w = (MapGenStronghold)net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(field_186004_w, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.STRONGHOLD);
+            field_186005_x = (MapGenVillage)net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(field_186005_x, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.VILLAGE);
+            field_186006_y = (MapGenMineshaft)net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(field_186006_y, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.MINESHAFT);
+            field_186007_z = (MapGenScatteredFeature)net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(field_186007_z, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.SCATTERED_FEATURE);
+            field_185979_A = net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(field_185979_A, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.RAVINE);
+            field_185980_B = (StructureOceanMonument)net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(field_185980_B, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.OCEAN_MONUMENT);
+            field_191060_C = (WoodlandMansion)net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(field_191060_C, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.WOODLAND_MANSION);
+        }
         this.field_185995_n = p_i46668_1_;
         this.field_185996_o = p_i46668_4_;
         this.field_185997_p = p_i46668_1_.func_72912_H().func_76067_t();
@@ -90,6 +100,17 @@
             this.field_186001_t = this.field_186000_s.field_177778_E ? Blocks.field_150353_l.func_176223_P() : Blocks.field_150355_j.func_176223_P();
             p_i46668_1_.func_181544_b(this.field_186000_s.field_177841_q);
         }
+
+        net.minecraftforge.event.terraingen.InitNoiseGensEvent.ContextOverworld ctx =
+                new net.minecraftforge.event.terraingen.InitNoiseGensEvent.ContextOverworld(field_185991_j, field_185992_k, field_185993_l, field_185994_m, field_185983_b, field_185984_c, field_185985_d);
+        ctx = net.minecraftforge.event.terraingen.TerrainGen.getModdedNoiseGenerators(p_i46668_1_, this.field_185990_i, ctx);
+        this.field_185991_j = ctx.getLPerlin1();
+        this.field_185992_k = ctx.getLPerlin2();
+        this.field_185993_l = ctx.getPerlin();
+        this.field_185994_m = ctx.getHeight();
+        this.field_185983_b = ctx.getScale();
+        this.field_185984_c = ctx.getDepth();
+        this.field_185985_d = ctx.getForest();
     }
 
     public void func_185976_a(int p_185976_1_, int p_185976_2_, ChunkPrimer p_185976_3_)
@@ -163,6 +184,7 @@
 
     public void func_185977_a(int p_185977_1_, int p_185977_2_, ChunkPrimer p_185977_3_, Biome[] p_185977_4_)
     {
+        if (!net.minecraftforge.event.ForgeEventFactory.onReplaceBiomeBlocks(this, p_185977_1_, p_185977_2_, p_185977_3_, this.field_185995_n)) return;
         double d0 = 0.03125D;
         this.field_186002_u = this.field_185994_m.func_151599_a(this.field_186002_u, (double)(p_185977_1_ * 16), (double)(p_185977_2_ * 16), 16, 16, 0.0625D, 0.0625D, 1.0D);
 
@@ -184,6 +206,7 @@
         this.field_185981_C = this.field_185995_n.func_72959_q().func_76933_b(this.field_185981_C, p_185932_1_ * 16, p_185932_2_ * 16, 16, 16);
         this.func_185977_a(p_185932_1_, p_185932_2_, chunkprimer, this.field_185981_C);
 
+        net.minecraftforge.event.ForgeEventFactory.onChunkGenerate(this, this.field_185995_n, p_185932_1_, p_185932_2_, chunkprimer);
         if (this.field_186000_s.field_177839_r)
         {
             this.field_186003_v.func_186125_a(this.field_185995_n, p_185932_1_, p_185932_2_, chunkprimer);
@@ -370,6 +393,8 @@
         boolean flag = false;
         ChunkPos chunkpos = new ChunkPos(p_185931_1_, p_185931_2_);
 
+        net.minecraftforge.event.ForgeEventFactory.onChunkPopulate(true, this, this.field_185995_n, this.field_185990_i, p_185931_1_, p_185931_2_, flag);
+
         if (this.field_185996_o)
         {
             if (this.field_186000_s.field_177829_w)
@@ -404,6 +429,7 @@
         }
 
         if (biome != Biomes.field_76769_d && biome != Biomes.field_76786_s && this.field_186000_s.field_177781_A && !flag && this.field_185990_i.nextInt(this.field_186000_s.field_177782_B) == 0)
+        if (net.minecraftforge.event.terraingen.TerrainGen.populate(this, this.field_185995_n, this.field_185990_i, p_185931_1_, p_185931_2_, flag, net.minecraftforge.event.terraingen.PopulateChunkEvent.Populate.EventType.LAKE))
         {
             int i1 = this.field_185990_i.nextInt(16) + 8;
             int j1 = this.field_185990_i.nextInt(256);
@@ -412,6 +438,7 @@
         }
 
         if (!flag && this.field_185990_i.nextInt(this.field_186000_s.field_177777_D / 10) == 0 && this.field_186000_s.field_177783_C)
+        if (net.minecraftforge.event.terraingen.TerrainGen.populate(this, this.field_185995_n, this.field_185990_i, p_185931_1_, p_185931_2_, flag, net.minecraftforge.event.terraingen.PopulateChunkEvent.Populate.EventType.LAVA))
         {
             int i2 = this.field_185990_i.nextInt(16) + 8;
             int l2 = this.field_185990_i.nextInt(this.field_185990_i.nextInt(248) + 8);
@@ -424,6 +451,7 @@
         }
 
         if (this.field_186000_s.field_177837_s)
+        if (net.minecraftforge.event.terraingen.TerrainGen.populate(this, this.field_185995_n, this.field_185990_i, p_185931_1_, p_185931_2_, flag, net.minecraftforge.event.terraingen.PopulateChunkEvent.Populate.EventType.DUNGEON))
         {
             for (int j2 = 0; j2 < this.field_186000_s.field_177835_t; ++j2)
             {
@@ -435,9 +463,12 @@
         }
 
         biome.func_180624_a(this.field_185995_n, this.field_185990_i, new BlockPos(i, 0, j));
+        if (net.minecraftforge.event.terraingen.TerrainGen.populate(this, this.field_185995_n, this.field_185990_i, p_185931_1_, p_185931_2_, flag, net.minecraftforge.event.terraingen.PopulateChunkEvent.Populate.EventType.ANIMALS))
         WorldEntitySpawner.func_77191_a(this.field_185995_n, biome, i + 8, j + 8, 16, 16, this.field_185990_i);
         blockpos = blockpos.func_177982_a(8, 0, 8);
 
+        if (net.minecraftforge.event.terraingen.TerrainGen.populate(this, this.field_185995_n, this.field_185990_i, p_185931_1_, p_185931_2_, flag, net.minecraftforge.event.terraingen.PopulateChunkEvent.Populate.EventType.ICE))
+        {
         for (int k2 = 0; k2 < 16; ++k2)
         {
             for (int j3 = 0; j3 < 16; ++j3)
@@ -456,7 +487,10 @@
                 }
             }
         }
+        }//Forge: End ICE
 
+        net.minecraftforge.event.ForgeEventFactory.onChunkPopulate(false, this, this.field_185995_n, this.field_185990_i, p_185931_1_, p_185931_2_, flag);
+
         BlockFalling.field_149832_M = false;
     }
 
@@ -476,6 +510,8 @@
     {
         Biome biome = this.field_185995_n.func_180494_b(p_177458_2_);
 
+        net.minecraftforge.event.terraingen.ChunkGeneratorEvent.GetPossibleCreatures event = new net.minecraftforge.event.terraingen.ChunkGeneratorEvent.GetPossibleCreatures(this, field_185995_n, p_177458_1_, p_177458_2_);
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return event.getPossibleCreatures();
         if (this.field_185996_o)
         {
             if (p_177458_1_ == EnumCreatureType.MONSTER && this.field_186007_z.func_175798_a(p_177458_2_))
@@ -494,6 +530,8 @@
 
     public boolean func_193414_a(World p_193414_1_, String p_193414_2_, BlockPos p_193414_3_)
     {
+        net.minecraftforge.event.terraingen.ChunkGeneratorEvent.StructureCheck event = new net.minecraftforge.event.terraingen.ChunkGeneratorEvent.StructureCheck(this, field_185995_n, p_193414_2_, p_193414_3_);
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return event.isInStructure();
         if (!this.field_185996_o)
         {
             return false;
@@ -527,6 +565,8 @@
     @Nullable
     public BlockPos func_180513_a(World p_180513_1_, String p_180513_2_, BlockPos p_180513_3_, boolean p_180513_4_)
     {
+        net.minecraftforge.event.terraingen.ChunkGeneratorEvent.FindStructure event = new net.minecraftforge.event.terraingen.ChunkGeneratorEvent.FindStructure(this, p_180513_1_, p_180513_2_, p_180513_3_, p_180513_4_);
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return event.getFoundPos();
         if (!this.field_185996_o)
         {
             return null;
@@ -559,6 +599,7 @@
 
     public void func_180514_a(Chunk p_180514_1_, int p_180514_2_, int p_180514_3_)
     {
+        net.minecraftforge.event.ForgeEventFactory.onChunkGenerate(this, this.field_185995_n, p_180514_2_, p_180514_3_, (ChunkPrimer) null);
         if (this.field_185996_o)
         {
             if (this.field_186000_s.field_177829_w)
