--- ../src-base/minecraft/net/minecraft/world/gen/ChunkProviderEnd.java
+++ ../src-work/minecraft/net/minecraft/world/gen/ChunkProviderEnd.java
@@ -32,6 +32,9 @@
     double[] field_73193_e;
     double[] field_73190_f;
     double[] field_73191_g;
+    // temporary variables used during event handling
+    private int chunkX = 0;
+    private int chunkZ = 0;
 
     public ChunkProviderEnd(World p_i2007_1_, long p_i2007_2_)
     {
@@ -42,6 +45,14 @@
         this.field_73199_l = new NoiseGeneratorOctaves(this.field_73204_i, 8);
         this.field_73196_a = new NoiseGeneratorOctaves(this.field_73204_i, 10);
         this.field_73194_b = new NoiseGeneratorOctaves(this.field_73204_i, 16);
+
+        NoiseGenerator[] noiseGens = {field_73201_j, field_73202_k, field_73199_l, field_73196_a, field_73194_b};
+        noiseGens = net.minecraftforge.event.terraingen.TerrainGen.getModdedNoiseGenerators(p_i2007_1_, this.field_73204_i, noiseGens);
+        this.field_73201_j = (NoiseGeneratorOctaves)noiseGens[0];
+        this.field_73202_k = (NoiseGeneratorOctaves)noiseGens[1];
+        this.field_73199_l = (NoiseGeneratorOctaves)noiseGens[2];
+        this.field_73196_a = (NoiseGeneratorOctaves)noiseGens[3];
+        this.field_73194_b = (NoiseGeneratorOctaves)noiseGens[4];
     }
 
     public void func_180520_a(int p_180520_1_, int p_180520_2_, ChunkPrimer p_180520_3_)
@@ -114,6 +125,9 @@
 
     public void func_180519_a(ChunkPrimer p_180519_1_)
     {
+        net.minecraftforge.event.terraingen.ChunkProviderEvent.ReplaceBiomeBlocks event = new net.minecraftforge.event.terraingen.ChunkProviderEvent.ReplaceBiomeBlocks(this, chunkX, chunkZ, p_180519_1_, this.field_73200_m);
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+        if (event.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) return;
         for (int i = 0; i < 16; ++i)
         {
             for (int j = 0; j < 16; ++j)
@@ -165,6 +179,7 @@
 
     public Chunk func_73154_d(int p_73154_1_, int p_73154_2_)
     {
+        chunkX = p_73154_1_; chunkZ = p_73154_2_;
         this.field_73204_i.setSeed((long)p_73154_1_ * 341873128712L + (long)p_73154_2_ * 132897987541L);
         ChunkPrimer chunkprimer = new ChunkPrimer();
         this.field_73198_o = this.field_73200_m.func_72959_q().func_76933_b(this.field_73198_o, p_73154_1_ * 16, p_73154_2_ * 16, 16, 16);
@@ -184,6 +199,10 @@
 
     private double[] func_73187_a(double[] p_73187_1_, int p_73187_2_, int p_73187_3_, int p_73187_4_, int p_73187_5_, int p_73187_6_, int p_73187_7_)
     {
+        net.minecraftforge.event.terraingen.ChunkProviderEvent.InitNoiseField event = new net.minecraftforge.event.terraingen.ChunkProviderEvent.InitNoiseField(this, p_73187_1_, p_73187_2_, p_73187_3_, p_73187_4_, p_73187_5_, p_73187_6_, p_73187_7_);
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+        if (event.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) return event.noisefield;
+
         if (p_73187_1_ == null)
         {
             p_73187_1_ = new double[p_73187_5_ * p_73187_6_ * p_73187_7_];
@@ -273,8 +292,10 @@
     public void func_73153_a(IChunkProvider p_73153_1_, int p_73153_2_, int p_73153_3_)
     {
         BlockFalling.field_149832_M = true;
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.terraingen.PopulateChunkEvent.Pre(p_73153_1_, field_73200_m, field_73200_m.field_73012_v, p_73153_2_, p_73153_3_, false));
         BlockPos blockpos = new BlockPos(p_73153_2_ * 16, 0, p_73153_3_ * 16);
         this.field_73200_m.func_180494_b(blockpos.func_177982_a(16, 0, 16)).func_180624_a(this.field_73200_m, this.field_73200_m.field_73012_v, blockpos);
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.terraingen.PopulateChunkEvent.Post(p_73153_1_, field_73200_m, field_73200_m.field_73012_v, p_73153_2_, p_73153_3_, false));
         BlockFalling.field_149832_M = false;
     }
 
