--- a/net/minecraft/world/gen/ChunkGeneratorOverworld.java
+++ b/net/minecraft/world/gen/ChunkGeneratorOverworld.java
@@ -28,8 +28,8 @@
    private NoiseGeneratorOctaves field_185993_l;
    private NoiseGeneratorPerlin field_185994_m;
    private final OverworldGenSettings field_186000_s;
-   private final NoiseGeneratorOctaves field_185983_b;
-   private final NoiseGeneratorOctaves field_185984_c;
+   private NoiseGeneratorOctaves field_185983_b;
+   private NoiseGeneratorOctaves field_185984_c;
    private final WorldType field_185997_p;
    private final float[] field_185999_r;
    private final PhantomSpawner field_203230_r = new PhantomSpawner();
@@ -58,6 +58,16 @@
       this.field_186000_s = p_i48957_3_;
       this.field_205475_r = this.field_186000_s.func_205532_l();
       this.field_205476_s = this.field_186000_s.func_205533_m();
+      
+      net.minecraftforge.event.terraingen.InitNoiseGensEvent.ContextOverworld ctx =
+              new net.minecraftforge.event.terraingen.InitNoiseGensEvent.ContextOverworld(field_185991_j, field_185992_k, field_185993_l, field_185994_m, field_185983_b, field_185984_c);
+      ctx = net.minecraftforge.event.terraingen.TerrainGen.getModdedNoiseGenerators(p_i48957_1_, sharedseedrandom, ctx);
+      this.field_185991_j = ctx.getLPerlin1();
+      this.field_185992_k = ctx.getLPerlin2();
+      this.field_185993_l = ctx.getPerlin();
+      this.field_185994_m = ctx.getHeight();
+      this.field_185983_b = ctx.getScale();
+      this.field_185984_c = ctx.getDepth();
    }
 
    public void func_202088_a(IChunk p_202088_1_) {
