--- a/net/minecraft/loot/LootContext.java
+++ b/net/minecraft/loot/LootContext.java
@@ -94,17 +_,47 @@
       return this.field_186499_b;
    }
 
+   // ============================== FORGE START ==============================
+   public int getLootingModifier() {
+      return net.minecraftforge.common.ForgeHooks.getLootingLevel(func_216031_c(LootParameters.field_216281_a), func_216031_c(LootParameters.field_216284_d), func_216031_c(LootParameters.field_216283_c));
+   }
+
+   private ResourceLocation queriedLootTableId;
+
+   private LootContext(Random rand, float luckIn, ServerWorld worldIn, Function<ResourceLocation, LootTable> lootTableManagerIn, Function<ResourceLocation, ILootCondition> p_i225885_5_, Map<LootParameter<?>, Object> parametersIn, Map<ResourceLocation, LootContext.IDynamicDropProvider> conditionsIn, ResourceLocation queriedLootTableId) {
+      this(rand, luckIn, worldIn, lootTableManagerIn, p_i225885_5_, parametersIn, conditionsIn);
+      if (queriedLootTableId != null) this.queriedLootTableId = queriedLootTableId;
+   }
+
+   public void setQueriedLootTableId(ResourceLocation queriedLootTableId) {
+      if (this.queriedLootTableId == null && queriedLootTableId != null) this.queriedLootTableId = queriedLootTableId;
+   }
+   public ResourceLocation getQueriedLootTableId() {
+      return this.queriedLootTableId == null? net.minecraftforge.common.loot.LootTableIdCondition.UNKNOWN_LOOT_TABLE : this.queriedLootTableId;
+   }
+   // =============================== FORGE END ===============================
+
    public static class Builder {
       private final ServerWorld field_186474_a;
       private final Map<LootParameter<?>, Object> field_216025_b = Maps.newIdentityHashMap();
       private final Map<ResourceLocation, LootContext.IDynamicDropProvider> field_216026_c = Maps.newHashMap();
       private Random field_216027_d;
       private float field_186475_b;
+      private ResourceLocation queriedLootTableId; // Forge: correctly pass around loot table ID with copy constructor
 
       public Builder(ServerWorld p_i46993_1_) {
          this.field_186474_a = p_i46993_1_;
       }
 
+      public Builder(LootContext context) {
+         this.field_186474_a = context.field_186499_b;
+         this.field_216025_b.putAll(context.field_216036_f);
+         this.field_216026_c.putAll(context.field_216037_g);
+         this.field_216027_d = context.field_216035_a;
+         this.field_186475_b = context.field_186498_a;
+         this.queriedLootTableId = context.queriedLootTableId;
+      }
+
       public LootContext.Builder func_216023_a(Random p_216023_1_) {
          this.field_216027_d = p_216023_1_;
          return this;
@@ -177,7 +_,7 @@
 
       public LootContext func_216022_a(LootParameterSet p_216022_1_) {
          Set<LootParameter<?>> set = Sets.difference(this.field_216025_b.keySet(), p_216022_1_.func_216276_b());
-         if (!set.isEmpty()) {
+         if (false && !set.isEmpty()) { // Forge: Allow mods to pass custom loot parameters (not part of the vanilla loot table) to the loot context.
             throw new IllegalArgumentException("Parameters not allowed in this parameter set: " + set);
          } else {
             Set<LootParameter<?>> set1 = Sets.difference(p_216022_1_.func_216277_a(), this.field_216025_b.keySet());
@@ -190,7 +_,7 @@
                }
 
                MinecraftServer minecraftserver = this.field_186474_a.func_73046_m();
-               return new LootContext(random, this.field_186475_b, this.field_186474_a, minecraftserver.func_200249_aQ()::func_186521_a, minecraftserver.func_229736_aP_()::func_227517_a_, this.field_216025_b, this.field_216026_c);
+               return new LootContext(random, this.field_186475_b, this.field_186474_a, minecraftserver.func_200249_aQ()::func_186521_a, minecraftserver.func_229736_aP_()::func_227517_a_, this.field_216025_b, this.field_216026_c, this.queriedLootTableId);
             }
          }
       }
