--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -236,6 +_,7 @@
    }
 
    public void m_37774_() {
+      if (!net.minecraftforge.event.ForgeEventFactory.shouldRaidStop(this)) return;
       this.f_37680_ = false;
       this.f_37682_.m_7706_();
       this.f_37687_ = Raid.RaidStatus.STOPPED;
@@ -493,37 +_,53 @@
       this.f_37678_ = 0.0F;
       DifficultyInstance difficultyinstance = this.f_37675_.m_6436_(p_37756_);
       boolean flag1 = this.m_37702_();
-
-      for(Raid.RaiderType raid$raidertype : Raid.RaiderType.f_37813_) {
-         int j = this.m_37730_(raid$raidertype, i, flag1) + this.m_37734_(raid$raidertype, this.f_37685_, i, difficultyinstance, flag1);
-         int k = 0;
-
-         for(int l = 0; l < j; ++l) {
-            Raider raider = raid$raidertype.f_37814_.m_20615_(this.f_37675_);
-            if (!flag && raider.m_7490_()) {
-               raider.m_33075_(true);
-               this.m_37710_(i, raider);
-               flag = true;
-            }
-
-            this.m_37713_(i, raider, p_37756_, false);
-            if (raid$raidertype.f_37814_ == EntityType.f_20518_) {
-               Raider raider1 = null;
-               if (i == this.m_37724_(Difficulty.NORMAL)) {
-                  raider1 = EntityType.f_20513_.m_20615_(this.f_37675_);
-               } else if (i >= this.m_37724_(Difficulty.HARD)) {
-                  if (k == 0) {
-                     raider1 = EntityType.f_20568_.m_20615_(this.f_37675_);
-                  } else {
-                     raider1 = EntityType.f_20493_.m_20615_(this.f_37675_);
-                  }
+      net.minecraftforge.event.world.RaidEvent.SpawnGroup event = net.minecraftforge.event.ForgeEventFactory.spawnGroupForRaid(this, p_37756_, flag1);
+      if (!event.isCanceled())
+      {
+         p_37756_ = event.getPos();
+         flag1 = event.shouldSpawnBonusGroup();
+         for (Raid.RaiderType raid$raidertype : Raid.RaiderType.f_37813_)
+         {
+            int j = this.m_37730_(raid$raidertype, i, flag1) + this.m_37734_(raid$raidertype, this.f_37685_, i, difficultyinstance, flag1);
+            int k = 0;
+
+            for (int l = 0; l < j; ++l)
+            {
+               Raider raider = raid$raidertype.f_37814_.m_20615_(this.f_37675_);
+               if (!flag && raider.m_7490_())
+               {
+                  raider.m_33075_(true);
+                  this.m_37710_(i, raider);
+                  flag = true;
                }
 
-               ++k;
-               if (raider1 != null) {
-                  this.m_37713_(i, raider1, p_37756_, false);
-                  raider1.m_20035_(p_37756_, 0.0F, 0.0F);
-                  raider1.m_20329_(raider);
+               this.m_37713_(i, raider, p_37756_, false);
+               if (raid$raidertype.f_37814_ == EntityType.f_20518_)
+               {
+                  Raider raider1 = null;
+                  if (i == this.m_37724_(Difficulty.NORMAL))
+                  {
+                     raider1 = EntityType.f_20513_.m_20615_(this.f_37675_);
+                  }
+                  else if (i >= this.m_37724_(Difficulty.HARD))
+                  {
+                     if (k == 0)
+                     {
+                        raider1 = EntityType.f_20568_.m_20615_(this.f_37675_);
+                     }
+                     else
+                     {
+                        raider1 = EntityType.f_20493_.m_20615_(this.f_37675_);
+                     }
+                  }
+
+                  ++k;
+                  if (raider1 != null)
+                  {
+                     this.m_37713_(i, raider1, p_37756_, false);
+                     raider1.m_20035_(p_37756_, 0.0F, 0.0F);
+                     raider1.m_20329_(raider);
+                  }
                }
             }
          }
@@ -786,6 +_,16 @@
       this.f_37672_.add(p_37727_.m_142081_());
    }
 
+   public ServerBossEvent getRaidEvent()
+   {
+      return f_37682_;
+   }
+
+   public Map<Integer, Set<Raider>> getGroupRaiderMap()
+   {
+      return new java.util.HashMap<>(f_37671_);
+   }
+
    static enum RaidStatus {
       ONGOING,
       VICTORY,
@@ -809,7 +_,7 @@
       }
    }
 
-   static enum RaiderType {
+   public static enum RaiderType implements net.minecraftforge.common.IExtensibleEnum {
       VINDICATOR(EntityType.f_20493_, new int[]{0, 0, 2, 0, 1, 4, 2, 5}),
       EVOKER(EntityType.f_20568_, new int[]{0, 0, 0, 0, 0, 1, 1, 2}),
       PILLAGER(EntityType.f_20513_, new int[]{0, 4, 3, 3, 4, 4, 4, 2}),
@@ -823,6 +_,20 @@
       private RaiderType(EntityType<? extends Raider> p_37821_, int[] p_37822_) {
          this.f_37814_ = p_37821_;
          this.f_37815_ = p_37822_;
+      }
+      
+      /**
+       * The waveCountsIn integer decides how many entities of the EntityType defined in typeIn will spawn in each wave.
+       * For example, one ravager will always spawn in wave 3.
+       */
+      public static RaiderType create(String name, EntityType<? extends Raider> typeIn, int[] waveCountsIn) {
+         throw new IllegalStateException("Enum not extended");
+      }
+      
+      @Override
+      @Deprecated
+      public void init() {
+         f_37813_ = values();
       }
    }
 }
