--- ../src-base/minecraft/net/minecraft/command/server/CommandSetBlock.java
+++ ../src-work/minecraft/net/minecraft/command/server/CommandSetBlock.java
@@ -52,6 +52,8 @@
                 i = func_175764_a(p_71515_2_[4], 0, 15);
             }
 
+            IBlockState state = block.func_176203_a(i);
+
             World world = p_71515_1_.func_130014_f_();
 
             if (!world.func_175667_e(blockpos))
@@ -63,7 +65,7 @@
                 NBTTagCompound nbttagcompound = new NBTTagCompound();
                 boolean flag = false;
 
-                if (p_71515_2_.length >= 7 && block.func_149716_u())
+                if (p_71515_2_.length >= 7 && block.hasTileEntity(state))
                 {
                     String s = func_147178_a(p_71515_1_, p_71515_2_, 6).func_150260_c();
 
