--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -55,6 +_,13 @@
                CauldronInteraction.m_175649_();
                ArgumentTypes.m_121586_();
                Registry.m_206101_();
+               net.minecraftforge.registries.GameData.vanillaSnapshot();
+               // Forge: Hacky fix to ensure that NetworkConstants is loaded before mods are constructed.
+               // Many older mods use network internals that shouldn't be used, yet are exposed so they get used anyways.
+               // This can cause class-loading issues with ForgeMod loading NetworkConstants and HandshakeResolver.
+               // To ensure that doesn't happen, we load it here and now. This is not an issue in 1.20.2 and newer.
+               net.minecraftforge.network.NetworkHooks.init();
+               if (false) // skip redirectOutputToLog, Forge already redirects stdout and stderr output to log so that they print with more context
                m_135890_();
             }
          }
@@ -128,7 +_,6 @@
          m_197757_();
       }
 
-      DefaultAttributes.m_22296_();
    }
 
    private static void m_197757_() {
