--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -85,6 +_,17 @@
         OptionSpec<Void> jfrProfilingOption = parser.accepts("jfrProfile");
         OptionSpec<Path> pidFile = parser.accepts("pidFile").withRequiredArg().withValuesConvertedBy(new PathConverter());
         OptionSpec<String> nonOptions = parser.nonOptions();
+        parser.accepts("allowUpdates").withRequiredArg().ofType(Boolean.class).defaultsTo(Boolean.TRUE); // Forge: allow mod updates to proceed
+        parser.accepts("gameDir").withRequiredArg().ofType(File.class).defaultsTo(new File(".")); //Forge: Consume this argument, we use it in the launcher, and the client side.
+        final OptionSpec<net.minecraft.core.BlockPos> spawnPosOpt;
+        OptionSpec<Void> uniqueWorld = null;
+        boolean gametestEnabled = Boolean.getBoolean("forge.gameTestServer");
+        if (gametestEnabled) {
+            spawnPosOpt = parser.accepts("spawnPos").withRequiredArg().withValuesConvertedBy(new net.minecraftforge.gametest.BlockPosValueConverter()).defaultsTo(new net.minecraft.core.BlockPos(0, 60, 0));
+            uniqueWorld = parser.accepts("uniqueWorld");
+        } else {
+            spawnPosOpt = null;
+        }
 
         try {
             OptionSet options = parser.parse(args);
@@ -93,6 +_,14 @@
                 return;
             }
 
+            Path eulaFile = Paths.get("eula.txt");
+            Eula eula = new Eula(eulaFile);
+
+            if (!eula.hasAgreedToEULA()) {
+               LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+               return;
+            }
+
             Path pidFilePath = options.valueOf(pidFile);
             if (pidFilePath != null) {
                 writePidFile(pidFilePath);
@@ -107,26 +_,30 @@
             Bootstrap.validate();
             Util.startTimerHackThread();
             Path settingsFile = Paths.get("server.properties");
+            if (!options.has(initSettings)) {
+                // Load mods before we load almost anything else anymore. Single spot now. Only loads if they haven't passed the initserver param
+                net.minecraftforge.server.loading.ServerModLoader.load();
+            }
             DedicatedServerSettings settings = new DedicatedServerSettings(settingsFile);
             settings.forceSave();
             RegionFileVersion.configure(settings.getProperties().regionFileComression);
-            Path eulaFile = Paths.get("eula.txt");
-            Eula eula = new Eula(eulaFile);
             if (options.has(initSettings)) {
                 LOGGER.info("Initialized '{}' and '{}'", settingsFile.toAbsolutePath(), eulaFile.toAbsolutePath());
                 return;
             }
 
-            if (!eula.hasAgreedToEULA()) {
-                LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
-                return;
-            }
-
             File universePath = new File(options.valueOf(universe));
             Services services = Services.create(new YggdrasilAuthenticationService(Proxy.NO_PROXY), universePath);
             NotificationManager notificationManager = new NotificationManager();
             ManagementServer jsonRpcServer = JsonRpc.create(settings, notificationManager);
             String levelName = Optional.ofNullable(options.valueOf(worldName)).orElse(settings.getProperties().levelName);
+            // Forge: make each gametest use a timestamped world name
+            if (uniqueWorld != null && options.has(uniqueWorld))
+                levelName = "gametest_world\\" + java.time.format.DateTimeFormatter.ofPattern("yyyy-MM-dd_HH.mm.ss.SSS").format(java.time.LocalDateTime.now());
+            if (levelName == null || levelName.isEmpty() || new File(universePath, levelName).getAbsolutePath().equals(new File(levelName).getAbsolutePath())) {
+                LOGGER.error("Invalid world directory specified, must not be null, empty or the same directory as your universe! " + levelName);
+                return;
+            }
             LevelStorageSource levelStorageSource = LevelStorageSource.createDefault(universePath.toPath());
             LevelStorageSource.LevelStorageAccess access = levelStorageSource.validateAndCreateAccess(levelName);
             Dynamic<?> levelDataTag;
@@ -162,6 +_,10 @@
 
             PackRepository packRepository = ServerPacksSource.createPackRepository(access);
 
+            if (levelDataTag != null) {
+                net.minecraftforge.common.ForgeHooks.readAdditionalLevelSaveData(access, access.getLevelDirectory());
+            }
+
             WorldStem worldStem;
             try {
                 WorldLoader.InitConfig worldLoadConfig = loadOrCreateConfig(settings.getProperties(), levelDataTag, safeModeEnabled, packRepository);
@@ -233,6 +_,7 @@
                 @Override
                 public void run() {
                     dedicatedServer.halt(true);
+                    org.apache.logging.log4j.LogManager.shutdown(); // we're manually managing the logging shutdown on the server. Make sure we do it here at the end.
                 }
             };
             shutdownThread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
@@ -268,6 +_,16 @@
             worldOptions = bonusChest ? properties.worldOptions.withBonusChest(true) : properties.worldOptions;
             dimensions = properties.createDimensions(context.datapackWorldgen());
         }
+
+        //Forge: Do a write-read-cycle to inject modded dimensions on first start of a dedicated server into its generated world dimensions list.
+        var dynamicops = net.minecraft.resources.RegistryOps.create(net.minecraft.nbt.NbtOps.INSTANCE, context.datapackWorldgen());
+        dimensions = WorldDimensions.CODEC.encoder()
+            .encodeStart(dynamicops, dimensions)
+            .flatMap((writtenPayloadWithModdedDimensions) ->
+                WorldDimensions.CODEC.decoder().parse(dynamicops, writtenPayloadWithModdedDimensions)
+            )
+            .resultOrPartial(LOGGER::error)
+            .orElse(dimensions);
 
         WorldDimensions.Complete finalDimensions = dimensions.bake(datapackDimensions);
         Lifecycle lifecycle = finalDimensions.lifecycle().add(context.datapackWorldgen().allRegistriesLifecycle());
