--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -296,7 +_,7 @@
 
     public static <S extends MinecraftServer> S spin(final Function<Thread, S> factory) {
         AtomicReference<S> serverReference = new AtomicReference<>();
-        Thread thread = new Thread(() -> serverReference.get().runServer(), "Server thread");
+        Thread thread = new Thread(net.minecraftforge.fml.util.thread.SidedThreadGroups.SERVER, () -> serverReference.get().runServer(), "Server thread");
         thread.setUncaughtExceptionHandler((t, e) -> LOGGER.error("Uncaught exception in server thread", e));
         if (Runtime.getRuntime().availableProcessors() > 4) {
             thread.setPriority(8);
@@ -436,6 +_,7 @@
         this.scoreboard.load(this.savedDataStorage.computeIfAbsent(ScoreboardSaveData.TYPE).getData());
         this.commandStorage = new CommandStorage(this.savedDataStorage);
         this.stopwatches = this.savedDataStorage.computeIfAbsent(Stopwatches.TYPE);
+        net.minecraftforge.event.ForgeEventFactory.onLevelLoad(levels.get(Level.OVERWORLD));
         if (!levelData.isInitialized()) {
             try {
                 setInitialSpawn(overworld, levelData, worldOptions.generateBonusChest(), isDebug, this.levelLoadListener);
@@ -470,6 +_,7 @@
                     this, this.executor, this.storageSource, derivedLevelData, dimension, entry.getValue(), isDebug, biomeZoomSeed, ImmutableList.of(), false
                 );
                 this.levels.put(dimension, level);
+                net.minecraftforge.event.ForgeEventFactory.onLevelLoad(level);
             } else {
                 level = overworld;
             }
@@ -492,6 +_,7 @@
             levelData.setSpawn(LevelData.RespawnData.of(level.dimension(), BlockPos.ZERO.above(80), 0.0F, 0.0F));
         } else {
             ServerChunkCache chunkSource = level.getChunkSource();
+            if (net.minecraftforge.event.ForgeEventFactory.onCreateWorldSpawn(level, levelData)) return;
             ChunkPos spawnChunk = ChunkPos.containing(chunkSource.randomState().sampler().findSpawnPosition());
             levelLoadListener.start(LevelLoadListener.Stage.PREPARE_GLOBAL_SPAWN, 0);
             levelLoadListener.updateFocus(level.dimension(), spawnChunk);
@@ -678,6 +_,7 @@
         for (ServerLevel level : this.getAllLevels()) {
             if (level != null) {
                 try {
+                    net.minecraftforge.event.ForgeEventFactory.onLevelUnload(level);
                     level.close();
                 } catch (IOException e) {
                     LOGGER.error("Exception closing the level", e);
@@ -725,9 +_,11 @@
                 throw new IllegalStateException("Failed to initialize server");
             }
 
+            net.minecraftforge.server.ServerLifecycleHooks.handleServerStarted(this);
             this.nextTickTimeNanos = Util.getNanos();
             this.statusIcon = this.loadStatusIcon().orElse(null);
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
 
             while (this.running) {
                 long thisTickNanos;
@@ -781,6 +_,8 @@
                 this.isReady = true;
                 JvmProfiler.INSTANCE.onServerTick(this.smoothedTickTimeMillis);
             }
+            net.minecraftforge.server.ServerLifecycleHooks.handleServerStopping(this);
+            net.minecraftforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
         } catch (Throwable t) {
             LOGGER.error("Encountered an unexpected exception", t);
             CrashReport report = constructOrExtractCrashReport(t);
@@ -792,6 +_,7 @@
                 LOGGER.error("We were unable to save this crash report to disk.");
             }
 
+            net.minecraftforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
             this.onServerCrash(report);
         } finally {
             try {
@@ -800,6 +_,7 @@
             } catch (Throwable t) {
                 LOGGER.error("Exception stopping the server", t);
             } finally {
+                net.minecraftforge.server.ServerLifecycleHooks.handleServerStopped(this);
                 this.onServerExit();
             }
         }
@@ -980,12 +_,14 @@
             }
         }
 
+        net.minecraftforge.event.ForgeEventFactory.onPreServerTick(haveTime, this);
         this.tickCount++;
         this.tickRateManager.tick();
         this.tickChildren(haveTime);
         if (nano - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) {
             this.lastServerStatus = nano;
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
         }
 
         this.ticksUntilAutosave--;
@@ -1003,6 +_,7 @@
         this.smoothedTickTimeMillis = this.smoothedTickTimeMillis * 0.8F + (float)tickTime / (float)TimeUtil.NANOSECONDS_PER_MILLISECOND * 0.19999999F;
         this.logTickMethodTime(nano);
         profiler.pop();
+        net.minecraftforge.event.ForgeEventFactory.onPostServerTick(haveTime, this);
     }
 
     protected void processPacketsAndTick(final boolean sprinting) {
@@ -1064,7 +_,8 @@
             Optional.of(players),
             Optional.of(ServerStatus.Version.current()),
             Optional.ofNullable(this.statusIcon),
-            this.enforceSecureProfile()
+            this.enforceSecureProfile(),
+            Optional.of(new net.minecraftforge.network.ServerStatusPing())
         );
     }
 
@@ -1109,9 +_,11 @@
         profiler.push("levels");
         this.updateEffectiveRespawnData();
 
-        for (ServerLevel level : this.getAllLevels()) {
+        for (ServerLevel level : this.getWorldArray()) {
+            long tickStart = Util.getNanos();
             profiler.push(() -> level + " " + level.dimension().identifier());
             profiler.push("tick");
+            net.minecraftforge.event.ForgeEventFactory.onPreLevelTick(level, haveTime);
 
             try {
                 level.tick(haveTime);
@@ -1120,9 +_,11 @@
                 level.fillReportDetails(report);
                 throw new ReportedException(report);
             }
+            net.minecraftforge.event.ForgeEventFactory.onPostLevelTick(level, haveTime);
 
             profiler.pop();
             profiler.pop();
+            perWorldTickTimes.computeIfAbsent(level.dimension(), k -> new long[100])[this.tickCount % 100] = Util.getNanos() - tickStart;
         }
 
         profiler.popPush("connection");
@@ -1131,7 +_,7 @@
         this.playerList.tick();
         profiler.popPush("debugSubscribers");
         this.debugSubscribers.tick();
-        if (this.tickRateManager.runsNormally()) {
+        if (net.minecraftforge.gametest.ForgeGameTestHooks.isGametestEnabled() && this.tickRateManager.runsNormally()) {
             profiler.popPush("gameTests");
             GameTestTicker.SINGLETON.tick();
         }
@@ -1217,7 +_,7 @@
     }
 
     public String getServerModName() {
-        return "vanilla";
+        return net.minecraftforge.internal.BrandingControl.getBranding();
     }
 
     public ServerClockManager clockManager() {
@@ -1561,6 +_,7 @@
                 this.functionManager.replaceLibrary(this.resources.managers.getFunctionLibrary());
                 this.structureTemplateManager.onResourceManagerReload(this.resources.resourceManager);
                 this.fuelValues = FuelValues.vanillaBurnTimes(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+                this.getPlayerList().getPlayers().forEach(this.getPlayerList()::sendPlayerPermissionLevel); //Forge: Fix newly added/modified commands not being sent to the client when commands reload.
             }, this);
         if (this.isSameThread()) {
             this.managedBlock(result::isDone);
@@ -1572,12 +_,15 @@
     public static WorldDataConfiguration configurePackRepository(
         final PackRepository packRepository, final WorldDataConfiguration initialDataConfig, final boolean initMode, final boolean safeMode
     ) {
+        net.minecraftforge.resource.ResourcePackLoader.loadResourcePacks(packRepository, false);
         DataPackConfig dataPackConfig = initialDataConfig.dataPacks();
         FeatureFlagSet forcedFeatures = initMode ? FeatureFlagSet.of() : initialDataConfig.enabledFeatures();
         FeatureFlagSet allowedFeatures = initMode ? FeatureFlags.REGISTRY.allFlags() : initialDataConfig.enabledFeatures();
         packRepository.reload();
+        DataPackConfig.DEFAULT.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
+        dataPackConfig.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
         if (safeMode) {
-            return configureRepositoryWithSelection(packRepository, List.of("vanilla"), forcedFeatures, false);
+            return configureRepositoryWithSelection(packRepository, net.minecraftforge.common.ForgeHooks.getModPacksWithVanilla(), forcedFeatures, false);
         }
 
         Set<String> selected = Sets.newLinkedHashSet();
@@ -2226,6 +_,48 @@
 
     public ServerLinks serverLinks() {
         return ServerLinks.EMPTY;
+    }
+
+    private static final com.google.gson.Gson GSON = new com.google.gson.Gson();
+    private String cachedServerStatus; // FORGE: cache the server status json in case a client spams requests
+    private void resetStatusCache(ServerStatus status) {
+       this.cachedServerStatus = GSON.toJson(ServerStatus.CODEC.encodeStart(com.mojang.serialization.JsonOps.INSTANCE, status).result().orElseThrow());
+    }
+
+    public String getStatusJson() {
+       return cachedServerStatus;
+    }
+
+    private Map<ResourceKey<Level>, long[]> perWorldTickTimes = Maps.newIdentityHashMap();
+    @Nullable
+    public long[] getTickTime(ResourceKey<Level> dim) {
+        return perWorldTickTimes.get(dim);
+    }
+
+    @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+    public synchronized Map<ResourceKey<Level>, ServerLevel> forgeGetWorldMap() {
+        return this.levels;
+    }
+
+    private int worldArrayMarker = 0;
+    private int worldArrayLast = -1;
+    private ServerLevel[] worldArray;
+
+    @Deprecated //Forge Internal use Only, use to protect against concurrent modifications in the world tick loop.
+    public synchronized void markWorldsDirty() {
+        worldArrayMarker++;
+    }
+
+    private ServerLevel[] getWorldArray() {
+        if (worldArrayMarker == worldArrayLast && worldArray != null)
+           return worldArray;
+        worldArray = this.levels.values().stream().toArray(ServerLevel[]::new);
+        worldArrayLast = worldArrayMarker;
+        return worldArray;
+    }
+
+    public MinecraftServer.ReloadableResources getServerResources() {
+        return resources;
     }
 
     protected int pauseWhenEmptySeconds() {
