--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -13,7 +13,6 @@
 import com.mojang.authlib.GameProfileRepository;
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
-import com.mojang.datafixers.DataFixer;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufOutputStream;
 import io.netty.buffer.Unpooled;
@@ -88,7 +87,6 @@
 import net.minecraft.util.ITickable;
 import net.minecraft.util.Util;
 import net.minecraft.util.WorldOptimizer;
-import net.minecraft.util.datafix.DataFixesManager;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.math.MathHelper;
@@ -133,7 +131,6 @@
    private final NetworkSystem field_147144_o;
    private final ServerStatusResponse field_147147_p = new ServerStatusResponse();
    private final Random field_147146_q = new Random();
-   private final DataFixer field_184112_s;
    private String field_71320_r;
    private int field_71319_s = -1;
    private final Map<DimensionType, WorldServer> field_71305_c = Maps.newIdentityHashMap();
@@ -194,17 +191,16 @@
    private boolean field_212205_ao;
    private float field_211152_ao;
 
-   public MinecraftServer(@Nullable File p_i49697_1_, Proxy p_i49697_2_, DataFixer p_i49697_3_, Commands p_i49697_4_, YggdrasilAuthenticationService p_i49697_5_, MinecraftSessionService p_i49697_6_, GameProfileRepository p_i49697_7_, PlayerProfileCache p_i49697_8_) {
-      this.field_110456_c = p_i49697_2_;
-      this.field_195579_af = p_i49697_4_;
-      this.field_152364_T = p_i49697_5_;
-      this.field_147143_S = p_i49697_6_;
-      this.field_152365_W = p_i49697_7_;
-      this.field_152366_X = p_i49697_8_;
-      this.field_71308_o = p_i49697_1_;
-      this.field_147144_o = p_i49697_1_ == null ? null : new NetworkSystem(this);
-      this.field_71310_m = p_i49697_1_ == null ? null : new AnvilSaveConverter(p_i49697_1_.toPath(), p_i49697_1_.toPath().resolve("../backups"), p_i49697_3_);
-      this.field_184112_s = p_i49697_3_;
+   public MinecraftServer(@Nullable File anvilFileIn, Proxy serverProxyIn, Commands commandManagerIn, YggdrasilAuthenticationService authServiceIn, MinecraftSessionService sessionServiceIn, GameProfileRepository profileRepoIn, PlayerProfileCache profileCacheIn) {
+      this.field_110456_c = serverProxyIn;
+      this.field_195579_af = commandManagerIn;
+      this.field_152364_T = authServiceIn;
+      this.field_147143_S = sessionServiceIn;
+      this.field_152365_W = profileRepoIn;
+      this.field_152366_X = profileCacheIn;
+      this.field_71308_o = anvilFileIn;
+      this.field_147144_o = anvilFileIn == null ? null : new NetworkSystem(this);
+      this.field_71310_m = anvilFileIn == null ? null : new AnvilSaveConverter(anvilFileIn.toPath(), anvilFileIn.toPath().resolve("../backups"));
       this.field_195576_ac.func_199006_a(this.field_199736_ah);
       this.field_195576_ac.func_199006_a(this.field_199530_ag);
       this.field_195576_ac.func_199006_a(this.field_200256_aj);
@@ -335,20 +331,19 @@
          worldserver.func_72912_H().func_76060_a(this.func_71265_f());
       }
 
-      WorldServerMulti worldservermulti = (new WorldServerMulti(this, p_212369_1_, DimensionType.NETHER, worldserver, this.field_71304_b)).func_212251_i__();
-      this.field_71305_c.put(DimensionType.NETHER, worldservermulti);
-      worldservermulti.func_72954_a(new ServerWorldEventHandler(this, worldservermulti));
-      if (!this.func_71264_H()) {
-         worldservermulti.func_72912_H().func_76060_a(this.func_71265_f());
+      for (DimensionType dim : DimensionType.func_212681_b()) {
+         WorldServer world = worldserver;
+         if (dim != DimensionType.OVERWORLD) {
+            world = (new WorldServerMulti(this, p_212369_1_, dim, worldserver, this.field_71304_b)).func_212251_i__();
+            this.field_71305_c.put(dim, world);
+            world.func_72954_a(new ServerWorldEventHandler(this, world));
+            if (!this.func_71264_H()) {
+               world.func_72912_H().func_76060_a(func_71265_f());
+            }
+         }
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Load(world));
       }
 
-      WorldServerMulti worldservermulti1 = (new WorldServerMulti(this, p_212369_1_, DimensionType.THE_END, worldserver, this.field_71304_b)).func_212251_i__();
-      this.field_71305_c.put(DimensionType.THE_END, worldservermulti1);
-      worldservermulti1.func_72954_a(new ServerWorldEventHandler(this, worldservermulti1));
-      if (!this.func_71264_H()) {
-         worldservermulti1.func_72912_H().func_76060_a(this.func_71265_f());
-      }
-
       this.func_184103_al().func_212504_a(worldserver);
       if (p_212369_3_.func_201357_P() != null) {
          this.func_201300_aS().func_201381_a(p_212369_3_.func_201357_P());
@@ -514,6 +509,7 @@
 
       for(WorldServer worldserver1 : this.func_212370_w()) {
          if (worldserver1 != null) {
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Unload(worldserver1));
             worldserver1.close();
          }
       }
@@ -547,6 +543,7 @@
    public void run() {
       try {
          if (this.func_71197_b()) {
+            net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStarted(this);
             this.field_211151_aa = Util.func_211177_b();
             this.field_147147_p.func_151315_a(new TextComponentString(this.field_71286_C));
             this.field_147147_p.func_151321_a(new ServerStatusResponse.Version("1.13.2", 404));
@@ -570,7 +567,10 @@
 
                this.field_71296_Q = true;
             }
+            net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStopping(this);
+            net.minecraftforge.fml.server.ServerLifecycleHooks.expectServerStopped(); // has to come before finalTick to avoid race conditions
          } else {
+            net.minecraftforge.fml.server.ServerLifecycleHooks.expectServerStopped(); // has to come before finalTick to avoid race conditions
             this.func_71228_a((CrashReport)null);
          }
       } catch (Throwable throwable1) {
@@ -589,6 +589,7 @@
             field_147145_h.error("We were unable to save this crash report to disk.");
          }
 
+         net.minecraftforge.fml.server.ServerLifecycleHooks.expectServerStopped(); // has to come before finalTick to avoid race conditions
          this.func_71228_a(crashreport);
       } finally {
          try {
@@ -597,6 +598,7 @@
          } catch (Throwable throwable) {
             field_147145_h.error("Exception stopping the server", throwable);
          } finally {
+            net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStopped(this);
             this.func_71240_o();
          }
 
@@ -652,6 +654,7 @@
 
    public void func_71217_p(BooleanSupplier p_71217_1_) {
       long i = Util.func_211178_c();
+      net.minecraftforge.fml.hooks.BasicEventHooks.onPreServerTick();
       ++this.field_71315_w;
       if (this.field_71295_T) {
          this.field_71295_T = false;
@@ -672,6 +675,7 @@
 
          Collections.shuffle(Arrays.asList(agameprofile));
          this.field_147147_p.func_151318_b().func_151330_a(agameprofile);
+         this.field_147147_p.invalidateJson();
       }
 
       if (this.field_71315_w % 900 == 0) {
@@ -696,6 +700,7 @@
       this.field_211152_ao = this.field_211152_ao * 0.8F + (float)l / 1000000.0F * 0.19999999F;
       this.field_71304_b.func_76319_b();
       this.field_71304_b.func_76319_b();
+      net.minecraftforge.fml.hooks.BasicEventHooks.onPostServerTick();
    }
 
    public void func_71190_q(BooleanSupplier p_71190_1_) {
@@ -723,6 +728,7 @@
             }
 
             this.field_71304_b.func_76320_a("tick");
+            net.minecraftforge.fml.hooks.BasicEventHooks.onPreWorldTick(worldserver);
 
             try {
                worldserver.func_72835_b(p_71190_1_);
@@ -740,6 +746,7 @@
                throw new ReportedException(crashreport1);
             }
 
+            net.minecraftforge.fml.hooks.BasicEventHooks.onPostWorldTick(worldserver);
             this.field_71304_b.func_76319_b();
             this.field_71304_b.func_76320_a("tracker");
             worldserver.func_73039_n().func_72788_a();
@@ -752,6 +759,8 @@
          }))[this.field_71315_w % 100] = Util.func_211178_c() - i;
       }
 
+      this.field_71304_b.func_76318_c("dim_unloading");
+      net.minecraftforge.common.DimensionManager.unloadWorlds(this, this.field_71315_w % 200 == 0);
       this.field_71304_b.func_76318_c("connection");
       this.func_147137_ag().func_151269_c();
       this.field_71304_b.func_76318_c("players");
@@ -774,6 +783,14 @@
    }
 
    public static void main(String[] p_main_0_) {
+      //Forge: Copied from DedicatedServer.init as to run as early as possible, Old code left in place intentionally.
+      //Done in good faith with permission: https://github.com/MinecraftForge/MinecraftForge/issues/3659#issuecomment-390467028
+      ServerEula eula = new ServerEula(new File("eula.txt"));
+      if (!eula.func_154346_a()) {
+          field_147145_h.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+          eula.func_154348_b();
+          return;
+      }
       Bootstrap.func_151354_b();
 
       try {
@@ -828,7 +845,7 @@
          MinecraftSessionService minecraftsessionservice = yggdrasilauthenticationservice.createMinecraftSessionService();
          GameProfileRepository gameprofilerepository = yggdrasilauthenticationservice.createProfileRepository();
          PlayerProfileCache playerprofilecache = new PlayerProfileCache(gameprofilerepository, new File(s1, field_152367_a.getName()));
-         final DedicatedServer dedicatedserver = new DedicatedServer(new File(s1), DataFixesManager.func_210901_a(), yggdrasilauthenticationservice, minecraftsessionservice, gameprofilerepository, playerprofilecache);
+         final DedicatedServer dedicatedserver = new DedicatedServer(new File(s1), yggdrasilauthenticationservice, minecraftsessionservice, gameprofilerepository, playerprofilecache);
          if (s != null) {
             dedicatedserver.func_71224_l(s);
          }
@@ -876,7 +893,7 @@
    }
 
    public void func_71256_s() {
-      this.field_175590_aa = new Thread(this, "Server thread");
+      this.field_175590_aa = new Thread(net.minecraftforge.fml.common.thread.SidedThreadGroups.SERVER, this, "Server thread");
       this.field_175590_aa.setUncaughtExceptionHandler((p_195574_0_, p_195574_1_) -> {
          field_147145_h.error(p_195574_1_);
       });
@@ -896,7 +913,7 @@
    }
 
    public WorldServer func_71218_a(DimensionType p_71218_1_) {
-      return this.field_71305_c.get(p_71218_1_);
+      return net.minecraftforge.common.DimensionManager.getWorld(this, p_71218_1_, true, true);
    }
 
    public Iterable<WorldServer> func_212370_w() {
@@ -935,7 +952,7 @@
    }
 
    public String getServerModName() {
-      return "vanilla";
+      return net.minecraftforge.fml.BrandingControl.getServerBranding();
    }
 
    public CrashReport func_71230_b(CrashReport p_71230_1_) {
@@ -1310,10 +1327,6 @@
       return this.field_175590_aa;
    }
 
-   public DataFixer func_195563_aC() {
-      return this.field_184112_s;
-   }
-
    public int func_184108_a(@Nullable WorldServer p_184108_1_) {
       return p_184108_1_ != null ? p_184108_1_.func_82736_K().func_180263_c("spawnRadius") : 10;
    }
@@ -1472,4 +1485,14 @@
          return 0;
       }
    }
+
+   @Nullable
+   public long[] getTickTime(DimensionType dim) {
+      return field_71312_k.get(dim);
+   }
+
+   @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+   public synchronized Map<DimensionType, WorldServer> forgeGetWorldMap() {
+      return this.field_71305_c;
+   }
 }
