--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -167,6 +_,8 @@
    }
 
    public boolean m_135988_(Advancement p_135989_, String p_135990_) {
+      // Forge: don't grant advancements for fake players
+      if (this.f_135968_ instanceof net.minecraftforge.common.util.FakePlayer) return false;
       boolean flag = false;
       AdvancementProgress advancementprogress = this.m_135996_(p_135989_);
       boolean flag1 = advancementprogress.m_8193_();
@@ -174,11 +_,13 @@
          this.m_136008_(p_135989_);
          this.f_135967_.add(p_135989_);
          flag = true;
+         net.minecraftforge.event.ForgeEventFactory.onAdvancementProgressedEvent(this.f_135968_, p_135989_, advancementprogress, p_135990_, net.minecraftforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent.ProgressType.GRANT);
          if (!flag1 && advancementprogress.m_8193_()) {
             p_135989_.m_138321_().m_9989_(this.f_135968_);
             if (p_135989_.m_138320_() != null && p_135989_.m_138320_().m_14996_() && this.f_135968_.m_9236_().m_46469_().m_46207_(GameRules.f_46153_)) {
                this.f_135962_.m_240416_(Component.m_237110_("chat.type.advancement." + p_135989_.m_138320_().m_14992_().m_15548_(), this.f_135968_.m_5446_(), p_135989_.m_138330_()), false);
             }
+            net.minecraftforge.event.ForgeEventFactory.onAdvancementEarnedEvent(this.f_135968_, p_135989_);
          }
       }
 
@@ -197,6 +_,7 @@
          this.m_136004_(p_135999_);
          this.f_135967_.add(p_135999_);
          flag = true;
+         net.minecraftforge.event.ForgeEventFactory.onAdvancementProgressedEvent(this.f_135968_, p_135999_, advancementprogress, p_136000_, net.minecraftforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent.ProgressType.REVOKE);
       }
 
       if (flag1 && !advancementprogress.m_8193_()) {
