--- a/net/minecraft/server/ServerAdvancementManager.java
+++ b/net/minecraft/server/ServerAdvancementManager.java
@@ -32,10 +_,18 @@
     private Map<ResourceLocation, AdvancementHolder> advancements = Map.of();
     private AdvancementTree tree = new AdvancementTree();
     private final HolderLookup.Provider registries;
+    private final net.minecraftforge.common.crafting.conditions.ICondition.IContext context; //Forge: add context
 
+    /** @deprecated Forge: use {@linkplain ServerAdvancementManager#ServerAdvancementManager(LootDataManager, net.minecraftforge.common.crafting.conditions.ICondition.IContext) constructor with context}. */
+    @Deprecated
     public ServerAdvancementManager(HolderLookup.Provider p_336198_) {
+        this(p_336198_, net.minecraftforge.common.crafting.conditions.ICondition.IContext.EMPTY);
+    }
+
+    public ServerAdvancementManager(HolderLookup.Provider p_336198_, net.minecraftforge.common.crafting.conditions.ICondition.IContext context) {
         super(GSON, Registries.elementsDirPath(Registries.ADVANCEMENT));
         this.registries = p_336198_;
+        this.context = context;
     }
 
     protected void apply(Map<ResourceLocation, JsonElement> p_136034_, ResourceManager p_136035_, ProfilerFiller p_136036_) {
@@ -43,7 +_,12 @@
         Builder<ResourceLocation, AdvancementHolder> builder = ImmutableMap.builder();
         p_136034_.forEach((p_326203_, p_326204_) -> {
             try {
-                Advancement advancement = Advancement.CODEC.parse(registryops, p_326204_).getOrThrow(JsonParseException::new);
+                var json = net.minecraftforge.common.ForgeHooks.readConditionalAdvancement(registryops, (com.google.gson.JsonObject)p_326204_);
+                if (json == null) {
+                    LOGGER.debug("Skipping loading advancement {} as its conditions were not met", p_326204_);
+                    return;
+                }
+                Advancement advancement = Advancement.CODEC.parse(registryops, json).getOrThrow(JsonParseException::new);
                 this.validate(p_326203_, advancement);
                 builder.put(p_326203_, new AdvancementHolder(p_326203_, advancement));
             } catch (Exception exception) {
