--- a/net/minecraft/world/entity/animal/equine/Llama.java
+++ b/net/minecraft/world/entity/animal/equine/Llama.java
@@ -370,14 +_,16 @@
 
     @Override
     public boolean causeFallDamage(final double fallDistance, final float damageModifier, final DamageSource damageSource) {
-        int dmg = this.calculateFallDamage(fallDistance, damageModifier);
+        var event = net.minecraftforge.event.ForgeEventFactory.onLivingFall(this, fallDistance, damageModifier);
+        if (event == null) return false;
+        int dmg = this.calculateFallDamage(event.getDistance(), event.getDamageMultiplier());
         if (dmg <= 0) {
             return false;
         }
 
-        if (fallDistance >= 6.0) {
+        if (event.getDistance() >= 6.0) {
             this.hurt(damageSource, dmg);
-            this.propagateFallToPassengers(fallDistance, damageModifier, damageSource);
+            this.propagateFallToPassengers(event.getDistance(), event.getDamageMultiplier(), damageSource);
         }
 
         this.playBlockFallSound();
