--- ../src_base/minecraft/cpw/mods/fml/common/asm/FMLSanityChecker.java
+++ ../src_work/minecraft/cpw/mods/fml/common/asm/FMLSanityChecker.java
@@ -206,10 +206,12 @@
                 FMLRelaunchLog.severe("Technical information: ClientBrandRetriever was at %s, there were %d certificates for it", codeSource.getLocation(), certCount);
             }
         }
-        if (!goodFML)
+        // MCPC+ start - disable message
+        /*if (!goodFML)
         {
             FMLRelaunchLog.severe("FML appears to be missing any signature data. This is not a good thing");
-        }
+        }*/
+        // MCPC+ end
         byte[] mlClass = cl.getClassBytes("ModLoader");
         // Only care in obfuscated env
         if (mlClass == null)
@@ -241,7 +243,7 @@
         cl = (LaunchClassLoader) data.get("classLoader");
         File mcDir = (File)data.get("mcLocation");
         fmlLocation = (File)data.get("coremodLocation");
-        ClassPatchManager.INSTANCE.setup(FMLLaunchHandler.side());
+        //ClassPatchManager.INSTANCE.setup(FMLLaunchHandler.side()); // MCPC+ - disabled since we do not use binpatches
         FMLDeobfuscatingRemapper.INSTANCE.setup(mcDir, cl, (String) data.get("deobfuscationFileName"));
     }
 
