--- a/net/minecraft/client/renderer/BlockFluidRenderer.java
+++ b/net/minecraft/client/renderer/BlockFluidRenderer.java
@@ -219,8 +219,8 @@
                BlockPos blockpos = p_205346_2_.func_177972_a(enumfacing);
                TextureAtlasSprite textureatlassprite2 = atextureatlassprite[1];
                if (!flag) {
-                  Block block = p_205346_1_.func_180495_p(blockpos).func_177230_c();
-                  if (block == Blocks.field_150359_w || block instanceof BlockStainedGlass) {
+                  IBlockState blockstate = p_205346_1_.func_180495_p(blockpos);
+                  if (blockstate.func_193401_d(p_205346_1_, blockpos, enumfacing) == net.minecraft.block.state.BlockFaceShape.SOLID) {
                      textureatlassprite2 = this.field_187501_d;
                   }
                }
