--- ../src-base/minecraft/net/minecraft/client/renderer/BlockFluidRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/BlockFluidRenderer.java
@@ -184,9 +184,9 @@
 
                 if (!flag)
                 {
-                    Block block = p_178270_1_.func_180495_p(blockpos).func_177230_c();
+                    IBlockState state = p_178270_1_.func_180495_p(blockpos);
 
-                    if (block == Blocks.field_150359_w || block == Blocks.field_150399_cn)
+                    if (state.func_193401_d(p_178270_1_, blockpos, EnumFacing.field_82609_l[i1+2].func_176734_d()) == net.minecraft.block.state.BlockFaceShape.SOLID)
                     {
                         textureatlassprite1 = this.field_187501_d;
                     }
