--- ../src-base/minecraft/net/minecraft/client/renderer/BlockRendererDispatcher.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/BlockRendererDispatcher.java
@@ -28,7 +28,7 @@
     public BlockRendererDispatcher(BlockModelShapes p_i46577_1_, BlockColors p_i46577_2_)
     {
         this.field_175028_a = p_i46577_1_;
-        this.field_175027_c = new BlockModelRenderer(p_i46577_2_);
+        this.field_175027_c = new net.minecraftforge.client.model.pipeline.ForgeBlockModelRenderer(p_i46577_2_);
         this.field_175025_e = new BlockFluidRenderer(p_i46577_2_);
     }
 
@@ -43,7 +43,7 @@
         {
             p_175020_1_ = p_175020_1_.func_185899_b(p_175020_4_, p_175020_2_);
             IBakedModel ibakedmodel = this.field_175028_a.func_178125_b(p_175020_1_);
-            IBakedModel ibakedmodel1 = (new SimpleBakedModel.Builder(p_175020_1_, ibakedmodel, p_175020_3_, p_175020_2_)).func_177645_b();
+            IBakedModel ibakedmodel1 = net.minecraftforge.client.ForgeHooksClient.getDamageModel(ibakedmodel, p_175020_3_, p_175020_1_, p_175020_4_, p_175020_2_);
             this.field_175027_c.func_178267_a(p_175020_4_, ibakedmodel1, p_175020_1_, p_175020_2_, Tessellator.func_178181_a().func_178180_c(), true);
         }
     }
@@ -75,6 +75,7 @@
                 switch (enumblockrendertype)
                 {
                     case MODEL:
+                        p_175018_1_ = p_175018_1_.func_177230_c().getExtendedState(p_175018_1_, p_175018_3_, p_175018_2_);
                         return this.field_175027_c.func_178267_a(p_175018_3_, this.func_184389_a(p_175018_1_), p_175018_1_, p_175018_2_, p_175018_4_, true);
                     case ENTITYBLOCK_ANIMATED:
                         return false;
