--- ../src-base/minecraft/net/minecraft/client/renderer/BufferBuilder.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/BufferBuilder.java
@@ -36,12 +36,21 @@
     private VertexFormat field_179011_q;
     private boolean field_179010_r;
 
+    @javax.annotation.Nullable
+    private final net.minecraft.util.BlockRenderLayer layer;
+
     public BufferBuilder(int p_i46275_1_)
     {
+        this(p_i46275_1_, null);
+    }
+
+    public BufferBuilder(int p_i46275_1_, @javax.annotation.Nullable net.minecraft.util.BlockRenderLayer layer)
+    {
         this.field_179001_a = GLAllocation.func_74524_c(p_i46275_1_ * 4);
         this.field_178999_b = this.field_179001_a.asIntBuffer();
         this.field_181676_c = this.field_179001_a.asShortBuffer();
         this.field_179000_c = this.field_179001_a.asFloatBuffer();
+        this.layer = layer;
     }
 
     private void func_181670_b(int p_181670_1_)
@@ -123,6 +132,8 @@
 
             bitset.set(i1);
         }
+        this.field_178999_b.limit(this.field_178999_b.capacity());
+        this.field_178999_b.position(this.func_181664_j());
     }
 
     public BufferBuilder.State func_181672_a()
@@ -415,7 +426,7 @@
 
     public void func_178981_a(int[] p_178981_1_)
     {
-        this.func_181670_b(p_178981_1_.length * 4);
+        this.func_181670_b(p_178981_1_.length * 4 + this.field_179011_q.func_177338_f());//Forge, fix MC-122110
         this.field_178999_b.position(this.func_181664_j());
         this.field_178999_b.put(p_178981_1_);
         this.field_178997_d += p_178981_1_.length / this.field_179011_q.func_181719_f();
@@ -506,15 +517,15 @@
                 break;
             case USHORT:
             case SHORT:
-                this.field_179001_a.putShort(i, (short)((int)p_181663_1_ * 32767 & 65535));
-                this.field_179001_a.putShort(i + 2, (short)((int)p_181663_2_ * 32767 & 65535));
-                this.field_179001_a.putShort(i + 4, (short)((int)p_181663_3_ * 32767 & 65535));
+                this.field_179001_a.putShort(i, (short)((int)(p_181663_1_ * 32767) & 65535));
+                this.field_179001_a.putShort(i + 2, (short)((int)(p_181663_2_ * 32767) & 65535));
+                this.field_179001_a.putShort(i + 4, (short)((int)(p_181663_3_ * 32767) & 65535));
                 break;
             case UBYTE:
             case BYTE:
-                this.field_179001_a.put(i, (byte)((int)p_181663_1_ * 127 & 255));
-                this.field_179001_a.put(i + 1, (byte)((int)p_181663_2_ * 127 & 255));
-                this.field_179001_a.put(i + 2, (byte)((int)p_181663_3_ * 127 & 255));
+                this.field_179001_a.put(i, (byte)((int)(p_181663_1_ * 127) & 255));
+                this.field_179001_a.put(i + 1, (byte)((int)(p_181663_2_ * 127) & 255));
+                this.field_179001_a.put(i + 2, (byte)((int)(p_181663_3_ * 127) & 255));
         }
 
         this.func_181667_k();
@@ -605,4 +616,25 @@
             return this.field_179018_e;
         }
     }
+
+
+    //For some unknown reason Mojang changed the vanilla function to hardcode alpha as 255.... So lets re-add the parameter -.-
+    public void putColorRGBA(int index, int red, int green, int blue, int alpha)
+    {
+        if (ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN)
+            this.field_178999_b.put(index, alpha << 24 | blue << 16 | green << 8 | red);
+        else
+            this.field_178999_b.put(index, red << 24 | green << 16 | blue << 8 | alpha);
+    }
+
+    public boolean isColorDisabled()
+    {
+        return this.field_78939_q;
+    }
+
+    @javax.annotation.Nullable
+    public net.minecraft.util.BlockRenderLayer getRenderLayer()
+    {
+        return layer;
+    }
 }
