--- a/net/minecraft/client/renderer/ChestRenderer.java
+++ b/net/minecraft/client/renderer/ChestRenderer.java
@@ -12,6 +12,7 @@
    public void func_178175_a(Block p_178175_1_, float p_178175_2_) {
       GlStateManager.color4f(p_178175_2_, p_178175_2_, p_178175_2_, 1.0F);
       GlStateManager.rotatef(90.0F, 0.0F, 1.0F, 0.0F);
-      ItemStackTileEntityRenderer.field_147719_a.func_179022_a(new ItemStack(p_178175_1_));
+      ItemStack stack = new ItemStack(p_178175_1_);
+      stack.func_77973_b().getTileEntityItemStackRenderer().func_179022_a(stack);
    }
 }
