--- a/net/minecraft/client/renderer/DimensionSpecialEffects.java
+++ b/net/minecraft/client/renderer/DimensionSpecialEffects.java
@@ -32,7 +_,7 @@
     }
 
     public static DimensionSpecialEffects forType(DimensionType p_108877_) {
-        return EFFECTS.get(p_108877_.effectsLocation());
+        return net.minecraftforge.client.DimensionSpecialEffectsManager.getForType(p_108877_.effectsLocation());
     }
 
     public boolean isSunriseOrSunset(float p_361903_) {
