--- a/net/minecraft/client/renderer/DimensionSpecialEffects.java
+++ b/net/minecraft/client/renderer/DimensionSpecialEffects.java
@@ -13,7 +_,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public abstract class DimensionSpecialEffects {
+public abstract class DimensionSpecialEffects implements net.minecraftforge.client.extensions.IForgeDimensionSpecialEffects {
    private static final Object2ObjectMap<ResourceLocation, DimensionSpecialEffects> f_108857_ = Util.m_137469_(new Object2ObjectArrayMap<>(), (p_108881_) -> {
       DimensionSpecialEffects.OverworldEffects dimensionspecialeffects$overworldeffects = new DimensionSpecialEffects.OverworldEffects();
       p_108881_.defaultReturnValue(dimensionspecialeffects$overworldeffects);
@@ -37,7 +_,7 @@
    }
 
    public static DimensionSpecialEffects m_108876_(DimensionType p_108877_) {
-      return f_108857_.get(p_108877_.f_63837_());
+      return net.minecraftforge.client.DimensionSpecialEffectsManager.getForType(p_108877_.f_63837_());
    }
 
    @Nullable
