--- ../src-base/minecraft/net/minecraft/client/renderer/EntityRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/EntityRenderer.java
@@ -211,6 +211,7 @@
             {
                 this.func_175069_a(new ResourceLocation("shaders/post/invert.json"));
             }
+            else net.minecraftforge.client.ForgeHooksClient.loadEntityShader(p_175066_1_, this);
         }
     }
 
@@ -290,7 +291,7 @@
             this.field_78531_r.func_175607_a(this.field_78531_r.field_71439_g);
         }
 
-        float f3 = this.field_78531_r.field_71441_e.func_175724_o(new BlockPos(this.field_78531_r.func_175606_aa()));
+        float f3 = this.field_78531_r.field_71441_e.func_175724_o(new BlockPos(this.field_78531_r.func_175606_aa().func_174824_e(1F))); // Forge: fix MC-51150
         float f4 = (float)this.field_78531_r.field_71474_y.field_151451_c / 32.0F;
         float f2 = f3 * (1.0F - f4) + f4;
         this.field_78539_ae += (f2 - this.field_78539_ae) * 0.1F;
@@ -412,7 +413,7 @@
 
                         if (d3 < d2 || d2 == 0.0D)
                         {
-                            if (entity1.func_184208_bv() == entity.func_184208_bv())
+                            if (entity1.func_184208_bv() == entity.func_184208_bv() && !entity1.canRiderInteract())
                             {
                                 if (d2 == 0.0D)
                                 {
@@ -505,7 +506,7 @@
                 f = f * 60.0F / 70.0F;
             }
 
-            return f;
+            return net.minecraftforge.client.ForgeHooksClient.getFOVModifier(this, entity, iblockstate, p_78481_1_, f);
         }
     }
 
@@ -569,14 +570,8 @@
             {
                 BlockPos blockpos = new BlockPos(entity);
                 IBlockState iblockstate = this.field_78531_r.field_71441_e.func_180495_p(blockpos);
-                Block block = iblockstate.func_177230_c();
+                net.minecraftforge.client.ForgeHooksClient.orientBedCamera(this.field_78531_r.field_71441_e, blockpos, iblockstate, entity);
 
-                if (block == Blocks.field_150324_C)
-                {
-                    int j = ((EnumFacing)iblockstate.func_177229_b(BlockBed.field_185512_D)).func_176736_b();
-                    GlStateManager.func_179114_b((float)(j * 90), 0.0F, 1.0F, 0.0F);
-                }
-
                 GlStateManager.func_179114_b(entity.field_70126_B + (entity.field_70177_z - entity.field_70126_B) * p_78467_1_ + 180.0F, 0.0F, -1.0F, 0.0F);
                 GlStateManager.func_179114_b(entity.field_70127_C + (entity.field_70125_A - entity.field_70127_C) * p_78467_1_, -1.0F, 0.0F, 0.0F);
             }
@@ -643,17 +638,20 @@
 
         if (!this.field_78531_r.field_71474_y.field_74325_U)
         {
-            GlStateManager.func_179114_b(entity.field_70127_C + (entity.field_70125_A - entity.field_70127_C) * p_78467_1_, 1.0F, 0.0F, 0.0F);
-
+            float yaw = entity.field_70126_B + (entity.field_70177_z - entity.field_70126_B) * p_78467_1_ + 180.0F;
+            float pitch = entity.field_70127_C + (entity.field_70125_A - entity.field_70127_C) * p_78467_1_;
+            float roll = 0.0F;
             if (entity instanceof EntityAnimal)
             {
                 EntityAnimal entityanimal = (EntityAnimal)entity;
-                GlStateManager.func_179114_b(entityanimal.field_70758_at + (entityanimal.field_70759_as - entityanimal.field_70758_at) * p_78467_1_ + 180.0F, 0.0F, 1.0F, 0.0F);
+                yaw = entityanimal.field_70758_at + (entityanimal.field_70759_as - entityanimal.field_70758_at) * p_78467_1_ + 180.0F;
             }
-            else
-            {
-                GlStateManager.func_179114_b(entity.field_70126_B + (entity.field_70177_z - entity.field_70126_B) * p_78467_1_ + 180.0F, 0.0F, 1.0F, 0.0F);
-            }
+            IBlockState state = ActiveRenderInfo.func_186703_a(this.field_78531_r.field_71441_e, entity, p_78467_1_);
+            net.minecraftforge.client.event.EntityViewRenderEvent.CameraSetup event = new net.minecraftforge.client.event.EntityViewRenderEvent.CameraSetup(this, entity, state, p_78467_1_, yaw, pitch, roll);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+            GlStateManager.func_179114_b(event.getRoll(), 0.0F, 0.0F, 1.0F);
+            GlStateManager.func_179114_b(event.getPitch(), 1.0F, 0.0F, 0.0F);
+            GlStateManager.func_179114_b(event.getYaw(), 0.0F, 1.0F, 0.0F);
         }
 
         GlStateManager.func_179109_b(0.0F, -f, 0.0F);
@@ -771,6 +769,7 @@
 
             boolean flag = this.field_78531_r.func_175606_aa() instanceof EntityLivingBase && ((EntityLivingBase)this.field_78531_r.func_175606_aa()).func_70608_bn();
 
+            if (!net.minecraftforge.client.ForgeHooksClient.renderFirstPersonHand(field_78531_r.field_71438_f, p_78476_1_, p_78476_2_))
             if (this.field_78531_r.field_71474_y.field_74320_O == 0 && !flag && !this.field_78531_r.field_71474_y.field_74319_N && !this.field_78531_r.field_71442_b.func_78747_a())
             {
                 this.func_180436_i();
@@ -875,6 +874,15 @@
                         f10 = 0.25F + f7 * 0.75F;
                     }
 
+                    float[] colors = {f8, f9, f10};
+                    world.field_73011_w.getLightmapColors(p_78472_1_, f, f2, f3, colors);
+                    f8 = colors[0]; f9 = colors[1]; f10 = colors[2];
+
+                    // Forge: fix MC-58177
+                    f8 = MathHelper.func_76131_a(f8, 0f, 1f);
+                    f9 = MathHelper.func_76131_a(f9, 0f, 1f);
+                    f10 = MathHelper.func_76131_a(f10, 0f, 1f);
+
                     if (this.field_78531_r.field_71439_g.func_70644_a(MobEffects.field_76439_r))
                     {
                         float f15 = this.func_180438_a(this.field_78531_r.field_71439_g, p_78472_1_);
@@ -1065,8 +1073,10 @@
 
                 if (OpenGlHelper.field_148824_g)
                 {
+                    if (!net.minecraftforge.client.ForgeHooksClient.onEntityOutlineRendering(p_181560_1_))
                     this.field_78531_r.field_71438_f.func_174975_c();
 
+                    if (!net.minecraftforge.client.ForgeHooksClient.onShaderRendering(p_181560_1_, this.field_147707_d, this.field_175083_ad))
                     if (this.field_147707_d != null && this.field_175083_ad)
                     {
                         GlStateManager.func_179128_n(5890);
@@ -1075,6 +1085,7 @@
                         this.field_147707_d.func_148018_a(p_181560_1_);
                         GlStateManager.func_179121_F();
                     }
+                    net.minecraftforge.client.ForgeHooksClient.onFramebufferRendering(p_181560_1_);
 
                     this.field_78531_r.func_147110_a().func_147610_a(true);
                 }
@@ -1101,6 +1112,10 @@
                 GlStateManager.func_179096_D();
                 this.func_78478_c();
                 this.field_78510_Z = System.nanoTime();
+                // Forge: Fix MC-112292
+                net.minecraft.client.renderer.tileentity.TileEntityRendererDispatcher.field_147556_a.field_147553_e = this.field_78531_r.func_110434_K();
+                // Forge: also fix rendering text before entering world (not part of MC-112292, but the same reason)
+                net.minecraft.client.renderer.tileentity.TileEntityRendererDispatcher.field_147556_a.field_147557_n = this.field_78531_r.field_71466_p;
             }
 
             if (this.field_78531_r.field_71462_r != null)
