--- ../src_base/minecraft/net/minecraft/client/renderer/EntityRenderer.java
+++ ../src_work/minecraft/net/minecraft/client/renderer/EntityRenderer.java
@@ -37,6 +37,11 @@
 import org.lwjgl.opengl.GLContext;
 import org.lwjgl.util.glu.GLU;
 
+import net.minecraftforge.client.ForgeHooksClient;
+import net.minecraftforge.client.event.DrawBlockHighlightEvent;
+import net.minecraftforge.client.event.RenderWorldLastEvent;
+import net.minecraftforge.common.MinecraftForge;
+
 @SideOnly(Side.CLIENT)
 public class EntityRenderer
 {
@@ -333,8 +338,15 @@
      */
     private void updateFovModifierHand()
     {
-        EntityPlayerSP var1 = (EntityPlayerSP)this.mc.renderViewEntity;
-        this.fovMultiplierTemp = var1.getFOVMultiplier();
+        if (mc.renderViewEntity instanceof EntityPlayerSP)
+        {
+            EntityPlayerSP var1 = (EntityPlayerSP)this.mc.renderViewEntity;
+            this.fovMultiplierTemp = var1.getFOVMultiplier();
+        }
+        else
+        {
+            this.fovMultiplierTemp = mc.thePlayer.getFOVMultiplier();
+        }
         this.fovModifierHandPrev = this.fovModifierHand;
         this.fovModifierHand += (this.fovMultiplierTemp - this.fovModifierHand) * 0.5F;
     }
@@ -350,7 +362,7 @@
         }
         else
         {
-            EntityPlayer var3 = (EntityPlayer)this.mc.renderViewEntity;
+            EntityLiving var3 = (EntityLiving)this.mc.renderViewEntity;
             float var4 = 70.0F;
 
             if (par2)
@@ -366,8 +378,9 @@
             }
 
             int var6 = ActiveRenderInfo.getBlockIdAtEntityViewpoint(this.mc.theWorld, var3, par1);
-
-            if (var6 != 0 && Block.blocksList[var6].blockMaterial == Material.water)
+            int meta = ActiveRenderInfo.getBlockMetadataAtEntityViewpoint(this.mc.theWorld, var3, par1);
+
+            if (var6 != 0 && Block.blocksList[var6].getBlockMaterial(meta) == Material.water)
             {
                 var4 = var4 * 60.0F / 70.0F;
             }
@@ -437,15 +450,7 @@
 
             if (!this.mc.gameSettings.debugCamEnable)
             {
-                int var10 = this.mc.theWorld.getBlockId(MathHelper.floor_double(var2.posX), MathHelper.floor_double(var2.posY), MathHelper.floor_double(var2.posZ));
-
-                if (var10 == Block.bed.blockID)
-                {
-                    int var11 = this.mc.theWorld.getBlockMetadata(MathHelper.floor_double(var2.posX), MathHelper.floor_double(var2.posY), MathHelper.floor_double(var2.posZ));
-                    int var12 = var11 & 3;
-                    GL11.glRotatef((float)(var12 * 90), 0.0F, 1.0F, 0.0F);
-                }
-
+                ForgeHooksClient.orientBedCamera(mc, var2);
                 GL11.glRotatef(var2.prevRotationYaw + (var2.rotationYaw - var2.prevRotationYaw) * par1 + 180.0F, 0.0F, -1.0F, 0.0F);
                 GL11.glRotatef(var2.prevRotationPitch + (var2.rotationPitch - var2.prevRotationPitch) * par1, -1.0F, 0.0F, 0.0F);
             }
@@ -1124,7 +1129,9 @@
             {
                 RenderHelper.enableStandardItemLighting();
                 this.mc.mcProfiler.endStartSection("entities");
+                ForgeHooksClient.setRenderPass(0);
                 var5.renderEntities(var4.getPosition(par1), var14, par1);
+                ForgeHooksClient.setRenderPass(-1);
                 this.enableLightmap((double)par1);
                 this.mc.mcProfiler.endStartSection("litParticles");
                 var6.renderLitParticles(var4, par1);
@@ -1139,8 +1146,11 @@
                     var17 = (EntityPlayer)var4;
                     GL11.glDisable(GL11.GL_ALPHA_TEST);
                     this.mc.mcProfiler.endStartSection("outline");
-                    var5.drawBlockBreaking(var17, this.mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1);
-                    var5.drawSelectionBox(var17, this.mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1);
+                    if (!ForgeHooksClient.onDrawBlockHighlight(var5, var17, mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1))
+                    {
+                        var5.drawBlockBreaking(var17, this.mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1);
+                        var5.drawSelectionBox(var17, this.mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1);
+                    }
                     GL11.glEnable(GL11.GL_ALPHA_TEST);
                 }
             }
@@ -1194,6 +1204,13 @@
                 this.mc.mcProfiler.endStartSection("water");
                 var5.sortAndRender(var4, 1, (double)par1);
             }
+            
+            RenderHelper.enableStandardItemLighting();
+            this.mc.mcProfiler.endStartSection("entities");
+            ForgeHooksClient.setRenderPass(1);
+            var5.renderEntities(var4.getPosition(par1), var14, par1);
+            ForgeHooksClient.setRenderPass(-1);
+            RenderHelper.disableStandardItemLighting();
 
             GL11.glDepthMask(true);
             GL11.glEnable(GL11.GL_CULL_FACE);
@@ -1204,15 +1221,18 @@
                 var17 = (EntityPlayer)var4;
                 GL11.glDisable(GL11.GL_ALPHA_TEST);
                 this.mc.mcProfiler.endStartSection("outline");
-                var5.drawBlockBreaking(var17, this.mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1);
-                var5.drawSelectionBox(var17, this.mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1);
+                if (!ForgeHooksClient.onDrawBlockHighlight(var5, var17, mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1))
+                {
+                    var5.drawBlockBreaking(var17, this.mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1);
+                    var5.drawSelectionBox(var17, this.mc.objectMouseOver, 0, var17.inventory.getCurrentItem(), par1);
+                }
                 GL11.glEnable(GL11.GL_ALPHA_TEST);
             }
 
             this.mc.mcProfiler.endStartSection("destroyProgress");
             GL11.glEnable(GL11.GL_BLEND);
             GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE);
-            var5.drawBlockDamageTexture(Tessellator.instance, (EntityPlayer)var4, par1);
+            var5.drawBlockDamageTexture(Tessellator.instance, var4, par1);
             GL11.glDisable(GL11.GL_BLEND);
             this.mc.mcProfiler.endStartSection("weather");
             this.renderRainSnow(par1);
@@ -1222,6 +1242,9 @@
             {
                 this.renderCloudsCheck(var5, par1);
             }
+
+            this.mc.mcProfiler.endStartSection("FRenderLast");
+            ForgeHooksClient.dispatchRenderLast(var5, par1);
 
             this.mc.mcProfiler.endStartSection("hand");
 
@@ -1299,6 +1322,7 @@
                 int var18 = var6 + this.random.nextInt(var7) - this.random.nextInt(var7);
                 int var19 = var3.getPrecipitationHeight(var17, var18);
                 int var20 = var3.getBlockId(var17, var19 - 1, var18);
+                int meta = var3.getBlockMetadata(var17, var19 - 1, var18);
                 BiomeGenBase var21 = var3.getBiomeGenForCoords(var17, var18);
 
                 if (var19 <= var5 + var7 && var19 >= var5 - var7 && var21.canSpawnLightningBolt() && var21.getFloatTemperature() >= 0.2F)
@@ -1308,7 +1332,7 @@
 
                     if (var20 > 0)
                     {
-                        if (Block.blocksList[var20].blockMaterial == Material.lava)
+                        if (Block.blocksList[var20].getBlockMaterial(meta) == Material.lava)
                         {
                             this.mc.effectRenderer.addEffect(new EntitySmokeFX(var3, (double)((float)var17 + var22), (double)((float)var19 + 0.1F) - Block.blocksList[var20].getBlockBoundsMinY(), (double)((float)var18 + var23), 0.0D, 0.0D, 0.0D));
                         }
@@ -1608,6 +1632,7 @@
         }
 
         int var21 = ActiveRenderInfo.getBlockIdAtEntityViewpoint(this.mc.theWorld, var3, par1);
+        int meta = ActiveRenderInfo.getBlockMetadataAtEntityViewpoint(this.mc.theWorld, var3, par1);
 
         if (this.cloudFog)
         {
@@ -1616,13 +1641,13 @@
             this.fogColorGreen = (float)var13.yCoord;
             this.fogColorBlue = (float)var13.zCoord;
         }
-        else if (var21 != 0 && Block.blocksList[var21].blockMaterial == Material.water)
+        else if (var21 != 0 && Block.blocksList[var21].getBlockMaterial(meta) == Material.water)
         {
             this.fogColorRed = 0.02F;
             this.fogColorGreen = 0.02F;
             this.fogColorBlue = 0.2F;
         }
-        else if (var21 != 0 && Block.blocksList[var21].blockMaterial == Material.lava)
+        else if (var21 != 0 && Block.blocksList[var21].getBlockMaterial(meta) == Material.lava)
         {
             this.fogColorRed = 0.6F;
             this.fogColorGreen = 0.1F;
@@ -1741,6 +1766,7 @@
             GL11.glNormal3f(0.0F, -1.0F, 0.0F);
             GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
             int var5 = ActiveRenderInfo.getBlockIdAtEntityViewpoint(this.mc.theWorld, var3, par2);
+            int meta = ActiveRenderInfo.getBlockMetadataAtEntityViewpoint(this.mc.theWorld, var3, par2);
             float var6;
 
             if (var3.isPotionActive(Potion.blindness))
@@ -1794,7 +1820,7 @@
                         var11 = (var6 * 30.0F + var8 * 70.0F) / 100.0F;
                     }
                 }
-                else if (var5 > 0 && Block.blocksList[var5].blockMaterial == Material.water)
+                else if (var5 > 0 && Block.blocksList[var5].getBlockMaterial(meta) == Material.water)
                 {
                     GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_EXP);
 
@@ -1818,7 +1844,7 @@
                         var11 = (var6 * 30.0F + var8 * 70.0F) / 100.0F;
                     }
                 }
-                else if (var5 > 0 && Block.blocksList[var5].blockMaterial == Material.lava)
+                else if (var5 > 0 && Block.blocksList[var5].getBlockMaterial(meta) == Material.lava)
                 {
                     GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_EXP);
                     GL11.glFogf(GL11.GL_FOG_DENSITY, 2.0F);
