--- a/net/minecraft/client/renderer/FirstPersonRenderer.java
+++ b/net/minecraft/client/renderer/FirstPersonRenderer.java
@@ -208,7 +_,7 @@
       boolean flag1 = true;
       if (p_228396_4_.isUsingItem()) {
          ItemStack itemstack = p_228396_4_.getUseItem();
-         if (itemstack.getItem() == Items.BOW || itemstack.getItem() == Items.CROSSBOW) {
+         if (itemstack.getItem() instanceof net.minecraft.item.ShootableItem) {
             flag = p_228396_4_.getUsedItemHand() == Hand.MAIN_HAND;
             flag1 = !flag;
          }
@@ -240,12 +_,14 @@
       if (flag) {
          float f5 = hand == Hand.MAIN_HAND ? f : 0.0F;
          float f2 = 1.0F - MathHelper.lerp(p_228396_1_, this.oMainHandHeight, this.mainHandHeight);
+         if(!net.minecraftforge.client.ForgeHooksClient.renderSpecificFirstPersonHand(Hand.MAIN_HAND, p_228396_2_, p_228396_3_, p_228396_5_, p_228396_1_, f1, f5, f2, this.mainHandItem))
          this.renderArmWithItem(p_228396_4_, p_228396_1_, f1, Hand.MAIN_HAND, f5, this.mainHandItem, f2, p_228396_2_, p_228396_3_, p_228396_5_);
       }
 
       if (flag1) {
          float f6 = hand == Hand.OFF_HAND ? f : 0.0F;
          float f7 = 1.0F - MathHelper.lerp(p_228396_1_, this.oOffHandHeight, this.offHandHeight);
+         if(!net.minecraftforge.client.ForgeHooksClient.renderSpecificFirstPersonHand(Hand.OFF_HAND, p_228396_2_, p_228396_3_, p_228396_5_, p_228396_1_, f1, f6, f7, this.offHandItem))
          this.renderArmWithItem(p_228396_4_, p_228396_1_, f1, Hand.OFF_HAND, f6, this.offHandItem, f7, p_228396_2_, p_228396_3_, p_228396_5_);
       }
 
@@ -410,8 +_,16 @@
          this.offHandHeight = MathHelper.clamp(this.offHandHeight - 0.4F, 0.0F, 1.0F);
       } else {
          float f = clientplayerentity.getAttackStrengthScale(1.0F);
-         this.mainHandHeight += MathHelper.clamp((this.mainHandItem == itemstack ? f * f * f : 0.0F) - this.mainHandHeight, -0.4F, 0.4F);
-         this.offHandHeight += MathHelper.clamp((float)(this.offHandItem == itemstack1 ? 1 : 0) - this.offHandHeight, -0.4F, 0.4F);
+         boolean requipM = net.minecraftforge.client.ForgeHooksClient.shouldCauseReequipAnimation(this.mainHandItem, itemstack, clientplayerentity.inventory.selected);
+         boolean requipO = net.minecraftforge.client.ForgeHooksClient.shouldCauseReequipAnimation(this.offHandItem, itemstack1, -1);
+
+         if (!requipM && this.mainHandItem != itemstack)
+            this.mainHandItem = itemstack;
+         if (!requipO && this.offHandItem != itemstack1)
+            this.offHandItem = itemstack1;
+
+         this.mainHandHeight += MathHelper.clamp((!requipM ? f * f * f : 0.0F) - this.mainHandHeight, -0.4F, 0.4F);
+         this.offHandHeight += MathHelper.clamp((float)(!requipO ? 1 : 0) - this.offHandHeight, -0.4F, 0.4F);
       }
 
       if (this.mainHandHeight < 0.1F) {
