--- a/net/minecraft/client/renderer/GameRenderer.java
+++ b/net/minecraft/client/renderer/GameRenderer.java
@@ -135,6 +_,8 @@
          this.func_175069_a(new ResourceLocation("shaders/post/spider.json"));
       } else if (p_175066_1_ instanceof EndermanEntity) {
          this.func_175069_a(new ResourceLocation("shaders/post/invert.json"));
+      } else {
+         net.minecraftforge.client.ForgeHooksClient.loadEntityShader(p_175066_1_, this);
       }
 
    }
@@ -230,10 +_,10 @@
             int i = 3;
             double d1 = d0;
             if (this.field_78531_r.field_71442_b.func_78749_i()) {
-               d1 = 6.0D;
+               d1 = this.field_78531_r.field_71439_g.func_233637_b_(net.minecraftforge.common.ForgeMod.REACH_DISTANCE.get()) + 1.0D; // Forge: Replace "6.0D" with the player's reach distance + 1.0D
                d0 = d1;
             } else {
-               if (d0 > 3.0D) {
+               if (d0 > this.field_78531_r.field_71439_g.func_233637_b_(net.minecraftforge.common.ForgeMod.REACH_DISTANCE.get()) - 2.0D) { // Forge: Replace "3.0D" with the player's reach distance - 2.0D
                   flag = true;
                }
 
@@ -256,7 +_,7 @@
                Entity entity1 = entityraytraceresult.func_216348_a();
                Vector3d vector3d3 = entityraytraceresult.func_216347_e();
                double d2 = vector3d.func_72436_e(vector3d3);
-               if (flag && d2 > 9.0D) {
+               if (flag && d2 > net.minecraft.util.math.MathHelper.func_233022_k_((float) (this.field_78531_r.field_71439_g.func_233637_b_(net.minecraftforge.common.ForgeMod.REACH_DISTANCE.get()) - 2.0D))) { // Forge: Replace "9.0D" with the player's reach distance - 2.0D, squared
                   this.field_78531_r.field_71476_x = BlockRayTraceResult.func_216352_a(vector3d3, Direction.func_210769_a(vector3d1.field_72450_a, vector3d1.field_72448_b, vector3d1.field_72449_c), new BlockPos(vector3d3));
                } else if (d2 < d1 || this.field_78531_r.field_71476_x == null) {
                   this.field_78531_r.field_71476_x = entityraytraceresult;
@@ -310,7 +_,7 @@
             d0 = d0 * 60.0D / 70.0D;
          }
 
-         return d0;
+         return net.minecraftforge.client.ForgeHooksClient.getFOVModifier(this, p_215311_1_, p_215311_2_, d0);
       }
    }
 
@@ -439,6 +_,7 @@
                RenderSystem.loadIdentity();
                this.field_147707_d.func_148018_a(p_195458_1_);
                RenderSystem.popMatrix();
+               RenderSystem.enableTexture(); //FORGE: Fix MC-194675
             }
 
             this.field_78531_r.func_147110_a().func_147610_a(true);
@@ -486,7 +_,7 @@
             }
          } else if (this.field_78531_r.field_71462_r != null) {
             try {
-               this.field_78531_r.field_71462_r.func_230430_a_(matrixstack, i, j, this.field_78531_r.func_193989_ak());
+               net.minecraftforge.client.ForgeHooksClient.drawScreen(this.field_78531_r.field_71462_r, matrixstack, i, j, this.field_78531_r.func_193989_ak());
             } catch (Throwable throwable) {
                CrashReport crashreport1 = CrashReport.func_85055_a(throwable, "Rendering screen");
                CrashReportCategory crashreportcategory1 = crashreport1.func_85058_a("Screen render details");
@@ -595,9 +_,16 @@
       Matrix4f matrix4f = matrixstack.func_227866_c_().func_227870_a_();
       this.func_228379_a_(matrix4f);
       activerenderinfo.func_216772_a(this.field_78531_r.field_71441_e, (Entity)(this.field_78531_r.func_175606_aa() == null ? this.field_78531_r.field_71439_g : this.field_78531_r.func_175606_aa()), !this.field_78531_r.field_71474_y.func_243230_g().func_243192_a(), this.field_78531_r.field_71474_y.func_243230_g().func_243193_b(), p_228378_1_);
+
+      net.minecraftforge.client.event.EntityViewRenderEvent.CameraSetup cameraSetup = net.minecraftforge.client.ForgeHooksClient.onCameraSetup(this, activerenderinfo, p_228378_1_);
+      activerenderinfo.setAnglesInternal(cameraSetup.getYaw(), cameraSetup.getPitch());
+      p_228378_4_.func_227863_a_(Vector3f.field_229183_f_.func_229187_a_(cameraSetup.getRoll()));
+
       p_228378_4_.func_227863_a_(Vector3f.field_229179_b_.func_229187_a_(activerenderinfo.func_216777_e()));
       p_228378_4_.func_227863_a_(Vector3f.field_229181_d_.func_229187_a_(activerenderinfo.func_216778_f() + 180.0F));
       this.field_78531_r.field_71438_f.func_228426_a_(p_228378_4_, p_228378_1_, p_228378_2_, flag, activerenderinfo, this, this.field_78513_d, matrix4f);
+      this.field_78531_r.func_213239_aq().func_219895_b("forge_render_last");
+      net.minecraftforge.client.ForgeHooksClient.dispatchRenderLast(this.field_78531_r.field_71438_f, p_228378_4_, p_228378_1_, matrix4f, p_228378_2_);
       this.field_78531_r.func_213239_aq().func_219895_b("hand");
       if (this.field_175074_C) {
          RenderSystem.clear(256, Minecraft.field_142025_a);
@@ -708,5 +_,10 @@
 
    public OverlayTexture func_228385_m_() {
       return this.field_228375_t_;
+   }
+
+   @Override
+   public net.minecraftforge.resource.IResourceType getResourceType() {
+      return net.minecraftforge.resource.VanillaResourceType.SHADERS;
    }
 }
