--- a/net/minecraft/client/renderer/GameRenderer.java
+++ b/net/minecraft/client/renderer/GameRenderer.java
@@ -220,6 +_,9 @@
             case null:
             default:
                 this.clearPostEffect();
+                var rl = cameraEntity == null ? null : net.minecraftforge.client.EntitySpectatorShaderManager.get(cameraEntity.getType());
+                if (rl != null)
+                    this.setPostEffect(rl);
         }
     }
 
@@ -515,6 +_,10 @@
                     Registry<Block> blockRegistry = this.minecraft.level.registryAccess().lookupOrThrow(Registries.BLOCK);
                     renderOutline = !itemStack.isEmpty()
                         && (itemStack.canBreakBlockInAdventureMode(blockInWorld) || itemStack.canPlaceOnBlockInAdventureMode(blockInWorld));
+                    if (!renderOutline && blockState.hasBlockEntity()) {
+                        var blockEntity = this.minecraft.level.getBlockEntity(pos);
+                        renderOutline = blockEntity != null && blockEntity.hasCustomOutlineRendering();
+                    }
                 }
             }
         }
@@ -538,6 +_,9 @@
         if (optionsState.bobView) {
             this.bobView(cameraState, bobStack);
         }
+
+        var cameraSetup = net.minecraftforge.client.event.ForgeEventFactoryClient.fireComputeCameraAngles(this, this.mainCamera, worldPartialTicks);
+        this.mainCamera.setRotation(cameraSetup.getYaw(), cameraSetup.getPitch(), cameraSetup.getRoll());
 
         projectionMatrix.mul(bobStack.last().pose());
         float screenEffectScale = optionsState.screenEffectScale;
