--- ../src-base/minecraft/net/minecraft/client/renderer/ItemRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/ItemRenderer.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.renderer;
 
+import cpw.mods.fml.client.TextureFXManager;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import net.minecraft.block.Block;
@@ -40,7 +41,12 @@
     public void func_78443_a(EntityLiving p_78443_1_, ItemStack p_78443_2_, int p_78443_3_)
     {
         GL11.glPushMatrix();
-        Block var4 = Block.field_71973_m[p_78443_2_.field_77993_c];
+
+        Block var4 = null;
+        if (p_78443_2_.field_77993_c < Block.field_71973_m.length)
+        {
+            var4 = Block.field_71973_m[p_78443_2_.field_77993_c];
+        }
 
         if (var4 != null && RenderBlocks.func_78597_b(var4.func_71857_b()))
         {
@@ -136,10 +142,17 @@
         float var9;
         float var10;
 
-        for (var7 = 0; var7 < 16; ++var7)
-        {
-            var8 = (float)var7 / 16.0F;
-            var9 = p_78439_1_ + (p_78439_3_ - p_78439_1_) * var8 - 0.001953125F;
+        /* Gets the width/16 of the currently bound texture, used
+         * to fix the side rendering issues on textures != 16 */
+        int tileSize = TextureFXManager.instance().getTextureDimensions(GL11.glGetInteger(GL11.GL_TEXTURE_BINDING_2D)).width / 16;
+
+        float tx = 1.0f / (32 * tileSize);
+        float tz = 1.0f /  tileSize;
+
+        for (var7 = 0; var7 < tileSize; ++var7)
+        {
+            var8 = (float)var7 / tileSize;
+            var9 = p_78439_1_ + (p_78439_3_ - p_78439_1_) * var8 - tx;
             var10 = var6 * var8;
             p_78439_0_.func_78374_a((double)var10, 0.0D, (double)(0.0F - p_78439_5_), (double)var9, (double)p_78439_4_);
             p_78439_0_.func_78374_a((double)var10, 0.0D, 0.0D, (double)var9, (double)p_78439_4_);
@@ -151,11 +164,11 @@
         p_78439_0_.func_78382_b();
         p_78439_0_.func_78375_b(1.0F, 0.0F, 0.0F);
 
-        for (var7 = 0; var7 < 16; ++var7)
-        {
-            var8 = (float)var7 / 16.0F;
-            var9 = p_78439_1_ + (p_78439_3_ - p_78439_1_) * var8 - 0.001953125F;
-            var10 = var6 * var8 + 0.0625F;
+        for (var7 = 0; var7 < tileSize; ++var7)
+        {
+            var8 = (float)var7 / tileSize;
+            var9 = p_78439_1_ + (p_78439_3_ - p_78439_1_) * var8 - tx;
+            var10 = var6 * var8 + tz;
             p_78439_0_.func_78374_a((double)var10, 1.0D, (double)(0.0F - p_78439_5_), (double)var9, (double)p_78439_2_);
             p_78439_0_.func_78374_a((double)var10, 1.0D, 0.0D, (double)var9, (double)p_78439_2_);
             p_78439_0_.func_78374_a((double)var10, 0.0D, 0.0D, (double)var9, (double)p_78439_4_);
@@ -166,11 +179,11 @@
         p_78439_0_.func_78382_b();
         p_78439_0_.func_78375_b(0.0F, 1.0F, 0.0F);
 
-        for (var7 = 0; var7 < 16; ++var7)
-        {
-            var8 = (float)var7 / 16.0F;
-            var9 = p_78439_4_ + (p_78439_2_ - p_78439_4_) * var8 - 0.001953125F;
-            var10 = var6 * var8 + 0.0625F;
+        for (var7 = 0; var7 < tileSize; ++var7)
+        {
+            var8 = (float)var7 / tileSize;
+            var9 = p_78439_4_ + (p_78439_2_ - p_78439_4_) * var8 - tx;
+            var10 = var6 * var8 + tz;
             p_78439_0_.func_78374_a(0.0D, (double)var10, 0.0D, (double)p_78439_1_, (double)var9);
             p_78439_0_.func_78374_a((double)var6, (double)var10, 0.0D, (double)p_78439_3_, (double)var9);
             p_78439_0_.func_78374_a((double)var6, (double)var10, (double)(0.0F - p_78439_5_), (double)p_78439_3_, (double)var9);
@@ -181,10 +194,10 @@
         p_78439_0_.func_78382_b();
         p_78439_0_.func_78375_b(0.0F, -1.0F, 0.0F);
 
-        for (var7 = 0; var7 < 16; ++var7)
-        {
-            var8 = (float)var7 / 16.0F;
-            var9 = p_78439_4_ + (p_78439_2_ - p_78439_4_) * var8 - 0.001953125F;
+        for (var7 = 0; var7 < tileSize; ++var7)
+        {
+            var8 = (float)var7 / tileSize;
+            var9 = p_78439_4_ + (p_78439_2_ - p_78439_4_) * var8 - tx;
             var10 = var6 * var8;
             p_78439_0_.func_78374_a((double)var6, (double)var10, 0.0D, (double)p_78439_3_, (double)var9);
             p_78439_0_.func_78374_a(0.0D, (double)var10, 0.0D, (double)p_78439_1_, (double)var9);
