--- ../src-base/minecraft/net/minecraft/client/renderer/ItemRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/ItemRenderer.java
@@ -308,7 +308,7 @@
 
         if (this.field_78453_b != null)
         {
-            if (this.field_78453_b.func_77973_b() == Items.field_151098_aY)
+            if (this.field_78453_b.func_77973_b() instanceof net.minecraft.item.ItemMap)
             {
                 this.func_178097_a(entityplayersp, f3, f1, f2);
             }
@@ -359,7 +359,8 @@
 
         if (this.field_78455_a.field_71439_g.func_70094_T())
         {
-            IBlockState iblockstate = this.field_78455_a.field_71441_e.func_180495_p(new BlockPos(this.field_78455_a.field_71439_g));
+            BlockPos blockpos = new BlockPos(this.field_78455_a.field_71439_g);
+            IBlockState iblockstate = this.field_78455_a.field_71441_e.func_180495_p(blockpos);
             EntityPlayerSP entityplayersp = this.field_78455_a.field_71439_g;
 
             for (int i = 0; i < 8; ++i)
@@ -367,7 +368,7 @@
                 double d0 = entityplayersp.field_70165_t + (double)(((float)((i >> 0) % 2) - 0.5F) * entityplayersp.field_70130_N * 0.8F);
                 double d1 = entityplayersp.field_70163_u + (double)(((float)((i >> 1) % 2) - 0.5F) * 0.1F);
                 double d2 = entityplayersp.field_70161_v + (double)(((float)((i >> 2) % 2) - 0.5F) * entityplayersp.field_70130_N * 0.8F);
-                BlockPos blockpos = new BlockPos(d0, d1 + (double)entityplayersp.func_70047_e(), d2);
+                blockpos = new BlockPos(d0, d1 + (double)entityplayersp.func_70047_e(), d2);
                 IBlockState iblockstate1 = this.field_78455_a.field_71441_e.func_180495_p(blockpos);
 
                 if (iblockstate1.func_177230_c().func_176214_u())
@@ -378,6 +379,7 @@
 
             if (iblockstate.func_177230_c().func_149645_b() != -1)
             {
+                if (!net.minecraftforge.event.ForgeEventFactory.renderBlockOverlay(field_78455_a.field_71439_g, p_78447_1_, net.minecraftforge.client.event.RenderBlockOverlayEvent.OverlayType.BLOCK, iblockstate, blockpos))
                 this.func_178108_a(p_78447_1_, this.field_78455_a.func_175602_ab().func_175023_a().func_178122_a(iblockstate));
             }
         }
@@ -386,11 +388,13 @@
         {
             if (this.field_78455_a.field_71439_g.func_70055_a(Material.field_151586_h))
             {
+                if (!net.minecraftforge.event.ForgeEventFactory.renderWaterOverlay(field_78455_a.field_71439_g, p_78447_1_))
                 this.func_78448_c(p_78447_1_);
             }
 
             if (this.field_78455_a.field_71439_g.func_70027_ad())
             {
+                if (!net.minecraftforge.event.ForgeEventFactory.renderFireOverlay(field_78455_a.field_71439_g, p_78447_1_))
                 this.func_78442_d(p_78447_1_);
             }
         }
@@ -507,6 +511,12 @@
         {
             if (!this.field_78453_b.func_179549_c(itemstack))
             {
+                if (!this.field_78453_b.func_77973_b().shouldCauseReequipAnimation(this.field_78453_b, itemstack, field_78450_g != entityplayersp.field_71071_by.field_70461_c))
+                {
+                    this.field_78453_b = itemstack;
+                    this.field_78450_g = entityplayersp.field_71071_by.field_70461_c;
+                    return;
+                }
                 flag = true;
             }
         }
