--- a/net/minecraft/client/renderer/ItemModelMesher.java
+++ b/net/minecraft/client/renderer/ItemModelMesher.java
@@ -31,7 +31,8 @@
 
    public TextureAtlasSprite func_199309_a(ItemStack p_199309_1_) {
       IBakedModel ibakedmodel = this.func_178089_a(p_199309_1_);
-      return (ibakedmodel == this.field_178090_d.func_174951_a() || ibakedmodel.func_188618_c()) && p_199309_1_.func_77973_b() instanceof ItemBlock ? this.field_178090_d.func_174954_c().func_178122_a(((ItemBlock)p_199309_1_.func_77973_b()).func_179223_d().func_176223_P()) : ibakedmodel.func_177554_e();
+      // FORGE: Make sure to call the item overrides
+      return (ibakedmodel == this.field_178090_d.func_174951_a() || ibakedmodel.func_188618_c()) && p_199309_1_.func_77973_b() instanceof ItemBlock ? this.field_178090_d.func_174954_c().func_178122_a(((ItemBlock)p_199309_1_.func_77973_b()).func_179223_d().func_176223_P()) : ibakedmodel.func_188617_f().func_209581_a(ibakedmodel, p_199309_1_, null, null).func_177554_e();
    }
 
    public IBakedModel func_178089_a(ItemStack p_178089_1_) {
