--- ../src_base/minecraft/net/minecraft/client/renderer/ItemRenderer.java
+++ ../src_work/minecraft/net/minecraft/client/renderer/ItemRenderer.java
@@ -15,11 +15,19 @@
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.item.EnumAction;
 import net.minecraft.item.Item;
+import net.minecraft.item.ItemBlock;
+import net.minecraft.item.ItemMap;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.storage.MapData;
 import org.lwjgl.opengl.GL11;
 import org.lwjgl.opengl.GL12;
+
+import net.minecraftforge.client.ForgeHooksClient;
+import net.minecraftforge.client.IItemRenderer;
+import net.minecraftforge.client.MinecraftForgeClient;
+import static net.minecraftforge.client.IItemRenderer.ItemRenderType.*;
+import static net.minecraftforge.client.IItemRenderer.ItemRendererHelper.*;
 
 @SideOnly(Side.CLIENT)
 public class ItemRenderer
@@ -54,27 +62,21 @@
     {
         GL11.glPushMatrix();
 
-        Block block = null;
-        if (par2ItemStack.itemID < Block.blocksList.length)
-        {
-            block = Block.blocksList[par2ItemStack.itemID];
-        }
-
-        if (block != null && RenderBlocks.renderItemIn3d(block.getRenderType()))
-        {
-            GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture("/terrain.png"));
-            this.renderBlocksInstance.renderBlockAsItem(block, par2ItemStack.getItemDamage(), 1.0F);
+        IItemRenderer customRenderer = MinecraftForgeClient.getItemRenderer(par2ItemStack, EQUIPPED);
+
+        if (customRenderer != null)
+        {
+            GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture(par2ItemStack.getItem().getTextureFile()));
+            ForgeHooksClient.renderEquippedItem(customRenderer, renderBlocksInstance, par1EntityLiving, par2ItemStack);
+        }
+        else if (par2ItemStack.getItem() instanceof ItemBlock && RenderBlocks.renderItemIn3d(Block.blocksList[par2ItemStack.itemID].getRenderType()))
+        {
+            GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture(par2ItemStack.getItem().getTextureFile()));
+            this.renderBlocksInstance.renderBlockAsItem(Block.blocksList[par2ItemStack.itemID], par2ItemStack.getItemDamage(), 1.0F);
         }
         else
         {
-            if (block != null)
-            {
-                GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture("/terrain.png"));
-            }
-            else
-            {
-                GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture("/gui/items.png"));
-            }
+            GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture(par2ItemStack.getItem().getTextureFile()));
 
             Tessellator tessellator = Tessellator.instance;
             int j = par1EntityLiving.getItemIcon(par2ItemStack, par3);
@@ -279,8 +281,9 @@
         Render render;
         RenderPlayer renderplayer;
 
-        if (itemstack != null && itemstack.itemID == Item.map.itemID)
-        {
+        if (itemstack != null && itemstack.getItem() instanceof ItemMap)
+        {
+            IItemRenderer custom = MinecraftForgeClient.getItemRenderer(itemstack, FIRST_PERSON_MAP);
             GL11.glPushMatrix();
             f4 = 0.8F;
             f7 = entityclientplayermp.getSwingProgress(par1);
@@ -346,11 +349,17 @@
             tessellator.addVertexWithUV((double)(128 + b0), (double)(0 - b0), 0.0D, 1.0D, 0.0D);
             tessellator.addVertexWithUV((double)(0 - b0), (double)(0 - b0), 0.0D, 0.0D, 0.0D);
             tessellator.draw();
-            MapData mapdata = Item.map.getMapData(itemstack, this.mc.theWorld);
-
-            if (mapdata != null)
-            {
-                this.mapItemRenderer.renderMap(this.mc.thePlayer, this.mc.renderEngine, mapdata);
+            MapData mapdata = ((ItemMap)itemstack.getItem()).getMapData(itemstack, this.mc.theWorld);
+            if (custom == null)
+            {
+                if (mapdata != null)
+                {
+                    this.mapItemRenderer.renderMap(this.mc.thePlayer, this.mc.renderEngine, mapdata);
+                }
+            }
+            else
+            {
+                custom.renderItem(FIRST_PERSON_MAP, itemstack, mc.thePlayer, mc.renderEngine, mapdata);
             }
 
             GL11.glPopMatrix();
@@ -453,12 +462,15 @@
             if (itemstack.getItem().requiresMultipleRenderPasses())
             {
                 this.renderItem(entityclientplayermp, itemstack, 0);
-                int i1 = Item.itemsList[itemstack.itemID].getColorFromItemStack(itemstack, 1);
-                f10 = (float)(i1 >> 16 & 255) / 255.0F;
-                f11 = (float)(i1 >> 8 & 255) / 255.0F;
-                f12 = (float)(i1 & 255) / 255.0F;
-                GL11.glColor4f(f3 * f10, f3 * f11, f3 * f12, 1.0F);
-                this.renderItem(entityclientplayermp, itemstack, 1);
+                for (int x = 1; x < itemstack.getItem().getRenderPasses(itemstack.getItemDamage()); x++)
+                {
+                    int i1 = Item.itemsList[itemstack.itemID].getColorFromItemStack(itemstack, x);
+                    f10 = (float)(i1 >> 16 & 255) / 255.0F;
+                    f11 = (float)(i1 >> 8 & 255) / 255.0F;
+                    f12 = (float)(i1 & 255) / 255.0F;
+                    GL11.glColor4f(f3 * f10, f3 * f11, f3 * f12, 1.0F);
+                    this.renderItem(entityclientplayermp, itemstack, x);
+                }
             }
             else
             {
