--- a/net/minecraft/client/renderer/ItemRenderer.java
+++ b/net/minecraft/client/renderer/ItemRenderer.java
@@ -49,7 +49,7 @@
 
    public ItemRenderer(TextureManager p_i46552_1_, ModelManager p_i46552_2_, ItemColors p_i46552_3_) {
       this.field_175057_n = p_i46552_1_;
-      this.field_175059_m = new ItemModelMesher(p_i46552_2_);
+      this.field_175059_m = new net.minecraftforge.client.ItemModelMesherForge(p_i46552_2_);
 
       for(Item item : Item.field_150901_e) {
          if (!field_195411_c.contains(item)) {
