--- ../src-base/minecraft/net/minecraft/client/renderer/ItemModelMesher.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/ItemModelMesher.java
@@ -32,7 +32,9 @@
 
     public TextureAtlasSprite func_178087_a(Item p_178087_1_, int p_178087_2_)
     {
-        return this.func_178089_a(new ItemStack(p_178087_1_, 1, p_178087_2_)).func_177554_e();
+        ItemStack stack = new ItemStack(p_178087_1_, 1, p_178087_2_);
+        IBakedModel model = this.func_178089_a(stack);
+        return model.func_188617_f().handleItemState(model, stack, null, null).func_177554_e();
     }
 
     public IBakedModel func_178089_a(ItemStack p_178089_1_)
