--- a/net/minecraft/client/renderer/LevelEventHandler.java
+++ b/net/minecraft/client/renderer/LevelEventHandler.java
@@ -305,7 +_,7 @@
             case 2001:
                 BlockState blockState = Block.stateById(data);
                 if (!blockState.isAir()) {
-                    SoundType soundType = blockState.getSoundType();
+                    SoundType soundType = blockState.getSoundType(this.level, pos, null);
                     this.level
                         .playLocalSound(
                             pos, soundType.getBreakSound(), SoundSource.BLOCKS, (soundType.getVolume() + 1.0F) / 2.0F, soundType.getPitch() * 0.8F, false
