--- a/net/minecraft/client/renderer/LightTexture.java
+++ b/net/minecraft/client/renderer/LightTexture.java
@@ -123,6 +123,8 @@
                      }
                   }
 
+                  clientworld.func_201675_m().getLightmapColors(p_205106_1_, f, f4, f5, vector3f1);
+
                   vector3f1.func_195901_a(0.0F, 1.0F);
                   if (f2 > 0.0F) {
                      float f10 = Math.max(vector3f1.func_195899_a(), Math.max(vector3f1.func_195900_b(), vector3f1.func_195902_c()));
