--- a/net/minecraft/client/renderer/Quaternion.java
+++ b/net/minecraft/client/renderer/Quaternion.java
@@ -176,4 +176,12 @@
    public Quaternion func_227068_g_() {
       return new Quaternion(this);
    }
+
+   // Forge start
+   public Quaternion(float[] values) {
+      this.field_227061_b_ = values[0];
+      this.field_227062_c_ = values[1];
+      this.field_227063_d_ = values[2];
+      this.field_227064_e_ = values[3];
+   }
 }
