--- ../src_base/minecraft/net/minecraft/client/renderer/RenderGlobal.java
+++ ../src_work/minecraft/net/minecraft/client/renderer/RenderGlobal.java
@@ -65,6 +65,8 @@
 import org.lwjgl.opengl.ARBOcclusionQuery;
 import org.lwjgl.opengl.GL11;
 
+import net.minecraftforge.client.IRenderHandler;
+
 @SideOnly(Side.CLIENT)
 public class RenderGlobal implements IWorldAccess
 {
@@ -677,8 +679,8 @@
                 this.sortedWorldRenderers[i1].isVisible = true;
             }
 
-            
             this.theWorld.theProfiler.endStartSection("render");
+            System.out.println("1"+this.theWorld.theProfiler.getNameOfLastSection());
             k = b0 + this.renderSortedRenderers(b1, l, par2, par3);
 
             do
@@ -776,6 +778,7 @@
                 GL11.glEnable(GL11.GL_ALPHA_TEST);
                 GL11.glEnable(GL11.GL_FOG);
                 this.theWorld.theProfiler.endStartSection("render");
+                System.out.println("2"+this.theWorld.theProfiler.getNameOfLastSection());
                 k += this.renderSortedRenderers(j1, l, par2, par3);
             }
             while (l < this.sortedWorldRenderers.length);
@@ -783,6 +786,7 @@
         else
         {
             this.theWorld.theProfiler.endStartSection("render");
+            System.out.println("3"+this.theWorld.theProfiler.getNameOfLastSection());
             k = b0 + this.renderSortedRenderers(0, this.sortedWorldRenderers.length, par2, par3);
         }
 
@@ -934,6 +938,12 @@
      */
     public void renderSky(float par1)
     {
+        IRenderHandler skyProvider = null;
+        if ((skyProvider = this.mc.theWorld.provider.getSkyRenderer()) != null)
+        {
+            skyProvider.render(par1, this.theWorld, mc);
+            return;
+        }
         if (this.mc.theWorld.provider.dimensionId == 1)
         {
             GL11.glDisable(GL11.GL_FOG);
@@ -1172,6 +1182,13 @@
 
     public void renderClouds(float par1)
     {
+        IRenderHandler renderer = null;
+        if ((renderer = theWorld.provider.getCloudRenderer()) != null)
+        {
+            renderer.render(par1, theWorld, mc);
+            return;
+        }
+
         if (this.mc.theWorld.provider.isSurfaceWorld())
         {
             if (this.mc.gameSettings.fancyGraphics)
@@ -1601,6 +1618,11 @@
     }
 
     public void drawBlockDamageTexture(Tessellator par1Tessellator, EntityPlayer par2EntityPlayer, float par3)
+    {
+        drawBlockDamageTexture(par1Tessellator, (EntityLiving)par2EntityPlayer, par3);
+    }
+
+    public void drawBlockDamageTexture(Tessellator par1Tessellator, EntityLiving par2EntityPlayer, float par3)
     {
         double d0 = par2EntityPlayer.lastTickPosX + (par2EntityPlayer.posX - par2EntityPlayer.lastTickPosX) * (double)par3;
         double d1 = par2EntityPlayer.lastTickPosY + (par2EntityPlayer.posY - par2EntityPlayer.lastTickPosY) * (double)par3;
