--- ../src-base/minecraft/net/minecraft/client/renderer/RenderEngine.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/RenderEngine.java
@@ -1,8 +1,11 @@
 package net.minecraft.client.renderer;
 
+import cpw.mods.fml.client.TextureFXManager;
+import cpw.mods.fml.common.FMLLog;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.awt.Color;
+import java.awt.Dimension;
 import java.awt.Graphics;
 import java.awt.image.BufferedImage;
 import java.awt.image.ImageObserver;
@@ -13,6 +16,9 @@
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
 import javax.imageio.ImageIO;
 import net.minecraft.client.renderer.texture.TextureMap;
 import net.minecraft.client.settings.GameSettings;
@@ -40,6 +46,8 @@
     private final TextureMap field_94155_m;
     private int field_94153_n;
 
+    public static Logger log = FMLLog.getLogger();
+
     public RenderEngine(TexturePackList p_i3192_1_, GameSettings p_i3192_2_)
     {
         this.field_78366_k = p_i3192_1_;
@@ -111,15 +119,18 @@
                     }
                     else
                     {
-                        aint1 = this.func_78348_b(this.func_78345_a(inputstream));
+                        BufferedImage bufferedimage = this.func_78345_a(inputstream);
+                        TextureFXManager.instance().fixTransparency(bufferedimage, p_78346_1_);
+                        aint1 = this.func_78348_b(bufferedimage);
                     }
                 }
 
                 this.field_78359_d.put(p_78346_1_, aint1);
                 return aint1;
             }
-            catch (IOException ioexception)
-            {
+            catch (Exception ioexception)
+            {
+                log.log(Level.INFO, String.format("An error occured reading texture file %s (getTexture)", p_78346_1_), ioexception);
                 ioexception.printStackTrace();
                 int[] aint2 = this.func_78348_b(this.field_78364_l);
                 this.field_78359_d.put(p_78346_1_, aint2);
@@ -209,7 +220,9 @@
                         }
                         else
                         {
-                            this.func_78351_a(this.func_78345_a(inputstream), i);
+                            BufferedImage bufferedimage = this.func_78345_a(inputstream);
+                            TextureFXManager.instance().fixTransparency(bufferedimage, p_78341_1_);
+                            this.func_78351_a(bufferedimage, i);
                         }
                     }
 
@@ -498,7 +511,9 @@
                 }
                 else
                 {
+                    String textureName = s;
                     bufferedimage = this.func_78345_a(itexturepack.func_77532_a(s));
+                    TextureFXManager.instance().fixTransparency(bufferedimage, textureName);
                 }
 
                 int j = ((Integer)this.field_78362_c.get(s)).intValue();
@@ -506,8 +521,9 @@
                 this.field_78361_b = false;
                 this.field_78363_a = false;
             }
-            catch (IOException ioexception)
-            {
+            catch (Exception ioexception)
+            {
+                log.log(Level.INFO,String.format("An error occured reading texture file %s (refreshTexture)", s), ioexception);
                 ioexception.printStackTrace();
             }
         }
@@ -536,15 +552,18 @@
                 }
                 else
                 {
+                    String textureName = s;
                     bufferedimage = this.func_78345_a(itexturepack.func_77532_a(s));
+                    TextureFXManager.instance().fixTransparency(bufferedimage, textureName);
                 }
 
                 this.func_78340_a(bufferedimage, (int[])this.field_78359_d.get(s));
                 this.field_78361_b = false;
                 this.field_78363_a = false;
             }
-            catch (IOException ioexception1)
-            {
+            catch (Exception ioexception1)
+            {
+                log.log(Level.INFO,String.format("An error occured reading texture file data %s (refreshTexture)", s), ioexception1);
                 ioexception1.printStackTrace();
             }
         }
