--- ../src-base/minecraft/net/minecraft/client/renderer/RenderBlocks.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/RenderBlocks.java
@@ -52,6 +52,8 @@
 import org.lwjgl.opengl.GL11;
 import org.lwjgl.opengl.GL12;
 
+import static net.minecraftforge.common.util.ForgeDirection.*;
+
 @SideOnly(Side.CLIENT)
 public class RenderBlocks
 {
@@ -339,9 +341,9 @@
     public boolean func_147773_v(Block p_147773_1_, int p_147773_2_, int p_147773_3_, int p_147773_4_)
     {
         Tessellator tessellator = Tessellator.field_78398_a;
-        int l = this.field_147845_a.func_72805_g(p_147773_2_, p_147773_3_, p_147773_4_);
-        int i1 = BlockBed.func_149895_l(l);
-        boolean flag = BlockBed.func_149975_b(l);
+        Block bed = this.field_147845_a.func_147439_a(p_147773_2_, p_147773_3_, p_147773_4_);
+        int i1 = bed.getBedDirection(field_147845_a, p_147773_2_, p_147773_3_, p_147773_4_);
+        boolean flag = bed.isBedFoot(field_147845_a, p_147773_2_, p_147773_3_, p_147773_4_);
         float f = 0.5F;
         float f1 = 1.0F;
         float f2 = 0.8F;
@@ -350,6 +352,7 @@
         tessellator.func_78380_c(j1);
         tessellator.func_78386_a(f, f, f);
         IIcon iicon = this.func_147793_a(p_147773_1_, this.field_147845_a, p_147773_2_, p_147773_3_, p_147773_4_, 0);
+        if (func_147744_b()) iicon = field_147840_d; //BugFix Proper breaking texture on underside
         double d0 = (double)iicon.func_94209_e();
         double d1 = (double)iicon.func_94212_f();
         double d2 = (double)iicon.func_94206_g();
@@ -366,6 +369,7 @@
         tessellator.func_78380_c(p_147773_1_.func_149677_c(this.field_147845_a, p_147773_2_, p_147773_3_ + 1, p_147773_4_));
         tessellator.func_78386_a(f1, f1, f1);
         iicon = this.func_147793_a(p_147773_1_, this.field_147845_a, p_147773_2_, p_147773_3_, p_147773_4_, 1);
+        if (func_147744_b()) iicon = field_147840_d; //BugFix Proper breaking texture on underside
         d0 = (double)iicon.func_94209_e();
         d1 = (double)iicon.func_94212_f();
         d2 = (double)iicon.func_94206_g();
@@ -2040,7 +2044,7 @@
         double d10;
         double d11;
 
-        if (!World.func_147466_a(this.field_147845_a, p_147801_2_, p_147801_3_ - 1, p_147801_4_) && !Blocks.field_150480_ab.func_149844_e(this.field_147845_a, p_147801_2_, p_147801_3_ - 1, p_147801_4_))
+        if (!World.func_147466_a(this.field_147845_a, p_147801_2_, p_147801_3_ - 1, p_147801_4_) && !Blocks.field_150480_ab.canCatchFire(this.field_147845_a, p_147801_2_, p_147801_3_ - 1, p_147801_4_, UP))
         {
             float f2 = 0.2F;
             float f1 = 0.0625F;
@@ -2060,7 +2064,7 @@
                 d0 = d5;
             }
 
-            if (Blocks.field_150480_ab.func_149844_e(this.field_147845_a, p_147801_2_ - 1, p_147801_3_, p_147801_4_))
+            if (Blocks.field_150480_ab.canCatchFire(this.field_147845_a, p_147801_2_ - 1, p_147801_3_, p_147801_4_, EAST))
             {
                 tessellator.func_78374_a((double)((float)p_147801_2_ + f2), (double)((float)p_147801_3_ + f + f1), (double)(p_147801_4_ + 1), d2, d1);
                 tessellator.func_78374_a((double)(p_147801_2_ + 0), (double)((float)(p_147801_3_ + 0) + f1), (double)(p_147801_4_ + 1), d2, d3);
@@ -2072,7 +2076,7 @@
                 tessellator.func_78374_a((double)((float)p_147801_2_ + f2), (double)((float)p_147801_3_ + f + f1), (double)(p_147801_4_ + 1), d2, d1);
             }
 
-            if (Blocks.field_150480_ab.func_149844_e(this.field_147845_a, p_147801_2_ + 1, p_147801_3_, p_147801_4_))
+            if (Blocks.field_150480_ab.canCatchFire(this.field_147845_a, p_147801_2_ + 1, p_147801_3_, p_147801_4_, WEST))
             {
                 tessellator.func_78374_a((double)((float)(p_147801_2_ + 1) - f2), (double)((float)p_147801_3_ + f + f1), (double)(p_147801_4_ + 0), d0, d1);
                 tessellator.func_78374_a((double)(p_147801_2_ + 1 - 0), (double)((float)(p_147801_3_ + 0) + f1), (double)(p_147801_4_ + 0), d0, d3);
@@ -2084,7 +2088,7 @@
                 tessellator.func_78374_a((double)((float)(p_147801_2_ + 1) - f2), (double)((float)p_147801_3_ + f + f1), (double)(p_147801_4_ + 0), d0, d1);
             }
 
-            if (Blocks.field_150480_ab.func_149844_e(this.field_147845_a, p_147801_2_, p_147801_3_, p_147801_4_ - 1))
+            if (Blocks.field_150480_ab.canCatchFire(this.field_147845_a, p_147801_2_, p_147801_3_, p_147801_4_ - 1, SOUTH))
             {
                 tessellator.func_78374_a((double)(p_147801_2_ + 0), (double)((float)p_147801_3_ + f + f1), (double)((float)p_147801_4_ + f2), d2, d1);
                 tessellator.func_78374_a((double)(p_147801_2_ + 0), (double)((float)(p_147801_3_ + 0) + f1), (double)(p_147801_4_ + 0), d2, d3);
@@ -2096,7 +2100,7 @@
                 tessellator.func_78374_a((double)(p_147801_2_ + 0), (double)((float)p_147801_3_ + f + f1), (double)((float)p_147801_4_ + f2), d2, d1);
             }
 
-            if (Blocks.field_150480_ab.func_149844_e(this.field_147845_a, p_147801_2_, p_147801_3_, p_147801_4_ + 1))
+            if (Blocks.field_150480_ab.canCatchFire(this.field_147845_a, p_147801_2_, p_147801_3_, p_147801_4_ + 1, NORTH))
             {
                 tessellator.func_78374_a((double)(p_147801_2_ + 1), (double)((float)p_147801_3_ + f + f1), (double)((float)(p_147801_4_ + 1) - f2), d0, d1);
                 tessellator.func_78374_a((double)(p_147801_2_ + 1), (double)((float)(p_147801_3_ + 0) + f1), (double)(p_147801_4_ + 1 - 0), d0, d3);
@@ -2108,7 +2112,7 @@
                 tessellator.func_78374_a((double)(p_147801_2_ + 1), (double)((float)p_147801_3_ + f + f1), (double)((float)(p_147801_4_ + 1) - f2), d0, d1);
             }
 
-            if (Blocks.field_150480_ab.func_149844_e(this.field_147845_a, p_147801_2_, p_147801_3_ + 1, p_147801_4_))
+            if (Blocks.field_150480_ab.canCatchFire(this.field_147845_a, p_147801_2_, p_147801_3_ + 1, p_147801_4_, DOWN))
             {
                 d5 = (double)p_147801_2_ + 0.5D + 0.5D;
                 d6 = (double)p_147801_2_ + 0.5D - 0.5D;
@@ -2724,10 +2728,10 @@
         double d16 = (double)p_147733_2_ + 0.5D + 0.0625D;
         double d17 = (double)p_147733_4_ + 0.5D - 0.0625D;
         double d18 = (double)p_147733_4_ + 0.5D + 0.0625D;
-        boolean flag = flag5 ? ((BlockStainedGlassPane)p_147733_1_).func_150098_a(this.field_147845_a.func_147439_a(p_147733_2_, p_147733_3_, p_147733_4_ - 1)) : ((BlockPane)p_147733_1_).func_150098_a(this.field_147845_a.func_147439_a(p_147733_2_, p_147733_3_, p_147733_4_ - 1));
-        boolean flag1 = flag5 ? ((BlockStainedGlassPane)p_147733_1_).func_150098_a(this.field_147845_a.func_147439_a(p_147733_2_, p_147733_3_, p_147733_4_ + 1)) : ((BlockPane)p_147733_1_).func_150098_a(this.field_147845_a.func_147439_a(p_147733_2_, p_147733_3_, p_147733_4_ + 1));
-        boolean flag2 = flag5 ? ((BlockStainedGlassPane)p_147733_1_).func_150098_a(this.field_147845_a.func_147439_a(p_147733_2_ - 1, p_147733_3_, p_147733_4_)) : ((BlockPane)p_147733_1_).func_150098_a(this.field_147845_a.func_147439_a(p_147733_2_ - 1, p_147733_3_, p_147733_4_));
-        boolean flag3 = flag5 ? ((BlockStainedGlassPane)p_147733_1_).func_150098_a(this.field_147845_a.func_147439_a(p_147733_2_ + 1, p_147733_3_, p_147733_4_)) : ((BlockPane)p_147733_1_).func_150098_a(this.field_147845_a.func_147439_a(p_147733_2_ + 1, p_147733_3_, p_147733_4_));
+        boolean flag  = ((BlockPane)p_147733_1_).canPaneConnectTo(this.field_147845_a, p_147733_2_, p_147733_3_, p_147733_4_ - 1, NORTH);
+        boolean flag1 = ((BlockPane)p_147733_1_).canPaneConnectTo(this.field_147845_a, p_147733_2_, p_147733_3_, p_147733_4_ + 1, SOUTH);
+        boolean flag2 = ((BlockPane)p_147733_1_).canPaneConnectTo(this.field_147845_a, p_147733_2_ - 1, p_147733_3_, p_147733_4_, WEST );
+        boolean flag3 = ((BlockPane)p_147733_1_).canPaneConnectTo(this.field_147845_a, p_147733_2_ + 1, p_147733_3_, p_147733_4_, EAST );
         double d19 = 0.001D;
         double d20 = 0.999D;
         double d21 = 0.001D;
@@ -3121,10 +3125,10 @@
         double d16 = (double)p_147767_2_ + 0.5D + 0.0625D;
         double d17 = (double)p_147767_4_ + 0.5D - 0.0625D;
         double d18 = (double)p_147767_4_ + 0.5D + 0.0625D;
-        boolean flag = p_147767_1_.func_150098_a(this.field_147845_a.func_147439_a(p_147767_2_, p_147767_3_, p_147767_4_ - 1));
-        boolean flag1 = p_147767_1_.func_150098_a(this.field_147845_a.func_147439_a(p_147767_2_, p_147767_3_, p_147767_4_ + 1));
-        boolean flag2 = p_147767_1_.func_150098_a(this.field_147845_a.func_147439_a(p_147767_2_ - 1, p_147767_3_, p_147767_4_));
-        boolean flag3 = p_147767_1_.func_150098_a(this.field_147845_a.func_147439_a(p_147767_2_ + 1, p_147767_3_, p_147767_4_));
+        boolean flag  = p_147767_1_.canPaneConnectTo(this.field_147845_a, p_147767_2_, p_147767_3_, p_147767_4_ - 1, NORTH);
+        boolean flag1 = p_147767_1_.canPaneConnectTo(this.field_147845_a, p_147767_2_, p_147767_3_, p_147767_4_ + 1, SOUTH);
+        boolean flag2 = p_147767_1_.canPaneConnectTo(this.field_147845_a, p_147767_2_ - 1, p_147767_3_, p_147767_4_, WEST );
+        boolean flag3 = p_147767_1_.canPaneConnectTo(this.field_147845_a, p_147767_2_ + 1, p_147767_3_, p_147767_4_, EAST );
         boolean flag4 = p_147767_1_.func_149646_a(this.field_147845_a, p_147767_2_, p_147767_3_ + 1, p_147767_4_, 1);
         boolean flag5 = p_147767_1_.func_149646_a(this.field_147845_a, p_147767_2_, p_147767_3_ - 1, p_147767_4_, 0);
         double d19 = 0.01D;
@@ -8247,4 +8251,18 @@
 
         return (IIcon)p_147758_1_;
     }
+
+    /*==================================== FORGE START ===========================================*/
+    private static RenderBlocks instance;
+    /**
+     * Returns a single lazy loaded instance of RenderBlocks, for use in mods who 
+     * don't care about the interaction of other objects on the current state of the RenderBlocks they are using.
+     * @return A global instance of RenderBlocks
+     */
+    public static RenderBlocks getInstance()
+    {
+        if (instance == null) instance = new RenderBlocks();
+        return instance;
+    }
+    /*==================================== FORGE END =============================================*/
 }
