--- a/net/minecraft/client/renderer/RenderState.java
+++ b/net/minecraft/client/renderer/RenderState.java
@@ -360,7 +360,7 @@
       private final OptionalDouble field_228587_Q_;
 
       public LineState(OptionalDouble p_i225982_1_) {
-         super("alpha", () -> {
+         super("line", () -> { // FORGE: fix MC-167447
             if (!Objects.equals(p_i225982_1_, OptionalDouble.of(1.0D))) {
                if (p_i225982_1_.isPresent()) {
                   RenderSystem.lineWidth((float)p_i225982_1_.getAsDouble());
