--- a/net/minecraft/client/renderer/ScreenEffectRenderer.java
+++ b/net/minecraft/client/renderer/ScreenEffectRenderer.java
@@ -55,18 +_,24 @@
         Player player = this.minecraft.player;
         if (this.minecraft.options.getCameraType().isFirstPerson() && !p_409640_) {
             if (!player.noPhysics) {
-                BlockState blockstate = getViewBlockingState(player);
+                var overlay = getOverlayBlock(player);
+                BlockState blockstate = overlay == null ? null : overlay.getLeft();
                 if (blockstate != null) {
+                    if (!net.minecraftforge.client.ForgeHooksClient.renderBlockOverlay(player, posestack, net.minecraftforge.client.event.RenderBlockScreenEffectEvent.OverlayType.BLOCK, overlay.getLeft(), overlay.getRight()))
                     renderTex(this.minecraft.getBlockRenderer().getBlockModelShaper().getParticleIcon(blockstate), posestack, this.bufferSource);
                 }
             }
 
             if (!this.minecraft.player.isSpectator()) {
                 if (this.minecraft.player.isEyeInFluid(FluidTags.WATER)) {
+                    if (!net.minecraftforge.client.ForgeHooksClient.renderWaterOverlay(player, posestack))
                     renderWater(this.minecraft, posestack, this.bufferSource);
+                } else if (!player.getEyeInFluidType().isAir()) {
+                    net.minecraftforge.client.extensions.common.IClientFluidTypeExtensions.of(player.getEyeInFluidType()).renderOverlay(this.minecraft, posestack, this.bufferSource);
                 }
 
                 if (this.minecraft.player.isOnFire()) {
+                    if (!net.minecraftforge.client.ForgeHooksClient.renderFireOverlay(player, posestack))
                     renderFire(posestack, this.bufferSource);
                 }
             }
@@ -116,6 +_,12 @@
 
     @Nullable
     private static BlockState getViewBlockingState(Player p_110717_) {
+        var ret = getOverlayBlock(p_110717_);
+        return ret == null ? null : ret.getLeft();
+    }
+
+    @Nullable
+    private static org.apache.commons.lang3.tuple.Pair<BlockState, BlockPos> getOverlayBlock(Player p_110717_) {
         BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
         for (int i = 0; i < 8; i++) {
@@ -125,7 +_,7 @@
             blockpos$mutableblockpos.set(d0, d1, d2);
             BlockState blockstate = p_110717_.level().getBlockState(blockpos$mutableblockpos);
             if (blockstate.getRenderShape() != RenderShape.INVISIBLE && blockstate.isViewBlocking(p_110717_.level(), blockpos$mutableblockpos)) {
-                return blockstate;
+                return org.apache.commons.lang3.tuple.Pair.of(blockstate, blockpos$mutableblockpos.immutable());
             }
         }
 
@@ -153,6 +_,10 @@
     }
 
     private static void renderWater(Minecraft p_110726_, PoseStack p_110727_, MultiBufferSource p_376402_) {
+        renderFluid(p_110726_, p_110727_, p_376402_, UNDERWATER_LOCATION);
+    }
+
+    public static void renderFluid(Minecraft p_110726_, PoseStack p_110727_, MultiBufferSource p_376402_, ResourceLocation texture) {
         BlockPos blockpos = BlockPos.containing(p_110726_.player.getX(), p_110726_.player.getEyeY(), p_110726_.player.getZ());
         float f = LightTexture.getBrightness(p_110726_.player.level().dimensionType(), p_110726_.player.level().getMaxLocalRawBrightness(blockpos));
         int i = ARGB.colorFromFloat(0.1F, f, f, f);
@@ -165,7 +_,7 @@
         float f7 = -p_110726_.player.getYRot() / 64.0F;
         float f8 = p_110726_.player.getXRot() / 64.0F;
         Matrix4f matrix4f = p_110727_.last().pose();
-        VertexConsumer vertexconsumer = p_376402_.getBuffer(RenderType.blockScreenEffect(UNDERWATER_LOCATION));
+        VertexConsumer vertexconsumer = p_376402_.getBuffer(RenderType.blockScreenEffect(texture));
         vertexconsumer.addVertex(matrix4f, -1.0F, -1.0F, -0.5F).setUv(4.0F + f7, 4.0F + f8).setColor(i);
         vertexconsumer.addVertex(matrix4f, 1.0F, -1.0F, -0.5F).setUv(0.0F + f7, 4.0F + f8).setColor(i);
         vertexconsumer.addVertex(matrix4f, 1.0F, 1.0F, -0.5F).setUv(0.0F + f7, 0.0F + f8).setColor(i);
