--- a/net/minecraft/client/renderer/SpriteCoordinateExpander.java
+++ b/net/minecraft/client/renderer/SpriteCoordinateExpander.java
@@ -17,32 +_,38 @@
 
     @Override
     public VertexConsumer addVertex(float p_342932_, float p_342886_, float p_342696_) {
-        return this.delegate.addVertex(p_342932_, p_342886_, p_342696_);
+        this.delegate.addVertex(p_342932_, p_342886_, p_342696_);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setColor(int p_344589_, int p_342555_, int p_344320_, int p_345258_) {
-        return this.delegate.setColor(p_344589_, p_342555_, p_344320_, p_345258_);
+        this.delegate.setColor(p_344589_, p_342555_, p_344320_, p_345258_);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv(float p_343856_, float p_344420_) {
-        return this.delegate.setUv(this.sprite.getU(p_343856_), this.sprite.getV(p_344420_));
+        this.delegate.setUv(this.sprite.getU(p_343856_), this.sprite.getV(p_344420_));
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv1(int p_343784_, int p_344827_) {
-        return this.delegate.setUv1(p_343784_, p_344827_);
+        this.delegate.setUv1(p_343784_, p_344827_);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv2(int p_345257_, int p_344124_) {
-        return this.delegate.setUv2(p_345257_, p_344124_);
+        this.delegate.setUv2(p_345257_, p_344124_);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setNormal(float p_342779_, float p_342534_, float p_344783_) {
-        return this.delegate.setNormal(p_342779_, p_342534_, p_344783_);
+        this.delegate.setNormal(p_342779_, p_342534_, p_344783_);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
