--- ../src_base/minecraft/net/minecraft/client/renderer/ThreadDownloadImageDataINNER1.java
+++ ../src_work/minecraft/net/minecraft/client/renderer/ThreadDownloadImageDataINNER1.java
@@ -29,23 +29,23 @@
             httpurlconnection.setDoOutput(false);
             httpurlconnection.connect();
 
-            if (httpurlconnection.getResponseCode() / 100 == 2)
+            if (httpurlconnection.getResponseCode() / 100 != 2)
             {
-                BufferedImage bufferedimage = ImageIO.read(httpurlconnection.getInputStream());
-
-                if (ThreadDownloadImageData.func_110555_b(this.field_110932_a) != null)
-                {
-                    bufferedimage = ThreadDownloadImageData.func_110555_b(this.field_110932_a).parseUserSkin(bufferedimage);
-                }
-
-                this.field_110932_a.func_110556_a(bufferedimage);
                 return;
             }
+
+            BufferedImage bufferedimage = ImageIO.read(httpurlconnection.getInputStream());
+
+            if (ThreadDownloadImageData.func_110555_b(this.field_110932_a) != null)
+            {
+                bufferedimage = ThreadDownloadImageData.func_110555_b(this.field_110932_a).parseUserSkin(bufferedimage);
+            }
+
+            this.field_110932_a.func_110556_a(bufferedimage);
         }
         catch (Exception exception)
         {
             exception.printStackTrace();
-            return;
         }
         finally
         {
