--- a/net/minecraft/client/renderer/WorldVertexBufferUploader.java
+++ b/net/minecraft/client/renderer/WorldVertexBufferUploader.java
@@ -22,25 +22,9 @@
             int k = vertexformatelement.func_177367_b().func_177397_c();
             int l = vertexformatelement.func_177369_e();
             bytebuffer.position(vertexformat.func_181720_d(j));
-            switch(vertexformatelement$enumusage) {
-            case POSITION:
-               GlStateManager.func_187427_b(vertexformatelement.func_177370_d(), k, i, bytebuffer);
-               GlStateManager.func_187410_q(32884);
-               break;
-            case UV:
-               OpenGlHelper.func_77472_b(OpenGlHelper.field_77478_a + l);
-               GlStateManager.func_187404_a(vertexformatelement.func_177370_d(), k, i, bytebuffer);
-               GlStateManager.func_187410_q(32888);
-               OpenGlHelper.func_77472_b(OpenGlHelper.field_77478_a);
-               break;
-            case COLOR:
-               GlStateManager.func_187400_c(vertexformatelement.func_177370_d(), k, i, bytebuffer);
-               GlStateManager.func_187410_q(32886);
-               break;
-            case NORMAL:
-               GlStateManager.func_187446_a(k, i, bytebuffer);
-               GlStateManager.func_187410_q(32885);
-            }
+
+            // moved to VertexFormatElement.preDraw
+            vertexformatelement.func_177375_c().preDraw(vertexformat, j, i, bytebuffer);
          }
 
          GlStateManager.func_187439_f(p_181679_1_.func_178979_i(), 0, p_181679_1_.func_178989_h());
@@ -50,22 +34,9 @@
             VertexFormatElement vertexformatelement1 = list.get(i1);
             VertexFormatElement.EnumUsage vertexformatelement$enumusage1 = vertexformatelement1.func_177375_c();
             int k1 = vertexformatelement1.func_177369_e();
-            switch(vertexformatelement$enumusage1) {
-            case POSITION:
-               GlStateManager.func_187429_p(32884);
-               break;
-            case UV:
-               OpenGlHelper.func_77472_b(OpenGlHelper.field_77478_a + k1);
-               GlStateManager.func_187429_p(32888);
-               OpenGlHelper.func_77472_b(OpenGlHelper.field_77478_a);
-               break;
-            case COLOR:
-               GlStateManager.func_187429_p(32886);
-               GlStateManager.func_179117_G();
-               break;
-            case NORMAL:
-               GlStateManager.func_187429_p(32885);
-            }
+
+            // moved to VertexFormatElement.postDraw
+            vertexformatelement1.func_177375_c().postDraw(vertexformat, i1, i, bytebuffer);
          }
       }
 
