--- ../src-base/minecraft/net/minecraft/client/renderer/WorldRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/WorldRenderer.java
@@ -351,6 +351,7 @@
     {
         this.field_78939_q = true;
     }
+    public boolean isNoColor() { return field_78939_q; }
 
     public WorldRenderer func_181666_a(float p_181666_1_, float p_181666_2_, float p_181666_3_, float p_181666_4_)
     {
@@ -578,6 +579,11 @@
         }
     }
 
+    public void checkAndGrow()
+    {
+        this.func_181670_b(field_179011_q.func_177338_f()/* / 4 * 4 */);
+    }
+
     @SideOnly(Side.CLIENT)
     public class State
     {
