--- a/net/minecraft/client/gui/render/GuiRenderer.java
+++ b/net/minecraft/client/gui/render/GuiRenderer.java
@@ -133,7 +_,7 @@
         for (PictureInPictureRenderer<?> pictureinpicturerenderer : p_410565_) {
             builder.put((Class<? extends PictureInPictureRenderState>)pictureinpicturerenderer.getRenderStateClass(), pictureinpicturerenderer);
         }
-
+        net.minecraftforge.client.ForgeHooksClient.onRegisterPictureInPictureRenderers(p_410441_, builder);
         this.pictureInPictureRenderers = builder.buildOrThrow();
     }
 
