--- a/net/minecraft/entity/item/minecart/AbstractMinecartEntity.java
+++ b/net/minecraft/entity/item/minecart/AbstractMinecartEntity.java
@@ -48,7 +_,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public abstract class AbstractMinecartEntity extends Entity {
+public abstract class AbstractMinecartEntity extends Entity implements net.minecraftforge.common.extensions.IForgeEntityMinecart {
    private static final DataParameter<Integer> DATA_ID_HURT = EntityDataManager.defineId(AbstractMinecartEntity.class, DataSerializers.INT);
    private static final DataParameter<Integer> DATA_ID_HURTDIR = EntityDataManager.defineId(AbstractMinecartEntity.class, DataSerializers.INT);
    private static final DataParameter<Float> DATA_ID_DAMAGE = EntityDataManager.defineId(AbstractMinecartEntity.class, DataSerializers.FLOAT);
@@ -77,6 +_,7 @@
       p_226574_0_.put(RailShape.NORTH_WEST, Pair.of(vector3i2, vector3i));
       p_226574_0_.put(RailShape.NORTH_EAST, Pair.of(vector3i2, vector3i1));
    });
+   private static net.minecraftforge.common.IMinecartCollisionHandler COLLISIONS = null;
    private int lSteps;
    private double lx;
    private double ly;
@@ -89,6 +_,7 @@
    private double lyd;
    @OnlyIn(Dist.CLIENT)
    private double lzd;
+   private boolean canBePushed = true;
 
    protected AbstractMinecartEntity(EntityType<?> p_i48538_1_, World p_i48538_2_) {
       super(p_i48538_1_, p_i48538_2_);
@@ -103,6 +_,14 @@
       this.yo = p_i48539_5_;
       this.zo = p_i48539_7_;
    }
+   
+   public net.minecraftforge.common.IMinecartCollisionHandler getCollisionHandler() {
+      return COLLISIONS;
+   }
+
+   public static void registerCollisionHandler(@Nullable net.minecraftforge.common.IMinecartCollisionHandler handler) {
+      COLLISIONS = handler;
+   }
 
    public static AbstractMinecartEntity createMinecart(World p_184263_0_, double p_184263_1_, double p_184263_3_, double p_184263_5_, AbstractMinecartEntity.Type p_184263_7_) {
       if (p_184263_7_ == AbstractMinecartEntity.Type.CHEST) {
@@ -138,7 +_,7 @@
    }
 
    public boolean isPushable() {
-      return true;
+      return canBePushed;
    }
 
    protected Vector3d getRelativePortalPosition(Direction.Axis p_241839_1_, TeleportationRepositioner.Result p_241839_2_) {
@@ -307,9 +_,9 @@
 
          BlockPos blockpos = new BlockPos(i, j, k);
          BlockState blockstate = this.level.getBlockState(blockpos);
-         if (AbstractRailBlock.isRail(blockstate)) {
+         if (canUseRail() && AbstractRailBlock.isRail(blockstate)) {
             this.moveAlongTrack(blockpos, blockstate);
-            if (blockstate.is(Blocks.ACTIVATOR_RAIL)) {
+            if (blockstate.getBlock() instanceof PoweredRailBlock && ((PoweredRailBlock) blockstate.getBlock()).isActivatorRail()) {
                this.activateMinecart(i, j, k, blockstate.getValue(PoweredRailBlock.POWERED));
             }
          } else {
@@ -334,8 +_,11 @@
          }
 
          this.setRot(this.yRot, this.xRot);
-         if (this.getMinecartType() == AbstractMinecartEntity.Type.RIDEABLE && getHorizontalDistanceSqr(this.getDeltaMovement()) > 0.01D) {
-            List<Entity> list = this.level.getEntities(this, this.getBoundingBox().inflate((double)0.2F, 0.0D, (double)0.2F), EntityPredicates.pushableBy(this));
+         AxisAlignedBB box;
+         if (getCollisionHandler() != null) box = getCollisionHandler().getMinecartCollisionBox(this);
+         else                               box = this.getBoundingBox().inflate(0.2F, 0.0D, 0.2F);
+         if (canBeRidden() && getHorizontalDistanceSqr(this.getDeltaMovement()) > 0.01D) {
+            List<Entity> list = this.level.getEntities(this, box, EntityPredicates.pushableBy(this));
             if (!list.isEmpty()) {
                for(int l = 0; l < list.size(); ++l) {
                   Entity entity1 = list.get(l);
@@ -347,7 +_,7 @@
                }
             }
          } else {
-            for(Entity entity : this.level.getEntities(this, this.getBoundingBox().inflate((double)0.2F, 0.0D, (double)0.2F))) {
+            for(Entity entity : this.level.getEntities(this, box)) {
                if (!this.hasPassenger(entity) && entity.isPushable() && entity instanceof AbstractMinecartEntity) {
                   entity.push(this);
                }
@@ -372,16 +_,23 @@
    }
 
    protected void comeOffTrack() {
-      double d0 = this.getMaxSpeed();
+      double d0 = onGround ? this.getMaxSpeed() : getMaxSpeedAirLateral();
       Vector3d vector3d = this.getDeltaMovement();
       this.setDeltaMovement(MathHelper.clamp(vector3d.x, -d0, d0), vector3d.y, MathHelper.clamp(vector3d.z, -d0, d0));
       if (this.onGround) {
          this.setDeltaMovement(this.getDeltaMovement().scale(0.5D));
       }
 
+      if (getMaxSpeedAirVertical() > 0 && getDeltaMovement().y > getMaxSpeedAirVertical()) {
+          if(Math.abs(getDeltaMovement().x) < 0.3f && Math.abs(getDeltaMovement().z) < 0.3f)
+              setDeltaMovement(new Vector3d(getDeltaMovement().x, 0.15f, getDeltaMovement().z));
+          else
+              setDeltaMovement(new Vector3d(getDeltaMovement().x, getMaxSpeedAirVertical(), getDeltaMovement().z));
+      }
+
       this.move(MoverType.SELF, this.getDeltaMovement());
       if (!this.onGround) {
-         this.setDeltaMovement(this.getDeltaMovement().scale(0.95D));
+         this.setDeltaMovement(this.getDeltaMovement().scale(getDragAir()));
       }
 
    }
@@ -396,29 +_,29 @@
       boolean flag = false;
       boolean flag1 = false;
       AbstractRailBlock abstractrailblock = (AbstractRailBlock)p_180460_2_.getBlock();
-      if (abstractrailblock == Blocks.POWERED_RAIL) {
+      if (abstractrailblock instanceof PoweredRailBlock && !((PoweredRailBlock) abstractrailblock).isActivatorRail()) {
          flag = p_180460_2_.getValue(PoweredRailBlock.POWERED);
          flag1 = !flag;
       }
 
       double d3 = 0.0078125D;
       Vector3d vector3d1 = this.getDeltaMovement();
-      RailShape railshape = p_180460_2_.getValue(abstractrailblock.getShapeProperty());
+      RailShape railshape = ((AbstractRailBlock)p_180460_2_.getBlock()).getRailDirection(p_180460_2_, this.level, p_180460_1_, this);
       switch(railshape) {
       case ASCENDING_EAST:
-         this.setDeltaMovement(vector3d1.add(-0.0078125D, 0.0D, 0.0D));
+         this.setDeltaMovement(vector3d1.add(-1 * getSlopeAdjustment(), 0.0D, 0.0D));
          ++d1;
          break;
       case ASCENDING_WEST:
-         this.setDeltaMovement(vector3d1.add(0.0078125D, 0.0D, 0.0D));
+         this.setDeltaMovement(vector3d1.add(getSlopeAdjustment(), 0.0D, 0.0D));
          ++d1;
          break;
       case ASCENDING_NORTH:
-         this.setDeltaMovement(vector3d1.add(0.0D, 0.0D, 0.0078125D));
+         this.setDeltaMovement(vector3d1.add(0.0D, 0.0D, getSlopeAdjustment()));
          ++d1;
          break;
       case ASCENDING_SOUTH:
-         this.setDeltaMovement(vector3d1.add(0.0D, 0.0D, -0.0078125D));
+         this.setDeltaMovement(vector3d1.add(0.0D, 0.0D, -1 * getSlopeAdjustment()));
          ++d1;
       }
 
@@ -449,7 +_,7 @@
          }
       }
 
-      if (flag1) {
+      if (flag1 && shouldDoRailFunctions()) {
          double d22 = Math.sqrt(getHorizontalDistanceSqr(this.getDeltaMovement()));
          if (d22 < 0.03D) {
             this.setDeltaMovement(Vector3d.ZERO);
@@ -478,10 +_,7 @@
       d0 = d23 + d4 * d14;
       d2 = d10 + d5 * d14;
       this.setPos(d0, d1, d2);
-      double d24 = this.isVehicle() ? 0.75D : 1.0D;
-      double d25 = this.getMaxSpeed();
-      vector3d1 = this.getDeltaMovement();
-      this.move(MoverType.SELF, new Vector3d(MathHelper.clamp(d24 * vector3d1.x, -d25, d25), 0.0D, MathHelper.clamp(d24 * vector3d1.z, -d25, d25)));
+      this.moveMinecartOnRail(p_180460_1_);
       if (vector3i.getY() != 0 && MathHelper.floor(this.getX()) - p_180460_1_.getX() == vector3i.getX() && MathHelper.floor(this.getZ()) - p_180460_1_.getZ() == vector3i.getZ()) {
          this.setPos(this.getX(), this.getY() + (double)vector3i.getY(), this.getZ());
       } else if (vector3i1.getY() != 0 && MathHelper.floor(this.getX()) - p_180460_1_.getX() == vector3i1.getX() && MathHelper.floor(this.getZ()) - p_180460_1_.getZ() == vector3i1.getZ()) {
@@ -509,7 +_,10 @@
          this.setDeltaMovement(d26 * (double)(j - p_180460_1_.getX()), vector3d5.y, d26 * (double)(i - p_180460_1_.getZ()));
       }
 
-      if (flag) {
+      if (shouldDoRailFunctions())
+          ((AbstractRailBlock)p_180460_2_.getBlock()).onMinecartPass(p_180460_2_, level, p_180460_1_, this);
+
+      if (flag && shouldDoRailFunctions()) {
          Vector3d vector3d6 = this.getDeltaMovement();
          double d27 = Math.sqrt(getHorizontalDistanceSqr(vector3d6));
          if (d27 > 0.01D) {
@@ -564,7 +_,7 @@
 
       BlockState blockstate = this.level.getBlockState(new BlockPos(i, j, k));
       if (AbstractRailBlock.isRail(blockstate)) {
-         RailShape railshape = blockstate.getValue(((AbstractRailBlock)blockstate.getBlock()).getShapeProperty());
+         RailShape railshape = ((AbstractRailBlock)blockstate.getBlock()).getRailDirection(blockstate, this.level, new BlockPos(i, j, k), this);
          p_70495_3_ = (double)j;
          if (railshape.isAscending()) {
             p_70495_3_ = (double)(j + 1);
@@ -603,7 +_,7 @@
 
       BlockState blockstate = this.level.getBlockState(new BlockPos(i, j, k));
       if (AbstractRailBlock.isRail(blockstate)) {
-         RailShape railshape = blockstate.getValue(((AbstractRailBlock)blockstate.getBlock()).getShapeProperty());
+         RailShape railshape = ((AbstractRailBlock)blockstate.getBlock()).getRailDirection(blockstate, this.level, new BlockPos(i, j, k), this);
          Pair<Vector3i, Vector3i> pair = exits(railshape);
          Vector3i vector3i = pair.getFirst();
          Vector3i vector3i1 = pair.getSecond();
@@ -666,6 +_,11 @@
    }
 
    public void push(Entity p_70108_1_) {
+      //net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.minecart.MinecartCollisionEvent(this, entityIn));
+      if (getCollisionHandler() != null) {
+         getCollisionHandler().onEntityCollision(this, p_70108_1_);
+         return;
+      }
       if (!this.level.isClientSide) {
          if (!p_70108_1_.noPhysics && !this.noPhysics) {
             if (!this.hasPassenger(p_70108_1_)) {
@@ -701,11 +_,11 @@
 
                      Vector3d vector3d2 = this.getDeltaMovement();
                      Vector3d vector3d3 = p_70108_1_.getDeltaMovement();
-                     if (((AbstractMinecartEntity)p_70108_1_).getMinecartType() == AbstractMinecartEntity.Type.FURNACE && this.getMinecartType() != AbstractMinecartEntity.Type.FURNACE) {
+                     if (((AbstractMinecartEntity)p_70108_1_).isPoweredCart() && !this.isPoweredCart()) {
                         this.setDeltaMovement(vector3d2.multiply(0.2D, 1.0D, 0.2D));
                         this.push(vector3d3.x - d0, 0.0D, vector3d3.z - d1);
                         p_70108_1_.setDeltaMovement(vector3d3.multiply(0.95D, 1.0D, 0.95D));
-                     } else if (((AbstractMinecartEntity)p_70108_1_).getMinecartType() != AbstractMinecartEntity.Type.FURNACE && this.getMinecartType() == AbstractMinecartEntity.Type.FURNACE) {
+                     } else if (!((AbstractMinecartEntity)p_70108_1_).isPoweredCart() && this.isPoweredCart()) {
                         p_70108_1_.setDeltaMovement(vector3d3.multiply(0.2D, 1.0D, 0.2D));
                         p_70108_1_.push(vector3d2.x + d0, 0.0D, vector3d2.z + d1);
                         this.setDeltaMovement(vector3d2.multiply(0.95D, 1.0D, 0.95D));
@@ -809,6 +_,40 @@
 
    public IPacket<?> getAddEntityPacket() {
       return new SSpawnObjectPacket(this);
+   }
+
+   private boolean canUseRail = true;
+   @Override public boolean canUseRail() { return canUseRail; }
+   @Override public void setCanUseRail(boolean value) { this.canUseRail = value; }
+   private float currentSpeedCapOnRail = getMaxCartSpeedOnRail();
+   @Override public float getCurrentCartSpeedCapOnRail() { return currentSpeedCapOnRail; }
+   @Override public void setCurrentCartSpeedCapOnRail(float value) { currentSpeedCapOnRail = Math.min(value, getMaxCartSpeedOnRail()); }
+   private float maxSpeedAirLateral = DEFAULT_MAX_SPEED_AIR_LATERAL;
+   @Override public float getMaxSpeedAirLateral() { return maxSpeedAirLateral; }
+   @Override public void setMaxSpeedAirLateral(float value) { maxSpeedAirLateral = value; }
+   private float maxSpeedAirVertical = DEFAULT_MAX_SPEED_AIR_VERTICAL;
+   @Override public float getMaxSpeedAirVertical() { return maxSpeedAirVertical; }
+   @Override public void setMaxSpeedAirVertical(float value) { maxSpeedAirVertical = value; }
+   private double dragAir = DEFAULT_AIR_DRAG;
+   @Override public double getDragAir() { return dragAir; }
+   @Override public void setDragAir(double value) { dragAir = value; }
+   @Override
+   public double getMaxSpeedWithRail() { //Non-default because getMaximumSpeed is protected
+      if (!canUseRail()) return getMaxSpeed();
+      BlockPos pos = this.getCurrentRailPosition();
+      BlockState state = getMinecart().level.getBlockState(pos);
+      if (!state.is(BlockTags.RAILS)) return getMaxSpeed();
+
+      float railMaxSpeed = ((AbstractRailBlock)state.getBlock()).getRailMaxSpeed(state, getMinecart().level, pos, getMinecart());
+      return Math.min(railMaxSpeed, getCurrentCartSpeedCapOnRail());
+   }
+   @Override
+   public void moveMinecartOnRail(BlockPos pos) { //Non-default because getMaximumSpeed is protected
+      AbstractMinecartEntity mc = getMinecart();
+      double d24 = mc.isVehicle() ? 0.75D : 1.0D;
+      double d25 = mc.getMaxSpeedWithRail();
+      Vector3d vec3d1 = mc.getDeltaMovement();
+      mc.move(MoverType.SELF, new Vector3d(MathHelper.clamp(d24 * vec3d1.x, -d25, d25), 0.0D, MathHelper.clamp(d24 * vec3d1.z, -d25, d25)));
    }
 
    public static enum Type {
