--- a/net/minecraft/entity/item/minecart/CommandBlockMinecartEntity.java
+++ b/net/minecraft/entity/item/minecart/CommandBlockMinecartEntity.java
@@ -73,6 +_,8 @@
    }
 
    public ActionResultType interact(PlayerEntity p_184230_1_, Hand p_184230_2_) {
+      ActionResultType ret = super.interact(p_184230_1_, p_184230_2_);
+      if (ret.consumesAction()) return ret;
       return this.commandBlock.usedBy(p_184230_1_);
    }
 
