--- a/net/minecraft/entity/item/minecart/MinecartEntity.java
+++ b/net/minecraft/entity/item/minecart/MinecartEntity.java
@@ -16,6 +16,8 @@
    }
 
    public ActionResultType func_184230_a(PlayerEntity p_184230_1_, Hand p_184230_2_) {
+      ActionResultType ret = super.func_184230_a(p_184230_1_, p_184230_2_);
+      if (ret.func_226246_a_()) return ret;
       if (p_184230_1_.func_226563_dT_()) {
          return ActionResultType.PASS;
       } else if (this.func_184207_aI()) {
