--- ../src-base/minecraft_server/net/minecraft/src/RConThreadClient.java	0000-00-00 00:00:00.000000000 -0000
+++ ../src-work/minecraft_server/net/minecraft/src/RConThreadClient.java	0000-00-00 00:00:00.000000000 -0000
@@ -30,7 +30,7 @@
             {
                 if (!this.field_40419_a)
                 {
-                    break;
+                    return;
                 }
 
                 try
@@ -113,7 +113,7 @@
             catch (Exception var19)
             {
                 System.out.println(var19);
-                break;
+                return;
             }
             finally
             {
