--- ../src-base/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java
+++ ../src-work/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java
@@ -167,6 +167,12 @@
                     }
                 }
 
+                boolean flag = block1.removedByPlayer(world, pos, mc.thePlayer, false);
+
+                if (flag)
+                {
+                    block1.onBlockDestroyedByPlayer(world, pos, iblockstate);
+                }
                 return flag;
             }
         }
