--- ../src-base/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java
+++ ../src-work/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java
@@ -162,11 +162,18 @@
 
                         if (itemstack1.stackSize == 0)
                         {
+                        	net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this.mc.thePlayer, itemstack1, EnumHand.MAIN_HAND);
                             this.mc.thePlayer.func_184611_a(EnumHand.MAIN_HAND, (ItemStack)null);
                         }
                     }
                 }
 
+                boolean flag = block1.removedByPlayer(world, pos, mc.thePlayer, false);
+
+                if (flag)
+                {
+                    block1.onBlockDestroyedByPlayer(world, pos, iblockstate);
+                }
                 return flag;
             }
         }
@@ -453,9 +460,10 @@
                 {
                     p_187101_1_.func_184611_a(p_187101_4_, itemstack);
 
-                    if (itemstack.stackSize == 0)
+                    if (itemstack.stackSize <= 0)
                     {
                         p_187101_1_.func_184611_a(p_187101_4_, (ItemStack)null);
+                        net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(p_187101_1_, itemstack, p_187101_4_);
                     }
                 }
 
