--- a/net/minecraft/util/datafix/schemas/V2832.java
+++ b/net/minecraft/util/datafix/schemas/V2832.java
@@ -52,7 +_,7 @@
                         DSL.constType(namespacedString()),
                         DSL.fields(
                             "generator",
-                            DSL.taggedChoiceLazy(
+                            DSL.or(DSL.taggedChoiceLazy(
                                 "type",
                                 DSL.string(),
                                 ImmutableMap.of(
@@ -97,7 +_,7 @@
                                             )
                                         )
                                 )
-                            )
+                            ), DSL.remainder()) // Keep data if the generator type can't be found
                         )
                     )
                 )
