--- a/net/minecraft/util/text/LanguageMap.java
+++ b/net/minecraft/util/text/LanguageMap.java
@@ -36,6 +36,7 @@
             this.field_74816_c.put(entry.getKey(), s);
          }
 
+         net.minecraftforge.fml.server.LanguageHook.captureLanguageMap(this.field_74816_c);
          this.field_150511_e = Util.func_211177_b();
       } catch (JsonParseException | IOException ioexception) {
          field_201045_a.error("Couldn't read strings from /assets/minecraft/lang/en_us.json", (Throwable)ioexception);
