--- a/net/minecraft/util/text/LanguageMap.java
+++ b/net/minecraft/util/text/LanguageMap.java
@@ -40,7 +40,8 @@
          field_201045_a.error("Couldn't read strings from /assets/minecraft/lang/en_us.json", (Throwable)ioexception);
       }
 
-      final Map<String, String> map = builder.build();
+      final Map<String, String> map = new java.util.HashMap<>(builder.build());
+      net.minecraftforge.fml.server.LanguageHook.captureLanguageMap(map);
       return new LanguageMap() {
          public String func_230503_a_(String p_230503_1_) {
             return map.getOrDefault(p_230503_1_, p_230503_1_);
@@ -63,6 +64,11 @@
                }, Style.field_240709_b_).isPresent();
             };
          }
+         
+         @Override
+         public Map<String, String> getLanguageData() {
+            return map;
+         }
       };
    }
 
@@ -84,6 +90,9 @@
    public static void func_240594_a_(LanguageMap p_240594_0_) {
       field_240592_d_ = p_240594_0_;
    }
+   
+   // FORGE START
+   public Map<String, String> getLanguageData() { return ImmutableMap.of(); }
 
    public abstract String func_230503_a_(String p_230503_1_);
 
