--- a/net/minecraft/util/text/Style.java
+++ b/net/minecraft/util/text/Style.java
@@ -125,6 +_,18 @@
       return new Style(this.color, this.bold, this.italic, p_244282_1_, this.strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
    }
 
+   public Style setUnderlined(@Nullable Boolean underlined) {
+      return new Style(this.color, this.bold, this.italic, underlined, this.strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
+   public Style setStrikethrough(@Nullable Boolean strikethrough) {
+      return new Style(this.color, this.bold, this.italic, this.underlined, strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
+   public Style setObfuscated(@Nullable Boolean obfuscated) {
+      return new Style(this.color, this.bold, this.italic, this.underlined, this.strikethrough, obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
    public Style withClickEvent(@Nullable ClickEvent p_240715_1_) {
       return new Style(this.color, this.bold, this.italic, this.underlined, this.strikethrough, this.obfuscated, p_240715_1_, this.hoverEvent, this.insertion, this.font);
    }
