--- a/net/minecraft/util/text/Style.java
+++ b/net/minecraft/util/text/Style.java
@@ -125,6 +_,18 @@
       return new Style(this.field_150247_b, this.field_150248_c, this.field_150245_d, p_244282_1_, this.field_150243_f, this.field_150244_g, this.field_150251_h, this.field_150252_i, this.field_179990_j, this.field_240710_l_);
    }
 
+   public Style setUnderlined(@Nullable Boolean underlined) {
+      return new Style(this.field_150247_b, this.field_150248_c, this.field_150245_d, underlined, this.field_150243_f, this.field_150244_g, this.field_150251_h, this.field_150252_i, this.field_179990_j, this.field_240710_l_);
+   }
+
+   public Style setStrikethrough(@Nullable Boolean strikethrough) {
+      return new Style(this.field_150247_b, this.field_150248_c, this.field_150245_d, this.field_150246_e, strikethrough, this.field_150244_g, this.field_150251_h, this.field_150252_i, this.field_179990_j, this.field_240710_l_);
+   }
+
+   public Style setObfuscated(@Nullable Boolean obfuscated) {
+      return new Style(this.field_150247_b, this.field_150248_c, this.field_150245_d, this.field_150246_e, this.field_150243_f, obfuscated, this.field_150251_h, this.field_150252_i, this.field_179990_j, this.field_240710_l_);
+   }
+
    public Style func_240715_a_(@Nullable ClickEvent p_240715_1_) {
       return new Style(this.field_150247_b, this.field_150248_c, this.field_150245_d, this.field_150246_e, this.field_150243_f, this.field_150244_g, p_240715_1_, this.field_150252_i, this.field_179990_j, this.field_240710_l_);
    }
