--- a/net/minecraft/client/gui/screen/AddServerScreen.java
+++ b/net/minecraft/client/gui/screen/AddServerScreen.java
@@ -53,7 +53,7 @@
    protected void init() {
       this.minecraft.field_195559_v.func_197967_a(true);
       this.field_146309_g = new TextFieldWidget(this.font, this.width / 2 - 100, 66, 200, 20, I18n.func_135052_a("addServer.enterName"));
-      this.field_146309_g.func_146195_b(true);
+      this.field_146309_g.setFocused(true);
       this.field_146309_g.func_146180_a(this.field_146311_h.field_78847_a);
       this.field_146309_g.func_212954_a(this::func_213028_a);
       this.children.add(this.field_146309_g);
