--- a/net/minecraft/client/gui/screen/ChatScreen.java
+++ b/net/minecraft/client/gui/screen/ChatScreen.java
@@ -89,7 +89,7 @@
          }
       } else {
          String s = this.field_146415_a.func_146179_b().trim();
-         if (!s.isEmpty()) {
+         if (!s.isEmpty() && net.minecraftforge.client.ForgeHooksClient.sendMessage(this.field_228174_e_.field_228106_o_)) {
             this.sendMessage(s);
          }
 
