--- a/net/minecraft/client/gui/screen/ServerListScreen.java
+++ b/net/minecraft/client/gui/screen/ServerListScreen.java
@@ -37,7 +37,7 @@
       }));
       this.field_146302_g = new TextFieldWidget(this.font, this.width / 2 - 100, 116, 200, 20, I18n.func_135052_a("addServer.enterIp"));
       this.field_146302_g.func_146203_f(128);
-      this.field_146302_g.func_146195_b(true);
+      this.field_146302_g.setFocused(true);
       this.field_146302_g.func_146180_a(this.minecraft.field_71474_y.field_74332_R);
       this.field_146302_g.func_212954_a((p_213024_1_) -> {
          this.func_195168_i();
