--- ../src-base/minecraft/net/minecraft/network/play/server/S3FPacketCustomPayload.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/S3FPacketCustomPayload.java
@@ -46,7 +46,11 @@
     public void func_148840_b(PacketBuffer p_148840_1_) throws IOException
     {
         p_148840_1_.func_180714_a(this.field_149172_a);
+        synchronized(this.field_149171_b) { //This may be access multiple times, from multiple threads, lets be safe.
+        this.field_149171_b.markReaderIndex();
         p_148840_1_.writeBytes((ByteBuf)this.field_149171_b);
+        this.field_149171_b.resetReaderIndex();
+        }
     }
 
     public void func_148833_a(INetHandlerPlayClient p_148833_1_)
