--- a/net/minecraft/network/play/server/SCommandListPacket.java
+++ b/net/minecraft/network/play/server/SCommandListPacket.java
@@ -119,6 +119,9 @@
          String s = p_197695_0_.func_150789_c(32767);
          ArgumentType<?> argumenttype = ArgumentTypes.func_197486_a(p_197695_0_);
          if (argumenttype == null) {
+            if ((i & 16) != 0) { // FORGE: Flush unused suggestion data
+               p_197695_0_.func_192575_l();
+            }
             return null;
          } else {
             RequiredArgumentBuilder<ISuggestionProvider, ?> requiredargumentbuilder = RequiredArgumentBuilder.argument(s, argumenttype);
