--- a/net/minecraft/network/play/server/SCustomPayloadPlayPacket.java
+++ b/net/minecraft/network/play/server/SCustomPayloadPlayPacket.java
@@ -8,7 +8,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;

-public class SCustomPayloadPlayPacket implements IPacket<IClientPlayNetHandler> {
+public class SCustomPayloadPlayPacket implements IPacket<IClientPlayNetHandler>, net.minecraftforge.fml.network.ICustomPacket<SCustomPayloadPlayPacket> {
    public static final ResourceLocation field_209911_b = new ResourceLocation("brand");
    public static final ResourceLocation field_209913_d = new ResourceLocation("debug/path");
    public static final ResourceLocation field_209914_e = new ResourceLocation("debug/neighbors_update");
@@ -68,4 +68,24 @@
    public PacketBuffer func_180735_b() {
       return new PacketBuffer(this.field_149171_b.copy());
    }
+
+    @Override
+    public PacketBuffer getInternalData() {
+        return field_149171_b;
+    }
+
+    @Override
+    public void setData(PacketBuffer data) {
+        this.field_149171_b = data;
+    }
+
+    @Override
+    public ResourceLocation getName() {
+    return field_149172_a;
+    }
+
+    @Override
+    public void setName(ResourceLocation name) {
+    this.field_149172_a = name;
+    }
 }
