--- a/net/minecraft/network/play/server/SEntityPropertiesPacket.java
+++ b/net/minecraft/network/play/server/SEntityPropertiesPacket.java
@@ -54,10 +54,14 @@
    }
 
    public void func_148840_b(PacketBuffer p_148840_1_) throws IOException {
+      writePacketData(p_148840_1_, net.minecraftforge.fml.network.ConnectionType.MODDED);
+   }
+   public void writePacketData(PacketBuffer p_148840_1_, net.minecraftforge.fml.network.ConnectionType connectionType) throws IOException {
+      List<SEntityPropertiesPacket.Snapshot> snapshots = connectionType.isVanilla() ? this.field_149444_b.stream().filter(sn -> Registry.field_239692_aP_.func_177774_c(sn.func_240834_a_()).func_110624_b().equals("minecraft")).collect(java.util.stream.Collectors.toList()) : this.field_149444_b;
       p_148840_1_.func_150787_b(this.field_149445_a);
-      p_148840_1_.writeInt(this.field_149444_b.size());
+      p_148840_1_.writeInt(snapshots.size());
 
-      for(SEntityPropertiesPacket.Snapshot sentitypropertiespacket$snapshot : this.field_149444_b) {
+      for(SEntityPropertiesPacket.Snapshot sentitypropertiespacket$snapshot : snapshots) {
          p_148840_1_.func_192572_a(Registry.field_239692_aP_.func_177774_c(sentitypropertiespacket$snapshot.func_240834_a_()));
          p_148840_1_.writeDouble(sentitypropertiespacket$snapshot.func_151410_b());
          p_148840_1_.func_150787_b(sentitypropertiespacket$snapshot.func_151408_c().size());
