--- a/net/minecraft/network/play/server/SPacketCustomPayload.java
+++ b/net/minecraft/network/play/server/SPacketCustomPayload.java
@@ -8,7 +8,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public class SPacketCustomPayload implements Packet<INetHandlerPlayClient> {
+public class SPacketCustomPayload implements Packet<INetHandlerPlayClient>, net.minecraftforge.fml.network.ICustomPacket<SPacketCustomPayload> {
    public static final ResourceLocation field_209910_a = new ResourceLocation("minecraft:trader_list");
    public static final ResourceLocation field_209911_b = new ResourceLocation("minecraft:brand");
    public static final ResourceLocation field_209912_c = new ResourceLocation("minecraft:book_open");
