--- a/net/minecraft/network/play/server/STagsListPacket.java
+++ b/net/minecraft/network/play/server/STagsListPacket.java
@@ -19,7 +19,7 @@
    }
 
    public void func_148837_a(PacketBuffer p_148837_1_) throws IOException {
-      this.field_199859_a = NetworkTagManager.func_199714_b(p_148837_1_);
+      this.field_199859_a = NetworkTagManager.readSafe(p_148837_1_);
    }
 
    public void func_148840_b(PacketBuffer p_148840_1_) throws IOException {
