--- ../src-base/minecraft/net/minecraft/block/BlockBanner.java
+++ ../src-work/minecraft/net/minecraft/block/BlockBanner.java
@@ -104,14 +104,7 @@
 
     public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune)
     {
-        ItemStack itemstack = this.func_185549_e(worldIn, pos, state);
-
-        if (itemstack != null)
         {
-            spawnAsEntity(worldIn, pos, itemstack);
-        }
-        else
-        {
             super.dropBlockAsItemWithChance(worldIn, pos, state, chance, fortune);
         }
     }
@@ -138,6 +131,28 @@
         }
     }
 
+    @Override
+    public java.util.List<ItemStack> getDrops(IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
+        TileEntity te = world.getTileEntity(pos);
+
+        java.util.List<ItemStack> ret = new java.util.ArrayList<ItemStack>();
+        if (te instanceof TileEntityBanner)
+        {
+            TileEntityBanner banner = (TileEntityBanner)te;
+            ItemStack item = new ItemStack(Items.banner, 1, banner.getBaseColor());
+            NBTTagCompound nbt = new NBTTagCompound();
+            TileEntityBanner.func_181020_a(nbt, banner.getBaseColor(), banner.func_181021_d());
+            item.setTagInfo("BlockEntityTag", nbt);
+            ret.add(item);
+        }
+        else
+        {
+            ret.add(new ItemStack(Items.banner, 1, 0));
+        }
+        return ret;
+    }
+
     public static class BlockBannerHanging extends BlockBanner
         {
             protected static final AxisAlignedBB field_185551_d = new AxisAlignedBB(0.0D, 0.0D, 0.875D, 1.0D, 0.78125D, 1.0D);
