--- ../src-base/minecraft/net/minecraft/block/BlockGrass.java
+++ ../src-work/minecraft/net/minecraft/block/BlockGrass.java
@@ -96,7 +96,7 @@
             {
                 if (j >= i / 16)
                 {
-                    if (worldIn.getBlockState(blockpos1).getBlock().blockMaterial == Material.air)
+                    if (worldIn.isAirBlock(blockpos1))
                     {
                         if (rand.nextInt(8) == 0)
                         {
