--- ../src-base/minecraft/net/minecraft/block/BlockRedstoneWire.java
+++ ../src-work/minecraft/net/minecraft/block/BlockRedstoneWire.java
@@ -60,10 +60,10 @@
         BlockPos blockpos1 = pos.offset(direction);
         Block block = worldIn.getBlockState(pos.offset(direction)).getBlock();
 
-        if (!canConnectTo(worldIn.getBlockState(blockpos1), direction) && (block.isSolidFullCube() || !canConnectUpwardsTo(worldIn.getBlockState(blockpos1.down()))))
+        if (!canRestoneConnect(worldIn, blockpos1, direction) && (block.isSolidFullCube() || !canRestoneConnect(worldIn, blockpos1.down(), null)))
         {
             Block block1 = worldIn.getBlockState(pos.up()).getBlock();
-            return !block1.isSolidFullCube() && block.isSolidFullCube() && canConnectUpwardsTo(worldIn.getBlockState(blockpos1.up())) ? BlockRedstoneWire.EnumAttachPosition.UP : BlockRedstoneWire.EnumAttachPosition.NONE;
+            return !block1.isSolidFullCube() && block.isSolidFullCube() && canRestoneConnect(worldIn, blockpos1.up(), null) ? BlockRedstoneWire.EnumAttachPosition.UP : BlockRedstoneWire.EnumAttachPosition.NONE;
         }
         else
         {
@@ -393,35 +393,25 @@
         Block block = iblockstate.getBlock();
         boolean flag = block.isNormalCube();
         boolean flag1 = worldIn.getBlockState(pos.up()).getBlock().isNormalCube();
-        return !flag1 && flag && canConnectUpwardsTo(worldIn, blockpos1.up()) ? true : (canConnectTo(iblockstate, side) ? true : (block == Blocks.powered_repeater && iblockstate.getValue(BlockRedstoneDiode.FACING) == side ? true : !flag && canConnectUpwardsTo(worldIn, blockpos1.down())));
+        return !flag1 && flag && canRestoneConnect(worldIn, blockpos1.up(), null) ? true : (canRestoneConnect(worldIn, blockpos1, side) ? true : (block == Blocks.powered_repeater && iblockstate.getValue(BlockRedstoneDiode.FACING) == side ? true : !flag && canRestoneConnect(worldIn, blockpos1.down(), null)));
     }
 
-    protected static boolean canConnectUpwardsTo(IBlockAccess worldIn, BlockPos pos)
+    protected static boolean canRestoneConnect(IBlockAccess world, BlockPos pos, EnumFacing side)
     {
-        return canConnectUpwardsTo(worldIn.getBlockState(pos));
-    }
+        IBlockState state = world.getBlockState(pos);
 
-    protected static boolean canConnectUpwardsTo(IBlockState state)
-    {
-        return canConnectTo(state, (EnumFacing)null);
-    }
-
-    protected static boolean canConnectTo(IBlockState blockState, EnumFacing side)
-    {
-        Block block = blockState.getBlock();
-
-        if (block == Blocks.redstone_wire)
+        if (state.getBlock() == Blocks.redstone_wire)
         {
             return true;
         }
-        else if (Blocks.unpowered_repeater.isAssociated(block))
+        else if (Blocks.unpowered_repeater.isAssociated(state.getBlock()))
         {
-            EnumFacing enumfacing1 = (EnumFacing)blockState.getValue(BlockRedstoneRepeater.FACING);
-            return enumfacing1 == side || enumfacing1.getOpposite() == side;
+            EnumFacing direction = (EnumFacing)state.getValue(BlockRedstoneRepeater.FACING);
+            return direction == side || direction.getOpposite() == side;
         }
         else
         {
-            return block.canProvidePower() && side != null;
+            return state.getBlock().canConnectRedstone(world, pos, side);
         }
     }
 
