--- ../src-base/minecraft/net/minecraft/block/BlockPistonBase.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPistonBase.java
@@ -200,7 +200,7 @@
                     }
                 }
 
-                if (!flag1 && block.getMaterial() != Material.air && canPush(block, worldIn, blockpos1, enumfacing.getOpposite(), false) && (block.getMobilityFlag() == 0 || block == Blocks.piston || block == Blocks.sticky_piston))
+                if (!flag1 && !block.isAir(worldIn, blockpos1) && canPush(block, worldIn, blockpos1, enumfacing.getOpposite(), false) && (block.getMobilityFlag() == 0 || block == Blocks.piston || block == Blocks.sticky_piston))
                 {
                     this.doMove(worldIn, pos, enumfacing, false);
                 }
@@ -344,7 +344,7 @@
                     return false;
                 }
 
-                return !(blockIn instanceof ITileEntityProvider);
+                return !(blockIn.hasTileEntity(worldIn.getBlockState(pos)));
             }
             else
             {
@@ -384,7 +384,9 @@
             {
                 blockpos1 = (BlockPos)list1.get(j);
                 Block block = worldIn.getBlockState(blockpos1).getBlock();
-                block.dropBlockAsItem(worldIn, blockpos1, worldIn.getBlockState(blockpos1), 0);
+                //With our change to how snowballs are dropped this needs to disallow to mimic vanilla behavior.
+                float chance = block instanceof BlockSnow ? -1.0f : 1.0f;
+                block.dropBlockAsItemWithChance(worldIn, blockpos1, worldIn.getBlockState(blockpos1), chance, 0);
                 worldIn.setBlockToAir(blockpos1);
                 --i;
                 ablock[i] = block;
