--- ../src-base/minecraft/net/minecraft/block/BlockPistonMoving.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPistonMoving.java
@@ -190,4 +190,16 @@
     {
         return new BlockStateContainer(this, new IProperty[] {FACING, TYPE});
     }
+
+    @Override
+    public java.util.List<net.minecraft.item.ItemStack> getDrops(IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
+        TileEntityPiston tileentitypiston = this.getTileEntity(world, pos);
+        if (tileentitypiston != null)
+        {
+            IBlockState pushed = tileentitypiston.getPistonState();
+            return pushed.getBlock().getDrops(world, pos, pushed, fortune);
+        }
+        return new java.util.ArrayList();
+    }
 }
