--- ../src-base/minecraft/net/minecraft/block/BlockLog.java
+++ ../src-work/minecraft/net/minecraft/block/BlockLog.java
@@ -38,9 +38,9 @@
                 BlockPos blockpos1 = (BlockPos)iterator.next();
                 IBlockState iblockstate1 = worldIn.getBlockState(blockpos1);
 
-                if (iblockstate1.getBlock().getMaterial() == Material.leaves && !((Boolean)iblockstate1.getValue(BlockLeaves.CHECK_DECAY)).booleanValue())
+                if (iblockstate1.getBlock().isLeaves(worldIn, blockpos1))
                 {
-                    worldIn.setBlockState(blockpos1, iblockstate1.withProperty(BlockLeaves.CHECK_DECAY, Boolean.valueOf(true)), 4);
+                    iblockstate1.getBlock().beginLeavesDecay(worldIn, blockpos1);
                 }
             }
         }
@@ -51,6 +51,9 @@
         return super.onBlockPlaced(worldIn, pos, facing, hitX, hitY, hitZ, meta, placer).withProperty(LOG_AXIS, BlockLog.EnumAxis.fromFacingAxis(facing.getAxis()));
     }
 
+    @Override public boolean canSustainLeaves(net.minecraft.world.IBlockAccess world, BlockPos pos){ return true; }
+    @Override public boolean isWood(net.minecraft.world.IBlockAccess world, BlockPos pos){ return true; }
+
     public static enum EnumAxis implements IStringSerializable
     {
         X("x"),
