--- a/net/minecraft/data/registries/RegistryPatchGenerator.java
+++ b/net/minecraft/data/registries/RegistryPatchGenerator.java
@@ -17,7 +_,7 @@
       return p_310881_.thenApply((p_309945_) -> {
          RegistryAccess.Frozen registryaccess$frozen = RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY);
          Cloner.Factory cloner$factory = new Cloner.Factory();
-         RegistryDataLoader.WORLDGEN_REGISTRIES.forEach((p_313050_) -> {
+         RegistryDataLoader.getWorldGenAndDimensionStream().forEach((p_313050_) -> {
             p_313050_.runWithArguments(cloner$factory::addCodec);
          });
          RegistrySetBuilder.PatchedRegistries registrysetbuilder$patchedregistries = p_310262_.buildPatch(registryaccess$frozen, p_309945_, cloner$factory);
