--- a/net/minecraft/data/registries/RegistriesDatapackGenerator.java
+++ b/net/minecraft/data/registries/RegistriesDatapackGenerator.java
@@ -31,7 +_,7 @@
    public CompletableFuture<?> m_213708_(CachedOutput p_255785_) {
       return this.f_254747_.thenCompose((p_256533_) -> {
          DynamicOps<JsonElement> dynamicops = RegistryOps.m_255058_(JsonOps.INSTANCE, p_256533_);
-         return CompletableFuture.allOf(RegistryDataLoader.f_243803_.stream().flatMap((p_256552_) -> {
+         return CompletableFuture.allOf(net.minecraftforge.registries.DataPackRegistriesHooks.getDataPackRegistries().stream().flatMap((p_256552_) -> {
             return this.m_254918_(p_255785_, p_256533_, dynamicops, p_256552_).stream();
          }).toArray((p_255809_) -> {
             return new CompletableFuture[p_255809_];
