--- ../src_base/minecraft/net/minecraft/dispenser/BehaviorDefaultDispenseItem.java
+++ ../src_work/minecraft/net/minecraft/dispenser/BehaviorDefaultDispenseItem.java
@@ -5,6 +5,10 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.world.World;
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
+// CraftBukkit end
 
 public class BehaviorDefaultDispenseItem implements IBehaviorDispenseItem
 {
@@ -27,10 +31,16 @@
         EnumFacing enumfacing = BlockDispenser.getFacing(par1IBlockSource.getBlockMetadata());
         IPosition iposition = BlockDispenser.getIPositionFromBlockSource(par1IBlockSource);
         ItemStack itemstack1 = par2ItemStack.splitStack(1);
-        doDispense(par1IBlockSource.getWorld(), itemstack1, 6, enumfacing, iposition);
+        // CraftBukkit start
+        if (!doDispense(par1IBlockSource.getWorld(), itemstack1, 6, enumfacing, par1IBlockSource))
+        {
+            par2ItemStack.stackSize++;
+        }
+        // CraftBukkit end
         return par2ItemStack;
     }
 
+    // MCPC+ start - vanilla compatibility
     public static void doDispense(World par0World, ItemStack par1ItemStack, int par2, EnumFacing par3EnumFacing, IPosition par4IPosition)
     {
         double d0 = par4IPosition.getX();
@@ -45,6 +55,67 @@
         entityitem.motionY += par0World.rand.nextGaussian() * 0.007499999832361937D * (double)par2;
         entityitem.motionZ += par0World.rand.nextGaussian() * 0.007499999832361937D * (double)par2;
         par0World.spawnEntityInWorld(entityitem);
+        // TODO: add CB event?
+    }
+    // MCPC+ end
+
+    // CraftBukkit start - void -> boolean return, IPosition -> ISourceBlock last argument
+    public static boolean doDispense(World par0World, ItemStack par1ItemStack, int par2, EnumFacing par3EnumFacing, IBlockSource par4IPosition)
+    {
+        IPosition iposition = BlockDispenser.getIPositionFromBlockSource(par4IPosition);
+        // CraftBukkit end
+        double d0 = iposition.getX();
+        double d1 = iposition.getY();
+        double d2 = iposition.getZ();
+        EntityItem entityitem = new EntityItem(par0World, d0, d1 - 0.3D, d2, par1ItemStack);
+        double d3 = par0World.rand.nextDouble() * 0.1D + 0.2D;
+        entityitem.motionX = (double)par3EnumFacing.getFrontOffsetX() * d3;
+        entityitem.motionY = 0.20000000298023224D;
+        entityitem.motionZ = (double)par3EnumFacing.getFrontOffsetZ() * d3;
+        entityitem.motionX += par0World.rand.nextGaussian() * 0.007499999832361937D * (double)par2;
+        entityitem.motionY += par0World.rand.nextGaussian() * 0.007499999832361937D * (double)par2;
+        entityitem.motionZ += par0World.rand.nextGaussian() * 0.007499999832361937D * (double)par2;
+        // CraftBukkit start
+        org.bukkit.block.Block block = par0World.getWorld().getBlockAt(par4IPosition.getXInt(), par4IPosition.getYInt(), par4IPosition.getZInt());
+        CraftItemStack craftItem = CraftItemStack.asCraftMirror(par1ItemStack);
+        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(entityitem.motionX, entityitem.motionY, entityitem.motionZ));
+
+        if (!BlockDispenser.eventFired)
+        {
+            par0World.getServer().getPluginManager().callEvent(event);
+        }
+
+        if (event.isCancelled())
+        {
+            return false;
+        }
+
+        entityitem.setEntityItemStack(CraftItemStack.asNMSCopy(event.getItem()));
+        entityitem.motionX = event.getVelocity().getX();
+        entityitem.motionY = event.getVelocity().getY();
+        entityitem.motionZ = event.getVelocity().getZ();
+
+        if (!event.getItem().equals(craftItem))
+        {
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            IBehaviorDispenseItem ibehaviordispenseitem = (IBehaviorDispenseItem) BlockDispenser.dispenseBehaviorRegistry.getObject(eventStack.getItem());
+
+            if (ibehaviordispenseitem != IBehaviorDispenseItem.itemDispenseBehaviorProvider && ibehaviordispenseitem.getClass() != BehaviorDefaultDispenseItem.class)
+            {
+                ibehaviordispenseitem.dispense(par4IPosition, eventStack);
+            }
+            else
+            {
+                par0World.spawnEntityInWorld(entityitem);
+            }
+
+            return false;
+        }
+
+        par0World.spawnEntityInWorld(entityitem);
+        return true;
+        // CraftBukkit end
     }
 
     /**
