--- ../src_base/minecraft/net/minecraft/dispenser/DispenserBehaviorFireball.java
+++ ../src_work/minecraft/net/minecraft/dispenser/DispenserBehaviorFireball.java
@@ -6,6 +6,10 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.world.World;
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
+// CraftBukkit end
 
 final class DispenserBehaviorFireball extends BehaviorDefaultDispenseItem
 {
@@ -24,8 +28,40 @@
         double d3 = random.nextGaussian() * 0.05D + (double)enumfacing.getFrontOffsetX();
         double d4 = random.nextGaussian() * 0.05D + (double)enumfacing.getFrontOffsetY();
         double d5 = random.nextGaussian() * 0.05D + (double)enumfacing.getFrontOffsetZ();
-        world.spawnEntityInWorld(new EntitySmallFireball(world, d0, d1, d2, d3, d4, d5));
-        par2ItemStack.splitStack(1);
+        // CraftBukkit start
+        ItemStack itemstack1 = par2ItemStack.splitStack(1);
+        org.bukkit.block.Block block = world.getWorld().getBlockAt(par1IBlockSource.getXInt(), par1IBlockSource.getYInt(), par1IBlockSource.getZInt());
+        CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d3, d4, d5));
+
+        if (!BlockDispenser.eventFired)
+        {
+            world.getServer().getPluginManager().callEvent(event);
+        }
+
+        if (event.isCancelled())
+        {
+            par2ItemStack.stackSize++;
+            return par2ItemStack;
+        }
+
+        if (!event.getItem().equals(craftItem))
+        {
+            par2ItemStack.stackSize++;
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            IBehaviorDispenseItem ibehaviordispenseitem = (IBehaviorDispenseItem) BlockDispenser.dispenseBehaviorRegistry.func_82594_a(eventStack.getItem());
+
+            if (ibehaviordispenseitem != IBehaviorDispenseItem.itemDispenseBehaviorProvider && ibehaviordispenseitem != this)
+            {
+                ibehaviordispenseitem.dispense(par1IBlockSource, eventStack);
+                return par2ItemStack;
+            }
+        }
+
+        world.spawnEntityInWorld(new EntitySmallFireball(world, d0, d1, d2, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ()));
+        // itemstack.a(1); // Handled during event processing
+        // CraftBukkit end
         return par2ItemStack;
     }
 
