--- ../src_base/minecraft/net/minecraft/dispenser/DispenserBehaviors.java
+++ ../src_work/minecraft/net/minecraft/dispenser/DispenserBehaviors.java
@@ -6,7 +6,7 @@
 
 public class DispenserBehaviors
 {
-    public static void registerDispenserBehaviours()
+    public static void registerVanillaDispenserBehaviors()
     {
         BlockDispenser.dispenseBehaviorRegistry.putObject(Item.arrow, new DispenserBehaviorArrow());
         BlockDispenser.dispenseBehaviorRegistry.putObject(Item.egg, new DispenserBehaviorEgg());
