--- a/net/minecraft/client/network/NetHandlerLoginClient.java
+++ b/net/minecraft/client/network/NetHandlerLoginClient.java
@@ -122,7 +122,9 @@
    }
 
    public void func_209521_a(SPacketCustomPayloadLogin p_209521_1_) {
+      if (!net.minecraftforge.fml.network.NetworkHooks.onCustomPayload(p_209521_1_, this.networkManager)) {
       this.field_209525_d.accept(new TextComponentTranslation("connect.negotiating", new Object[0]));
       this.networkManager.sendPacket(new CPacketCustomPayloadLogin(p_209521_1_.func_209918_a(), (PacketBuffer)null));
    }
 }
+}
