--- ../src-base/minecraft/net/minecraft/client/network/NetHandlerPlayClient.java
+++ ../src-work/minecraft/net/minecraft/client/network/NetHandlerPlayClient.java
@@ -191,6 +191,8 @@
 import net.minecraft.world.storage.ISaveHandler;
 import net.minecraft.world.storage.MapData;
 import net.minecraft.world.storage.MapStorage;
+import net.minecraftforge.client.event.ClientChatReceivedEvent;
+import net.minecraftforge.common.MinecraftForge;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -706,7 +708,11 @@
     // JAVADOC METHOD $$ func_147251_a
     public void handleChat(S02PacketChat p_147251_1_)
     {
-        this.gameController.ingameGUI.getChatGUI().printChatMessage(p_147251_1_.func_148915_c());
+        ClientChatReceivedEvent event = new ClientChatReceivedEvent(p_147251_1_.func_148915_c());
+        if (!MinecraftForge.EVENT_BUS.post(event) && event.message != null)
+        {
+            this.gameController.ingameGUI.getChatGUI().printChatMessage(event.message);
+        }
     }
 
     // JAVADOC METHOD $$ func_147279_a
@@ -1160,6 +1166,10 @@
                 {
                     tileentity.readFromNBT(p_147273_1_.func_148857_g());
                 }
+                else
+                {
+                    tileentity.onDataPacket(netManager, p_147273_1_);
+                }
             }
         }
     }
