--- ../src-base/minecraft/net/minecraft/client/network/NetHandlerPlayClient.java
+++ ../src-work/minecraft/net/minecraft/client/network/NetHandlerPlayClient.java
@@ -196,6 +196,8 @@
 import net.minecraft.world.storage.ISaveHandler;
 import net.minecraft.world.storage.MapData;
 import net.minecraft.world.storage.MapStorage;
+import net.minecraftforge.client.event.ClientChatReceivedEvent;
+import net.minecraftforge.common.MinecraftForge;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -390,7 +392,8 @@
 
     public void func_147286_a(S11PacketSpawnExperienceOrb p_147286_1_)
     {
-        EntityXPOrb entityxporb = new EntityXPOrb(this.field_147300_g, (double)p_147286_1_.func_148984_d(), (double)p_147286_1_.func_148983_e(), (double)p_147286_1_.func_148982_f(), p_147286_1_.func_148986_g());
+        EntityXPOrb entityxporb = new EntityXPOrb(this.field_147300_g, (double)p_147286_1_.func_148984_d() / 32.0D, (double)p_147286_1_.func_148983_e() / 32.0D, (double)p_147286_1_.func_148982_f() / 32.0D, p_147286_1_.func_148986_g());
+        // FORGE: BugFix MC-12013 Wrong XP orb clientside spawn position
         entityxporb.field_70118_ct = p_147286_1_.func_148984_d();
         entityxporb.field_70117_cu = p_147286_1_.func_148983_e();
         entityxporb.field_70116_cv = p_147286_1_.func_148982_f();
@@ -687,7 +690,11 @@
 
     public void func_147251_a(S02PacketChat p_147251_1_)
     {
-        this.field_147299_f.field_71456_v.func_146158_b().func_146227_a(p_147251_1_.func_148915_c());
+        ClientChatReceivedEvent event = new ClientChatReceivedEvent(p_147251_1_.func_148915_c());
+        if (!MinecraftForge.EVENT_BUS.post(event) && event.message != null)
+        {
+            this.field_147299_f.field_71456_v.func_146158_b().func_146227_a(event.message);
+        }
     }
 
     public void func_147279_a(S0BPacketAnimation p_147279_1_)
@@ -735,6 +742,11 @@
         float f = (float)(p_147281_1_.func_149028_l() * 360) / 256.0F;
         float f1 = (float)(p_147281_1_.func_149030_m() * 360) / 256.0F;
         EntityLivingBase entitylivingbase = (EntityLivingBase)EntityList.func_75616_a(p_147281_1_.func_149025_e(), this.field_147299_f.field_71441_e);
+        if (entitylivingbase == null)
+        {
+            cpw.mods.fml.common.FMLLog.info("Server attempted to spawn an unknown entity using ID: {0} at ({1}, {2}, {3}) Skipping!", p_147281_1_.func_149025_e(), d0, d1, d2);
+            return;
+        }
         entitylivingbase.field_70118_ct = p_147281_1_.func_149023_f();
         entitylivingbase.field_70117_cu = p_147281_1_.func_149034_g();
         entitylivingbase.field_70116_cv = p_147281_1_.func_149029_h();
@@ -1127,6 +1139,10 @@
                 {
                     tileentity.func_145839_a(p_147273_1_.func_148857_g());
                 }
+                else
+                {
+                    tileentity.onDataPacket(field_147302_e, p_147273_1_);
+                }
             }
         }
     }
