--- a/net/minecraft/util/math/RayTraceResult.java
+++ b/net/minecraft/util/math/RayTraceResult.java
@@ -10,6 +10,17 @@
    public Vec3d field_72307_f;
    public Entity field_72308_g;
 
+   /* ======================================== FORGE START =====================================*/
+   private cpw.mods.modlauncher.api.TypesafeMap extraInfo = null;
+
+   /** Use this to add extra info to the result */
+   public cpw.mods.modlauncher.api.TypesafeMap extraInfo() {
+      if (extraInfo == null)
+         extraInfo = new cpw.mods.modlauncher.api.TypesafeMap();
+      return extraInfo;
+   }
+   /* ======================================== FORGE END   =====================================*/
+
    public RayTraceResult(Vec3d p_i47094_1_, EnumFacing p_i47094_2_, BlockPos p_i47094_3_) {
       this(RayTraceResult.Type.BLOCK, p_i47094_1_, p_i47094_2_, p_i47094_3_);
    }
