--- ../src-base/minecraft/net/minecraft/util/math/Vec3d.java
+++ ../src-work/minecraft/net/minecraft/util/math/Vec3d.java
@@ -1,9 +1,6 @@
 package net.minecraft.util.math;
 
 import javax.annotation.Nullable;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
-
 public class Vec3d
 {
     public static final Vec3d field_186680_a = new Vec3d(0.0D, 0.0D, 0.0D);
@@ -54,7 +51,6 @@
         return this.field_72450_a * p_72430_1_.field_72450_a + this.field_72448_b * p_72430_1_.field_72448_b + this.field_72449_c * p_72430_1_.field_72449_c;
     }
 
-    @SideOnly(Side.CLIENT)
     public Vec3d func_72431_c(Vec3d p_72431_1_)
     {
         return new Vec3d(this.field_72448_b * p_72431_1_.field_72449_c - this.field_72449_c * p_72431_1_.field_72448_b, this.field_72449_c * p_72431_1_.field_72450_a - this.field_72450_a * p_72431_1_.field_72449_c, this.field_72450_a * p_72431_1_.field_72448_b - this.field_72448_b * p_72431_1_.field_72450_a);
@@ -114,7 +110,6 @@
         return (double)MathHelper.func_76133_a(this.field_72450_a * this.field_72450_a + this.field_72448_b * this.field_72448_b + this.field_72449_c * this.field_72449_c);
     }
 
-    @SideOnly(Side.CLIENT)
     public double func_189985_c()
     {
         return this.field_72450_a * this.field_72450_a + this.field_72448_b * this.field_72448_b + this.field_72449_c * this.field_72449_c;
@@ -227,13 +222,11 @@
         return new Vec3d(d0, d1, d2);
     }
 
-    @SideOnly(Side.CLIENT)
     public static Vec3d func_189984_a(Vec2f p_189984_0_)
     {
         return func_189986_a(p_189984_0_.field_189982_i, p_189984_0_.field_189983_j);
     }
 
-    @SideOnly(Side.CLIENT)
     public static Vec3d func_189986_a(float p_189986_0_, float p_189986_1_)
     {
         float f = MathHelper.func_76134_b(-p_189986_1_ * 0.017453292F - (float)Math.PI);
