--- a/com/mojang/blaze3d/opengl/GlCommandEncoder.java
+++ b/com/mojang/blaze3d/opengl/GlCommandEncoder.java
@@ -132,6 +_,9 @@
             GlStateManager._colorMask(true, true, true, true);
             GlStateManager._clear(16384);
             GlStateManager._glFramebufferTexture2D(36160, 36064, 3553, 0, 0);
+            // Forge: if using stencil, clear it as well (3553 = GlConst.GL_TEXTURE_2D)
+            if (p_394273_.isStencilEnabled())
+                GlStateManager._glFramebufferTexture2D(GlConst.GL_FRAMEBUFFER, org.lwjgl.opengl.GL30.GL_STENCIL_ATTACHMENT, GlConst.GL_TEXTURE_2D, 0, 0);
             GlStateManager._glBindFramebuffer(36160, 0);
         }
     }
