--- ../src-base/minecraft/net/minecraft/command/server/CommandTestForBlock.java
+++ ../src-work/minecraft/net/minecraft/command/server/CommandTestForBlock.java
@@ -72,7 +72,7 @@
                     NBTTagCompound nbttagcompound = new NBTTagCompound();
                     boolean flag = false;
 
-                    if (p_184881_3_.length >= 6 && block.hasTileEntity())
+                    if (p_184881_3_.length >= 6 && block.hasTileEntity(block.getStateFromMeta(i)))
                     {
                         String s = getChatComponentFromNthArg(p_184881_2_, p_184881_3_, 5).getUnformattedText();
 
