--- a/net/minecraft/client/gui/widget/TextFieldWidget.java
+++ b/net/minecraft/client/gui/widget/TextFieldWidget.java
@@ -349,7 +349,7 @@
       } else {
          boolean flag = p_mouseClicked_1_ >= (double)this.x && p_mouseClicked_1_ < (double)(this.x + this.width) && p_mouseClicked_3_ >= (double)this.y && p_mouseClicked_3_ < (double)(this.y + this.height);
          if (this.field_146212_n) {
-            this.func_146195_b(flag);
+            this.setFocused(flag);
          }
 
          if (this.isFocused() && flag && p_mouseClicked_5_ == 0) {
@@ -358,7 +358,7 @@
                i -= 4;
             }
 
-            String s = this.field_146211_a.func_78269_a(this.field_146216_j.substring(this.field_146225_q), this.func_146200_o());
+            String s = this.field_146211_a.func_78269_a(this.field_146216_j.substring(this.field_146225_q), this.getWidth());
             this.func_146190_e(this.field_146211_a.func_78269_a(s, i).length() + this.field_146225_q);
             return true;
          } else {
@@ -367,7 +367,7 @@
       }
    }
 
-   public void func_146195_b(boolean p_146195_1_) {
+   public void setFocused(boolean p_146195_1_) {
       super.setFocused(p_146195_1_);
    }
 
@@ -381,7 +381,7 @@
          int i = this.field_146226_p ? this.field_146222_t : this.field_146221_u;
          int j = this.field_146224_r - this.field_146225_q;
          int k = this.field_146223_s - this.field_146225_q;
-         String s = this.field_146211_a.func_78269_a(this.field_146216_j.substring(this.field_146225_q), this.func_146200_o());
+         String s = this.field_146211_a.func_78269_a(this.field_146216_j.substring(this.field_146225_q), this.getWidth());
          boolean flag = j >= 0 && j <= s.length();
          boolean flag1 = this.isFocused() && this.field_146214_l / 6 % 2 == 0 && flag;
          int l = this.field_146215_m ? this.x + 4 : this.x;
@@ -522,7 +522,7 @@
       this.field_146226_p = p_146184_1_;
    }
 
-   public int func_146200_o() {
+   public int getWidth() {
       return this.func_146181_i() ? this.width - 8 : this.width;
    }
 
@@ -534,7 +534,7 @@
             this.field_146225_q = i;
          }
 
-         int j = this.func_146200_o();
+         int j = this.getWidth();
          String s = this.field_146211_a.func_78269_a(this.field_146216_j.substring(this.field_146225_q), j);
          int k = s.length() + this.field_146225_q;
          if (this.field_146223_s == this.field_146225_q) {
