--- ../src-base/minecraft/net/minecraft/network/packet/NetHandler.java
+++ ../src-work/minecraft/net/minecraft/network/packet/NetHandler.java
@@ -1,4 +1,6 @@
 package net.minecraft.network.packet;
+
+import net.minecraft.entity.player.EntityPlayer;
 
 public abstract class NetHandler
 {
@@ -365,4 +367,8 @@
     {
         this.func_72509_a(p_98182_1_);
     }
+
+    public abstract void handleVanilla250Packet(Packet250CustomPayload payload);
+
+    public abstract EntityPlayer getPlayer();
 }
