--- a/net/minecraft/resources/DataPackRegistries.java
+++ b/net/minecraft/resources/DataPackRegistries.java
@@ -22,6 +22,8 @@
    private final AdvancementManager field_240958_h_ = new AdvancementManager(this.field_240956_f_);
    private final FunctionReloader field_240959_i_;
 
+   private final net.minecraftforge.common.loot.LootModifierManager lootManager = new net.minecraftforge.common.loot.LootModifierManager();
+
    public DataPackRegistries(Commands.EnvironmentType p_i232598_1_, int p_i232598_2_) {
       this.field_240953_c_ = new Commands(p_i232598_1_);
       this.field_240959_i_ = new FunctionReloader(p_i232598_2_, this.field_240953_c_.func_197054_a());
@@ -31,6 +33,7 @@
       this.field_240952_b_.func_219534_a(this.field_240957_g_);
       this.field_240952_b_.func_219534_a(this.field_240959_i_);
       this.field_240952_b_.func_219534_a(this.field_240958_h_);
+      this.field_240952_b_.func_219534_a(this.lootManager);
    }
 
    public FunctionReloader func_240960_a_() {
@@ -85,4 +88,8 @@
    public void close() {
       this.field_240952_b_.close();
    }
+
+   public net.minecraftforge.common.loot.LootModifierManager getLootModifierManager() {
+      return lootManager;
+   }
 }
