--- a/net/minecraft/resources/IResourcePack.java
+++ b/net/minecraft/resources/IResourcePack.java
@@ -11,7 +_,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public interface IResourcePack extends AutoCloseable {
+public interface IResourcePack extends AutoCloseable, net.minecraftforge.common.extensions.IForgeResourcePack {
    @OnlyIn(Dist.CLIENT)
    InputStream func_195763_b(String p_195763_1_) throws IOException;
 
