--- a/net/minecraft/resources/IResourceManager.java
+++ b/net/minecraft/resources/IResourceManager.java
@@ -10,18 +10,15 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public interface IResourceManager {
-   @OnlyIn(Dist.CLIENT)
    Set<String> func_199001_a();
 
    IResource func_199002_a(ResourceLocation p_199002_1_) throws IOException;
 
-   @OnlyIn(Dist.CLIENT)
    boolean func_219533_b(ResourceLocation p_219533_1_);
 
    List<IResource> func_199004_b(ResourceLocation p_199004_1_) throws IOException;
 
    Collection<ResourceLocation> func_199003_a(String p_199003_1_, Predicate<String> p_199003_2_);
 
-   @OnlyIn(Dist.CLIENT)
    void func_199021_a(IResourcePack p_199021_1_);
 }
