--- a/net/minecraft/resources/ResourcePackList.java
+++ b/net/minecraft/resources/ResourcePackList.java
@@ -22,7 +22,7 @@
 
    public ResourcePackList(ResourcePackInfo.IFactory p_i231423_1_, IPackFinder... p_i231423_2_) {
       this.field_198990_d = p_i231423_1_;
-      this.field_198987_a = ImmutableSet.copyOf(p_i231423_2_);
+      this.field_198987_a = new java.util.HashSet<>(java.util.Arrays.asList(p_i231423_2_));
    }
 
    public ResourcePackList(IPackFinder... p_i241886_1_) {
@@ -89,6 +89,10 @@
       return this.field_198988_b.get(p_198981_1_);
    }
 
+   public void addPackFinder(IPackFinder packFinder) {
+      this.field_198987_a.add(packFinder);
+   }
+
    public void close() {
       this.field_198988_b.values().forEach(ResourcePackInfo::close);
    }
