--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -57,7 +_,7 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class LevelChunk implements ChunkAccess {
+public class LevelChunk extends net.minecraftforge.common.capabilities.CapabilityProvider<LevelChunk> implements ChunkAccess, net.minecraftforge.common.extensions.IForgeLevelChunk {
    static final Logger f_62771_ = LogManager.getLogger();
    private static final TickingBlockEntity f_156361_ = new TickingBlockEntity() {
       public void m_142224_() {
@@ -106,6 +_,7 @@
    }
 
    public LevelChunk(Level p_62800_, ChunkPos p_62801_, ChunkBiomeContainer p_62802_, UpgradeData p_62803_, TickList<Block> p_62804_, TickList<Fluid> p_62805_, long p_62806_, @Nullable LevelChunkSection[] p_62807_, @Nullable Consumer<LevelChunk> p_62808_) {
+      super(LevelChunk.class);
       this.f_62776_ = p_62800_;
       this.f_62792_ = p_62801_;
       this.f_62778_ = p_62803_;
@@ -132,6 +_,7 @@
       }
 
       this.f_62783_ = new ShortList[p_62800_.m_151559_()];
+      this.gatherCapabilities();
    }
 
    public LevelChunk(ServerLevel p_156365_, ProtoChunk p_156366_, @Nullable Consumer<LevelChunk> p_156367_) {
@@ -279,14 +_,14 @@
          boolean flag2 = blockstate.m_155947_();
          if (!this.f_62776_.f_46443_) {
             blockstate.m_60753_(this.f_62776_, p_62865_, p_62866_, p_62867_);
-         } else if (!blockstate.m_60713_(block) && flag2) {
+         } else if ((!blockstate.m_60713_(block) || !p_62866_.m_155947_()) && flag2) {
             this.m_8114_(p_62865_);
          }
 
          if (!levelchunksection.m_62982_(k, l, i1).m_60713_(block)) {
             return null;
          } else {
-            if (!this.f_62776_.f_46443_) {
+            if (!this.f_62776_.f_46443_ && !this.f_62776_.captureBlockSnapshots) {
                p_62866_.m_60696_(this.f_62776_, p_62865_, blockstate, p_62867_);
             }
 
@@ -349,6 +_,10 @@
    @Nullable
    public BlockEntity m_5685_(BlockPos p_62868_, LevelChunk.EntityCreationType p_62869_) {
       BlockEntity blockentity = this.f_62779_.get(p_62868_);
+      if (blockentity != null && blockentity.m_58901_()) {
+         f_62779_.remove(p_62868_);
+         blockentity = null;
+      }
       if (blockentity == null) {
          CompoundTag compoundtag = this.f_62774_.remove(p_62868_);
          if (compoundtag != null) {
@@ -366,9 +_,6 @@
                this.m_142170_(blockentity);
             }
          }
-      } else if (blockentity.m_58901_()) {
-         this.f_62779_.remove(p_62868_);
-         return null;
       }
 
       return blockentity;
@@ -379,6 +_,7 @@
       if (this.m_156370_()) {
          this.m_156404_(p_156391_);
          this.m_156406_(p_156391_);
+         p_156391_.onLoad();
       }
 
    }
@@ -418,9 +_,14 @@
    public CompoundTag m_8051_(BlockPos p_62932_) {
       BlockEntity blockentity = this.m_7702_(p_62932_);
       if (blockentity != null && !blockentity.m_58901_()) {
+         try {
          CompoundTag compoundtag1 = blockentity.m_6945_(new CompoundTag());
          compoundtag1.m_128379_("keepPacked", false);
          return compoundtag1;
+         } catch (Exception e) {
+            LogManager.getLogger().error("A BlockEntity type {} has thrown an exception trying to write state. It will not persist, Report this to the mod author", blockentity.getClass().getName(), e);
+            return null;
+         }
       } else {
          CompoundTag compoundtag = this.f_62774_.get(p_62932_);
          if (compoundtag != null) {
@@ -567,7 +_,7 @@
 
    public Stream<BlockPos> m_6267_() {
       return StreamSupport.stream(BlockPos.m_121976_(this.f_62792_.m_45604_(), this.m_141937_(), this.f_62792_.m_45605_(), this.f_62792_.m_45608_(), this.m_151558_() - 1, this.f_62792_.m_45609_()).spliterator(), false).filter((p_156419_) -> {
-         return this.m_8055_(p_156419_).m_60791_() != 0;
+         return this.m_8055_(p_156419_).getLightEmission(m_62953_(), p_156419_) != 0;
       });
    }
 
@@ -763,6 +_,7 @@
    }
 
    public void m_156369_() {
+      this.f_62776_.addFreshBlockEntities(this.f_62779_.values());
       this.f_62779_.values().forEach((p_156409_) -> {
          this.m_156404_(p_156409_);
          this.m_156406_(p_156409_);
@@ -826,6 +_,7 @@
             if (LevelChunk.this.m_156410_(blockpos)) {
                try {
                   ProfilerFiller profilerfiller = LevelChunk.this.f_62776_.m_46473_();
+                  net.minecraftforge.server.timings.TimeTracker.BLOCK_ENTITY_UPDATE.trackStart(f_156428_);
                   profilerfiller.m_6521_(this::m_142280_);
                   BlockState blockstate = LevelChunk.this.m_8055_(blockpos);
                   if (this.f_156428_.m_58903_().m_155262_(blockstate)) {
@@ -843,6 +_,12 @@
                   CrashReport crashreport = CrashReport.m_127521_(throwable, "Ticking block entity");
                   CrashReportCategory crashreportcategory = crashreport.m_127514_("Block entity being ticked");
                   this.f_156428_.m_58886_(crashreportcategory);
+
+                  if (net.minecraftforge.common.ForgeConfig.SERVER.removeErroringBlockEntities.get()) {
+                     LogManager.getLogger().fatal("{}", crashreport.m_127526_());
+                     f_156428_.m_7651_();
+                     LevelChunk.this.m_8114_(f_156428_.m_58899_());
+                  } else
                   throw new ReportedException(crashreport);
                }
             }
@@ -871,6 +_,32 @@
       IMMEDIATE,
       QUEUED,
       CHECK;
+   }
+
+   /**
+    * <strong>FOR INTERNAL USE ONLY</strong>
+    * <p>
+    * Only public for use in {@link net.minecraft.world.level.chunk.storage.ChunkSerializer}.
+    */
+   @java.lang.Deprecated
+   @javax.annotation.Nullable
+   public final CompoundTag writeCapsToNBT() {
+      return this.serializeCaps();
+   }
+
+   /**
+    * <strong>FOR INTERNAL USE ONLY</strong>
+    * <p>
+    * Only public for use in {@link net.minecraft.world.level.chunk.storage.ChunkSerializer}.
+    */
+   @java.lang.Deprecated
+   public final void readCapsFromNBT(CompoundTag tag) {
+      this.deserializeCaps(tag);
+   }
+
+   @Override
+   public Level getWorldForge() {
+      return m_62953_();
    }
 
    class RebindableTickingBlockEntityWrapper implements TickingBlockEntity {
