--- ../src-base/minecraft/net/minecraft/entity/passive/AbstractHorse.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/AbstractHorse.java
@@ -604,10 +604,9 @@
         this.field_110278_bp = 1;
     }
 
-    public void func_70645_a(DamageSource p_70645_1_)
+    public boolean onDeath(DamageSource cause)
     {
-        super.func_70645_a(p_70645_1_);
-
+        if(super.onDeath(cause)) return true;
         if (!this.field_70170_p.field_72995_K && this.field_110296_bG != null)
         {
             for (int i = 0; i < this.field_110296_bG.func_70302_i_(); ++i)
@@ -620,6 +619,7 @@
                 }
             }
         }
+        return false;
     }
 
     public void func_70636_d()
@@ -981,6 +981,7 @@
         }
 
         this.func_110232_cE();
+        this.itemHandler = new net.minecraftforge.items.wrapper.InvWrapper(this.field_110296_bG);
     }
 
     public boolean func_70878_b(EntityAnimal p_70878_1_)
@@ -1219,4 +1220,22 @@
 
         return p_180482_2_;
     }
+
+    // FORGE
+    private net.minecraftforge.items.IItemHandler itemHandler = null; // Initialized by initHorseChest above.
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable net.minecraft.util.EnumFacing facing)
+    {
+        if (capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) return (T) itemHandler;
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @Nullable net.minecraft.util.EnumFacing facing)
+    {
+        return capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing);
+    }
 }
