--- a/net/minecraft/entity/passive/BeeEntity.java
+++ b/net/minecraft/entity/passive/BeeEntity.java
@@ -358,7 +358,7 @@
          return false;
       } else {
          TileEntity tileentity = this.field_70170_p.func_175625_s(this.field_226369_bI_);
-         return tileentity instanceof BeehiveTileEntity && ((BeehiveTileEntity)tileentity).func_226968_d_();
+         return tileentity instanceof net.minecraftforge.common.extensions.IForgeBeehiveTileEntity && ((net.minecraftforge.common.extensions.IForgeBeehiveTileEntity)tileentity).func_226968_d_();
       }
    }
 
@@ -376,8 +376,8 @@
 
    private boolean func_226435_i_(BlockPos p_226435_1_) {
       TileEntity tileentity = this.field_70170_p.func_175625_s(p_226435_1_);
-      if (tileentity instanceof BeehiveTileEntity) {
-         return !((BeehiveTileEntity)tileentity).func_226970_h_();
+      if (tileentity instanceof net.minecraftforge.common.extensions.IForgeBeehiveTileEntity) {
+         return !((net.minecraftforge.common.extensions.IForgeBeehiveTileEntity)tileentity).func_226970_h_();
       } else {
          return false;
       }
@@ -662,8 +662,8 @@
       public boolean func_225506_g_() {
          if (BeeEntity.this.func_226409_eA_() && BeeEntity.this.func_226415_eI_() && BeeEntity.this.field_226369_bI_.func_218137_a(BeeEntity.this.func_213303_ch(), 2.0D)) {
             TileEntity tileentity = BeeEntity.this.field_70170_p.func_175625_s(BeeEntity.this.field_226369_bI_);
-            if (tileentity instanceof BeehiveTileEntity) {
-               BeehiveTileEntity beehivetileentity = (BeehiveTileEntity)tileentity;
+            if (tileentity instanceof net.minecraftforge.common.extensions.IForgeBeehiveTileEntity) {
+               net.minecraftforge.common.extensions.IForgeBeehiveTileEntity beehivetileentity = (net.minecraftforge.common.extensions.IForgeBeehiveTileEntity)tileentity;
                if (!beehivetileentity.func_226970_h_()) {
                   return true;
                }
@@ -681,8 +681,8 @@
 
       public void func_75249_e() {
          TileEntity tileentity = BeeEntity.this.field_70170_p.func_175625_s(BeeEntity.this.field_226369_bI_);
-         if (tileentity instanceof BeehiveTileEntity) {
-            BeehiveTileEntity beehivetileentity = (BeehiveTileEntity)tileentity;
+         if (tileentity instanceof net.minecraftforge.common.extensions.IForgeBeehiveTileEntity) {
+            net.minecraftforge.common.extensions.IForgeBeehiveTileEntity beehivetileentity = (net.minecraftforge.common.extensions.IForgeBeehiveTileEntity)tileentity;
             beehivetileentity.func_226961_a_(BeeEntity.this, BeeEntity.this.func_226411_eD_());
          }
 
