--- a/net/minecraft/entity/passive/BeeEntity.java
+++ b/net/minecraft/entity/passive/BeeEntity.java
@@ -184,6 +_,7 @@
       this.field_226363_bC_ = p_70037_1_.func_74762_e("TicksSincePollination");
       this.field_226364_bD_ = p_70037_1_.func_74762_e("CannotEnterHiveTicks");
       this.field_226365_bE_ = p_70037_1_.func_74762_e("CropsGrownSincePollination");
+      if(!field_70170_p.field_72995_K) //FORGE: allow this entity to be read from nbt on client. (Fixes MC-189565)
       this.func_241358_a_((ServerWorld)this.field_70170_p, p_70037_1_);
    }
 
@@ -425,7 +_,7 @@
          return false;
       } else {
          TileEntity tileentity = this.field_70170_p.func_175625_s(this.field_226369_bI_);
-         return tileentity != null && tileentity.func_200662_C() == TileEntityType.field_226985_G_;
+         return tileentity instanceof BeehiveTileEntity;
       }
    }
 
