--- ../src_base/minecraft/net/minecraft/entity/passive/EntitySquid.java
+++ ../src_work/minecraft/net/minecraft/entity/passive/EntitySquid.java
@@ -33,11 +33,6 @@
         this.texture = "/mob/squid.png";
         this.setSize(0.95F, 0.95F);
         this.field_70864_bA = 1.0F / (this.rand.nextFloat() + 1.0F) * 0.2F;
-    }
-
-    public int getMaxHealth()
-    {
-        return 10;
     }
 
     /**
