--- ../src-base/minecraft/net/minecraft/entity/passive/EntityHorse.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityHorse.java
@@ -34,6 +34,7 @@
     private static final UUID field_184786_bD = UUID.fromString("556E1665-8B10-40C8-8F9D-CF9B1667F295");
     private static final DataParameter<Integer> field_184789_bG = EntityDataManager.<Integer>func_187226_a(EntityHorse.class, DataSerializers.field_187192_b);
     private static final DataParameter<Integer> field_184791_bI = EntityDataManager.<Integer>func_187226_a(EntityHorse.class, DataSerializers.field_187192_b);
+    private static final DataParameter<String> HORSE_ARMOR_STRING = EntityDataManager.<String>func_187226_a(EntityHorse.class, DataSerializers.field_187194_d);
     private static final String[] field_110268_bz = new String[] {"textures/entity/horse/horse_white.png", "textures/entity/horse/horse_creamy.png", "textures/entity/horse/horse_chestnut.png", "textures/entity/horse/horse_brown.png", "textures/entity/horse/horse_black.png", "textures/entity/horse/horse_gray.png", "textures/entity/horse/horse_darkbrown.png"};
     private static final String[] field_110269_bA = new String[] {"hwh", "hcr", "hch", "hbr", "hbl", "hgr", "hdb"};
     private static final String[] field_110291_bB = new String[] {null, "textures/entity/horse/horse_markings_white.png", "textures/entity/horse/horse_markings_whitefield.png", "textures/entity/horse/horse_markings_whitedots.png", "textures/entity/horse/horse_markings_blackdots.png"};
@@ -51,6 +52,7 @@
         super.func_70088_a();
         this.field_70180_af.func_187214_a(field_184789_bG, Integer.valueOf(0));
         this.field_70180_af.func_187214_a(field_184791_bI, Integer.valueOf(HorseArmorType.NONE.func_188579_a()));
+        this.field_70180_af.func_187214_a(HORSE_ARMOR_STRING, HorseArmorType.NONE.name());
     }
 
     public static void func_189803_b(DataFixer p_189803_0_)
@@ -79,7 +81,7 @@
         {
             ItemStack itemstack = new ItemStack(p_70037_1_.func_74775_l("ArmorItem"));
 
-            if (!itemstack.func_190926_b() && HorseArmorType.func_188577_b(itemstack.func_77973_b()))
+            if (!itemstack.func_190926_b() && HorseArmorType.isHorseArmor(itemstack))
             {
                 this.field_110296_bG.func_70299_a(1, itemstack);
             }
@@ -149,6 +151,7 @@
     {
         HorseArmorType horsearmortype = HorseArmorType.func_188580_a(p_146086_1_);
         this.field_70180_af.func_187227_b(field_184791_bI, Integer.valueOf(horsearmortype.func_188579_a()));
+        this.field_70180_af.func_187227_b(HORSE_ARMOR_STRING, horsearmortype.name());
         this.func_110230_cF();
 
         if (!this.field_70170_p.field_72995_K)
@@ -165,7 +168,9 @@
 
     public HorseArmorType func_184783_dl()
     {
-        return HorseArmorType.func_188575_a(((Integer)this.field_70180_af.func_187225_a(field_184791_bI)).intValue());
+        HorseArmorType armor = HorseArmorType.getByName(this.field_70180_af.func_187225_a(HORSE_ARMOR_STRING)); //First check the Forge armor DataParameter
+        if (armor == HorseArmorType.NONE) armor = HorseArmorType.func_188575_a(this.field_70180_af.func_187225_a(field_184791_bI)); //If the Forge armor DataParameter returns NONE, fallback to the vanilla armor DataParameter. This is necessary to prevent issues with Forge clients connected to vanilla servers.
+        return armor;
     }
 
     public void func_76316_a(IInventory p_76316_1_)
@@ -207,6 +212,8 @@
             this.field_70180_af.func_187230_e();
             this.func_110230_cF();
         }
+        ItemStack armor = this.field_110296_bG.func_70301_a(1);
+        if (armor.func_77973_b() instanceof net.minecraftforge.common.IHorseArmor) ((net.minecraftforge.common.IHorseArmor) armor.func_77973_b()).onHorseArmorTick(field_70170_p, this, armor);
     }
 
     protected SoundEvent func_184639_G()
@@ -381,7 +388,7 @@
 
     public boolean func_190682_f(ItemStack p_190682_1_)
     {
-        return HorseArmorType.func_188577_b(p_190682_1_.func_77973_b());
+        return HorseArmorType.isHorseArmor(p_190682_1_);
     }
 
     @Nullable
