--- ../src-base/minecraft/net/minecraft/entity/passive/EntitySheep.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntitySheep.java
@@ -2,6 +2,7 @@
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
+import java.util.ArrayList;
 import java.util.Random;
 import net.minecraft.block.Block;
 import net.minecraft.entity.EntityAgeable;
@@ -27,9 +28,11 @@
 import net.minecraft.item.crafting.CraftingManager;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.MathHelper;
+import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
+import net.minecraftforge.common.IShearable;
 
-public class EntitySheep extends EntityAnimal
+public class EntitySheep extends EntityAnimal implements IShearable
 {
     private final InventoryCrafting field_90016_e = new InventoryCrafting(new Container()
     {
@@ -124,28 +127,6 @@
 
     public boolean func_70085_c(EntityPlayer p_70085_1_)
     {
-        ItemStack itemstack = p_70085_1_.field_71071_by.func_70448_g();
-
-        if (itemstack != null && itemstack.func_77973_b() == Items.field_151097_aZ && !this.func_70892_o() && !this.func_70631_g_())
-        {
-            if (!this.field_70170_p.field_72995_K)
-            {
-                this.func_70893_e(true);
-                int i = 1 + this.field_70146_Z.nextInt(3);
-
-                for (int j = 0; j < i; ++j)
-                {
-                    EntityItem entityitem = this.func_70099_a(new ItemStack(Item.func_150898_a(Blocks.field_150325_L), 1, this.func_70896_n()), 1.0F);
-                    entityitem.field_70181_x += (double)(this.field_70146_Z.nextFloat() * 0.05F);
-                    entityitem.field_70159_w += (double)((this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.1F);
-                    entityitem.field_70179_y += (double)((this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.1F);
-                }
-            }
-
-            itemstack.func_77972_a(1, p_70085_1_);
-            this.func_85030_a("mob.sheep.shear", 1.0F, 1.0F);
-        }
-
         return super.func_70085_c(p_70085_1_);
     }
 
@@ -290,4 +271,24 @@
     {
         return 15 - ((EntitySheep)p_90013_1_).func_70896_n();
     }
+
+    @Override
+    public boolean isShearable(ItemStack item, IBlockAccess world, int x, int y, int z)
+    {
+        return !func_70892_o() && !func_70631_g_();
+    }
+
+    @Override
+    public ArrayList<ItemStack> onSheared(ItemStack item, IBlockAccess world, int x, int y, int z, int fortune)
+    {
+        ArrayList<ItemStack> ret = new ArrayList<ItemStack>();
+        func_70893_e(true);
+        int i = 1 + field_70146_Z.nextInt(3);
+        for (int j = 0; j < i; j++)
+        {
+            ret.add(new ItemStack(Blocks.field_150325_L, 1, func_70896_n()));
+        }
+        this.func_85030_a("mob.sheep.shear", 1.0F, 1.0F);
+        return ret;
+    }
 }
