--- a/net/minecraft/entity/passive/EntityParrot.java
+++ b/net/minecraft/entity/passive/EntityParrot.java
@@ -219,7 +219,7 @@
          }
 
          if (!this.field_70170_p.field_72995_K) {
-            if (this.field_70146_Z.nextInt(10) == 0) {
+            if (this.field_70146_Z.nextInt(10) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_184645_1_)) {
                this.func_193101_c(p_184645_1_);
                this.func_70908_e(true);
                this.field_70170_p.func_72960_a(this, (byte)7);
@@ -260,7 +260,7 @@
       int k = MathHelper.func_76128_c(this.field_70161_v);
       BlockPos blockpos = new BlockPos(i, j, k);
       Block block = p_205020_1_.func_180495_p(blockpos.func_177977_b()).func_177230_c();
-      return block instanceof BlockLeaves || block == Blocks.field_150349_c || block instanceof BlockLog || block == Blocks.field_150350_a && super.func_205020_a(p_205020_1_, p_205020_2_);
+      return block instanceof BlockLeaves || block == Blocks.field_196658_i || block instanceof BlockLog || block == Blocks.field_150350_a && super.func_205020_a(p_205020_1_, p_205020_2_);
    }
 
    public void func_180430_e(float p_180430_1_, float p_180430_2_) {
