--- ../src-base/minecraft/net/minecraft/entity/passive/EntityTameable.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityTameable.java
@@ -137,6 +137,11 @@
         return (((Byte)this.field_70180_af.func_187225_a(field_184755_bv)).byteValue() & 4) != 0;
     }
 
+    public boolean isTamingItem(net.minecraft.item.ItemStack stack)
+    {
+        return stack != null && ((stack.func_77973_b() == net.minecraft.init.Items.field_151103_aS && this instanceof EntityWolf) || (stack.func_77973_b() == net.minecraft.init.Items.field_151115_aP && this instanceof EntityOcelot));
+    }
+
     public void func_70903_f(boolean p_70903_1_)
     {
         byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184755_bv)).byteValue();
