--- ../src-base/minecraft/net/minecraft/entity/passive/EntityOcelot.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityOcelot.java
@@ -281,7 +281,7 @@
 
                 Block block = this.field_70170_p.func_147439_a(i, j - 1, k);
 
-                if (block == Blocks.field_150349_c || block.func_149688_o() == Material.field_151584_j)
+                if (block == Blocks.field_150349_c || block.isLeaves(field_70170_p, i, j - 1, k))
                 {
                     return true;
                 }
