--- ../src-base/minecraft/net/minecraft/entity/passive/EntityOcelot.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityOcelot.java
@@ -288,7 +288,7 @@
             IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos.func_177977_b());
             Block block = iblockstate.func_177230_c();
 
-            if (block == Blocks.field_150349_c || iblockstate.func_185904_a() == Material.field_151584_j)
+            if (block == Blocks.field_150349_c || block.isLeaves(iblockstate, this.field_70170_p, blockpos.func_177977_b()))
             {
                 return true;
             }
