--- ../src-base/minecraft/net/minecraft/entity/passive/EntityVillager.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityVillager.java
@@ -109,7 +109,7 @@
     private boolean field_70959_by;
     private boolean field_175565_bs;
     private int field_70956_bz;
-    private String field_82189_bL;
+    private java.util.UUID field_82189_bL;
     private int field_175563_bv;
     private int field_175562_bw;
     private boolean field_82190_bM;
@@ -234,7 +234,7 @@
                     if (this.field_70954_d != null && this.field_82189_bL != null)
                     {
                         this.field_70170_p.func_72960_a(this, (byte)14);
-                        this.field_70954_d.func_82688_a(this.field_82189_bL, 1);
+                        this.field_70954_d.modifyPlayerReputation(this.field_82189_bL, 1);
                     }
                 }
 
@@ -255,7 +255,7 @@
             itemstack.func_111282_a(p_184645_1_, this, p_184645_2_);
             return true;
         }
-        else if (!this.func_190669_a(itemstack, this.getClass()) && this.func_70089_S() && !this.func_70940_q() && !this.func_70631_g_())
+        else if (!this.func_190669_a(itemstack, this.getClass()) && this.func_70089_S() && !this.func_70940_q() && !this.func_70631_g_() && !p_184645_1_.func_70093_af())
         {
             if (this.field_70963_i == null)
             {
@@ -327,6 +327,7 @@
     {
         super.func_70014_b(p_70014_1_);
         p_70014_1_.func_74768_a("Profession", this.func_70946_n());
+        p_70014_1_.func_74778_a("ProfessionName", this.getProfessionForge().getRegistryName().toString());
         p_70014_1_.func_74768_a("Riches", this.field_70956_bz);
         p_70014_1_.func_74768_a("Career", this.field_175563_bv);
         p_70014_1_.func_74768_a("CareerLevel", this.field_175562_bw);
@@ -356,6 +357,14 @@
     {
         super.func_70037_a(p_70037_1_);
         this.func_70938_b(p_70037_1_.func_74762_e("Profession"));
+        if (p_70037_1_.func_74764_b("ProfessionName"))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession p =
+                net.minecraftforge.fml.common.registry.ForgeRegistries.VILLAGER_PROFESSIONS.getValue(new net.minecraft.util.ResourceLocation(p_70037_1_.func_74779_i("ProfessionName")));
+            if (p == null)
+                p = net.minecraftforge.fml.common.registry.ForgeRegistries.VILLAGER_PROFESSIONS.getValue(new net.minecraft.util.ResourceLocation("minecraft:farmer"));
+            this.setProfession(p);
+        }
         this.field_70956_bz = p_70037_1_.func_74762_e("Riches");
         this.field_175563_bv = p_70037_1_.func_74762_e("Career");
         this.field_175562_bw = p_70037_1_.func_74762_e("CareerLevel");
@@ -412,13 +421,43 @@
     public void func_70938_b(int p_70938_1_)
     {
         this.field_70180_af.func_187227_b(field_184752_bw, Integer.valueOf(p_70938_1_));
+        net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, p_70938_1_);
     }
 
+    @Deprecated //Use Forge Variant below
     public int func_70946_n()
     {
-        return Math.max(((Integer)this.field_70180_af.func_187225_a(field_184752_bw)).intValue() % 6, 0);
+        return Math.max(((Integer)this.field_70180_af.func_187225_a(field_184752_bw)).intValue(), 0);
     }
 
+    private net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof;
+    public void setProfession(net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof)
+    {
+        this.prof = prof;
+        this.func_70938_b(net.minecraftforge.fml.common.registry.VillagerRegistry.getId(prof));
+    }
+
+    public net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession getProfessionForge()
+    {
+        if (this.prof == null)
+        {
+            this.prof = net.minecraftforge.fml.common.registry.VillagerRegistry.getById(this.func_70946_n());
+            if (this.prof == null)
+                return net.minecraftforge.fml.common.registry.VillagerRegistry.getById(0); //Farmer
+        }
+        return this.prof;
+    }
+
+    @Override
+    public void func_184206_a(DataParameter<?> key)
+    {
+        super.func_184206_a(key);
+        if (key.equals(field_184752_bw))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, this.field_70180_af.func_187225_a(field_184752_bw));
+        }
+    }
+
     public boolean func_70941_o()
     {
         return this.field_70952_f;
@@ -456,7 +495,7 @@
                     i = -3;
                 }
 
-                this.field_70954_d.func_82688_a(p_70604_1_.func_70005_c_(), i);
+                this.field_70954_d.modifyPlayerReputation(p_70604_1_.func_110124_au(), i);
 
                 if (this.func_70089_S())
                 {
@@ -476,7 +515,7 @@
             {
                 if (entity instanceof EntityPlayer)
                 {
-                    this.field_70954_d.func_82688_a(entity.func_70005_c_(), -2);
+                    this.field_70954_d.modifyPlayerReputation(entity.func_110124_au(), -2);
                 }
                 else if (entity instanceof IMob)
                 {
@@ -569,7 +608,7 @@
 
             if (this.field_70962_h != null)
             {
-                this.field_82189_bL = this.field_70962_h.func_70005_c_();
+                this.field_82189_bL = this.field_70962_h.func_110124_au();
             }
             else
             {
@@ -612,20 +651,18 @@
             this.func_175554_cu();
         }
 
-        return this.field_70963_i;
+        return net.minecraftforge.event.ForgeEventFactory.listTradeOffers(this, p_70934_1_, field_70963_i);
     }
 
     private void func_175554_cu()
     {
-        EntityVillager.ITradeList[][][] aentityvillager$itradelist = field_175561_bA[this.func_70946_n()];
-
         if (this.field_175563_bv != 0 && this.field_175562_bw != 0)
         {
             ++this.field_175562_bw;
         }
         else
         {
-            this.field_175563_bv = this.field_70146_Z.nextInt(aentityvillager$itradelist.length) + 1;
+            this.field_175563_bv = this.getProfessionForge().getRandomCareer(this.field_70146_Z) + 1;
             this.field_175562_bw = 1;
         }
 
@@ -636,19 +673,13 @@
 
         int i = this.field_175563_bv - 1;
         int j = this.field_175562_bw - 1;
+        java.util.List<EntityVillager.ITradeList> trades = this.getProfessionForge().getCareer(i).getTrades(j);
 
-        if (i >= 0 && i < aentityvillager$itradelist.length)
+        if (trades != null)
         {
-            EntityVillager.ITradeList[][] aentityvillager$itradelist1 = aentityvillager$itradelist[i];
-
-            if (j >= 0 && j < aentityvillager$itradelist1.length)
+            for (EntityVillager.ITradeList entityvillager$itradelist : trades)
             {
-                EntityVillager.ITradeList[] aentityvillager$itradelist2 = aentityvillager$itradelist1[j];
-
-                for (EntityVillager.ITradeList entityvillager$itradelist : aentityvillager$itradelist2)
-                {
-                    entityvillager$itradelist.func_190888_a(this, this.field_70963_i, this.field_70146_Z);
-                }
+                entityvillager$itradelist.func_190888_a(this, this.field_70963_i, this.field_70146_Z);
             }
         }
     }
@@ -758,7 +789,7 @@
                     s1 = "nitwit";
             }
 
-            if (s1 != null)
+            s1 = this.getProfessionForge().getCareer(this.field_175563_bv-1).getName();
             {
                 ITextComponent itextcomponent = new TextComponentTranslation("entity.Villager." + s1, new Object[0]);
                 itextcomponent.func_150256_b().func_150209_a(this.func_174823_aP());
@@ -771,10 +802,6 @@
 
                 return itextcomponent;
             }
-            else
-            {
-                return super.func_145748_c_();
-            }
         }
     }
 
@@ -828,7 +855,7 @@
 
         if (p_190672_3_)
         {
-            this.func_70938_b(this.field_70170_p.field_73012_v.nextInt(6));
+            net.minecraftforge.fml.common.registry.VillagerRegistry.setRandomProfession(this, this.field_70170_p.field_73012_v);
         }
 
         this.func_175552_ct();
@@ -1089,6 +1116,9 @@
                 p_190888_2_.add(new MerchantRecipe(itemstack, itemstack1));
             }
         }
+    //MODDERS DO NOT USE OR EDIT THIS IN ANY WAY IT WILL HAVE NO EFFECT, THIS IS JUST IN HERE TO ALLOW FORGE TO ACCESS IT
+    @Deprecated
+    public static ITradeList[][][][] GET_TRADES_DONT_USE(){ return field_175561_bA; }
 
     public static class ListItemForEmeralds implements EntityVillager.ITradeList
         {
