--- ../src-base/minecraft/net/minecraft/entity/passive/EntityRabbit.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityRabbit.java
@@ -85,6 +85,7 @@
         this.field_70714_bg.func_75776_a(5, new EntityRabbit.AIRaidFarm(this));
         this.field_70714_bg.func_75776_a(6, new EntityAIWanderAvoidWater(this, 0.6D));
         this.field_70714_bg.func_75776_a(11, new EntityAIWatchClosest(this, EntityPlayer.class, 10.0F));
+        super.func_184651_r();
     }
 
     protected float func_175134_bD()
