--- a/net/minecraft/entity/passive/EntitySheep.java
+++ b/net/minecraft/entity/passive/EntitySheep.java
@@ -50,7 +50,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public class EntitySheep extends EntityAnimal {
+public class EntitySheep extends EntityAnimal implements net.minecraftforge.common.IShearable {
    private static final DataParameter<Byte> field_184774_bv = EntityDataManager.func_187226_a(EntitySheep.class, DataSerializers.field_187191_a);
    private final InventoryCrafting field_90016_e = new InventoryCrafting(new Container() {
       public boolean func_75145_c(EntityPlayer p_75145_1_) {
@@ -214,7 +214,7 @@
 
    public boolean func_184645_a(EntityPlayer p_184645_1_, EnumHand p_184645_2_) {
       ItemStack itemstack = p_184645_1_.func_184586_b(p_184645_2_);
-      if (itemstack.func_77973_b() == Items.field_151097_aZ && !this.func_70892_o() && !this.func_70631_g_()) {
+      if (false && itemstack.func_77973_b() == Items.field_151097_aZ && !this.func_70892_o() && !this.func_70631_g_()) { //Forge: Moved to onSheared
          if (!this.field_70170_p.field_72995_K) {
             this.func_70893_e(true);
             int i = 1 + this.field_70146_Z.nextInt(3);
@@ -329,7 +329,7 @@
       EnumDyeColor enumdyecolor1 = ((EntitySheep)p_175511_2_).func_175509_cj();
       this.field_90016_e.func_70299_a(0, new ItemStack(ItemDye.func_195961_a(enumdyecolor)));
       this.field_90016_e.func_70299_a(1, new ItemStack(ItemDye.func_195961_a(enumdyecolor1)));
-      ItemStack itemstack = p_175511_1_.field_70170_p.func_199532_z().func_199514_a(this.field_90016_e, ((EntitySheep)p_175511_1_).field_70170_p);
+      ItemStack itemstack = p_175511_1_.field_70170_p.func_199532_z().getResult(this.field_90016_e, ((EntitySheep)p_175511_1_).field_70170_p, net.minecraftforge.common.crafting.VanillaRecipeTypes.CRAFTING);
       Item item = itemstack.func_77973_b();
       EnumDyeColor enumdyecolor2;
       if (item instanceof ItemDye) {
@@ -344,4 +344,23 @@
    public float func_70047_e() {
       return 0.95F * this.field_70131_O;
    }
+
+   @Override
+   public boolean isShearable(ItemStack item, net.minecraft.world.IWorldReader world, BlockPos pos) {
+      return !this.func_70892_o() && !this.func_70631_g_();
+   }
+
+   @Override
+   public java.util.List<ItemStack> onSheared(ItemStack item, net.minecraft.world.IWorld world, BlockPos pos, int fortune) {
+      java.util.List<ItemStack> ret = new java.util.ArrayList<>();
+      if (!this.field_70170_p.field_72995_K) {
+         this.func_70893_e(true);
+         int i = 1 + this.field_70146_Z.nextInt(3);
+
+         for(int j = 0; j < i; ++j) {
+            ret.add(new ItemStack(field_200206_bz.get(this.func_175509_cj())));
+         }
+      }
+      return ret;
+   }
 }
