--- ../src-base/minecraft/net/minecraft/entity/passive/EntityOcelot.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityOcelot.java
@@ -298,7 +298,7 @@
 
                 Block block = this.worldObj.getBlock(i, j - 1, k);
 
-                if (block == Blocks.grass || block.getMaterial() == Material.leaves)
+                if (block == Blocks.grass || block.isLeaves(worldObj, i, j - 1, k))
                 {
                     return true;
                 }
