--- ../src-base/minecraft/net/minecraft/entity/passive/EntityHorse.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityHorse.java
@@ -436,6 +436,7 @@
 
         this.field_110296_bG.func_110134_a(this);
         this.func_110232_cE();
+        this.itemHandler = new net.minecraftforge.items.wrapper.InvWrapper(this.field_110296_bG);
     }
 
     private void func_110232_cE()
@@ -1204,6 +1205,7 @@
                 }
 
                 this.field_110277_bt = 0.0F;
+                net.minecraftforge.common.ForgeHooks.onLivingJump(this);
             }
 
             this.field_70747_aH = this.func_70689_ay() * 0.1F;
@@ -1798,4 +1800,21 @@
                 this.field_188477_b = p_i46589_2_;
             }
         }
+
+    // FORGE
+    private net.minecraftforge.items.IItemHandler itemHandler = null; // Initialized by initHorseChest above.
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, net.minecraft.util.EnumFacing facing)
+    {
+        if (capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) return (T) itemHandler;
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, net.minecraft.util.EnumFacing facing)
+    {
+        return capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing);
+    }
 }
