--- ../src-base/minecraft/net/minecraft/entity/passive/HorseArmorType.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/HorseArmorType.java
@@ -25,6 +25,14 @@
         this.field_188587_f = "";
     }
 
+    //Allows for textures located outside the vanilla horse armor folder
+    private  HorseArmorType(String textureLocation, String hash, int armorStrengthIn) 
+    {
+        this.field_188588_g = armorStrengthIn;
+        this.field_188586_e = textureLocation;
+        this.field_188587_f = hash; 
+    }
+    
     private HorseArmorType(int p_i46800_3_, String p_i46800_4_, String p_i46800_5_)
     {
         this.field_188588_g = p_i46800_3_;
@@ -55,34 +63,42 @@
         return this.field_188586_e;
     }
 
+    /**Exists only for vanilla compat purposes. Modders should never call this. Use getByName instead.*/
     public static HorseArmorType func_188575_a(int p_188575_0_)
     {
         return values()[p_188575_0_];
     }
-
-    public static HorseArmorType func_188580_a(ItemStack p_188580_0_)
+    
+    public static HorseArmorType getByName(String name)
     {
-        return p_188580_0_.func_190926_b() ? NONE : func_188576_a(p_188580_0_.func_77973_b());
+        HorseArmorType type = HorseArmorType.valueOf(name);
+        return type != null ? type : NONE;
     }
 
-    public static HorseArmorType func_188576_a(Item p_188576_0_)
+    public static HorseArmorType func_188580_a(ItemStack p_188580_0_)
     {
-        if (p_188576_0_ == Items.field_151138_bX)
+        if(p_188580_0_.func_190926_b())
         {
+            return NONE;
+        }
+        else if (p_188580_0_.func_77973_b() == Items.field_151138_bX)
+        {
             return IRON;
         }
-        else if (p_188576_0_ == Items.field_151136_bY)
+        else if (p_188580_0_.func_77973_b() == Items.field_151136_bY)
         {
             return GOLD;
         }
-        else
+        else if (p_188580_0_.func_77973_b() == Items.field_151125_bZ)
         {
-            return p_188576_0_ == Items.field_151125_bZ ? DIAMOND : NONE;
+            return DIAMOND;
         }
+        else
+            return p_188580_0_.func_77973_b().isValidHorseArmor(p_188580_0_) ? p_188580_0_.func_77973_b().getHorseArmorType(p_188580_0_) : NONE;
     }
-
-    public static boolean func_188577_b(Item p_188577_0_)
+    
+    public static boolean isHorseArmor(ItemStack stack)
     {
-        return func_188576_a(p_188577_0_) != NONE;
+        return func_188580_a(stack) != NONE;
     }
 }
