--- a/net/minecraft/entity/passive/TurtleEntity.java
+++ b/net/minecraft/entity/passive/TurtleEntity.java
@@ -233,8 +233,8 @@
       this.func_184185_a(soundevent, 0.15F, 1.0F);
    }
 
-   public boolean func_204701_dC() {
-      return super.func_204701_dC() && !this.func_203020_dx();
+   public boolean func_213743_em() {
+      return super.func_213743_em() && !this.func_203020_dx();
    }
 
    protected float func_203009_ad() {
