--- a/net/minecraft/client/gui/components/AbstractSelectionList.java
+++ b/net/minecraft/client/gui/components/AbstractSelectionList.java
@@ -514,6 +_,13 @@
       return new ScreenRectangle(this.f_93393_, this.f_93390_, this.f_93392_ - this.f_93393_, this.f_93391_ - this.f_93390_);
    }
 
+   public int getWidth() { return this.f_93388_; }
+   public int getHeight() { return this.f_93389_; }
+   public int getTop() { return this.f_93390_; }
+   public int getBottom() { return this.f_93391_; }
+   public int getLeft() { return this.f_93393_; }
+   public int getRight() { return this.f_93392_; }
+
    @OnlyIn(Dist.CLIENT)
    protected abstract static class Entry<E extends AbstractSelectionList.Entry<E>> implements GuiEventListener {
       /** @deprecated */
