--- a/net/minecraft/client/gui/components/BossHealthOverlay.java
+++ b/net/minecraft/client/gui/components/BossHealthOverlay.java
@@ -36,13 +_,16 @@
 
          for(LerpingBossEvent lerpingbossevent : this.events.values()) {
             int k = i / 2 - 91;
+            var event = net.minecraftforge.client.ForgeHooksClient.onCustomizeBossEventProgress(p_283175_, this.minecraft.getWindow(), lerpingbossevent, k, j, 10 + this.minecraft.font.lineHeight);
+            if (!event.isCanceled()) {
             this.drawBar(p_283175_, k, j, lerpingbossevent);
             Component component = lerpingbossevent.getName();
             int l = this.minecraft.font.width(component);
             int i1 = i / 2 - l / 2;
             int j1 = j - 9;
             p_283175_.drawString(this.minecraft.font, component, i1, j1, 16777215);
-            j += 10 + 9;
+            }
+            j += event.getIncrement();
             if (j >= p_283175_.guiHeight() / 3) {
                break;
             }
