--- a/net/minecraft/client/gui/components/Button.java
+++ b/net/minecraft/client/gui/components/Button.java
@@ -31,6 +_,12 @@
         this.createNarration = p_259552_;
     }
 
+    // Forge (GH-9238)
+    protected Button(Builder builder) {
+        this(builder.x, builder.y, builder.width, builder.height, builder.message, builder.onPress, builder.createNarration);
+        this.setTooltip(builder.tooltip);
+    }
+
     @Override
     public void onPress() {
         this.onPress.onPress(this);
@@ -95,9 +_,12 @@
         }
 
         public Button build() {
-            Button button = new Button(this.x, this.y, this.width, this.height, this.message, this.onPress, this.createNarration);
-            button.setTooltip(this.tooltip);
-            return button;
+            return build(Button::new);
+        }
+
+        // Forge (GH-9238)
+        public Button build(java.util.function.Function<Builder, Button> builder) {
+            return builder.apply(this);
         }
     }
 
