--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -136,8 +_,10 @@
         return panel;
     }
 
+    private final java.util.concurrent.CountDownLatch latch = new java.util.concurrent.CountDownLatch(1);
     public void start() {
         this.logAppenderThread.start();
+        this.latch.countDown();
     }
 
     public void close() {
@@ -151,6 +_,9 @@
     }
 
     public void print(final JTextArea console, final JScrollPane scrollPane, final String line) {
+        try {
+            this.latch.await();
+        } catch (InterruptedException ignored) {} // Prevent logging until after constructor has ended
         if (!SwingUtilities.isEventDispatchThread()) {
             SwingUtilities.invokeLater(() -> this.print(console, scrollPane, line));
         } else {
