--- ../src-base/minecraft/net/minecraft/server/gui/MinecraftServerGui.java
+++ ../src-work/minecraft/net/minecraft/server/gui/MinecraftServerGui.java
@@ -38,7 +38,9 @@
     private static final Font field_164249_a = new Font("Monospaced", 0, 12);
     private static final Logger field_164248_b = LogManager.getLogger();
     private final DedicatedServer field_120021_b;
+    private static JFrame jframe;
 
+    public static JFrame getJFrame() { return jframe; }
     public static void func_120016_a(final DedicatedServer p_120016_0_)
     {
         try
@@ -51,7 +53,7 @@
         }
 
         MinecraftServerGui minecraftservergui = new MinecraftServerGui(p_120016_0_);
-        JFrame jframe = new JFrame("Minecraft server");
+        jframe = new JFrame("Minecraft server");
         jframe.add(minecraftservergui);
         jframe.pack();
         jframe.setLocationRelativeTo((Component)null);
@@ -77,6 +79,7 @@
                 System.exit(0);
             }
         });
+        minecraftservergui.latch.countDown();
     }
 
     public MinecraftServerGui(DedicatedServer p_i2362_1_)
@@ -161,8 +164,13 @@
         return jpanel;
     }
 
+    private java.util.concurrent.CountDownLatch latch = new java.util.concurrent.CountDownLatch(1);
     public void func_164247_a(final JTextArea p_164247_1_, final JScrollPane p_164247_2_, final String p_164247_3_)
     {
+        try
+        {
+            latch.await();
+        } catch (InterruptedException e){} //Prevent logging until after constructor has ended.
         if (!SwingUtilities.isEventDispatchThread())
         {
             SwingUtilities.invokeLater(new Runnable()
