--- ../src-base/minecraft/net/minecraft/entity/monster/EntityEnderman.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityEnderman.java
@@ -259,7 +259,9 @@
 
     private boolean teleportTo(double x, double y, double z)
     {
-        boolean flag = this.func_184595_k(x, y, z);
+        net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(this, x, y, z, 0);
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return false;
+        boolean flag = this.func_184595_k(event.targetX, event.targetY, event.targetZ);
 
         if (flag)
         {
@@ -342,6 +344,18 @@
         }
     }
 
+    /*===================================== Forge Start ==============================*/
+    public static void setCarriable(Block block, boolean canCarry)
+    {
+        if (canCarry) carriableBlocks.add(block);
+        else          carriableBlocks.remove(block);
+    }
+    public static boolean getCarriable(Block block)
+    {
+        return carriableBlocks.contains(block);
+    }
+    /*===================================== Forge End ==============================*/
+
     public boolean isScreaming()
     {
         return ((Boolean)this.dataWatcher.func_187225_a(field_184719_bw)).booleanValue();
