--- a/net/minecraft/client/gui/screens/options/controls/KeyBindsList.java
+++ b/net/minecraft/client/gui/screens/options/controls/KeyBindsList.java
@@ -116,9 +_,10 @@
             this.name = name;
             this.changeButton = Button.builder(name, button -> {
                     KeyBindsList.this.keyBindsScreen.selectedKey = key;
+                    key.setKeyModifierAndCode(null, com.mojang.blaze3d.platform.InputConstants.UNKNOWN);
                     KeyBindsList.this.resetMappingAndUpdateButtons();
                 })
-                .bounds(0, 0, 75, 20)
+                .bounds(0, 0, 75 + 20 /* Forge: Add space */, 20)
                 .createNarration(
                     defaultNarrationSupplier -> key.isUnbound()
                         ? Component.translatable("narrator.controls.unbound", name)
@@ -126,7 +_,7 @@
                 )
                 .build();
             this.resetButton = Button.builder(RESET_BUTTON_TITLE, button -> {
-                key.setKey(key.getDefaultKey());
+                this.key.setToDefault();
                 KeyBindsList.this.resetMappingAndUpdateButtons();
             }).bounds(0, 0, 50, 20).createNarration(defaultNarrationSupplier -> Component.translatable("narrator.controls.reset", name)).build();
             this.refreshEntry();
@@ -167,7 +_,8 @@
             MutableComponent tooltip = Component.empty();
             if (!this.key.isUnbound()) {
                 for (KeyMapping otherKey : KeyBindsList.this.minecraft.options.keyMappings) {
-                    if (otherKey != this.key && this.key.same(otherKey) && (!otherKey.isDefault() || !this.key.isDefault())) {
+                    var vanillConflict = otherKey != this.key && this.key.same(otherKey) && (!otherKey.isDefault() || !this.key.isDefault());
+                    if (vanillConflict || otherKey.hasKeyModifierConflict(this.key)) { // FORGE: gracefully handle conflicts like SHIFT vs SHIFT+G
                         if (this.hasCollision) {
                             tooltip.append(", ");
                         }
