--- ../src-base/minecraft/net/minecraft/network/packet/NetHandler.java
+++ ../src-work/minecraft/net/minecraft/network/packet/NetHandler.java
@@ -1,4 +1,6 @@
 package net.minecraft.network.packet;
+
+import net.minecraft.entity.player.EntityPlayer;
 
 public abstract class NetHandler
 {
@@ -382,4 +384,8 @@
     {
         return false;
     }
+
+    public abstract void handleVanilla250Packet(Packet250CustomPayload payload);
+
+    public abstract EntityPlayer getPlayer();
 }
