--- a/net/minecraft/data/worldgen/Carvers.java
+++ b/net/minecraft/data/worldgen/Carvers.java
@@ -15,6 +_,7 @@
 import net.minecraft.world.level.levelgen.heightproviders.BiasedToBottomHeight;
 import net.minecraft.world.level.levelgen.heightproviders.UniformHeight;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Carvers {
    public static final ConfiguredWorldCarver<CaveCarverConfiguration> f_126848_ = m_126855_("cave", WorldCarver.f_64974_.m_65063_(new CaveCarverConfiguration(0.14285715F, BiasedToBottomHeight.m_161931_(VerticalAnchor.m_158922_(0), VerticalAnchor.m_158922_(127), 8), ConstantFloat.m_146458_(0.5F), VerticalAnchor.m_158930_(10), false, CarverDebugSettings.m_159136_(false, Blocks.f_50669_.m_49966_()), ConstantFloat.m_146458_(1.0F), ConstantFloat.m_146458_(1.0F), ConstantFloat.m_146458_(-0.7F))));
    public static final ConfiguredWorldCarver<CaveCarverConfiguration> f_176865_ = m_126855_("prototype_cave", WorldCarver.f_64974_.m_65063_(new CaveCarverConfiguration(0.33333334F, UniformHeight.m_162034_(VerticalAnchor.m_158930_(8), VerticalAnchor.m_158922_(126)), UniformFloat.m_146605_(0.1F, 0.9F), VerticalAnchor.m_158930_(8), false, CarverDebugSettings.m_159136_(false, Blocks.f_50669_.m_49966_()), UniformFloat.m_146605_(0.7F, 1.4F), UniformFloat.m_146605_(0.8F, 1.3F), UniformFloat.m_146605_(-1.0F, -0.4F))));
