--- a/net/minecraft/data/worldgen/SurfaceBuilders.java
+++ b/net/minecraft/data/worldgen/SurfaceBuilders.java
@@ -7,6 +_,7 @@
 import net.minecraft.world.level.levelgen.surfacebuilders.SurfaceBuilderBaseConfiguration;
 import net.minecraft.world.level.levelgen.surfacebuilders.SurfaceBuilderConfiguration;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class SurfaceBuilders {
    public static final ConfiguredSurfaceBuilder<SurfaceBuilderBaseConfiguration> f_127276_ = m_127300_("badlands", SurfaceBuilder.f_75168_.m_75223_(SurfaceBuilder.f_75206_));
    public static final ConfiguredSurfaceBuilder<SurfaceBuilderBaseConfiguration> f_127277_ = m_127300_("basalt_deltas", SurfaceBuilder.f_75175_.m_75223_(SurfaceBuilder.f_75213_));
