--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -90,7 +_,8 @@
       this.m_6478_(MoverType.SELF, this.m_20184_());
       float f = 0.98F;
       if (this.m_20096_()) {
-         f = this.m_9236_().m_8055_(this.m_20099_()).m_60734_().m_49958_() * 0.98F;
+         BlockPos pos = m_20099_();
+         f = this.m_9236_().m_8055_(pos).getFriction(this.m_9236_(), pos, this) * 0.98F;
       }
 
       this.m_20256_(this.m_20184_().m_82542_((double)f, 0.98D, (double)f));
@@ -172,6 +_,7 @@
    }
 
    public boolean m_6469_(DamageSource p_20785_, float p_20786_) {
+      if (this.m_9236_().f_46443_ || this.m_213877_()) return false; //Forge: Fixes MC-53850
       if (this.m_6673_(p_20785_)) {
          return false;
       } else if (this.m_9236_().f_46443_) {
@@ -204,6 +_,7 @@
    public void m_6123_(Player p_20792_) {
       if (!this.m_9236_().f_46443_) {
          if (p_20792_.f_36101_ == 0) {
+            if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerXpEvent.PickupXp(p_20792_, this))) return;
             p_20792_.f_36101_ = 2;
             p_20792_.m_7938_(this, 1);
             int i = this.m_147092_(p_20792_, this.f_20770_);
@@ -224,7 +_,7 @@
       Map.Entry<EquipmentSlot, ItemStack> entry = EnchantmentHelper.m_44839_(Enchantments.f_44962_, p_147093_, ItemStack::m_41768_);
       if (entry != null) {
          ItemStack itemstack = entry.getValue();
-         int i = Math.min(this.m_20798_(this.f_20770_), itemstack.m_41773_());
+         int i = Math.min((int) (this.f_20770_ * itemstack.getXpRepairRatio()), itemstack.m_41773_());
          itemstack.m_41721_(itemstack.m_41773_() - i);
          int j = p_147094_ - this.m_20793_(i);
          return j > 0 ? this.m_147092_(p_147093_, j) : 0;
