--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -138,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public abstract class LivingEntity extends Entity implements Attackable, WaypointTransmitter {
+public abstract class LivingEntity extends Entity implements Attackable, WaypointTransmitter, net.minecraftforge.common.extensions.IForgeLivingEntity {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final String TAG_ACTIVE_EFFECTS = "active_effects";
     public static final String TAG_ATTRIBUTES = "attributes";
@@ -209,6 +_,8 @@
     public static final float ELYTRA_VERTICAL_AIR_DRAG = 0.98F;
     public static final float BASE_SWIM_SPEED = 0.02F;
     private int currentImpulseContextResetGraceTime = 0;
+    /** Forge: Use a variant that calls {@link ItemStack#isMonsterDisguise(Player, net.minecraft.world.entity.monster.Monster)} and {@link net.minecraftforge.event.entity.living.MonsterDisguiseEvent} */
+    @Deprecated
     public static final Predicate<LivingEntity> PLAYER_NOT_WEARING_DISGUISE_ITEM = livingEntity -> {
         if (livingEntity instanceof Player player) {
             ItemStack helmet = player.getItemBySlot(EquipmentSlot.HEAD);
@@ -292,7 +_,7 @@
         this.reapplyPosition();
         this.setYRot(this.random.nextFloat() * (float) (Math.PI * 2));
         this.yHeadRot = this.getYRot();
-        this.brain = this.makeBrain(Brain.Packed.EMPTY);
+        this.brain = net.minecraftforge.common.ForgeHooks.onLivingMakeBrain(this, this.makeBrain(Brain.Packed.EMPTY), Brain.Packed.EMPTY);
     }
 
     @Override
@@ -356,7 +_,8 @@
             .add(Attributes.AIR_DRAG_MODIFIER)
             .add(Attributes.FRICTION_MODIFIER)
             .add(Attributes.NAME_TAG_DISTANCE)
-            .add(Attributes.BELOW_NAME_DISTANCE);
+            .add(Attributes.BELOW_NAME_DISTANCE)
+            .add(Attributes.JUMP_STRENGTH);
     }
 
     @Override
@@ -383,7 +_,8 @@
 
                 double scale = Math.min(0.2F + power / 15.0, 2.5);
                 int particles = (int)(150.0 * scale);
-                level.sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, onState), x, y, z, particles, 0.0, 0.0, 0.0, 0.15F);
+                if (!onState.addLandingEffects((ServerLevel) this.level(), pos, onState, this, particles))
+                ((ServerLevel)this.level()).sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, onState).setPos(pos), x, y, z, particles, 0.0, 0.0, 0.0, 0.15F);
             }
         }
 
@@ -393,6 +_,7 @@
         }
     }
 
+    @Deprecated //FORGE: Use canDrownInFluid instead
     public boolean canBreatheUnderwater() {
         return this.is(EntityTypeTags.CAN_BREATHE_UNDER_WATER);
     }
@@ -433,6 +_,9 @@
                 }
             }
 
+            int airSupply = this.getAirSupply();
+            net.minecraftforge.common.ForgeHooks.onLivingBreathe(this, airSupply - decreaseAirSupply(airSupply), increaseAirSupply(airSupply) - airSupply);
+            if (false) // Forge: Handled in ForgeHooks#onLivingBreathe(LivingEntity, int, int)
             if (this.isEyeInFluid(FluidTags.WATER)
                 && !level.getBlockState(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())).is(Blocks.BUBBLE_COLUMN)) {
                 boolean canDrownInWater = !this.canBreatheUnderwater()
@@ -788,6 +_,9 @@
 
         ItemEntity entity = this.createItemStackToDrop(itemStack, randomly, thrownFromHand);
         if (entity != null) {
+                if (captureDrops() != null)
+                    captureDrops().add(entity);
+                else
             this.level().addFreshEntity(entity);
         }
 
@@ -828,7 +_,7 @@
                 this.setPosToBed(sleepingPos);
             }
         }, this::clearSleepingPos);
-        input.read("Brain", Brain.Packed.CODEC).ifPresent(packedBrain -> this.brain = this.makeBrain(packedBrain));
+        input.read("Brain", Brain.Packed.CODEC).ifPresent(packedBrain -> this.brain = net.minecraftforge.common.ForgeHooks.onLivingMakeBrain(this, this.makeBrain(packedBrain), packedBrain));
         this.lastHurtByPlayer = EntityReference.read(input, "last_hurt_by_player");
         this.lastHurtByPlayerMemoryTime = input.getIntOr("last_hurt_by_player_memory_time", 0);
         this.lastHurtByMob = EntityReference.read(input, "last_hurt_by_mob");
@@ -854,8 +_,10 @@
                     Holder<MobEffect> mobEffect = iterator.next();
                     MobEffectInstance effect = this.activeEffects.get(mobEffect);
                     if (!effect.tickServer(serverLevel, this, () -> this.onEffectUpdated(effect, true, null))) {
+                    if (!net.minecraftforge.event.ForgeEventFactory.onLivingEffectExpire(this, effect)) {
                         iterator.remove();
                         this.onEffectsRemoved(List.of(effect));
+                    }
                     } else if (effect.getDuration() % 600 == 0) {
                         this.onEffectUpdated(effect, false, null);
                     }
@@ -942,6 +_,7 @@
             }
         }
 
+        visibilityPercent = net.minecraftforge.common.ForgeHooks.getEntityVisibilityMultiplier(this, targetingEntity, visibilityPercent);
         return visibilityPercent;
     }
 
@@ -1018,6 +_,7 @@
 
         MobEffectInstance effect = this.activeEffects.get(newEffect.getEffect());
         boolean changed = false;
+        net.minecraftforge.event.ForgeEventFactory.onLivingEffectAdd(this, effect, newEffect, source);
         if (effect == null) {
             this.activeEffects.put(newEffect.getEffect(), newEffect);
             this.onEffectAdded(newEffect, source);
@@ -1033,6 +_,10 @@
     }
 
     public boolean canBeAffected(final MobEffectInstance newEffect) {
+        var eventResult = net.minecraftforge.event.ForgeEventFactory.onLivingEffectCanApply(this, newEffect).getResult();
+        if (!eventResult.isDefault()) {
+            return eventResult.isAllowed();
+        }
         if (this.is(EntityTypeTags.IMMUNE_TO_INFESTED)) {
             return !newEffect.is(MobEffects.INFESTED);
         } else if (this.is(EntityTypeTags.IMMUNE_TO_OOZING)) {
@@ -1063,6 +_,9 @@
     }
 
     public boolean removeEffect(final Holder<MobEffect> effect) {
+        if (net.minecraftforge.event.ForgeEventFactory.onLivingEffectRemove(this, effect.get())) {
+            return false;
+        }
         MobEffectInstance effectInstance = this.removeEffectNoUpdate(effect);
         if (effectInstance != null) {
             this.onEffectsRemoved(List.of(effectInstance));
@@ -1107,6 +_,9 @@
             this.effectsDirty = true;
 
             for (MobEffectInstance effect : effects) {
+                if (net.minecraftforge.event.ForgeEventFactory.onLivingEffectRemove(this, effect)) {
+                    continue;
+                }
                 effect.getEffect().value().removeAttributeModifiers(this.getAttributes());
 
                 for (Entity passenger : this.getPassengers()) {
@@ -1154,9 +_,13 @@
     }
 
     public void heal(final float heal) {
+        var ammount = net.minecraftforge.event.ForgeEventFactory.onLivingHeal(this, heal);
+        if (ammount <= 0) {
+            return;
+        }
         float health = this.getHealth();
         if (health > 0.0F) {
-            this.setHealth(health + heal);
+            this.setHealth(health + ammount);
         }
     }
 
@@ -1174,6 +_,9 @@
 
     @Override
     public boolean hurtServer(final ServerLevel level, final DamageSource source, float damage) {
+        if (!net.minecraftforge.common.ForgeHooks.onLivingAttack(this, source, damage)) {
+            return false;
+        }
         if (this.isInvulnerableTo(level, source)) {
             return false;
         }
@@ -1332,6 +_,10 @@
                 }
 
                 float damageBlocked = blocksAttacks.resolveBlockedDamage(source, damage, angle);
+                var ev = net.minecraftforge.event.ForgeEventFactory.onShieldBlock(this, source, damageBlocked, blockingWith);
+                if (ev == null) return 0.0F;
+                damageBlocked = ev.getBlockedDamage();
+                if (ev.shieldTakesDamage())
                 blocksAttacks.hurtBlockingItem(this.level(), blockingWith, this, this.getUsedItemHand(), damageBlocked);
                 if (damageBlocked > 0.0F && !source.is(DamageTypeTags.IS_PROJECTILE) && source.getDirectEntity() instanceof LivingEntity livingEntity) {
                     this.blockUsingItem(level, livingEntity, source, damage);
@@ -1363,7 +_,7 @@
         Entity sourceEntity = source.getEntity();
         if (sourceEntity instanceof Player playerSource) {
             this.setLastHurtByPlayer(playerSource, 100);
-        } else if (sourceEntity instanceof Wolf wolf && wolf.isTame()) {
+        } else if (sourceEntity instanceof net.minecraft.world.entity.TamableAnimal wolf && wolf.isTame()) {
             if (wolf.getOwnerReference() != null) {
                 this.setLastHurtByPlayer(wolf.getOwnerReference().getUUID(), 100);
             } else {
@@ -1394,7 +_,7 @@
         for (InteractionHand hand : InteractionHand.values()) {
             ItemStack itemStack = this.getItemInHand(hand);
             protection = itemStack.get(DataComponents.DEATH_PROTECTION);
-            if (protection != null) {
+            if (protection != null && net.minecraftforge.common.ForgeHooks.onLivingUseTotem(this, killingDamage, itemStack, hand)) {
                 protectionItem = itemStack.copy();
                 itemStack.shrink(1);
                 break;
@@ -1449,6 +_,7 @@
     }
 
     public void die(final DamageSource source) {
+        if (net.minecraftforge.event.ForgeEventFactory.onLivingDeath(this, source)) return;
         if (!this.isRemoved() && !this.dead) {
             Entity sourceEntity = source.getEntity();
             LivingEntity killer = this.getKillCredit();
@@ -1489,10 +_,10 @@
         if (this.level() instanceof ServerLevel serverLevel) {
             boolean var6 = false;
             if (killer instanceof WitherBoss) {
-                if (serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+                if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(serverLevel, killer)) {
                     BlockPos pos = this.blockPosition();
                     BlockState state = Blocks.WITHER_ROSE.defaultBlockState();
-                    if (this.level().getBlockState(pos).isAir() && state.canSurvive(this.level(), pos)) {
+                    if (this.level().isEmptyBlock(pos) && state.canSurvive(this.level(), pos)) {
                         this.level().setBlock(pos, state, 3);
                         var6 = true;
                     }
@@ -1507,6 +_,7 @@
     }
 
     protected void dropAllDeathLoot(final ServerLevel level, final DamageSource source) {
+        this.captureDrops(new java.util.ArrayList<>());
         boolean playerKilled = this.lastHurtByPlayerMemoryTime > 0;
         if (this.shouldDropLoot(level)) {
             this.dropFromLootTable(level, source, playerKilled);
@@ -1515,6 +_,11 @@
 
         this.dropEquipment(level);
         this.dropExperience(level, source.getEntity());
+
+        var drops = captureDrops(null);
+        if (!net.minecraftforge.event.ForgeEventFactory.onLivingDrops(this, source, drops, playerKilled)) {
+            drops.forEach(e -> level().addFreshEntity(e));
+        }
     }
 
     protected void dropEquipment(final ServerLevel level) {
@@ -1526,7 +_,8 @@
                 this.isAlwaysExperienceDropper()
                     || this.lastHurtByPlayerMemoryTime > 0 && this.shouldDropExperience() && level.getGameRules().get(GameRules.MOB_DROPS)
             )) {
-            ExperienceOrb.award(level, this.position(), this.getExperienceReward(level, killer));
+            int reward = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.getLastHurtByPlayer(), this.getExperienceReward(level, killer));
+            ExperienceOrb.award(level, this.position(), reward);
         }
     }
 
@@ -1639,6 +_,11 @@
     }
 
     public void knockback(double power, double xd, double zd, final DamageSource source, final float damage, final boolean comesFromEffect) {
+        var event = net.minecraftforge.event.ForgeEventFactory.onLivingKnockBack(this, (float)power, xd, zd);
+        if (event == null) return;
+        power = event.getStrength();
+        xd = event.getRatioX();
+        zd = event.getRatioZ();
         power *= 1.0 - this.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE);
         if (!(power <= 0.0)) {
             this.needsSync = true;
@@ -1724,6 +_,13 @@
         } else {
             BlockPos ladderCheckPos = this.blockPosition();
             BlockState state = this.getInBlockState();
+            var ladderPos = net.minecraftforge.common.ForgeHooks.isLivingOnLadder(state, level(), ladderCheckPos, this);
+            if (ladderPos.isPresent()) {
+                this.lastClimbablePos = ladderPos;
+                return true;
+            } else if (ladderPos != null) {
+                return false;
+            }
             if (this.isFallFlying() && state.is(BlockTags.CAN_GLIDE_THROUGH)) {
                 return false;
             } else if (state.is(BlockTags.CLIMBABLE)) {
@@ -1788,9 +_,11 @@
 
     @Override
     public boolean causeFallDamage(final double fallDistance, final float damageModifier, final DamageSource damageSource) {
+        var event = net.minecraftforge.event.ForgeEventFactory.onLivingFall(this, fallDistance, damageModifier);
+        if (event == null) return false;
         double effectiveFallDistance;
         if (this.isIgnoringFallDamageFromCurrentImpulse()) {
-            effectiveFallDistance = Math.min(fallDistance, this.currentImpulseImpactPos.y - this.getY());
+            effectiveFallDistance = Math.min(event.getDistance(), this.currentImpulseImpactPos.y - this.getY());
             boolean hasLandedAboveCurrentImpulseImpactPosY = effectiveFallDistance <= 0.0;
             if (hasLandedAboveCurrentImpulseImpactPosY) {
                 this.resetCurrentImpulseContext();
@@ -1798,11 +_,11 @@
                 this.tryResetCurrentImpulseContext();
             }
         } else {
-            effectiveFallDistance = fallDistance;
+            effectiveFallDistance = event.getDistance();
         }
 
-        boolean damaged = super.causeFallDamage(effectiveFallDistance, damageModifier, damageSource);
-        int dmg = this.calculateFallDamage(effectiveFallDistance, damageModifier);
+        boolean damaged = super.causeFallDamage(effectiveFallDistance, event.getDamageMultiplier(), damageSource);
+        int dmg = this.calculateFallDamage(effectiveFallDistance, event.getDamageMultiplier());
         if (dmg > 0) {
             this.resetCurrentImpulseContext();
             this.playSound(this.getFallDamageSound(dmg), 1.0F, 1.0F);
@@ -1864,9 +_,10 @@
             int xx = Mth.floor(this.getX());
             int yy = Mth.floor(this.getY() - 0.2F);
             int zz = Mth.floor(this.getZ());
-            BlockState state = this.level().getBlockState(new BlockPos(xx, yy, zz));
+            BlockPos pos = new BlockPos(xx, yy, zz);
+            BlockState state = this.level().getBlockState(pos);
             if (!state.isAir()) {
-                SoundType soundType = state.getSoundType();
+                SoundType soundType = state.getSoundType(level(), pos, this);
                 this.playSound(soundType.getFallSound(), soundType.getVolume() * 0.5F, soundType.getPitch() * 0.75F);
             }
         }
@@ -1927,9 +_,9 @@
             float damageResisted = oldDamage - damage;
             if (damageResisted > 0.0F && damageResisted < 3.4028235E37F) {
                 if (this instanceof ServerPlayer serverPlayer) {
-                    serverPlayer.awardStat(Stats.DAMAGE_RESISTED, Math.round(damageResisted * 10.0F));
+                    serverPlayer.awardStat(Stats.CUSTOM.get(Stats.DAMAGE_RESISTED), Math.round(damageResisted * 10.0F));
                 } else if (damageSource.getEntity() instanceof ServerPlayer) {
-                    ((ServerPlayer)damageSource.getEntity()).awardStat(Stats.DAMAGE_DEALT_RESISTED, Math.round(damageResisted * 10.0F));
+                    ((ServerPlayer)damageSource.getEntity()).awardStat(Stats.CUSTOM.get(Stats.DAMAGE_DEALT_RESISTED), Math.round(damageResisted * 10.0F));
                 }
             }
         }
@@ -1958,6 +_,8 @@
 
     protected void actuallyHurt(final ServerLevel level, final DamageSource source, float dmg) {
         if (!this.isInvulnerableTo(level, source)) {
+            dmg = net.minecraftforge.common.ForgeHooks.onLivingHurt(this, source, dmg);
+            if (dmg <= 0) return;
             dmg = this.getDamageAfterArmorAbsorb(source, dmg);
             dmg = this.getDamageAfterMagicAbsorb(source, dmg);
             float originalDamage = dmg;
@@ -1968,6 +_,7 @@
                 serverPlayer.awardStat(Stats.DAMAGE_DEALT_ABSORBED, Math.round(absorbedDamage * 10.0F));
             }
 
+            dmg = net.minecraftforge.common.ForgeHooks.onLivingDamage(this, source, dmg);
             if (dmg != 0.0F) {
                 this.getCombatTracker().recordDamage(source, dmg);
                 this.setHealth(this.getHealth() - dmg);
@@ -2031,6 +_,8 @@
     }
 
     public void swing(final InteractionHand hand, final boolean sendToSwingingEntity) {
+        ItemStack stack = this.getItemInHand(hand);
+        if (!stack.isEmpty() && stack.onEntitySwing(this)) return;
         if (!this.swinging || this.swingTime >= this.getCurrentSwingDuration() / 2 || this.swingTime < 0) {
             this.swingTime = -1;
             this.swinging = true;
@@ -2172,9 +_,10 @@
     }
 
     private void swapHandItems() {
-        ItemStack tmp = this.getItemBySlot(EquipmentSlot.OFFHAND);
-        this.setItemSlot(EquipmentSlot.OFFHAND, this.getItemBySlot(EquipmentSlot.MAINHAND));
-        this.setItemSlot(EquipmentSlot.MAINHAND, tmp);
+        var event = net.minecraftforge.event.ForgeEventFactory.onLivingSwapHandItems(this);
+        if (event == null) return;
+        this.setItemSlot(EquipmentSlot.OFFHAND, event.getItemSwappedToOffHand());
+        this.setItemSlot(EquipmentSlot.MAINHAND, event.getItemSwappedToMainHand());
     }
 
     @Override
@@ -2396,15 +_,18 @@
             }
 
             this.needsSync = true;
+            net.minecraftforge.common.ForgeHooks.onLivingJump(this);
         }
     }
 
+    @Deprecated // FORGE: use sinkInFluid instead
     protected void goDownInWater() {
-        this.setDeltaMovement(this.getDeltaMovement().add(0.0, -0.04F, 0.0));
+        this.sinkInFluid(net.minecraftforge.common.ForgeMod.WATER_TYPE.get());
     }
 
+    @Deprecated // FORGE: use jumpInFluid instead
     protected void jumpInLiquid(final TagKey<Fluid> type) {
-        this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.04F, 0.0));
+        this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.04F * this.getAttributeValue(net.minecraftforge.common.ForgeMod.SWIM_SPEED.getHolder().get()), 0.0));
     }
 
     protected float getWaterSlowDown() {
@@ -2427,8 +_,9 @@
     }
 
     public void travel(final Vec3 input) {
-        if (this.shouldTravelInFluid(this.level().getFluidState(this.blockPosition()))) {
-            this.travelInFluid(input);
+        var fluidstate = this.level().getFluidState(this.blockPosition());
+        if (this.shouldTravelInFluid(fluidstate)) {
+            this.travelInFluid(input, fluidstate);
         } else if (this.isFallFlying()) {
             this.travelFallFlying(input);
         } else {
@@ -2441,7 +_,7 @@
     }
 
     protected boolean shouldTravelInFluid(final FluidState fluidState) {
-        return (this.isInWater() || this.isInLava()) && this.isAffectedByFluids() && !this.canStandOnFluid(fluidState);
+        return (this.isInWater() || this.isInLava() || this.isInFluidType(fluidState)) && this.isAffectedByFluids() && !this.canStandOnFluid(fluidState);
     }
 
     protected void travelFlying(final Vec3 input, final float speed) {
@@ -2468,7 +_,7 @@
         BlockPos posBelow = this.getBlockPosBelowThatAffectsMyMovement();
         float blockFriction = this.onGround()
             ? computeModifiedFriction(
-                this.level().getBlockState(posBelow).getBlock().getFriction(), (float)this.getAttributeValue(Attributes.FRICTION_MODIFIER)
+                this.level().getBlockState(posBelow).getFriction(level(), posBelow, this), (float)this.getAttributeValue(Attributes.FRICTION_MODIFIER)
             )
             : 1.0F;
         Vec3 movement = this.handleRelativeFrictionAndCalculateMovement(input, blockFriction);
@@ -2500,10 +_,18 @@
         return computeModifiedFriction(this.omnidirectionalAirMover() ? 0.91F : 0.98F, (float)this.getAttributeValue(Attributes.AIR_DRAG_MODIFIER));
     }
 
-    protected void travelInFluid(final Vec3 input) {
+    @Deprecated // FORGE: Use the version that takes a FluidState
+    protected void travelInFluid(Vec3 input) {
+        this.travelInFluid(input, net.minecraft.world.level.material.Fluids.WATER.defaultFluidState());
+    }
+
+    protected void travelInFluid(Vec3 input, FluidState fluidstate) {
         boolean isFalling = this.getDeltaMovement().y <= 0.0;
         double oldY = this.getY();
         double baseGravity = this.getEffectiveGravity();
+        if (this.isInFluidType(fluidstate) && this.moveInFluid(fluidstate, input, oldY)) {
+            // Modded fluid handled it
+        } else
         if (this.isInWater()) {
             this.travelInWater(input, baseGravity, isFalling, oldY);
             this.floatInWaterWhileRidden();
@@ -2529,6 +_,7 @@
             slowDown = 0.96F;
         }
 
+        speed *= this.getAttributeValue(net.minecraftforge.common.ForgeMod.SWIM_SPEED.getHolder().get());
         this.moveRelative(speed, input);
         this.move(MoverType.SELF, this.getDeltaMovement());
         Vec3 movement = this.getDeltaMovement();
@@ -2706,7 +_,7 @@
             double xd = Mth.clamp(delta.x, -0.15F, 0.15F);
             double zd = Mth.clamp(delta.z, -0.15F, 0.15F);
             double yd = Math.max(delta.y, -0.15F);
-            if (yd < 0.0 && !this.getInBlockState().is(Blocks.SCAFFOLDING) && this.isSuppressingSlidingDownLadder() && this instanceof Player) {
+            if (yd < 0.0 && !this.getInBlockState().isScaffolding(this) && this.isSuppressingSlidingDownLadder() && this instanceof Player) {
                 yd = 0.0;
             }
 
@@ -2762,6 +_,7 @@
 
     @Override
     public void tick() {
+        if (net.minecraftforge.event.ForgeEventFactory.onLivingTick(this)) return;
         super.tick();
         this.updatingUsingItem();
         this.updateSwimAmount();
@@ -2962,6 +_,7 @@
             ItemStack previous = lastEquipmentItems.get(slot);
             ItemStack current = this.getItemBySlot(slot);
             if (this.equipmentHasChanged(previous, current)) {
+                net.minecraftforge.event.ForgeEventFactory.onLivingEquipmentChange(this, slot, previous, current);
                 if (changedItems == null) {
                     changedItems = Maps.newEnumMap(EquipmentSlot.class);
                 }
@@ -3097,6 +_,10 @@
         profiler.push("jump");
         if (this.jumping && this.isAffectedByFluids()) {
             double fluidHeight;
+            var fluidType = this.getMaxHeightFluidType();
+            if (!fluidType.isAir()) {
+                fluidHeight = this.getFluidTypeHeight(fluidType);
+            } else
             if (this.isInLava()) {
                 fluidHeight = this.getFluidHeight(FluidTags.LAVA);
             } else {
@@ -3107,15 +_,19 @@
             double fluidJumpThreshold = this.getFluidJumpThreshold();
             if (!inWaterAndHasFluidHeight || this.onGround() && !(fluidHeight > fluidJumpThreshold)) {
                 if (!this.isInLava() || this.onGround() && this.isInShallowFluid(FluidTags.LAVA)) {
+                    if (fluidType.isAir() || this.onGround() && !(fluidHeight > fluidJumpThreshold)) {
                     if ((this.onGround() || inWaterAndHasFluidHeight && fluidHeight <= fluidJumpThreshold) && this.noJumpDelay == 0) {
                         this.jumpFromGround();
                         this.noJumpDelay = 10;
                     }
+                    } else {
+                        this.jumpInFluid(fluidType);
+                    }
                 } else {
-                    this.jumpInLiquid(FluidTags.LAVA);
+                    this.jumpInFluid(net.minecraftforge.common.ForgeMod.LAVA_TYPE.get());
                 }
             } else {
-                this.jumpInLiquid(FluidTags.WATER);
+                this.jumpInFluid(net.minecraftforge.common.ForgeMod.WATER_TYPE.get());
             }
         } else {
             this.noJumpDelay = 0;
@@ -3434,8 +_,11 @@
 
     private void updatingUsingItem() {
         if (this.isUsingItem()) {
-            if (ItemStack.isSameItem(this.getItemInHand(this.getUsedItemHand()), this.useItem)) {
+            var current = this.getItemInHand(this.getUsedItemHand());
+            if (net.minecraftforge.common.ForgeHooks.canContinueUsing(this.useItem, current)) {
                 this.useItem = this.getItemInHand(this.getUsedItemHand());
+            }
+            if (this.useItem == current) {
                 this.updateUsingItem(this.useItem);
             } else {
                 this.stopUsingItem();
@@ -3478,8 +_,12 @@
     }
 
     protected void updateUsingItem(final ItemStack useItem) {
+        if (!useItem.isEmpty()) {
+            this.useItemRemaining = net.minecraftforge.event.ForgeEventFactory.onItemUseTick(this, useItem, this.getUseItemRemainingTicks());
+        }
+        if (this.getUseItemRemainingTicks() > 0)
         useItem.onUseTick(this.level(), this, this.getUseItemRemainingTicks());
-        if (--this.useItemRemaining == 0 && !this.level().isClientSide() && !useItem.useOnRelease()) {
+        if (--this.useItemRemaining <= 0 && !this.level().isClientSide() && !useItem.useOnRelease()) {
             this.completeUsingItem();
         }
     }
@@ -3507,8 +_,10 @@
     public void startUsingItem(final InteractionHand hand) {
         ItemStack itemStack = this.getItemInHand(hand);
         if (!itemStack.isEmpty() && !this.isUsingItem()) {
+            int duration = net.minecraftforge.event.ForgeEventFactory.onItemUseStart(this, itemStack, itemStack.getUseDuration(this));
+            if (duration < 0) return;
             this.useItem = itemStack;
-            this.useItemRemaining = itemStack.getUseDuration(this);
+            this.useItemRemaining = duration;
             if (!this.level().isClientSide()) {
                 this.setLivingEntityFlag(1, true);
                 this.setLivingEntityFlag(2, hand == InteractionHand.OFF_HAND);
@@ -3566,6 +_,9 @@
                 p = p.xRot(-this.getXRot() * (float) (Math.PI / 180.0));
                 p = p.yRot(-this.getYRot() * (float) (Math.PI / 180.0));
                 p = p.add(this.getX(), this.getEyeY(), this.getZ());
+                if (this.level() instanceof ServerLevel serverLevel) //Forge: Fix MC-2518 spawnParticle is nooped on server, need to use server specific variant
+                    serverLevel.sendParticles(breakParticle, p.x, p.y, p.z, 1, d.x, d.y + 0.05D, d.z, 0.0D);
+                else
                 this.level().addParticle(breakParticle, p.x, p.y, p.z, d.x, d.y + 0.05, d.z);
             }
         }
@@ -3578,7 +_,9 @@
                 this.releaseUsingItem();
             } else {
                 if (!this.useItem.isEmpty() && this.isUsingItem()) {
+                    ItemStack copy = this.useItem.copy();
                     ItemStack result = this.useItem.finishUsingItem(this.level(), this);
+                    result = net.minecraftforge.event.ForgeEventFactory.onItemUseFinish(this, copy, getUseItemRemainingTicks(), result);
                     if (result != this.useItem) {
                         this.setItemInHand(hand, result);
                     }
@@ -3612,7 +_,13 @@
         ItemStack itemInUsedHand = this.getItemInHand(this.getUsedItemHand());
         if (!this.useItem.isEmpty() && ItemStack.isSameItem(itemInUsedHand, this.useItem)) {
             this.useItem = itemInUsedHand;
+            if (!net.minecraftforge.event.ForgeEventFactory.onUseItemStop(this, useItem, this.getUseItemRemainingTicks())) {
+               ItemStack copy = this instanceof Player ? useItem.copy() : null;
             this.useItem.releaseUsing(this.level(), this, this.getUseItemRemainingTicks());
+               if (copy != null && useItem.isEmpty()) {
+                   net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem((Player)this, copy, getUsedItemHand());
+               }
+            }
             if (this.useItem.useOnRelease()) {
                 this.updatingUsingItem();
             }
@@ -3622,6 +_,7 @@
     }
 
     public void stopUsingItem() {
+        if (this.isUsingItem() && !this.useItem.isEmpty()) this.useItem.onStopUsing(this, useItemRemaining);
         if (!this.level().isClientSide()) {
             boolean wasUsingItem = this.isUsingItem();
             this.recentKineticEnemies = null;
@@ -3783,8 +_,8 @@
         }
 
         BlockState blockState = this.level().getBlockState(bedPosition);
-        if (blockState.getBlock() instanceof BedBlock) {
-            this.level().setBlock(bedPosition, blockState.setValue(BedBlock.OCCUPIED, true), 3);
+        if (blockState.isBed(level(), bedPosition, this)) {
+            blockState.setBedOccupied(level(), bedPosition, this, true);
         }
 
         this.setPose(Pose.SLEEPING);
@@ -3799,15 +_,15 @@
     }
 
     private boolean checkBedExists() {
-        return this.getSleepingPos().map(bedPosition -> this.level().getBlockState(bedPosition).getBlock() instanceof BedBlock).orElse(false);
+        return this.getSleepingPos().map(bedPosition -> net.minecraftforge.event.ForgeEventFactory.fireSleepingLocationCheck(this, bedPosition)).orElse(false);
     }
 
     public void stopSleeping() {
         this.getSleepingPos().filter(this.level()::hasChunkAt).ifPresent(bedPosition -> {
             BlockState state = this.level().getBlockState(bedPosition);
-            if (state.getBlock() instanceof BedBlock) {
+            if (state.isBed(level(), bedPosition, this)) {
                 Direction facing = state.getValue(BedBlock.FACING);
-                this.level().setBlock(bedPosition, state.setValue(BedBlock.OCCUPIED, false), 3);
+                state.setBedOccupied(level(), bedPosition, this, false);
                 Vec3 standUp = BedBlock.findStandUpPosition(this.getType(), this.level(), bedPosition, facing, this.getYRot()).orElseGet(() -> {
                     BlockPos above = bedPosition.above();
                     return new Vec3(above.getX() + 0.5, above.getY() + 0.1, above.getZ() + 0.5);
@@ -3827,7 +_,9 @@
 
     public @Nullable Direction getBedOrientation() {
         BlockPos bedPos = this.getSleepingPos().orElse(null);
-        return bedPos != null ? BedBlock.getBedOrientation(this.level(), bedPos) : null;
+        if (bedPos == null) return Direction.UP;
+        BlockState state = this.level().getBlockState(bedPos);
+        return !state.isBed(level(), bedPos, this) ? Direction.UP : state.getBedDirection(level(), bedPos);
     }
 
     @Override
@@ -3836,7 +_,7 @@
     }
 
     public ItemStack getProjectile(final ItemStack heldWeapon) {
-        return ItemStack.EMPTY;
+        return net.minecraftforge.common.ForgeHooks.getProjectile(this, heldWeapon, ItemStack.EMPTY);
     }
 
     private static byte entityEventForEquipmentBreak(final EquipmentSlot equipmentSlot) {
@@ -3888,6 +_,8 @@
     }
 
     public EquipmentSlot getEquipmentSlotForItem(final ItemStack itemStack) {
+        final EquipmentSlot slot = itemStack.getEquipmentSlot();
+        if (slot != null) return slot; // FORGE: Allow modders to set a non-default equipment slot for a stack; e.g. a non-armor chestplate-slot item
         Equippable equippable = itemStack.get(DataComponents.EQUIPPABLE);
         return equippable != null && this.canUseSlot(equippable.slot()) ? equippable.slot() : EquipmentSlot.MAINHAND;
     }
@@ -4055,5 +_,42 @@
     }
 
     public record Fallsounds(SoundEvent small, SoundEvent big) {
+    }
+
+    /**
+     * Returns true if the entity's rider (EntityPlayer) should face forward when mounted.
+     * currently only used in vanilla code by pigs.
+     *
+     * @param player The player who is riding the entity.
+     * @return If the player should orient the same direction as this entity.
+     */
+    public boolean shouldRiderFaceForward(Player player) {
+        return this instanceof net.minecraft.world.entity.animal.pig.Pig;
+    }
+
+    private net.minecraftforge.common.util.LazyOptional<?>[] handlers = net.minecraftforge.items.wrapper.EntityEquipmentInvWrapper.create(this);
+
+    @Override
+    public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+        if (capability == net.minecraftforge.common.capabilities.ForgeCapabilities.ITEM_HANDLER && this.isAlive()) {
+             if (facing == null) return handlers[2].cast();
+             else if (facing.getAxis().isVertical()) return handlers[0].cast();
+             else if (facing.getAxis().isHorizontal()) return handlers[1].cast();
+        }
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public void invalidateCaps() {
+        super.invalidateCaps();
+        for (int x = 0; x < handlers.length; x++) {
+             handlers[x].invalidate();
+        }
+    }
+
+    @Override
+    public void reviveCaps() {
+        super.reviveCaps();
+        handlers = net.minecraftforge.items.wrapper.EntityEquipmentInvWrapper.create(this);
     }
 }
