--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -186,11 +_,15 @@
    }
 
    public void die(DamageSource p_21809_) {
+      // FORGE: Super moved to top so that death message would be cancelled properly
+      net.minecraft.network.chat.Component deathMessage = this.getCombatTracker().getDeathMessage();
+      super.die(p_21809_);
+
+      if (this.dead)
       if (!this.level().isClientSide && this.level().getGameRules().getBoolean(GameRules.RULE_SHOWDEATHMESSAGES) && this.getOwner() instanceof ServerPlayer) {
-         this.getOwner().sendSystemMessage(this.getCombatTracker().getDeathMessage());
+         this.getOwner().sendSystemMessage(deathMessage);
       }
 
-      super.die(p_21809_);
    }
 
    public boolean isOrderedToSit() {
