--- ../src-base/minecraft/net/minecraft/world/storage/loot/LootContext.java
+++ ../src-work/minecraft/net/minecraft/world/storage/loot/LootContext.java
@@ -89,6 +89,16 @@
         }
     }
 
+    public WorldServer getWorld()
+    {
+        return field_186499_b;
+    }
+
+    public int getLootingModifier()
+    {
+        return net.minecraftforge.common.ForgeHooks.getLootingLevel(func_186493_a(), func_186492_c(), field_186503_f);
+    }
+
     public static class Builder
         {
             private final WorldServer field_186474_a;
