--- a/net/minecraft/world/storage/loot/LootContext.java
+++ b/net/minecraft/world/storage/loot/LootContext.java
@@ -76,6 +76,10 @@
       return this.field_186499_b;
    }
 
+   public int getLootingModifier() {
+      return net.minecraftforge.common.ForgeHooks.getLootingLevel(func_216031_c(LootParameters.field_216281_a), func_216031_c(LootParameters.field_216284_d), func_216031_c(LootParameters.field_216283_c));
+   }
+
    public static class Builder {
       private final ServerWorld field_186474_a;
       private final Map<LootParameter<?>, Object> field_216025_b = Maps.newIdentityHashMap();
