--- ../src-base/minecraft/net/minecraft/client/renderer/texture/TextureManager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/texture/TextureManager.java
@@ -143,12 +143,15 @@
 
     public void onResourceManagerReload(IResourceManager resourceManager)
     {
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar bar = net.minecraftforge.fml.common.ProgressManager.push("Reloading Texture Manager", this.mapTextureObjects.keySet().size(), true);
         Iterator iterator = this.mapTextureObjects.entrySet().iterator();
 
         while (iterator.hasNext())
         {
             Entry entry = (Entry)iterator.next();
+            bar.step(entry.getKey().toString());
             this.loadTexture((ResourceLocation)entry.getKey(), (ITextureObject)entry.getValue());
         }
+        net.minecraftforge.fml.common.ProgressManager.pop(bar);
     }
 }
