--- ../src-base/minecraft/net/minecraft/client/renderer/texture/TextureManager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/texture/TextureManager.java
@@ -137,9 +137,12 @@
 
     public void onResourceManagerReload(IResourceManager resourceManager)
     {
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar bar = net.minecraftforge.fml.common.ProgressManager.push("Reloading Texture Manager", this.mapTextureObjects.keySet().size(), true);
         for (Entry<ResourceLocation, ITextureObject> entry : this.mapTextureObjects.entrySet())
         {
+            bar.step(entry.getKey().toString());
             this.loadTexture((ResourceLocation)entry.getKey(), (ITextureObject)entry.getValue());
         }
+        net.minecraftforge.fml.common.ProgressManager.pop(bar);
     }
 }
