--- a/net/minecraft/world/entity/animal/golem/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/golem/SnowGolem.java
@@ -95,7 +_,7 @@
                 this.hurtServer(serverLevel, this.damageSources().onFire(), 1.0F);
             }
 
-            if (!serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+            if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(serverLevel, this)) {
                 return;
             }
 
@@ -106,7 +_,7 @@
                 int yy = Mth.floor(this.getY());
                 int zz = Mth.floor(this.getZ() + (i / 2 % 2 * 2 - 1) * 0.25F);
                 BlockPos snowPos = new BlockPos(xx, yy, zz);
-                if (this.level().getBlockState(snowPos).isAir() && snow.canSurvive(this.level(), snowPos)) {
+                if (this.level().isEmptyBlock(snowPos) && snow.canSurvive(this.level(), snowPos)) {
                     this.level().setBlockAndUpdate(snowPos, snow);
                     this.level().gameEvent(GameEvent.BLOCK_PLACE, snowPos, GameEvent.Context.of(this, snow));
                 }
@@ -136,7 +_,7 @@
     @Override
     protected InteractionResult mobInteract(final Player player, final InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.is(Items.SHEARS) && this.readyForShearing()) {
+        if (itemStack.canPerformAction(net.minecraftforge.common.ToolActions.SHEARS_HARVEST) && this.readyForShearing()) {
             if (this.level() instanceof ServerLevel level) {
                 this.shear(level, SoundSource.PLAYERS, itemStack);
                 this.gameEvent(GameEvent.SHEAR, player);
