--- a/net/minecraft/client/gui/screen/inventory/ContainerScreen.java
+++ b/net/minecraft/client/gui/screen/inventory/ContainerScreen.java
@@ -73,6 +73,7 @@
       int i = this.field_147003_i;
       int j = this.field_147009_r;
       this.func_146976_a(p_render_3_, p_render_1_, p_render_2_);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.GuiContainerEvent.DrawBackground(this, p_render_1_, p_render_2_));
       GlStateManager.disableRescaleNormal();
       RenderHelper.func_74518_a();
       GlStateManager.disableLighting();
@@ -102,7 +103,8 @@
             int j1 = slot.field_75223_e;
             int k1 = slot.field_75221_f;
             GlStateManager.colorMask(true, true, true, false);
-            this.fillGradient(j1, k1, j1 + 16, k1 + 16, -2130706433, -2130706433);
+            int slotColor = this.getSlotColor(i1);
+            this.fillGradient(j1, k1, j1 + 16, k1 + 16, slotColor, slotColor);
             GlStateManager.colorMask(true, true, true, true);
             GlStateManager.enableLighting();
             GlStateManager.enableDepthTest();
@@ -112,6 +114,7 @@
       RenderHelper.func_74518_a();
       this.func_146979_b(p_render_1_, p_render_2_);
       RenderHelper.func_74520_c();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.GuiContainerEvent.DrawForeground(this, p_render_1_, p_render_2_));
       PlayerInventory playerinventory = this.minecraft.field_71439_g.field_71071_by;
       ItemStack itemstack = this.field_147012_x.func_190926_b() ? playerinventory.func_70445_o() : this.field_147012_x;
       if (!itemstack.func_190926_b()) {
@@ -163,8 +166,10 @@
       GlStateManager.translatef(0.0F, 0.0F, 32.0F);
       this.blitOffset = 200;
       this.itemRenderer.field_77023_b = 200.0F;
+      net.minecraft.client.gui.FontRenderer font = p_146982_1_.func_77973_b().getFontRenderer(p_146982_1_);
+      if (font == null) font = this.font;
       this.itemRenderer.func_180450_b(p_146982_1_, p_146982_2_, p_146982_3_);
-      this.itemRenderer.func_180453_a(this.font, p_146982_1_, p_146982_2_, p_146982_3_ - (this.field_147012_x.func_190926_b() ? 0 : 8), p_146982_4_);
+      this.itemRenderer.func_180453_a(font, p_146982_1_, p_146982_2_, p_146982_3_ - (this.field_147012_x.func_190926_b() ? 0 : 8), p_146982_4_);
       this.blitOffset = 0;
       this.itemRenderer.field_77023_b = 0.0F;
    }
@@ -208,11 +213,10 @@
       this.blitOffset = 100;
       this.itemRenderer.field_77023_b = 100.0F;
       if (itemstack.func_190926_b() && p_146977_1_.func_111238_b()) {
-         String s1 = p_146977_1_.func_178171_c();
-         if (s1 != null) {
-            TextureAtlasSprite textureatlassprite = this.minecraft.func_147117_R().func_110572_b(s1);
+         TextureAtlasSprite textureatlassprite = p_146977_1_.getBackgroundSprite();
+         if (textureatlassprite != null) {
             GlStateManager.disableLighting();
-            this.minecraft.func_110434_K().func_110577_a(AtlasTexture.field_110575_b);
+            this.minecraft.func_110434_K().func_110577_a(p_146977_1_.getBackgroundLocation());
             blit(i, j, this.blitOffset, 16, 16, textureatlassprite);
             GlStateManager.enableLighting();
             flag1 = true;
@@ -273,7 +277,8 @@
       if (super.mouseClicked(p_mouseClicked_1_, p_mouseClicked_3_, p_mouseClicked_5_)) {
          return true;
       } else {
-         boolean flag = this.minecraft.field_71474_y.field_74322_I.func_197984_a(p_mouseClicked_5_);
+         InputMappings.Input mouseKey = InputMappings.Type.MOUSE.func_197944_a(p_mouseClicked_5_);
+         boolean flag = this.minecraft.field_71474_y.field_74322_I.isActiveAndMatches(mouseKey);
          Slot slot = this.func_195360_a(p_mouseClicked_1_, p_mouseClicked_3_);
          long i = Util.func_211177_b();
          this.field_146993_M = this.field_146998_K == slot && i - this.field_146997_J < 250L && this.field_146992_L == p_mouseClicked_5_;
@@ -282,6 +287,7 @@
             int j = this.field_147003_i;
             int k = this.field_147009_r;
             boolean flag1 = this.func_195361_a(p_mouseClicked_1_, p_mouseClicked_3_, j, k, p_mouseClicked_5_);
+            if (slot != null) flag1 = false; // Forge, prevent dropping of items through slots outside of GUI boundaries
             int l = -1;
             if (slot != null) {
                l = slot.field_75222_d;
@@ -307,7 +313,7 @@
                   }
                } else if (!this.field_147007_t) {
                   if (this.minecraft.field_71439_g.field_71071_by.func_70445_o().func_190926_b()) {
-                     if (this.minecraft.field_71474_y.field_74322_I.func_197984_a(p_mouseClicked_5_)) {
+                     if (this.minecraft.field_71474_y.field_74322_I.isActiveAndMatches(mouseKey)) {
                         this.func_184098_a(slot, l, p_mouseClicked_5_, ClickType.CLONE);
                      } else {
                         boolean flag2 = l != -999 && (InputMappings.func_216506_a(Minecraft.func_71410_x().field_195558_d.func_198092_i(), 340) || InputMappings.func_216506_a(Minecraft.func_71410_x().field_195558_d.func_198092_i(), 344));
@@ -331,7 +337,7 @@
                         this.field_146987_F = 0;
                      } else if (p_mouseClicked_5_ == 1) {
                         this.field_146987_F = 1;
-                     } else if (this.minecraft.field_71474_y.field_74322_I.func_197984_a(p_mouseClicked_5_)) {
+                     } else if (this.minecraft.field_71474_y.field_74322_I.isActiveAndMatches(mouseKey)) {
                         this.field_146987_F = 2;
                      }
                   }
@@ -384,10 +390,13 @@
    }
 
    public boolean mouseReleased(double p_mouseReleased_1_, double p_mouseReleased_3_, int p_mouseReleased_5_) {
+      super.mouseReleased(p_mouseReleased_1_, p_mouseReleased_3_, p_mouseReleased_5_); //Forge, Call parent to release buttons
       Slot slot = this.func_195360_a(p_mouseReleased_1_, p_mouseReleased_3_);
       int i = this.field_147003_i;
       int j = this.field_147009_r;
       boolean flag = this.func_195361_a(p_mouseReleased_1_, p_mouseReleased_3_, i, j, p_mouseReleased_5_);
+      if (slot != null) flag = false; // Forge, prevent dropping of items through slots outside of GUI boundaries
+      InputMappings.Input mouseKey = InputMappings.Type.MOUSE.func_197944_a(p_mouseReleased_5_);
       int k = -1;
       if (slot != null) {
          k = slot.field_75222_d;
@@ -401,7 +410,7 @@
          if (hasShiftDown()) {
             if (!this.field_146994_N.func_190926_b()) {
                for(Slot slot2 : this.field_147002_h.field_75151_b) {
-                  if (slot2 != null && slot2.func_82869_a(this.minecraft.field_71439_g) && slot2.func_75216_d() && slot2.field_75224_c == slot.field_75224_c && Container.func_94527_a(slot2, this.field_146994_N, true)) {
+                  if (slot2 != null && slot2.func_82869_a(this.minecraft.field_71439_g) && slot2.func_75216_d() && slot2.isSameInventory(slot) && Container.func_94527_a(slot2, this.field_146994_N, true)) {
                      this.func_184098_a(slot2, slot2.field_75222_d, p_mouseReleased_5_, ClickType.QUICK_MOVE);
                   }
                }
@@ -465,7 +474,7 @@
 
             this.func_184098_a((Slot)null, -999, Container.func_94534_d(2, this.field_146987_F), ClickType.QUICK_CRAFT);
          } else if (!this.minecraft.field_71439_g.field_71071_by.func_70445_o().func_190926_b()) {
-            if (this.minecraft.field_71474_y.field_74322_I.func_197984_a(p_mouseReleased_5_)) {
+            if (this.minecraft.field_71474_y.field_74322_I.isActiveAndMatches(mouseKey)) {
                this.func_184098_a(slot, k, p_mouseReleased_5_, ClickType.CLONE);
             } else {
                boolean flag1 = k != -999 && (InputMappings.func_216506_a(Minecraft.func_71410_x().field_195558_d.func_198092_i(), 340) || InputMappings.func_216506_a(Minecraft.func_71410_x().field_195558_d.func_198092_i(), 344));
@@ -514,27 +523,32 @@
       if (super.keyPressed(p_keyPressed_1_, p_keyPressed_2_, p_keyPressed_3_)) {
          return true;
       } else {
-         if (p_keyPressed_1_ == 256 || this.minecraft.field_71474_y.field_151445_Q.func_197976_a(p_keyPressed_1_, p_keyPressed_2_)) {
+         InputMappings.Input mouseKey = InputMappings.func_197954_a(p_keyPressed_1_, p_keyPressed_2_);
+         if (p_keyPressed_1_ == 256 || this.minecraft.field_71474_y.field_151445_Q.isActiveAndMatches(mouseKey)) {
             this.minecraft.field_71439_g.func_71053_j();
+            return true; // Forge MC-146650: Needs to return true when the key is handled.
          }
 
-         this.func_195363_d(p_keyPressed_1_, p_keyPressed_2_);
+         if (this.func_195363_d(p_keyPressed_1_, p_keyPressed_2_))
+            return true; // Forge MC-146650: Needs to return true when the key is handled.
          if (this.field_147006_u != null && this.field_147006_u.func_75216_d()) {
-            if (this.minecraft.field_71474_y.field_74322_I.func_197976_a(p_keyPressed_1_, p_keyPressed_2_)) {
+            if (this.minecraft.field_71474_y.field_74322_I.isActiveAndMatches(mouseKey)) {
                this.func_184098_a(this.field_147006_u, this.field_147006_u.field_75222_d, 0, ClickType.CLONE);
-            } else if (this.minecraft.field_71474_y.field_74316_C.func_197976_a(p_keyPressed_1_, p_keyPressed_2_)) {
+               return true; // Forge MC-146650: Needs to return true when the key is handled.
+            } else if (this.minecraft.field_71474_y.field_74316_C.isActiveAndMatches(mouseKey)) {
                this.func_184098_a(this.field_147006_u, this.field_147006_u.field_75222_d, hasControlDown() ? 1 : 0, ClickType.THROW);
+               return true; // Forge MC-146650: Needs to return true when the key is handled.
             }
          }
 
-         return true;
+         return false; // Forge MC-146650: Needs to return false when the key is not handled.
       }
    }
 
    protected boolean func_195363_d(int p_195363_1_, int p_195363_2_) {
       if (this.minecraft.field_71439_g.field_71071_by.func_70445_o().func_190926_b() && this.field_147006_u != null) {
          for(int i = 0; i < 9; ++i) {
-            if (this.minecraft.field_71474_y.field_151456_ac[i].func_197976_a(p_195363_1_, p_195363_2_)) {
+            if (this.minecraft.field_71474_y.field_151456_ac[i].isActiveAndMatches(InputMappings.func_197954_a(p_195363_1_, p_195363_2_))) {
                this.func_184098_a(this.field_147006_u, this.field_147006_u.field_75222_d, i, ClickType.SWAP);
                return true;
             }
@@ -565,4 +579,16 @@
    public T func_212873_a_() {
       return this.field_147002_h;
    }
+
+   @javax.annotation.Nullable
+   public Slot getSlotUnderMouse() { return this.field_147006_u; }
+   public int getGuiLeft() { return field_147003_i; }
+   public int getGuiTop() { return field_147009_r; }
+   public int getXSize() { return field_146999_f; }
+   public int getYSize() { return field_147000_g; }
+
+   protected int slotColor = -2130706433;
+   public int getSlotColor(int index) {
+      return slotColor;
+   }
 }
