--- a/net/minecraft/client/gui/screen/inventory/ContainerScreen.java
+++ b/net/minecraft/client/gui/screen/inventory/ContainerScreen.java
@@ -86,6 +_,7 @@
       int i = this.leftPos;
       int j = this.topPos;
       this.renderBg(p_230430_1_, p_230430_4_, p_230430_2_, p_230430_3_);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.GuiContainerEvent.DrawBackground(this, p_230430_1_, p_230430_2_, p_230430_3_));
       RenderSystem.disableRescaleNormal();
       RenderSystem.disableDepthTest();
       super.render(p_230430_1_, p_230430_2_, p_230430_3_, p_230430_4_);
@@ -111,13 +_,15 @@
             int j1 = slot.x;
             int k1 = slot.y;
             RenderSystem.colorMask(true, true, true, false);
-            this.fillGradient(p_230430_1_, j1, k1, j1 + 16, k1 + 16, -2130706433, -2130706433);
+            int slotColor = this.getSlotColor(i1);
+            this.fillGradient(p_230430_1_, j1, k1, j1 + 16, k1 + 16, slotColor, slotColor);
             RenderSystem.colorMask(true, true, true, true);
             RenderSystem.enableDepthTest();
          }
       }
 
       this.renderLabels(p_230430_1_, p_230430_2_, p_230430_3_);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.GuiContainerEvent.DrawForeground(this, p_230430_1_, p_230430_2_, p_230430_3_));
       PlayerInventory playerinventory = this.minecraft.player.inventory;
       ItemStack itemstack = this.draggingItem.isEmpty() ? playerinventory.getCarried() : this.draggingItem;
       if (!itemstack.isEmpty()) {
@@ -167,8 +_,10 @@
       RenderSystem.translatef(0.0F, 0.0F, 32.0F);
       this.setBlitOffset(200);
       this.itemRenderer.blitOffset = 200.0F;
+      net.minecraft.client.gui.FontRenderer font = p_146982_1_.getItem().getFontRenderer(p_146982_1_);
+      if (font == null) font = this.font;
       this.itemRenderer.renderAndDecorateItem(p_146982_1_, p_146982_2_, p_146982_3_);
-      this.itemRenderer.renderGuiItemDecorations(this.font, p_146982_1_, p_146982_2_, p_146982_3_ - (this.draggingItem.isEmpty() ? 0 : 8), p_146982_4_);
+      this.itemRenderer.renderGuiItemDecorations(font, p_146982_1_, p_146982_2_, p_146982_3_ - (this.draggingItem.isEmpty() ? 0 : 8), p_146982_4_);
       this.setBlitOffset(0);
       this.itemRenderer.blitOffset = 0.0F;
    }
@@ -278,7 +_,8 @@
       if (super.mouseClicked(p_231044_1_, p_231044_3_, p_231044_5_)) {
          return true;
       } else {
-         boolean flag = this.minecraft.options.keyPickItem.matchesMouse(p_231044_5_);
+         InputMappings.Input mouseKey = InputMappings.Type.MOUSE.getOrCreate(p_231044_5_);
+         boolean flag = this.minecraft.options.keyPickItem.isActiveAndMatches(mouseKey);
          Slot slot = this.findSlot(p_231044_1_, p_231044_3_);
          long i = Util.getMillis();
          this.doubleclick = this.lastClickSlot == slot && i - this.lastClickTime < 250L && this.lastClickButton == p_231044_5_;
@@ -289,6 +_,7 @@
             int j = this.leftPos;
             int k = this.topPos;
             boolean flag1 = this.hasClickedOutside(p_231044_1_, p_231044_3_, j, k, p_231044_5_);
+            if (slot != null) flag1 = false; // Forge, prevent dropping of items through slots outside of GUI boundaries
             int l = -1;
             if (slot != null) {
                l = slot.index;
@@ -314,7 +_,7 @@
                   }
                } else if (!this.isQuickCrafting) {
                   if (this.minecraft.player.inventory.getCarried().isEmpty()) {
-                     if (this.minecraft.options.keyPickItem.matchesMouse(p_231044_5_)) {
+                     if (this.minecraft.options.keyPickItem.isActiveAndMatches(mouseKey)) {
                         this.slotClicked(slot, l, p_231044_5_, ClickType.CLONE);
                      } else {
                         boolean flag2 = l != -999 && (InputMappings.isKeyDown(Minecraft.getInstance().getWindow().getWindow(), 340) || InputMappings.isKeyDown(Minecraft.getInstance().getWindow().getWindow(), 344));
@@ -338,7 +_,7 @@
                         this.quickCraftingType = 0;
                      } else if (p_231044_5_ == 1) {
                         this.quickCraftingType = 1;
-                     } else if (this.minecraft.options.keyPickItem.matchesMouse(p_231044_5_)) {
+                     } else if (this.minecraft.options.keyPickItem.isActiveAndMatches(mouseKey)) {
                         this.quickCraftingType = 2;
                      }
                   }
@@ -407,10 +_,13 @@
    }
 
    public boolean mouseReleased(double p_231048_1_, double p_231048_3_, int p_231048_5_) {
+      super.mouseReleased(p_231048_1_, p_231048_3_, p_231048_5_); //Forge, Call parent to release buttons
       Slot slot = this.findSlot(p_231048_1_, p_231048_3_);
       int i = this.leftPos;
       int j = this.topPos;
       boolean flag = this.hasClickedOutside(p_231048_1_, p_231048_3_, i, j, p_231048_5_);
+      if (slot != null) flag = false; // Forge, prevent dropping of items through slots outside of GUI boundaries
+      InputMappings.Input mouseKey = InputMappings.Type.MOUSE.getOrCreate(p_231048_5_);
       int k = -1;
       if (slot != null) {
          k = slot.index;
@@ -424,7 +_,7 @@
          if (hasShiftDown()) {
             if (!this.lastQuickMoved.isEmpty()) {
                for(Slot slot2 : this.menu.slots) {
-                  if (slot2 != null && slot2.mayPickup(this.minecraft.player) && slot2.hasItem() && slot2.container == slot.container && Container.canItemQuickReplace(slot2, this.lastQuickMoved, true)) {
+                  if (slot2 != null && slot2.mayPickup(this.minecraft.player) && slot2.hasItem() && slot2.isSameInventory(slot) && Container.canItemQuickReplace(slot2, this.lastQuickMoved, true)) {
                      this.slotClicked(slot2, slot2.index, p_231048_5_, ClickType.QUICK_MOVE);
                   }
                }
@@ -488,7 +_,7 @@
 
             this.slotClicked((Slot)null, -999, Container.getQuickcraftMask(2, this.quickCraftingType), ClickType.QUICK_CRAFT);
          } else if (!this.minecraft.player.inventory.getCarried().isEmpty()) {
-            if (this.minecraft.options.keyPickItem.matchesMouse(p_231048_5_)) {
+            if (this.minecraft.options.keyPickItem.isActiveAndMatches(mouseKey)) {
                this.slotClicked(slot, k, p_231048_5_, ClickType.CLONE);
             } else {
                boolean flag1 = k != -999 && (InputMappings.isKeyDown(Minecraft.getInstance().getWindow().getWindow(), 340) || InputMappings.isKeyDown(Minecraft.getInstance().getWindow().getWindow(), 344));
@@ -530,34 +_,39 @@
    }
 
    public boolean keyPressed(int p_231046_1_, int p_231046_2_, int p_231046_3_) {
+      InputMappings.Input mouseKey = InputMappings.getKey(p_231046_1_, p_231046_2_);
       if (super.keyPressed(p_231046_1_, p_231046_2_, p_231046_3_)) {
          return true;
-      } else if (this.minecraft.options.keyInventory.matches(p_231046_1_, p_231046_2_)) {
+      } else if (this.minecraft.options.keyInventory.isActiveAndMatches(mouseKey)) {
          this.onClose();
          return true;
       } else {
-         this.checkHotbarKeyPressed(p_231046_1_, p_231046_2_);
+         boolean handled = this.checkHotbarKeyPressed(p_231046_1_, p_231046_2_);// Forge MC-146650: Needs to return true when the key is handled
          if (this.hoveredSlot != null && this.hoveredSlot.hasItem()) {
-            if (this.minecraft.options.keyPickItem.matches(p_231046_1_, p_231046_2_)) {
+            if (this.minecraft.options.keyPickItem.isActiveAndMatches(mouseKey)) {
                this.slotClicked(this.hoveredSlot, this.hoveredSlot.index, 0, ClickType.CLONE);
-            } else if (this.minecraft.options.keyDrop.matches(p_231046_1_, p_231046_2_)) {
+               handled = true;
+            } else if (this.minecraft.options.keyDrop.isActiveAndMatches(mouseKey)) {
                this.slotClicked(this.hoveredSlot, this.hoveredSlot.index, hasControlDown() ? 1 : 0, ClickType.THROW);
+               handled = true;
             }
+         } else if (this.minecraft.options.keyDrop.isActiveAndMatches(mouseKey)) {
+             handled = true; // Forge MC-146650: Emulate MC bug, so we don't drop from hotbar when pressing drop without hovering over a item.
          }
 
-         return true;
+         return handled;
       }
    }
 
    protected boolean checkHotbarKeyPressed(int p_195363_1_, int p_195363_2_) {
       if (this.minecraft.player.inventory.getCarried().isEmpty() && this.hoveredSlot != null) {
-         if (this.minecraft.options.keySwapOffhand.matches(p_195363_1_, p_195363_2_)) {
+         if (this.minecraft.options.keySwapOffhand.isActiveAndMatches(InputMappings.getKey(p_195363_1_, p_195363_2_))) {
             this.slotClicked(this.hoveredSlot, this.hoveredSlot.index, 40, ClickType.SWAP);
             return true;
          }
 
          for(int i = 0; i < 9; ++i) {
-            if (this.minecraft.options.keyHotbarSlots[i].matches(p_195363_1_, p_195363_2_)) {
+            if (this.minecraft.options.keyHotbarSlots[i].isActiveAndMatches(InputMappings.getKey(p_195363_1_, p_195363_2_))) {
                this.slotClicked(this.hoveredSlot, this.hoveredSlot.index, i, ClickType.SWAP);
                return true;
             }
@@ -587,6 +_,18 @@
 
    public T getMenu() {
       return this.menu;
+   }
+
+   @javax.annotation.Nullable
+   public Slot getSlotUnderMouse() { return this.hoveredSlot; }
+   public int getGuiLeft() { return leftPos; }
+   public int getGuiTop() { return topPos; }
+   public int getXSize() { return imageWidth; }
+   public int getYSize() { return imageHeight; }
+
+   protected int slotColor = -2130706433;
+   public int getSlotColor(int index) {
+      return slotColor;
    }
 
    public void onClose() {
