--- a/net/minecraft/client/renderer/model/multipart/Multipart.java
+++ b/net/minecraft/client/renderer/model/multipart/Multipart.java
@@ -82,11 +82,12 @@
    }
 
    @Nullable
-   public IBakedModel func_217641_a(ModelBakery p_217641_1_, Function<ResourceLocation, TextureAtlasSprite> p_217641_2_, ISprite p_217641_3_) {
+   @Override
+   public IBakedModel bake(ModelBakery p_217641_1_, Function<ResourceLocation, TextureAtlasSprite> p_217641_2_, ISprite p_217641_3_, net.minecraft.client.renderer.vertex.VertexFormat format) {
       MultipartBakedModel.Builder multipartbakedmodel$builder = new MultipartBakedModel.Builder();
 
       for(Selector selector : this.func_188136_a()) {
-         IBakedModel ibakedmodel = selector.func_188165_a().func_217641_a(p_217641_1_, p_217641_2_, p_217641_3_);
+         IBakedModel ibakedmodel = selector.func_188165_a().bake(p_217641_1_, p_217641_2_, p_217641_3_, format);
          if (ibakedmodel != null) {
             multipartbakedmodel$builder.func_188648_a(selector.func_188166_a(this.field_188140_b), ibakedmodel);
          }
