--- ../src-base/minecraft/net/minecraft/server/integrated/IntegratedServerListenThread.java
+++ ../src-work/minecraft/net/minecraft/server/integrated/IntegratedServerListenThread.java
@@ -1,5 +1,6 @@
 package net.minecraft.server.integrated;
 
+import cpw.mods.fml.common.network.FMLNetworkHandler;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.io.IOException;
