--- ../src-base/minecraft/net/minecraft/inventory/ContainerPlayer.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerPlayer.java
@@ -48,8 +48,7 @@
                     }
                     else
                     {
-                        EntityEquipmentSlot entityequipmentslot1 = EntityLiving.func_184640_d(p_75214_1_);
-                        return entityequipmentslot1 == entityequipmentslot;
+                        return p_75214_1_.func_77973_b().isValidArmor(p_75214_1_, entityequipmentslot, field_82862_h);
                     }
                 }
                 @SideOnly(Side.CLIENT)
