--- a/net/minecraft/inventory/ContainerRepair.java
+++ b/net/minecraft/inventory/ContainerRepair.java
@@ -61,6 +61,8 @@
                p_190901_1_.func_82242_a(-ContainerRepair.this.field_82854_e);
             }
 
+            float breakChance = net.minecraftforge.common.ForgeHooks.onAnvilRepair(p_190901_1_, p_190901_2_, ContainerRepair.this.field_82853_g.func_70301_a(0), ContainerRepair.this.field_82853_g.func_70301_a(1));
+
             ContainerRepair.this.field_82853_g.func_70299_a(0, ItemStack.field_190927_a);
             if (ContainerRepair.this.field_82856_l > 0) {
                ItemStack itemstack = ContainerRepair.this.field_82853_g.func_70301_a(1);
@@ -77,7 +79,7 @@
             ContainerRepair.this.field_82854_e = 0;
             IBlockState iblockstate1 = p_i45807_2_.func_180495_p(p_i45807_3_);
             if (!p_i45807_2_.field_72995_K) {
-               if (!p_190901_1_.field_71075_bZ.field_75098_d && iblockstate1.func_203425_a(BlockTags.field_200572_k) && p_190901_1_.func_70681_au().nextFloat() < 0.12F) {
+               if (!p_190901_1_.field_71075_bZ.field_75098_d && iblockstate1.func_203425_a(BlockTags.field_200572_k) && p_190901_1_.func_70681_au().nextFloat() < breakChance) {
                   IBlockState iblockstate = BlockAnvil.func_196433_f(iblockstate1);
                   if (iblockstate == null) {
                      p_i45807_2_.func_175698_g(p_i45807_3_);
@@ -130,8 +132,11 @@
          Map<Enchantment, Integer> map = EnchantmentHelper.func_82781_a(itemstack1);
          j = j + itemstack.func_82838_A() + (itemstack2.func_190926_b() ? 0 : itemstack2.func_82838_A());
          this.field_82856_l = 0;
+         boolean flag = false;
+
          if (!itemstack2.func_190926_b()) {
-            boolean flag = itemstack2.func_77973_b() == Items.field_151134_bR && !ItemEnchantedBook.func_92110_g(itemstack2).isEmpty();
+            if (!net.minecraftforge.common.ForgeHooks.onAnvilChange(this, itemstack, itemstack2, field_82852_f, field_82857_m, j)) return;
+            flag = itemstack2.func_77973_b() == Items.field_151134_bR && !ItemEnchantedBook.func_92110_g(itemstack2).isEmpty();
             if (itemstack1.func_77984_f() && itemstack1.func_77973_b().func_82789_a(itemstack, itemstack2)) {
                int l2 = Math.min(itemstack1.func_77952_i(), itemstack1.func_77958_k() / 4);
                if (l2 <= 0) {
@@ -248,6 +253,7 @@
             i += k;
             itemstack1.func_200302_a(new TextComponentString(this.field_82857_m));
          }
+         if (flag && !itemstack1.isBookEnchantable(itemstack2)) itemstack1 = ItemStack.field_190927_a;
 
          this.field_82854_e = j + i;
          if (i <= 0) {
