--- ../src-base/minecraft/net/minecraft/inventory/ContainerBeacon.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerBeacon.java
@@ -145,8 +145,7 @@
 
         public boolean func_75214_a(ItemStack p_75214_1_)
         {
-            Item item = p_75214_1_.func_77973_b();
-            return item == Items.field_151166_bC || item == Items.field_151045_i || item == Items.field_151043_k || item == Items.field_151042_j;
+            return p_75214_1_.func_77973_b().isBeaconPayment(p_75214_1_);
         }
 
         public int func_75219_a()
