--- ../src_base/minecraft/net/minecraft/inventory/ContainerFurnace.java
+++ ../src_work/minecraft/net/minecraft/inventory/ContainerFurnace.java
@@ -95,6 +95,14 @@
             this.furnace.currentItemBurnTime = par2;
         }
     }
+    
+    /**
+     * Returns the Tile Entity behind this furnace inventory / container
+     */
+    public TileEntityFurnace getFurnace()
+    {
+        return this.furnace;
+    }
 
     public boolean canInteractWith(EntityPlayer par1EntityPlayer)
     {
@@ -125,7 +133,7 @@
             }
             else if (par2 != 1 && par2 != 0)
             {
-                if (FurnaceRecipes.smelting().getSmeltingResult(var5.getItem().shiftedIndex) != null)
+                if (FurnaceRecipes.smelting().getSmeltingResult(var5) != null)
                 {
                     if (!this.mergeItemStack(var5, 0, 1, false))
                     {
