--- ../src-base/minecraft/net/minecraft/inventory/Container.java
+++ ../src-work/minecraft/net/minecraft/inventory/Container.java
@@ -708,7 +708,7 @@
                 p_94525_2_.func_190920_e(1);
                 break;
             case 2:
-                p_94525_2_.func_190920_e(p_94525_2_.func_77973_b().func_77639_j());
+                p_94525_2_.func_190920_e(p_94525_2_.func_77976_d());
         }
 
         p_94525_2_.func_190917_f(p_94525_3_);
