--- ../src_base/minecraft/net/minecraft/inventory/ContainerDispenser.java
+++ ../src_work/minecraft/net/minecraft/inventory/ContainerDispenser.java
@@ -3,6 +3,9 @@
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.tileentity.TileEntityDispenser;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event;
+import net.minecraftforge.event.container.ContainerInteractEvent;
 
 public class ContainerDispenser extends Container
 {
@@ -38,7 +41,13 @@
 
     public boolean canInteractWith(EntityPlayer par1EntityPlayer)
     {
-        return this.tileEntityDispenser.isUseableByPlayer(par1EntityPlayer);
+        ContainerInteractEvent event = new ContainerInteractEvent(this, par1EntityPlayer);
+        MinecraftForge.EVENT_BUS.post(event);
+        if (event.isCanceled())
+        {
+            return false;
+        }
+        return event.interact == Event.Result.ALLOW ? true : event.interact == Event.Result.DEFAULT ? this.tileEntityDispenser.isUseableByPlayer(par1EntityPlayer) : false;
     }
 
     /**
