--- ../src_base/minecraft/net/minecraft/inventory/ContainerEnchantment.java
+++ ../src_work/minecraft/net/minecraft/inventory/ContainerEnchantment.java
@@ -12,6 +12,9 @@
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event;
+import net.minecraftforge.event.container.ContainerInteractEvent;
 
 public class ContainerEnchantment extends Container
 {
@@ -244,7 +247,14 @@
 
     public boolean canInteractWith(EntityPlayer par1EntityPlayer)
     {
-        return this.worldPointer.getBlockId(this.posX, this.posY, this.posZ) != Block.enchantmentTable.blockID ? false : par1EntityPlayer.getDistanceSq((double)this.posX + 0.5D, (double)this.posY + 0.5D, (double)this.posZ + 0.5D) <= 64.0D;
+        ContainerInteractEvent event = new ContainerInteractEvent(this, par1EntityPlayer);
+        MinecraftForge.EVENT_BUS.post(event);
+        if (event.isCanceled())
+        {
+            return false;
+        }
+        boolean flag = this.worldPointer.getBlockId(this.posX, this.posY, this.posZ) != Block.enchantmentTable.blockID ? false : par1EntityPlayer.getDistanceSq((double)this.posX + 0.5D, (double)this.posY + 0.5D, (double)this.posZ + 0.5D) <= 64.0D;
+        return event.interact == Event.Result.ALLOW ? true : event.interact == Event.Result.DEFAULT ? flag : false;
     }
 
     /**
