--- ../src-base/minecraft/net/minecraft/inventory/ContainerEnchantment.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerEnchantment.java
@@ -12,6 +12,7 @@
 import net.minecraft.init.Items;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+import net.minecraftforge.common.ForgeHooks;
 
 public class ContainerEnchantment extends Container
 {
@@ -116,6 +117,7 @@
                 {
                     i = 0;
                     int j;
+                    float power = 0;
 
                     for (j = -1; j <= 1; ++j)
                     {
@@ -123,37 +125,15 @@
                         {
                             if ((j != 0 || k != 0) && this.field_75172_h.func_147437_c(this.field_75173_i + k, this.field_75170_j, this.field_75171_k + j) && this.field_75172_h.func_147437_c(this.field_75173_i + k, this.field_75170_j + 1, this.field_75171_k + j))
                             {
-                                if (this.field_75172_h.func_147439_a(this.field_75173_i + k * 2, this.field_75170_j, this.field_75171_k + j * 2) == Blocks.field_150342_X)
-                                {
-                                    ++i;
-                                }
+                                power += ForgeHooks.getEnchantPower(field_75172_h, field_75173_i + k * 2, field_75170_j,     field_75171_k + j * 2);
+                                power += ForgeHooks.getEnchantPower(field_75172_h, field_75173_i + k * 2, field_75170_j + 1, field_75171_k + j * 2);
 
-                                if (this.field_75172_h.func_147439_a(this.field_75173_i + k * 2, this.field_75170_j + 1, this.field_75171_k + j * 2) == Blocks.field_150342_X)
-                                {
-                                    ++i;
-                                }
-
                                 if (k != 0 && j != 0)
                                 {
-                                    if (this.field_75172_h.func_147439_a(this.field_75173_i + k * 2, this.field_75170_j, this.field_75171_k + j) == Blocks.field_150342_X)
-                                    {
-                                        ++i;
-                                    }
-
-                                    if (this.field_75172_h.func_147439_a(this.field_75173_i + k * 2, this.field_75170_j + 1, this.field_75171_k + j) == Blocks.field_150342_X)
-                                    {
-                                        ++i;
-                                    }
-
-                                    if (this.field_75172_h.func_147439_a(this.field_75173_i + k, this.field_75170_j, this.field_75171_k + j * 2) == Blocks.field_150342_X)
-                                    {
-                                        ++i;
-                                    }
-
-                                    if (this.field_75172_h.func_147439_a(this.field_75173_i + k, this.field_75170_j + 1, this.field_75171_k + j * 2) == Blocks.field_150342_X)
-                                    {
-                                        ++i;
-                                    }
+                                    power += ForgeHooks.getEnchantPower(field_75172_h, field_75173_i + k * 2, field_75170_j,     field_75171_k + j    );
+                                    power += ForgeHooks.getEnchantPower(field_75172_h, field_75173_i + k * 2, field_75170_j + 1, field_75171_k + j    );
+                                    power += ForgeHooks.getEnchantPower(field_75172_h, field_75173_i + k,     field_75170_j,     field_75171_k + j * 2);
+                                    power += ForgeHooks.getEnchantPower(field_75172_h, field_75173_i + k,     field_75170_j + 1, field_75171_k + j * 2);
                                 }
                             }
                         }
@@ -161,7 +141,7 @@
 
                     for (j = 0; j < 3; ++j)
                     {
-                        this.field_75167_g[j] = EnchantmentHelper.func_77514_a(this.field_75169_l, j, i, itemstack);
+                        this.field_75167_g[j] = EnchantmentHelper.func_77514_a(this.field_75169_l, j, (int)power, itemstack);
                     }
 
                     this.func_75142_b();
