--- ../src_base/minecraft/net/minecraft/inventory/ContainerMerchant.java
+++ ../src_work/minecraft/net/minecraft/inventory/ContainerMerchant.java
@@ -7,6 +7,9 @@
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event;
+import net.minecraftforge.event.container.ContainerInteractEvent;
 
 public class ContainerMerchant extends Container
 {
@@ -78,7 +81,13 @@
 
     public boolean canInteractWith(EntityPlayer par1EntityPlayer)
     {
-        return this.theMerchant.getCustomer() == par1EntityPlayer;
+        ContainerInteractEvent event = new ContainerInteractEvent(this, par1EntityPlayer);
+        MinecraftForge.EVENT_BUS.post(event);
+        if (event.isCanceled())
+        {
+            return false;
+        }
+        return event.interact == Event.Result.ALLOW ? true : event.interact == Event.Result.DEFAULT ? this.theMerchant.getCustomer() == par1EntityPlayer : false;
     }
 
     /**
