--- ../src-base/minecraft/net/minecraft/inventory/ContainerBrewingStand.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerBrewingStand.java
@@ -5,6 +5,7 @@
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.init.Items;
+import net.minecraft.item.ItemPotion;
 import net.minecraft.item.ItemStack;
 import net.minecraft.stats.AchievementList;
 import net.minecraft.tileentity.TileEntityBrewingStand;
@@ -194,7 +195,7 @@
 
             public void func_82870_a(EntityPlayer p_82870_1_, ItemStack p_82870_2_)
             {
-                if (p_82870_2_.func_77973_b() == Items.field_151068_bn && p_82870_2_.func_77960_j() > 0)
+                if (p_82870_2_.func_77973_b() instanceof ItemPotion && p_82870_2_.func_77960_j() > 0)
                 {
                     this.field_75244_a.func_71064_a(AchievementList.field_76001_A, 1);
                 }
@@ -204,7 +205,7 @@
 
             public static boolean func_75243_a_(ItemStack p_75243_0_)
             {
-                return p_75243_0_ != null && (p_75243_0_.func_77973_b() == Items.field_151068_bn || p_75243_0_.func_77973_b() == Items.field_151069_bo);
+                return p_75243_0_ != null && (p_75243_0_.func_77973_b() instanceof ItemPotion || p_75243_0_.func_77973_b() == Items.field_151069_bo);
             }
         }
 }
