--- ../src-base/minecraft/net/minecraft/inventory/ContainerPlayer.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerPlayer.java
@@ -44,7 +44,7 @@
                 }
                 public boolean func_75214_a(ItemStack p_75214_1_)
                 {
-                    return entityequipmentslot == EntityLiving.func_184640_d(p_75214_1_);
+                    return p_75214_1_.func_77973_b().isValidArmor(p_75214_1_, entityequipmentslot, field_82862_h);
                 }
                 public boolean func_82869_a(EntityPlayer p_82869_1_)
                 {
