--- a/net/minecraft/inventory/ContainerEnchantment.java
+++ b/net/minecraft/inventory/ContainerEnchantment.java
@@ -61,7 +61,7 @@
       });
       this.func_75146_a(new Slot(this.field_75168_e, 1, 35, 47) {
          public boolean func_75214_a(ItemStack p_75214_1_) {
-            return p_75214_1_.func_77973_b() == Items.field_196128_bn;
+            return net.minecraftforge.common.Tags.Items.GEMS_LAPIS.func_199685_a_(p_75214_1_.func_77973_b());
          }
       });
 
@@ -121,40 +121,29 @@
 
    }
 
+   private float getPower(BlockPos pos) {
+      return field_75172_h.func_180495_p(pos).getEnchantPowerBonus(field_75172_h, pos);
+   }
+
    public void func_75130_a(IInventory p_75130_1_) {
       if (p_75130_1_ == this.field_75168_e) {
          ItemStack itemstack = p_75130_1_.func_70301_a(0);
          if (!itemstack.func_190926_b() && itemstack.func_77956_u()) {
             if (!this.field_75172_h.field_72995_K) {
                int l = 0;
+               float power = 0;
 
                for(int j = -1; j <= 1; ++j) {
                   for(int k = -1; k <= 1; ++k) {
                      if ((j != 0 || k != 0) && this.field_75172_h.func_175623_d(this.field_178150_j.func_177982_a(k, 0, j)) && this.field_75172_h.func_175623_d(this.field_178150_j.func_177982_a(k, 1, j))) {
-                        if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k * 2, 0, j * 2)).func_177230_c() == Blocks.field_150342_X) {
-                           ++l;
-                        }
+                        power += getPower(field_178150_j.func_177982_a(k * 2, 0, j * 2));
+                        power += getPower(field_178150_j.func_177982_a(k * 2, 1, j * 2));
 
-                        if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k * 2, 1, j * 2)).func_177230_c() == Blocks.field_150342_X) {
-                           ++l;
-                        }
-
                         if (k != 0 && j != 0) {
-                           if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k * 2, 0, j)).func_177230_c() == Blocks.field_150342_X) {
-                              ++l;
-                           }
-
-                           if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k * 2, 1, j)).func_177230_c() == Blocks.field_150342_X) {
-                              ++l;
-                           }
-
-                           if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k, 0, j * 2)).func_177230_c() == Blocks.field_150342_X) {
-                              ++l;
-                           }
-
-                           if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k, 1, j * 2)).func_177230_c() == Blocks.field_150342_X) {
-                              ++l;
-                           }
+                           power += getPower(field_178150_j.func_177982_a(k * 2, 0, j));
+                           power += getPower(field_178150_j.func_177982_a(k * 2, 1, j));
+                           power += getPower(field_178150_j.func_177982_a(k, 0, j * 2));
+                           power += getPower(field_178150_j.func_177982_a(k, 1, j * 2));
                         }
                      }
                   }
@@ -163,12 +152,13 @@
                this.field_75169_l.setSeed((long)this.field_178149_f);
 
                for(int i1 = 0; i1 < 3; ++i1) {
-                  this.field_75167_g[i1] = EnchantmentHelper.func_77514_a(this.field_75169_l, i1, l, itemstack);
+                  this.field_75167_g[i1] = EnchantmentHelper.func_77514_a(this.field_75169_l, i1, (int)power, itemstack);
                   this.field_185001_h[i1] = -1;
                   this.field_185002_i[i1] = -1;
                   if (this.field_75167_g[i1] < i1 + 1) {
                      this.field_75167_g[i1] = 0;
                   }
+                  this.field_75167_g[i1] = net.minecraftforge.event.ForgeEventFactory.onEnchantmentLevelSet(field_75172_h, field_178150_j, i1, (int)power, itemstack, field_75167_g[i1]);
                }
 
                for(int j1 = 0; j1 < 3; ++j1) {
@@ -300,9 +290,8 @@
                return ItemStack.field_190927_a;
             }
 
-            if (itemstack1.func_77942_o() && itemstack1.func_190916_E() == 1) {
-               this.field_75151_b.get(0).func_75215_d(itemstack1.func_77946_l());
-               itemstack1.func_190920_e(0);
+            if (itemstack1.func_77942_o()) { // Forge: Fix MC-17431
+               ((Slot)this.field_75151_b.get(0)).func_75215_d(itemstack1.func_77979_a(1));
             } else if (!itemstack1.func_190926_b()) {
                this.field_75151_b.get(0).func_75215_d(new ItemStack(itemstack1.func_77973_b()));
                itemstack1.func_190918_g(1);
