--- a/net/minecraft/inventory/Container.java
+++ b/net/minecraft/inventory/Container.java
@@ -67,9 +67,11 @@
          ItemStack itemstack = this.field_75151_b.get(i).func_75211_c();
          ItemStack itemstack1 = this.field_75153_a.get(i);
          if (!ItemStack.func_77989_b(itemstack1, itemstack)) {
+            boolean clientStackChanged = !itemstack1.equals(itemstack, true);
             itemstack1 = itemstack.func_190926_b() ? ItemStack.field_190927_a : itemstack.func_77946_l();
             this.field_75153_a.set(i, itemstack1);
 
+            if (clientStackChanged)
             for(int j = 0; j < this.field_75149_d.size(); ++j) {
                this.field_75149_d.get(j).func_71111_a(this, i, itemstack1);
             }
@@ -430,14 +432,15 @@
             ItemStack itemstack = slot.func_75211_c();
             if (!itemstack.func_190926_b() && func_195929_a(p_75135_1_, itemstack)) {
                int j = itemstack.func_190916_E() + p_75135_1_.func_190916_E();
-               if (j <= p_75135_1_.func_77976_d()) {
+               int maxSize = Math.min(slot.func_75219_a(), p_75135_1_.func_77976_d());
+               if (j <= maxSize) {
                   p_75135_1_.func_190920_e(0);
                   itemstack.func_190920_e(j);
                   slot.func_75218_e();
                   flag = true;
-               } else if (itemstack.func_190916_E() < p_75135_1_.func_77976_d()) {
-                  p_75135_1_.func_190918_g(p_75135_1_.func_77976_d() - itemstack.func_190916_E());
-                  itemstack.func_190920_e(p_75135_1_.func_77976_d());
+               } else if (itemstack.func_190916_E() < maxSize) {
+                  p_75135_1_.func_190918_g(maxSize - itemstack.func_190916_E());
+                  itemstack.func_190920_e(maxSize);
                   slot.func_75218_e();
                   flag = true;
                }
@@ -538,7 +541,7 @@
          p_94525_2_.func_190920_e(1);
          break;
       case 2:
-         p_94525_2_.func_190920_e(p_94525_2_.func_77973_b().func_77639_j());
+         p_94525_2_.func_190920_e(p_94525_2_.func_77976_d());
       }
 
       p_94525_2_.func_190917_f(p_94525_3_);
@@ -576,7 +579,7 @@
       if (!p_192389_1_.field_72995_K) {
          EntityPlayerMP entityplayermp = (EntityPlayerMP)p_192389_2_;
          ItemStack itemstack = ItemStack.field_190927_a;
-         IRecipe irecipe = p_192389_1_.func_73046_m().func_199529_aN().func_199515_b(p_192389_3_, p_192389_1_);
+         IRecipe irecipe = p_192389_1_.func_73046_m().func_199529_aN().getRecipe(p_192389_3_, p_192389_1_, net.minecraftforge.common.crafting.VanillaRecipeTypes.CRAFTING);
          if (p_192389_4_.func_201561_a(p_192389_1_, entityplayermp, irecipe) && irecipe != null) {
             itemstack = irecipe.func_77572_b(p_192389_3_);
          }
