--- ../src_base/minecraft/net/minecraft/inventory/SlotCrafting.java
+++ ../src_work/minecraft/net/minecraft/inventory/SlotCrafting.java
@@ -6,6 +6,10 @@
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.stats.AchievementList;
+
+import net.minecraftforge.common.ForgeHooks;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.entity.player.PlayerDestroyItemEvent;
 
 public class SlotCrafting extends Slot
 {
@@ -124,9 +128,15 @@
 
                 if (var4.getItem().hasContainerItem())
                 {
-                    ItemStack var5 = new ItemStack(var4.getItem().getContainerItem());
+                    ItemStack var5 = var4.getItem().getContainerItemStack(var4);
+                    
+                    if (var5.isItemStackDamageable() && var5.getItemDamage() > var5.getMaxDamage())
+                    {
+                        MinecraftForge.EVENT_BUS.post(new PlayerDestroyItemEvent(thePlayer, var5));
+                        var5 = null;
+                    }
 
-                    if (!var4.getItem().doesContainerItemLeaveCraftingGrid(var4) || !this.thePlayer.inventory.addItemStackToInventory(var5))
+                    if (var5 != null && (!var4.getItem().doesContainerItemLeaveCraftingGrid(var4) || !this.thePlayer.inventory.addItemStackToInventory(var5)))
                     {
                         if (this.craftMatrix.getStackInSlot(var3) == null)
                         {
