--- a/net/minecraft/inventory/SlotCrafting.java
+++ b/net/minecraft/inventory/SlotCrafting.java
@@ -39,6 +39,7 @@
    protected void func_75208_c(ItemStack p_75208_1_) {
       if (this.field_75237_g > 0) {
          p_75208_1_.func_77980_a(this.field_75238_b.field_70170_p, this.field_75238_b, this.field_75237_g);
+         net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerCraftingEvent(this.field_75238_b, p_75208_1_, this.field_75239_a);
       }
 
       ((IRecipeHolder)this.field_75224_c).func_201560_d(this.field_75238_b);
@@ -47,7 +48,9 @@
 
    public ItemStack func_190901_a(EntityPlayer p_190901_1_, ItemStack p_190901_2_) {
       this.func_75208_c(p_190901_2_);
-      NonNullList<ItemStack> nonnulllist = p_190901_1_.field_70170_p.func_199532_z().func_199513_c(this.field_75239_a, p_190901_1_.field_70170_p);
+      net.minecraftforge.common.ForgeHooks.setCraftingPlayer(p_190901_1_);
+      NonNullList<ItemStack> nonnulllist = p_190901_1_.field_70170_p.func_199532_z().getRemainingItems(this.field_75239_a, p_190901_1_.field_70170_p, net.minecraftforge.common.crafting.VanillaRecipeTypes.CRAFTING);
+      net.minecraftforge.common.ForgeHooks.setCraftingPlayer(null);
 
       for(int i = 0; i < nonnulllist.size(); ++i) {
          ItemStack itemstack = this.field_75239_a.func_70301_a(i);
