--- ../src_base/minecraft/net/minecraft/inventory/SlotFurnace.java
+++ ../src_work/minecraft/net/minecraft/inventory/SlotFurnace.java
@@ -69,7 +69,7 @@
         if (!this.thePlayer.worldObj.isRemote)
         {
             int var2 = this.field_75228_b;
-            float var3 = FurnaceRecipes.smelting().getExperience(par1ItemStack.itemID);
+            float var3 = FurnaceRecipes.smelting().getExperience(par1ItemStack);
             int var4;
 
             if (var3 == 0.0F)
