--- ../src-base/minecraft/net/minecraft/entity/EntitySpawnPlacementRegistry.java
+++ ../src-work/minecraft/net/minecraft/entity/EntitySpawnPlacementRegistry.java
@@ -44,6 +44,12 @@
         return (EntityLiving.SpawnPlacementType)ENTITY_PLACEMENTS.get(entityClass);
     }
 
+    public static void setPlacementType(Class<? extends Entity> entityClass, EntityLiving.SpawnPlacementType placementType)
+    {
+        if(!ENTITY_PLACEMENTS.containsKey(entityClass))
+            ENTITY_PLACEMENTS.put(entityClass, placementType);
+    }
+
     static
     {
         ENTITY_PLACEMENTS.put(EntityBat.class, EntityLiving.SpawnPlacementType.ON_GROUND);
