--- ../src-base/minecraft/net/minecraft/entity/EntityTrackerEntry.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityTrackerEntry.java
@@ -477,6 +477,9 @@
             logger.warn("Fetching addPacket for removed entity");
         }
 
+        Packet pkt = net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.getEntitySpawningPacket(this.trackedEntity);
+        if (pkt != null) return pkt;
+
         if (this.trackedEntity instanceof EntityItem)
         {
             return new SPacketSpawnObject(this.trackedEntity, 2, 1);
