--- a/net/minecraft/client/model/geom/ModelLayers.java
+++ b/net/minecraft/client/model/geom/ModelLayers.java
@@ -220,11 +_,13 @@
    }
 
    public static ModelLayerLocation m_171291_(WoodType p_171292_) {
-      return m_171300_("sign/" + p_171292_.f_61839_(), "main");
+      ResourceLocation location = new ResourceLocation(p_171292_.f_61839_());
+      return new ModelLayerLocation(new ResourceLocation(location.m_135827_(), "sign/" + location.m_135815_()), "main");
    }
 
    public static ModelLayerLocation m_247439_(WoodType p_252225_) {
-      return m_171300_("hanging_sign/" + p_252225_.f_61839_(), "main");
+      ResourceLocation location = new ResourceLocation(p_252225_.f_61839_());
+      return new ModelLayerLocation(new ResourceLocation(location.m_135827_(), "hanging_sign/" + location.m_135815_()), "main");
    }
 
    public static Stream<ModelLayerLocation> m_171288_() {
