--- a/net/minecraft/data/FluidTagsProvider.java
+++ b/net/minecraft/data/FluidTagsProvider.java
@@ -8,8 +_,12 @@
 import net.minecraft.util.registry.Registry;
 
 public class FluidTagsProvider extends TagsProvider<Fluid> {
+   @Deprecated
    public FluidTagsProvider(DataGenerator p_i49156_1_) {
       super(p_i49156_1_, Registry.field_212619_h);
+   }
+   public FluidTagsProvider(DataGenerator p_i49156_1_, String modId, @javax.annotation.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+      super(p_i49156_1_, Registry.field_212619_h, modId, existingFileHelper);
    }
 
    protected void func_200432_c() {
