--- a/net/minecraft/data/Main.java
+++ b/net/minecraft/data/Main.java
@@ -76,14 +_,17 @@
         OptionSpec<Void> optionspec4 = optionparser.accepts("all", "Include all generators");
         OptionSpec<String> optionspec5 = optionparser.accepts("output", "Output folder").withRequiredArg().defaultsTo("generated");
         OptionSpec<String> optionspec6 = optionparser.accepts("input", "Input folder").withRequiredArg();
+        var loader = net.minecraftforge.data.loading.DatagenModLoader.setup(optionparser, false);
         OptionSet optionset = optionparser.parse(p_129669_);
-        if (!optionset.has(optionspec) && optionset.hasOptions()) {
+        if (!optionset.has(optionspec) && optionset.hasOptions() && loader.hasArgs(optionset)) {
             Path path = Paths.get(optionspec5.value(optionset));
             boolean flag = optionset.has(optionspec4);
             boolean flag1 = flag || optionset.has(optionspec1);
             boolean flag2 = flag || optionset.has(optionspec2);
             boolean flag3 = flag || optionset.has(optionspec3);
             Collection<Path> collection = optionset.valuesOf(optionspec6).stream().map(p_129659_ -> Paths.get(p_129659_)).toList();
+            if (!loader.run(optionset, path, collection, flag1, flag, flag2, flag3, true))
+                return;
             DataGenerator datagenerator = new DataGenerator(path, SharedConstants.getCurrentVersion(), true);
             addServerProviders(datagenerator, collection, flag1, flag2, flag3);
             datagenerator.run();
