--- a/net/minecraft/data/RecipeProvider.java
+++ b/net/minecraft/data/RecipeProvider.java
@@ -57,6 +57,7 @@
 
          }
       });
+      if (this.getClass() == RecipeProvider.class) //Forge: Subclasses don't need this.
       this.func_208310_b(p_200398_1_, Advancement.Builder.func_200278_a().func_200275_a("impossible", new ImpossibleTrigger.Instance()).func_200273_b(), path.resolve("data/minecraft/advancements/recipes/root.json"));
    }
 
