--- a/net/minecraft/data/TagsProvider.java
+++ b/net/minecraft/data/TagsProvider.java
@@ -53,6 +53,7 @@
          } else {
             JsonObject jsonobject = p_240524_5_.func_232965_c_();
             Path path = this.func_200431_a(p_240524_4_);
+            if (path == null) return; //Forge: Allow running this data provider without writing it. Recipe provider needs valid tags.
 
             try {
                String s = field_200437_e.toJson((JsonElement)jsonobject);
@@ -87,7 +88,7 @@
       });
    }
 
-   public static class Builder<T> {
+   public static class Builder<T> implements net.minecraftforge.common.extensions.IForgeTagBuilder<T> {
       private final ITag.Builder field_240528_a_;
       private final Registry<T> field_240529_b_;
       private final String field_240530_c_;
