--- ../src_base/minecraft/net/minecraft/entity/ai/EntityAIMate.java
+++ ../src_work/minecraft/net/minecraft/entity/ai/EntityAIMate.java
@@ -119,7 +119,7 @@
             this.targetMate.resetInLove();
             entityageable.setGrowingAge(-24000);
             entityageable.setLocationAndAngles(this.theAnimal.posX, this.theAnimal.posY, this.theAnimal.posZ, 0.0F, 0.0F);
-            this.theWorld.spawnEntityInWorld(entityageable);
+            this.theWorld.addEntity(entityageable, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
             Random random = this.theAnimal.getRNG();
 
             for (int i = 0; i < 7; ++i)
