--- a/net/minecraft/entity/ai/EntityAIMate.java
+++ b/net/minecraft/entity/ai/EntityAIMate.java
@@ -76,6 +76,17 @@
 
    protected void func_75388_i() {
       EntityAgeable entityageable = this.field_75390_d.func_90011_a(this.field_75391_e);
+      final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(field_75390_d, field_75391_e, entityageable);
+      final boolean cancelled = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+      entityageable = event.getChild();
+      if (cancelled) {
+         //Reset the "inLove" state for the animals
+         this.field_75390_d.func_70873_a(6000);
+         this.field_75391_e.func_70873_a(6000);
+         this.field_75390_d.func_70875_t();
+         this.field_75391_e.func_70875_t();
+         return;
+      }
       if (entityageable != null) {
          EntityPlayerMP entityplayermp = this.field_75390_d.func_191993_do();
          if (entityplayermp == null && this.field_75391_e.func_191993_do() != null) {
