--- a/net/minecraft/entity/ai/EntityAIVillagerMate.java
+++ b/net/minecraft/entity/ai/EntityAIVillagerMate.java
@@ -82,11 +82,14 @@
    }
 
    private void func_75447_i() {
-      EntityVillager entityvillager = this.field_75450_b.func_90011_a(this.field_75451_c);
+      net.minecraft.entity.EntityAgeable entityvillager = this.field_75450_b.func_90011_a(this.field_75451_c);
       this.field_75451_c.func_70873_a(6000);
       this.field_75450_b.func_70873_a(6000);
       this.field_75451_c.func_175549_o(false);
       this.field_75450_b.func_175549_o(false);
+      final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(field_75450_b, field_75451_c, entityvillager);
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event) || event.getChild() == null) return;
+      entityvillager = event.getChild();
       entityvillager.func_70873_a(-24000);
       entityvillager.func_70012_b(this.field_75450_b.field_70165_t, this.field_75450_b.field_70163_u, this.field_75450_b.field_70161_v, 0.0F, 0.0F);
       this.field_75448_d.func_72838_d(entityvillager);
