--- ../src_base/minecraft/net/minecraft/entity/ai/EntityAIRunAroundLikeCrazy.java
+++ ../src_work/minecraft/net/minecraft/entity/ai/EntityAIRunAroundLikeCrazy.java
@@ -85,8 +85,20 @@
                 this.horseHost.increaseTemper(5);
             }
 
-            this.horseHost.riddenByEntity.mountEntity((Entity)null);
-            this.horseHost.riddenByEntity = null;
+            // CraftBukkit start - Handle dismounting to account for VehicleExitEvent being fired.
+            if (this.horseHost.riddenByEntity != null)
+            {
+                this.horseHost.riddenByEntity.mountEntity((Entity)null);
+
+                // If the entity still has a passenger, then a plugin cancelled the event.
+                if (this.horseHost.riddenByEntity != null)
+                {
+                    return;
+                }
+            }
+
+            // this.field_111180_a.riddenByEntity = null;
+            // CraftBukkit end
             this.horseHost.makeHorseRearWithSound();
             this.horseHost.worldObj.setEntityState(this.horseHost, (byte)6);
         }
