--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIMate.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIMate.java
@@ -86,6 +86,18 @@
     {
         EntityAgeable entityageable = this.field_75390_d.func_90011_a(this.field_75391_e);
 
+        final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(field_75390_d, field_75391_e, entityageable);
+        final boolean cancelled = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+        entityageable = event.getChild();
+        if (cancelled) {
+            //Reset the "inLove" state for the animals
+            this.field_75390_d.func_70873_a(6000);
+            this.field_75391_e.func_70873_a(6000);
+            this.field_75390_d.func_70875_t();
+            this.field_75391_e.func_70875_t();
+            return;
+        }
+
         if (entityageable != null)
         {
             EntityPlayer entityplayer = this.field_75390_d.func_146083_cb();
