--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIAttackOnCollide.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIAttackOnCollide.java
@@ -3,6 +3,7 @@
 import net.minecraft.entity.EntityCreature;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.pathfinding.PathEntity;
+import net.minecraft.pathfinding.PathPoint;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
 
@@ -21,6 +22,8 @@
     private double field_151496_k;
     private static final String __OBFID = "CL_00001595";
 
+    private int failedPathFindingPenalty;
+
     public EntityAIAttackOnCollide(EntityCreature p_i1635_1_, Class p_i1635_2_, double p_i1635_3_, boolean p_i1635_5_)
     {
         this(p_i1635_1_, p_i1635_3_, p_i1635_5_);
@@ -54,8 +57,16 @@
         }
         else
         {
-            this.field_75438_g = this.field_75441_b.func_70661_as().func_75494_a(entitylivingbase);
-            return this.field_75438_g != null;
+            if (-- this.field_75445_i <= 0)
+            {
+                this.field_75438_g = this.field_75441_b.func_70661_as().func_75494_a(entitylivingbase);
+               this.field_75445_i = 4 + this.field_75441_b.func_70681_au().nextInt(7);
+                return this.field_75438_g != null;
+            }
+            else
+            {
+                return true;
+            }
         }
     }
 
@@ -89,8 +100,25 @@
             this.field_151497_i = entitylivingbase.field_70165_t;
             this.field_151495_j = entitylivingbase.field_70121_D.field_72338_b;
             this.field_151496_k = entitylivingbase.field_70161_v;
-            this.field_75445_i = 4 + this.field_75441_b.func_70681_au().nextInt(7);
+            this.field_75445_i = failedPathFindingPenalty + 4 + this.field_75441_b.func_70681_au().nextInt(7);
 
+            if (this.field_75441_b.func_70661_as().func_75505_d() != null)
+            {
+                PathPoint finalPathPoint = this.field_75441_b.func_70661_as().func_75505_d().func_75870_c();
+                if (finalPathPoint != null && entitylivingbase.func_70092_e(finalPathPoint.field_75839_a, finalPathPoint.field_75837_b, finalPathPoint.field_75838_c) < 1)
+                {
+                    failedPathFindingPenalty = 0;
+                }
+                else
+                {
+                    failedPathFindingPenalty += 10;
+                }
+            }
+            else
+            {
+                failedPathFindingPenalty += 10;
+            }
+
             if (d0 > 1024.0D)
             {
                 this.field_75445_i += 10;
