--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIEatGrass.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIEatGrass.java
@@ -70,7 +70,7 @@
 
             if (field_179505_b.apply(this.field_151501_c.func_180495_p(blockpos)))
             {
-                if (this.field_151501_c.func_82736_K().func_82766_b("mobGriefing"))
+                if (this.field_151501_c.func_82736_K().func_82766_b("mobGriefing") && !net.minecraftforge.event.ForgeEventFactory.onLivingUpdateBlock(field_151500_b, blockpos, Blocks.field_150350_a.func_176223_P()))
                 {
                     this.field_151501_c.func_175655_b(blockpos, false);
                 }
@@ -83,7 +83,7 @@
 
                 if (this.field_151501_c.func_180495_p(blockpos1).func_177230_c() == Blocks.field_150349_c)
                 {
-                    if (this.field_151501_c.func_82736_K().func_82766_b("mobGriefing"))
+                    if (this.field_151501_c.func_82736_K().func_82766_b("mobGriefing") && !net.minecraftforge.event.ForgeEventFactory.onLivingUpdateBlock(field_151500_b, blockpos, Blocks.field_150346_d.func_176223_P()))
                     {
                         this.field_151501_c.func_175718_b(2001, blockpos1, Block.func_149682_b(Blocks.field_150349_c));
                         this.field_151501_c.func_180501_a(blockpos1, Blocks.field_150346_d.func_176223_P(), 2);
