--- ../src-base/minecraft/net/minecraft/client/renderer/block/model/Variant.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/block/model/Variant.java
@@ -12,7 +12,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class Variant
+public class Variant implements net.minecraftforge.client.model.ISmartVariant
 {
     private final ResourceLocation field_188050_a;
     private final ModelRotation field_188051_b;
@@ -32,11 +32,17 @@
         return this.field_188050_a;
     }
 
+    @Deprecated
     public ModelRotation func_188048_b()
     {
         return this.field_188051_b;
     }
 
+    public net.minecraftforge.common.model.IModelState getState()
+    {
+        return this.field_188051_b;
+    }
+
     public boolean func_188049_c()
     {
         return this.field_188052_c;
@@ -138,4 +144,9 @@
                 }
             }
         }
+
+    public net.minecraftforge.client.model.IModel process(net.minecraftforge.client.model.IModel base)
+    {
+        return base.uvlock(func_188049_c());
+    }
 }
