--- a/net/minecraft/client/resources/model/cuboid/CuboidModel.java
+++ b/net/minecraft/client/resources/model/cuboid/CuboidModel.java
@@ -28,8 +_,20 @@
     @Nullable Boolean ambientOcclusion,
     @Nullable ItemTransforms transforms,
     TextureSlots.Data textureSlots,
-    @Nullable Identifier parent
+    @Nullable Identifier parent,
+    net.minecraftforge.client.model.@Nullable ForgeBlockModelData forgeData
 ) implements UnbakedModel {
+    public CuboidModel(
+        @Nullable UnbakedGeometry geometry,
+        UnbakedModel.@Nullable GuiLight guiLight,
+        @Nullable Boolean ambientOcclusion,
+        @Nullable ItemTransforms transforms,
+        TextureSlots.Data textureSlots,
+        @Nullable Identifier parent
+    ) {
+        this(geometry, guiLight, ambientOcclusion, transforms, textureSlots, parent, null);
+    }
+
     @VisibleForTesting
     static final Gson GSON = new GsonBuilder()
         .registerTypeAdapter(CuboidModel.class, new CuboidModel.Deserializer())
@@ -62,8 +_,9 @@
                 guiLight = UnbakedModel.GuiLight.getByName(GsonHelper.getAsString(object, "gui_light"));
             }
 
+            var forgeData = net.minecraftforge.client.ForgeHooksClient.deserializeBlockModel(object, context);
             Identifier parentLocation = parentName.isEmpty() ? null : Identifier.parse(parentName);
-            return new CuboidModel(elements, guiLight, hasAmbientOcclusion, transforms, textureMap, parentLocation);
+            return new CuboidModel(elements, guiLight, hasAmbientOcclusion, transforms, textureMap, parentLocation, forgeData);
         }
 
         private TextureSlots.Data getTextureMap(final JsonObject object) {
@@ -84,6 +_,9 @@
         }
 
         protected @Nullable UnbakedGeometry getElements(final JsonDeserializationContext context, final JsonObject object) {
+            var geo = net.minecraftforge.client.ForgeHooksClient.deserializeBlockModelGeometry(object, context);
+            if (geo != null)
+                return geo;
             if (!object.has("elements")) {
                 return null;
             }
