--- a/net/minecraft/world/level/levelgen/structure/templatesystem/PosRuleTestType.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/PosRuleTestType.java
@@ -3,6 +_,7 @@
 import com.mojang.serialization.Codec;
 import net.minecraft.core.Registry;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public interface PosRuleTestType<P extends PosRuleTest> {
    PosRuleTestType<PosAlwaysTrueTest> f_74205_ = m_74211_("always_true", PosAlwaysTrueTest.f_74187_);
    PosRuleTestType<LinearPosTest> f_74206_ = m_74211_("linear_pos", LinearPosTest.f_74147_);
