--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -249,6 +_,10 @@
         return transform(pos, settings.getMirror(), settings.getRotation(), settings.getRotationPivot());
     }
 
+    public static Vec3 transformedVec3d(StructurePlaceSettings placementIn, Vec3 pos) {
+        return transform(pos, placementIn.getMirror(), placementIn.getRotation(), placementIn.getRotationPivot());
+    }
+
     public boolean placeInWorld(
         final ServerLevelAccessor level,
         final BlockPos position,
@@ -276,7 +_,7 @@
             int maxX = Integer.MIN_VALUE;
             int maxY = Integer.MIN_VALUE;
             int maxZ = Integer.MIN_VALUE;
-            List<StructureTemplate.StructureBlockInfo> processedBlockInfoList = processBlockInfos(level, position, referencePos, settings, blockInfoList);
+            List<StructureTemplate.StructureBlockInfo> processedBlockInfoList = processBlockInfos(level, position, referencePos, settings, blockInfoList, this);
 
             try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(LOGGER)) {
                 for (StructureTemplate.StructureBlockInfo blockInfo : processedBlockInfoList) {
@@ -395,10 +_,10 @@
                         position,
                         settings.getMirror(),
                         settings.getRotation(),
-                        settings.getRotationPivot(),
                         boundingBox,
                         settings.shouldFinalizeEntities(),
-                        reporter
+                        reporter,
+                        settings
                     );
                 }
             }
@@ -440,12 +_,21 @@
         });
     }
 
+    /**
+     * @deprecated Forge: Use {@link #processBlockInfos(ServerLevelAccessor, BlockPos, BlockPos, StructurePlaceSettings, List, StructureTemplate)} instead.
+     */
+    @Deprecated
+    public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(final ServerLevelAccessor level, final BlockPos position, final BlockPos referencePos, final StructurePlaceSettings settings, final List<StructureTemplate.StructureBlockInfo> blockInfoList) {
+        return processBlockInfos(level, position, referencePos, settings, blockInfoList, null);
+    }
+
     public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(
         final ServerLevelAccessor level,
         final BlockPos position,
         final BlockPos referencePos,
         final StructurePlaceSettings settings,
-        final List<StructureTemplate.StructureBlockInfo> blockInfoList
+        List<StructureTemplate.StructureBlockInfo> blockInfoList,
+        @Nullable StructureTemplate template
     ) {
         List<StructureTemplate.StructureBlockInfo> originalBlockInfoList = new ArrayList<>();
         List<StructureTemplate.StructureBlockInfo> processedBlockInfoList = new ArrayList<>();
@@ -469,7 +_,7 @@
                 Iterator<StructureProcessor> iterator = settings.getProcessors().iterator();
 
                 while (processedBlockInfo != null && iterator.hasNext()) {
-                    processedBlockInfo = iterator.next().processBlock(level, position, referencePos, blockInfo.pos, processedBlockInfo, settings);
+                    processedBlockInfo = iterator.next().processBlock(level, position, referencePos, blockInfo.pos, processedBlockInfo, settings, template);
                 }
 
                 if (processedBlockInfo != null) {
@@ -491,17 +_,17 @@
         final BlockPos position,
         final Mirror mirror,
         final Rotation rotation,
-        final BlockPos pivot,
         final @Nullable BoundingBox boundingBox,
         final boolean finalizeEntities,
-        final ProblemReporter problemReporter
+        final ProblemReporter problemReporter,
+        final StructurePlaceSettings placementIn
     ) {
-        for (StructureTemplate.StructureEntityInfo entityInfo : this.entityInfoList) {
-            BlockPos blockPos = transform(entityInfo.blockPos, mirror, rotation, pivot).offset(position);
+        var entities = processEntityInfos(this, level, position, placementIn, this.entityInfoList);
+        for (StructureTemplate.StructureEntityInfo entityInfo : entities) {
+            BlockPos blockPos = entityInfo.blockPos; // FORGE: Position will have already been transformed by processEntityInfos
             if (boundingBox == null || boundingBox.isInside(blockPos)) {
                 CompoundTag tag = entityInfo.nbt.copy();
-                Vec3 relativePos = transform(entityInfo.pos, mirror, rotation, pivot);
-                Vec3 pos = relativePos.add(position.getX(), position.getY(), position.getZ());
+                Vec3 pos = entityInfo.pos; // FORGE: Position will have already been transformed by processEntityInfos
                 ListTag posTag = new ListTag();
                 posTag.add(DoubleTag.valueOf(pos.x));
                 posTag.add(DoubleTag.valueOf(pos.y));
@@ -523,6 +_,30 @@
             }
         }
     }
+
+    public static List<StructureTemplate.StructureEntityInfo> processEntityInfos(
+        @Nullable StructureTemplate template,
+        final LevelAccessor level,
+        final BlockPos targetPos,
+        final StructurePlaceSettings settings,
+        final List<StructureTemplate.StructureEntityInfo> entities
+    ) {
+        var list = new ArrayList<StructureTemplate.StructureEntityInfo>();
+        for (var entityInfo : entities) {
+           var pos = transformedVec3d(settings, entityInfo.pos).add(Vec3.atLowerCornerOf(targetPos));
+           var blockpos = calculateRelativePosition(settings, entityInfo.blockPos).offset(targetPos);
+           var info = new StructureTemplate.StructureEntityInfo(pos, blockpos, entityInfo.nbt);
+           for (var proc : settings.getProcessors()) {
+              info = proc.processEntity(level, targetPos, entityInfo, info, settings, template);
+              if (info == null)
+                 break;
+           }
+           if (info != null)
+              list.add(info);
+        }
+        return list;
+     }
+
 
     private static Optional<Entity> createEntityIgnoreException(final ProblemReporter reporter, final ServerLevelAccessor level, final CompoundTag tag) {
         try {
