--- a/net/minecraft/entity/ai/attributes/AttributeModifierMap.java
+++ b/net/minecraft/entity/ai/attributes/AttributeModifierMap.java
@@ -70,6 +70,16 @@
       private final Map<Attribute, ModifiableAttributeInstance> field_233811_a_ = Maps.newHashMap();
       private boolean field_233812_b_;
 
+      public MutableAttribute() { }
+
+      public MutableAttribute(AttributeModifierMap attributeMap) {
+         this.field_233811_a_.putAll(attributeMap.field_233802_a_);
+      }
+
+      public void combine(MutableAttribute other) {
+         this.field_233811_a_.putAll(other.field_233811_a_);
+      }
+
       private ModifiableAttributeInstance func_233817_b_(Attribute p_233817_1_) {
          ModifiableAttributeInstance modifiableattributeinstance = new ModifiableAttributeInstance(p_233817_1_, (p_233816_2_) -> {
             if (this.field_233812_b_) {
