--- ../src-base/minecraft/net/minecraft/server/management/ServerConfigurationManager.java
+++ ../src-work/minecraft/net/minecraft/server/management/ServerConfigurationManager.java
@@ -1,6 +1,9 @@
 package net.minecraft.server.management;
 
 import com.google.common.base.Charsets;
+
+import cpw.mods.fml.common.network.FMLNetworkHandler;
+import cpw.mods.fml.common.registry.GameRegistry;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.io.File;
@@ -124,6 +127,8 @@
 
         p_72355_2_.func_71116_b();
 
+        FMLNetworkHandler.handlePlayerLogin(p_72355_2_, netserverhandler, p_72355_1_);
+
         if (nbttagcompound != null && nbttagcompound.func_74764_b("Riding"))
         {
             Entity entity = EntityList.func_75615_a(nbttagcompound.func_74775_l("Riding"), worldserver);
@@ -238,6 +243,7 @@
 
     public void func_72367_e(EntityPlayerMP p_72367_1_)
     {
+        GameRegistry.onPlayerLogout(p_72367_1_);
         this.func_72391_b(p_72367_1_);
         WorldServer worldserver = p_72367_1_.func_71121_q();
 
@@ -395,6 +401,7 @@
         this.field_72404_b.add(entityplayermp1);
         entityplayermp1.func_71116_b();
         entityplayermp1.func_70606_j(entityplayermp1.func_110143_aJ());
+        GameRegistry.onPlayerRespawn(entityplayermp1);
         return entityplayermp1;
     }
 
@@ -420,6 +427,8 @@
             PotionEffect potioneffect = (PotionEffect)iterator.next();
             p_72356_1_.field_71135_a.func_72567_b(new Packet41EntityEffect(p_72356_1_.field_70157_k, potioneffect));
         }
+
+        GameRegistry.onPlayerChangedDimension(p_72356_1_);
     }
 
     public void func_82448_a(Entity p_82448_1_, int p_82448_2_, WorldServer p_82448_3_, WorldServer p_82448_4_)
