--- a/net/minecraft/world/level/storage/loot/entries/LootPoolEntryType.java
+++ b/net/minecraft/world/level/storage/loot/entries/LootPoolEntryType.java
@@ -3,7 +_,7 @@
 import net.minecraft.world.level.storage.loot.Serializer;
 import net.minecraft.world.level.storage.loot.SerializerType;
 
-public class LootPoolEntryType extends SerializerType<LootPoolEntryContainer> {
+public class LootPoolEntryType extends net.minecraftforge.registries.ForgeSerializerTypeRegistryEntry<LootPoolEntryType, LootPoolEntryContainer> {
    public LootPoolEntryType(Serializer<? extends LootPoolEntryContainer> p_79674_) {
       super(p_79674_);
    }
