--- a/net/minecraft/client/renderer/texture/LayeredColorMaskTexture.java
+++ b/net/minecraft/client/renderer/texture/LayeredColorMaskTexture.java
@@ -38,8 +38,7 @@
             String s = this.field_174949_h.get(i);
             if (s != null) {
                try (
-                  IResource iresource1 = p_195413_1_.func_199002_a(new ResourceLocation(s));
-                  NativeImage nativeimage2 = NativeImage.func_195713_a(iresource1.func_199027_b());
+                  NativeImage nativeimage2 = net.minecraftforge.client.MinecraftForgeClient.getImageLayer(new ResourceLocation(s), p_195413_1_);
                ) {
                   int j = this.field_174950_i.get(i).func_196057_c();
                   if (nativeimage2.func_195702_a() == nativeimage1.func_195702_a() && nativeimage2.func_195714_b() == nativeimage1.func_195714_b()) {
