--- ../src-base/minecraft/net/minecraft/client/renderer/texture/Stitcher.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/texture/Stitcher.java
@@ -56,18 +56,24 @@
     {
         Stitcher.Holder[] astitcher$holder = (Stitcher.Holder[])this.field_94319_a.toArray(new Stitcher.Holder[this.field_94319_a.size()]);
         Arrays.sort((Object[])astitcher$holder);
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar bar = net.minecraftforge.fml.common.ProgressManager.push("Texture stitching", astitcher$holder.length);
 
         for (Stitcher.Holder stitcher$holder : astitcher$holder)
         {
+            bar.step(stitcher$holder.func_98150_a().func_94215_i());
             if (!this.func_94310_b(stitcher$holder))
             {
                 String s = String.format("Unable to fit: %s - size: %dx%d - Maybe try a lowerresolution resourcepack?", new Object[] {stitcher$holder.func_98150_a().func_94215_i(), Integer.valueOf(stitcher$holder.func_98150_a().func_94211_a()), Integer.valueOf(stitcher$holder.func_98150_a().func_94216_b())});
+                net.minecraftforge.fml.common.FMLLog.info(s);
+                for (Stitcher.Holder h : astitcher$holder)
+                    net.minecraftforge.fml.common.FMLLog.info("  %s", h);
                 throw new StitcherException(stitcher$holder, s);
             }
         }
 
         this.field_94318_c = MathHelper.func_151236_b(this.field_94318_c);
         this.field_94315_d = MathHelper.func_151236_b(this.field_94315_d);
+        net.minecraftforge.fml.common.ProgressManager.pop(bar);
     }
 
     public List<TextureAtlasSprite> func_94309_g()
@@ -200,7 +206,7 @@
                 this.field_94204_c = p_i45094_1_.func_94211_a();
                 this.field_94201_d = p_i45094_1_.func_94216_b();
                 this.field_147968_d = p_i45094_2_;
-                this.field_94202_e = Stitcher.func_147969_b(this.field_94201_d, p_i45094_2_) > Stitcher.func_147969_b(this.field_94204_c, p_i45094_2_);
+                this.field_94202_e = false;
             }
 
             public TextureAtlasSprite func_98150_a()
@@ -220,7 +226,7 @@
 
             public void func_94194_d()
             {
-                this.field_94202_e = !this.field_94202_e;
+                // rotation during stitching doesn't work as MC won't rotate textures when putting them on the sheet
             }
 
             public boolean func_94195_e()
@@ -238,7 +244,7 @@
 
             public String toString()
             {
-                return "Holder{width=" + this.field_94204_c + ", height=" + this.field_94201_d + '}';
+                return "Holder{width=" + this.field_94204_c + ", height=" + this.field_94201_d + ", name=" + this.field_98151_a.func_94215_i() + '}';
             }
 
             public int compareTo(Stitcher.Holder p_compareTo_1_)
