--- a/net/minecraft/client/renderer/texture/Stitcher.java
+++ b/net/minecraft/client/renderer/texture/Stitcher.java
@@ -13,6 +13,8 @@
 
 @OnlyIn(Dist.CLIENT)
 public class Stitcher {
+   private static final org.apache.logging.log4j.Logger LOGGER = org.apache.logging.log4j.LogManager.getLogger();
+
    private static final Comparator<Stitcher.Holder> field_217797_a = Comparator.<Stitcher.Holder, Integer>comparing((p_217793_0_) -> {
       return -p_217793_0_.field_94201_d;
    }).thenComparing((p_217795_0_) -> {
@@ -50,15 +52,21 @@
    public void func_94305_f() {
       List<Stitcher.Holder> list = Lists.newArrayList(this.field_94319_a);
       list.sort(field_217797_a);
+      try(net.minecraftforge.fml.common.progress.ProgressBar bar = net.minecraftforge.fml.common.progress.StartupProgressManager.start("Texture stitching", list.size())) {
 
       for(Stitcher.Holder stitcher$holder : list) {
+         bar.step(stitcher$holder.field_98151_a.func_195668_m().toString());
          if (!this.func_94310_b(stitcher$holder)) {
+            LOGGER.info(String.format("Unable to fit: %s - size: %dx%d - Maybe try a lower resolution resourcepack?", stitcher$holder.field_98151_a.func_195668_m(), stitcher$holder.field_98151_a.func_94211_a(), stitcher$holder.field_98151_a.func_94216_b()));
+            for (Stitcher.Holder h : list)
+                LOGGER.info("  {}", h);
             throw new StitcherException(stitcher$holder.field_98151_a);
          }
       }
 
       this.field_94318_c = MathHelper.func_151236_b(this.field_94318_c);
       this.field_94315_d = MathHelper.func_151236_b(this.field_94315_d);
+      }; // Forge: end progress bar
    }
 
    public List<TextureAtlasSprite> func_94309_g() {
@@ -104,7 +112,7 @@
          boolean flag4 = flag2 && j != l;
          boolean flag;
          if (flag3 ^ flag4) {
-            flag = flag3;
+            flag = !flag3 && flag1; // Forge: Fix stitcher not expanding entire height before growing width, and (potentially) growing larger then the max size.
          } else {
             flag = flag1 && i <= j;
          }
@@ -141,7 +149,7 @@
       }
 
       public String toString() {
-         return "Holder{width=" + this.field_94204_c + ", height=" + this.field_94201_d + '}';
+         return "Holder{width=" + this.field_94204_c + ", height=" + this.field_94201_d + ", name=" + this.field_98151_a.func_195668_m() + '}';
       }
    }
 
