--- ../src-base/minecraft/net/minecraft/client/renderer/texture/Stitcher.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/texture/Stitcher.java
@@ -58,12 +58,17 @@
     {
         Stitcher.Holder[] astitcher$holder = (Stitcher.Holder[])this.field_94319_a.toArray(new Stitcher.Holder[this.field_94319_a.size()]);
         Arrays.sort((Object[])astitcher$holder);
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar bar = net.minecraftforge.fml.common.ProgressManager.push("Texture stitching", astitcher$holder.length);
 
         for (Stitcher.Holder stitcher$holder : astitcher$holder)
         {
+            bar.step(stitcher$holder.func_98150_a().func_94215_i());
             if (!this.func_94310_b(stitcher$holder))
             {
                 String s = String.format("Unable to fit: %s - size: %dx%d - Maybe try a lowerresolution resourcepack?", new Object[] {stitcher$holder.func_98150_a().func_94215_i(), Integer.valueOf(stitcher$holder.func_98150_a().func_94211_a()), Integer.valueOf(stitcher$holder.func_98150_a().func_94216_b())});
+                net.minecraftforge.fml.common.FMLLog.info(s);
+                for (Stitcher.Holder h : astitcher$holder)
+                    net.minecraftforge.fml.common.FMLLog.info("  %s", h);
                 throw new StitcherException(stitcher$holder, s);
             }
         }
@@ -73,6 +78,7 @@
             this.field_94318_c = MathHelper.func_151236_b(this.field_94318_c);
             this.field_94315_d = MathHelper.func_151236_b(this.field_94315_d);
         }
+        net.minecraftforge.fml.common.ProgressManager.pop(bar);
     }
 
     public List<TextureAtlasSprite> func_94309_g()
@@ -260,7 +266,7 @@
 
             public String toString()
             {
-                return "Holder{width=" + this.field_94204_c + ", height=" + this.field_94201_d + '}';
+                return "Holder{width=" + this.field_94204_c + ", height=" + this.field_94201_d + ", name=" + this.field_98151_a.func_94215_i() + '}';
             }
 
             public int compareTo(Stitcher.Holder p_compareTo_1_)
