--- a/net/minecraft/client/renderer/texture/TextureManager.java
+++ b/net/minecraft/client/renderer/texture/TextureManager.java
@@ -167,6 +_,7 @@
    public void release(ResourceLocation p_118514_) {
       AbstractTexture abstracttexture = this.byPath.remove(p_118514_);
       if (abstracttexture != null) {
+         this.byPath.remove(p_118514_); // Forge: fix MC-98707
          this.safeClose(p_118514_, abstracttexture);
       }
 
