--- ../src_base/minecraft/net/minecraft/client/renderer/texture/TextureMap.java
+++ ../src_work/minecraft/net/minecraft/client/renderer/texture/TextureMap.java
@@ -95,8 +95,18 @@
         while (iterator.hasNext())
         {
             String s = (String)iterator.next();
-            String s1 = this.field_94254_c + s + this.field_94251_d;
-            List list = TextureManager.func_94267_b().func_94266_e(s1);
+            String s1;
+            if (s.indexOf(':')==-1)
+            {
+                s1 = this.field_94254_c + s + this.field_94251_d;
+            }
+            else
+            {
+                String domain = s.substring(0,s.indexOf(':'));
+                String file = s.substring(s.indexOf(':')+1);
+                s1 = "mods/"+domain +"/" + this.field_94254_c + file + this.field_94251_d;
+            }
+            List list = TextureManager.func_94267_b().createNewTexture(s, s1);
 
             if (!list.isEmpty())
             {
@@ -150,7 +160,17 @@
             if (list1.size() > 1)
             {
                 this.field_94258_i.add(texturestitched);
-                String s3 = this.field_94254_c + s2 + ".txt";
+                String s3;
+                if (s2.indexOf(':') == -1)
+                {
+                    s3 = field_94254_c + s2 + ".txt";
+                }
+                else
+                {
+                    String domain = s2.substring(0, s2.indexOf(':'));
+                    String file = s2.substring(s2.indexOf(':') + 1);
+                    s3 = "mods/" + domain + "/" + field_94254_c + file + ".txt";
+                }
                 ITexturePack itexturepack = Minecraft.getMinecraft().texturePackList.getSelectedTexturePack();
                 boolean flag1 = !itexturepack.func_98138_b("/" + this.field_94254_c + s2 + ".png", false);
 
