--- ../src-base/minecraft/net/minecraft/client/renderer/texture/TextureManager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/texture/TextureManager.java
@@ -137,17 +137,20 @@
 
         if (itextureobject != null)
         {
+            this.field_110585_a.remove(p_147645_1_); // Forge: fix MC-98707
             TextureUtil.func_147942_a(itextureobject.func_110552_b());
         }
     }
 
     public void func_110549_a(IResourceManager p_110549_1_)
     {
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar bar = net.minecraftforge.fml.common.ProgressManager.push("Reloading Texture Manager", this.field_110585_a.keySet().size(), true);
         Iterator<Entry<ResourceLocation, ITextureObject>> iterator = this.field_110585_a.entrySet().iterator();
 
         while (iterator.hasNext())
         {
             Entry<ResourceLocation, ITextureObject> entry = (Entry)iterator.next();
+            bar.step(entry.getKey().toString());
             ITextureObject itextureobject = entry.getValue();
 
             if (itextureobject == TextureUtil.field_111001_a)
@@ -159,5 +162,6 @@
                 this.func_110579_a(entry.getKey(), itextureobject);
             }
         }
+        net.minecraftforge.fml.common.ProgressManager.pop(bar);
     }
 }
