--- a/net/minecraft/client/renderer/texture/TextureManager.java
+++ b/net/minecraft/client/renderer/texture/TextureManager.java
@@ -157,6 +157,7 @@
    public void func_147645_c(ResourceLocation p_147645_1_) {
       Texture texture = this.func_229267_b_(p_147645_1_);
       if (texture != null) {
+         this.field_110585_a.remove(p_147645_1_); // Forge: MC-98707 Texture manager is not deleting entry in mapTextureObjects when texture is deleted
          TextureUtil.func_225679_a_(texture.func_110552_b());
       }
 
