--- a/net/minecraft/client/renderer/texture/TextureAtlasSprite.java
+++ b/net/minecraft/client/renderer/texture/TextureAtlasSprite.java
@@ -145,6 +_,7 @@
 
    void m_118375_(int p_118376_, int p_118377_, NativeImage[] p_118378_) {
       for(int i = 0; i < this.f_118342_.length; ++i) {
+         if ((this.f_174723_ >> i <= 0) || (this.f_174724_ >> i <= 0)) break;
          p_118378_[i].m_85003_(i, this.f_118349_ >> i, this.f_118350_ >> i, p_118376_ >> i, p_118377_ >> i, this.f_174723_ >> i, this.f_174724_ >> i, this.f_118342_.length > 1, false);
       }
 
@@ -438,5 +_,15 @@
          }
 
       }
+   }
+
+   // Forge Start
+   public int getPixelRGBA(int frameIndex, int x, int y) {
+       if (this.f_174725_ != null) {
+           x += this.f_174725_.m_174759_(frameIndex) * this.f_174723_;
+           y += this.f_174725_.m_174764_(frameIndex) * this.f_174724_;
+       }
+
+       return this.f_118342_[0].m_84985_(x, y);
    }
 }
