--- a/net/minecraft/client/renderer/texture/TextureManager.java
+++ b/net/minecraft/client/renderer/texture/TextureManager.java
@@ -168,6 +_,7 @@
    public void m_118513_(ResourceLocation p_118514_) {
       AbstractTexture abstracttexture = this.f_118468_.remove(p_118514_);
       if (abstracttexture != null) {
+         this.f_118468_.remove(p_118514_); // Forge: fix MC-98707
          this.m_118508_(p_118514_, abstracttexture);
       }
 
